// lodash/assignIn.js
var copyObject = require('./_copyObject'),
    createAssigner = require('./_createAssigner'),
    keysIn = require('./keysIn');

/**
 * This method is like `_.assign` except that it iterates over own and
 * inherited source properties.
 *
 * **Note:** This method mutates `object`.
 *
 * @static
 * @memberOf _
 * @since 4.0.0
 * @alias extend
 * @category Object
 * @param ***REMOVED***Object} object The destination object.
 * @param ***REMOVED***...Object} [sources] The source objects.
 * @returns ***REMOVED***Object} Returns `object`.
 * @see _.assign
 * @example
 *
 * function Foo() ***REMOVED***
 *   this.a = 1;
 * }
 *
 * function Bar() ***REMOVED***
 *   this.c = 3;
 * }
 *
 * Foo.prototype.b = 2;
 * Bar.prototype.d = 4;
 *
 * _.assignIn(***REMOVED*** 'a': 0 }, new Foo, new Bar);
 * // => ***REMOVED*** 'a': 1, 'b': 2, 'c': 3, 'd': 4 }
 */
var assignIn = createAssigner(function(object, source) ***REMOVED***
  copyObject(source, keysIn(source), object);
});

module.exports = assignIn;
