// lodash/forOwn.js
var baseForOwn = require('./_baseForOwn'),
    castFunction = require('./_castFunction');

/**
 * Iterates over own enumerable string keyed properties of an object and
 * invokes `iteratee` for each property. The iteratee is invoked with three
 * arguments: (value, key, object). Iteratee functions may exit iteration
 * early by explicitly returning `false`.
 *
 * @static
 * @memberOf _
 * @since 0.3.0
 * @category Object
 * @param ***REMOVED***Object} object The object to iterate over.
 * @param ***REMOVED***Function} [iteratee=_.identity] The function invoked per iteration.
 * @returns ***REMOVED***Object} Returns `object`.
 * @see _.forOwnRight
 * @example
 *
 * function Foo() ***REMOVED***
 *   this.a = 1;
 *   this.b = 2;
 * }
 *
 * Foo.prototype.c = 3;
 *
 * _.forOwn(new Foo, function(value, key) ***REMOVED***
 *   console.log(key);
 * });
 * // => Logs 'a' then 'b' (iteration order is not guaranteed).
 */
function forOwn(object, iteratee) ***REMOVED***
  return object && baseForOwn(object, castFunction(iteratee));
}

module.exports = forOwn;
