// http-parser-js/http-parser.js
/*jshint node:true */

var assert = require('assert');

exports.HTTPParser = HTTPParser;
function HTTPParser(type) ***REMOVED***
  assert.ok(type === HTTPParser.REQUEST || type === HTTPParser.RESPONSE || type === undefined);
  if (type === undefined) ***REMOVED***
    // Node v12+
***REMOVED*** else ***REMOVED***
    this.initialize(type);
***REMOVED***
  this.maxHeaderSize=HTTPParser.maxHeaderSize
}
HTTPParser.prototype.initialize = function (type, async_resource) ***REMOVED***
  assert.ok(type === HTTPParser.REQUEST || type === HTTPParser.RESPONSE);
  this.type = type;
  this.state = type + '_LINE';
  this.info = ***REMOVED***
    headers: [],
    upgrade: false
***REMOVED***;
  this.trailers = [];
  this.line = '';
  this.isChunked = false;
  this.connection = '';
  this.headerSize = 0; // for preventing too big headers
  this.body_bytes = null;
  this.isUserCall = false;
  this.hadError = false;
};

HTTPParser.encoding = 'ascii';
HTTPParser.maxHeaderSize = 80 * 1024; // maxHeaderSize (in bytes) is configurable, but 80kb by default;
HTTPParser.REQUEST = 'REQUEST';
HTTPParser.RESPONSE = 'RESPONSE';

// Note: *not* starting with kOnHeaders=0 line the Node parser, because any
//   newly added constants (kOnTimeout in Node v12.19.0) will overwrite 0!
var kOnHeaders = HTTPParser.kOnHeaders = 1;
var kOnHeadersComplete = HTTPParser.kOnHeadersComplete = 2;
var kOnBody = HTTPParser.kOnBody = 3;
var kOnMessageComplete = HTTPParser.kOnMessageComplete = 4;

// Some handler stubs, needed for compatibility
HTTPParser.prototype[kOnHeaders] =
HTTPParser.prototype[kOnHeadersComplete] =
HTTPParser.prototype[kOnBody] =
HTTPParser.prototype[kOnMessageComplete] = function () ***REMOVED***};

var compatMode0_12 = true;
Object.defineProperty(HTTPParser, 'kOnExecute', ***REMOVED***
    get: function () ***REMOVED***
      // hack for backward compatibility
      compatMode0_12 = false;
      return 99;
  ***REMOVED***
***REMOVED***);

var methods = exports.methods = HTTPParser.methods = [
  'DELETE',
  'GET',
  'HEAD',
  'POST',
  'PUT',
  'CONNECT',
  'OPTIONS',
  'TRACE',
  'COPY',
  'LOCK',
  'MKCOL',
  'MOVE',
  'PROPFIND',
  'PROPPATCH',
  'SEARCH',
  'UNLOCK',
  'BIND',
  'REBIND',
  'UNBIND',
  'ACL',
  'REPORT',
  'MKACTIVITY',
  'CHECKOUT',
  'MERGE',
  'M-SEARCH',
  'NOTIFY',
  'SUBSCRIBE',
  'UNSUBSCRIBE',
  'PATCH',
  'PURGE',
  'MKCALENDAR',
  'LINK',
  'UNLINK',
  'SOURCE',
];
var method_connect = methods.indexOf('CONNECT');
HTTPParser.prototype.reinitialize = HTTPParser;
HTTPParser.prototype.close =
HTTPParser.prototype.pause =
HTTPParser.prototype.resume =
HTTPParser.prototype.free = function () ***REMOVED***};
HTTPParser.prototype._compatMode0_11 = false;
HTTPParser.prototype.getAsyncId = function() ***REMOVED*** return 0; };

var headerState = ***REMOVED***
  REQUEST_LINE: true,
  RESPONSE_LINE: true,
  HEADER: true
};
HTTPParser.prototype.execute = function (chunk, start, length) ***REMOVED***
  if (!(this instanceof HTTPParser)) ***REMOVED***
    throw new TypeError('not a HTTPParser');
***REMOVED***

  // backward compat to node < 0.11.4
  // Note: the start and length params were removed in newer version
  start = start || 0;
  length = typeof length === 'number' ? length : chunk.length;

  this.chunk = chunk;
  this.offset = start;
  var end = this.end = start + length;
  try ***REMOVED***
    while (this.offset < end) ***REMOVED***
      if (this[this.state]()) ***REMOVED***
        break;
    ***REMOVED***
  ***REMOVED***
***REMOVED*** catch (err) ***REMOVED***
    if (this.isUserCall) ***REMOVED***
      throw err;
  ***REMOVED***
    this.hadError = true;
    return err;
***REMOVED***
  this.chunk = null;
  length = this.offset - start;
  if (headerState[this.state]) ***REMOVED***
    this.headerSize += length;
    if (this.headerSize > (this.maxHeaderSize||HTTPParser.maxHeaderSize)) ***REMOVED***
      return new Error('max header size exceeded');
  ***REMOVED***
***REMOVED***
  return length;
};

var stateFinishAllowed = ***REMOVED***
  REQUEST_LINE: true,
  RESPONSE_LINE: true,
  BODY_RAW: true
};
HTTPParser.prototype.finish = function () ***REMOVED***
  if (this.hadError) ***REMOVED***
    return;
***REMOVED***
  if (!stateFinishAllowed[this.state]) ***REMOVED***
    return new Error('invalid state for EOF');
***REMOVED***
  if (this.state === 'BODY_RAW') ***REMOVED***
    this.userCall()(this[kOnMessageComplete]());
***REMOVED***
};

// These three methods are used for an internal speed optimization, and it also
// works if theses are noops. Basically consume() asks us to read the bytes
// ourselves, but if we don't do it we get them through execute().
HTTPParser.prototype.consume =
HTTPParser.prototype.unconsume =
HTTPParser.prototype.getCurrentBuffer = function () ***REMOVED***};

//For correct error handling - see HTTPParser#execute
//Usage: this.userCall()(userFunction('arg'));
HTTPParser.prototype.userCall = function () ***REMOVED***
  this.isUserCall = true;
  var self = this;
  return function (ret) ***REMOVED***
    self.isUserCall = false;
    return ret;
***REMOVED***;
};

HTTPParser.prototype.nextRequest = function () ***REMOVED***
  this.userCall()(this[kOnMessageComplete]());
  this.reinitialize(this.type);
};

HTTPParser.prototype.consumeLine = function () ***REMOVED***
  var end = this.end,
      chunk = this.chunk;
  for (var i = this.offset; i < end; i++) ***REMOVED***
    if (chunk[i] === 0x0a) ***REMOVED*** // \n
      var line = this.line + chunk.toString(HTTPParser.encoding, this.offset, i);
      if (line.charAt(line.length - 1) === '\r') ***REMOVED***
        line = line.substr(0, line.length - 1);
    ***REMOVED***
      this.line = '';
      this.offset = i + 1;
      return line;
  ***REMOVED***
***REMOVED***
  //line split over multiple chunks
  this.line += chunk.toString(HTTPParser.encoding, this.offset, this.end);
  this.offset = this.end;
};

var headerExp = /^([^: \t]+):[ \t]*((?:.*[^ \t])|)/;
var headerContinueExp = /^[ \t]+(.*[^ \t])/;
HTTPParser.prototype.parseHeader = function (line, headers) ***REMOVED***
  if (line.indexOf('\r') !== -1) ***REMOVED***
    throw parseErrorCode('HPE_LF_EXPECTED');
***REMOVED***

  var match = headerExp.exec(line);
  var k = match && match[1];
  if (k) ***REMOVED*** // skip empty string (malformed header)
    headers.push(k);
    headers.push(match[2]);
***REMOVED*** else ***REMOVED***
    var matchContinue = headerContinueExp.exec(line);
    if (matchContinue && headers.length) ***REMOVED***
      if (headers[headers.length - 1]) ***REMOVED***
        headers[headers.length - 1] += ' ';
    ***REMOVED***
      headers[headers.length - 1] += matchContinue[1];
  ***REMOVED***
***REMOVED***
};

var requestExp = /^([A-Z-]+) ([^ ]+) HTTP\/(\d)\.(\d)$/;
HTTPParser.prototype.REQUEST_LINE = function () ***REMOVED***
  var line = this.consumeLine();
  if (!line) ***REMOVED***
    return;
***REMOVED***
  var match = requestExp.exec(line);
  if (match === null) ***REMOVED***
    throw parseErrorCode('HPE_INVALID_CONSTANT');
***REMOVED***
  this.info.method = this._compatMode0_11 ? match[1] : methods.indexOf(match[1]);
  if (this.info.method === -1) ***REMOVED***
    throw new Error('invalid request method');
***REMOVED***
  this.info.url = match[2];
  this.info.versionMajor = +match[3];
  this.info.versionMinor = +match[4];
  this.body_bytes = 0;
  this.state = 'HEADER';
};

var responseExp = /^HTTP\/(\d)\.(\d) (\d***REMOVED***3}) ?(.*)$/;
HTTPParser.prototype.RESPONSE_LINE = function () ***REMOVED***
  var line = this.consumeLine();
  if (!line) ***REMOVED***
    return;
***REMOVED***
  var match = responseExp.exec(line);
  if (match === null) ***REMOVED***
    throw parseErrorCode('HPE_INVALID_CONSTANT');
***REMOVED***
  this.info.versionMajor = +match[1];
  this.info.versionMinor = +match[2];
  var statusCode = this.info.statusCode = +match[3];
  this.info.statusMessage = match[4];
  // Implied zero length.
  if ((statusCode / 100 | 0) === 1 || statusCode === 204 || statusCode === 304) ***REMOVED***
    this.body_bytes = 0;
***REMOVED***
  this.state = 'HEADER';
};

HTTPParser.prototype.shouldKeepAlive = function () ***REMOVED***
  if (this.info.versionMajor > 0 && this.info.versionMinor > 0) ***REMOVED***
    if (this.connection.indexOf('close') !== -1) ***REMOVED***
      return false;
  ***REMOVED***
***REMOVED*** else if (this.connection.indexOf('keep-alive') === -1) ***REMOVED***
    return false;
***REMOVED***
  if (this.body_bytes !== null || this.isChunked) ***REMOVED*** // || skipBody
    return true;
***REMOVED***
  return false;
};

HTTPParser.prototype.HEADER = function () ***REMOVED***
  var line = this.consumeLine();
  if (line === undefined) ***REMOVED***
    return;
***REMOVED***
  var info = this.info;
  if (line) ***REMOVED***
    this.parseHeader(line, info.headers);
***REMOVED*** else ***REMOVED***
    var headers = info.headers;
    var hasContentLength = false;
    var currentContentLengthValue;
    var hasUpgradeHeader = false;
    for (var i = 0; i < headers.length; i += 2) ***REMOVED***
      switch (headers[i].toLowerCase()) ***REMOVED***
        case 'transfer-encoding':
          this.isChunked = headers[i + 1].toLowerCase() === 'chunked';
          break;
        case 'content-length':
          currentContentLengthValue = +headers[i + 1];
          if (hasContentLength) ***REMOVED***
            // Fix duplicate Content-Length header with same values.
            // Throw error only if values are different.
            // Known issues:
            // https://github.com/request/request/issues/2091#issuecomment-328715113
            // https://github.com/nodejs/node/issues/6517#issuecomment-216263771
            if (currentContentLengthValue !== this.body_bytes) ***REMOVED***
              throw parseErrorCode('HPE_UNEXPECTED_CONTENT_LENGTH');
          ***REMOVED***
        ***REMOVED*** else ***REMOVED***
            hasContentLength = true;
            this.body_bytes = currentContentLengthValue;
        ***REMOVED***
          break;
        case 'connection':
          this.connection += headers[i + 1].toLowerCase();
          break;
        case 'upgrade':
          hasUpgradeHeader = true;
          break;
    ***REMOVED***
  ***REMOVED***

    // if both isChunked and hasContentLength, isChunked wins
    // This is required so the body is parsed using the chunked method, and matches
    // Chrome's behavior.  We could, maybe, ignore them both (would get chunked
    // encoding into the body), and/or disable shouldKeepAlive to be more
    // resilient.
    if (this.isChunked && hasContentLength) ***REMOVED***
      hasContentLength = false;
      this.body_bytes = null;
  ***REMOVED***

    // Logic from https://github.com/nodejs/http-parser/blob/921d5585515a153fa00e411cf144280c59b41f90/http_parser.c#L1727-L1737
    // "For responses, "Upgrade: foo" and "Connection: upgrade" are
    //   mandatory only when it is a 101 Switching Protocols response,
    //   otherwise it is purely informational, to announce support.
    if (hasUpgradeHeader && this.connection.indexOf('upgrade') != -1) ***REMOVED***
      info.upgrade = this.type === HTTPParser.REQUEST || info.statusCode === 101;
  ***REMOVED*** else ***REMOVED***
      info.upgrade = info.method === method_connect;
  ***REMOVED***

    if (this.isChunked && info.upgrade) ***REMOVED***
      this.isChunked = false;
  ***REMOVED***

    info.shouldKeepAlive = this.shouldKeepAlive();
    //problem which also exists in original node: we should know skipBody before calling onHeadersComplete
    var skipBody;
    if (compatMode0_12) ***REMOVED***
      skipBody = this.userCall()(this[kOnHeadersComplete](info));
  ***REMOVED*** else ***REMOVED***
      skipBody = this.userCall()(this[kOnHeadersComplete](info.versionMajor,
          info.versionMinor, info.headers, info.method, info.url, info.statusCode,
          info.statusMessage, info.upgrade, info.shouldKeepAlive));
  ***REMOVED***
    if (skipBody === 2) ***REMOVED***
      this.nextRequest();
      return true;
  ***REMOVED*** else if (this.isChunked && !skipBody) ***REMOVED***
      this.state = 'BODY_CHUNKHEAD';
  ***REMOVED*** else if (skipBody || this.body_bytes === 0) ***REMOVED***
      this.nextRequest();
      // For older versions of node (v6.x and older?), that return skipBody=1 or skipBody=true,
      //   need this "return true;" if it's an upgrade request.
      return info.upgrade;
  ***REMOVED*** else if (this.body_bytes === null) ***REMOVED***
      this.state = 'BODY_RAW';
  ***REMOVED*** else ***REMOVED***
      this.state = 'BODY_SIZED';
  ***REMOVED***
***REMOVED***
};

HTTPParser.prototype.BODY_CHUNKHEAD = function () ***REMOVED***
  var line = this.consumeLine();
  if (line === undefined) ***REMOVED***
    return;
***REMOVED***
  this.body_bytes = parseInt(line, 16);
  if (!this.body_bytes) ***REMOVED***
    this.state = 'BODY_CHUNKTRAILERS';
***REMOVED*** else ***REMOVED***
    this.state = 'BODY_CHUNK';
***REMOVED***
};

HTTPParser.prototype.BODY_CHUNK = function () ***REMOVED***
  var length = Math.min(this.end - this.offset, this.body_bytes);
  this.userCall()(this[kOnBody](this.chunk, this.offset, length));
  this.offset += length;
  this.body_bytes -= length;
  if (!this.body_bytes) ***REMOVED***
    this.state = 'BODY_CHUNKEMPTYLINE';
***REMOVED***
};

HTTPParser.prototype.BODY_CHUNKEMPTYLINE = function () ***REMOVED***
  var line = this.consumeLine();
  if (line === undefined) ***REMOVED***
    return;
***REMOVED***
  assert.equal(line, '');
  this.state = 'BODY_CHUNKHEAD';
};

HTTPParser.prototype.BODY_CHUNKTRAILERS = function () ***REMOVED***
  var line = this.consumeLine();
  if (line === undefined) ***REMOVED***
    return;
***REMOVED***
  if (line) ***REMOVED***
    this.parseHeader(line, this.trailers);
***REMOVED*** else ***REMOVED***
    if (this.trailers.length) ***REMOVED***
      this.userCall()(this[kOnHeaders](this.trailers, ''));
  ***REMOVED***
    this.nextRequest();
***REMOVED***
};

HTTPParser.prototype.BODY_RAW = function () ***REMOVED***
  var length = this.end - this.offset;
  this.userCall()(this[kOnBody](this.chunk, this.offset, length));
  this.offset = this.end;
};

HTTPParser.prototype.BODY_SIZED = function () ***REMOVED***
  var length = Math.min(this.end - this.offset, this.body_bytes);
  this.userCall()(this[kOnBody](this.chunk, this.offset, length));
  this.offset += length;
  this.body_bytes -= length;
  if (!this.body_bytes) ***REMOVED***
    this.nextRequest();
***REMOVED***
};

// backward compat to node < 0.11.6
['Headers', 'HeadersComplete', 'Body', 'MessageComplete'].forEach(function (name) ***REMOVED***
  var k = HTTPParser['kOn' + name];
  Object.defineProperty(HTTPParser.prototype, 'on' + name, ***REMOVED***
    get: function () ***REMOVED***
      return this[k];
  ***REMOVED***,
    set: function (to) ***REMOVED***
      // hack for backward compatibility
      this._compatMode0_11 = true;
      method_connect = 'CONNECT';
      return (this[k] = to);
  ***REMOVED***
***REMOVED***);
});

function parseErrorCode(code) ***REMOVED***
  var err = new Error('Parse Error');
  err.code = code;
  return err;
}
