// readable-stream/lib/internal/streams/end-of-stream.js
// Ported from https://github.com/mafintosh/end-of-stream with
// permission from the author, Mathias Buus (@mafintosh).
'use strict';

var ERR_STREAM_PREMATURE_CLOSE = require('../../../errors').codes.ERR_STREAM_PREMATURE_CLOSE;

function once(callback) ***REMOVED***
  var called = false;
  return function () ***REMOVED***
    if (called) return;
    called = true;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) ***REMOVED***
      args[_key] = arguments[_key];
  ***REMOVED***

    callback.apply(this, args);
***REMOVED***;
}

function noop() ***REMOVED***}

function isRequest(stream) ***REMOVED***
  return stream.setHeader && typeof stream.abort === 'function';
}

function eos(stream, opts, callback) ***REMOVED***
  if (typeof opts === 'function') return eos(stream, null, opts);
  if (!opts) opts = ***REMOVED***};
  callback = once(callback || noop);
  var readable = opts.readable || opts.readable !== false && stream.readable;
  var writable = opts.writable || opts.writable !== false && stream.writable;

  var onlegacyfinish = function onlegacyfinish() ***REMOVED***
    if (!stream.writable) onfinish();
***REMOVED***;

  var writableEnded = stream._writableState && stream._writableState.finished;

  var onfinish = function onfinish() ***REMOVED***
    writable = false;
    writableEnded = true;
    if (!readable) callback.call(stream);
***REMOVED***;

  var readableEnded = stream._readableState && stream._readableState.endEmitted;

  var onend = function onend() ***REMOVED***
    readable = false;
    readableEnded = true;
    if (!writable) callback.call(stream);
***REMOVED***;

  var onerror = function onerror(err) ***REMOVED***
    callback.call(stream, err);
***REMOVED***;

  var onclose = function onclose() ***REMOVED***
    var err;

    if (readable && !readableEnded) ***REMOVED***
      if (!stream._readableState || !stream._readableState.ended) err = new ERR_STREAM_PREMATURE_CLOSE();
      return callback.call(stream, err);
  ***REMOVED***

    if (writable && !writableEnded) ***REMOVED***
      if (!stream._writableState || !stream._writableState.ended) err = new ERR_STREAM_PREMATURE_CLOSE();
      return callback.call(stream, err);
  ***REMOVED***
***REMOVED***;

  var onrequest = function onrequest() ***REMOVED***
    stream.req.on('finish', onfinish);
***REMOVED***;

  if (isRequest(stream)) ***REMOVED***
    stream.on('complete', onfinish);
    stream.on('abort', onclose);
    if (stream.req) onrequest();else stream.on('request', onrequest);
***REMOVED*** else if (writable && !stream._writableState) ***REMOVED***
    // legacy streams
    stream.on('end', onlegacyfinish);
    stream.on('close', onlegacyfinish);
***REMOVED***

  stream.on('end', onend);
  stream.on('finish', onfinish);
  if (opts.error !== false) stream.on('error', onerror);
  stream.on('close', onclose);
  return function () ***REMOVED***
    stream.removeListener('complete', onfinish);
    stream.removeListener('abort', onclose);
    stream.removeListener('request', onrequest);
    if (stream.req) stream.req.removeListener('finish', onfinish);
    stream.removeListener('end', onlegacyfinish);
    stream.removeListener('close', onlegacyfinish);
    stream.removeListener('finish', onfinish);
    stream.removeListener('end', onend);
    stream.removeListener('error', onerror);
    stream.removeListener('close', onclose);
***REMOVED***;
}

module.exports = eos;
