// jose/dist/node/cjs/lib/buffer_utils.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.concatKdf = exports.lengthAndInput = exports.uint32be = exports.uint64be = exports.p2s = exports.concat = exports.decoder = exports.encoder = void 0;
const digest_js_1 = require("../runtime/digest.js");
exports.encoder = new TextEncoder();
exports.decoder = new TextDecoder();
const MAX_INT32 = 2 ** 32;
function concat(...buffers) ***REMOVED***
    const size = buffers.reduce((acc, ***REMOVED*** length }) => acc + length, 0);
    const buf = new Uint8Array(size);
    let i = 0;
    buffers.forEach((buffer) => ***REMOVED***
        buf.set(buffer, i);
        i += buffer.length;
  ***REMOVED***);
    return buf;
}
exports.concat = concat;
function p2s(alg, p2sInput) ***REMOVED***
    return concat(exports.encoder.encode(alg), new Uint8Array([0]), p2sInput);
}
exports.p2s = p2s;
function writeUInt32BE(buf, value, offset) ***REMOVED***
    if (value < 0 || value >= MAX_INT32) ***REMOVED***
        throw new RangeError(`value must be >= 0 and <= $***REMOVED***MAX_INT32 - 1}. Received $***REMOVED***value}`);
  ***REMOVED***
    buf.set([value >>> 24, value >>> 16, value >>> 8, value & 0xff], offset);
}
function uint64be(value) ***REMOVED***
    const high = Math.floor(value / MAX_INT32);
    const low = value % MAX_INT32;
    const buf = new Uint8Array(8);
    writeUInt32BE(buf, high, 0);
    writeUInt32BE(buf, low, 4);
    return buf;
}
exports.uint64be = uint64be;
function uint32be(value) ***REMOVED***
    const buf = new Uint8Array(4);
    writeUInt32BE(buf, value);
    return buf;
}
exports.uint32be = uint32be;
function lengthAndInput(input) ***REMOVED***
    return concat(uint32be(input.length), input);
}
exports.lengthAndInput = lengthAndInput;
async function concatKdf(secret, bits, value) ***REMOVED***
    const iterations = Math.ceil((bits >> 3) / 32);
    const res = new Uint8Array(iterations * 32);
    for (let iter = 0; iter < iterations; iter++) ***REMOVED***
        const buf = new Uint8Array(4 + secret.length + value.length);
        buf.set(uint32be(iter + 1));
        buf.set(secret, 4);
        buf.set(value, 4 + secret.length);
        res.set(await (0, digest_js_1.default)('sha256', buf), iter * 32);
  ***REMOVED***
    return res.slice(0, bits >> 3);
}
exports.concatKdf = concatKdf;
