// finalhandler/index.js
/*!
 * finalhandler
 * Copyright(c) 2014-2022 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module dependencies.
 * @private
 */

var debug = require('debug')('finalhandler')
var encodeUrl = require('encodeurl')
var escapeHtml = require('escape-html')
var onFinished = require('on-finished')
var parseUrl = require('parseurl')
var statuses = require('statuses')
var unpipe = require('unpipe')

/**
 * Module variables.
 * @private
 */

var DOUBLE_SPACE_REGEXP = /\x20***REMOVED***2}/g
var NEWLINE_REGEXP = /\n/g

/* istanbul ignore next */
var defer = typeof setImmediate === 'function'
  ? setImmediate
  : function (fn) ***REMOVED*** process.nextTick(fn.bind.apply(fn, arguments)) }
var isFinished = onFinished.isFinished

/**
 * Create a minimal HTML document.
 *
 * @param ***REMOVED***string} message
 * @private
 */

function createHtmlDocument (message) ***REMOVED***
  var body = escapeHtml(message)
    .replace(NEWLINE_REGEXP, '<br>')
    .replace(DOUBLE_SPACE_REGEXP, ' &nbsp;')

  return '<!DOCTYPE html>\n' +
    '<html lang="en">\n' +
    '<head>\n' +
    '<meta charset="utf-8">\n' +
    '<title>Error</title>\n' +
    '</head>\n' +
    '<body>\n' +
    '<pre>' + body + '</pre>\n' +
    '</body>\n' +
    '</html>\n'
}

/**
 * Module exports.
 * @public
 */

module.exports = finalhandler

/**
 * Create a function to handle the final response.
 *
 * @param ***REMOVED***Request} req
 * @param ***REMOVED***Response} res
 * @param ***REMOVED***Object} [options]
 * @return ***REMOVED***Function}
 * @public
 */

function finalhandler (req, res, options) ***REMOVED***
  var opts = options || ***REMOVED***}

  // get environment
  var env = opts.env || process.env.NODE_ENV || 'development'

  // get error callback
  var onerror = opts.onerror

  return function (err) ***REMOVED***
    var headers
    var msg
    var status

    // ignore 404 on in-flight response
    if (!err && headersSent(res)) ***REMOVED***
      debug('cannot 404 after headers sent')
      return
  ***REMOVED***

    // unhandled error
    if (err) ***REMOVED***
      // respect status code from error
      status = getErrorStatusCode(err)

      if (status === undefined) ***REMOVED***
        // fallback to status code on response
        status = getResponseStatusCode(res)
    ***REMOVED*** else ***REMOVED***
        // respect headers from error
        headers = getErrorHeaders(err)
    ***REMOVED***

      // get error message
      msg = getErrorMessage(err, status, env)
  ***REMOVED*** else ***REMOVED***
      // not found
      status = 404
      msg = 'Cannot ' + req.method + ' ' + encodeUrl(getResourceName(req))
  ***REMOVED***

    debug('default %s', status)

    // schedule onerror callback
    if (err && onerror) ***REMOVED***
      defer(onerror, err, req, res)
  ***REMOVED***

    // cannot actually respond
    if (headersSent(res)) ***REMOVED***
      debug('cannot %d after headers sent', status)
      req.socket.destroy()
      return
  ***REMOVED***

    // send response
    send(req, res, status, headers, msg)
***REMOVED***
}

/**
 * Get headers from Error object.
 *
 * @param ***REMOVED***Error} err
 * @return ***REMOVED***object}
 * @private
 */

function getErrorHeaders (err) ***REMOVED***
  if (!err.headers || typeof err.headers !== 'object') ***REMOVED***
    return undefined
***REMOVED***

  var headers = Object.create(null)
  var keys = Object.keys(err.headers)

  for (var i = 0; i < keys.length; i++) ***REMOVED***
    var key = keys[i]
    headers[key] = err.headers[key]
***REMOVED***

  return headers
}

/**
 * Get message from Error object, fallback to status message.
 *
 * @param ***REMOVED***Error} err
 * @param ***REMOVED***number} status
 * @param ***REMOVED***string} env
 * @return ***REMOVED***string}
 * @private
 */

function getErrorMessage (err, status, env) ***REMOVED***
  var msg

  if (env !== 'production') ***REMOVED***
    // use err.stack, which typically includes err.message
    msg = err.stack

    // fallback to err.toString() when possible
    if (!msg && typeof err.toString === 'function') ***REMOVED***
      msg = err.toString()
  ***REMOVED***
***REMOVED***

  return msg || statuses.message[status]
}

/**
 * Get status code from Error object.
 *
 * @param ***REMOVED***Error} err
 * @return ***REMOVED***number}
 * @private
 */

function getErrorStatusCode (err) ***REMOVED***
  // check err.status
  if (typeof err.status === 'number' && err.status >= 400 && err.status < 600) ***REMOVED***
    return err.status
***REMOVED***

  // check err.statusCode
  if (typeof err.statusCode === 'number' && err.statusCode >= 400 && err.statusCode < 600) ***REMOVED***
    return err.statusCode
***REMOVED***

  return undefined
}

/**
 * Get resource name for the request.
 *
 * This is typically just the original pathname of the request
 * but will fallback to "resource" is that cannot be determined.
 *
 * @param ***REMOVED***IncomingMessage} req
 * @return ***REMOVED***string}
 * @private
 */

function getResourceName (req) ***REMOVED***
  try ***REMOVED***
    return parseUrl.original(req).pathname
***REMOVED*** catch (e) ***REMOVED***
    return 'resource'
***REMOVED***
}

/**
 * Get status code from response.
 *
 * @param ***REMOVED***OutgoingMessage} res
 * @return ***REMOVED***number}
 * @private
 */

function getResponseStatusCode (res) ***REMOVED***
  var status = res.statusCode

  // default status code to 500 if outside valid range
  if (typeof status !== 'number' || status < 400 || status > 599) ***REMOVED***
    status = 500
***REMOVED***

  return status
}

/**
 * Determine if the response headers have been sent.
 *
 * @param ***REMOVED***object} res
 * @returns ***REMOVED***boolean}
 * @private
 */

function headersSent (res) ***REMOVED***
  return typeof res.headersSent !== 'boolean'
    ? Boolean(res._header)
    : res.headersSent
}

/**
 * Send response.
 *
 * @param ***REMOVED***IncomingMessage} req
 * @param ***REMOVED***OutgoingMessage} res
 * @param ***REMOVED***number} status
 * @param ***REMOVED***object} headers
 * @param ***REMOVED***string} message
 * @private
 */

function send (req, res, status, headers, message) ***REMOVED***
  function write () ***REMOVED***
    // response body
    var body = createHtmlDocument(message)

    // response status
    res.statusCode = status
    res.statusMessage = statuses.message[status]

    // remove any content headers
    res.removeHeader('Content-Encoding')
    res.removeHeader('Content-Language')
    res.removeHeader('Content-Range')

    // response headers
    setHeaders(res, headers)

    // security headers
    res.setHeader('Content-Security-Policy', "default-src 'none'")
    res.setHeader('X-Content-Type-Options', 'nosniff')

    // standard headers
    res.setHeader('Content-Type', 'text/html; charset=utf-8')
    res.setHeader('Content-Length', Buffer.byteLength(body, 'utf8'))

    if (req.method === 'HEAD') ***REMOVED***
      res.end()
      return
  ***REMOVED***

    res.end(body, 'utf8')
***REMOVED***

  if (isFinished(req)) ***REMOVED***
    write()
    return
***REMOVED***

  // unpipe everything from the request
  unpipe(req)

  // flush the request
  onFinished(req, write)
  req.resume()
}

/**
 * Set response headers from an object.
 *
 * @param ***REMOVED***OutgoingMessage} res
 * @param ***REMOVED***object} headers
 * @private
 */

function setHeaders (res, headers) ***REMOVED***
  if (!headers) ***REMOVED***
    return
***REMOVED***

  var keys = Object.keys(headers)
  for (var i = 0; i < keys.length; i++) ***REMOVED***
    var key = keys[i]
    res.setHeader(key, headers[key])
***REMOVED***
}
