// @babel/types/lib/retrievers/getBindingIdentifiers.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
});
exports.default = getBindingIdentifiers;
var _generated = require("../validators/generated");
function getBindingIdentifiers(node, duplicates, outerOnly) ***REMOVED***
  const search = [].concat(node);
  const ids = Object.create(null);
  while (search.length) ***REMOVED***
    const id = search.shift();
    if (!id) continue;
    const keys =
    getBindingIdentifiers.keys[id.type];
    if ((0, _generated.isIdentifier)(id)) ***REMOVED***
      if (duplicates) ***REMOVED***
        const _ids = ids[id.name] = ids[id.name] || [];
        _ids.push(id);
    ***REMOVED*** else ***REMOVED***
        ids[id.name] = id;
    ***REMOVED***
      continue;
  ***REMOVED***
    if ((0, _generated.isExportDeclaration)(id) && !(0, _generated.isExportAllDeclaration)(id)) ***REMOVED***
      if ((0, _generated.isDeclaration)(id.declaration)) ***REMOVED***
        search.push(id.declaration);
    ***REMOVED***
      continue;
  ***REMOVED***
    if (outerOnly) ***REMOVED***
      if ((0, _generated.isFunctionDeclaration)(id)) ***REMOVED***
        search.push(id.id);
        continue;
    ***REMOVED***
      if ((0, _generated.isFunctionExpression)(id)) ***REMOVED***
        continue;
    ***REMOVED***
  ***REMOVED***
    if (keys) ***REMOVED***
      for (let i = 0; i < keys.length; i++) ***REMOVED***
        const key = keys[i];
        const nodes =
        id[key];
        if (nodes) ***REMOVED***
          Array.isArray(nodes) ? search.push(...nodes) : search.push(nodes);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

  return ids;
}

getBindingIdentifiers.keys = ***REMOVED***
  DeclareClass: ["id"],
  DeclareFunction: ["id"],
  DeclareModule: ["id"],
  DeclareVariable: ["id"],
  DeclareInterface: ["id"],
  DeclareTypeAlias: ["id"],
  DeclareOpaqueType: ["id"],
  InterfaceDeclaration: ["id"],
  TypeAlias: ["id"],
  OpaqueType: ["id"],
  CatchClause: ["param"],
  LabeledStatement: ["label"],
  UnaryExpression: ["argument"],
  AssignmentExpression: ["left"],
  ImportSpecifier: ["local"],
  ImportNamespaceSpecifier: ["local"],
  ImportDefaultSpecifier: ["local"],
  ImportDeclaration: ["specifiers"],
  ExportSpecifier: ["exported"],
  ExportNamespaceSpecifier: ["exported"],
  ExportDefaultSpecifier: ["exported"],
  FunctionDeclaration: ["id", "params"],
  FunctionExpression: ["id", "params"],
  ArrowFunctionExpression: ["params"],
  ObjectMethod: ["params"],
  ClassMethod: ["params"],
  ClassPrivateMethod: ["params"],
  ForInStatement: ["left"],
  ForOfStatement: ["left"],
  ClassDeclaration: ["id"],
  ClassExpression: ["id"],
  RestElement: ["argument"],
  UpdateExpression: ["argument"],
  ObjectProperty: ["value"],
  AssignmentPattern: ["left"],
  ArrayPattern: ["elements"],
  ObjectPattern: ["properties"],
  VariableDeclaration: ["declarations"],
  VariableDeclarator: ["id"]
};
