// css-to-react-native/index.js
'use strict';

Object.defineProperty(exports, '__esModule', ***REMOVED***
  value: true
});

function _interopDefault(ex) ***REMOVED***
  return ex && typeof ex === 'object' && 'default' in ex ? ex['default'] : ex;
}

var parse = require('postcss-value-parser');

var parse__default = _interopDefault(parse);

var camelizeStyleName = _interopDefault(require('camelize'));

var cssColorKeywords = _interopDefault(require('css-color-keywords'));

var matchString = function matchString(node) ***REMOVED***
  if (node.type !== 'string') return null;
  return node.value.replace(/\\([0-9a-f]***REMOVED***1,6})(?:\s|$)/gi, function (match, charCode) ***REMOVED***
    return String.fromCharCode(parseInt(charCode, 16));
***REMOVED***).replace(/\\/g, '');
};

var hexColorRe = /^(#(?:[0-9a-f]***REMOVED***3,4})***REMOVED***1,2})$/i;
var cssFunctionNameRe = /^(rgba?|hsla?|hwb|lab|lch|gray|color)$/;

var matchColor = function matchColor(node) ***REMOVED***
  if (node.type === 'word' && (hexColorRe.test(node.value) || node.value in cssColorKeywords || node.value === 'transparent')) ***REMOVED***
    return node.value;
***REMOVED*** else if (node.type === 'function' && cssFunctionNameRe.test(node.value)) ***REMOVED***
    return parse.stringify(node);
***REMOVED***

  return null;
};

var noneRe = /^(none)$/i;
var autoRe = /^(auto)$/i;
var identRe = /(^-?[_a-z][_a-z0-9-]*$)/i; // Note if these are wrong, you'll need to change index.js too

var numberRe = /^([+-]?(?:\d*\.)?\d+(?:e[+-]?\d+)?)$/i; // Note lengthRe is sneaky: you can omit units for 0

var lengthRe = /^(0$|(?:[+-]?(?:\d*\.)?\d+(?:e[+-]?\d+)?)(?=px$))/i;
var unsupportedUnitRe = /^([+-]?(?:\d*\.)?\d+(?:e[+-]?\d+)?(ch|em|ex|rem|vh|vw|vmin|vmax|cm|mm|in|pc|pt))$/i;
var angleRe = /^([+-]?(?:\d*\.)?\d+(?:e[+-]?\d+)?(?:deg|rad))$/i;
var percentRe = /^([+-]?(?:\d*\.)?\d+(?:e[+-]?\d+)?%)$/i;

var noopToken = function noopToken(predicate) ***REMOVED***
  return function (node) ***REMOVED***
    return predicate(node) ? '<token>' : null;
***REMOVED***;
};

var valueForTypeToken = function valueForTypeToken(type) ***REMOVED***
  return function (node) ***REMOVED***
    return node.type === type ? node.value : null;
***REMOVED***;
};

var regExpToken = function regExpToken(regExp, transform) ***REMOVED***
  if (transform === void 0) ***REMOVED***
    transform = String;
***REMOVED***

  return function (node) ***REMOVED***
    if (node.type !== 'word') return null;
    var match = node.value.match(regExp);
    if (match === null) return null;
    var value = transform(match[1]);
    return value;
***REMOVED***;
};

var SPACE = noopToken(function (node) ***REMOVED***
  return node.type === 'space';
});
var SLASH = noopToken(function (node) ***REMOVED***
  return node.type === 'div' && node.value === '/';
});
var COMMA = noopToken(function (node) ***REMOVED***
  return node.type === 'div' && node.value === ',';
});
var WORD = valueForTypeToken('word');
var NONE = regExpToken(noneRe);
var AUTO = regExpToken(autoRe);
var NUMBER = regExpToken(numberRe, Number);
var LENGTH = regExpToken(lengthRe, Number);
var UNSUPPORTED_LENGTH_UNIT = regExpToken(unsupportedUnitRe);
var ANGLE = regExpToken(angleRe, function (angle) ***REMOVED***
  return angle.toLowerCase();
});
var PERCENT = regExpToken(percentRe);
var IDENT = regExpToken(identRe);
var STRING = matchString;
var COLOR = matchColor;
var LINE = regExpToken(/^(none|underline|line-through)$/i);
var BORDER_STYLE = regExpToken(/^(solid|dashed|dotted)$/);
var defaultBorderWidth = 1;
var defaultBorderColor = 'black';
var defaultBorderStyle = 'solid';

var border = function border(tokenStream) ***REMOVED***
  var borderWidth;
  var borderColor;
  var borderStyle;

  if (tokenStream.matches(NONE)) ***REMOVED***
    tokenStream.expectEmpty();
    return ***REMOVED***
      borderWidth: 0,
      borderColor: 'black',
      borderStyle: 'solid'
  ***REMOVED***;
***REMOVED***

  var partsParsed = 0;

  while (partsParsed < 3 && tokenStream.hasTokens()) ***REMOVED***
    if (partsParsed !== 0) tokenStream.expect(SPACE);

    if (borderWidth === undefined && tokenStream.matches(LENGTH, UNSUPPORTED_LENGTH_UNIT)) ***REMOVED***
      borderWidth = tokenStream.lastValue;
  ***REMOVED*** else if (borderColor === undefined && tokenStream.matches(COLOR)) ***REMOVED***
      borderColor = tokenStream.lastValue;
  ***REMOVED*** else if (borderStyle === undefined && tokenStream.matches(BORDER_STYLE)) ***REMOVED***
      borderStyle = tokenStream.lastValue;
  ***REMOVED*** else ***REMOVED***
      tokenStream["throw"]();
  ***REMOVED***

    partsParsed += 1;
***REMOVED***

  tokenStream.expectEmpty();
  if (borderWidth === undefined) borderWidth = defaultBorderWidth;
  if (borderColor === undefined) borderColor = defaultBorderColor;
  if (borderStyle === undefined) borderStyle = defaultBorderStyle;
  return ***REMOVED***
    borderWidth: borderWidth,
    borderColor: borderColor,
    borderStyle: borderStyle
***REMOVED***;
};

var directionFactory = function directionFactory(_ref) ***REMOVED***
  var _ref$types = _ref.types,
      types = _ref$types === void 0 ? [LENGTH, UNSUPPORTED_LENGTH_UNIT, PERCENT] : _ref$types,
      _ref$directions = _ref.directions,
      directions = _ref$directions === void 0 ? ['Top', 'Right', 'Bottom', 'Left'] : _ref$directions,
      _ref$prefix = _ref.prefix,
      prefix = _ref$prefix === void 0 ? '' : _ref$prefix,
      _ref$suffix = _ref.suffix,
      suffix = _ref$suffix === void 0 ? '' : _ref$suffix;
  return function (tokenStream) ***REMOVED***
    var _ref2;

    var values = []; // borderWidth doesn't currently allow a percent value, but may do in the future

    values.push(tokenStream.expect.apply(tokenStream, types));

    while (values.length < 4 && tokenStream.hasTokens()) ***REMOVED***
      tokenStream.expect(SPACE);
      values.push(tokenStream.expect.apply(tokenStream, types));
  ***REMOVED***

    tokenStream.expectEmpty();
    var top = values[0],
        _values$ = values[1],
        right = _values$ === void 0 ? top : _values$,
        _values$2 = values[2],
        bottom = _values$2 === void 0 ? top : _values$2,
        _values$3 = values[3],
        left = _values$3 === void 0 ? right : _values$3;

    var keyFor = function keyFor(n) ***REMOVED***
      return "" + prefix + directions[n] + suffix;
  ***REMOVED***;

    return _ref2 = ***REMOVED***}, _ref2[keyFor(0)] = top, _ref2[keyFor(1)] = right, _ref2[keyFor(2)] = bottom, _ref2[keyFor(3)] = left, _ref2;
***REMOVED***;
};

var parseShadowOffset = function parseShadowOffset(tokenStream) ***REMOVED***
  var width = tokenStream.expect(LENGTH);
  var height = tokenStream.matches(SPACE) ? tokenStream.expect(LENGTH) : width;
  tokenStream.expectEmpty();
  return ***REMOVED***
    width: width,
    height: height
***REMOVED***;
};

var parseShadow = function parseShadow(tokenStream) ***REMOVED***
  var offsetX;
  var offsetY;
  var radius;
  var color;

  if (tokenStream.matches(NONE)) ***REMOVED***
    tokenStream.expectEmpty();
    return ***REMOVED***
      offset: ***REMOVED***
        width: 0,
        height: 0
    ***REMOVED***,
      radius: 0,
      color: 'black'
  ***REMOVED***;
***REMOVED***

  var didParseFirst = false;

  while (tokenStream.hasTokens()) ***REMOVED***
    if (didParseFirst) tokenStream.expect(SPACE);

    if (offsetX === undefined && tokenStream.matches(LENGTH, UNSUPPORTED_LENGTH_UNIT)) ***REMOVED***
      offsetX = tokenStream.lastValue;
      tokenStream.expect(SPACE);
      offsetY = tokenStream.expect(LENGTH, UNSUPPORTED_LENGTH_UNIT);
      tokenStream.saveRewindPoint();

      if (tokenStream.matches(SPACE) && tokenStream.matches(LENGTH, UNSUPPORTED_LENGTH_UNIT)) ***REMOVED***
        radius = tokenStream.lastValue;
    ***REMOVED*** else ***REMOVED***
        tokenStream.rewind();
    ***REMOVED***
  ***REMOVED*** else if (color === undefined && tokenStream.matches(COLOR)) ***REMOVED***
      color = tokenStream.lastValue;
  ***REMOVED*** else ***REMOVED***
      tokenStream["throw"]();
  ***REMOVED***

    didParseFirst = true;
***REMOVED***

  if (offsetX === undefined) tokenStream["throw"]();
  return ***REMOVED***
    offset: ***REMOVED***
      width: offsetX,
      height: offsetY
  ***REMOVED***,
    radius: radius !== undefined ? radius : 0,
    color: color !== undefined ? color : 'black'
***REMOVED***;
};

var boxShadow = function boxShadow(tokenStream) ***REMOVED***
  var _parseShadow = parseShadow(tokenStream),
      offset = _parseShadow.offset,
      radius = _parseShadow.radius,
      color = _parseShadow.color;

  return ***REMOVED***
    shadowOffset: offset,
    shadowRadius: radius,
    shadowColor: color,
    shadowOpacity: 1
***REMOVED***;
};

var defaultFlexGrow = 1;
var defaultFlexShrink = 1;
var defaultFlexBasis = 0;

var flex = function flex(tokenStream) ***REMOVED***
  var flexGrow;
  var flexShrink;
  var flexBasis;

  if (tokenStream.matches(NONE)) ***REMOVED***
    tokenStream.expectEmpty();
    return ***REMOVED***
      flexGrow: 0,
      flexShrink: 0,
      flexBasis: 'auto'
  ***REMOVED***;
***REMOVED***

  tokenStream.saveRewindPoint();

  if (tokenStream.matches(AUTO) && !tokenStream.hasTokens()) ***REMOVED***
    return ***REMOVED***
      flexGrow: 1,
      flexShrink: 1,
      flexBasis: 'auto'
  ***REMOVED***;
***REMOVED***

  tokenStream.rewind();
  var partsParsed = 0;

  while (partsParsed < 2 && tokenStream.hasTokens()) ***REMOVED***
    if (partsParsed !== 0) tokenStream.expect(SPACE);

    if (flexGrow === undefined && tokenStream.matches(NUMBER)) ***REMOVED***
      flexGrow = tokenStream.lastValue;
      tokenStream.saveRewindPoint();

      if (tokenStream.matches(SPACE) && tokenStream.matches(NUMBER)) ***REMOVED***
        flexShrink = tokenStream.lastValue;
    ***REMOVED*** else ***REMOVED***
        tokenStream.rewind();
    ***REMOVED***
  ***REMOVED*** else if (flexBasis === undefined && tokenStream.matches(LENGTH, UNSUPPORTED_LENGTH_UNIT, PERCENT)) ***REMOVED***
      flexBasis = tokenStream.lastValue;
  ***REMOVED*** else if (flexBasis === undefined && tokenStream.matches(AUTO)) ***REMOVED***
      flexBasis = 'auto';
  ***REMOVED*** else ***REMOVED***
      tokenStream["throw"]();
  ***REMOVED***

    partsParsed += 1;
***REMOVED***

  tokenStream.expectEmpty();
  if (flexGrow === undefined) flexGrow = defaultFlexGrow;
  if (flexShrink === undefined) flexShrink = defaultFlexShrink;
  if (flexBasis === undefined) flexBasis = defaultFlexBasis;
  return ***REMOVED***
    flexGrow: flexGrow,
    flexShrink: flexShrink,
    flexBasis: flexBasis
***REMOVED***;
};

var FLEX_WRAP = regExpToken(/(nowrap|wrap|wrap-reverse)/);
var FLEX_DIRECTION = regExpToken(/(row|row-reverse|column|column-reverse)/);
var defaultFlexWrap = 'nowrap';
var defaultFlexDirection = 'row';

var flexFlow = function flexFlow(tokenStream) ***REMOVED***
  var flexWrap;
  var flexDirection;
  var partsParsed = 0;

  while (partsParsed < 2 && tokenStream.hasTokens()) ***REMOVED***
    if (partsParsed !== 0) tokenStream.expect(SPACE);

    if (flexWrap === undefined && tokenStream.matches(FLEX_WRAP)) ***REMOVED***
      flexWrap = tokenStream.lastValue;
  ***REMOVED*** else if (flexDirection === undefined && tokenStream.matches(FLEX_DIRECTION)) ***REMOVED***
      flexDirection = tokenStream.lastValue;
  ***REMOVED*** else ***REMOVED***
      tokenStream["throw"]();
  ***REMOVED***

    partsParsed += 1;
***REMOVED***

  tokenStream.expectEmpty();
  if (flexWrap === undefined) flexWrap = defaultFlexWrap;
  if (flexDirection === undefined) flexDirection = defaultFlexDirection;
  return ***REMOVED***
    flexWrap: flexWrap,
    flexDirection: flexDirection
***REMOVED***;
};

var fontFamily = function fontFamily(tokenStream) ***REMOVED***
  var fontFamily;

  if (tokenStream.matches(STRING)) ***REMOVED***
    fontFamily = tokenStream.lastValue;
***REMOVED*** else ***REMOVED***
    fontFamily = tokenStream.expect(IDENT);

    while (tokenStream.hasTokens()) ***REMOVED***
      tokenStream.expect(SPACE);
      var nextIdent = tokenStream.expect(IDENT);
      fontFamily += " " + nextIdent;
  ***REMOVED***
***REMOVED***

  tokenStream.expectEmpty();
  return ***REMOVED***
    fontFamily: fontFamily
***REMOVED***;
};

var NORMAL = regExpToken(/^(normal)$/);
var STYLE = regExpToken(/^(italic)$/);
var WEIGHT = regExpToken(/^([1-9]00|bold)$/);
var VARIANT = regExpToken(/^(small-caps)$/);
var defaultFontStyle = 'normal';
var defaultFontWeight = 'normal';
var defaultFontVariant = [];

var font = function font(tokenStream) ***REMOVED***
  var fontStyle;
  var fontWeight;
  var fontVariant; // let fontSize;

  var lineHeight; // let fontFamily;

  var numStyleWeightVariantMatched = 0;

  while (numStyleWeightVariantMatched < 3 && tokenStream.hasTokens()) ***REMOVED***
    if (tokenStream.matches(NORMAL)) ;else if (fontStyle === undefined && tokenStream.matches(STYLE)) ***REMOVED***
      fontStyle = tokenStream.lastValue;
  ***REMOVED*** else if (fontWeight === undefined && tokenStream.matches(WEIGHT)) ***REMOVED***
      fontWeight = tokenStream.lastValue;
  ***REMOVED*** else if (fontVariant === undefined && tokenStream.matches(VARIANT)) ***REMOVED***
      fontVariant = [tokenStream.lastValue];
  ***REMOVED*** else ***REMOVED***
      break;
  ***REMOVED***
    tokenStream.expect(SPACE);
    numStyleWeightVariantMatched += 1;
***REMOVED***

  var fontSize = tokenStream.expect(LENGTH, UNSUPPORTED_LENGTH_UNIT);

  if (tokenStream.matches(SLASH)) ***REMOVED***
    lineHeight = tokenStream.expect(LENGTH, UNSUPPORTED_LENGTH_UNIT);
***REMOVED***

  tokenStream.expect(SPACE);

  var _fontFamily = fontFamily(tokenStream),
      fontFamily$1 = _fontFamily.fontFamily;

  if (fontStyle === undefined) fontStyle = defaultFontStyle;
  if (fontWeight === undefined) fontWeight = defaultFontWeight;
  if (fontVariant === undefined) fontVariant = defaultFontVariant;
  var out = ***REMOVED***
    fontStyle: fontStyle,
    fontWeight: fontWeight,
    fontVariant: fontVariant,
    fontSize: fontSize,
    fontFamily: fontFamily$1
***REMOVED***;
  if (lineHeight !== undefined) out.lineHeight = lineHeight;
  return out;
};

var fontVariant = function fontVariant(tokenStream) ***REMOVED***
  var values = [tokenStream.expect(IDENT)];

  while (tokenStream.hasTokens()) ***REMOVED***
    tokenStream.expect(SPACE);
    values.push(tokenStream.expect(IDENT));
***REMOVED***

  return ***REMOVED***
    fontVariant: values
***REMOVED***;
};

var ALIGN_CONTENT = regExpToken(/(flex-(?:start|end)|center|stretch|space-(?:between|around))/);
var JUSTIFY_CONTENT = regExpToken(/(flex-(?:start|end)|center|space-(?:between|around|evenly))/);

var placeContent = function placeContent(tokenStream) ***REMOVED***
  var alignContent = tokenStream.expect(ALIGN_CONTENT);
  var justifyContent;

  if (tokenStream.hasTokens()) ***REMOVED***
    tokenStream.expect(SPACE);
    justifyContent = tokenStream.expect(JUSTIFY_CONTENT);
***REMOVED*** else ***REMOVED***
    justifyContent = 'stretch';
***REMOVED***

  tokenStream.expectEmpty();
  return ***REMOVED***
    alignContent: alignContent,
    justifyContent: justifyContent
***REMOVED***;
};

var STYLE$1 = regExpToken(/^(solid|double|dotted|dashed)$/);
var defaultTextDecorationLine = 'none';
var defaultTextDecorationStyle = 'solid';
var defaultTextDecorationColor = 'black';

var textDecoration = function textDecoration(tokenStream) ***REMOVED***
  var line;
  var style;
  var color;
  var didParseFirst = false;

  while (tokenStream.hasTokens()) ***REMOVED***
    if (didParseFirst) tokenStream.expect(SPACE);

    if (line === undefined && tokenStream.matches(LINE)) ***REMOVED***
      var lines = [tokenStream.lastValue.toLowerCase()];
      tokenStream.saveRewindPoint();

      if (lines[0] !== 'none' && tokenStream.matches(SPACE) && tokenStream.matches(LINE)) ***REMOVED***
        lines.push(tokenStream.lastValue.toLowerCase()); // Underline comes before line-through

        lines.sort().reverse();
    ***REMOVED*** else ***REMOVED***
        tokenStream.rewind();
    ***REMOVED***

      line = lines.join(' ');
  ***REMOVED*** else if (style === undefined && tokenStream.matches(STYLE$1)) ***REMOVED***
      style = tokenStream.lastValue;
  ***REMOVED*** else if (color === undefined && tokenStream.matches(COLOR)) ***REMOVED***
      color = tokenStream.lastValue;
  ***REMOVED*** else ***REMOVED***
      tokenStream["throw"]();
  ***REMOVED***

    didParseFirst = true;
***REMOVED***

  return ***REMOVED***
    textDecorationLine: line !== undefined ? line : defaultTextDecorationLine,
    textDecorationColor: color !== undefined ? color : defaultTextDecorationColor,
    textDecorationStyle: style !== undefined ? style : defaultTextDecorationStyle
***REMOVED***;
};

var textDecorationLine = function textDecorationLine(tokenStream) ***REMOVED***
  var lines = [];
  var didParseFirst = false;

  while (tokenStream.hasTokens()) ***REMOVED***
    if (didParseFirst) tokenStream.expect(SPACE);
    lines.push(tokenStream.expect(LINE).toLowerCase());
    didParseFirst = true;
***REMOVED***

  lines.sort().reverse();
  return ***REMOVED***
    textDecorationLine: lines.join(' ')
***REMOVED***;
};

var textShadow = function textShadow(tokenStream) ***REMOVED***
  var _parseShadow2 = parseShadow(tokenStream),
      offset = _parseShadow2.offset,
      radius = _parseShadow2.radius,
      color = _parseShadow2.color;

  return ***REMOVED***
    textShadowOffset: offset,
    textShadowRadius: radius,
    textShadowColor: color
***REMOVED***;
};

var oneOfType = function oneOfType(tokenType) ***REMOVED***
  return function (functionStream) ***REMOVED***
    var value = functionStream.expect(tokenType);
    functionStream.expectEmpty();
    return value;
***REMOVED***;
};

var singleNumber = oneOfType(NUMBER);
var singleLength = oneOfType(LENGTH);
var singleAngle = oneOfType(ANGLE);

var xyTransformFactory = function xyTransformFactory(tokenType) ***REMOVED***
  return function (key, valueIfOmitted) ***REMOVED***
    return function (functionStream) ***REMOVED***
      var _ref3, _ref4;

      var x = functionStream.expect(tokenType);
      var y;

      if (functionStream.hasTokens()) ***REMOVED***
        functionStream.expect(COMMA);
        y = functionStream.expect(tokenType);
    ***REMOVED*** else if (valueIfOmitted !== undefined) ***REMOVED***
        y = valueIfOmitted;
    ***REMOVED*** else ***REMOVED***
        // Assumption, if x === y, then we can omit XY
        // I.e. scale(5) => [***REMOVED*** scale: 5 }] rather than [***REMOVED*** scaleX: 5 }, ***REMOVED*** scaleY: 5 }]
        return x;
    ***REMOVED***

      functionStream.expectEmpty();
      return [(_ref3 = ***REMOVED***}, _ref3[key + "Y"] = y, _ref3), (_ref4 = ***REMOVED***}, _ref4[key + "X"] = x, _ref4)];
  ***REMOVED***;
***REMOVED***;
};

var xyNumber = xyTransformFactory(NUMBER);
var xyLength = xyTransformFactory(LENGTH);
var xyAngle = xyTransformFactory(ANGLE);
var partTransforms = ***REMOVED***
  perspective: singleNumber,
  scale: xyNumber('scale'),
  scaleX: singleNumber,
  scaleY: singleNumber,
  translate: xyLength('translate', 0),
  translateX: singleLength,
  translateY: singleLength,
  rotate: singleAngle,
  rotateX: singleAngle,
  rotateY: singleAngle,
  rotateZ: singleAngle,
  skewX: singleAngle,
  skewY: singleAngle,
  skew: xyAngle('skew', '0deg')
};

var transform = function transform(tokenStream) ***REMOVED***
  var transforms = [];
  var didParseFirst = false;

  while (tokenStream.hasTokens()) ***REMOVED***
    if (didParseFirst) tokenStream.expect(SPACE);
    var functionStream = tokenStream.expectFunction();
    var functionName = functionStream.functionName;
    var transformedValues = partTransforms[functionName](functionStream);

    if (!Array.isArray(transformedValues)) ***REMOVED***
      var _ref5;

      transformedValues = [(_ref5 = ***REMOVED***}, _ref5[functionName] = transformedValues, _ref5)];
  ***REMOVED***

    transforms = transformedValues.concat(transforms);
    didParseFirst = true;
***REMOVED***

  return ***REMOVED***
    transform: transforms
***REMOVED***;
};

var background = function background(tokenStream) ***REMOVED***
  return ***REMOVED***
    backgroundColor: tokenStream.expect(COLOR)
***REMOVED***;
};

var borderColor = directionFactory(***REMOVED***
  types: [COLOR],
  prefix: 'border',
  suffix: 'Color'
});
var borderRadius = directionFactory(***REMOVED***
  directions: ['TopLeft', 'TopRight', 'BottomRight', 'BottomLeft'],
  prefix: 'border',
  suffix: 'Radius'
});
var borderWidth = directionFactory(***REMOVED***
  prefix: 'border',
  suffix: 'Width'
});
var margin = directionFactory(***REMOVED***
  types: [LENGTH, UNSUPPORTED_LENGTH_UNIT, PERCENT, AUTO],
  prefix: 'margin'
});
var padding = directionFactory(***REMOVED***
  prefix: 'padding'
});

var fontWeight = function fontWeight(tokenStream) ***REMOVED***
  return ***REMOVED***
    fontWeight: tokenStream.expect(WORD) // Also match numbers as strings

***REMOVED***;
};

var shadowOffset = function shadowOffset(tokenStream) ***REMOVED***
  return ***REMOVED***
    shadowOffset: parseShadowOffset(tokenStream)
***REMOVED***;
};

var textShadowOffset = function textShadowOffset(tokenStream) ***REMOVED***
  return ***REMOVED***
    textShadowOffset: parseShadowOffset(tokenStream)
***REMOVED***;
};

var transforms = ***REMOVED***
  background: background,
  border: border,
  borderColor: borderColor,
  borderRadius: borderRadius,
  borderWidth: borderWidth,
  boxShadow: boxShadow,
  flex: flex,
  flexFlow: flexFlow,
  font: font,
  fontFamily: fontFamily,
  fontVariant: fontVariant,
  fontWeight: fontWeight,
  margin: margin,
  padding: padding,
  placeContent: placeContent,
  shadowOffset: shadowOffset,
  textShadow: textShadow,
  textShadowOffset: textShadowOffset,
  textDecoration: textDecoration,
  textDecorationLine: textDecorationLine,
  transform: transform
};
var propertiesWithoutUnits;

if (process.env.NODE_ENV !== 'production') ***REMOVED***
  propertiesWithoutUnits = ['aspectRatio', 'elevation', 'flexGrow', 'flexShrink', 'opacity', 'shadowOpacity', 'zIndex'];
}

var devPropertiesWithUnitsRegExp = propertiesWithoutUnits != null ? new RegExp(propertiesWithoutUnits.join('|')) : null;
var SYMBOL_MATCH = 'SYMBOL_MATCH';

var TokenStream =
/*#__PURE__*/
function () ***REMOVED***
  function TokenStream(nodes, parent) ***REMOVED***
    this.index = 0;
    this.nodes = nodes;
    this.functionName = parent != null ? parent.value : null;
    this.lastValue = null;
    this.rewindIndex = -1;
***REMOVED***

  var _proto = TokenStream.prototype;

  _proto.hasTokens = function hasTokens() ***REMOVED***
    return this.index <= this.nodes.length - 1;
***REMOVED***;

  _proto[SYMBOL_MATCH] = function () ***REMOVED***
    if (!this.hasTokens()) return null;
    var node = this.nodes[this.index];

    for (var i = 0; i < arguments.length; i += 1) ***REMOVED***
      var tokenDescriptor = i < 0 || arguments.length <= i ? undefined : arguments[i];
      var value = tokenDescriptor(node);

      if (value !== null) ***REMOVED***
        this.index += 1;
        this.lastValue = value;
        return value;
    ***REMOVED***
  ***REMOVED***

    return null;
***REMOVED***;

  _proto.matches = function matches() ***REMOVED***
    return this[SYMBOL_MATCH].apply(this, arguments) !== null;
***REMOVED***;

  _proto.expect = function expect() ***REMOVED***
    var value = this[SYMBOL_MATCH].apply(this, arguments);
    return value !== null ? value : this["throw"]();
***REMOVED***;

  _proto.matchesFunction = function matchesFunction() ***REMOVED***
    var node = this.nodes[this.index];
    if (node.type !== 'function') return null;
    var value = new TokenStream(node.nodes, node);
    this.index += 1;
    this.lastValue = null;
    return value;
***REMOVED***;

  _proto.expectFunction = function expectFunction() ***REMOVED***
    var value = this.matchesFunction();
    return value !== null ? value : this["throw"]();
***REMOVED***;

  _proto.expectEmpty = function expectEmpty() ***REMOVED***
    if (this.hasTokens()) this["throw"]();
***REMOVED***;

  _proto["throw"] = function _throw() ***REMOVED***
    throw new Error("Unexpected token type: " + this.nodes[this.index].type);
***REMOVED***;

  _proto.saveRewindPoint = function saveRewindPoint() ***REMOVED***
    this.rewindIndex = this.index;
***REMOVED***;

  _proto.rewind = function rewind() ***REMOVED***
    if (this.rewindIndex === -1) throw new Error('Internal error');
    this.index = this.rewindIndex;
    this.lastValue = null;
***REMOVED***;

  return TokenStream;
}();
/* eslint-disable no-param-reassign */
// Note if this is wrong, you'll need to change tokenTypes.js too


var numberOrLengthRe = /^([+-]?(?:\d*\.)?\d+(?:e[+-]?\d+)?)(?:px)?$/i;
var numberOnlyRe = /^[+-]?(?:\d*\.\d*|[1-9]\d*)(?:e[+-]?\d+)?$/i;
var boolRe = /^true|false$/i;
var nullRe = /^null$/i;
var undefinedRe = /^undefined$/i; // Undocumented export

var transformRawValue = function transformRawValue(propName, value) ***REMOVED***
  if (process.env.NODE_ENV !== 'production') ***REMOVED***
    var needsUnit = !devPropertiesWithUnitsRegExp.test(propName);
    var isNumberWithoutUnit = numberOnlyRe.test(value);

    if (needsUnit && isNumberWithoutUnit) ***REMOVED***
      // eslint-disable-next-line no-console
      console.warn("Expected style \"" + propName + ": " + value + "\" to contain units");
  ***REMOVED***

    if (!needsUnit && value !== '0' && !isNumberWithoutUnit) ***REMOVED***
      // eslint-disable-next-line no-console
      console.warn("Expected style \"" + propName + ": " + value + "\" to be unitless");
  ***REMOVED***
***REMOVED***

  var numberMatch = value.match(numberOrLengthRe);
  if (numberMatch !== null) return Number(numberMatch[1]);
  var boolMatch = value.match(boolRe);
  if (boolMatch !== null) return boolMatch[0].toLowerCase() === 'true';
  var nullMatch = value.match(nullRe);
  if (nullMatch !== null) return null;
  var undefinedMatch = value.match(undefinedRe);
  if (undefinedMatch !== null) return undefined;
  return value;
};

var baseTransformShorthandValue = function baseTransformShorthandValue(propName, value) ***REMOVED***
  var ast = parse__default(value);
  var tokenStream = new TokenStream(ast.nodes);
  return transforms[propName](tokenStream);
};

var transformShorthandValue = process.env.NODE_ENV === 'production' ? baseTransformShorthandValue : function (propName, value) ***REMOVED***
  try ***REMOVED***
    return baseTransformShorthandValue(propName, value);
***REMOVED*** catch (e) ***REMOVED***
    throw new Error("Failed to parse declaration \"" + propName + ": " + value + "\"");
***REMOVED***
};

var getStylesForProperty = function getStylesForProperty(propName, inputValue, allowShorthand) ***REMOVED***
  var _ref6;

  var isRawValue = allowShorthand === false || !(propName in transforms);
  var value = inputValue.trim();
  var propValues = isRawValue ? (_ref6 = ***REMOVED***}, _ref6[propName] = transformRawValue(propName, value), _ref6) : transformShorthandValue(propName, value);
  return propValues;
};

var getPropertyName = function getPropertyName(propName) ***REMOVED***
  var isCustomProp = /^--\w+/.test(propName);

  if (isCustomProp) ***REMOVED***
    return propName;
***REMOVED***

  return camelizeStyleName(propName);
};

var index = function index(rules, shorthandBlacklist) ***REMOVED***
  if (shorthandBlacklist === void 0) ***REMOVED***
    shorthandBlacklist = [];
***REMOVED***

  return rules.reduce(function (accum, rule) ***REMOVED***
    var propertyName = getPropertyName(rule[0]);
    var value = rule[1];
    var allowShorthand = shorthandBlacklist.indexOf(propertyName) === -1;
    return Object.assign(accum, getStylesForProperty(propertyName, value, allowShorthand));
***REMOVED***, ***REMOVED***});
};

exports["default"] = index;
exports.getPropertyName = getPropertyName;
exports.getStylesForProperty = getStylesForProperty;
exports.transformRawValue = transformRawValue;
