// @google-cloud/firestore/build/src/field-value.js
"use strict";
/*!
 * Copyright 2018 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.DeleteTransform = exports.FieldTransform = exports.FieldValue = void 0;
const deepEqual = require("fast-deep-equal");
const serializer_1 = require("./serializer");
const validate_1 = require("./validate");
/**
 * Sentinel values that can be used when writing documents with set(), create()
 * or update().
 *
 * @class FieldValue
 */
class FieldValue ***REMOVED***
    /** @private */
    constructor() ***REMOVED*** }
    /**
     * Returns a sentinel for use with update() or set() with ***REMOVED***merge:true} to mark
     * a field for deletion.
     *
     * @returns ***REMOVED***FieldValue} The sentinel value to use in your objects.
     *
     * @example
     * ```
     * let documentRef = firestore.doc('col/doc');
     * let data = ***REMOVED*** a: 'b', c: 'd' };
     *
     * documentRef.set(data).then(() => ***REMOVED***
     *   return documentRef.update(***REMOVED***a: Firestore.FieldValue.delete()});
     * }).then(() => ***REMOVED***
     *   // Document now only contains ***REMOVED*** c: 'd' }
     * });
     * ```
     */
    static delete() ***REMOVED***
        return DeleteTransform.DELETE_SENTINEL;
  ***REMOVED***
    /**
     * Returns a sentinel used with set(), create() or update() to include a
     * server-generated timestamp in the written data.
     *
     * @return ***REMOVED***FieldValue} The FieldValue sentinel for use in a call to set(),
     * create() or update().
     *
     * @example
     * ```
     * let documentRef = firestore.doc('col/doc');
     *
     * documentRef.set(***REMOVED***
     *   time: Firestore.FieldValue.serverTimestamp()
     * }).then(() => ***REMOVED***
     *   return documentRef.get();
     * }).then(doc => ***REMOVED***
     *   console.log(`Server time set to $***REMOVED***doc.get('time')}`);
     * });
     * ```
     */
    static serverTimestamp() ***REMOVED***
        return ServerTimestampTransform.SERVER_TIMESTAMP_SENTINEL;
  ***REMOVED***
    /**
     * Returns a special value that can be used with set(), create() or update()
     * that tells the server to increment the the field's current value by the
     * given value.
     *
     * If either current field value or the operand uses floating point
     * precision, both values will be interpreted as floating point numbers and
     * all arithmetic will follow IEEE 754 semantics. Otherwise, integer
     * precision is kept and the result is capped between -2^63 and 2^63-1.
     *
     * If the current field value is not of type 'number', or if the field does
     * not yet exist, the transformation will set the field to the given value.
     *
     * @param ***REMOVED***number} n The value to increment by.
     * @return ***REMOVED***FieldValue} The FieldValue sentinel for use in a call to set(),
     * create() or update().
     *
     * @example
     * ```
     * let documentRef = firestore.doc('col/doc');
     *
     * documentRef.update(
     *   'counter', Firestore.FieldValue.increment(1)
     * ).then(() => ***REMOVED***
     *   return documentRef.get();
     * }).then(doc => ***REMOVED***
     *   // doc.get('counter') was incremented
     * });
     * ```
     */
    static increment(n) ***REMOVED***
        // eslint-disable-next-line prefer-rest-params
        (0, validate_1.validateMinNumberOfArguments)('FieldValue.increment', arguments, 1);
        return new NumericIncrementTransform(n);
  ***REMOVED***
    /**
     * Returns a special value that can be used with set(), create() or update()
     * that tells the server to union the given elements with any array value that
     * already exists on the server. Each specified element that doesn't already
     * exist in the array will be added to the end. If the field being modified is
     * not already an array it will be overwritten with an array containing
     * exactly the specified elements.
     *
     * @param ***REMOVED***...*} elements The elements to union into the array.
     * @return ***REMOVED***FieldValue} The FieldValue sentinel for use in a call to set(),
     * create() or update().
     *
     * @example
     * ```
     * let documentRef = firestore.doc('col/doc');
     *
     * documentRef.update(
     *   'array', Firestore.FieldValue.arrayUnion('foo')
     * ).then(() => ***REMOVED***
     *   return documentRef.get();
     * }).then(doc => ***REMOVED***
     *   // doc.get('array') contains field 'foo'
     * });
     * ```
     */
    static arrayUnion(...elements) ***REMOVED***
        (0, validate_1.validateMinNumberOfArguments)('FieldValue.arrayUnion', elements, 1);
        return new ArrayUnionTransform(elements);
  ***REMOVED***
    /**
     * Returns a special value that can be used with set(), create() or update()
     * that tells the server to remove the given elements from any array value
     * that already exists on the server. All instances of each element specified
     * will be removed from the array. If the field being modified is not already
     * an array it will be overwritten with an empty array.
     *
     * @param ***REMOVED***...*} elements The elements to remove from the array.
     * @return ***REMOVED***FieldValue} The FieldValue sentinel for use in a call to set(),
     * create() or update().
     *
     * @example
     * ```
     * let documentRef = firestore.doc('col/doc');
     *
     * documentRef.update(
     *   'array', Firestore.FieldValue.arrayRemove('foo')
     * ).then(() => ***REMOVED***
     *   return documentRef.get();
     * }).then(doc => ***REMOVED***
     *   // doc.get('array') no longer contains field 'foo'
     * });
     * ```
     */
    static arrayRemove(...elements) ***REMOVED***
        (0, validate_1.validateMinNumberOfArguments)('FieldValue.arrayRemove', elements, 1);
        return new ArrayRemoveTransform(elements);
  ***REMOVED***
    /**
     * Returns true if this `FieldValue` is equal to the provided value.
     *
     * @param ***REMOVED****} other The value to compare against.
     * @return ***REMOVED***boolean} true if this `FieldValue` is equal to the provided value.
     *
     * @example
     * ```
     * let fieldValues = [
     *   Firestore.FieldValue.increment(-1.0),
     *   Firestore.FieldValue.increment(-1),
     *   Firestore.FieldValue.increment(-0.0),
     *   Firestore.FieldValue.increment(-0),
     *   Firestore.FieldValue.increment(0),
     *   Firestore.FieldValue.increment(0.0),
     *   Firestore.FieldValue.increment(1),
     *   Firestore.FieldValue.increment(1.0)
     * ];
     *
     * let equal = 0;
     * for (let i = 0; i < fieldValues.length; ++i) ***REMOVED***
     *   for (let j = i + 1; j < fieldValues.length; ++j) ***REMOVED***
     *     if (fieldValues[i].isEqual(fieldValues[j])) ***REMOVED***
     *       ++equal;
     *   ***REMOVED***
     * ***REMOVED***
     * }
     * console.log(`Found $***REMOVED***equal} equalities.`);
     * ```
     */
    isEqual(other) ***REMOVED***
        return this === other;
  ***REMOVED***
}
exports.FieldValue = FieldValue;
/**
 * An internal interface shared by all field transforms.
 *
 * A 'FieldTransform` subclass should implement '.includeInDocumentMask',
 * '.includeInDocumentTransform' and 'toProto' (if '.includeInDocumentTransform'
 * is 'true').
 *
 * @private
 * @internal
 * @abstract
 */
class FieldTransform extends FieldValue ***REMOVED***
}
exports.FieldTransform = FieldTransform;
/**
 * A transform that deletes a field from a Firestore document.
 *
 * @private
 * @internal
 */
class DeleteTransform extends FieldTransform ***REMOVED***
    constructor() ***REMOVED***
        super();
  ***REMOVED***
    /**
     * Deletes are included in document masks.
     * @private
     * @internal
     */
    get includeInDocumentMask() ***REMOVED***
        return true;
  ***REMOVED***
    /**
     * Deletes are are omitted from document transforms.
     * @private
     * @internal
     */
    get includeInDocumentTransform() ***REMOVED***
        return false;
  ***REMOVED***
    get methodName() ***REMOVED***
        return 'FieldValue.delete';
  ***REMOVED***
    validate() ***REMOVED*** }
    toProto() ***REMOVED***
        throw new Error('FieldValue.delete() should not be included in a FieldTransform');
  ***REMOVED***
}
exports.DeleteTransform = DeleteTransform;
/**
 * Sentinel value for a field delete.
 * @private
 * @internal
 */
DeleteTransform.DELETE_SENTINEL = new DeleteTransform();
/**
 * A transform that sets a field to the Firestore server time.
 *
 * @private
 * @internal
 */
class ServerTimestampTransform extends FieldTransform ***REMOVED***
    constructor() ***REMOVED***
        super();
  ***REMOVED***
    /**
     * Server timestamps are omitted from document masks.
     *
     * @private
     * @internal
     */
    get includeInDocumentMask() ***REMOVED***
        return false;
  ***REMOVED***
    /**
     * Server timestamps are included in document transforms.
     *
     * @private
     * @internal
     */
    get includeInDocumentTransform() ***REMOVED***
        return true;
  ***REMOVED***
    get methodName() ***REMOVED***
        return 'FieldValue.serverTimestamp';
  ***REMOVED***
    validate() ***REMOVED*** }
    toProto(serializer, fieldPath) ***REMOVED***
        return ***REMOVED***
            fieldPath: fieldPath.formattedName,
            setToServerValue: 'REQUEST_TIME',
      ***REMOVED***;
  ***REMOVED***
}
/**
 * Sentinel value for a server timestamp.
 *
 * @private
 * @internal
 */
ServerTimestampTransform.SERVER_TIMESTAMP_SENTINEL = new ServerTimestampTransform();
/**
 * Increments a field value on the backend.
 *
 * @private
 * @internal
 */
class NumericIncrementTransform extends FieldTransform ***REMOVED***
    constructor(operand) ***REMOVED***
        super();
        this.operand = operand;
  ***REMOVED***
    /**
     * Numeric transforms are omitted from document masks.
     *
     * @private
     * @internal
     */
    get includeInDocumentMask() ***REMOVED***
        return false;
  ***REMOVED***
    /**
     * Numeric transforms are included in document transforms.
     *
     * @private
     * @internal
     */
    get includeInDocumentTransform() ***REMOVED***
        return true;
  ***REMOVED***
    get methodName() ***REMOVED***
        return 'FieldValue.increment';
  ***REMOVED***
    validate() ***REMOVED***
        (0, validate_1.validateNumber)('FieldValue.increment()', this.operand);
  ***REMOVED***
    toProto(serializer, fieldPath) ***REMOVED***
        const encodedOperand = serializer.encodeValue(this.operand);
        return ***REMOVED*** fieldPath: fieldPath.formattedName, increment: encodedOperand };
  ***REMOVED***
    isEqual(other) ***REMOVED***
        return (this === other ||
            (other instanceof NumericIncrementTransform &&
                this.operand === other.operand));
  ***REMOVED***
}
/**
 * Transforms an array value via a union operation.
 *
 * @private
 * @internal
 */
class ArrayUnionTransform extends FieldTransform ***REMOVED***
    constructor(elements) ***REMOVED***
        super();
        this.elements = elements;
  ***REMOVED***
    /**
     * Array transforms are omitted from document masks.
     * @private
     * @internal
     */
    get includeInDocumentMask() ***REMOVED***
        return false;
  ***REMOVED***
    /**
     * Array transforms are included in document transforms.
     * @private
     * @internal
     */
    get includeInDocumentTransform() ***REMOVED***
        return true;
  ***REMOVED***
    get methodName() ***REMOVED***
        return 'FieldValue.arrayUnion';
  ***REMOVED***
    validate(allowUndefined) ***REMOVED***
        for (let i = 0; i < this.elements.length; ++i) ***REMOVED***
            validateArrayElement(i, this.elements[i], allowUndefined);
      ***REMOVED***
  ***REMOVED***
    toProto(serializer, fieldPath) ***REMOVED***
        const encodedElements = serializer.encodeValue(this.elements).arrayValue;
        return ***REMOVED***
            fieldPath: fieldPath.formattedName,
            appendMissingElements: encodedElements,
      ***REMOVED***;
  ***REMOVED***
    isEqual(other) ***REMOVED***
        return (this === other ||
            (other instanceof ArrayUnionTransform &&
                deepEqual(this.elements, other.elements)));
  ***REMOVED***
}
/**
 * Transforms an array value via a remove operation.
 *
 * @private
 * @internal
 */
class ArrayRemoveTransform extends FieldTransform ***REMOVED***
    constructor(elements) ***REMOVED***
        super();
        this.elements = elements;
  ***REMOVED***
    /**
     * Array transforms are omitted from document masks.
     * @private
     * @internal
     */
    get includeInDocumentMask() ***REMOVED***
        return false;
  ***REMOVED***
    /**
     * Array transforms are included in document transforms.
     * @private
     * @internal
     */
    get includeInDocumentTransform() ***REMOVED***
        return true;
  ***REMOVED***
    get methodName() ***REMOVED***
        return 'FieldValue.arrayRemove';
  ***REMOVED***
    validate(allowUndefined) ***REMOVED***
        for (let i = 0; i < this.elements.length; ++i) ***REMOVED***
            validateArrayElement(i, this.elements[i], allowUndefined);
      ***REMOVED***
  ***REMOVED***
    toProto(serializer, fieldPath) ***REMOVED***
        const encodedElements = serializer.encodeValue(this.elements).arrayValue;
        return ***REMOVED***
            fieldPath: fieldPath.formattedName,
            removeAllFromArray: encodedElements,
      ***REMOVED***;
  ***REMOVED***
    isEqual(other) ***REMOVED***
        return (this === other ||
            (other instanceof ArrayRemoveTransform &&
                deepEqual(this.elements, other.elements)));
  ***REMOVED***
}
/**
 * Validates that `value` can be used as an element inside of an array. Certain
 * field values (such as ServerTimestamps) are rejected. Nested arrays are also
 * rejected.
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param value The value to validate.
 * @param allowUndefined Whether to allow nested properties that are `undefined`.
 */
function validateArrayElement(arg, value, allowUndefined) ***REMOVED***
    if (Array.isArray(value)) ***REMOVED***
        throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'array element')} Nested arrays are not supported.`);
  ***REMOVED***
    (0, serializer_1.validateUserInput)(arg, value, 'array element', 
    /*path=*/ ***REMOVED*** allowDeletes: 'none', allowTransforms: false, allowUndefined }, 
    /*path=*/ undefined, 
    /*level=*/ 0, 
    /*inArray=*/ true);
}
