// @google-cloud/firestore/build/src/order.js
"use strict";
/*!
 * Copyright 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.compare = exports.compareArrays = exports.primitiveComparator = void 0;
const convert_1 = require("./convert");
const path_1 = require("./path");
/*!
 * The type order as defined by the backend.
 */
var TypeOrder;
(function (TypeOrder) ***REMOVED***
    TypeOrder[TypeOrder["NULL"] = 0] = "NULL";
    TypeOrder[TypeOrder["BOOLEAN"] = 1] = "BOOLEAN";
    TypeOrder[TypeOrder["NUMBER"] = 2] = "NUMBER";
    TypeOrder[TypeOrder["TIMESTAMP"] = 3] = "TIMESTAMP";
    TypeOrder[TypeOrder["STRING"] = 4] = "STRING";
    TypeOrder[TypeOrder["BLOB"] = 5] = "BLOB";
    TypeOrder[TypeOrder["REF"] = 6] = "REF";
    TypeOrder[TypeOrder["GEO_POINT"] = 7] = "GEO_POINT";
    TypeOrder[TypeOrder["ARRAY"] = 8] = "ARRAY";
    TypeOrder[TypeOrder["OBJECT"] = 9] = "OBJECT";
})(TypeOrder || (TypeOrder = ***REMOVED***}));
/*!
 * @private
 * @internal
 */
function typeOrder(val) ***REMOVED***
    const valueType = (0, convert_1.detectValueType)(val);
    switch (valueType) ***REMOVED***
        case 'nullValue':
            return TypeOrder.NULL;
        case 'integerValue':
            return TypeOrder.NUMBER;
        case 'doubleValue':
            return TypeOrder.NUMBER;
        case 'stringValue':
            return TypeOrder.STRING;
        case 'booleanValue':
            return TypeOrder.BOOLEAN;
        case 'arrayValue':
            return TypeOrder.ARRAY;
        case 'timestampValue':
            return TypeOrder.TIMESTAMP;
        case 'geoPointValue':
            return TypeOrder.GEO_POINT;
        case 'bytesValue':
            return TypeOrder.BLOB;
        case 'referenceValue':
            return TypeOrder.REF;
        case 'mapValue':
            return TypeOrder.OBJECT;
        default:
            throw new Error('Unexpected value type: ' + valueType);
  ***REMOVED***
}
/*!
 * @private
 * @internal
 */
function primitiveComparator(left, right) ***REMOVED***
    if (left < right) ***REMOVED***
        return -1;
  ***REMOVED***
    if (left > right) ***REMOVED***
        return 1;
  ***REMOVED***
    return 0;
}
exports.primitiveComparator = primitiveComparator;
/*!
 * Utility function to compare doubles (using Firestore semantics for NaN).
 * @private
 * @internal
 */
function compareNumbers(left, right) ***REMOVED***
    if (left < right) ***REMOVED***
        return -1;
  ***REMOVED***
    if (left > right) ***REMOVED***
        return 1;
  ***REMOVED***
    if (left === right) ***REMOVED***
        return 0;
  ***REMOVED***
    // one or both are NaN.
    if (isNaN(left)) ***REMOVED***
        return isNaN(right) ? 0 : -1;
  ***REMOVED***
    return 1;
}
/*!
 * @private
 * @internal
 */
function compareNumberProtos(left, right) ***REMOVED***
    let leftValue, rightValue;
    if (left.integerValue !== undefined) ***REMOVED***
        leftValue = Number(left.integerValue);
  ***REMOVED***
    else ***REMOVED***
        leftValue = Number(left.doubleValue);
  ***REMOVED***
    if (right.integerValue !== undefined) ***REMOVED***
        rightValue = Number(right.integerValue);
  ***REMOVED***
    else ***REMOVED***
        rightValue = Number(right.doubleValue);
  ***REMOVED***
    return compareNumbers(leftValue, rightValue);
}
/*!
 * @private
 * @internal
 */
function compareTimestamps(left, right) ***REMOVED***
    const seconds = primitiveComparator(left.seconds || 0, right.seconds || 0);
    if (seconds !== 0) ***REMOVED***
        return seconds;
  ***REMOVED***
    return primitiveComparator(left.nanos || 0, right.nanos || 0);
}
/*!
 * @private
 * @internal
 */
function compareBlobs(left, right) ***REMOVED***
    if (!(left instanceof Buffer) || !(right instanceof Buffer)) ***REMOVED***
        throw new Error('Blobs can only be compared if they are Buffers.');
  ***REMOVED***
    return Buffer.compare(left, right);
}
/*!
 * @private
 * @internal
 */
function compareReferenceProtos(left, right) ***REMOVED***
    const leftPath = path_1.QualifiedResourcePath.fromSlashSeparatedString(left.referenceValue);
    const rightPath = path_1.QualifiedResourcePath.fromSlashSeparatedString(right.referenceValue);
    return leftPath.compareTo(rightPath);
}
/*!
 * @private
 * @internal
 */
function compareGeoPoints(left, right) ***REMOVED***
    return (primitiveComparator(left.latitude || 0, right.latitude || 0) ||
        primitiveComparator(left.longitude || 0, right.longitude || 0));
}
/*!
 * @private
 * @internal
 */
function compareArrays(left, right) ***REMOVED***
    for (let i = 0; i < left.length && i < right.length; i++) ***REMOVED***
        const valueComparison = compare(left[i], right[i]);
        if (valueComparison !== 0) ***REMOVED***
            return valueComparison;
      ***REMOVED***
  ***REMOVED***
    // If all the values matched so far, just check the length.
    return primitiveComparator(left.length, right.length);
}
exports.compareArrays = compareArrays;
/*!
 * @private
 * @internal
 */
function compareObjects(left, right) ***REMOVED***
    // This requires iterating over the keys in the object in order and doing a
    // deep comparison.
    const leftKeys = Object.keys(left);
    const rightKeys = Object.keys(right);
    leftKeys.sort();
    rightKeys.sort();
    for (let i = 0; i < leftKeys.length && i < rightKeys.length; i++) ***REMOVED***
        const keyComparison = primitiveComparator(leftKeys[i], rightKeys[i]);
        if (keyComparison !== 0) ***REMOVED***
            return keyComparison;
      ***REMOVED***
        const key = leftKeys[i];
        const valueComparison = compare(left[key], right[key]);
        if (valueComparison !== 0) ***REMOVED***
            return valueComparison;
      ***REMOVED***
  ***REMOVED***
    // If all the keys matched so far, just check the length.
    return primitiveComparator(leftKeys.length, rightKeys.length);
}
/*!
 * @private
 * @internal
 */
function compare(left, right) ***REMOVED***
    // First compare the types.
    const leftType = typeOrder(left);
    const rightType = typeOrder(right);
    const typeComparison = primitiveComparator(leftType, rightType);
    if (typeComparison !== 0) ***REMOVED***
        return typeComparison;
  ***REMOVED***
    // So they are the same type.
    switch (leftType) ***REMOVED***
        case TypeOrder.NULL:
            // Nulls are all equal.
            return 0;
        case TypeOrder.BOOLEAN:
            return primitiveComparator(left.booleanValue, right.booleanValue);
        case TypeOrder.STRING:
            return primitiveComparator(left.stringValue, right.stringValue);
        case TypeOrder.NUMBER:
            return compareNumberProtos(left, right);
        case TypeOrder.TIMESTAMP:
            return compareTimestamps(left.timestampValue, right.timestampValue);
        case TypeOrder.BLOB:
            return compareBlobs(left.bytesValue, right.bytesValue);
        case TypeOrder.REF:
            return compareReferenceProtos(left, right);
        case TypeOrder.GEO_POINT:
            return compareGeoPoints(left.geoPointValue, right.geoPointValue);
        case TypeOrder.ARRAY:
            return compareArrays(left.arrayValue.values || [], right.arrayValue.values || []);
        case TypeOrder.OBJECT:
            return compareObjects(left.mapValue.fields || ***REMOVED***}, right.mapValue.fields || ***REMOVED***});
        default:
            throw new Error(`Encountered unknown type order: $***REMOVED***leftType}`);
  ***REMOVED***
}
exports.compare = compare;
