// @google-cloud/firestore/build/src/path.js
"use strict";
/*!
 * Copyright 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.validateFieldPath = exports.FieldPath = exports.validateResourcePath = exports.QualifiedResourcePath = exports.ResourcePath = exports.DEFAULT_DATABASE_ID = void 0;
const util_1 = require("./util");
const validate_1 = require("./validate");
/*!
 * The default database ID for this Firestore client. We do not yet expose the
 * ability to use different databases.
 */
exports.DEFAULT_DATABASE_ID = '(default)';
/*!
 * A regular expression to verify an absolute Resource Path in Firestore. It
 * extracts the project ID, the database name and the relative resource path
 * if available.
 *
 * @type ***REMOVED***RegExp}
 */
const RESOURCE_PATH_RE = 
// Note: [\s\S] matches all characters including newlines.
/^projects\/([^/]*)\/databases\/([^/]*)(?:\/documents\/)?([\s\S]*)$/;
/*!
 * A regular expression to verify whether a field name can be passed to the
 * backend without escaping.
 *
 * @type ***REMOVED***RegExp}
 */
const UNESCAPED_FIELD_NAME_RE = /^[_a-zA-Z][_a-zA-Z0-9]*$/;
/*!
 * A regular expression to verify field paths that are passed to the API as
 * strings. Field paths that do not match this expression have to be provided
 * as a [FieldPath]***REMOVED***@link FieldPath} object.
 *
 * @type ***REMOVED***RegExp}
 */
const FIELD_PATH_RE = /^[^*~/[\]]+$/;
/**
 * An abstract class representing a Firestore path.
 *
 * Subclasses have to implement `split()` and `canonicalString()`.
 *
 * @private
 * @internal
 * @class
 */
class Path ***REMOVED***
    /**
     * Creates a new Path with the given segments.
     *
     * @private
     * @internal
     * @private
     * @param segments Sequence of parts of a path.
     */
    constructor(segments) ***REMOVED***
        this.segments = segments;
  ***REMOVED***
    /**
     * Returns the number of segments of this field path.
     *
     * @private
     * @internal
     */
    get size() ***REMOVED***
        return this.segments.length;
  ***REMOVED***
    /**
     * Create a child path beneath the current level.
     *
     * @private
     * @internal
     * @param relativePath Relative path to append to the current path.
     * @returns The new path.
     */
    append(relativePath) ***REMOVED***
        if (relativePath instanceof Path) ***REMOVED***
            return this.construct(this.segments.concat(relativePath.segments));
      ***REMOVED***
        return this.construct(this.segments.concat(this.split(relativePath)));
  ***REMOVED***
    /**
     * Returns the path of the parent node.
     *
     * @private
     * @internal
     * @returns The new path or null if we are already at the root.
     */
    parent() ***REMOVED***
        if (this.segments.length === 0) ***REMOVED***
            return null;
      ***REMOVED***
        return this.construct(this.segments.slice(0, this.segments.length - 1));
  ***REMOVED***
    /**
     * Checks whether the current path is a prefix of the specified path.
     *
     * @private
     * @internal
     * @param other The path to check against.
     * @returns 'true' iff the current path is a prefix match with 'other'.
     */
    isPrefixOf(other) ***REMOVED***
        if (other.segments.length < this.segments.length) ***REMOVED***
            return false;
      ***REMOVED***
        for (let i = 0; i < this.segments.length; i++) ***REMOVED***
            if (this.segments[i] !== other.segments[i]) ***REMOVED***
                return false;
          ***REMOVED***
      ***REMOVED***
        return true;
  ***REMOVED***
    /**
     * Compare the current path against another Path object.
     *
     * @private
     * @internal
     * @param other The path to compare to.
     * @returns -1 if current < other, 1 if current > other, 0 if equal
     */
    compareTo(other) ***REMOVED***
        const len = Math.min(this.segments.length, other.segments.length);
        for (let i = 0; i < len; i++) ***REMOVED***
            if (this.segments[i] < other.segments[i]) ***REMOVED***
                return -1;
          ***REMOVED***
            if (this.segments[i] > other.segments[i]) ***REMOVED***
                return 1;
          ***REMOVED***
      ***REMOVED***
        if (this.segments.length < other.segments.length) ***REMOVED***
            return -1;
      ***REMOVED***
        if (this.segments.length > other.segments.length) ***REMOVED***
            return 1;
      ***REMOVED***
        return 0;
  ***REMOVED***
    /**
     * Returns a copy of the underlying segments.
     *
     * @private
     * @internal
     * @returns A copy of the segments that make up this path.
     */
    toArray() ***REMOVED***
        return this.segments.slice();
  ***REMOVED***
    /**
     * Pops the last segment from this `Path` and returns a newly constructed
     * `Path`.
     *
     * @private
     * @internal
     * @returns The newly created Path.
     */
    popLast() ***REMOVED***
        this.segments.pop();
        return this.construct(this.segments);
  ***REMOVED***
    /**
     * Returns true if this `Path` is equal to the provided value.
     *
     * @private
     * @internal
     * @param other The value to compare against.
     * @return true if this `Path` is equal to the provided value.
     */
    isEqual(other) ***REMOVED***
        return this === other || this.compareTo(other) === 0;
  ***REMOVED***
}
/**
 * A slash-separated path for navigating resources within the current Firestore
 * instance.
 *
 * @private
 * @internal
 */
class ResourcePath extends Path ***REMOVED***
    /**
     * Constructs a ResourcePath.
     *
     * @private
     * @internal
     * @param segments Sequence of names of the parts of the path.
     */
    constructor(...segments) ***REMOVED***
        super(segments);
  ***REMOVED***
    /**
     * Indicates whether this path points to a document.
     * @private
     * @internal
     */
    get isDocument() ***REMOVED***
        return this.segments.length > 0 && this.segments.length % 2 === 0;
  ***REMOVED***
    /**
     * Indicates whether this path points to a collection.
     * @private
     * @internal
     */
    get isCollection() ***REMOVED***
        return this.segments.length % 2 === 1;
  ***REMOVED***
    /**
     * The last component of the path.
     * @private
     * @internal
     */
    get id() ***REMOVED***
        if (this.segments.length > 0) ***REMOVED***
            return this.segments[this.segments.length - 1];
      ***REMOVED***
        return null;
  ***REMOVED***
    /**
     * Returns the location of this path relative to the root of the project's
     * database.
     * @private
     * @internal
     */
    get relativeName() ***REMOVED***
        return this.segments.join('/');
  ***REMOVED***
    /**
     * Constructs a new instance of ResourcePath.
     *
     * @private
     * @internal
     * @param segments Sequence of parts of the path.
     * @returns The newly created ResourcePath.
     */
    construct(segments) ***REMOVED***
        return new ResourcePath(...segments);
  ***REMOVED***
    /**
     * Splits a string into path segments, using slashes as separators.
     *
     * @private
     * @internal
     * @param relativePath The path to split.
     * @returns The split path segments.
     */
    split(relativePath) ***REMOVED***
        // We may have an empty segment at the beginning or end if they had a
        // leading or trailing slash (which we allow).
        return relativePath.split('/').filter(segment => segment.length > 0);
  ***REMOVED***
    /**
     * Converts this path to a fully qualified ResourcePath.
     *
     * @private
     * @internal
     * @param projectIdIfMissing The project ID of the current Firestore project.
     * The project ID is only used if it's not provided as part of this
     * ResourcePath.
     * @return A fully-qualified resource path pointing to the same element.
     */
    toQualifiedResourcePath(projectIdIfMissing) ***REMOVED***
        return new QualifiedResourcePath(projectIdIfMissing, exports.DEFAULT_DATABASE_ID, ...this.segments);
  ***REMOVED***
}
exports.ResourcePath = ResourcePath;
/**
 * A default instance pointing to the root collection.
 * @private
 * @internal
 */
ResourcePath.EMPTY = new ResourcePath();
/**
 * A slash-separated path that includes a project and database ID for referring
 * to resources in any Firestore project.
 *
 * @private
 * @internal
 */
class QualifiedResourcePath extends ResourcePath ***REMOVED***
    /**
     * Constructs a Firestore Resource Path.
     *
     * @private
     * @internal
     * @param projectId The Firestore project id.
     * @param databaseId The Firestore database id.
     * @param segments Sequence of names of the parts of the path.
     */
    constructor(projectId, databaseId, ...segments) ***REMOVED***
        super(...segments);
        this.projectId = projectId;
        this.databaseId = databaseId;
  ***REMOVED***
    /**
     * String representation of the path relative to the database root.
     * @private
     * @internal
     */
    get relativeName() ***REMOVED***
        return this.segments.join('/');
  ***REMOVED***
    /**
     * Creates a resource path from an absolute Firestore path.
     *
     * @private
     * @internal
     * @param absolutePath A string representation of a Resource Path.
     * @returns The new ResourcePath.
     */
    static fromSlashSeparatedString(absolutePath) ***REMOVED***
        const elements = RESOURCE_PATH_RE.exec(absolutePath);
        if (elements) ***REMOVED***
            const project = elements[1];
            const database = elements[2];
            const path = elements[3];
            return new QualifiedResourcePath(project, database).append(path);
      ***REMOVED***
        throw new Error(`Resource name '$***REMOVED***absolutePath}' is not valid.`);
  ***REMOVED***
    /**
     * Create a child path beneath the current level.
     *
     * @private
     * @internal
     * @param relativePath Relative path to append to the current path.
     * @returns The new path.
     */
    append(relativePath) ***REMOVED***
        // `super.append()` calls `QualifiedResourcePath.construct()` when invoked
        // from here and returns a QualifiedResourcePath.
        return super.append(relativePath);
  ***REMOVED***
    /**
     * Create a child path beneath the current level.
     *
     * @private
     * @internal
     * @returns The new path.
     */
    parent() ***REMOVED***
        return super.parent();
  ***REMOVED***
    /**
     * String representation of a ResourcePath as expected by the API.
     *
     * @private
     * @internal
     * @returns The representation as expected by the API.
     */
    get formattedName() ***REMOVED***
        const components = [
            'projects',
            this.projectId,
            'databases',
            this.databaseId,
            'documents',
            ...this.segments,
        ];
        return components.join('/');
  ***REMOVED***
    /**
     * Constructs a new instance of ResourcePath. We need this instead of using
     * the normal constructor because polymorphic 'this' doesn't work on static
     * methods.
     *
     * @private
     * @internal
     * @param segments Sequence of names of the parts of the path.
     * @returns The newly created QualifiedResourcePath.
     */
    construct(segments) ***REMOVED***
        return new QualifiedResourcePath(this.projectId, this.databaseId, ...segments);
  ***REMOVED***
    /**
     * Convenience method to match the ResourcePath API. This method always
     * returns the current instance.
     *
     * @private
     * @internal
     */
    toQualifiedResourcePath() ***REMOVED***
        return this;
  ***REMOVED***
    /**
     * Compare the current path against another ResourcePath object.
     *
     * @private
     * @internal
     * @param other The path to compare to.
     * @returns -1 if current < other, 1 if current > other, 0 if equal
     */
    compareTo(other) ***REMOVED***
        if (other instanceof QualifiedResourcePath) ***REMOVED***
            if (this.projectId < other.projectId) ***REMOVED***
                return -1;
          ***REMOVED***
            if (this.projectId > other.projectId) ***REMOVED***
                return 1;
          ***REMOVED***
            if (this.databaseId < other.databaseId) ***REMOVED***
                return -1;
          ***REMOVED***
            if (this.databaseId > other.databaseId) ***REMOVED***
                return 1;
          ***REMOVED***
      ***REMOVED***
        return super.compareTo(other);
  ***REMOVED***
    /**
     * Converts this ResourcePath to the Firestore Proto representation.
     * @private
     * @internal
     */
    toProto() ***REMOVED***
        return ***REMOVED***
            referenceValue: this.formattedName,
      ***REMOVED***;
  ***REMOVED***
}
exports.QualifiedResourcePath = QualifiedResourcePath;
/**
 * Validates that the given string can be used as a relative or absolute
 * resource path.
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param resourcePath The path to validate.
 * @throws if the string can't be used as a resource path.
 */
function validateResourcePath(arg, resourcePath) ***REMOVED***
    if (typeof resourcePath !== 'string' || resourcePath === '') ***REMOVED***
        throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'resource path')} Path must be a non-empty string.`);
  ***REMOVED***
    if (resourcePath.indexOf('//') >= 0) ***REMOVED***
        throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'resource path')} Paths must not contain //.`);
  ***REMOVED***
}
exports.validateResourcePath = validateResourcePath;
/**
 * A dot-separated path for navigating sub-objects (e.g. nested maps) within a document.
 *
 * @class
 */
class FieldPath extends Path ***REMOVED***
    /**
     * Constructs a Firestore Field Path.
     *
     * @param ***REMOVED***...string} segments Sequence of field names that form this path.
     *
     * @example
     * ```
     * let query = firestore.collection('col');
     * let fieldPath = new FieldPath('f.o.o', 'bar');
     *
     * query.where(fieldPath, '==', 42).get().then(snapshot => ***REMOVED***
     *   snapshot.forEach(document => ***REMOVED***
     *     console.log(`Document contains ***REMOVED***'f.o.o' : ***REMOVED***'bar' : 42}}`);
     * ***REMOVED***);
     * });
     * ```
     */
    constructor(...segments) ***REMOVED***
        if (Array.isArray(segments[0])) ***REMOVED***
            throw new Error('The FieldPath constructor no longer supports an array as its first argument. ' +
                'Please unpack your array and call FieldPath() with individual arguments.');
      ***REMOVED***
        (0, validate_1.validateMinNumberOfArguments)('FieldPath', segments, 1);
        for (let i = 0; i < segments.length; ++i) ***REMOVED***
            (0, validate_1.validateString)(i, segments[i]);
            if (segments[i].length === 0) ***REMOVED***
                throw new Error(`Element at index $***REMOVED***i} should not be an empty string.`);
          ***REMOVED***
      ***REMOVED***
        super(segments);
  ***REMOVED***
    /**
     * A special FieldPath value to refer to the ID of a document. It can be used
     * in queries to sort or filter by the document ID.
     *
     * @returns ***REMOVED***FieldPath}
     */
    static documentId() ***REMOVED***
        return FieldPath._DOCUMENT_ID;
  ***REMOVED***
    /**
     * Turns a field path argument into a [FieldPath]***REMOVED***@link FieldPath}.
     * Supports FieldPaths as input (which are passed through) and dot-separated
     * strings.
     *
     * @private
     * @internal
     * @param ***REMOVED***string|FieldPath} fieldPath The FieldPath to create.
     * @returns ***REMOVED***FieldPath} A field path representation.
     */
    static fromArgument(fieldPath) ***REMOVED***
        // validateFieldPath() is used in all public API entry points to validate
        // that fromArgument() is only called with a Field Path or a string.
        return fieldPath instanceof FieldPath
            ? fieldPath
            : new FieldPath(...fieldPath.split('.'));
  ***REMOVED***
    /**
     * String representation of a FieldPath as expected by the API.
     *
     * @private
     * @internal
     * @override
     * @returns ***REMOVED***string} The representation as expected by the API.
     */
    get formattedName() ***REMOVED***
        return this.segments
            .map(str => ***REMOVED***
            return UNESCAPED_FIELD_NAME_RE.test(str)
                ? str
                : '`' + str.replace('\\', '\\\\').replace('`', '\\`') + '`';
      ***REMOVED***)
            .join('.');
  ***REMOVED***
    /**
     * Returns a string representation of this path.
     *
     * @private
     * @internal
     * @returns A string representing this path.
     */
    toString() ***REMOVED***
        return this.formattedName;
  ***REMOVED***
    /**
     * Splits a string into path segments, using dots as separators.
     *
     * @private
     * @internal
     * @override
     * @param ***REMOVED***string} fieldPath The path to split.
     * @returns ***REMOVED***Array.<string>} - The split path segments.
     */
    split(fieldPath) ***REMOVED***
        return fieldPath.split('.');
  ***REMOVED***
    /**
     * Constructs a new instance of FieldPath. We need this instead of using
     * the normal constructor because polymorphic 'this' doesn't work on static
     * methods.
     *
     * @private
     * @internal
     * @override
     * @param segments Sequence of field names.
     * @returns The newly created FieldPath.
     */
    construct(segments) ***REMOVED***
        return new FieldPath(...segments);
  ***REMOVED***
    /**
     * Returns true if this `FieldPath` is equal to the provided value.
     *
     * @param ***REMOVED****} other The value to compare against.
     * @return ***REMOVED***boolean} true if this `FieldPath` is equal to the provided value.
     */
    isEqual(other) ***REMOVED***
        return super.isEqual(other);
  ***REMOVED***
}
exports.FieldPath = FieldPath;
/**
 * A special sentinel value to refer to the ID of a document.
 *
 * @private
 * @internal
 */
FieldPath._DOCUMENT_ID = new FieldPath('__name__');
/**
 * Validates that the provided value can be used as a field path argument.
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param fieldPath The value to verify.
 * @throws if the string can't be used as a field path.
 */
function validateFieldPath(arg, fieldPath) ***REMOVED***
    if (fieldPath instanceof FieldPath) ***REMOVED***
        return;
  ***REMOVED***
    if (fieldPath === undefined) ***REMOVED***
        throw new Error((0, validate_1.invalidArgumentMessage)(arg, 'field path') + ' The path cannot be omitted.');
  ***REMOVED***
    if ((0, util_1.isObject)(fieldPath) && fieldPath.constructor.name === 'FieldPath') ***REMOVED***
        throw new Error((0, validate_1.customObjectMessage)(arg, fieldPath));
  ***REMOVED***
    if (typeof fieldPath !== 'string') ***REMOVED***
        throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'field path')} Paths can only be specified as strings or via a FieldPath object.`);
  ***REMOVED***
    if (fieldPath.indexOf('..') >= 0) ***REMOVED***
        throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'field path')} Paths must not contain ".." in them.`);
  ***REMOVED***
    if (fieldPath.startsWith('.') || fieldPath.endsWith('.')) ***REMOVED***
        throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'field path')} Paths must not start or end with ".".`);
  ***REMOVED***
    if (!FIELD_PATH_RE.test(fieldPath)) ***REMOVED***
        throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'field path')} Paths can't be empty and must not contain
    "*~/[]".`);
  ***REMOVED***
}
exports.validateFieldPath = validateFieldPath;
