// @google-cloud/firestore/build/src/watch.js
"use strict";
/*!
 * Copyright 2017 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.QueryWatch = exports.DocumentWatch = exports.WATCH_IDLE_TIMEOUT_MS = void 0;
const assert = require("assert");
const rbtree = require("functional-red-black-tree");
const google_gax_1 = require("google-gax");
const backoff_1 = require("./backoff");
const document_1 = require("./document");
const document_change_1 = require("./document-change");
const logger_1 = require("./logger");
const path_1 = require("./path");
const timestamp_1 = require("./timestamp");
const types_1 = require("./types");
const util_1 = require("./util");
/*!
 * Target ID used by watch. Watch uses a fixed target id since we only support
 * one target per stream.
 * @type ***REMOVED***number}
 */
const WATCH_TARGET_ID = 0x1;
/*!
 * Idle timeout used to detect Watch streams that stall (see
 * https://github.com/googleapis/nodejs-firestore/issues/1057, b/156308554).
 * Under normal load, the Watch backend will send a TARGET_CHANGE message
 * roughly every 30 seconds. As discussed with the backend team, we reset the
 * Watch stream if we do not receive any message within 120 seconds.
 */
exports.WATCH_IDLE_TIMEOUT_MS = 120 * 1000;
/*!
 * Sentinel value for a document remove.
 */
const REMOVED = ***REMOVED***};
/*!
 * The change type for document change events.
 */
// tslint:disable-next-line:variable-name
const ChangeType = ***REMOVED***
    added: 'added',
    modified: 'modified',
    removed: 'removed',
};
/*!
 * The comparator used for document watches (which should always get called with
 * the same document).
 */
const DOCUMENT_WATCH_COMPARATOR = (doc1, doc2) => ***REMOVED***
    assert(doc1 === doc2, 'Document watches only support one document.');
    return 0;
};
const EMPTY_FUNCTION = () => ***REMOVED*** };
/**
 * Watch provides listen functionality and exposes the 'onSnapshot' observer. It
 * can be used with a valid Firestore Listen target.
 *
 * @class
 * @private
 * @internal
 */
class Watch ***REMOVED***
    /**
     * @private
     * @internal
     *
     * @param firestore The Firestore Database client.
     */
    constructor(firestore, _converter = (0, types_1.defaultConverter)()) ***REMOVED***
        this._converter = _converter;
        /**
         * Indicates whether we are interested in data from the stream. Set to false in the
         * 'unsubscribe()' callback.
         * @private
         * @internal
         */
        this.isActive = true;
        /**
         * The current stream to the backend.
         * @private
         * @internal
         */
        this.currentStream = null;
        /**
         * The server assigns and updates the resume token.
         * @private
         * @internal
         */
        this.resumeToken = undefined;
        /**
         * A map of document names to QueryDocumentSnapshots for the last sent snapshot.
         * @private
         * @internal
         */
        this.docMap = new Map();
        /**
         * The accumulated map of document changes (keyed by document name) for the
         * current snapshot.
         * @private
         * @internal
         */
        this.changeMap = new Map();
        /**
         * The current state of the query results. *
         * @private
         * @internal
         */
        this.current = false;
        /**
         * We need this to track whether we've pushed an initial set of changes,
         * since we should push those even when there are no changes, if there
         * aren't docs.
         * @private
         * @internal
         */
        this.hasPushed = false;
        this.firestore = firestore;
        this.backoff = new backoff_1.ExponentialBackoff();
        this.requestTag = (0, util_1.requestTag)();
        this.onNext = EMPTY_FUNCTION;
        this.onError = EMPTY_FUNCTION;
  ***REMOVED***
    /**
     * Starts a watch and attaches a listener for document change events.
     *
     * @private
     * @internal
     * @param onNext A callback to be called every time a new snapshot is
     * available.
     * @param onError A callback to be called if the listen fails or is cancelled.
     * No further callbacks will occur.
     *
     * @returns An unsubscribe function that can be called to cancel the snapshot
     * listener.
     */
    onSnapshot(onNext, onError) ***REMOVED***
        assert(this.onNext === EMPTY_FUNCTION, 'onNext should not already be defined.');
        assert(this.onError === EMPTY_FUNCTION, 'onError should not already be defined.');
        assert(this.docTree === undefined, 'docTree should not already be defined.');
        this.onNext = onNext;
        this.onError = onError;
        this.docTree = rbtree(this.getComparator());
        this.initStream();
        const unsubscribe = () => ***REMOVED***
            (0, logger_1.logger)('Watch.onSnapshot', this.requestTag, 'Unsubscribe called');
            // Prevent further callbacks.
            this.onNext = () => ***REMOVED*** };
            this.onError = () => ***REMOVED*** };
            this.shutdown();
      ***REMOVED***;
        this.firestore.registerListener();
        return unsubscribe;
  ***REMOVED***
    /**
     * Returns the current count of all documents, including the changes from
     * the current changeMap.
     * @private
     * @internal
     */
    currentSize() ***REMOVED***
        const changes = this.extractCurrentChanges(timestamp_1.Timestamp.now());
        return this.docMap.size + changes.adds.length - changes.deletes.length;
  ***REMOVED***
    /**
     * Splits up document changes into removals, additions, and updates.
     * @private
     * @internal
     */
    extractCurrentChanges(readTime) ***REMOVED***
        const deletes = [];
        const adds = [];
        const updates = [];
        this.changeMap.forEach((value, name) => ***REMOVED***
            if (value === REMOVED) ***REMOVED***
                if (this.docMap.has(name)) ***REMOVED***
                    deletes.push(name);
              ***REMOVED***
          ***REMOVED***
            else if (this.docMap.has(name)) ***REMOVED***
                value.readTime = readTime;
                updates.push(value.build());
          ***REMOVED***
            else ***REMOVED***
                value.readTime = readTime;
                adds.push(value.build());
          ***REMOVED***
      ***REMOVED***);
        return ***REMOVED*** deletes, adds, updates };
  ***REMOVED***
    /**
     * Helper to clear the docs on RESET or filter mismatch.
     * @private
     * @internal
     */
    resetDocs() ***REMOVED***
        (0, logger_1.logger)('Watch.resetDocs', this.requestTag, 'Resetting documents');
        this.changeMap.clear();
        this.resumeToken = undefined;
        this.docTree.forEach((snapshot) => ***REMOVED***
            // Mark each document as deleted. If documents are not deleted, they
            // will be send again by the server.
            this.changeMap.set(snapshot.ref.path, REMOVED);
      ***REMOVED***);
        this.current = false;
  ***REMOVED***
    /**
     * Closes the stream and calls onError() if the stream is still active.
     * @private
     * @internal
     */
    closeStream(err) ***REMOVED***
        if (this.isActive) ***REMOVED***
            (0, logger_1.logger)('Watch.closeStream', this.requestTag, 'Invoking onError: ', err);
            this.onError(err);
      ***REMOVED***
        this.shutdown();
  ***REMOVED***
    /**
     * Re-opens the stream unless the specified error is considered permanent.
     * Clears the change map.
     * @private
     * @internal
     */
    maybeReopenStream(err) ***REMOVED***
        if (this.isActive && !this.isPermanentWatchError(err)) ***REMOVED***
            (0, logger_1.logger)('Watch.maybeReopenStream', this.requestTag, 'Stream ended, re-opening after retryable error:', err);
            this.changeMap.clear();
            if (this.isResourceExhaustedError(err)) ***REMOVED***
                this.backoff.resetToMax();
          ***REMOVED***
            this.initStream();
      ***REMOVED***
        else ***REMOVED***
            this.closeStream(err);
      ***REMOVED***
  ***REMOVED***
    /**
     * Cancels the current idle timeout and reschedules a new timer.
     *
     * @private
     * @internal
     */
    resetIdleTimeout() ***REMOVED***
        if (this.idleTimeoutHandle) ***REMOVED***
            clearTimeout(this.idleTimeoutHandle);
      ***REMOVED***
        this.idleTimeoutHandle = (0, backoff_1.delayExecution)(() => ***REMOVED***
            var _a;
            (0, logger_1.logger)('Watch.resetIdleTimeout', this.requestTag, 'Resetting stream after idle timeout');
            (_a = this.currentStream) === null || _a === void 0 ? void 0 : _a.end();
            this.currentStream = null;
            const error = new google_gax_1.GoogleError('Watch stream idle timeout');
            error.code = google_gax_1.Status.UNKNOWN;
            this.maybeReopenStream(error);
      ***REMOVED***, exports.WATCH_IDLE_TIMEOUT_MS);
  ***REMOVED***
    /**
     * Helper to restart the outgoing stream to the backend.
     * @private
     * @internal
     */
    resetStream() ***REMOVED***
        (0, logger_1.logger)('Watch.resetStream', this.requestTag, 'Restarting stream');
        if (this.currentStream) ***REMOVED***
            this.currentStream.end();
            this.currentStream = null;
      ***REMOVED***
        this.initStream();
  ***REMOVED***
    /**
     * Initializes a new stream to the backend with backoff.
     * @private
     * @internal
     */
    initStream() ***REMOVED***
        this.backoff
            .backoffAndWait()
            .then(async () => ***REMOVED***
            if (!this.isActive) ***REMOVED***
                (0, logger_1.logger)('Watch.initStream', this.requestTag, 'Not initializing inactive stream');
                return;
          ***REMOVED***
            await this.firestore.initializeIfNeeded(this.requestTag);
            const request = ***REMOVED***};
            request.database = this.firestore.formattedName;
            request.addTarget = this.getTarget(this.resumeToken);
            // Note that we need to call the internal _listen API to pass additional
            // header values in readWriteStream.
            return this.firestore
                .requestStream('listen', 
            /* bidirectional= */ true, request, this.requestTag)
                .then(backendStream => ***REMOVED***
                if (!this.isActive) ***REMOVED***
                    (0, logger_1.logger)('Watch.initStream', this.requestTag, 'Closing inactive stream');
                    backendStream.emit('end');
                    return;
              ***REMOVED***
                (0, logger_1.logger)('Watch.initStream', this.requestTag, 'Opened new stream');
                this.currentStream = backendStream;
                this.resetIdleTimeout();
                this.currentStream.on('data', (proto) => ***REMOVED***
                    this.resetIdleTimeout();
                    this.onData(proto);
              ***REMOVED***)
                    .on('error', err => ***REMOVED***
                    if (this.currentStream === backendStream) ***REMOVED***
                        this.currentStream = null;
                        this.maybeReopenStream(err);
                  ***REMOVED***
              ***REMOVED***)
                    .on('end', () => ***REMOVED***
                    if (this.currentStream === backendStream) ***REMOVED***
                        this.currentStream = null;
                        const err = new google_gax_1.GoogleError('Stream ended unexpectedly');
                        err.code = google_gax_1.Status.UNKNOWN;
                        this.maybeReopenStream(err);
                  ***REMOVED***
              ***REMOVED***);
                this.currentStream.resume();
          ***REMOVED***);
      ***REMOVED***)
            .catch(err => ***REMOVED***
            this.closeStream(err);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Handles 'data' events and closes the stream if the response type is
     * invalid.
     * @private
     * @internal
     */
    onData(proto) ***REMOVED***
        if (proto.targetChange) ***REMOVED***
            (0, logger_1.logger)('Watch.onData', this.requestTag, 'Processing target change');
            const change = proto.targetChange;
            const noTargetIds = !change.targetIds || change.targetIds.length === 0;
            if (change.targetChangeType === 'NO_CHANGE') ***REMOVED***
                if (noTargetIds && change.readTime && this.current) ***REMOVED***
                    // This means everything is up-to-date, so emit the current
                    // set of docs as a snapshot, if there were changes.
                    this.pushSnapshot(timestamp_1.Timestamp.fromProto(change.readTime), change.resumeToken);
              ***REMOVED***
          ***REMOVED***
            else if (change.targetChangeType === 'ADD') ***REMOVED***
                if (WATCH_TARGET_ID !== change.targetIds[0]) ***REMOVED***
                    this.closeStream(Error('Unexpected target ID sent by server'));
              ***REMOVED***
          ***REMOVED***
            else if (change.targetChangeType === 'REMOVE') ***REMOVED***
                let code = google_gax_1.Status.INTERNAL;
                let message = 'internal error';
                if (change.cause) ***REMOVED***
                    code = change.cause.code;
                    message = change.cause.message;
              ***REMOVED***
                // @todo: Surface a .code property on the exception.
                this.closeStream(new Error('Error ' + code + ': ' + message));
          ***REMOVED***
            else if (change.targetChangeType === 'RESET') ***REMOVED***
                // Whatever changes have happened so far no longer matter.
                this.resetDocs();
          ***REMOVED***
            else if (change.targetChangeType === 'CURRENT') ***REMOVED***
                this.current = true;
          ***REMOVED***
            else ***REMOVED***
                this.closeStream(new Error('Unknown target change type: ' + JSON.stringify(change)));
          ***REMOVED***
            if (change.resumeToken &&
                this.affectsTarget(change.targetIds, WATCH_TARGET_ID)) ***REMOVED***
                this.backoff.reset();
          ***REMOVED***
      ***REMOVED***
        else if (proto.documentChange) ***REMOVED***
            (0, logger_1.logger)('Watch.onData', this.requestTag, 'Processing change event');
            // No other targetIds can show up here, but we still need to see
            // if the targetId was in the added list or removed list.
            const targetIds = proto.documentChange.targetIds || [];
            const removedTargetIds = proto.documentChange.removedTargetIds || [];
            let changed = false;
            let removed = false;
            for (let i = 0; i < targetIds.length; i++) ***REMOVED***
                if (targetIds[i] === WATCH_TARGET_ID) ***REMOVED***
                    changed = true;
              ***REMOVED***
          ***REMOVED***
            for (let i = 0; i < removedTargetIds.length; i++) ***REMOVED***
                if (removedTargetIds[i] === WATCH_TARGET_ID) ***REMOVED***
                    removed = true;
              ***REMOVED***
          ***REMOVED***
            const document = proto.documentChange.document;
            const name = document.name;
            const relativeName = path_1.QualifiedResourcePath.fromSlashSeparatedString(name).relativeName;
            if (changed) ***REMOVED***
                (0, logger_1.logger)('Watch.onData', this.requestTag, 'Received document change');
                const ref = this.firestore.doc(relativeName);
                const snapshot = new document_1.DocumentSnapshotBuilder(ref.withConverter(this._converter));
                snapshot.fieldsProto = document.fields || ***REMOVED***};
                snapshot.createTime = timestamp_1.Timestamp.fromProto(document.createTime);
                snapshot.updateTime = timestamp_1.Timestamp.fromProto(document.updateTime);
                this.changeMap.set(relativeName, snapshot);
          ***REMOVED***
            else if (removed) ***REMOVED***
                (0, logger_1.logger)('Watch.onData', this.requestTag, 'Received document remove');
                this.changeMap.set(relativeName, REMOVED);
          ***REMOVED***
      ***REMOVED***
        else if (proto.documentDelete || proto.documentRemove) ***REMOVED***
            (0, logger_1.logger)('Watch.onData', this.requestTag, 'Processing remove event');
            const name = (proto.documentDelete || proto.documentRemove).document;
            const relativeName = path_1.QualifiedResourcePath.fromSlashSeparatedString(name).relativeName;
            this.changeMap.set(relativeName, REMOVED);
      ***REMOVED***
        else if (proto.filter) ***REMOVED***
            (0, logger_1.logger)('Watch.onData', this.requestTag, 'Processing filter update');
            if (proto.filter.count !== this.currentSize()) ***REMOVED***
                // We need to remove all the current results.
                this.resetDocs();
                // The filter didn't match, so re-issue the query.
                this.resetStream();
          ***REMOVED***
      ***REMOVED***
        else ***REMOVED***
            this.closeStream(new Error('Unknown listen response type: ' + JSON.stringify(proto)));
      ***REMOVED***
  ***REMOVED***
    /**
     * Checks if the current target id is included in the list of target ids.
     * If no targetIds are provided, returns true.
     * @private
     * @internal
     */
    affectsTarget(targetIds, currentId) ***REMOVED***
        if (targetIds === undefined || targetIds.length === 0) ***REMOVED***
            return true;
      ***REMOVED***
        for (const targetId of targetIds) ***REMOVED***
            if (targetId === currentId) ***REMOVED***
                return true;
          ***REMOVED***
      ***REMOVED***
        return false;
  ***REMOVED***
    /**
     * Assembles a new snapshot from the current set of changes and invokes the
     * user's callback. Clears the current changes on completion.
     * @private
     * @internal
     */
    pushSnapshot(readTime, nextResumeToken) ***REMOVED***
        const appliedChanges = this.computeSnapshot(readTime);
        if (!this.hasPushed || appliedChanges.length > 0) ***REMOVED***
            (0, logger_1.logger)('Watch.pushSnapshot', this.requestTag, 'Sending snapshot with %d changes and %d documents', String(appliedChanges.length), this.docTree.length);
            // We pass the current set of changes, even if `docTree` is modified later.
            const currentTree = this.docTree;
            this.onNext(readTime, currentTree.length, () => currentTree.keys, () => appliedChanges);
            this.hasPushed = true;
      ***REMOVED***
        this.changeMap.clear();
        this.resumeToken = nextResumeToken;
  ***REMOVED***
    /**
     * Applies a document delete to the document tree and the document map.
     * Returns the corresponding DocumentChange event.
     * @private
     * @internal
     */
    deleteDoc(name) ***REMOVED***
        assert(this.docMap.has(name), 'Document to delete does not exist');
        const oldDocument = this.docMap.get(name);
        const existing = this.docTree.find(oldDocument);
        const oldIndex = existing.index;
        this.docTree = existing.remove();
        this.docMap.delete(name);
        return new document_change_1.DocumentChange(ChangeType.removed, oldDocument, oldIndex, -1);
  ***REMOVED***
    /**
     * Applies a document add to the document tree and the document map. Returns
     * the corresponding DocumentChange event.
     * @private
     * @internal
     */
    addDoc(newDocument) ***REMOVED***
        const name = newDocument.ref.path;
        assert(!this.docMap.has(name), 'Document to add already exists');
        this.docTree = this.docTree.insert(newDocument, null);
        const newIndex = this.docTree.find(newDocument).index;
        this.docMap.set(name, newDocument);
        return new document_change_1.DocumentChange(ChangeType.added, newDocument, -1, newIndex);
  ***REMOVED***
    /**
     * Applies a document modification to the document tree and the document map.
     * Returns the DocumentChange event for successful modifications.
     * @private
     * @internal
     */
    modifyDoc(newDocument) ***REMOVED***
        const name = newDocument.ref.path;
        assert(this.docMap.has(name), 'Document to modify does not exist');
        const oldDocument = this.docMap.get(name);
        if (!oldDocument.updateTime.isEqual(newDocument.updateTime)) ***REMOVED***
            const removeChange = this.deleteDoc(name);
            const addChange = this.addDoc(newDocument);
            return new document_change_1.DocumentChange(ChangeType.modified, newDocument, removeChange.oldIndex, addChange.newIndex);
      ***REMOVED***
        return null;
  ***REMOVED***
    /**
     * Applies the mutations in changeMap to both the document tree and the
     * document lookup map. Modified docMap in-place and returns the updated
     * state.
     * @private
     * @internal
     */
    computeSnapshot(readTime) ***REMOVED***
        const changeSet = this.extractCurrentChanges(readTime);
        const appliedChanges = [];
        // Process the sorted changes in the order that is expected by our clients
        // (removals, additions, and then modifications). We also need to sort the
        // individual changes to assure that oldIndex/newIndex keep incrementing.
        changeSet.deletes.sort((name1, name2) => ***REMOVED***
            // Deletes are sorted based on the order of the existing document.
            return this.getComparator()(this.docMap.get(name1), this.docMap.get(name2));
      ***REMOVED***);
        changeSet.deletes.forEach(name => ***REMOVED***
            const change = this.deleteDoc(name);
            appliedChanges.push(change);
      ***REMOVED***);
        changeSet.adds.sort(this.getComparator());
        changeSet.adds.forEach(snapshot => ***REMOVED***
            const change = this.addDoc(snapshot);
            appliedChanges.push(change);
      ***REMOVED***);
        changeSet.updates.sort(this.getComparator());
        changeSet.updates.forEach(snapshot => ***REMOVED***
            const change = this.modifyDoc(snapshot);
            if (change) ***REMOVED***
                appliedChanges.push(change);
          ***REMOVED***
      ***REMOVED***);
        assert(this.docTree.length === this.docMap.size, 'The update document ' +
            'tree and document map should have the same number of entries.');
        return appliedChanges;
  ***REMOVED***
    /**
     * Determines whether a watch error is considered permanent and should not be
     * retried. Errors that don't provide a GRPC error code are always considered
     * transient in this context.
     *
     * @private
     * @internal
     * @param error An error object.
     * @return Whether the error is permanent.
     */
    isPermanentWatchError(error) ***REMOVED***
        if (error.code === undefined) ***REMOVED***
            (0, logger_1.logger)('Watch.isPermanentError', this.requestTag, 'Unable to determine error code: ', error);
            return false;
      ***REMOVED***
        switch (error.code) ***REMOVED***
            case google_gax_1.Status.ABORTED:
            case google_gax_1.Status.CANCELLED:
            case google_gax_1.Status.UNKNOWN:
            case google_gax_1.Status.DEADLINE_EXCEEDED:
            case google_gax_1.Status.RESOURCE_EXHAUSTED:
            case google_gax_1.Status.INTERNAL:
            case google_gax_1.Status.UNAVAILABLE:
            case google_gax_1.Status.UNAUTHENTICATED:
                return false;
            default:
                return true;
      ***REMOVED***
  ***REMOVED***
    /**
     * Determines whether we need to initiate a longer backoff due to system
     * overload.
     *
     * @private
     * @internal
     * @param error A GRPC Error object that exposes an error code.
     * @return Whether we need to back off our retries.
     */
    isResourceExhaustedError(error) ***REMOVED***
        return error.code === google_gax_1.Status.RESOURCE_EXHAUSTED;
  ***REMOVED***
    /** Closes the stream and clears all timeouts. */
    shutdown() ***REMOVED***
        var _a;
        if (this.isActive) ***REMOVED***
            this.isActive = false;
            if (this.idleTimeoutHandle) ***REMOVED***
                clearTimeout(this.idleTimeoutHandle);
                this.idleTimeoutHandle = undefined;
          ***REMOVED***
            this.firestore.unregisterListener();
      ***REMOVED***
        (_a = this.currentStream) === null || _a === void 0 ? void 0 : _a.end();
        this.currentStream = null;
  ***REMOVED***
}
/**
 * Creates a new Watch instance to listen on DocumentReferences.
 *
 * @private
 * @internal
 */
class DocumentWatch extends Watch ***REMOVED***
    constructor(firestore, ref) ***REMOVED***
        super(firestore, ref._converter);
        this.ref = ref;
  ***REMOVED***
    getComparator() ***REMOVED***
        return DOCUMENT_WATCH_COMPARATOR;
  ***REMOVED***
    getTarget(resumeToken) ***REMOVED***
        const formattedName = this.ref.formattedName;
        return ***REMOVED***
            documents: ***REMOVED***
                documents: [formattedName],
          ***REMOVED***,
            targetId: WATCH_TARGET_ID,
            resumeToken,
      ***REMOVED***;
  ***REMOVED***
}
exports.DocumentWatch = DocumentWatch;
/**
 * Creates a new Watch instance to listen on Queries.
 *
 * @private
 * @internal
 */
class QueryWatch extends Watch ***REMOVED***
    constructor(firestore, query, converter) ***REMOVED***
        super(firestore, converter);
        this.query = query;
        this.comparator = query.comparator();
  ***REMOVED***
    getComparator() ***REMOVED***
        return this.query.comparator();
  ***REMOVED***
    getTarget(resumeToken) ***REMOVED***
        const query = this.query.toProto();
        return ***REMOVED*** query, targetId: WATCH_TARGET_ID, resumeToken };
  ***REMOVED***
}
exports.QueryWatch = QueryWatch;
