// @google-cloud/firestore/build/src/write-batch.js
"use strict";
/*!
 * Copyright 2019 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.validateFieldValue = exports.validateDocumentData = exports.validateSetOptions = exports.WriteBatch = exports.WriteResult = void 0;
const document_1 = require("./document");
const logger_1 = require("./logger");
const path_1 = require("./path");
const reference_1 = require("./reference");
const serializer_1 = require("./serializer");
const timestamp_1 = require("./timestamp");
const util_1 = require("./util");
const validate_1 = require("./validate");
/**
 * A WriteResult wraps the write time set by the Firestore servers on sets(),
 * updates(), and creates().
 *
 * @class WriteResult
 */
class WriteResult ***REMOVED***
    /**
     * @private
     *
     * @param _writeTime The time of the corresponding document write.
     */
    constructor(_writeTime) ***REMOVED***
        this._writeTime = _writeTime;
  ***REMOVED***
    /**
     * The write time as set by the Firestore servers.
     *
     * @type ***REMOVED***Timestamp}
     * @name WriteResult#writeTime
     * @readonly
     *
     * @example
     * ```
     * let documentRef = firestore.doc('col/doc');
     *
     * documentRef.set(***REMOVED***foo: 'bar'}).then(writeResult => ***REMOVED***
     *   console.log(`Document written at: $***REMOVED***writeResult.writeTime.toDate()}`);
     * });
     * ```
     */
    get writeTime() ***REMOVED***
        return this._writeTime;
  ***REMOVED***
    /**
     * Returns true if this `WriteResult` is equal to the provided value.
     *
     * @param ***REMOVED****} other The value to compare against.
     * @return true if this `WriteResult` is equal to the provided value.
     */
    isEqual(other) ***REMOVED***
        return (this === other ||
            (other instanceof WriteResult &&
                this._writeTime.isEqual(other._writeTime)));
  ***REMOVED***
}
exports.WriteResult = WriteResult;
/**
 * A Firestore WriteBatch that can be used to atomically commit multiple write
 * operations at once.
 *
 * @class WriteBatch
 */
class WriteBatch ***REMOVED***
    /** @private */
    constructor(firestore) ***REMOVED***
        /**
         * An array of document paths and the corresponding write operations that are
         * executed as part of the commit. The resulting `api.IWrite` will be sent to
         * the backend.
         *
         * @private
         * @internal
         */
        this._ops = [];
        this._committed = false;
        this._firestore = firestore;
        this._serializer = new serializer_1.Serializer(firestore);
        this._allowUndefined = !!firestore._settings.ignoreUndefinedProperties;
  ***REMOVED***
    /**
     * The number of writes in this batch.
     * @private
     */
    get _opCount() ***REMOVED***
        return this._ops.length;
  ***REMOVED***
    /**
     * Checks if this write batch has any pending operations.
     *
     * @private
     * @internal
     */
    get isEmpty() ***REMOVED***
        return this._ops.length === 0;
  ***REMOVED***
    /**
     * Throws an error if this batch has already been committed.
     *
     * @private
     * @internal
     */
    verifyNotCommitted() ***REMOVED***
        if (this._committed) ***REMOVED***
            throw new Error('Cannot modify a WriteBatch that has been committed.');
      ***REMOVED***
  ***REMOVED***
    /**
     * Create a document with the provided object values. This will fail the batch
     * if a document exists at its location.
     *
     * @param ***REMOVED***DocumentReference} documentRef A reference to the document to be
     * created.
     * @param ***REMOVED***T} data The object to serialize as the document.
     * @throws ***REMOVED***Error} If the provided input is not a valid Firestore document.
     * @returns ***REMOVED***WriteBatch} This WriteBatch instance. Used for chaining
     * method calls.
     *
     * @example
     * ```
     * let writeBatch = firestore.batch();
     * let documentRef = firestore.collection('col').doc();
     *
     * writeBatch.create(documentRef, ***REMOVED***foo: 'bar'});
     *
     * writeBatch.commit().then(() => ***REMOVED***
     *   console.log('Successfully executed batch.');
     * });
     * ```
     */
    create(documentRef, data) ***REMOVED***
        const ref = (0, reference_1.validateDocumentReference)('documentRef', documentRef);
        const firestoreData = ref._converter.toFirestore(data);
        validateDocumentData('data', firestoreData, 
        /* allowDeletes= */ false, this._allowUndefined);
        this.verifyNotCommitted();
        const transform = document_1.DocumentTransform.fromObject(ref, firestoreData);
        transform.validate();
        const precondition = new document_1.Precondition(***REMOVED*** exists: false });
        const op = () => ***REMOVED***
            const document = document_1.DocumentSnapshot.fromObject(ref, firestoreData);
            const write = document.toWriteProto();
            if (!transform.isEmpty) ***REMOVED***
                write.updateTransforms = transform.toProto(this._serializer);
          ***REMOVED***
            write.currentDocument = precondition.toProto();
            return write;
      ***REMOVED***;
        this._ops.push(***REMOVED*** docPath: documentRef.path, op });
        return this;
  ***REMOVED***
    /**
     * Deletes a document from the database.
     *
     * @param ***REMOVED***DocumentReference} documentRef A reference to the document to be
     * deleted.
     * @param ***REMOVED***Precondition=} precondition A precondition to enforce for this
     * delete.
     * @param ***REMOVED***Timestamp=} precondition.lastUpdateTime If set, enforces that the
     * document was last updated at lastUpdateTime. Fails the batch if the
     * document doesn't exist or was last updated at a different time.
     * @param ***REMOVED***boolean= } precondition.exists If set to true, enforces that the target
     * document must or must not exist.
     * @returns ***REMOVED***WriteBatch} This WriteBatch instance. Used for chaining
     * method calls.
     *
     * @example
     * ```
     * let writeBatch = firestore.batch();
     * let documentRef = firestore.doc('col/doc');
     *
     * writeBatch.delete(documentRef);
     *
     * writeBatch.commit().then(() => ***REMOVED***
     *   console.log('Successfully executed batch.');
     * });
     * ```
     */
    delete(documentRef, precondition) ***REMOVED***
        const ref = (0, reference_1.validateDocumentReference)('documentRef', documentRef);
        validateDeletePrecondition('precondition', precondition, ***REMOVED*** optional: true });
        this.verifyNotCommitted();
        const conditions = new document_1.Precondition(precondition);
        const op = () => ***REMOVED***
            const write = ***REMOVED*** delete: ref.formattedName };
            if (!conditions.isEmpty) ***REMOVED***
                write.currentDocument = conditions.toProto();
          ***REMOVED***
            return write;
      ***REMOVED***;
        this._ops.push(***REMOVED*** docPath: documentRef.path, op });
        return this;
  ***REMOVED***
    /**
     * Write to the document referred to by the provided
     * [DocumentReference]***REMOVED***@link DocumentReference}. If the document does not
     * exist yet, it will be created. If you pass [SetOptions]***REMOVED***@link SetOptions},
     * the provided data can be merged into the existing document.
     *
     * @param ***REMOVED***DocumentReference} documentRef A reference to the document to be
     * set.
     * @param ***REMOVED***T|Partial<T>} data The object to serialize as the document.
     * @param ***REMOVED***SetOptions=} options An object to configure the set behavior.
     * @param ***REMOVED***boolean=} options.merge - If true, set() merges the values
     * specified in its data argument. Fields omitted from this set() call
     * remain untouched. If your input sets any field to an empty map, all nested
     * fields are overwritten.
     * @param ***REMOVED***Array.<string|FieldPath>=} options.mergeFields - If provided,
     * set() only replaces the specified field paths. Any field path that is no
     * specified is ignored and remains untouched. If your input sets any field to
     * an empty map, all nested fields are overwritten.
     * @throws ***REMOVED***Error} If the provided input is not a valid Firestore document.
     * @returns ***REMOVED***WriteBatch} This WriteBatch instance. Used for chaining
     * method calls.
     *
     * @example
     * ```
     * let writeBatch = firestore.batch();
     * let documentRef = firestore.doc('col/doc');
     *
     * writeBatch.set(documentRef, ***REMOVED***foo: 'bar'});
     *
     * writeBatch.commit().then(() => ***REMOVED***
     *   console.log('Successfully executed batch.');
     * });
     * ```
     */
    set(documentRef, data, options) ***REMOVED***
        validateSetOptions('options', options, ***REMOVED*** optional: true });
        const mergeLeaves = options && 'merge' in options && options.merge;
        const mergePaths = options && 'mergeFields' in options;
        const ref = (0, reference_1.validateDocumentReference)('documentRef', documentRef);
        let firestoreData;
        if (mergeLeaves || mergePaths) ***REMOVED***
            // Cast to any in order to satisfy the union type constraint on
            // toFirestore().
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            firestoreData = ref._converter.toFirestore(data, options);
      ***REMOVED***
        else ***REMOVED***
            firestoreData = ref._converter.toFirestore(data);
      ***REMOVED***
        validateDocumentData('data', firestoreData, 
        /* allowDeletes= */ !!(mergePaths || mergeLeaves), this._allowUndefined);
        this.verifyNotCommitted();
        let documentMask;
        if (mergePaths) ***REMOVED***
            documentMask = document_1.DocumentMask.fromFieldMask(options.mergeFields);
            firestoreData = documentMask.applyTo(firestoreData);
      ***REMOVED***
        const transform = document_1.DocumentTransform.fromObject(documentRef, firestoreData);
        transform.validate();
        const op = () => ***REMOVED***
            const document = document_1.DocumentSnapshot.fromObject(documentRef, firestoreData);
            if (mergePaths) ***REMOVED***
                documentMask.removeFields(transform.fields);
          ***REMOVED***
            else if (mergeLeaves) ***REMOVED***
                documentMask = document_1.DocumentMask.fromObject(firestoreData);
          ***REMOVED***
            const write = document.toWriteProto();
            if (!transform.isEmpty) ***REMOVED***
                write.updateTransforms = transform.toProto(this._serializer);
          ***REMOVED***
            if (mergePaths || mergeLeaves) ***REMOVED***
                write.updateMask = documentMask.toProto();
          ***REMOVED***
            return write;
      ***REMOVED***;
        this._ops.push(***REMOVED*** docPath: documentRef.path, op });
        return this;
  ***REMOVED***
    /**
     * Update fields of the document referred to by the provided
     * [DocumentReference]***REMOVED***@link DocumentReference}. If the document
     * doesn't yet exist, the update fails and the entire batch will be rejected.
     *
     * The update() method accepts either an object with field paths encoded as
     * keys and field values encoded as values, or a variable number of arguments
     * that alternate between field paths and field values. Nested fields can be
     * updated by providing dot-separated field path strings or by providing
     * FieldPath objects.
     *
     * A Precondition restricting this update can be specified as the last
     * argument.
     *
     * @param ***REMOVED***DocumentReference} documentRef A reference to the document to be
     * updated.
     * @param ***REMOVED***UpdateData|string|FieldPath} dataOrField An object
     * containing the fields and values with which to update the document
     * or the path of the first field to update.
     * @param ***REMOVED***
     * ...(Precondition|*|string|FieldPath)} preconditionOrValues -
     * An alternating list of field paths and values to update or a Precondition
     * to restrict this update.
     * @throws ***REMOVED***Error} If the provided input is not valid Firestore data.
     * @returns ***REMOVED***WriteBatch} This WriteBatch instance. Used for chaining
     * method calls.
     *
     * @example
     * ```
     * let writeBatch = firestore.batch();
     * let documentRef = firestore.doc('col/doc');
     *
     * writeBatch.update(documentRef, ***REMOVED***foo: 'bar'});
     *
     * writeBatch.commit().then(() => ***REMOVED***
     *   console.log('Successfully executed batch.');
     * });
     * ```
     */
    update(documentRef, dataOrField, ...preconditionOrValues) ***REMOVED***
        // eslint-disable-next-line prefer-rest-params
        (0, validate_1.validateMinNumberOfArguments)('WriteBatch.update', arguments, 2);
        (0, reference_1.validateDocumentReference)('documentRef', documentRef);
        this.verifyNotCommitted();
        const updateMap = new Map();
        let precondition = new document_1.Precondition(***REMOVED*** exists: true });
        const argumentError = 'Update() requires either a single JavaScript ' +
            'object or an alternating list of field/value pairs that can be ' +
            'followed by an optional precondition.';
        const usesVarargs = typeof dataOrField === 'string' || dataOrField instanceof path_1.FieldPath;
        if (usesVarargs) ***REMOVED***
            const argumentOffset = 1; // Respect 'documentRef' in the error message
            const fieldOrValues = [dataOrField, ...preconditionOrValues];
            try ***REMOVED***
                for (let i = 0; i < fieldOrValues.length; i += 2) ***REMOVED***
                    if (i === fieldOrValues.length - 1) ***REMOVED***
                        const maybePrecondition = fieldOrValues[i];
                        validateUpdatePrecondition(i + argumentOffset, maybePrecondition);
                        precondition = new document_1.Precondition(maybePrecondition);
                  ***REMOVED***
                    else ***REMOVED***
                        const maybeFieldPath = fieldOrValues[i];
                        (0, path_1.validateFieldPath)(i + argumentOffset, maybeFieldPath);
                        // Unlike the `validateMinNumberOfArguments` invocation above, this
                        // validation can be triggered both from `WriteBatch.update()` and
                        // `DocumentReference.update()`. Hence, we don't use the fully
                        // qualified API name in the error message.
                        (0, validate_1.validateMinNumberOfArguments)('update', fieldOrValues, i + 1);
                        const fieldPath = path_1.FieldPath.fromArgument(maybeFieldPath);
                        validateFieldValue(i + argumentOffset, fieldOrValues[i + 1], this._allowUndefined, fieldPath);
                        updateMap.set(fieldPath, fieldOrValues[i + 1]);
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
            catch (err) ***REMOVED***
                (0, logger_1.logger)('WriteBatch.update', null, 'Varargs validation failed:', err);
                // We catch the validation error here and re-throw to provide a better
                // error message.
                throw new Error(`$***REMOVED***argumentError} $***REMOVED***err.message}`);
          ***REMOVED***
      ***REMOVED***
        else ***REMOVED***
            try ***REMOVED***
                validateUpdateMap('dataOrField', dataOrField, this._allowUndefined);
                // eslint-disable-next-line prefer-rest-params
                (0, validate_1.validateMaxNumberOfArguments)('update', arguments, 3);
                Object.entries(dataOrField).forEach(([key, value]) => ***REMOVED***
                    // Skip `undefined` values (can be hit if `ignoreUndefinedProperties`
                    // is set)
                    if (value !== undefined) ***REMOVED***
                        (0, path_1.validateFieldPath)(key, key);
                        updateMap.set(path_1.FieldPath.fromArgument(key), value);
                  ***REMOVED***
              ***REMOVED***);
                if (preconditionOrValues.length > 0) ***REMOVED***
                    validateUpdatePrecondition('preconditionOrValues', preconditionOrValues[0]);
                    precondition = new document_1.Precondition(preconditionOrValues[0]);
              ***REMOVED***
          ***REMOVED***
            catch (err) ***REMOVED***
                (0, logger_1.logger)('WriteBatch.update', null, 'Non-varargs validation failed:', err);
                // We catch the validation error here and prefix the error with a custom
                // message to describe the usage of update() better.
                throw new Error(`$***REMOVED***argumentError} $***REMOVED***err.message}`);
          ***REMOVED***
      ***REMOVED***
        validateNoConflictingFields('dataOrField', updateMap);
        const transform = document_1.DocumentTransform.fromUpdateMap(documentRef, updateMap);
        transform.validate();
        const documentMask = document_1.DocumentMask.fromUpdateMap(updateMap);
        const op = () => ***REMOVED***
            const document = document_1.DocumentSnapshot.fromUpdateMap(documentRef, updateMap);
            const write = document.toWriteProto();
            write.updateMask = documentMask.toProto();
            if (!transform.isEmpty) ***REMOVED***
                write.updateTransforms = transform.toProto(this._serializer);
          ***REMOVED***
            write.currentDocument = precondition.toProto();
            return write;
      ***REMOVED***;
        this._ops.push(***REMOVED*** docPath: documentRef.path, op });
        return this;
  ***REMOVED***
    /**
     * Atomically commits all pending operations to the database and verifies all
     * preconditions. Fails the entire write if any precondition is not met.
     *
     * @returns ***REMOVED***Promise.<Array.<WriteResult>>} A Promise that resolves
     * when this batch completes.
     *
     * @example
     * ```
     * let writeBatch = firestore.batch();
     * let documentRef = firestore.doc('col/doc');
     *
     * writeBatch.set(documentRef, ***REMOVED***foo: 'bar'});
     *
     * writeBatch.commit().then(() => ***REMOVED***
     *   console.log('Successfully executed batch.');
     * });
     * ```
     */
    commit() ***REMOVED***
        // Capture the error stack to preserve stack tracing across async calls.
        const stack = Error().stack;
        // Commits should also be retried when they fail with status code ABORTED.
        const retryCodes = [10 /* StatusCode.ABORTED */, ...(0, util_1.getRetryCodes)('commit')];
        return this._commit(***REMOVED*** retryCodes })
            .then(response => ***REMOVED***
            return (response.writeResults || []).map(writeResult => new WriteResult(timestamp_1.Timestamp.fromProto(writeResult.updateTime || response.commitTime)));
      ***REMOVED***)
            .catch(err => ***REMOVED***
            throw (0, util_1.wrapError)(err, stack);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Commit method that takes an optional transaction ID.
     *
     * @private
     * @internal
     * @param commitOptions Options to use for this commit.
     * @param commitOptions.transactionId The transaction ID of this commit.
     * @param commitOptions.requestTag A unique client-assigned identifier for
     * this request.
     * @returns  A Promise that resolves when this batch completes.
     */
    async _commit(commitOptions) ***REMOVED***
        var _a;
        // Note: We don't call `verifyNotCommitted()` to allow for retries.
        this._committed = true;
        const tag = (_a = commitOptions === null || commitOptions === void 0 ? void 0 : commitOptions.requestTag) !== null && _a !== void 0 ? _a : (0, util_1.requestTag)();
        await this._firestore.initializeIfNeeded(tag);
        // Note that the request may not always be of type ICommitRequest. This is
        // just here to ensure type safety.
        const request = ***REMOVED***
            database: this._firestore.formattedName,
            writes: this._ops.map(op => op.op()),
      ***REMOVED***;
        if (commitOptions === null || commitOptions === void 0 ? void 0 : commitOptions.transactionId) ***REMOVED***
            request.transaction = commitOptions.transactionId;
      ***REMOVED***
        (0, logger_1.logger)('WriteBatch.commit', tag, 'Sending %d writes', request.writes.length);
        return this._firestore.request((commitOptions === null || commitOptions === void 0 ? void 0 : commitOptions.methodName) || 'commit', request, tag, commitOptions === null || commitOptions === void 0 ? void 0 : commitOptions.retryCodes);
  ***REMOVED***
    /**
     * Resets the WriteBatch and dequeues all pending operations.
     * @private
     * @internal
     */
    _reset() ***REMOVED***
        this._ops.splice(0);
        this._committed = false;
  ***REMOVED***
}
exports.WriteBatch = WriteBatch;
/**
 * Validates the use of 'value' as a Precondition and enforces that 'exists'
 * and 'lastUpdateTime' use valid types.
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param value The object to validate
 * @param allowExists Whether to allow the 'exists' preconditions.
 */
function validatePrecondition(arg, value, allowExists) ***REMOVED***
    if (typeof value !== 'object' || value === null) ***REMOVED***
        throw new Error('Input is not an object.');
  ***REMOVED***
    const precondition = value;
    let conditions = 0;
    if (precondition.exists !== undefined) ***REMOVED***
        ++conditions;
        if (!allowExists) ***REMOVED***
            throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'precondition')} "exists" is not an allowed precondition.`);
      ***REMOVED***
        if (typeof precondition.exists !== 'boolean') ***REMOVED***
            throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'precondition')} "exists" is not a boolean.'`);
      ***REMOVED***
  ***REMOVED***
    if (precondition.lastUpdateTime !== undefined) ***REMOVED***
        ++conditions;
        if (!(precondition.lastUpdateTime instanceof timestamp_1.Timestamp)) ***REMOVED***
            throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'precondition')} "lastUpdateTime" is not a Firestore Timestamp.`);
      ***REMOVED***
  ***REMOVED***
    if (conditions > 1) ***REMOVED***
        throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'precondition')} Input specifies more than one precondition.`);
  ***REMOVED***
}
/**
 * Validates the use of 'value' as an update Precondition.
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param value The object to validate.
 * @param options Optional validation options specifying whether the value can
 * be omitted.
 */
function validateUpdatePrecondition(arg, value, options) ***REMOVED***
    if (!(0, validate_1.validateOptional)(value, options)) ***REMOVED***
        validatePrecondition(arg, value, /* allowExists= */ false);
  ***REMOVED***
}
/**
 * Validates the use of 'value' as a delete Precondition.
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param value The object to validate.
 * @param options Optional validation options specifying whether the value can
 * be omitted.
 */
function validateDeletePrecondition(arg, value, options) ***REMOVED***
    if (!(0, validate_1.validateOptional)(value, options)) ***REMOVED***
        validatePrecondition(arg, value, /* allowExists= */ true);
  ***REMOVED***
}
/**
 * Validates the use of 'value' as SetOptions and enforces that 'merge' is a
 * boolean.
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param value The object to validate.
 * @param options Optional validation options specifying whether the value can
 * be omitted.
 * @throws if the input is not a valid SetOptions object.
 */
function validateSetOptions(arg, value, options) ***REMOVED***
    if (!(0, validate_1.validateOptional)(value, options)) ***REMOVED***
        if (!(0, util_1.isObject)(value)) ***REMOVED***
            throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'set() options argument')} Input is not an object.`);
      ***REMOVED***
        const setOptions = value;
        if ('mergeFields' in setOptions) ***REMOVED***
            for (let i = 0; i < setOptions.mergeFields.length; ++i) ***REMOVED***
                try ***REMOVED***
                    (0, path_1.validateFieldPath)(i, setOptions.mergeFields[i]);
              ***REMOVED***
                catch (err) ***REMOVED***
                    throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'set() options argument')} "mergeFields" is not valid: $***REMOVED***err.message}`);
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        if ('merge' in setOptions && 'mergeFields' in setOptions) ***REMOVED***
            throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'set() options argument')} You cannot specify both "merge" and "mergeFields".`);
      ***REMOVED***
  ***REMOVED***
}
exports.validateSetOptions = validateSetOptions;
/**
 * Validates a JavaScript object for usage as a Firestore document.
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param obj JavaScript object to validate.
 * @param allowDeletes Whether to allow FieldValue.delete() sentinels.
 * @param allowUndefined Whether to allow nested properties that are `undefined`.
 * @throws when the object is invalid.
 */
function validateDocumentData(arg, obj, allowDeletes, allowUndefined) ***REMOVED***
    if (!(0, util_1.isPlainObject)(obj)) ***REMOVED***
        throw new Error((0, validate_1.customObjectMessage)(arg, obj));
  ***REMOVED***
    (0, serializer_1.validateUserInput)(arg, obj, 'Firestore document', ***REMOVED***
        allowDeletes: allowDeletes ? 'all' : 'none',
        allowTransforms: true,
        allowUndefined,
  ***REMOVED***);
}
exports.validateDocumentData = validateDocumentData;
/**
 * Validates that a value can be used as field value during an update.
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param val The value to verify.
 * @param allowUndefined Whether to allow nested properties that are `undefined`.
 * @param path The path to show in the error message.
 */
function validateFieldValue(arg, val, allowUndefined, path) ***REMOVED***
    (0, serializer_1.validateUserInput)(arg, val, 'Firestore value', ***REMOVED*** allowDeletes: 'root', allowTransforms: true, allowUndefined }, path);
}
exports.validateFieldValue = validateFieldValue;
/**
 * Validates that the update data does not contain any ambiguous field
 * definitions (such as 'a.b' and 'a').
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param data An update map with field/value pairs.
 */
function validateNoConflictingFields(arg, data) ***REMOVED***
    const fields = [];
    data.forEach((value, key) => ***REMOVED***
        fields.push(key);
  ***REMOVED***);
    fields.sort((left, right) => left.compareTo(right));
    for (let i = 1; i < fields.length; ++i) ***REMOVED***
        if (fields[i - 1].isPrefixOf(fields[i])) ***REMOVED***
            throw new Error(`$***REMOVED***(0, validate_1.invalidArgumentMessage)(arg, 'update map')} Field "$***REMOVED***fields[i - 1]}" was specified multiple times.`);
      ***REMOVED***
  ***REMOVED***
}
/**
 * Validates that a JavaScript object is a map of field paths to field values.
 *
 * @private
 * @internal
 * @param arg The argument name or argument index (for varargs methods).
 * @param obj JavaScript object to validate.
 * @param allowUndefined Whether to allow nested properties that are `undefined`.
 * @throws when the object is invalid.
 */
function validateUpdateMap(arg, obj, allowUndefined) ***REMOVED***
    if (!(0, util_1.isPlainObject)(obj)) ***REMOVED***
        throw new Error((0, validate_1.customObjectMessage)(arg, obj));
  ***REMOVED***
    if (Object.keys(obj).length === 0) ***REMOVED***
        throw new Error('At least one field must be updated.');
  ***REMOVED***
    validateFieldValue(arg, obj, allowUndefined);
}
