// babel-plugin-styled-components/lib/utils/detectors.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
});
exports.isWithThemeHelper = exports.isValidTopLevelImport = exports.isUseTheme = exports.isStyled = exports.isPureHelper = exports.isKeyframesHelper = exports.isInjectGlobalHelper = exports.isHelper = exports.isCreateGlobalStyleHelper = exports.isCSSHelper = exports.importLocalName = void 0;

var _options = require("./options");

const VALID_TOP_LEVEL_IMPORT_PATH_MATCHERS = ['styled-components', 'styled-components/no-tags', 'styled-components/native', 'styled-components/primitives'].map(literal => x => x === literal);

const isValidTopLevelImport = (x, state) => ***REMOVED***
  return [...VALID_TOP_LEVEL_IMPORT_PATH_MATCHERS, ...(0, _options.useTopLevelImportPathMatchers)(state)].some(isMatch => isMatch(x));
};

exports.isValidTopLevelImport = isValidTopLevelImport;
const localNameCache = ***REMOVED***};

const importLocalName = (name, state, options = ***REMOVED***}) => ***REMOVED***
  const ***REMOVED***
    cacheIdentifier,
    bypassCache = false
***REMOVED*** = options;
  const cacheKeyAffix = cacheIdentifier ? `|$***REMOVED***cacheIdentifier}` : '';
  const cacheKey = name + state.file.opts.filename + cacheKeyAffix;

  if (!bypassCache && cacheKey in localNameCache) ***REMOVED***
    return localNameCache[cacheKey]; // state.customImportName is injected by the babel macro
***REMOVED*** else if (state.customImportName) ***REMOVED***
    return state.customImportName.name;
***REMOVED***

  let localName = state.styledRequired ? name === 'default' ? 'styled' : name : false;
  state.file.path.traverse(***REMOVED***
    ImportDeclaration: ***REMOVED***
      exit(path) ***REMOVED***
        const ***REMOVED***
          node
      ***REMOVED*** = path;

        if (isValidTopLevelImport(node.source.value, state)) ***REMOVED***
          for (const specifier of path.get('specifiers')) ***REMOVED***
            if (specifier.isImportSpecifier() && specifier.node.imported.name === 'styled') ***REMOVED***
              localName = 'styled';
          ***REMOVED***

            if (specifier.isImportDefaultSpecifier()) ***REMOVED***
              localName = specifier.node.local.name;
          ***REMOVED***

            if (specifier.isImportSpecifier() && specifier.node.imported.name === name) ***REMOVED***
              localName = specifier.node.local.name;
          ***REMOVED***

            if (specifier.isImportNamespaceSpecifier()) ***REMOVED***
              localName = name === 'default' ? specifier.node.local.name : name;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

  ***REMOVED***
***REMOVED***);
  localNameCache[cacheKey] = localName;
  return localName;
};

exports.importLocalName = importLocalName;

const isStyled = t => (tag, state) => ***REMOVED***
  if (t.isCallExpression(tag) && t.isMemberExpression(tag.callee) && tag.callee.property.name !== 'default'
  /** ignore default for #93 below */
  ) ***REMOVED***
    // styled.something()
    return isStyled(t)(tag.callee.object, state);
***REMOVED*** else ***REMOVED***
    return t.isMemberExpression(tag) && tag.object.name === importLocalName('default', state, ***REMOVED***
      cacheIdentifier: tag.object.name
  ***REMOVED***) && !isHelper(t)(tag.property, state) || t.isCallExpression(tag) && tag.callee.name === importLocalName('default', state, ***REMOVED***
      cacheIdentifier: tag.callee.name
  ***REMOVED***) ||
    /**
     * #93 Support require()
     * styled-components might be imported using a require()
     * call and assigned to a variable of any name.
     * - styled.default.div``
     * - styled.default.something()
     */
    state.styledRequired && t.isMemberExpression(tag) && t.isMemberExpression(tag.object) && tag.object.property.name === 'default' && tag.object.object.name === state.styledRequired || state.styledRequired && t.isCallExpression(tag) && t.isMemberExpression(tag.callee) && tag.callee.property.name === 'default' && tag.callee.object.name === state.styledRequired || importLocalName('default', state) && t.isMemberExpression(tag) && t.isMemberExpression(tag.object) && tag.object.property.name === 'default' && tag.object.object.name === importLocalName('default', state) || importLocalName('default', state) && t.isCallExpression(tag) && t.isMemberExpression(tag.callee) && tag.object.property.name === 'default' && tag.object.object.name === importLocalName('default', state);
***REMOVED***
};

exports.isStyled = isStyled;

const isCSSHelper = t => (tag, state) => t.isIdentifier(tag) && tag.name === importLocalName('css', state);

exports.isCSSHelper = isCSSHelper;

const isCreateGlobalStyleHelper = t => (tag, state) => t.isIdentifier(tag) && tag.name === importLocalName('createGlobalStyle', state);

exports.isCreateGlobalStyleHelper = isCreateGlobalStyleHelper;

const isInjectGlobalHelper = t => (tag, state) => t.isIdentifier(tag) && tag.name === importLocalName('injectGlobal', state);

exports.isInjectGlobalHelper = isInjectGlobalHelper;

const isKeyframesHelper = t => (tag, state) => t.isIdentifier(tag) && tag.name === importLocalName('keyframes', state);

exports.isKeyframesHelper = isKeyframesHelper;

const isWithThemeHelper = t => (tag, state) => t.isIdentifier(tag) && tag.name === importLocalName('withTheme', state);

exports.isWithThemeHelper = isWithThemeHelper;

const isUseTheme = t => (tag, state) => t.isIdentifier(tag) && tag.name === importLocalName('useTheme', state);

exports.isUseTheme = isUseTheme;

const isHelper = t => (tag, state) => isCreateGlobalStyleHelper(t)(tag, state) || isCSSHelper(t)(tag, state) || isInjectGlobalHelper(t)(tag, state) || isUseTheme(t)(tag, state) || isKeyframesHelper(t)(tag, state) || isWithThemeHelper(t)(tag, state);

exports.isHelper = isHelper;

const isPureHelper = t => (tag, state) => isCreateGlobalStyleHelper(t)(tag, state) || isCSSHelper(t)(tag, state) || isKeyframesHelper(t)(tag, state) || isUseTheme(t)(tag, state) || isWithThemeHelper(t)(tag, state);

exports.isPureHelper = isPureHelper;
