// @babel/traverse/lib/path/context.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
});
exports._call = _call;
exports._getQueueContexts = _getQueueContexts;
exports._resyncKey = _resyncKey;
exports._resyncList = _resyncList;
exports._resyncParent = _resyncParent;
exports._resyncRemoved = _resyncRemoved;
exports.call = call;
exports.isBlacklisted = exports.isDenylisted = isDenylisted;
exports.popContext = popContext;
exports.pushContext = pushContext;
exports.requeue = requeue;
exports.resync = resync;
exports.setContext = setContext;
exports.setKey = setKey;
exports.setScope = setScope;
exports.setup = setup;
exports.skip = skip;
exports.skipKey = skipKey;
exports.stop = stop;
exports.visit = visit;
var _traverseNode = require("../traverse-node");
var _index = require("./index");
function call(key) ***REMOVED***
  const opts = this.opts;
  this.debug(key);
  if (this.node) ***REMOVED***
    if (this._call(opts[key])) return true;
***REMOVED***
  if (this.node) ***REMOVED***
    return this._call(opts[this.node.type] && opts[this.node.type][key]);
***REMOVED***
  return false;
}
function _call(fns) ***REMOVED***
  if (!fns) return false;
  for (const fn of fns) ***REMOVED***
    if (!fn) continue;
    const node = this.node;
    if (!node) return true;
    const ret = fn.call(this.state, this, this.state);
    if (ret && typeof ret === "object" && typeof ret.then === "function") ***REMOVED***
      throw new Error(`You appear to be using a plugin with an async traversal visitor, ` + `which your current version of Babel does not support. ` + `If you're using a published plugin, you may need to upgrade ` + `your @babel/core version.`);
  ***REMOVED***
    if (ret) ***REMOVED***
      throw new Error(`Unexpected return value from visitor method $***REMOVED***fn}`);
  ***REMOVED***
    if (this.node !== node) return true;
    if (this._traverseFlags > 0) return true;
***REMOVED***
  return false;
}
function isDenylisted() ***REMOVED***
  var _this$opts$denylist;
  const denylist = (_this$opts$denylist = this.opts.denylist) != null ? _this$opts$denylist : this.opts.blacklist;
  return denylist && denylist.indexOf(this.node.type) > -1;
}
function restoreContext(path, context) ***REMOVED***
  if (path.context !== context) ***REMOVED***
    path.context = context;
    path.state = context.state;
    path.opts = context.opts;
***REMOVED***
}
function visit() ***REMOVED***
  if (!this.node) ***REMOVED***
    return false;
***REMOVED***
  if (this.isDenylisted()) ***REMOVED***
    return false;
***REMOVED***
  if (this.opts.shouldSkip && this.opts.shouldSkip(this)) ***REMOVED***
    return false;
***REMOVED***
  const currentContext = this.context;
  if (this.shouldSkip || this.call("enter")) ***REMOVED***
    this.debug("Skip...");
    return this.shouldStop;
***REMOVED***
  restoreContext(this, currentContext);
  this.debug("Recursing into...");
  this.shouldStop = (0, _traverseNode.traverseNode)(this.node, this.opts, this.scope, this.state, this, this.skipKeys);
  restoreContext(this, currentContext);
  this.call("exit");
  return this.shouldStop;
}
function skip() ***REMOVED***
  this.shouldSkip = true;
}
function skipKey(key) ***REMOVED***
  if (this.skipKeys == null) ***REMOVED***
    this.skipKeys = ***REMOVED***};
***REMOVED***
  this.skipKeys[key] = true;
}
function stop() ***REMOVED***
  this._traverseFlags |= _index.SHOULD_SKIP | _index.SHOULD_STOP;
}
function setScope() ***REMOVED***
  if (this.opts && this.opts.noScope) return;
  let path = this.parentPath;
  if ((this.key === "key" || this.listKey === "decorators") && path.isMethod() || this.key === "discriminant" && path.isSwitchStatement()) ***REMOVED***
    path = path.parentPath;
***REMOVED***
  let target;
  while (path && !target) ***REMOVED***
    if (path.opts && path.opts.noScope) return;
    target = path.scope;
    path = path.parentPath;
***REMOVED***
  this.scope = this.getScope(target);
  if (this.scope) this.scope.init();
}
function setContext(context) ***REMOVED***
  if (this.skipKeys != null) ***REMOVED***
    this.skipKeys = ***REMOVED***};
***REMOVED***
  this._traverseFlags = 0;
  if (context) ***REMOVED***
    this.context = context;
    this.state = context.state;
    this.opts = context.opts;
***REMOVED***
  this.setScope();
  return this;
}
function resync() ***REMOVED***
  if (this.removed) return;
  this._resyncParent();
  this._resyncList();
  this._resyncKey();
}
function _resyncParent() ***REMOVED***
  if (this.parentPath) ***REMOVED***
    this.parent = this.parentPath.node;
***REMOVED***
}
function _resyncKey() ***REMOVED***
  if (!this.container) return;
  if (this.node === this.container[this.key]) ***REMOVED***
    return;
***REMOVED***
  if (Array.isArray(this.container)) ***REMOVED***
    for (let i = 0; i < this.container.length; i++) ***REMOVED***
      if (this.container[i] === this.node) ***REMOVED***
        return this.setKey(i);
    ***REMOVED***
  ***REMOVED***
***REMOVED*** else ***REMOVED***
    for (const key of Object.keys(this.container)) ***REMOVED***
      if (this.container[key] === this.node) ***REMOVED***
        return this.setKey(key);
    ***REMOVED***
  ***REMOVED***
***REMOVED***
  this.key = null;
}
function _resyncList() ***REMOVED***
  if (!this.parent || !this.inList) return;
  const newContainer = this.parent[this.listKey];
  if (this.container === newContainer) return;
  this.container = newContainer || null;
}
function _resyncRemoved() ***REMOVED***
  if (this.key == null || !this.container || this.container[this.key] !== this.node) ***REMOVED***
    this._markRemoved();
***REMOVED***
}
function popContext() ***REMOVED***
  this.contexts.pop();
  if (this.contexts.length > 0) ***REMOVED***
    this.setContext(this.contexts[this.contexts.length - 1]);
***REMOVED*** else ***REMOVED***
    this.setContext(undefined);
***REMOVED***
}
function pushContext(context) ***REMOVED***
  this.contexts.push(context);
  this.setContext(context);
}
function setup(parentPath, container, listKey, key) ***REMOVED***
  this.listKey = listKey;
  this.container = container;
  this.parentPath = parentPath || this.parentPath;
  this.setKey(key);
}
function setKey(key) ***REMOVED***
  var _this$node;
  this.key = key;
  this.node = this.container[this.key];
  this.type = (_this$node = this.node) == null ? void 0 : _this$node.type;
}
function requeue(pathToQueue = this) ***REMOVED***
  if (pathToQueue.removed) return;
  ;
  const contexts = this.contexts;
  for (const context of contexts) ***REMOVED***
    context.maybeQueue(pathToQueue);
***REMOVED***
}
function _getQueueContexts() ***REMOVED***
  let path = this;
  let contexts = this.contexts;
  while (!contexts.length) ***REMOVED***
    path = path.parentPath;
    if (!path) break;
    contexts = path.contexts;
***REMOVED***
  return contexts;
}
