// yargs-parser/browser.js
// Main entrypoint for ESM web browser environments. Avoids using Node.js
// specific libraries, such as "path".
//
// TODO: figure out reasonable web equivalents for "resolve", "normalize", etc.
import ***REMOVED*** camelCase, decamelize, looksLikeNumber } from './build/lib/string-utils.js'
import ***REMOVED*** YargsParser } from './build/lib/yargs-parser.js'
const parser = new YargsParser(***REMOVED***
  cwd: () => ***REMOVED*** return '' },
  format: (str, arg) => ***REMOVED*** return str.replace('%s', arg) },
  normalize: (str) => ***REMOVED*** return str },
  resolve: (str) => ***REMOVED*** return str },
  require: () => ***REMOVED***
    throw Error('loading config from files not currently supported in browser')
***REMOVED***,
  env: () => ***REMOVED***}
})

const yargsParser = function Parser (args, opts) ***REMOVED***
  const result = parser.parse(args.slice(), opts)
  return result.argv
}
yargsParser.detailed = function (args, opts) ***REMOVED***
  return parser.parse(args.slice(), opts)
}
yargsParser.camelCase = camelCase
yargsParser.decamelize = decamelize
yargsParser.looksLikeNumber = looksLikeNumber

export default yargsParser
