// jose/dist/node/esm/util/errors.js
export class JOSEError extends Error ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JOSE_GENERIC';
  ***REMOVED***
    constructor(message) ***REMOVED***
        var _a;
        super(message);
        this.code = 'ERR_JOSE_GENERIC';
        this.name = this.constructor.name;
        (_a = Error.captureStackTrace) === null || _a === void 0 ? void 0 : _a.call(Error, this, this.constructor);
  ***REMOVED***
}
export class JWTClaimValidationFailed extends JOSEError ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWT_CLAIM_VALIDATION_FAILED';
  ***REMOVED***
    constructor(message, claim = 'unspecified', reason = 'unspecified') ***REMOVED***
        super(message);
        this.code = 'ERR_JWT_CLAIM_VALIDATION_FAILED';
        this.claim = claim;
        this.reason = reason;
  ***REMOVED***
}
export class JWTExpired extends JOSEError ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWT_EXPIRED';
  ***REMOVED***
    constructor(message, claim = 'unspecified', reason = 'unspecified') ***REMOVED***
        super(message);
        this.code = 'ERR_JWT_EXPIRED';
        this.claim = claim;
        this.reason = reason;
  ***REMOVED***
}
export class JOSEAlgNotAllowed extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JOSE_ALG_NOT_ALLOWED';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JOSE_ALG_NOT_ALLOWED';
  ***REMOVED***
}
export class JOSENotSupported extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JOSE_NOT_SUPPORTED';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JOSE_NOT_SUPPORTED';
  ***REMOVED***
}
export class JWEDecryptionFailed extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWE_DECRYPTION_FAILED';
        this.message = 'decryption operation failed';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWE_DECRYPTION_FAILED';
  ***REMOVED***
}
export class JWEInvalid extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWE_INVALID';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWE_INVALID';
  ***REMOVED***
}
export class JWSInvalid extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWS_INVALID';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWS_INVALID';
  ***REMOVED***
}
export class JWTInvalid extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWT_INVALID';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWT_INVALID';
  ***REMOVED***
}
export class JWKInvalid extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWK_INVALID';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWK_INVALID';
  ***REMOVED***
}
export class JWKSInvalid extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWKS_INVALID';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWKS_INVALID';
  ***REMOVED***
}
export class JWKSNoMatchingKey extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWKS_NO_MATCHING_KEY';
        this.message = 'no applicable key found in the JSON Web Key Set';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWKS_NO_MATCHING_KEY';
  ***REMOVED***
}
export class JWKSMultipleMatchingKeys extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWKS_MULTIPLE_MATCHING_KEYS';
        this.message = 'multiple matching keys found in the JSON Web Key Set';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWKS_MULTIPLE_MATCHING_KEYS';
  ***REMOVED***
}
export class JWKSTimeout extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWKS_TIMEOUT';
        this.message = 'request timed out';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWKS_TIMEOUT';
  ***REMOVED***
}
export class JWSSignatureVerificationFailed extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWS_SIGNATURE_VERIFICATION_FAILED';
        this.message = 'signature verification failed';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWS_SIGNATURE_VERIFICATION_FAILED';
  ***REMOVED***
}
