// jose/dist/node/cjs/jwe/general/decrypt.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.generalDecrypt = void 0;
const decrypt_js_1 = require("../flattened/decrypt.js");
const errors_js_1 = require("../../util/errors.js");
const is_object_js_1 = require("../../lib/is_object.js");
async function generalDecrypt(jwe, key, options) ***REMOVED***
    if (!(0, is_object_js_1.default)(jwe)) ***REMOVED***
        throw new errors_js_1.JWEInvalid('General JWE must be an object');
  ***REMOVED***
    if (!Array.isArray(jwe.recipients) || !jwe.recipients.every(is_object_js_1.default)) ***REMOVED***
        throw new errors_js_1.JWEInvalid('JWE Recipients missing or incorrect type');
  ***REMOVED***
    if (!jwe.recipients.length) ***REMOVED***
        throw new errors_js_1.JWEInvalid('JWE Recipients has no members');
  ***REMOVED***
    for (const recipient of jwe.recipients) ***REMOVED***
        try ***REMOVED***
            return await (0, decrypt_js_1.flattenedDecrypt)(***REMOVED***
                aad: jwe.aad,
                ciphertext: jwe.ciphertext,
                encrypted_key: recipient.encrypted_key,
                header: recipient.header,
                iv: jwe.iv,
                protected: jwe.protected,
                tag: jwe.tag,
                unprotected: jwe.unprotected,
          ***REMOVED***, key, options);
      ***REMOVED***
        catch ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    throw new errors_js_1.JWEDecryptionFailed();
}
exports.generalDecrypt = generalDecrypt;
