// jsdoc/lib/jsdoc/util/markdown.js
/**
 * Provides access to Markdown-related functions.
 * @module jsdoc/util/markdown
 */
const env = require('jsdoc/env');
const logger = require('jsdoc/util/logger');
const MarkdownIt = require('markdown-it');
const ***REMOVED*** marked } = require('marked');
const mda = require('markdown-it-anchor');
const path = require('jsdoc/path');
const util = require('util');

/**
 * Enumeration of Markdown parsers that are available.
 * @enum ***REMOVED***String}
 */
const parserNames = ***REMOVED***
    /**
     * The [`markdown-js`](https://github.com/evilstreak/markdown-js) (aka "evilstreak") parser.
     *
     * @deprecated Replaced by `markdown-it`.
     */
    evilstreak: 'markdownit',
    /**
     * The "GitHub-flavored Markdown" parser.
     *
     * @deprecated Replaced by `markdown-it`.
     */
    gfm: 'markdownit',
    /**
     * The `markdown-it` parser.
     */
    markdownit: 'markdownit',
    /**
     * The [Marked](https://github.com/chjj/marked) parser.
     *
     * @deprecated Will be replaced by `markdown-it` in JSDoc 3.7.0.
     */
    marked: 'marked'
};

/**
 * Escape underscores that occur within an inline tag in order to protect them from the `marked`
 * parser.
 *
 * @param ***REMOVED***string} source - The source text to sanitize.
 * @return ***REMOVED***string} The source text, where underscores within inline tags have been protected with a
 * preceding backslash (e.g., `\_`). The `marked` parser will strip the backslash and protect the
 * underscore.
 */
function escapeUnderscores(source) ***REMOVED***
    return source.replace(/\***REMOVED***@[^}\r\n]+\}/g, wholeMatch => wholeMatch.replace(/(^|[^\\])_/g, '$1\\_'));
}

/**
 * Escape HTTP/HTTPS URLs so that they are not automatically converted to HTML links.
 *
 * @param ***REMOVED***string} source - The source text to escape.
 * @return ***REMOVED***string} The source text with escape characters added to HTTP/HTTPS URLs.
 */
function escapeUrls(source) ***REMOVED***
    return source.replace(/(https?):\/\//g, '$1:\\/\\/');
}

/**
 * Unescape HTTP/HTTPS URLs after Markdown parsing is complete.
 *
 * @param ***REMOVED***string} source - The source text to unescape.
 * @return ***REMOVED***string} The source text with escape characters removed from HTTP/HTTPS URLs.
 */
function unescapeUrls(source) ***REMOVED***
    return source.replace(/(https?):\\\/\\\//g, '$1://');
}

/**
 * Escape backslashes within inline tags so that they are not stripped.
 *
 * @param ***REMOVED***string} source - The source text to escape.
 * @return ***REMOVED***string} The source text with backslashes escaped within inline tags.
 */
function escapeInlineTagBackslashes(source) ***REMOVED***
    return source.replace(/\***REMOVED***@[^}\r\n]+\}/g, wholeMatch => wholeMatch.replace(/\\/g, '\\\\'));
}

/**
 * Escape characters in text within a code block.
 *
 * @param ***REMOVED***string} source - The source text to escape.
 * @return ***REMOVED***string} The escaped source text.
 */
function escapeCode(source) ***REMOVED***
    return source.replace(/</g, '&lt;')
        .replace(/"/g, '&quot;')
        .replace(/'/g, '&#39;');
}

/**
 * Wrap a code snippet in HTML tags that enable syntax highlighting.
 *
 * @param ***REMOVED***string} code - The code snippet.
 * @param ***REMOVED***string?} language - The language of the code snippet.
 * @return ***REMOVED***string} The wrapped code snippet.
 */
function highlight(code, language) ***REMOVED***
    let classString;
    let langClass = '';

    if (language && (language !== 'plain')) ***REMOVED***
        langClass = ` lang-$***REMOVED***language}`;
  ***REMOVED***

    if (language !== 'plain') ***REMOVED***
        classString = util.format(' class="prettyprint source%s"', langClass);
  ***REMOVED***
    else ***REMOVED***
        classString = ' class="source"';
  ***REMOVED***

    return util.format('<pre%s><code>%s</code></pre>', classString, escapeCode(code));
}

/**
 * Unencode quotes that occur within ***REMOVED***@ ... } after the Markdown parser has turned them into HTML
 * entities.
 *
 * @param ***REMOVED***string} source - The source text to unencode.
 * @return ***REMOVED***string} The source text with HTML entity `&quot;` converted back to standard quotes.
 */
function unencodeQuotes(source) ***REMOVED***
    return source.replace(/\***REMOVED***@[^}\r\n]+\}/g, wholeMatch => wholeMatch.replace(/&quot;/g, '"'));
}

/**
 * Get the appropriate function for applying syntax highlighting to text, based on the user's
 * Markdown configuration settings.
 *
 * @param ***REMOVED***Object} conf - The user's Markdown configuration settings.
 * @return ***REMOVED***function} The highlighter function.
 */
function getHighlighter(conf) ***REMOVED***
    let highlighter;
    let highlighterPath;

    switch (typeof conf.highlight) ***REMOVED***
        case 'string':
            highlighterPath = path.getResourcePath(conf.highlight);

            if (highlighterPath) ***REMOVED***
                highlighter = require(highlighterPath).highlight;

                if (typeof highlighter !== 'function') ***REMOVED***
                    logger.error('The syntax highlighting module "%s" does not assign a method ' +
                        'to exports.highlight. Using the default syntax highlighter.',
                    conf.highlight);
                    highlighter = highlight;
              ***REMOVED***
          ***REMOVED***
            else ***REMOVED***
                logger.error('Unable to find the syntax highlighting module "%s". Using the ' +
                    'default syntax highlighter.', conf.highlight);
                highlighter = highlight;
          ***REMOVED***

            break;

        case 'function':
            highlighter = conf.highlight;

            break;

        default:
            highlighter = highlight;
  ***REMOVED***

    return highlighter;
}

/**
 * Retrieve a function that accepts a single parameter containing Markdown source. The function uses
 * the specified parser to transform the Markdown source to HTML, then returns the HTML as a string.
 *
 * @private
 * @param ***REMOVED***String} parserName The name of the selected parser.
 * @param ***REMOVED***Object} [conf] Configuration for the selected parser, if any.
 * @returns ***REMOVED***Function} A function that accepts Markdown source, feeds it to the selected parser, and
 * returns the resulting HTML.
 */
function getParseFunction(parserName, conf) ***REMOVED***
    let highlighter;
    let parserFunction;
    let renderer;

    conf = conf || ***REMOVED***};
    highlighter = getHighlighter(conf);

    switch (parserName) ***REMOVED***
        case parserNames.marked:
            if (conf.hardwrap) ***REMOVED***
                marked.setOptions(***REMOVED***breaks: true});
          ***REMOVED***

            // Marked generates an "id" attribute for headers; this custom renderer suppresses it
            renderer = new marked.Renderer();

            if (!conf.idInHeadings) ***REMOVED***
                renderer.heading = (text, level) => util.format('<h%s>%s</h%s>', level, text, level);
          ***REMOVED***

            renderer.code = highlighter;

            parserFunction = source => ***REMOVED***
                let result;

                source = escapeUnderscores(source);
                source = escapeUrls(source);

                result = marked(source, ***REMOVED*** renderer: renderer })
                    .replace(/\s+$/, '')
                    .replace(/&#39;/g, "'");

                result = unescapeUrls(result);
                result = unencodeQuotes(result);

                return result;
          ***REMOVED***;
            parserFunction._parser = parserNames.marked;

            return parserFunction;

        case parserNames.markdownit:
            renderer = new MarkdownIt(***REMOVED***
                breaks: Boolean(conf.hardwrap),
                highlight: highlighter,
                html: true
          ***REMOVED***);

            if (conf.idInHeadings) ***REMOVED***
                renderer.use(mda, ***REMOVED*** tabIndex: false });
          ***REMOVED***

            parserFunction = source => ***REMOVED***
                let result;

                source = escapeUrls(source);
                source = escapeInlineTagBackslashes(source);

                result = renderer.render(source)
                    .replace(/\s+$/, '')
                    .replace(/&#39;/g, "'");

                result = unescapeUrls(result);
                result = unencodeQuotes(result);

                return result;
          ***REMOVED***;
            parserFunction._parser = parserNames.markdownit;

            return parserFunction;

        default:
            logger.error('Unrecognized Markdown parser "%s". Markdown support is disabled.',
                parserName);

            return undefined;
  ***REMOVED***
}

/**
 * Retrieve a Markdown parsing function based on the value of the `conf.json` file's
 * `env.conf.markdown` property. The parsing function accepts a single parameter containing Markdown
 * source. The function uses the parser specified in `conf.json` to transform the Markdown source to
 * HTML, then returns the HTML as a string.
 *
 * @returns ***REMOVED***function} A function that accepts Markdown source, feeds it to the selected parser, and
 * returns the resulting HTML.
 */
exports.getParser = () => ***REMOVED***
    const conf = env.conf.markdown;
    const parser = (conf && conf.parser) ? parserNames[conf.parser] : parserNames.markdownit;

    return getParseFunction(parser, conf);
};
