// jsdoc/lib/jsdoc/util/templateHelper.js
/**
 * @module jsdoc/util/templateHelper
 */
const catharsis = require('catharsis');
let dictionary = require('jsdoc/tag/dictionary');
const env = require('jsdoc/env');
const inline = require('jsdoc/tag/inline');
const logger = require('jsdoc/util/logger');
const name = require('jsdoc/name');
const util = require('util');

const hasOwnProp = Object.prototype.hasOwnProperty;

const MODULE_NAMESPACE = 'module:';

const files = ***REMOVED***};
const ids = ***REMOVED***};

// each container gets its own html file
const containers = ['class', 'module', 'external', 'namespace', 'mixin', 'interface'];

let tutorials;

/** Sets tutorials map.
    @param ***REMOVED***jsdoc.tutorial.Tutorial} root - Root tutorial node.
 */
exports.setTutorials = root => ***REMOVED***
    tutorials = root;
};

exports.globalName = name.SCOPE.NAMES.GLOBAL;
exports.fileExtension = '.html';
exports.scopeToPunc = name.scopeToPunc;

const linkMap = ***REMOVED***
    // two-way lookup
    longnameToUrl: ***REMOVED***},
    urlToLongname: ***REMOVED***},

    // one-way lookup (IDs are only unique per file)
    longnameToId: ***REMOVED***}
};

// two-way lookup
const tutorialLinkMap = ***REMOVED***
    nameToUrl: ***REMOVED***},
    urlToName: ***REMOVED***}
};

const longnameToUrl = exports.longnameToUrl = linkMap.longnameToUrl;
const longnameToId = exports.longnameToId = linkMap.longnameToId;

const registerLink = exports.registerLink = (longname, fileUrl) => ***REMOVED***
    linkMap.longnameToUrl[longname] = fileUrl;
    linkMap.urlToLongname[fileUrl] = longname;
};

const registerId = exports.registerId = (longname, fragment) => ***REMOVED***
    linkMap.longnameToId[longname] = fragment;
};

function getNamespace(kind) ***REMOVED***
    if (dictionary.isNamespace(kind)) ***REMOVED***
        return `$***REMOVED***kind}:`;
  ***REMOVED***

    return '';
}

function formatNameForLink(doclet) ***REMOVED***
    let newName = getNamespace(doclet.kind) + (doclet.name || '') + (doclet.variation || '');
    const scopePunc = exports.scopeToPunc[doclet.scope] || '';

    // Only prepend the scope punctuation if it's not the same character that marks the start of a
    // fragment ID. Using `#` in HTML5 fragment IDs is legal, but URLs like `foo.html##bar` are
    // just confusing.
    if (scopePunc !== '#') ***REMOVED***
        newName = scopePunc + newName;
  ***REMOVED***

    return newName;
}

function makeUniqueFilename(filename, str) ***REMOVED***
    let key = filename.toLowerCase();
    let nonUnique = true;

    // don't allow filenames to begin with an underscore
    if (!filename.length || filename[0] === '_') ***REMOVED***
        filename = `-$***REMOVED***filename}`;
        key = filename.toLowerCase();
  ***REMOVED***

    // append enough underscores to make the filename unique
    while (nonUnique) ***REMOVED***
        if ( hasOwnProp.call(files, key) ) ***REMOVED***
            filename += '_';
            key = filename.toLowerCase();
      ***REMOVED*** else ***REMOVED***
            nonUnique = false;
      ***REMOVED***
  ***REMOVED***

    files[key] = str;

    return filename;
}

/**
 * Convert a string to a unique filename, including an extension.
 *
 * Filenames are cached to ensure that they are used only once. For example, if the same string is
 * passed in twice, two different filenames will be returned.
 *
 * Also, filenames are not considered unique if they are capitalized differently but are otherwise
 * identical.
 *
 * @function
 * @param ***REMOVED***string} str The string to convert.
 * @return ***REMOVED***string} The filename to use for the string.
 */
const getUniqueFilename = exports.getUniqueFilename = str => ***REMOVED***
    const namespaces = dictionary.getNamespaces().join('|');
    let basename = (str || '')
        // use - instead of : in namespace prefixes
        .replace(new RegExp(`^($***REMOVED***namespaces}):`), '$1-')
        // replace characters that can cause problems on some filesystems
        .replace(/[\\/?*:|'"<>]/g, '_')
        // use - instead of ~ to denote 'inner'
        .replace(/~/g, '-')
        // use _ instead of # to denote 'instance'
        .replace(/#/g, '_')
        // use _ instead of / (for example, in module names)
        .replace(/\//g, '_')
        // remove the variation, if any
        .replace(/\([\s\S]*\)$/, '')
        // make sure we don't create hidden files, or files whose names start with a dash
        .replace(/^[.-]/, '');

    // in case we've now stripped the entire basename (uncommon, but possible):
    basename = basename.length ? basename : '_';

    return makeUniqueFilename(basename, str) + exports.fileExtension;
};

/**
 * Get a longname's filename if one has been registered; otherwise, generate a unique filename, then
 * register the filename.
 * @private
 */
function getFilename(longname) ***REMOVED***
    let fileUrl;

    if ( hasOwnProp.call(longnameToUrl, longname) ) ***REMOVED***
        fileUrl = longnameToUrl[longname];
  ***REMOVED***
    else ***REMOVED***
        fileUrl = getUniqueFilename(longname);
        registerLink(longname, fileUrl);
  ***REMOVED***

    return fileUrl;
}

/**
 * Check whether a symbol is the only symbol exported by a module (as in
 * `module.exports = function() ***REMOVED***};`).
 *
 * @private
 * @param ***REMOVED***module:jsdoc/doclet.Doclet} doclet - The doclet for the symbol.
 * @return ***REMOVED***boolean} `true` if the symbol is the only symbol exported by a module; otherwise,
 * `false`.
 */
function isModuleExports(doclet) ***REMOVED***
    return doclet.longname && doclet.longname === doclet.name &&
        doclet.longname.indexOf(MODULE_NAMESPACE) === 0 && doclet.kind !== 'module';
}

function makeUniqueId(filename, id) ***REMOVED***
    let key;
    let nonUnique = true;

    key = id.toLowerCase();

    // HTML5 IDs cannot contain whitespace characters
    id = id.replace(/\s/g, '');

    // append enough underscores to make the identifier unique
    while (nonUnique) ***REMOVED***
        if ( hasOwnProp.call(ids, filename) && hasOwnProp.call(ids[filename], key) ) ***REMOVED***
            id += '_';
            key = id.toLowerCase();
      ***REMOVED***
        else ***REMOVED***
            nonUnique = false;
      ***REMOVED***
  ***REMOVED***

    ids[filename] = ids[filename] || ***REMOVED***};
    ids[filename][key] = id;

    return id;
}

/**
 * Get a doclet's ID if one has been registered; otherwise, generate a unique ID, then register
 * the ID.
 * @private
 */
function getId(longname, id) ***REMOVED***
    if ( hasOwnProp.call(longnameToId, longname) ) ***REMOVED***
        id = longnameToId[longname];
  ***REMOVED***
    else if (!id) ***REMOVED***
        // no ID required
        return '';
  ***REMOVED***
    else ***REMOVED***
        id = makeUniqueId(longname, id);
        registerId(longname, id);
  ***REMOVED***

    return id;
}

/**
 * Convert a doclet to an identifier that is unique for a specified filename.
 *
 * Identifiers are not considered unique if they are capitalized differently but are otherwise
 * identical.
 *
 * @method
 * @param ***REMOVED***string} filename - The file in which the identifier will be used.
 * @param ***REMOVED***string} doclet - The doclet to convert.
 * @return ***REMOVED***string} A unique identifier based on the file and doclet.
 */
exports.getUniqueId = makeUniqueId;

const htmlsafe = exports.htmlsafe = str => ***REMOVED***
    if (typeof str !== 'string') ***REMOVED***
        str = String(str);
  ***REMOVED***

    return str.replace(/&/g, '&amp;')
        .replace(/</g, '&lt;');
};

function parseType(longname) ***REMOVED***
    let err;

    try ***REMOVED***
        return catharsis.parse(longname, ***REMOVED***jsdoc: true});
  ***REMOVED***
    catch (e) ***REMOVED***
        err = new Error(`unable to parse $***REMOVED***longname}: $***REMOVED***e.message}`);
        logger.error(err);

        return longname;
  ***REMOVED***
}

function stringifyType(parsedType, cssClass, stringifyLinkMap) ***REMOVED***
    return require('catharsis').stringify(parsedType, ***REMOVED***
        cssClass: cssClass,
        htmlSafe: true,
        links: stringifyLinkMap
  ***REMOVED***);
}

function hasUrlPrefix(text) ***REMOVED***
    return (/^(http|ftp)s?:\/\//).test(text);
}

function isComplexTypeExpression(expr) ***REMOVED***
    // record types, type unions, and type applications all count as "complex"
    return /^***REMOVED***.+}$/.test(expr) || /^.+\|.+$/.test(expr) || /^.+<.+>$/.test(expr);
}

function fragmentHash(fragmentId) ***REMOVED***
    if (!fragmentId) ***REMOVED***
        return '';
  ***REMOVED***

    return `#$***REMOVED***fragmentId}`;
}

function getShortName(longname) ***REMOVED***
    return name.shorten(longname).name;
}

/**
 * Build an HTML link to the symbol with the specified longname. If the longname is not
 * associated with a URL, this method simply returns the link text, if provided, or the longname.
 *
 * The `longname` parameter can also contain a URL rather than a symbol's longname.
 *
 * This method supports type applications that can contain one or more types, such as
 * `Array.<MyClass>` or `Array.<(MyClass|YourClass)>`. In these examples, the method attempts to
 * replace `Array`, `MyClass`, and `YourClass` with links to the appropriate types. The link text
 * is ignored for type applications.
 *
 * @param ***REMOVED***string} longname - The longname (or URL) that is the target of the link.
 * @param ***REMOVED***string=} linkText - The text to display for the link, or `longname` if no text is
 * provided.
 * @param ***REMOVED***Object} options - Options for building the link.
 * @param ***REMOVED***string=} options.cssClass - The CSS class (or classes) to include in the link's `<a>`
 * tag.
 * @param ***REMOVED***string=} options.fragmentId - The fragment identifier (for example, `name` in
 * `foo.html#name`) to append to the link target.
 * @param ***REMOVED***string=} options.linkMap - The link map in which to look up the longname.
 * @param ***REMOVED***boolean=} options.monospace - Indicates whether to display the link text in a monospace
 * font.
 * @param ***REMOVED***boolean=} options.shortenName - Indicates whether to extract the short name from the
 * longname and display the short name in the link text. Ignored if `linkText` is specified.
 * @return ***REMOVED***string} The HTML link, or the link text if the link is not available.
 */
function buildLink(longname, linkText, options) ***REMOVED***
    const classString = options.cssClass ? util.format(' class="%s"', options.cssClass) : '';
    let fileUrl;
    const fragmentString = fragmentHash(options.fragmentId);
    let stripped;
    let text;

    let parsedType;

    // handle cases like:
    // @see <http://example.org>
    // @see http://example.org
    stripped = longname ? longname.replace(/^<|>$/g, '') : '';
    if ( hasUrlPrefix(stripped) ) ***REMOVED***
        fileUrl = stripped;
        text = linkText || stripped;
  ***REMOVED***
    // handle complex type expressions that may require multiple links
    // (but skip anything that looks like an inline tag or HTML tag)
    else if (longname && isComplexTypeExpression(longname) && /\***REMOVED***@.+\}/.test(longname) === false &&
        /^<[\s\S]+>/.test(longname) === false) ***REMOVED***
        parsedType = parseType(longname);

        return stringifyType(parsedType, options.cssClass, options.linkMap);
  ***REMOVED***
    else ***REMOVED***
        fileUrl = hasOwnProp.call(options.linkMap, longname) ? options.linkMap[longname] : '';
        text = linkText || (options.shortenName ? getShortName(longname) : longname);
  ***REMOVED***

    text = options.monospace ? `<code>$***REMOVED***text}</code>` : text;

    if (!fileUrl) ***REMOVED***
        return text;
  ***REMOVED***
    else ***REMOVED***
        return util.format('<a href="%s"%s>%s</a>', encodeURI(fileUrl + fragmentString),
            classString, text);
  ***REMOVED***
}

/**
 * Retrieve an HTML link to the symbol with the specified longname. If the longname is not
 * associated with a URL, this method simply returns the link text, if provided, or the longname.
 *
 * The `longname` parameter can also contain a URL rather than a symbol's longname.
 *
 * This method supports type applications that can contain one or more types, such as
 * `Array.<MyClass>` or `Array.<(MyClass|YourClass)>`. In these examples, the method attempts to
 * replace `Array`, `MyClass`, and `YourClass` with links to the appropriate types. The link text
 * is ignored for type applications.
 *
 * @function
 * @param ***REMOVED***string} longname - The longname (or URL) that is the target of the link.
 * @param ***REMOVED***string=} linkText - The text to display for the link, or `longname` if no text is
 * provided.
 * @param ***REMOVED***string=} cssClass - The CSS class (or classes) to include in the link's `<a>` tag.
 * @param ***REMOVED***string=} fragmentId - The fragment identifier (for example, `name` in `foo.html#name`) to
 * append to the link target.
 * @return ***REMOVED***string} The HTML link, or a plain-text string if the link is not available.
 */
const linkto = exports.linkto = (longname, linkText, cssClass, fragmentId) => buildLink(longname, linkText, ***REMOVED***
    cssClass: cssClass,
    fragmentId: fragmentId,
    linkMap: longnameToUrl
});

function useMonospace(tag, text) ***REMOVED***
    let cleverLinks;
    let monospaceLinks;
    let result;

    if ( hasUrlPrefix(text) ) ***REMOVED***
        result = false;
  ***REMOVED***
    else if (tag === 'linkplain') ***REMOVED***
        result = false;
  ***REMOVED***
    else if (tag === 'linkcode') ***REMOVED***
        result = true;
  ***REMOVED***
    else ***REMOVED***
        cleverLinks = env.conf.templates.cleverLinks;
        monospaceLinks = env.conf.templates.monospaceLinks;

        if (monospaceLinks || cleverLinks) ***REMOVED***
            result = true;
      ***REMOVED***
  ***REMOVED***

    return result || false;
}

function splitLinkText(text) ***REMOVED***
    let linkText;
    let target;
    let splitIndex;

    // if a pipe is not present, we split on the first space
    splitIndex = text.indexOf('|');
    if (splitIndex === -1) ***REMOVED***
        splitIndex = text.search(/\s/);
  ***REMOVED***

    if (splitIndex !== -1) ***REMOVED***
        linkText = text.substr(splitIndex + 1);
        // Normalize subsequent newlines to a single space.
        linkText = linkText.replace(/\n+/, ' ');
        target = text.substr(0, splitIndex);
  ***REMOVED***

    return ***REMOVED***
        linkText: linkText,
        target: target || text
  ***REMOVED***;
}

const tutorialToUrl = exports.tutorialToUrl = tutorial => ***REMOVED***
    let fileUrl;
    const node = tutorials.getByName(tutorial);

    // no such tutorial
    if (!node) ***REMOVED***
        logger.error( new Error(`No such tutorial: $***REMOVED***tutorial}`) );

        return null;
  ***REMOVED***

    // define the URL if necessary
    if (!hasOwnProp.call(tutorialLinkMap.nameToUrl, node.name)) ***REMOVED***
        fileUrl = `tutorial-$***REMOVED***getUniqueFilename(node.name)}`;
        tutorialLinkMap.nameToUrl[node.name] = fileUrl;
        tutorialLinkMap.urlToName[fileUrl] = node.name;
  ***REMOVED***

    return tutorialLinkMap.nameToUrl[node.name];
};

/**
 * Retrieve a link to a tutorial, or the name of the tutorial if the tutorial is missing. If the
 * `missingOpts` parameter is supplied, the names of missing tutorials will be prefixed by the
 * specified text and wrapped in the specified HTML tag and CSS class.
 *
 * @function
 * @todo Deprecate missingOpts once we have a better error-reporting mechanism.
 * @param ***REMOVED***string} tutorial The name of the tutorial.
 * @param ***REMOVED***string} content The link text to use.
 * @param ***REMOVED***object} [missingOpts] Options for displaying the name of a missing tutorial.
 * @param ***REMOVED***string} missingOpts.classname The CSS class to wrap around the tutorial name.
 * @param ***REMOVED***string} missingOpts.prefix The prefix to add to the tutorial name.
 * @param ***REMOVED***string} missingOpts.tag The tag to wrap around the tutorial name.
 * @return ***REMOVED***string} An HTML link to the tutorial, or the name of the tutorial with the specified
 * options.
 */
const toTutorial = exports.toTutorial = (tutorial, content, missingOpts) => ***REMOVED***
    let classname;
    let link;
    let node;
    let tag;


    if (!tutorial) ***REMOVED***
        logger.error( new Error('Missing required parameter: tutorial') );

        return null;
  ***REMOVED***

    node = tutorials.getByName(tutorial);
    // no such tutorial
    if (!node) ***REMOVED***
        missingOpts = missingOpts || ***REMOVED***};
        tag = missingOpts.tag;
        classname = missingOpts.classname;

        link = tutorial;
        if (missingOpts.prefix) ***REMOVED***
            link = missingOpts.prefix + link;
      ***REMOVED***
        if (tag) ***REMOVED***
            link = `<$***REMOVED***tag}$***REMOVED***classname ? (` class="$***REMOVED***classname}">`) : '>'}$***REMOVED***link}`;
            link += `</$***REMOVED***tag}>`;
      ***REMOVED***

        return link;
  ***REMOVED***

    content = content || node.title;

    return `<a href="$***REMOVED***tutorialToUrl(tutorial)}">$***REMOVED***content}</a>`;
};

function shouldShortenLongname() ***REMOVED***
    if (env.conf && env.conf.templates && env.conf.templates.useShortNamesInLinks) ***REMOVED***
        return true;
  ***REMOVED***

    return false;
}

/**
 * Find `***REMOVED***@link ...}` and `***REMOVED***@tutorial ...}` inline tags and turn them into HTML links.
 *
 * @param ***REMOVED***string} str - The string to search for `***REMOVED***@link ...}` and `***REMOVED***@tutorial ...}` tags.
 * @return ***REMOVED***string} The linkified text.
 */
exports.resolveLinks = str => ***REMOVED***
    let replacers;

    function extractLeadingText(string, completeTag) ***REMOVED***
        const tagIndex = string.indexOf(completeTag);
        let leadingText = null;
        const leadingTextRegExp = /\[(.+?)\]/g;
        let leadingTextInfo = leadingTextRegExp.exec(string);

        // did we find leading text, and if so, does it immediately precede the tag?
        while (leadingTextInfo && leadingTextInfo.length) ***REMOVED***
            if (leadingTextInfo.index + leadingTextInfo[0].length === tagIndex) ***REMOVED***
                string = string.replace(leadingTextInfo[0], '');
                leadingText = leadingTextInfo[1];
                break;
          ***REMOVED***

            leadingTextInfo = leadingTextRegExp.exec(string);
      ***REMOVED***

        return ***REMOVED***
            leadingText: leadingText,
            string: string
      ***REMOVED***;
  ***REMOVED***

    function processLink(string, ***REMOVED***completeTag, text, tag}) ***REMOVED***
        const leading = extractLeadingText(string, completeTag);
        let linkText = leading.leadingText;
        let monospace;
        let split;
        let target;

        string = leading.string;

        split = splitLinkText(text);
        target = split.target;
        linkText = linkText || split.linkText;

        monospace = useMonospace(tag, text);

        return string.replace( completeTag, buildLink(target, linkText, ***REMOVED***
            linkMap: longnameToUrl,
            monospace: monospace,
            shortenName: shouldShortenLongname()
      ***REMOVED***) );
  ***REMOVED***

    function processTutorial(string, ***REMOVED***completeTag, text}) ***REMOVED***
        const leading = extractLeadingText(string, completeTag);

        string = leading.string;

        return string.replace( completeTag, toTutorial(text, leading.leadingText) );
  ***REMOVED***

    replacers = ***REMOVED***
        link: processLink,
        linkcode: processLink,
        linkplain: processLink,
        tutorial: processTutorial
  ***REMOVED***;

    return inline.replaceInlineTags(str, replacers).newString;
};

/**
 * Convert tag text like `Jane Doe <jdoe@example.org>` into a `mailto:` link.
 *
 * @param ***REMOVED***string} str - The tag text.
 * @return ***REMOVED***string} The linkified text.
 */
exports.resolveAuthorLinks = str => ***REMOVED***
    let author = '';
    let matches;

    if (str) ***REMOVED***
        matches = str.match(/^\s?([\s\S]+)\b\s+<(\S+@\S+)>\s?$/);

        if (matches && matches.length === 3) ***REMOVED***
            author = `<a href="mailto:$***REMOVED***matches[2]}">$***REMOVED***htmlsafe(matches[1])}</a>`;
      ***REMOVED***
        else ***REMOVED***
            author = htmlsafe(str);
      ***REMOVED***
  ***REMOVED***

    return author;
};

/**
 * Find items in a TaffyDB database that match the specified key-value pairs.
 *
 * @function
 * @param ***REMOVED***TAFFY} data The TaffyDB database to search.
 * @param ***REMOVED***object|function} spec Key-value pairs to match against (for example,
 * `***REMOVED*** longname: 'foo' }`), or a function that returns `true` if a value matches or `false` if it
 * does not match.
 * @return ***REMOVED***array<object>} The matching items.
 */
const find = exports.find = (data, spec) => data(spec).get();

/**
 * Retrieve all of the following types of members from a set of doclets:
 *
 * + Classes
 * + Externals
 * + Globals
 * + Mixins
 * + Modules
 * + Namespaces
 * + Events
 * @param ***REMOVED***TAFFY} data The TaffyDB database to search.
 * @return ***REMOVED***object} An object with `classes`, `externals`, `globals`, `mixins`, `modules`,
 * `events`, and `namespaces` properties. Each property contains an array of objects.
 */
exports.getMembers = data => ***REMOVED***
    const members = ***REMOVED***
        classes: find( data, ***REMOVED***kind: 'class'} ),
        externals: find( data, ***REMOVED***kind: 'external'} ),
        events: find( data, ***REMOVED***kind: 'event'} ),
        globals: find(data, ***REMOVED***
            kind: ['member', 'function', 'constant', 'typedef'],
            memberof: ***REMOVED*** isUndefined: true }
      ***REMOVED***),
        mixins: find( data, ***REMOVED***kind: 'mixin'} ),
        modules: find( data, ***REMOVED***kind: 'module'} ),
        namespaces: find( data, ***REMOVED***kind: 'namespace'} ),
        interfaces: find( data, ***REMOVED***kind: 'interface'} )
  ***REMOVED***;

    // strip quotes from externals, since we allow quoted names that would normally indicate a
    // namespace hierarchy (as in `@external "jquery.fn"`)
    // TODO: we should probably be doing this for other types of symbols, here or elsewhere; see
    // jsdoc3/jsdoc#396
    members.externals = members.externals.map(doclet => ***REMOVED***
        doclet.name = doclet.name.replace(/(^"|"$)/g, '');

        return doclet;
  ***REMOVED***);

    // functions that are also modules (as in `module.exports = function() ***REMOVED***};`) are not globals
    members.globals = members.globals.filter(doclet => !isModuleExports(doclet));

    return members;
};

/**
 * Retrieve the member attributes for a doclet (for example, `virtual`, `static`, and
 * `readonly`).
 * @param ***REMOVED***object} d The doclet whose attributes will be retrieved.
 * @return ***REMOVED***array<string>} The member attributes for the doclet.
 */
exports.getAttribs = d => ***REMOVED***
    const attribs = [];

    if (!d) ***REMOVED***
        return attribs;
  ***REMOVED***

    if (d.async) ***REMOVED***
        attribs.push('async');
  ***REMOVED***

    if (d.generator) ***REMOVED***
        attribs.push('generator');
  ***REMOVED***

    if (d.virtual) ***REMOVED***
        attribs.push('abstract');
  ***REMOVED***

    if (d.access && d.access !== 'public') ***REMOVED***
        attribs.push(d.access);
  ***REMOVED***

    if (d.scope && d.scope !== 'instance' && d.scope !== name.SCOPE.NAMES.GLOBAL) ***REMOVED***
        if (d.kind === 'function' || d.kind === 'member' || d.kind === 'constant') ***REMOVED***
            attribs.push(d.scope);
      ***REMOVED***
  ***REMOVED***

    if (d.readonly === true) ***REMOVED***
        if (d.kind === 'member') ***REMOVED***
            attribs.push('readonly');
      ***REMOVED***
  ***REMOVED***

    if (d.kind === 'constant') ***REMOVED***
        attribs.push('constant');
  ***REMOVED***

    if (d.nullable === true) ***REMOVED***
        attribs.push('nullable');
  ***REMOVED***
    else if (d.nullable === false) ***REMOVED***
        attribs.push('non-null');
  ***REMOVED***

    return attribs;
};

/**
 * Retrieve links to allowed types for the member.
 *
 * @param ***REMOVED***Object} d - The doclet whose types will be retrieved.
 * @param ***REMOVED***string} [cssClass] - The CSS class to include in the `class` attribute for each link.
 * @return ***REMOVED***Array.<string>} HTML links to allowed types for the member.
 */
exports.getSignatureTypes = (***REMOVED***type}, cssClass) => ***REMOVED***
    let types = [];

    if (type && type.names) ***REMOVED***
        types = type.names;
  ***REMOVED***

    if (types && types.length) ***REMOVED***
        types = types.map(t => linkto(t, htmlsafe(t), cssClass));
  ***REMOVED***

    return types;
};

/**
 * Retrieve names of the parameters that the member accepts. If a value is provided for `optClass`,
 * the names of optional parameters will be wrapped in a `<span>` tag with that class.
 * @param ***REMOVED***object} d The doclet whose parameter names will be retrieved.
 * @param ***REMOVED***string} [optClass] The class to assign to the `<span>` tag that is wrapped around the
 * names of optional parameters. If a value is not provided, optional parameter names will not be
 * wrapped with a `<span>` tag. Must be a legal value for a CSS class name.
 * @return ***REMOVED***array<string>} An array of parameter names, with or without `<span>` tags wrapping the
 * names of optional parameters.
 */
exports.getSignatureParams = (***REMOVED***params}, optClass) => ***REMOVED***
    const pnames = [];

    if (params) ***REMOVED***
        params.forEach(p => ***REMOVED***
            if (p.name && !p.name.includes('.')) ***REMOVED***
                if (p.optional && optClass) ***REMOVED***
                    pnames.push(`<span class="$***REMOVED***optClass}">$***REMOVED***p.name}</span>`);
              ***REMOVED***
                else ***REMOVED***
                    pnames.push(p.name);
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***

    return pnames;
};

/**
 * Retrieve links to types that the member can return or yield.
 *
 * @param ***REMOVED***Object} d - The doclet whose types will be retrieved.
 * @param ***REMOVED***string} [cssClass] - The CSS class to include in the `class` attribute for each link.
 * @return ***REMOVED***Array.<string>} HTML links to types that the member can return or yield.
 */
exports.getSignatureReturns = (***REMOVED***yields, returns}, cssClass) => ***REMOVED***
    let returnTypes = [];

    if (yields || returns) ***REMOVED***
        (yields || returns).forEach(r => ***REMOVED***
            if (r && r.type && r.type.names) ***REMOVED***
                if (!returnTypes.length) ***REMOVED***
                    returnTypes = r.type.names;
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***

    if (returnTypes && returnTypes.length) ***REMOVED***
        returnTypes = returnTypes.map(r => linkto(r, htmlsafe(r), cssClass));
  ***REMOVED***

    return returnTypes;
};

/**
 * Retrieve an ordered list of doclets for a symbol's ancestors.
 *
 * @param ***REMOVED***TAFFY} data - The TaffyDB database to search.
 * @param ***REMOVED***Object} doclet - The doclet whose ancestors will be retrieved.
 * @return ***REMOVED***Array.<module:jsdoc/doclet.Doclet>} A array of ancestor doclets, sorted from most to
 * least distant.
 */
exports.getAncestors = (data, doclet) => ***REMOVED***
    const ancestors = [];
    let doc = doclet;
    let previousDoc;

    while (doc) ***REMOVED***
        previousDoc = doc;
        doc = find(data, ***REMOVED***longname: doc.memberof})[0];

        // prevent infinite loop that can be caused by duplicated module definitions
        if (previousDoc === doc) ***REMOVED***
            break;
      ***REMOVED***

        if (doc) ***REMOVED***
            ancestors.unshift(doc);
      ***REMOVED***
  ***REMOVED***

    return ancestors;
};

/**
 * Retrieve links to a member's ancestors.
 *
 * @param ***REMOVED***TAFFY} data - The TaffyDB database to search.
 * @param ***REMOVED***Object} doclet - The doclet whose ancestors will be retrieved.
 * @param ***REMOVED***string} [cssClass] - The CSS class to include in the `class` attribute for each link.
 * @return ***REMOVED***Array.<string>} HTML links to a member's ancestors.
 */
exports.getAncestorLinks = (data, doclet, cssClass) => ***REMOVED***
    const ancestors = exports.getAncestors(data, doclet);
    const links = [];

    ancestors.forEach(ancestor => ***REMOVED***
        const linkText = (exports.scopeToPunc[ancestor.scope] || '') + ancestor.name;
        const link = linkto(ancestor.longname, linkText, cssClass);

        links.push(link);
  ***REMOVED***);

    if (links.length) ***REMOVED***
        links[links.length - 1] += (exports.scopeToPunc[doclet.scope] || '');
  ***REMOVED***

    return links;
};

/**
 * Iterates through all the doclets in `data`, ensuring that if a method `@listens` to an event,
 * then that event has a `listeners` array with the longname of the listener in it.
 *
 * @param ***REMOVED***TAFFY} data - The TaffyDB database to search.
 */
exports.addEventListeners = data => ***REMOVED***
    // just a cache to prevent me doing so many lookups
    const _events = ***REMOVED***};
    let doc;
    let l;
    // TODO: do this on the *pruned* data
    // find all doclets that @listen to something.
    /* eslint-disable no-invalid-this */
    const listeners = find(data, function() ***REMOVED***
        return this.listens && this.listens.length;
  ***REMOVED***);
    /* eslint-enable no-invalid-this */

    if (!listeners.length) ***REMOVED***
        return;
  ***REMOVED***

    listeners.forEach((***REMOVED***listens, longname}) => ***REMOVED***
        l = listens;
        l.forEach(eventLongname => ***REMOVED***
            doc = _events[eventLongname] || find(data, ***REMOVED***
                longname: eventLongname,
                kind: 'event'
          ***REMOVED***)[0];
            if (doc) ***REMOVED***
                if (!doc.listeners) ***REMOVED***
                    doc.listeners = [longname];
              ***REMOVED*** else ***REMOVED***
                    doc.listeners.push(longname);
              ***REMOVED***
                _events[eventLongname] = _events[eventLongname] || doc;
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***);
};

/**
 * Remove members that will not be included in the output, including:
 *
 * + Undocumented members.
 * + Members tagged `@ignore`.
 * + Members of anonymous classes.
 * + Members tagged `@private`, unless the `private` option is enabled.
 * + Members tagged with anything other than specified by the `access` options.
 * @param ***REMOVED***TAFFY} data The TaffyDB database to prune.
 * @return ***REMOVED***TAFFY} The pruned database.
 */
exports.prune = data => ***REMOVED***
    data(***REMOVED***undocumented: true}).remove();
    data(***REMOVED***ignore: true}).remove();
    data(***REMOVED***memberof: '<anonymous>'}).remove();

    if (!env.opts.access || (env.opts.access && !env.opts.access.includes('all'))) ***REMOVED***
        if (env.opts.access && !env.opts.access.includes('package')) ***REMOVED***
            data(***REMOVED***access: 'package'}).remove();
      ***REMOVED***
        if (env.opts.access && !env.opts.access.includes('public')) ***REMOVED***
            data(***REMOVED***access: 'public'}).remove();
      ***REMOVED***
        if (env.opts.access && !env.opts.access.includes('protected')) ***REMOVED***
            data(***REMOVED***access: 'protected'}).remove();
      ***REMOVED***
        if (!env.opts.private && (!env.opts.access || (env.opts.access && !env.opts.access.includes('private')))) ***REMOVED***
            data(***REMOVED***access: 'private'}).remove();
      ***REMOVED***
        if (env.opts.access && !env.opts.access.includes('undefined')) ***REMOVED***
            data(***REMOVED***access: ***REMOVED***isUndefined: true}}).remove();
      ***REMOVED***
  ***REMOVED***

    return data;
};

/**
 * Create a URL that points to the generated documentation for the doclet.
 *
 * If a doclet corresponds to an output file (for example, if the doclet represents a class), the
 * URL will consist of a filename.
 *
 * If a doclet corresponds to a smaller portion of an output file (for example, if the doclet
 * represents a method), the URL will consist of a filename and a fragment ID.
 *
 * @param ***REMOVED***module:jsdoc/doclet.Doclet} doclet - The doclet that will be used to create the URL.
 * @return ***REMOVED***string} The URL to the generated documentation for the doclet.
 */
exports.createLink = doclet => ***REMOVED***
    let fakeContainer;
    let filename;
    let fileUrl;
    let fragment = '';
    const longname = doclet.longname;
    let match;

    // handle doclets in which doclet.longname implies that the doclet gets its own HTML file, but
    // doclet.kind says otherwise. this happens due to mistagged JSDoc (for example, a module that
    // somehow has doclet.kind set to `member`).
    // TODO: generate a warning (ideally during parsing!)
    if (!containers.includes(doclet.kind)) ***REMOVED***
        match = /(\S+):/.exec(longname);
        if (match && containers.includes(match[1])) ***REMOVED***
            fakeContainer = match[1];
      ***REMOVED***
  ***REMOVED***

    // the doclet gets its own HTML file
    if ( containers.includes(doclet.kind) || isModuleExports(doclet) ) ***REMOVED***
        filename = getFilename(longname);
  ***REMOVED***
    // mistagged version of a doclet that gets its own HTML file
    else if ( !containers.includes(doclet.kind) && fakeContainer ) ***REMOVED***
        filename = getFilename(doclet.memberof || longname);
        if (doclet.name !== doclet.longname) ***REMOVED***
            fragment = formatNameForLink(doclet);
            fragment = getId(longname, fragment);
      ***REMOVED***
  ***REMOVED***
    // the doclet is within another HTML file
    else ***REMOVED***
        filename = getFilename(doclet.memberof || exports.globalName);
        if ( (doclet.name !== doclet.longname) || (doclet.scope === name.SCOPE.NAMES.GLOBAL) ) ***REMOVED***
            fragment = formatNameForLink(doclet);
            fragment = getId(longname, fragment);
      ***REMOVED***
  ***REMOVED***

    fileUrl = encodeURI( filename + fragmentHash(fragment) );

    return fileUrl;
};

/**
 * Convert an array of doclet longnames into a tree structure, optionally attaching doclets to the
 * tree.
 *
 * @function
 * @see module:jsdoc/name.longnamesToTree
 * @param ***REMOVED***Array<string>} longnames - The longnames to convert into a tree.
 * @param ***REMOVED***Object<string, module:jsdoc/doclet.Doclet>} doclets - The doclets to attach to a tree.
 * Each property should be the longname of a doclet, and each value should be the doclet for that
 * longname.
 * @return ***REMOVED***Object} A tree with information about each longname.
 */
exports.longnamesToTree = name.longnamesToTree;

/**
 * Replace the existing tag dictionary with a new tag dictionary.
 *
 * Used for testing only. Do not call this method directly. Instead, call
 * ***REMOVED***@link module:jsdoc/doclet._replaceDictionary}, which also updates this module's tag dictionary.
 *
 * @private
 * @param ***REMOVED***module:jsdoc/tag/dictionary.Dictionary} dict - The new tag dictionary.
 */
exports._replaceDictionary = function _replaceDictionary(dict) ***REMOVED***
    dictionary = dict;
};
