// css-to-react-native/src/transforms/index.js
import ***REMOVED***
  AUTO,
  COLOR,
  LENGTH,
  PERCENT,
  UNSUPPORTED_LENGTH_UNIT,
  WORD,
} from '../tokenTypes'
import border from './border'
import boxShadow from './boxShadow'
import flex from './flex'
import flexFlow from './flexFlow'
import font from './font'
import fontFamily from './fontFamily'
import fontVariant from './fontVariant'
import placeContent from './placeContent'
import textDecoration from './textDecoration'
import textDecorationLine from './textDecorationLine'
import textShadow from './textShadow'
import transform from './transform'
import ***REMOVED*** directionFactory, parseShadowOffset } from './util'

const background = tokenStream => (***REMOVED***
  backgroundColor: tokenStream.expect(COLOR),
})
const borderColor = directionFactory(***REMOVED***
  types: [COLOR],
  prefix: 'border',
  suffix: 'Color',
})
const borderRadius = directionFactory(***REMOVED***
  directions: ['TopLeft', 'TopRight', 'BottomRight', 'BottomLeft'],
  prefix: 'border',
  suffix: 'Radius',
})
const borderWidth = directionFactory(***REMOVED*** prefix: 'border', suffix: 'Width' })
const margin = directionFactory(***REMOVED***
  types: [LENGTH, UNSUPPORTED_LENGTH_UNIT, PERCENT, AUTO],
  prefix: 'margin',
})
const padding = directionFactory(***REMOVED*** prefix: 'padding' })

const fontWeight = tokenStream => (***REMOVED***
  fontWeight: tokenStream.expect(WORD), // Also match numbers as strings
})
const shadowOffset = tokenStream => (***REMOVED***
  shadowOffset: parseShadowOffset(tokenStream),
})
const textShadowOffset = tokenStream => (***REMOVED***
  textShadowOffset: parseShadowOffset(tokenStream),
})

export default ***REMOVED***
  background,
  border,
  borderColor,
  borderRadius,
  borderWidth,
  boxShadow,
  flex,
  flexFlow,
  font,
  fontFamily,
  fontVariant,
  fontWeight,
  margin,
  padding,
  placeContent,
  shadowOffset,
  textShadow,
  textShadowOffset,
  textDecoration,
  textDecorationLine,
  transform,
}
