// css-to-react-native/src/transforms/util.js
import ***REMOVED***
  LENGTH,
  UNSUPPORTED_LENGTH_UNIT,
  PERCENT,
  COLOR,
  SPACE,
  NONE,
} from '../tokenTypes'

export const directionFactory = (***REMOVED***
  types = [LENGTH, UNSUPPORTED_LENGTH_UNIT, PERCENT],
  directions = ['Top', 'Right', 'Bottom', 'Left'],
  prefix = '',
  suffix = '',
}) => tokenStream => ***REMOVED***
  const values = []

  // borderWidth doesn't currently allow a percent value, but may do in the future
  values.push(tokenStream.expect(...types))

  while (values.length < 4 && tokenStream.hasTokens()) ***REMOVED***
    tokenStream.expect(SPACE)
    values.push(tokenStream.expect(...types))
***REMOVED***

  tokenStream.expectEmpty()

  const [top, right = top, bottom = top, left = right] = values

  const keyFor = n => `$***REMOVED***prefix}$***REMOVED***directions[n]}$***REMOVED***suffix}`

  return ***REMOVED***
    [keyFor(0)]: top,
    [keyFor(1)]: right,
    [keyFor(2)]: bottom,
    [keyFor(3)]: left,
***REMOVED***
}

export const parseShadowOffset = tokenStream => ***REMOVED***
  const width = tokenStream.expect(LENGTH)
  const height = tokenStream.matches(SPACE) ? tokenStream.expect(LENGTH) : width
  tokenStream.expectEmpty()
  return ***REMOVED*** width, height }
}

export const parseShadow = tokenStream => ***REMOVED***
  let offsetX
  let offsetY
  let radius
  let color

  if (tokenStream.matches(NONE)) ***REMOVED***
    tokenStream.expectEmpty()
    return ***REMOVED***
      offset: ***REMOVED*** width: 0, height: 0 },
      radius: 0,
      color: 'black',
  ***REMOVED***
***REMOVED***

  let didParseFirst = false
  while (tokenStream.hasTokens()) ***REMOVED***
    if (didParseFirst) tokenStream.expect(SPACE)

    if (
      offsetX === undefined &&
      tokenStream.matches(LENGTH, UNSUPPORTED_LENGTH_UNIT)
    ) ***REMOVED***
      offsetX = tokenStream.lastValue
      tokenStream.expect(SPACE)
      offsetY = tokenStream.expect(LENGTH, UNSUPPORTED_LENGTH_UNIT)

      tokenStream.saveRewindPoint()
      if (
        tokenStream.matches(SPACE) &&
        tokenStream.matches(LENGTH, UNSUPPORTED_LENGTH_UNIT)
      ) ***REMOVED***
        radius = tokenStream.lastValue
    ***REMOVED*** else ***REMOVED***
        tokenStream.rewind()
    ***REMOVED***
  ***REMOVED*** else if (color === undefined && tokenStream.matches(COLOR)) ***REMOVED***
      color = tokenStream.lastValue
  ***REMOVED*** else ***REMOVED***
      tokenStream.throw()
  ***REMOVED***

    didParseFirst = true
***REMOVED***

  if (offsetX === undefined) tokenStream.throw()

  return ***REMOVED***
    offset: ***REMOVED*** width: offsetX, height: offsetY },
    radius: radius !== undefined ? radius : 0,
    color: color !== undefined ? color : 'black',
***REMOVED***
}
