// text-decoding/src/utils.js
//
// Utilities
//
/**
 * @param ***REMOVED***number} a The number to test.
 * @param ***REMOVED***number} min The minimum value in the range, inclusive.
 * @param ***REMOVED***number} max The maximum value in the range, inclusive.
 * @return ***REMOVED***boolean} True if a >= min and a <= max.
 */
export function inRange(a, min, max) ***REMOVED***
  return min <= a && a <= max
}

export const floor = Math.floor

/**
 * @param ***REMOVED***string} string Input string of UTF-16 code units.
 * @return ***REMOVED***!Array.<number>} Code points.
 */
export function stringToCodePoints(string) ***REMOVED***
  // https://heycam.github.io/webidl/#dfn-obtain-unicode

  // 1. Let S be the DOMString value.
  var s = String(string)

  // 2. Let n be the length of S.
  var n = s.length

  // 3. Initialize i to 0.
  var i = 0

  // 4. Initialize U to be an empty sequence of Unicode characters.
  var u = []

  // 5. While i < n:
  while (i < n) ***REMOVED***
    // 1. Let c be the code unit in S at index i.
    var c = s.charCodeAt(i)

    // 2. Depending on the value of c:

    // c < 0xD800 or c > 0xDFFF
    if (c < 0xD800 || c > 0xDFFF) ***REMOVED***
      // Append to U the Unicode character with code point c.
      u.push(c)
  ***REMOVED***

    // 0xDC00 ≤ c ≤ 0xDFFF
    else if (0xDC00 <= c && c <= 0xDFFF) ***REMOVED***
      // Append to U a U+FFFD REPLACEMENT CHARACTER.
      u.push(0xFFFD)
  ***REMOVED***

    // 0xD800 ≤ c ≤ 0xDBFF
    else if (0xD800 <= c && c <= 0xDBFF) ***REMOVED***
      // 1. If i = n−1, then append to U a U+FFFD REPLACEMENT
      // CHARACTER.
      if (i === n - 1) ***REMOVED***
        u.push(0xFFFD)
    ***REMOVED***
      // 2. Otherwise, i < n−1:
      else ***REMOVED***
        // 1. Let d be the code unit in S at index i+1.
        var d = s.charCodeAt(i + 1)

        // 2. If 0xDC00 ≤ d ≤ 0xDFFF, then:
        if (0xDC00 <= d && d <= 0xDFFF) ***REMOVED***
          // 1. Let a be c & 0x3FF.
          var a = c & 0x3FF

          // 2. Let b be d & 0x3FF.
          var b = d & 0x3FF

          // 3. Append to U the Unicode character with code point
          // 2^16+2^10*a+b.
          u.push(0x10000 + (a << 10) + b)

          // 4. Set i to i+1.
          i += 1
      ***REMOVED***

        // 3. Otherwise, d < 0xDC00 or d > 0xDFFF. Append to U a
        // U+FFFD REPLACEMENT CHARACTER.
        else  ***REMOVED***
          u.push(0xFFFD)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

    // 3. Set i to i+1.
    i += 1
***REMOVED***

  // 6. Return U.
  return u
}

/**
 * @param ***REMOVED***!Array.<number>} code_points Array of code points.
 * @return ***REMOVED***string} string String of UTF-16 code units.
 */
export function codePointsToString(code_points) ***REMOVED***
  var s = ''
  for (var i = 0; i < code_points.length; ++i) ***REMOVED***
    var cp = code_points[i]
    if (cp <= 0xFFFF) ***REMOVED***
      s += String.fromCharCode(cp)
  ***REMOVED*** else ***REMOVED***
      cp -= 0x10000
      s += String.fromCharCode((cp >> 10) + 0xD800,
        (cp & 0x3FF) + 0xDC00)
  ***REMOVED***
***REMOVED***
  return s
}

/**
 * @param ***REMOVED***boolean} fatal If true, decoding errors raise an exception.
 * @param ***REMOVED***number=} opt_code_point Override the standard fallback code point.
 * @return The code point to insert on a decoding error.
 */
export function decoderError(fatal, opt_code_point) ***REMOVED***
  if (fatal)
    throw TypeError('Decoder error')
  return opt_code_point || 0xFFFD
}

/**
 * @param ***REMOVED***number} code_point The code point that could not be encoded.
 * @return ***REMOVED***number} Always throws, no value is actually returned.
 */
export function encoderError(code_point) ***REMOVED***
  throw TypeError('The code point ' + code_point + ' could not be encoded.')
}

/**
 * @param ***REMOVED***number} code_unit
 * @param ***REMOVED***boolean} utf16be
 */
export function convertCodeUnitToBytes(code_unit, utf16be) ***REMOVED***
  // 1. Let byte1 be code unit >> 8.
  const byte1 = code_unit >> 8

  // 2. Let byte2 be code unit & 0x00FF.
  const byte2 = code_unit & 0x00FF

  // 3. Then return the bytes in order:
  // utf-16be flag is set: byte1, then byte2.
  if (utf16be)
    return [byte1, byte2]
  // utf-16be flag is unset: byte2, then byte1.
  return [byte2, byte1]
}


//
// 4. Terminology
//

/**
 * An ASCII byte is a byte in the range 0x00 to 0x7F, inclusive.
 * @param ***REMOVED***number} a The number to test.
 * @return ***REMOVED***boolean} True if a is in the range 0x00 to 0x7F, inclusive.
 */
export function isASCIIByte(a) ***REMOVED***
  return 0x00 <= a && a <= 0x7F
}

/**
 * An ASCII code point is a code point in the range U+0000 to
 * U+007F, inclusive.
 */
export const isASCIICodePoint = isASCIIByte

/**
 * End-of-stream is a special token that signifies no more tokens are in the stream.
 */
export const end_of_stream = -1

export const finished = -1
