// https-proxy-agent/dist/parse-proxy-response.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) ***REMOVED***
    return (mod && mod.__esModule) ? mod : ***REMOVED*** "default": mod };
};
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
const debug_1 = __importDefault(require("debug"));
const debug = debug_1.default('https-proxy-agent:parse-proxy-response');
function parseProxyResponse(socket) ***REMOVED***
    return new Promise((resolve, reject) => ***REMOVED***
        // we need to buffer any HTTP traffic that happens with the proxy before we get
        // the CONNECT response, so that if the response is anything other than an "200"
        // response code, then we can re-play the "data" events on the socket once the
        // HTTP parser is hooked up...
        let buffersLength = 0;
        const buffers = [];
        function read() ***REMOVED***
            const b = socket.read();
            if (b)
                ondata(b);
            else
                socket.once('readable', read);
      ***REMOVED***
        function cleanup() ***REMOVED***
            socket.removeListener('end', onend);
            socket.removeListener('error', onerror);
            socket.removeListener('close', onclose);
            socket.removeListener('readable', read);
      ***REMOVED***
        function onclose(err) ***REMOVED***
            debug('onclose had error %o', err);
      ***REMOVED***
        function onend() ***REMOVED***
            debug('onend');
      ***REMOVED***
        function onerror(err) ***REMOVED***
            cleanup();
            debug('onerror %o', err);
            reject(err);
      ***REMOVED***
        function ondata(b) ***REMOVED***
            buffers.push(b);
            buffersLength += b.length;
            const buffered = Buffer.concat(buffers, buffersLength);
            const endOfHeaders = buffered.indexOf('\r\n\r\n');
            if (endOfHeaders === -1) ***REMOVED***
                // keep buffering
                debug('have not received end of HTTP headers yet...');
                read();
                return;
          ***REMOVED***
            const firstLine = buffered.toString('ascii', 0, buffered.indexOf('\r\n'));
            const statusCode = +firstLine.split(' ')[1];
            debug('got proxy server response: %o', firstLine);
            resolve(***REMOVED***
                statusCode,
                buffered
          ***REMOVED***);
      ***REMOVED***
        socket.on('error', onerror);
        socket.on('close', onclose);
        socket.on('end', onend);
        read();
  ***REMOVED***);
}
exports.default = parseProxyResponse;
