// markdown-it/lib/rules_core/replacements.js
// Simple typographic replacements
//
// (c) (C) → ©
// (tm) (TM) → ™
// (r) (R) → ®
// +- → ±
// (p) (P) -> §
// ... → … (also ?.... → ?.., !.... → !..)
// ???????? → ???, !!!!! → !!!, `,,` → `,`
// -- → &ndash;, --- → &mdash;
//
'use strict';

// TODO:
// - fractionals 1/2, 1/4, 3/4 -> ½, ¼, ¾
// - miltiplication 2 x 4 -> 2 × 4

var RARE_RE = /\+-|\.\.|\?\?\?\?|!!!!|,,|--/;

// Workaround for phantomjs - need regex without /g flag,
// or root check will fail every second time
var SCOPED_ABBR_TEST_RE = /\((c|tm|r|p)\)/i;

var SCOPED_ABBR_RE = /\((c|tm|r|p)\)/ig;
var SCOPED_ABBR = ***REMOVED***
  c: '©',
  r: '®',
  p: '§',
  tm: '™'
};

function replaceFn(match, name) ***REMOVED***
  return SCOPED_ABBR[name.toLowerCase()];
}

function replace_scoped(inlineTokens) ***REMOVED***
  var i, token, inside_autolink = 0;

  for (i = inlineTokens.length - 1; i >= 0; i--) ***REMOVED***
    token = inlineTokens[i];

    if (token.type === 'text' && !inside_autolink) ***REMOVED***
      token.content = token.content.replace(SCOPED_ABBR_RE, replaceFn);
  ***REMOVED***

    if (token.type === 'link_open' && token.info === 'auto') ***REMOVED***
      inside_autolink--;
  ***REMOVED***

    if (token.type === 'link_close' && token.info === 'auto') ***REMOVED***
      inside_autolink++;
  ***REMOVED***
***REMOVED***
}

function replace_rare(inlineTokens) ***REMOVED***
  var i, token, inside_autolink = 0;

  for (i = inlineTokens.length - 1; i >= 0; i--) ***REMOVED***
    token = inlineTokens[i];

    if (token.type === 'text' && !inside_autolink) ***REMOVED***
      if (RARE_RE.test(token.content)) ***REMOVED***
        token.content = token.content
          .replace(/\+-/g, '±')
          // .., ..., ....... -> …
          // but ?..... & !..... -> ?.. & !..
          .replace(/\.***REMOVED***2,}/g, '…').replace(/([?!])…/g, '$1..')
          .replace(/([?!])***REMOVED***4,}/g, '$1$1$1').replace(/,***REMOVED***2,}/g, ',')
          // em-dash
          .replace(/(^|[^-])---(?=[^-]|$)/mg, '$1\u2014')
          // en-dash
          .replace(/(^|\s)--(?=\s|$)/mg, '$1\u2013')
          .replace(/(^|[^-\s])--(?=[^-\s]|$)/mg, '$1\u2013');
    ***REMOVED***
  ***REMOVED***

    if (token.type === 'link_open' && token.info === 'auto') ***REMOVED***
      inside_autolink--;
  ***REMOVED***

    if (token.type === 'link_close' && token.info === 'auto') ***REMOVED***
      inside_autolink++;
  ***REMOVED***
***REMOVED***
}


module.exports = function replace(state) ***REMOVED***
  var blkIdx;

  if (!state.md.options.typographer) ***REMOVED*** return; }

  for (blkIdx = state.tokens.length - 1; blkIdx >= 0; blkIdx--) ***REMOVED***

    if (state.tokens[blkIdx].type !== 'inline') ***REMOVED*** continue; }

    if (SCOPED_ABBR_TEST_RE.test(state.tokens[blkIdx].content)) ***REMOVED***
      replace_scoped(state.tokens[blkIdx].children);
  ***REMOVED***

    if (RARE_RE.test(state.tokens[blkIdx].content)) ***REMOVED***
      replace_rare(state.tokens[blkIdx].children);
  ***REMOVED***

***REMOVED***
};
