// firebase-admin/lib/machine-learning/machine-learning-api-client.js
/*! firebase-admin v11.5.0 */
"use strict";
/*!
 * Copyright 2020 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.MachineLearningApiClient = exports.isGcsTfliteModelOptions = void 0;
const api_request_1 = require("../utils/api-request");
const error_1 = require("../utils/error");
const utils = require("../utils/index");
const validator = require("../utils/validator");
const machine_learning_utils_1 = require("./machine-learning-utils");
const ML_V1BETA2_API = 'https://firebaseml.googleapis.com/v1beta2';
const FIREBASE_VERSION_HEADER = ***REMOVED***
    'X-Firebase-Client': `fire-admin-node/$***REMOVED***utils.getSdkVersion()}`,
};
// Operation polling defaults
const POLL_DEFAULT_MAX_TIME_MILLISECONDS = 120000; // Maximum overall 2 minutes
const POLL_BASE_WAIT_TIME_MILLISECONDS = 3000; // Start with 3 second delay
const POLL_MAX_WAIT_TIME_MILLISECONDS = 30000; // Maximum 30 second delay
function isGcsTfliteModelOptions(options) ***REMOVED***
    const gcsUri = options?.tfliteModel?.gcsTfliteUri;
    return typeof gcsUri !== 'undefined';
}
exports.isGcsTfliteModelOptions = isGcsTfliteModelOptions;
/**
 * Class that facilitates sending requests to the Firebase ML backend API.
 *
 * @internal
 */
class MachineLearningApiClient ***REMOVED***
    constructor(app) ***REMOVED***
        this.app = app;
        if (!validator.isNonNullObject(app) || !('options' in app)) ***REMOVED***
            throw new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'First argument passed to admin.machineLearning() must be a valid '
                + 'Firebase app instance.');
      ***REMOVED***
        this.httpClient = new api_request_1.AuthorizedHttpClient(app);
  ***REMOVED***
    createModel(model) ***REMOVED***
        if (!validator.isNonNullObject(model) ||
            !validator.isNonEmptyString(model.displayName)) ***REMOVED***
            const err = new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'Invalid model content.');
            return Promise.reject(err);
      ***REMOVED***
        return this.getProjectUrl()
            .then((url) => ***REMOVED***
            const request = ***REMOVED***
                method: 'POST',
                url: `$***REMOVED***url}/models`,
                data: model,
          ***REMOVED***;
            return this.sendRequest(request);
      ***REMOVED***);
  ***REMOVED***
    updateModel(modelId, model, updateMask) ***REMOVED***
        if (!validator.isNonEmptyString(modelId) ||
            !validator.isNonNullObject(model) ||
            !validator.isNonEmptyArray(updateMask)) ***REMOVED***
            const err = new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'Invalid model or mask content.');
            return Promise.reject(err);
      ***REMOVED***
        return this.getProjectUrl()
            .then((url) => ***REMOVED***
            const request = ***REMOVED***
                method: 'PATCH',
                url: `$***REMOVED***url}/models/$***REMOVED***modelId}?updateMask=$***REMOVED***updateMask.join()}`,
                data: model,
          ***REMOVED***;
            return this.sendRequest(request);
      ***REMOVED***);
  ***REMOVED***
    getModel(modelId) ***REMOVED***
        return Promise.resolve()
            .then(() => ***REMOVED***
            return this.getModelName(modelId);
      ***REMOVED***)
            .then((modelName) => ***REMOVED***
            return this.getResourceWithShortName(modelName);
      ***REMOVED***);
  ***REMOVED***
    getOperation(operationName) ***REMOVED***
        return Promise.resolve()
            .then(() => ***REMOVED***
            return this.getResourceWithFullName(operationName);
      ***REMOVED***);
  ***REMOVED***
    listModels(options = ***REMOVED***}) ***REMOVED***
        if (!validator.isNonNullObject(options)) ***REMOVED***
            const err = new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'Invalid ListModelsOptions');
            return Promise.reject(err);
      ***REMOVED***
        if (typeof options.filter !== 'undefined' && !validator.isNonEmptyString(options.filter)) ***REMOVED***
            const err = new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'Invalid list filter.');
            return Promise.reject(err);
      ***REMOVED***
        if (typeof options.pageSize !== 'undefined') ***REMOVED***
            if (!validator.isNumber(options.pageSize)) ***REMOVED***
                const err = new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'Invalid page size.');
                return Promise.reject(err);
          ***REMOVED***
            if (options.pageSize < 1 || options.pageSize > 100) ***REMOVED***
                const err = new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'Page size must be between 1 and 100.');
                return Promise.reject(err);
          ***REMOVED***
      ***REMOVED***
        if (typeof options.pageToken !== 'undefined' && !validator.isNonEmptyString(options.pageToken)) ***REMOVED***
            const err = new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'Next page token must be a non-empty string.');
            return Promise.reject(err);
      ***REMOVED***
        return this.getProjectUrl()
            .then((url) => ***REMOVED***
            const request = ***REMOVED***
                method: 'GET',
                url: `$***REMOVED***url}/models`,
                data: options,
          ***REMOVED***;
            return this.sendRequest(request);
      ***REMOVED***);
  ***REMOVED***
    deleteModel(modelId) ***REMOVED***
        return this.getProjectUrl()
            .then((url) => ***REMOVED***
            const modelName = this.getModelName(modelId);
            const request = ***REMOVED***
                method: 'DELETE',
                url: `$***REMOVED***url}/$***REMOVED***modelName}`,
          ***REMOVED***;
            return this.sendRequest(request);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Handles a Long Running Operation coming back from the server.
     *
     * @param op - The operation to handle
     * @param options - The options for polling
     */
    handleOperation(op, options) ***REMOVED***
        if (op.done) ***REMOVED***
            if (op.response) ***REMOVED***
                return Promise.resolve(op.response);
          ***REMOVED***
            else if (op.error) ***REMOVED***
                const err = machine_learning_utils_1.FirebaseMachineLearningError.fromOperationError(op.error.code, op.error.message);
                return Promise.reject(err);
          ***REMOVED***
            // Done operations must have either a response or an error.
            throw new machine_learning_utils_1.FirebaseMachineLearningError('invalid-server-response', 'Invalid operation response.');
      ***REMOVED***
        // Operation is not done
        if (options?.wait) ***REMOVED***
            return this.pollOperationWithExponentialBackoff(op.name, options);
      ***REMOVED***
        const metadata = op.metadata || ***REMOVED***};
        const metadataType = metadata['@type'] || '';
        if (!metadataType.includes('ModelOperationMetadata')) ***REMOVED***
            throw new machine_learning_utils_1.FirebaseMachineLearningError('invalid-server-response', `Unknown Metadata type: $***REMOVED***JSON.stringify(metadata)}`);
      ***REMOVED***
        return this.getModel(extractModelId(metadata.name));
  ***REMOVED***
    // baseWaitMillis and maxWaitMillis should only ever be modified by unit tests to run faster.
    pollOperationWithExponentialBackoff(opName, options) ***REMOVED***
        const maxTimeMilliseconds = options?.maxTimeMillis ?? POLL_DEFAULT_MAX_TIME_MILLISECONDS;
        const baseWaitMillis = options?.baseWaitMillis ?? POLL_BASE_WAIT_TIME_MILLISECONDS;
        const maxWaitMillis = options?.maxWaitMillis ?? POLL_MAX_WAIT_TIME_MILLISECONDS;
        const poller = new api_request_1.ExponentialBackoffPoller(baseWaitMillis, maxWaitMillis, maxTimeMilliseconds);
        return poller.poll(() => ***REMOVED***
            return this.getOperation(opName)
                .then((responseData) => ***REMOVED***
                if (!responseData.done) ***REMOVED***
                    return null;
              ***REMOVED***
                if (responseData.error) ***REMOVED***
                    const err = machine_learning_utils_1.FirebaseMachineLearningError.fromOperationError(responseData.error.code, responseData.error.message);
                    throw err;
              ***REMOVED***
                return responseData.response;
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Gets the specified resource from the ML API. Resource names must be the short names without project
     * ID prefix (e.g. `models/123456789`).
     *
     * @param ***REMOVED***string} name Short name of the resource to get. e.g. 'models/12345'
     * @returns ***REMOVED***Promise<T>} A promise that fulfills with the resource.
     */
    getResourceWithShortName(name) ***REMOVED***
        return this.getProjectUrl()
            .then((url) => ***REMOVED***
            const request = ***REMOVED***
                method: 'GET',
                url: `$***REMOVED***url}/$***REMOVED***name}`,
          ***REMOVED***;
            return this.sendRequest(request);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Gets the specified resource from the ML API. Resource names must be the full names including project
     * number prefix.
     * @param fullName - Full resource name of the resource to get. e.g. projects/123465/operations/987654
     * @returns ***REMOVED***Promise<T>} A promise that fulfulls with the resource.
     */
    getResourceWithFullName(fullName) ***REMOVED***
        const request = ***REMOVED***
            method: 'GET',
            url: `$***REMOVED***ML_V1BETA2_API}/$***REMOVED***fullName}`
      ***REMOVED***;
        return this.sendRequest(request);
  ***REMOVED***
    sendRequest(request) ***REMOVED***
        request.headers = FIREBASE_VERSION_HEADER;
        return this.httpClient.send(request)
            .then((resp) => ***REMOVED***
            return resp.data;
      ***REMOVED***)
            .catch((err) => ***REMOVED***
            throw this.toFirebaseError(err);
      ***REMOVED***);
  ***REMOVED***
    toFirebaseError(err) ***REMOVED***
        if (err instanceof error_1.PrefixedFirebaseError) ***REMOVED***
            return err;
      ***REMOVED***
        const response = err.response;
        if (!response.isJson()) ***REMOVED***
            return new machine_learning_utils_1.FirebaseMachineLearningError('unknown-error', `Unexpected response with status: $***REMOVED***response.status} and body: $***REMOVED***response.text}`);
      ***REMOVED***
        const error = response.data.error || ***REMOVED***};
        let code = 'unknown-error';
        if (error.status && error.status in ERROR_CODE_MAPPING) ***REMOVED***
            code = ERROR_CODE_MAPPING[error.status];
      ***REMOVED***
        const message = error.message || `Unknown server error: $***REMOVED***response.text}`;
        return new machine_learning_utils_1.FirebaseMachineLearningError(code, message);
  ***REMOVED***
    getProjectUrl() ***REMOVED***
        return this.getProjectIdPrefix()
            .then((projectIdPrefix) => ***REMOVED***
            return `$***REMOVED***ML_V1BETA2_API}/$***REMOVED***projectIdPrefix}`;
      ***REMOVED***);
  ***REMOVED***
    getProjectIdPrefix() ***REMOVED***
        if (this.projectIdPrefix) ***REMOVED***
            return Promise.resolve(this.projectIdPrefix);
      ***REMOVED***
        return utils.findProjectId(this.app)
            .then((projectId) => ***REMOVED***
            if (!validator.isNonEmptyString(projectId)) ***REMOVED***
                throw new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'Failed to determine project ID. Initialize the SDK with service account credentials, or '
                    + 'set project ID as an app option. Alternatively, set the GOOGLE_CLOUD_PROJECT '
                    + 'environment variable.');
          ***REMOVED***
            this.projectIdPrefix = `projects/$***REMOVED***projectId}`;
            return this.projectIdPrefix;
      ***REMOVED***);
  ***REMOVED***
    getModelName(modelId) ***REMOVED***
        if (!validator.isNonEmptyString(modelId)) ***REMOVED***
            throw new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'Model ID must be a non-empty string.');
      ***REMOVED***
        if (modelId.indexOf('/') !== -1) ***REMOVED***
            throw new machine_learning_utils_1.FirebaseMachineLearningError('invalid-argument', 'Model ID must not contain any "/" characters.');
      ***REMOVED***
        return `models/$***REMOVED***modelId}`;
  ***REMOVED***
}
exports.MachineLearningApiClient = MachineLearningApiClient;
const ERROR_CODE_MAPPING = ***REMOVED***
    INVALID_ARGUMENT: 'invalid-argument',
    NOT_FOUND: 'not-found',
    RESOURCE_EXHAUSTED: 'resource-exhausted',
    UNAUTHENTICATED: 'authentication-error',
    UNKNOWN: 'unknown-error',
};
function extractModelId(resourceName) ***REMOVED***
    return resourceName.split('/').pop();
}
