// jsdoc/lib/jsdoc/augment.js
/**
 * Provides methods for augmenting the parse results based on their content.
 * @module jsdoc/augment
 */

const doop = require('jsdoc/util/doop');
const jsdoc = ***REMOVED***
    doclet: require('jsdoc/doclet')
};
const name = require('jsdoc/name');

const hasOwnProp = Object.prototype.hasOwnProperty;

function mapDependencies(index, propertyName) ***REMOVED***
    const dependencies = ***REMOVED***};
    let doc;
    let doclets;
    const kinds = ['class', 'external', 'interface', 'mixin'];
    let len = 0;

    Object.keys(index).forEach(indexName => ***REMOVED***
        doclets = index[indexName];
        for (let i = 0, ii = doclets.length; i < ii; i++) ***REMOVED***
            doc = doclets[i];
            if (kinds.includes(doc.kind)) ***REMOVED***
                dependencies[indexName] = ***REMOVED***};
                if (hasOwnProp.call(doc, propertyName)) ***REMOVED***
                    len = doc[propertyName].length;
                    for (let j = 0; j < len; j++) ***REMOVED***
                        dependencies[indexName][doc[propertyName][j]] = true;
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***);

    return dependencies;
}

class Sorter ***REMOVED***
    constructor(dependencies) ***REMOVED***
        this.dependencies = dependencies;
        this.visited = ***REMOVED***};
        this.sorted = [];
  ***REMOVED***

    visit(key) ***REMOVED***
        if (!(key in this.visited)) ***REMOVED***
            this.visited[key] = true;

            if (this.dependencies[key]) ***REMOVED***
                Object.keys(this.dependencies[key]).forEach(path => ***REMOVED***
                    this.visit(path);
              ***REMOVED***);
          ***REMOVED***

            this.sorted.push(key);
      ***REMOVED***
  ***REMOVED***

    sort() ***REMOVED***
        Object.keys(this.dependencies).forEach(key => ***REMOVED***
            this.visit(key);
      ***REMOVED***);

        return this.sorted;
  ***REMOVED***
}

function sort(dependencies) ***REMOVED***
    const sorter = new Sorter(dependencies);

    return sorter.sort();
}

function getMembers(longname, ***REMOVED***index}, scopes) ***REMOVED***
    const memberof = index.memberof[longname] || [];
    const members = [];

    memberof.forEach(candidate => ***REMOVED***
        if (scopes.includes(candidate.scope)) ***REMOVED***
            members.push(candidate);
      ***REMOVED***
  ***REMOVED***);

    return members;
}

function getDocumentedLongname(longname, ***REMOVED***index}) ***REMOVED***
    const doclets = index.documented[longname] || [];

    return doclets[doclets.length - 1];
}

function addDocletProperty(doclets, propName, value) ***REMOVED***
    for (let i = 0, l = doclets.length; i < l; i++) ***REMOVED***
        doclets[i][propName] = value;
  ***REMOVED***
}

function reparentDoclet(***REMOVED***longname}, child) ***REMOVED***
    const parts = name.shorten(child.longname);

    parts.memberof = longname;
    child.memberof = longname;
    child.longname = name.combine(parts);
}

function parentIsClass(***REMOVED***kind}) ***REMOVED***
    return kind === 'class';
}

function staticToInstance(doclet) ***REMOVED***
    const parts = name.shorten(doclet.longname);

    parts.scope = name.SCOPE.PUNC.INSTANCE;
    doclet.longname = name.combine(parts);
    doclet.scope = name.SCOPE.NAMES.INSTANCE;
}

/**
 * Update the list of doclets to be added to another symbol.
 *
 * We add only one doclet per longname. For example: If `ClassA` inherits from two classes that both
 * use the same method name, `ClassA` gets docs for one method rather than two.
 *
 * Also, the last symbol wins for any given longname. For example: If you write `@extends Class1
 * @extends Class2`, and both classes have an instance method called `myMethod`, you get the docs
 * from `Class2#myMethod`.
 *
 * @private
 * @param ***REMOVED***module:jsdoc/doclet.Doclet} doclet - The doclet to be added.
 * @param ***REMOVED***Array.<module:jsdoc/doclet.Doclet>} additions - An array of doclets that will be added to
 * another symbol.
 * @param ***REMOVED***Object.<string, number>} indexes - A dictionary of indexes into the `additions` array.
 * Each key is a longname, and each value is the index of the longname's doclet.
 * @return ***REMOVED***void}
 */
function updateAddedDoclets(doclet, additions, indexes) ***REMOVED***
    if (typeof indexes[doclet.longname] !== 'undefined') ***REMOVED***
        // replace the existing doclet
        additions[indexes[doclet.longname]] = doclet;
  ***REMOVED***
    else ***REMOVED***
        // add the doclet to the array, and track its index
        additions.push(doclet);
        indexes[doclet.longname] = additions.length - 1;
  ***REMOVED***
}

/**
 * Update the index of doclets whose `undocumented` property is not `true`.
 *
 * @private
 * @param ***REMOVED***module:jsdoc/doclet.Doclet} doclet - The doclet to be added to the index.
 * @param ***REMOVED***Object.<string, Array.<module:jsdoc/doclet.Doclet>>} documented - The index of doclets
 * whose `undocumented` property is not `true`.
 * @return ***REMOVED***void}
 */
function updateDocumentedDoclets(doclet, documented) ***REMOVED***
    if ( !hasOwnProp.call(documented, doclet.longname) ) ***REMOVED***
        documented[doclet.longname] = [];
  ***REMOVED***

    documented[doclet.longname].push(doclet);
}

/**
 * Update the index of doclets with a `memberof` value.
 *
 * @private
 * @param ***REMOVED***module:jsdoc/doclet.Doclet} doclet - The doclet to be added to the index.
 * @param ***REMOVED***Object.<string, Array.<module:jsdoc/doclet.Doclet>>} memberof - The index of doclets
 * with a `memberof` value.
 * @return ***REMOVED***void}
 */
function updateMemberofDoclets(doclet, memberof) ***REMOVED***
    if (doclet.memberof) ***REMOVED***
        if ( !hasOwnProp.call(memberof, doclet.memberof) ) ***REMOVED***
            memberof[doclet.memberof] = [];
      ***REMOVED***

        memberof[doclet.memberof].push(doclet);
  ***REMOVED***
}

function explicitlyInherits(doclets) ***REMOVED***
    let doclet;
    let inherits = false;

    for (let i = 0, l = doclets.length; i < l; i++) ***REMOVED***
        doclet = doclets[i];
        if (typeof doclet.inheritdoc !== 'undefined' || typeof doclet.override !== 'undefined') ***REMOVED***
            inherits = true;
            break;
      ***REMOVED***
  ***REMOVED***

    return inherits;
}

function changeMemberof(longname, newMemberof) ***REMOVED***
    const atoms = name.shorten(longname);

    atoms.memberof = newMemberof;

    return name.combine(atoms);
}

// TODO: try to reduce overlap with similar methods
function getInheritedAdditions(doclets, docs, ***REMOVED***documented, memberof}) ***REMOVED***
    let additionIndexes;
    const additions = [];
    let childDoclet;
    let childLongname;
    let doc;
    let parentDoclet;
    let parentMembers;
    let parents;
    let member;
    let parts;

    // doclets will be undefined if the inherited symbol isn't documented
    doclets = doclets || [];

    for (let i = 0, ii = doclets.length; i < ii; i++) ***REMOVED***
        doc = doclets[i];
        parents = doc.augments;

        if ( parents && (doc.kind === 'class' || doc.kind === 'interface') ) ***REMOVED***
            // reset the lookup table of added doclet indexes by longname
            additionIndexes = ***REMOVED***};

            for (let j = 0, jj = parents.length; j < jj; j++) ***REMOVED***
                parentMembers = getMembers(parents[j], docs, ['instance']);

                for (let k = 0, kk = parentMembers.length; k < kk; k++) ***REMOVED***
                    parentDoclet = parentMembers[k];

                    // We only care about symbols that are documented.
                    if (parentDoclet.undocumented) ***REMOVED***
                        continue;
                  ***REMOVED***

                    childLongname = changeMemberof(parentDoclet.longname, doc.longname);
                    childDoclet = getDocumentedLongname(childLongname, docs) || ***REMOVED***};

                    // We don't want to fold in properties from the child doclet if it had an
                    // `@inheritdoc` tag.
                    if (hasOwnProp.call(childDoclet, 'inheritdoc')) ***REMOVED***
                        childDoclet = ***REMOVED***};
                  ***REMOVED***

                    member = jsdoc.doclet.combine(childDoclet, parentDoclet);

                    if (!member.inherited) ***REMOVED***
                        member.inherits = member.longname;
                  ***REMOVED***
                    member.inherited = true;

                    member.memberof = doc.longname;
                    parts = name.shorten(member.longname);
                    parts.memberof = doc.longname;
                    member.longname = name.combine(parts);

                    // Indicate what the descendant is overriding. (We only care about the closest
                    // ancestor. For classes A > B > C, if B#a overrides A#a, and C#a inherits B#a,
                    // we don't want the doclet for C#a to say that it overrides A#a.)
                    if ( hasOwnProp.call(docs.index.longname, member.longname) ) ***REMOVED***
                        member.overrides = parentDoclet.longname;
                  ***REMOVED***
                    else ***REMOVED***
                        delete member.overrides;
                  ***REMOVED***

                    // Add the ancestor's docs unless the descendant overrides the ancestor AND
                    // documents the override.
                    if ( !hasOwnProp.call(documented, member.longname) ) ***REMOVED***
                        updateAddedDoclets(member, additions, additionIndexes);
                        updateDocumentedDoclets(member, documented);
                        updateMemberofDoclets(member, memberof);
                  ***REMOVED***
                    // If the descendant used an @inheritdoc or @override tag, add the ancestor's
                    // docs, and ignore the existing doclets.
                    else if ( explicitlyInherits(documented[member.longname]) ) ***REMOVED***
                        // Ignore any existing doclets. (This is safe because we only get here if
                        // `member.longname` is an own property of `documented`.)
                        addDocletProperty(documented[member.longname], 'ignore', true);

                        updateAddedDoclets(member, additions, additionIndexes);
                        updateDocumentedDoclets(member, documented);
                        updateMemberofDoclets(member, memberof);

                        // Remove property that's no longer accurate.
                        if (member.virtual) ***REMOVED***
                            delete member.virtual;
                      ***REMOVED***
                        // Remove properties that we no longer need.
                        if (member.inheritdoc) ***REMOVED***
                            delete member.inheritdoc;
                      ***REMOVED***
                        if (member.override) ***REMOVED***
                            delete member.override;
                      ***REMOVED***
                  ***REMOVED***
                    // If the descendant overrides the ancestor and documents the override,
                    // update the doclets to indicate what the descendant is overriding.
                    else ***REMOVED***
                        addDocletProperty(documented[member.longname], 'overrides',
                            parentDoclet.longname);
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***

    return additions;
}

function updateMixes(mixedDoclet, mixedLongname) ***REMOVED***
    let idx;
    let mixedName;
    let names;

    // take the fast path if there's no array of mixed-in longnames
    if (!mixedDoclet.mixes) ***REMOVED***
        mixedDoclet.mixes = [mixedLongname];
  ***REMOVED***
    else ***REMOVED***
        // find the short name of the longname we're mixing in
        mixedName = name.shorten(mixedLongname).name;
        // find the short name of each previously mixed-in symbol
        // TODO: why do we run a map if we always shorten the same value? this looks like a bug...
        names = mixedDoclet.mixes.map(() => name.shorten(mixedDoclet.longname).name);

        // if we're mixing `myMethod` into `MixinC` from `MixinB`, and `MixinB` had the method mixed
        // in from `MixinA`, don't show `MixinA.myMethod` in the `mixes` list
        idx = names.indexOf(mixedName);
        if (idx !== -1) ***REMOVED***
            mixedDoclet.mixes.splice(idx, 1);
      ***REMOVED***

        mixedDoclet.mixes.push(mixedLongname);
  ***REMOVED***
}

// TODO: try to reduce overlap with similar methods
function getMixedInAdditions(mixinDoclets, allDoclets, ***REMOVED***documented, memberof}) ***REMOVED***
    let additionIndexes;
    const additions = [];
    const commentedDoclets = documented;
    let doclet;
    let mixedDoclet;
    let mixedDoclets;
    let mixes;

    // mixinDoclets will be undefined if the mixed-in symbol isn't documented
    mixinDoclets = mixinDoclets || [];

    for (let i = 0, ii = mixinDoclets.length; i < ii; i++) ***REMOVED***
        doclet = mixinDoclets[i];
        mixes = doclet.mixes;

        if (mixes) ***REMOVED***
            // reset the lookup table of added doclet indexes by longname
            additionIndexes = ***REMOVED***};

            for (let j = 0, jj = mixes.length; j < jj; j++) ***REMOVED***
                mixedDoclets = getMembers(mixes[j], allDoclets, ['static']);

                for (let k = 0, kk = mixedDoclets.length; k < kk; k++) ***REMOVED***
                    // We only care about symbols that are documented.
                    if (mixedDoclets[k].undocumented) ***REMOVED***
                        continue;
                  ***REMOVED***

                    mixedDoclet = doop(mixedDoclets[k]);

                    updateMixes(mixedDoclet, mixedDoclet.longname);
                    mixedDoclet.mixed = true;

                    reparentDoclet(doclet, mixedDoclet);

                    // if we're mixing into a class, treat the mixed-in symbol as an instance member
                    if (parentIsClass(doclet)) ***REMOVED***
                        staticToInstance(mixedDoclet);
                  ***REMOVED***

                    updateAddedDoclets(mixedDoclet, additions, additionIndexes);
                    updateDocumentedDoclets(mixedDoclet, commentedDoclets);
                    updateMemberofDoclets(mixedDoclet, memberof);
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***

    return additions;
}

function updateImplements(implDoclets, implementedLongname) ***REMOVED***
    if ( !Array.isArray(implDoclets) ) ***REMOVED***
        implDoclets = [implDoclets];
  ***REMOVED***

    implDoclets.forEach(implDoclet => ***REMOVED***
        if ( !hasOwnProp.call(implDoclet, 'implements') ) ***REMOVED***
            implDoclet.implements = [];
      ***REMOVED***

        if (!implDoclet.implements.includes(implementedLongname)) ***REMOVED***
            implDoclet.implements.push(implementedLongname);
      ***REMOVED***
  ***REMOVED***);
}

// TODO: try to reduce overlap with similar methods
function getImplementedAdditions(implDoclets, allDoclets, ***REMOVED***documented, memberof}) ***REMOVED***
    let additionIndexes;
    const additions = [];
    let childDoclet;
    let childLongname;
    const commentedDoclets = documented;
    let doclet;
    let implementations;
    let implExists;
    let implementationDoclet;
    let interfaceDoclets;
    let parentDoclet;

    // interfaceDoclets will be undefined if the implemented symbol isn't documented
    implDoclets = implDoclets || [];

    for (let i = 0, ii = implDoclets.length; i < ii; i++) ***REMOVED***
        doclet = implDoclets[i];
        implementations = doclet.implements;

        if (implementations) ***REMOVED***
            // reset the lookup table of added doclet indexes by longname
            additionIndexes = ***REMOVED***};

            for (let j = 0, jj = implementations.length; j < jj; j++) ***REMOVED***
                interfaceDoclets = getMembers(implementations[j], allDoclets, ['instance']);

                for (let k = 0, kk = interfaceDoclets.length; k < kk; k++) ***REMOVED***
                    parentDoclet = interfaceDoclets[k];

                    // We only care about symbols that are documented.
                    if (parentDoclet.undocumented) ***REMOVED***
                        continue;
                  ***REMOVED***

                    childLongname = changeMemberof(parentDoclet.longname, doclet.longname);
                    childDoclet = getDocumentedLongname(childLongname, allDoclets) || ***REMOVED***};

                    // We don't want to fold in properties from the child doclet if it had an
                    // `@inheritdoc` tag.
                    if (hasOwnProp.call(childDoclet, 'inheritdoc')) ***REMOVED***
                        childDoclet = ***REMOVED***};
                  ***REMOVED***

                    implementationDoclet = jsdoc.doclet.combine(childDoclet, parentDoclet);

                    reparentDoclet(doclet, implementationDoclet);
                    updateImplements(implementationDoclet, parentDoclet.longname);

                    // If there's no implementation, move along.
                    implExists = hasOwnProp.call(allDoclets.index.longname,
                        implementationDoclet.longname);
                    if (!implExists) ***REMOVED***
                        continue;
                  ***REMOVED***

                    // Add the interface's docs unless the implementation is already documented.
                    if ( !hasOwnProp.call(commentedDoclets, implementationDoclet.longname) ) ***REMOVED***
                        updateAddedDoclets(implementationDoclet, additions, additionIndexes);
                        updateDocumentedDoclets(implementationDoclet, commentedDoclets);
                        updateMemberofDoclets(implementationDoclet, memberof);
                  ***REMOVED***
                    // If the implementation used an @inheritdoc or @override tag, add the
                    // interface's docs, and ignore the existing doclets.
                    else if ( explicitlyInherits(commentedDoclets[implementationDoclet.longname]) ) ***REMOVED***
                        // Ignore any existing doclets. (This is safe because we only get here if
                        // `implementationDoclet.longname` is an own property of
                        // `commentedDoclets`.)
                        addDocletProperty(commentedDoclets[implementationDoclet.longname], 'ignore',
                            true);

                        updateAddedDoclets(implementationDoclet, additions, additionIndexes);
                        updateDocumentedDoclets(implementationDoclet, commentedDoclets);
                        updateMemberofDoclets(implementationDoclet, memberof);

                        // Remove property that's no longer accurate.
                        if (implementationDoclet.virtual) ***REMOVED***
                            delete implementationDoclet.virtual;
                      ***REMOVED***
                        // Remove properties that we no longer need.
                        if (implementationDoclet.inheritdoc) ***REMOVED***
                            delete implementationDoclet.inheritdoc;
                      ***REMOVED***
                        if (implementationDoclet.override) ***REMOVED***
                            delete implementationDoclet.override;
                      ***REMOVED***
                  ***REMOVED***
                    // If there's an implementation, and it's documented, update the doclets to
                    // indicate what the implementation is implementing.
                    else ***REMOVED***
                        updateImplements(commentedDoclets[implementationDoclet.longname],
                            parentDoclet.longname);
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***

    return additions;
}

function augment(doclets, propertyName, docletFinder) ***REMOVED***
    const index = doclets.index.longname;
    const dependencies = sort( mapDependencies(index, propertyName) );

    dependencies.forEach(depName => ***REMOVED***
        const additions = docletFinder(index[depName], doclets, doclets.index);

        additions.forEach(addition => ***REMOVED***
            const longname = addition.longname;

            if ( !hasOwnProp.call(index, longname) ) ***REMOVED***
                index[longname] = [];
          ***REMOVED***
            index[longname].push(addition);
            doclets.push(addition);
      ***REMOVED***);
  ***REMOVED***);
}

/**
 * Add doclets to reflect class inheritance.
 *
 * For example, if `ClassA` has the instance method `myMethod`, and `ClassB` inherits from `ClassA`,
 * calling this method creates a new doclet for `ClassB#myMethod`.
 *
 * @param ***REMOVED***!Array.<module:jsdoc/doclet.Doclet>} doclets - The doclets generated by JSDoc.
 * @param ***REMOVED***!Object} doclets.index - The doclet index.
 * @return ***REMOVED***void}
 */
exports.addInherited = doclets => ***REMOVED***
    augment(doclets, 'augments', getInheritedAdditions);
};

/**
 * Add doclets to reflect mixins. When a symbol is mixed into a class, the class' version of the
 * mixed-in symbol is treated as an instance member.
 *
 * For example:
 *
 * + If `MixinA` has the static method `myMethod`, and `MixinB` mixes `MixinA`, calling this method
 * creates a new doclet for the static method `MixinB.myMethod`.
 * + If `MixinA` has the static method `myMethod`, and `ClassA` mixes `MixinA`, calling this method
 * creates a new doclet for the instance method `ClassA#myMethod`.
 *
 * @param ***REMOVED***!Array.<module:jsdoc/doclet.Doclet>} doclets - The doclets generated by JSDoc.
 * @param ***REMOVED***!Object} doclets.index - The doclet index.
 * @return ***REMOVED***void}
 */
exports.addMixedIn = doclets => ***REMOVED***
    augment(doclets, 'mixes', getMixedInAdditions);
};

/**
 * Add and update doclets to reflect implementations of interfaces.
 *
 * For example, if `InterfaceA` has the instance method `myMethod`, and `ClassA` implements
 * `InterfaceA`, calling this method does the following:
 *
 * + Updates `InterfaceA` to indicate that it is implemented by `ClassA`
 * + Updates `InterfaceA#myMethod` to indicate that it is implemented by `ClassA#myMethod`
 * + Updates `ClassA#myMethod` to indicate that it implements `InterfaceA#myMethod`
 *
 * If `ClassA#myMethod` used the `@override` or `@inheritdoc` tag, calling this method would also
 * generate a new doclet that reflects the interface's documentation for `InterfaceA#myMethod`.
 *
 * @param ***REMOVED***!Array.<module:jsdoc/doclet.Doclet>} docs - The doclets generated by JSDoc.
 * @param ***REMOVED***!Object} doclets.index - The doclet index.
 * @return ***REMOVED***void}
 */
exports.addImplemented = doclets => ***REMOVED***
    augment(doclets, 'implements', getImplementedAdditions);
};

/**
 * Add and update doclets to reflect all of the following:
 *
 * + Inherited classes
 * + Mixins
 * + Interface implementations
 *
 * Calling this method is equivalent to calling all other methods exported by this module.
 *
 * @return ***REMOVED***void}
 */
exports.augmentAll = doclets => ***REMOVED***
    exports.addMixedIn(doclets);
    exports.addImplemented(doclets);
    exports.addInherited(doclets);
    // look for implemented doclets again, in case we inherited an interface
    exports.addImplemented(doclets);
};
