// jsdoc/lib/jsdoc/config.js
/**
 * @module jsdoc/config
 */
const stripBom = require('jsdoc/util/stripbom');
const stripJsonComments = require('strip-json-comments');

function mergeRecurse(target, source) ***REMOVED***
    Object.keys(source).forEach(p => ***REMOVED***
        if ( source[p].constructor === Object ) ***REMOVED***
            if ( !target[p] ) ***REMOVED***
                target[p] = ***REMOVED***};
          ***REMOVED***
            mergeRecurse(target[p], source[p]);
      ***REMOVED***
        else ***REMOVED***
            target[p] = source[p];
      ***REMOVED***
  ***REMOVED***);

    return target;
}

// required config values, override these defaults in your config.json if necessary
const defaults = ***REMOVED***
    plugins: [],
    recurseDepth: 10,
    source: ***REMOVED***
        includePattern: '.+\\.js(doc|x)?$',
        excludePattern: ''
  ***REMOVED***,
    sourceType: 'module',
    tags: ***REMOVED***
        allowUnknownTags: true,
        dictionaries: ['jsdoc', 'closure']
  ***REMOVED***,
    templates: ***REMOVED***
        monospaceLinks: false,
        cleverLinks: false
  ***REMOVED***
};

/**
 * Represents a JSDoc application configuration.
 */
class Config ***REMOVED***
    /**
     * @param ***REMOVED***(string|object)} [jsonOrObject] - The contents of config.json, or a JavaScript object
     * exported from a .js config file.
     */
    constructor(jsonOrObject) ***REMOVED***
        if (typeof jsonOrObject === 'undefined') ***REMOVED***
            jsonOrObject = ***REMOVED***};
      ***REMOVED***

        if (typeof jsonOrObject === 'string') ***REMOVED***
            jsonOrObject = JSON.parse( (stripJsonComments(stripBom.strip(jsonOrObject)) || '***REMOVED***}') );
      ***REMOVED***

        if (typeof jsonOrObject !== 'object') ***REMOVED***
            jsonOrObject = ***REMOVED***};
      ***REMOVED***

        this._config = mergeRecurse(defaults, jsonOrObject);
  ***REMOVED***

    /**
     * Get the merged configuration values.
     */
    get() ***REMOVED***
        return this._config;
  ***REMOVED***
}

module.exports = Config;
