// jsdoc/lib/jsdoc/env.js
/**
 * Data about the environment in which JSDoc is running, including the configuration settings that
 * were used to run JSDoc.
 *
 * @module jsdoc/env
 */
module.exports = ***REMOVED***
    /**
     * The times at which JSDoc started and finished.
     *
     * @type ***REMOVED***Object}
     * @property ***REMOVED***Date} start - The time at which JSDoc started running.
     * @property ***REMOVED***Date} finish - The time at which JSDoc finished running.
     */
    run: ***REMOVED***
        start: new Date(),
        finish: null
  ***REMOVED***,

    /**
     * The command-line arguments passed to JSDoc.
     *
     * @type ***REMOVED***Array<*>}
     */
    args: [],

    /**
     * The data parsed from JSDoc's configuration file.
     *
     * @type Object<string, *>
     */
    conf: ***REMOVED***},

    /**
     * The absolute path to the base directory in which JSDoc is located. Set at startup.
     *
     * @private
     * @type ***REMOVED***string}
     */
    dirname: null,

    /**
     * The user's working directory at the time when JSDoc started running.
     *
     * @private
     * @type ***REMOVED***string}
     */
    pwd: null,

    /**
     * The command-line arguments, parsed into a key/value hash.
     *
     * @type ***REMOVED***Object}
     * @example if (global.env.opts.help) ***REMOVED*** console.log('Helpful message.'); }
    */
    opts: ***REMOVED***},

    /**
     * The source files that JSDoc will parse.
     *
     * @type ***REMOVED***Array<string>}
     * @memberof env
     */
    sourceFiles: [],

    /**
     * The JSDoc version number and revision date.
     *
     * @type ***REMOVED***Object<string, string>}
     * @property ***REMOVED***string} number - The JSDoc version number.
     * @property ***REMOVED***string} revision - The JSDoc revision number, expressed as a UTC date string.
     */
    version: ***REMOVED***
        number: null,
        revision: null
  ***REMOVED***
};
