// jsdoc/lib/jsdoc/package.js
const logger = require('jsdoc/util/logger');
const stripBom = require('jsdoc/util/stripbom');

/**
 * Provides access to information about a JavaScript package.
 *
 * @module jsdoc/package
 * @see https://www.npmjs.org/doc/files/package.json.html
 */

// Collect all of the license information from a `package.json` file.
function getLicenses(packageInfo) ***REMOVED***
    const licenses = packageInfo.licenses ? packageInfo.licenses.slice(0) : [];

    if (packageInfo.license) ***REMOVED***
        licenses.push(***REMOVED*** type: packageInfo.license });
  ***REMOVED***

    return licenses;
}

/**
 * Information about where to report bugs in the package.
 *
 * @typedef ***REMOVED***Object} module:jsdoc/package.Package~BugInfo
 * @property ***REMOVED***string} email - The email address for reporting bugs.
 * @property ***REMOVED***string} url - The URL for reporting bugs.
 */

/**
 * Information about a package's software license.
 *
 * @typedef ***REMOVED***Object} module:jsdoc/package.Package~LicenseInfo
 * @property ***REMOVED***string} type - An identifier for the type of license.
 * @property ***REMOVED***string} url - The URL for the complete text of the license.
 */

/**
 * Information about a package author or contributor.
 *
 * @typedef ***REMOVED***Object} module:jsdoc/package.Package~PersonInfo
 * @property ***REMOVED***string} name - The person's full name.
 * @property ***REMOVED***string} email - The person's email address.
 * @property ***REMOVED***string} url - The URL of the person's website.
 */

/**
 * Information about a package's version-control repository.
 *
 * @typedef ***REMOVED***Object} module:jsdoc/package.Package~RepositoryInfo
 * @property ***REMOVED***string} type - The type of version-control system that the repository uses (for
 * example, `git` or `svn`).
 * @property ***REMOVED***string} url - The URL for the repository.
 */

/**
 * Information about a JavaScript package. JSDoc can extract package information from
 * `package.json` files that follow the
 * [npm specification](https://www.npmjs.org/doc/files/package.json.html).
 *
 * **Note**: JSDoc does not validate or normalize the contents of `package.json` files. If your
 * `package.json` file does not follow the npm specification, some properties of the `Package`
 * object may not use the format documented here.
 */
class Package ***REMOVED***
    /**
     * @param ***REMOVED***string} json - The contents of the `package.json` file.
     */
    constructor(json) ***REMOVED***
        let packageInfo;

        /**
         * The string identifier that is shared by all `Package` objects.
         *
         * @readonly
         * @default
         * @type ***REMOVED***string}
         */
        this.kind = 'package';

        try ***REMOVED***
            packageInfo = JSON.parse(stripBom.strip(json) || '***REMOVED***}');
      ***REMOVED***
        catch (e) ***REMOVED***
            logger.error('Unable to parse the package file: %s', e.message);
            packageInfo = ***REMOVED***};
      ***REMOVED***

        if (packageInfo.name) ***REMOVED***
            /**
             * The package name.
             *
             * @type ***REMOVED***string}
             */
            this.name = packageInfo.name;
      ***REMOVED***

        /**
         * The unique longname for this `Package` object.
         *
         * @type ***REMOVED***string}
         */
        this.longname = `$***REMOVED***this.kind}:$***REMOVED***this.name}`;

        if (packageInfo.author) ***REMOVED***
            /**
             * The author of this package. Contains either a
             * ***REMOVED***@link module:jsdoc/package.Package~PersonInfo PersonInfo} object or a string with
             * information about the author.
             *
             * @type ***REMOVED***(module:jsdoc/package.Package~PersonInfo|string)}
             * @since 3.3.0
             */
            this.author = packageInfo.author;
      ***REMOVED***

        if (packageInfo.bugs) ***REMOVED***
            /**
             * Information about where to report bugs in the project. May contain a URL, a string, or an
             * object with more detailed information.
             *
             * @type ***REMOVED***(string|module:jsdoc/package.Package~BugInfo)}
             * @since 3.3.0
             */
            this.bugs = packageInfo.bugs;
      ***REMOVED***

        if (packageInfo.contributors) ***REMOVED***
            /**
             * The contributors to this package.
             *
             * @type ***REMOVED***Array.<(module:jsdoc/package.Package~PersonInfo|string)>}
             * @since 3.3.0
             */
            this.contributors = packageInfo.contributors;
      ***REMOVED***

        if (packageInfo.dependencies) ***REMOVED***
            /**
             * The dependencies for this package.
             *
             * @type ***REMOVED***Object}
             * @since 3.3.0
             */
            this.dependencies = packageInfo.dependencies;
      ***REMOVED***

        if (packageInfo.description) ***REMOVED***
            /**
             * A brief description of the package.
             *
             * @type ***REMOVED***string}
             */
            this.description = packageInfo.description;
      ***REMOVED***

        if (packageInfo.devDependencies) ***REMOVED***
            /**
             * The development dependencies for this package.
             *
             * @type ***REMOVED***Object}
             * @since 3.3.0
             */
            this.devDependencies = packageInfo.devDependencies;
      ***REMOVED***

        if (packageInfo.engines) ***REMOVED***
            /**
             * The JavaScript engines that this package supports. Each key is a string that identifies
             * the engine (for example, `node`). Each value is a
             * [semver](https://www.npmjs.org/doc/misc/semver.html)-compliant version number for the
             * engine.
             *
             * @type ***REMOVED***Object}
             * @since 3.3.0
             */
            this.engines = packageInfo.engines;
      ***REMOVED***

        /**
         * The source files associated with the package.
         *
         * New `Package` objects always contain an empty array, regardless of whether the `package.json`
         * file includes a `files` property.
         *
         * After JSDoc parses your input files, it sets this property to a list of paths to your input
         * files.
         *
         * @type ***REMOVED***Array.<string>}
         */
        this.files = [];

        if (packageInfo.homepage) ***REMOVED***
            /**
             * The URL for the package's homepage.
             *
             * @type ***REMOVED***string}
             * @since 3.3.0
             */
            this.homepage = packageInfo.homepage;
      ***REMOVED***

        if (packageInfo.keywords) ***REMOVED***
            /**
             * Keywords to help users find the package.
             *
             * @type ***REMOVED***Array.<string>}
             * @since 3.3.0
             */
            this.keywords = packageInfo.keywords;
      ***REMOVED***

        if (packageInfo.license || packageInfo.licenses) ***REMOVED***
            /**
             * The licenses used by this package. Combines information from the `package.json` file's
             * `license` property and the deprecated `licenses` property.
             *
             * @type ***REMOVED***Array.<module:jsdoc/package.Package~LicenseInfo>}
             */
            this.licenses = getLicenses(packageInfo);
      ***REMOVED***

        if (packageInfo.main) ***REMOVED***
            /**
             * The module ID that provides the primary entry point to the package. For example, if your
             * package is a CommonJS module, and the value of this property is `foo`, users should be
             * able to load your module with `require('foo')`.
             *
             * @type ***REMOVED***string}
             * @since 3.3.0
             */
            this.main = packageInfo.main;
      ***REMOVED***

        if (packageInfo.repository) ***REMOVED***
            /**
             * The version-control repository for the package.
             *
             * @type ***REMOVED***module:jsdoc/package.Package~RepositoryInfo}
             * @since 3.3.0
             */
            this.repository = packageInfo.repository;
      ***REMOVED***

        if (packageInfo.version) ***REMOVED***
            /**
             * The [semver](https://www.npmjs.org/doc/misc/semver.html)-compliant version number of the
             * package.
             *
             * @type ***REMOVED***string}
             * @since 3.2.0
             */
            this.version = packageInfo.version;
      ***REMOVED***
  ***REMOVED***
}
exports.Package = Package;
