// jsdoc/lib/jsdoc/schema.js
/**
 * Schema for validating JSDoc doclets.
 * @module jsdoc/schema
 * @see <https://trac.tools.ietf.org/html/draft-wright-json-schema-validation-01>
 */
// JSON schema types
const ARRAY = 'array';
const BOOLEAN = 'boolean';
const NULL = 'null';
const NUMBER = 'number';
const OBJECT = 'object';
const STRING = 'string';

const BOOLEAN_OPTIONAL = [BOOLEAN, NULL];
const STRING_OPTIONAL = [STRING, NULL];

const EVENT_REGEXP = 'event:[\\S]+';
const PACKAGE_REGEXP = 'package:[\\S]+';

const STRING_SCHEMA = ***REMOVED***
    type: STRING
};

// information about the code associated with a doclet
const META_SCHEMA = exports.META_SCHEMA = ***REMOVED***
    type: OBJECT,
    additionalProperties: false,
    properties: ***REMOVED***
        code: ***REMOVED***
            type: OBJECT,
            additionalProperties: false,
            properties: ***REMOVED***
                funcscope: ***REMOVED***
                    type: STRING
              ***REMOVED***,
                id: ***REMOVED***
                    type: STRING
              ***REMOVED***,
                name: ***REMOVED***},
                node: ***REMOVED***
                    type: OBJECT
              ***REMOVED***,
                paramnames: ***REMOVED***
                    type: ARRAY,
                    uniqueItems: true,
                    items: ***REMOVED***
                        type: STRING
                  ***REMOVED***
              ***REMOVED***,
                type: ***REMOVED***
                    type: STRING
              ***REMOVED***,
                value: ***REMOVED***}
          ***REMOVED***
      ***REMOVED***,
        columnno: ***REMOVED***
            title: 'The column number of the code associated with this doclet.',
            type: NUMBER
      ***REMOVED***,
        filename: ***REMOVED***
            title: 'The name of the file that contains the code associated with this doclet.',
            type: STRING
      ***REMOVED***,
        lineno: ***REMOVED***
            title: 'The line number of the code associated with this doclet.',
            type: NUMBER
      ***REMOVED***,
        path: ***REMOVED***
            title: 'The path in which the code associated with this doclet is located.',
            type: STRING
      ***REMOVED***,
        range: ***REMOVED***
            title: 'The positions of the first and last characters of the code associated with ' +
                'this doclet.',
            type: ARRAY,
            minItems: 2,
            maxItems: 2,
            items: ***REMOVED***
                type: NUMBER
          ***REMOVED***
      ***REMOVED***,
        vars: ***REMOVED***
            type: OBJECT
      ***REMOVED***
  ***REMOVED***
};

// type property containing type names
const TYPE_PROPERTY_SCHEMA = exports.TYPE_PROPERTY_SCHEMA = ***REMOVED***
    type: OBJECT,
    additionalProperties: false,
    properties: ***REMOVED***
        names: ***REMOVED***
            type: ARRAY,
            minItems: 1,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        // type parser output
        parsedType: ***REMOVED***
            type: OBJECT,
            additionalProperties: true
      ***REMOVED***
  ***REMOVED***
};

// enumeration properties
const ENUM_PROPERTY_SCHEMA = exports.ENUM_PROPERTY_SCHEMA = ***REMOVED***
    type: OBJECT,
    additionalProperties: false,
    properties: ***REMOVED***
        comment: ***REMOVED***
            type: STRING
      ***REMOVED***,
        defaultvalue: ***REMOVED***},
        description: ***REMOVED***
            type: STRING_OPTIONAL
      ***REMOVED***,
        kind: ***REMOVED***
            type: STRING,
            enum: ['member']
      ***REMOVED***,
        longname: ***REMOVED***
            type: STRING
      ***REMOVED***,
        memberof: ***REMOVED***
            type: STRING
      ***REMOVED***,
        meta: META_SCHEMA,
        name: ***REMOVED***
            type: STRING
      ***REMOVED***,
        // is this member nullable? (derived from the type expression)
        nullable: ***REMOVED***
            type: BOOLEAN_OPTIONAL
      ***REMOVED***,
        // is this member optional? (derived from the type expression)
        optional: ***REMOVED***
            type: BOOLEAN_OPTIONAL
      ***REMOVED***,
        scope: ***REMOVED***
            type: STRING,
            enum: ['static']
      ***REMOVED***,
        type: TYPE_PROPERTY_SCHEMA,
        // can this member be provided more than once? (derived from the type expression)
        variable: ***REMOVED***
            type: BOOLEAN_OPTIONAL
      ***REMOVED***
  ***REMOVED***
};

// function parameter, or object property defined with @property tag
const PARAM_SCHEMA = exports.PARAM_SCHEMA = ***REMOVED***
    type: OBJECT,
    additionalProperties: false,
    properties: ***REMOVED***
        // what is the default value for this parameter?
        defaultvalue: ***REMOVED***},
        // a description of the parameter
        description: ***REMOVED***
            type: STRING_OPTIONAL
      ***REMOVED***,
        // what name does this parameter have within the function?
        name: ***REMOVED***
            type: STRING
      ***REMOVED***,
        // can the value for this parameter be null?
        nullable: ***REMOVED***
            type: BOOLEAN_OPTIONAL
      ***REMOVED***,
        // is a value for this parameter optional?
        optional: ***REMOVED***
            type: BOOLEAN_OPTIONAL
      ***REMOVED***,
        // what are the types of value expected for this parameter?
        type: TYPE_PROPERTY_SCHEMA,
        // can this parameter be repeated?
        variable: ***REMOVED***
            type: BOOLEAN_OPTIONAL
      ***REMOVED***
  ***REMOVED***
};

const DOCLET_SCHEMA = exports.DOCLET_SCHEMA = ***REMOVED***
    type: OBJECT,
    additionalProperties: false,
    properties: ***REMOVED***
        // what access privileges are allowed
        access: ***REMOVED***
            type: STRING,
            enum: [
                'package',
                'private',
                'protected',
                'public'
            ]
      ***REMOVED***,
        alias: ***REMOVED***
            type: STRING
      ***REMOVED***,
        async: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        augments: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        author: ***REMOVED***
            type: ARRAY,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        borrowed: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            items: ***REMOVED***
                type: OBJECT,
                additionalProperties: false,
                properties: ***REMOVED***
                    // name of the target
                    as: ***REMOVED***
                        type: STRING
                  ***REMOVED***,
                    // name of the source
                    from: ***REMOVED***
                        type: STRING
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***,
        // a description of the class that this constructor belongs to
        classdesc: ***REMOVED***
            type: STRING
      ***REMOVED***,
        comment: ***REMOVED***
            type: STRING
      ***REMOVED***,
        copyright: ***REMOVED***
            type: STRING
      ***REMOVED***,
        defaultvalue: ***REMOVED***},
        defaultvaluetype: ***REMOVED***
            type: STRING,
            enum: [OBJECT, ARRAY]
      ***REMOVED***,
        // is usage of this symbol deprecated?
        deprecated: ***REMOVED***
            type: [STRING, BOOLEAN]
      ***REMOVED***,
        // a description
        description: ***REMOVED***
            type: STRING_OPTIONAL
      ***REMOVED***,
        // something else to consider
        examples: ***REMOVED***
            type: ARRAY,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        exceptions: ***REMOVED***
            type: ARRAY,
            items: PARAM_SCHEMA
      ***REMOVED***,
        // the path to another constructor
        extends: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        // the path to another doc object
        fires: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            items: ***REMOVED***
                type: STRING,
                pattern: EVENT_REGEXP
          ***REMOVED***
      ***REMOVED***,
        forceMemberof: ***REMOVED***
            type: BOOLEAN_OPTIONAL
      ***REMOVED***,
        generator: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        hideconstructor: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        ignore: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        implementations: ***REMOVED***
            type: ARRAY,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        implements: ***REMOVED***
            type: ARRAY,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        inheritdoc: ***REMOVED***
            type: STRING
      ***REMOVED***,
        inherited: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        inherits: ***REMOVED***
            type: STRING,
            dependency: ***REMOVED***
                inherited: true
          ***REMOVED***
      ***REMOVED***,
        isEnum: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        // what kind of symbol is this?
        kind: ***REMOVED***
            type: STRING,
            enum: [
                'class',
                'constant',
                'event',
                'external',
                'file',
                'function',
                'interface',
                'member',
                'mixin',
                'module',
                'namespace',
                'package',
                'param',
                'typedef'
            ]
      ***REMOVED***,
        license: ***REMOVED***
            type: STRING
      ***REMOVED***,
        listens: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            items: ***REMOVED***
                type: STRING,
                pattern: EVENT_REGEXP
          ***REMOVED***
      ***REMOVED***,
        longname: ***REMOVED***
            type: STRING
      ***REMOVED***,
        // probably a leading substring of the path
        memberof: ***REMOVED***
            type: STRING
      ***REMOVED***,
        // information about this doc
        meta: META_SCHEMA,
        // was this doclet mixed in?
        mixed: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        mixes: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        modifies: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            items: PARAM_SCHEMA
      ***REMOVED***,
        // probably a trailing substring of the path
        name: ***REMOVED***
            type: STRING
      ***REMOVED***,
        // is this member nullable? (derived from the type expression)
        nullable: ***REMOVED***
            type: BOOLEAN_OPTIONAL
      ***REMOVED***,
        // is this member optional? (derived from the type expression)
        optional: ***REMOVED***
            type: BOOLEAN_OPTIONAL
      ***REMOVED***,
        // does this member explicitly override the parent?
        override: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        overrides: ***REMOVED***
            type: STRING
      ***REMOVED***,
        // are there function parameters associated with this doc?
        params: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            items: PARAM_SCHEMA
      ***REMOVED***,
        preserveName: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        properties: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            minItems: 1,
            items: ***REMOVED***
                anyOf: [ENUM_PROPERTY_SCHEMA, PARAM_SCHEMA]
          ***REMOVED***
      ***REMOVED***,
        readonly: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        // the symbol being documented requires another symbol
        requires: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            minItems: 1,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        returns: ***REMOVED***
            type: ARRAY,
            minItems: 1,
            items: PARAM_SCHEMA
      ***REMOVED***,
        // what sort of parent scope does this symbol have?
        scope: ***REMOVED***
            type: STRING,
            enum: [
                'global',
                'inner',
                'instance',
                'static'
            ]
      ***REMOVED***,
        // something else to consider
        see: ***REMOVED***
            type: ARRAY,
            minItems: 1,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        // at what previous version was this doc added?
        since: ***REMOVED***
            type: STRING
      ***REMOVED***,
        summary: ***REMOVED***
            type: STRING
      ***REMOVED***,
        // arbitrary tags associated with this doc
        tags: ***REMOVED***
            type: ARRAY,
            minItems: 1,
            items: ***REMOVED***
                type: OBJECT,
                additionalProperties: false,
                properties: ***REMOVED***
                    originalTitle: ***REMOVED***
                        type: STRING
                  ***REMOVED***,
                    text: ***REMOVED***
                        type: STRING
                  ***REMOVED***,
                    title: ***REMOVED***
                        type: STRING
                  ***REMOVED***,
                    value: ***REMOVED***
                        oneOf: [STRING_SCHEMA, PARAM_SCHEMA]
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***,
        'this': ***REMOVED***
            type: STRING
      ***REMOVED***,
        todo: ***REMOVED***
            type: ARRAY,
            minItems: 1,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        // extended tutorials
        tutorials: ***REMOVED***
            type: ARRAY,
            minItems: 1,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        // what type is the value that this doc is associated with, like `number`
        type: TYPE_PROPERTY_SCHEMA,
        undocumented: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        // can this member be provided more than once? (derived from the type expression)
        variable: ***REMOVED***
            type: BOOLEAN_OPTIONAL
      ***REMOVED***,
        variation: ***REMOVED***
            type: STRING
      ***REMOVED***,
        // what is the version of this doc
        version: ***REMOVED***
            type: STRING
      ***REMOVED***,
        // is a member left to be implemented during inheritance?
        virtual: ***REMOVED***
            type: BOOLEAN
      ***REMOVED***,
        yields: ***REMOVED***
            type: ARRAY,
            minItems: 1,
            items: PARAM_SCHEMA
      ***REMOVED***
  ***REMOVED***
};

const CONTACT_INFO_SCHEMA = exports.CONTACT_INFO_SCHEMA = ***REMOVED***
    type: OBJECT,
    additionalProperties: false,
    properties: ***REMOVED***
        email: ***REMOVED***
            type: STRING
      ***REMOVED***,
        name: ***REMOVED***
            type: STRING
      ***REMOVED***,
        url: ***REMOVED***
            type: STRING,
            format: 'uri'
      ***REMOVED***
  ***REMOVED***
};

const BUGS_SCHEMA = exports.BUGS_SCHEMA = ***REMOVED***
    type: OBJECT,
    additionalProperties: false,
    properties: ***REMOVED***
        email: ***REMOVED***
            type: STRING
      ***REMOVED***,
        url: ***REMOVED***
            type: STRING,
            format: 'uri'
      ***REMOVED***
  ***REMOVED***
};

const PACKAGE_SCHEMA = exports.PACKAGE_SCHEMA = ***REMOVED***
    type: OBJECT,
    additionalProperties: false,
    properties: ***REMOVED***
        author: ***REMOVED***
            anyOf: [STRING_SCHEMA, CONTACT_INFO_SCHEMA]
      ***REMOVED***,
        bugs: ***REMOVED***
            anyOf: [STRING_SCHEMA, BUGS_SCHEMA]
      ***REMOVED***,
        contributors: ***REMOVED***
            type: ARRAY,
            minItems: 0,
            items: ***REMOVED***
                anyOf: [STRING_SCHEMA, CONTACT_INFO_SCHEMA]
          ***REMOVED***
      ***REMOVED***,
        dependencies: ***REMOVED***
            type: OBJECT
      ***REMOVED***,
        description: ***REMOVED***
            type: STRING
      ***REMOVED***,
        devDependencies: ***REMOVED***
            type: OBJECT
      ***REMOVED***,
        engines: ***REMOVED***
            type: OBJECT
      ***REMOVED***,
        files: ***REMOVED***
            type: ARRAY,
            uniqueItems: true,
            minItems: 0,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        homepage: ***REMOVED***
            type: STRING,
            format: 'uri'
      ***REMOVED***,
        keywords: ***REMOVED***
            type: ARRAY,
            minItems: 0,
            items: ***REMOVED***
                type: STRING
          ***REMOVED***
      ***REMOVED***,
        kind: ***REMOVED***
            type: STRING,
            enum: ['package']
      ***REMOVED***,
        licenses: ***REMOVED***
            type: ARRAY,
            minItems: 1,
            items: ***REMOVED***
                type: OBJECT,
                additionalProperties: false,
                properties: ***REMOVED***
                    type: ***REMOVED***
                        type: STRING
                  ***REMOVED***,
                    url: ***REMOVED***
                        type: STRING,
                        format: 'uri'
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***,
        longname: ***REMOVED***
            type: STRING,
            pattern: PACKAGE_REGEXP
      ***REMOVED***,
        main: ***REMOVED***
            type: STRING
      ***REMOVED***,
        name: ***REMOVED***
            type: STRING
      ***REMOVED***,
        repository: ***REMOVED***
            type: OBJECT,
            additionalProperties: false,
            properties: ***REMOVED***
                type: ***REMOVED***
                    type: STRING
              ***REMOVED***,
                // we don't use `format: 'uri'` here because repo URLs are atypical
                url: ***REMOVED***
                    type: STRING
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***,
        version: ***REMOVED***
            type: STRING
      ***REMOVED***
  ***REMOVED***
};

exports.DOCLETS_SCHEMA = ***REMOVED***
    type: ARRAY,
    items: ***REMOVED***
        anyOf: [DOCLET_SCHEMA, PACKAGE_SCHEMA]
  ***REMOVED***
};
