// jsdoc/lib/jsdoc/tag.js
/**
 * Functionality related to JSDoc tags.
 * @module jsdoc/tag
 * @requires module:jsdoc/env
 * @requires module:jsdoc/path
 * @requires module:jsdoc/tag/dictionary
 * @requires module:jsdoc/tag/validator
 * @requires module:jsdoc/tag/type
 * @requires module:jsdoc/util/logger
 * @requires module:util
 */
const jsdoc = ***REMOVED***
    env: require('jsdoc/env'),
    tag: ***REMOVED***
        dictionary: require('jsdoc/tag/dictionary'),
        validator: require('jsdoc/tag/validator'),
        type: require('jsdoc/tag/type')
  ***REMOVED***,
    util: ***REMOVED***
        logger: require('jsdoc/util/logger')
  ***REMOVED***
};
const path = require('jsdoc/path');
const util = require('util');

// Check whether the text is the same as a symbol name with leading or trailing whitespace. If so,
// the whitespace must be preserved, and the text cannot be trimmed.
function mustPreserveWhitespace(text, meta) ***REMOVED***
    return meta && meta.code && meta.code.name === text && text.match(/(?:^\s+)|(?:\s+$)/);
}

function trim(text, opts, meta) ***REMOVED***
    let indentMatcher;
    let match;

    opts = opts || ***REMOVED***};
    text = String(typeof text === 'undefined' ? '' : text);

    if ( mustPreserveWhitespace(text, meta) ) ***REMOVED***
        text = util.format('"%s"', text);
  ***REMOVED***
    else if (opts.keepsWhitespace) ***REMOVED***
        text = text.replace(/^[\n\r\f]+|[\n\r\f]+$/g, '');
        if (opts.removesIndent) ***REMOVED***
            match = text.match(/^([ \t]+)/);
            if (match && match[1]) ***REMOVED***
                indentMatcher = new RegExp(`^$***REMOVED***match[1]}`, 'gm');
                text = text.replace(indentMatcher, '');
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    else ***REMOVED***
        text = text.replace(/^\s+|\s+$/g, '');
  ***REMOVED***

    return text;
}

function addHiddenProperty(obj, propName, propValue) ***REMOVED***
    Object.defineProperty(obj, propName, ***REMOVED***
        value: propValue,
        writable: true,
        enumerable: Boolean(jsdoc.env.opts.debug),
        configurable: true
  ***REMOVED***);
}

function parseType(***REMOVED***text, originalTitle}, ***REMOVED***canHaveName, canHaveType}, meta) ***REMOVED***
    try ***REMOVED***
        return jsdoc.tag.type.parse(text, canHaveName, canHaveType);
  ***REMOVED***
    catch (e) ***REMOVED***
        jsdoc.util.logger.error(
            'Unable to parse a tag\'s type expression%s with tag title "%s" and text "%s": %s',
            meta.filename ? ( ` for source file $***REMOVED***path.join(meta.path, meta.filename)}$***REMOVED***meta.lineno ? (` in line $***REMOVED***meta.lineno}`) : ''}` ) : '',
            originalTitle,
            text,
            e.message
        );

        return ***REMOVED***};
  ***REMOVED***
}

function processTagText(tag, tagDef, meta) ***REMOVED***
    let tagType;

    if (tagDef.onTagText) ***REMOVED***
        tag.text = tagDef.onTagText(tag.text);
  ***REMOVED***

    if (tagDef.canHaveType || tagDef.canHaveName) ***REMOVED***
        /** The value property represents the result of parsing the tag text. */
        tag.value = ***REMOVED***};

        tagType = parseType(tag, tagDef, meta);

        // It is possible for a tag to *not* have a type but still have
        // optional or defaultvalue, e.g. '@param [foo]'.
        // Although tagType.type.length == 0 we should still copy the other properties.
        if (tagType.type) ***REMOVED***
            if (tagType.type.length) ***REMOVED***
                tag.value.type = ***REMOVED***
                    names: tagType.type
              ***REMOVED***;
                addHiddenProperty(tag.value.type, 'parsedType', tagType.parsedType);
          ***REMOVED***

            ['optional', 'nullable', 'variable', 'defaultvalue'].forEach(prop => ***REMOVED***
                if (typeof tagType[prop] !== 'undefined') ***REMOVED***
                    tag.value[prop] = tagType[prop];
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***

        if (tagType.text && tagType.text.length) ***REMOVED***
            tag.value.description = tagType.text;
      ***REMOVED***

        if (tagDef.canHaveName) ***REMOVED***
            // note the dash is a special case: as a param name it means "no name"
            if (tagType.name && tagType.name !== '-') ***REMOVED*** tag.value.name = tagType.name; }
      ***REMOVED***
  ***REMOVED***
    else ***REMOVED***
        tag.value = tag.text;
  ***REMOVED***
}

/**
 * Replace the existing tag dictionary with a new tag dictionary.
 *
 * Used for testing only. Do not call this method directly. Instead, call
 * ***REMOVED***@link module:jsdoc/doclet._replaceDictionary}, which also updates this module's tag dictionary.
 *
 * @private
 * @param ***REMOVED***module:jsdoc/tag/dictionary.Dictionary} dict - The new tag dictionary.
 */
exports._replaceDictionary = function _replaceDictionary(dict) ***REMOVED***
    jsdoc.tag.dictionary = dict;
};

/**
 * Represents a single doclet tag.
 */
class Tag ***REMOVED***
    /**
     * Constructs a new tag object. Calls the tag validator.
     *
     * @param ***REMOVED***string} tagTitle
     * @param ***REMOVED***string=} tagBody
     * @param ***REMOVED***object=} meta
     */
    constructor(tagTitle, tagBody, meta) ***REMOVED***
        let tagDef;
        let trimOpts;

        meta = meta || ***REMOVED***};

        this.originalTitle = trim(tagTitle);

        /** The title of the tag (for example, `title` in `@title text`). */
        this.title = jsdoc.tag.dictionary.normalise(this.originalTitle);

        tagDef = jsdoc.tag.dictionary.lookUp(this.title);
        trimOpts = ***REMOVED***
            keepsWhitespace: tagDef.keepsWhitespace,
            removesIndent: tagDef.removesIndent
      ***REMOVED***;

        /**
         * The text following the tag (for example, `text` in `@title text`).
         *
         * Whitespace is trimmed from the tag text as follows:
         *
         * + If the tag's `keepsWhitespace` option is falsy, all leading and trailing whitespace are
         * removed.
         * + If the tag's `keepsWhitespace` option is set to `true`, leading and trailing whitespace are
         * not trimmed, unless the `removesIndent` option is also enabled.
         * + If the tag's `removesIndent` option is set to `true`, any indentation that is shared by
         * every line in the string is removed. This option is ignored unless `keepsWhitespace` is set
         * to `true`.
         *
         * **Note**: If the tag text is the name of a symbol, and the symbol's name includes leading or
         * trailing whitespace (for example, the property names in `***REMOVED*** ' ': true, ' foo ': false }`),
         * the tag text is not trimmed. Instead, the tag text is wrapped in double quotes to prevent the
         * whitespace from being trimmed.
         */
        this.text = trim(tagBody, trimOpts, meta);

        if (this.text) ***REMOVED***
            processTagText(this, tagDef, meta);
      ***REMOVED***

        jsdoc.tag.validator.validate(this, tagDef, meta);
  ***REMOVED***
}
exports.Tag = Tag;
