// jsdoc/lib/jsdoc/template.js
/**
 * Wrapper for underscore's template utility to allow loading templates from files.
 * @module jsdoc/template
 */
const _ = require('underscore');
const fs = require('jsdoc/fs');
const path = require('path');

/**
 * Underscore template helper.
 */
class Template ***REMOVED***
    /**
     * @param ***REMOVED***string} filepath - Templates directory.
     */
    constructor(filepath) ***REMOVED***
        this.path = filepath;
        this.layout = null;
        this.cache = ***REMOVED***};
        // override default template tag settings
        this.settings = ***REMOVED***
            evaluate: /<\?js([\s\S]+?)\?>/g,
            interpolate: /<\?js=([\s\S]+?)\?>/g,
            escape: /<\?js~([\s\S]+?)\?>/g
      ***REMOVED***;
  ***REMOVED***

    /**
     * Loads template from given file.
     * @param ***REMOVED***string} file - Template filename.
     * @return ***REMOVED***function} Returns template closure.
     */
    load(file) ***REMOVED***
        return _.template(fs.readFileSync(file, 'utf8'), null, this.settings);
  ***REMOVED***

    /**
     * Renders template using given data.
     *
     * This is low-level function, for rendering full templates use ***REMOVED***@link Template.render()}.
     *
     * @param ***REMOVED***string} file - Template filename.
     * @param ***REMOVED***object} data - Template variables (doesn't have to be object, but passing variables dictionary is best way and most common use).
     * @return ***REMOVED***string} Rendered template.
     */
    partial(file, data) ***REMOVED***
        file = path.resolve(this.path, file);

        // load template into cache
        if (!(file in this.cache)) ***REMOVED***
            this.cache[file] = this.load(file);
      ***REMOVED***

        // keep template helper context
        return this.cache[file].call(this, data);
  ***REMOVED***

    /**
     * Renders template with given data.
     *
     * This method automaticaly applies layout if set.
     *
     * @param ***REMOVED***string} file - Template filename.
     * @param ***REMOVED***object} data - Template variables (doesn't have to be object, but passing variables dictionary is best way and most common use).
     * @return ***REMOVED***string} Rendered template.
     */
    render(file, data) ***REMOVED***
        // main content
        let content = this.partial(file, data);

        // apply layout
        if (this.layout) ***REMOVED***
            data.content = content;
            content = this.partial(this.layout, data);
      ***REMOVED***

        return content;
  ***REMOVED***
}
exports.Template = Template;
