// @firebase/app-check-compat/dist/esm/index.esm.js
import firebase from '@firebase/app-compat';
import ***REMOVED*** Component } from '@firebase/component';
import ***REMOVED*** ReCaptchaV3Provider, ReCaptchaEnterpriseProvider, CustomProvider, initializeAppCheck, setTokenAutoRefreshEnabled, getToken, onTokenChanged } from '@firebase/app-check';
import ***REMOVED*** ErrorFactory } from '@firebase/util';

var name = "@firebase/app-check-compat";
var version = "0.3.1";

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var _a;
var ERRORS = (_a = ***REMOVED***},
    _a["use-before-activation" /* AppCheckError.USE_BEFORE_ACTIVATION */] = 'App Check is being used before activate() is called for FirebaseApp ***REMOVED***$appName}. ' +
        'Call activate() before instantiating other Firebase services.',
    _a);
var ERROR_FACTORY = new ErrorFactory('appCheck', 'AppCheck', ERRORS);

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var AppCheckService = /** @class */ (function () ***REMOVED***
    function AppCheckService(app) ***REMOVED***
        this.app = app;
  ***REMOVED***
    AppCheckService.prototype.activate = function (siteKeyOrProvider, isTokenAutoRefreshEnabled) ***REMOVED***
        var provider;
        if (typeof siteKeyOrProvider === 'string') ***REMOVED***
            provider = new ReCaptchaV3Provider(siteKeyOrProvider);
      ***REMOVED***
        else if (siteKeyOrProvider instanceof ReCaptchaEnterpriseProvider ||
            siteKeyOrProvider instanceof ReCaptchaV3Provider ||
            siteKeyOrProvider instanceof CustomProvider) ***REMOVED***
            provider = siteKeyOrProvider;
      ***REMOVED***
        else ***REMOVED***
            provider = new CustomProvider(***REMOVED*** getToken: siteKeyOrProvider.getToken });
      ***REMOVED***
        this._delegate = initializeAppCheck(this.app, ***REMOVED***
            provider: provider,
            isTokenAutoRefreshEnabled: isTokenAutoRefreshEnabled
      ***REMOVED***);
  ***REMOVED***;
    AppCheckService.prototype.setTokenAutoRefreshEnabled = function (isTokenAutoRefreshEnabled) ***REMOVED***
        if (!this._delegate) ***REMOVED***
            throw ERROR_FACTORY.create("use-before-activation" /* AppCheckError.USE_BEFORE_ACTIVATION */, ***REMOVED***
                appName: this.app.name
          ***REMOVED***);
      ***REMOVED***
        setTokenAutoRefreshEnabled(this._delegate, isTokenAutoRefreshEnabled);
  ***REMOVED***;
    AppCheckService.prototype.getToken = function (forceRefresh) ***REMOVED***
        if (!this._delegate) ***REMOVED***
            throw ERROR_FACTORY.create("use-before-activation" /* AppCheckError.USE_BEFORE_ACTIVATION */, ***REMOVED***
                appName: this.app.name
          ***REMOVED***);
      ***REMOVED***
        return getToken(this._delegate, forceRefresh);
  ***REMOVED***;
    AppCheckService.prototype.onTokenChanged = function (onNextOrObserver, onError, onCompletion) ***REMOVED***
        if (!this._delegate) ***REMOVED***
            throw ERROR_FACTORY.create("use-before-activation" /* AppCheckError.USE_BEFORE_ACTIVATION */, ***REMOVED***
                appName: this.app.name
          ***REMOVED***);
      ***REMOVED***
        return onTokenChanged(this._delegate, 
        /**
         * Exp onTokenChanged() will handle both overloads but we need
         * to specify one to not confuse Typescript.
         */
        onNextOrObserver, onError, onCompletion);
  ***REMOVED***;
    return AppCheckService;
}());

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var factory = function (container) ***REMOVED***
    // Dependencies
    var app = container.getProvider('app-compat').getImmediate();
    return new AppCheckService(app);
};
function registerAppCheck() ***REMOVED***
    firebase.INTERNAL.registerComponent(new Component('appCheck-compat', factory, "PUBLIC" /* ComponentType.PUBLIC */).setServiceProps(***REMOVED***
        ReCaptchaEnterpriseProvider: ReCaptchaEnterpriseProvider,
        ReCaptchaV3Provider: ReCaptchaV3Provider,
        CustomProvider: CustomProvider
  ***REMOVED***));
}
registerAppCheck();
firebase.registerVersion(name, version);

export ***REMOVED*** registerAppCheck };
