// @grpc/grpc-js/src/channel-credentials.ts
/*
 * Copyright 2019 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import ***REMOVED*** ConnectionOptions, createSecureContext, PeerCertificate, SecureContext } from 'tls';

import ***REMOVED*** CallCredentials } from './call-credentials';
import ***REMOVED*** CIPHER_SUITES, getDefaultRootsData } from './tls-helpers';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function verifyIsBufferOrNull(obj: any, friendlyName: string): void ***REMOVED***
  if (obj && !(obj instanceof Buffer)) ***REMOVED***
    throw new TypeError(`$***REMOVED***friendlyName}, if provided, must be a Buffer.`);
***REMOVED***
}

/**
 * A callback that will receive the expected hostname and presented peer
 * certificate as parameters. The callback should return an error to
 * indicate that the presented certificate is considered invalid and
 * otherwise returned undefined.
 */
export type CheckServerIdentityCallback = (
  hostname: string,
  cert: PeerCertificate
) => Error | undefined;

function bufferOrNullEqual(buf1: Buffer | null, buf2: Buffer | null) ***REMOVED***
  if (buf1 === null && buf2 === null) ***REMOVED***
    return true;
***REMOVED*** else ***REMOVED***
    return buf1 !== null && buf2 !== null && buf1.equals(buf2);
***REMOVED***
}

/**
 * Additional peer verification options that can be set when creating
 * SSL credentials.
 */
export interface VerifyOptions ***REMOVED***
  /**
   * If set, this callback will be invoked after the usual hostname verification
   * has been performed on the peer certificate.
   */
  checkServerIdentity?: CheckServerIdentityCallback;
}

/**
 * A class that contains credentials for communicating over a channel, as well
 * as a set of per-call credentials, which are applied to every method call made
 * over a channel initialized with an instance of this class.
 */
export abstract class ChannelCredentials ***REMOVED***
  protected callCredentials: CallCredentials;

  protected constructor(callCredentials?: CallCredentials) ***REMOVED***
    this.callCredentials = callCredentials || CallCredentials.createEmpty();
***REMOVED***
  /**
   * Returns a copy of this object with the included set of per-call credentials
   * expanded to include callCredentials.
   * @param callCredentials A CallCredentials object to associate with this
   * instance.
   */
  abstract compose(callCredentials: CallCredentials): ChannelCredentials;

  /**
   * Gets the set of per-call credentials associated with this instance.
   */
  _getCallCredentials(): CallCredentials ***REMOVED***
    return this.callCredentials;
***REMOVED***

  /**
   * Gets a SecureContext object generated from input parameters if this
   * instance was created with createSsl, or null if this instance was created
   * with createInsecure.
   */
  abstract _getConnectionOptions(): ConnectionOptions | null;

  /**
   * Indicates whether this credentials object creates a secure channel.
   */
  abstract _isSecure(): boolean;

  /**
   * Check whether two channel credentials objects are equal. Two secure
   * credentials are equal if they were constructed with the same parameters.
   * @param other The other ChannelCredentials Object
   */
  abstract _equals(other: ChannelCredentials): boolean;

  /**
   * Return a new ChannelCredentials instance with a given set of credentials.
   * The resulting instance can be used to construct a Channel that communicates
   * over TLS.
   * @param rootCerts The root certificate data.
   * @param privateKey The client certificate private key, if available.
   * @param certChain The client certificate key chain, if available.
   * @param verifyOptions Additional options to modify certificate verification
   */
  static createSsl(
    rootCerts?: Buffer | null,
    privateKey?: Buffer | null,
    certChain?: Buffer | null,
    verifyOptions?: VerifyOptions
  ): ChannelCredentials ***REMOVED***
    verifyIsBufferOrNull(rootCerts, 'Root certificate');
    verifyIsBufferOrNull(privateKey, 'Private key');
    verifyIsBufferOrNull(certChain, 'Certificate chain');
    if (privateKey && !certChain) ***REMOVED***
      throw new Error(
        'Private key must be given with accompanying certificate chain'
      );
  ***REMOVED***
    if (!privateKey && certChain) ***REMOVED***
      throw new Error(
        'Certificate chain must be given with accompanying private key'
      );
  ***REMOVED***
    const secureContext = createSecureContext(***REMOVED***
      ca: rootCerts ?? getDefaultRootsData() ?? undefined,
      key: privateKey ?? undefined,
      cert: certChain ?? undefined,
      ciphers: CIPHER_SUITES,
  ***REMOVED***);
    return new SecureChannelCredentialsImpl(
      secureContext,
      verifyOptions ?? ***REMOVED***}
    );
***REMOVED***

  /**
   * Return a new ChannelCredentials instance with credentials created using
   * the provided secureContext. The resulting instances can be used to
   * construct a Channel that communicates over TLS. gRPC will not override
   * anything in the provided secureContext, so the environment variables
   * GRPC_SSL_CIPHER_SUITES and GRPC_DEFAULT_SSL_ROOTS_FILE_PATH will
   * not be applied.
   * @param secureContext The return value of tls.createSecureContext()
   * @param verifyOptions Additional options to modify certificate verification
   */
  static createFromSecureContext(secureContext: SecureContext, verifyOptions?: VerifyOptions): ChannelCredentials ***REMOVED***
    return new SecureChannelCredentialsImpl(
      secureContext,
      verifyOptions ?? ***REMOVED***}
    )
***REMOVED***

  /**
   * Return a new ChannelCredentials instance with no credentials.
   */
  static createInsecure(): ChannelCredentials ***REMOVED***
    return new InsecureChannelCredentialsImpl();
***REMOVED***
}

class InsecureChannelCredentialsImpl extends ChannelCredentials ***REMOVED***
  constructor(callCredentials?: CallCredentials) ***REMOVED***
    super(callCredentials);
***REMOVED***

  compose(callCredentials: CallCredentials): never ***REMOVED***
    throw new Error('Cannot compose insecure credentials');
***REMOVED***

  _getConnectionOptions(): ConnectionOptions | null ***REMOVED***
    return null;
***REMOVED***
  _isSecure(): boolean ***REMOVED***
    return false;
***REMOVED***
  _equals(other: ChannelCredentials): boolean ***REMOVED***
    return other instanceof InsecureChannelCredentialsImpl;
***REMOVED***
}

class SecureChannelCredentialsImpl extends ChannelCredentials ***REMOVED***
  connectionOptions: ConnectionOptions;

  constructor(
    private secureContext: SecureContext,
    private verifyOptions: VerifyOptions
  ) ***REMOVED***
    super();
    this.connectionOptions = ***REMOVED*** 
      secureContext
  ***REMOVED***;
    // Node asserts that this option is a function, so we cannot pass undefined
    if (verifyOptions?.checkServerIdentity) ***REMOVED***
      this.connectionOptions.checkServerIdentity = verifyOptions.checkServerIdentity;
  ***REMOVED***
***REMOVED***

  compose(callCredentials: CallCredentials): ChannelCredentials ***REMOVED***
    const combinedCallCredentials = this.callCredentials.compose(
      callCredentials
    );
    return new ComposedChannelCredentialsImpl(this, combinedCallCredentials);
***REMOVED***

  _getConnectionOptions(): ConnectionOptions | null ***REMOVED***
    // Copy to prevent callers from mutating this.connectionOptions
    return ***REMOVED*** ...this.connectionOptions };
***REMOVED***
  _isSecure(): boolean ***REMOVED***
    return true;
***REMOVED***
  _equals(other: ChannelCredentials): boolean ***REMOVED***
    if (this === other) ***REMOVED***
      return true;
  ***REMOVED***
    if (other instanceof SecureChannelCredentialsImpl) ***REMOVED***
      return (
        this.secureContext === other.secureContext && 
        this.verifyOptions.checkServerIdentity === other.verifyOptions.checkServerIdentity
        );
  ***REMOVED*** else ***REMOVED***
      return false;
  ***REMOVED***
***REMOVED***
}

class ComposedChannelCredentialsImpl extends ChannelCredentials ***REMOVED***
  constructor(
    private channelCredentials: SecureChannelCredentialsImpl,
    callCreds: CallCredentials
  ) ***REMOVED***
    super(callCreds);
***REMOVED***
  compose(callCredentials: CallCredentials) ***REMOVED***
    const combinedCallCredentials = this.callCredentials.compose(
      callCredentials
    );
    return new ComposedChannelCredentialsImpl(
      this.channelCredentials,
      combinedCallCredentials
    );
***REMOVED***

  _getConnectionOptions(): ConnectionOptions | null ***REMOVED***
    return this.channelCredentials._getConnectionOptions();
***REMOVED***
  _isSecure(): boolean ***REMOVED***
    return true;
***REMOVED***
  _equals(other: ChannelCredentials): boolean ***REMOVED***
    if (this === other) ***REMOVED***
      return true;
  ***REMOVED***
    if (other instanceof ComposedChannelCredentialsImpl) ***REMOVED***
      return (
        this.channelCredentials._equals(other.channelCredentials) &&
        this.callCredentials._equals(other.callCredentials)
      );
  ***REMOVED*** else ***REMOVED***
      return false;
  ***REMOVED***
***REMOVED***
}
