// @grpc/grpc-js/src/index.ts
/*
 * Copyright 2019 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import ***REMOVED***
  ClientDuplexStream,
  ClientReadableStream,
  ClientUnaryCall,
  ClientWritableStream,
  ServiceError,
} from './call';
import ***REMOVED*** CallCredentials, OAuth2Client } from './call-credentials';
import ***REMOVED*** Deadline, StatusObject } from './call-stream';
import ***REMOVED*** Channel, ChannelImplementation } from './channel';
import ***REMOVED*** CompressionAlgorithms } from './compression-algorithms';
import ***REMOVED*** ConnectivityState } from './connectivity-state';
import ***REMOVED*** ChannelCredentials } from './channel-credentials';
import ***REMOVED***
  CallOptions,
  Client,
  ClientOptions,
  CallInvocationTransformer,
  CallProperties,
  UnaryCallback,
} from './client';
import ***REMOVED*** LogVerbosity, Status, Propagate } from './constants';
import * as logging from './logging';
import ***REMOVED***
  Deserialize,
  loadPackageDefinition,
  makeClientConstructor,
  MethodDefinition,
  ProtobufTypeDefinition,
  Serialize,
  ServiceClientConstructor,
  ServiceDefinition,
} from './make-client';
import ***REMOVED*** Metadata, MetadataOptions, MetadataValue } from './metadata';
import ***REMOVED***
  Server,
  UntypedHandleCall,
  UntypedServiceImplementation,
} from './server';
import ***REMOVED*** KeyCertPair, ServerCredentials } from './server-credentials';
import ***REMOVED*** StatusBuilder } from './status-builder';
import ***REMOVED***
  handleBidiStreamingCall,
  handleServerStreamingCall,
  handleClientStreamingCall,
  handleUnaryCall,
  sendUnaryData,
  ServerUnaryCall,
  ServerReadableStream,
  ServerWritableStream,
  ServerDuplexStream,
  ServerErrorResponse,
} from './server-call';

export ***REMOVED*** OAuth2Client };

/**** Client Credentials ****/

// Using assign only copies enumerable properties, which is what we want
export const credentials = ***REMOVED***
  /**
   * Combine a ChannelCredentials with any number of CallCredentials into a
   * single ChannelCredentials object.
   * @param channelCredentials The ChannelCredentials object.
   * @param callCredentials Any number of CallCredentials objects.
   * @return The resulting ChannelCredentials object.
   */
  combineChannelCredentials: (
    channelCredentials: ChannelCredentials,
    ...callCredentials: CallCredentials[]
  ): ChannelCredentials => ***REMOVED***
    return callCredentials.reduce(
      (acc, other) => acc.compose(other),
      channelCredentials
    );
***REMOVED***,

  /**
   * Combine any number of CallCredentials into a single CallCredentials
   * object.
   * @param first The first CallCredentials object.
   * @param additional Any number of additional CallCredentials objects.
   * @return The resulting CallCredentials object.
   */
  combineCallCredentials: (
    first: CallCredentials,
    ...additional: CallCredentials[]
  ): CallCredentials => ***REMOVED***
    return additional.reduce((acc, other) => acc.compose(other), first);
***REMOVED***,

  // from channel-credentials.ts
  createInsecure: ChannelCredentials.createInsecure,
  createSsl: ChannelCredentials.createSsl,
  createFromSecureContext: ChannelCredentials.createFromSecureContext,

  // from call-credentials.ts
  createFromMetadataGenerator: CallCredentials.createFromMetadataGenerator,
  createFromGoogleCredential: CallCredentials.createFromGoogleCredential,
  createEmpty: CallCredentials.createEmpty,
};

/**** Metadata ****/

export ***REMOVED*** Metadata, MetadataOptions, MetadataValue };

/**** Constants ****/

export ***REMOVED***
  LogVerbosity as logVerbosity,
  Status as status,
  ConnectivityState as connectivityState,
  Propagate as propagate,
  CompressionAlgorithms as compressionAlgorithms
  // TODO: Other constants as well
};

/**** Client ****/

export ***REMOVED***
  Client,
  ClientOptions,
  loadPackageDefinition,
  makeClientConstructor,
  makeClientConstructor as makeGenericClientConstructor,
  CallProperties,
  CallInvocationTransformer,
  ChannelImplementation as Channel,
  Channel as ChannelInterface,
  UnaryCallback as requestCallback,
};

/**
 * Close a Client object.
 * @param client The client to close.
 */
export const closeClient = (client: Client) => client.close();

export const waitForClientReady = (
  client: Client,
  deadline: Date | number,
  callback: (error?: Error) => void
) => client.waitForReady(deadline, callback);

/* Interfaces */

export ***REMOVED***
  sendUnaryData,
  ChannelCredentials,
  CallCredentials,
  Deadline,
  Serialize as serialize,
  Deserialize as deserialize,
  ClientUnaryCall,
  ClientReadableStream,
  ClientWritableStream,
  ClientDuplexStream,
  CallOptions,
  MethodDefinition,
  StatusObject,
  ServiceError,
  ServerUnaryCall,
  ServerReadableStream,
  ServerWritableStream,
  ServerDuplexStream,
  ServerErrorResponse,
  ServiceDefinition,
  UntypedHandleCall,
  UntypedServiceImplementation,
};

/**** Server ****/

export ***REMOVED***
  handleBidiStreamingCall,
  handleServerStreamingCall,
  handleUnaryCall,
  handleClientStreamingCall,
};

/* eslint-disable @typescript-eslint/no-explicit-any */
export type Call =
  | ClientUnaryCall
  | ClientReadableStream<any>
  | ClientWritableStream<any>
  | ClientDuplexStream<any, any>;
/* eslint-enable @typescript-eslint/no-explicit-any */

/**** Unimplemented function stubs ****/

/* eslint-disable @typescript-eslint/no-explicit-any */

export const loadObject = (value: any, options: any): never => ***REMOVED***
  throw new Error(
    'Not available in this library. Use @grpc/proto-loader and loadPackageDefinition instead'
  );
};

export const load = (filename: any, format: any, options: any): never => ***REMOVED***
  throw new Error(
    'Not available in this library. Use @grpc/proto-loader and loadPackageDefinition instead'
  );
};

export const setLogger = (logger: Partial<Console>): void => ***REMOVED***
  logging.setLogger(logger);
};

export const setLogVerbosity = (verbosity: LogVerbosity): void => ***REMOVED***
  logging.setLoggerVerbosity(verbosity);
};

export ***REMOVED*** Server };
export ***REMOVED*** ServerCredentials };
export ***REMOVED*** KeyCertPair };

export const getClientChannel = (client: Client) => ***REMOVED***
  return Client.prototype.getChannel.call(client);
};

export ***REMOVED*** StatusBuilder };

export ***REMOVED*** Listener } from './call-stream';

export ***REMOVED***
  Requester,
  ListenerBuilder,
  RequesterBuilder,
  Interceptor,
  InterceptorOptions,
  InterceptorProvider,
  InterceptingCall,
  InterceptorConfigurationError,
} from './client-interceptors';

export ***REMOVED***
  GrpcObject,
  ServiceClientConstructor,
  ProtobufTypeDefinition
} from './make-client';

export ***REMOVED*** ChannelOptions } from './channel-options';

export ***REMOVED***
  getChannelzServiceDefinition,
  getChannelzHandlers
} from './channelz';

export ***REMOVED*** addAdminServicesToServer } from './admin';

import * as experimental from './experimental';
export ***REMOVED*** experimental };

import * as resolver_dns from './resolver-dns';
import * as resolver_uds from './resolver-uds';
import * as resolver_ip from './resolver-ip';
import * as load_balancer_pick_first from './load-balancer-pick-first';
import * as load_balancer_round_robin from './load-balancer-round-robin';
import * as load_balancer_outlier_detection from './load-balancer-outlier-detection';
import * as channelz from './channelz';

const clientVersion = require('../../package.json').version;

(() => ***REMOVED***
  logging.trace(LogVerbosity.DEBUG, 'index', 'Loading @grpc/grpc-js version ' + clientVersion);
  resolver_dns.setup();
  resolver_uds.setup();
  resolver_ip.setup();
  load_balancer_pick_first.setup();
  load_balancer_round_robin.setup();
  load_balancer_outlier_detection.setup();
  channelz.setup();
})();
