// @grpc/grpc-js/src/resolver.ts
/*
 * Copyright 2019 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import ***REMOVED*** MethodConfig, ServiceConfig } from './service-config';
import ***REMOVED*** StatusObject } from './call-stream';
import ***REMOVED*** SubchannelAddress } from './subchannel-address';
import ***REMOVED*** GrpcUri, uriToString } from './uri-parser';
import ***REMOVED*** ChannelOptions } from './channel-options';
import ***REMOVED*** Metadata } from './metadata';
import ***REMOVED*** Status } from './constants';
import ***REMOVED*** Filter, FilterFactory } from './filter';

export interface CallConfig ***REMOVED***
  methodConfig: MethodConfig;
  onCommitted?: () => void;
  pickInformation: ***REMOVED*** [key: string]: string };
  status: Status;
  dynamicFilterFactories: FilterFactory<Filter>[];
}

/**
 * Selects a configuration for a method given the name and metadata. Defined in
 * https://github.com/grpc/proposal/blob/master/A31-xds-timeout-support-and-config-selector.md#new-functionality-in-grpc
 */
export interface ConfigSelector ***REMOVED***
  (methodName: string, metadata: Metadata): CallConfig;
}

/**
 * A listener object passed to the resolver's constructor that provides name
 * resolution updates back to the resolver's owner.
 */
export interface ResolverListener ***REMOVED***
  /**
   * Called whenever the resolver has new name resolution results to report
   * @param addressList The new list of backend addresses
   * @param serviceConfig The new service configuration corresponding to the
   *     `addressList`. Will be `null` if no service configuration was
   *     retrieved or if the service configuration was invalid
   * @param serviceConfigError If non-`null`, indicates that the retrieved
   *     service configuration was invalid
   */
  onSuccessfulResolution(
    addressList: SubchannelAddress[],
    serviceConfig: ServiceConfig | null,
    serviceConfigError: StatusObject | null,
    configSelector: ConfigSelector | null,
    attributes: ***REMOVED*** [key: string]: unknown }
  ): void;
  /**
   * Called whenever a name resolution attempt fails.
   * @param error Describes how resolution failed
   */
  onError(error: StatusObject): void;
}

/**
 * A resolver class that handles one or more of the name syntax schemes defined
 * in the [gRPC Name Resolution document](https://github.com/grpc/grpc/blob/master/doc/naming.md)
 */
export interface Resolver ***REMOVED***
  /**
   * Indicates that the caller wants new name resolution data. Calling this
   * function may eventually result in calling one of the `ResolverListener`
   * functions, but that is not guaranteed. Those functions will never be
   * called synchronously with the constructor or updateResolution.
   */
  updateResolution(): void;

  /**
   * Destroy the resolver. Should be called when the owning channel shuts down.
   */
  destroy(): void;
}

export interface ResolverConstructor ***REMOVED***
  new (
    target: GrpcUri,
    listener: ResolverListener,
    channelOptions: ChannelOptions
  ): Resolver;
  /**
   * Get the default authority for a target. This loosely corresponds to that
   * target's hostname. Throws an error if this resolver class cannot parse the
   * `target`.
   * @param target
   */
  getDefaultAuthority(target: GrpcUri): string;
}

const registeredResolvers: ***REMOVED*** [scheme: string]: ResolverConstructor } = ***REMOVED***};
let defaultScheme: string | null = null;

/**
 * Register a resolver class to handle target names prefixed with the `prefix`
 * string. This prefix should correspond to a URI scheme name listed in the
 * [gRPC Name Resolution document](https://github.com/grpc/grpc/blob/master/doc/naming.md)
 * @param prefix
 * @param resolverClass
 */
export function registerResolver(
  scheme: string,
  resolverClass: ResolverConstructor
) ***REMOVED***
  registeredResolvers[scheme] = resolverClass;
}

/**
 * Register a default resolver to handle target names that do not start with
 * any registered prefix.
 * @param resolverClass
 */
export function registerDefaultScheme(scheme: string) ***REMOVED***
  defaultScheme = scheme;
}

/**
 * Create a name resolver for the specified target, if possible. Throws an
 * error if no such name resolver can be created.
 * @param target
 * @param listener
 */
export function createResolver(
  target: GrpcUri,
  listener: ResolverListener,
  options: ChannelOptions
): Resolver ***REMOVED***
  if (target.scheme !== undefined && target.scheme in registeredResolvers) ***REMOVED***
    return new registeredResolvers[target.scheme](target, listener, options);
***REMOVED*** else ***REMOVED***
    throw new Error(
      `No resolver could be created for target $***REMOVED***uriToString(target)}`
    );
***REMOVED***
}

/**
 * Get the default authority for the specified target, if possible. Throws an
 * error if no registered name resolver can parse that target string.
 * @param target
 */
export function getDefaultAuthority(target: GrpcUri): string ***REMOVED***
  if (target.scheme !== undefined && target.scheme in registeredResolvers) ***REMOVED***
    return registeredResolvers[target.scheme].getDefaultAuthority(target);
***REMOVED*** else ***REMOVED***
    throw new Error(`Invalid target $***REMOVED***uriToString(target)}`);
***REMOVED***
}

export function mapUriDefaultScheme(target: GrpcUri): GrpcUri | null ***REMOVED***
  if (target.scheme === undefined || !(target.scheme in registeredResolvers)) ***REMOVED***
    if (defaultScheme !== null) ***REMOVED***
      return ***REMOVED***
        scheme: defaultScheme,
        authority: undefined,
        path: uriToString(target),
    ***REMOVED***;
  ***REMOVED*** else ***REMOVED***
      return null;
  ***REMOVED***
***REMOVED***
  return target;
}
