// uglify-js/lib/mozilla-ast.js
/***********************************************************************

  A JavaScript tokenizer / parser / beautifier / compressor.
  https://github.com/mishoo/UglifyJS

  -------------------------------- (C) ---------------------------------

                           Author: Mihai Bazon
                         <mihai.bazon@gmail.com>
                       http://mihai.bazon.net/blog

  Distributed under the BSD license:

    Copyright 2012 (c) Mihai Bazon <mihai.bazon@gmail.com>

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

        * Redistributions of source code must retain the above
          copyright notice, this list of conditions and the following
          disclaimer.

        * Redistributions in binary form must reproduce the above
          copyright notice, this list of conditions and the following
          disclaimer in the documentation and/or other materials
          provided with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER “AS IS” AND ANY
    EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
    PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE
    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
    PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
    TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
    THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
    SUCH DAMAGE.

 ***********************************************************************/

"use strict";

(function() ***REMOVED***
    var MOZ_TO_ME = ***REMOVED***
        Program: function(M) ***REMOVED***
            return new AST_Toplevel(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                body: normalize_directives(M.body.map(from_moz)),
          ***REMOVED***);
      ***REMOVED***,
        ArrowFunctionExpression: function(M) ***REMOVED***
            var argnames = [], rest = null;
            M.params.forEach(function(param) ***REMOVED***
                if (param.type == "RestElement") ***REMOVED***
                    rest = from_moz(param.argument);
              ***REMOVED*** else ***REMOVED***
                    argnames.push(from_moz(param));
              ***REMOVED***
          ***REMOVED***);
            var fn = new (M.async ? AST_AsyncArrow : AST_Arrow)(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                argnames: argnames,
                rest: rest,
          ***REMOVED***);
            var node = from_moz(M.body);
            if (node instanceof AST_BlockStatement) ***REMOVED***
                fn.body = normalize_directives(node.body);
                fn.value = null;
          ***REMOVED*** else ***REMOVED***
                fn.body = [];
                fn.value = node;
          ***REMOVED***
            return fn;
      ***REMOVED***,
        FunctionDeclaration: function(M) ***REMOVED***
            var ctor;
            if (M.async) ***REMOVED***
                ctor = M.generator ? AST_AsyncGeneratorDefun : AST_AsyncDefun;
          ***REMOVED*** else ***REMOVED***
                ctor = M.generator ? AST_GeneratorDefun : AST_Defun;
          ***REMOVED***
            var argnames = [], rest = null;
            M.params.forEach(function(param) ***REMOVED***
                if (param.type == "RestElement") ***REMOVED***
                    rest = from_moz(param.argument);
              ***REMOVED*** else ***REMOVED***
                    argnames.push(from_moz(param));
              ***REMOVED***
          ***REMOVED***);
            return new ctor(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                name: from_moz(M.id),
                argnames: argnames,
                rest: rest,
                body: normalize_directives(from_moz(M.body).body),
          ***REMOVED***);
      ***REMOVED***,
        FunctionExpression: function(M) ***REMOVED***
            var ctor;
            if (M.async) ***REMOVED***
                ctor = M.generator ? AST_AsyncGeneratorFunction : AST_AsyncFunction;
          ***REMOVED*** else ***REMOVED***
                ctor = M.generator ? AST_GeneratorFunction : AST_Function;
          ***REMOVED***
            var argnames = [], rest = null;
            M.params.forEach(function(param) ***REMOVED***
                if (param.type == "RestElement") ***REMOVED***
                    rest = from_moz(param.argument);
              ***REMOVED*** else ***REMOVED***
                    argnames.push(from_moz(param));
              ***REMOVED***
          ***REMOVED***);
            return new ctor(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                name: from_moz(M.id),
                argnames: argnames,
                rest: rest,
                body: normalize_directives(from_moz(M.body).body),
          ***REMOVED***);
      ***REMOVED***,
        ClassDeclaration: function(M) ***REMOVED***
            return new AST_DefClass(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                name: from_moz(M.id),
                extends: from_moz(M.superClass),
                properties: M.body.body.map(from_moz),
          ***REMOVED***);
      ***REMOVED***,
        ClassExpression: function(M) ***REMOVED***
            return new AST_ClassExpression(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                name: from_moz(M.id),
                extends: from_moz(M.superClass),
                properties: M.body.body.map(from_moz),
          ***REMOVED***);
      ***REMOVED***,
        MethodDefinition: function(M) ***REMOVED***
            var key = M.key, internal = false;
            if (M.computed) ***REMOVED***
                key = from_moz(key);
          ***REMOVED*** else if (key.type == "PrivateIdentifier") ***REMOVED***
                internal = true;
                key = "#" + key.name;
          ***REMOVED*** else ***REMOVED***
                key = read_name(key);
          ***REMOVED***
            var ctor = AST_ClassMethod, value = from_moz(M.value);
            switch (M.kind) ***REMOVED***
              case "get":
                ctor = AST_ClassGetter;
                value = new AST_Accessor(value);
                break;
              case "set":
                ctor = AST_ClassSetter;
                value = new AST_Accessor(value);
                break;
          ***REMOVED***
            return new ctor(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                key: key,
                private: internal,
                static: M.static,
                value: value,
          ***REMOVED***);
      ***REMOVED***,
        PropertyDefinition: function(M) ***REMOVED***
            var key = M.key, internal = false;
            if (M.computed) ***REMOVED***
                key = from_moz(key);
          ***REMOVED*** else if (key.type == "PrivateIdentifier") ***REMOVED***
                internal = true;
                key = "#" + key.name;
          ***REMOVED*** else ***REMOVED***
                key = read_name(key);
          ***REMOVED***
            return new AST_ClassField(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                key: key,
                private: internal,
                static: M.static,
                value: from_moz(M.value),
          ***REMOVED***);
      ***REMOVED***,
        StaticBlock: function(M) ***REMOVED***
            var start = my_start_token(M);
            var end = my_end_token(M);
            return new AST_ClassInit(***REMOVED***
                start: start,
                end: end,
                value: new AST_ClassInitBlock(***REMOVED***
                    start: start,
                    end: end,
                    body: normalize_directives(M.body.map(from_moz)),
              ***REMOVED***),
          ***REMOVED***);
      ***REMOVED***,
        ForOfStatement: function(M) ***REMOVED***
            return new (M.await ? AST_ForAwaitOf : AST_ForOf)(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                init: from_moz(M.left),
                object: from_moz(M.right),
                body: from_moz(M.body),
          ***REMOVED***);
      ***REMOVED***,
        TryStatement: function(M) ***REMOVED***
            var handlers = M.handlers || [M.handler];
            if (handlers.length > 1 || M.guardedHandlers && M.guardedHandlers.length) ***REMOVED***
                throw new Error("Multiple catch clauses are not supported.");
          ***REMOVED***
            return new AST_Try(***REMOVED***
                start    : my_start_token(M),
                end      : my_end_token(M),
                body     : from_moz(M.block).body,
                bcatch   : from_moz(handlers[0]),
                bfinally : M.finalizer ? new AST_Finally(from_moz(M.finalizer)) : null,
          ***REMOVED***);
      ***REMOVED***,
        Property: function(M) ***REMOVED***
            var key = M.computed ? from_moz(M.key) : read_name(M.key);
            var args = ***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                key: key,
                value: from_moz(M.value),
          ***REMOVED***;
            if (M.kind == "init") return new (M.method ? AST_ObjectMethod : AST_ObjectKeyVal)(args);
            args.value = new AST_Accessor(args.value);
            if (M.kind == "get") return new AST_ObjectGetter(args);
            if (M.kind == "set") return new AST_ObjectSetter(args);
      ***REMOVED***,
        ArrayExpression: function(M) ***REMOVED***
            return new AST_Array(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                elements: M.elements.map(function(elem) ***REMOVED***
                    return elem === null ? new AST_Hole() : from_moz(elem);
              ***REMOVED***),
          ***REMOVED***);
      ***REMOVED***,
        ArrayPattern: function(M) ***REMOVED***
            var elements = [], rest = null;
            M.elements.forEach(function(el) ***REMOVED***
                if (el === null) ***REMOVED***
                    elements.push(new AST_Hole());
              ***REMOVED*** else if (el.type == "RestElement") ***REMOVED***
                    rest = from_moz(el.argument);
              ***REMOVED*** else ***REMOVED***
                    elements.push(from_moz(el));
              ***REMOVED***
          ***REMOVED***);
            return new AST_DestructuredArray(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                elements: elements,
                rest: rest,
          ***REMOVED***);
      ***REMOVED***,
        ObjectPattern: function(M) ***REMOVED***
            var props = [], rest = null;
            M.properties.forEach(function(prop) ***REMOVED***
                if (prop.type == "RestElement") ***REMOVED***
                    rest = from_moz(prop.argument);
              ***REMOVED*** else ***REMOVED***
                    props.push(new AST_DestructuredKeyVal(from_moz(prop)));
              ***REMOVED***
          ***REMOVED***);
            return new AST_DestructuredObject(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                properties: props,
                rest: rest,
          ***REMOVED***);
      ***REMOVED***,
        MemberExpression: function(M) ***REMOVED***
            return new (M.computed ? AST_Sub : AST_Dot)(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                optional: M.optional,
                expression: from_moz(M.object),
                property: M.computed ? from_moz(M.property) : M.property.name,
          ***REMOVED***);
      ***REMOVED***,
        MetaProperty: function(M) ***REMOVED***
            var expr = from_moz(M.meta);
            var prop = read_name(M.property);
            if (expr.name == "new" && prop == "target") return new AST_NewTarget(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                name: "new.target",
          ***REMOVED***);
            return new AST_Dot(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                expression: expr,
                property: prop,
          ***REMOVED***);
      ***REMOVED***,
        SwitchCase: function(M) ***REMOVED***
            return new (M.test ? AST_Case : AST_Default)(***REMOVED***
                start      : my_start_token(M),
                end        : my_end_token(M),
                expression : from_moz(M.test),
                body       : M.consequent.map(from_moz),
          ***REMOVED***);
      ***REMOVED***,
        ExportAllDeclaration: function(M) ***REMOVED***
            var start = my_start_token(M);
            var end = my_end_token(M);
            return new AST_ExportForeign(***REMOVED***
                start: start,
                end: end,
                aliases: [ M.exported ? from_moz_alias(M.exported) : new AST_String(***REMOVED***
                    start: start,
                    value: "*",
                    end: end,
              ***REMOVED***) ],
                keys: [ new AST_String(***REMOVED***
                    start: start,
                    value: "*",
                    end: end,
              ***REMOVED***) ],
                path: from_moz(M.source),
          ***REMOVED***);
      ***REMOVED***,
        ExportDefaultDeclaration: function(M) ***REMOVED***
            var decl = from_moz(M.declaration);
            if (!decl.name) switch (decl.CTOR) ***REMOVED***
              case AST_AsyncDefun:
                decl = new AST_AsyncFunction(decl);
                break;
              case AST_AsyncGeneratorDefun:
                decl = new AST_AsyncGeneratorFunction(decl);
                break;
              case AST_DefClass:
                decl = new AST_ClassExpression(decl);
                break;
              case AST_Defun:
                decl = new AST_Function(decl);
                break;
              case AST_GeneratorDefun:
                decl = new AST_GeneratorFunction(decl);
                break;
          ***REMOVED***
            return new AST_ExportDefault(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                body: decl,
          ***REMOVED***);
      ***REMOVED***,
        ExportNamedDeclaration: function(M) ***REMOVED***
            if (M.declaration) return new AST_ExportDeclaration(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                body: from_moz(M.declaration),
          ***REMOVED***);
            if (M.source) ***REMOVED***
                var aliases = [], keys = [];
                M.specifiers.forEach(function(prop) ***REMOVED***
                    aliases.push(from_moz_alias(prop.exported));
                    keys.push(from_moz_alias(prop.local));
              ***REMOVED***);
                return new AST_ExportForeign(***REMOVED***
                    start: my_start_token(M),
                    end: my_end_token(M),
                    aliases: aliases,
                    keys: keys,
                    path: from_moz(M.source),
              ***REMOVED***);
          ***REMOVED***
            return new AST_ExportReferences(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                properties: M.specifiers.map(function(prop) ***REMOVED***
                    var sym = new AST_SymbolExport(from_moz(prop.local));
                    sym.alias = from_moz_alias(prop.exported);
                    return sym;
              ***REMOVED***),
          ***REMOVED***);
      ***REMOVED***,
        ImportDeclaration: function(M) ***REMOVED***
            var start = my_start_token(M);
            var end = my_end_token(M);
            var all = null, def = null, props = null;
            M.specifiers.forEach(function(prop) ***REMOVED***
                var sym = new AST_SymbolImport(from_moz(prop.local));
                switch (prop.type) ***REMOVED***
                  case "ImportDefaultSpecifier":
                    def = sym;
                    def.key = new AST_String(***REMOVED***
                        start: start,
                        value: "",
                        end: end,
                  ***REMOVED***);
                    break;
                  case "ImportNamespaceSpecifier":
                    all = sym;
                    all.key = new AST_String(***REMOVED***
                        start: start,
                        value: "*",
                        end: end,
                  ***REMOVED***);
                    break;
                  default:
                    sym.key = from_moz_alias(prop.imported);
                    if (!props) props = [];
                    props.push(sym);
                    break;
              ***REMOVED***
          ***REMOVED***);
            return new AST_Import(***REMOVED***
                start: start,
                end: end,
                all: all,
                default: def,
                properties: props,
                path: from_moz(M.source),
          ***REMOVED***);
      ***REMOVED***,
        ImportExpression: function(M) ***REMOVED***
            var start = my_start_token(M);
            var arg = from_moz(M.source);
            return new AST_Call(***REMOVED***
                start: start,
                end: my_end_token(M),
                expression: new AST_SymbolRef(***REMOVED***
                    start: start,
                    end: arg.start,
                    name: "import",
              ***REMOVED***),
                args: [ arg ],
          ***REMOVED***);
      ***REMOVED***,
        VariableDeclaration: function(M) ***REMOVED***
            return new (***REMOVED***
                const: AST_Const,
                let: AST_Let,
          ***REMOVED***[M.kind] || AST_Var)(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                definitions: M.declarations.map(from_moz),
          ***REMOVED***);
      ***REMOVED***,
        Literal: function(M) ***REMOVED***
            var args = ***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
          ***REMOVED***;
            if (M.bigint) ***REMOVED***
                args.value = M.bigint.toLowerCase() + "n";
                return new AST_BigInt(args);
          ***REMOVED***
            var val = M.value;
            if (val === null) return new AST_Null(args);
            var rx = M.regex;
            if (rx && rx.pattern) ***REMOVED***
                // RegExpLiteral as per ESTree AST spec
                args.value = new RegExp(rx.pattern, rx.flags);
                args.value.raw_source = rx.pattern;
                return new AST_RegExp(args);
          ***REMOVED*** else if (rx) ***REMOVED***
                // support legacy RegExp
                args.value = M.regex && M.raw ? M.raw : val;
                return new AST_RegExp(args);
          ***REMOVED***
            switch (typeof val) ***REMOVED***
              case "string":
                args.value = val;
                return new AST_String(args);
              case "number":
                if (isNaN(val)) return new AST_NaN(args);
                var negate, node;
                if (isFinite(val)) ***REMOVED***
                    negate = 1 / val < 0;
                    args.value = negate ? -val : val;
                    node = new AST_Number(args);
              ***REMOVED*** else ***REMOVED***
                    negate = val < 0;
                    node = new AST_Infinity(args);
              ***REMOVED***
                return negate ? new AST_UnaryPrefix(***REMOVED***
                    start: args.start,
                    end: args.end,
                    operator: "-",
                    expression: node,
              ***REMOVED***) : node;
              case "boolean":
                return new (val ? AST_True : AST_False)(args);
          ***REMOVED***
      ***REMOVED***,
        TemplateLiteral: function(M) ***REMOVED***
            return new AST_Template(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                expressions: M.expressions.map(from_moz),
                strings: M.quasis.map(function(el) ***REMOVED***
                    return el.value.raw;
              ***REMOVED***),
          ***REMOVED***);
      ***REMOVED***,
        TaggedTemplateExpression: function(M) ***REMOVED***
            var tmpl = from_moz(M.quasi);
            tmpl.start = my_start_token(M);
            tmpl.end = my_end_token(M);
            tmpl.tag = from_moz(M.tag);
            return tmpl;
      ***REMOVED***,
        Identifier: function(M) ***REMOVED***
            var p, level = FROM_MOZ_STACK.length - 1;
            do ***REMOVED***
                p = FROM_MOZ_STACK[--level];
          ***REMOVED*** while (p.type == "ArrayPattern"
                || p.type == "AssignmentPattern" && p.left === FROM_MOZ_STACK[level + 1]
                || p.type == "ObjectPattern"
                || p.type == "Property" && p.value === FROM_MOZ_STACK[level + 1]
                || p.type == "VariableDeclarator" && p.id === FROM_MOZ_STACK[level + 1]);
            var ctor = AST_SymbolRef;
            switch (p.type) ***REMOVED***
              case "ArrowFunctionExpression":
                if (p.body !== FROM_MOZ_STACK[level + 1]) ctor = AST_SymbolFunarg;
                break;
              case "BreakStatement":
              case "ContinueStatement":
                ctor = AST_LabelRef;
                break;
              case "CatchClause":
                ctor = AST_SymbolCatch;
                break;
              case "ClassDeclaration":
                if (p.id === FROM_MOZ_STACK[level + 1]) ctor = AST_SymbolDefClass;
                break;
              case "ClassExpression":
                if (p.id === FROM_MOZ_STACK[level + 1]) ctor = AST_SymbolClass;
                break;
              case "FunctionDeclaration":
                ctor = p.id === FROM_MOZ_STACK[level + 1] ? AST_SymbolDefun : AST_SymbolFunarg;
                break;
              case "FunctionExpression":
                ctor = p.id === FROM_MOZ_STACK[level + 1] ? AST_SymbolLambda : AST_SymbolFunarg;
                break;
              case "LabeledStatement":
                ctor = AST_Label;
                break;
              case "VariableDeclaration":
                ctor = ***REMOVED***
                    const: AST_SymbolConst,
                    let: AST_SymbolLet,
              ***REMOVED***[p.kind] || AST_SymbolVar;
                break;
          ***REMOVED***
            return new ctor(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                name: M.name,
          ***REMOVED***);
      ***REMOVED***,
        Super: function(M) ***REMOVED***
            return new AST_Super(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                name: "super",
          ***REMOVED***);
      ***REMOVED***,
        ThisExpression: function(M) ***REMOVED***
            return new AST_This(***REMOVED***
                start: my_start_token(M),
                end: my_end_token(M),
                name: "this",
          ***REMOVED***);
      ***REMOVED***,
        ParenthesizedExpression: function(M) ***REMOVED***
            var node = from_moz(M.expression);
            if (!node.start.parens) node.start.parens = [];
            node.start.parens.push(my_start_token(M));
            if (!node.end.parens) node.end.parens = [];
            node.end.parens.push(my_end_token(M));
            return node;
      ***REMOVED***,
        ChainExpression: function(M) ***REMOVED***
            var node = from_moz(M.expression);
            node.terminal = true;
            return node;
      ***REMOVED***,
  ***REMOVED***;

    MOZ_TO_ME.UpdateExpression =
    MOZ_TO_ME.UnaryExpression = function To_Moz_Unary(M) ***REMOVED***
        var prefix = "prefix" in M ? M.prefix
            : M.type == "UnaryExpression" ? true : false;
        return new (prefix ? AST_UnaryPrefix : AST_UnaryPostfix)(***REMOVED***
            start      : my_start_token(M),
            end        : my_end_token(M),
            operator   : M.operator,
            expression : from_moz(M.argument)
      ***REMOVED***);
  ***REMOVED***;

    map("EmptyStatement", AST_EmptyStatement);
    map("ExpressionStatement", AST_SimpleStatement, "expression>body");
    map("BlockStatement", AST_BlockStatement, "body@body");
    map("IfStatement", AST_If, "test>condition, consequent>body, alternate>alternative");
    map("LabeledStatement", AST_LabeledStatement, "label>label, body>body");
    map("BreakStatement", AST_Break, "label>label");
    map("ContinueStatement", AST_Continue, "label>label");
    map("WithStatement", AST_With, "object>expression, body>body");
    map("SwitchStatement", AST_Switch, "discriminant>expression, cases@body");
    map("ReturnStatement", AST_Return, "argument>value");
    map("ThrowStatement", AST_Throw, "argument>value");
    map("WhileStatement", AST_While, "test>condition, body>body");
    map("DoWhileStatement", AST_Do, "test>condition, body>body");
    map("ForStatement", AST_For, "init>init, test>condition, update>step, body>body");
    map("ForInStatement", AST_ForIn, "left>init, right>object, body>body");
    map("DebuggerStatement", AST_Debugger);
    map("VariableDeclarator", AST_VarDef, "id>name, init>value");
    map("CatchClause", AST_Catch, "param>argname, body%body");

    map("BinaryExpression", AST_Binary, "operator=operator, left>left, right>right");
    map("LogicalExpression", AST_Binary, "operator=operator, left>left, right>right");
    map("AssignmentExpression", AST_Assign, "operator=operator, left>left, right>right");
    map("AssignmentPattern", AST_DefaultValue, "left>name, right>value");
    map("ConditionalExpression", AST_Conditional, "test>condition, consequent>consequent, alternate>alternative");
    map("NewExpression", AST_New, "callee>expression, arguments@args, pure=pure");
    map("CallExpression", AST_Call, "callee>expression, arguments@args, optional=optional, pure=pure");
    map("SequenceExpression", AST_Sequence, "expressions@expressions");
    map("SpreadElement", AST_Spread, "argument>expression");
    map("ObjectExpression", AST_Object, "properties@properties");
    map("AwaitExpression", AST_Await, "argument>expression");
    map("YieldExpression", AST_Yield, "argument>expression, delegate=nested");

    def_to_moz(AST_Toplevel, function To_Moz_Program(M) ***REMOVED***
        return to_moz_scope("Program", M);
  ***REMOVED***);

    def_to_moz(AST_LambdaDefinition, function To_Moz_FunctionDeclaration(M) ***REMOVED***
        var params = M.argnames.map(to_moz);
        if (M.rest) params.push(***REMOVED***
            type: "RestElement",
            argument: to_moz(M.rest),
      ***REMOVED***);
        return ***REMOVED***
            type: "FunctionDeclaration",
            id: to_moz(M.name),
            async: is_async(M),
            generator: is_generator(M),
            params: params,
            body: to_moz_scope("BlockStatement", M),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_Lambda, function To_Moz_FunctionExpression(M) ***REMOVED***
        var params = M.argnames.map(to_moz);
        if (M.rest) params.push(***REMOVED***
            type: "RestElement",
            argument: to_moz(M.rest),
      ***REMOVED***);
        if (is_arrow(M)) return ***REMOVED***
            type: "ArrowFunctionExpression",
            async: is_async(M),
            params: params,
            body: M.value ? to_moz(M.value) : to_moz_scope("BlockStatement", M),
      ***REMOVED***;
        return ***REMOVED***
            type: "FunctionExpression",
            id: to_moz(M.name),
            async: is_async(M),
            generator: is_generator(M),
            params: params,
            body: to_moz_scope("BlockStatement", M),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_DefClass, function To_Moz_ClassDeclaration(M) ***REMOVED***
        return ***REMOVED***
            type: "ClassDeclaration",
            id: to_moz(M.name),
            superClass: to_moz(M.extends),
            body: ***REMOVED***
                type: "ClassBody",
                body: M.properties.map(to_moz),
          ***REMOVED***,
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_ClassExpression, function To_Moz_ClassExpression(M) ***REMOVED***
        return ***REMOVED***
            type: "ClassExpression",
            id: to_moz(M.name),
            superClass: to_moz(M.extends),
            body: ***REMOVED***
                type: "ClassBody",
                body: M.properties.map(to_moz),
          ***REMOVED***,
      ***REMOVED***;
  ***REMOVED***);

    function To_Moz_MethodDefinition(kind) ***REMOVED***
        return function(M) ***REMOVED***
            var computed = M.key instanceof AST_Node;
            var key = computed ? to_moz(M.key) : M.private ? ***REMOVED***
                type: "PrivateIdentifier",
                name: M.key.slice(1),
          ***REMOVED*** : ***REMOVED***
                type: "Literal",
                value: M.key,
          ***REMOVED***;
            return ***REMOVED***
                type: "MethodDefinition",
                kind: kind,
                computed: computed,
                key: key,
                static: M.static,
                value: to_moz(M.value),
          ***REMOVED***;
      ***REMOVED***;
  ***REMOVED***
    def_to_moz(AST_ClassGetter, To_Moz_MethodDefinition("get"));
    def_to_moz(AST_ClassSetter, To_Moz_MethodDefinition("set"));
    def_to_moz(AST_ClassMethod, To_Moz_MethodDefinition("method"));

    def_to_moz(AST_ClassField, function To_Moz_PropertyDefinition(M) ***REMOVED***
        var computed = M.key instanceof AST_Node;
        var key = computed ? to_moz(M.key) : M.private ? ***REMOVED***
            type: "PrivateIdentifier",
            name: M.key.slice(1),
      ***REMOVED*** : ***REMOVED***
            type: "Literal",
            value: M.key,
      ***REMOVED***;
        return ***REMOVED***
            type: "PropertyDefinition",
            computed: computed,
            key: key,
            static: M.static,
            value: to_moz(M.value),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_ClassInit, function To_Moz_StaticBlock(M) ***REMOVED***
        return to_moz_scope("StaticBlock", M.value);
  ***REMOVED***);

    function To_Moz_ForOfStatement(is_await) ***REMOVED***
        return function(M) ***REMOVED***
            return ***REMOVED***
                type: "ForOfStatement",
                await: is_await,
                left: to_moz(M.init),
                right: to_moz(M.object),
                body: to_moz(M.body),
          ***REMOVED***;
      ***REMOVED***;
  ***REMOVED***
    def_to_moz(AST_ForAwaitOf, To_Moz_ForOfStatement(true));
    def_to_moz(AST_ForOf, To_Moz_ForOfStatement(false));

    def_to_moz(AST_Directive, function To_Moz_Directive(M) ***REMOVED***
        return ***REMOVED***
            type: "ExpressionStatement",
            expression: set_moz_loc(M, ***REMOVED***
                type: "Literal",
                value: M.value,
          ***REMOVED***),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_SwitchBranch, function To_Moz_SwitchCase(M) ***REMOVED***
        return ***REMOVED***
            type: "SwitchCase",
            test: to_moz(M.expression),
            consequent: M.body.map(to_moz),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_Try, function To_Moz_TryStatement(M) ***REMOVED***
        return ***REMOVED***
            type: "TryStatement",
            block: to_moz_block(M),
            handler: to_moz(M.bcatch),
            guardedHandlers: [],
            finalizer: to_moz(M.bfinally),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_Catch, function To_Moz_CatchClause(M) ***REMOVED***
        return ***REMOVED***
            type: "CatchClause",
            param: to_moz(M.argname),
            guard: null,
            body: to_moz_block(M),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_ExportDeclaration, function To_Moz_ExportNamedDeclaration_declaration(M) ***REMOVED***
        return ***REMOVED***
            type: "ExportNamedDeclaration",
            declaration: to_moz(M.body),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_ExportDefault, function To_Moz_ExportDefaultDeclaration(M) ***REMOVED***
        return ***REMOVED***
            type: "ExportDefaultDeclaration",
            declaration: to_moz(M.body),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_ExportForeign, function To_Moz_ExportAllDeclaration_ExportNamedDeclaration(M) ***REMOVED***
        if (M.keys[0].value == "*") return ***REMOVED***
            type: "ExportAllDeclaration",
            exported: M.aliases[0].value == "*" ? null : to_moz_alias(M.aliases[0]),
            source: to_moz(M.path),
      ***REMOVED***;
        var specifiers = [];
        for (var i = 0; i < M.aliases.length; i++) ***REMOVED***
            specifiers.push(set_moz_loc(***REMOVED***
                start: M.keys[i].start,
                end: M.aliases[i].end,
          ***REMOVED***, ***REMOVED***
                type: "ExportSpecifier",
                local: to_moz_alias(M.keys[i]),
                exported: to_moz_alias(M.aliases[i]),
          ***REMOVED***));
      ***REMOVED***
        return ***REMOVED***
            type: "ExportNamedDeclaration",
            specifiers: specifiers,
            source: to_moz(M.path),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_ExportReferences, function To_Moz_ExportNamedDeclaration_specifiers(M) ***REMOVED***
        return ***REMOVED***
            type: "ExportNamedDeclaration",
            specifiers: M.properties.map(function(prop) ***REMOVED***
                return set_moz_loc(***REMOVED***
                    start: prop.start,
                    end: prop.alias.end,
              ***REMOVED***, ***REMOVED***
                    type: "ExportSpecifier",
                    local: to_moz(prop),
                    exported: to_moz_alias(prop.alias),
              ***REMOVED***);
          ***REMOVED***),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_Import, function To_Moz_ImportDeclaration(M) ***REMOVED***
        var specifiers = M.properties ? M.properties.map(function(prop) ***REMOVED***
            return set_moz_loc(***REMOVED***
                start: prop.key.start,
                end: prop.end,
          ***REMOVED***, ***REMOVED***
                type: "ImportSpecifier",
                local: to_moz(prop),
                imported: to_moz_alias(prop.key),
          ***REMOVED***);
      ***REMOVED***) : [];
        if (M.all) specifiers.unshift(set_moz_loc(M.all, ***REMOVED***
            type: "ImportNamespaceSpecifier",
            local: to_moz(M.all),
      ***REMOVED***));
        if (M.default) specifiers.unshift(set_moz_loc(M.default, ***REMOVED***
            type: "ImportDefaultSpecifier",
            local: to_moz(M.default),
      ***REMOVED***));
        return ***REMOVED***
            type: "ImportDeclaration",
            specifiers: specifiers,
            source: to_moz(M.path),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_Definitions, function To_Moz_VariableDeclaration(M) ***REMOVED***
        return ***REMOVED***
            type: "VariableDeclaration",
            kind: M.TYPE.toLowerCase(),
            declarations: M.definitions.map(to_moz),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_PropAccess, function To_Moz_MemberExpression(M) ***REMOVED***
        var computed = M instanceof AST_Sub;
        var expr = ***REMOVED***
            type: "MemberExpression",
            object: to_moz(M.expression),
            computed: computed,
            optional: M.optional,
            property: computed ? to_moz(M.property) : ***REMOVED***
                type: "Identifier",
                name: M.property,
          ***REMOVED***,
      ***REMOVED***;
        return M.terminal ? ***REMOVED***
            type: "ChainExpression",
            expression: expr,
      ***REMOVED*** : expr;
  ***REMOVED***);

    def_to_moz(AST_Unary, function To_Moz_Unary(M) ***REMOVED***
        return ***REMOVED***
            type: M.operator == "++" || M.operator == "--" ? "UpdateExpression" : "UnaryExpression",
            operator: M.operator,
            prefix: M instanceof AST_UnaryPrefix,
            argument: to_moz(M.expression)
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_Binary, function To_Moz_BinaryExpression(M) ***REMOVED***
        return ***REMOVED***
            type: M.operator == "&&" || M.operator == "||" ? "LogicalExpression" : "BinaryExpression",
            left: to_moz(M.left),
            operator: M.operator,
            right: to_moz(M.right)
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_Array, function To_Moz_ArrayExpression(M) ***REMOVED***
        return ***REMOVED***
            type: "ArrayExpression",
            elements: M.elements.map(to_moz),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_DestructuredArray, function To_Moz_ArrayPattern(M) ***REMOVED***
        var elements = M.elements.map(to_moz);
        if (M.rest) elements.push(***REMOVED***
            type: "RestElement",
            argument: to_moz(M.rest),
      ***REMOVED***);
        return ***REMOVED***
            type: "ArrayPattern",
            elements: elements,
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_DestructuredKeyVal, function To_Moz_Property(M) ***REMOVED***
        var computed = M.key instanceof AST_Node;
        var key = computed ? to_moz(M.key) : ***REMOVED***
            type: "Literal",
            value: M.key,
      ***REMOVED***;
        return ***REMOVED***
            type: "Property",
            kind: "init",
            computed: computed,
            key: key,
            value: to_moz(M.value),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_DestructuredObject, function To_Moz_ObjectPattern(M) ***REMOVED***
        var props = M.properties.map(to_moz);
        if (M.rest) props.push(***REMOVED***
            type: "RestElement",
            argument: to_moz(M.rest),
      ***REMOVED***);
        return ***REMOVED***
            type: "ObjectPattern",
            properties: props,
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_ObjectProperty, function To_Moz_Property(M) ***REMOVED***
        var computed = M.key instanceof AST_Node;
        var key = computed ? to_moz(M.key) : ***REMOVED***
            type: "Literal",
            value: M.key,
      ***REMOVED***;
        var kind;
        if (M instanceof AST_ObjectKeyVal) ***REMOVED***
            kind = "init";
      ***REMOVED*** else if (M instanceof AST_ObjectGetter) ***REMOVED***
            kind = "get";
      ***REMOVED*** else if (M instanceof AST_ObjectSetter) ***REMOVED***
            kind = "set";
      ***REMOVED***
        return ***REMOVED***
            type: "Property",
            kind: kind,
            computed: computed,
            method: M instanceof AST_ObjectMethod,
            key: key,
            value: to_moz(M.value),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_Symbol, function To_Moz_Identifier(M) ***REMOVED***
        var def = M.definition();
        return ***REMOVED***
            type: "Identifier",
            name: def && def.mangled_name || M.name,
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_Super, function To_Moz_Super() ***REMOVED***
        return ***REMOVED*** type: "Super" };
  ***REMOVED***);

    def_to_moz(AST_This, function To_Moz_ThisExpression() ***REMOVED***
        return ***REMOVED*** type: "ThisExpression" };
  ***REMOVED***);

    def_to_moz(AST_NewTarget, function To_Moz_MetaProperty() ***REMOVED***
        return ***REMOVED***
            type: "MetaProperty",
            meta: ***REMOVED***
                type: "Identifier",
                name: "new",
          ***REMOVED***,
            property: ***REMOVED***
                type: "Identifier",
                name: "target",
          ***REMOVED***,
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_RegExp, function To_Moz_RegExpLiteral(M) ***REMOVED***
        var flags = M.value.toString().match(/\/([gimuy]*)$/)[1];
        var value = "/" + M.value.raw_source + "/" + flags;
        return ***REMOVED***
            type: "Literal",
            value: value,
            raw: value,
            regex: ***REMOVED***
                pattern: M.value.raw_source,
                flags: flags,
          ***REMOVED***,
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_BigInt, function To_Moz_BigInt(M) ***REMOVED***
        var value = M.value;
        return ***REMOVED***
            type: "Literal",
            bigint: value.slice(0, -1),
            raw: value,
      ***REMOVED***;
  ***REMOVED***);

    function To_Moz_Literal(M) ***REMOVED***
        var value = M.value;
        if (typeof value === "number" && (value < 0 || (value === 0 && 1 / value < 0))) ***REMOVED***
            return ***REMOVED***
                type: "UnaryExpression",
                operator: "-",
                prefix: true,
                argument: ***REMOVED***
                    type: "Literal",
                    value: -value,
                    raw: M.start.raw,
              ***REMOVED***,
          ***REMOVED***;
      ***REMOVED***
        return ***REMOVED***
            type: "Literal",
            value: value,
            raw: M.start.raw,
      ***REMOVED***;
  ***REMOVED***
    def_to_moz(AST_Boolean, To_Moz_Literal);
    def_to_moz(AST_Constant, To_Moz_Literal);
    def_to_moz(AST_Null, To_Moz_Literal);

    def_to_moz(AST_Atom, function To_Moz_Atom(M) ***REMOVED***
        return ***REMOVED***
            type: "Identifier",
            name: String(M.value),
      ***REMOVED***;
  ***REMOVED***);

    def_to_moz(AST_Template, function To_Moz_TemplateLiteral_TaggedTemplateExpression(M) ***REMOVED***
        var last = M.strings.length - 1;
        var tmpl = ***REMOVED***
            type: "TemplateLiteral",
            expressions: M.expressions.map(to_moz),
            quasis: M.strings.map(function(str, index) ***REMOVED***
                return ***REMOVED***
                    type: "TemplateElement",
                    tail: index == last,
                    value: ***REMOVED*** raw: str },
              ***REMOVED***;
          ***REMOVED***),
      ***REMOVED***;
        if (!M.tag) return tmpl;
        return ***REMOVED***
            type: "TaggedTemplateExpression",
            tag: to_moz(M.tag),
            quasi: tmpl,
      ***REMOVED***;
  ***REMOVED***);

    AST_Block.DEFMETHOD("to_mozilla_ast", AST_BlockStatement.prototype.to_mozilla_ast);
    AST_Hole.DEFMETHOD("to_mozilla_ast", return_null);
    AST_Node.DEFMETHOD("to_mozilla_ast", function() ***REMOVED***
        throw new Error("Cannot convert AST_" + this.TYPE);
  ***REMOVED***);

    /* -----[ tools ]----- */

    function normalize_directives(body) ***REMOVED***
        for (var i = 0; i < body.length; i++) ***REMOVED***
            var stat = body[i];
            if (!(stat instanceof AST_SimpleStatement)) break;
            var node = stat.body;
            if (!(node instanceof AST_String)) break;
            if (stat.start.pos !== node.start.pos) break;
            body[i] = new AST_Directive(node);
      ***REMOVED***
        return body;
  ***REMOVED***

    function raw_token(moznode) ***REMOVED***
        if (moznode.type == "Literal") ***REMOVED***
            return moznode.raw != null ? moznode.raw : moznode.value + "";
      ***REMOVED***
  ***REMOVED***

    function my_start_token(moznode) ***REMOVED***
        var loc = moznode.loc, start = loc && loc.start;
        var range = moznode.range;
        return new AST_Token(***REMOVED***
            file    : loc && loc.source,
            line    : start && start.line,
            col     : start && start.column,
            pos     : range ? range[0] : moznode.start,
            endline : start && start.line,
            endcol  : start && start.column,
            endpos  : range ? range[0] : moznode.start,
            raw     : raw_token(moznode),
      ***REMOVED***);
  ***REMOVED***

    function my_end_token(moznode) ***REMOVED***
        var loc = moznode.loc, end = loc && loc.end;
        var range = moznode.range;
        return new AST_Token(***REMOVED***
            file    : loc && loc.source,
            line    : end && end.line,
            col     : end && end.column,
            pos     : range ? range[1] : moznode.end,
            endline : end && end.line,
            endcol  : end && end.column,
            endpos  : range ? range[1] : moznode.end,
            raw     : raw_token(moznode),
      ***REMOVED***);
  ***REMOVED***

    function read_name(M) ***REMOVED***
        return "" + M[M.type == "Identifier" ? "name" : "value"];
  ***REMOVED***

    function map(moztype, mytype, propmap) ***REMOVED***
        var moz_to_me = [
            "start: my_start_token(M)",
            "end: my_end_token(M)",
        ];
        var me_to_moz = [
            "type: " + JSON.stringify(moztype),
        ];

        if (propmap) propmap.split(/\s*,\s*/).forEach(function(prop) ***REMOVED***
            var m = /([a-z0-9$_]+)(=|@|>|%)([a-z0-9$_]+)/i.exec(prop);
            if (!m) throw new Error("Can't understand property map: " + prop);
            var moz = m[1], how = m[2], my = m[3];
            switch (how) ***REMOVED***
              case "@":
                moz_to_me.push(my + ": M." + moz + ".map(from_moz)");
                me_to_moz.push(moz + ": M." +  my + ".map(to_moz)");
                break;
              case ">":
                moz_to_me.push(my + ": from_moz(M." + moz + ")");
                me_to_moz.push(moz + ": to_moz(M." + my + ")");
                break;
              case "=":
                moz_to_me.push(my + ": M." + moz);
                me_to_moz.push(moz + ": M." + my);
                break;
              case "%":
                moz_to_me.push(my + ": from_moz(M." + moz + ").body");
                me_to_moz.push(moz + ": to_moz_block(M)");
                break;
              default:
                throw new Error("Can't understand operator in propmap: " + prop);
          ***REMOVED***
      ***REMOVED***);

        MOZ_TO_ME[moztype] = new Function("U2", "my_start_token", "my_end_token", "from_moz", [
            "return function From_Moz_" + moztype + "(M) ***REMOVED***",
            "    return new U2.AST_" + mytype.TYPE + "(***REMOVED***",
            moz_to_me.join(",\n"),
            "  ***REMOVED***);",
            "};",
        ].join("\n"))(exports, my_start_token, my_end_token, from_moz);
        def_to_moz(mytype, new Function("to_moz", "to_moz_block", "to_moz_scope", [
            "return function To_Moz_" + moztype + "(M) ***REMOVED***",
            "    return ***REMOVED***",
            me_to_moz.join(",\n"),
            "  ***REMOVED***;",
            "};",
        ].join("\n"))(to_moz, to_moz_block, to_moz_scope));
  ***REMOVED***

    var FROM_MOZ_STACK = null;

    function from_moz(moz) ***REMOVED***
        FROM_MOZ_STACK.push(moz);
        var node = null;
        if (moz) ***REMOVED***
            if (!HOP(MOZ_TO_ME, moz.type)) throw new Error("Unsupported type: " + moz.type);
            node = MOZ_TO_ME[moz.type](moz);
      ***REMOVED***
        FROM_MOZ_STACK.pop();
        return node;
  ***REMOVED***

    function from_moz_alias(moz) ***REMOVED***
        return new AST_String(***REMOVED***
            start: my_start_token(moz),
            value: read_name(moz),
            end: my_end_token(moz),
      ***REMOVED***);
  ***REMOVED***

    AST_Node.from_mozilla_ast = function(node) ***REMOVED***
        var save_stack = FROM_MOZ_STACK;
        FROM_MOZ_STACK = [];
        var ast = from_moz(node);
        FROM_MOZ_STACK = save_stack;
        ast.walk(new TreeWalker(function(node) ***REMOVED***
            if (node instanceof AST_LabelRef) ***REMOVED***
                for (var level = 0, parent; parent = this.parent(level); level++) ***REMOVED***
                    if (parent instanceof AST_Scope) break;
                    if (parent instanceof AST_LabeledStatement && parent.label.name == node.name) ***REMOVED***
                        node.thedef = parent.label;
                        break;
                  ***REMOVED***
              ***REMOVED***
                if (!node.thedef) ***REMOVED***
                    var s = node.start;
                    js_error("Undefined label " + node.name, s.file, s.line, s.col, s.pos);
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***));
        return ast;
  ***REMOVED***;

    function set_moz_loc(mynode, moznode) ***REMOVED***
        var start = mynode.start;
        var end = mynode.end;
        if (start.pos != null && end.endpos != null) ***REMOVED***
            moznode.range = [start.pos, end.endpos];
      ***REMOVED***
        if (start.line) ***REMOVED***
            moznode.loc = ***REMOVED***
                start: ***REMOVED***line: start.line, column: start.col},
                end: end.endline ? ***REMOVED***line: end.endline, column: end.endcol} : null,
          ***REMOVED***;
            if (start.file) ***REMOVED***
                moznode.loc.source = start.file;
          ***REMOVED***
      ***REMOVED***
        return moznode;
  ***REMOVED***

    function def_to_moz(mytype, handler) ***REMOVED***
        mytype.DEFMETHOD("to_mozilla_ast", function() ***REMOVED***
            return set_moz_loc(this, handler(this));
      ***REMOVED***);
  ***REMOVED***

    function to_moz(node) ***REMOVED***
        return node != null ? node.to_mozilla_ast() : null;
  ***REMOVED***

    function to_moz_alias(alias) ***REMOVED***
        return is_identifier_string(alias.value) ? set_moz_loc(alias, ***REMOVED***
            type: "Identifier",
            name: alias.value,
      ***REMOVED***) : to_moz(alias);
  ***REMOVED***

    function to_moz_block(node) ***REMOVED***
        return ***REMOVED***
            type: "BlockStatement",
            body: node.body.map(to_moz),
      ***REMOVED***;
  ***REMOVED***

    function to_moz_scope(type, node) ***REMOVED***
        var body = node.body.map(to_moz);
        if (node.body[0] instanceof AST_SimpleStatement && node.body[0].body instanceof AST_String) ***REMOVED***
            body.unshift(to_moz(new AST_EmptyStatement(node.body[0])));
      ***REMOVED***
        return ***REMOVED***
            type: type,
            body: body,
      ***REMOVED***;
  ***REMOVED***
})();
