// uglify-js/lib/output.js
/***********************************************************************

  A JavaScript tokenizer / parser / beautifier / compressor.
  https://github.com/mishoo/UglifyJS

  -------------------------------- (C) ---------------------------------

                           Author: Mihai Bazon
                         <mihai.bazon@gmail.com>
                       http://mihai.bazon.net/blog

  Distributed under the BSD license:

    Copyright 2012 (c) Mihai Bazon <mihai.bazon@gmail.com>

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

        * Redistributions of source code must retain the above
          copyright notice, this list of conditions and the following
          disclaimer.

        * Redistributions in binary form must reproduce the above
          copyright notice, this list of conditions and the following
          disclaimer in the documentation and/or other materials
          provided with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER “AS IS” AND ANY
    EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
    PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE
    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
    PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
    TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
    THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
    SUCH DAMAGE.

 ***********************************************************************/

"use strict";

function is_some_comments(comment) ***REMOVED***
    // multiline comment
    return comment.type == "comment2" && /@preserve|@license|@cc_on/i.test(comment.value);
}

function OutputStream(options) ***REMOVED***
    options = defaults(options, ***REMOVED***
        annotations      : false,
        ascii_only       : false,
        beautify         : false,
        braces           : false,
        comments         : false,
        extendscript     : false,
        galio            : false,
        ie               : false,
        indent_level     : 4,
        indent_start     : 0,
        inline_script    : true,
        keep_quoted_props: false,
        max_line_len     : false,
        preamble         : null,
        preserve_line    : false,
        quote_keys       : false,
        quote_style      : 0,
        semicolons       : true,
        shebang          : true,
        source_map       : null,
        v8               : false,
        webkit           : false,
        width            : 80,
        wrap_iife        : false,
  ***REMOVED***, true);

    // Convert comment option to RegExp if necessary and set up comments filter
    var comment_filter = return_false; // Default case, throw all comments away
    if (options.comments) ***REMOVED***
        var comments = options.comments;
        if (typeof options.comments === "string" && /^\/.*\/[a-zA-Z]*$/.test(options.comments)) ***REMOVED***
            var regex_pos = options.comments.lastIndexOf("/");
            comments = new RegExp(
                options.comments.substr(1, regex_pos - 1),
                options.comments.substr(regex_pos + 1)
            );
      ***REMOVED***
        if (comments instanceof RegExp) ***REMOVED***
            comment_filter = function(comment) ***REMOVED***
                return comment.type != "comment5" && comments.test(comment.value);
          ***REMOVED***;
      ***REMOVED*** else if (typeof comments === "function") ***REMOVED***
            comment_filter = function(comment) ***REMOVED***
                return comment.type != "comment5" && comments(this, comment);
          ***REMOVED***;
      ***REMOVED*** else if (comments === "some") ***REMOVED***
            comment_filter = is_some_comments;
      ***REMOVED*** else ***REMOVED*** // NOTE includes "all" option
            comment_filter = return_true;
      ***REMOVED***
  ***REMOVED***

    function make_indent(value) ***REMOVED***
        if (typeof value == "number") return new Array(value + 1).join(" ");
        if (!value) return "";
        if (!/^\s*$/.test(value)) throw new Error("unsupported indentation: " + JSON.stringify("" + value));
        return value;
  ***REMOVED***

    var current_col = 0;
    var current_line = 1;
    var current_indent = make_indent(options.indent_start);
    var full_indent = make_indent(options.indent_level);
    var half_indent = full_indent.length + 1 >> 1;
    var last;
    var line_end = 0;
    var line_fixed = true;
    var mappings = options.source_map && [];
    var mapping_name;
    var mapping_token;
    var might_need_space;
    var might_need_semicolon;
    var need_newline_indented = false;
    var need_space = false;
    var output;
    var stack;
    var stored = "";

    function reset() ***REMOVED***
        last = "";
        might_need_space = false;
        might_need_semicolon = false;
        stack = [];
        var str = output;
        output = "";
        return str;
  ***REMOVED***

    reset();
    var to_utf8 = options.ascii_only ? function(str, identifier) ***REMOVED***
        if (identifier) str = str.replace(/[\ud800-\udbff][\udc00-\udfff]/g, function(ch) ***REMOVED***
            return "\\u***REMOVED***" + (ch.charCodeAt(0) - 0xd7c0 << 10 | ch.charCodeAt(1) - 0xdc00).toString(16) + "}";
      ***REMOVED***);
        return str.replace(/[\u0000-\u001f\u007f-\uffff]/g, function(ch) ***REMOVED***
            var code = ch.charCodeAt(0).toString(16);
            if (code.length <= 2 && !identifier) ***REMOVED***
                while (code.length < 2) code = "0" + code;
                return "\\x" + code;
          ***REMOVED*** else ***REMOVED***
                while (code.length < 4) code = "0" + code;
                return "\\u" + code;
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED*** : function(str) ***REMOVED***
        var s = "";
        for (var i = 0, j = 0; i < str.length; i++) ***REMOVED***
            var code = str.charCodeAt(i);
            if (is_surrogate_pair_head(code)) ***REMOVED***
                if (is_surrogate_pair_tail(str.charCodeAt(i + 1))) ***REMOVED***
                    i++;
                    continue;
              ***REMOVED***
          ***REMOVED*** else if (!is_surrogate_pair_tail(code)) ***REMOVED***
                continue;
          ***REMOVED***
            s += str.slice(j, i) + "\\u" + code.toString(16);
            j = i + 1;
      ***REMOVED***
        return j == 0 ? str : s + str.slice(j);
  ***REMOVED***;

    function quote_single(str) ***REMOVED***
        return "'" + str.replace(/\x27/g, "\\'") + "'";
  ***REMOVED***

    function quote_double(str) ***REMOVED***
        return '"' + str.replace(/\x22/g, '\\"') + '"';
  ***REMOVED***

    var quote_string = [
        null,
        quote_single,
        quote_double,
        function(str, quote) ***REMOVED***
            return quote == "'" ? quote_single(str) : quote_double(str);
      ***REMOVED***,
    ][options.quote_style] || function(str, quote, dq, sq) ***REMOVED***
        return dq > sq ? quote_single(str) : quote_double(str);
  ***REMOVED***;

    function make_string(str, quote) ***REMOVED***
        var dq = 0, sq = 0;
        str = str.replace(/[\\\b\f\n\r\v\t\x22\x27\u2028\u2029\0\ufeff]/g, function(s, i) ***REMOVED***
            switch (s) ***REMOVED***
              case '"': ++dq; return '"';
              case "'": ++sq; return "'";
              case "\\": return "\\\\";
              case "\n": return "\\n";
              case "\r": return "\\r";
              case "\t": return "\\t";
              case "\b": return "\\b";
              case "\f": return "\\f";
              case "\x0B": return options.ie ? "\\x0B" : "\\v";
              case "\u2028": return "\\u2028";
              case "\u2029": return "\\u2029";
              case "\ufeff": return "\\ufeff";
              case "\0":
                  return /[0-9]/.test(str.charAt(i+1)) ? "\\x00" : "\\0";
          ***REMOVED***
            return s;
      ***REMOVED***);
        return quote_string(to_utf8(str), quote, dq, sq);
  ***REMOVED***

    /* -----[ beautification/minification ]----- */

    var adjust_mappings = mappings ? function(line, col) ***REMOVED***
        mappings.forEach(function(mapping) ***REMOVED***
            mapping.line += line;
            mapping.col += col;
      ***REMOVED***);
  ***REMOVED*** : noop;

    var flush_mappings = mappings ? function() ***REMOVED***
        mappings.forEach(function(mapping) ***REMOVED***
            options.source_map.add(
                mapping.token.file,
                mapping.line, mapping.col,
                mapping.token.line, mapping.token.col,
                !mapping.name && mapping.token.type == "name" ? mapping.token.value : mapping.name
            );
      ***REMOVED***);
        mappings = [];
  ***REMOVED*** : noop;

    function insert_newlines(count) ***REMOVED***
        stored += output.slice(0, line_end);
        output = output.slice(line_end);
        var new_col = output.length;
        adjust_mappings(count, new_col - current_col);
        current_line += count;
        current_col = new_col;
        while (count--) stored += "\n";
  ***REMOVED***

    var fix_line = options.max_line_len ? function(flush) ***REMOVED***
        if (line_fixed) ***REMOVED***
            if (current_col > options.max_line_len) ***REMOVED***
                AST_Node.warn("Output exceeds ***REMOVED***max_line_len} characters", options);
          ***REMOVED***
            return;
      ***REMOVED***
        if (current_col > options.max_line_len) ***REMOVED***
            insert_newlines(1);
            line_fixed = true;
      ***REMOVED***
        if (line_fixed || flush) flush_mappings();
  ***REMOVED*** : noop;

    var require_semicolon = makePredicate("( [ + * / - , .");

    function require_space(prev, ch, str) ***REMOVED***
        return is_identifier_char(prev) && (is_identifier_char(ch) || ch == "\\")
            || (ch == "/" && ch == prev)
            || ((ch == "+" || ch == "-") && ch == last)
            || last == "--" && ch == ">"
            || last == "!" && str == "--"
            || prev == "/" && (str == "in" || str == "instanceof");
  ***REMOVED***

    var print = options.beautify
        || options.comments
        || options.max_line_len
        || options.preserve_line
        || options.shebang
        || !options.semicolons
        || options.source_map
        || options.width ? function(str) ***REMOVED***
        var ch = str.charAt(0);
        if (need_newline_indented && ch) ***REMOVED***
            need_newline_indented = false;
            if (ch != "\n") ***REMOVED***
                print("\n");
                indent();
          ***REMOVED***
      ***REMOVED***
        if (need_space && ch) ***REMOVED***
            need_space = false;
            if (!/[\s;})]/.test(ch)) ***REMOVED***
                space();
          ***REMOVED***
      ***REMOVED***
        var prev = last.slice(-1);
        if (might_need_semicolon) ***REMOVED***
            might_need_semicolon = false;
            if (prev == ":" && ch == "}" || prev != ";" && (!ch || ";}".indexOf(ch) < 0)) ***REMOVED***
                var need_semicolon = require_semicolon[ch];
                if (need_semicolon || options.semicolons) ***REMOVED***
                    output += ";";
                    current_col++;
                    if (!line_fixed) ***REMOVED***
                        fix_line();
                        if (line_fixed && !need_semicolon && output == ";") ***REMOVED***
                            output = "";
                            current_col = 0;
                      ***REMOVED***
                  ***REMOVED***
                    if (line_end == output.length - 1) line_end++;
              ***REMOVED*** else ***REMOVED***
                    fix_line();
                    output += "\n";
                    current_line++;
                    current_col = 0;
                    // reset the semicolon flag, since we didn't print one
                    // now and might still have to later
                    if (/^\s+$/.test(str)) might_need_semicolon = true;
              ***REMOVED***
                if (!options.beautify) might_need_space = false;
          ***REMOVED***
      ***REMOVED***

        if (might_need_space) ***REMOVED***
            if (require_space(prev, ch, str)) ***REMOVED***
                output += " ";
                current_col++;
          ***REMOVED***
            if (prev != "<" || str != "!") might_need_space = false;
      ***REMOVED***

        if (mapping_token) ***REMOVED***
            mappings.push(***REMOVED***
                token: mapping_token,
                name: mapping_name,
                line: current_line,
                col: current_col,
          ***REMOVED***);
            mapping_token = false;
            if (line_fixed) flush_mappings();
      ***REMOVED***

        output += str;
        var a = str.split(/\r?\n/), n = a.length - 1;
        current_line += n;
        current_col += a[0].length;
        if (n > 0) ***REMOVED***
            fix_line();
            current_col = a[n].length;
      ***REMOVED***
        last = str;
  ***REMOVED*** : function(str) ***REMOVED***
        var ch = str.charAt(0);
        var prev = last.slice(-1);
        if (might_need_semicolon) ***REMOVED***
            might_need_semicolon = false;
            if (prev == ":" && ch == "}" || (!ch || ";}".indexOf(ch) < 0) && prev != ";") ***REMOVED***
                output += ";";
                might_need_space = false;
          ***REMOVED***
      ***REMOVED***
        if (might_need_space) ***REMOVED***
            if (require_space(prev, ch, str)) output += " ";
            if (prev != "<" || str != "!") might_need_space = false;
      ***REMOVED***
        output += str;
        last = str;
  ***REMOVED***;

    var space = options.beautify ? function() ***REMOVED***
        print(" ");
  ***REMOVED*** : function() ***REMOVED***
        might_need_space = true;
  ***REMOVED***;

    var indent = options.beautify ? function(half) ***REMOVED***
        if (need_newline_indented) print("\n");
        print(half ? current_indent.slice(0, -half_indent) : current_indent);
  ***REMOVED*** : noop;

    var with_indent = options.beautify ? function(cont) ***REMOVED***
        var save_indentation = current_indent;
        current_indent += full_indent;
        cont();
        current_indent = save_indentation;
  ***REMOVED*** : function(cont) ***REMOVED*** cont() };

    var may_add_newline = options.max_line_len || options.preserve_line ? function() ***REMOVED***
        fix_line();
        line_end = output.length;
        line_fixed = false;
  ***REMOVED*** : noop;

    var newline = options.beautify ? function() ***REMOVED***
        print("\n");
        line_end = output.length;
  ***REMOVED*** : may_add_newline;

    var semicolon = options.beautify ? function() ***REMOVED***
        print(";");
  ***REMOVED*** : function() ***REMOVED***
        might_need_semicolon = true;
  ***REMOVED***;

    function force_semicolon() ***REMOVED***
        if (might_need_semicolon) print(";");
        print(";");
  ***REMOVED***

    function with_block(cont, end) ***REMOVED***
        print("***REMOVED***");
        newline();
        with_indent(cont);
        add_mapping(end);
        indent();
        print("}");
  ***REMOVED***

    function with_parens(cont) ***REMOVED***
        print("(");
        may_add_newline();
        cont();
        may_add_newline();
        print(")");
  ***REMOVED***

    function with_square(cont) ***REMOVED***
        print("[");
        may_add_newline();
        cont();
        may_add_newline();
        print("]");
  ***REMOVED***

    function comma() ***REMOVED***
        may_add_newline();
        print(",");
        may_add_newline();
        space();
  ***REMOVED***

    function colon() ***REMOVED***
        print(":");
        space();
  ***REMOVED***

    var add_mapping = mappings ? function(token, name) ***REMOVED***
        mapping_token = token;
        mapping_name = name;
  ***REMOVED*** : noop;

    function get() ***REMOVED***
        if (!line_fixed) fix_line(true);
        return stored + output;
  ***REMOVED***

    function has_nlb() ***REMOVED***
        return /(^|\n) *$/.test(output);
  ***REMOVED***

    function pad_comment(token, force) ***REMOVED***
        if (need_newline_indented) return;
        if (token.nlb && (force || !has_nlb())) ***REMOVED***
            need_newline_indented = true;
      ***REMOVED*** else if (force) ***REMOVED***
            need_space = true;
      ***REMOVED***
  ***REMOVED***

    function print_comment(comment) ***REMOVED***
        var value = comment.value.replace(/[@#]__PURE__/g, " ");
        if (/^\s*$/.test(value) && !/^\s*$/.test(comment.value)) return false;
        if (/comment[134]/.test(comment.type)) ***REMOVED***
            print("//" + value);
            need_newline_indented = true;
      ***REMOVED*** else if (comment.type == "comment2") ***REMOVED***
            print("/*" + value + "*/");
      ***REMOVED***
        return true;
  ***REMOVED***

    function should_merge_comments(node, parent) ***REMOVED***
        if (parent instanceof AST_Binary) return parent.left === node;
        if (parent.TYPE == "Call") return parent.expression === node;
        if (parent instanceof AST_Conditional) return parent.condition === node;
        if (parent instanceof AST_Dot) return parent.expression === node;
        if (parent instanceof AST_Exit) return true;
        if (parent instanceof AST_Sequence) return parent.expressions[0] === node;
        if (parent instanceof AST_Sub) return parent.expression === node;
        if (parent instanceof AST_UnaryPostfix) return true;
        if (parent instanceof AST_Yield) return true;
  ***REMOVED***

    function prepend_comments(node) ***REMOVED***
        var self = this;
        var scan;
        if (node instanceof AST_Exit) ***REMOVED***
            scan = node.value;
      ***REMOVED*** else if (node instanceof AST_Yield) ***REMOVED***
            scan = node.expression;
      ***REMOVED***
        var comments = dump(node);
        if (!comments) comments = [];

        if (scan) ***REMOVED***
            var tw = new TreeWalker(function(node) ***REMOVED***
                if (!should_merge_comments(node, tw.parent())) return true;
                var before = dump(node);
                if (before) comments = comments.concat(before);
          ***REMOVED***);
            tw.push(node);
            scan.walk(tw);
      ***REMOVED***

        if (current_line == 1 && current_col == 0) ***REMOVED***
            if (comments.length > 0 && options.shebang && comments[0].type == "comment5") ***REMOVED***
                print("#!" + comments.shift().value + "\n");
                indent();
          ***REMOVED***
            var preamble = options.preamble;
            if (preamble) print(preamble.replace(/\r\n?|\u2028|\u2029|(^|\S)\s*$/g, "$1\n"));
      ***REMOVED***

        comments = comments.filter(comment_filter, node);
        var printed = false;
        comments.forEach(function(comment, index) ***REMOVED***
            pad_comment(comment, index);
            if (print_comment(comment)) printed = true;
      ***REMOVED***);
        if (printed) pad_comment(node.start, true);

        function dump(node) ***REMOVED***
            var token = node.start;
            if (!token) ***REMOVED***
                if (!scan) return;
                node.start = token = new AST_Token();
          ***REMOVED***
            var comments = token.comments_before;
            if (!comments) ***REMOVED***
                if (!scan) return;
                token.comments_before = comments = [];
          ***REMOVED***
            if (comments._dumped === self) return;
            comments._dumped = self;
            return comments;
      ***REMOVED***
  ***REMOVED***

    function append_comments(node, tail) ***REMOVED***
        var self = this;
        var token = node.end;
        if (!token) return;
        var comments = token[tail ? "comments_before" : "comments_after"];
        if (!comments || comments._dumped === self) return;
        if (!(node instanceof AST_Statement || all(comments, function(c) ***REMOVED***
            return !/comment[134]/.test(c.type);
      ***REMOVED***))) return;
        comments._dumped = self;
        comments.filter(comment_filter, node).forEach(function(comment, index) ***REMOVED***
            pad_comment(comment, index || !tail);
            print_comment(comment);
      ***REMOVED***);
  ***REMOVED***

    return ***REMOVED***
        get             : get,
        reset           : reset,
        indent          : indent,
        should_break    : options.beautify && options.width ? function() ***REMOVED***
            return current_col >= options.width;
      ***REMOVED*** : return_false,
        has_parens      : function() ***REMOVED*** return last.slice(-1) == "(" },
        newline         : newline,
        print           : print,
        space           : space,
        comma           : comma,
        colon           : colon,
        last            : function() ***REMOVED*** return last },
        semicolon       : semicolon,
        force_semicolon : force_semicolon,
        to_utf8         : to_utf8,
        print_name      : function(name) ***REMOVED*** print(to_utf8(name.toString(), true)) },
        print_string    : options.inline_script ? function(str, quote) ***REMOVED***
            str = make_string(str, quote).replace(/<\x2f(script)([>\/\t\n\f\r ])/gi, "<\\/$1$2");
            print(str.replace(/\x3c!--/g, "\\x3c!--").replace(/--\x3e/g, "--\\x3e"));
      ***REMOVED*** : function(str, quote) ***REMOVED***
            print(make_string(str, quote));
      ***REMOVED***,
        with_indent     : with_indent,
        with_block      : with_block,
        with_parens     : with_parens,
        with_square     : with_square,
        add_mapping     : add_mapping,
        option          : function(opt) ***REMOVED*** return options[opt] },
        prepend_comments: options.comments || options.shebang ? prepend_comments : noop,
        append_comments : options.comments ? append_comments : noop,
        push_node       : function(node) ***REMOVED*** stack.push(node) },
        pop_node        : options.preserve_line ? function() ***REMOVED***
            var node = stack.pop();
            if (node.start && node.start.line > current_line) ***REMOVED***
                insert_newlines(node.start.line - current_line);
          ***REMOVED***
      ***REMOVED*** : function() ***REMOVED***
            stack.pop();
      ***REMOVED***,
        parent          : function(n) ***REMOVED***
            return stack[stack.length - 2 - (n || 0)];
      ***REMOVED***,
  ***REMOVED***;
}

/* -----[ code generators ]----- */

(function() ***REMOVED***

    /* -----[ utils ]----- */

    function DEFPRINT(nodetype, generator) ***REMOVED***
        nodetype.DEFMETHOD("_codegen", generator);
  ***REMOVED***

    var use_asm = false;

    AST_Node.DEFMETHOD("print", function(stream, force_parens) ***REMOVED***
        var self = this;
        stream.push_node(self);
        if (force_parens || self.needs_parens(stream)) ***REMOVED***
            stream.with_parens(doit);
      ***REMOVED*** else ***REMOVED***
            doit();
      ***REMOVED***
        stream.pop_node();

        function doit() ***REMOVED***
            stream.prepend_comments(self);
            self.add_source_map(stream);
            self._codegen(stream);
            stream.append_comments(self);
      ***REMOVED***
  ***REMOVED***);
    var readonly = OutputStream(***REMOVED***
        inline_script: false,
        shebang: false,
        width: false,
  ***REMOVED***);
    AST_Node.DEFMETHOD("print_to_string", function(options) ***REMOVED***
        if (options) ***REMOVED***
            var stream = OutputStream(options);
            this.print(stream);
            return stream.get();
      ***REMOVED***
        this.print(readonly);
        return readonly.reset();
  ***REMOVED***);

    /* -----[ PARENTHESES ]----- */

    function PARENS(nodetype, func) ***REMOVED***
        nodetype.DEFMETHOD("needs_parens", func);
  ***REMOVED***

    PARENS(AST_Node, return_false);

    // a function expression needs parens around it when it's provably
    // the first token to appear in a statement.
    function needs_parens_function(output) ***REMOVED***
        var p = output.parent();
        if (!output.has_parens() && first_in_statement(output, false, true)) ***REMOVED***
            // export default function() ***REMOVED***}
            // export default (function foo() ***REMOVED***});
            // export default (function() ***REMOVED***})(foo);
            // export default (function() ***REMOVED***})`foo`;
            // export default (function() ***REMOVED***}) ? foo : bar;
            return this.name || !(p instanceof AST_ExportDefault);
      ***REMOVED***
        if (output.option("webkit") && p instanceof AST_PropAccess && p.expression === this) return true;
        if (output.option("wrap_iife") && p instanceof AST_Call && p.expression === this) return true;
  ***REMOVED***
    PARENS(AST_AsyncFunction, needs_parens_function);
    PARENS(AST_AsyncGeneratorFunction, needs_parens_function);
    PARENS(AST_ClassExpression, needs_parens_function);
    PARENS(AST_Function, needs_parens_function);
    PARENS(AST_GeneratorFunction, needs_parens_function);

    // same goes for an object literal, because otherwise it would be
    // interpreted as a block of code.
    function needs_parens_obj(output) ***REMOVED***
        return !output.has_parens() && first_in_statement(output, true);
  ***REMOVED***
    PARENS(AST_Object, needs_parens_obj);

    function needs_parens_unary(output) ***REMOVED***
        var p = output.parent();
        // (-x) ** y
        if (p instanceof AST_Binary) return p.operator == "**" && p.left === this;
        // (await x)(y)
        // new (await x)
        if (p instanceof AST_Call) return p.expression === this;
        // class extends (x++) ***REMOVED***}
        // class x extends (typeof y) ***REMOVED***}
        if (p instanceof AST_Class) return true;
        // (x++)[y]
        // (typeof x).y
        // https://github.com/mishoo/UglifyJS/issues/115
        if (p instanceof AST_PropAccess) return p.expression === this;
        // (~x)`foo`
        if (p instanceof AST_Template) return p.tag === this;
  ***REMOVED***
    PARENS(AST_Await, needs_parens_unary);
    PARENS(AST_Unary, needs_parens_unary);

    PARENS(AST_Sequence, function(output) ***REMOVED***
        var p = output.parent();
            // [ 1, (2, 3), 4 ] ---> [ 1, 3, 4 ]
        return p instanceof AST_Array
            // () ---> (foo, bar)
            || is_arrow(p) && p.value === this
            // await (foo, bar)
            || p instanceof AST_Await
            // 1 + (2, 3) + 4 ---> 8
            || p instanceof AST_Binary
            // new (foo, bar) or foo(1, (2, 3), 4)
            || p instanceof AST_Call
            // class extends (foo, bar) ***REMOVED***}
            // class foo extends (bar, baz) ***REMOVED***}
            || p instanceof AST_Class
            // class ***REMOVED*** foo = (bar, baz) }
            // class ***REMOVED*** [(foo, bar)]() ***REMOVED***} }
            || p instanceof AST_ClassProperty
            // (false, true) ? (a = 10, b = 20) : (c = 30)
            // ---> 20 (side effect, set a := 10 and b := 20)
            || p instanceof AST_Conditional
            // [ a = (1, 2) ] = [] ---> a == 2
            || p instanceof AST_DefaultValue
            // ***REMOVED*** [(1, 2)]: foo } = bar
            // ***REMOVED*** 1: (2, foo) } = bar
            || p instanceof AST_DestructuredKeyVal
            // export default (foo, bar)
            || p instanceof AST_ExportDefault
            // for (foo of (bar, baz));
            || p instanceof AST_ForOf
            // ***REMOVED*** [(1, 2)]: 3 }[2] ---> 3
            // ***REMOVED*** foo: (1, 2) }.foo ---> 2
            || p instanceof AST_ObjectProperty
            // (1, ***REMOVED***foo:2}).foo or (1, ***REMOVED***foo:2})["foo"] ---> 2
            || p instanceof AST_PropAccess && p.expression === this
            // ...(foo, bar, baz)
            || p instanceof AST_Spread
            // (foo, bar)`baz`
            || p instanceof AST_Template && p.tag === this
            // !(foo, bar, baz)
            || p instanceof AST_Unary
            // var a = (1, 2), b = a + a; ---> b == 4
            || p instanceof AST_VarDef
            // yield (foo, bar)
            || p instanceof AST_Yield;
  ***REMOVED***);

    PARENS(AST_Binary, function(output) ***REMOVED***
        var p = output.parent();
        // await (foo && bar)
        if (p instanceof AST_Await) return true;
        // this deals with precedence:
        //   3 * (2 + 1)
        //   3 - (2 - 1)
        //   (1 ** 2) ** 3
        if (p instanceof AST_Binary) ***REMOVED***
            var po = p.operator, pp = PRECEDENCE[po];
            var so = this.operator, sp = PRECEDENCE[so];
            return pp > sp
                || po == "??" && (so == "&&" || so == "||")
                || (pp == sp && this === p[po == "**" ? "left" : "right"]);
      ***REMOVED***
        // (foo && bar)()
        if (p instanceof AST_Call) return p.expression === this;
        // class extends (foo && bar) ***REMOVED***}
        // class foo extends (bar || null) ***REMOVED***}
        if (p instanceof AST_Class) return true;
        // (foo && bar)["prop"], (foo && bar).prop
        if (p instanceof AST_PropAccess) return p.expression === this;
        // (foo && bar)``
        if (p instanceof AST_Template) return p.tag === this;
        // typeof (foo && bar)
        if (p instanceof AST_Unary) return true;
  ***REMOVED***);

    function need_chain_parens(node, parent) ***REMOVED***
        if (!node.terminal) return false;
        if (!(parent instanceof AST_Call || parent instanceof AST_PropAccess)) return false;
        return parent.expression === node;
  ***REMOVED***

    PARENS(AST_PropAccess, function(output) ***REMOVED***
        var node = this;
        var p = output.parent();
        // i.e. new (foo().bar)
        //
        // if there's one call into this subtree, then we need
        // parens around it too, otherwise the call will be
        // interpreted as passing the arguments to the upper New
        // expression.
        if (p instanceof AST_New && p.expression === node && root_expr(node).TYPE == "Call") return true;
        // (foo?.bar)()
        // (foo?.bar).baz
        // new (foo?.bar)()
        return need_chain_parens(node, p);
  ***REMOVED***);

    PARENS(AST_Call, function(output) ***REMOVED***
        var node = this;
        var p = output.parent();
        if (p instanceof AST_New) return p.expression === node;
        // https://bugs.webkit.org/show_bug.cgi?id=123506
        if (output.option("webkit")
            && node.expression instanceof AST_Function
            && p instanceof AST_PropAccess
            && p.expression === node) ***REMOVED***
            var g = output.parent(1);
            if (g instanceof AST_Assign && g.left === p) return true;
      ***REMOVED***
        // (foo?.())()
        // (foo?.()).bar
        // new (foo?.())()
        return need_chain_parens(node, p);
  ***REMOVED***);

    PARENS(AST_New, function(output) ***REMOVED***
        if (need_constructor_parens(this, output)) return false;
        var p = output.parent();
        // (new foo)(bar)
        if (p instanceof AST_Call) return p.expression === this;
        // (new Date).getTime(), (new Date)["getTime"]()
        if (p instanceof AST_PropAccess) return true;
        // (new foo)`bar`
        if (p instanceof AST_Template) return p.tag === this;
  ***REMOVED***);

    PARENS(AST_Number, function(output) ***REMOVED***
        if (!output.option("galio")) return false;
        // https://github.com/mishoo/UglifyJS/pull/1009
        var p = output.parent();
        return p instanceof AST_PropAccess && p.expression === this && /^0/.test(make_num(this.value));
  ***REMOVED***);

    function needs_parens_assign_cond(self, output) ***REMOVED***
        var p = output.parent();
        // await (a = foo)
        if (p instanceof AST_Await) return true;
        // 1 + (a = 2) + 3 → 6, side effect setting a = 2
        if (p instanceof AST_Binary) return !(p instanceof AST_Assign);
        // (a = func)() —or— new (a = Object)()
        if (p instanceof AST_Call) return p.expression === self;
        // class extends (a = foo) ***REMOVED***}
        // class foo extends (bar ? baz : moo) ***REMOVED***}
        if (p instanceof AST_Class) return true;
        // (a = foo) ? bar : baz
        if (p instanceof AST_Conditional) return p.condition === self;
        // (a = foo)["prop"] —or— (a = foo).prop
        if (p instanceof AST_PropAccess) return p.expression === self;
        // (a = foo)`bar`
        if (p instanceof AST_Template) return p.tag === self;
        // !(a = false) → true
        if (p instanceof AST_Unary) return true;
  ***REMOVED***
    PARENS(AST_Arrow, function(output) ***REMOVED***
        return needs_parens_assign_cond(this, output);
  ***REMOVED***);
    PARENS(AST_Assign, function(output) ***REMOVED***
        if (needs_parens_assign_cond(this, output)) return true;
        //  v8 parser bug   --->     workaround
        // f([1], [a] = []) ---> f([1], ([a] = []))
        if (output.option("v8")) return this.left instanceof AST_Destructured;
        // (***REMOVED*** p: a } = o);
        if (this.left instanceof AST_DestructuredObject) return needs_parens_obj(output);
  ***REMOVED***);
    PARENS(AST_AsyncArrow, function(output) ***REMOVED***
        return needs_parens_assign_cond(this, output);
  ***REMOVED***);
    PARENS(AST_Conditional, function(output) ***REMOVED***
        return needs_parens_assign_cond(this, output)
            // https://github.com/mishoo/UglifyJS/issues/1144
            || output.option("extendscript") && output.parent() instanceof AST_Conditional;
  ***REMOVED***);
    PARENS(AST_Yield, function(output) ***REMOVED***
        return needs_parens_assign_cond(this, output);
  ***REMOVED***);

    /* -----[ PRINTERS ]----- */

    DEFPRINT(AST_Directive, function(output) ***REMOVED***
        var quote = this.quote;
        var value = this.value;
        switch (output.option("quote_style")) ***REMOVED***
          case 0:
          case 2:
            if (value.indexOf('"') == -1) quote = '"';
            break;
          case 1:
            if (value.indexOf("'") == -1) quote = "'";
            break;
      ***REMOVED***
        output.print(quote + value + quote);
        output.semicolon();
  ***REMOVED***);
    DEFPRINT(AST_Debugger, function(output) ***REMOVED***
        output.print("debugger");
        output.semicolon();
  ***REMOVED***);

    /* -----[ statements ]----- */

    function display_body(body, is_toplevel, output, allow_directives) ***REMOVED***
        var last = body.length - 1;
        var in_directive = allow_directives;
        var was_asm = use_asm;
        body.forEach(function(stmt, i) ***REMOVED***
            if (in_directive) ***REMOVED***
                if (stmt instanceof AST_Directive) ***REMOVED***
                    if (stmt.value == "use asm") use_asm = true;
              ***REMOVED*** else if (!(stmt instanceof AST_EmptyStatement)) ***REMOVED***
                    if (stmt instanceof AST_SimpleStatement && stmt.body instanceof AST_String) ***REMOVED***
                        output.force_semicolon();
                  ***REMOVED***
                    in_directive = false;
              ***REMOVED***
          ***REMOVED***
            if (stmt instanceof AST_EmptyStatement) return;
            output.indent();
            stmt.print(output);
            if (i == last && is_toplevel) return;
            output.newline();
            if (is_toplevel) output.newline();
      ***REMOVED***);
        use_asm = was_asm;
  ***REMOVED***

    DEFPRINT(AST_Toplevel, function(output) ***REMOVED***
        display_body(this.body, true, output, true);
        output.print("");
  ***REMOVED***);
    DEFPRINT(AST_LabeledStatement, function(output) ***REMOVED***
        this.label.print(output);
        output.colon();
        this.body.print(output);
  ***REMOVED***);
    DEFPRINT(AST_SimpleStatement, function(output) ***REMOVED***
        this.body.print(output);
        output.semicolon();
  ***REMOVED***);
    function print_braced_empty(self, output) ***REMOVED***
        output.print("***REMOVED***");
        output.with_indent(function() ***REMOVED***
            output.append_comments(self, true);
      ***REMOVED***);
        output.print("}");
  ***REMOVED***
    function print_braced(self, output, allow_directives) ***REMOVED***
        if (self.body.length > 0) ***REMOVED***
            output.with_block(function() ***REMOVED***
                display_body(self.body, false, output, allow_directives);
          ***REMOVED***, self.end);
      ***REMOVED*** else print_braced_empty(self, output);
  ***REMOVED***
    DEFPRINT(AST_BlockStatement, function(output) ***REMOVED***
        print_braced(this, output);
  ***REMOVED***);
    DEFPRINT(AST_EmptyStatement, function(output) ***REMOVED***
        output.semicolon();
  ***REMOVED***);
    DEFPRINT(AST_Do, function(output) ***REMOVED***
        var self = this;
        output.print("do");
        make_block(self.body, output);
        output.space();
        output.print("while");
        output.space();
        output.with_parens(function() ***REMOVED***
            self.condition.print(output);
      ***REMOVED***);
        output.semicolon();
  ***REMOVED***);
    DEFPRINT(AST_While, function(output) ***REMOVED***
        var self = this;
        output.print("while");
        output.space();
        output.with_parens(function() ***REMOVED***
            self.condition.print(output);
      ***REMOVED***);
        force_statement(self.body, output);
  ***REMOVED***);
    DEFPRINT(AST_For, function(output) ***REMOVED***
        var self = this;
        output.print("for");
        output.space();
        output.with_parens(function() ***REMOVED***
            if (self.init) ***REMOVED***
                if (self.init instanceof AST_Definitions) ***REMOVED***
                    self.init.print(output);
              ***REMOVED*** else ***REMOVED***
                    parenthesize_for_no_in(self.init, output, true);
              ***REMOVED***
                output.print(";");
                output.space();
          ***REMOVED*** else ***REMOVED***
                output.print(";");
          ***REMOVED***
            if (self.condition) ***REMOVED***
                self.condition.print(output);
                output.print(";");
                output.space();
          ***REMOVED*** else ***REMOVED***
                output.print(";");
          ***REMOVED***
            if (self.step) ***REMOVED***
                self.step.print(output);
          ***REMOVED***
      ***REMOVED***);
        force_statement(self.body, output);
  ***REMOVED***);
    function print_for_enum(prefix, infix) ***REMOVED***
        return function(output) ***REMOVED***
            var self = this;
            output.print(prefix);
            output.space();
            output.with_parens(function() ***REMOVED***
                self.init.print(output);
                output.space();
                output.print(infix);
                output.space();
                self.object.print(output);
          ***REMOVED***);
            force_statement(self.body, output);
      ***REMOVED***;
  ***REMOVED***
    DEFPRINT(AST_ForAwaitOf, print_for_enum("for await", "of"));
    DEFPRINT(AST_ForIn, print_for_enum("for", "in"));
    DEFPRINT(AST_ForOf, print_for_enum("for", "of"));
    DEFPRINT(AST_With, function(output) ***REMOVED***
        var self = this;
        output.print("with");
        output.space();
        output.with_parens(function() ***REMOVED***
            self.expression.print(output);
      ***REMOVED***);
        force_statement(self.body, output);
  ***REMOVED***);
    DEFPRINT(AST_ExportDeclaration, function(output) ***REMOVED***
        output.print("export");
        output.space();
        this.body.print(output);
  ***REMOVED***);
    DEFPRINT(AST_ExportDefault, function(output) ***REMOVED***
        output.print("export");
        output.space();
        output.print("default");
        output.space();
        var body = this.body;
        body.print(output);
        if (body instanceof AST_ClassExpression) ***REMOVED***
            if (!body.name) return;
      ***REMOVED***
        if (body instanceof AST_DefClass) return;
        if (body instanceof AST_LambdaDefinition) return;
        if (body instanceof AST_LambdaExpression) ***REMOVED***
            if (!body.name && !is_arrow(body)) return;
      ***REMOVED***
        output.semicolon();
  ***REMOVED***);
    function print_alias(alias, output) ***REMOVED***
        var value = alias.value;
        if (value == "*" || is_identifier_string(value)) ***REMOVED***
            output.print_name(value);
      ***REMOVED*** else ***REMOVED***
            output.print_string(value, alias.quote);
      ***REMOVED***
  ***REMOVED***
    DEFPRINT(AST_ExportForeign, function(output) ***REMOVED***
        var self = this;
        output.print("export");
        output.space();
        var len = self.keys.length;
        if (len == 0) ***REMOVED***
            print_braced_empty(self, output);
      ***REMOVED*** else if (self.keys[0].value == "*") ***REMOVED***
            print_entry(0);
      ***REMOVED*** else output.with_block(function() ***REMOVED***
            output.indent();
            print_entry(0);
            for (var i = 1; i < len; i++) ***REMOVED***
                output.print(",");
                output.newline();
                output.indent();
                print_entry(i);
          ***REMOVED***
            output.newline();
      ***REMOVED***, self.end);
        output.space();
        output.print("from");
        output.space();
        self.path.print(output);
        output.semicolon();

        function print_entry(index) ***REMOVED***
            var alias = self.aliases[index];
            var key = self.keys[index];
            print_alias(key, output);
            if (alias.value != key.value) ***REMOVED***
                output.space();
                output.print("as");
                output.space();
                print_alias(alias, output);
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***);
    DEFPRINT(AST_ExportReferences, function(output) ***REMOVED***
        var self = this;
        output.print("export");
        output.space();
        print_properties(self, output);
        output.semicolon();
  ***REMOVED***);
    DEFPRINT(AST_Import, function(output) ***REMOVED***
        var self = this;
        output.print("import");
        output.space();
        if (self.default) self.default.print(output);
        if (self.all) ***REMOVED***
            if (self.default) output.comma();
            self.all.print(output);
      ***REMOVED***
        if (self.properties) ***REMOVED***
            if (self.default) output.comma();
            print_properties(self, output);
      ***REMOVED***
        if (self.all || self.default || self.properties) ***REMOVED***
            output.space();
            output.print("from");
            output.space();
      ***REMOVED***
        self.path.print(output);
        output.semicolon();
  ***REMOVED***);

    /* -----[ functions ]----- */
    function print_funargs(self, output) ***REMOVED***
        output.with_parens(function() ***REMOVED***
            self.argnames.forEach(function(arg, i) ***REMOVED***
                if (i) output.comma();
                arg.print(output);
          ***REMOVED***);
            if (self.rest) ***REMOVED***
                if (self.argnames.length) output.comma();
                output.print("...");
                self.rest.print(output);
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***
    function print_arrow(self, output) ***REMOVED***
        var argname = self.argnames.length == 1 && !self.rest && self.argnames[0];
        if (argname instanceof AST_SymbolFunarg && argname.name != "yield") ***REMOVED***
            argname.print(output);
      ***REMOVED*** else ***REMOVED***
            print_funargs(self, output);
      ***REMOVED***
        output.space();
        output.print("=>");
        output.space();
        if (self.value) ***REMOVED***
            self.value.print(output);
      ***REMOVED*** else ***REMOVED***
            print_braced(self, output, true);
      ***REMOVED***
  ***REMOVED***
    DEFPRINT(AST_Arrow, function(output) ***REMOVED***
        print_arrow(this, output);
  ***REMOVED***);
    DEFPRINT(AST_AsyncArrow, function(output) ***REMOVED***
        output.print("async");
        output.space();
        print_arrow(this, output);
  ***REMOVED***);
    function print_lambda(self, output) ***REMOVED***
        if (self.name) ***REMOVED***
            output.space();
            self.name.print(output);
      ***REMOVED***
        print_funargs(self, output);
        output.space();
        print_braced(self, output, true);
  ***REMOVED***
    DEFPRINT(AST_Lambda, function(output) ***REMOVED***
        output.print("function");
        print_lambda(this, output);
  ***REMOVED***);
    function print_async(output) ***REMOVED***
        output.print("async");
        output.space();
        output.print("function");
        print_lambda(this, output);
  ***REMOVED***
    DEFPRINT(AST_AsyncDefun, print_async);
    DEFPRINT(AST_AsyncFunction, print_async);
    function print_async_generator(output) ***REMOVED***
        output.print("async");
        output.space();
        output.print("function*");
        print_lambda(this, output);
  ***REMOVED***
    DEFPRINT(AST_AsyncGeneratorDefun, print_async_generator);
    DEFPRINT(AST_AsyncGeneratorFunction, print_async_generator);
    function print_generator(output) ***REMOVED***
        output.print("function*");
        print_lambda(this, output);
  ***REMOVED***
    DEFPRINT(AST_GeneratorDefun, print_generator);
    DEFPRINT(AST_GeneratorFunction, print_generator);

    /* -----[ classes ]----- */
    DEFPRINT(AST_Class, function(output) ***REMOVED***
        var self = this;
        output.print("class");
        if (self.name) ***REMOVED***
            output.space();
            self.name.print(output);
      ***REMOVED***
        if (self.extends) ***REMOVED***
            output.space();
            output.print("extends");
            output.space();
            self.extends.print(output);
      ***REMOVED***
        output.space();
        print_properties(self, output, true);
  ***REMOVED***);
    DEFPRINT(AST_ClassField, function(output) ***REMOVED***
        var self = this;
        if (self.static) ***REMOVED***
            output.print("static");
            output.space();
      ***REMOVED***
        print_property_key(self, output);
        if (self.value) ***REMOVED***
            output.space();
            output.print("=");
            output.space();
            self.value.print(output);
      ***REMOVED***
        output.semicolon();
  ***REMOVED***);
    DEFPRINT(AST_ClassGetter, print_accessor("get"));
    DEFPRINT(AST_ClassSetter, print_accessor("set"));
    function print_method(self, output) ***REMOVED***
        var fn = self.value;
        if (is_async(fn)) ***REMOVED***
            output.print("async");
            output.space();
      ***REMOVED***
        if (is_generator(fn)) output.print("*");
        print_property_key(self, output);
        print_lambda(self.value, output);
  ***REMOVED***
    DEFPRINT(AST_ClassMethod, function(output) ***REMOVED***
        var self = this;
        if (self.static) ***REMOVED***
            output.print("static");
            output.space();
      ***REMOVED***
        print_method(self, output);
  ***REMOVED***);
    DEFPRINT(AST_ClassInit, function(output) ***REMOVED***
        output.print("static");
        output.space();
        print_braced(this.value, output);
  ***REMOVED***);

    /* -----[ jumps ]----- */
    function print_jump(kind, prop) ***REMOVED***
        return function(output) ***REMOVED***
            output.print(kind);
            var target = this[prop];
            if (target) ***REMOVED***
                output.space();
                target.print(output);
          ***REMOVED***
            output.semicolon();
      ***REMOVED***;
  ***REMOVED***
    DEFPRINT(AST_Return, print_jump("return", "value"));
    DEFPRINT(AST_Throw, print_jump("throw", "value"));
    DEFPRINT(AST_Break, print_jump("break", "label"));
    DEFPRINT(AST_Continue, print_jump("continue", "label"));

    /* -----[ if ]----- */
    function make_then(self, output) ***REMOVED***
        var b = self.body;
        if (output.option("braces") && !(b instanceof AST_Const || b instanceof AST_Let)
            || output.option("ie") && b instanceof AST_Do)
            return make_block(b, output);
        // The squeezer replaces "block"-s that contain only a single
        // statement with the statement itself; technically, the AST
        // is correct, but this can create problems when we output an
        // IF having an ELSE clause where the THEN clause ends in an
        // IF *without* an ELSE block (then the outer ELSE would refer
        // to the inner IF).  This function checks for this case and
        // adds the block braces if needed.
        if (!b) return output.force_semicolon();
        while (true) ***REMOVED***
            if (b instanceof AST_If) ***REMOVED***
                if (!b.alternative) ***REMOVED***
                    make_block(self.body, output);
                    return;
              ***REMOVED***
                b = b.alternative;
          ***REMOVED*** else if (b instanceof AST_StatementWithBody) ***REMOVED***
                b = b.body;
          ***REMOVED*** else break;
      ***REMOVED***
        force_statement(self.body, output);
  ***REMOVED***
    DEFPRINT(AST_If, function(output) ***REMOVED***
        var self = this;
        output.print("if");
        output.space();
        output.with_parens(function() ***REMOVED***
            self.condition.print(output);
      ***REMOVED***);
        if (self.alternative) ***REMOVED***
            make_then(self, output);
            output.space();
            output.print("else");
            if (self.alternative instanceof AST_If) ***REMOVED***
                output.space();
                self.alternative.print(output);
          ***REMOVED*** else ***REMOVED***
                force_statement(self.alternative, output);
          ***REMOVED***
      ***REMOVED*** else ***REMOVED***
            force_statement(self.body, output);
      ***REMOVED***
  ***REMOVED***);

    /* -----[ switch ]----- */
    DEFPRINT(AST_Switch, function(output) ***REMOVED***
        var self = this;
        output.print("switch");
        output.space();
        output.with_parens(function() ***REMOVED***
            self.expression.print(output);
      ***REMOVED***);
        output.space();
        var last = self.body.length - 1;
        if (last < 0) print_braced_empty(self, output);
        else output.with_block(function() ***REMOVED***
            self.body.forEach(function(branch, i) ***REMOVED***
                output.indent(true);
                branch.print(output);
                if (i < last && branch.body.length > 0)
                    output.newline();
          ***REMOVED***);
      ***REMOVED***, self.end);
  ***REMOVED***);
    function print_branch_body(self, output) ***REMOVED***
        output.newline();
        self.body.forEach(function(stmt) ***REMOVED***
            output.indent();
            stmt.print(output);
            output.newline();
      ***REMOVED***);
  ***REMOVED***
    DEFPRINT(AST_Default, function(output) ***REMOVED***
        output.print("default:");
        print_branch_body(this, output);
  ***REMOVED***);
    DEFPRINT(AST_Case, function(output) ***REMOVED***
        var self = this;
        output.print("case");
        output.space();
        self.expression.print(output);
        output.print(":");
        print_branch_body(self, output);
  ***REMOVED***);

    /* -----[ exceptions ]----- */
    DEFPRINT(AST_Try, function(output) ***REMOVED***
        var self = this;
        output.print("try");
        output.space();
        print_braced(self, output);
        if (self.bcatch) ***REMOVED***
            output.space();
            self.bcatch.print(output);
      ***REMOVED***
        if (self.bfinally) ***REMOVED***
            output.space();
            self.bfinally.print(output);
      ***REMOVED***
  ***REMOVED***);
    DEFPRINT(AST_Catch, function(output) ***REMOVED***
        var self = this;
        output.print("catch");
        if (self.argname) ***REMOVED***
            output.space();
            output.with_parens(function() ***REMOVED***
                self.argname.print(output);
          ***REMOVED***);
      ***REMOVED***
        output.space();
        print_braced(self, output);
  ***REMOVED***);
    DEFPRINT(AST_Finally, function(output) ***REMOVED***
        output.print("finally");
        output.space();
        print_braced(this, output);
  ***REMOVED***);

    function print_definitions(type) ***REMOVED***
        return function(output) ***REMOVED***
            var self = this;
            output.print(type);
            output.space();
            self.definitions.forEach(function(def, i) ***REMOVED***
                if (i) output.comma();
                def.print(output);
          ***REMOVED***);
            var p = output.parent();
            if (!(p instanceof AST_IterationStatement && p.init === self)) output.semicolon();
      ***REMOVED***;
  ***REMOVED***
    DEFPRINT(AST_Const, print_definitions("const"));
    DEFPRINT(AST_Let, print_definitions("let"));
    DEFPRINT(AST_Var, print_definitions("var"));

    function parenthesize_for_no_in(node, output, no_in) ***REMOVED***
        var parens = false;
        // need to take some precautions here:
        //    https://github.com/mishoo/UglifyJS/issues/60
        if (no_in) node.walk(new TreeWalker(function(node) ***REMOVED***
            if (parens) return true;
            if (node instanceof AST_Binary && node.operator == "in") return parens = true;
            if (node instanceof AST_Scope && !(is_arrow(node) && node.value)) return true;
      ***REMOVED***));
        node.print(output, parens);
  ***REMOVED***

    DEFPRINT(AST_VarDef, function(output) ***REMOVED***
        var self = this;
        self.name.print(output);
        if (self.value) ***REMOVED***
            output.space();
            output.print("=");
            output.space();
            var p = output.parent(1);
            var no_in = p instanceof AST_For || p instanceof AST_ForEnumeration;
            parenthesize_for_no_in(self.value, output, no_in);
      ***REMOVED***
  ***REMOVED***);

    DEFPRINT(AST_DefaultValue, function(output) ***REMOVED***
        var self = this;
        self.name.print(output);
        output.space();
        output.print("=");
        output.space();
        self.value.print(output);
  ***REMOVED***);

    /* -----[ other expressions ]----- */
    function print_annotation(self, output) ***REMOVED***
        if (!output.option("annotations")) return;
        if (!self.pure) return;
        var level = 0, parent = self, node;
        do ***REMOVED***
            node = parent;
            parent = output.parent(level++);
            if (parent instanceof AST_Call && parent.expression === node) return;
      ***REMOVED*** while (parent instanceof AST_PropAccess && parent.expression === node);
        output.print("/*@__PURE__*/");
  ***REMOVED***
    function print_call_args(self, output) ***REMOVED***
        output.with_parens(function() ***REMOVED***
            self.args.forEach(function(expr, i) ***REMOVED***
                if (i) output.comma();
                expr.print(output);
          ***REMOVED***);
            output.add_mapping(self.end);
      ***REMOVED***);
  ***REMOVED***
    DEFPRINT(AST_Call, function(output) ***REMOVED***
        var self = this;
        print_annotation(self, output);
        self.expression.print(output);
        if (self.optional) output.print("?.");
        print_call_args(self, output);
  ***REMOVED***);
    DEFPRINT(AST_New, function(output) ***REMOVED***
        var self = this;
        print_annotation(self, output);
        output.print("new");
        output.space();
        self.expression.print(output);
        if (need_constructor_parens(self, output)) print_call_args(self, output);
  ***REMOVED***);
    DEFPRINT(AST_Sequence, function(output) ***REMOVED***
        this.expressions.forEach(function(node, index) ***REMOVED***
            if (index > 0) ***REMOVED***
                output.comma();
                if (output.should_break()) ***REMOVED***
                    output.newline();
                    output.indent();
              ***REMOVED***
          ***REMOVED***
            node.print(output);
      ***REMOVED***);
  ***REMOVED***);
    DEFPRINT(AST_Dot, function(output) ***REMOVED***
        var self = this;
        var expr = self.expression;
        expr.print(output);
        var prop = self.property;
        if (output.option("ie") && RESERVED_WORDS[prop] || self.quoted && output.option("keep_quoted_props")) ***REMOVED***
            if (self.optional) output.print("?.");
            output.with_square(function() ***REMOVED***
                output.add_mapping(self.end);
                output.print_string(prop);
          ***REMOVED***);
      ***REMOVED*** else ***REMOVED***
            if (expr instanceof AST_Number && !/[ex.)]/i.test(output.last())) output.print(".");
            output.print(self.optional ? "?." : ".");
            // the name after dot would be mapped about here.
            output.add_mapping(self.end);
            output.print_name(prop);
      ***REMOVED***
  ***REMOVED***);
    DEFPRINT(AST_Sub, function(output) ***REMOVED***
        var self = this;
        self.expression.print(output);
        if (self.optional) output.print("?.");
        output.with_square(function() ***REMOVED***
            self.property.print(output);
      ***REMOVED***);
  ***REMOVED***);
    DEFPRINT(AST_Spread, function(output) ***REMOVED***
        output.print("...");
        this.expression.print(output);
  ***REMOVED***);
    DEFPRINT(AST_UnaryPrefix, function(output) ***REMOVED***
        var op = this.operator;
        var exp = this.expression;
        output.print(op);
        if (/^[a-z]/i.test(op)
            || (/[+-]$/.test(op)
                && exp instanceof AST_UnaryPrefix
                && /^[+-]/.test(exp.operator))) ***REMOVED***
            output.space();
      ***REMOVED***
        exp.print(output);
  ***REMOVED***);
    DEFPRINT(AST_UnaryPostfix, function(output) ***REMOVED***
        var self = this;
        self.expression.print(output);
        output.add_mapping(self.end);
        output.print(self.operator);
  ***REMOVED***);
    DEFPRINT(AST_Binary, function(output) ***REMOVED***
        var self = this;
        self.left.print(output);
        output.space();
        output.print(self.operator);
        output.space();
        self.right.print(output);
  ***REMOVED***);
    DEFPRINT(AST_Conditional, function(output) ***REMOVED***
        var self = this;
        self.condition.print(output);
        output.space();
        output.print("?");
        output.space();
        self.consequent.print(output);
        output.space();
        output.colon();
        self.alternative.print(output);
  ***REMOVED***);
    DEFPRINT(AST_Await, function(output) ***REMOVED***
        output.print("await");
        output.space();
        this.expression.print(output);
  ***REMOVED***);
    DEFPRINT(AST_Yield, function(output) ***REMOVED***
        output.print(this.nested ? "yield*" : "yield");
        if (this.expression) ***REMOVED***
            output.space();
            this.expression.print(output);
      ***REMOVED***
  ***REMOVED***);

    /* -----[ literals ]----- */
    DEFPRINT(AST_Array, function(output) ***REMOVED***
        var a = this.elements, len = a.length;
        output.with_square(len > 0 ? function() ***REMOVED***
            output.space();
            a.forEach(function(exp, i) ***REMOVED***
                if (i) output.comma();
                exp.print(output);
                // If the final element is a hole, we need to make sure it
                // doesn't look like a trailing comma, by inserting an actual
                // trailing comma.
                if (i === len - 1 && exp instanceof AST_Hole)
                  output.comma();
          ***REMOVED***);
            output.space();
      ***REMOVED*** : noop);
  ***REMOVED***);
    DEFPRINT(AST_DestructuredArray, function(output) ***REMOVED***
        var a = this.elements, len = a.length, rest = this.rest;
        output.with_square(len || rest ? function() ***REMOVED***
            output.space();
            a.forEach(function(exp, i) ***REMOVED***
                if (i) output.comma();
                exp.print(output);
          ***REMOVED***);
            if (rest) ***REMOVED***
                if (len) output.comma();
                output.print("...");
                rest.print(output);
          ***REMOVED*** else if (a[len - 1] instanceof AST_Hole) ***REMOVED***
                // If the final element is a hole, we need to make sure it
                // doesn't look like a trailing comma, by inserting an actual
                // trailing comma.
                output.comma();
          ***REMOVED***
            output.space();
      ***REMOVED*** : noop);
  ***REMOVED***);
    DEFPRINT(AST_DestructuredKeyVal, function(output) ***REMOVED***
        var self = this;
        var key = print_property_key(self, output);
        var value = self.value;
        if (key) ***REMOVED***
            if (value instanceof AST_DefaultValue) ***REMOVED***
                if (value.name instanceof AST_Symbol && key == get_symbol_name(value.name)) ***REMOVED***
                    output.space();
                    output.print("=");
                    output.space();
                    value.value.print(output);
                    return;
              ***REMOVED***
          ***REMOVED*** else if (value instanceof AST_Symbol) ***REMOVED***
                if (key == get_symbol_name(value)) return;
          ***REMOVED***
      ***REMOVED***
        output.colon();
        value.print(output);
  ***REMOVED***);
    DEFPRINT(AST_DestructuredObject, function(output) ***REMOVED***
        var self = this;
        var props = self.properties, len = props.length, rest = self.rest;
        if (len || rest) output.with_block(function() ***REMOVED***
            props.forEach(function(prop, i) ***REMOVED***
                if (i) ***REMOVED***
                    output.print(",");
                    output.newline();
              ***REMOVED***
                output.indent();
                prop.print(output);
          ***REMOVED***);
            if (rest) ***REMOVED***
                if (len) ***REMOVED***
                    output.print(",");
                    output.newline();
              ***REMOVED***
                output.indent();
                output.print("...");
                rest.print(output);
          ***REMOVED***
            output.newline();
      ***REMOVED***, self.end);
        else print_braced_empty(self, output);
  ***REMOVED***);
    function print_properties(self, output, no_comma) ***REMOVED***
        var props = self.properties;
        if (props.length > 0) output.with_block(function() ***REMOVED***
            props.forEach(function(prop, i) ***REMOVED***
                if (i) ***REMOVED***
                    if (!no_comma) output.print(",");
                    output.newline();
              ***REMOVED***
                output.indent();
                prop.print(output);
          ***REMOVED***);
            output.newline();
      ***REMOVED***, self.end);
        else print_braced_empty(self, output);
  ***REMOVED***
    DEFPRINT(AST_Object, function(output) ***REMOVED***
        print_properties(this, output);
  ***REMOVED***);

    function print_property_key(self, output) ***REMOVED***
        var key = self.key;
        if (key instanceof AST_Node) return output.with_square(function() ***REMOVED***
            key.print(output);
      ***REMOVED***);
        var quote = self.start && self.start.quote;
        if (output.option("quote_keys") || quote && output.option("keep_quoted_props")) ***REMOVED***
            output.print_string(key, quote);
      ***REMOVED*** else if ("" + +key == key && key >= 0) ***REMOVED***
            output.print(make_num(key));
      ***REMOVED*** else if (self.private) ***REMOVED***
            output.print_name(key);
      ***REMOVED*** else if (RESERVED_WORDS[key] ? !output.option("ie") : is_identifier_string(key)) ***REMOVED***
            output.print_name(key);
            return key;
      ***REMOVED*** else ***REMOVED***
            output.print_string(key, quote);
      ***REMOVED***
  ***REMOVED***
    DEFPRINT(AST_ObjectKeyVal, function(output) ***REMOVED***
        var self = this;
        print_property_key(self, output);
        output.colon();
        self.value.print(output);
  ***REMOVED***);
    DEFPRINT(AST_ObjectMethod, function(output) ***REMOVED***
        print_method(this, output);
  ***REMOVED***);
    function print_accessor(type) ***REMOVED***
        return function(output) ***REMOVED***
            var self = this;
            if (self.static) ***REMOVED***
                output.print("static");
                output.space();
          ***REMOVED***
            output.print(type);
            output.space();
            print_property_key(self, output);
            print_lambda(self.value, output);
      ***REMOVED***;
  ***REMOVED***
    DEFPRINT(AST_ObjectGetter, print_accessor("get"));
    DEFPRINT(AST_ObjectSetter, print_accessor("set"));
    function get_symbol_name(sym) ***REMOVED***
        var def = sym.definition();
        return def && def.mangled_name || sym.name;
  ***REMOVED***
    DEFPRINT(AST_Symbol, function(output) ***REMOVED***
        output.print_name(get_symbol_name(this));
  ***REMOVED***);
    DEFPRINT(AST_SymbolExport, function(output) ***REMOVED***
        var self = this;
        var name = get_symbol_name(self);
        output.print_name(name);
        var alias = self.alias;
        if (alias.value != name) ***REMOVED***
            output.space();
            output.print("as");
            output.space();
            print_alias(alias, output);
      ***REMOVED***
  ***REMOVED***);
    DEFPRINT(AST_SymbolImport, function(output) ***REMOVED***
        var self = this;
        var name = get_symbol_name(self);
        var key = self.key;
        if (key.value && key.value != name) ***REMOVED***
            print_alias(key, output);
            output.space();
            output.print("as");
            output.space();
      ***REMOVED***
        output.print_name(name);
  ***REMOVED***);
    DEFPRINT(AST_Hole, noop);
    DEFPRINT(AST_Template, function(output) ***REMOVED***
        var self = this;
        if (self.tag) self.tag.print(output);
        output.print("`");
        for (var i = 0; i < self.expressions.length; i++) ***REMOVED***
            output.print(self.strings[i]);
            output.print("$***REMOVED***");
            self.expressions[i].print(output);
            output.print("}");
      ***REMOVED***
        output.print(self.strings[i]);
        output.print("`");
  ***REMOVED***);
    DEFPRINT(AST_Constant, function(output) ***REMOVED***
        output.print("" + this.value);
  ***REMOVED***);
    DEFPRINT(AST_String, function(output) ***REMOVED***
        output.print_string(this.value, this.quote);
  ***REMOVED***);
    DEFPRINT(AST_Number, function(output) ***REMOVED***
        var start = this.start;
        if (use_asm && start && start.raw != null) ***REMOVED***
            output.print(start.raw);
      ***REMOVED*** else ***REMOVED***
            output.print(make_num(this.value));
      ***REMOVED***
  ***REMOVED***);

    DEFPRINT(AST_RegExp, function(output) ***REMOVED***
        var regexp = this.value;
        var str = regexp.toString();
        var end = str.lastIndexOf("/");
        if (regexp.raw_source) ***REMOVED***
            str = "/" + regexp.raw_source + str.slice(end);
      ***REMOVED*** else if (end == 1) ***REMOVED***
            str = "/(?:)" + str.slice(end);
      ***REMOVED*** else if (str.indexOf("/", 1) < end) ***REMOVED***
            str = "/" + str.slice(1, end).replace(/\\\\|[^/]?\//g, function(match) ***REMOVED***
                return match[0] == "\\" ? match : match.slice(0, -1) + "\\/";
          ***REMOVED***) + str.slice(end);
      ***REMOVED***
        output.print(output.to_utf8(str).replace(/\\(?:\0(?![0-9])|[^\0])/g, function(match) ***REMOVED***
            switch (match[1]) ***REMOVED***
              case "\n": return "\\n";
              case "\r": return "\\r";
              case "\t": return "\t";
              case "\b": return "\b";
              case "\f": return "\f";
              case "\0": return "\0";
              case "\x0B": return "\v";
              case "\u2028": return "\\u2028";
              case "\u2029": return "\\u2029";
              default: return match;
          ***REMOVED***
      ***REMOVED***).replace(/[\n\r\u2028\u2029]/g, function(c) ***REMOVED***
            switch (c) ***REMOVED***
              case "\n": return "\\n";
              case "\r": return "\\r";
              case "\u2028": return "\\u2028";
              case "\u2029": return "\\u2029";
          ***REMOVED***
      ***REMOVED***));
  ***REMOVED***);

    function force_statement(stat, output) ***REMOVED***
        if (output.option("braces") && !(stat instanceof AST_Const || stat instanceof AST_Let)) ***REMOVED***
            make_block(stat, output);
      ***REMOVED*** else if (stat instanceof AST_EmptyStatement) ***REMOVED***
            output.force_semicolon();
      ***REMOVED*** else ***REMOVED***
            output.space();
            stat.print(output);
      ***REMOVED***
  ***REMOVED***

    // self should be AST_New.  decide if we want to show parens or not.
    function need_constructor_parens(self, output) ***REMOVED***
        // Always print parentheses with arguments
        if (self.args.length > 0) return true;

        return output.option("beautify");
  ***REMOVED***

    function best_of(a) ***REMOVED***
        var best = a[0], len = best.length;
        for (var i = 1; i < a.length; ++i) ***REMOVED***
            if (a[i].length < len) ***REMOVED***
                best = a[i];
                len = best.length;
          ***REMOVED***
      ***REMOVED***
        return best;
  ***REMOVED***

    function make_num(num) ***REMOVED***
        var str = num.toString(10).replace(/^0\./, ".").replace("e+", "e");
        var candidates = [ str ];
        if (Math.floor(num) === num) ***REMOVED***
            if (num < 0) ***REMOVED***
                candidates.push("-0x" + (-num).toString(16).toLowerCase());
          ***REMOVED*** else ***REMOVED***
                candidates.push("0x" + num.toString(16).toLowerCase());
          ***REMOVED***
      ***REMOVED***
        var match, len, digits;
        if (match = /^\.0+/.exec(str)) ***REMOVED***
            len = match[0].length;
            digits = str.slice(len);
            candidates.push(digits + "e-" + (digits.length + len - 1));
      ***REMOVED*** else if (match = /[^0]0+$/.exec(str)) ***REMOVED***
            len = match[0].length - 1;
            candidates.push(str.slice(0, -len) + "e" + len);
      ***REMOVED*** else if (match = /^(\d)\.(\d+)e(-?\d+)$/.exec(str)) ***REMOVED***
            candidates.push(match[1] + match[2] + "e" + (match[3] - match[2].length));
      ***REMOVED***
        return best_of(candidates);
  ***REMOVED***

    function make_block(stmt, output) ***REMOVED***
        output.space();
        if (stmt instanceof AST_EmptyStatement) ***REMOVED***
            print_braced_empty(stmt, output);
      ***REMOVED*** else if (stmt instanceof AST_BlockStatement) ***REMOVED***
            stmt.print(output);
      ***REMOVED*** else output.with_block(function() ***REMOVED***
            output.indent();
            stmt.print(output);
            output.newline();
      ***REMOVED***, stmt.end);
  ***REMOVED***

    /* -----[ source map generators ]----- */

    function DEFMAP(nodetype, generator) ***REMOVED***
        nodetype.forEach(function(nodetype) ***REMOVED***
            nodetype.DEFMETHOD("add_source_map", generator);
      ***REMOVED***);
  ***REMOVED***

    DEFMAP([
        // We could easily add info for ALL nodes, but it seems to me that
        // would be quite wasteful, hence this noop in the base class.
        AST_Node,
        // since the label symbol will mark it
        AST_LabeledStatement,
    ], noop);

    // XXX: I'm not exactly sure if we need it for all of these nodes,
    // or if we should add even more.
    DEFMAP([
        AST_Array,
        AST_Await,
        AST_BlockStatement,
        AST_Catch,
        AST_Constant,
        AST_Debugger,
        AST_Definitions,
        AST_Destructured,
        AST_Directive,
        AST_Finally,
        AST_Jump,
        AST_Lambda,
        AST_New,
        AST_Object,
        AST_Spread,
        AST_StatementWithBody,
        AST_Symbol,
        AST_Switch,
        AST_SwitchBranch,
        AST_Try,
        AST_UnaryPrefix,
        AST_Yield,
    ], function(output) ***REMOVED***
        output.add_mapping(this.start);
  ***REMOVED***);

    DEFMAP([
        AST_ClassProperty,
        AST_DestructuredKeyVal,
        AST_ObjectProperty,
    ], function(output) ***REMOVED***
        if (typeof this.key == "string") output.add_mapping(this.start, this.key);
  ***REMOVED***);
})();
