// uglify-js/lib/parse.js
/***********************************************************************

  A JavaScript tokenizer / parser / beautifier / compressor.
  https://github.com/mishoo/UglifyJS

  -------------------------------- (C) ---------------------------------

                           Author: Mihai Bazon
                         <mihai.bazon@gmail.com>
                       http://mihai.bazon.net/blog

  Distributed under the BSD license:

    Copyright 2012 (c) Mihai Bazon <mihai.bazon@gmail.com>
    Parser based on parse-js (http://marijn.haverbeke.nl/parse-js/).

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

        * Redistributions of source code must retain the above
          copyright notice, this list of conditions and the following
          disclaimer.

        * Redistributions in binary form must reproduce the above
          copyright notice, this list of conditions and the following
          disclaimer in the documentation and/or other materials
          provided with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER “AS IS” AND ANY
    EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
    PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE
    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
    PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
    TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
    THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
    SUCH DAMAGE.

 ***********************************************************************/

"use strict";

var KEYWORDS = "break case catch class const continue debugger default delete do else extends finally for function if in instanceof new return switch throw try typeof var void while with";
var KEYWORDS_ATOM = "false null true";
var RESERVED_WORDS = [
    "abstract async await boolean byte char double enum export final float goto implements import int interface let long native package private protected public short static super synchronized this throws transient volatile yield",
    KEYWORDS_ATOM,
    KEYWORDS,
].join(" ");
var KEYWORDS_BEFORE_EXPRESSION = "return new delete throw else case";

KEYWORDS = makePredicate(KEYWORDS);
RESERVED_WORDS = makePredicate(RESERVED_WORDS);
KEYWORDS_BEFORE_EXPRESSION = makePredicate(KEYWORDS_BEFORE_EXPRESSION);
KEYWORDS_ATOM = makePredicate(KEYWORDS_ATOM);

var RE_BIN_NUMBER = /^0b([01]+)$/i;
var RE_HEX_NUMBER = /^0x([0-9a-f]+)$/i;
var RE_OCT_NUMBER = /^0o?([0-7]+)$/i;

var OPERATORS = makePredicate([
    "in",
    "instanceof",
    "typeof",
    "new",
    "void",
    "delete",
    "++",
    "--",
    "+",
    "-",
    "!",
    "~",
    "&",
    "|",
    "^",
    "*",
    "/",
    "%",
    "**",
    ">>",
    "<<",
    ">>>",
    "<",
    ">",
    "<=",
    ">=",
    "==",
    "===",
    "!=",
    "!==",
    "?",
    "=",
    "+=",
    "-=",
    "/=",
    "*=",
    "%=",
    "**=",
    ">>=",
    "<<=",
    ">>>=",
    "&=",
    "|=",
    "^=",
    "&&",
    "||",
    "??",
    "&&=",
    "||=",
    "??=",
]);

var NEWLINE_CHARS = "\n\r\u2028\u2029";
var OPERATOR_CHARS = "+-*&%=<>!?|~^";
var PUNC_OPENERS = "[***REMOVED***(";
var PUNC_SEPARATORS = ",;:";
var PUNC_CLOSERS = ")}]";
var PUNC_AFTER_EXPRESSION = PUNC_SEPARATORS + PUNC_CLOSERS;
var PUNC_BEFORE_EXPRESSION = PUNC_OPENERS + PUNC_SEPARATORS;
var PUNC_CHARS = PUNC_BEFORE_EXPRESSION + "`" + PUNC_CLOSERS;
var WHITESPACE_CHARS = NEWLINE_CHARS + " \u00a0\t\f\u000b\u200b\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u202f\u205f\u3000\uFEFF";
var NON_IDENTIFIER_CHARS = makePredicate(characters("./'\"#" + OPERATOR_CHARS + PUNC_CHARS + WHITESPACE_CHARS));

NEWLINE_CHARS = makePredicate(characters(NEWLINE_CHARS));
OPERATOR_CHARS = makePredicate(characters(OPERATOR_CHARS));
PUNC_AFTER_EXPRESSION = makePredicate(characters(PUNC_AFTER_EXPRESSION));
PUNC_BEFORE_EXPRESSION = makePredicate(characters(PUNC_BEFORE_EXPRESSION));
PUNC_CHARS = makePredicate(characters(PUNC_CHARS));
WHITESPACE_CHARS = makePredicate(characters(WHITESPACE_CHARS));

/* -----[ Tokenizer ]----- */

function is_surrogate_pair_head(code) ***REMOVED***
    return code >= 0xd800 && code <= 0xdbff;
}

function is_surrogate_pair_tail(code) ***REMOVED***
    return code >= 0xdc00 && code <= 0xdfff;
}

function is_digit(code) ***REMOVED***
    return code >= 48 && code <= 57;
}

function is_identifier_char(ch) ***REMOVED***
    return !NON_IDENTIFIER_CHARS[ch];
}

function is_identifier_string(str) ***REMOVED***
    return /^[a-z_$][a-z0-9_$]*$/i.test(str);
}

function decode_escape_sequence(seq) ***REMOVED***
    switch (seq[0]) ***REMOVED***
      case "b": return "\b";
      case "f": return "\f";
      case "n": return "\n";
      case "r": return "\r";
      case "t": return "\t";
      case "u":
        var code;
        if (seq[1] == "***REMOVED***" && seq.slice(-1) == "}") ***REMOVED***
            code = seq.slice(2, -1);
      ***REMOVED*** else if (seq.length == 5) ***REMOVED***
            code = seq.slice(1);
      ***REMOVED*** else ***REMOVED***
            return;
      ***REMOVED***
        var num = parseInt(code, 16);
        if (num < 0 || isNaN(num)) return;
        if (num < 0x10000) return String.fromCharCode(num);
        if (num > 0x10ffff) return;
        return String.fromCharCode((num >> 10) + 0xd7c0) + String.fromCharCode((num & 0x03ff) + 0xdc00);
      case "v": return "\u000b";
      case "x":
        if (seq.length != 3) return;
        var num = parseInt(seq.slice(1), 16);
        if (num < 0 || isNaN(num)) return;
        return String.fromCharCode(num);
      case "\r":
      case "\n":
        return "";
      default:
        if (seq == "0") return "\0";
        if (seq[0] >= "0" && seq[0] <= "9") return;
        return seq;
  ***REMOVED***
}

function parse_js_number(num) ***REMOVED***
    var match;
    if (match = RE_BIN_NUMBER.exec(num)) return parseInt(match[1], 2);
    if (match = RE_HEX_NUMBER.exec(num)) return parseInt(match[1], 16);
    if (match = RE_OCT_NUMBER.exec(num)) return parseInt(match[1], 8);
    var val = parseFloat(num);
    if (val == num) return val;
}

function JS_Parse_Error(message, filename, line, col, pos) ***REMOVED***
    this.message = message;
    this.filename = filename;
    this.line = line;
    this.col = col;
    this.pos = pos;
}
JS_Parse_Error.prototype = Object.create(Error.prototype);
JS_Parse_Error.prototype.constructor = JS_Parse_Error;
JS_Parse_Error.prototype.name = "SyntaxError";
configure_error_stack(JS_Parse_Error);

function js_error(message, filename, line, col, pos) ***REMOVED***
    throw new JS_Parse_Error(message, filename, line, col, pos);
}

function is_token(token, type, val) ***REMOVED***
    return token.type == type && (val == null || token.value == val);
}

var EX_EOF = ***REMOVED***};

function tokenizer($TEXT, filename, html5_comments, shebang) ***REMOVED***

    var S = ***REMOVED***
        text            : $TEXT,
        filename        : filename,
        pos             : 0,
        tokpos          : 0,
        line            : 1,
        tokline         : 0,
        col             : 0,
        tokcol          : 0,
        newline_before  : false,
        regex_allowed   : false,
        comments_before : [],
        directives      : Object.create(null),
        read_template   : with_eof_error("Unterminated template literal", function(strings) ***REMOVED***
            var s = "";
            for (;;) ***REMOVED***
                var ch = read();
                switch (ch) ***REMOVED***
                  case "\\":
                    ch += read();
                    break;
                  case "`":
                    strings.push(s);
                    return;
                  case "$":
                    if (peek() == "***REMOVED***") ***REMOVED***
                        next();
                        strings.push(s);
                        S.regex_allowed = true;
                        return true;
                  ***REMOVED***
              ***REMOVED***
                s += ch;
          ***REMOVED***

            function read() ***REMOVED***
                var ch = next(true, true);
                return ch == "\r" ? "\n" : ch;
          ***REMOVED***
      ***REMOVED***),
  ***REMOVED***;
    var prev_was_dot = false;

    function peek() ***REMOVED***
        return S.text.charAt(S.pos);
  ***REMOVED***

    function next(signal_eof, in_string) ***REMOVED***
        var ch = S.text.charAt(S.pos++);
        if (signal_eof && !ch)
            throw EX_EOF;
        if (NEWLINE_CHARS[ch]) ***REMOVED***
            S.col = 0;
            S.line++;
            if (!in_string) S.newline_before = true;
            if (ch == "\r" && peek() == "\n") ***REMOVED***
                // treat `\r\n` as `\n`
                S.pos++;
                ch = "\n";
          ***REMOVED***
      ***REMOVED*** else ***REMOVED***
            S.col++;
      ***REMOVED***
        return ch;
  ***REMOVED***

    function forward(i) ***REMOVED***
        while (i-- > 0) next();
  ***REMOVED***

    function looking_at(str) ***REMOVED***
        return S.text.substr(S.pos, str.length) == str;
  ***REMOVED***

    function find_eol() ***REMOVED***
        var text = S.text;
        for (var i = S.pos; i < S.text.length; ++i) ***REMOVED***
            if (NEWLINE_CHARS[text[i]]) return i;
      ***REMOVED***
        return -1;
  ***REMOVED***

    function find(what, signal_eof) ***REMOVED***
        var pos = S.text.indexOf(what, S.pos);
        if (signal_eof && pos == -1) throw EX_EOF;
        return pos;
  ***REMOVED***

    function start_token() ***REMOVED***
        S.tokline = S.line;
        S.tokcol = S.col;
        S.tokpos = S.pos;
  ***REMOVED***

    function token(type, value, is_comment) ***REMOVED***
        S.regex_allowed = type == "operator" && !UNARY_POSTFIX[value]
            || type == "keyword" && KEYWORDS_BEFORE_EXPRESSION[value]
            || type == "punc" && PUNC_BEFORE_EXPRESSION[value];
        if (type == "punc" && value == ".") prev_was_dot = true;
        else if (!is_comment) prev_was_dot = false;
        var ret = ***REMOVED***
            type    : type,
            value   : value,
            line    : S.tokline,
            col     : S.tokcol,
            pos     : S.tokpos,
            endline : S.line,
            endcol  : S.col,
            endpos  : S.pos,
            nlb     : S.newline_before,
            file    : filename
      ***REMOVED***;
        if (/^(?:num|string|regexp)$/i.test(type)) ***REMOVED***
            ret.raw = $TEXT.substring(ret.pos, ret.endpos);
      ***REMOVED***
        if (!is_comment) ***REMOVED***
            ret.comments_before = S.comments_before;
            ret.comments_after = S.comments_before = [];
      ***REMOVED***
        S.newline_before = false;
        return new AST_Token(ret);
  ***REMOVED***

    function skip_whitespace() ***REMOVED***
        while (WHITESPACE_CHARS[peek()])
            next();
  ***REMOVED***

    function read_while(pred) ***REMOVED***
        var ret = "", ch;
        while ((ch = peek()) && pred(ch, ret)) ret += next();
        return ret;
  ***REMOVED***

    function parse_error(err) ***REMOVED***
        js_error(err, filename, S.tokline, S.tokcol, S.tokpos);
  ***REMOVED***

    function is_octal(num) ***REMOVED***
        return /^0[0-7_]+$/.test(num);
  ***REMOVED***

    function read_num(prefix) ***REMOVED***
        var has_e = false, after_e = false, has_x = false, has_dot = prefix == ".";
        var num = read_while(function(ch, str) ***REMOVED***
            switch (ch) ***REMOVED***
              case "x": case "X":
                return has_x ? false : (has_x = true);
              case "e": case "E":
                return has_x ? true : has_e ? false : (has_e = after_e = true);
              case "+": case "-":
                return after_e;
              case (after_e = false, "."):
                return has_dot || has_e || has_x || is_octal(str) ? false : (has_dot = true);
          ***REMOVED***
            return /[_0-9a-dfo]/i.test(ch);
      ***REMOVED***);
        if (prefix) num = prefix + num;
        if (is_octal(num)) ***REMOVED***
            if (next_token.has_directive("use strict")) parse_error("Legacy octal literals are not allowed in strict mode");
      ***REMOVED*** else ***REMOVED***
            num = num.replace(has_x ? /([1-9a-f]|.0)_(?=[0-9a-f])/gi : /([1-9]|.0)_(?=[0-9])/gi, "$1");
      ***REMOVED***
        var valid = parse_js_number(num);
        if (isNaN(valid)) parse_error("Invalid syntax: " + num);
        if (has_dot || has_e || peek() != "n") return token("num", valid);
        return token("bigint", num.toLowerCase() + next());
  ***REMOVED***

    function read_escaped_char(in_string) ***REMOVED***
        var seq = next(true, in_string);
        if (seq >= "0" && seq <= "7") return read_octal_escape_sequence(seq);
        if (seq == "u") ***REMOVED***
            var ch = next(true, in_string);
            seq += ch;
            if (ch != "***REMOVED***") ***REMOVED***
                seq += next(true, in_string) + next(true, in_string) + next(true, in_string);
          ***REMOVED*** else do ***REMOVED***
                ch = next(true, in_string);
                seq += ch;
          ***REMOVED*** while (ch != "}");
      ***REMOVED*** else if (seq == "x") ***REMOVED***
            seq += next(true, in_string) + next(true, in_string);
      ***REMOVED***
        var str = decode_escape_sequence(seq);
        if (typeof str != "string") parse_error("Invalid escape sequence: \\" + seq);
        return str;
  ***REMOVED***

    function read_octal_escape_sequence(ch) ***REMOVED***
        // Read
        var p = peek();
        if (p >= "0" && p <= "7") ***REMOVED***
            ch += next(true);
            if (ch[0] <= "3" && (p = peek()) >= "0" && p <= "7")
                ch += next(true);
      ***REMOVED***

        // Parse
        if (ch === "0") return "\0";
        if (ch.length > 0 && next_token.has_directive("use strict"))
            parse_error("Legacy octal escape sequences are not allowed in strict mode");
        return String.fromCharCode(parseInt(ch, 8));
  ***REMOVED***

    var read_string = with_eof_error("Unterminated string constant", function(quote_char) ***REMOVED***
        var quote = next(), ret = "";
        for (;;) ***REMOVED***
            var ch = next(true, true);
            if (ch == "\\") ch = read_escaped_char(true);
            else if (NEWLINE_CHARS[ch]) parse_error("Unterminated string constant");
            else if (ch == quote) break;
            ret += ch;
      ***REMOVED***
        var tok = token("string", ret);
        tok.quote = quote_char;
        return tok;
  ***REMOVED***);

    function skip_line_comment(type) ***REMOVED***
        var regex_allowed = S.regex_allowed;
        var i = find_eol(), ret;
        if (i == -1) ***REMOVED***
            ret = S.text.substr(S.pos);
            S.pos = S.text.length;
      ***REMOVED*** else ***REMOVED***
            ret = S.text.substring(S.pos, i);
            S.pos = i;
      ***REMOVED***
        S.col = S.tokcol + (S.pos - S.tokpos);
        S.comments_before.push(token(type, ret, true));
        S.regex_allowed = regex_allowed;
        return next_token;
  ***REMOVED***

    var skip_multiline_comment = with_eof_error("Unterminated multiline comment", function() ***REMOVED***
        var regex_allowed = S.regex_allowed;
        var i = find("*/", true);
        var text = S.text.substring(S.pos, i).replace(/\r\n|\r|\u2028|\u2029/g, "\n");
        // update stream position
        forward(text.length /* doesn't count \r\n as 2 char while S.pos - i does */ + 2);
        S.comments_before.push(token("comment2", text, true));
        S.regex_allowed = regex_allowed;
        return next_token;
  ***REMOVED***);

    function read_name() ***REMOVED***
        var backslash = false, ch, escaped = false, name = peek() == "#" ? next() : "";
        while (ch = peek()) ***REMOVED***
            if (!backslash) ***REMOVED***
                if (ch == "\\") escaped = backslash = true, next();
                else if (is_identifier_char(ch)) name += next();
                else break;
          ***REMOVED*** else ***REMOVED***
                if (ch != "u") parse_error("Expecting UnicodeEscapeSequence -- uXXXX");
                ch = read_escaped_char();
                if (!is_identifier_char(ch)) parse_error("Unicode char: " + ch.charCodeAt(0) + " is not valid in identifier");
                name += ch;
                backslash = false;
          ***REMOVED***
      ***REMOVED***
        if (KEYWORDS[name] && escaped) ***REMOVED***
            var hex = name.charCodeAt(0).toString(16).toUpperCase();
            name = "\\u" + "0000".substr(hex.length) + hex + name.slice(1);
      ***REMOVED***
        return name;
  ***REMOVED***

    var read_regexp = with_eof_error("Unterminated regular expression", function(source) ***REMOVED***
        var prev_backslash = false, ch, in_class = false;
        while ((ch = next(true))) if (NEWLINE_CHARS[ch]) ***REMOVED***
            parse_error("Unexpected line terminator");
      ***REMOVED*** else if (prev_backslash) ***REMOVED***
            source += "\\" + ch;
            prev_backslash = false;
      ***REMOVED*** else if (ch == "[") ***REMOVED***
            in_class = true;
            source += ch;
      ***REMOVED*** else if (ch == "]" && in_class) ***REMOVED***
            in_class = false;
            source += ch;
      ***REMOVED*** else if (ch == "/" && !in_class) ***REMOVED***
            break;
      ***REMOVED*** else if (ch == "\\") ***REMOVED***
            prev_backslash = true;
      ***REMOVED*** else ***REMOVED***
            source += ch;
      ***REMOVED***
        var mods = read_name();
        try ***REMOVED***
            var regexp = new RegExp(source, mods);
            regexp.raw_source = source;
            return token("regexp", regexp);
      ***REMOVED*** catch (e) ***REMOVED***
            parse_error(e.message);
      ***REMOVED***
  ***REMOVED***);

    function read_operator(prefix) ***REMOVED***
        function grow(op) ***REMOVED***
            if (!peek()) return op;
            var bigger = op + peek();
            if (OPERATORS[bigger]) ***REMOVED***
                next();
                return grow(bigger);
          ***REMOVED*** else ***REMOVED***
                return op;
          ***REMOVED***
      ***REMOVED***
        return token("operator", grow(prefix || next()));
  ***REMOVED***

    function handle_slash() ***REMOVED***
        next();
        switch (peek()) ***REMOVED***
          case "/":
            next();
            return skip_line_comment("comment1");
          case "*":
            next();
            return skip_multiline_comment();
      ***REMOVED***
        return S.regex_allowed ? read_regexp("") : read_operator("/");
  ***REMOVED***

    function handle_dot() ***REMOVED***
        next();
        if (looking_at("..")) return token("operator", "." + next() + next());
        return is_digit(peek().charCodeAt(0)) ? read_num(".") : token("punc", ".");
  ***REMOVED***

    function read_word() ***REMOVED***
        var word = read_name();
        if (prev_was_dot) return token("name", word);
        return KEYWORDS_ATOM[word] ? token("atom", word)
            : !KEYWORDS[word] ? token("name", word)
            : OPERATORS[word] ? token("operator", word)
            : token("keyword", word);
  ***REMOVED***

    function with_eof_error(eof_error, cont) ***REMOVED***
        return function(x) ***REMOVED***
            try ***REMOVED***
                return cont(x);
          ***REMOVED*** catch (ex) ***REMOVED***
                if (ex === EX_EOF) parse_error(eof_error);
                else throw ex;
          ***REMOVED***
      ***REMOVED***;
  ***REMOVED***

    function next_token(force_regexp) ***REMOVED***
        if (force_regexp != null)
            return read_regexp(force_regexp);
        if (shebang && S.pos == 0 && looking_at("#!")) ***REMOVED***
            start_token();
            forward(2);
            skip_line_comment("comment5");
      ***REMOVED***
        for (;;) ***REMOVED***
            skip_whitespace();
            start_token();
            if (html5_comments) ***REMOVED***
                if (looking_at("<!--")) ***REMOVED***
                    forward(4);
                    skip_line_comment("comment3");
                    continue;
              ***REMOVED***
                if (looking_at("-->") && S.newline_before) ***REMOVED***
                    forward(3);
                    skip_line_comment("comment4");
                    continue;
              ***REMOVED***
          ***REMOVED***
            var ch = peek();
            if (!ch) return token("eof");
            var code = ch.charCodeAt(0);
            switch (code) ***REMOVED***
              case 34: case 39: return read_string(ch);
              case 46: return handle_dot();
              case 47:
                var tok = handle_slash();
                if (tok === next_token) continue;
                return tok;
          ***REMOVED***
            if (is_digit(code)) return read_num();
            if (PUNC_CHARS[ch]) return token("punc", next());
            if (looking_at("=>")) return token("punc", next() + next());
            if (OPERATOR_CHARS[ch]) return read_operator();
            if (code == 35 || code == 92 || !NON_IDENTIFIER_CHARS[ch]) return read_word();
            break;
      ***REMOVED***
        parse_error("Unexpected character '" + ch + "'");
  ***REMOVED***

    next_token.context = function(nc) ***REMOVED***
        if (nc) S = nc;
        return S;
  ***REMOVED***;

    next_token.add_directive = function(directive) ***REMOVED***
        S.directives[directive] = true;
  ***REMOVED***

    next_token.push_directives_stack = function() ***REMOVED***
        S.directives = Object.create(S.directives);
  ***REMOVED***

    next_token.pop_directives_stack = function() ***REMOVED***
        S.directives = Object.getPrototypeOf(S.directives);
  ***REMOVED***

    next_token.has_directive = function(directive) ***REMOVED***
        return !!S.directives[directive];
  ***REMOVED***

    return next_token;
}

/* -----[ Parser (constants) ]----- */

var UNARY_PREFIX = makePredicate("typeof void delete -- ++ ! ~ - +");

var UNARY_POSTFIX = makePredicate("-- ++");

var ASSIGNMENT = makePredicate("= += -= /= *= %= **= >>= <<= >>>= &= |= ^= &&= ||= ??=");

var PRECEDENCE = function(a, ret) ***REMOVED***
    for (var i = 0; i < a.length;) ***REMOVED***
        var b = a[i++];
        for (var j = 0; j < b.length; j++) ***REMOVED***
            ret[b[j]] = i;
      ***REMOVED***
  ***REMOVED***
    return ret;
}([
    ["??"],
    ["||"],
    ["&&"],
    ["|"],
    ["^"],
    ["&"],
    ["==", "===", "!=", "!=="],
    ["<", ">", "<=", ">=", "in", "instanceof"],
    [">>", "<<", ">>>"],
    ["+", "-"],
    ["*", "/", "%"],
    ["**"],
], ***REMOVED***});

var ATOMIC_START_TOKEN = makePredicate("atom bigint num regexp string");

/* -----[ Parser ]----- */

function parse($TEXT, options) ***REMOVED***
    options = defaults(options, ***REMOVED***
        bare_returns   : false,
        expression     : false,
        filename       : null,
        html5_comments : true,
        module         : false,
        shebang        : true,
        strict         : false,
        toplevel       : null,
  ***REMOVED***, true);

    var S = ***REMOVED***
        input         : typeof $TEXT == "string"
                        ? tokenizer($TEXT, options.filename, options.html5_comments, options.shebang)
                        : $TEXT,
        in_async      : false,
        in_directives : true,
        in_funarg     : -1,
        in_function   : 0,
        in_generator  : false,
        in_loop       : 0,
        labels        : [],
        peeked        : null,
        prev          : null,
        token         : null,
  ***REMOVED***;

    S.token = next();

    function is(type, value) ***REMOVED***
        return is_token(S.token, type, value);
  ***REMOVED***

    function peek() ***REMOVED***
        return S.peeked || (S.peeked = S.input());
  ***REMOVED***

    function next() ***REMOVED***
        S.prev = S.token;
        if (S.peeked) ***REMOVED***
            S.token = S.peeked;
            S.peeked = null;
      ***REMOVED*** else ***REMOVED***
            S.token = S.input();
      ***REMOVED***
        S.in_directives = S.in_directives && (
            S.token.type == "string" || is("punc", ";")
        );
        return S.token;
  ***REMOVED***

    function prev() ***REMOVED***
        return S.prev;
  ***REMOVED***

    function croak(msg, line, col, pos) ***REMOVED***
        var ctx = S.input.context();
        js_error(msg,
                 ctx.filename,
                 line != null ? line : ctx.tokline,
                 col != null ? col : ctx.tokcol,
                 pos != null ? pos : ctx.tokpos);
  ***REMOVED***

    function token_error(token, msg) ***REMOVED***
        croak(msg, token.line, token.col);
  ***REMOVED***

    function token_to_string(type, value) ***REMOVED***
        return type + (value === undefined ? "" : " «" + value + "»");
  ***REMOVED***

    function unexpected(token) ***REMOVED***
        if (token == null) token = S.token;
        token_error(token, "Unexpected token: " + token_to_string(token.type, token.value));
  ***REMOVED***

    function expect_token(type, val) ***REMOVED***
        if (is(type, val)) return next();
        token_error(S.token, "Unexpected token: " + token_to_string(S.token.type, S.token.value) + ", expected: " + token_to_string(type, val));
  ***REMOVED***

    function expect(punc) ***REMOVED***
        return expect_token("punc", punc);
  ***REMOVED***

    function has_newline_before(token) ***REMOVED***
        return token.nlb || !all(token.comments_before, function(comment) ***REMOVED***
            return !comment.nlb;
      ***REMOVED***);
  ***REMOVED***

    function can_insert_semicolon() ***REMOVED***
        return !options.strict
            && (is("eof") || is("punc", "}") || has_newline_before(S.token));
  ***REMOVED***

    function semicolon(optional) ***REMOVED***
        if (is("punc", ";")) next();
        else if (!optional && !can_insert_semicolon()) expect(";");
  ***REMOVED***

    function parenthesized() ***REMOVED***
        expect("(");
        var exp = expression();
        expect(")");
        return exp;
  ***REMOVED***

    function embed_tokens(parser) ***REMOVED***
        return function() ***REMOVED***
            var start = S.token;
            var expr = parser.apply(null, arguments);
            var end = prev();
            expr.start = start;
            expr.end = end;
            return expr;
      ***REMOVED***;
  ***REMOVED***

    function handle_regexp() ***REMOVED***
        if (is("operator", "/") || is("operator", "/=")) ***REMOVED***
            S.peeked = null;
            S.token = S.input(S.token.value.substr(1)); // force regexp
      ***REMOVED***
  ***REMOVED***

    var statement = embed_tokens(function(toplevel) ***REMOVED***
        handle_regexp();
        switch (S.token.type) ***REMOVED***
          case "string":
            var dir = S.in_directives;
            var body = expression();
            if (dir) ***REMOVED***
                if (body instanceof AST_String) ***REMOVED***
                    var value = body.start.raw.slice(1, -1);
                    S.input.add_directive(value);
                    body.value = value;
              ***REMOVED*** else ***REMOVED***
                    S.in_directives = dir = false;
              ***REMOVED***
          ***REMOVED***
            semicolon();
            return dir ? new AST_Directive(body) : new AST_SimpleStatement(***REMOVED*** body: body });
          case "num":
          case "bigint":
          case "regexp":
          case "operator":
          case "atom":
            return simple_statement();

          case "name":
            switch (S.token.value) ***REMOVED***
              case "async":
                if (is_token(peek(), "keyword", "function")) ***REMOVED***
                    next();
                    next();
                    if (!is("operator", "*")) return function_(AST_AsyncDefun);
                    next();
                    return function_(AST_AsyncGeneratorDefun);
              ***REMOVED***
                break;
              case "await":
                if (S.in_async) return simple_statement();
                break;
              case "export":
                if (!toplevel && options.module !== "") unexpected();
                next();
                return export_();
              case "import":
                var token = peek();
                if (token.type == "punc" && /^[(.]$/.test(token.value)) break;
                if (!toplevel && options.module !== "") unexpected();
                next();
                return import_();
              case "let":
                if (is_vardefs()) ***REMOVED***
                    next();
                    var node = let_();
                    semicolon();
                    return node;
              ***REMOVED***
                break;
              case "yield":
                if (S.in_generator) return simple_statement();
                break;
          ***REMOVED***
            return is_token(peek(), "punc", ":")
                ? labeled_statement()
                : simple_statement();

          case "punc":
            switch (S.token.value) ***REMOVED***
              case "***REMOVED***":
                return new AST_BlockStatement(***REMOVED***
                    start : S.token,
                    body  : block_(),
                    end   : prev()
              ***REMOVED***);
              case "[":
              case "(":
              case "`":
                return simple_statement();
              case ";":
                S.in_directives = false;
                next();
                return new AST_EmptyStatement();
              default:
                unexpected();
          ***REMOVED***

          case "keyword":
            switch (S.token.value) ***REMOVED***
              case "break":
                next();
                return break_cont(AST_Break);

              case "class":
                next();
                return class_(AST_DefClass);

              case "const":
                next();
                var node = const_();
                semicolon();
                return node;

              case "continue":
                next();
                return break_cont(AST_Continue);

              case "debugger":
                next();
                semicolon();
                return new AST_Debugger();

              case "do":
                next();
                var body = in_loop(statement);
                expect_token("keyword", "while");
                var condition = parenthesized();
                semicolon(true);
                return new AST_Do(***REMOVED***
                    body      : body,
                    condition : condition,
              ***REMOVED***);

              case "while":
                next();
                return new AST_While(***REMOVED***
                    condition : parenthesized(),
                    body      : in_loop(statement),
              ***REMOVED***);

              case "for":
                next();
                return for_();

              case "function":
                next();
                if (!is("operator", "*")) return function_(AST_Defun);
                next();
                return function_(AST_GeneratorDefun);

              case "if":
                next();
                return if_();

              case "return":
                if (S.in_function == 0 && !options.bare_returns)
                    croak("'return' outside of function");
                next();
                var value = null;
                if (is("punc", ";")) ***REMOVED***
                    next();
              ***REMOVED*** else if (!can_insert_semicolon()) ***REMOVED***
                    value = expression();
                    semicolon();
              ***REMOVED***
                return new AST_Return(***REMOVED*** value: value });

              case "switch":
                next();
                return new AST_Switch(***REMOVED***
                    expression : parenthesized(),
                    body       : in_loop(switch_body_),
              ***REMOVED***);

              case "throw":
                next();
                if (has_newline_before(S.token))
                    croak("Illegal newline after 'throw'");
                var value = expression();
                semicolon();
                return new AST_Throw(***REMOVED*** value: value });

              case "try":
                next();
                return try_();

              case "var":
                next();
                var node = var_();
                semicolon();
                return node;

              case "with":
                if (S.input.has_directive("use strict")) ***REMOVED***
                    croak("Strict mode may not include a with statement");
              ***REMOVED***
                next();
                return new AST_With(***REMOVED***
                    expression : parenthesized(),
                    body       : statement(),
              ***REMOVED***);
          ***REMOVED***
      ***REMOVED***
        unexpected();
  ***REMOVED***);

    function labeled_statement() ***REMOVED***
        var label = as_symbol(AST_Label);
        if (!all(S.labels, function(l) ***REMOVED***
            return l.name != label.name;
      ***REMOVED***)) ***REMOVED***
            // ECMA-262, 12.12: An ECMAScript program is considered
            // syntactically incorrect if it contains a
            // LabelledStatement that is enclosed by a
            // LabelledStatement with the same Identifier as label.
            croak("Label " + label.name + " defined twice");
      ***REMOVED***
        expect(":");
        S.labels.push(label);
        var stat = statement();
        S.labels.pop();
        if (!(stat instanceof AST_IterationStatement)) ***REMOVED***
            // check for `continue` that refers to this label.
            // those should be reported as syntax errors.
            // https://github.com/mishoo/UglifyJS/issues/287
            label.references.forEach(function(ref) ***REMOVED***
                if (ref instanceof AST_Continue) ***REMOVED***
                    token_error(ref.label.start, "Continue label `" + label.name + "` must refer to IterationStatement");
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***
        return new AST_LabeledStatement(***REMOVED*** body: stat, label: label });
  ***REMOVED***

    function simple_statement() ***REMOVED***
        var body = expression();
        semicolon();
        return new AST_SimpleStatement(***REMOVED*** body: body });
  ***REMOVED***

    function break_cont(type) ***REMOVED***
        var label = null, ldef;
        if (!can_insert_semicolon()) ***REMOVED***
            label = as_symbol(AST_LabelRef, true);
      ***REMOVED***
        if (label != null) ***REMOVED***
            ldef = find_if(function(l) ***REMOVED***
                return l.name == label.name;
          ***REMOVED***, S.labels);
            if (!ldef) token_error(label.start, "Undefined label " + label.name);
            label.thedef = ldef;
      ***REMOVED*** else if (S.in_loop == 0) croak(type.TYPE + " not inside a loop or switch");
        semicolon();
        var stat = new type(***REMOVED*** label: label });
        if (ldef) ldef.references.push(stat);
        return stat;
  ***REMOVED***

    function has_modifier(name, no_nlb) ***REMOVED***
        if (!is("name", name)) return;
        var token = peek();
        if (!token) return;
        if (is_token(token, "operator", "=")) return;
        if (token.type == "punc" && /^[(;}]$/.test(token.value)) return;
        if (no_nlb && has_newline_before(token)) return;
        return next();
  ***REMOVED***

    function class_(ctor) ***REMOVED***
        var was_async = S.in_async;
        var was_gen = S.in_generator;
        S.input.push_directives_stack();
        S.input.add_directive("use strict");
        var name;
        if (ctor === AST_DefClass) ***REMOVED***
            name = as_symbol(AST_SymbolDefClass);
      ***REMOVED*** else ***REMOVED***
            name = as_symbol(AST_SymbolClass, true);
      ***REMOVED***
        var parent = null;
        if (is("keyword", "extends")) ***REMOVED***
            next();
            handle_regexp();
            parent = expr_atom(true);
      ***REMOVED***
        expect("***REMOVED***");
        var props = [];
        while (!is("punc", "}")) ***REMOVED***
            if (is("punc", ";")) ***REMOVED***
                next();
                continue;
          ***REMOVED***
            var start = S.token;
            var fixed = !!has_modifier("static");
            var async = has_modifier("async", true);
            if (is("operator", "*")) ***REMOVED***
                next();
                var internal = is("name") && /^#/.test(S.token.value);
                var key = as_property_key();
                var gen_start = S.token;
                var gen = function_(async ? AST_AsyncGeneratorFunction : AST_GeneratorFunction);
                gen.start = gen_start;
                gen.end = prev();
                props.push(new AST_ClassMethod(***REMOVED***
                    start: start,
                    static: fixed,
                    private: internal,
                    key: key,
                    value: gen,
                    end: prev(),
              ***REMOVED***));
                continue;
          ***REMOVED***
            if (fixed && is("punc", "***REMOVED***")) ***REMOVED***
                props.push(new AST_ClassInit(***REMOVED***
                    start: start,
                    value: new AST_ClassInitBlock(***REMOVED***
                        start: start,
                        body: block_(),
                        end: prev(),
                  ***REMOVED***),
                    end: prev(),
              ***REMOVED***));
                continue;
          ***REMOVED***
            var internal = is("name") && /^#/.test(S.token.value);
            var key = as_property_key();
            if (is("punc", "(")) ***REMOVED***
                var func_start = S.token;
                var func = function_(async ? AST_AsyncFunction : AST_Function);
                func.start = func_start;
                func.end = prev();
                props.push(new AST_ClassMethod(***REMOVED***
                    start: start,
                    static: fixed,
                    private: internal,
                    key: key,
                    value: func,
                    end: prev(),
              ***REMOVED***));
                continue;
          ***REMOVED***
            if (async) unexpected(async);
            var value = null;
            if (is("operator", "=")) ***REMOVED***
                next();
                S.in_async = false;
                S.in_generator = false;
                value = maybe_assign();
                S.in_generator = was_gen;
                S.in_async = was_async;
          ***REMOVED*** else if (!(is("punc", ";") || is("punc", "}"))) ***REMOVED***
                var type = null;
                switch (key) ***REMOVED***
                  case "get":
                    type = AST_ClassGetter;
                    break;
                  case "set":
                    type = AST_ClassSetter;
                    break;
              ***REMOVED***
                if (type) ***REMOVED***
                    props.push(new type(***REMOVED***
                        start: start,
                        static: fixed,
                        private: is("name") && /^#/.test(S.token.value),
                        key: as_property_key(),
                        value: create_accessor(),
                        end: prev(),
                  ***REMOVED***));
                    continue;
              ***REMOVED***
          ***REMOVED***
            semicolon();
            props.push(new AST_ClassField(***REMOVED***
                start: start,
                static: fixed,
                private: internal,
                key: key,
                value: value,
                end: prev(),
          ***REMOVED***));
      ***REMOVED***
        next();
        S.input.pop_directives_stack();
        S.in_generator = was_gen;
        S.in_async = was_async;
        return new ctor(***REMOVED***
            extends: parent,
            name: name,
            properties: props,
      ***REMOVED***);
  ***REMOVED***

    function for_() ***REMOVED***
        var await_token = is("name", "await") && next();
        expect("(");
        var init = null;
        if (await_token || !is("punc", ";")) ***REMOVED***
            init = is("keyword", "const")
                ? (next(), const_(true))
                : is("name", "let") && is_vardefs()
                ? (next(), let_(true))
                : is("keyword", "var")
                ? (next(), var_(true))
                : expression(true);
            var ctor;
            if (await_token) ***REMOVED***
                expect_token("name", "of");
                ctor = AST_ForAwaitOf;
          ***REMOVED*** else if (is("operator", "in")) ***REMOVED***
                next();
                ctor = AST_ForIn;
          ***REMOVED*** else if (is("name", "of")) ***REMOVED***
                next();
                ctor = AST_ForOf;
          ***REMOVED***
            if (ctor) ***REMOVED***
                if (init instanceof AST_Definitions) ***REMOVED***
                    if (init.definitions.length > 1) ***REMOVED***
                        token_error(init.start, "Only one variable declaration allowed in for..in/of loop");
                  ***REMOVED***
                    if (ctor !== AST_ForIn && init.definitions[0].value) ***REMOVED***
                        token_error(init.definitions[0].value.start, "No initializers allowed in for..of loop");
                  ***REMOVED***
              ***REMOVED*** else if (!(is_assignable(init) || (init = to_destructured(init)) instanceof AST_Destructured)) ***REMOVED***
                    token_error(init.start, "Invalid left-hand side in for..in/of loop");
              ***REMOVED***
                return for_enum(ctor, init);
          ***REMOVED***
      ***REMOVED***
        return regular_for(init);
  ***REMOVED***

    function regular_for(init) ***REMOVED***
        expect(";");
        var test = is("punc", ";") ? null : expression();
        expect(";");
        var step = is("punc", ")") ? null : expression();
        expect(")");
        return new AST_For(***REMOVED***
            init      : init,
            condition : test,
            step      : step,
            body      : in_loop(statement)
      ***REMOVED***);
  ***REMOVED***

    function for_enum(ctor, init) ***REMOVED***
        handle_regexp();
        var obj = expression();
        expect(")");
        return new ctor(***REMOVED***
            init   : init,
            object : obj,
            body   : in_loop(statement)
      ***REMOVED***);
  ***REMOVED***

    function to_funarg(node) ***REMOVED***
        if (node instanceof AST_Array) ***REMOVED***
            var rest = null;
            if (node.elements[node.elements.length - 1] instanceof AST_Spread) ***REMOVED***
                rest = to_funarg(node.elements.pop().expression);
          ***REMOVED***
            return new AST_DestructuredArray(***REMOVED***
                start: node.start,
                elements: node.elements.map(to_funarg),
                rest: rest,
                end: node.end,
          ***REMOVED***);
      ***REMOVED***
        if (node instanceof AST_Assign) return new AST_DefaultValue(***REMOVED***
            start: node.start,
            name: to_funarg(node.left),
            value: node.right,
            end: node.end,
      ***REMOVED***);
        if (node instanceof AST_DefaultValue) ***REMOVED***
            node.name = to_funarg(node.name);
            return node;
      ***REMOVED***
        if (node instanceof AST_DestructuredArray) ***REMOVED***
            node.elements = node.elements.map(to_funarg);
            if (node.rest) node.rest = to_funarg(node.rest);
            return node;
      ***REMOVED***
        if (node instanceof AST_DestructuredObject) ***REMOVED***
            node.properties.forEach(function(prop) ***REMOVED***
                prop.value = to_funarg(prop.value);
          ***REMOVED***);
            if (node.rest) node.rest = to_funarg(node.rest);
            return node;
      ***REMOVED***
        if (node instanceof AST_Hole) return node;
        if (node instanceof AST_Object) ***REMOVED***
            var rest = null;
            if (node.properties[node.properties.length - 1] instanceof AST_Spread) ***REMOVED***
                rest = to_funarg(node.properties.pop().expression);
          ***REMOVED***
            return new AST_DestructuredObject(***REMOVED***
                start: node.start,
                properties: node.properties.map(function(prop) ***REMOVED***
                    if (!(prop instanceof AST_ObjectKeyVal)) token_error(prop.start, "Invalid destructuring assignment");
                    return new AST_DestructuredKeyVal(***REMOVED***
                        start: prop.start,
                        key: prop.key,
                        value: to_funarg(prop.value),
                        end: prop.end,
                  ***REMOVED***);
              ***REMOVED***),
                rest: rest,
                end: node.end,
          ***REMOVED***);
      ***REMOVED***
        if (node instanceof AST_SymbolFunarg) return node;
        if (node instanceof AST_SymbolRef) return new AST_SymbolFunarg(node);
        if (node instanceof AST_Yield) return new AST_SymbolFunarg(***REMOVED***
            start: node.start,
            name: "yield",
            end: node.end,
      ***REMOVED***);
        token_error(node.start, "Invalid arrow parameter");
  ***REMOVED***

    function arrow(exprs, start, async) ***REMOVED***
        var was_async = S.in_async;
        var was_gen = S.in_generator;
        S.in_async = async;
        S.in_generator = false;
        var was_funarg = S.in_funarg;
        S.in_funarg = S.in_function;
        var argnames = exprs.map(to_funarg);
        var rest = exprs.rest || null;
        if (rest) rest = to_funarg(rest);
        S.in_funarg = was_funarg;
        expect("=>");
        var body, value;
        var loop = S.in_loop;
        var labels = S.labels;
        ++S.in_function;
        S.input.push_directives_stack();
        S.in_loop = 0;
        S.labels = [];
        if (is("punc", "***REMOVED***")) ***REMOVED***
            S.in_directives = true;
            body = block_();
            value = null;
      ***REMOVED*** else ***REMOVED***
            body = [];
            handle_regexp();
            value = maybe_assign();
      ***REMOVED***
        var is_strict = S.input.has_directive("use strict");
        S.input.pop_directives_stack();
        --S.in_function;
        S.in_loop = loop;
        S.labels = labels;
        S.in_generator = was_gen;
        S.in_async = was_async;
        var node = new (async ? AST_AsyncArrow : AST_Arrow)(***REMOVED***
            start: start,
            argnames: argnames,
            rest: rest,
            body: body,
            value: value,
            end: prev(),
      ***REMOVED***);
        if (is_strict) node.each_argname(strict_verify_symbol);
        return node;
  ***REMOVED***

    var function_ = function(ctor) ***REMOVED***
        var was_async = S.in_async;
        var was_gen = S.in_generator;
        var name;
        if (/Defun$/.test(ctor.TYPE)) ***REMOVED***
            name = as_symbol(AST_SymbolDefun);
            S.in_async = /^Async/.test(ctor.TYPE);
            S.in_generator = /Generator/.test(ctor.TYPE);
      ***REMOVED*** else ***REMOVED***
            S.in_async = /^Async/.test(ctor.TYPE);
            S.in_generator = /Generator/.test(ctor.TYPE);
            name = as_symbol(AST_SymbolLambda, true);
      ***REMOVED***
        if (name && ctor !== AST_Accessor && !(name instanceof AST_SymbolDeclaration))
            unexpected(prev());
        expect("(");
        var was_funarg = S.in_funarg;
        S.in_funarg = S.in_function;
        var argnames = expr_list(")", !options.strict, false, function() ***REMOVED***
            return maybe_default(AST_SymbolFunarg);
      ***REMOVED***);
        S.in_funarg = was_funarg;
        var loop = S.in_loop;
        var labels = S.labels;
        ++S.in_function;
        S.in_directives = true;
        S.input.push_directives_stack();
        S.in_loop = 0;
        S.labels = [];
        var body = block_();
        var is_strict = S.input.has_directive("use strict");
        S.input.pop_directives_stack();
        --S.in_function;
        S.in_loop = loop;
        S.labels = labels;
        S.in_generator = was_gen;
        S.in_async = was_async;
        var node = new ctor(***REMOVED***
            name: name,
            argnames: argnames,
            rest: argnames.rest || null,
            body: body,
      ***REMOVED***);
        if (is_strict) ***REMOVED***
            if (name) strict_verify_symbol(name);
            node.each_argname(strict_verify_symbol);
      ***REMOVED***
        return node;
  ***REMOVED***;

    function if_() ***REMOVED***
        var cond = parenthesized(), body = statement(), alt = null;
        if (is("keyword", "else")) ***REMOVED***
            next();
            alt = statement();
      ***REMOVED***
        return new AST_If(***REMOVED***
            condition   : cond,
            body        : body,
            alternative : alt,
      ***REMOVED***);
  ***REMOVED***

    function is_alias() ***REMOVED***
        return is("name") || is("string") || is_identifier_string(S.token.value);
  ***REMOVED***

    function make_string(token) ***REMOVED***
        return new AST_String(***REMOVED***
            start: token,
            quote: token.quote,
            value: token.value,
            end: token,
      ***REMOVED***);
  ***REMOVED***

    function as_path() ***REMOVED***
        var path = S.token;
        expect_token("string");
        semicolon();
        return make_string(path);
  ***REMOVED***

    function export_() ***REMOVED***
        if (is("operator", "*")) ***REMOVED***
            var key = S.token;
            var alias = key;
            next();
            if (is("name", "as")) ***REMOVED***
                next();
                if (!is_alias()) expect_token("name");
                alias = S.token;
                next();
          ***REMOVED***
            expect_token("name", "from");
            return new AST_ExportForeign(***REMOVED***
                aliases: [ make_string(alias) ],
                keys: [ make_string(key) ],
                path: as_path(),
          ***REMOVED***);
      ***REMOVED***
        if (is("punc", "***REMOVED***")) ***REMOVED***
            next();
            var aliases = [];
            var keys = [];
            while (is_alias()) ***REMOVED***
                var key = S.token;
                next();
                keys.push(key);
                if (is("name", "as")) ***REMOVED***
                    next();
                    if (!is_alias()) expect_token("name");
                    aliases.push(S.token);
                    next();
              ***REMOVED*** else ***REMOVED***
                    aliases.push(key);
              ***REMOVED***
                if (!is("punc", "}")) expect(",");
          ***REMOVED***
            expect("}");
            if (is("name", "from")) ***REMOVED***
                next();
                return new AST_ExportForeign(***REMOVED***
                    aliases: aliases.map(make_string),
                    keys: keys.map(make_string),
                    path: as_path(),
              ***REMOVED***);
          ***REMOVED***
            semicolon();
            return new AST_ExportReferences(***REMOVED***
                properties: keys.map(function(token, index) ***REMOVED***
                    if (!is_token(token, "name")) token_error(token, "Name expected");
                    var sym = _make_symbol(AST_SymbolExport, token);
                    sym.alias = make_string(aliases[index]);
                    return sym;
              ***REMOVED***),
          ***REMOVED***);
      ***REMOVED***
        if (is("keyword", "default")) ***REMOVED***
            next();
            var start = S.token;
            var body = export_default_decl();
            if (body) ***REMOVED***
                body.start = start;
                body.end = prev();
          ***REMOVED*** else ***REMOVED***
                handle_regexp();
                body = expression();
                semicolon();
          ***REMOVED***
            return new AST_ExportDefault(***REMOVED*** body: body });
      ***REMOVED***
        return new AST_ExportDeclaration(***REMOVED*** body: export_decl() });
  ***REMOVED***

    function maybe_named(def, expr) ***REMOVED***
        if (expr.name) ***REMOVED***
            expr = new def(expr);
            expr.name = new (def === AST_DefClass ? AST_SymbolDefClass : AST_SymbolDefun)(expr.name);
      ***REMOVED***
        return expr;
  ***REMOVED***

    function export_default_decl() ***REMOVED***
        if (is("name", "async")) ***REMOVED***
            if (!is_token(peek(), "keyword", "function")) return;
            next();
            next();
            if (!is("operator", "*")) return maybe_named(AST_AsyncDefun, function_(AST_AsyncFunction));
            next();
            return maybe_named(AST_AsyncGeneratorDefun, function_(AST_AsyncGeneratorFunction));
      ***REMOVED*** else if (is("keyword")) switch (S.token.value) ***REMOVED***
          case "class":
            next();
            return maybe_named(AST_DefClass, class_(AST_ClassExpression));
          case "function":
            next();
            if (!is("operator", "*")) return maybe_named(AST_Defun, function_(AST_Function));
            next();
            return maybe_named(AST_GeneratorDefun, function_(AST_GeneratorFunction));
      ***REMOVED***
  ***REMOVED***

    var export_decl = embed_tokens(function() ***REMOVED***
        if (is("name")) switch (S.token.value) ***REMOVED***
          case "async":
            next();
            expect_token("keyword", "function");
            if (!is("operator", "*")) return function_(AST_AsyncDefun);
            next();
            return function_(AST_AsyncGeneratorDefun);
          case "let":
            next();
            var node = let_();
            semicolon();
            return node;
      ***REMOVED*** else if (is("keyword")) switch (S.token.value) ***REMOVED***
          case "class":
            next();
            return class_(AST_DefClass);
          case "const":
            next();
            var node = const_();
            semicolon();
            return node;
          case "function":
            next();
            if (!is("operator", "*")) return function_(AST_Defun);
            next();
            return function_(AST_GeneratorDefun);
          case "var":
            next();
            var node = var_();
            semicolon();
            return node;
      ***REMOVED***
        unexpected();
  ***REMOVED***);

    function import_() ***REMOVED***
        var all = null;
        var def = as_symbol(AST_SymbolImport, true);
        var props = null;
        var cont;
        if (def) ***REMOVED***
            def.key = new AST_String(***REMOVED***
                start: def.start,
                value: "",
                end: def.end,
          ***REMOVED***);
            if (cont = is("punc", ",")) next();
      ***REMOVED*** else ***REMOVED***
            cont = !is("string");
      ***REMOVED***
        if (cont) ***REMOVED***
            if (is("operator", "*")) ***REMOVED***
                var key = S.token;
                next();
                expect_token("name", "as");
                all = as_symbol(AST_SymbolImport);
                all.key = make_string(key);
          ***REMOVED*** else ***REMOVED***
                expect("***REMOVED***");
                props = [];
                while (is_alias()) ***REMOVED***
                    var alias;
                    if (is_token(peek(), "name", "as")) ***REMOVED***
                        var key = S.token;
                        next();
                        next();
                        alias = as_symbol(AST_SymbolImport);
                        alias.key = make_string(key);
                  ***REMOVED*** else ***REMOVED***
                        alias = as_symbol(AST_SymbolImport);
                        alias.key = new AST_String(***REMOVED***
                            start: alias.start,
                            value: alias.name,
                            end: alias.end,
                      ***REMOVED***);
                  ***REMOVED***
                    props.push(alias);
                    if (!is("punc", "}")) expect(",");
              ***REMOVED***
                expect("}");
          ***REMOVED***
      ***REMOVED***
        if (all || def || props) expect_token("name", "from");
        return new AST_Import(***REMOVED***
            all: all,
            default: def,
            path: as_path(),
            properties: props,
      ***REMOVED***);
  ***REMOVED***

    function block_() ***REMOVED***
        expect("***REMOVED***");
        var a = [];
        while (!is("punc", "}")) ***REMOVED***
            if (is("eof")) expect("}");
            a.push(statement());
      ***REMOVED***
        next();
        return a;
  ***REMOVED***

    function switch_body_() ***REMOVED***
        expect("***REMOVED***");
        var a = [], branch, cur, default_branch, tmp;
        while (!is("punc", "}")) ***REMOVED***
            if (is("eof")) expect("}");
            if (is("keyword", "case")) ***REMOVED***
                if (branch) branch.end = prev();
                cur = [];
                branch = new AST_Case(***REMOVED***
                    start      : (tmp = S.token, next(), tmp),
                    expression : expression(),
                    body       : cur
              ***REMOVED***);
                a.push(branch);
                expect(":");
          ***REMOVED*** else if (is("keyword", "default")) ***REMOVED***
                if (branch) branch.end = prev();
                if (default_branch) croak("More than one default clause in switch statement");
                cur = [];
                branch = new AST_Default(***REMOVED***
                    start : (tmp = S.token, next(), expect(":"), tmp),
                    body  : cur
              ***REMOVED***);
                a.push(branch);
                default_branch = branch;
          ***REMOVED*** else ***REMOVED***
                if (!cur) unexpected();
                cur.push(statement());
          ***REMOVED***
      ***REMOVED***
        if (branch) branch.end = prev();
        next();
        return a;
  ***REMOVED***

    function try_() ***REMOVED***
        var body = block_(), bcatch = null, bfinally = null;
        if (is("keyword", "catch")) ***REMOVED***
            var start = S.token;
            next();
            var name = null;
            if (is("punc", "(")) ***REMOVED***
                next();
                name = maybe_destructured(AST_SymbolCatch);
                expect(")");
          ***REMOVED***
            bcatch = new AST_Catch(***REMOVED***
                start   : start,
                argname : name,
                body    : block_(),
                end     : prev()
          ***REMOVED***);
      ***REMOVED***
        if (is("keyword", "finally")) ***REMOVED***
            var start = S.token;
            next();
            bfinally = new AST_Finally(***REMOVED***
                start : start,
                body  : block_(),
                end   : prev()
          ***REMOVED***);
      ***REMOVED***
        if (!bcatch && !bfinally)
            croak("Missing catch/finally blocks");
        return new AST_Try(***REMOVED***
            body     : body,
            bcatch   : bcatch,
            bfinally : bfinally
      ***REMOVED***);
  ***REMOVED***

    function vardefs(type, no_in) ***REMOVED***
        var a = [];
        for (;;) ***REMOVED***
            var start = S.token;
            var name = maybe_destructured(type);
            var value = null;
            if (is("operator", "=")) ***REMOVED***
                next();
                value = maybe_assign(no_in);
          ***REMOVED*** else if (!no_in && (type === AST_SymbolConst || name instanceof AST_Destructured)) ***REMOVED***
                croak("Missing initializer in declaration");
          ***REMOVED***
            a.push(new AST_VarDef(***REMOVED***
                start : start,
                name  : name,
                value : value,
                end   : prev()
          ***REMOVED***));
            if (!is("punc", ","))
                break;
            next();
      ***REMOVED***
        return a;
  ***REMOVED***

    function is_vardefs() ***REMOVED***
        var token = peek();
        return is_token(token, "name") || is_token(token, "punc", "[") || is_token(token, "punc", "***REMOVED***");
  ***REMOVED***

    var const_ = function(no_in) ***REMOVED***
        return new AST_Const(***REMOVED***
            start       : prev(),
            definitions : vardefs(AST_SymbolConst, no_in),
            end         : prev()
      ***REMOVED***);
  ***REMOVED***;

    var let_ = function(no_in) ***REMOVED***
        return new AST_Let(***REMOVED***
            start       : prev(),
            definitions : vardefs(AST_SymbolLet, no_in),
            end         : prev()
      ***REMOVED***);
  ***REMOVED***;

    var var_ = function(no_in) ***REMOVED***
        return new AST_Var(***REMOVED***
            start       : prev(),
            definitions : vardefs(AST_SymbolVar, no_in),
            end         : prev()
      ***REMOVED***);
  ***REMOVED***;

    var new_ = function(allow_calls) ***REMOVED***
        var start = S.token;
        expect_token("operator", "new");
        var call;
        if (is("punc", ".") && is_token(peek(), "name", "target")) ***REMOVED***
            next();
            next();
            call = new AST_NewTarget();
      ***REMOVED*** else ***REMOVED***
            var exp = expr_atom(false), args;
            if (is("punc", "(")) ***REMOVED***
                next();
                args = expr_list(")", !options.strict);
          ***REMOVED*** else ***REMOVED***
                args = [];
          ***REMOVED***
            call = new AST_New(***REMOVED*** expression: exp, args: args });
      ***REMOVED***
        call.start = start;
        call.end = prev();
        return subscripts(call, allow_calls);
  ***REMOVED***;

    function as_atom_node() ***REMOVED***
        var ret, tok = S.token, value = tok.value;
        switch (tok.type) ***REMOVED***
          case "num":
            if (isFinite(value)) ***REMOVED***
                ret = new AST_Number(***REMOVED*** value: value });
          ***REMOVED*** else ***REMOVED***
                ret = new AST_Infinity();
                if (value < 0) ret = new AST_UnaryPrefix(***REMOVED*** operator: "-", expression: ret });
          ***REMOVED***
            break;
          case "bigint":
            ret = new AST_BigInt(***REMOVED*** value: value });
            break;
          case "string":
            ret = new AST_String(***REMOVED*** value: value, quote: tok.quote });
            break;
          case "regexp":
            ret = new AST_RegExp(***REMOVED*** value: value });
            break;
          case "atom":
            switch (value) ***REMOVED***
              case "false":
                ret = new AST_False();
                break;
              case "true":
                ret = new AST_True();
                break;
              case "null":
                ret = new AST_Null();
                break;
              default:
                unexpected();
          ***REMOVED***
            break;
          default:
            unexpected();
      ***REMOVED***
        next();
        ret.start = ret.end = tok;
        return ret;
  ***REMOVED***

    var expr_atom = function(allow_calls) ***REMOVED***
        if (is("operator", "new")) ***REMOVED***
            return new_(allow_calls);
      ***REMOVED***
        var start = S.token;
        if (is("punc")) ***REMOVED***
            switch (start.value) ***REMOVED***
              case "`":
                return subscripts(template(null), allow_calls);
              case "(":
                next();
                if (is("punc", ")")) ***REMOVED***
                    next();
                    return arrow([], start);
              ***REMOVED***
                var ex = expression(false, true);
                var len = start.comments_before.length;
                [].unshift.apply(ex.start.comments_before, start.comments_before);
                start.comments_before.length = 0;
                start.comments_before = ex.start.comments_before;
                start.comments_before_length = len;
                if (len == 0 && start.comments_before.length > 0) ***REMOVED***
                    var comment = start.comments_before[0];
                    if (!comment.nlb) ***REMOVED***
                        comment.nlb = start.nlb;
                        start.nlb = false;
                  ***REMOVED***
              ***REMOVED***
                start.comments_after = ex.start.comments_after;
                ex.start = start;
                expect(")");
                var end = prev();
                end.comments_before = ex.end.comments_before;
                end.comments_after.forEach(function(comment) ***REMOVED***
                    ex.end.comments_after.push(comment);
                    if (comment.nlb) S.token.nlb = true;
              ***REMOVED***);
                end.comments_after.length = 0;
                end.comments_after = ex.end.comments_after;
                ex.end = end;
                if (is("punc", "=>")) return arrow(ex instanceof AST_Sequence ? ex.expressions : [ ex ], start);
                return subscripts(ex, allow_calls);
              case "[":
                return subscripts(array_(), allow_calls);
              case "***REMOVED***":
                return subscripts(object_(), allow_calls);
          ***REMOVED***
            unexpected();
      ***REMOVED***
        if (is("keyword")) switch (start.value) ***REMOVED***
          case "class":
            next();
            var clazz = class_(AST_ClassExpression);
            clazz.start = start;
            clazz.end = prev();
            return subscripts(clazz, allow_calls);
          case "function":
            next();
            var func;
            if (is("operator", "*")) ***REMOVED***
                next();
                func = function_(AST_GeneratorFunction);
          ***REMOVED*** else ***REMOVED***
                func = function_(AST_Function);
          ***REMOVED***
            func.start = start;
            func.end = prev();
            return subscripts(func, allow_calls);
      ***REMOVED***
        if (is("name")) ***REMOVED***
            var sym = _make_symbol(AST_SymbolRef, start);
            next();
            if (sym.name == "async") ***REMOVED***
                if (is("keyword", "function")) ***REMOVED***
                    next();
                    var func;
                    if (is("operator", "*")) ***REMOVED***
                        next();
                        func = function_(AST_AsyncGeneratorFunction);
                  ***REMOVED*** else ***REMOVED***
                        func = function_(AST_AsyncFunction);
                  ***REMOVED***
                    func.start = start;
                    func.end = prev();
                    return subscripts(func, allow_calls);
              ***REMOVED***
                if (is("name") && is_token(peek(), "punc", "=>")) ***REMOVED***
                    start = S.token;
                    sym = _make_symbol(AST_SymbolRef, start);
                    next();
                    return arrow([ sym ], start, true);
              ***REMOVED***
                if (is("punc", "(")) ***REMOVED***
                    var call = subscripts(sym, allow_calls);
                    if (!is("punc", "=>")) return call;
                    var args = call.args;
                    if (args[args.length - 1] instanceof AST_Spread) ***REMOVED***
                        args.rest = args.pop().expression;
                  ***REMOVED***
                    return arrow(args, start, true);
              ***REMOVED***
          ***REMOVED***
            return is("punc", "=>") ? arrow([ sym ], start) : subscripts(sym, allow_calls);
      ***REMOVED***
        if (ATOMIC_START_TOKEN[S.token.type]) ***REMOVED***
            return subscripts(as_atom_node(), allow_calls);
      ***REMOVED***
        unexpected();
  ***REMOVED***;

    function expr_list(closing, allow_trailing_comma, allow_empty, parser) ***REMOVED***
        if (!parser) parser = maybe_assign;
        var first = true, a = [];
        while (!is("punc", closing)) ***REMOVED***
            if (first) first = false; else expect(",");
            if (allow_trailing_comma && is("punc", closing)) break;
            if (allow_empty && is("punc", ",")) ***REMOVED***
                a.push(new AST_Hole(***REMOVED*** start: S.token, end: S.token }));
          ***REMOVED*** else if (!is("operator", "...")) ***REMOVED***
                a.push(parser());
          ***REMOVED*** else if (parser === maybe_assign) ***REMOVED***
                a.push(new AST_Spread(***REMOVED***
                    start: S.token,
                    expression: (next(), parser()),
                    end: prev(),
              ***REMOVED***));
          ***REMOVED*** else ***REMOVED***
                next();
                a.rest = parser();
                if (a.rest instanceof AST_DefaultValue) token_error(a.rest.start, "Invalid rest parameter");
                break;
          ***REMOVED***
      ***REMOVED***
        expect(closing);
        return a;
  ***REMOVED***

    var array_ = embed_tokens(function() ***REMOVED***
        expect("[");
        return new AST_Array(***REMOVED***
            elements: expr_list("]", !options.strict, true)
      ***REMOVED***);
  ***REMOVED***);

    var create_accessor = embed_tokens(function() ***REMOVED***
        return function_(AST_Accessor);
  ***REMOVED***);

    var object_ = embed_tokens(function() ***REMOVED***
        expect("***REMOVED***");
        var first = true, a = [];
        while (!is("punc", "}")) ***REMOVED***
            if (first) first = false; else expect(",");
            // allow trailing comma
            if (!options.strict && is("punc", "}")) break;
            var start = S.token;
            if (is("operator", "*")) ***REMOVED***
                next();
                var key = as_property_key();
                var gen_start = S.token;
                var gen = function_(AST_GeneratorFunction);
                gen.start = gen_start;
                gen.end = prev();
                a.push(new AST_ObjectMethod(***REMOVED***
                    start: start,
                    key: key,
                    value: gen,
                    end: prev(),
              ***REMOVED***));
                continue;
          ***REMOVED***
            if (is("operator", "...")) ***REMOVED***
                next();
                a.push(new AST_Spread(***REMOVED***
                    start: start,
                    expression: maybe_assign(),
                    end: prev(),
              ***REMOVED***));
                continue;
          ***REMOVED***
            if (is_token(peek(), "operator", "=")) ***REMOVED***
                var name = as_symbol(AST_SymbolRef);
                next();
                a.push(new AST_ObjectKeyVal(***REMOVED***
                    start: start,
                    key: start.value,
                    value: new AST_Assign(***REMOVED***
                        start: start,
                        left: name,
                        operator: "=",
                        right: maybe_assign(),
                        end: prev(),
                  ***REMOVED***),
                    end: prev(),
              ***REMOVED***));
                continue;
          ***REMOVED***
            if (is_token(peek(), "punc", ",") || is_token(peek(), "punc", "}")) ***REMOVED***
                a.push(new AST_ObjectKeyVal(***REMOVED***
                    start: start,
                    key: start.value,
                    value: as_symbol(AST_SymbolRef),
                    end: prev(),
              ***REMOVED***));
                continue;
          ***REMOVED***
            var key = as_property_key();
            if (is("punc", "(")) ***REMOVED***
                var func_start = S.token;
                var func = function_(AST_Function);
                func.start = func_start;
                func.end = prev();
                a.push(new AST_ObjectMethod(***REMOVED***
                    start: start,
                    key: key,
                    value: func,
                    end: prev(),
              ***REMOVED***));
                continue;
          ***REMOVED***
            if (is("punc", ":")) ***REMOVED***
                next();
                a.push(new AST_ObjectKeyVal(***REMOVED***
                    start: start,
                    key: key,
                    value: maybe_assign(),
                    end: prev(),
              ***REMOVED***));
                continue;
          ***REMOVED***
            if (start.type == "name") switch (key) ***REMOVED***
              case "async":
                var is_gen = is("operator", "*") && next();
                key = as_property_key();
                var func_start = S.token;
                var func = function_(is_gen ? AST_AsyncGeneratorFunction : AST_AsyncFunction);
                func.start = func_start;
                func.end = prev();
                a.push(new AST_ObjectMethod(***REMOVED***
                    start: start,
                    key: key,
                    value: func,
                    end: prev(),
              ***REMOVED***));
                continue;
              case "get":
                a.push(new AST_ObjectGetter(***REMOVED***
                    start: start,
                    key: as_property_key(),
                    value: create_accessor(),
                    end: prev(),
              ***REMOVED***));
                continue;
              case "set":
                a.push(new AST_ObjectSetter(***REMOVED***
                    start: start,
                    key: as_property_key(),
                    value: create_accessor(),
                    end: prev(),
              ***REMOVED***));
                continue;
          ***REMOVED***
            unexpected();
      ***REMOVED***
        next();
        return new AST_Object(***REMOVED*** properties: a });
  ***REMOVED***);

    function as_property_key() ***REMOVED***
        var tmp = S.token;
        switch (tmp.type) ***REMOVED***
          case "operator":
            if (!KEYWORDS[tmp.value]) unexpected();
          case "num":
          case "string":
          case "name":
          case "keyword":
          case "atom":
            next();
            return "" + tmp.value;
          case "punc":
            expect("[");
            var key = maybe_assign();
            expect("]");
            return key;
          default:
            unexpected();
      ***REMOVED***
  ***REMOVED***

    function as_name() ***REMOVED***
        var name = S.token.value;
        expect_token("name");
        return name;
  ***REMOVED***

    function _make_symbol(type, token) ***REMOVED***
        var name = token.value;
        switch (name) ***REMOVED***
          case "await":
            if (S.in_async) unexpected(token);
            break;
          case "super":
            type = AST_Super;
            break;
          case "this":
            type = AST_This;
            break;
          case "yield":
            if (S.in_generator) unexpected(token);
            break;
      ***REMOVED***
        return new type(***REMOVED***
            name: "" + name,
            start: token,
            end: token,
      ***REMOVED***);
  ***REMOVED***

    function strict_verify_symbol(sym) ***REMOVED***
        if (sym.name == "arguments" || sym.name == "eval" || sym.name == "let")
            token_error(sym.start, "Unexpected " + sym.name + " in strict mode");
  ***REMOVED***

    function as_symbol(type, no_error) ***REMOVED***
        if (!is("name")) ***REMOVED***
            if (!no_error) croak("Name expected");
            return null;
      ***REMOVED***
        var sym = _make_symbol(type, S.token);
        if (S.input.has_directive("use strict") && sym instanceof AST_SymbolDeclaration) ***REMOVED***
            strict_verify_symbol(sym);
      ***REMOVED***
        next();
        return sym;
  ***REMOVED***

    function maybe_destructured(type) ***REMOVED***
        var start = S.token;
        if (is("punc", "[")) ***REMOVED***
            next();
            var elements = expr_list("]", !options.strict, true, function() ***REMOVED***
                return maybe_default(type);
          ***REMOVED***);
            return new AST_DestructuredArray(***REMOVED***
                start: start,
                elements: elements,
                rest: elements.rest || null,
                end: prev(),
          ***REMOVED***);
      ***REMOVED***
        if (is("punc", "***REMOVED***")) ***REMOVED***
            next();
            var first = true, a = [], rest = null;
            while (!is("punc", "}")) ***REMOVED***
                if (first) first = false; else expect(",");
                // allow trailing comma
                if (!options.strict && is("punc", "}")) break;
                var key_start = S.token;
                if (is("punc", "[") || is_token(peek(), "punc", ":")) ***REMOVED***
                    var key = as_property_key();
                    expect(":");
                    a.push(new AST_DestructuredKeyVal(***REMOVED***
                        start: key_start,
                        key: key,
                        value: maybe_default(type),
                        end: prev(),
                  ***REMOVED***));
                    continue;
              ***REMOVED***
                if (is("operator", "...")) ***REMOVED***
                    next();
                    rest = maybe_destructured(type);
                    break;
              ***REMOVED***
                var name = as_symbol(type);
                if (is("operator", "=")) ***REMOVED***
                    next();
                    name = new AST_DefaultValue(***REMOVED***
                        start: name.start,
                        name: name,
                        value: maybe_assign(),
                        end: prev(),
                  ***REMOVED***);
              ***REMOVED***
                a.push(new AST_DestructuredKeyVal(***REMOVED***
                    start: key_start,
                    key: key_start.value,
                    value: name,
                    end: prev(),
              ***REMOVED***));
          ***REMOVED***
            expect("}");
            return new AST_DestructuredObject(***REMOVED***
                start: start,
                properties: a,
                rest: rest,
                end: prev(),
          ***REMOVED***);
      ***REMOVED***
        return as_symbol(type);
  ***REMOVED***

    function maybe_default(type) ***REMOVED***
        var start = S.token;
        var name = maybe_destructured(type);
        if (!is("operator", "=")) return name;
        next();
        return new AST_DefaultValue(***REMOVED***
            start: start,
            name: name,
            value: maybe_assign(),
            end: prev(),
      ***REMOVED***);
  ***REMOVED***

    function template(tag) ***REMOVED***
        var start = tag ? tag.start : S.token;
        var read = S.input.context().read_template;
        var strings = [];
        var expressions = [];
        while (read(strings)) ***REMOVED***
            next();
            expressions.push(expression());
            if (!is("punc", "}")) unexpected();
      ***REMOVED***
        next();
        return new AST_Template(***REMOVED***
            start: start,
            expressions: expressions,
            strings: strings,
            tag: tag,
            end: prev(),
      ***REMOVED***);
  ***REMOVED***

    function subscripts(expr, allow_calls) ***REMOVED***
        var start = expr.start;
        var optional = null;
        while (true) ***REMOVED***
            if (is("operator", "?") && is_token(peek(), "punc", ".")) ***REMOVED***
                next();
                next();
                optional = expr;
          ***REMOVED***
            if (is("punc", "[")) ***REMOVED***
                next();
                var prop = expression();
                expect("]");
                expr = new AST_Sub(***REMOVED***
                    start: start,
                    optional: optional === expr,
                    expression: expr,
                    property: prop,
                    end: prev(),
              ***REMOVED***);
          ***REMOVED*** else if (allow_calls && is("punc", "(")) ***REMOVED***
                next();
                expr = new AST_Call(***REMOVED***
                    start: start,
                    optional: optional === expr,
                    expression: expr,
                    args: expr_list(")", !options.strict),
                    end: prev(),
              ***REMOVED***);
          ***REMOVED*** else if (optional === expr || is("punc", ".")) ***REMOVED***
                if (optional !== expr) next();
                expr = new AST_Dot(***REMOVED***
                    start: start,
                    optional: optional === expr,
                    expression: expr,
                    property: as_name(),
                    end: prev(),
              ***REMOVED***);
          ***REMOVED*** else if (is("punc", "`")) ***REMOVED***
                if (optional) croak("Invalid template on optional chain");
                expr = template(expr);
          ***REMOVED*** else ***REMOVED***
                break;
          ***REMOVED***
      ***REMOVED***
        if (optional) expr.terminal = true;
        if (expr instanceof AST_Call && !expr.pure) ***REMOVED***
            var start = expr.start;
            var comments = start.comments_before;
            var i = HOP(start, "comments_before_length") ? start.comments_before_length : comments.length;
            while (--i >= 0) ***REMOVED***
                if (/[@#]__PURE__/.test(comments[i].value)) ***REMOVED***
                    expr.pure = true;
                    break;
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        return expr;
  ***REMOVED***

    function maybe_unary(no_in) ***REMOVED***
        var start = S.token;
        if (S.in_async && is("name", "await")) ***REMOVED***
            if (S.in_funarg === S.in_function) croak("Invalid use of await in function argument");
            S.input.context().regex_allowed = true;
            next();
            return new AST_Await(***REMOVED***
                start: start,
                expression: maybe_unary(no_in),
                end: prev(),
          ***REMOVED***);
      ***REMOVED***
        if (S.in_generator && is("name", "yield")) ***REMOVED***
            if (S.in_funarg === S.in_function) croak("Invalid use of yield in function argument");
            S.input.context().regex_allowed = true;
            next();
            var exp = null;
            var nested = false;
            if (is("operator", "*")) ***REMOVED***
                next();
                exp = maybe_assign(no_in);
                nested = true;
          ***REMOVED*** else if (is("punc") ? !PUNC_AFTER_EXPRESSION[S.token.value] : !can_insert_semicolon()) ***REMOVED***
                exp = maybe_assign(no_in);
          ***REMOVED***
            return new AST_Yield(***REMOVED***
                start: start,
                expression: exp,
                nested: nested,
                end: prev(),
          ***REMOVED***);
      ***REMOVED***
        if (is("operator") && UNARY_PREFIX[start.value]) ***REMOVED***
            next();
            handle_regexp();
            var ex = make_unary(AST_UnaryPrefix, start, maybe_unary(no_in));
            ex.start = start;
            ex.end = prev();
            return ex;
      ***REMOVED***
        var val = expr_atom(true);
        while (is("operator") && UNARY_POSTFIX[S.token.value] && !has_newline_before(S.token)) ***REMOVED***
            val = make_unary(AST_UnaryPostfix, S.token, val);
            val.start = start;
            val.end = S.token;
            next();
      ***REMOVED***
        return val;
  ***REMOVED***

    function make_unary(ctor, token, expr) ***REMOVED***
        var op = token.value;
        switch (op) ***REMOVED***
          case "++":
          case "--":
            if (!is_assignable(expr))
                token_error(token, "Invalid use of " + op + " operator");
            break;
          case "delete":
            if (expr instanceof AST_SymbolRef && S.input.has_directive("use strict"))
                token_error(expr.start, "Calling delete on expression not allowed in strict mode");
            break;
      ***REMOVED***
        return new ctor(***REMOVED*** operator: op, expression: expr });
  ***REMOVED***

    var expr_op = function(left, min_precision, no_in) ***REMOVED***
        var op = is("operator") ? S.token.value : null;
        if (op == "in" && no_in) op = null;
        var precision = op != null ? PRECEDENCE[op] : null;
        if (precision != null && precision > min_precision) ***REMOVED***
            next();
            var right = expr_op(maybe_unary(no_in), op == "**" ? precision - 1 : precision, no_in);
            return expr_op(new AST_Binary(***REMOVED***
                start    : left.start,
                left     : left,
                operator : op,
                right    : right,
                end      : right.end,
          ***REMOVED***), min_precision, no_in);
      ***REMOVED***
        return left;
  ***REMOVED***;

    function expr_ops(no_in) ***REMOVED***
        return expr_op(maybe_unary(no_in), 0, no_in);
  ***REMOVED***

    var maybe_conditional = function(no_in) ***REMOVED***
        var start = S.token;
        var expr = expr_ops(no_in);
        if (is("operator", "?")) ***REMOVED***
            next();
            var yes = maybe_assign();
            expect(":");
            return new AST_Conditional(***REMOVED***
                start       : start,
                condition   : expr,
                consequent  : yes,
                alternative : maybe_assign(no_in),
                end         : prev()
          ***REMOVED***);
      ***REMOVED***
        return expr;
  ***REMOVED***;

    function is_assignable(expr) ***REMOVED***
        return expr instanceof AST_PropAccess && !expr.optional || expr instanceof AST_SymbolRef;
  ***REMOVED***

    function to_destructured(node) ***REMOVED***
        if (node instanceof AST_Array) ***REMOVED***
            var rest = null;
            if (node.elements[node.elements.length - 1] instanceof AST_Spread) ***REMOVED***
                rest = to_destructured(node.elements.pop().expression);
                if (!(rest instanceof AST_Destructured || is_assignable(rest))) return node;
          ***REMOVED***
            var elements = node.elements.map(to_destructured);
            return all(elements, function(node) ***REMOVED***
                return node instanceof AST_DefaultValue
                    || node instanceof AST_Destructured
                    || node instanceof AST_Hole
                    || is_assignable(node);
          ***REMOVED***) ? new AST_DestructuredArray(***REMOVED***
                start: node.start,
                elements: elements,
                rest: rest,
                end: node.end,
          ***REMOVED***) : node;
      ***REMOVED***
        if (node instanceof AST_Assign) ***REMOVED***
            var name = to_destructured(node.left);
            return name instanceof AST_Destructured || is_assignable(name) ? new AST_DefaultValue(***REMOVED***
                start: node.start,
                name: name,
                value: node.right,
                end: node.end,
          ***REMOVED***) : node;
      ***REMOVED***
        if (!(node instanceof AST_Object)) return node;
        var rest = null;
        if (node.properties[node.properties.length - 1] instanceof AST_Spread) ***REMOVED***
            rest = to_destructured(node.properties.pop().expression);
            if (!(rest instanceof AST_Destructured || is_assignable(rest))) return node;
      ***REMOVED***
        var props = [];
        for (var i = 0; i < node.properties.length; i++) ***REMOVED***
            var prop = node.properties[i];
            if (!(prop instanceof AST_ObjectKeyVal)) return node;
            var value = to_destructured(prop.value);
            if (!(value instanceof AST_DefaultValue || value instanceof AST_Destructured || is_assignable(value))) ***REMOVED***
                return node;
          ***REMOVED***
            props.push(new AST_DestructuredKeyVal(***REMOVED***
                start: prop.start,
                key: prop.key,
                value: value,
                end: prop.end,
          ***REMOVED***));
      ***REMOVED***
        return new AST_DestructuredObject(***REMOVED***
            start: node.start,
            properties: props,
            rest: rest,
            end: node.end,
      ***REMOVED***);
  ***REMOVED***

    function maybe_assign(no_in) ***REMOVED***
        var start = S.token;
        var left = maybe_conditional(no_in), val = S.token.value;
        if (is("operator") && ASSIGNMENT[val]) ***REMOVED***
            if (is_assignable(left) || val == "=" && (left = to_destructured(left)) instanceof AST_Destructured) ***REMOVED***
                next();
                return new AST_Assign(***REMOVED***
                    start    : start,
                    left     : left,
                    operator : val,
                    right    : maybe_assign(no_in),
                    end      : prev()
              ***REMOVED***);
          ***REMOVED***
            croak("Invalid assignment");
      ***REMOVED***
        return left;
  ***REMOVED***

    function expression(no_in, maybe_arrow) ***REMOVED***
        var start = S.token;
        var exprs = [];
        while (true) ***REMOVED***
            if (maybe_arrow && is("operator", "...")) ***REMOVED***
                next();
                exprs.rest = maybe_destructured(AST_SymbolFunarg);
                break;
          ***REMOVED***
            exprs.push(maybe_assign(no_in));
            if (!is("punc", ",")) break;
            next();
            if (maybe_arrow && is("punc", ")") && is_token(peek(), "punc", "=>")) break;
      ***REMOVED***
        return exprs.length == 1 && !exprs.rest ? exprs[0] : new AST_Sequence(***REMOVED***
            start: start,
            expressions: exprs,
            end: prev(),
      ***REMOVED***);
  ***REMOVED***

    function in_loop(cont) ***REMOVED***
        ++S.in_loop;
        var ret = cont();
        --S.in_loop;
        return ret;
  ***REMOVED***

    if (options.expression) ***REMOVED***
        handle_regexp();
        var exp = expression();
        expect_token("eof");
        return exp;
  ***REMOVED***

    return function() ***REMOVED***
        var start = S.token;
        var body = [];
        if (options.module) ***REMOVED***
            S.in_async = true;
            S.input.add_directive("use strict");
      ***REMOVED***
        S.input.push_directives_stack();
        while (!is("eof"))
            body.push(statement(true));
        S.input.pop_directives_stack();
        var end = prev() || start;
        var toplevel = options.toplevel;
        if (toplevel) ***REMOVED***
            toplevel.body = toplevel.body.concat(body);
            toplevel.end = end;
      ***REMOVED*** else ***REMOVED***
            toplevel = new AST_Toplevel(***REMOVED*** start: start, body: body, end: end });
      ***REMOVED***
        return toplevel;
  ***REMOVED***();
}
