// @firebase/functions/dist/index.esm2017.js
import ***REMOVED*** _registerComponent, registerVersion, _getProvider, getApp } from '@firebase/app';
import ***REMOVED*** FirebaseError, getModularInstance, getDefaultEmulatorHostnameAndPort } from '@firebase/util';
import ***REMOVED*** Component } from '@firebase/component';

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const LONG_TYPE = 'type.googleapis.com/google.protobuf.Int64Value';
const UNSIGNED_LONG_TYPE = 'type.googleapis.com/google.protobuf.UInt64Value';
function mapValues(
// ***REMOVED*** [k: string]: unknown } is no longer a wildcard assignment target after typescript 3.5
// eslint-disable-next-line @typescript-eslint/no-explicit-any
o, f) ***REMOVED***
    const result = ***REMOVED***};
    for (const key in o) ***REMOVED***
        if (o.hasOwnProperty(key)) ***REMOVED***
            result[key] = f(o[key]);
      ***REMOVED***
  ***REMOVED***
    return result;
}
/**
 * Takes data and encodes it in a JSON-friendly way, such that types such as
 * Date are preserved.
 * @internal
 * @param data - Data to encode.
 */
function encode(data) ***REMOVED***
    if (data == null) ***REMOVED***
        return null;
  ***REMOVED***
    if (data instanceof Number) ***REMOVED***
        data = data.valueOf();
  ***REMOVED***
    if (typeof data === 'number' && isFinite(data)) ***REMOVED***
        // Any number in JS is safe to put directly in JSON and parse as a double
        // without any loss of precision.
        return data;
  ***REMOVED***
    if (data === true || data === false) ***REMOVED***
        return data;
  ***REMOVED***
    if (Object.prototype.toString.call(data) === '[object String]') ***REMOVED***
        return data;
  ***REMOVED***
    if (data instanceof Date) ***REMOVED***
        return data.toISOString();
  ***REMOVED***
    if (Array.isArray(data)) ***REMOVED***
        return data.map(x => encode(x));
  ***REMOVED***
    if (typeof data === 'function' || typeof data === 'object') ***REMOVED***
        return mapValues(data, x => encode(x));
  ***REMOVED***
    // If we got this far, the data is not encodable.
    throw new Error('Data cannot be encoded in JSON: ' + data);
}
/**
 * Takes data that's been encoded in a JSON-friendly form and returns a form
 * with richer datatypes, such as Dates, etc.
 * @internal
 * @param json - JSON to convert.
 */
function decode(json) ***REMOVED***
    if (json == null) ***REMOVED***
        return json;
  ***REMOVED***
    if (json['@type']) ***REMOVED***
        switch (json['@type']) ***REMOVED***
            case LONG_TYPE:
            // Fall through and handle this the same as unsigned.
            case UNSIGNED_LONG_TYPE: ***REMOVED***
                // Technically, this could work return a valid number for malformed
                // data if there was a number followed by garbage. But it's just not
                // worth all the extra code to detect that case.
                const value = Number(json['value']);
                if (isNaN(value)) ***REMOVED***
                    throw new Error('Data cannot be decoded from JSON: ' + json);
              ***REMOVED***
                return value;
          ***REMOVED***
            default: ***REMOVED***
                throw new Error('Data cannot be decoded from JSON: ' + json);
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    if (Array.isArray(json)) ***REMOVED***
        return json.map(x => decode(x));
  ***REMOVED***
    if (typeof json === 'function' || typeof json === 'object') ***REMOVED***
        return mapValues(json, x => decode(x));
  ***REMOVED***
    // Anything else is safe to return.
    return json;
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Type constant for Firebase Functions.
 */
const FUNCTIONS_TYPE = 'functions';

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Standard error codes for different ways a request can fail, as defined by:
 * https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto
 *
 * This map is used primarily to convert from a backend error code string to
 * a client SDK error code string, and make sure it's in the supported set.
 */
const errorCodeMap = ***REMOVED***
    OK: 'ok',
    CANCELLED: 'cancelled',
    UNKNOWN: 'unknown',
    INVALID_ARGUMENT: 'invalid-argument',
    DEADLINE_EXCEEDED: 'deadline-exceeded',
    NOT_FOUND: 'not-found',
    ALREADY_EXISTS: 'already-exists',
    PERMISSION_DENIED: 'permission-denied',
    UNAUTHENTICATED: 'unauthenticated',
    RESOURCE_EXHAUSTED: 'resource-exhausted',
    FAILED_PRECONDITION: 'failed-precondition',
    ABORTED: 'aborted',
    OUT_OF_RANGE: 'out-of-range',
    UNIMPLEMENTED: 'unimplemented',
    INTERNAL: 'internal',
    UNAVAILABLE: 'unavailable',
    DATA_LOSS: 'data-loss'
};
/**
 * An explicit error that can be thrown from a handler to send an error to the
 * client that called the function.
 */
class FunctionsError extends FirebaseError ***REMOVED***
    constructor(
    /**
     * A standard error code that will be returned to the client. This also
     * determines the HTTP status code of the response, as defined in code.proto.
     */
    code, message, 
    /**
     * Extra data to be converted to JSON and included in the error response.
     */
    details) ***REMOVED***
        super(`$***REMOVED***FUNCTIONS_TYPE}/$***REMOVED***code}`, message || '');
        this.details = details;
  ***REMOVED***
}
/**
 * Takes an HTTP status code and returns the corresponding ErrorCode.
 * This is the standard HTTP status code -> error mapping defined in:
 * https://github.com/googleapis/googleapis/blob/master/google/rpc/code.proto
 *
 * @param status An HTTP status code.
 * @return The corresponding ErrorCode, or ErrorCode.UNKNOWN if none.
 */
function codeForHTTPStatus(status) ***REMOVED***
    // Make sure any successful status is OK.
    if (status >= 200 && status < 300) ***REMOVED***
        return 'ok';
  ***REMOVED***
    switch (status) ***REMOVED***
        case 0:
            // This can happen if the server returns 500.
            return 'internal';
        case 400:
            return 'invalid-argument';
        case 401:
            return 'unauthenticated';
        case 403:
            return 'permission-denied';
        case 404:
            return 'not-found';
        case 409:
            return 'aborted';
        case 429:
            return 'resource-exhausted';
        case 499:
            return 'cancelled';
        case 500:
            return 'internal';
        case 501:
            return 'unimplemented';
        case 503:
            return 'unavailable';
        case 504:
            return 'deadline-exceeded';
  ***REMOVED***
    return 'unknown';
}
/**
 * Takes an HTTP response and returns the corresponding Error, if any.
 */
function _errorForResponse(status, bodyJSON) ***REMOVED***
    let code = codeForHTTPStatus(status);
    // Start with reasonable defaults from the status code.
    let description = code;
    let details = undefined;
    // Then look through the body for explicit details.
    try ***REMOVED***
        const errorJSON = bodyJSON && bodyJSON.error;
        if (errorJSON) ***REMOVED***
            const status = errorJSON.status;
            if (typeof status === 'string') ***REMOVED***
                if (!errorCodeMap[status]) ***REMOVED***
                    // They must've included an unknown error code in the body.
                    return new FunctionsError('internal', 'internal');
              ***REMOVED***
                code = errorCodeMap[status];
                // TODO(klimt): Add better default descriptions for error enums.
                // The default description needs to be updated for the new code.
                description = status;
          ***REMOVED***
            const message = errorJSON.message;
            if (typeof message === 'string') ***REMOVED***
                description = message;
          ***REMOVED***
            details = errorJSON.details;
            if (details !== undefined) ***REMOVED***
                details = decode(details);
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    catch (e) ***REMOVED***
        // If we couldn't parse explicit error data, that's fine.
  ***REMOVED***
    if (code === 'ok') ***REMOVED***
        // Technically, there's an edge case where a developer could explicitly
        // return an error code of OK, and we will treat it as success, but that
        // seems reasonable.
        return null;
  ***REMOVED***
    return new FunctionsError(code, description, details);
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Helper class to get metadata that should be included with a function call.
 * @internal
 */
class ContextProvider ***REMOVED***
    constructor(authProvider, messagingProvider, appCheckProvider) ***REMOVED***
        this.auth = null;
        this.messaging = null;
        this.appCheck = null;
        this.auth = authProvider.getImmediate(***REMOVED*** optional: true });
        this.messaging = messagingProvider.getImmediate(***REMOVED***
            optional: true
      ***REMOVED***);
        if (!this.auth) ***REMOVED***
            authProvider.get().then(auth => (this.auth = auth), () => ***REMOVED***
                /* get() never rejects */
          ***REMOVED***);
      ***REMOVED***
        if (!this.messaging) ***REMOVED***
            messagingProvider.get().then(messaging => (this.messaging = messaging), () => ***REMOVED***
                /* get() never rejects */
          ***REMOVED***);
      ***REMOVED***
        if (!this.appCheck) ***REMOVED***
            appCheckProvider.get().then(appCheck => (this.appCheck = appCheck), () => ***REMOVED***
                /* get() never rejects */
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***
    async getAuthToken() ***REMOVED***
        if (!this.auth) ***REMOVED***
            return undefined;
      ***REMOVED***
        try ***REMOVED***
            const token = await this.auth.getToken();
            return token === null || token === void 0 ? void 0 : token.accessToken;
      ***REMOVED***
        catch (e) ***REMOVED***
            // If there's any error when trying to get the auth token, leave it off.
            return undefined;
      ***REMOVED***
  ***REMOVED***
    async getMessagingToken() ***REMOVED***
        if (!this.messaging ||
            !('Notification' in self) ||
            Notification.permission !== 'granted') ***REMOVED***
            return undefined;
      ***REMOVED***
        try ***REMOVED***
            return await this.messaging.getToken();
      ***REMOVED***
        catch (e) ***REMOVED***
            // We don't warn on this, because it usually means messaging isn't set up.
            // console.warn('Failed to retrieve instance id token.', e);
            // If there's any error when trying to get the token, leave it off.
            return undefined;
      ***REMOVED***
  ***REMOVED***
    async getAppCheckToken() ***REMOVED***
        if (this.appCheck) ***REMOVED***
            const result = await this.appCheck.getToken();
            if (result.error) ***REMOVED***
                // Do not send the App Check header to the functions endpoint if
                // there was an error from the App Check exchange endpoint. The App
                // Check SDK will already have logged the error to console.
                return null;
          ***REMOVED***
            return result.token;
      ***REMOVED***
        return null;
  ***REMOVED***
    async getContext() ***REMOVED***
        const authToken = await this.getAuthToken();
        const messagingToken = await this.getMessagingToken();
        const appCheckToken = await this.getAppCheckToken();
        return ***REMOVED*** authToken, messagingToken, appCheckToken };
  ***REMOVED***
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const DEFAULT_REGION = 'us-central1';
/**
 * Returns a Promise that will be rejected after the given duration.
 * The error will be of type FunctionsError.
 *
 * @param millis Number of milliseconds to wait before rejecting.
 */
function failAfter(millis) ***REMOVED***
    // Node timers and browser timers are fundamentally incompatible, but we
    // don't care about the value here
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let timer = null;
    return ***REMOVED***
        promise: new Promise((_, reject) => ***REMOVED***
            timer = setTimeout(() => ***REMOVED***
                reject(new FunctionsError('deadline-exceeded', 'deadline-exceeded'));
          ***REMOVED***, millis);
      ***REMOVED***),
        cancel: () => ***REMOVED***
            if (timer) ***REMOVED***
                clearTimeout(timer);
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***;
}
/**
 * The main class for the Firebase Functions SDK.
 * @internal
 */
class FunctionsService ***REMOVED***
    /**
     * Creates a new Functions service for the given app.
     * @param app - The FirebaseApp to use.
     */
    constructor(app, authProvider, messagingProvider, appCheckProvider, regionOrCustomDomain = DEFAULT_REGION, fetchImpl) ***REMOVED***
        this.app = app;
        this.fetchImpl = fetchImpl;
        this.emulatorOrigin = null;
        this.contextProvider = new ContextProvider(authProvider, messagingProvider, appCheckProvider);
        // Cancels all ongoing requests when resolved.
        this.cancelAllRequests = new Promise(resolve => ***REMOVED***
            this.deleteService = () => ***REMOVED***
                return Promise.resolve(resolve());
          ***REMOVED***;
      ***REMOVED***);
        // Resolve the region or custom domain overload by attempting to parse it.
        try ***REMOVED***
            const url = new URL(regionOrCustomDomain);
            this.customDomain = url.origin;
            this.region = DEFAULT_REGION;
      ***REMOVED***
        catch (e) ***REMOVED***
            this.customDomain = null;
            this.region = regionOrCustomDomain;
      ***REMOVED***
  ***REMOVED***
    _delete() ***REMOVED***
        return this.deleteService();
  ***REMOVED***
    /**
     * Returns the URL for a callable with the given name.
     * @param name - The name of the callable.
     * @internal
     */
    _url(name) ***REMOVED***
        const projectId = this.app.options.projectId;
        if (this.emulatorOrigin !== null) ***REMOVED***
            const origin = this.emulatorOrigin;
            return `$***REMOVED***origin}/$***REMOVED***projectId}/$***REMOVED***this.region}/$***REMOVED***name}`;
      ***REMOVED***
        if (this.customDomain !== null) ***REMOVED***
            return `$***REMOVED***this.customDomain}/$***REMOVED***name}`;
      ***REMOVED***
        return `https://$***REMOVED***this.region}-$***REMOVED***projectId}.cloudfunctions.net/$***REMOVED***name}`;
  ***REMOVED***
}
/**
 * Modify this instance to communicate with the Cloud Functions emulator.
 *
 * Note: this must be called before this instance has been used to do any operations.
 *
 * @param host The emulator host (ex: localhost)
 * @param port The emulator port (ex: 5001)
 * @public
 */
function connectFunctionsEmulator$1(functionsInstance, host, port) ***REMOVED***
    functionsInstance.emulatorOrigin = `http://$***REMOVED***host}:$***REMOVED***port}`;
}
/**
 * Returns a reference to the callable https trigger with the given name.
 * @param name - The name of the trigger.
 * @public
 */
function httpsCallable$1(functionsInstance, name, options) ***REMOVED***
    return (data => ***REMOVED***
        return call(functionsInstance, name, data, options || ***REMOVED***});
  ***REMOVED***);
}
/**
 * Returns a reference to the callable https trigger with the given url.
 * @param url - The url of the trigger.
 * @public
 */
function httpsCallableFromURL$1(functionsInstance, url, options) ***REMOVED***
    return (data => ***REMOVED***
        return callAtURL(functionsInstance, url, data, options || ***REMOVED***});
  ***REMOVED***);
}
/**
 * Does an HTTP POST and returns the completed response.
 * @param url The url to post to.
 * @param body The JSON body of the post.
 * @param headers The HTTP headers to include in the request.
 * @return A Promise that will succeed when the request finishes.
 */
async function postJSON(url, body, headers, fetchImpl) ***REMOVED***
    headers['Content-Type'] = 'application/json';
    let response;
    try ***REMOVED***
        response = await fetchImpl(url, ***REMOVED***
            method: 'POST',
            body: JSON.stringify(body),
            headers
      ***REMOVED***);
  ***REMOVED***
    catch (e) ***REMOVED***
        // This could be an unhandled error on the backend, or it could be a
        // network error. There's no way to know, since an unhandled error on the
        // backend will fail to set the proper CORS header, and thus will be
        // treated as a network error by fetch.
        return ***REMOVED***
            status: 0,
            json: null
      ***REMOVED***;
  ***REMOVED***
    let json = null;
    try ***REMOVED***
        json = await response.json();
  ***REMOVED***
    catch (e) ***REMOVED***
        // If we fail to parse JSON, it will fail the same as an empty body.
  ***REMOVED***
    return ***REMOVED***
        status: response.status,
        json
  ***REMOVED***;
}
/**
 * Calls a callable function asynchronously and returns the result.
 * @param name The name of the callable trigger.
 * @param data The data to pass as params to the function.s
 */
function call(functionsInstance, name, data, options) ***REMOVED***
    const url = functionsInstance._url(name);
    return callAtURL(functionsInstance, url, data, options);
}
/**
 * Calls a callable function asynchronously and returns the result.
 * @param url The url of the callable trigger.
 * @param data The data to pass as params to the function.s
 */
async function callAtURL(functionsInstance, url, data, options) ***REMOVED***
    // Encode any special types, such as dates, in the input data.
    data = encode(data);
    const body = ***REMOVED*** data };
    // Add a header for the authToken.
    const headers = ***REMOVED***};
    const context = await functionsInstance.contextProvider.getContext();
    if (context.authToken) ***REMOVED***
        headers['Authorization'] = 'Bearer ' + context.authToken;
  ***REMOVED***
    if (context.messagingToken) ***REMOVED***
        headers['Firebase-Instance-ID-Token'] = context.messagingToken;
  ***REMOVED***
    if (context.appCheckToken !== null) ***REMOVED***
        headers['X-Firebase-AppCheck'] = context.appCheckToken;
  ***REMOVED***
    // Default timeout to 70s, but let the options override it.
    const timeout = options.timeout || 70000;
    const failAfterHandle = failAfter(timeout);
    const response = await Promise.race([
        postJSON(url, body, headers, functionsInstance.fetchImpl),
        failAfterHandle.promise,
        functionsInstance.cancelAllRequests
    ]);
    // Always clear the failAfter timeout
    failAfterHandle.cancel();
    // If service was deleted, interrupted response throws an error.
    if (!response) ***REMOVED***
        throw new FunctionsError('cancelled', 'Firebase Functions instance was deleted.');
  ***REMOVED***
    // Check for an error status, regardless of http status.
    const error = _errorForResponse(response.status, response.json);
    if (error) ***REMOVED***
        throw error;
  ***REMOVED***
    if (!response.json) ***REMOVED***
        throw new FunctionsError('internal', 'Response is not valid JSON object.');
  ***REMOVED***
    let responseData = response.json.data;
    // TODO(klimt): For right now, allow "result" instead of "data", for
    // backwards compatibility.
    if (typeof responseData === 'undefined') ***REMOVED***
        responseData = response.json.result;
  ***REMOVED***
    if (typeof responseData === 'undefined') ***REMOVED***
        // Consider the response malformed.
        throw new FunctionsError('internal', 'Response is missing data field.');
  ***REMOVED***
    // Decode any special types, such as dates, in the returned data.
    const decodedData = decode(responseData);
    return ***REMOVED*** data: decodedData };
}

const name = "@firebase/functions";
const version = "0.9.1";

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const AUTH_INTERNAL_NAME = 'auth-internal';
const APP_CHECK_INTERNAL_NAME = 'app-check-internal';
const MESSAGING_INTERNAL_NAME = 'messaging-internal';
function registerFunctions(fetchImpl, variant) ***REMOVED***
    const factory = (container, ***REMOVED*** instanceIdentifier: regionOrCustomDomain }) => ***REMOVED***
        // Dependencies
        const app = container.getProvider('app').getImmediate();
        const authProvider = container.getProvider(AUTH_INTERNAL_NAME);
        const messagingProvider = container.getProvider(MESSAGING_INTERNAL_NAME);
        const appCheckProvider = container.getProvider(APP_CHECK_INTERNAL_NAME);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return new FunctionsService(app, authProvider, messagingProvider, appCheckProvider, regionOrCustomDomain, fetchImpl);
  ***REMOVED***;
    _registerComponent(new Component(FUNCTIONS_TYPE, factory, "PUBLIC" /* ComponentType.PUBLIC */).setMultipleInstances(true));
    registerVersion(name, version, variant);
    // BUILD_TARGET will be replaced by values like esm5, esm2017, cjs5, etc during the compilation
    registerVersion(name, version, 'esm2017');
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Returns a ***REMOVED***@link Functions} instance for the given app.
 * @param app - The ***REMOVED***@link @firebase/app#FirebaseApp} to use.
 * @param regionOrCustomDomain - one of:
 *   a) The region the callable functions are located in (ex: us-central1)
 *   b) A custom domain hosting the callable functions (ex: https://mydomain.com)
 * @public
 */
function getFunctions(app = getApp(), regionOrCustomDomain = DEFAULT_REGION) ***REMOVED***
    // Dependencies
    const functionsProvider = _getProvider(getModularInstance(app), FUNCTIONS_TYPE);
    const functionsInstance = functionsProvider.getImmediate(***REMOVED***
        identifier: regionOrCustomDomain
  ***REMOVED***);
    const emulator = getDefaultEmulatorHostnameAndPort('functions');
    if (emulator) ***REMOVED***
        connectFunctionsEmulator(functionsInstance, ...emulator);
  ***REMOVED***
    return functionsInstance;
}
/**
 * Modify this instance to communicate with the Cloud Functions emulator.
 *
 * Note: this must be called before this instance has been used to do any operations.
 *
 * @param host - The emulator host (ex: localhost)
 * @param port - The emulator port (ex: 5001)
 * @public
 */
function connectFunctionsEmulator(functionsInstance, host, port) ***REMOVED***
    connectFunctionsEmulator$1(getModularInstance(functionsInstance), host, port);
}
/**
 * Returns a reference to the callable HTTPS trigger with the given name.
 * @param name - The name of the trigger.
 * @public
 */
function httpsCallable(functionsInstance, name, options) ***REMOVED***
    return httpsCallable$1(getModularInstance(functionsInstance), name, options);
}
/**
 * Returns a reference to the callable HTTPS trigger with the specified url.
 * @param url - The url of the trigger.
 * @public
 */
function httpsCallableFromURL(functionsInstance, url, options) ***REMOVED***
    return httpsCallableFromURL$1(getModularInstance(functionsInstance), url, options);
}

/**
 * Cloud Functions for Firebase
 *
 * @packageDocumentation
 */
registerFunctions(fetch.bind(self));

export ***REMOVED*** connectFunctionsEmulator, getFunctions, httpsCallable, httpsCallableFromURL };
