// @babel/generator/lib/buffer.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
});
exports.default = void 0;
class Buffer ***REMOVED***
  constructor(map) ***REMOVED***
    this._map = null;
    this._buf = "";
    this._str = "";
    this._appendCount = 0;
    this._last = 0;
    this._queue = [];
    this._queueCursor = 0;
    this._position = ***REMOVED***
      line: 1,
      column: 0
  ***REMOVED***;
    this._sourcePosition = ***REMOVED***
      identifierName: undefined,
      line: undefined,
      column: undefined,
      filename: undefined
  ***REMOVED***;
    this._map = map;
    this._allocQueue();
***REMOVED***
  _allocQueue() ***REMOVED***
    const queue = this._queue;
    for (let i = 0; i < 16; i++) ***REMOVED***
      queue.push(***REMOVED***
        char: 0,
        repeat: 1,
        line: undefined,
        column: undefined,
        identifierName: undefined,
        filename: ""
    ***REMOVED***);
  ***REMOVED***
***REMOVED***
  _pushQueue(char, repeat, line, column, identifierName, filename) ***REMOVED***
    const cursor = this._queueCursor;
    if (cursor === this._queue.length) ***REMOVED***
      this._allocQueue();
  ***REMOVED***
    const item = this._queue[cursor];
    item.char = char;
    item.repeat = repeat;
    item.line = line;
    item.column = column;
    item.identifierName = identifierName;
    item.filename = filename;
    this._queueCursor++;
***REMOVED***
  _popQueue() ***REMOVED***
    if (this._queueCursor === 0) ***REMOVED***
      throw new Error("Cannot pop from empty queue");
  ***REMOVED***
    return this._queue[--this._queueCursor];
***REMOVED***

  get() ***REMOVED***
    this._flush();
    const map = this._map;
    const result = ***REMOVED***
      code: (this._buf + this._str).trimRight(),
      decodedMap: map == null ? void 0 : map.getDecoded(),
      get map() ***REMOVED***
        const resultMap = map ? map.get() : null;
        result.map = resultMap;
        return resultMap;
    ***REMOVED***,
      set map(value) ***REMOVED***
        Object.defineProperty(result, "map", ***REMOVED***
          value,
          writable: true
      ***REMOVED***);
    ***REMOVED***,
      get rawMappings() ***REMOVED***
        const mappings = map == null ? void 0 : map.getRawMappings();
        result.rawMappings = mappings;
        return mappings;
    ***REMOVED***,
      set rawMappings(value) ***REMOVED***
        Object.defineProperty(result, "rawMappings", ***REMOVED***
          value,
          writable: true
      ***REMOVED***);
    ***REMOVED***
  ***REMOVED***;
    return result;
***REMOVED***

  append(str, maybeNewline) ***REMOVED***
    this._flush();
    this._append(str, this._sourcePosition, maybeNewline);
***REMOVED***
  appendChar(char) ***REMOVED***
    this._flush();
    this._appendChar(char, 1, this._sourcePosition);
***REMOVED***

  queue(char) ***REMOVED***
    if (char === 10) ***REMOVED***
      while (this._queueCursor !== 0) ***REMOVED***
        const char = this._queue[this._queueCursor - 1].char;
        if (char !== 32 && char !== 9) ***REMOVED***
          break;
      ***REMOVED***
        this._queueCursor--;
    ***REMOVED***
  ***REMOVED***
    const sourcePosition = this._sourcePosition;
    this._pushQueue(char, 1, sourcePosition.line, sourcePosition.column, sourcePosition.identifierName, sourcePosition.filename);
***REMOVED***

  queueIndentation(char, repeat) ***REMOVED***
    this._pushQueue(char, repeat, undefined, undefined, undefined, undefined);
***REMOVED***
  _flush() ***REMOVED***
    const queueCursor = this._queueCursor;
    const queue = this._queue;
    for (let i = 0; i < queueCursor; i++) ***REMOVED***
      const item = queue[i];
      this._appendChar(item.char, item.repeat, item);
  ***REMOVED***
    this._queueCursor = 0;
***REMOVED***
  _appendChar(char, repeat, sourcePos) ***REMOVED***
    this._last = char;
    this._str += repeat > 1 ? String.fromCharCode(char).repeat(repeat) : String.fromCharCode(char);
    if (char !== 10) ***REMOVED***
      this._mark(sourcePos.line, sourcePos.column, sourcePos.identifierName, sourcePos.filename);
      this._position.column += repeat;
  ***REMOVED*** else ***REMOVED***
      this._position.line++;
      this._position.column = 0;
  ***REMOVED***
***REMOVED***
  _append(str, sourcePos, maybeNewline) ***REMOVED***
    const len = str.length;
    const position = this._position;
    this._last = str.charCodeAt(len - 1);
    if (++this._appendCount > 4096) ***REMOVED***
      +this._str;
      this._buf += this._str;
      this._str = str;
      this._appendCount = 0;
  ***REMOVED*** else ***REMOVED***
      this._str += str;
  ***REMOVED***
    if (!maybeNewline && !this._map) ***REMOVED***
      position.column += len;
      return;
  ***REMOVED***
    const ***REMOVED***
      column,
      identifierName,
      filename
  ***REMOVED*** = sourcePos;
    let line = sourcePos.line;

    let i = str.indexOf("\n");
    let last = 0;

    if (i !== 0) ***REMOVED***
      this._mark(line, column, identifierName, filename);
  ***REMOVED***

    while (i !== -1) ***REMOVED***
      position.line++;
      position.column = 0;
      last = i + 1;

      if (last < len) ***REMOVED***
        this._mark(++line, 0, identifierName, filename);
    ***REMOVED***
      i = str.indexOf("\n", last);
  ***REMOVED***
    position.column += len - last;
***REMOVED***
  _mark(line, column, identifierName, filename) ***REMOVED***
    var _this$_map;
    (_this$_map = this._map) == null ? void 0 : _this$_map.mark(this._position, line, column, identifierName, filename);
***REMOVED***
  removeTrailingNewline() ***REMOVED***
    const queueCursor = this._queueCursor;
    if (queueCursor !== 0 && this._queue[queueCursor - 1].char === 10) ***REMOVED***
      this._queueCursor--;
  ***REMOVED***
***REMOVED***
  removeLastSemicolon() ***REMOVED***
    const queueCursor = this._queueCursor;
    if (queueCursor !== 0 && this._queue[queueCursor - 1].char === 59) ***REMOVED***
      this._queueCursor--;
  ***REMOVED***
***REMOVED***
  getLastChar() ***REMOVED***
    const queueCursor = this._queueCursor;
    return queueCursor !== 0 ? this._queue[queueCursor - 1].char : this._last;
***REMOVED***

  getNewlineCount() ***REMOVED***
    const queueCursor = this._queueCursor;
    let count = 0;
    if (queueCursor === 0) return this._last === 10 ? 1 : 0;
    for (let i = queueCursor - 1; i >= 0; i--) ***REMOVED***
      if (this._queue[i].char !== 10) ***REMOVED***
        break;
    ***REMOVED***
      count++;
  ***REMOVED***
    return count === queueCursor && this._last === 10 ? count + 1 : count;
***REMOVED***

  endsWithCharAndNewline() ***REMOVED***
    const queue = this._queue;
    const queueCursor = this._queueCursor;
    if (queueCursor !== 0) ***REMOVED***
      const lastCp = queue[queueCursor - 1].char;
      if (lastCp !== 10) return;
      if (queueCursor > 1) ***REMOVED***
        return queue[queueCursor - 2].char;
    ***REMOVED*** else ***REMOVED***
        return this._last;
    ***REMOVED***
  ***REMOVED***
***REMOVED***

  hasContent() ***REMOVED***
    return this._queueCursor !== 0 || !!this._last;
***REMOVED***

  exactSource(loc, cb) ***REMOVED***
    if (!this._map) return cb();
    this.source("start", loc);
    cb();
    this.source("end", loc);
***REMOVED***

  source(prop, loc) ***REMOVED***
    if (!this._map) return;

    this._normalizePosition(prop, loc, 0, 0);
***REMOVED***
  sourceWithOffset(prop, loc, lineOffset, columnOffset) ***REMOVED***
    if (!this._map) return;
    this._normalizePosition(prop, loc, lineOffset, columnOffset);
***REMOVED***

  withSource(prop, loc, cb) ***REMOVED***
    if (!this._map) return cb();
    this.source(prop, loc);
    cb();
***REMOVED***
  _normalizePosition(prop, loc, lineOffset, columnOffset) ***REMOVED***
    const pos = loc[prop];
    const target = this._sourcePosition;
    target.identifierName = prop === "start" && loc.identifierName || undefined;
    if (pos) ***REMOVED***
      target.line = pos.line + lineOffset;
      target.column = pos.column + columnOffset;
      target.filename = loc.filename;
  ***REMOVED***
***REMOVED***
  getCurrentColumn() ***REMOVED***
    const queue = this._queue;
    const queueCursor = this._queueCursor;
    let lastIndex = -1;
    let len = 0;
    for (let i = 0; i < queueCursor; i++) ***REMOVED***
      const item = queue[i];
      if (item.char === 10) ***REMOVED***
        lastIndex = len;
    ***REMOVED***
      len += item.repeat;
  ***REMOVED***
    return lastIndex === -1 ? this._position.column + len : len - 1 - lastIndex;
***REMOVED***
  getCurrentLine() ***REMOVED***
    let count = 0;
    const queue = this._queue;
    for (let i = 0; i < this._queueCursor; i++) ***REMOVED***
      if (queue[i].char === 10) ***REMOVED***
        count++;
    ***REMOVED***
  ***REMOVED***
    return this._position.line + count;
***REMOVED***
}
exports.default = Buffer;
