// @google-cloud/storage/build/src/nodejs-common/service.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.Service = exports.DEFAULT_PROJECT_ID_TOKEN = void 0;
/*!
 * Copyright 2022 Google LLC. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const extend = require("extend");
const uuid = require("uuid");
const util_1 = require("./util");
exports.DEFAULT_PROJECT_ID_TOKEN = '***REMOVED******REMOVED***projectId}}';
class Service ***REMOVED***
    /**
     * Service is a base class, meant to be inherited from by a "service," like
     * BigQuery or Storage.
     *
     * This handles making authenticated requests by exposing a `makeReq_`
     * function.
     *
     * @constructor
     * @alias module:common/service
     *
     * @param ***REMOVED***object} config - Configuration object.
     * @param ***REMOVED***string} config.baseUrl - The base URL to make API requests to.
     * @param ***REMOVED***string[]} config.scopes - The scopes required for the request.
     * @param ***REMOVED***object=} options - [Configuration object](#/docs).
     */
    constructor(config, options = ***REMOVED***}) ***REMOVED***
        this.baseUrl = config.baseUrl;
        this.apiEndpoint = config.apiEndpoint;
        this.timeout = options.timeout;
        this.globalInterceptors = Array.isArray(options.interceptors_)
            ? options.interceptors_
            : [];
        this.interceptors = [];
        this.packageJson = config.packageJson;
        this.projectId = options.projectId || exports.DEFAULT_PROJECT_ID_TOKEN;
        this.projectIdRequired = config.projectIdRequired !== false;
        this.providedUserAgent = options.userAgent;
        const reqCfg = extend(***REMOVED***}, config, ***REMOVED***
            projectIdRequired: this.projectIdRequired,
            projectId: this.projectId,
            authClient: options.authClient,
            credentials: options.credentials,
            keyFile: options.keyFilename,
            email: options.email,
            token: options.token,
      ***REMOVED***);
        this.makeAuthenticatedRequest =
            util_1.util.makeAuthenticatedRequestFactory(reqCfg);
        this.authClient = this.makeAuthenticatedRequest.authClient;
        this.getCredentials = this.makeAuthenticatedRequest.getCredentials;
        const isCloudFunctionEnv = !!process.env.FUNCTION_NAME;
        if (isCloudFunctionEnv) ***REMOVED***
            this.interceptors.push(***REMOVED***
                request(reqOpts) ***REMOVED***
                    reqOpts.forever = false;
                    return reqOpts;
              ***REMOVED***,
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***
    /**
     * Return the user's custom request interceptors.
     */
    getRequestInterceptors() ***REMOVED***
        // Interceptors should be returned in the order they were assigned.
        return [].slice
            .call(this.globalInterceptors)
            .concat(this.interceptors)
            .filter(interceptor => typeof interceptor.request === 'function')
            .map(interceptor => interceptor.request);
  ***REMOVED***
    getProjectId(callback) ***REMOVED***
        if (!callback) ***REMOVED***
            return this.getProjectIdAsync();
      ***REMOVED***
        this.getProjectIdAsync().then(p => callback(null, p), callback);
  ***REMOVED***
    async getProjectIdAsync() ***REMOVED***
        const projectId = await this.authClient.getProjectId();
        if (this.projectId === exports.DEFAULT_PROJECT_ID_TOKEN && projectId) ***REMOVED***
            this.projectId = projectId;
      ***REMOVED***
        return this.projectId;
  ***REMOVED***
    request_(reqOpts, callback) ***REMOVED***
        reqOpts = extend(true, ***REMOVED***}, reqOpts, ***REMOVED*** timeout: this.timeout });
        const isAbsoluteUrl = reqOpts.uri.indexOf('http') === 0;
        const uriComponents = [this.baseUrl];
        if (this.projectIdRequired) ***REMOVED***
            if (reqOpts.projectId) ***REMOVED***
                uriComponents.push('projects');
                uriComponents.push(reqOpts.projectId);
          ***REMOVED***
            else ***REMOVED***
                uriComponents.push('projects');
                uriComponents.push(this.projectId);
          ***REMOVED***
      ***REMOVED***
        uriComponents.push(reqOpts.uri);
        if (isAbsoluteUrl) ***REMOVED***
            uriComponents.splice(0, uriComponents.indexOf(reqOpts.uri));
      ***REMOVED***
        reqOpts.uri = uriComponents
            .map(uriComponent => ***REMOVED***
            const trimSlashesRegex = /^\/*|\/*$/g;
            return uriComponent.replace(trimSlashesRegex, '');
      ***REMOVED***)
            .join('/')
            // Some URIs have colon separators.
            // Bad: https://.../projects/:list
            // Good: https://.../projects:list
            .replace(/\/:/g, ':');
        const requestInterceptors = this.getRequestInterceptors();
        const interceptorArray = Array.isArray(reqOpts.interceptors_)
            ? reqOpts.interceptors_
            : [];
        interceptorArray.forEach(interceptor => ***REMOVED***
            if (typeof interceptor.request === 'function') ***REMOVED***
                requestInterceptors.push(interceptor.request);
          ***REMOVED***
      ***REMOVED***);
        requestInterceptors.forEach(requestInterceptor => ***REMOVED***
            reqOpts = requestInterceptor(reqOpts);
      ***REMOVED***);
        delete reqOpts.interceptors_;
        const pkg = this.packageJson;
        let userAgent = util_1.util.getUserAgentFromPackageJson(pkg);
        if (this.providedUserAgent) ***REMOVED***
            userAgent = `$***REMOVED***this.providedUserAgent} $***REMOVED***userAgent}`;
      ***REMOVED***
        reqOpts.headers = extend(***REMOVED***}, reqOpts.headers, ***REMOVED***
            'User-Agent': userAgent,
            'x-goog-api-client': `gl-node/$***REMOVED***process.versions.node} gccl/$***REMOVED***pkg.version} gccl-invocation-id/$***REMOVED***uuid.v4()}`,
      ***REMOVED***);
        if (reqOpts.shouldReturnStream) ***REMOVED***
            return this.makeAuthenticatedRequest(reqOpts);
      ***REMOVED***
        else ***REMOVED***
            this.makeAuthenticatedRequest(reqOpts, callback);
      ***REMOVED***
  ***REMOVED***
    /**
     * Make an authenticated API request.
     *
     * @param ***REMOVED***object} reqOpts - Request options that are passed to `request`.
     * @param ***REMOVED***string} reqOpts.uri - A URI relative to the baseUrl.
     * @param ***REMOVED***function} callback - The callback function passed to `request`.
     */
    request(reqOpts, callback) ***REMOVED***
        Service.prototype.request_.call(this, reqOpts, callback);
  ***REMOVED***
    /**
     * Make an authenticated API request.
     *
     * @param ***REMOVED***object} reqOpts - Request options that are passed to `request`.
     * @param ***REMOVED***string} reqOpts.uri - A URI relative to the baseUrl.
     */
    requestStream(reqOpts) ***REMOVED***
        const opts = extend(true, reqOpts, ***REMOVED*** shouldReturnStream: true });
        return Service.prototype.request_.call(this, opts);
  ***REMOVED***
}
exports.Service = Service;
