// @google-cloud/storage/build/src/nodejs-common/service-object.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.ServiceObject = void 0;
/*!
 * Copyright 2022 Google LLC. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const promisify_1 = require("@google-cloud/promisify");
const events_1 = require("events");
const extend = require("extend");
const util_1 = require("./util");
/**
 * ServiceObject is a base class, meant to be inherited from by a "service
 * object," like a BigQuery dataset or Storage bucket.
 *
 * Most of the time, these objects share common functionality; they can be
 * created or deleted, and you can get or set their metadata.
 *
 * By inheriting from this class, a service object will be extended with these
 * shared behaviors. Note that any method can be overridden when the service
 * object requires specific behavior.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
class ServiceObject extends events_1.EventEmitter ***REMOVED***
    /*
     * @constructor
     * @alias module:common/service-object
     *
     * @private
     *
     * @param ***REMOVED***object} config - Configuration object.
     * @param ***REMOVED***string} config.baseUrl - The base URL to make API requests to.
     * @param ***REMOVED***string} config.createMethod - The method which creates this object.
     * @param ***REMOVED***string=} config.id - The identifier of the object. For example, the
     *     name of a Storage bucket or Pub/Sub topic.
     * @param ***REMOVED***object=} config.methods - A map of each method name that should be inherited.
     * @param ***REMOVED***object} config.methods[].reqOpts - Default request options for this
     *     particular method. A common use case is when `setMetadata` requires a
     *     `PUT` method to override the default `PATCH`.
     * @param ***REMOVED***object} config.parent - The parent service instance. For example, an
     *     instance of Storage if the object is Bucket.
     */
    constructor(config) ***REMOVED***
        super();
        this.metadata = ***REMOVED***};
        this.baseUrl = config.baseUrl;
        this.parent = config.parent; // Parent class.
        this.id = config.id; // Name or ID (e.g. dataset ID, bucket name, etc).
        this.createMethod = config.createMethod;
        this.methods = config.methods || ***REMOVED***};
        this.interceptors = [];
        this.projectId = config.projectId;
        if (config.methods) ***REMOVED***
            // This filters the ServiceObject instance (e.g. a "File") to only have
            // the configured methods. We make a couple of exceptions for core-
            // functionality ("request()" and "getRequestInterceptors()")
            Object.getOwnPropertyNames(ServiceObject.prototype)
                .filter(methodName => ***REMOVED***
                return (
                // All ServiceObjects need `request` and `getRequestInterceptors`.
                // clang-format off
                !/^request/.test(methodName) &&
                    !/^getRequestInterceptors/.test(methodName) &&
                    // clang-format on
                    // The ServiceObject didn't redefine the method.
                    // eslint-disable-next-line @typescript-eslint/no-explicit-any
                    this[methodName] ===
                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
                        ServiceObject.prototype[methodName] &&
                    // This method isn't wanted.
                    !config.methods[methodName]);
          ***REMOVED***)
                .forEach(methodName => ***REMOVED***
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                this[methodName] = undefined;
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***
    create(optionsOrCallback, callback) ***REMOVED***
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const self = this;
        const args = [this.id];
        if (typeof optionsOrCallback === 'function') ***REMOVED***
            callback = optionsOrCallback;
      ***REMOVED***
        if (typeof optionsOrCallback === 'object') ***REMOVED***
            args.push(optionsOrCallback);
      ***REMOVED***
        // Wrap the callback to return *this* instance of the object, not the
        // newly-created one.
        // tslint: disable-next-line no-any
        function onCreate(...args) ***REMOVED***
            const [err, instance] = args;
            if (!err) ***REMOVED***
                self.metadata = instance.metadata;
                if (self.id && instance.metadata) ***REMOVED***
                    self.id = instance.metadata.id;
              ***REMOVED***
                args[1] = self; // replace the created `instance` with this one.
          ***REMOVED***
            callback(...args);
      ***REMOVED***
        args.push(onCreate);
        // eslint-disable-next-line prefer-spread
        this.createMethod.apply(null, args);
  ***REMOVED***
    delete(optionsOrCallback, cb) ***REMOVED***
        const [options, callback] = util_1.util.maybeOptionsOrCallback(optionsOrCallback, cb);
        const ignoreNotFound = options.ignoreNotFound;
        delete options.ignoreNotFound;
        const methodConfig = (typeof this.methods.delete === 'object' && this.methods.delete) || ***REMOVED***};
        const reqOpts = extend(true, ***REMOVED***
            method: 'DELETE',
            uri: '',
      ***REMOVED***, methodConfig.reqOpts, ***REMOVED***
            qs: options,
      ***REMOVED***);
        // The `request` method may have been overridden to hold any special
        // behavior. Ensure we call the original `request` method.
        ServiceObject.prototype.request.call(this, reqOpts, (err, ...args) => ***REMOVED***
            if (err) ***REMOVED***
                if (err.code === 404 && ignoreNotFound) ***REMOVED***
                    err = null;
              ***REMOVED***
          ***REMOVED***
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            callback(err, ...args);
      ***REMOVED***);
  ***REMOVED***
    exists(optionsOrCallback, cb) ***REMOVED***
        const [options, callback] = util_1.util.maybeOptionsOrCallback(optionsOrCallback, cb);
        this.get(options, err => ***REMOVED***
            if (err) ***REMOVED***
                if (err.code === 404) ***REMOVED***
                    callback(null, false);
              ***REMOVED***
                else ***REMOVED***
                    callback(err);
              ***REMOVED***
                return;
          ***REMOVED***
            callback(null, true);
      ***REMOVED***);
  ***REMOVED***
    get(optionsOrCallback, cb) ***REMOVED***
        // eslint-disable-next-line @typescript-eslint/no-this-alias
        const self = this;
        const [opts, callback] = util_1.util.maybeOptionsOrCallback(optionsOrCallback, cb);
        const options = Object.assign(***REMOVED***}, opts);
        const autoCreate = options.autoCreate && typeof this.create === 'function';
        delete options.autoCreate;
        function onCreate(err, instance, apiResponse) ***REMOVED***
            if (err) ***REMOVED***
                if (err.code === 409) ***REMOVED***
                    self.get(options, callback);
                    return;
              ***REMOVED***
                callback(err, null, apiResponse);
                return;
          ***REMOVED***
            callback(null, instance, apiResponse);
      ***REMOVED***
        this.getMetadata(options, (err, metadata) => ***REMOVED***
            if (err) ***REMOVED***
                if (err.code === 404 && autoCreate) ***REMOVED***
                    const args = [];
                    if (Object.keys(options).length > 0) ***REMOVED***
                        args.push(options);
                  ***REMOVED***
                    args.push(onCreate);
                    self.create(...args);
                    return;
              ***REMOVED***
                callback(err, null, metadata);
                return;
          ***REMOVED***
            callback(null, self, metadata);
      ***REMOVED***);
  ***REMOVED***
    getMetadata(optionsOrCallback, cb) ***REMOVED***
        const [options, callback] = util_1.util.maybeOptionsOrCallback(optionsOrCallback, cb);
        const methodConfig = (typeof this.methods.getMetadata === 'object' &&
            this.methods.getMetadata) ||
            ***REMOVED***};
        const reqOpts = extend(true, ***REMOVED***
            uri: '',
      ***REMOVED***, methodConfig.reqOpts, ***REMOVED***
            qs: options,
      ***REMOVED***);
        // The `request` method may have been overridden to hold any special
        // behavior. Ensure we call the original `request` method.
        ServiceObject.prototype.request.call(this, reqOpts, (err, body, res) => ***REMOVED***
            this.metadata = body;
            callback(err, this.metadata, res);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Return the user's custom request interceptors.
     */
    getRequestInterceptors() ***REMOVED***
        // Interceptors should be returned in the order they were assigned.
        const localInterceptors = this.interceptors
            .filter(interceptor => typeof interceptor.request === 'function')
            .map(interceptor => interceptor.request);
        return this.parent.getRequestInterceptors().concat(localInterceptors);
  ***REMOVED***
    setMetadata(metadata, optionsOrCallback, cb) ***REMOVED***
        const [options, callback] = util_1.util.maybeOptionsOrCallback(optionsOrCallback, cb);
        const methodConfig = (typeof this.methods.setMetadata === 'object' &&
            this.methods.setMetadata) ||
            ***REMOVED***};
        const reqOpts = extend(true, ***REMOVED***}, ***REMOVED***
            method: 'PATCH',
            uri: '',
      ***REMOVED***, methodConfig.reqOpts, ***REMOVED***
            json: metadata,
            qs: options,
      ***REMOVED***);
        // The `request` method may have been overridden to hold any special
        // behavior. Ensure we call the original `request` method.
        ServiceObject.prototype.request.call(this, reqOpts, (err, body, res) => ***REMOVED***
            this.metadata = body;
            callback(err, this.metadata, res);
      ***REMOVED***);
  ***REMOVED***
    request_(reqOpts, callback) ***REMOVED***
        reqOpts = extend(true, ***REMOVED***}, reqOpts);
        if (this.projectId) ***REMOVED***
            reqOpts.projectId = this.projectId;
      ***REMOVED***
        const isAbsoluteUrl = reqOpts.uri.indexOf('http') === 0;
        const uriComponents = [this.baseUrl, this.id || '', reqOpts.uri];
        if (isAbsoluteUrl) ***REMOVED***
            uriComponents.splice(0, uriComponents.indexOf(reqOpts.uri));
      ***REMOVED***
        reqOpts.uri = uriComponents
            .filter(x => x.trim()) // Limit to non-empty strings.
            .map(uriComponent => ***REMOVED***
            const trimSlashesRegex = /^\/*|\/*$/g;
            return uriComponent.replace(trimSlashesRegex, '');
      ***REMOVED***)
            .join('/');
        const childInterceptors = Array.isArray(reqOpts.interceptors_)
            ? reqOpts.interceptors_
            : [];
        const localInterceptors = [].slice.call(this.interceptors);
        reqOpts.interceptors_ = childInterceptors.concat(localInterceptors);
        if (reqOpts.shouldReturnStream) ***REMOVED***
            return this.parent.requestStream(reqOpts);
      ***REMOVED***
        this.parent.request(reqOpts, callback);
  ***REMOVED***
    request(reqOpts, callback) ***REMOVED***
        this.request_(reqOpts, callback);
  ***REMOVED***
    /**
     * Make an authenticated API request.
     *
     * @param ***REMOVED***object} reqOpts - Request options that are passed to `request`.
     * @param ***REMOVED***string} reqOpts.uri - A URI relative to the baseUrl.
     */
    requestStream(reqOpts) ***REMOVED***
        const opts = extend(true, reqOpts, ***REMOVED*** shouldReturnStream: true });
        return this.request_(opts);
  ***REMOVED***
}
exports.ServiceObject = ServiceObject;
(0, promisify_1.promisifyAll)(ServiceObject, ***REMOVED*** exclude: ['getRequestInterceptors'] });
