// gcp-metadata/build/src/gcp-residency.js
"use strict";
/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.detectGCPResidency = exports.isGoogleComputeEngine = exports.isGoogleComputeEngineMACAddress = exports.isGoogleComputeEngineLinux = exports.isGoogleCloudServerless = exports.GCE_LINUX_BIOS_PATHS = void 0;
const fs_1 = require("fs");
const os_1 = require("os");
/**
 * Known paths unique to Google Compute Engine Linux instances
 */
exports.GCE_LINUX_BIOS_PATHS = ***REMOVED***
    BIOS_DATE: '/sys/class/dmi/id/bios_date',
    BIOS_VENDOR: '/sys/class/dmi/id/bios_vendor',
};
const GCE_MAC_ADDRESS_REGEX = /^42:01/;
/**
 * Determines if the process is running on a Google Cloud Serverless environment (Cloud Run or Cloud Functions instance).
 *
 * Uses the:
 * - ***REMOVED***@link https://cloud.google.com/run/docs/container-contract#env-vars Cloud Run environment variables}.
 * - ***REMOVED***@link https://cloud.google.com/functions/docs/env-var Cloud Functions environment variables}.
 *
 * @returns ***REMOVED***boolean} `true` if the process is running on GCP serverless, `false` otherwise.
 */
function isGoogleCloudServerless() ***REMOVED***
    /**
     * `CLOUD_RUN_JOB` is used for Cloud Run Jobs
     * - See ***REMOVED***@link https://cloud.google.com/run/docs/container-contract#env-vars Cloud Run environment variables}.
     *
     * `FUNCTION_NAME` is used in older Cloud Functions environments:
     * - See ***REMOVED***@link https://cloud.google.com/functions/docs/env-var Python 3.7 and Go 1.11}.
     *
     * `K_SERVICE` is used in Cloud Run and newer Cloud Functions environments:
     * - See ***REMOVED***@link https://cloud.google.com/run/docs/container-contract#env-vars Cloud Run environment variables}.
     * - See ***REMOVED***@link https://cloud.google.com/functions/docs/env-var Cloud Functions newer runtimes}.
     */
    const isGFEnvironment = process.env.CLOUD_RUN_JOB ||
        process.env.FUNCTION_NAME ||
        process.env.K_SERVICE;
    return !!isGFEnvironment;
}
exports.isGoogleCloudServerless = isGoogleCloudServerless;
/**
 * Determines if the process is running on a Linux Google Compute Engine instance.
 *
 * @returns ***REMOVED***boolean} `true` if the process is running on Linux GCE, `false` otherwise.
 */
function isGoogleComputeEngineLinux() ***REMOVED***
    if ((0, os_1.platform)() !== 'linux')
        return false;
    try ***REMOVED***
        // ensure this file exist
        (0, fs_1.statSync)(exports.GCE_LINUX_BIOS_PATHS.BIOS_DATE);
        // ensure this file exist and matches
        const biosVendor = (0, fs_1.readFileSync)(exports.GCE_LINUX_BIOS_PATHS.BIOS_VENDOR, 'utf8');
        return /Google/.test(biosVendor);
  ***REMOVED***
    catch (_a) ***REMOVED***
        return false;
  ***REMOVED***
}
exports.isGoogleComputeEngineLinux = isGoogleComputeEngineLinux;
/**
 * Determines if the process is running on a Google Compute Engine instance with a known
 * MAC address.
 *
 * @returns ***REMOVED***boolean} `true` if the process is running on GCE (as determined by MAC address), `false` otherwise.
 */
function isGoogleComputeEngineMACAddress() ***REMOVED***
    const interfaces = (0, os_1.networkInterfaces)();
    for (const item of Object.values(interfaces)) ***REMOVED***
        if (!item)
            continue;
        for (const ***REMOVED*** mac } of item) ***REMOVED***
            if (GCE_MAC_ADDRESS_REGEX.test(mac)) ***REMOVED***
                return true;
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    return false;
}
exports.isGoogleComputeEngineMACAddress = isGoogleComputeEngineMACAddress;
/**
 * Determines if the process is running on a Google Compute Engine instance.
 *
 * @returns ***REMOVED***boolean} `true` if the process is running on GCE, `false` otherwise.
 */
function isGoogleComputeEngine() ***REMOVED***
    return isGoogleComputeEngineLinux() || isGoogleComputeEngineMACAddress();
}
exports.isGoogleComputeEngine = isGoogleComputeEngine;
/**
 * Determines if the process is running on Google Cloud Platform.
 *
 * @returns ***REMOVED***boolean} `true` if the process is running on GCP, `false` otherwise.
 */
function detectGCPResidency() ***REMOVED***
    return isGoogleCloudServerless() || isGoogleComputeEngine();
}
exports.detectGCPResidency = detectGCPResidency;
