// js2xmlparser/lib/main.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.parse = exports.parseToExistingElement = exports.Absent = void 0;
/**
 * Copyright (C) 2016-2020 Michael Kourlas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var xmlcreate_1 = require("xmlcreate");
var options_1 = require("./options");
var utils_1 = require("./utils");
/**
 * Indicates that an object of a particular type should be suppressed from the
 * XML output.
 *
 * See the `typeHandlers` property in ***REMOVED***@link IOptions} for more details.
 */
var Absent = /** @class */ (function () ***REMOVED***
    function Absent() ***REMOVED***
  ***REMOVED***
    Object.defineProperty(Absent, "instance", ***REMOVED***
        /**
         * Returns the sole instance of Absent.
         */
        get: function () ***REMOVED***
            return Absent._instance;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Absent._instance = new Absent();
    return Absent;
}());
exports.Absent = Absent;
/**
 * Gets the type handler associated with a value.
 */
function getHandler(value, options) ***REMOVED***
    var type = Object.prototype.toString.call(value);
    var handler;
    if (Object.prototype.hasOwnProperty.call(options.typeHandlers, "*")) ***REMOVED***
        handler = options.typeHandlers["*"];
  ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(options.typeHandlers, type)) ***REMOVED***
        handler = options.typeHandlers[type];
  ***REMOVED***
    return handler;
}
/**
 * Parses a string into XML and adds it to the parent element or attribute.
 */
function parseString(str, parentElement, options) ***REMOVED***
    var requiresCdata = function (s) ***REMOVED***
        return ((options.cdataInvalidChars &&
            (s.indexOf("<") !== -1 || s.indexOf("&") !== -1)) ||
            options.cdataKeys.indexOf(parentElement.name) !== -1 ||
            options.cdataKeys.indexOf("*") !== -1);
  ***REMOVED***;
    if (parentElement instanceof xmlcreate_1.XmlElement) ***REMOVED***
        if (requiresCdata(str)) ***REMOVED***
            var cdataStrs = str.split("]]>");
            for (var i = 0; i < cdataStrs.length; i++) ***REMOVED***
                if (requiresCdata(cdataStrs[i])) ***REMOVED***
                    parentElement.cdata(***REMOVED***
                        charData: cdataStrs[i],
                        replaceInvalidCharsInCharData: options.replaceInvalidChars,
                  ***REMOVED***);
              ***REMOVED***
                else ***REMOVED***
                    parentElement.charData(***REMOVED***
                        charData: cdataStrs[i],
                        replaceInvalidCharsInCharData: options.replaceInvalidChars,
                  ***REMOVED***);
              ***REMOVED***
                if (i < cdataStrs.length - 1) ***REMOVED***
                    parentElement.charData(***REMOVED***
                        charData: "]]>",
                        replaceInvalidCharsInCharData: options.replaceInvalidChars,
                  ***REMOVED***);
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        else ***REMOVED***
            parentElement.charData(***REMOVED***
                charData: str,
                replaceInvalidCharsInCharData: options.replaceInvalidChars,
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***
    else ***REMOVED***
        parentElement.text(***REMOVED***
            charData: str,
            replaceInvalidCharsInCharData: options.replaceInvalidChars,
      ***REMOVED***);
  ***REMOVED***
}
/**
 * Parses an attribute into XML and adds it to the parent element.
 */
function parseAttribute(name, value, parentElement, options) ***REMOVED***
    var attribute = parentElement.attribute(***REMOVED***
        name: name,
        replaceInvalidCharsInName: options.replaceInvalidChars,
  ***REMOVED***);
    parseString((0, utils_1.stringify)(value), attribute, options);
}
/**
 * Parses an object or Map entry into XML and adds it to the parent element.
 */
function parseObjectOrMapEntry(key, value, parentElement, options) ***REMOVED***
    // Alias key
    if (key === options.aliasString) ***REMOVED***
        parentElement.name = (0, utils_1.stringify)(value);
        return;
  ***REMOVED***
    // Attributes key
    if (key.indexOf(options.attributeString) === 0 && (0, utils_1.isObject)(value)) ***REMOVED***
        for (var _i = 0, _a = Object.keys(value); _i < _a.length; _i++) ***REMOVED***
            var subkey = _a[_i];
            parseAttribute(subkey, (0, utils_1.stringify)(value[subkey]), parentElement, options);
      ***REMOVED***
        return;
  ***REMOVED***
    // Value key
    if (key.indexOf(options.valueString) === 0) ***REMOVED***
        parseValue(key, (0, utils_1.stringify)(value), parentElement, options);
        return;
  ***REMOVED***
    // Standard handling (create new element for entry)
    var element = parentElement;
    if (!(0, utils_1.isArray)(value) && !(0, utils_1.isSet)(value)) ***REMOVED***
        // If handler for value returns absent, then do not add element
        var handler = getHandler(value, options);
        if (!(0, utils_1.isUndefined)(handler)) ***REMOVED***
            if (handler(value) === Absent.instance) ***REMOVED***
                return;
          ***REMOVED***
      ***REMOVED***
        element = parentElement.element(***REMOVED***
            name: key,
            replaceInvalidCharsInName: options.replaceInvalidChars,
            useSelfClosingTagIfEmpty: options.useSelfClosingTagIfEmpty,
      ***REMOVED***);
  ***REMOVED***
    parseValue(key, value, element, options);
}
/**
 * Parses an Object or Map into XML and adds it to the parent element.
 */
function parseObjectOrMap(objectOrMap, parentElement, options) ***REMOVED***
    if ((0, utils_1.isMap)(objectOrMap)) ***REMOVED***
        objectOrMap.forEach(function (value, key) ***REMOVED***
            parseObjectOrMapEntry((0, utils_1.stringify)(key), value, parentElement, options);
      ***REMOVED***);
  ***REMOVED***
    else ***REMOVED***
        for (var _i = 0, _a = Object.keys(objectOrMap); _i < _a.length; _i++) ***REMOVED***
            var key = _a[_i];
            parseObjectOrMapEntry(key, objectOrMap[key], parentElement, options);
      ***REMOVED***
  ***REMOVED***
}
/**
 * Parses an array or Set into XML and adds it to the parent element.
 */
function parseArrayOrSet(key, arrayOrSet, parentElement, options) ***REMOVED***
    var arrayNameFunc;
    if (Object.prototype.hasOwnProperty.call(options.wrapHandlers, "*")) ***REMOVED***
        arrayNameFunc = options.wrapHandlers["*"];
  ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(options.wrapHandlers, key)) ***REMOVED***
        arrayNameFunc = options.wrapHandlers[key];
  ***REMOVED***
    var arrayKey = key;
    var arrayElement = parentElement;
    if (!(0, utils_1.isUndefined)(arrayNameFunc)) ***REMOVED***
        var arrayNameFuncKey = arrayNameFunc(arrayKey, arrayOrSet);
        if (!(0, utils_1.isNull)(arrayNameFuncKey)) ***REMOVED***
            arrayKey = arrayNameFuncKey;
            arrayElement = parentElement.element(***REMOVED***
                name: key,
                replaceInvalidCharsInName: options.replaceInvalidChars,
                useSelfClosingTagIfEmpty: options.useSelfClosingTagIfEmpty,
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***
    arrayOrSet.forEach(function (item) ***REMOVED***
        var element = arrayElement;
        if (!(0, utils_1.isArray)(item) && !(0, utils_1.isSet)(item)) ***REMOVED***
            // If handler for value returns absent, then do not add element
            var handler = getHandler(item, options);
            if (!(0, utils_1.isUndefined)(handler)) ***REMOVED***
                if (handler(item) === Absent.instance) ***REMOVED***
                    return;
              ***REMOVED***
          ***REMOVED***
            element = arrayElement.element(***REMOVED***
                name: arrayKey,
                replaceInvalidCharsInName: options.replaceInvalidChars,
                useSelfClosingTagIfEmpty: options.useSelfClosingTagIfEmpty,
          ***REMOVED***);
      ***REMOVED***
        parseValue(arrayKey, item, element, options);
  ***REMOVED***);
}
/**
 * Parses an arbitrary JavaScript value into XML and adds it to the parent
 * element.
 */
function parseValue(key, value, parentElement, options) ***REMOVED***
    // If a handler for a particular type is user-defined, use that handler
    // instead of the defaults
    var handler = getHandler(value, options);
    if (!(0, utils_1.isUndefined)(handler)) ***REMOVED***
        value = handler(value);
  ***REMOVED***
    if ((0, utils_1.isObject)(value) || (0, utils_1.isMap)(value)) ***REMOVED***
        parseObjectOrMap(value, parentElement, options);
        return;
  ***REMOVED***
    if ((0, utils_1.isArray)(value) || (0, utils_1.isSet)(value)) ***REMOVED***
        parseArrayOrSet(key, value, parentElement, options);
        return;
  ***REMOVED***
    parseString((0, utils_1.stringify)(value), parentElement, options);
}
/**
 * Converts the specified object to XML and adds the XML representation to the
 * specified XmlElement object using the specified options.
 *
 * This function does not add a root element. In addition, it does not add an
 * XML declaration or DTD, and the associated options in ***REMOVED***@link IOptions} are
 * ignored. If desired, these must be added manually.
 */
function parseToExistingElement(element, object, options) ***REMOVED***
    var opts = new options_1.Options(options);
    parseValue(element.name, object, element, opts);
}
exports.parseToExistingElement = parseToExistingElement;
/**
 * Returns a XML string representation of the specified object using the
 * specified options.
 *
 * `root` is the name of the root XML element. When the object is converted
 * to XML, it will be a child of this root element.
 */
function parse(root, object, options) ***REMOVED***
    var opts = new options_1.Options(options);
    var document = new xmlcreate_1.XmlDocument(***REMOVED***
        validation: opts.validation,
  ***REMOVED***);
    if (opts.declaration.include) ***REMOVED***
        document.decl(opts.declaration);
  ***REMOVED***
    if (opts.dtd.include) ***REMOVED***
        document.dtd(***REMOVED***
            // Validated in options.ts
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            name: opts.dtd.name,
            pubId: opts.dtd.pubId,
            sysId: opts.dtd.sysId,
      ***REMOVED***);
  ***REMOVED***
    var rootElement = document.element(***REMOVED***
        name: root,
        replaceInvalidCharsInName: opts.replaceInvalidChars,
        useSelfClosingTagIfEmpty: opts.useSelfClosingTagIfEmpty,
  ***REMOVED***);
    parseToExistingElement(rootElement, object, options);
    return document.toString(opts.format);
}
exports.parse = parse;
