// uuid/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
});
Object.defineProperty(exports, "v1", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _v.default;
***REMOVED***
});
Object.defineProperty(exports, "v3", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _v2.default;
***REMOVED***
});
Object.defineProperty(exports, "v4", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _v3.default;
***REMOVED***
});
Object.defineProperty(exports, "v5", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _v4.default;
***REMOVED***
});
Object.defineProperty(exports, "NIL", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _nil.default;
***REMOVED***
});
Object.defineProperty(exports, "version", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _version.default;
***REMOVED***
});
Object.defineProperty(exports, "validate", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _validate.default;
***REMOVED***
});
Object.defineProperty(exports, "stringify", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _stringify.default;
***REMOVED***
});
Object.defineProperty(exports, "parse", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _parse.default;
***REMOVED***
});

var _v = _interopRequireDefault(require("./v1.js"));

var _v2 = _interopRequireDefault(require("./v3.js"));

var _v3 = _interopRequireDefault(require("./v4.js"));

var _v4 = _interopRequireDefault(require("./v5.js"));

var _nil = _interopRequireDefault(require("./nil.js"));

var _version = _interopRequireDefault(require("./version.js"));

var _validate = _interopRequireDefault(require("./validate.js"));

var _stringify = _interopRequireDefault(require("./stringify.js"));

var _parse = _interopRequireDefault(require("./parse.js"));

function _interopRequireDefault(obj) ***REMOVED*** return obj && obj.__esModule ? obj : ***REMOVED*** default: obj }; }
