// @firebase/app-check-compat/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', ***REMOVED*** value: true });

var firebase = require('@firebase/app-compat');
var component = require('@firebase/component');
var appCheck = require('@firebase/app-check');
var util = require('@firebase/util');

function _interopDefaultLegacy (e) ***REMOVED*** return e && typeof e === 'object' && 'default' in e ? e : ***REMOVED*** 'default': e }; }

var firebase__default = /*#__PURE__*/_interopDefaultLegacy(firebase);

var name = "@firebase/app-check-compat";
var version = "0.3.1";

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var _a;
var ERRORS = (_a = ***REMOVED***},
    _a["use-before-activation" /* AppCheckError.USE_BEFORE_ACTIVATION */] = 'App Check is being used before activate() is called for FirebaseApp ***REMOVED***$appName}. ' +
        'Call activate() before instantiating other Firebase services.',
    _a);
var ERROR_FACTORY = new util.ErrorFactory('appCheck', 'AppCheck', ERRORS);

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var AppCheckService = /** @class */ (function () ***REMOVED***
    function AppCheckService(app) ***REMOVED***
        this.app = app;
  ***REMOVED***
    AppCheckService.prototype.activate = function (siteKeyOrProvider, isTokenAutoRefreshEnabled) ***REMOVED***
        var provider;
        if (typeof siteKeyOrProvider === 'string') ***REMOVED***
            provider = new appCheck.ReCaptchaV3Provider(siteKeyOrProvider);
      ***REMOVED***
        else if (siteKeyOrProvider instanceof appCheck.ReCaptchaEnterpriseProvider ||
            siteKeyOrProvider instanceof appCheck.ReCaptchaV3Provider ||
            siteKeyOrProvider instanceof appCheck.CustomProvider) ***REMOVED***
            provider = siteKeyOrProvider;
      ***REMOVED***
        else ***REMOVED***
            provider = new appCheck.CustomProvider(***REMOVED*** getToken: siteKeyOrProvider.getToken });
      ***REMOVED***
        this._delegate = appCheck.initializeAppCheck(this.app, ***REMOVED***
            provider: provider,
            isTokenAutoRefreshEnabled: isTokenAutoRefreshEnabled
      ***REMOVED***);
  ***REMOVED***;
    AppCheckService.prototype.setTokenAutoRefreshEnabled = function (isTokenAutoRefreshEnabled) ***REMOVED***
        if (!this._delegate) ***REMOVED***
            throw ERROR_FACTORY.create("use-before-activation" /* AppCheckError.USE_BEFORE_ACTIVATION */, ***REMOVED***
                appName: this.app.name
          ***REMOVED***);
      ***REMOVED***
        appCheck.setTokenAutoRefreshEnabled(this._delegate, isTokenAutoRefreshEnabled);
  ***REMOVED***;
    AppCheckService.prototype.getToken = function (forceRefresh) ***REMOVED***
        if (!this._delegate) ***REMOVED***
            throw ERROR_FACTORY.create("use-before-activation" /* AppCheckError.USE_BEFORE_ACTIVATION */, ***REMOVED***
                appName: this.app.name
          ***REMOVED***);
      ***REMOVED***
        return appCheck.getToken(this._delegate, forceRefresh);
  ***REMOVED***;
    AppCheckService.prototype.onTokenChanged = function (onNextOrObserver, onError, onCompletion) ***REMOVED***
        if (!this._delegate) ***REMOVED***
            throw ERROR_FACTORY.create("use-before-activation" /* AppCheckError.USE_BEFORE_ACTIVATION */, ***REMOVED***
                appName: this.app.name
          ***REMOVED***);
      ***REMOVED***
        return appCheck.onTokenChanged(this._delegate, 
        /**
         * Exp onTokenChanged() will handle both overloads but we need
         * to specify one to not confuse Typescript.
         */
        onNextOrObserver, onError, onCompletion);
  ***REMOVED***;
    return AppCheckService;
}());

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var factory = function (container) ***REMOVED***
    // Dependencies
    var app = container.getProvider('app-compat').getImmediate();
    return new AppCheckService(app);
};
function registerAppCheck() ***REMOVED***
    firebase__default["default"].INTERNAL.registerComponent(new component.Component('appCheck-compat', factory, "PUBLIC" /* ComponentType.PUBLIC */).setServiceProps(***REMOVED***
        ReCaptchaEnterpriseProvider: appCheck.ReCaptchaEnterpriseProvider,
        ReCaptchaV3Provider: appCheck.ReCaptchaV3Provider,
        CustomProvider: appCheck.CustomProvider
  ***REMOVED***));
}
registerAppCheck();
firebase__default["default"].registerVersion(name, version);

exports.registerAppCheck = registerAppCheck;
