// retry-request/index.js
'use strict';

const ***REMOVED***PassThrough} = require('stream');
const debug = require('debug')('retry-request');
const extend = require('extend');

const DEFAULTS = ***REMOVED***
  objectMode: false,
  retries: 2,

  /*
    The maximum time to delay in seconds. If retryDelayMultiplier results in a
    delay greater than maxRetryDelay, retries should delay by maxRetryDelay
    seconds instead.
  */
  maxRetryDelay: 64,

  /*
    The multiplier by which to increase the delay time between the completion of
    failed requests, and the initiation of the subsequent retrying request.
  */
  retryDelayMultiplier: 2,

  /*
    The length of time to keep retrying in seconds. The last sleep period will
    be shortened as necessary, so that the last retry runs at deadline (and not
    considerably beyond it).  The total time starting from when the initial
    request is sent, after which an error will be returned, regardless of the
    retrying attempts made meanwhile.
   */
  totalTimeout: 600,

  noResponseRetries: 2,
  currentRetryAttempt: 0,
  shouldRetryFn: function (response) ***REMOVED***
    const retryRanges = [
      // https://en.wikipedia.org/wiki/List_of_HTTP_status_codes
      // 1xx - Retry (Informational, request still processing)
      // 2xx - Do not retry (Success)
      // 3xx - Do not retry (Redirect)
      // 4xx - Do not retry (Client errors)
      // 429 - Retry ("Too Many Requests")
      // 5xx - Retry (Server errors)
      [100, 199],
      [429, 429],
      [500, 599],
    ];

    const statusCode = response.statusCode;
    debug(`Response status: $***REMOVED***statusCode}`);

    let range;
    while ((range = retryRanges.shift())) ***REMOVED***
      if (statusCode >= range[0] && statusCode <= range[1]) ***REMOVED***
        // Not a successful status or redirect.
        return true;
    ***REMOVED***
  ***REMOVED***
***REMOVED***,
};

function retryRequest(requestOpts, opts, callback) ***REMOVED***
  const streamMode = typeof arguments[arguments.length - 1] !== 'function';

  if (typeof opts === 'function') ***REMOVED***
    callback = opts;
***REMOVED***

  const manualCurrentRetryAttemptWasSet =
    opts && typeof opts.currentRetryAttempt === 'number';
  opts = extend(***REMOVED***}, DEFAULTS, opts);

  if (typeof opts.request === 'undefined') ***REMOVED***
    try ***REMOVED***
      // eslint-disable-next-line node/no-unpublished-require
      opts.request = require('request');
  ***REMOVED*** catch (e) ***REMOVED***
      throw new Error('A request library must be provided to retry-request.');
  ***REMOVED***
***REMOVED***

  let currentRetryAttempt = opts.currentRetryAttempt;

  let numNoResponseAttempts = 0;
  let streamResponseHandled = false;

  let retryStream;
  let requestStream;
  let delayStream;

  let activeRequest;
  const retryRequest = ***REMOVED***
    abort: function () ***REMOVED***
      if (activeRequest && activeRequest.abort) ***REMOVED***
        activeRequest.abort();
    ***REMOVED***
  ***REMOVED***,
***REMOVED***;

  if (streamMode) ***REMOVED***
    retryStream = new PassThrough(***REMOVED***objectMode: opts.objectMode});
    retryStream.abort = resetStreams;
***REMOVED***

  const timeOfFirstRequest = Date.now();
  if (currentRetryAttempt > 0) ***REMOVED***
    retryAfterDelay(currentRetryAttempt);
***REMOVED*** else ***REMOVED***
    makeRequest();
***REMOVED***

  if (streamMode) ***REMOVED***
    return retryStream;
***REMOVED*** else ***REMOVED***
    return retryRequest;
***REMOVED***

  function resetStreams() ***REMOVED***
    delayStream = null;

    if (requestStream) ***REMOVED***
      requestStream.abort && requestStream.abort();
      requestStream.cancel && requestStream.cancel();

      if (requestStream.destroy) ***REMOVED***
        requestStream.destroy();
    ***REMOVED*** else if (requestStream.end) ***REMOVED***
        requestStream.end();
    ***REMOVED***
  ***REMOVED***
***REMOVED***

  function makeRequest() ***REMOVED***
    currentRetryAttempt++;
    debug(`Current retry attempt: $***REMOVED***currentRetryAttempt}`);

    if (streamMode) ***REMOVED***
      streamResponseHandled = false;

      delayStream = new PassThrough(***REMOVED***objectMode: opts.objectMode});
      requestStream = opts.request(requestOpts);

      setImmediate(() => ***REMOVED***
        retryStream.emit('request');
    ***REMOVED***);

      requestStream
        // gRPC via google-cloud-node can emit an `error` as well as a `response`
        // Whichever it emits, we run with-- we can't run with both. That's what
        // is up with the `streamResponseHandled` tracking.
        .on('error', err => ***REMOVED***
          if (streamResponseHandled) ***REMOVED***
            return;
        ***REMOVED***

          streamResponseHandled = true;
          onResponse(err);
      ***REMOVED***)
        .on('response', (resp, body) => ***REMOVED***
          if (streamResponseHandled) ***REMOVED***
            return;
        ***REMOVED***

          streamResponseHandled = true;
          onResponse(null, resp, body);
      ***REMOVED***)
        .on('complete', retryStream.emit.bind(retryStream, 'complete'));

      requestStream.pipe(delayStream);
  ***REMOVED*** else ***REMOVED***
      activeRequest = opts.request(requestOpts, onResponse);
  ***REMOVED***
***REMOVED***

  function retryAfterDelay(currentRetryAttempt) ***REMOVED***
    if (streamMode) ***REMOVED***
      resetStreams();
  ***REMOVED***

    const nextRetryDelay = getNextRetryDelay(***REMOVED***
      maxRetryDelay: opts.maxRetryDelay,
      retryDelayMultiplier: opts.retryDelayMultiplier,
      retryNumber: currentRetryAttempt,
      timeOfFirstRequest,
      totalTimeout: opts.totalTimeout,
  ***REMOVED***);
    debug(`Next retry delay: $***REMOVED***nextRetryDelay}`);

    if (nextRetryDelay <= 0) ***REMOVED***
      numNoResponseAttempts = opts.noResponseRetries + 1;
      return;
  ***REMOVED***

    setTimeout(makeRequest, nextRetryDelay);
***REMOVED***

  function onResponse(err, response, body) ***REMOVED***
    // An error such as DNS resolution.
    if (err) ***REMOVED***
      numNoResponseAttempts++;

      if (numNoResponseAttempts <= opts.noResponseRetries) ***REMOVED***
        retryAfterDelay(numNoResponseAttempts);
    ***REMOVED*** else ***REMOVED***
        if (streamMode) ***REMOVED***
          retryStream.emit('error', err);
          retryStream.end();
      ***REMOVED*** else ***REMOVED***
          callback(err, response, body);
      ***REMOVED***
    ***REMOVED***

      return;
  ***REMOVED***

    // Send the response to see if we should try again.
    // NOTE: "currentRetryAttempt" isn't accurate by default, as it counts
    // the very first request sent as the first "retry". It is only accurate
    // when a user provides their own "currentRetryAttempt" option at
    // instantiation.
    const adjustedCurrentRetryAttempt = manualCurrentRetryAttemptWasSet
      ? currentRetryAttempt
      : currentRetryAttempt - 1;
    if (
      adjustedCurrentRetryAttempt < opts.retries &&
      opts.shouldRetryFn(response)
    ) ***REMOVED***
      retryAfterDelay(currentRetryAttempt);
      return;
  ***REMOVED***

    // No more attempts need to be made, just continue on.
    if (streamMode) ***REMOVED***
      retryStream.emit('response', response);
      delayStream.pipe(retryStream);
      requestStream.on('error', err => ***REMOVED***
        retryStream.destroy(err);
    ***REMOVED***);
  ***REMOVED*** else ***REMOVED***
      callback(err, response, body);
  ***REMOVED***
***REMOVED***
}

module.exports = retryRequest;

function getNextRetryDelay(config) ***REMOVED***
  const ***REMOVED***
    maxRetryDelay,
    retryDelayMultiplier,
    retryNumber,
    timeOfFirstRequest,
    totalTimeout,
***REMOVED*** = config;

  const maxRetryDelayMs = maxRetryDelay * 1000;
  const totalTimeoutMs = totalTimeout * 1000;

  const jitter = Math.floor(Math.random() * 1000);
  const calculatedNextRetryDelay =
    Math.pow(retryDelayMultiplier, retryNumber) * 1000 + jitter;

  const maxAllowableDelayMs =
    totalTimeoutMs - (Date.now() - timeOfFirstRequest);

  return Math.min(
    calculatedNextRetryDelay,
    maxAllowableDelayMs,
    maxRetryDelayMs
  );
}

module.exports.getNextRetryDelay = getNextRetryDelay;
