// @firebase/performance/dist/esm/index.esm.js
import ***REMOVED*** ErrorFactory, areCookiesEnabled, isIndexedDBAvailable, validateIndexedDBOpenable, getModularInstance, deepEqual } from '@firebase/util';
import ***REMOVED*** Logger, LogLevel } from '@firebase/logger';
import ***REMOVED*** __spreadArray, __assign } from 'tslib';
import ***REMOVED*** getApp, _getProvider, _registerComponent, registerVersion } from '@firebase/app';
import ***REMOVED*** Component } from '@firebase/component';
import '@firebase/installations';

var name = "@firebase/performance";
var version = "0.6.1";

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var SDK_VERSION = version;
/** The prefix for start User Timing marks used for creating Traces. */
var TRACE_START_MARK_PREFIX = 'FB-PERF-TRACE-START';
/** The prefix for stop User Timing marks used for creating Traces. */
var TRACE_STOP_MARK_PREFIX = 'FB-PERF-TRACE-STOP';
/** The prefix for User Timing measure used for creating Traces. */
var TRACE_MEASURE_PREFIX = 'FB-PERF-TRACE-MEASURE';
/** The prefix for out of the box page load Trace name. */
var OOB_TRACE_PAGE_LOAD_PREFIX = '_wt_';
var FIRST_PAINT_COUNTER_NAME = '_fp';
var FIRST_CONTENTFUL_PAINT_COUNTER_NAME = '_fcp';
var FIRST_INPUT_DELAY_COUNTER_NAME = '_fid';
var CONFIG_LOCAL_STORAGE_KEY = '@firebase/performance/config';
var CONFIG_EXPIRY_LOCAL_STORAGE_KEY = '@firebase/performance/configexpire';
var SERVICE = 'performance';
var SERVICE_NAME = 'Performance';

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var _a;
var ERROR_DESCRIPTION_MAP = (_a = ***REMOVED***},
    _a["trace started" /* ErrorCode.TRACE_STARTED_BEFORE */] = 'Trace ***REMOVED***$traceName} was started before.',
    _a["trace stopped" /* ErrorCode.TRACE_STOPPED_BEFORE */] = 'Trace ***REMOVED***$traceName} is not running.',
    _a["nonpositive trace startTime" /* ErrorCode.NONPOSITIVE_TRACE_START_TIME */] = 'Trace ***REMOVED***$traceName} startTime should be positive.',
    _a["nonpositive trace duration" /* ErrorCode.NONPOSITIVE_TRACE_DURATION */] = 'Trace ***REMOVED***$traceName} duration should be positive.',
    _a["no window" /* ErrorCode.NO_WINDOW */] = 'Window is not available.',
    _a["no app id" /* ErrorCode.NO_APP_ID */] = 'App id is not available.',
    _a["no project id" /* ErrorCode.NO_PROJECT_ID */] = 'Project id is not available.',
    _a["no api key" /* ErrorCode.NO_API_KEY */] = 'Api key is not available.',
    _a["invalid cc log" /* ErrorCode.INVALID_CC_LOG */] = 'Attempted to queue invalid cc event',
    _a["FB not default" /* ErrorCode.FB_NOT_DEFAULT */] = 'Performance can only start when Firebase app instance is the default one.',
    _a["RC response not ok" /* ErrorCode.RC_NOT_OK */] = 'RC response is not ok',
    _a["invalid attribute name" /* ErrorCode.INVALID_ATTRIBUTE_NAME */] = 'Attribute name ***REMOVED***$attributeName} is invalid.',
    _a["invalid attribute value" /* ErrorCode.INVALID_ATTRIBUTE_VALUE */] = 'Attribute value ***REMOVED***$attributeValue} is invalid.',
    _a["invalid custom metric name" /* ErrorCode.INVALID_CUSTOM_METRIC_NAME */] = 'Custom metric name ***REMOVED***$customMetricName} is invalid',
    _a["invalid String merger input" /* ErrorCode.INVALID_STRING_MERGER_PARAMETER */] = 'Input for String merger is invalid, contact support team to resolve.',
    _a["already initialized" /* ErrorCode.ALREADY_INITIALIZED */] = 'initializePerformance() has already been called with ' +
        'different options. To avoid this error, call initializePerformance() with the ' +
        'same options as when it was originally called, or call getPerformance() to return the' +
        ' already initialized instance.',
    _a);
var ERROR_FACTORY = new ErrorFactory(SERVICE, SERVICE_NAME, ERROR_DESCRIPTION_MAP);

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var consoleLogger = new Logger(SERVICE_NAME);
consoleLogger.logLevel = LogLevel.INFO;

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var apiInstance;
var windowInstance;
/**
 * This class holds a reference to various browser related objects injected by
 * set methods.
 */
var Api = /** @class */ (function () ***REMOVED***
    function Api(window) ***REMOVED***
        this.window = window;
        if (!window) ***REMOVED***
            throw ERROR_FACTORY.create("no window" /* ErrorCode.NO_WINDOW */);
      ***REMOVED***
        this.performance = window.performance;
        this.PerformanceObserver = window.PerformanceObserver;
        this.windowLocation = window.location;
        this.navigator = window.navigator;
        this.document = window.document;
        if (this.navigator && this.navigator.cookieEnabled) ***REMOVED***
            // If user blocks cookies on the browser, accessing localStorage will
            // throw an exception.
            this.localStorage = window.localStorage;
      ***REMOVED***
        if (window.perfMetrics && window.perfMetrics.onFirstInputDelay) ***REMOVED***
            this.onFirstInputDelay = window.perfMetrics.onFirstInputDelay;
      ***REMOVED***
  ***REMOVED***
    Api.prototype.getUrl = function () ***REMOVED***
        // Do not capture the string query part of url.
        return this.windowLocation.href.split('?')[0];
  ***REMOVED***;
    Api.prototype.mark = function (name) ***REMOVED***
        if (!this.performance || !this.performance.mark) ***REMOVED***
            return;
      ***REMOVED***
        this.performance.mark(name);
  ***REMOVED***;
    Api.prototype.measure = function (measureName, mark1, mark2) ***REMOVED***
        if (!this.performance || !this.performance.measure) ***REMOVED***
            return;
      ***REMOVED***
        this.performance.measure(measureName, mark1, mark2);
  ***REMOVED***;
    Api.prototype.getEntriesByType = function (type) ***REMOVED***
        if (!this.performance || !this.performance.getEntriesByType) ***REMOVED***
            return [];
      ***REMOVED***
        return this.performance.getEntriesByType(type);
  ***REMOVED***;
    Api.prototype.getEntriesByName = function (name) ***REMOVED***
        if (!this.performance || !this.performance.getEntriesByName) ***REMOVED***
            return [];
      ***REMOVED***
        return this.performance.getEntriesByName(name);
  ***REMOVED***;
    Api.prototype.getTimeOrigin = function () ***REMOVED***
        // Polyfill the time origin with performance.timing.navigationStart.
        return (this.performance &&
            (this.performance.timeOrigin || this.performance.timing.navigationStart));
  ***REMOVED***;
    Api.prototype.requiredApisAvailable = function () ***REMOVED***
        if (!fetch || !Promise || !areCookiesEnabled()) ***REMOVED***
            consoleLogger.info('Firebase Performance cannot start if browser does not support fetch and Promise or cookie is disabled.');
            return false;
      ***REMOVED***
        if (!isIndexedDBAvailable()) ***REMOVED***
            consoleLogger.info('IndexedDB is not supported by current browswer');
            return false;
      ***REMOVED***
        return true;
  ***REMOVED***;
    Api.prototype.setupObserver = function (entryType, callback) ***REMOVED***
        if (!this.PerformanceObserver) ***REMOVED***
            return;
      ***REMOVED***
        var observer = new this.PerformanceObserver(function (list) ***REMOVED***
            for (var _i = 0, _a = list.getEntries(); _i < _a.length; _i++) ***REMOVED***
                var entry = _a[_i];
                // `entry` is a PerformanceEntry instance.
                callback(entry);
          ***REMOVED***
      ***REMOVED***);
        // Start observing the entry types you care about.
        observer.observe(***REMOVED*** entryTypes: [entryType] });
  ***REMOVED***;
    Api.getInstance = function () ***REMOVED***
        if (apiInstance === undefined) ***REMOVED***
            apiInstance = new Api(windowInstance);
      ***REMOVED***
        return apiInstance;
  ***REMOVED***;
    return Api;
}());
function setupApi(window) ***REMOVED***
    windowInstance = window;
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var iid;
function getIidPromise(installationsService) ***REMOVED***
    var iidPromise = installationsService.getId();
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    iidPromise.then(function (iidVal) ***REMOVED***
        iid = iidVal;
  ***REMOVED***);
    return iidPromise;
}
// This method should be used after the iid is retrieved by getIidPromise method.
function getIid() ***REMOVED***
    return iid;
}
function getAuthTokenPromise(installationsService) ***REMOVED***
    var authTokenPromise = installationsService.getToken();
    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    authTokenPromise.then(function (authTokenVal) ***REMOVED***
  ***REMOVED***);
    return authTokenPromise;
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function mergeStrings(part1, part2) ***REMOVED***
    var sizeDiff = part1.length - part2.length;
    if (sizeDiff < 0 || sizeDiff > 1) ***REMOVED***
        throw ERROR_FACTORY.create("invalid String merger input" /* ErrorCode.INVALID_STRING_MERGER_PARAMETER */);
  ***REMOVED***
    var resultArray = [];
    for (var i = 0; i < part1.length; i++) ***REMOVED***
        resultArray.push(part1.charAt(i));
        if (part2.length > i) ***REMOVED***
            resultArray.push(part2.charAt(i));
      ***REMOVED***
  ***REMOVED***
    return resultArray.join('');
}

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var settingsServiceInstance;
var SettingsService = /** @class */ (function () ***REMOVED***
    function SettingsService() ***REMOVED***
        // The variable which controls logging of automatic traces and HTTP/S network monitoring.
        this.instrumentationEnabled = true;
        // The variable which controls logging of custom traces.
        this.dataCollectionEnabled = true;
        // Configuration flags set through remote config.
        this.loggingEnabled = false;
        // Sampling rate between 0 and 1.
        this.tracesSamplingRate = 1;
        this.networkRequestsSamplingRate = 1;
        // Address of logging service.
        this.logEndPointUrl = 'https://firebaselogging.googleapis.com/v0cc/log?format=json_proto';
        // Performance event transport endpoint URL which should be compatible with proto3.
        // New Address for transport service, not configurable via Remote Config.
        this.flTransportEndpointUrl = mergeStrings('hts/frbslgigp.ogepscmv/ieo/eaylg', 'tp:/ieaeogn-agolai.o/1frlglgc/o');
        this.transportKey = mergeStrings('AzSC8r6ReiGqFMyfvgow', 'Iayx0u-XT3vksVM-pIV');
        // Source type for performance event logs.
        this.logSource = 462;
        // Flags which control per session logging of traces and network requests.
        this.logTraceAfterSampling = false;
        this.logNetworkAfterSampling = false;
        // TTL of config retrieved from remote config in hours.
        this.configTimeToLive = 12;
  ***REMOVED***
    SettingsService.prototype.getFlTransportFullUrl = function () ***REMOVED***
        return this.flTransportEndpointUrl.concat('?key=', this.transportKey);
  ***REMOVED***;
    SettingsService.getInstance = function () ***REMOVED***
        if (settingsServiceInstance === undefined) ***REMOVED***
            settingsServiceInstance = new SettingsService();
      ***REMOVED***
        return settingsServiceInstance;
  ***REMOVED***;
    return SettingsService;
}());

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var VisibilityState;
(function (VisibilityState) ***REMOVED***
    VisibilityState[VisibilityState["UNKNOWN"] = 0] = "UNKNOWN";
    VisibilityState[VisibilityState["VISIBLE"] = 1] = "VISIBLE";
    VisibilityState[VisibilityState["HIDDEN"] = 2] = "HIDDEN";
})(VisibilityState || (VisibilityState = ***REMOVED***}));
var RESERVED_ATTRIBUTE_PREFIXES = ['firebase_', 'google_', 'ga_'];
var ATTRIBUTE_FORMAT_REGEX = new RegExp('^[a-zA-Z]\\w*$');
var MAX_ATTRIBUTE_NAME_LENGTH = 40;
var MAX_ATTRIBUTE_VALUE_LENGTH = 100;
function getServiceWorkerStatus() ***REMOVED***
    var navigator = Api.getInstance().navigator;
    if (navigator === null || navigator === void 0 ? void 0 : navigator.serviceWorker) ***REMOVED***
        if (navigator.serviceWorker.controller) ***REMOVED***
            return 2 /* ServiceWorkerStatus.CONTROLLED */;
      ***REMOVED***
        else ***REMOVED***
            return 3 /* ServiceWorkerStatus.UNCONTROLLED */;
      ***REMOVED***
  ***REMOVED***
    else ***REMOVED***
        return 1 /* ServiceWorkerStatus.UNSUPPORTED */;
  ***REMOVED***
}
function getVisibilityState() ***REMOVED***
    var document = Api.getInstance().document;
    var visibilityState = document.visibilityState;
    switch (visibilityState) ***REMOVED***
        case 'visible':
            return VisibilityState.VISIBLE;
        case 'hidden':
            return VisibilityState.HIDDEN;
        default:
            return VisibilityState.UNKNOWN;
  ***REMOVED***
}
function getEffectiveConnectionType() ***REMOVED***
    var navigator = Api.getInstance().navigator;
    var navigatorConnection = navigator.connection;
    var effectiveType = navigatorConnection && navigatorConnection.effectiveType;
    switch (effectiveType) ***REMOVED***
        case 'slow-2g':
            return 1 /* EffectiveConnectionType.CONNECTION_SLOW_2G */;
        case '2g':
            return 2 /* EffectiveConnectionType.CONNECTION_2G */;
        case '3g':
            return 3 /* EffectiveConnectionType.CONNECTION_3G */;
        case '4g':
            return 4 /* EffectiveConnectionType.CONNECTION_4G */;
        default:
            return 0 /* EffectiveConnectionType.UNKNOWN */;
  ***REMOVED***
}
function isValidCustomAttributeName(name) ***REMOVED***
    if (name.length === 0 || name.length > MAX_ATTRIBUTE_NAME_LENGTH) ***REMOVED***
        return false;
  ***REMOVED***
    var matchesReservedPrefix = RESERVED_ATTRIBUTE_PREFIXES.some(function (prefix) ***REMOVED***
        return name.startsWith(prefix);
  ***REMOVED***);
    return !matchesReservedPrefix && !!name.match(ATTRIBUTE_FORMAT_REGEX);
}
function isValidCustomAttributeValue(value) ***REMOVED***
    return value.length !== 0 && value.length <= MAX_ATTRIBUTE_VALUE_LENGTH;
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function getAppId(firebaseApp) ***REMOVED***
    var _a;
    var appId = (_a = firebaseApp.options) === null || _a === void 0 ? void 0 : _a.appId;
    if (!appId) ***REMOVED***
        throw ERROR_FACTORY.create("no app id" /* ErrorCode.NO_APP_ID */);
  ***REMOVED***
    return appId;
}
function getProjectId(firebaseApp) ***REMOVED***
    var _a;
    var projectId = (_a = firebaseApp.options) === null || _a === void 0 ? void 0 : _a.projectId;
    if (!projectId) ***REMOVED***
        throw ERROR_FACTORY.create("no project id" /* ErrorCode.NO_PROJECT_ID */);
  ***REMOVED***
    return projectId;
}
function getApiKey(firebaseApp) ***REMOVED***
    var _a;
    var apiKey = (_a = firebaseApp.options) === null || _a === void 0 ? void 0 : _a.apiKey;
    if (!apiKey) ***REMOVED***
        throw ERROR_FACTORY.create("no api key" /* ErrorCode.NO_API_KEY */);
  ***REMOVED***
    return apiKey;
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var REMOTE_CONFIG_SDK_VERSION = '0.0.1';
// These values will be used if the remote config object is successfully
// retrieved, but the template does not have these fields.
var DEFAULT_CONFIGS = ***REMOVED***
    loggingEnabled: true
};
var FIS_AUTH_PREFIX = 'FIREBASE_INSTALLATIONS_AUTH';
function getConfig(performanceController, iid) ***REMOVED***
    var config = getStoredConfig();
    if (config) ***REMOVED***
        processConfig(config);
        return Promise.resolve();
  ***REMOVED***
    return getRemoteConfig(performanceController, iid)
        .then(processConfig)
        .then(function (config) ***REMOVED*** return storeConfig(config); }, 
    /** Do nothing for error, use defaults set in settings service. */
    function () ***REMOVED*** });
}
function getStoredConfig() ***REMOVED***
    var localStorage = Api.getInstance().localStorage;
    if (!localStorage) ***REMOVED***
        return;
  ***REMOVED***
    var expiryString = localStorage.getItem(CONFIG_EXPIRY_LOCAL_STORAGE_KEY);
    if (!expiryString || !configValid(expiryString)) ***REMOVED***
        return;
  ***REMOVED***
    var configStringified = localStorage.getItem(CONFIG_LOCAL_STORAGE_KEY);
    if (!configStringified) ***REMOVED***
        return;
  ***REMOVED***
    try ***REMOVED***
        var configResponse = JSON.parse(configStringified);
        return configResponse;
  ***REMOVED***
    catch (_a) ***REMOVED***
        return;
  ***REMOVED***
}
function storeConfig(config) ***REMOVED***
    var localStorage = Api.getInstance().localStorage;
    if (!config || !localStorage) ***REMOVED***
        return;
  ***REMOVED***
    localStorage.setItem(CONFIG_LOCAL_STORAGE_KEY, JSON.stringify(config));
    localStorage.setItem(CONFIG_EXPIRY_LOCAL_STORAGE_KEY, String(Date.now() +
        SettingsService.getInstance().configTimeToLive * 60 * 60 * 1000));
}
var COULD_NOT_GET_CONFIG_MSG = 'Could not fetch config, will use default configs';
function getRemoteConfig(performanceController, iid) ***REMOVED***
    // Perf needs auth token only to retrieve remote config.
    return getAuthTokenPromise(performanceController.installations)
        .then(function (authToken) ***REMOVED***
        var projectId = getProjectId(performanceController.app);
        var apiKey = getApiKey(performanceController.app);
        var configEndPoint = "https://firebaseremoteconfig.googleapis.com/v1/projects/".concat(projectId, "/namespaces/fireperf:fetch?key=").concat(apiKey);
        var request = new Request(configEndPoint, ***REMOVED***
            method: 'POST',
            headers: ***REMOVED*** Authorization: "".concat(FIS_AUTH_PREFIX, " ").concat(authToken) },
            /* eslint-disable camelcase */
            body: JSON.stringify(***REMOVED***
                app_instance_id: iid,
                app_instance_id_token: authToken,
                app_id: getAppId(performanceController.app),
                app_version: SDK_VERSION,
                sdk_version: REMOTE_CONFIG_SDK_VERSION
          ***REMOVED***)
            /* eslint-enable camelcase */
      ***REMOVED***);
        return fetch(request).then(function (response) ***REMOVED***
            if (response.ok) ***REMOVED***
                return response.json();
          ***REMOVED***
            // In case response is not ok. This will be caught by catch.
            throw ERROR_FACTORY.create("RC response not ok" /* ErrorCode.RC_NOT_OK */);
      ***REMOVED***);
  ***REMOVED***)
        .catch(function () ***REMOVED***
        consoleLogger.info(COULD_NOT_GET_CONFIG_MSG);
        return undefined;
  ***REMOVED***);
}
/**
 * Processes config coming either from calling RC or from local storage.
 * This method only runs if call is successful or config in storage
 * is valid.
 */
function processConfig(config) ***REMOVED***
    if (!config) ***REMOVED***
        return config;
  ***REMOVED***
    var settingsServiceInstance = SettingsService.getInstance();
    var entries = config.entries || ***REMOVED***};
    if (entries.fpr_enabled !== undefined) ***REMOVED***
        // TODO: Change the assignment of loggingEnabled once the received type is
        // known.
        settingsServiceInstance.loggingEnabled =
            String(entries.fpr_enabled) === 'true';
  ***REMOVED***
    else ***REMOVED***
        // Config retrieved successfully, but there is no fpr_enabled in template.
        // Use secondary configs value.
        settingsServiceInstance.loggingEnabled = DEFAULT_CONFIGS.loggingEnabled;
  ***REMOVED***
    if (entries.fpr_log_source) ***REMOVED***
        settingsServiceInstance.logSource = Number(entries.fpr_log_source);
  ***REMOVED***
    else if (DEFAULT_CONFIGS.logSource) ***REMOVED***
        settingsServiceInstance.logSource = DEFAULT_CONFIGS.logSource;
  ***REMOVED***
    if (entries.fpr_log_endpoint_url) ***REMOVED***
        settingsServiceInstance.logEndPointUrl = entries.fpr_log_endpoint_url;
  ***REMOVED***
    else if (DEFAULT_CONFIGS.logEndPointUrl) ***REMOVED***
        settingsServiceInstance.logEndPointUrl = DEFAULT_CONFIGS.logEndPointUrl;
  ***REMOVED***
    // Key from Remote Config has to be non-empty string, otherwsie use local value.
    if (entries.fpr_log_transport_key) ***REMOVED***
        settingsServiceInstance.transportKey = entries.fpr_log_transport_key;
  ***REMOVED***
    else if (DEFAULT_CONFIGS.transportKey) ***REMOVED***
        settingsServiceInstance.transportKey = DEFAULT_CONFIGS.transportKey;
  ***REMOVED***
    if (entries.fpr_vc_network_request_sampling_rate !== undefined) ***REMOVED***
        settingsServiceInstance.networkRequestsSamplingRate = Number(entries.fpr_vc_network_request_sampling_rate);
  ***REMOVED***
    else if (DEFAULT_CONFIGS.networkRequestsSamplingRate !== undefined) ***REMOVED***
        settingsServiceInstance.networkRequestsSamplingRate =
            DEFAULT_CONFIGS.networkRequestsSamplingRate;
  ***REMOVED***
    if (entries.fpr_vc_trace_sampling_rate !== undefined) ***REMOVED***
        settingsServiceInstance.tracesSamplingRate = Number(entries.fpr_vc_trace_sampling_rate);
  ***REMOVED***
    else if (DEFAULT_CONFIGS.tracesSamplingRate !== undefined) ***REMOVED***
        settingsServiceInstance.tracesSamplingRate =
            DEFAULT_CONFIGS.tracesSamplingRate;
  ***REMOVED***
    // Set the per session trace and network logging flags.
    settingsServiceInstance.logTraceAfterSampling = shouldLogAfterSampling(settingsServiceInstance.tracesSamplingRate);
    settingsServiceInstance.logNetworkAfterSampling = shouldLogAfterSampling(settingsServiceInstance.networkRequestsSamplingRate);
    return config;
}
function configValid(expiry) ***REMOVED***
    return Number(expiry) > Date.now();
}
function shouldLogAfterSampling(samplingRate) ***REMOVED***
    return Math.random() <= samplingRate;
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var initializationStatus = 1 /* InitializationStatus.notInitialized */;
var initializationPromise;
function getInitializationPromise(performanceController) ***REMOVED***
    initializationStatus = 2 /* InitializationStatus.initializationPending */;
    initializationPromise =
        initializationPromise || initializePerf(performanceController);
    return initializationPromise;
}
function isPerfInitialized() ***REMOVED***
    return initializationStatus === 3 /* InitializationStatus.initialized */;
}
function initializePerf(performanceController) ***REMOVED***
    return getDocumentReadyComplete()
        .then(function () ***REMOVED*** return getIidPromise(performanceController.installations); })
        .then(function (iid) ***REMOVED*** return getConfig(performanceController, iid); })
        .then(function () ***REMOVED*** return changeInitializationStatus(); }, function () ***REMOVED*** return changeInitializationStatus(); });
}
/**
 * Returns a promise which resolves whenever the document readystate is complete or
 * immediately if it is called after page load complete.
 */
function getDocumentReadyComplete() ***REMOVED***
    var document = Api.getInstance().document;
    return new Promise(function (resolve) ***REMOVED***
        if (document && document.readyState !== 'complete') ***REMOVED***
            var handler_1 = function () ***REMOVED***
                if (document.readyState === 'complete') ***REMOVED***
                    document.removeEventListener('readystatechange', handler_1);
                    resolve();
              ***REMOVED***
          ***REMOVED***;
            document.addEventListener('readystatechange', handler_1);
      ***REMOVED***
        else ***REMOVED***
            resolve();
      ***REMOVED***
  ***REMOVED***);
}
function changeInitializationStatus() ***REMOVED***
    initializationStatus = 3 /* InitializationStatus.initialized */;
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var DEFAULT_SEND_INTERVAL_MS = 10 * 1000;
var INITIAL_SEND_TIME_DELAY_MS = 5.5 * 1000;
// If end point does not work, the call will be tried for these many times.
var DEFAULT_REMAINING_TRIES = 3;
var MAX_EVENT_COUNT_PER_REQUEST = 1000;
var remainingTries = DEFAULT_REMAINING_TRIES;
/* eslint-enable camelcase */
var queue = [];
var isTransportSetup = false;
function setupTransportService() ***REMOVED***
    if (!isTransportSetup) ***REMOVED***
        processQueue(INITIAL_SEND_TIME_DELAY_MS);
        isTransportSetup = true;
  ***REMOVED***
}
function processQueue(timeOffset) ***REMOVED***
    setTimeout(function () ***REMOVED***
        // If there is no remainingTries left, stop retrying.
        if (remainingTries === 0) ***REMOVED***
            return;
      ***REMOVED***
        // If there are no events to process, wait for DEFAULT_SEND_INTERVAL_MS and try again.
        if (!queue.length) ***REMOVED***
            return processQueue(DEFAULT_SEND_INTERVAL_MS);
      ***REMOVED***
        dispatchQueueEvents();
  ***REMOVED***, timeOffset);
}
function dispatchQueueEvents() ***REMOVED***
    // Extract events up to the maximum cap of single logRequest from top of "official queue".
    // The staged events will be used for current logRequest attempt, remaining events will be kept
    // for next attempt.
    var staged = queue.splice(0, MAX_EVENT_COUNT_PER_REQUEST);
    /* eslint-disable camelcase */
    // We will pass the JSON serialized event to the backend.
    var log_event = staged.map(function (evt) ***REMOVED*** return (***REMOVED***
        source_extension_json_proto3: evt.message,
        event_time_ms: String(evt.eventTime)
  ***REMOVED***); });
    var data = ***REMOVED***
        request_time_ms: String(Date.now()),
        client_info: ***REMOVED***
            client_type: 1,
            js_client_info: ***REMOVED***}
      ***REMOVED***,
        log_source: SettingsService.getInstance().logSource,
        log_event: log_event
  ***REMOVED***;
    /* eslint-enable camelcase */
    sendEventsToFl(data, staged).catch(function () ***REMOVED***
        // If the request fails for some reason, add the events that were attempted
        // back to the primary queue to retry later.
        queue = __spreadArray(__spreadArray([], staged, true), queue, true);
        remainingTries--;
        consoleLogger.info("Tries left: ".concat(remainingTries, "."));
        processQueue(DEFAULT_SEND_INTERVAL_MS);
  ***REMOVED***);
}
function sendEventsToFl(data, staged) ***REMOVED***
    return postToFlEndpoint(data)
        .then(function (res) ***REMOVED***
        if (!res.ok) ***REMOVED***
            consoleLogger.info('Call to Firebase backend failed.');
      ***REMOVED***
        return res.json();
  ***REMOVED***)
        .then(function (res) ***REMOVED***
        // Find the next call wait time from the response.
        var transportWait = Number(res.nextRequestWaitMillis);
        var requestOffset = DEFAULT_SEND_INTERVAL_MS;
        if (!isNaN(transportWait)) ***REMOVED***
            requestOffset = Math.max(transportWait, requestOffset);
      ***REMOVED***
        // Delete request if response include RESPONSE_ACTION_UNKNOWN or DELETE_REQUEST action.
        // Otherwise, retry request using normal scheduling if response include RETRY_REQUEST_LATER.
        var logResponseDetails = res.logResponseDetails;
        if (Array.isArray(logResponseDetails) &&
            logResponseDetails.length > 0 &&
            logResponseDetails[0].responseAction === 'RETRY_REQUEST_LATER') ***REMOVED***
            queue = __spreadArray(__spreadArray([], staged, true), queue, true);
            consoleLogger.info("Retry transport request later.");
      ***REMOVED***
        remainingTries = DEFAULT_REMAINING_TRIES;
        // Schedule the next process.
        processQueue(requestOffset);
  ***REMOVED***);
}
function postToFlEndpoint(data) ***REMOVED***
    var flTransportFullUrl = SettingsService.getInstance().getFlTransportFullUrl();
    return fetch(flTransportFullUrl, ***REMOVED***
        method: 'POST',
        body: JSON.stringify(data)
  ***REMOVED***);
}
function addToQueue(evt) ***REMOVED***
    if (!evt.eventTime || !evt.message) ***REMOVED***
        throw ERROR_FACTORY.create("invalid cc log" /* ErrorCode.INVALID_CC_LOG */);
  ***REMOVED***
    // Add the new event to the queue.
    queue = __spreadArray(__spreadArray([], queue, true), [evt], false);
}
/** Log handler for cc service to send the performance logs to the server. */
function transportHandler(
// eslint-disable-next-line @typescript-eslint/no-explicit-any
serializer) ***REMOVED***
    return function () ***REMOVED***
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        var message = serializer.apply(void 0, args);
        addToQueue(***REMOVED***
            message: message,
            eventTime: Date.now()
      ***REMOVED***);
  ***REMOVED***;
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/* eslint-enble camelcase */
var logger;
// This method is not called before initialization.
function sendLog(resource, resourceType) ***REMOVED***
    if (!logger) ***REMOVED***
        logger = transportHandler(serializer);
  ***REMOVED***
    logger(resource, resourceType);
}
function logTrace(trace) ***REMOVED***
    var settingsService = SettingsService.getInstance();
    // Do not log if trace is auto generated and instrumentation is disabled.
    if (!settingsService.instrumentationEnabled && trace.isAuto) ***REMOVED***
        return;
  ***REMOVED***
    // Do not log if trace is custom and data collection is disabled.
    if (!settingsService.dataCollectionEnabled && !trace.isAuto) ***REMOVED***
        return;
  ***REMOVED***
    // Do not log if required apis are not available.
    if (!Api.getInstance().requiredApisAvailable()) ***REMOVED***
        return;
  ***REMOVED***
    // Only log the page load auto traces if page is visible.
    if (trace.isAuto && getVisibilityState() !== VisibilityState.VISIBLE) ***REMOVED***
        return;
  ***REMOVED***
    if (isPerfInitialized()) ***REMOVED***
        sendTraceLog(trace);
  ***REMOVED***
    else ***REMOVED***
        // Custom traces can be used before the initialization but logging
        // should wait until after.
        getInitializationPromise(trace.performanceController).then(function () ***REMOVED*** return sendTraceLog(trace); }, function () ***REMOVED*** return sendTraceLog(trace); });
  ***REMOVED***
}
function sendTraceLog(trace) ***REMOVED***
    if (!getIid()) ***REMOVED***
        return;
  ***REMOVED***
    var settingsService = SettingsService.getInstance();
    if (!settingsService.loggingEnabled ||
        !settingsService.logTraceAfterSampling) ***REMOVED***
        return;
  ***REMOVED***
    setTimeout(function () ***REMOVED*** return sendLog(trace, 1 /* ResourceType.Trace */); }, 0);
}
function logNetworkRequest(networkRequest) ***REMOVED***
    var settingsService = SettingsService.getInstance();
    // Do not log network requests if instrumentation is disabled.
    if (!settingsService.instrumentationEnabled) ***REMOVED***
        return;
  ***REMOVED***
    // Do not log the js sdk's call to transport service domain to avoid unnecessary cycle.
    // Need to blacklist both old and new endpoints to avoid migration gap.
    var networkRequestUrl = networkRequest.url;
    // Blacklist old log endpoint and new transport endpoint.
    // Because Performance SDK doesn't instrument requests sent from SDK itself.
    var logEndpointUrl = settingsService.logEndPointUrl.split('?')[0];
    var flEndpointUrl = settingsService.flTransportEndpointUrl.split('?')[0];
    if (networkRequestUrl === logEndpointUrl ||
        networkRequestUrl === flEndpointUrl) ***REMOVED***
        return;
  ***REMOVED***
    if (!settingsService.loggingEnabled ||
        !settingsService.logNetworkAfterSampling) ***REMOVED***
        return;
  ***REMOVED***
    setTimeout(function () ***REMOVED*** return sendLog(networkRequest, 0 /* ResourceType.NetworkRequest */); }, 0);
}
function serializer(resource, resourceType) ***REMOVED***
    if (resourceType === 0 /* ResourceType.NetworkRequest */) ***REMOVED***
        return serializeNetworkRequest(resource);
  ***REMOVED***
    return serializeTrace(resource);
}
function serializeNetworkRequest(networkRequest) ***REMOVED***
    var networkRequestMetric = ***REMOVED***
        url: networkRequest.url,
        http_method: networkRequest.httpMethod || 0,
        http_response_code: 200,
        response_payload_bytes: networkRequest.responsePayloadBytes,
        client_start_time_us: networkRequest.startTimeUs,
        time_to_response_initiated_us: networkRequest.timeToResponseInitiatedUs,
        time_to_response_completed_us: networkRequest.timeToResponseCompletedUs
  ***REMOVED***;
    var perfMetric = ***REMOVED***
        application_info: getApplicationInfo(networkRequest.performanceController.app),
        network_request_metric: networkRequestMetric
  ***REMOVED***;
    return JSON.stringify(perfMetric);
}
function serializeTrace(trace) ***REMOVED***
    var traceMetric = ***REMOVED***
        name: trace.name,
        is_auto: trace.isAuto,
        client_start_time_us: trace.startTimeUs,
        duration_us: trace.durationUs
  ***REMOVED***;
    if (Object.keys(trace.counters).length !== 0) ***REMOVED***
        traceMetric.counters = trace.counters;
  ***REMOVED***
    var customAttributes = trace.getAttributes();
    if (Object.keys(customAttributes).length !== 0) ***REMOVED***
        traceMetric.custom_attributes = customAttributes;
  ***REMOVED***
    var perfMetric = ***REMOVED***
        application_info: getApplicationInfo(trace.performanceController.app),
        trace_metric: traceMetric
  ***REMOVED***;
    return JSON.stringify(perfMetric);
}
function getApplicationInfo(firebaseApp) ***REMOVED***
    return ***REMOVED***
        google_app_id: getAppId(firebaseApp),
        app_instance_id: getIid(),
        web_app_info: ***REMOVED***
            sdk_version: SDK_VERSION,
            page_url: Api.getInstance().getUrl(),
            service_worker_status: getServiceWorkerStatus(),
            visibility_state: getVisibilityState(),
            effective_connection_type: getEffectiveConnectionType()
      ***REMOVED***,
        application_process_state: 0
  ***REMOVED***;
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var MAX_METRIC_NAME_LENGTH = 100;
var RESERVED_AUTO_PREFIX = '_';
var oobMetrics = [
    FIRST_PAINT_COUNTER_NAME,
    FIRST_CONTENTFUL_PAINT_COUNTER_NAME,
    FIRST_INPUT_DELAY_COUNTER_NAME
];
/**
 * Returns true if the metric is custom and does not start with reserved prefix, or if
 * the metric is one of out of the box page load trace metrics.
 */
function isValidMetricName(name, traceName) ***REMOVED***
    if (name.length === 0 || name.length > MAX_METRIC_NAME_LENGTH) ***REMOVED***
        return false;
  ***REMOVED***
    return ((traceName &&
        traceName.startsWith(OOB_TRACE_PAGE_LOAD_PREFIX) &&
        oobMetrics.indexOf(name) > -1) ||
        !name.startsWith(RESERVED_AUTO_PREFIX));
}
/**
 * Converts the provided value to an integer value to be used in case of a metric.
 * @param providedValue Provided number value of the metric that needs to be converted to an integer.
 *
 * @returns Converted integer number to be set for the metric.
 */
function convertMetricValueToInteger(providedValue) ***REMOVED***
    var valueAsInteger = Math.floor(providedValue);
    if (valueAsInteger < providedValue) ***REMOVED***
        consoleLogger.info("Metric value should be an Integer, setting the value as : ".concat(valueAsInteger, "."));
  ***REMOVED***
    return valueAsInteger;
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var Trace = /** @class */ (function () ***REMOVED***
    /**
     * @param performanceController The performance controller running.
     * @param name The name of the trace.
     * @param isAuto If the trace is auto-instrumented.
     * @param traceMeasureName The name of the measure marker in user timing specification. This field
     * is only set when the trace is built for logging when the user directly uses the user timing
     * api (performance.mark and performance.measure).
     */
    function Trace(performanceController, name, isAuto, traceMeasureName) ***REMOVED***
        if (isAuto === void 0) ***REMOVED*** isAuto = false; }
        this.performanceController = performanceController;
        this.name = name;
        this.isAuto = isAuto;
        this.state = 1 /* TraceState.UNINITIALIZED */;
        this.customAttributes = ***REMOVED***};
        this.counters = ***REMOVED***};
        this.api = Api.getInstance();
        this.randomId = Math.floor(Math.random() * 1000000);
        if (!this.isAuto) ***REMOVED***
            this.traceStartMark = "".concat(TRACE_START_MARK_PREFIX, "-").concat(this.randomId, "-").concat(this.name);
            this.traceStopMark = "".concat(TRACE_STOP_MARK_PREFIX, "-").concat(this.randomId, "-").concat(this.name);
            this.traceMeasure =
                traceMeasureName ||
                    "".concat(TRACE_MEASURE_PREFIX, "-").concat(this.randomId, "-").concat(this.name);
            if (traceMeasureName) ***REMOVED***
                // For the case of direct user timing traces, no start stop will happen. The measure object
                // is already available.
                this.calculateTraceMetrics();
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    /**
     * Starts a trace. The measurement of the duration starts at this point.
     */
    Trace.prototype.start = function () ***REMOVED***
        if (this.state !== 1 /* TraceState.UNINITIALIZED */) ***REMOVED***
            throw ERROR_FACTORY.create("trace started" /* ErrorCode.TRACE_STARTED_BEFORE */, ***REMOVED***
                traceName: this.name
          ***REMOVED***);
      ***REMOVED***
        this.api.mark(this.traceStartMark);
        this.state = 2 /* TraceState.RUNNING */;
  ***REMOVED***;
    /**
     * Stops the trace. The measurement of the duration of the trace stops at this point and trace
     * is logged.
     */
    Trace.prototype.stop = function () ***REMOVED***
        if (this.state !== 2 /* TraceState.RUNNING */) ***REMOVED***
            throw ERROR_FACTORY.create("trace stopped" /* ErrorCode.TRACE_STOPPED_BEFORE */, ***REMOVED***
                traceName: this.name
          ***REMOVED***);
      ***REMOVED***
        this.state = 3 /* TraceState.TERMINATED */;
        this.api.mark(this.traceStopMark);
        this.api.measure(this.traceMeasure, this.traceStartMark, this.traceStopMark);
        this.calculateTraceMetrics();
        logTrace(this);
  ***REMOVED***;
    /**
     * Records a trace with predetermined values. If this method is used a trace is created and logged
     * directly. No need to use start and stop methods.
     * @param startTime Trace start time since epoch in millisec
     * @param duration The duraction of the trace in millisec
     * @param options An object which can optionally hold maps of custom metrics and custom attributes
     */
    Trace.prototype.record = function (startTime, duration, options) ***REMOVED***
        if (startTime <= 0) ***REMOVED***
            throw ERROR_FACTORY.create("nonpositive trace startTime" /* ErrorCode.NONPOSITIVE_TRACE_START_TIME */, ***REMOVED***
                traceName: this.name
          ***REMOVED***);
      ***REMOVED***
        if (duration <= 0) ***REMOVED***
            throw ERROR_FACTORY.create("nonpositive trace duration" /* ErrorCode.NONPOSITIVE_TRACE_DURATION */, ***REMOVED***
                traceName: this.name
          ***REMOVED***);
      ***REMOVED***
        this.durationUs = Math.floor(duration * 1000);
        this.startTimeUs = Math.floor(startTime * 1000);
        if (options && options.attributes) ***REMOVED***
            this.customAttributes = __assign(***REMOVED***}, options.attributes);
      ***REMOVED***
        if (options && options.metrics) ***REMOVED***
            for (var _i = 0, _a = Object.keys(options.metrics); _i < _a.length; _i++) ***REMOVED***
                var metricName = _a[_i];
                if (!isNaN(Number(options.metrics[metricName]))) ***REMOVED***
                    this.counters[metricName] = Math.floor(Number(options.metrics[metricName]));
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        logTrace(this);
  ***REMOVED***;
    /**
     * Increments a custom metric by a certain number or 1 if number not specified. Will create a new
     * custom metric if one with the given name does not exist. The value will be floored down to an
     * integer.
     * @param counter Name of the custom metric
     * @param numAsInteger Increment by value
     */
    Trace.prototype.incrementMetric = function (counter, numAsInteger) ***REMOVED***
        if (numAsInteger === void 0) ***REMOVED*** numAsInteger = 1; }
        if (this.counters[counter] === undefined) ***REMOVED***
            this.putMetric(counter, numAsInteger);
      ***REMOVED***
        else ***REMOVED***
            this.putMetric(counter, this.counters[counter] + numAsInteger);
      ***REMOVED***
  ***REMOVED***;
    /**
     * Sets a custom metric to a specified value. Will create a new custom metric if one with the
     * given name does not exist. The value will be floored down to an integer.
     * @param counter Name of the custom metric
     * @param numAsInteger Set custom metric to this value
     */
    Trace.prototype.putMetric = function (counter, numAsInteger) ***REMOVED***
        if (isValidMetricName(counter, this.name)) ***REMOVED***
            this.counters[counter] = convertMetricValueToInteger(numAsInteger !== null && numAsInteger !== void 0 ? numAsInteger : 0);
      ***REMOVED***
        else ***REMOVED***
            throw ERROR_FACTORY.create("invalid custom metric name" /* ErrorCode.INVALID_CUSTOM_METRIC_NAME */, ***REMOVED***
                customMetricName: counter
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***;
    /**
     * Returns the value of the custom metric by that name. If a custom metric with that name does
     * not exist will return zero.
     * @param counter
     */
    Trace.prototype.getMetric = function (counter) ***REMOVED***
        return this.counters[counter] || 0;
  ***REMOVED***;
    /**
     * Sets a custom attribute of a trace to a certain value.
     * @param attr
     * @param value
     */
    Trace.prototype.putAttribute = function (attr, value) ***REMOVED***
        var isValidName = isValidCustomAttributeName(attr);
        var isValidValue = isValidCustomAttributeValue(value);
        if (isValidName && isValidValue) ***REMOVED***
            this.customAttributes[attr] = value;
            return;
      ***REMOVED***
        // Throw appropriate error when the attribute name or value is invalid.
        if (!isValidName) ***REMOVED***
            throw ERROR_FACTORY.create("invalid attribute name" /* ErrorCode.INVALID_ATTRIBUTE_NAME */, ***REMOVED***
                attributeName: attr
          ***REMOVED***);
      ***REMOVED***
        if (!isValidValue) ***REMOVED***
            throw ERROR_FACTORY.create("invalid attribute value" /* ErrorCode.INVALID_ATTRIBUTE_VALUE */, ***REMOVED***
                attributeValue: value
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***;
    /**
     * Retrieves the value a custom attribute of a trace is set to.
     * @param attr
     */
    Trace.prototype.getAttribute = function (attr) ***REMOVED***
        return this.customAttributes[attr];
  ***REMOVED***;
    Trace.prototype.removeAttribute = function (attr) ***REMOVED***
        if (this.customAttributes[attr] === undefined) ***REMOVED***
            return;
      ***REMOVED***
        delete this.customAttributes[attr];
  ***REMOVED***;
    Trace.prototype.getAttributes = function () ***REMOVED***
        return __assign(***REMOVED***}, this.customAttributes);
  ***REMOVED***;
    Trace.prototype.setStartTime = function (startTime) ***REMOVED***
        this.startTimeUs = startTime;
  ***REMOVED***;
    Trace.prototype.setDuration = function (duration) ***REMOVED***
        this.durationUs = duration;
  ***REMOVED***;
    /**
     * Calculates and assigns the duration and start time of the trace using the measure performance
     * entry.
     */
    Trace.prototype.calculateTraceMetrics = function () ***REMOVED***
        var perfMeasureEntries = this.api.getEntriesByName(this.traceMeasure);
        var perfMeasureEntry = perfMeasureEntries && perfMeasureEntries[0];
        if (perfMeasureEntry) ***REMOVED***
            this.durationUs = Math.floor(perfMeasureEntry.duration * 1000);
            this.startTimeUs = Math.floor((perfMeasureEntry.startTime + this.api.getTimeOrigin()) * 1000);
      ***REMOVED***
  ***REMOVED***;
    /**
     * @param navigationTimings A single element array which contains the navigationTIming object of
     * the page load
     * @param paintTimings A array which contains paintTiming object of the page load
     * @param firstInputDelay First input delay in millisec
     */
    Trace.createOobTrace = function (performanceController, navigationTimings, paintTimings, firstInputDelay) ***REMOVED***
        var route = Api.getInstance().getUrl();
        if (!route) ***REMOVED***
            return;
      ***REMOVED***
        var trace = new Trace(performanceController, OOB_TRACE_PAGE_LOAD_PREFIX + route, true);
        var timeOriginUs = Math.floor(Api.getInstance().getTimeOrigin() * 1000);
        trace.setStartTime(timeOriginUs);
        // navigationTimings includes only one element.
        if (navigationTimings && navigationTimings[0]) ***REMOVED***
            trace.setDuration(Math.floor(navigationTimings[0].duration * 1000));
            trace.putMetric('domInteractive', Math.floor(navigationTimings[0].domInteractive * 1000));
            trace.putMetric('domContentLoadedEventEnd', Math.floor(navigationTimings[0].domContentLoadedEventEnd * 1000));
            trace.putMetric('loadEventEnd', Math.floor(navigationTimings[0].loadEventEnd * 1000));
      ***REMOVED***
        var FIRST_PAINT = 'first-paint';
        var FIRST_CONTENTFUL_PAINT = 'first-contentful-paint';
        if (paintTimings) ***REMOVED***
            var firstPaint = paintTimings.find(function (paintObject) ***REMOVED*** return paintObject.name === FIRST_PAINT; });
            if (firstPaint && firstPaint.startTime) ***REMOVED***
                trace.putMetric(FIRST_PAINT_COUNTER_NAME, Math.floor(firstPaint.startTime * 1000));
          ***REMOVED***
            var firstContentfulPaint = paintTimings.find(function (paintObject) ***REMOVED*** return paintObject.name === FIRST_CONTENTFUL_PAINT; });
            if (firstContentfulPaint && firstContentfulPaint.startTime) ***REMOVED***
                trace.putMetric(FIRST_CONTENTFUL_PAINT_COUNTER_NAME, Math.floor(firstContentfulPaint.startTime * 1000));
          ***REMOVED***
            if (firstInputDelay) ***REMOVED***
                trace.putMetric(FIRST_INPUT_DELAY_COUNTER_NAME, Math.floor(firstInputDelay * 1000));
          ***REMOVED***
      ***REMOVED***
        logTrace(trace);
  ***REMOVED***;
    Trace.createUserTimingTrace = function (performanceController, measureName) ***REMOVED***
        var trace = new Trace(performanceController, measureName, false, measureName);
        logTrace(trace);
  ***REMOVED***;
    return Trace;
}());

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function createNetworkRequestEntry(performanceController, entry) ***REMOVED***
    var performanceEntry = entry;
    if (!performanceEntry || performanceEntry.responseStart === undefined) ***REMOVED***
        return;
  ***REMOVED***
    var timeOrigin = Api.getInstance().getTimeOrigin();
    var startTimeUs = Math.floor((performanceEntry.startTime + timeOrigin) * 1000);
    var timeToResponseInitiatedUs = performanceEntry.responseStart
        ? Math.floor((performanceEntry.responseStart - performanceEntry.startTime) * 1000)
        : undefined;
    var timeToResponseCompletedUs = Math.floor((performanceEntry.responseEnd - performanceEntry.startTime) * 1000);
    // Remove the query params from logged network request url.
    var url = performanceEntry.name && performanceEntry.name.split('?')[0];
    var networkRequest = ***REMOVED***
        performanceController: performanceController,
        url: url,
        responsePayloadBytes: performanceEntry.transferSize,
        startTimeUs: startTimeUs,
        timeToResponseInitiatedUs: timeToResponseInitiatedUs,
        timeToResponseCompletedUs: timeToResponseCompletedUs
  ***REMOVED***;
    logNetworkRequest(networkRequest);
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var FID_WAIT_TIME_MS = 5000;
function setupOobResources(performanceController) ***REMOVED***
    // Do not initialize unless iid is available.
    if (!getIid()) ***REMOVED***
        return;
  ***REMOVED***
    // The load event might not have fired yet, and that means performance navigation timing
    // object has a duration of 0. The setup should run after all current tasks in js queue.
    setTimeout(function () ***REMOVED*** return setupOobTraces(performanceController); }, 0);
    setTimeout(function () ***REMOVED*** return setupNetworkRequests(performanceController); }, 0);
    setTimeout(function () ***REMOVED*** return setupUserTimingTraces(performanceController); }, 0);
}
function setupNetworkRequests(performanceController) ***REMOVED***
    var api = Api.getInstance();
    var resources = api.getEntriesByType('resource');
    for (var _i = 0, resources_1 = resources; _i < resources_1.length; _i++) ***REMOVED***
        var resource = resources_1[_i];
        createNetworkRequestEntry(performanceController, resource);
  ***REMOVED***
    api.setupObserver('resource', function (entry) ***REMOVED***
        return createNetworkRequestEntry(performanceController, entry);
  ***REMOVED***);
}
function setupOobTraces(performanceController) ***REMOVED***
    var api = Api.getInstance();
    var navigationTimings = api.getEntriesByType('navigation');
    var paintTimings = api.getEntriesByType('paint');
    // If First Input Desly polyfill is added to the page, report the fid value.
    // https://github.com/GoogleChromeLabs/first-input-delay
    if (api.onFirstInputDelay) ***REMOVED***
        // If the fid call back is not called for certain time, continue without it.
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        var timeoutId_1 = setTimeout(function () ***REMOVED***
            Trace.createOobTrace(performanceController, navigationTimings, paintTimings);
            timeoutId_1 = undefined;
      ***REMOVED***, FID_WAIT_TIME_MS);
        api.onFirstInputDelay(function (fid) ***REMOVED***
            if (timeoutId_1) ***REMOVED***
                clearTimeout(timeoutId_1);
                Trace.createOobTrace(performanceController, navigationTimings, paintTimings, fid);
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***
    else ***REMOVED***
        Trace.createOobTrace(performanceController, navigationTimings, paintTimings);
  ***REMOVED***
}
function setupUserTimingTraces(performanceController) ***REMOVED***
    var api = Api.getInstance();
    // Run through the measure performance entries collected up to this point.
    var measures = api.getEntriesByType('measure');
    for (var _i = 0, measures_1 = measures; _i < measures_1.length; _i++) ***REMOVED***
        var measure = measures_1[_i];
        createUserTimingTrace(performanceController, measure);
  ***REMOVED***
    // Setup an observer to capture the measures from this point on.
    api.setupObserver('measure', function (entry) ***REMOVED***
        return createUserTimingTrace(performanceController, entry);
  ***REMOVED***);
}
function createUserTimingTrace(performanceController, measure) ***REMOVED***
    var measureName = measure.name;
    // Do not create a trace, if the user timing marks and measures are created by the sdk itself.
    if (measureName.substring(0, TRACE_MEASURE_PREFIX.length) ===
        TRACE_MEASURE_PREFIX) ***REMOVED***
        return;
  ***REMOVED***
    Trace.createUserTimingTrace(performanceController, measureName);
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var PerformanceController = /** @class */ (function () ***REMOVED***
    function PerformanceController(app, installations) ***REMOVED***
        this.app = app;
        this.installations = installations;
        this.initialized = false;
  ***REMOVED***
    /**
     * This method *must* be called internally as part of creating a
     * PerformanceController instance.
     *
     * Currently it's not possible to pass the settings object through the
     * constructor using Components, so this method exists to be called with the
     * desired settings, to ensure nothing is collected without the user's
     * consent.
     */
    PerformanceController.prototype._init = function (settings) ***REMOVED***
        var _this = this;
        if (this.initialized) ***REMOVED***
            return;
      ***REMOVED***
        if ((settings === null || settings === void 0 ? void 0 : settings.dataCollectionEnabled) !== undefined) ***REMOVED***
            this.dataCollectionEnabled = settings.dataCollectionEnabled;
      ***REMOVED***
        if ((settings === null || settings === void 0 ? void 0 : settings.instrumentationEnabled) !== undefined) ***REMOVED***
            this.instrumentationEnabled = settings.instrumentationEnabled;
      ***REMOVED***
        if (Api.getInstance().requiredApisAvailable()) ***REMOVED***
            validateIndexedDBOpenable()
                .then(function (isAvailable) ***REMOVED***
                if (isAvailable) ***REMOVED***
                    setupTransportService();
                    getInitializationPromise(_this).then(function () ***REMOVED*** return setupOobResources(_this); }, function () ***REMOVED*** return setupOobResources(_this); });
                    _this.initialized = true;
              ***REMOVED***
          ***REMOVED***)
                .catch(function (error) ***REMOVED***
                consoleLogger.info("Environment doesn't support IndexedDB: ".concat(error));
          ***REMOVED***);
      ***REMOVED***
        else ***REMOVED***
            consoleLogger.info('Firebase Performance cannot start if the browser does not support ' +
                '"Fetch" and "Promise", or cookies are disabled.');
      ***REMOVED***
  ***REMOVED***;
    Object.defineProperty(PerformanceController.prototype, "instrumentationEnabled", ***REMOVED***
        get: function () ***REMOVED***
            return SettingsService.getInstance().instrumentationEnabled;
      ***REMOVED***,
        set: function (val) ***REMOVED***
            SettingsService.getInstance().instrumentationEnabled = val;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(PerformanceController.prototype, "dataCollectionEnabled", ***REMOVED***
        get: function () ***REMOVED***
            return SettingsService.getInstance().dataCollectionEnabled;
      ***REMOVED***,
        set: function (val) ***REMOVED***
            SettingsService.getInstance().dataCollectionEnabled = val;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    return PerformanceController;
}());

/**
 * Firebase Performance Monitoring
 *
 * @packageDocumentation
 */
var DEFAULT_ENTRY_NAME = '[DEFAULT]';
/**
 * Returns a ***REMOVED***@link FirebasePerformance} instance for the given app.
 * @param app - The ***REMOVED***@link @firebase/app#FirebaseApp} to use.
 * @public
 */
function getPerformance(app) ***REMOVED***
    if (app === void 0) ***REMOVED*** app = getApp(); }
    app = getModularInstance(app);
    var provider = _getProvider(app, 'performance');
    var perfInstance = provider.getImmediate();
    return perfInstance;
}
/**
 * Returns a ***REMOVED***@link FirebasePerformance} instance for the given app. Can only be called once.
 * @param app - The ***REMOVED***@link @firebase/app#FirebaseApp} to use.
 * @param settings - Optional settings for the ***REMOVED***@link FirebasePerformance} instance.
 * @public
 */
function initializePerformance(app, settings) ***REMOVED***
    app = getModularInstance(app);
    var provider = _getProvider(app, 'performance');
    // throw if an instance was already created.
    // It could happen if initializePerformance() is called more than once, or getPerformance() is called first.
    if (provider.isInitialized()) ***REMOVED***
        var existingInstance = provider.getImmediate();
        var initialSettings = provider.getOptions();
        if (deepEqual(initialSettings, settings !== null && settings !== void 0 ? settings : ***REMOVED***})) ***REMOVED***
            return existingInstance;
      ***REMOVED***
        else ***REMOVED***
            throw ERROR_FACTORY.create("already initialized" /* ErrorCode.ALREADY_INITIALIZED */);
      ***REMOVED***
  ***REMOVED***
    var perfInstance = provider.initialize(***REMOVED***
        options: settings
  ***REMOVED***);
    return perfInstance;
}
/**
 * Returns a new `PerformanceTrace` instance.
 * @param performance - The ***REMOVED***@link FirebasePerformance} instance to use.
 * @param name - The name of the trace.
 * @public
 */
function trace(performance, name) ***REMOVED***
    performance = getModularInstance(performance);
    return new Trace(performance, name);
}
var factory = function (container, _a) ***REMOVED***
    var settings = _a.options;
    // Dependencies
    var app = container.getProvider('app').getImmediate();
    var installations = container
        .getProvider('installations-internal')
        .getImmediate();
    if (app.name !== DEFAULT_ENTRY_NAME) ***REMOVED***
        throw ERROR_FACTORY.create("FB not default" /* ErrorCode.FB_NOT_DEFAULT */);
  ***REMOVED***
    if (typeof window === 'undefined') ***REMOVED***
        throw ERROR_FACTORY.create("no window" /* ErrorCode.NO_WINDOW */);
  ***REMOVED***
    setupApi(window);
    var perfInstance = new PerformanceController(app, installations);
    perfInstance._init(settings);
    return perfInstance;
};
function registerPerformance() ***REMOVED***
    _registerComponent(new Component('performance', factory, "PUBLIC" /* ComponentType.PUBLIC */));
    registerVersion(name, version);
    // BUILD_TARGET will be replaced by values like esm5, esm2017, cjs5, etc during the compilation
    registerVersion(name, version, 'esm5');
}
registerPerformance();

export ***REMOVED*** getPerformance, initializePerformance, trace };
