// split2/index.js
/*
Copyright (c) 2014-2021, Matteo Collina <hello@matteocollina.com>

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
*/

'use strict'

const ***REMOVED*** Transform } = require('stream')
const ***REMOVED*** StringDecoder } = require('string_decoder')
const kLast = Symbol('last')
const kDecoder = Symbol('decoder')

function transform (chunk, enc, cb) ***REMOVED***
  let list
  if (this.overflow) ***REMOVED*** // Line buffer is full. Skip to start of next line.
    const buf = this[kDecoder].write(chunk)
    list = buf.split(this.matcher)

    if (list.length === 1) return cb() // Line ending not found. Discard entire chunk.

    // Line ending found. Discard trailing fragment of previous line and reset overflow state.
    list.shift()
    this.overflow = false
***REMOVED*** else ***REMOVED***
    this[kLast] += this[kDecoder].write(chunk)
    list = this[kLast].split(this.matcher)
***REMOVED***

  this[kLast] = list.pop()

  for (let i = 0; i < list.length; i++) ***REMOVED***
    try ***REMOVED***
      push(this, this.mapper(list[i]))
  ***REMOVED*** catch (error) ***REMOVED***
      return cb(error)
  ***REMOVED***
***REMOVED***

  this.overflow = this[kLast].length > this.maxLength
  if (this.overflow && !this.skipOverflow) ***REMOVED***
    cb(new Error('maximum buffer reached'))
    return
***REMOVED***

  cb()
}

function flush (cb) ***REMOVED***
  // forward any gibberish left in there
  this[kLast] += this[kDecoder].end()

  if (this[kLast]) ***REMOVED***
    try ***REMOVED***
      push(this, this.mapper(this[kLast]))
  ***REMOVED*** catch (error) ***REMOVED***
      return cb(error)
  ***REMOVED***
***REMOVED***

  cb()
}

function push (self, val) ***REMOVED***
  if (val !== undefined) ***REMOVED***
    self.push(val)
***REMOVED***
}

function noop (incoming) ***REMOVED***
  return incoming
}

function split (matcher, mapper, options) ***REMOVED***
  // Set defaults for any arguments not supplied.
  matcher = matcher || /\r?\n/
  mapper = mapper || noop
  options = options || ***REMOVED***}

  // Test arguments explicitly.
  switch (arguments.length) ***REMOVED***
    case 1:
      // If mapper is only argument.
      if (typeof matcher === 'function') ***REMOVED***
        mapper = matcher
        matcher = /\r?\n/
      // If options is only argument.
    ***REMOVED*** else if (typeof matcher === 'object' && !(matcher instanceof RegExp)) ***REMOVED***
        options = matcher
        matcher = /\r?\n/
    ***REMOVED***
      break

    case 2:
      // If mapper and options are arguments.
      if (typeof matcher === 'function') ***REMOVED***
        options = mapper
        mapper = matcher
        matcher = /\r?\n/
      // If matcher and options are arguments.
    ***REMOVED*** else if (typeof mapper === 'object') ***REMOVED***
        options = mapper
        mapper = noop
    ***REMOVED***
***REMOVED***

  options = Object.assign(***REMOVED***}, options)
  options.autoDestroy = true
  options.transform = transform
  options.flush = flush
  options.readableObjectMode = true

  const stream = new Transform(options)

  stream[kLast] = ''
  stream[kDecoder] = new StringDecoder('utf8')
  stream.matcher = matcher
  stream.mapper = mapper
  stream.maxLength = options.maxLength
  stream.skipOverflow = options.skipOverflow || false
  stream.overflow = false
  stream._destroy = function (err, cb) ***REMOVED***
    // Weird Node v12 bug that we need to work around
    this._writableState.errorEmitted = false
    cb(err)
***REMOVED***

  return stream
}

module.exports = split
