// @firebase/logger/dist/esm/index.esm2017.js
/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * A container for all of the Logger instances
 */
const instances = [];
/**
 * The JS SDK supports 5 log levels and also allows a user the ability to
 * silence the logs altogether.
 *
 * The order is a follows:
 * DEBUG < VERBOSE < INFO < WARN < ERROR
 *
 * All of the log types above the current log level will be captured (i.e. if
 * you set the log level to `INFO`, errors will still be logged, but `DEBUG` and
 * `VERBOSE` logs will not)
 */
var LogLevel;
(function (LogLevel) ***REMOVED***
    LogLevel[LogLevel["DEBUG"] = 0] = "DEBUG";
    LogLevel[LogLevel["VERBOSE"] = 1] = "VERBOSE";
    LogLevel[LogLevel["INFO"] = 2] = "INFO";
    LogLevel[LogLevel["WARN"] = 3] = "WARN";
    LogLevel[LogLevel["ERROR"] = 4] = "ERROR";
    LogLevel[LogLevel["SILENT"] = 5] = "SILENT";
})(LogLevel || (LogLevel = ***REMOVED***}));
const levelStringToEnum = ***REMOVED***
    'debug': LogLevel.DEBUG,
    'verbose': LogLevel.VERBOSE,
    'info': LogLevel.INFO,
    'warn': LogLevel.WARN,
    'error': LogLevel.ERROR,
    'silent': LogLevel.SILENT
};
/**
 * The default log level
 */
const defaultLogLevel = LogLevel.INFO;
/**
 * By default, `console.debug` is not displayed in the developer console (in
 * chrome). To avoid forcing users to have to opt-in to these logs twice
 * (i.e. once for firebase, and once in the console), we are sending `DEBUG`
 * logs to the `console.log` function.
 */
const ConsoleMethod = ***REMOVED***
    [LogLevel.DEBUG]: 'log',
    [LogLevel.VERBOSE]: 'log',
    [LogLevel.INFO]: 'info',
    [LogLevel.WARN]: 'warn',
    [LogLevel.ERROR]: 'error'
};
/**
 * The default log handler will forward DEBUG, VERBOSE, INFO, WARN, and ERROR
 * messages on to their corresponding console counterparts (if the log method
 * is supported by the current log level)
 */
const defaultLogHandler = (instance, logType, ...args) => ***REMOVED***
    if (logType < instance.logLevel) ***REMOVED***
        return;
  ***REMOVED***
    const now = new Date().toISOString();
    const method = ConsoleMethod[logType];
    if (method) ***REMOVED***
        console[method](`[$***REMOVED***now}]  $***REMOVED***instance.name}:`, ...args);
  ***REMOVED***
    else ***REMOVED***
        throw new Error(`Attempted to log a message with an invalid logType (value: $***REMOVED***logType})`);
  ***REMOVED***
};
class Logger ***REMOVED***
    /**
     * Gives you an instance of a Logger to capture messages according to
     * Firebase's logging scheme.
     *
     * @param name The name that the logs will be associated with
     */
    constructor(name) ***REMOVED***
        this.name = name;
        /**
         * The log level of the given Logger instance.
         */
        this._logLevel = defaultLogLevel;
        /**
         * The main (internal) log handler for the Logger instance.
         * Can be set to a new function in internal package code but not by user.
         */
        this._logHandler = defaultLogHandler;
        /**
         * The optional, additional, user-defined log handler for the Logger instance.
         */
        this._userLogHandler = null;
        /**
         * Capture the current instance for later use
         */
        instances.push(this);
  ***REMOVED***
    get logLevel() ***REMOVED***
        return this._logLevel;
  ***REMOVED***
    set logLevel(val) ***REMOVED***
        if (!(val in LogLevel)) ***REMOVED***
            throw new TypeError(`Invalid value "$***REMOVED***val}" assigned to \`logLevel\``);
      ***REMOVED***
        this._logLevel = val;
  ***REMOVED***
    // Workaround for setter/getter having to be the same type.
    setLogLevel(val) ***REMOVED***
        this._logLevel = typeof val === 'string' ? levelStringToEnum[val] : val;
  ***REMOVED***
    get logHandler() ***REMOVED***
        return this._logHandler;
  ***REMOVED***
    set logHandler(val) ***REMOVED***
        if (typeof val !== 'function') ***REMOVED***
            throw new TypeError('Value assigned to `logHandler` must be a function');
      ***REMOVED***
        this._logHandler = val;
  ***REMOVED***
    get userLogHandler() ***REMOVED***
        return this._userLogHandler;
  ***REMOVED***
    set userLogHandler(val) ***REMOVED***
        this._userLogHandler = val;
  ***REMOVED***
    /**
     * The functions below are all based on the `console` interface
     */
    debug(...args) ***REMOVED***
        this._userLogHandler && this._userLogHandler(this, LogLevel.DEBUG, ...args);
        this._logHandler(this, LogLevel.DEBUG, ...args);
  ***REMOVED***
    log(...args) ***REMOVED***
        this._userLogHandler &&
            this._userLogHandler(this, LogLevel.VERBOSE, ...args);
        this._logHandler(this, LogLevel.VERBOSE, ...args);
  ***REMOVED***
    info(...args) ***REMOVED***
        this._userLogHandler && this._userLogHandler(this, LogLevel.INFO, ...args);
        this._logHandler(this, LogLevel.INFO, ...args);
  ***REMOVED***
    warn(...args) ***REMOVED***
        this._userLogHandler && this._userLogHandler(this, LogLevel.WARN, ...args);
        this._logHandler(this, LogLevel.WARN, ...args);
  ***REMOVED***
    error(...args) ***REMOVED***
        this._userLogHandler && this._userLogHandler(this, LogLevel.ERROR, ...args);
        this._logHandler(this, LogLevel.ERROR, ...args);
  ***REMOVED***
}
function setLogLevel(level) ***REMOVED***
    instances.forEach(inst => ***REMOVED***
        inst.setLogLevel(level);
  ***REMOVED***);
}
function setUserLogHandler(logCallback, options) ***REMOVED***
    for (const instance of instances) ***REMOVED***
        let customLogLevel = null;
        if (options && options.level) ***REMOVED***
            customLogLevel = levelStringToEnum[options.level];
      ***REMOVED***
        if (logCallback === null) ***REMOVED***
            instance.userLogHandler = null;
      ***REMOVED***
        else ***REMOVED***
            instance.userLogHandler = (instance, level, ...args) => ***REMOVED***
                const message = args
                    .map(arg => ***REMOVED***
                    if (arg == null) ***REMOVED***
                        return null;
                  ***REMOVED***
                    else if (typeof arg === 'string') ***REMOVED***
                        return arg;
                  ***REMOVED***
                    else if (typeof arg === 'number' || typeof arg === 'boolean') ***REMOVED***
                        return arg.toString();
                  ***REMOVED***
                    else if (arg instanceof Error) ***REMOVED***
                        return arg.message;
                  ***REMOVED***
                    else ***REMOVED***
                        try ***REMOVED***
                            return JSON.stringify(arg);
                      ***REMOVED***
                        catch (ignored) ***REMOVED***
                            return null;
                      ***REMOVED***
                  ***REMOVED***
              ***REMOVED***)
                    .filter(arg => arg)
                    .join(' ');
                if (level >= (customLogLevel !== null && customLogLevel !== void 0 ? customLogLevel : instance.logLevel)) ***REMOVED***
                    logCallback(***REMOVED***
                        level: LogLevel[level].toLowerCase(),
                        message,
                        args,
                        type: instance.name
                  ***REMOVED***);
              ***REMOVED***
          ***REMOVED***;
      ***REMOVED***
  ***REMOVED***
}

export ***REMOVED*** LogLevel, Logger, setLogLevel, setUserLogHandler };
