// @fastify/busboy/deps/dicer/lib/HeaderParser.js
'use strict'

const EventEmitter = require('events').EventEmitter
const inherits = require('util').inherits
const getLimit = require('../../../lib/utils/getLimit')

const StreamSearch = require('../../streamsearch/sbmh')

const B_DCRLF = Buffer.from('\r\n\r\n')
const RE_CRLF = /\r\n/g
const RE_HDR = /^([^:]+):[ \t]?([\x00-\xFF]+)?$/ // eslint-disable-line no-control-regex

function HeaderParser (cfg) ***REMOVED***
  EventEmitter.call(this)

  cfg = cfg || ***REMOVED***}
  const self = this
  this.nread = 0
  this.maxed = false
  this.npairs = 0
  this.maxHeaderPairs = getLimit(cfg, 'maxHeaderPairs', 2000)
  this.maxHeaderSize = getLimit(cfg, 'maxHeaderSize', 80 * 1024)
  this.buffer = ''
  this.header = ***REMOVED***}
  this.finished = false
  this.ss = new StreamSearch(B_DCRLF)
  this.ss.on('info', function (isMatch, data, start, end) ***REMOVED***
    if (data && !self.maxed) ***REMOVED***
      if (self.nread + end - start >= self.maxHeaderSize) ***REMOVED***
        end = self.maxHeaderSize - self.nread + start
        self.nread = self.maxHeaderSize
        self.maxed = true
    ***REMOVED*** else ***REMOVED*** self.nread += (end - start) }

      self.buffer += data.toString('binary', start, end)
  ***REMOVED***
    if (isMatch) ***REMOVED*** self._finish() }
***REMOVED***)
}
inherits(HeaderParser, EventEmitter)

HeaderParser.prototype.push = function (data) ***REMOVED***
  const r = this.ss.push(data)
  if (this.finished) ***REMOVED*** return r }
}

HeaderParser.prototype.reset = function () ***REMOVED***
  this.finished = false
  this.buffer = ''
  this.header = ***REMOVED***}
  this.ss.reset()
}

HeaderParser.prototype._finish = function () ***REMOVED***
  if (this.buffer) ***REMOVED*** this._parseHeader() }
  this.ss.matches = this.ss.maxMatches
  const header = this.header
  this.header = ***REMOVED***}
  this.buffer = ''
  this.finished = true
  this.nread = this.npairs = 0
  this.maxed = false
  this.emit('header', header)
}

HeaderParser.prototype._parseHeader = function () ***REMOVED***
  if (this.npairs === this.maxHeaderPairs) ***REMOVED*** return }

  const lines = this.buffer.split(RE_CRLF)
  const len = lines.length
  let m, h

  for (var i = 0; i < len; ++i) ***REMOVED*** // eslint-disable-line no-var
    if (lines[i].length === 0) ***REMOVED*** continue }
    if (lines[i][0] === '\t' || lines[i][0] === ' ') ***REMOVED***
      // folded header content
      // RFC2822 says to just remove the CRLF and not the whitespace following
      // it, so we follow the RFC and include the leading whitespace ...
      if (h) ***REMOVED***
        this.header[h][this.header[h].length - 1] += lines[i]
        continue
    ***REMOVED***
  ***REMOVED***

    const posColon = lines[i].indexOf(':')
    if (
      posColon === -1 ||
      posColon === 0
    ) ***REMOVED***
      return
  ***REMOVED***
    m = RE_HDR.exec(lines[i])
    h = m[1].toLowerCase()
    this.header[h] = this.header[h] || []
    this.header[h].push((m[2] || ''))
    if (++this.npairs === this.maxHeaderPairs) ***REMOVED*** break }
***REMOVED***
}

module.exports = HeaderParser
