// jose/dist/node/cjs/util/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.JWSSignatureVerificationFailed = exports.JWKSTimeout = exports.JWKSMultipleMatchingKeys = exports.JWKSNoMatchingKey = exports.JWKSInvalid = exports.JWKInvalid = exports.JWTInvalid = exports.JWSInvalid = exports.JWEInvalid = exports.JWEDecryptionFailed = exports.JOSENotSupported = exports.JOSEAlgNotAllowed = exports.JWTExpired = exports.JWTClaimValidationFailed = exports.JOSEError = void 0;
class JOSEError extends Error ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JOSE_GENERIC';
  ***REMOVED***
    constructor(message) ***REMOVED***
        var _a;
        super(message);
        this.code = 'ERR_JOSE_GENERIC';
        this.name = this.constructor.name;
        (_a = Error.captureStackTrace) === null || _a === void 0 ? void 0 : _a.call(Error, this, this.constructor);
  ***REMOVED***
}
exports.JOSEError = JOSEError;
class JWTClaimValidationFailed extends JOSEError ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWT_CLAIM_VALIDATION_FAILED';
  ***REMOVED***
    constructor(message, claim = 'unspecified', reason = 'unspecified') ***REMOVED***
        super(message);
        this.code = 'ERR_JWT_CLAIM_VALIDATION_FAILED';
        this.claim = claim;
        this.reason = reason;
  ***REMOVED***
}
exports.JWTClaimValidationFailed = JWTClaimValidationFailed;
class JWTExpired extends JOSEError ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWT_EXPIRED';
  ***REMOVED***
    constructor(message, claim = 'unspecified', reason = 'unspecified') ***REMOVED***
        super(message);
        this.code = 'ERR_JWT_EXPIRED';
        this.claim = claim;
        this.reason = reason;
  ***REMOVED***
}
exports.JWTExpired = JWTExpired;
class JOSEAlgNotAllowed extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JOSE_ALG_NOT_ALLOWED';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JOSE_ALG_NOT_ALLOWED';
  ***REMOVED***
}
exports.JOSEAlgNotAllowed = JOSEAlgNotAllowed;
class JOSENotSupported extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JOSE_NOT_SUPPORTED';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JOSE_NOT_SUPPORTED';
  ***REMOVED***
}
exports.JOSENotSupported = JOSENotSupported;
class JWEDecryptionFailed extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWE_DECRYPTION_FAILED';
        this.message = 'decryption operation failed';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWE_DECRYPTION_FAILED';
  ***REMOVED***
}
exports.JWEDecryptionFailed = JWEDecryptionFailed;
class JWEInvalid extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWE_INVALID';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWE_INVALID';
  ***REMOVED***
}
exports.JWEInvalid = JWEInvalid;
class JWSInvalid extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWS_INVALID';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWS_INVALID';
  ***REMOVED***
}
exports.JWSInvalid = JWSInvalid;
class JWTInvalid extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWT_INVALID';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWT_INVALID';
  ***REMOVED***
}
exports.JWTInvalid = JWTInvalid;
class JWKInvalid extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWK_INVALID';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWK_INVALID';
  ***REMOVED***
}
exports.JWKInvalid = JWKInvalid;
class JWKSInvalid extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWKS_INVALID';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWKS_INVALID';
  ***REMOVED***
}
exports.JWKSInvalid = JWKSInvalid;
class JWKSNoMatchingKey extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWKS_NO_MATCHING_KEY';
        this.message = 'no applicable key found in the JSON Web Key Set';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWKS_NO_MATCHING_KEY';
  ***REMOVED***
}
exports.JWKSNoMatchingKey = JWKSNoMatchingKey;
class JWKSMultipleMatchingKeys extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWKS_MULTIPLE_MATCHING_KEYS';
        this.message = 'multiple matching keys found in the JSON Web Key Set';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWKS_MULTIPLE_MATCHING_KEYS';
  ***REMOVED***
}
exports.JWKSMultipleMatchingKeys = JWKSMultipleMatchingKeys;
class JWKSTimeout extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWKS_TIMEOUT';
        this.message = 'request timed out';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWKS_TIMEOUT';
  ***REMOVED***
}
exports.JWKSTimeout = JWKSTimeout;
class JWSSignatureVerificationFailed extends JOSEError ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.code = 'ERR_JWS_SIGNATURE_VERIFICATION_FAILED';
        this.message = 'signature verification failed';
  ***REMOVED***
    static get code() ***REMOVED***
        return 'ERR_JWS_SIGNATURE_VERIFICATION_FAILED';
  ***REMOVED***
}
exports.JWSSignatureVerificationFailed = JWSSignatureVerificationFailed;
