// @firebase/database-compat/dist/index.esm5.js
import firebase from '@firebase/app-compat';
import ***REMOVED*** Provider, ComponentContainer, Component } from '@firebase/component';
import ***REMOVED*** _validatePathString, onChildMoved, onChildChanged, onChildRemoved, onChildAdded, onValue, off, get, query, limitToFirst, limitToLast, orderByChild, orderByKey, orderByPriority, orderByValue, startAt, startAfter, endAt, endBefore, equalTo, _ReferenceImpl, _QueryImpl, _QueryParams, child, set, _validateWritablePath, update, setWithPriority, remove, runTransaction, setPriority, push, OnDisconnect as OnDisconnect$1, connectDatabaseEmulator, refFromURL, ref, goOffline, goOnline, serverTimestamp, increment, forceWebSockets, forceLongPolling, _setSDKVersion, _repoManagerDatabaseFromApp, enableLogging } from '@firebase/database';
import ***REMOVED*** errorPrefix, validateArgCount, validateCallback, validateContextObject, Deferred } from '@firebase/util';
import ***REMOVED*** __extends } from 'tslib';
import ***REMOVED*** Logger } from '@firebase/logger';

var name = "@firebase/database-compat";
var version = "0.3.1";

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var logClient = new Logger('@firebase/database-compat');
var warn = function (msg) ***REMOVED***
    var message = 'FIREBASE WARNING: ' + msg;
    logClient.warn(message);
};

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var validateBoolean = function (fnName, argumentName, bool, optional) ***REMOVED***
    if (optional && bool === undefined) ***REMOVED***
        return;
  ***REMOVED***
    if (typeof bool !== 'boolean') ***REMOVED***
        throw new Error(errorPrefix(fnName, argumentName) + 'must be a boolean.');
  ***REMOVED***
};
var validateEventType = function (fnName, eventType, optional) ***REMOVED***
    if (optional && eventType === undefined) ***REMOVED***
        return;
  ***REMOVED***
    switch (eventType) ***REMOVED***
        case 'value':
        case 'child_added':
        case 'child_removed':
        case 'child_changed':
        case 'child_moved':
            break;
        default:
            throw new Error(errorPrefix(fnName, 'eventType') +
                'must be a valid event type = "value", "child_added", "child_removed", ' +
                '"child_changed", or "child_moved".');
  ***REMOVED***
};

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var OnDisconnect = /** @class */ (function () ***REMOVED***
    function OnDisconnect(_delegate) ***REMOVED***
        this._delegate = _delegate;
  ***REMOVED***
    OnDisconnect.prototype.cancel = function (onComplete) ***REMOVED***
        validateArgCount('OnDisconnect.cancel', 0, 1, arguments.length);
        validateCallback('OnDisconnect.cancel', 'onComplete', onComplete, true);
        var result = this._delegate.cancel();
        if (onComplete) ***REMOVED***
            result.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    OnDisconnect.prototype.remove = function (onComplete) ***REMOVED***
        validateArgCount('OnDisconnect.remove', 0, 1, arguments.length);
        validateCallback('OnDisconnect.remove', 'onComplete', onComplete, true);
        var result = this._delegate.remove();
        if (onComplete) ***REMOVED***
            result.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    OnDisconnect.prototype.set = function (value, onComplete) ***REMOVED***
        validateArgCount('OnDisconnect.set', 1, 2, arguments.length);
        validateCallback('OnDisconnect.set', 'onComplete', onComplete, true);
        var result = this._delegate.set(value);
        if (onComplete) ***REMOVED***
            result.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    OnDisconnect.prototype.setWithPriority = function (value, priority, onComplete) ***REMOVED***
        validateArgCount('OnDisconnect.setWithPriority', 2, 3, arguments.length);
        validateCallback('OnDisconnect.setWithPriority', 'onComplete', onComplete, true);
        var result = this._delegate.setWithPriority(value, priority);
        if (onComplete) ***REMOVED***
            result.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    OnDisconnect.prototype.update = function (objectToMerge, onComplete) ***REMOVED***
        validateArgCount('OnDisconnect.update', 1, 2, arguments.length);
        if (Array.isArray(objectToMerge)) ***REMOVED***
            var newObjectToMerge = ***REMOVED***};
            for (var i = 0; i < objectToMerge.length; ++i) ***REMOVED***
                newObjectToMerge['' + i] = objectToMerge[i];
          ***REMOVED***
            objectToMerge = newObjectToMerge;
            warn('Passing an Array to firebase.database.onDisconnect().update() is deprecated. Use set() if you want to overwrite the ' +
                'existing data, or an Object with integer keys if you really do want to only update some of the children.');
      ***REMOVED***
        validateCallback('OnDisconnect.update', 'onComplete', onComplete, true);
        var result = this._delegate.update(objectToMerge);
        if (onComplete) ***REMOVED***
            result.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    return OnDisconnect;
}());

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var TransactionResult = /** @class */ (function () ***REMOVED***
    /**
     * A type for the resolve value of Firebase.transaction.
     */
    function TransactionResult(committed, snapshot) ***REMOVED***
        this.committed = committed;
        this.snapshot = snapshot;
  ***REMOVED***
    // Do not create public documentation. This is intended to make JSON serialization work but is otherwise unnecessary
    // for end-users
    TransactionResult.prototype.toJSON = function () ***REMOVED***
        validateArgCount('TransactionResult.toJSON', 0, 1, arguments.length);
        return ***REMOVED*** committed: this.committed, snapshot: this.snapshot.toJSON() };
  ***REMOVED***;
    return TransactionResult;
}());

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Class representing a firebase data snapshot.  It wraps a SnapshotNode and
 * surfaces the public methods (val, forEach, etc.) we want to expose.
 */
var DataSnapshot = /** @class */ (function () ***REMOVED***
    function DataSnapshot(_database, _delegate) ***REMOVED***
        this._database = _database;
        this._delegate = _delegate;
  ***REMOVED***
    /**
     * Retrieves the snapshot contents as JSON.  Returns null if the snapshot is
     * empty.
     *
     * @returns JSON representation of the DataSnapshot contents, or null if empty.
     */
    DataSnapshot.prototype.val = function () ***REMOVED***
        validateArgCount('DataSnapshot.val', 0, 0, arguments.length);
        return this._delegate.val();
  ***REMOVED***;
    /**
     * Returns the snapshot contents as JSON, including priorities of node.  Suitable for exporting
     * the entire node contents.
     * @returns JSON representation of the DataSnapshot contents, or null if empty.
     */
    DataSnapshot.prototype.exportVal = function () ***REMOVED***
        validateArgCount('DataSnapshot.exportVal', 0, 0, arguments.length);
        return this._delegate.exportVal();
  ***REMOVED***;
    // Do not create public documentation. This is intended to make JSON serialization work but is otherwise unnecessary
    // for end-users
    DataSnapshot.prototype.toJSON = function () ***REMOVED***
        // Optional spacer argument is unnecessary because we're depending on recursion rather than stringifying the content
        validateArgCount('DataSnapshot.toJSON', 0, 1, arguments.length);
        return this._delegate.toJSON();
  ***REMOVED***;
    /**
     * Returns whether the snapshot contains a non-null value.
     *
     * @returns Whether the snapshot contains a non-null value, or is empty.
     */
    DataSnapshot.prototype.exists = function () ***REMOVED***
        validateArgCount('DataSnapshot.exists', 0, 0, arguments.length);
        return this._delegate.exists();
  ***REMOVED***;
    /**
     * Returns a DataSnapshot of the specified child node's contents.
     *
     * @param path - Path to a child.
     * @returns DataSnapshot for child node.
     */
    DataSnapshot.prototype.child = function (path) ***REMOVED***
        validateArgCount('DataSnapshot.child', 0, 1, arguments.length);
        // Ensure the childPath is a string (can be a number)
        path = String(path);
        _validatePathString('DataSnapshot.child', 'path', path, false);
        return new DataSnapshot(this._database, this._delegate.child(path));
  ***REMOVED***;
    /**
     * Returns whether the snapshot contains a child at the specified path.
     *
     * @param path - Path to a child.
     * @returns Whether the child exists.
     */
    DataSnapshot.prototype.hasChild = function (path) ***REMOVED***
        validateArgCount('DataSnapshot.hasChild', 1, 1, arguments.length);
        _validatePathString('DataSnapshot.hasChild', 'path', path, false);
        return this._delegate.hasChild(path);
  ***REMOVED***;
    /**
     * Returns the priority of the object, or null if no priority was set.
     *
     * @returns The priority.
     */
    DataSnapshot.prototype.getPriority = function () ***REMOVED***
        validateArgCount('DataSnapshot.getPriority', 0, 0, arguments.length);
        return this._delegate.priority;
  ***REMOVED***;
    /**
     * Iterates through child nodes and calls the specified action for each one.
     *
     * @param action - Callback function to be called
     * for each child.
     * @returns True if forEach was canceled by action returning true for
     * one of the child nodes.
     */
    DataSnapshot.prototype.forEach = function (action) ***REMOVED***
        var _this = this;
        validateArgCount('DataSnapshot.forEach', 1, 1, arguments.length);
        validateCallback('DataSnapshot.forEach', 'action', action, false);
        return this._delegate.forEach(function (expDataSnapshot) ***REMOVED***
            return action(new DataSnapshot(_this._database, expDataSnapshot));
      ***REMOVED***);
  ***REMOVED***;
    /**
     * Returns whether this DataSnapshot has children.
     * @returns True if the DataSnapshot contains 1 or more child nodes.
     */
    DataSnapshot.prototype.hasChildren = function () ***REMOVED***
        validateArgCount('DataSnapshot.hasChildren', 0, 0, arguments.length);
        return this._delegate.hasChildren();
  ***REMOVED***;
    Object.defineProperty(DataSnapshot.prototype, "key", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.key;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    /**
     * Returns the number of children for this DataSnapshot.
     * @returns The number of children that this DataSnapshot contains.
     */
    DataSnapshot.prototype.numChildren = function () ***REMOVED***
        validateArgCount('DataSnapshot.numChildren', 0, 0, arguments.length);
        return this._delegate.size;
  ***REMOVED***;
    /**
     * @returns The Firebase reference for the location this snapshot's data came
     * from.
     */
    DataSnapshot.prototype.getRef = function () ***REMOVED***
        validateArgCount('DataSnapshot.ref', 0, 0, arguments.length);
        return new Reference(this._database, this._delegate.ref);
  ***REMOVED***;
    Object.defineProperty(DataSnapshot.prototype, "ref", ***REMOVED***
        get: function () ***REMOVED***
            return this.getRef();
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    return DataSnapshot;
}());
/**
 * A Query represents a filter to be applied to a firebase location.  This object purely represents the
 * query expression (and exposes our public API to build the query).  The actual query logic is in ViewBase.js.
 *
 * Since every Firebase reference is a query, Firebase inherits from this object.
 */
var Query = /** @class */ (function () ***REMOVED***
    function Query(database, _delegate) ***REMOVED***
        this.database = database;
        this._delegate = _delegate;
  ***REMOVED***
    Query.prototype.on = function (eventType, callback, cancelCallbackOrContext, context) ***REMOVED***
        var _this = this;
        var _a;
        validateArgCount('Query.on', 2, 4, arguments.length);
        validateCallback('Query.on', 'callback', callback, false);
        var ret = Query.getCancelAndContextArgs_('Query.on', cancelCallbackOrContext, context);
        var valueCallback = function (expSnapshot, previousChildName) ***REMOVED***
            callback.call(ret.context, new DataSnapshot(_this.database, expSnapshot), previousChildName);
      ***REMOVED***;
        valueCallback.userCallback = callback;
        valueCallback.context = ret.context;
        var cancelCallback = (_a = ret.cancel) === null || _a === void 0 ? void 0 : _a.bind(ret.context);
        switch (eventType) ***REMOVED***
            case 'value':
                onValue(this._delegate, valueCallback, cancelCallback);
                return callback;
            case 'child_added':
                onChildAdded(this._delegate, valueCallback, cancelCallback);
                return callback;
            case 'child_removed':
                onChildRemoved(this._delegate, valueCallback, cancelCallback);
                return callback;
            case 'child_changed':
                onChildChanged(this._delegate, valueCallback, cancelCallback);
                return callback;
            case 'child_moved':
                onChildMoved(this._delegate, valueCallback, cancelCallback);
                return callback;
            default:
                throw new Error(errorPrefix('Query.on', 'eventType') +
                    'must be a valid event type = "value", "child_added", "child_removed", ' +
                    '"child_changed", or "child_moved".');
      ***REMOVED***
  ***REMOVED***;
    Query.prototype.off = function (eventType, callback, context) ***REMOVED***
        validateArgCount('Query.off', 0, 3, arguments.length);
        validateEventType('Query.off', eventType, true);
        validateCallback('Query.off', 'callback', callback, true);
        validateContextObject('Query.off', 'context', context, true);
        if (callback) ***REMOVED***
            var valueCallback = function () ***REMOVED*** };
            valueCallback.userCallback = callback;
            valueCallback.context = context;
            off(this._delegate, eventType, valueCallback);
      ***REMOVED***
        else ***REMOVED***
            off(this._delegate, eventType);
      ***REMOVED***
  ***REMOVED***;
    /**
     * Get the server-value for this query, or return a cached value if not connected.
     */
    Query.prototype.get = function () ***REMOVED***
        var _this = this;
        return get(this._delegate).then(function (expSnapshot) ***REMOVED***
            return new DataSnapshot(_this.database, expSnapshot);
      ***REMOVED***);
  ***REMOVED***;
    /**
     * Attaches a listener, waits for the first event, and then removes the listener
     */
    Query.prototype.once = function (eventType, callback, failureCallbackOrContext, context) ***REMOVED***
        var _this = this;
        validateArgCount('Query.once', 1, 4, arguments.length);
        validateCallback('Query.once', 'callback', callback, true);
        var ret = Query.getCancelAndContextArgs_('Query.once', failureCallbackOrContext, context);
        var deferred = new Deferred();
        var valueCallback = function (expSnapshot, previousChildName) ***REMOVED***
            var result = new DataSnapshot(_this.database, expSnapshot);
            if (callback) ***REMOVED***
                callback.call(ret.context, result, previousChildName);
          ***REMOVED***
            deferred.resolve(result);
      ***REMOVED***;
        valueCallback.userCallback = callback;
        valueCallback.context = ret.context;
        var cancelCallback = function (error) ***REMOVED***
            if (ret.cancel) ***REMOVED***
                ret.cancel.call(ret.context, error);
          ***REMOVED***
            deferred.reject(error);
      ***REMOVED***;
        switch (eventType) ***REMOVED***
            case 'value':
                onValue(this._delegate, valueCallback, cancelCallback, ***REMOVED***
                    onlyOnce: true
              ***REMOVED***);
                break;
            case 'child_added':
                onChildAdded(this._delegate, valueCallback, cancelCallback, ***REMOVED***
                    onlyOnce: true
              ***REMOVED***);
                break;
            case 'child_removed':
                onChildRemoved(this._delegate, valueCallback, cancelCallback, ***REMOVED***
                    onlyOnce: true
              ***REMOVED***);
                break;
            case 'child_changed':
                onChildChanged(this._delegate, valueCallback, cancelCallback, ***REMOVED***
                    onlyOnce: true
              ***REMOVED***);
                break;
            case 'child_moved':
                onChildMoved(this._delegate, valueCallback, cancelCallback, ***REMOVED***
                    onlyOnce: true
              ***REMOVED***);
                break;
            default:
                throw new Error(errorPrefix('Query.once', 'eventType') +
                    'must be a valid event type = "value", "child_added", "child_removed", ' +
                    '"child_changed", or "child_moved".');
      ***REMOVED***
        return deferred.promise;
  ***REMOVED***;
    /**
     * Set a limit and anchor it to the start of the window.
     */
    Query.prototype.limitToFirst = function (limit) ***REMOVED***
        validateArgCount('Query.limitToFirst', 1, 1, arguments.length);
        return new Query(this.database, query(this._delegate, limitToFirst(limit)));
  ***REMOVED***;
    /**
     * Set a limit and anchor it to the end of the window.
     */
    Query.prototype.limitToLast = function (limit) ***REMOVED***
        validateArgCount('Query.limitToLast', 1, 1, arguments.length);
        return new Query(this.database, query(this._delegate, limitToLast(limit)));
  ***REMOVED***;
    /**
     * Given a child path, return a new query ordered by the specified grandchild path.
     */
    Query.prototype.orderByChild = function (path) ***REMOVED***
        validateArgCount('Query.orderByChild', 1, 1, arguments.length);
        return new Query(this.database, query(this._delegate, orderByChild(path)));
  ***REMOVED***;
    /**
     * Return a new query ordered by the KeyIndex
     */
    Query.prototype.orderByKey = function () ***REMOVED***
        validateArgCount('Query.orderByKey', 0, 0, arguments.length);
        return new Query(this.database, query(this._delegate, orderByKey()));
  ***REMOVED***;
    /**
     * Return a new query ordered by the PriorityIndex
     */
    Query.prototype.orderByPriority = function () ***REMOVED***
        validateArgCount('Query.orderByPriority', 0, 0, arguments.length);
        return new Query(this.database, query(this._delegate, orderByPriority()));
  ***REMOVED***;
    /**
     * Return a new query ordered by the ValueIndex
     */
    Query.prototype.orderByValue = function () ***REMOVED***
        validateArgCount('Query.orderByValue', 0, 0, arguments.length);
        return new Query(this.database, query(this._delegate, orderByValue()));
  ***REMOVED***;
    Query.prototype.startAt = function (value, name) ***REMOVED***
        if (value === void 0) ***REMOVED*** value = null; }
        validateArgCount('Query.startAt', 0, 2, arguments.length);
        return new Query(this.database, query(this._delegate, startAt(value, name)));
  ***REMOVED***;
    Query.prototype.startAfter = function (value, name) ***REMOVED***
        if (value === void 0) ***REMOVED*** value = null; }
        validateArgCount('Query.startAfter', 0, 2, arguments.length);
        return new Query(this.database, query(this._delegate, startAfter(value, name)));
  ***REMOVED***;
    Query.prototype.endAt = function (value, name) ***REMOVED***
        if (value === void 0) ***REMOVED*** value = null; }
        validateArgCount('Query.endAt', 0, 2, arguments.length);
        return new Query(this.database, query(this._delegate, endAt(value, name)));
  ***REMOVED***;
    Query.prototype.endBefore = function (value, name) ***REMOVED***
        if (value === void 0) ***REMOVED*** value = null; }
        validateArgCount('Query.endBefore', 0, 2, arguments.length);
        return new Query(this.database, query(this._delegate, endBefore(value, name)));
  ***REMOVED***;
    /**
     * Load the selection of children with exactly the specified value, and, optionally,
     * the specified name.
     */
    Query.prototype.equalTo = function (value, name) ***REMOVED***
        validateArgCount('Query.equalTo', 1, 2, arguments.length);
        return new Query(this.database, query(this._delegate, equalTo(value, name)));
  ***REMOVED***;
    /**
     * @returns URL for this location.
     */
    Query.prototype.toString = function () ***REMOVED***
        validateArgCount('Query.toString', 0, 0, arguments.length);
        return this._delegate.toString();
  ***REMOVED***;
    // Do not create public documentation. This is intended to make JSON serialization work but is otherwise unnecessary
    // for end-users.
    Query.prototype.toJSON = function () ***REMOVED***
        // An optional spacer argument is unnecessary for a string.
        validateArgCount('Query.toJSON', 0, 1, arguments.length);
        return this._delegate.toJSON();
  ***REMOVED***;
    /**
     * Return true if this query and the provided query are equivalent; otherwise, return false.
     */
    Query.prototype.isEqual = function (other) ***REMOVED***
        validateArgCount('Query.isEqual', 1, 1, arguments.length);
        if (!(other instanceof Query)) ***REMOVED***
            var error = 'Query.isEqual failed: First argument must be an instance of firebase.database.Query.';
            throw new Error(error);
      ***REMOVED***
        return this._delegate.isEqual(other._delegate);
  ***REMOVED***;
    /**
     * Helper used by .on and .once to extract the context and or cancel arguments.
     * @param fnName - The function name (on or once)
     *
     */
    Query.getCancelAndContextArgs_ = function (fnName, cancelOrContext, context) ***REMOVED***
        var ret = ***REMOVED*** cancel: undefined, context: undefined };
        if (cancelOrContext && context) ***REMOVED***
            ret.cancel = cancelOrContext;
            validateCallback(fnName, 'cancel', ret.cancel, true);
            ret.context = context;
            validateContextObject(fnName, 'context', ret.context, true);
      ***REMOVED***
        else if (cancelOrContext) ***REMOVED***
            // we have either a cancel callback or a context.
            if (typeof cancelOrContext === 'object' && cancelOrContext !== null) ***REMOVED***
                // it's a context!
                ret.context = cancelOrContext;
          ***REMOVED***
            else if (typeof cancelOrContext === 'function') ***REMOVED***
                ret.cancel = cancelOrContext;
          ***REMOVED***
            else ***REMOVED***
                throw new Error(errorPrefix(fnName, 'cancelOrContext') +
                    ' must either be a cancel callback or a context object.');
          ***REMOVED***
      ***REMOVED***
        return ret;
  ***REMOVED***;
    Object.defineProperty(Query.prototype, "ref", ***REMOVED***
        get: function () ***REMOVED***
            return new Reference(this.database, new _ReferenceImpl(this._delegate._repo, this._delegate._path));
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    return Query;
}());
var Reference = /** @class */ (function (_super) ***REMOVED***
    __extends(Reference, _super);
    /**
     * Call options:
     *   new Reference(Repo, Path) or
     *   new Reference(url: string, string|RepoManager)
     *
     * Externally - this is the firebase.database.Reference type.
     */
    function Reference(database, _delegate) ***REMOVED***
        var _this = _super.call(this, database, new _QueryImpl(_delegate._repo, _delegate._path, new _QueryParams(), false)) || this;
        _this.database = database;
        _this._delegate = _delegate;
        return _this;
  ***REMOVED***
    /** @returns ***REMOVED***?string} */
    Reference.prototype.getKey = function () ***REMOVED***
        validateArgCount('Reference.key', 0, 0, arguments.length);
        return this._delegate.key;
  ***REMOVED***;
    Reference.prototype.child = function (pathString) ***REMOVED***
        validateArgCount('Reference.child', 1, 1, arguments.length);
        if (typeof pathString === 'number') ***REMOVED***
            pathString = String(pathString);
      ***REMOVED***
        return new Reference(this.database, child(this._delegate, pathString));
  ***REMOVED***;
    /** @returns ***REMOVED***?Reference} */
    Reference.prototype.getParent = function () ***REMOVED***
        validateArgCount('Reference.parent', 0, 0, arguments.length);
        var parent = this._delegate.parent;
        return parent ? new Reference(this.database, parent) : null;
  ***REMOVED***;
    /** @returns ***REMOVED***!Reference} */
    Reference.prototype.getRoot = function () ***REMOVED***
        validateArgCount('Reference.root', 0, 0, arguments.length);
        return new Reference(this.database, this._delegate.root);
  ***REMOVED***;
    Reference.prototype.set = function (newVal, onComplete) ***REMOVED***
        validateArgCount('Reference.set', 1, 2, arguments.length);
        validateCallback('Reference.set', 'onComplete', onComplete, true);
        var result = set(this._delegate, newVal);
        if (onComplete) ***REMOVED***
            result.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    Reference.prototype.update = function (values, onComplete) ***REMOVED***
        validateArgCount('Reference.update', 1, 2, arguments.length);
        if (Array.isArray(values)) ***REMOVED***
            var newObjectToMerge = ***REMOVED***};
            for (var i = 0; i < values.length; ++i) ***REMOVED***
                newObjectToMerge['' + i] = values[i];
          ***REMOVED***
            values = newObjectToMerge;
            warn('Passing an Array to Firebase.update() is deprecated. ' +
                'Use set() if you want to overwrite the existing data, or ' +
                'an Object with integer keys if you really do want to ' +
                'only update some of the children.');
      ***REMOVED***
        _validateWritablePath('Reference.update', this._delegate._path);
        validateCallback('Reference.update', 'onComplete', onComplete, true);
        var result = update(this._delegate, values);
        if (onComplete) ***REMOVED***
            result.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    Reference.prototype.setWithPriority = function (newVal, newPriority, onComplete) ***REMOVED***
        validateArgCount('Reference.setWithPriority', 2, 3, arguments.length);
        validateCallback('Reference.setWithPriority', 'onComplete', onComplete, true);
        var result = setWithPriority(this._delegate, newVal, newPriority);
        if (onComplete) ***REMOVED***
            result.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    Reference.prototype.remove = function (onComplete) ***REMOVED***
        validateArgCount('Reference.remove', 0, 1, arguments.length);
        validateCallback('Reference.remove', 'onComplete', onComplete, true);
        var result = remove(this._delegate);
        if (onComplete) ***REMOVED***
            result.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    Reference.prototype.transaction = function (transactionUpdate, onComplete, applyLocally) ***REMOVED***
        var _this = this;
        validateArgCount('Reference.transaction', 1, 3, arguments.length);
        validateCallback('Reference.transaction', 'transactionUpdate', transactionUpdate, false);
        validateCallback('Reference.transaction', 'onComplete', onComplete, true);
        validateBoolean('Reference.transaction', 'applyLocally', applyLocally, true);
        var result = runTransaction(this._delegate, transactionUpdate, ***REMOVED***
            applyLocally: applyLocally
      ***REMOVED***).then(function (transactionResult) ***REMOVED***
            return new TransactionResult(transactionResult.committed, new DataSnapshot(_this.database, transactionResult.snapshot));
      ***REMOVED***);
        if (onComplete) ***REMOVED***
            result.then(function (transactionResult) ***REMOVED***
                return onComplete(null, transactionResult.committed, transactionResult.snapshot);
          ***REMOVED***, function (error) ***REMOVED*** return onComplete(error, false, null); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    Reference.prototype.setPriority = function (priority, onComplete) ***REMOVED***
        validateArgCount('Reference.setPriority', 1, 2, arguments.length);
        validateCallback('Reference.setPriority', 'onComplete', onComplete, true);
        var result = setPriority(this._delegate, priority);
        if (onComplete) ***REMOVED***
            result.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        return result;
  ***REMOVED***;
    Reference.prototype.push = function (value, onComplete) ***REMOVED***
        var _this = this;
        validateArgCount('Reference.push', 0, 2, arguments.length);
        validateCallback('Reference.push', 'onComplete', onComplete, true);
        var expPromise = push(this._delegate, value);
        var promise = expPromise.then(function (expRef) ***REMOVED*** return new Reference(_this.database, expRef); });
        if (onComplete) ***REMOVED***
            promise.then(function () ***REMOVED*** return onComplete(null); }, function (error) ***REMOVED*** return onComplete(error); });
      ***REMOVED***
        var result = new Reference(this.database, expPromise);
        result.then = promise.then.bind(promise);
        result.catch = promise.catch.bind(promise, undefined);
        return result;
  ***REMOVED***;
    Reference.prototype.onDisconnect = function () ***REMOVED***
        _validateWritablePath('Reference.onDisconnect', this._delegate._path);
        return new OnDisconnect(new OnDisconnect$1(this._delegate._repo, this._delegate._path));
  ***REMOVED***;
    Object.defineProperty(Reference.prototype, "key", ***REMOVED***
        get: function () ***REMOVED***
            return this.getKey();
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(Reference.prototype, "parent", ***REMOVED***
        get: function () ***REMOVED***
            return this.getParent();
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(Reference.prototype, "root", ***REMOVED***
        get: function () ***REMOVED***
            return this.getRoot();
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    return Reference;
}(Query));

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Class representing a firebase database.
 */
var Database = /** @class */ (function () ***REMOVED***
    /**
     * The constructor should not be called by users of our public API.
     */
    function Database(_delegate, app) ***REMOVED***
        var _this = this;
        this._delegate = _delegate;
        this.app = app;
        this.INTERNAL = ***REMOVED***
            delete: function () ***REMOVED*** return _this._delegate._delete(); },
            forceWebSockets: forceWebSockets,
            forceLongPolling: forceLongPolling
      ***REMOVED***;
  ***REMOVED***
    /**
     * Modify this instance to communicate with the Realtime Database emulator.
     *
     * <p>Note: This method must be called before performing any other operation.
     *
     * @param host - the emulator host (ex: localhost)
     * @param port - the emulator port (ex: 8080)
     * @param options.mockUserToken - the mock auth token to use for unit testing Security Rules
     */
    Database.prototype.useEmulator = function (host, port, options) ***REMOVED***
        if (options === void 0) ***REMOVED*** options = ***REMOVED***}; }
        connectDatabaseEmulator(this._delegate, host, port, options);
  ***REMOVED***;
    Database.prototype.ref = function (path) ***REMOVED***
        validateArgCount('database.ref', 0, 1, arguments.length);
        if (path instanceof Reference) ***REMOVED***
            var childRef = refFromURL(this._delegate, path.toString());
            return new Reference(this, childRef);
      ***REMOVED***
        else ***REMOVED***
            var childRef = ref(this._delegate, path);
            return new Reference(this, childRef);
      ***REMOVED***
  ***REMOVED***;
    /**
     * Returns a reference to the root or the path specified in url.
     * We throw a exception if the url is not in the same domain as the
     * current repo.
     * @returns Firebase reference.
     */
    Database.prototype.refFromURL = function (url) ***REMOVED***
        var apiName = 'database.refFromURL';
        validateArgCount(apiName, 1, 1, arguments.length);
        var childRef = refFromURL(this._delegate, url);
        return new Reference(this, childRef);
  ***REMOVED***;
    // Make individual repo go offline.
    Database.prototype.goOffline = function () ***REMOVED***
        validateArgCount('database.goOffline', 0, 0, arguments.length);
        return goOffline(this._delegate);
  ***REMOVED***;
    Database.prototype.goOnline = function () ***REMOVED***
        validateArgCount('database.goOnline', 0, 0, arguments.length);
        return goOnline(this._delegate);
  ***REMOVED***;
    Database.ServerValue = ***REMOVED***
        TIMESTAMP: serverTimestamp(),
        increment: function (delta) ***REMOVED*** return increment(delta); }
  ***REMOVED***;
    return Database;
}());

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Used by console to create a database based on the app,
 * passed database URL and a custom auth implementation.
 *
 * @param app - A valid FirebaseApp-like object
 * @param url - A valid Firebase databaseURL
 * @param version - custom version e.g. firebase-admin version
 * @param customAuthImpl - custom auth implementation
 */
function initStandalone(_a) ***REMOVED***
    var app = _a.app, url = _a.url, version = _a.version, customAuthImpl = _a.customAuthImpl, namespace = _a.namespace, _b = _a.nodeAdmin, nodeAdmin = _b === void 0 ? false : _b;
    _setSDKVersion(version);
    /**
     * ComponentContainer('database-standalone') is just a placeholder that doesn't perform
     * any actual function.
     */
    var authProvider = new Provider('auth-internal', new ComponentContainer('database-standalone'));
    authProvider.setComponent(new Component('auth-internal', function () ***REMOVED*** return customAuthImpl; }, "PRIVATE" /* ComponentType.PRIVATE */));
    return ***REMOVED***
        instance: new Database(_repoManagerDatabaseFromApp(app, authProvider, 
        /* appCheckProvider= */ undefined, url, nodeAdmin), app),
        namespace: namespace
  ***REMOVED***;
}

var INTERNAL = /*#__PURE__*/Object.freeze(***REMOVED***
  __proto__: null,
  initStandalone: initStandalone
});

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var ServerValue = Database.ServerValue;
function registerDatabase(instance) ***REMOVED***
    // Register the Database Service with the 'firebase' namespace.
    instance.INTERNAL.registerComponent(new Component('database-compat', function (container, _a) ***REMOVED***
        var url = _a.instanceIdentifier;
        /* Dependencies */
        // getImmediate for FirebaseApp will always succeed
        var app = container.getProvider('app-compat').getImmediate();
        var databaseExp = container
            .getProvider('database')
            .getImmediate(***REMOVED*** identifier: url });
        return new Database(databaseExp, app);
  ***REMOVED***, "PUBLIC" /* ComponentType.PUBLIC */)
        .setServiceProps(
    // firebase.database namespace properties
    ***REMOVED***
        Reference: Reference,
        Query: Query,
        Database: Database,
        DataSnapshot: DataSnapshot,
        enableLogging: enableLogging,
        INTERNAL: INTERNAL,
        ServerValue: ServerValue
  ***REMOVED***)
        .setMultipleInstances(true));
    instance.registerVersion(name, version);
}
registerDatabase(firebase);

export ***REMOVED*** registerDatabase };
