// xmlcreate/lib/nodes/XmlAttribute.js
"use strict";
/**
 * Copyright (C) 2016-2019 Michael Kourlas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) ***REMOVED***
    return (mod && mod.__esModule) ? mod : ***REMOVED*** "default": mod };
};
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
var error_1 = require("../error");
var escape_1 = require("../escape");
var options_1 = require("../options");
var validate_1 = require("../validate");
var XmlAttributeText_1 = __importDefault(require("./XmlAttributeText"));
var XmlCharRef_1 = __importDefault(require("./XmlCharRef"));
var XmlEntityRef_1 = __importDefault(require("./XmlEntityRef"));
/**
 * Represents an attribute.
 *
 * An attribute is part of the start tag of an element and is
 * structured as follows, where `***REMOVED***name}` is the name of the attribute and
 * `***REMOVED***value}` is the value of the attribute:
 *
 * ```xml
 * <element ***REMOVED***name}="***REMOVED***value}">
 * ```
 *
 * The `***REMOVED***name}` value is a property of this node, while the `***REMOVED***value}` property
 * consists of the children of this node.
 *
 * Attributes can have an unlimited number of attribute text, character
 * references, and entity references.
 */
var XmlAttribute = /** @class */ (function () ***REMOVED***
    function XmlAttribute(parent, validation, options) ***REMOVED***
        this._validation = validation;
        if (!(0, validate_1.isUndefined)(options.replaceInvalidCharsInName)) ***REMOVED***
            this._replaceInvalidCharsInName = options.replaceInvalidCharsInName;
      ***REMOVED***
        else ***REMOVED***
            this._replaceInvalidCharsInName = false;
      ***REMOVED***
        this._children = [];
        this._parent = parent;
        this.name = options.name;
  ***REMOVED***
    Object.defineProperty(XmlAttribute.prototype, "name", ***REMOVED***
        /**
         * Gets the name of this attribute.
         */
        get: function () ***REMOVED***
            return this._name;
      ***REMOVED***,
        /**
         * Sets the name of this attribute.
         */
        set: function (name) ***REMOVED***
            if (this._replaceInvalidCharsInName) ***REMOVED***
                name = (0, validate_1.fixName)(name);
                if (name.length === 0) ***REMOVED***
                    throw new Error((0, error_1.getContext)(this.up()) + ": attribute name"
                        + " should not be empty");
              ***REMOVED***
          ***REMOVED***
            else if (this._validation && !(0, validate_1.validateName)(name)) ***REMOVED***
                if (name.length === 0) ***REMOVED***
                    throw new Error((0, error_1.getContext)(this.up()) + ": attribute name"
                        + " should not be empty");
              ***REMOVED***
                else ***REMOVED***
                    throw new Error((0, error_1.getContext)(this.up()) + ": attribute name"
                        + (" \"" + name + "\" should not contain characters not")
                        + " allowed in XML names");
              ***REMOVED***
          ***REMOVED***
            this._name = name;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    /**
     * Adds a character reference to this attribute and returns the new
     * character reference.
     */
    XmlAttribute.prototype.charRef = function (options) ***REMOVED***
        var charRef = new XmlCharRef_1.default(this, this._validation, options);
        this._children.push(charRef);
        return charRef;
  ***REMOVED***;
    /**
     * Adds an entity reference to this attribute and returns the new entity
     * reference.
     */
    XmlAttribute.prototype.entityRef = function (options) ***REMOVED***
        var charRef = new XmlEntityRef_1.default(this, this._validation, options);
        this._children.push(charRef);
        return charRef;
  ***REMOVED***;
    /**
     * Adds attribute text to this attribute and returns the new text.
     */
    XmlAttribute.prototype.text = function (options) ***REMOVED***
        var textNode = new XmlAttributeText_1.default(this, this._validation, options);
        this._children.push(textNode);
        return textNode;
  ***REMOVED***;
    /**
     * Returns an XML string representation of this attribute.
     */
    XmlAttribute.prototype.toString = function (options) ***REMOVED***
        if (options === void 0) ***REMOVED*** options = ***REMOVED***}; }
        var optionsObj = new options_1.StringOptions(options);
        var quote = optionsObj.doubleQuotes ? "\"" : "'";
        var str = this._name + "=" + quote;
        for (var _i = 0, _a = this._children; _i < _a.length; _i++) ***REMOVED***
            var child = _a[_i];
            if (optionsObj.doubleQuotes) ***REMOVED***
                str += (0, escape_1.escapeDoubleQuotes)(child.toString());
          ***REMOVED***
            else ***REMOVED***
                str += (0, escape_1.escapeSingleQuotes)(child.toString());
          ***REMOVED***
      ***REMOVED***
        str += quote;
        return str;
  ***REMOVED***;
    /**
     * Returns the parent of this attribute.
     */
    XmlAttribute.prototype.up = function () ***REMOVED***
        return this._parent;
  ***REMOVED***;
    return XmlAttribute;
}());
exports.default = XmlAttribute;
