// xmlcreate/lib/nodes/XmlElement.js
"use strict";
/**
 * Copyright (C) 2016-2019 Michael Kourlas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __importDefault = (this && this.__importDefault) || function (mod) ***REMOVED***
    return (mod && mod.__esModule) ? mod : ***REMOVED*** "default": mod };
};
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
var error_1 = require("../error");
var options_1 = require("../options");
var validate_1 = require("../validate");
var XmlAttribute_1 = __importDefault(require("./XmlAttribute"));
var XmlCdata_1 = __importDefault(require("./XmlCdata"));
var XmlCharData_1 = __importDefault(require("./XmlCharData"));
var XmlCharRef_1 = __importDefault(require("./XmlCharRef"));
var XmlComment_1 = __importDefault(require("./XmlComment"));
var XmlEntityRef_1 = __importDefault(require("./XmlEntityRef"));
var XmlProcInst_1 = __importDefault(require("./XmlProcInst"));
/**
 * Represents an XML element.
 *
 * A sample element is structured as follows, where `***REMOVED***name}` is the name
 * of the element:
 *
 * ```xml
 * <***REMOVED***name} attname="attvalue">
 *     <subelem/>
 *     <?pitarget picontent?>
 *     text
 * </***REMOVED***name}></pre>
 * ```
 *
 * XML elements can have an unlimited number of attributes, CDATA sections,
 * character references, comments, elements, entity references, processing
 * instructions, and character data.
 *
 * An element with no content will be represented using an empty element tag:
 *
 * ```xml
 * <***REMOVED***name}/>
 * ```
 */
var XmlElement = /** @class */ (function () ***REMOVED***
    function XmlElement(parent, validation, options) ***REMOVED***
        this._validation = validation;
        if (!(0, validate_1.isUndefined)(options.replaceInvalidCharsInName)) ***REMOVED***
            this._replaceInvalidCharsInName = options.replaceInvalidCharsInName;
      ***REMOVED***
        else ***REMOVED***
            this._replaceInvalidCharsInName = false;
      ***REMOVED***
        if (!(0, validate_1.isUndefined)(options.useSelfClosingTagIfEmpty)) ***REMOVED***
            this._useSelfClosingTagIfEmpty = options.useSelfClosingTagIfEmpty;
      ***REMOVED***
        else ***REMOVED***
            this._useSelfClosingTagIfEmpty = true;
      ***REMOVED***
        this._children = [];
        this._attributeNames = [];
        this._parent = parent;
        this.name = options.name;
  ***REMOVED***
    Object.defineProperty(XmlElement.prototype, "name", ***REMOVED***
        /**
         * Gets the name of this element.
         */
        get: function () ***REMOVED***
            return this._name;
      ***REMOVED***,
        /**
         * Sets the name of this element.
         */
        set: function (name) ***REMOVED***
            if (this._replaceInvalidCharsInName) ***REMOVED***
                name = (0, validate_1.fixName)(name);
                if (name.length === 0) ***REMOVED***
                    throw new Error((0, error_1.getContext)(this.up()) + ": element name should"
                        + " not be empty");
              ***REMOVED***
          ***REMOVED***
            else if (this._validation && !(0, validate_1.validateName)(name)) ***REMOVED***
                if (name.length === 0) ***REMOVED***
                    throw new Error((0, error_1.getContext)(this.up()) + ": element name should"
                        + " not be empty");
              ***REMOVED***
                else ***REMOVED***
                    throw new Error((0, error_1.getContext)(this.up()) + ": element name"
                        + (" \"" + name + "\" should not contain characters not")
                        + " allowed in XML names");
              ***REMOVED***
          ***REMOVED***
            this._name = name;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    /**
     * Adds an attribute to this element and returns the new attribute.
     */
    XmlElement.prototype.attribute = function (options) ***REMOVED***
        if (this._validation
            && this._attributeNames.indexOf(options.name) !== -1) ***REMOVED***
            throw new Error((0, error_1.getContext)(this.up()) + ": element \"" + this.name + "\""
                + " already contains an attribute with the"
                + (" name \"" + options.name + "\""));
      ***REMOVED***
        var attribute = new XmlAttribute_1.default(this, this._validation, options);
        this._children.push(attribute);
        this._attributeNames.push(options.name);
        return attribute;
  ***REMOVED***;
    /**
     * Adds a CDATA section to this element and returns the new CDATA section.
     */
    XmlElement.prototype.cdata = function (options) ***REMOVED***
        var cdata = new XmlCdata_1.default(this, this._validation, options);
        this._children.push(cdata);
        return cdata;
  ***REMOVED***;
    /**
     * Adds character data to this element and returns the new character data.
     */
    XmlElement.prototype.charData = function (options) ***REMOVED***
        var charDataNode = new XmlCharData_1.default(this, this._validation, options);
        this._children.push(charDataNode);
        return charDataNode;
  ***REMOVED***;
    /**
     * Adds a character reference to this element and returns the new
     * character reference.
     */
    XmlElement.prototype.charRef = function (options) ***REMOVED***
        var charRef = new XmlCharRef_1.default(this, this._validation, options);
        this._children.push(charRef);
        return charRef;
  ***REMOVED***;
    /**
     * Adds a comment to this element and returns the new comment.
     */
    XmlElement.prototype.comment = function (options) ***REMOVED***
        var comment = new XmlComment_1.default(this, this._validation, options);
        this._children.push(comment);
        return comment;
  ***REMOVED***;
    /**
     * Adds an element to this element and returns the new element.
     */
    XmlElement.prototype.element = function (options) ***REMOVED***
        var element = new XmlElement(this, this._validation, options);
        this._children.push(element);
        return element;
  ***REMOVED***;
    /**
     * Adds an entity reference to this element and returns the new entity
     * reference.
     */
    XmlElement.prototype.entityRef = function (options) ***REMOVED***
        var entityRef = new XmlEntityRef_1.default(this, this._validation, options);
        this._children.push(entityRef);
        return entityRef;
  ***REMOVED***;
    /**
     * Adds a processing instruction to this element and returns the new
     * processing instruction.
     */
    XmlElement.prototype.procInst = function (options) ***REMOVED***
        var procInst = new XmlProcInst_1.default(this, this._validation, options);
        this._children.push(procInst);
        return procInst;
  ***REMOVED***;
    /**
     * Returns an XML string representation of this element using the specified
     * options.
     */
    XmlElement.prototype.toString = function (options) ***REMOVED***
        if (options === void 0) ***REMOVED*** options = ***REMOVED***}; }
        return this.toStringWithIndent(options, "");
  ***REMOVED***;
    /**
     * Returns the parent of this element.
     */
    XmlElement.prototype.up = function () ***REMOVED***
        return this._parent;
  ***REMOVED***;
    /**
     * Returns an XML string representation of this element using the specified
     * options and initial indent.
     */
    XmlElement.prototype.toStringWithIndent = function (options, indent) ***REMOVED***
        var optionsObj = new options_1.StringOptions(options);
        var newIndent = indent + optionsObj.indent;
        // Element tag start
        var str = "<" + this._name;
        // Attributes and other nodes
        var nodes = [];
        for (var _i = 0, _a = this._children; _i < _a.length; _i++) ***REMOVED***
            var node = _a[_i];
            if (node instanceof XmlAttribute_1.default) ***REMOVED***
                str += " " + node.toString(options);
          ***REMOVED***
            else ***REMOVED***
                nodes.push(node);
          ***REMOVED***
      ***REMOVED***
        // Child nodes
        if (nodes.length > 0) ***REMOVED***
            var childStr = "";
            for (var i = 0; i < nodes.length; i++) ***REMOVED***
                var next = nodes[i];
                var nextStr = "";
                if (next instanceof XmlElement) ***REMOVED***
                    nextStr += next.toStringWithIndent(optionsObj, newIndent);
              ***REMOVED***
                else ***REMOVED***
                    nextStr += next.toString();
              ***REMOVED***
                var prev = i > 0 ? nodes[i - 1] : undefined;
                // Skip empty nodes
                if (next instanceof XmlCharData_1.default && next.toString() === "") ***REMOVED***
                    continue;
              ***REMOVED***
                // Line break before child nodes unless all nodes, or at least
                // the most recent two, are of type XmlCharacterReference,
                // XmlEntityReference, or XmlCharData
                if (optionsObj.pretty) ***REMOVED***
                    if (!this.allSameLineNodes(nodes)) ***REMOVED***
                        if (!(i > 0 && this.onSameLine(next, prev))) ***REMOVED***
                            nextStr = optionsObj.newline + newIndent + nextStr;
                      ***REMOVED***
                  ***REMOVED***
              ***REMOVED***
                childStr += nextStr;
          ***REMOVED***
            // Line break before end tag unless all nodes are of type
            // XmlCharacterReference, XmlEntityReference, or XmlCharData
            if (optionsObj.pretty) ***REMOVED***
                if (!this.allSameLineNodes(nodes)) ***REMOVED***
                    childStr += optionsObj.newline + indent;
              ***REMOVED***
          ***REMOVED***
            if (childStr.length === 0 && this._useSelfClosingTagIfEmpty) ***REMOVED***
                // Element empty tag end
                str += "/>";
          ***REMOVED***
            else ***REMOVED***
                // Element start and end tags
                str += ">" + childStr + "</" + this._name + ">";
          ***REMOVED***
      ***REMOVED***
        else if (this._useSelfClosingTagIfEmpty) ***REMOVED***
            // Element empty tag end
            str += "/>";
      ***REMOVED***
        else ***REMOVED***
            // Element start and end tags
            str += "></" + this._name + ">";
      ***REMOVED***
        return str;
  ***REMOVED***;
    /**
     * Returns true if the specified nodes are all character references,
     * entity references, or character data.
     */
    XmlElement.prototype.allSameLineNodes = function (nodes) ***REMOVED***
        for (var _i = 0, nodes_1 = nodes; _i < nodes_1.length; _i++) ***REMOVED***
            var node = nodes_1[_i];
            if (!((node instanceof XmlCharRef_1.default
                || node instanceof XmlEntityRef_1.default
                || node instanceof XmlCharData_1.default))) ***REMOVED***
                return false;
          ***REMOVED***
      ***REMOVED***
        return true;
  ***REMOVED***;
    /**
     * Returns true if the specified nodes are all character references,
     * entity references, or character data.
     */
    XmlElement.prototype.onSameLine = function (prev, next) ***REMOVED***
        return (prev instanceof XmlCharRef_1.default
            || prev instanceof XmlEntityRef_1.default
            || prev instanceof XmlCharData_1.default)
            && (!(0, validate_1.isUndefined)(next)
                && (next instanceof XmlCharRef_1.default
                    || next instanceof XmlEntityRef_1.default
                    || next instanceof XmlCharData_1.default));
  ***REMOVED***;
    return XmlElement;
}());
exports.default = XmlElement;
