// proto3-json-serializer/build/src/value.js
"use strict";
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.googleProtobufValueFromProto3JSON = exports.googleProtobufListValueFromProto3JSON = exports.googleProtobufStructFromProto3JSON = exports.googleProtobufValueToProto3JSON = exports.googleProtobufListValueToProto3JSON = exports.googleProtobufStructToProto3JSON = void 0;
const util_1 = require("./util");
function googleProtobufStructToProto3JSON(obj) ***REMOVED***
    const result = ***REMOVED***};
    const fields = obj.fields;
    for (const [key, value] of Object.entries(fields)) ***REMOVED***
        result[key] = googleProtobufValueToProto3JSON(value);
  ***REMOVED***
    return result;
}
exports.googleProtobufStructToProto3JSON = googleProtobufStructToProto3JSON;
function googleProtobufListValueToProto3JSON(obj) ***REMOVED***
    (0, util_1.assert)(Array.isArray(obj.values), 'ListValue internal representation must contain array of values');
    return obj.values.map(googleProtobufValueToProto3JSON);
}
exports.googleProtobufListValueToProto3JSON = googleProtobufListValueToProto3JSON;
function googleProtobufValueToProto3JSON(obj) ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(obj, 'nullValue')) ***REMOVED***
        return null;
  ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(obj, 'numberValue') &&
        typeof obj.numberValue === 'number') ***REMOVED***
        if (!Number.isFinite(obj.numberValue)) ***REMOVED***
            return obj.numberValue.toString();
      ***REMOVED***
        return obj.numberValue;
  ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(obj, 'stringValue') &&
        typeof obj.stringValue === 'string') ***REMOVED***
        return obj.stringValue;
  ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(obj, 'boolValue') &&
        typeof obj.boolValue === 'boolean') ***REMOVED***
        return obj.boolValue;
  ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(obj, 'structValue') &&
        typeof obj.structValue === 'object') ***REMOVED***
        return googleProtobufStructToProto3JSON(obj.structValue);
  ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(obj, 'listValue') &&
        typeof obj === 'object' &&
        typeof obj.listValue === 'object') ***REMOVED***
        return googleProtobufListValueToProto3JSON(obj.listValue);
  ***REMOVED***
    // Assuming empty Value to be null
    return null;
}
exports.googleProtobufValueToProto3JSON = googleProtobufValueToProto3JSON;
function googleProtobufStructFromProto3JSON(json) ***REMOVED***
    const fields = ***REMOVED***};
    for (const [key, value] of Object.entries(json)) ***REMOVED***
        fields[key] = googleProtobufValueFromProto3JSON(value);
  ***REMOVED***
    return ***REMOVED*** fields };
}
exports.googleProtobufStructFromProto3JSON = googleProtobufStructFromProto3JSON;
function googleProtobufListValueFromProto3JSON(json) ***REMOVED***
    return ***REMOVED***
        values: json.map(element => googleProtobufValueFromProto3JSON(element)),
  ***REMOVED***;
}
exports.googleProtobufListValueFromProto3JSON = googleProtobufListValueFromProto3JSON;
function googleProtobufValueFromProto3JSON(json) ***REMOVED***
    if (json === null) ***REMOVED***
        return ***REMOVED*** nullValue: 'NULL_VALUE' };
  ***REMOVED***
    if (typeof json === 'number') ***REMOVED***
        return ***REMOVED*** numberValue: json };
  ***REMOVED***
    if (typeof json === 'string') ***REMOVED***
        return ***REMOVED*** stringValue: json };
  ***REMOVED***
    if (typeof json === 'boolean') ***REMOVED***
        return ***REMOVED*** boolValue: json };
  ***REMOVED***
    if (Array.isArray(json)) ***REMOVED***
        return ***REMOVED***
            listValue: googleProtobufListValueFromProto3JSON(json),
      ***REMOVED***;
  ***REMOVED***
    if (typeof json === 'object') ***REMOVED***
        return ***REMOVED***
            structValue: googleProtobufStructFromProto3JSON(json),
      ***REMOVED***;
  ***REMOVED***
    throw new Error(`googleProtobufValueFromProto3JSON: incorrect parameter type: $***REMOVED***typeof json}`);
}
exports.googleProtobufValueFromProto3JSON = googleProtobufValueFromProto3JSON;
