// google-auth-library/build/src/auth/impersonated.js
"use strict";
/**
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.Impersonated = exports.IMPERSONATED_ACCOUNT_TYPE = void 0;
const oauth2client_1 = require("./oauth2client");
const gaxios_1 = require("gaxios");
exports.IMPERSONATED_ACCOUNT_TYPE = 'impersonated_service_account';
class Impersonated extends oauth2client_1.OAuth2Client ***REMOVED***
    /**
     * Impersonated service account credentials.
     *
     * Create a new access token by impersonating another service account.
     *
     * Impersonated Credentials allowing credentials issued to a user or
     * service account to impersonate another. The source project using
     * Impersonated Credentials must enable the "IAMCredentials" API.
     * Also, the target service account must grant the orginating principal
     * the "Service Account Token Creator" IAM role.
     *
     * @param ***REMOVED***object} options - The configuration object.
     * @param ***REMOVED***object} [options.sourceClient] the source credential used as to
     * acquire the impersonated credentials.
     * @param ***REMOVED***string} [options.targetPrincipal] the service account to
     * impersonate.
     * @param ***REMOVED***string[]} [options.delegates] the chained list of delegates
     * required to grant the final access_token. If set, the sequence of
     * identities must have "Service Account Token Creator" capability granted to
     * the preceding identity. For example, if set to [serviceAccountB,
     * serviceAccountC], the sourceCredential must have the Token Creator role on
     * serviceAccountB. serviceAccountB must have the Token Creator on
     * serviceAccountC. Finally, C must have Token Creator on target_principal.
     * If left unset, sourceCredential must have that role on targetPrincipal.
     * @param ***REMOVED***string[]} [options.targetScopes] scopes to request during the
     * authorization grant.
     * @param ***REMOVED***number} [options.lifetime] number of seconds the delegated
     * credential should be valid for up to 3600 seconds by default, or 43,200
     * seconds by extending the token's lifetime, see:
     * https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials#sa-credentials-oauth
     * @param ***REMOVED***string} [options.endpoint] api endpoint override.
     */
    constructor(options = ***REMOVED***}) ***REMOVED***
        var _a, _b, _c, _d, _e, _f;
        super(options);
        this.credentials = ***REMOVED***
            expiry_date: 1,
            refresh_token: 'impersonated-placeholder',
      ***REMOVED***;
        this.sourceClient = (_a = options.sourceClient) !== null && _a !== void 0 ? _a : new oauth2client_1.OAuth2Client();
        this.targetPrincipal = (_b = options.targetPrincipal) !== null && _b !== void 0 ? _b : '';
        this.delegates = (_c = options.delegates) !== null && _c !== void 0 ? _c : [];
        this.targetScopes = (_d = options.targetScopes) !== null && _d !== void 0 ? _d : [];
        this.lifetime = (_e = options.lifetime) !== null && _e !== void 0 ? _e : 3600;
        this.endpoint = (_f = options.endpoint) !== null && _f !== void 0 ? _f : 'https://iamcredentials.googleapis.com';
  ***REMOVED***
    /**
     * Refreshes the access token.
     * @param refreshToken Unused parameter
     */
    async refreshToken(refreshToken) ***REMOVED***
        var _a, _b, _c, _d, _e, _f;
        try ***REMOVED***
            await this.sourceClient.getAccessToken();
            const name = 'projects/-/serviceAccounts/' + this.targetPrincipal;
            const u = `$***REMOVED***this.endpoint}/v1/$***REMOVED***name}:generateAccessToken`;
            const body = ***REMOVED***
                delegates: this.delegates,
                scope: this.targetScopes,
                lifetime: this.lifetime + 's',
          ***REMOVED***;
            const res = await this.sourceClient.request(***REMOVED***
                url: u,
                data: body,
                method: 'POST',
          ***REMOVED***);
            const tokenResponse = res.data;
            this.credentials.access_token = tokenResponse.accessToken;
            this.credentials.expiry_date = Date.parse(tokenResponse.expireTime);
            return ***REMOVED***
                tokens: this.credentials,
                res,
          ***REMOVED***;
      ***REMOVED***
        catch (error) ***REMOVED***
            if (!(error instanceof Error))
                throw error;
            let status = 0;
            let message = '';
            if (error instanceof gaxios_1.GaxiosError) ***REMOVED***
                status = (_c = (_b = (_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.error) === null || _c === void 0 ? void 0 : _c.status;
                message = (_f = (_e = (_d = error === null || error === void 0 ? void 0 : error.response) === null || _d === void 0 ? void 0 : _d.data) === null || _e === void 0 ? void 0 : _e.error) === null || _f === void 0 ? void 0 : _f.message;
          ***REMOVED***
            if (status && message) ***REMOVED***
                error.message = `$***REMOVED***status}: unable to impersonate: $***REMOVED***message}`;
                throw error;
          ***REMOVED***
            else ***REMOVED***
                error.message = `unable to impersonate: $***REMOVED***error}`;
                throw error;
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    /**
     * Generates an OpenID Connect ID token for a service account.
     *
     * ***REMOVED***@link https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateIdToken Reference Documentation}
     *
     * @param targetAudience the audience for the fetched ID token.
     * @param options the for the request
     * @return an OpenID Connect ID token
     */
    async fetchIdToken(targetAudience, options) ***REMOVED***
        var _a;
        await this.sourceClient.getAccessToken();
        const name = `projects/-/serviceAccounts/$***REMOVED***this.targetPrincipal}`;
        const u = `$***REMOVED***this.endpoint}/v1/$***REMOVED***name}:generateIdToken`;
        const body = ***REMOVED***
            delegates: this.delegates,
            audience: targetAudience,
            includeEmail: (_a = options === null || options === void 0 ? void 0 : options.includeEmail) !== null && _a !== void 0 ? _a : true,
      ***REMOVED***;
        const res = await this.sourceClient.request(***REMOVED***
            url: u,
            data: body,
            method: 'POST',
      ***REMOVED***);
        return res.data.token;
  ***REMOVED***
}
exports.Impersonated = Impersonated;
