// protobufjs/src/namespace.js
"use strict";
module.exports = Namespace;

// extends ReflectionObject
var ReflectionObject = require("./object");
((Namespace.prototype = Object.create(ReflectionObject.prototype)).constructor = Namespace).className = "Namespace";

var Field    = require("./field"),
    OneOf    = require("./oneof"),
    util     = require("./util");

var Type,    // cyclic
    Service,
    Enum;

/**
 * Constructs a new namespace instance.
 * @name Namespace
 * @classdesc Reflected namespace.
 * @extends NamespaceBase
 * @constructor
 * @param ***REMOVED***string} name Namespace name
 * @param ***REMOVED***Object.<string,*>} [options] Declared options
 */

/**
 * Constructs a namespace from JSON.
 * @memberof Namespace
 * @function
 * @param ***REMOVED***string} name Namespace name
 * @param ***REMOVED***Object.<string,*>} json JSON object
 * @returns ***REMOVED***Namespace} Created namespace
 * @throws ***REMOVED***TypeError} If arguments are invalid
 */
Namespace.fromJSON = function fromJSON(name, json) ***REMOVED***
    return new Namespace(name, json.options).addJSON(json.nested);
};

/**
 * Converts an array of reflection objects to JSON.
 * @memberof Namespace
 * @param ***REMOVED***ReflectionObject[]} array Object array
 * @param ***REMOVED***IToJSONOptions} [toJSONOptions] JSON conversion options
 * @returns ***REMOVED***Object.<string,*>|undefined} JSON object or `undefined` when array is empty
 */
function arrayToJSON(array, toJSONOptions) ***REMOVED***
    if (!(array && array.length))
        return undefined;
    var obj = ***REMOVED***};
    for (var i = 0; i < array.length; ++i)
        obj[array[i].name] = array[i].toJSON(toJSONOptions);
    return obj;
}

Namespace.arrayToJSON = arrayToJSON;

/**
 * Tests if the specified id is reserved.
 * @param ***REMOVED***Array.<number[]|string>|undefined} reserved Array of reserved ranges and names
 * @param ***REMOVED***number} id Id to test
 * @returns ***REMOVED***boolean} `true` if reserved, otherwise `false`
 */
Namespace.isReservedId = function isReservedId(reserved, id) ***REMOVED***
    if (reserved)
        for (var i = 0; i < reserved.length; ++i)
            if (typeof reserved[i] !== "string" && reserved[i][0] <= id && reserved[i][1] > id)
                return true;
    return false;
};

/**
 * Tests if the specified name is reserved.
 * @param ***REMOVED***Array.<number[]|string>|undefined} reserved Array of reserved ranges and names
 * @param ***REMOVED***string} name Name to test
 * @returns ***REMOVED***boolean} `true` if reserved, otherwise `false`
 */
Namespace.isReservedName = function isReservedName(reserved, name) ***REMOVED***
    if (reserved)
        for (var i = 0; i < reserved.length; ++i)
            if (reserved[i] === name)
                return true;
    return false;
};

/**
 * Not an actual constructor. Use ***REMOVED***@link Namespace} instead.
 * @classdesc Base class of all reflection objects containing nested objects. This is not an actual class but here for the sake of having consistent type definitions.
 * @exports NamespaceBase
 * @extends ReflectionObject
 * @abstract
 * @constructor
 * @param ***REMOVED***string} name Namespace name
 * @param ***REMOVED***Object.<string,*>} [options] Declared options
 * @see ***REMOVED***@link Namespace}
 */
function Namespace(name, options) ***REMOVED***
    ReflectionObject.call(this, name, options);

    /**
     * Nested objects by name.
     * @type ***REMOVED***Object.<string,ReflectionObject>|undefined}
     */
    this.nested = undefined; // toJSON

    /**
     * Cached nested objects as an array.
     * @type ***REMOVED***ReflectionObject[]|null}
     * @private
     */
    this._nestedArray = null;
}

function clearCache(namespace) ***REMOVED***
    namespace._nestedArray = null;
    return namespace;
}

/**
 * Nested objects of this namespace as an array for iteration.
 * @name NamespaceBase#nestedArray
 * @type ***REMOVED***ReflectionObject[]}
 * @readonly
 */
Object.defineProperty(Namespace.prototype, "nestedArray", ***REMOVED***
    get: function() ***REMOVED***
        return this._nestedArray || (this._nestedArray = util.toArray(this.nested));
  ***REMOVED***
});

/**
 * Namespace descriptor.
 * @interface INamespace
 * @property ***REMOVED***Object.<string,*>} [options] Namespace options
 * @property ***REMOVED***Object.<string,AnyNestedObject>} [nested] Nested object descriptors
 */

/**
 * Any extension field descriptor.
 * @typedef AnyExtensionField
 * @type ***REMOVED***IExtensionField|IExtensionMapField}
 */

/**
 * Any nested object descriptor.
 * @typedef AnyNestedObject
 * @type ***REMOVED***IEnum|IType|IService|AnyExtensionField|INamespace}
 */
// ^ BEWARE: VSCode hangs forever when using more than 5 types (that's why AnyExtensionField exists in the first place)

/**
 * Converts this namespace to a namespace descriptor.
 * @param ***REMOVED***IToJSONOptions} [toJSONOptions] JSON conversion options
 * @returns ***REMOVED***INamespace} Namespace descriptor
 */
Namespace.prototype.toJSON = function toJSON(toJSONOptions) ***REMOVED***
    return util.toObject([
        "options" , this.options,
        "nested"  , arrayToJSON(this.nestedArray, toJSONOptions)
    ]);
};

/**
 * Adds nested objects to this namespace from nested object descriptors.
 * @param ***REMOVED***Object.<string,AnyNestedObject>} nestedJson Any nested object descriptors
 * @returns ***REMOVED***Namespace} `this`
 */
Namespace.prototype.addJSON = function addJSON(nestedJson) ***REMOVED***
    var ns = this;
    /* istanbul ignore else */
    if (nestedJson) ***REMOVED***
        for (var names = Object.keys(nestedJson), i = 0, nested; i < names.length; ++i) ***REMOVED***
            nested = nestedJson[names[i]];
            ns.add( // most to least likely
                ( nested.fields !== undefined
                ? Type.fromJSON
                : nested.values !== undefined
                ? Enum.fromJSON
                : nested.methods !== undefined
                ? Service.fromJSON
                : nested.id !== undefined
                ? Field.fromJSON
                : Namespace.fromJSON )(names[i], nested)
            );
      ***REMOVED***
  ***REMOVED***
    return this;
};

/**
 * Gets the nested object of the specified name.
 * @param ***REMOVED***string} name Nested object name
 * @returns ***REMOVED***ReflectionObject|null} The reflection object or `null` if it doesn't exist
 */
Namespace.prototype.get = function get(name) ***REMOVED***
    return this.nested && this.nested[name]
        || null;
};

/**
 * Gets the values of the nested ***REMOVED***@link Enum|enum} of the specified name.
 * This methods differs from ***REMOVED***@link Namespace#get|get} in that it returns an enum's values directly and throws instead of returning `null`.
 * @param ***REMOVED***string} name Nested enum name
 * @returns ***REMOVED***Object.<string,number>} Enum values
 * @throws ***REMOVED***Error} If there is no such enum
 */
Namespace.prototype.getEnum = function getEnum(name) ***REMOVED***
    if (this.nested && this.nested[name] instanceof Enum)
        return this.nested[name].values;
    throw Error("no such enum: " + name);
};

/**
 * Adds a nested object to this namespace.
 * @param ***REMOVED***ReflectionObject} object Nested object to add
 * @returns ***REMOVED***Namespace} `this`
 * @throws ***REMOVED***TypeError} If arguments are invalid
 * @throws ***REMOVED***Error} If there is already a nested object with this name
 */
Namespace.prototype.add = function add(object) ***REMOVED***

    if (!(object instanceof Field && object.extend !== undefined || object instanceof Type || object instanceof Enum || object instanceof Service || object instanceof Namespace || object instanceof OneOf))
        throw TypeError("object must be a valid nested object");

    if (!this.nested)
        this.nested = ***REMOVED***};
    else ***REMOVED***
        var prev = this.get(object.name);
        if (prev) ***REMOVED***
            if (prev instanceof Namespace && object instanceof Namespace && !(prev instanceof Type || prev instanceof Service)) ***REMOVED***
                // replace plain namespace but keep existing nested elements and options
                var nested = prev.nestedArray;
                for (var i = 0; i < nested.length; ++i)
                    object.add(nested[i]);
                this.remove(prev);
                if (!this.nested)
                    this.nested = ***REMOVED***};
                object.setOptions(prev.options, true);

          ***REMOVED*** else
                throw Error("duplicate name '" + object.name + "' in " + this);
      ***REMOVED***
  ***REMOVED***
    this.nested[object.name] = object;
    object.onAdd(this);
    return clearCache(this);
};

/**
 * Removes a nested object from this namespace.
 * @param ***REMOVED***ReflectionObject} object Nested object to remove
 * @returns ***REMOVED***Namespace} `this`
 * @throws ***REMOVED***TypeError} If arguments are invalid
 * @throws ***REMOVED***Error} If `object` is not a member of this namespace
 */
Namespace.prototype.remove = function remove(object) ***REMOVED***

    if (!(object instanceof ReflectionObject))
        throw TypeError("object must be a ReflectionObject");
    if (object.parent !== this)
        throw Error(object + " is not a member of " + this);

    delete this.nested[object.name];
    if (!Object.keys(this.nested).length)
        this.nested = undefined;

    object.onRemove(this);
    return clearCache(this);
};

/**
 * Defines additial namespaces within this one if not yet existing.
 * @param ***REMOVED***string|string[]} path Path to create
 * @param ***REMOVED****} [json] Nested types to create from JSON
 * @returns ***REMOVED***Namespace} Pointer to the last namespace created or `this` if path is empty
 */
Namespace.prototype.define = function define(path, json) ***REMOVED***

    if (util.isString(path))
        path = path.split(".");
    else if (!Array.isArray(path))
        throw TypeError("illegal path");
    if (path && path.length && path[0] === "")
        throw Error("path must be relative");

    var ptr = this;
    while (path.length > 0) ***REMOVED***
        var part = path.shift();
        if (ptr.nested && ptr.nested[part]) ***REMOVED***
            ptr = ptr.nested[part];
            if (!(ptr instanceof Namespace))
                throw Error("path conflicts with non-namespace objects");
      ***REMOVED*** else
            ptr.add(ptr = new Namespace(part));
  ***REMOVED***
    if (json)
        ptr.addJSON(json);
    return ptr;
};

/**
 * Resolves this namespace's and all its nested objects' type references. Useful to validate a reflection tree, but comes at a cost.
 * @returns ***REMOVED***Namespace} `this`
 */
Namespace.prototype.resolveAll = function resolveAll() ***REMOVED***
    var nested = this.nestedArray, i = 0;
    while (i < nested.length)
        if (nested[i] instanceof Namespace)
            nested[i++].resolveAll();
        else
            nested[i++].resolve();
    return this.resolve();
};

/**
 * Recursively looks up the reflection object matching the specified path in the scope of this namespace.
 * @param ***REMOVED***string|string[]} path Path to look up
 * @param ***REMOVED****|Array.<*>} filterTypes Filter types, any combination of the constructors of `protobuf.Type`, `protobuf.Enum`, `protobuf.Service` etc.
 * @param ***REMOVED***boolean} [parentAlreadyChecked=false] If known, whether the parent has already been checked
 * @returns ***REMOVED***ReflectionObject|null} Looked up object or `null` if none could be found
 */
Namespace.prototype.lookup = function lookup(path, filterTypes, parentAlreadyChecked) ***REMOVED***

    /* istanbul ignore next */
    if (typeof filterTypes === "boolean") ***REMOVED***
        parentAlreadyChecked = filterTypes;
        filterTypes = undefined;
  ***REMOVED*** else if (filterTypes && !Array.isArray(filterTypes))
        filterTypes = [ filterTypes ];

    if (util.isString(path) && path.length) ***REMOVED***
        if (path === ".")
            return this.root;
        path = path.split(".");
  ***REMOVED*** else if (!path.length)
        return this;

    // Start at root if path is absolute
    if (path[0] === "")
        return this.root.lookup(path.slice(1), filterTypes);

    // Test if the first part matches any nested object, and if so, traverse if path contains more
    var found = this.get(path[0]);
    if (found) ***REMOVED***
        if (path.length === 1) ***REMOVED***
            if (!filterTypes || filterTypes.indexOf(found.constructor) > -1)
                return found;
      ***REMOVED*** else if (found instanceof Namespace && (found = found.lookup(path.slice(1), filterTypes, true)))
            return found;

    // Otherwise try each nested namespace
  ***REMOVED*** else
        for (var i = 0; i < this.nestedArray.length; ++i)
            if (this._nestedArray[i] instanceof Namespace && (found = this._nestedArray[i].lookup(path, filterTypes, true)))
                return found;

    // If there hasn't been a match, try again at the parent
    if (this.parent === null || parentAlreadyChecked)
        return null;
    return this.parent.lookup(path, filterTypes);
};

/**
 * Looks up the reflection object at the specified path, relative to this namespace.
 * @name NamespaceBase#lookup
 * @function
 * @param ***REMOVED***string|string[]} path Path to look up
 * @param ***REMOVED***boolean} [parentAlreadyChecked=false] Whether the parent has already been checked
 * @returns ***REMOVED***ReflectionObject|null} Looked up object or `null` if none could be found
 * @variation 2
 */
// lookup(path: string, [parentAlreadyChecked: boolean])

/**
 * Looks up the ***REMOVED***@link Type|type} at the specified path, relative to this namespace.
 * Besides its signature, this methods differs from ***REMOVED***@link Namespace#lookup|lookup} in that it throws instead of returning `null`.
 * @param ***REMOVED***string|string[]} path Path to look up
 * @returns ***REMOVED***Type} Looked up type
 * @throws ***REMOVED***Error} If `path` does not point to a type
 */
Namespace.prototype.lookupType = function lookupType(path) ***REMOVED***
    var found = this.lookup(path, [ Type ]);
    if (!found)
        throw Error("no such type: " + path);
    return found;
};

/**
 * Looks up the values of the ***REMOVED***@link Enum|enum} at the specified path, relative to this namespace.
 * Besides its signature, this methods differs from ***REMOVED***@link Namespace#lookup|lookup} in that it throws instead of returning `null`.
 * @param ***REMOVED***string|string[]} path Path to look up
 * @returns ***REMOVED***Enum} Looked up enum
 * @throws ***REMOVED***Error} If `path` does not point to an enum
 */
Namespace.prototype.lookupEnum = function lookupEnum(path) ***REMOVED***
    var found = this.lookup(path, [ Enum ]);
    if (!found)
        throw Error("no such Enum '" + path + "' in " + this);
    return found;
};

/**
 * Looks up the ***REMOVED***@link Type|type} or ***REMOVED***@link Enum|enum} at the specified path, relative to this namespace.
 * Besides its signature, this methods differs from ***REMOVED***@link Namespace#lookup|lookup} in that it throws instead of returning `null`.
 * @param ***REMOVED***string|string[]} path Path to look up
 * @returns ***REMOVED***Type} Looked up type or enum
 * @throws ***REMOVED***Error} If `path` does not point to a type or enum
 */
Namespace.prototype.lookupTypeOrEnum = function lookupTypeOrEnum(path) ***REMOVED***
    var found = this.lookup(path, [ Type, Enum ]);
    if (!found)
        throw Error("no such Type or Enum '" + path + "' in " + this);
    return found;
};

/**
 * Looks up the ***REMOVED***@link Service|service} at the specified path, relative to this namespace.
 * Besides its signature, this methods differs from ***REMOVED***@link Namespace#lookup|lookup} in that it throws instead of returning `null`.
 * @param ***REMOVED***string|string[]} path Path to look up
 * @returns ***REMOVED***Service} Looked up service
 * @throws ***REMOVED***Error} If `path` does not point to a service
 */
Namespace.prototype.lookupService = function lookupService(path) ***REMOVED***
    var found = this.lookup(path, [ Service ]);
    if (!found)
        throw Error("no such Service '" + path + "' in " + this);
    return found;
};

// Sets up cyclic dependencies (called in index-light)
Namespace._configure = function(Type_, Service_, Enum_) ***REMOVED***
    Type    = Type_;
    Service = Service_;
    Enum    = Enum_;
};
