// jsdoc/lib/jsdoc/src/astbuilder.js
const babelParser = require('@babel/parser');
const env = require('jsdoc/env');
const logger = require('jsdoc/util/logger');

// exported so we can use them in tests
const parserOptions = exports.parserOptions = ***REMOVED***
    allowAwaitOutsideFunction: true,
    allowImportExportEverywhere: true,
    allowReturnOutsideFunction: true,
    allowSuperOutsideMethod: true,
    plugins: [
        'asyncGenerators',
        'bigInt',
        'classPrivateMethods',
        'classPrivateProperties',
        'classProperties',
        ['decorators', ***REMOVED***
            decoratorsBeforeExport: true
      ***REMOVED***],
        'doExpressions',
        'dynamicImport',
        'estree',
        'exportDefaultFrom',
        'exportNamespaceFrom',
        'functionBind',
        'functionSent',
        'importMeta',
        'jsx',
        'logicalAssignment',
        'nullishCoalescingOperator',
        'numericSeparator',
        'objectRestSpread',
        'optionalCatchBinding',
        'optionalChaining',
        ['pipelineOperator', ***REMOVED***
            proposal: 'minimal'
      ***REMOVED***],
        'throwExpressions'
    ],
    ranges: true,
    sourceType: env.conf.sourceType
};

function parse(source, filename) ***REMOVED***
    let ast;

    try ***REMOVED***
        ast = babelParser.parse(source, parserOptions);
        // console.log(JSON.stringify(ast, null, 2));
  ***REMOVED***
    catch (e) ***REMOVED***
        logger.error('Unable to parse %s: %s', filename, e.message);
  ***REMOVED***

    return ast;
}

// TODO: docs
class AstBuilder ***REMOVED***
    // TODO: docs
    /* eslint-disable no-empty-function */
    constructor() ***REMOVED***}
    /* eslint-enable no-empty-function */

    // TODO: docs
    /* eslint-disable class-methods-use-this */
    build(source, filename) ***REMOVED***
        return parse(source, filename);
  ***REMOVED***
    /* eslint-enable class-methods-use-this */
}
exports.AstBuilder = AstBuilder;
