// jsdoc/lib/jsdoc/src/walker.js
/**
 * Traversal utilities for ASTs that are compatible with the ESTree API.
 *
 * @module jsdoc/src/walker
 */
const astnode = require('jsdoc/src/astnode');
const logger = require('jsdoc/util/logger');
const Syntax = require('jsdoc/src/syntax').Syntax;

// TODO: docs
function getCurrentScope(scopes) ***REMOVED***
    return scopes[scopes.length - 1] || null;
}

// TODO: docs
function moveLeadingComments(source, target, count) ***REMOVED***
    if (source.leadingComments) ***REMOVED***
        if (count === undefined) ***REMOVED***
            count = source.leadingComments.length;
      ***REMOVED***

        target.leadingComments = source.leadingComments.slice(0, count);
        source.leadingComments = source.leadingComments.slice(count);
  ***REMOVED***
}

// TODO: docs
function moveTrailingComments(source, target, count) ***REMOVED***
    if (source.trailingComments) ***REMOVED***
        if (count === undefined) ***REMOVED***
            count = source.trailingComments.length;
      ***REMOVED***

        target.trailingComments = source.trailingComments.slice(
            source.trailingComments.length - count, count
        );
        source.trailingComments = source.trailingComments.slice(0);
  ***REMOVED***
}

/* eslint-disable no-empty-function, no-unused-vars */
function leafNode(node, parent, state, cb) ***REMOVED***}
/* eslint-enable no-empty-function, no-unused-vars */

// TODO: docs
const walkers = exports.walkers = ***REMOVED***};

walkers[Syntax.ArrayExpression] = (node, parent, state, cb) => ***REMOVED***
    for (let element of node.elements) ***REMOVED***
        if (element) ***REMOVED***
            cb(element, node, state);
      ***REMOVED***
  ***REMOVED***
};

// TODO: verify correctness
walkers[Syntax.ArrayPattern] = (node, parent, state, cb) => ***REMOVED***
    for (let element of node.elements) ***REMOVED***
        // must be an identifier or an expression
        if (element && element.type !== Syntax.Identifier) ***REMOVED***
            cb(element, node, state);
      ***REMOVED***
  ***REMOVED***
};

walkers[Syntax.ArrowFunctionExpression] = (node, parent, state, cb) => ***REMOVED***
    if (node.id) ***REMOVED***
        cb(node.id, node, state);
  ***REMOVED***

    for (let param of node.params) ***REMOVED***
        cb(param, node, state);
  ***REMOVED***

    cb(node.body, node, state);
};

walkers[Syntax.AssignmentExpression] = (node, parent, state, cb) => ***REMOVED***
    cb(node.left, node, state);
    cb(node.right, node, state);
};

walkers[Syntax.AssignmentPattern] = walkers[Syntax.AssignmentExpression];

walkers[Syntax.AwaitExpression] = (node, parent, state, cb) => ***REMOVED***
    cb(node.argument, node, state);
};

walkers[Syntax.BigIntLiteral] = leafNode;

walkers[Syntax.BinaryExpression] = (node, parent, state, cb) => ***REMOVED***
    cb(node.left, node, state);
    cb(node.right, node, state);
};

walkers[Syntax.BindExpression] = (node, parent, state, cb) => ***REMOVED***
    if (node.object) ***REMOVED***
        cb(node.object, node, state);
  ***REMOVED***

    cb(node.callee, node, state);
};

walkers[Syntax.BlockStatement] = (node, parent, state, cb) => ***REMOVED***
    for (let bodyItem of node.body) ***REMOVED***
        cb(bodyItem, node, state);
  ***REMOVED***
};

walkers[Syntax.BreakStatement] = leafNode;

walkers[Syntax.CallExpression] = function(node, parent, state, cb) ***REMOVED***
    cb(node.callee, node, state);

    if (node.arguments) ***REMOVED***
        for (let arg of node.arguments) ***REMOVED***
            cb(arg, node, state);
      ***REMOVED***
  ***REMOVED***
};

walkers[Syntax.CatchClause] = leafNode;

walkers[Syntax.ClassBody] = walkers[Syntax.BlockStatement];

walkers[Syntax.ClassDeclaration] = (node, parent, state, cb) => ***REMOVED***
    if (node.id) ***REMOVED***
        cb(node.id, node, state);
  ***REMOVED***

    if (node.superClass) ***REMOVED***
        cb(node.superClass, node, state);
  ***REMOVED***

    if (node.body) ***REMOVED***
        cb(node.body, node, state);
  ***REMOVED***

    if (node.decorators) ***REMOVED***
        for (let decorator of node.decorators) ***REMOVED***
            cb(decorator, node, state);
      ***REMOVED***
  ***REMOVED***
};

walkers[Syntax.ClassExpression] = walkers[Syntax.ClassDeclaration];

// walkers[Syntax.ClassPrivateProperty] is defined later

// walkers[Syntax.ClassProperty] is defined later

// TODO: verify correctness
walkers[Syntax.ComprehensionBlock] = walkers[Syntax.AssignmentExpression];

// TODO: verify correctness
walkers[Syntax.ComprehensionExpression] = (node, parent, state, cb) => ***REMOVED***
    cb(node.body, node, state);

    if (node.filter) ***REMOVED***
        cb(node.filter, node, state);
  ***REMOVED***

    for (let block of node.blocks) ***REMOVED***
        cb(block, node, state);
  ***REMOVED***
};

walkers[Syntax.ConditionalExpression] = (node, parent, state, cb) => ***REMOVED***
    cb(node.test, node, state);
    cb(node.consequent, node, state);
    cb(node.alternate, node, state);
};

walkers[Syntax.ContinueStatement] = leafNode;

walkers[Syntax.DebuggerStatement] = leafNode;

walkers[Syntax.Decorator] = (node, parent, state, cb) => ***REMOVED***
    cb(node.expression, node, state);
};

walkers[Syntax.DoExpression] = (node, parent, state, cb) => ***REMOVED***
    cb(node.body, node, state);
};

walkers[Syntax.DoWhileStatement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.test, node, state);
    cb(node.body, node, state);
};

walkers[Syntax.EmptyStatement] = leafNode;

walkers[Syntax.ExperimentalRestProperty] = (node, parent, state, cb) => ***REMOVED***
    cb(node.argument, node, state);
};

walkers[Syntax.ExperimentalSpreadProperty] = walkers[Syntax.ExperimentalRestProperty];

walkers[Syntax.ExportAllDeclaration] = (node, parent, state, cb) => ***REMOVED***
    if (node.source) ***REMOVED***
        cb(node.source, node, state);
  ***REMOVED***
};

walkers[Syntax.ExportDefaultDeclaration] = (node, parent, state, cb) => ***REMOVED***
    // if the declaration target is a class, move leading comments to the declaration target
    if (node.declaration && node.declaration.type === Syntax.ClassDeclaration) ***REMOVED***
        moveLeadingComments(node, node.declaration);
  ***REMOVED***

    if (node.declaration) ***REMOVED***
        cb(node.declaration, node, state);
  ***REMOVED***
};

walkers[Syntax.ExportDefaultSpecifier] = (node, parent, state, cb) => ***REMOVED***
    cb(node.exported, node, state);
};

walkers[Syntax.ExportNamedDeclaration] = (node, parent, state, cb) => ***REMOVED***
    if (node.declaration) ***REMOVED***
        cb(node.declaration, node, state);
  ***REMOVED***

    for (let specifier of node.specifiers) ***REMOVED***
        cb(specifier, node, state);
  ***REMOVED***

    if (node.source) ***REMOVED***
        cb(node.source, node, state);
  ***REMOVED***
};

walkers[Syntax.ExportNamespaceSpecifier] = (node, parent, state, cb) => ***REMOVED***
    cb(node.exported, node, state);
};

walkers[Syntax.ExportSpecifier] = (node, parent, state, cb) => ***REMOVED***
    if (node.exported) ***REMOVED***
        cb(node.exported, node, state);
  ***REMOVED***

    if (node.local) ***REMOVED***
        cb(node.local, node, state);
  ***REMOVED***
};

walkers[Syntax.ExpressionStatement] = (node, parent, state, cb) => ***REMOVED***
    moveLeadingComments(node, node.expression);

    cb(node.expression, node, state);
};

walkers[Syntax.File] = (node, parent, state, cb) => ***REMOVED***
    cb(node.program, node, state);
};

walkers[Syntax.ForInStatement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.left, node, state);
    cb(node.right, node, state);
    cb(node.body, node, state);
};

walkers[Syntax.ForOfStatement] = walkers[Syntax.ForInStatement];

walkers[Syntax.ForStatement] = (node, parent, state, cb) => ***REMOVED***
    if (node.init) ***REMOVED***
        cb(node.init, node, state);
  ***REMOVED***

    if (node.test) ***REMOVED***
        cb(node.test, node, state);
  ***REMOVED***

    if (node.update) ***REMOVED***
        cb(node.update, node, state);
  ***REMOVED***

    cb(node.body, node, state);
};

walkers[Syntax.FunctionDeclaration] = walkers[Syntax.ArrowFunctionExpression];

walkers[Syntax.FunctionExpression] = walkers[Syntax.ArrowFunctionExpression];

walkers[Syntax.Identifier] = leafNode;

walkers[Syntax.IfStatement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.test, node, state);
    cb(node.consequent, node, state);
    if (node.alternate) ***REMOVED***
        cb(node.alternate, node, state);
  ***REMOVED***
};

walkers[Syntax.Import] = leafNode;

walkers[Syntax.ImportDeclaration] = (node, parent, state, cb) => ***REMOVED***
    if (node.specifiers) ***REMOVED***
        for (let specifier of node.specifiers) ***REMOVED***
            cb(specifier, node, state);
      ***REMOVED***
  ***REMOVED***

    if (node.source) ***REMOVED***
        cb(node.source, node, state);
  ***REMOVED***
};

walkers[Syntax.ImportDefaultSpecifier] = (node, parent, state, cb) => ***REMOVED***
    if (node.local) ***REMOVED***
        cb(node.local, node, state);
  ***REMOVED***
};

walkers[Syntax.ImportNamespaceSpecifier] = walkers[Syntax.ImportDefaultSpecifier];

walkers[Syntax.ImportSpecifier] = walkers[Syntax.ExportSpecifier];

walkers[Syntax.JSXAttribute] = (node, parent, state, cb) => ***REMOVED***
    cb(node.name, node, state);

    if (node.value) ***REMOVED***
        cb(node.value, node, state);
  ***REMOVED***
};

walkers[Syntax.JSXClosingElement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.name, node, state);
};

walkers[Syntax.JSXElement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.openingElement, node, state);

    if (node.closingElement) ***REMOVED***
        cb(node.closingElement, node, state);
  ***REMOVED***

    for (let child of node.children) ***REMOVED***
        cb(child, node, state);
  ***REMOVED***
};

walkers[Syntax.JSXEmptyExpression] = leafNode;

walkers[Syntax.JSXExpressionContainer] = (node, parent, state, cb) => ***REMOVED***
    cb(node.expression, node, state);
};

walkers[Syntax.JSXIdentifier] = leafNode;

walkers[Syntax.JSXMemberExpression] = (node, parent, state, cb) => ***REMOVED***
    cb(node.object, node, state);

    cb(node.property, node, state);
};

walkers[Syntax.JSXNamespacedName] = (node, parent, state, cb) => ***REMOVED***
    cb(node.namespace, node, state);

    cb(node.name, node, state);
};

walkers[Syntax.JSXOpeningElement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.name, node, state);

    for (let attribute of node.attributes) ***REMOVED***
        cb(attribute, node, state);
  ***REMOVED***
};

walkers[Syntax.JSXSpreadAttribute] = (node, parent, state, cb) => ***REMOVED***
    cb(node.argument, node, state);
};

walkers[Syntax.JSXText] = leafNode;

walkers[Syntax.LabeledStatement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.body, node, state);
};

// TODO: add scope info??
walkers[Syntax.LetStatement] = (node, parent, state, cb) => ***REMOVED***
    for (let headItem of node.head) ***REMOVED***
        cb(headItem.id, node, state);
        if (headItem.init) ***REMOVED***
            cb(headItem.init, node, state);
      ***REMOVED***
  ***REMOVED***

    cb(node.body, node, state);
};

walkers[Syntax.Literal] = leafNode;

walkers[Syntax.LogicalExpression] = walkers[Syntax.BinaryExpression];

walkers[Syntax.MemberExpression] = (node, parent, state, cb) => ***REMOVED***
    cb(node.object, node, state);
    if (node.property) ***REMOVED***
        cb(node.property, node, state);
  ***REMOVED***
};

walkers[Syntax.MetaProperty] = leafNode;

walkers[Syntax.MethodDefinition] = (node, parent, state, cb) => ***REMOVED***
    if (node.key) ***REMOVED***
        cb(node.key, node, state);
  ***REMOVED***

    if (node.value) ***REMOVED***
        cb(node.value, node, state);
  ***REMOVED***

    if (node.decorators) ***REMOVED***
        for (let decorator of node.decorators) ***REMOVED***
            cb(decorator, node, state);
      ***REMOVED***
  ***REMOVED***
};

walkers[Syntax.ModuleDeclaration] = (node, parent, state, cb) => ***REMOVED***
    if (node.id) ***REMOVED***
        cb(node.id, node, state);
  ***REMOVED***

    if (node.source) ***REMOVED***
        cb(node.source, node, state);
  ***REMOVED***

    if (node.body) ***REMOVED***
        cb(node.body, node, state);
  ***REMOVED***
};

walkers[Syntax.NewExpression] = walkers[Syntax.CallExpression];

walkers[Syntax.ObjectExpression] = (node, parent, state, cb) => ***REMOVED***
    for (let property of node.properties) ***REMOVED***
        cb(property, node, state);
  ***REMOVED***
};

walkers[Syntax.ObjectPattern] = walkers[Syntax.ObjectExpression];

walkers[Syntax.PrivateName] = (node, parent, state, cb) => ***REMOVED***
    cb(node.id, node, state);
};

walkers[Syntax.Program] = (node, parent, state, cb) => ***REMOVED***
    // if the first item in the body has multiple leading comments, move all but the last one to
    // this node. this happens, for example, when a file has a /** @module */ standalone comment
    // followed by one or more other comments.
    if (node.body[0] && node.body[0].leadingComments && node.body[0].leadingComments.length > 1) ***REMOVED***
        moveLeadingComments(node.body[0], node, node.body[0].leadingComments.length - 1);
  ***REMOVED***

    // if the last item in the body has trailing comments, move them to this node
    if (node.body.length && node.body[node.body.length - 1].trailingComments) ***REMOVED***
        moveTrailingComments(node.body[node.body.length - 1], node);
  ***REMOVED***

    for (let bodyItem of node.body) ***REMOVED***
        cb(bodyItem, node, state);
  ***REMOVED***
};

walkers[Syntax.Property] = (node, parent, state, cb) => ***REMOVED***
    // move leading comments from key to property node
    moveLeadingComments(node.key, node);

    if (node.value) ***REMOVED***
        cb(node.value, node, state);
  ***REMOVED***

    if (node.decorators) ***REMOVED***
        for (let decorator of node.decorators) ***REMOVED***
            cb(decorator, node, state);
      ***REMOVED***
  ***REMOVED***
};

walkers[Syntax.ClassPrivateProperty] = (node, parent, state, cb) => ***REMOVED***
    // move leading comments from key to property node
    moveLeadingComments(node.key, node);

    // add `name` property to key, so we don't have to give this type of node special treatment
    // when we resolve its name
    node.key.name = node.key.id.name;

    if (node.value) ***REMOVED***
        cb(node.value, node, state);
  ***REMOVED***

    if (node.decorators) ***REMOVED***
        for (let decorator of node.decorators) ***REMOVED***
            cb(decorator, node, state);
      ***REMOVED***
  ***REMOVED***
};

walkers[Syntax.ClassProperty] = walkers[Syntax.Property];

walkers[Syntax.RestElement] = (node, parent, state, cb) => ***REMOVED***
    if (node.argument) ***REMOVED***
        cb(node.argument, node, state);
  ***REMOVED***
};

walkers[Syntax.ReturnStatement] = (node, parent, state, cb) => ***REMOVED***
    if (node.argument) ***REMOVED***
        cb(node.argument, node, state);
  ***REMOVED***
};

walkers[Syntax.SequenceExpression] = (node, parent, state, cb) => ***REMOVED***
    for (let expression of node.expressions) ***REMOVED***
        cb(expression, node, state);
  ***REMOVED***
};

walkers[Syntax.SpreadElement] = (node, parent, state, cb) => ***REMOVED***
    if (node.argument) ***REMOVED***
        cb(node.argument, node, state);
  ***REMOVED***
};

walkers[Syntax.Super] = leafNode;

walkers[Syntax.SwitchCase] = (node, parent, state, cb) => ***REMOVED***
    if (node.test) ***REMOVED***
        cb(node.test, node, state);
  ***REMOVED***

    for (let consequentItem of node.consequent) ***REMOVED***
        cb(consequentItem, node, state);
  ***REMOVED***
};

walkers[Syntax.SwitchStatement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.discriminant, node, state);

    for (let caseItem of node.cases) ***REMOVED***
        cb(caseItem, node, state);
  ***REMOVED***
};

walkers[Syntax.TaggedTemplateExpression] = (node, parent, state, cb) => ***REMOVED***
    if (node.tag) ***REMOVED***
        cb(node.tag, node, state);
  ***REMOVED***
    if (node.quasi) ***REMOVED***
        cb(node.quasi, node, state);
  ***REMOVED***
};

walkers[Syntax.TemplateElement] = leafNode;

walkers[Syntax.TemplateLiteral] = (node, parent, state, cb) => ***REMOVED***
    if (node.quasis && node.quasis.length) ***REMOVED***
        for (let quasi of node.quasis) ***REMOVED***
            cb(quasi, node, state);
      ***REMOVED***
  ***REMOVED***

    if (node.expressions && node.expressions.length) ***REMOVED***
        for (let expression of node.expressions) ***REMOVED***
            cb(expression, node, state);
      ***REMOVED***
  ***REMOVED***
};

walkers[Syntax.ThisExpression] = leafNode;

walkers[Syntax.ThrowStatement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.argument, node, state);
};

walkers[Syntax.TryStatement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.block, node, state);

    if (node.handler) ***REMOVED***
        cb(node.handler.body, node, state);
  ***REMOVED***

    if (node.finalizer) ***REMOVED***
        cb(node.finalizer, node, state);
  ***REMOVED***
};

walkers[Syntax.UnaryExpression] = (node, parent, state, cb) => ***REMOVED***
    cb(node.argument, node, state);
};

walkers[Syntax.UpdateExpression] = walkers[Syntax.UnaryExpression];

walkers[Syntax.VariableDeclaration] = (node, parent, state, cb) => ***REMOVED***
    // move leading comments to first declarator
    moveLeadingComments(node, node.declarations[0]);

    for (let declaration of node.declarations) ***REMOVED***
        cb(declaration, node, state);
  ***REMOVED***
};

walkers[Syntax.VariableDeclarator] = (node, parent, state, cb) => ***REMOVED***
    cb(node.id, node, state);

    if (node.init) ***REMOVED***
        cb(node.init, node, state);
  ***REMOVED***
};

walkers[Syntax.WhileStatement] = walkers[Syntax.DoWhileStatement];

walkers[Syntax.WithStatement] = (node, parent, state, cb) => ***REMOVED***
    cb(node.object, node, state);
    cb(node.body, node, state);
};

walkers[Syntax.YieldExpression] = (node, parent, state, cb) => ***REMOVED***
    if (node.argument) ***REMOVED***
        cb(node.argument, node, state);
  ***REMOVED***
};

/**
 * Create a walker that can traverse an AST that is consistent with the Mozilla Parser API.
 *
 * @memberof module:jsdoc/src/walker
 */
class Walker ***REMOVED***
    // TODO: docs
    constructor(walkerFuncs = walkers) ***REMOVED***
        this._walkers = walkerFuncs;
  ***REMOVED***

    // TODO: docs
    _recurse(filename, ast) ***REMOVED***
        const self = this;
        const state = ***REMOVED***
            filename: filename,
            nodes: [],
            scopes: []
      ***REMOVED***;

        function logUnknownNodeType(***REMOVED***type}) ***REMOVED***
            logger.debug('Found a node with unrecognized type %s. Ignoring the node and its ' +
                'descendants.', type);
      ***REMOVED***

        function cb(node, parent, cbState) ***REMOVED***
            let currentScope;

            const isScope = astnode.isScope(node);

            astnode.addNodeProperties(node);
            node.parent = parent || null;

            currentScope = getCurrentScope(cbState.scopes);
            if (currentScope) ***REMOVED***
                node.enclosingScope = currentScope;
          ***REMOVED***

            if (isScope) ***REMOVED***
                cbState.scopes.push(node);
          ***REMOVED***
            cbState.nodes.push(node);

            if (!self._walkers[node.type]) ***REMOVED***
                logUnknownNodeType(node);
          ***REMOVED*** else ***REMOVED***
                self._walkers[node.type](node, parent, cbState, cb);
          ***REMOVED***

            if (isScope) ***REMOVED***
                cbState.scopes.pop();
          ***REMOVED***
      ***REMOVED***

        cb(ast, null, state);

        return state;
  ***REMOVED***

    // TODO: docs
    recurse(ast, visitor, filename) ***REMOVED***
        let shouldContinue;
        const state = this._recurse(filename, ast);

        if (visitor) ***REMOVED***
            for (let node of state.nodes) ***REMOVED***
                shouldContinue = visitor.visit(node, filename);
                if (!shouldContinue) ***REMOVED***
                    break;
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***

        return ast;
  ***REMOVED***
}
exports.Walker = Walker;
