// @remix-run/router/history.ts
////////////////////////////////////////////////////////////////////////////////
//#region Types and Constants
////////////////////////////////////////////////////////////////////////////////

/**
 * Actions represent the type of change to a location value.
 */
export enum Action ***REMOVED***
  /**
   * A POP indicates a change to an arbitrary index in the history stack, such
   * as a back or forward navigation. It does not describe the direction of the
   * navigation, only that the current index changed.
   *
   * Note: This is the default action for newly created history objects.
   */
  Pop = "POP",

  /**
   * A PUSH indicates a new entry being added to the history stack, such as when
   * a link is clicked and a new page loads. When this happens, all subsequent
   * entries in the stack are lost.
   */
  Push = "PUSH",

  /**
   * A REPLACE indicates the entry at the current index in the history stack
   * being replaced by a new one.
   */
  Replace = "REPLACE",
}

/**
 * The pathname, search, and hash values of a URL.
 */
export interface Path ***REMOVED***
  /**
   * A URL pathname, beginning with a /.
   */
  pathname: string;

  /**
   * A URL search string, beginning with a ?.
   */
  search: string;

  /**
   * A URL fragment identifier, beginning with a #.
   */
  hash: string;
}

/**
 * An entry in a history stack. A location contains information about the
 * URL path, as well as possibly some arbitrary state and a key.
 */
export interface Location extends Path ***REMOVED***
  /**
   * A value of arbitrary data associated with this location.
   */
  state: any;

  /**
   * A unique string associated with this location. May be used to safely store
   * and retrieve data in some other storage API, like `localStorage`.
   *
   * Note: This value is always "default" on the initial location.
   */
  key: string;
}

/**
 * A change to the current location.
 */
export interface Update ***REMOVED***
  /**
   * The action that triggered the change.
   */
  action: Action;

  /**
   * The new location.
   */
  location: Location;

  /**
   * The delta between this location and the former location in the history stack
   */
  delta: number;
}

/**
 * A function that receives notifications about location changes.
 */
export interface Listener ***REMOVED***
  (update: Update): void;
}

/**
 * Describes a location that is the destination of some navigation, either via
 * `history.push` or `history.replace`. May be either a URL or the pieces of a
 * URL path.
 */
export type To = string | Partial<Path>;

/**
 * A history is an interface to the navigation stack. The history serves as the
 * source of truth for the current location, as well as provides a set of
 * methods that may be used to change it.
 *
 * It is similar to the DOM's `window.history` object, but with a smaller, more
 * focused API.
 */
export interface History ***REMOVED***
  /**
   * The last action that modified the current location. This will always be
   * Action.Pop when a history instance is first created. This value is mutable.
   */
  readonly action: Action;

  /**
   * The current location. This value is mutable.
   */
  readonly location: Location;

  /**
   * Returns a valid href for the given `to` value that may be used as
   * the value of an <a href> attribute.
   *
   * @param to - The destination URL
   */
  createHref(to: To): string;

  /**
   * Returns a URL for the given `to` value
   *
   * @param to - The destination URL
   */
  createURL(to: To): URL;

  /**
   * Encode a location the same way window.history would do (no-op for memory
   * history) so we ensure our PUSH/REPLACE navigations for data routers
   * behave the same as POP
   *
   * @param to Unencoded path
   */
  encodeLocation(to: To): Path;

  /**
   * Pushes a new location onto the history stack, increasing its length by one.
   * If there were any entries in the stack after the current one, they are
   * lost.
   *
   * @param to - The new URL
   * @param state - Data to associate with the new location
   */
  push(to: To, state?: any): void;

  /**
   * Replaces the current location in the history stack with a new one.  The
   * location that was replaced will no longer be available.
   *
   * @param to - The new URL
   * @param state - Data to associate with the new location
   */
  replace(to: To, state?: any): void;

  /**
   * Navigates `n` entries backward/forward in the history stack relative to the
   * current index. For example, a "back" navigation would use go(-1).
   *
   * @param delta - The delta in the stack index
   */
  go(delta: number): void;

  /**
   * Sets up a listener that will be called whenever the current location
   * changes.
   *
   * @param listener - A function that will be called when the location changes
   * @returns unlisten - A function that may be used to stop listening
   */
  listen(listener: Listener): () => void;
}

type HistoryState = ***REMOVED***
  usr: any;
  key?: string;
  idx: number;
};

const PopStateEventType = "popstate";
//#endregion

////////////////////////////////////////////////////////////////////////////////
//#region Memory History
////////////////////////////////////////////////////////////////////////////////

/**
 * A user-supplied object that describes a location. Used when providing
 * entries to `createMemoryHistory` via its `initialEntries` option.
 */
export type InitialEntry = string | Partial<Location>;

export type MemoryHistoryOptions = ***REMOVED***
  initialEntries?: InitialEntry[];
  initialIndex?: number;
  v5Compat?: boolean;
};

/**
 * A memory history stores locations in memory. This is useful in stateful
 * environments where there is no web browser, such as node tests or React
 * Native.
 */
export interface MemoryHistory extends History ***REMOVED***
  /**
   * The current index in the history stack.
   */
  readonly index: number;
}

/**
 * Memory history stores the current location in memory. It is designed for use
 * in stateful non-browser environments like tests and React Native.
 */
export function createMemoryHistory(
  options: MemoryHistoryOptions = ***REMOVED***}
): MemoryHistory ***REMOVED***
  let ***REMOVED*** initialEntries = ["/"], initialIndex, v5Compat = false } = options;
  let entries: Location[]; // Declare so we can access from createMemoryLocation
  entries = initialEntries.map((entry, index) =>
    createMemoryLocation(
      entry,
      typeof entry === "string" ? null : entry.state,
      index === 0 ? "default" : undefined
    )
  );
  let index = clampIndex(
    initialIndex == null ? entries.length - 1 : initialIndex
  );
  let action = Action.Pop;
  let listener: Listener | null = null;

  function clampIndex(n: number): number ***REMOVED***
    return Math.min(Math.max(n, 0), entries.length - 1);
***REMOVED***
  function getCurrentLocation(): Location ***REMOVED***
    return entries[index];
***REMOVED***
  function createMemoryLocation(
    to: To,
    state: any = null,
    key?: string
  ): Location ***REMOVED***
    let location = createLocation(
      entries ? getCurrentLocation().pathname : "/",
      to,
      state,
      key
    );
    warning(
      location.pathname.charAt(0) === "/",
      `relative pathnames are not supported in memory history: $***REMOVED***JSON.stringify(
        to
      )}`
    );
    return location;
***REMOVED***

  function createHref(to: To) ***REMOVED***
    return typeof to === "string" ? to : createPath(to);
***REMOVED***

  let history: MemoryHistory = ***REMOVED***
    get index() ***REMOVED***
      return index;
  ***REMOVED***,
    get action() ***REMOVED***
      return action;
  ***REMOVED***,
    get location() ***REMOVED***
      return getCurrentLocation();
  ***REMOVED***,
    createHref,
    createURL(to) ***REMOVED***
      return new URL(createHref(to), "http://localhost");
  ***REMOVED***,
    encodeLocation(to: To) ***REMOVED***
      let path = typeof to === "string" ? parsePath(to) : to;
      return ***REMOVED***
        pathname: path.pathname || "",
        search: path.search || "",
        hash: path.hash || "",
    ***REMOVED***;
  ***REMOVED***,
    push(to, state) ***REMOVED***
      action = Action.Push;
      let nextLocation = createMemoryLocation(to, state);
      index += 1;
      entries.splice(index, entries.length, nextLocation);
      if (v5Compat && listener) ***REMOVED***
        listener(***REMOVED*** action, location: nextLocation, delta: 1 });
    ***REMOVED***
  ***REMOVED***,
    replace(to, state) ***REMOVED***
      action = Action.Replace;
      let nextLocation = createMemoryLocation(to, state);
      entries[index] = nextLocation;
      if (v5Compat && listener) ***REMOVED***
        listener(***REMOVED*** action, location: nextLocation, delta: 0 });
    ***REMOVED***
  ***REMOVED***,
    go(delta) ***REMOVED***
      action = Action.Pop;
      let nextIndex = clampIndex(index + delta);
      let nextLocation = entries[nextIndex];
      index = nextIndex;
      if (listener) ***REMOVED***
        listener(***REMOVED*** action, location: nextLocation, delta });
    ***REMOVED***
  ***REMOVED***,
    listen(fn: Listener) ***REMOVED***
      listener = fn;
      return () => ***REMOVED***
        listener = null;
    ***REMOVED***;
  ***REMOVED***,
***REMOVED***;

  return history;
}
//#endregion

////////////////////////////////////////////////////////////////////////////////
//#region Browser History
////////////////////////////////////////////////////////////////////////////////

/**
 * A browser history stores the current location in regular URLs in a web
 * browser environment. This is the standard for most web apps and provides the
 * cleanest URLs the browser's address bar.
 *
 * @see https://github.com/remix-run/history/tree/main/docs/api-reference.md#browserhistory
 */
export interface BrowserHistory extends UrlHistory ***REMOVED***}

export type BrowserHistoryOptions = UrlHistoryOptions;

/**
 * Browser history stores the location in regular URLs. This is the standard for
 * most web apps, but it requires some configuration on the server to ensure you
 * serve the same app at multiple URLs.
 *
 * @see https://github.com/remix-run/history/tree/main/docs/api-reference.md#createbrowserhistory
 */
export function createBrowserHistory(
  options: BrowserHistoryOptions = ***REMOVED***}
): BrowserHistory ***REMOVED***
  function createBrowserLocation(
    window: Window,
    globalHistory: Window["history"]
  ) ***REMOVED***
    let ***REMOVED*** pathname, search, hash } = window.location;
    return createLocation(
      "",
      ***REMOVED*** pathname, search, hash },
      // state defaults to `null` because `window.history.state` does
      (globalHistory.state && globalHistory.state.usr) || null,
      (globalHistory.state && globalHistory.state.key) || "default"
    );
***REMOVED***

  function createBrowserHref(window: Window, to: To) ***REMOVED***
    return typeof to === "string" ? to : createPath(to);
***REMOVED***

  return getUrlBasedHistory(
    createBrowserLocation,
    createBrowserHref,
    null,
    options
  );
}
//#endregion

////////////////////////////////////////////////////////////////////////////////
//#region Hash History
////////////////////////////////////////////////////////////////////////////////

/**
 * A hash history stores the current location in the fragment identifier portion
 * of the URL in a web browser environment.
 *
 * This is ideal for apps that do not control the server for some reason
 * (because the fragment identifier is never sent to the server), including some
 * shared hosting environments that do not provide fine-grained controls over
 * which pages are served at which URLs.
 *
 * @see https://github.com/remix-run/history/tree/main/docs/api-reference.md#hashhistory
 */
export interface HashHistory extends UrlHistory ***REMOVED***}

export type HashHistoryOptions = UrlHistoryOptions;

/**
 * Hash history stores the location in window.location.hash. This makes it ideal
 * for situations where you don't want to send the location to the server for
 * some reason, either because you do cannot configure it or the URL space is
 * reserved for something else.
 *
 * @see https://github.com/remix-run/history/tree/main/docs/api-reference.md#createhashhistory
 */
export function createHashHistory(
  options: HashHistoryOptions = ***REMOVED***}
): HashHistory ***REMOVED***
  function createHashLocation(
    window: Window,
    globalHistory: Window["history"]
  ) ***REMOVED***
    let ***REMOVED***
      pathname = "/",
      search = "",
      hash = "",
  ***REMOVED*** = parsePath(window.location.hash.substr(1));
    return createLocation(
      "",
      ***REMOVED*** pathname, search, hash },
      // state defaults to `null` because `window.history.state` does
      (globalHistory.state && globalHistory.state.usr) || null,
      (globalHistory.state && globalHistory.state.key) || "default"
    );
***REMOVED***

  function createHashHref(window: Window, to: To) ***REMOVED***
    let base = window.document.querySelector("base");
    let href = "";

    if (base && base.getAttribute("href")) ***REMOVED***
      let url = window.location.href;
      let hashIndex = url.indexOf("#");
      href = hashIndex === -1 ? url : url.slice(0, hashIndex);
  ***REMOVED***

    return href + "#" + (typeof to === "string" ? to : createPath(to));
***REMOVED***

  function validateHashLocation(location: Location, to: To) ***REMOVED***
    warning(
      location.pathname.charAt(0) === "/",
      `relative pathnames are not supported in hash history.push($***REMOVED***JSON.stringify(
        to
      )})`
    );
***REMOVED***

  return getUrlBasedHistory(
    createHashLocation,
    createHashHref,
    validateHashLocation,
    options
  );
}
//#endregion

////////////////////////////////////////////////////////////////////////////////
//#region UTILS
////////////////////////////////////////////////////////////////////////////////

/**
 * @private
 */
export function invariant(value: boolean, message?: string): asserts value;
export function invariant<T>(
  value: T | null | undefined,
  message?: string
): asserts value is T;
export function invariant(value: any, message?: string) ***REMOVED***
  if (value === false || value === null || typeof value === "undefined") ***REMOVED***
    throw new Error(message);
***REMOVED***
}

function warning(cond: any, message: string) ***REMOVED***
  if (!cond) ***REMOVED***
    // eslint-disable-next-line no-console
    if (typeof console !== "undefined") console.warn(message);

    try ***REMOVED***
      // Welcome to debugging history!
      //
      // This error is thrown as a convenience so you can more easily
      // find the source for a warning that appears in the console by
      // enabling "pause on exceptions" in your JavaScript debugger.
      throw new Error(message);
      // eslint-disable-next-line no-empty
  ***REMOVED*** catch (e) ***REMOVED***}
***REMOVED***
}

function createKey() ***REMOVED***
  return Math.random().toString(36).substr(2, 8);
}

/**
 * For browser-based histories, we combine the state and key into an object
 */
function getHistoryState(location: Location, index: number): HistoryState ***REMOVED***
  return ***REMOVED***
    usr: location.state,
    key: location.key,
    idx: index,
***REMOVED***;
}

/**
 * Creates a Location object with a unique key from the given Path
 */
export function createLocation(
  current: string | Location,
  to: To,
  state: any = null,
  key?: string
): Readonly<Location> ***REMOVED***
  let location: Readonly<Location> = ***REMOVED***
    pathname: typeof current === "string" ? current : current.pathname,
    search: "",
    hash: "",
    ...(typeof to === "string" ? parsePath(to) : to),
    state,
    // TODO: This could be cleaned up.  push/replace should probably just take
    // full Locations now and avoid the need to run through this flow at all
    // But that's a pretty big refactor to the current test suite so going to
    // keep as is for the time being and just let any incoming keys take precedence
    key: (to && (to as Location).key) || key || createKey(),
***REMOVED***;
  return location;
}

/**
 * Creates a string URL path from the given pathname, search, and hash components.
 */
export function createPath(***REMOVED***
  pathname = "/",
  search = "",
  hash = "",
}: Partial<Path>) ***REMOVED***
  if (search && search !== "?")
    pathname += search.charAt(0) === "?" ? search : "?" + search;
  if (hash && hash !== "#")
    pathname += hash.charAt(0) === "#" ? hash : "#" + hash;
  return pathname;
}

/**
 * Parses a string URL path into its separate pathname, search, and hash components.
 */
export function parsePath(path: string): Partial<Path> ***REMOVED***
  let parsedPath: Partial<Path> = ***REMOVED***};

  if (path) ***REMOVED***
    let hashIndex = path.indexOf("#");
    if (hashIndex >= 0) ***REMOVED***
      parsedPath.hash = path.substr(hashIndex);
      path = path.substr(0, hashIndex);
  ***REMOVED***

    let searchIndex = path.indexOf("?");
    if (searchIndex >= 0) ***REMOVED***
      parsedPath.search = path.substr(searchIndex);
      path = path.substr(0, searchIndex);
  ***REMOVED***

    if (path) ***REMOVED***
      parsedPath.pathname = path;
  ***REMOVED***
***REMOVED***

  return parsedPath;
}

export interface UrlHistory extends History ***REMOVED***}

export type UrlHistoryOptions = ***REMOVED***
  window?: Window;
  v5Compat?: boolean;
};

function getUrlBasedHistory(
  getLocation: (window: Window, globalHistory: Window["history"]) => Location,
  createHref: (window: Window, to: To) => string,
  validateLocation: ((location: Location, to: To) => void) | null,
  options: UrlHistoryOptions = ***REMOVED***}
): UrlHistory ***REMOVED***
  let ***REMOVED*** window = document.defaultView!, v5Compat = false } = options;
  let globalHistory = window.history;
  let action = Action.Pop;
  let listener: Listener | null = null;

  let index = getIndex()!;
  // Index should only be null when we initialize. If not, it's because the
  // user called history.pushState or history.replaceState directly, in which
  // case we should log a warning as it will result in bugs.
  if (index == null) ***REMOVED***
    index = 0;
    globalHistory.replaceState(***REMOVED*** ...globalHistory.state, idx: index }, "");
***REMOVED***

  function getIndex(): number ***REMOVED***
    let state = globalHistory.state || ***REMOVED*** idx: null };
    return state.idx;
***REMOVED***

  function handlePop() ***REMOVED***
    let nextAction = Action.Pop;
    let nextIndex = getIndex();

    if (nextIndex != null) ***REMOVED***
      let delta = nextIndex - index;
      action = nextAction;
      index = nextIndex;
      if (listener) ***REMOVED***
        listener(***REMOVED*** action, location: history.location, delta });
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
      warning(
        false,
        // TODO: Write up a doc that explains our blocking strategy in detail
        // and link to it here so people can understand better what is going on
        // and how to avoid it.
        `You are trying to block a POP navigation to a location that was not ` +
          `created by @remix-run/router. The block will fail silently in ` +
          `production, but in general you should do all navigation with the ` +
          `router (instead of using window.history.pushState directly) ` +
          `to avoid this situation.`
      );
  ***REMOVED***
***REMOVED***

  function push(to: To, state?: any) ***REMOVED***
    action = Action.Push;
    let location = createLocation(history.location, to, state);
    if (validateLocation) validateLocation(location, to);

    index = getIndex() + 1;
    let historyState = getHistoryState(location, index);
    let url = history.createHref(location);

    // try...catch because iOS limits us to 100 pushState calls :/
    try ***REMOVED***
      globalHistory.pushState(historyState, "", url);
  ***REMOVED*** catch (error) ***REMOVED***
      // They are going to lose state here, but there is no real
      // way to warn them about it since the page will refresh...
      window.location.assign(url);
  ***REMOVED***

    if (v5Compat && listener) ***REMOVED***
      listener(***REMOVED*** action, location: history.location, delta: 1 });
  ***REMOVED***
***REMOVED***

  function replace(to: To, state?: any) ***REMOVED***
    action = Action.Replace;
    let location = createLocation(history.location, to, state);
    if (validateLocation) validateLocation(location, to);

    index = getIndex();
    let historyState = getHistoryState(location, index);
    let url = history.createHref(location);
    globalHistory.replaceState(historyState, "", url);

    if (v5Compat && listener) ***REMOVED***
      listener(***REMOVED*** action, location: history.location, delta: 0 });
  ***REMOVED***
***REMOVED***

  function createURL(to: To): URL ***REMOVED***
    // window.location.origin is "null" (the literal string value) in Firefox
    // under certain conditions, notably when serving from a local HTML file
    // See https://bugzilla.mozilla.org/show_bug.cgi?id=878297
    let base =
      window.location.origin !== "null"
        ? window.location.origin
        : window.location.href;

    let href = typeof to === "string" ? to : createPath(to);
    invariant(
      base,
      `No window.location.(origin|href) available to create URL for href: $***REMOVED***href}`
    );
    return new URL(href, base);
***REMOVED***

  let history: History = ***REMOVED***
    get action() ***REMOVED***
      return action;
  ***REMOVED***,
    get location() ***REMOVED***
      return getLocation(window, globalHistory);
  ***REMOVED***,
    listen(fn: Listener) ***REMOVED***
      if (listener) ***REMOVED***
        throw new Error("A history only accepts one active listener");
    ***REMOVED***
      window.addEventListener(PopStateEventType, handlePop);
      listener = fn;

      return () => ***REMOVED***
        window.removeEventListener(PopStateEventType, handlePop);
        listener = null;
    ***REMOVED***;
  ***REMOVED***,
    createHref(to) ***REMOVED***
      return createHref(window, to);
  ***REMOVED***,
    createURL,
    encodeLocation(to) ***REMOVED***
      // Encode a Location the same way window.location would
      let url = createURL(to);
      return ***REMOVED***
        pathname: url.pathname,
        search: url.search,
        hash: url.hash,
    ***REMOVED***;
  ***REMOVED***,
    push,
    replace,
    go(n) ***REMOVED***
      return globalHistory.go(n);
  ***REMOVED***,
***REMOVED***;

  return history;
}

//#endregion
