// @remix-run/router/utils.ts
import type ***REMOVED*** Location, Path, To } from "./history";
import ***REMOVED*** invariant, parsePath } from "./history";

/**
 * Map of routeId -> data returned from a loader/action/error
 */
export interface RouteData ***REMOVED***
  [routeId: string]: any;
}

export enum ResultType ***REMOVED***
  data = "data",
  deferred = "deferred",
  redirect = "redirect",
  error = "error",
}

/**
 * Successful result from a loader or action
 */
export interface SuccessResult ***REMOVED***
  type: ResultType.data;
  data: any;
  statusCode?: number;
  headers?: Headers;
}

/**
 * Successful defer() result from a loader or action
 */
export interface DeferredResult ***REMOVED***
  type: ResultType.deferred;
  deferredData: DeferredData;
  statusCode?: number;
  headers?: Headers;
}

/**
 * Redirect result from a loader or action
 */
export interface RedirectResult ***REMOVED***
  type: ResultType.redirect;
  status: number;
  location: string;
  revalidate: boolean;
}

/**
 * Unsuccessful result from a loader or action
 */
export interface ErrorResult ***REMOVED***
  type: ResultType.error;
  error: any;
  headers?: Headers;
}

/**
 * Result from a loader or action - potentially successful or unsuccessful
 */
export type DataResult =
  | SuccessResult
  | DeferredResult
  | RedirectResult
  | ErrorResult;

export type MutationFormMethod = "post" | "put" | "patch" | "delete";
export type FormMethod = "get" | MutationFormMethod;

export type FormEncType =
  | "application/x-www-form-urlencoded"
  | "multipart/form-data";

/**
 * @private
 * Internal interface to pass around for action submissions, not intended for
 * external consumption
 */
export interface Submission ***REMOVED***
  formMethod: FormMethod;
  formAction: string;
  formEncType: FormEncType;
  formData: FormData;
}

/**
 * @private
 * Arguments passed to route loader/action functions.  Same for now but we keep
 * this as a private implementation detail in case they diverge in the future.
 */
interface DataFunctionArgs ***REMOVED***
  request: Request;
  params: Params;
  context?: any;
}

/**
 * Arguments passed to loader functions
 */
export interface LoaderFunctionArgs extends DataFunctionArgs ***REMOVED***}

/**
 * Arguments passed to action functions
 */
export interface ActionFunctionArgs extends DataFunctionArgs ***REMOVED***}

/**
 * Route loader function signature
 */
export interface LoaderFunction ***REMOVED***
  (args: LoaderFunctionArgs): Promise<Response> | Response | Promise<any> | any;
}

/**
 * Route action function signature
 */
export interface ActionFunction ***REMOVED***
  (args: ActionFunctionArgs): Promise<Response> | Response | Promise<any> | any;
}

/**
 * Route shouldRevalidate function signature.  This runs after any submission
 * (navigation or fetcher), so we flatten the navigation/fetcher submission
 * onto the arguments.  It shouldn't matter whether it came from a navigation
 * or a fetcher, what really matters is the URLs and the formData since loaders
 * have to re-run based on the data models that were potentially mutated.
 */
export interface ShouldRevalidateFunction ***REMOVED***
  (args: ***REMOVED***
    currentUrl: URL;
    currentParams: AgnosticDataRouteMatch["params"];
    nextUrl: URL;
    nextParams: AgnosticDataRouteMatch["params"];
    formMethod?: Submission["formMethod"];
    formAction?: Submission["formAction"];
    formEncType?: Submission["formEncType"];
    formData?: Submission["formData"];
    actionResult?: DataResult;
    defaultShouldRevalidate: boolean;
***REMOVED***): boolean;
}

/**
 * Base RouteObject with common props shared by all types of routes
 */
type AgnosticBaseRouteObject = ***REMOVED***
  caseSensitive?: boolean;
  path?: string;
  id?: string;
  loader?: LoaderFunction;
  action?: ActionFunction;
  hasErrorBoundary?: boolean;
  shouldRevalidate?: ShouldRevalidateFunction;
  handle?: any;
};

/**
 * Index routes must not have children
 */
export type AgnosticIndexRouteObject = AgnosticBaseRouteObject & ***REMOVED***
  children?: undefined;
  index: true;
};

/**
 * Non-index routes may have children, but cannot have index
 */
export type AgnosticNonIndexRouteObject = AgnosticBaseRouteObject & ***REMOVED***
  children?: AgnosticRouteObject[];
  index?: false;
};

/**
 * A route object represents a logical route, with (optionally) its child
 * routes organized in a tree-like structure.
 */
export type AgnosticRouteObject =
  | AgnosticIndexRouteObject
  | AgnosticNonIndexRouteObject;

export type AgnosticDataIndexRouteObject = AgnosticIndexRouteObject & ***REMOVED***
  id: string;
};

export type AgnosticDataNonIndexRouteObject = AgnosticNonIndexRouteObject & ***REMOVED***
  children?: AgnosticDataRouteObject[];
  id: string;
};

/**
 * A data route object, which is just a RouteObject with a required unique ID
 */
export type AgnosticDataRouteObject =
  | AgnosticDataIndexRouteObject
  | AgnosticDataNonIndexRouteObject;

// Recursive helper for finding path parameters in the absence of wildcards
type _PathParam<Path extends string> =
  // split path into individual path segments
  Path extends `$***REMOVED***infer L}/$***REMOVED***infer R}`
    ? _PathParam<L> | _PathParam<R>
    : // find params after `:`
    Path extends `:$***REMOVED***infer Param}`
    ? Param extends `$***REMOVED***infer Optional}?`
      ? Optional
      : Param
    : // otherwise, there aren't any params present
      never;

/**
 * Examples:
 * "/a/b/*" -> "*"
 * ":a" -> "a"
 * "/a/:b" -> "b"
 * "/a/blahblahblah:b" -> "b"
 * "/:a/:b" -> "a" | "b"
 * "/:a/b/:c/*" -> "a" | "c" | "*"
 */
type PathParam<Path extends string> =
  // check if path is just a wildcard
  Path extends "*"
    ? "*"
    : // look for wildcard at the end of the path
    Path extends `$***REMOVED***infer Rest}/*`
    ? "*" | _PathParam<Rest>
    : // look for params in the absence of wildcards
      _PathParam<Path>;

// Attempt to parse the given string segment. If it fails, then just return the
// plain string type as a default fallback. Otherwise return the union of the
// parsed string literals that were referenced as dynamic segments in the route.
export type ParamParseKey<Segment extends string> =
  // if could not find path params, fallback to `string`
  [PathParam<Segment>] extends [never] ? string : PathParam<Segment>;

/**
 * The parameters that were parsed from the URL path.
 */
export type Params<Key extends string = string> = ***REMOVED***
  readonly [key in Key]: string | undefined;
};

/**
 * A RouteMatch contains info about how a route matched a URL.
 */
export interface AgnosticRouteMatch<
  ParamKey extends string = string,
  RouteObjectType extends AgnosticRouteObject = AgnosticRouteObject
> ***REMOVED***
  /**
   * The names and values of dynamic parameters in the URL.
   */
  params: Params<ParamKey>;
  /**
   * The portion of the URL pathname that was matched.
   */
  pathname: string;
  /**
   * The portion of the URL pathname that was matched before child routes.
   */
  pathnameBase: string;
  /**
   * The route object that was used to match.
   */
  route: RouteObjectType;
}

export interface AgnosticDataRouteMatch
  extends AgnosticRouteMatch<string, AgnosticDataRouteObject> ***REMOVED***}

function isIndexRoute(
  route: AgnosticRouteObject
): route is AgnosticIndexRouteObject ***REMOVED***
  return route.index === true;
}

// Walk the route tree generating unique IDs where necessary so we are working
// solely with AgnosticDataRouteObject's within the Router
export function convertRoutesToDataRoutes(
  routes: AgnosticRouteObject[],
  parentPath: number[] = [],
  allIds: Set<string> = new Set<string>()
): AgnosticDataRouteObject[] ***REMOVED***
  return routes.map((route, index) => ***REMOVED***
    let treePath = [...parentPath, index];
    let id = typeof route.id === "string" ? route.id : treePath.join("-");
    invariant(
      route.index !== true || !route.children,
      `Cannot specify children on an index route`
    );
    invariant(
      !allIds.has(id),
      `Found a route id collision on id "$***REMOVED***id}".  Route ` +
        "id's must be globally unique within Data Router usages"
    );
    allIds.add(id);

    if (isIndexRoute(route)) ***REMOVED***
      let indexRoute: AgnosticDataIndexRouteObject = ***REMOVED*** ...route, id };
      return indexRoute;
  ***REMOVED*** else ***REMOVED***
      let pathOrLayoutRoute: AgnosticDataNonIndexRouteObject = ***REMOVED***
        ...route,
        id,
        children: route.children
          ? convertRoutesToDataRoutes(route.children, treePath, allIds)
          : undefined,
    ***REMOVED***;
      return pathOrLayoutRoute;
  ***REMOVED***
***REMOVED***);
}

/**
 * Matches the given routes to a location and returns the match data.
 *
 * @see https://reactrouter.com/utils/match-routes
 */
export function matchRoutes<
  RouteObjectType extends AgnosticRouteObject = AgnosticRouteObject
>(
  routes: RouteObjectType[],
  locationArg: Partial<Location> | string,
  basename = "/"
): AgnosticRouteMatch<string, RouteObjectType>[] | null ***REMOVED***
  let location =
    typeof locationArg === "string" ? parsePath(locationArg) : locationArg;

  let pathname = stripBasename(location.pathname || "/", basename);

  if (pathname == null) ***REMOVED***
    return null;
***REMOVED***

  let branches = flattenRoutes(routes);
  rankRouteBranches(branches);

  let matches = null;
  for (let i = 0; matches == null && i < branches.length; ++i) ***REMOVED***
    matches = matchRouteBranch<string, RouteObjectType>(
      branches[i],
      // Incoming pathnames are generally encoded from either window.location
      // or from router.navigate, but we want to match against the unencoded
      // paths in the route definitions.  Memory router locations won't be
      // encoded here but there also shouldn't be anything to decode so this
      // should be a safe operation.  This avoids needing matchRoutes to be
      // history-aware.
      safelyDecodeURI(pathname)
    );
***REMOVED***

  return matches;
}

interface RouteMeta<
  RouteObjectType extends AgnosticRouteObject = AgnosticRouteObject
> ***REMOVED***
  relativePath: string;
  caseSensitive: boolean;
  childrenIndex: number;
  route: RouteObjectType;
}

interface RouteBranch<
  RouteObjectType extends AgnosticRouteObject = AgnosticRouteObject
> ***REMOVED***
  path: string;
  score: number;
  routesMeta: RouteMeta<RouteObjectType>[];
}

function flattenRoutes<
  RouteObjectType extends AgnosticRouteObject = AgnosticRouteObject
>(
  routes: RouteObjectType[],
  branches: RouteBranch<RouteObjectType>[] = [],
  parentsMeta: RouteMeta<RouteObjectType>[] = [],
  parentPath = ""
): RouteBranch<RouteObjectType>[] ***REMOVED***
  let flattenRoute = (
    route: RouteObjectType,
    index: number,
    relativePath?: string
  ) => ***REMOVED***
    let meta: RouteMeta<RouteObjectType> = ***REMOVED***
      relativePath:
        relativePath === undefined ? route.path || "" : relativePath,
      caseSensitive: route.caseSensitive === true,
      childrenIndex: index,
      route,
  ***REMOVED***;

    if (meta.relativePath.startsWith("/")) ***REMOVED***
      invariant(
        meta.relativePath.startsWith(parentPath),
        `Absolute route path "$***REMOVED***meta.relativePath}" nested under path ` +
          `"$***REMOVED***parentPath}" is not valid. An absolute child route path ` +
          `must start with the combined path of all its parent routes.`
      );

      meta.relativePath = meta.relativePath.slice(parentPath.length);
  ***REMOVED***

    let path = joinPaths([parentPath, meta.relativePath]);
    let routesMeta = parentsMeta.concat(meta);

    // Add the children before adding this route to the array so we traverse the
    // route tree depth-first and child routes appear before their parents in
    // the "flattened" version.
    if (route.children && route.children.length > 0) ***REMOVED***
      invariant(
        // Our types know better, but runtime JS may not!
        // @ts-expect-error
        route.index !== true,
        `Index routes must not have child routes. Please remove ` +
          `all child routes from route path "$***REMOVED***path}".`
      );

      flattenRoutes(route.children, branches, routesMeta, path);
  ***REMOVED***

    // Routes without a path shouldn't ever match by themselves unless they are
    // index routes, so don't add them to the list of possible branches.
    if (route.path == null && !route.index) ***REMOVED***
      return;
  ***REMOVED***

    branches.push(***REMOVED***
      path,
      score: computeScore(path, route.index),
      routesMeta,
  ***REMOVED***);
***REMOVED***;
  routes.forEach((route, index) => ***REMOVED***
    // coarse-grain check for optional params
    if (route.path === "" || !route.path?.includes("?")) ***REMOVED***
      flattenRoute(route, index);
  ***REMOVED*** else ***REMOVED***
      for (let exploded of explodeOptionalSegments(route.path)) ***REMOVED***
        flattenRoute(route, index, exploded);
    ***REMOVED***
  ***REMOVED***
***REMOVED***);

  return branches;
}

/**
 * Computes all combinations of optional path segments for a given path,
 * excluding combinations that are ambiguous and of lower priority.
 *
 * For example, `/one/:two?/three/:four?/:five?` explodes to:
 * - `/one/three`
 * - `/one/:two/three`
 * - `/one/three/:four`
 * - `/one/three/:five`
 * - `/one/:two/three/:four`
 * - `/one/:two/three/:five`
 * - `/one/three/:four/:five`
 * - `/one/:two/three/:four/:five`
 */
function explodeOptionalSegments(path: string): string[] ***REMOVED***
  let segments = path.split("/");
  if (segments.length === 0) return [];

  let [first, ...rest] = segments;

  // Optional path segments are denoted by a trailing `?`
  let isOptional = first.endsWith("?");
  // Compute the corresponding required segment: `foo?` -> `foo`
  let required = first.replace(/\?$/, "");

  if (rest.length === 0) ***REMOVED***
    // Intepret empty string as omitting an optional segment
    // `["one", "", "three"]` corresponds to omitting `:two` from `/one/:two?/three` -> `/one/three`
    return isOptional ? [required, ""] : [required];
***REMOVED***

  let restExploded = explodeOptionalSegments(rest.join("/"));

  let result: string[] = [];

  // All child paths with the prefix.  Do this for all children before the
  // optional version for all children so we get consistent ordering where the
  // parent optional aspect is preferred as required.  Otherwise, we can get
  // child sections interspersed where deeper optional segments are higher than
  // parent optional segments, where for example, /:two would explodes _earlier_
  // then /:one.  By always including the parent as required _for all children_
  // first, we avoid this issue
  result.push(
    ...restExploded.map((subpath) =>
      subpath === "" ? required : [required, subpath].join("/")
    )
  );

  // Then if this is an optional value, add all child versions without
  if (isOptional) ***REMOVED***
    result.push(...restExploded);
***REMOVED***

  // for absolute paths, ensure `/` instead of empty segment
  return result.map((exploded) =>
    path.startsWith("/") && exploded === "" ? "/" : exploded
  );
}

function rankRouteBranches(branches: RouteBranch[]): void ***REMOVED***
  branches.sort((a, b) =>
    a.score !== b.score
      ? b.score - a.score // Higher score first
      : compareIndexes(
          a.routesMeta.map((meta) => meta.childrenIndex),
          b.routesMeta.map((meta) => meta.childrenIndex)
        )
  );
}

const paramRe = /^:\w+$/;
const dynamicSegmentValue = 3;
const indexRouteValue = 2;
const emptySegmentValue = 1;
const staticSegmentValue = 10;
const splatPenalty = -2;
const isSplat = (s: string) => s === "*";

function computeScore(path: string, index: boolean | undefined): number ***REMOVED***
  let segments = path.split("/");
  let initialScore = segments.length;
  if (segments.some(isSplat)) ***REMOVED***
    initialScore += splatPenalty;
***REMOVED***

  if (index) ***REMOVED***
    initialScore += indexRouteValue;
***REMOVED***

  return segments
    .filter((s) => !isSplat(s))
    .reduce(
      (score, segment) =>
        score +
        (paramRe.test(segment)
          ? dynamicSegmentValue
          : segment === ""
          ? emptySegmentValue
          : staticSegmentValue),
      initialScore
    );
}

function compareIndexes(a: number[], b: number[]): number ***REMOVED***
  let siblings =
    a.length === b.length && a.slice(0, -1).every((n, i) => n === b[i]);

  return siblings
    ? // If two routes are siblings, we should try to match the earlier sibling
      // first. This allows people to have fine-grained control over the matching
      // behavior by simply putting routes with identical paths in the order they
      // want them tried.
      a[a.length - 1] - b[b.length - 1]
    : // Otherwise, it doesn't really make sense to rank non-siblings by index,
      // so they sort equally.
      0;
}

function matchRouteBranch<
  ParamKey extends string = string,
  RouteObjectType extends AgnosticRouteObject = AgnosticRouteObject
>(
  branch: RouteBranch<RouteObjectType>,
  pathname: string
): AgnosticRouteMatch<ParamKey, RouteObjectType>[] | null ***REMOVED***
  let ***REMOVED*** routesMeta } = branch;

  let matchedParams = ***REMOVED***};
  let matchedPathname = "/";
  let matches: AgnosticRouteMatch<ParamKey, RouteObjectType>[] = [];
  for (let i = 0; i < routesMeta.length; ++i) ***REMOVED***
    let meta = routesMeta[i];
    let end = i === routesMeta.length - 1;
    let remainingPathname =
      matchedPathname === "/"
        ? pathname
        : pathname.slice(matchedPathname.length) || "/";
    let match = matchPath(
      ***REMOVED*** path: meta.relativePath, caseSensitive: meta.caseSensitive, end },
      remainingPathname
    );

    if (!match) return null;

    Object.assign(matchedParams, match.params);

    let route = meta.route;

    matches.push(***REMOVED***
      // TODO: Can this as be avoided?
      params: matchedParams as Params<ParamKey>,
      pathname: joinPaths([matchedPathname, match.pathname]),
      pathnameBase: normalizePathname(
        joinPaths([matchedPathname, match.pathnameBase])
      ),
      route,
  ***REMOVED***);

    if (match.pathnameBase !== "/") ***REMOVED***
      matchedPathname = joinPaths([matchedPathname, match.pathnameBase]);
  ***REMOVED***
***REMOVED***

  return matches;
}

/**
 * Returns a path with params interpolated.
 *
 * @see https://reactrouter.com/utils/generate-path
 */
export function generatePath<Path extends string>(
  originalPath: Path,
  params: ***REMOVED***
    [key in PathParam<Path>]: string | null;
***REMOVED*** = ***REMOVED***} as any
): string ***REMOVED***
  let path = originalPath;
  if (path.endsWith("*") && path !== "*" && !path.endsWith("/*")) ***REMOVED***
    warning(
      false,
      `Route path "$***REMOVED***path}" will be treated as if it were ` +
        `"$***REMOVED***path.replace(/\*$/, "/*")}" because the \`*\` character must ` +
        `always follow a \`/\` in the pattern. To get rid of this warning, ` +
        `please change the route path to "$***REMOVED***path.replace(/\*$/, "/*")}".`
    );
    path = path.replace(/\*$/, "/*") as Path;
***REMOVED***

  return (
    path
      .replace(
        /^:(\w+)(\??)/g,
        (_, key: PathParam<Path>, optional: string | undefined) => ***REMOVED***
          let param = params[key];
          if (optional === "?") ***REMOVED***
            return param == null ? "" : param;
        ***REMOVED***
          if (param == null) ***REMOVED***
            invariant(false, `Missing ":$***REMOVED***key}" param`);
        ***REMOVED***
          return param;
      ***REMOVED***
      )
      .replace(
        /\/:(\w+)(\??)/g,
        (_, key: PathParam<Path>, optional: string | undefined) => ***REMOVED***
          let param = params[key];
          if (optional === "?") ***REMOVED***
            return param == null ? "" : `/$***REMOVED***param}`;
        ***REMOVED***
          if (param == null) ***REMOVED***
            invariant(false, `Missing ":$***REMOVED***key}" param`);
        ***REMOVED***
          return `/$***REMOVED***param}`;
      ***REMOVED***
      )
      // Remove any optional markers from optional static segments
      .replace(/\?/g, "")
      .replace(/(\/?)\*/, (_, prefix, __, str) => ***REMOVED***
        const star = "*" as PathParam<Path>;

        if (params[star] == null) ***REMOVED***
          // If no splat was provided, trim the trailing slash _unless_ it's
          // the entire path
          return str === "/*" ? "/" : "";
      ***REMOVED***

        // Apply the splat
        return `$***REMOVED***prefix}$***REMOVED***params[star]}`;
    ***REMOVED***)
  );
}

/**
 * A PathPattern is used to match on some portion of a URL pathname.
 */
export interface PathPattern<Path extends string = string> ***REMOVED***
  /**
   * A string to match against a URL pathname. May contain `:id`-style segments
   * to indicate placeholders for dynamic parameters. May also end with `/*` to
   * indicate matching the rest of the URL pathname.
   */
  path: Path;
  /**
   * Should be `true` if the static portions of the `path` should be matched in
   * the same case.
   */
  caseSensitive?: boolean;
  /**
   * Should be `true` if this pattern should match the entire URL pathname.
   */
  end?: boolean;
}

/**
 * A PathMatch contains info about how a PathPattern matched on a URL pathname.
 */
export interface PathMatch<ParamKey extends string = string> ***REMOVED***
  /**
   * The names and values of dynamic parameters in the URL.
   */
  params: Params<ParamKey>;
  /**
   * The portion of the URL pathname that was matched.
   */
  pathname: string;
  /**
   * The portion of the URL pathname that was matched before child routes.
   */
  pathnameBase: string;
  /**
   * The pattern that was used to match.
   */
  pattern: PathPattern;
}

type Mutable<T> = ***REMOVED***
  -readonly [P in keyof T]: T[P];
};

/**
 * Performs pattern matching on a URL pathname and returns information about
 * the match.
 *
 * @see https://reactrouter.com/utils/match-path
 */
export function matchPath<
  ParamKey extends ParamParseKey<Path>,
  Path extends string
>(
  pattern: PathPattern<Path> | Path,
  pathname: string
): PathMatch<ParamKey> | null ***REMOVED***
  if (typeof pattern === "string") ***REMOVED***
    pattern = ***REMOVED*** path: pattern, caseSensitive: false, end: true };
***REMOVED***

  let [matcher, paramNames] = compilePath(
    pattern.path,
    pattern.caseSensitive,
    pattern.end
  );

  let match = pathname.match(matcher);
  if (!match) return null;

  let matchedPathname = match[0];
  let pathnameBase = matchedPathname.replace(/(.)\/+$/, "$1");
  let captureGroups = match.slice(1);
  let params: Params = paramNames.reduce<Mutable<Params>>(
    (memo, paramName, index) => ***REMOVED***
      // We need to compute the pathnameBase here using the raw splat value
      // instead of using params["*"] later because it will be decoded then
      if (paramName === "*") ***REMOVED***
        let splatValue = captureGroups[index] || "";
        pathnameBase = matchedPathname
          .slice(0, matchedPathname.length - splatValue.length)
          .replace(/(.)\/+$/, "$1");
    ***REMOVED***

      memo[paramName] = safelyDecodeURIComponent(
        captureGroups[index] || "",
        paramName
      );
      return memo;
  ***REMOVED***,
    ***REMOVED***}
  );

  return ***REMOVED***
    params,
    pathname: matchedPathname,
    pathnameBase,
    pattern,
***REMOVED***;
}

function compilePath(
  path: string,
  caseSensitive = false,
  end = true
): [RegExp, string[]] ***REMOVED***
  warning(
    path === "*" || !path.endsWith("*") || path.endsWith("/*"),
    `Route path "$***REMOVED***path}" will be treated as if it were ` +
      `"$***REMOVED***path.replace(/\*$/, "/*")}" because the \`*\` character must ` +
      `always follow a \`/\` in the pattern. To get rid of this warning, ` +
      `please change the route path to "$***REMOVED***path.replace(/\*$/, "/*")}".`
  );

  let paramNames: string[] = [];
  let regexpSource =
    "^" +
    path
      .replace(/\/*\*?$/, "") // Ignore trailing / and /*, we'll handle it below
      .replace(/^\/*/, "/") // Make sure it has a leading /
      .replace(/[\\.*+^$?***REMOVED***}|()[\]]/g, "\\$&") // Escape special regex chars
      .replace(/\/:(\w+)/g, (_: string, paramName: string) => ***REMOVED***
        paramNames.push(paramName);
        return "/([^\\/]+)";
    ***REMOVED***);

  if (path.endsWith("*")) ***REMOVED***
    paramNames.push("*");
    regexpSource +=
      path === "*" || path === "/*"
        ? "(.*)$" // Already matched the initial /, just match the rest
        : "(?:\\/(.+)|\\/*)$"; // Don't include the / in params["*"]
***REMOVED*** else if (end) ***REMOVED***
    // When matching to the end, ignore trailing slashes
    regexpSource += "\\/*$";
***REMOVED*** else if (path !== "" && path !== "/") ***REMOVED***
    // If our path is non-empty and contains anything beyond an initial slash,
    // then we have _some_ form of path in our regex so we should expect to
    // match only if we find the end of this path segment.  Look for an optional
    // non-captured trailing slash (to match a portion of the URL) or the end
    // of the path (if we've matched to the end).  We used to do this with a
    // word boundary but that gives false positives on routes like
    // /user-preferences since `-` counts as a word boundary.
    regexpSource += "(?:(?=\\/|$))";
***REMOVED*** else ***REMOVED***
    // Nothing to match for "" or "/"
***REMOVED***

  let matcher = new RegExp(regexpSource, caseSensitive ? undefined : "i");

  return [matcher, paramNames];
}

function safelyDecodeURI(value: string) ***REMOVED***
  try ***REMOVED***
    return decodeURI(value);
***REMOVED*** catch (error) ***REMOVED***
    warning(
      false,
      `The URL path "$***REMOVED***value}" could not be decoded because it is is a ` +
        `malformed URL segment. This is probably due to a bad percent ` +
        `encoding ($***REMOVED***error}).`
    );

    return value;
***REMOVED***
}

function safelyDecodeURIComponent(value: string, paramName: string) ***REMOVED***
  try ***REMOVED***
    return decodeURIComponent(value);
***REMOVED*** catch (error) ***REMOVED***
    warning(
      false,
      `The value for the URL param "$***REMOVED***paramName}" will not be decoded because` +
        ` the string "$***REMOVED***value}" is a malformed URL segment. This is probably` +
        ` due to a bad percent encoding ($***REMOVED***error}).`
    );

    return value;
***REMOVED***
}

/**
 * @private
 */
export function stripBasename(
  pathname: string,
  basename: string
): string | null ***REMOVED***
  if (basename === "/") return pathname;

  if (!pathname.toLowerCase().startsWith(basename.toLowerCase())) ***REMOVED***
    return null;
***REMOVED***

  // We want to leave trailing slash behavior in the user's control, so if they
  // specify a basename with a trailing slash, we should support it
  let startIndex = basename.endsWith("/")
    ? basename.length - 1
    : basename.length;
  let nextChar = pathname.charAt(startIndex);
  if (nextChar && nextChar !== "/") ***REMOVED***
    // pathname does not start with basename/
    return null;
***REMOVED***

  return pathname.slice(startIndex) || "/";
}

/**
 * @private
 */
export function warning(cond: any, message: string): void ***REMOVED***
  if (!cond) ***REMOVED***
    // eslint-disable-next-line no-console
    if (typeof console !== "undefined") console.warn(message);

    try ***REMOVED***
      // Welcome to debugging @remix-run/router!
      //
      // This error is thrown as a convenience so you can more easily
      // find the source for a warning that appears in the console by
      // enabling "pause on exceptions" in your JavaScript debugger.
      throw new Error(message);
      // eslint-disable-next-line no-empty
  ***REMOVED*** catch (e) ***REMOVED***}
***REMOVED***
}

/**
 * Returns a resolved path object relative to the given pathname.
 *
 * @see https://reactrouter.com/utils/resolve-path
 */
export function resolvePath(to: To, fromPathname = "/"): Path ***REMOVED***
  let ***REMOVED***
    pathname: toPathname,
    search = "",
    hash = "",
***REMOVED*** = typeof to === "string" ? parsePath(to) : to;

  let pathname = toPathname
    ? toPathname.startsWith("/")
      ? toPathname
      : resolvePathname(toPathname, fromPathname)
    : fromPathname;

  return ***REMOVED***
    pathname,
    search: normalizeSearch(search),
    hash: normalizeHash(hash),
***REMOVED***;
}

function resolvePathname(relativePath: string, fromPathname: string): string ***REMOVED***
  let segments = fromPathname.replace(/\/+$/, "").split("/");
  let relativeSegments = relativePath.split("/");

  relativeSegments.forEach((segment) => ***REMOVED***
    if (segment === "..") ***REMOVED***
      // Keep the root "" segment so the pathname starts at /
      if (segments.length > 1) segments.pop();
  ***REMOVED*** else if (segment !== ".") ***REMOVED***
      segments.push(segment);
  ***REMOVED***
***REMOVED***);

  return segments.length > 1 ? segments.join("/") : "/";
}

function getInvalidPathError(
  char: string,
  field: string,
  dest: string,
  path: Partial<Path>
) ***REMOVED***
  return (
    `Cannot include a '$***REMOVED***char}' character in a manually specified ` +
    `\`to.$***REMOVED***field}\` field [$***REMOVED***JSON.stringify(
      path
    )}].  Please separate it out to the ` +
    `\`to.$***REMOVED***dest}\` field. Alternatively you may provide the full path as ` +
    `a string in <Link to="..."> and the router will parse it for you.`
  );
}

/**
 * @private
 *
 * When processing relative navigation we want to ignore ancestor routes that
 * do not contribute to the path, such that index/pathless layout routes don't
 * interfere.
 *
 * For example, when moving a route element into an index route and/or a
 * pathless layout route, relative link behavior contained within should stay
 * the same.  Both of the following examples should link back to the root:
 *
 *   <Route path="/">
 *     <Route path="accounts" element=***REMOVED***<Link to=".."}>
 *   </Route>
 *
 *   <Route path="/">
 *     <Route path="accounts">
 *       <Route element=***REMOVED***<AccountsLayout />}>       // <-- Does not contribute
 *         <Route index element=***REMOVED***<Link to=".."} />  // <-- Does not contribute
 *       </Route
 *     </Route>
 *   </Route>
 */
export function getPathContributingMatches<
  T extends AgnosticRouteMatch = AgnosticRouteMatch
>(matches: T[]) ***REMOVED***
  return matches.filter(
    (match, index) =>
      index === 0 || (match.route.path && match.route.path.length > 0)
  );
}

/**
 * @private
 */
export function resolveTo(
  toArg: To,
  routePathnames: string[],
  locationPathname: string,
  isPathRelative = false
): Path ***REMOVED***
  let to: Partial<Path>;
  if (typeof toArg === "string") ***REMOVED***
    to = parsePath(toArg);
***REMOVED*** else ***REMOVED***
    to = ***REMOVED*** ...toArg };

    invariant(
      !to.pathname || !to.pathname.includes("?"),
      getInvalidPathError("?", "pathname", "search", to)
    );
    invariant(
      !to.pathname || !to.pathname.includes("#"),
      getInvalidPathError("#", "pathname", "hash", to)
    );
    invariant(
      !to.search || !to.search.includes("#"),
      getInvalidPathError("#", "search", "hash", to)
    );
***REMOVED***

  let isEmptyPath = toArg === "" || to.pathname === "";
  let toPathname = isEmptyPath ? "/" : to.pathname;

  let from: string;

  // Routing is relative to the current pathname if explicitly requested.
  //
  // If a pathname is explicitly provided in `to`, it should be relative to the
  // route context. This is explained in `Note on `<Link to>` values` in our
  // migration guide from v5 as a means of disambiguation between `to` values
  // that begin with `/` and those that do not. However, this is problematic for
  // `to` values that do not provide a pathname. `to` can simply be a search or
  // hash string, in which case we should assume that the navigation is relative
  // to the current location's pathname and *not* the route pathname.
  if (isPathRelative || toPathname == null) ***REMOVED***
    from = locationPathname;
***REMOVED*** else ***REMOVED***
    let routePathnameIndex = routePathnames.length - 1;

    if (toPathname.startsWith("..")) ***REMOVED***
      let toSegments = toPathname.split("/");

      // Each leading .. segment means "go up one route" instead of "go up one
      // URL segment".  This is a key difference from how <a href> works and a
      // major reason we call this a "to" value instead of a "href".
      while (toSegments[0] === "..") ***REMOVED***
        toSegments.shift();
        routePathnameIndex -= 1;
    ***REMOVED***

      to.pathname = toSegments.join("/");
  ***REMOVED***

    // If there are more ".." segments than parent routes, resolve relative to
    // the root / URL.
    from = routePathnameIndex >= 0 ? routePathnames[routePathnameIndex] : "/";
***REMOVED***

  let path = resolvePath(to, from);

  // Ensure the pathname has a trailing slash if the original "to" had one
  let hasExplicitTrailingSlash =
    toPathname && toPathname !== "/" && toPathname.endsWith("/");
  // Or if this was a link to the current path which has a trailing slash
  let hasCurrentTrailingSlash =
    (isEmptyPath || toPathname === ".") && locationPathname.endsWith("/");
  if (
    !path.pathname.endsWith("/") &&
    (hasExplicitTrailingSlash || hasCurrentTrailingSlash)
  ) ***REMOVED***
    path.pathname += "/";
***REMOVED***

  return path;
}

/**
 * @private
 */
export function getToPathname(to: To): string | undefined ***REMOVED***
  // Empty strings should be treated the same as / paths
  return to === "" || (to as Path).pathname === ""
    ? "/"
    : typeof to === "string"
    ? parsePath(to).pathname
    : to.pathname;
}

/**
 * @private
 */
export const joinPaths = (paths: string[]): string =>
  paths.join("/").replace(/\/\/+/g, "/");

/**
 * @private
 */
export const normalizePathname = (pathname: string): string =>
  pathname.replace(/\/+$/, "").replace(/^\/*/, "/");

/**
 * @private
 */
export const normalizeSearch = (search: string): string =>
  !search || search === "?"
    ? ""
    : search.startsWith("?")
    ? search
    : "?" + search;

/**
 * @private
 */
export const normalizeHash = (hash: string): string =>
  !hash || hash === "#" ? "" : hash.startsWith("#") ? hash : "#" + hash;

export type JsonFunction = <Data>(
  data: Data,
  init?: number | ResponseInit
) => Response;

/**
 * This is a shortcut for creating `application/json` responses. Converts `data`
 * to JSON and sets the `Content-Type` header.
 */
export const json: JsonFunction = (data, init = ***REMOVED***}) => ***REMOVED***
  let responseInit = typeof init === "number" ? ***REMOVED*** status: init } : init;

  let headers = new Headers(responseInit.headers);
  if (!headers.has("Content-Type")) ***REMOVED***
    headers.set("Content-Type", "application/json; charset=utf-8");
***REMOVED***

  return new Response(JSON.stringify(data), ***REMOVED***
    ...responseInit,
    headers,
***REMOVED***);
};

export interface TrackedPromise extends Promise<any> ***REMOVED***
  _tracked?: boolean;
  _data?: any;
  _error?: any;
}

export class AbortedDeferredError extends Error ***REMOVED***}

export class DeferredData ***REMOVED***
  private pendingKeysSet: Set<string> = new Set<string>();
  private controller: AbortController;
  private abortPromise: Promise<void>;
  private unlistenAbortSignal: () => void;
  private subscribers: Set<(aborted: boolean, settledKey?: string) => void> =
    new Set();
  data: Record<string, unknown>;
  init?: ResponseInit;
  deferredKeys: string[] = [];

  constructor(data: Record<string, unknown>, responseInit?: ResponseInit) ***REMOVED***
    invariant(
      data && typeof data === "object" && !Array.isArray(data),
      "defer() only accepts plain objects"
    );

    // Set up an AbortController + Promise we can race against to exit early
    // cancellation
    let reject: (e: AbortedDeferredError) => void;
    this.abortPromise = new Promise((_, r) => (reject = r));
    this.controller = new AbortController();
    let onAbort = () =>
      reject(new AbortedDeferredError("Deferred data aborted"));
    this.unlistenAbortSignal = () =>
      this.controller.signal.removeEventListener("abort", onAbort);
    this.controller.signal.addEventListener("abort", onAbort);

    this.data = Object.entries(data).reduce(
      (acc, [key, value]) =>
        Object.assign(acc, ***REMOVED***
          [key]: this.trackPromise(key, value),
      ***REMOVED***),
      ***REMOVED***}
    );

    this.init = responseInit;
***REMOVED***

  private trackPromise(
    key: string,
    value: Promise<unknown> | unknown
  ): TrackedPromise | unknown ***REMOVED***
    if (!(value instanceof Promise)) ***REMOVED***
      return value;
  ***REMOVED***

    this.deferredKeys.push(key);
    this.pendingKeysSet.add(key);

    // We store a little wrapper promise that will be extended with
    // _data/_error props upon resolve/reject
    let promise: TrackedPromise = Promise.race([value, this.abortPromise]).then(
      (data) => this.onSettle(promise, key, null, data as unknown),
      (error) => this.onSettle(promise, key, error as unknown)
    );

    // Register rejection listeners to avoid uncaught promise rejections on
    // errors or aborted deferred values
    promise.catch(() => ***REMOVED***});

    Object.defineProperty(promise, "_tracked", ***REMOVED*** get: () => true });
    return promise;
***REMOVED***

  private onSettle(
    promise: TrackedPromise,
    key: string,
    error: unknown,
    data?: unknown
  ): unknown ***REMOVED***
    if (
      this.controller.signal.aborted &&
      error instanceof AbortedDeferredError
    ) ***REMOVED***
      this.unlistenAbortSignal();
      Object.defineProperty(promise, "_error", ***REMOVED*** get: () => error });
      return Promise.reject(error);
  ***REMOVED***

    this.pendingKeysSet.delete(key);

    if (this.done) ***REMOVED***
      // Nothing left to abort!
      this.unlistenAbortSignal();
  ***REMOVED***

    if (error) ***REMOVED***
      Object.defineProperty(promise, "_error", ***REMOVED*** get: () => error });
      this.emit(false, key);
      return Promise.reject(error);
  ***REMOVED***

    Object.defineProperty(promise, "_data", ***REMOVED*** get: () => data });
    this.emit(false, key);
    return data;
***REMOVED***

  private emit(aborted: boolean, settledKey?: string) ***REMOVED***
    this.subscribers.forEach((subscriber) => subscriber(aborted, settledKey));
***REMOVED***

  subscribe(fn: (aborted: boolean, settledKey?: string) => void) ***REMOVED***
    this.subscribers.add(fn);
    return () => this.subscribers.delete(fn);
***REMOVED***

  cancel() ***REMOVED***
    this.controller.abort();
    this.pendingKeysSet.forEach((v, k) => this.pendingKeysSet.delete(k));
    this.emit(true);
***REMOVED***

  async resolveData(signal: AbortSignal) ***REMOVED***
    let aborted = false;
    if (!this.done) ***REMOVED***
      let onAbort = () => this.cancel();
      signal.addEventListener("abort", onAbort);
      aborted = await new Promise((resolve) => ***REMOVED***
        this.subscribe((aborted) => ***REMOVED***
          signal.removeEventListener("abort", onAbort);
          if (aborted || this.done) ***REMOVED***
            resolve(aborted);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***);
  ***REMOVED***
    return aborted;
***REMOVED***

  get done() ***REMOVED***
    return this.pendingKeysSet.size === 0;
***REMOVED***

  get unwrappedData() ***REMOVED***
    invariant(
      this.data !== null && this.done,
      "Can only unwrap data on initialized and settled deferreds"
    );

    return Object.entries(this.data).reduce(
      (acc, [key, value]) =>
        Object.assign(acc, ***REMOVED***
          [key]: unwrapTrackedPromise(value),
      ***REMOVED***),
      ***REMOVED***}
    );
***REMOVED***

  get pendingKeys() ***REMOVED***
    return Array.from(this.pendingKeysSet);
***REMOVED***
}

function isTrackedPromise(value: any): value is TrackedPromise ***REMOVED***
  return (
    value instanceof Promise && (value as TrackedPromise)._tracked === true
  );
}

function unwrapTrackedPromise(value: any) ***REMOVED***
  if (!isTrackedPromise(value)) ***REMOVED***
    return value;
***REMOVED***

  if (value._error) ***REMOVED***
    throw value._error;
***REMOVED***
  return value._data;
}

export type DeferFunction = (
  data: Record<string, unknown>,
  init?: number | ResponseInit
) => DeferredData;

export const defer: DeferFunction = (data, init = ***REMOVED***}) => ***REMOVED***
  let responseInit = typeof init === "number" ? ***REMOVED*** status: init } : init;

  return new DeferredData(data, responseInit);
};

export type RedirectFunction = (
  url: string,
  init?: number | ResponseInit
) => Response;

/**
 * A redirect response. Sets the status code and the `Location` header.
 * Defaults to "302 Found".
 */
export const redirect: RedirectFunction = (url, init = 302) => ***REMOVED***
  let responseInit = init;
  if (typeof responseInit === "number") ***REMOVED***
    responseInit = ***REMOVED*** status: responseInit };
***REMOVED*** else if (typeof responseInit.status === "undefined") ***REMOVED***
    responseInit.status = 302;
***REMOVED***

  let headers = new Headers(responseInit.headers);
  headers.set("Location", url);

  return new Response(null, ***REMOVED***
    ...responseInit,
    headers,
***REMOVED***);
};

/**
 * @private
 * Utility class we use to hold auto-unwrapped 4xx/5xx Response bodies
 */
export class ErrorResponse ***REMOVED***
  status: number;
  statusText: string;
  data: any;
  error?: Error;
  internal: boolean;

  constructor(
    status: number,
    statusText: string | undefined,
    data: any,
    internal = false
  ) ***REMOVED***
    this.status = status;
    this.statusText = statusText || "";
    this.internal = internal;
    if (data instanceof Error) ***REMOVED***
      this.data = data.toString();
      this.error = data;
  ***REMOVED*** else ***REMOVED***
      this.data = data;
  ***REMOVED***
***REMOVED***
}

/**
 * Check if the given error is an ErrorResponse generated from a 4xx/5xx
 * Response throw from an action/loader
 */
export function isRouteErrorResponse(e: any): e is ErrorResponse ***REMOVED***
  return e instanceof ErrorResponse;
}
