// pg/lib/index.js
'use strict'

var Client = require('./client')
var defaults = require('./defaults')
var Connection = require('./connection')
var Pool = require('pg-pool')
const ***REMOVED*** DatabaseError } = require('pg-protocol')

const poolFactory = (Client) => ***REMOVED***
  return class BoundPool extends Pool ***REMOVED***
    constructor(options) ***REMOVED***
      super(options, Client)
  ***REMOVED***
***REMOVED***
}

var PG = function (clientConstructor) ***REMOVED***
  this.defaults = defaults
  this.Client = clientConstructor
  this.Query = this.Client.Query
  this.Pool = poolFactory(this.Client)
  this._pools = []
  this.Connection = Connection
  this.types = require('pg-types')
  this.DatabaseError = DatabaseError
}

if (typeof process.env.NODE_PG_FORCE_NATIVE !== 'undefined') ***REMOVED***
  module.exports = new PG(require('./native'))
} else ***REMOVED***
  module.exports = new PG(Client)

  // lazy require native module...the native module may not have installed
  Object.defineProperty(module.exports, 'native', ***REMOVED***
    configurable: true,
    enumerable: false,
    get() ***REMOVED***
      var native = null
      try ***REMOVED***
        native = new PG(require('./native'))
    ***REMOVED*** catch (err) ***REMOVED***
        if (err.code !== 'MODULE_NOT_FOUND') ***REMOVED***
          throw err
      ***REMOVED***
    ***REMOVED***

      // overwrite module.exports.native so that getter is never called again
      Object.defineProperty(module.exports, 'native', ***REMOVED***
        value: native,
    ***REMOVED***)

      return native
  ***REMOVED***,
***REMOVED***)
}
