// pg/lib/result.js
'use strict'

var types = require('pg-types')

var matchRegexp = /^([A-Za-z]+)(?: (\d+))?(?: (\d+))?/

// result object returned from query
// in the 'end' event and also
// passed as second argument to provided callback
class Result ***REMOVED***
  constructor(rowMode, types) ***REMOVED***
    this.command = null
    this.rowCount = null
    this.oid = null
    this.rows = []
    this.fields = []
    this._parsers = undefined
    this._types = types
    this.RowCtor = null
    this.rowAsArray = rowMode === 'array'
    if (this.rowAsArray) ***REMOVED***
      this.parseRow = this._parseRowAsArray
  ***REMOVED***
***REMOVED***

  // adds a command complete message
  addCommandComplete(msg) ***REMOVED***
    var match
    if (msg.text) ***REMOVED***
      // pure javascript
      match = matchRegexp.exec(msg.text)
  ***REMOVED*** else ***REMOVED***
      // native bindings
      match = matchRegexp.exec(msg.command)
  ***REMOVED***
    if (match) ***REMOVED***
      this.command = match[1]
      if (match[3]) ***REMOVED***
        // COMMMAND OID ROWS
        this.oid = parseInt(match[2], 10)
        this.rowCount = parseInt(match[3], 10)
    ***REMOVED*** else if (match[2]) ***REMOVED***
        // COMMAND ROWS
        this.rowCount = parseInt(match[2], 10)
    ***REMOVED***
  ***REMOVED***
***REMOVED***

  _parseRowAsArray(rowData) ***REMOVED***
    var row = new Array(rowData.length)
    for (var i = 0, len = rowData.length; i < len; i++) ***REMOVED***
      var rawValue = rowData[i]
      if (rawValue !== null) ***REMOVED***
        row[i] = this._parsers[i](rawValue)
    ***REMOVED*** else ***REMOVED***
        row[i] = null
    ***REMOVED***
  ***REMOVED***
    return row
***REMOVED***

  parseRow(rowData) ***REMOVED***
    var row = ***REMOVED***}
    for (var i = 0, len = rowData.length; i < len; i++) ***REMOVED***
      var rawValue = rowData[i]
      var field = this.fields[i].name
      if (rawValue !== null) ***REMOVED***
        row[field] = this._parsers[i](rawValue)
    ***REMOVED*** else ***REMOVED***
        row[field] = null
    ***REMOVED***
  ***REMOVED***
    return row
***REMOVED***

  addRow(row) ***REMOVED***
    this.rows.push(row)
***REMOVED***

  addFields(fieldDescriptions) ***REMOVED***
    // clears field definitions
    // multiple query statements in 1 action can result in multiple sets
    // of rowDescriptions...eg: 'select NOW(); select 1::int;'
    // you need to reset the fields
    this.fields = fieldDescriptions
    if (this.fields.length) ***REMOVED***
      this._parsers = new Array(fieldDescriptions.length)
  ***REMOVED***
    for (var i = 0; i < fieldDescriptions.length; i++) ***REMOVED***
      var desc = fieldDescriptions[i]
      if (this._types) ***REMOVED***
        this._parsers[i] = this._types.getTypeParser(desc.dataTypeID, desc.format || 'text')
    ***REMOVED*** else ***REMOVED***
        this._parsers[i] = types.getTypeParser(desc.dataTypeID, desc.format || 'text')
    ***REMOVED***
  ***REMOVED***
***REMOVED***
}

module.exports = Result
