// pg/lib/utils.js
'use strict'

const crypto = require('crypto')

const defaults = require('./defaults')

function escapeElement(elementRepresentation) ***REMOVED***
  var escaped = elementRepresentation.replace(/\\/g, '\\\\').replace(/"/g, '\\"')

  return '"' + escaped + '"'
}

// convert a JS array to a postgres array literal
// uses comma separator so won't work for types like box that use
// a different array separator.
function arrayString(val) ***REMOVED***
  var result = '***REMOVED***'
  for (var i = 0; i < val.length; i++) ***REMOVED***
    if (i > 0) ***REMOVED***
      result = result + ','
  ***REMOVED***
    if (val[i] === null || typeof val[i] === 'undefined') ***REMOVED***
      result = result + 'NULL'
  ***REMOVED*** else if (Array.isArray(val[i])) ***REMOVED***
      result = result + arrayString(val[i])
  ***REMOVED*** else if (val[i] instanceof Buffer) ***REMOVED***
      result += '\\\\x' + val[i].toString('hex')
  ***REMOVED*** else ***REMOVED***
      result += escapeElement(prepareValue(val[i]))
  ***REMOVED***
***REMOVED***
  result = result + '}'
  return result
}

// converts values from javascript types
// to their 'raw' counterparts for use as a postgres parameter
// note: you can override this function to provide your own conversion mechanism
// for complex types, etc...
var prepareValue = function (val, seen) ***REMOVED***
  // null and undefined are both null for postgres
  if (val == null) ***REMOVED***
    return null
***REMOVED***
  if (val instanceof Buffer) ***REMOVED***
    return val
***REMOVED***
  if (ArrayBuffer.isView(val)) ***REMOVED***
    var buf = Buffer.from(val.buffer, val.byteOffset, val.byteLength)
    if (buf.length === val.byteLength) ***REMOVED***
      return buf
  ***REMOVED***
    return buf.slice(val.byteOffset, val.byteOffset + val.byteLength) // Node.js v4 does not support those Buffer.from params
***REMOVED***
  if (val instanceof Date) ***REMOVED***
    if (defaults.parseInputDatesAsUTC) ***REMOVED***
      return dateToStringUTC(val)
  ***REMOVED*** else ***REMOVED***
      return dateToString(val)
  ***REMOVED***
***REMOVED***
  if (Array.isArray(val)) ***REMOVED***
    return arrayString(val)
***REMOVED***
  if (typeof val === 'object') ***REMOVED***
    return prepareObject(val, seen)
***REMOVED***
  return val.toString()
}

function prepareObject(val, seen) ***REMOVED***
  if (val && typeof val.toPostgres === 'function') ***REMOVED***
    seen = seen || []
    if (seen.indexOf(val) !== -1) ***REMOVED***
      throw new Error('circular reference detected while preparing "' + val + '" for query')
  ***REMOVED***
    seen.push(val)

    return prepareValue(val.toPostgres(prepareValue), seen)
***REMOVED***
  return JSON.stringify(val)
}

function pad(number, digits) ***REMOVED***
  number = '' + number
  while (number.length < digits) ***REMOVED***
    number = '0' + number
***REMOVED***
  return number
}

function dateToString(date) ***REMOVED***
  var offset = -date.getTimezoneOffset()

  var year = date.getFullYear()
  var isBCYear = year < 1
  if (isBCYear) year = Math.abs(year) + 1 // negative years are 1 off their BC representation

  var ret =
    pad(year, 4) +
    '-' +
    pad(date.getMonth() + 1, 2) +
    '-' +
    pad(date.getDate(), 2) +
    'T' +
    pad(date.getHours(), 2) +
    ':' +
    pad(date.getMinutes(), 2) +
    ':' +
    pad(date.getSeconds(), 2) +
    '.' +
    pad(date.getMilliseconds(), 3)

  if (offset < 0) ***REMOVED***
    ret += '-'
    offset *= -1
***REMOVED*** else ***REMOVED***
    ret += '+'
***REMOVED***

  ret += pad(Math.floor(offset / 60), 2) + ':' + pad(offset % 60, 2)
  if (isBCYear) ret += ' BC'
  return ret
}

function dateToStringUTC(date) ***REMOVED***
  var year = date.getUTCFullYear()
  var isBCYear = year < 1
  if (isBCYear) year = Math.abs(year) + 1 // negative years are 1 off their BC representation

  var ret =
    pad(year, 4) +
    '-' +
    pad(date.getUTCMonth() + 1, 2) +
    '-' +
    pad(date.getUTCDate(), 2) +
    'T' +
    pad(date.getUTCHours(), 2) +
    ':' +
    pad(date.getUTCMinutes(), 2) +
    ':' +
    pad(date.getUTCSeconds(), 2) +
    '.' +
    pad(date.getUTCMilliseconds(), 3)

  ret += '+00:00'
  if (isBCYear) ret += ' BC'
  return ret
}

function normalizeQueryConfig(config, values, callback) ***REMOVED***
  // can take in strings or config objects
  config = typeof config === 'string' ? ***REMOVED*** text: config } : config
  if (values) ***REMOVED***
    if (typeof values === 'function') ***REMOVED***
      config.callback = values
  ***REMOVED*** else ***REMOVED***
      config.values = values
  ***REMOVED***
***REMOVED***
  if (callback) ***REMOVED***
    config.callback = callback
***REMOVED***
  return config
}

const md5 = function (string) ***REMOVED***
  return crypto.createHash('md5').update(string, 'utf-8').digest('hex')
}

// See AuthenticationMD5Password at https://www.postgresql.org/docs/current/static/protocol-flow.html
const postgresMd5PasswordHash = function (user, password, salt) ***REMOVED***
  var inner = md5(password + user)
  var outer = md5(Buffer.concat([Buffer.from(inner), salt]))
  return 'md5' + outer
}

module.exports = ***REMOVED***
  prepareValue: function prepareValueWrapper(value) ***REMOVED***
    // this ensures that extra arguments do not get passed into prepareValue
    // by accident, eg: from calling values.map(utils.prepareValue)
    return prepareValue(value)
***REMOVED***,
  normalizeQueryConfig,
  postgresMd5PasswordHash,
  md5,
}
