// google-gax/build/protos/http.js
/*eslint-disable block-scoped-var, id-length, no-control-regex, no-magic-numbers, no-prototype-builtins, no-redeclare, no-shadow, no-var, sort-vars*/
(function(global, factory) ***REMOVED*** /* global define, require, module */

    /* AMD */ if (typeof define === 'function' && define.amd)
        define(["protobufjs/minimal"], factory);

    /* CommonJS */ else if (typeof require === 'function' && typeof module === 'object' && module && module.exports)
        module.exports = factory(require("protobufjs/minimal"));

})(this, function($protobuf) ***REMOVED***
    "use strict";

    // Common aliases
    var $Reader = $protobuf.Reader, $Writer = $protobuf.Writer, $util = $protobuf.util;
    
    // Exported root namespace
    var $root = $protobuf.roots["default"] || ($protobuf.roots["default"] = ***REMOVED***});
    
    $root.google = (function() ***REMOVED***
    
        /**
         * Namespace google.
         * @exports google
         * @namespace
         */
        var google = ***REMOVED***};
    
        google.api = (function() ***REMOVED***
    
            /**
             * Namespace api.
             * @memberof google
             * @namespace
             */
            var api = ***REMOVED***};
    
            api.Http = (function() ***REMOVED***
    
                /**
                 * Properties of a Http.
                 * @memberof google.api
                 * @interface IHttp
                 * @property ***REMOVED***Array.<google.api.IHttpRule>|null} [rules] Http rules
                 * @property ***REMOVED***boolean|null} [fully_decode_reserved_expansion] Http fully_decode_reserved_expansion
                 */
    
                /**
                 * Constructs a new Http.
                 * @memberof google.api
                 * @classdesc Represents a Http.
                 * @implements IHttp
                 * @constructor
                 * @param ***REMOVED***google.api.IHttp=} [properties] Properties to set
                 */
                function Http(properties) ***REMOVED***
                    this.rules = [];
                    if (properties)
                        for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                            if (properties[keys[i]] != null)
                                this[keys[i]] = properties[keys[i]];
              ***REMOVED***
    
                /**
                 * Http rules.
                 * @member ***REMOVED***Array.<google.api.IHttpRule>} rules
                 * @memberof google.api.Http
                 * @instance
                 */
                Http.prototype.rules = $util.emptyArray;
    
                /**
                 * Http fully_decode_reserved_expansion.
                 * @member ***REMOVED***boolean} fully_decode_reserved_expansion
                 * @memberof google.api.Http
                 * @instance
                 */
                Http.prototype.fully_decode_reserved_expansion = false;
    
                /**
                 * Creates a new Http instance using the specified properties.
                 * @function create
                 * @memberof google.api.Http
                 * @static
                 * @param ***REMOVED***google.api.IHttp=} [properties] Properties to set
                 * @returns ***REMOVED***google.api.Http} Http instance
                 */
                Http.create = function create(properties) ***REMOVED***
                    return new Http(properties);
              ***REMOVED***;
    
                /**
                 * Encodes the specified Http message. Does not implicitly ***REMOVED***@link google.api.Http.verify|verify} messages.
                 * @function encode
                 * @memberof google.api.Http
                 * @static
                 * @param ***REMOVED***google.api.IHttp} message Http message or plain object to encode
                 * @param ***REMOVED***$protobuf.Writer} [writer] Writer to encode to
                 * @returns ***REMOVED***$protobuf.Writer} Writer
                 */
                Http.encode = function encode(message, writer) ***REMOVED***
                    if (!writer)
                        writer = $Writer.create();
                    if (message.rules != null && message.rules.length)
                        for (var i = 0; i < message.rules.length; ++i)
                            $root.google.api.HttpRule.encode(message.rules[i], writer.uint32(/* id 1, wireType 2 =*/10).fork()).ldelim();
                    if (message.fully_decode_reserved_expansion != null && Object.hasOwnProperty.call(message, "fully_decode_reserved_expansion"))
                        writer.uint32(/* id 2, wireType 0 =*/16).bool(message.fully_decode_reserved_expansion);
                    return writer;
              ***REMOVED***;
    
                /**
                 * Encodes the specified Http message, length delimited. Does not implicitly ***REMOVED***@link google.api.Http.verify|verify} messages.
                 * @function encodeDelimited
                 * @memberof google.api.Http
                 * @static
                 * @param ***REMOVED***google.api.IHttp} message Http message or plain object to encode
                 * @param ***REMOVED***$protobuf.Writer} [writer] Writer to encode to
                 * @returns ***REMOVED***$protobuf.Writer} Writer
                 */
                Http.encodeDelimited = function encodeDelimited(message, writer) ***REMOVED***
                    return this.encode(message, writer).ldelim();
              ***REMOVED***;
    
                /**
                 * Decodes a Http message from the specified reader or buffer.
                 * @function decode
                 * @memberof google.api.Http
                 * @static
                 * @param ***REMOVED***$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
                 * @param ***REMOVED***number} [length] Message length if known beforehand
                 * @returns ***REMOVED***google.api.Http} Http
                 * @throws ***REMOVED***Error} If the payload is not a reader or valid buffer
                 * @throws ***REMOVED***$protobuf.util.ProtocolError} If required fields are missing
                 */
                Http.decode = function decode(reader, length) ***REMOVED***
                    if (!(reader instanceof $Reader))
                        reader = $Reader.create(reader);
                    var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.api.Http();
                    while (reader.pos < end) ***REMOVED***
                        var tag = reader.uint32();
                        switch (tag >>> 3) ***REMOVED***
                        case 1:
                            if (!(message.rules && message.rules.length))
                                message.rules = [];
                            message.rules.push($root.google.api.HttpRule.decode(reader, reader.uint32()));
                            break;
                        case 2:
                            message.fully_decode_reserved_expansion = reader.bool();
                            break;
                        default:
                            reader.skipType(tag & 7);
                            break;
                      ***REMOVED***
                  ***REMOVED***
                    return message;
              ***REMOVED***;
    
                /**
                 * Decodes a Http message from the specified reader or buffer, length delimited.
                 * @function decodeDelimited
                 * @memberof google.api.Http
                 * @static
                 * @param ***REMOVED***$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
                 * @returns ***REMOVED***google.api.Http} Http
                 * @throws ***REMOVED***Error} If the payload is not a reader or valid buffer
                 * @throws ***REMOVED***$protobuf.util.ProtocolError} If required fields are missing
                 */
                Http.decodeDelimited = function decodeDelimited(reader) ***REMOVED***
                    if (!(reader instanceof $Reader))
                        reader = new $Reader(reader);
                    return this.decode(reader, reader.uint32());
              ***REMOVED***;
    
                /**
                 * Verifies a Http message.
                 * @function verify
                 * @memberof google.api.Http
                 * @static
                 * @param ***REMOVED***Object.<string,*>} message Plain object to verify
                 * @returns ***REMOVED***string|null} `null` if valid, otherwise the reason why it is not
                 */
                Http.verify = function verify(message) ***REMOVED***
                    if (typeof message !== "object" || message === null)
                        return "object expected";
                    if (message.rules != null && message.hasOwnProperty("rules")) ***REMOVED***
                        if (!Array.isArray(message.rules))
                            return "rules: array expected";
                        for (var i = 0; i < message.rules.length; ++i) ***REMOVED***
                            var error = $root.google.api.HttpRule.verify(message.rules[i]);
                            if (error)
                                return "rules." + error;
                      ***REMOVED***
                  ***REMOVED***
                    if (message.fully_decode_reserved_expansion != null && message.hasOwnProperty("fully_decode_reserved_expansion"))
                        if (typeof message.fully_decode_reserved_expansion !== "boolean")
                            return "fully_decode_reserved_expansion: boolean expected";
                    return null;
              ***REMOVED***;
    
                /**
                 * Creates a Http message from a plain object. Also converts values to their respective internal types.
                 * @function fromObject
                 * @memberof google.api.Http
                 * @static
                 * @param ***REMOVED***Object.<string,*>} object Plain object
                 * @returns ***REMOVED***google.api.Http} Http
                 */
                Http.fromObject = function fromObject(object) ***REMOVED***
                    if (object instanceof $root.google.api.Http)
                        return object;
                    var message = new $root.google.api.Http();
                    if (object.rules) ***REMOVED***
                        if (!Array.isArray(object.rules))
                            throw TypeError(".google.api.Http.rules: array expected");
                        message.rules = [];
                        for (var i = 0; i < object.rules.length; ++i) ***REMOVED***
                            if (typeof object.rules[i] !== "object")
                                throw TypeError(".google.api.Http.rules: object expected");
                            message.rules[i] = $root.google.api.HttpRule.fromObject(object.rules[i]);
                      ***REMOVED***
                  ***REMOVED***
                    if (object.fully_decode_reserved_expansion != null)
                        message.fully_decode_reserved_expansion = Boolean(object.fully_decode_reserved_expansion);
                    return message;
              ***REMOVED***;
    
                /**
                 * Creates a plain object from a Http message. Also converts values to other types if specified.
                 * @function toObject
                 * @memberof google.api.Http
                 * @static
                 * @param ***REMOVED***google.api.Http} message Http
                 * @param ***REMOVED***$protobuf.IConversionOptions} [options] Conversion options
                 * @returns ***REMOVED***Object.<string,*>} Plain object
                 */
                Http.toObject = function toObject(message, options) ***REMOVED***
                    if (!options)
                        options = ***REMOVED***};
                    var object = ***REMOVED***};
                    if (options.arrays || options.defaults)
                        object.rules = [];
                    if (options.defaults)
                        object.fully_decode_reserved_expansion = false;
                    if (message.rules && message.rules.length) ***REMOVED***
                        object.rules = [];
                        for (var j = 0; j < message.rules.length; ++j)
                            object.rules[j] = $root.google.api.HttpRule.toObject(message.rules[j], options);
                  ***REMOVED***
                    if (message.fully_decode_reserved_expansion != null && message.hasOwnProperty("fully_decode_reserved_expansion"))
                        object.fully_decode_reserved_expansion = message.fully_decode_reserved_expansion;
                    return object;
              ***REMOVED***;
    
                /**
                 * Converts this Http to JSON.
                 * @function toJSON
                 * @memberof google.api.Http
                 * @instance
                 * @returns ***REMOVED***Object.<string,*>} JSON object
                 */
                Http.prototype.toJSON = function toJSON() ***REMOVED***
                    return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
              ***REMOVED***;
    
                return Http;
          ***REMOVED***)();
    
            api.HttpRule = (function() ***REMOVED***
    
                /**
                 * Properties of a HttpRule.
                 * @memberof google.api
                 * @interface IHttpRule
                 * @property ***REMOVED***string|null} [selector] HttpRule selector
                 * @property ***REMOVED***string|null} [get] HttpRule get
                 * @property ***REMOVED***string|null} [put] HttpRule put
                 * @property ***REMOVED***string|null} [post] HttpRule post
                 * @property ***REMOVED***string|null} ["delete"] HttpRule delete
                 * @property ***REMOVED***string|null} [patch] HttpRule patch
                 * @property ***REMOVED***google.api.ICustomHttpPattern|null} [custom] HttpRule custom
                 * @property ***REMOVED***string|null} [body] HttpRule body
                 * @property ***REMOVED***string|null} [response_body] HttpRule response_body
                 * @property ***REMOVED***Array.<google.api.IHttpRule>|null} [additional_bindings] HttpRule additional_bindings
                 */
    
                /**
                 * Constructs a new HttpRule.
                 * @memberof google.api
                 * @classdesc Represents a HttpRule.
                 * @implements IHttpRule
                 * @constructor
                 * @param ***REMOVED***google.api.IHttpRule=} [properties] Properties to set
                 */
                function HttpRule(properties) ***REMOVED***
                    this.additional_bindings = [];
                    if (properties)
                        for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                            if (properties[keys[i]] != null)
                                this[keys[i]] = properties[keys[i]];
              ***REMOVED***
    
                /**
                 * HttpRule selector.
                 * @member ***REMOVED***string} selector
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                HttpRule.prototype.selector = "";
    
                /**
                 * HttpRule get.
                 * @member ***REMOVED***string} get
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                HttpRule.prototype.get = "";
    
                /**
                 * HttpRule put.
                 * @member ***REMOVED***string} put
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                HttpRule.prototype.put = "";
    
                /**
                 * HttpRule post.
                 * @member ***REMOVED***string} post
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                HttpRule.prototype.post = "";
    
                /**
                 * HttpRule delete.
                 * @member ***REMOVED***string} delete
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                HttpRule.prototype["delete"] = "";
    
                /**
                 * HttpRule patch.
                 * @member ***REMOVED***string} patch
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                HttpRule.prototype.patch = "";
    
                /**
                 * HttpRule custom.
                 * @member ***REMOVED***google.api.ICustomHttpPattern|null|undefined} custom
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                HttpRule.prototype.custom = null;
    
                /**
                 * HttpRule body.
                 * @member ***REMOVED***string} body
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                HttpRule.prototype.body = "";
    
                /**
                 * HttpRule response_body.
                 * @member ***REMOVED***string} response_body
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                HttpRule.prototype.response_body = "";
    
                /**
                 * HttpRule additional_bindings.
                 * @member ***REMOVED***Array.<google.api.IHttpRule>} additional_bindings
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                HttpRule.prototype.additional_bindings = $util.emptyArray;
    
                // OneOf field names bound to virtual getters and setters
                var $oneOfFields;
    
                /**
                 * HttpRule pattern.
                 * @member ***REMOVED***"get"|"put"|"post"|"delete"|"patch"|"custom"|undefined} pattern
                 * @memberof google.api.HttpRule
                 * @instance
                 */
                Object.defineProperty(HttpRule.prototype, "pattern", ***REMOVED***
                    get: $util.oneOfGetter($oneOfFields = ["get", "put", "post", "delete", "patch", "custom"]),
                    set: $util.oneOfSetter($oneOfFields)
              ***REMOVED***);
    
                /**
                 * Creates a new HttpRule instance using the specified properties.
                 * @function create
                 * @memberof google.api.HttpRule
                 * @static
                 * @param ***REMOVED***google.api.IHttpRule=} [properties] Properties to set
                 * @returns ***REMOVED***google.api.HttpRule} HttpRule instance
                 */
                HttpRule.create = function create(properties) ***REMOVED***
                    return new HttpRule(properties);
              ***REMOVED***;
    
                /**
                 * Encodes the specified HttpRule message. Does not implicitly ***REMOVED***@link google.api.HttpRule.verify|verify} messages.
                 * @function encode
                 * @memberof google.api.HttpRule
                 * @static
                 * @param ***REMOVED***google.api.IHttpRule} message HttpRule message or plain object to encode
                 * @param ***REMOVED***$protobuf.Writer} [writer] Writer to encode to
                 * @returns ***REMOVED***$protobuf.Writer} Writer
                 */
                HttpRule.encode = function encode(message, writer) ***REMOVED***
                    if (!writer)
                        writer = $Writer.create();
                    if (message.selector != null && Object.hasOwnProperty.call(message, "selector"))
                        writer.uint32(/* id 1, wireType 2 =*/10).string(message.selector);
                    if (message.get != null && Object.hasOwnProperty.call(message, "get"))
                        writer.uint32(/* id 2, wireType 2 =*/18).string(message.get);
                    if (message.put != null && Object.hasOwnProperty.call(message, "put"))
                        writer.uint32(/* id 3, wireType 2 =*/26).string(message.put);
                    if (message.post != null && Object.hasOwnProperty.call(message, "post"))
                        writer.uint32(/* id 4, wireType 2 =*/34).string(message.post);
                    if (message["delete"] != null && Object.hasOwnProperty.call(message, "delete"))
                        writer.uint32(/* id 5, wireType 2 =*/42).string(message["delete"]);
                    if (message.patch != null && Object.hasOwnProperty.call(message, "patch"))
                        writer.uint32(/* id 6, wireType 2 =*/50).string(message.patch);
                    if (message.body != null && Object.hasOwnProperty.call(message, "body"))
                        writer.uint32(/* id 7, wireType 2 =*/58).string(message.body);
                    if (message.custom != null && Object.hasOwnProperty.call(message, "custom"))
                        $root.google.api.CustomHttpPattern.encode(message.custom, writer.uint32(/* id 8, wireType 2 =*/66).fork()).ldelim();
                    if (message.additional_bindings != null && message.additional_bindings.length)
                        for (var i = 0; i < message.additional_bindings.length; ++i)
                            $root.google.api.HttpRule.encode(message.additional_bindings[i], writer.uint32(/* id 11, wireType 2 =*/90).fork()).ldelim();
                    if (message.response_body != null && Object.hasOwnProperty.call(message, "response_body"))
                        writer.uint32(/* id 12, wireType 2 =*/98).string(message.response_body);
                    return writer;
              ***REMOVED***;
    
                /**
                 * Encodes the specified HttpRule message, length delimited. Does not implicitly ***REMOVED***@link google.api.HttpRule.verify|verify} messages.
                 * @function encodeDelimited
                 * @memberof google.api.HttpRule
                 * @static
                 * @param ***REMOVED***google.api.IHttpRule} message HttpRule message or plain object to encode
                 * @param ***REMOVED***$protobuf.Writer} [writer] Writer to encode to
                 * @returns ***REMOVED***$protobuf.Writer} Writer
                 */
                HttpRule.encodeDelimited = function encodeDelimited(message, writer) ***REMOVED***
                    return this.encode(message, writer).ldelim();
              ***REMOVED***;
    
                /**
                 * Decodes a HttpRule message from the specified reader or buffer.
                 * @function decode
                 * @memberof google.api.HttpRule
                 * @static
                 * @param ***REMOVED***$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
                 * @param ***REMOVED***number} [length] Message length if known beforehand
                 * @returns ***REMOVED***google.api.HttpRule} HttpRule
                 * @throws ***REMOVED***Error} If the payload is not a reader or valid buffer
                 * @throws ***REMOVED***$protobuf.util.ProtocolError} If required fields are missing
                 */
                HttpRule.decode = function decode(reader, length) ***REMOVED***
                    if (!(reader instanceof $Reader))
                        reader = $Reader.create(reader);
                    var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.api.HttpRule();
                    while (reader.pos < end) ***REMOVED***
                        var tag = reader.uint32();
                        switch (tag >>> 3) ***REMOVED***
                        case 1:
                            message.selector = reader.string();
                            break;
                        case 2:
                            message.get = reader.string();
                            break;
                        case 3:
                            message.put = reader.string();
                            break;
                        case 4:
                            message.post = reader.string();
                            break;
                        case 5:
                            message["delete"] = reader.string();
                            break;
                        case 6:
                            message.patch = reader.string();
                            break;
                        case 8:
                            message.custom = $root.google.api.CustomHttpPattern.decode(reader, reader.uint32());
                            break;
                        case 7:
                            message.body = reader.string();
                            break;
                        case 12:
                            message.response_body = reader.string();
                            break;
                        case 11:
                            if (!(message.additional_bindings && message.additional_bindings.length))
                                message.additional_bindings = [];
                            message.additional_bindings.push($root.google.api.HttpRule.decode(reader, reader.uint32()));
                            break;
                        default:
                            reader.skipType(tag & 7);
                            break;
                      ***REMOVED***
                  ***REMOVED***
                    return message;
              ***REMOVED***;
    
                /**
                 * Decodes a HttpRule message from the specified reader or buffer, length delimited.
                 * @function decodeDelimited
                 * @memberof google.api.HttpRule
                 * @static
                 * @param ***REMOVED***$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
                 * @returns ***REMOVED***google.api.HttpRule} HttpRule
                 * @throws ***REMOVED***Error} If the payload is not a reader or valid buffer
                 * @throws ***REMOVED***$protobuf.util.ProtocolError} If required fields are missing
                 */
                HttpRule.decodeDelimited = function decodeDelimited(reader) ***REMOVED***
                    if (!(reader instanceof $Reader))
                        reader = new $Reader(reader);
                    return this.decode(reader, reader.uint32());
              ***REMOVED***;
    
                /**
                 * Verifies a HttpRule message.
                 * @function verify
                 * @memberof google.api.HttpRule
                 * @static
                 * @param ***REMOVED***Object.<string,*>} message Plain object to verify
                 * @returns ***REMOVED***string|null} `null` if valid, otherwise the reason why it is not
                 */
                HttpRule.verify = function verify(message) ***REMOVED***
                    if (typeof message !== "object" || message === null)
                        return "object expected";
                    var properties = ***REMOVED***};
                    if (message.selector != null && message.hasOwnProperty("selector"))
                        if (!$util.isString(message.selector))
                            return "selector: string expected";
                    if (message.get != null && message.hasOwnProperty("get")) ***REMOVED***
                        properties.pattern = 1;
                        if (!$util.isString(message.get))
                            return "get: string expected";
                  ***REMOVED***
                    if (message.put != null && message.hasOwnProperty("put")) ***REMOVED***
                        if (properties.pattern === 1)
                            return "pattern: multiple values";
                        properties.pattern = 1;
                        if (!$util.isString(message.put))
                            return "put: string expected";
                  ***REMOVED***
                    if (message.post != null && message.hasOwnProperty("post")) ***REMOVED***
                        if (properties.pattern === 1)
                            return "pattern: multiple values";
                        properties.pattern = 1;
                        if (!$util.isString(message.post))
                            return "post: string expected";
                  ***REMOVED***
                    if (message["delete"] != null && message.hasOwnProperty("delete")) ***REMOVED***
                        if (properties.pattern === 1)
                            return "pattern: multiple values";
                        properties.pattern = 1;
                        if (!$util.isString(message["delete"]))
                            return "delete: string expected";
                  ***REMOVED***
                    if (message.patch != null && message.hasOwnProperty("patch")) ***REMOVED***
                        if (properties.pattern === 1)
                            return "pattern: multiple values";
                        properties.pattern = 1;
                        if (!$util.isString(message.patch))
                            return "patch: string expected";
                  ***REMOVED***
                    if (message.custom != null && message.hasOwnProperty("custom")) ***REMOVED***
                        if (properties.pattern === 1)
                            return "pattern: multiple values";
                        properties.pattern = 1;
                        ***REMOVED***
                            var error = $root.google.api.CustomHttpPattern.verify(message.custom);
                            if (error)
                                return "custom." + error;
                      ***REMOVED***
                  ***REMOVED***
                    if (message.body != null && message.hasOwnProperty("body"))
                        if (!$util.isString(message.body))
                            return "body: string expected";
                    if (message.response_body != null && message.hasOwnProperty("response_body"))
                        if (!$util.isString(message.response_body))
                            return "response_body: string expected";
                    if (message.additional_bindings != null && message.hasOwnProperty("additional_bindings")) ***REMOVED***
                        if (!Array.isArray(message.additional_bindings))
                            return "additional_bindings: array expected";
                        for (var i = 0; i < message.additional_bindings.length; ++i) ***REMOVED***
                            var error = $root.google.api.HttpRule.verify(message.additional_bindings[i]);
                            if (error)
                                return "additional_bindings." + error;
                      ***REMOVED***
                  ***REMOVED***
                    return null;
              ***REMOVED***;
    
                /**
                 * Creates a HttpRule message from a plain object. Also converts values to their respective internal types.
                 * @function fromObject
                 * @memberof google.api.HttpRule
                 * @static
                 * @param ***REMOVED***Object.<string,*>} object Plain object
                 * @returns ***REMOVED***google.api.HttpRule} HttpRule
                 */
                HttpRule.fromObject = function fromObject(object) ***REMOVED***
                    if (object instanceof $root.google.api.HttpRule)
                        return object;
                    var message = new $root.google.api.HttpRule();
                    if (object.selector != null)
                        message.selector = String(object.selector);
                    if (object.get != null)
                        message.get = String(object.get);
                    if (object.put != null)
                        message.put = String(object.put);
                    if (object.post != null)
                        message.post = String(object.post);
                    if (object["delete"] != null)
                        message["delete"] = String(object["delete"]);
                    if (object.patch != null)
                        message.patch = String(object.patch);
                    if (object.custom != null) ***REMOVED***
                        if (typeof object.custom !== "object")
                            throw TypeError(".google.api.HttpRule.custom: object expected");
                        message.custom = $root.google.api.CustomHttpPattern.fromObject(object.custom);
                  ***REMOVED***
                    if (object.body != null)
                        message.body = String(object.body);
                    if (object.response_body != null)
                        message.response_body = String(object.response_body);
                    if (object.additional_bindings) ***REMOVED***
                        if (!Array.isArray(object.additional_bindings))
                            throw TypeError(".google.api.HttpRule.additional_bindings: array expected");
                        message.additional_bindings = [];
                        for (var i = 0; i < object.additional_bindings.length; ++i) ***REMOVED***
                            if (typeof object.additional_bindings[i] !== "object")
                                throw TypeError(".google.api.HttpRule.additional_bindings: object expected");
                            message.additional_bindings[i] = $root.google.api.HttpRule.fromObject(object.additional_bindings[i]);
                      ***REMOVED***
                  ***REMOVED***
                    return message;
              ***REMOVED***;
    
                /**
                 * Creates a plain object from a HttpRule message. Also converts values to other types if specified.
                 * @function toObject
                 * @memberof google.api.HttpRule
                 * @static
                 * @param ***REMOVED***google.api.HttpRule} message HttpRule
                 * @param ***REMOVED***$protobuf.IConversionOptions} [options] Conversion options
                 * @returns ***REMOVED***Object.<string,*>} Plain object
                 */
                HttpRule.toObject = function toObject(message, options) ***REMOVED***
                    if (!options)
                        options = ***REMOVED***};
                    var object = ***REMOVED***};
                    if (options.arrays || options.defaults)
                        object.additional_bindings = [];
                    if (options.defaults) ***REMOVED***
                        object.selector = "";
                        object.body = "";
                        object.response_body = "";
                  ***REMOVED***
                    if (message.selector != null && message.hasOwnProperty("selector"))
                        object.selector = message.selector;
                    if (message.get != null && message.hasOwnProperty("get")) ***REMOVED***
                        object.get = message.get;
                        if (options.oneofs)
                            object.pattern = "get";
                  ***REMOVED***
                    if (message.put != null && message.hasOwnProperty("put")) ***REMOVED***
                        object.put = message.put;
                        if (options.oneofs)
                            object.pattern = "put";
                  ***REMOVED***
                    if (message.post != null && message.hasOwnProperty("post")) ***REMOVED***
                        object.post = message.post;
                        if (options.oneofs)
                            object.pattern = "post";
                  ***REMOVED***
                    if (message["delete"] != null && message.hasOwnProperty("delete")) ***REMOVED***
                        object["delete"] = message["delete"];
                        if (options.oneofs)
                            object.pattern = "delete";
                  ***REMOVED***
                    if (message.patch != null && message.hasOwnProperty("patch")) ***REMOVED***
                        object.patch = message.patch;
                        if (options.oneofs)
                            object.pattern = "patch";
                  ***REMOVED***
                    if (message.body != null && message.hasOwnProperty("body"))
                        object.body = message.body;
                    if (message.custom != null && message.hasOwnProperty("custom")) ***REMOVED***
                        object.custom = $root.google.api.CustomHttpPattern.toObject(message.custom, options);
                        if (options.oneofs)
                            object.pattern = "custom";
                  ***REMOVED***
                    if (message.additional_bindings && message.additional_bindings.length) ***REMOVED***
                        object.additional_bindings = [];
                        for (var j = 0; j < message.additional_bindings.length; ++j)
                            object.additional_bindings[j] = $root.google.api.HttpRule.toObject(message.additional_bindings[j], options);
                  ***REMOVED***
                    if (message.response_body != null && message.hasOwnProperty("response_body"))
                        object.response_body = message.response_body;
                    return object;
              ***REMOVED***;
    
                /**
                 * Converts this HttpRule to JSON.
                 * @function toJSON
                 * @memberof google.api.HttpRule
                 * @instance
                 * @returns ***REMOVED***Object.<string,*>} JSON object
                 */
                HttpRule.prototype.toJSON = function toJSON() ***REMOVED***
                    return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
              ***REMOVED***;
    
                return HttpRule;
          ***REMOVED***)();
    
            api.CustomHttpPattern = (function() ***REMOVED***
    
                /**
                 * Properties of a CustomHttpPattern.
                 * @memberof google.api
                 * @interface ICustomHttpPattern
                 * @property ***REMOVED***string|null} [kind] CustomHttpPattern kind
                 * @property ***REMOVED***string|null} [path] CustomHttpPattern path
                 */
    
                /**
                 * Constructs a new CustomHttpPattern.
                 * @memberof google.api
                 * @classdesc Represents a CustomHttpPattern.
                 * @implements ICustomHttpPattern
                 * @constructor
                 * @param ***REMOVED***google.api.ICustomHttpPattern=} [properties] Properties to set
                 */
                function CustomHttpPattern(properties) ***REMOVED***
                    if (properties)
                        for (var keys = Object.keys(properties), i = 0; i < keys.length; ++i)
                            if (properties[keys[i]] != null)
                                this[keys[i]] = properties[keys[i]];
              ***REMOVED***
    
                /**
                 * CustomHttpPattern kind.
                 * @member ***REMOVED***string} kind
                 * @memberof google.api.CustomHttpPattern
                 * @instance
                 */
                CustomHttpPattern.prototype.kind = "";
    
                /**
                 * CustomHttpPattern path.
                 * @member ***REMOVED***string} path
                 * @memberof google.api.CustomHttpPattern
                 * @instance
                 */
                CustomHttpPattern.prototype.path = "";
    
                /**
                 * Creates a new CustomHttpPattern instance using the specified properties.
                 * @function create
                 * @memberof google.api.CustomHttpPattern
                 * @static
                 * @param ***REMOVED***google.api.ICustomHttpPattern=} [properties] Properties to set
                 * @returns ***REMOVED***google.api.CustomHttpPattern} CustomHttpPattern instance
                 */
                CustomHttpPattern.create = function create(properties) ***REMOVED***
                    return new CustomHttpPattern(properties);
              ***REMOVED***;
    
                /**
                 * Encodes the specified CustomHttpPattern message. Does not implicitly ***REMOVED***@link google.api.CustomHttpPattern.verify|verify} messages.
                 * @function encode
                 * @memberof google.api.CustomHttpPattern
                 * @static
                 * @param ***REMOVED***google.api.ICustomHttpPattern} message CustomHttpPattern message or plain object to encode
                 * @param ***REMOVED***$protobuf.Writer} [writer] Writer to encode to
                 * @returns ***REMOVED***$protobuf.Writer} Writer
                 */
                CustomHttpPattern.encode = function encode(message, writer) ***REMOVED***
                    if (!writer)
                        writer = $Writer.create();
                    if (message.kind != null && Object.hasOwnProperty.call(message, "kind"))
                        writer.uint32(/* id 1, wireType 2 =*/10).string(message.kind);
                    if (message.path != null && Object.hasOwnProperty.call(message, "path"))
                        writer.uint32(/* id 2, wireType 2 =*/18).string(message.path);
                    return writer;
              ***REMOVED***;
    
                /**
                 * Encodes the specified CustomHttpPattern message, length delimited. Does not implicitly ***REMOVED***@link google.api.CustomHttpPattern.verify|verify} messages.
                 * @function encodeDelimited
                 * @memberof google.api.CustomHttpPattern
                 * @static
                 * @param ***REMOVED***google.api.ICustomHttpPattern} message CustomHttpPattern message or plain object to encode
                 * @param ***REMOVED***$protobuf.Writer} [writer] Writer to encode to
                 * @returns ***REMOVED***$protobuf.Writer} Writer
                 */
                CustomHttpPattern.encodeDelimited = function encodeDelimited(message, writer) ***REMOVED***
                    return this.encode(message, writer).ldelim();
              ***REMOVED***;
    
                /**
                 * Decodes a CustomHttpPattern message from the specified reader or buffer.
                 * @function decode
                 * @memberof google.api.CustomHttpPattern
                 * @static
                 * @param ***REMOVED***$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
                 * @param ***REMOVED***number} [length] Message length if known beforehand
                 * @returns ***REMOVED***google.api.CustomHttpPattern} CustomHttpPattern
                 * @throws ***REMOVED***Error} If the payload is not a reader or valid buffer
                 * @throws ***REMOVED***$protobuf.util.ProtocolError} If required fields are missing
                 */
                CustomHttpPattern.decode = function decode(reader, length) ***REMOVED***
                    if (!(reader instanceof $Reader))
                        reader = $Reader.create(reader);
                    var end = length === undefined ? reader.len : reader.pos + length, message = new $root.google.api.CustomHttpPattern();
                    while (reader.pos < end) ***REMOVED***
                        var tag = reader.uint32();
                        switch (tag >>> 3) ***REMOVED***
                        case 1:
                            message.kind = reader.string();
                            break;
                        case 2:
                            message.path = reader.string();
                            break;
                        default:
                            reader.skipType(tag & 7);
                            break;
                      ***REMOVED***
                  ***REMOVED***
                    return message;
              ***REMOVED***;
    
                /**
                 * Decodes a CustomHttpPattern message from the specified reader or buffer, length delimited.
                 * @function decodeDelimited
                 * @memberof google.api.CustomHttpPattern
                 * @static
                 * @param ***REMOVED***$protobuf.Reader|Uint8Array} reader Reader or buffer to decode from
                 * @returns ***REMOVED***google.api.CustomHttpPattern} CustomHttpPattern
                 * @throws ***REMOVED***Error} If the payload is not a reader or valid buffer
                 * @throws ***REMOVED***$protobuf.util.ProtocolError} If required fields are missing
                 */
                CustomHttpPattern.decodeDelimited = function decodeDelimited(reader) ***REMOVED***
                    if (!(reader instanceof $Reader))
                        reader = new $Reader(reader);
                    return this.decode(reader, reader.uint32());
              ***REMOVED***;
    
                /**
                 * Verifies a CustomHttpPattern message.
                 * @function verify
                 * @memberof google.api.CustomHttpPattern
                 * @static
                 * @param ***REMOVED***Object.<string,*>} message Plain object to verify
                 * @returns ***REMOVED***string|null} `null` if valid, otherwise the reason why it is not
                 */
                CustomHttpPattern.verify = function verify(message) ***REMOVED***
                    if (typeof message !== "object" || message === null)
                        return "object expected";
                    if (message.kind != null && message.hasOwnProperty("kind"))
                        if (!$util.isString(message.kind))
                            return "kind: string expected";
                    if (message.path != null && message.hasOwnProperty("path"))
                        if (!$util.isString(message.path))
                            return "path: string expected";
                    return null;
              ***REMOVED***;
    
                /**
                 * Creates a CustomHttpPattern message from a plain object. Also converts values to their respective internal types.
                 * @function fromObject
                 * @memberof google.api.CustomHttpPattern
                 * @static
                 * @param ***REMOVED***Object.<string,*>} object Plain object
                 * @returns ***REMOVED***google.api.CustomHttpPattern} CustomHttpPattern
                 */
                CustomHttpPattern.fromObject = function fromObject(object) ***REMOVED***
                    if (object instanceof $root.google.api.CustomHttpPattern)
                        return object;
                    var message = new $root.google.api.CustomHttpPattern();
                    if (object.kind != null)
                        message.kind = String(object.kind);
                    if (object.path != null)
                        message.path = String(object.path);
                    return message;
              ***REMOVED***;
    
                /**
                 * Creates a plain object from a CustomHttpPattern message. Also converts values to other types if specified.
                 * @function toObject
                 * @memberof google.api.CustomHttpPattern
                 * @static
                 * @param ***REMOVED***google.api.CustomHttpPattern} message CustomHttpPattern
                 * @param ***REMOVED***$protobuf.IConversionOptions} [options] Conversion options
                 * @returns ***REMOVED***Object.<string,*>} Plain object
                 */
                CustomHttpPattern.toObject = function toObject(message, options) ***REMOVED***
                    if (!options)
                        options = ***REMOVED***};
                    var object = ***REMOVED***};
                    if (options.defaults) ***REMOVED***
                        object.kind = "";
                        object.path = "";
                  ***REMOVED***
                    if (message.kind != null && message.hasOwnProperty("kind"))
                        object.kind = message.kind;
                    if (message.path != null && message.hasOwnProperty("path"))
                        object.path = message.path;
                    return object;
              ***REMOVED***;
    
                /**
                 * Converts this CustomHttpPattern to JSON.
                 * @function toJSON
                 * @memberof google.api.CustomHttpPattern
                 * @instance
                 * @returns ***REMOVED***Object.<string,*>} JSON object
                 */
                CustomHttpPattern.prototype.toJSON = function toJSON() ***REMOVED***
                    return this.constructor.toObject(this, $protobuf.util.toJSONOptions);
              ***REMOVED***;
    
                return CustomHttpPattern;
          ***REMOVED***)();
    
            return api;
      ***REMOVED***)();
    
        return google;
  ***REMOVED***)();

    return $root;
});
