// yargs/lib/platform-shims/esm.mjs
'use strict'

import ***REMOVED*** notStrictEqual, strictEqual } from 'assert'
import cliui from 'cliui'
import escalade from 'escalade/sync'
import ***REMOVED*** format, inspect } from 'util'
import ***REMOVED*** readFileSync } from 'fs'
import ***REMOVED*** fileURLToPath } from 'url';
import Parser from 'yargs-parser'
import ***REMOVED*** basename, dirname, extname, relative, resolve } from 'path'
import ***REMOVED*** getProcessArgvBin } from '../../build/lib/utils/process-argv.js'
import ***REMOVED*** YError } from '../../build/lib/yerror.js'
import y18n from 'y18n'

const REQUIRE_ERROR = 'require is not supported by ESM'
const REQUIRE_DIRECTORY_ERROR = 'loading a directory of commands is not supported yet for ESM'

const mainFilename = fileURLToPath(import.meta.url).split('node_modules')[0]
const __dirname = fileURLToPath(import.meta.url)

export default ***REMOVED***
  assert: ***REMOVED***
    notStrictEqual,
    strictEqual
***REMOVED***,
  cliui,
  findUp: escalade,
  getEnv: (key) => ***REMOVED***
    return process.env[key]
***REMOVED***,
  inspect,
  getCallerFile: () => ***REMOVED***
    throw new YError(REQUIRE_DIRECTORY_ERROR)
***REMOVED***,
  getProcessArgvBin,
  mainFilename: mainFilename || process.cwd(),
  Parser,
  path: ***REMOVED***
    basename,
    dirname,
    extname,
    relative,
    resolve
***REMOVED***,
  process: ***REMOVED***
    argv: () => process.argv,
    cwd: process.cwd,
    execPath: () => process.execPath,
    exit: process.exit,
    nextTick: process.nextTick,
    stdColumns: typeof process.stdout.columns !== 'undefined' ? process.stdout.columns : null
***REMOVED***,
  readFileSync,
  require: () => ***REMOVED***
    throw new YError(REQUIRE_ERROR)
***REMOVED***,
  requireDirectory: () => ***REMOVED***
    throw new YError(REQUIRE_DIRECTORY_ERROR)
***REMOVED***,
  stringWidth: (str) => ***REMOVED***
    return [...str].length
***REMOVED***,
  y18n: y18n(***REMOVED***
    directory: resolve(__dirname, '../../../locales'),
    updateFiles: false
***REMOVED***)
}
