// scheduler/cjs/scheduler-unstable_mock.development.js
/**
 * @license React
 * scheduler-unstable_mock.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

if (process.env.NODE_ENV !== "production") ***REMOVED***
  (function() ***REMOVED***
'use strict';

var enableSchedulerDebugging = false;
var enableProfiling = false;

function push(heap, node) ***REMOVED***
  var index = heap.length;
  heap.push(node);
  siftUp(heap, node, index);
}
function peek(heap) ***REMOVED***
  return heap.length === 0 ? null : heap[0];
}
function pop(heap) ***REMOVED***
  if (heap.length === 0) ***REMOVED***
    return null;
***REMOVED***

  var first = heap[0];
  var last = heap.pop();

  if (last !== first) ***REMOVED***
    heap[0] = last;
    siftDown(heap, last, 0);
***REMOVED***

  return first;
}

function siftUp(heap, node, i) ***REMOVED***
  var index = i;

  while (index > 0) ***REMOVED***
    var parentIndex = index - 1 >>> 1;
    var parent = heap[parentIndex];

    if (compare(parent, node) > 0) ***REMOVED***
      // The parent is larger. Swap positions.
      heap[parentIndex] = node;
      heap[index] = parent;
      index = parentIndex;
  ***REMOVED*** else ***REMOVED***
      // The parent is smaller. Exit.
      return;
  ***REMOVED***
***REMOVED***
}

function siftDown(heap, node, i) ***REMOVED***
  var index = i;
  var length = heap.length;
  var halfLength = length >>> 1;

  while (index < halfLength) ***REMOVED***
    var leftIndex = (index + 1) * 2 - 1;
    var left = heap[leftIndex];
    var rightIndex = leftIndex + 1;
    var right = heap[rightIndex]; // If the left or right node is smaller, swap with the smaller of those.

    if (compare(left, node) < 0) ***REMOVED***
      if (rightIndex < length && compare(right, left) < 0) ***REMOVED***
        heap[index] = right;
        heap[rightIndex] = node;
        index = rightIndex;
    ***REMOVED*** else ***REMOVED***
        heap[index] = left;
        heap[leftIndex] = node;
        index = leftIndex;
    ***REMOVED***
  ***REMOVED*** else if (rightIndex < length && compare(right, node) < 0) ***REMOVED***
      heap[index] = right;
      heap[rightIndex] = node;
      index = rightIndex;
  ***REMOVED*** else ***REMOVED***
      // Neither child is smaller. Exit.
      return;
  ***REMOVED***
***REMOVED***
}

function compare(a, b) ***REMOVED***
  // Compare sort index first, then task id.
  var diff = a.sortIndex - b.sortIndex;
  return diff !== 0 ? diff : a.id - b.id;
}

// TODO: Use symbols?
var ImmediatePriority = 1;
var UserBlockingPriority = 2;
var NormalPriority = 3;
var LowPriority = 4;
var IdlePriority = 5;

function markTaskErrored(task, ms) ***REMOVED***
}

/* eslint-disable no-var */
// Math.pow(2, 30) - 1
// 0b111111111111111111111111111111

var maxSigned31BitInt = 1073741823; // Times out immediately

var IMMEDIATE_PRIORITY_TIMEOUT = -1; // Eventually times out

var USER_BLOCKING_PRIORITY_TIMEOUT = 250;
var NORMAL_PRIORITY_TIMEOUT = 5000;
var LOW_PRIORITY_TIMEOUT = 10000; // Never times out

var IDLE_PRIORITY_TIMEOUT = maxSigned31BitInt; // Tasks are stored on a min heap

var taskQueue = [];
var timerQueue = []; // Incrementing id counter. Used to maintain insertion order.

var taskIdCounter = 1; // Pausing the scheduler is useful for debugging.
var currentTask = null;
var currentPriorityLevel = NormalPriority; // This is set while performing work, to prevent re-entrance.

var isPerformingWork = false;
var isHostCallbackScheduled = false;
var isHostTimeoutScheduled = false;
var currentMockTime = 0;
var scheduledCallback = null;
var scheduledTimeout = null;
var timeoutTime = -1;
var yieldedValues = null;
var expectedNumberOfYields = -1;
var didStop = false;
var isFlushing = false;
var needsPaint = false;
var shouldYieldForPaint = false;
var disableYieldValue = false;

function setDisableYieldValue(newValue) ***REMOVED***
  disableYieldValue = newValue;
}

function advanceTimers(currentTime) ***REMOVED***
  // Check for tasks that are no longer delayed and add them to the queue.
  var timer = peek(timerQueue);

  while (timer !== null) ***REMOVED***
    if (timer.callback === null) ***REMOVED***
      // Timer was cancelled.
      pop(timerQueue);
  ***REMOVED*** else if (timer.startTime <= currentTime) ***REMOVED***
      // Timer fired. Transfer to the task queue.
      pop(timerQueue);
      timer.sortIndex = timer.expirationTime;
      push(taskQueue, timer);
  ***REMOVED*** else ***REMOVED***
      // Remaining timers are pending.
      return;
  ***REMOVED***

    timer = peek(timerQueue);
***REMOVED***
}

function handleTimeout(currentTime) ***REMOVED***
  isHostTimeoutScheduled = false;
  advanceTimers(currentTime);

  if (!isHostCallbackScheduled) ***REMOVED***
    if (peek(taskQueue) !== null) ***REMOVED***
      isHostCallbackScheduled = true;
      requestHostCallback(flushWork);
  ***REMOVED*** else ***REMOVED***
      var firstTimer = peek(timerQueue);

      if (firstTimer !== null) ***REMOVED***
        requestHostTimeout(handleTimeout, firstTimer.startTime - currentTime);
    ***REMOVED***
  ***REMOVED***
***REMOVED***
}

function flushWork(hasTimeRemaining, initialTime) ***REMOVED***


  isHostCallbackScheduled = false;

  if (isHostTimeoutScheduled) ***REMOVED***
    // We scheduled a timeout but it's no longer needed. Cancel it.
    isHostTimeoutScheduled = false;
    cancelHostTimeout();
***REMOVED***

  isPerformingWork = true;
  var previousPriorityLevel = currentPriorityLevel;

  try ***REMOVED***
    if (enableProfiling) ***REMOVED***
      try ***REMOVED***
        return workLoop(hasTimeRemaining, initialTime);
    ***REMOVED*** catch (error) ***REMOVED***
        if (currentTask !== null) ***REMOVED***
          var currentTime = getCurrentTime();
          markTaskErrored(currentTask, currentTime);
          currentTask.isQueued = false;
      ***REMOVED***

        throw error;
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
      // No catch in prod code path.
      return workLoop(hasTimeRemaining, initialTime);
  ***REMOVED***
***REMOVED*** finally ***REMOVED***
    currentTask = null;
    currentPriorityLevel = previousPriorityLevel;
    isPerformingWork = false;
***REMOVED***
}

function workLoop(hasTimeRemaining, initialTime) ***REMOVED***
  var currentTime = initialTime;
  advanceTimers(currentTime);
  currentTask = peek(taskQueue);

  while (currentTask !== null && !(enableSchedulerDebugging )) ***REMOVED***
    if (currentTask.expirationTime > currentTime && (!hasTimeRemaining || shouldYieldToHost())) ***REMOVED***
      // This currentTask hasn't expired, and we've reached the deadline.
      break;
  ***REMOVED***

    var callback = currentTask.callback;

    if (typeof callback === 'function') ***REMOVED***
      currentTask.callback = null;
      currentPriorityLevel = currentTask.priorityLevel;
      var didUserCallbackTimeout = currentTask.expirationTime <= currentTime;

      var continuationCallback = callback(didUserCallbackTimeout);
      currentTime = getCurrentTime();

      if (typeof continuationCallback === 'function') ***REMOVED***
        currentTask.callback = continuationCallback;
    ***REMOVED*** else ***REMOVED***

        if (currentTask === peek(taskQueue)) ***REMOVED***
          pop(taskQueue);
      ***REMOVED***
    ***REMOVED***

      advanceTimers(currentTime);
  ***REMOVED*** else ***REMOVED***
      pop(taskQueue);
  ***REMOVED***

    currentTask = peek(taskQueue);
***REMOVED*** // Return whether there's additional work


  if (currentTask !== null) ***REMOVED***
    return true;
***REMOVED*** else ***REMOVED***
    var firstTimer = peek(timerQueue);

    if (firstTimer !== null) ***REMOVED***
      requestHostTimeout(handleTimeout, firstTimer.startTime - currentTime);
  ***REMOVED***

    return false;
***REMOVED***
}

function unstable_runWithPriority(priorityLevel, eventHandler) ***REMOVED***
  switch (priorityLevel) ***REMOVED***
    case ImmediatePriority:
    case UserBlockingPriority:
    case NormalPriority:
    case LowPriority:
    case IdlePriority:
      break;

    default:
      priorityLevel = NormalPriority;
***REMOVED***

  var previousPriorityLevel = currentPriorityLevel;
  currentPriorityLevel = priorityLevel;

  try ***REMOVED***
    return eventHandler();
***REMOVED*** finally ***REMOVED***
    currentPriorityLevel = previousPriorityLevel;
***REMOVED***
}

function unstable_next(eventHandler) ***REMOVED***
  var priorityLevel;

  switch (currentPriorityLevel) ***REMOVED***
    case ImmediatePriority:
    case UserBlockingPriority:
    case NormalPriority:
      // Shift down to normal priority
      priorityLevel = NormalPriority;
      break;

    default:
      // Anything lower than normal priority should remain at the current level.
      priorityLevel = currentPriorityLevel;
      break;
***REMOVED***

  var previousPriorityLevel = currentPriorityLevel;
  currentPriorityLevel = priorityLevel;

  try ***REMOVED***
    return eventHandler();
***REMOVED*** finally ***REMOVED***
    currentPriorityLevel = previousPriorityLevel;
***REMOVED***
}

function unstable_wrapCallback(callback) ***REMOVED***
  var parentPriorityLevel = currentPriorityLevel;
  return function () ***REMOVED***
    // This is a fork of runWithPriority, inlined for performance.
    var previousPriorityLevel = currentPriorityLevel;
    currentPriorityLevel = parentPriorityLevel;

    try ***REMOVED***
      return callback.apply(this, arguments);
  ***REMOVED*** finally ***REMOVED***
      currentPriorityLevel = previousPriorityLevel;
  ***REMOVED***
***REMOVED***;
}

function unstable_scheduleCallback(priorityLevel, callback, options) ***REMOVED***
  var currentTime = getCurrentTime();
  var startTime;

  if (typeof options === 'object' && options !== null) ***REMOVED***
    var delay = options.delay;

    if (typeof delay === 'number' && delay > 0) ***REMOVED***
      startTime = currentTime + delay;
  ***REMOVED*** else ***REMOVED***
      startTime = currentTime;
  ***REMOVED***
***REMOVED*** else ***REMOVED***
    startTime = currentTime;
***REMOVED***

  var timeout;

  switch (priorityLevel) ***REMOVED***
    case ImmediatePriority:
      timeout = IMMEDIATE_PRIORITY_TIMEOUT;
      break;

    case UserBlockingPriority:
      timeout = USER_BLOCKING_PRIORITY_TIMEOUT;
      break;

    case IdlePriority:
      timeout = IDLE_PRIORITY_TIMEOUT;
      break;

    case LowPriority:
      timeout = LOW_PRIORITY_TIMEOUT;
      break;

    case NormalPriority:
    default:
      timeout = NORMAL_PRIORITY_TIMEOUT;
      break;
***REMOVED***

  var expirationTime = startTime + timeout;
  var newTask = ***REMOVED***
    id: taskIdCounter++,
    callback: callback,
    priorityLevel: priorityLevel,
    startTime: startTime,
    expirationTime: expirationTime,
    sortIndex: -1
***REMOVED***;

  if (startTime > currentTime) ***REMOVED***
    // This is a delayed task.
    newTask.sortIndex = startTime;
    push(timerQueue, newTask);

    if (peek(taskQueue) === null && newTask === peek(timerQueue)) ***REMOVED***
      // All tasks are delayed, and this is the task with the earliest delay.
      if (isHostTimeoutScheduled) ***REMOVED***
        // Cancel an existing timeout.
        cancelHostTimeout();
    ***REMOVED*** else ***REMOVED***
        isHostTimeoutScheduled = true;
    ***REMOVED*** // Schedule a timeout.


      requestHostTimeout(handleTimeout, startTime - currentTime);
  ***REMOVED***
***REMOVED*** else ***REMOVED***
    newTask.sortIndex = expirationTime;
    push(taskQueue, newTask);
    // wait until the next time we yield.


    if (!isHostCallbackScheduled && !isPerformingWork) ***REMOVED***
      isHostCallbackScheduled = true;
      requestHostCallback(flushWork);
  ***REMOVED***
***REMOVED***

  return newTask;
}

function unstable_pauseExecution() ***REMOVED***
}

function unstable_continueExecution() ***REMOVED***

  if (!isHostCallbackScheduled && !isPerformingWork) ***REMOVED***
    isHostCallbackScheduled = true;
    requestHostCallback(flushWork);
***REMOVED***
}

function unstable_getFirstCallbackNode() ***REMOVED***
  return peek(taskQueue);
}

function unstable_cancelCallback(task) ***REMOVED***
  // remove from the queue because you can't remove arbitrary nodes from an
  // array based heap, only the first one.)


  task.callback = null;
}

function unstable_getCurrentPriorityLevel() ***REMOVED***
  return currentPriorityLevel;
}

function requestHostCallback(callback) ***REMOVED***
  scheduledCallback = callback;
}

function requestHostTimeout(callback, ms) ***REMOVED***
  scheduledTimeout = callback;
  timeoutTime = currentMockTime + ms;
}

function cancelHostTimeout() ***REMOVED***
  scheduledTimeout = null;
  timeoutTime = -1;
}

function shouldYieldToHost() ***REMOVED***
  if (expectedNumberOfYields === 0 && yieldedValues === null || expectedNumberOfYields !== -1 && yieldedValues !== null && yieldedValues.length >= expectedNumberOfYields || shouldYieldForPaint && needsPaint) ***REMOVED***
    // We yielded at least as many values as expected. Stop flushing.
    didStop = true;
    return true;
***REMOVED***

  return false;
}

function getCurrentTime() ***REMOVED***
  return currentMockTime;
}

function forceFrameRate() ***REMOVED***// No-op
}

function reset() ***REMOVED***
  if (isFlushing) ***REMOVED***
    throw new Error('Cannot reset while already flushing work.');
***REMOVED***

  currentMockTime = 0;
  scheduledCallback = null;
  scheduledTimeout = null;
  timeoutTime = -1;
  yieldedValues = null;
  expectedNumberOfYields = -1;
  didStop = false;
  isFlushing = false;
  needsPaint = false;
} // Should only be used via an assertion helper that inspects the yielded values.


function unstable_flushNumberOfYields(count) ***REMOVED***
  if (isFlushing) ***REMOVED***
    throw new Error('Already flushing work.');
***REMOVED***

  if (scheduledCallback !== null) ***REMOVED***
    var cb = scheduledCallback;
    expectedNumberOfYields = count;
    isFlushing = true;

    try ***REMOVED***
      var hasMoreWork = true;

      do ***REMOVED***
        hasMoreWork = cb(true, currentMockTime);
    ***REMOVED*** while (hasMoreWork && !didStop);

      if (!hasMoreWork) ***REMOVED***
        scheduledCallback = null;
    ***REMOVED***
  ***REMOVED*** finally ***REMOVED***
      expectedNumberOfYields = -1;
      didStop = false;
      isFlushing = false;
  ***REMOVED***
***REMOVED***
}

function unstable_flushUntilNextPaint() ***REMOVED***
  if (isFlushing) ***REMOVED***
    throw new Error('Already flushing work.');
***REMOVED***

  if (scheduledCallback !== null) ***REMOVED***
    var cb = scheduledCallback;
    shouldYieldForPaint = true;
    needsPaint = false;
    isFlushing = true;

    try ***REMOVED***
      var hasMoreWork = true;

      do ***REMOVED***
        hasMoreWork = cb(true, currentMockTime);
    ***REMOVED*** while (hasMoreWork && !didStop);

      if (!hasMoreWork) ***REMOVED***
        scheduledCallback = null;
    ***REMOVED***
  ***REMOVED*** finally ***REMOVED***
      shouldYieldForPaint = false;
      didStop = false;
      isFlushing = false;
  ***REMOVED***
***REMOVED***
}

function unstable_flushExpired() ***REMOVED***
  if (isFlushing) ***REMOVED***
    throw new Error('Already flushing work.');
***REMOVED***

  if (scheduledCallback !== null) ***REMOVED***
    isFlushing = true;

    try ***REMOVED***
      var hasMoreWork = scheduledCallback(false, currentMockTime);

      if (!hasMoreWork) ***REMOVED***
        scheduledCallback = null;
    ***REMOVED***
  ***REMOVED*** finally ***REMOVED***
      isFlushing = false;
  ***REMOVED***
***REMOVED***
}

function unstable_flushAllWithoutAsserting() ***REMOVED***
  // Returns false if no work was flushed.
  if (isFlushing) ***REMOVED***
    throw new Error('Already flushing work.');
***REMOVED***

  if (scheduledCallback !== null) ***REMOVED***
    var cb = scheduledCallback;
    isFlushing = true;

    try ***REMOVED***
      var hasMoreWork = true;

      do ***REMOVED***
        hasMoreWork = cb(true, currentMockTime);
    ***REMOVED*** while (hasMoreWork);

      if (!hasMoreWork) ***REMOVED***
        scheduledCallback = null;
    ***REMOVED***

      return true;
  ***REMOVED*** finally ***REMOVED***
      isFlushing = false;
  ***REMOVED***
***REMOVED*** else ***REMOVED***
    return false;
***REMOVED***
}

function unstable_clearYields() ***REMOVED***
  if (yieldedValues === null) ***REMOVED***
    return [];
***REMOVED***

  var values = yieldedValues;
  yieldedValues = null;
  return values;
}

function unstable_flushAll() ***REMOVED***
  if (yieldedValues !== null) ***REMOVED***
    throw new Error('Log is not empty. Assert on the log of yielded values before ' + 'flushing additional work.');
***REMOVED***

  unstable_flushAllWithoutAsserting();

  if (yieldedValues !== null) ***REMOVED***
    throw new Error('While flushing work, something yielded a value. Use an ' + 'assertion helper to assert on the log of yielded values, e.g. ' + 'expect(Scheduler).toFlushAndYield([...])');
***REMOVED***
}

function unstable_yieldValue(value) ***REMOVED***
  // eslint-disable-next-line react-internal/no-production-logging
  if (console.log.name === 'disabledLog' || disableYieldValue) ***REMOVED***
    // If console.log has been patched, we assume we're in render
    // replaying and we ignore any values yielding in the second pass.
    return;
***REMOVED***

  if (yieldedValues === null) ***REMOVED***
    yieldedValues = [value];
***REMOVED*** else ***REMOVED***
    yieldedValues.push(value);
***REMOVED***
}

function unstable_advanceTime(ms) ***REMOVED***
  // eslint-disable-next-line react-internal/no-production-logging
  if (console.log.name === 'disabledLog' || disableYieldValue) ***REMOVED***
    // If console.log has been patched, we assume we're in render
    // replaying and we ignore any time advancing in the second pass.
    return;
***REMOVED***

  currentMockTime += ms;

  if (scheduledTimeout !== null && timeoutTime <= currentMockTime) ***REMOVED***
    scheduledTimeout(currentMockTime);
    timeoutTime = -1;
    scheduledTimeout = null;
***REMOVED***
}

function requestPaint() ***REMOVED***
  needsPaint = true;
}
var unstable_Profiling =  null;

exports.reset = reset;
exports.unstable_IdlePriority = IdlePriority;
exports.unstable_ImmediatePriority = ImmediatePriority;
exports.unstable_LowPriority = LowPriority;
exports.unstable_NormalPriority = NormalPriority;
exports.unstable_Profiling = unstable_Profiling;
exports.unstable_UserBlockingPriority = UserBlockingPriority;
exports.unstable_advanceTime = unstable_advanceTime;
exports.unstable_cancelCallback = unstable_cancelCallback;
exports.unstable_clearYields = unstable_clearYields;
exports.unstable_continueExecution = unstable_continueExecution;
exports.unstable_flushAll = unstable_flushAll;
exports.unstable_flushAllWithoutAsserting = unstable_flushAllWithoutAsserting;
exports.unstable_flushExpired = unstable_flushExpired;
exports.unstable_flushNumberOfYields = unstable_flushNumberOfYields;
exports.unstable_flushUntilNextPaint = unstable_flushUntilNextPaint;
exports.unstable_forceFrameRate = forceFrameRate;
exports.unstable_getCurrentPriorityLevel = unstable_getCurrentPriorityLevel;
exports.unstable_getFirstCallbackNode = unstable_getFirstCallbackNode;
exports.unstable_next = unstable_next;
exports.unstable_now = getCurrentTime;
exports.unstable_pauseExecution = unstable_pauseExecution;
exports.unstable_requestPaint = requestPaint;
exports.unstable_runWithPriority = unstable_runWithPriority;
exports.unstable_scheduleCallback = unstable_scheduleCallback;
exports.unstable_setDisableYieldValue = setDisableYieldValue;
exports.unstable_shouldYield = shouldYieldToHost;
exports.unstable_wrapCallback = unstable_wrapCallback;
exports.unstable_yieldValue = unstable_yieldValue;
***REMOVED***)();
}
