// pg-protocol/dist/inbound-parser.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) ***REMOVED***
    function adopt(value) ***REMOVED*** return value instanceof P ? value : new P(function (resolve) ***REMOVED*** resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) ***REMOVED***
        function fulfilled(value) ***REMOVED*** try ***REMOVED*** step(generator.next(value)); } catch (e) ***REMOVED*** reject(e); } }
        function rejected(value) ***REMOVED*** try ***REMOVED*** step(generator["throw"](value)); } catch (e) ***REMOVED*** reject(e); } }
        function step(result) ***REMOVED*** result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
  ***REMOVED***);
};
var __importDefault = (this && this.__importDefault) || function (mod) ***REMOVED***
    return (mod && mod.__esModule) ? mod : ***REMOVED*** "default": mod };
};
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
const test_buffers_1 = __importDefault(require("./testing/test-buffers"));
const buffer_list_1 = __importDefault(require("./testing/buffer-list"));
const _1 = require(".");
const assert_1 = __importDefault(require("assert"));
const stream_1 = require("stream");
var authOkBuffer = test_buffers_1.default.authenticationOk();
var paramStatusBuffer = test_buffers_1.default.parameterStatus('client_encoding', 'UTF8');
var readyForQueryBuffer = test_buffers_1.default.readyForQuery();
var backendKeyDataBuffer = test_buffers_1.default.backendKeyData(1, 2);
var commandCompleteBuffer = test_buffers_1.default.commandComplete('SELECT 3');
var parseCompleteBuffer = test_buffers_1.default.parseComplete();
var bindCompleteBuffer = test_buffers_1.default.bindComplete();
var portalSuspendedBuffer = test_buffers_1.default.portalSuspended();
var addRow = function (bufferList, name, offset) ***REMOVED***
    return bufferList
        .addCString(name) // field name
        .addInt32(offset++) // table id
        .addInt16(offset++) // attribute of column number
        .addInt32(offset++) // objectId of field's data type
        .addInt16(offset++) // datatype size
        .addInt32(offset++) // type modifier
        .addInt16(0); // format code, 0 => text
};
var row1 = ***REMOVED***
    name: 'id',
    tableID: 1,
    attributeNumber: 2,
    dataTypeID: 3,
    dataTypeSize: 4,
    typeModifier: 5,
    formatCode: 0,
};
var oneRowDescBuff = test_buffers_1.default.rowDescription([row1]);
row1.name = 'bang';
var twoRowBuf = test_buffers_1.default.rowDescription([
    row1,
    ***REMOVED***
        name: 'whoah',
        tableID: 10,
        attributeNumber: 11,
        dataTypeID: 12,
        dataTypeSize: 13,
        typeModifier: 14,
        formatCode: 0,
  ***REMOVED***,
]);
var emptyRowFieldBuf = new buffer_list_1.default().addInt16(0).join(true, 'D');
var emptyRowFieldBuf = test_buffers_1.default.dataRow([]);
var oneFieldBuf = new buffer_list_1.default()
    .addInt16(1) // number of fields
    .addInt32(5) // length of bytes of fields
    .addCString('test')
    .join(true, 'D');
var oneFieldBuf = test_buffers_1.default.dataRow(['test']);
var expectedAuthenticationOkayMessage = ***REMOVED***
    name: 'authenticationOk',
    length: 8,
};
var expectedParameterStatusMessage = ***REMOVED***
    name: 'parameterStatus',
    parameterName: 'client_encoding',
    parameterValue: 'UTF8',
    length: 25,
};
var expectedBackendKeyDataMessage = ***REMOVED***
    name: 'backendKeyData',
    processID: 1,
    secretKey: 2,
};
var expectedReadyForQueryMessage = ***REMOVED***
    name: 'readyForQuery',
    length: 5,
    status: 'I',
};
var expectedCommandCompleteMessage = ***REMOVED***
    name: 'commandComplete',
    length: 13,
    text: 'SELECT 3',
};
var emptyRowDescriptionBuffer = new buffer_list_1.default()
    .addInt16(0) // number of fields
    .join(true, 'T');
var expectedEmptyRowDescriptionMessage = ***REMOVED***
    name: 'rowDescription',
    length: 6,
    fieldCount: 0,
    fields: [],
};
var expectedOneRowMessage = ***REMOVED***
    name: 'rowDescription',
    length: 27,
    fieldCount: 1,
    fields: [
        ***REMOVED***
            name: 'id',
            tableID: 1,
            columnID: 2,
            dataTypeID: 3,
            dataTypeSize: 4,
            dataTypeModifier: 5,
            format: 'text',
      ***REMOVED***,
    ],
};
var expectedTwoRowMessage = ***REMOVED***
    name: 'rowDescription',
    length: 53,
    fieldCount: 2,
    fields: [
        ***REMOVED***
            name: 'bang',
            tableID: 1,
            columnID: 2,
            dataTypeID: 3,
            dataTypeSize: 4,
            dataTypeModifier: 5,
            format: 'text',
      ***REMOVED***,
        ***REMOVED***
            name: 'whoah',
            tableID: 10,
            columnID: 11,
            dataTypeID: 12,
            dataTypeSize: 13,
            dataTypeModifier: 14,
            format: 'text',
      ***REMOVED***,
    ],
};
var emptyParameterDescriptionBuffer = new buffer_list_1.default()
    .addInt16(0) // number of parameters
    .join(true, 't');
var oneParameterDescBuf = test_buffers_1.default.parameterDescription([1111]);
var twoParameterDescBuf = test_buffers_1.default.parameterDescription([2222, 3333]);
var expectedEmptyParameterDescriptionMessage = ***REMOVED***
    name: 'parameterDescription',
    length: 6,
    parameterCount: 0,
    dataTypeIDs: [],
};
var expectedOneParameterMessage = ***REMOVED***
    name: 'parameterDescription',
    length: 10,
    parameterCount: 1,
    dataTypeIDs: [1111],
};
var expectedTwoParameterMessage = ***REMOVED***
    name: 'parameterDescription',
    length: 14,
    parameterCount: 2,
    dataTypeIDs: [2222, 3333],
};
var testForMessage = function (buffer, expectedMessage) ***REMOVED***
    it('recieves and parses ' + expectedMessage.name, () => __awaiter(this, void 0, void 0, function* () ***REMOVED***
        const messages = yield parseBuffers([buffer]);
        const [lastMessage] = messages;
        for (const key in expectedMessage) ***REMOVED***
            assert_1.default.deepEqual(lastMessage[key], expectedMessage[key]);
      ***REMOVED***
  ***REMOVED***));
};
var plainPasswordBuffer = test_buffers_1.default.authenticationCleartextPassword();
var md5PasswordBuffer = test_buffers_1.default.authenticationMD5Password();
var SASLBuffer = test_buffers_1.default.authenticationSASL();
var SASLContinueBuffer = test_buffers_1.default.authenticationSASLContinue();
var SASLFinalBuffer = test_buffers_1.default.authenticationSASLFinal();
var expectedPlainPasswordMessage = ***REMOVED***
    name: 'authenticationCleartextPassword',
};
var expectedMD5PasswordMessage = ***REMOVED***
    name: 'authenticationMD5Password',
    salt: Buffer.from([1, 2, 3, 4]),
};
var expectedSASLMessage = ***REMOVED***
    name: 'authenticationSASL',
    mechanisms: ['SCRAM-SHA-256'],
};
var expectedSASLContinueMessage = ***REMOVED***
    name: 'authenticationSASLContinue',
    data: 'data',
};
var expectedSASLFinalMessage = ***REMOVED***
    name: 'authenticationSASLFinal',
    data: 'data',
};
var notificationResponseBuffer = test_buffers_1.default.notification(4, 'hi', 'boom');
var expectedNotificationResponseMessage = ***REMOVED***
    name: 'notification',
    processId: 4,
    channel: 'hi',
    payload: 'boom',
};
const parseBuffers = (buffers) => __awaiter(void 0, void 0, void 0, function* () ***REMOVED***
    const stream = new stream_1.PassThrough();
    for (const buffer of buffers) ***REMOVED***
        stream.write(buffer);
  ***REMOVED***
    stream.end();
    const msgs = [];
    yield _1.parse(stream, (msg) => msgs.push(msg));
    return msgs;
});
describe('PgPacketStream', function () ***REMOVED***
    testForMessage(authOkBuffer, expectedAuthenticationOkayMessage);
    testForMessage(plainPasswordBuffer, expectedPlainPasswordMessage);
    testForMessage(md5PasswordBuffer, expectedMD5PasswordMessage);
    testForMessage(SASLBuffer, expectedSASLMessage);
    testForMessage(SASLContinueBuffer, expectedSASLContinueMessage);
    // this exercises a found bug in the parser:
    // https://github.com/brianc/node-postgres/pull/2210#issuecomment-627626084
    // and adds a test which is deterministic, rather than relying on network packet chunking
    const extendedSASLContinueBuffer = Buffer.concat([SASLContinueBuffer, Buffer.from([1, 2, 3, 4])]);
    testForMessage(extendedSASLContinueBuffer, expectedSASLContinueMessage);
    testForMessage(SASLFinalBuffer, expectedSASLFinalMessage);
    // this exercises a found bug in the parser:
    // https://github.com/brianc/node-postgres/pull/2210#issuecomment-627626084
    // and adds a test which is deterministic, rather than relying on network packet chunking
    const extendedSASLFinalBuffer = Buffer.concat([SASLFinalBuffer, Buffer.from([1, 2, 4, 5])]);
    testForMessage(extendedSASLFinalBuffer, expectedSASLFinalMessage);
    testForMessage(paramStatusBuffer, expectedParameterStatusMessage);
    testForMessage(backendKeyDataBuffer, expectedBackendKeyDataMessage);
    testForMessage(readyForQueryBuffer, expectedReadyForQueryMessage);
    testForMessage(commandCompleteBuffer, expectedCommandCompleteMessage);
    testForMessage(notificationResponseBuffer, expectedNotificationResponseMessage);
    testForMessage(test_buffers_1.default.emptyQuery(), ***REMOVED***
        name: 'emptyQuery',
        length: 4,
  ***REMOVED***);
    testForMessage(Buffer.from([0x6e, 0, 0, 0, 4]), ***REMOVED***
        name: 'noData',
  ***REMOVED***);
    describe('rowDescription messages', function () ***REMOVED***
        testForMessage(emptyRowDescriptionBuffer, expectedEmptyRowDescriptionMessage);
        testForMessage(oneRowDescBuff, expectedOneRowMessage);
        testForMessage(twoRowBuf, expectedTwoRowMessage);
  ***REMOVED***);
    describe('parameterDescription messages', function () ***REMOVED***
        testForMessage(emptyParameterDescriptionBuffer, expectedEmptyParameterDescriptionMessage);
        testForMessage(oneParameterDescBuf, expectedOneParameterMessage);
        testForMessage(twoParameterDescBuf, expectedTwoParameterMessage);
  ***REMOVED***);
    describe('parsing rows', function () ***REMOVED***
        describe('parsing empty row', function () ***REMOVED***
            testForMessage(emptyRowFieldBuf, ***REMOVED***
                name: 'dataRow',
                fieldCount: 0,
          ***REMOVED***);
      ***REMOVED***);
        describe('parsing data row with fields', function () ***REMOVED***
            testForMessage(oneFieldBuf, ***REMOVED***
                name: 'dataRow',
                fieldCount: 1,
                fields: ['test'],
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***);
    describe('notice message', function () ***REMOVED***
        // this uses the same logic as error message
        var buff = test_buffers_1.default.notice([***REMOVED*** type: 'C', value: 'code' }]);
        testForMessage(buff, ***REMOVED***
            name: 'notice',
            code: 'code',
      ***REMOVED***);
  ***REMOVED***);
    testForMessage(test_buffers_1.default.error([]), ***REMOVED***
        name: 'error',
  ***REMOVED***);
    describe('with all the fields', function () ***REMOVED***
        var buffer = test_buffers_1.default.error([
            ***REMOVED***
                type: 'S',
                value: 'ERROR',
          ***REMOVED***,
            ***REMOVED***
                type: 'C',
                value: 'code',
          ***REMOVED***,
            ***REMOVED***
                type: 'M',
                value: 'message',
          ***REMOVED***,
            ***REMOVED***
                type: 'D',
                value: 'details',
          ***REMOVED***,
            ***REMOVED***
                type: 'H',
                value: 'hint',
          ***REMOVED***,
            ***REMOVED***
                type: 'P',
                value: '100',
          ***REMOVED***,
            ***REMOVED***
                type: 'p',
                value: '101',
          ***REMOVED***,
            ***REMOVED***
                type: 'q',
                value: 'query',
          ***REMOVED***,
            ***REMOVED***
                type: 'W',
                value: 'where',
          ***REMOVED***,
            ***REMOVED***
                type: 'F',
                value: 'file',
          ***REMOVED***,
            ***REMOVED***
                type: 'L',
                value: 'line',
          ***REMOVED***,
            ***REMOVED***
                type: 'R',
                value: 'routine',
          ***REMOVED***,
            ***REMOVED***
                type: 'Z',
                value: 'alsdkf',
          ***REMOVED***,
        ]);
        testForMessage(buffer, ***REMOVED***
            name: 'error',
            severity: 'ERROR',
            code: 'code',
            message: 'message',
            detail: 'details',
            hint: 'hint',
            position: '100',
            internalPosition: '101',
            internalQuery: 'query',
            where: 'where',
            file: 'file',
            line: 'line',
            routine: 'routine',
      ***REMOVED***);
  ***REMOVED***);
    testForMessage(parseCompleteBuffer, ***REMOVED***
        name: 'parseComplete',
  ***REMOVED***);
    testForMessage(bindCompleteBuffer, ***REMOVED***
        name: 'bindComplete',
  ***REMOVED***);
    testForMessage(bindCompleteBuffer, ***REMOVED***
        name: 'bindComplete',
  ***REMOVED***);
    testForMessage(test_buffers_1.default.closeComplete(), ***REMOVED***
        name: 'closeComplete',
  ***REMOVED***);
    describe('parses portal suspended message', function () ***REMOVED***
        testForMessage(portalSuspendedBuffer, ***REMOVED***
            name: 'portalSuspended',
      ***REMOVED***);
  ***REMOVED***);
    describe('parses replication start message', function () ***REMOVED***
        testForMessage(Buffer.from([0x57, 0x00, 0x00, 0x00, 0x04]), ***REMOVED***
            name: 'replicationStart',
            length: 4,
      ***REMOVED***);
  ***REMOVED***);
    describe('copy', () => ***REMOVED***
        testForMessage(test_buffers_1.default.copyIn(0), ***REMOVED***
            name: 'copyInResponse',
            length: 7,
            binary: false,
            columnTypes: [],
      ***REMOVED***);
        testForMessage(test_buffers_1.default.copyIn(2), ***REMOVED***
            name: 'copyInResponse',
            length: 11,
            binary: false,
            columnTypes: [0, 1],
      ***REMOVED***);
        testForMessage(test_buffers_1.default.copyOut(0), ***REMOVED***
            name: 'copyOutResponse',
            length: 7,
            binary: false,
            columnTypes: [],
      ***REMOVED***);
        testForMessage(test_buffers_1.default.copyOut(3), ***REMOVED***
            name: 'copyOutResponse',
            length: 13,
            binary: false,
            columnTypes: [0, 1, 2],
      ***REMOVED***);
        testForMessage(test_buffers_1.default.copyDone(), ***REMOVED***
            name: 'copyDone',
            length: 4,
      ***REMOVED***);
        testForMessage(test_buffers_1.default.copyData(Buffer.from([5, 6, 7])), ***REMOVED***
            name: 'copyData',
            length: 7,
            chunk: Buffer.from([5, 6, 7]),
      ***REMOVED***);
  ***REMOVED***);
    // since the data message on a stream can randomly divide the incomming
    // tcp packets anywhere, we need to make sure we can parse every single
    // split on a tcp message
    describe('split buffer, single message parsing', function () ***REMOVED***
        var fullBuffer = test_buffers_1.default.dataRow([null, 'bang', 'zug zug', null, '!']);
        it('parses when full buffer comes in', function () ***REMOVED***
            return __awaiter(this, void 0, void 0, function* () ***REMOVED***
                const messages = yield parseBuffers([fullBuffer]);
                const message = messages[0];
                assert_1.default.equal(message.fields.length, 5);
                assert_1.default.equal(message.fields[0], null);
                assert_1.default.equal(message.fields[1], 'bang');
                assert_1.default.equal(message.fields[2], 'zug zug');
                assert_1.default.equal(message.fields[3], null);
                assert_1.default.equal(message.fields[4], '!');
          ***REMOVED***);
      ***REMOVED***);
        var testMessageRecievedAfterSpiltAt = function (split) ***REMOVED***
            return __awaiter(this, void 0, void 0, function* () ***REMOVED***
                var firstBuffer = Buffer.alloc(fullBuffer.length - split);
                var secondBuffer = Buffer.alloc(fullBuffer.length - firstBuffer.length);
                fullBuffer.copy(firstBuffer, 0, 0);
                fullBuffer.copy(secondBuffer, 0, firstBuffer.length);
                const messages = yield parseBuffers([fullBuffer]);
                const message = messages[0];
                assert_1.default.equal(message.fields.length, 5);
                assert_1.default.equal(message.fields[0], null);
                assert_1.default.equal(message.fields[1], 'bang');
                assert_1.default.equal(message.fields[2], 'zug zug');
                assert_1.default.equal(message.fields[3], null);
                assert_1.default.equal(message.fields[4], '!');
          ***REMOVED***);
      ***REMOVED***;
        it('parses when split in the middle', function () ***REMOVED***
            testMessageRecievedAfterSpiltAt(6);
      ***REMOVED***);
        it('parses when split at end', function () ***REMOVED***
            testMessageRecievedAfterSpiltAt(2);
      ***REMOVED***);
        it('parses when split at beginning', function () ***REMOVED***
            testMessageRecievedAfterSpiltAt(fullBuffer.length - 2);
            testMessageRecievedAfterSpiltAt(fullBuffer.length - 1);
            testMessageRecievedAfterSpiltAt(fullBuffer.length - 5);
      ***REMOVED***);
  ***REMOVED***);
    describe('split buffer, multiple message parsing', function () ***REMOVED***
        var dataRowBuffer = test_buffers_1.default.dataRow(['!']);
        var readyForQueryBuffer = test_buffers_1.default.readyForQuery();
        var fullBuffer = Buffer.alloc(dataRowBuffer.length + readyForQueryBuffer.length);
        dataRowBuffer.copy(fullBuffer, 0, 0);
        readyForQueryBuffer.copy(fullBuffer, dataRowBuffer.length, 0);
        var verifyMessages = function (messages) ***REMOVED***
            assert_1.default.strictEqual(messages.length, 2);
            assert_1.default.deepEqual(messages[0], ***REMOVED***
                name: 'dataRow',
                fieldCount: 1,
                length: 11,
                fields: ['!'],
          ***REMOVED***);
            assert_1.default.equal(messages[0].fields[0], '!');
            assert_1.default.deepEqual(messages[1], ***REMOVED***
                name: 'readyForQuery',
                length: 5,
                status: 'I',
          ***REMOVED***);
      ***REMOVED***;
        // sanity check
        it('recieves both messages when packet is not split', function () ***REMOVED***
            return __awaiter(this, void 0, void 0, function* () ***REMOVED***
                const messages = yield parseBuffers([fullBuffer]);
                verifyMessages(messages);
          ***REMOVED***);
      ***REMOVED***);
        var splitAndVerifyTwoMessages = function (split) ***REMOVED***
            return __awaiter(this, void 0, void 0, function* () ***REMOVED***
                var firstBuffer = Buffer.alloc(fullBuffer.length - split);
                var secondBuffer = Buffer.alloc(fullBuffer.length - firstBuffer.length);
                fullBuffer.copy(firstBuffer, 0, 0);
                fullBuffer.copy(secondBuffer, 0, firstBuffer.length);
                const messages = yield parseBuffers([firstBuffer, secondBuffer]);
                verifyMessages(messages);
          ***REMOVED***);
      ***REMOVED***;
        describe('recieves both messages when packet is split', function () ***REMOVED***
            it('in the middle', function () ***REMOVED***
                return splitAndVerifyTwoMessages(11);
          ***REMOVED***);
            it('at the front', function () ***REMOVED***
                return Promise.all([
                    splitAndVerifyTwoMessages(fullBuffer.length - 1),
                    splitAndVerifyTwoMessages(fullBuffer.length - 4),
                    splitAndVerifyTwoMessages(fullBuffer.length - 6),
                ]);
          ***REMOVED***);
            it('at the end', function () ***REMOVED***
                return Promise.all([splitAndVerifyTwoMessages(8), splitAndVerifyTwoMessages(1)]);
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***);
});
