// pg-protocol/dist/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.NoticeMessage = exports.DataRowMessage = exports.CommandCompleteMessage = exports.ReadyForQueryMessage = exports.NotificationResponseMessage = exports.BackendKeyDataMessage = exports.AuthenticationMD5Password = exports.ParameterStatusMessage = exports.ParameterDescriptionMessage = exports.RowDescriptionMessage = exports.Field = exports.CopyResponse = exports.CopyDataMessage = exports.DatabaseError = exports.copyDone = exports.emptyQuery = exports.replicationStart = exports.portalSuspended = exports.noData = exports.closeComplete = exports.bindComplete = exports.parseComplete = void 0;
exports.parseComplete = ***REMOVED***
    name: 'parseComplete',
    length: 5,
};
exports.bindComplete = ***REMOVED***
    name: 'bindComplete',
    length: 5,
};
exports.closeComplete = ***REMOVED***
    name: 'closeComplete',
    length: 5,
};
exports.noData = ***REMOVED***
    name: 'noData',
    length: 5,
};
exports.portalSuspended = ***REMOVED***
    name: 'portalSuspended',
    length: 5,
};
exports.replicationStart = ***REMOVED***
    name: 'replicationStart',
    length: 4,
};
exports.emptyQuery = ***REMOVED***
    name: 'emptyQuery',
    length: 4,
};
exports.copyDone = ***REMOVED***
    name: 'copyDone',
    length: 4,
};
class DatabaseError extends Error ***REMOVED***
    constructor(message, length, name) ***REMOVED***
        super(message);
        this.length = length;
        this.name = name;
  ***REMOVED***
}
exports.DatabaseError = DatabaseError;
class CopyDataMessage ***REMOVED***
    constructor(length, chunk) ***REMOVED***
        this.length = length;
        this.chunk = chunk;
        this.name = 'copyData';
  ***REMOVED***
}
exports.CopyDataMessage = CopyDataMessage;
class CopyResponse ***REMOVED***
    constructor(length, name, binary, columnCount) ***REMOVED***
        this.length = length;
        this.name = name;
        this.binary = binary;
        this.columnTypes = new Array(columnCount);
  ***REMOVED***
}
exports.CopyResponse = CopyResponse;
class Field ***REMOVED***
    constructor(name, tableID, columnID, dataTypeID, dataTypeSize, dataTypeModifier, format) ***REMOVED***
        this.name = name;
        this.tableID = tableID;
        this.columnID = columnID;
        this.dataTypeID = dataTypeID;
        this.dataTypeSize = dataTypeSize;
        this.dataTypeModifier = dataTypeModifier;
        this.format = format;
  ***REMOVED***
}
exports.Field = Field;
class RowDescriptionMessage ***REMOVED***
    constructor(length, fieldCount) ***REMOVED***
        this.length = length;
        this.fieldCount = fieldCount;
        this.name = 'rowDescription';
        this.fields = new Array(this.fieldCount);
  ***REMOVED***
}
exports.RowDescriptionMessage = RowDescriptionMessage;
class ParameterDescriptionMessage ***REMOVED***
    constructor(length, parameterCount) ***REMOVED***
        this.length = length;
        this.parameterCount = parameterCount;
        this.name = 'parameterDescription';
        this.dataTypeIDs = new Array(this.parameterCount);
  ***REMOVED***
}
exports.ParameterDescriptionMessage = ParameterDescriptionMessage;
class ParameterStatusMessage ***REMOVED***
    constructor(length, parameterName, parameterValue) ***REMOVED***
        this.length = length;
        this.parameterName = parameterName;
        this.parameterValue = parameterValue;
        this.name = 'parameterStatus';
  ***REMOVED***
}
exports.ParameterStatusMessage = ParameterStatusMessage;
class AuthenticationMD5Password ***REMOVED***
    constructor(length, salt) ***REMOVED***
        this.length = length;
        this.salt = salt;
        this.name = 'authenticationMD5Password';
  ***REMOVED***
}
exports.AuthenticationMD5Password = AuthenticationMD5Password;
class BackendKeyDataMessage ***REMOVED***
    constructor(length, processID, secretKey) ***REMOVED***
        this.length = length;
        this.processID = processID;
        this.secretKey = secretKey;
        this.name = 'backendKeyData';
  ***REMOVED***
}
exports.BackendKeyDataMessage = BackendKeyDataMessage;
class NotificationResponseMessage ***REMOVED***
    constructor(length, processId, channel, payload) ***REMOVED***
        this.length = length;
        this.processId = processId;
        this.channel = channel;
        this.payload = payload;
        this.name = 'notification';
  ***REMOVED***
}
exports.NotificationResponseMessage = NotificationResponseMessage;
class ReadyForQueryMessage ***REMOVED***
    constructor(length, status) ***REMOVED***
        this.length = length;
        this.status = status;
        this.name = 'readyForQuery';
  ***REMOVED***
}
exports.ReadyForQueryMessage = ReadyForQueryMessage;
class CommandCompleteMessage ***REMOVED***
    constructor(length, text) ***REMOVED***
        this.length = length;
        this.text = text;
        this.name = 'commandComplete';
  ***REMOVED***
}
exports.CommandCompleteMessage = CommandCompleteMessage;
class DataRowMessage ***REMOVED***
    constructor(length, fields) ***REMOVED***
        this.length = length;
        this.fields = fields;
        this.name = 'dataRow';
        this.fieldCount = fields.length;
  ***REMOVED***
}
exports.DataRowMessage = DataRowMessage;
class NoticeMessage ***REMOVED***
    constructor(length, message) ***REMOVED***
        this.length = length;
        this.message = message;
        this.name = 'notice';
  ***REMOVED***
}
exports.NoticeMessage = NoticeMessage;
