// pg-protocol/dist/outbound-serializer.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) ***REMOVED***
    return (mod && mod.__esModule) ? mod : ***REMOVED*** "default": mod };
};
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
const assert_1 = __importDefault(require("assert"));
const serializer_1 = require("./serializer");
const buffer_list_1 = __importDefault(require("./testing/buffer-list"));
describe('serializer', () => ***REMOVED***
    it('builds startup message', function () ***REMOVED***
        const actual = serializer_1.serialize.startup(***REMOVED***
            user: 'brian',
            database: 'bang',
      ***REMOVED***);
        assert_1.default.deepEqual(actual, new buffer_list_1.default()
            .addInt16(3)
            .addInt16(0)
            .addCString('user')
            .addCString('brian')
            .addCString('database')
            .addCString('bang')
            .addCString('client_encoding')
            .addCString('UTF8')
            .addCString('')
            .join(true));
  ***REMOVED***);
    it('builds password message', function () ***REMOVED***
        const actual = serializer_1.serialize.password('!');
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addCString('!').join(true, 'p'));
  ***REMOVED***);
    it('builds request ssl message', function () ***REMOVED***
        const actual = serializer_1.serialize.requestSsl();
        const expected = new buffer_list_1.default().addInt32(80877103).join(true);
        assert_1.default.deepEqual(actual, expected);
  ***REMOVED***);
    it('builds SASLInitialResponseMessage message', function () ***REMOVED***
        const actual = serializer_1.serialize.sendSASLInitialResponseMessage('mech', 'data');
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addCString('mech').addInt32(4).addString('data').join(true, 'p'));
  ***REMOVED***);
    it('builds SCRAMClientFinalMessage message', function () ***REMOVED***
        const actual = serializer_1.serialize.sendSCRAMClientFinalMessage('data');
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addString('data').join(true, 'p'));
  ***REMOVED***);
    it('builds query message', function () ***REMOVED***
        var txt = 'select * from boom';
        const actual = serializer_1.serialize.query(txt);
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addCString(txt).join(true, 'Q'));
  ***REMOVED***);
    describe('parse message', () => ***REMOVED***
        it('builds parse message', function () ***REMOVED***
            const actual = serializer_1.serialize.parse(***REMOVED*** text: '!' });
            var expected = new buffer_list_1.default().addCString('').addCString('!').addInt16(0).join(true, 'P');
            assert_1.default.deepEqual(actual, expected);
      ***REMOVED***);
        it('builds parse message with named query', function () ***REMOVED***
            const actual = serializer_1.serialize.parse(***REMOVED***
                name: 'boom',
                text: 'select * from boom',
                types: [],
          ***REMOVED***);
            var expected = new buffer_list_1.default().addCString('boom').addCString('select * from boom').addInt16(0).join(true, 'P');
            assert_1.default.deepEqual(actual, expected);
      ***REMOVED***);
        it('with multiple parameters', function () ***REMOVED***
            const actual = serializer_1.serialize.parse(***REMOVED***
                name: 'force',
                text: 'select * from bang where name = $1',
                types: [1, 2, 3, 4],
          ***REMOVED***);
            var expected = new buffer_list_1.default()
                .addCString('force')
                .addCString('select * from bang where name = $1')
                .addInt16(4)
                .addInt32(1)
                .addInt32(2)
                .addInt32(3)
                .addInt32(4)
                .join(true, 'P');
            assert_1.default.deepEqual(actual, expected);
      ***REMOVED***);
  ***REMOVED***);
    describe('bind messages', function () ***REMOVED***
        it('with no values', function () ***REMOVED***
            const actual = serializer_1.serialize.bind();
            var expectedBuffer = new buffer_list_1.default()
                .addCString('')
                .addCString('')
                .addInt16(0)
                .addInt16(0)
                .addInt16(0)
                .join(true, 'B');
            assert_1.default.deepEqual(actual, expectedBuffer);
      ***REMOVED***);
        it('with named statement, portal, and values', function () ***REMOVED***
            const actual = serializer_1.serialize.bind(***REMOVED***
                portal: 'bang',
                statement: 'woo',
                values: ['1', 'hi', null, 'zing'],
          ***REMOVED***);
            var expectedBuffer = new buffer_list_1.default()
                .addCString('bang') // portal name
                .addCString('woo') // statement name
                .addInt16(4)
                .addInt16(0)
                .addInt16(0)
                .addInt16(0)
                .addInt16(0)
                .addInt16(4)
                .addInt32(1)
                .add(Buffer.from('1'))
                .addInt32(2)
                .add(Buffer.from('hi'))
                .addInt32(-1)
                .addInt32(4)
                .add(Buffer.from('zing'))
                .addInt16(0)
                .join(true, 'B');
            assert_1.default.deepEqual(actual, expectedBuffer);
      ***REMOVED***);
  ***REMOVED***);
    it('with custom valueMapper', function () ***REMOVED***
        const actual = serializer_1.serialize.bind(***REMOVED***
            portal: 'bang',
            statement: 'woo',
            values: ['1', 'hi', null, 'zing'],
            valueMapper: () => null,
      ***REMOVED***);
        var expectedBuffer = new buffer_list_1.default()
            .addCString('bang') // portal name
            .addCString('woo') // statement name
            .addInt16(4)
            .addInt16(0)
            .addInt16(0)
            .addInt16(0)
            .addInt16(0)
            .addInt16(4)
            .addInt32(-1)
            .addInt32(-1)
            .addInt32(-1)
            .addInt32(-1)
            .addInt16(0)
            .join(true, 'B');
        assert_1.default.deepEqual(actual, expectedBuffer);
  ***REMOVED***);
    it('with named statement, portal, and buffer value', function () ***REMOVED***
        const actual = serializer_1.serialize.bind(***REMOVED***
            portal: 'bang',
            statement: 'woo',
            values: ['1', 'hi', null, Buffer.from('zing', 'utf8')],
      ***REMOVED***);
        var expectedBuffer = new buffer_list_1.default()
            .addCString('bang') // portal name
            .addCString('woo') // statement name
            .addInt16(4) // value count
            .addInt16(0) // string
            .addInt16(0) // string
            .addInt16(0) // string
            .addInt16(1) // binary
            .addInt16(4)
            .addInt32(1)
            .add(Buffer.from('1'))
            .addInt32(2)
            .add(Buffer.from('hi'))
            .addInt32(-1)
            .addInt32(4)
            .add(Buffer.from('zing', 'utf-8'))
            .addInt16(0)
            .join(true, 'B');
        assert_1.default.deepEqual(actual, expectedBuffer);
  ***REMOVED***);
    describe('builds execute message', function () ***REMOVED***
        it('for unamed portal with no row limit', function () ***REMOVED***
            const actual = serializer_1.serialize.execute();
            var expectedBuffer = new buffer_list_1.default().addCString('').addInt32(0).join(true, 'E');
            assert_1.default.deepEqual(actual, expectedBuffer);
      ***REMOVED***);
        it('for named portal with row limit', function () ***REMOVED***
            const actual = serializer_1.serialize.execute(***REMOVED***
                portal: 'my favorite portal',
                rows: 100,
          ***REMOVED***);
            var expectedBuffer = new buffer_list_1.default().addCString('my favorite portal').addInt32(100).join(true, 'E');
            assert_1.default.deepEqual(actual, expectedBuffer);
      ***REMOVED***);
  ***REMOVED***);
    it('builds flush command', function () ***REMOVED***
        const actual = serializer_1.serialize.flush();
        var expected = new buffer_list_1.default().join(true, 'H');
        assert_1.default.deepEqual(actual, expected);
  ***REMOVED***);
    it('builds sync command', function () ***REMOVED***
        const actual = serializer_1.serialize.sync();
        var expected = new buffer_list_1.default().join(true, 'S');
        assert_1.default.deepEqual(actual, expected);
  ***REMOVED***);
    it('builds end command', function () ***REMOVED***
        const actual = serializer_1.serialize.end();
        var expected = Buffer.from([0x58, 0, 0, 0, 4]);
        assert_1.default.deepEqual(actual, expected);
  ***REMOVED***);
    describe('builds describe command', function () ***REMOVED***
        it('describe statement', function () ***REMOVED***
            const actual = serializer_1.serialize.describe(***REMOVED*** type: 'S', name: 'bang' });
            var expected = new buffer_list_1.default().addChar('S').addCString('bang').join(true, 'D');
            assert_1.default.deepEqual(actual, expected);
      ***REMOVED***);
        it('describe unnamed portal', function () ***REMOVED***
            const actual = serializer_1.serialize.describe(***REMOVED*** type: 'P' });
            var expected = new buffer_list_1.default().addChar('P').addCString('').join(true, 'D');
            assert_1.default.deepEqual(actual, expected);
      ***REMOVED***);
  ***REMOVED***);
    describe('builds close command', function () ***REMOVED***
        it('describe statement', function () ***REMOVED***
            const actual = serializer_1.serialize.close(***REMOVED*** type: 'S', name: 'bang' });
            var expected = new buffer_list_1.default().addChar('S').addCString('bang').join(true, 'C');
            assert_1.default.deepEqual(actual, expected);
      ***REMOVED***);
        it('describe unnamed portal', function () ***REMOVED***
            const actual = serializer_1.serialize.close(***REMOVED*** type: 'P' });
            var expected = new buffer_list_1.default().addChar('P').addCString('').join(true, 'C');
            assert_1.default.deepEqual(actual, expected);
      ***REMOVED***);
  ***REMOVED***);
    describe('copy messages', function () ***REMOVED***
        it('builds copyFromChunk', () => ***REMOVED***
            const actual = serializer_1.serialize.copyData(Buffer.from([1, 2, 3]));
            const expected = new buffer_list_1.default().add(Buffer.from([1, 2, 3])).join(true, 'd');
            assert_1.default.deepEqual(actual, expected);
      ***REMOVED***);
        it('builds copy fail', () => ***REMOVED***
            const actual = serializer_1.serialize.copyFail('err!');
            const expected = new buffer_list_1.default().addCString('err!').join(true, 'f');
            assert_1.default.deepEqual(actual, expected);
      ***REMOVED***);
        it('builds copy done', () => ***REMOVED***
            const actual = serializer_1.serialize.copyDone();
            const expected = new buffer_list_1.default().join(true, 'c');
            assert_1.default.deepEqual(actual, expected);
      ***REMOVED***);
  ***REMOVED***);
    it('builds cancel message', () => ***REMOVED***
        const actual = serializer_1.serialize.cancel(3, 4);
        const expected = new buffer_list_1.default().addInt16(1234).addInt16(5678).addInt32(3).addInt32(4).join(true);
        assert_1.default.deepEqual(actual, expected);
  ***REMOVED***);
});
