// pg-protocol/dist/parser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) ***REMOVED***
    return (mod && mod.__esModule) ? mod : ***REMOVED*** "default": mod };
};
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.Parser = void 0;
const messages_1 = require("./messages");
const buffer_reader_1 = require("./buffer-reader");
const assert_1 = __importDefault(require("assert"));
// every message is prefixed with a single bye
const CODE_LENGTH = 1;
// every message has an int32 length which includes itself but does
// NOT include the code in the length
const LEN_LENGTH = 4;
const HEADER_LENGTH = CODE_LENGTH + LEN_LENGTH;
const emptyBuffer = Buffer.allocUnsafe(0);
class Parser ***REMOVED***
    constructor(opts) ***REMOVED***
        this.buffer = emptyBuffer;
        this.bufferLength = 0;
        this.bufferOffset = 0;
        this.reader = new buffer_reader_1.BufferReader();
        if ((opts === null || opts === void 0 ? void 0 : opts.mode) === 'binary') ***REMOVED***
            throw new Error('Binary mode not supported yet');
      ***REMOVED***
        this.mode = (opts === null || opts === void 0 ? void 0 : opts.mode) || 'text';
  ***REMOVED***
    parse(buffer, callback) ***REMOVED***
        this.mergeBuffer(buffer);
        const bufferFullLength = this.bufferOffset + this.bufferLength;
        let offset = this.bufferOffset;
        while (offset + HEADER_LENGTH <= bufferFullLength) ***REMOVED***
            // code is 1 byte long - it identifies the message type
            const code = this.buffer[offset];
            // length is 1 Uint32BE - it is the length of the message EXCLUDING the code
            const length = this.buffer.readUInt32BE(offset + CODE_LENGTH);
            const fullMessageLength = CODE_LENGTH + length;
            if (fullMessageLength + offset <= bufferFullLength) ***REMOVED***
                const message = this.handlePacket(offset + HEADER_LENGTH, code, length, this.buffer);
                callback(message);
                offset += fullMessageLength;
          ***REMOVED***
            else ***REMOVED***
                break;
          ***REMOVED***
      ***REMOVED***
        if (offset === bufferFullLength) ***REMOVED***
            // No more use for the buffer
            this.buffer = emptyBuffer;
            this.bufferLength = 0;
            this.bufferOffset = 0;
      ***REMOVED***
        else ***REMOVED***
            // Adjust the cursors of remainingBuffer
            this.bufferLength = bufferFullLength - offset;
            this.bufferOffset = offset;
      ***REMOVED***
  ***REMOVED***
    mergeBuffer(buffer) ***REMOVED***
        if (this.bufferLength > 0) ***REMOVED***
            const newLength = this.bufferLength + buffer.byteLength;
            const newFullLength = newLength + this.bufferOffset;
            if (newFullLength > this.buffer.byteLength) ***REMOVED***
                // We can't concat the new buffer with the remaining one
                let newBuffer;
                if (newLength <= this.buffer.byteLength && this.bufferOffset >= this.bufferLength) ***REMOVED***
                    // We can move the relevant part to the beginning of the buffer instead of allocating a new buffer
                    newBuffer = this.buffer;
              ***REMOVED***
                else ***REMOVED***
                    // Allocate a new larger buffer
                    let newBufferLength = this.buffer.byteLength * 2;
                    while (newLength >= newBufferLength) ***REMOVED***
                        newBufferLength *= 2;
                  ***REMOVED***
                    newBuffer = Buffer.allocUnsafe(newBufferLength);
              ***REMOVED***
                // Move the remaining buffer to the new one
                this.buffer.copy(newBuffer, 0, this.bufferOffset, this.bufferOffset + this.bufferLength);
                this.buffer = newBuffer;
                this.bufferOffset = 0;
          ***REMOVED***
            // Concat the new buffer with the remaining one
            buffer.copy(this.buffer, this.bufferOffset + this.bufferLength);
            this.bufferLength = newLength;
      ***REMOVED***
        else ***REMOVED***
            this.buffer = buffer;
            this.bufferOffset = 0;
            this.bufferLength = buffer.byteLength;
      ***REMOVED***
  ***REMOVED***
    handlePacket(offset, code, length, bytes) ***REMOVED***
        switch (code) ***REMOVED***
            case 50 /* BindComplete */:
                return messages_1.bindComplete;
            case 49 /* ParseComplete */:
                return messages_1.parseComplete;
            case 51 /* CloseComplete */:
                return messages_1.closeComplete;
            case 110 /* NoData */:
                return messages_1.noData;
            case 115 /* PortalSuspended */:
                return messages_1.portalSuspended;
            case 99 /* CopyDone */:
                return messages_1.copyDone;
            case 87 /* ReplicationStart */:
                return messages_1.replicationStart;
            case 73 /* EmptyQuery */:
                return messages_1.emptyQuery;
            case 68 /* DataRow */:
                return this.parseDataRowMessage(offset, length, bytes);
            case 67 /* CommandComplete */:
                return this.parseCommandCompleteMessage(offset, length, bytes);
            case 90 /* ReadyForQuery */:
                return this.parseReadyForQueryMessage(offset, length, bytes);
            case 65 /* NotificationResponse */:
                return this.parseNotificationMessage(offset, length, bytes);
            case 82 /* AuthenticationResponse */:
                return this.parseAuthenticationResponse(offset, length, bytes);
            case 83 /* ParameterStatus */:
                return this.parseParameterStatusMessage(offset, length, bytes);
            case 75 /* BackendKeyData */:
                return this.parseBackendKeyData(offset, length, bytes);
            case 69 /* ErrorMessage */:
                return this.parseErrorMessage(offset, length, bytes, 'error');
            case 78 /* NoticeMessage */:
                return this.parseErrorMessage(offset, length, bytes, 'notice');
            case 84 /* RowDescriptionMessage */:
                return this.parseRowDescriptionMessage(offset, length, bytes);
            case 116 /* ParameterDescriptionMessage */:
                return this.parseParameterDescriptionMessage(offset, length, bytes);
            case 71 /* CopyIn */:
                return this.parseCopyInMessage(offset, length, bytes);
            case 72 /* CopyOut */:
                return this.parseCopyOutMessage(offset, length, bytes);
            case 100 /* CopyData */:
                return this.parseCopyData(offset, length, bytes);
            default:
                assert_1.default.fail(`unknown message code: $***REMOVED***code.toString(16)}`);
      ***REMOVED***
  ***REMOVED***
    parseReadyForQueryMessage(offset, length, bytes) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const status = this.reader.string(1);
        return new messages_1.ReadyForQueryMessage(length, status);
  ***REMOVED***
    parseCommandCompleteMessage(offset, length, bytes) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const text = this.reader.cstring();
        return new messages_1.CommandCompleteMessage(length, text);
  ***REMOVED***
    parseCopyData(offset, length, bytes) ***REMOVED***
        const chunk = bytes.slice(offset, offset + (length - 4));
        return new messages_1.CopyDataMessage(length, chunk);
  ***REMOVED***
    parseCopyInMessage(offset, length, bytes) ***REMOVED***
        return this.parseCopyMessage(offset, length, bytes, 'copyInResponse');
  ***REMOVED***
    parseCopyOutMessage(offset, length, bytes) ***REMOVED***
        return this.parseCopyMessage(offset, length, bytes, 'copyOutResponse');
  ***REMOVED***
    parseCopyMessage(offset, length, bytes, messageName) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const isBinary = this.reader.byte() !== 0;
        const columnCount = this.reader.int16();
        const message = new messages_1.CopyResponse(length, messageName, isBinary, columnCount);
        for (let i = 0; i < columnCount; i++) ***REMOVED***
            message.columnTypes[i] = this.reader.int16();
      ***REMOVED***
        return message;
  ***REMOVED***
    parseNotificationMessage(offset, length, bytes) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const processId = this.reader.int32();
        const channel = this.reader.cstring();
        const payload = this.reader.cstring();
        return new messages_1.NotificationResponseMessage(length, processId, channel, payload);
  ***REMOVED***
    parseRowDescriptionMessage(offset, length, bytes) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const fieldCount = this.reader.int16();
        const message = new messages_1.RowDescriptionMessage(length, fieldCount);
        for (let i = 0; i < fieldCount; i++) ***REMOVED***
            message.fields[i] = this.parseField();
      ***REMOVED***
        return message;
  ***REMOVED***
    parseField() ***REMOVED***
        const name = this.reader.cstring();
        const tableID = this.reader.int32();
        const columnID = this.reader.int16();
        const dataTypeID = this.reader.int32();
        const dataTypeSize = this.reader.int16();
        const dataTypeModifier = this.reader.int32();
        const mode = this.reader.int16() === 0 ? 'text' : 'binary';
        return new messages_1.Field(name, tableID, columnID, dataTypeID, dataTypeSize, dataTypeModifier, mode);
  ***REMOVED***
    parseParameterDescriptionMessage(offset, length, bytes) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const parameterCount = this.reader.int16();
        const message = new messages_1.ParameterDescriptionMessage(length, parameterCount);
        for (let i = 0; i < parameterCount; i++) ***REMOVED***
            message.dataTypeIDs[i] = this.reader.int32();
      ***REMOVED***
        return message;
  ***REMOVED***
    parseDataRowMessage(offset, length, bytes) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const fieldCount = this.reader.int16();
        const fields = new Array(fieldCount);
        for (let i = 0; i < fieldCount; i++) ***REMOVED***
            const len = this.reader.int32();
            // a -1 for length means the value of the field is null
            fields[i] = len === -1 ? null : this.reader.string(len);
      ***REMOVED***
        return new messages_1.DataRowMessage(length, fields);
  ***REMOVED***
    parseParameterStatusMessage(offset, length, bytes) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const name = this.reader.cstring();
        const value = this.reader.cstring();
        return new messages_1.ParameterStatusMessage(length, name, value);
  ***REMOVED***
    parseBackendKeyData(offset, length, bytes) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const processID = this.reader.int32();
        const secretKey = this.reader.int32();
        return new messages_1.BackendKeyDataMessage(length, processID, secretKey);
  ***REMOVED***
    parseAuthenticationResponse(offset, length, bytes) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const code = this.reader.int32();
        // TODO(bmc): maybe better types here
        const message = ***REMOVED***
            name: 'authenticationOk',
            length,
      ***REMOVED***;
        switch (code) ***REMOVED***
            case 0: // AuthenticationOk
                break;
            case 3: // AuthenticationCleartextPassword
                if (message.length === 8) ***REMOVED***
                    message.name = 'authenticationCleartextPassword';
              ***REMOVED***
                break;
            case 5: // AuthenticationMD5Password
                if (message.length === 12) ***REMOVED***
                    message.name = 'authenticationMD5Password';
                    const salt = this.reader.bytes(4);
                    return new messages_1.AuthenticationMD5Password(length, salt);
              ***REMOVED***
                break;
            case 10: // AuthenticationSASL
                message.name = 'authenticationSASL';
                message.mechanisms = [];
                let mechanism;
                do ***REMOVED***
                    mechanism = this.reader.cstring();
                    if (mechanism) ***REMOVED***
                        message.mechanisms.push(mechanism);
                  ***REMOVED***
              ***REMOVED*** while (mechanism);
                break;
            case 11: // AuthenticationSASLContinue
                message.name = 'authenticationSASLContinue';
                message.data = this.reader.string(length - 8);
                break;
            case 12: // AuthenticationSASLFinal
                message.name = 'authenticationSASLFinal';
                message.data = this.reader.string(length - 8);
                break;
            default:
                throw new Error('Unknown authenticationOk message type ' + code);
      ***REMOVED***
        return message;
  ***REMOVED***
    parseErrorMessage(offset, length, bytes, name) ***REMOVED***
        this.reader.setBuffer(offset, bytes);
        const fields = ***REMOVED***};
        let fieldType = this.reader.string(1);
        while (fieldType !== '\0') ***REMOVED***
            fields[fieldType] = this.reader.cstring();
            fieldType = this.reader.string(1);
      ***REMOVED***
        const messageValue = fields.M;
        const message = name === 'notice' ? new messages_1.NoticeMessage(length, messageValue) : new messages_1.DatabaseError(messageValue, length, name);
        message.severity = fields.S;
        message.code = fields.C;
        message.detail = fields.D;
        message.hint = fields.H;
        message.position = fields.P;
        message.internalPosition = fields.p;
        message.internalQuery = fields.q;
        message.where = fields.W;
        message.schema = fields.s;
        message.table = fields.t;
        message.column = fields.c;
        message.dataType = fields.d;
        message.constraint = fields.n;
        message.file = fields.F;
        message.line = fields.L;
        message.routine = fields.R;
        return message;
  ***REMOVED***
}
exports.Parser = Parser;
