// on-finished/index.js
/*!
 * on-finished
 * Copyright(c) 2013 Jonathan Ong
 * Copyright(c) 2014 Douglas Christopher Wilson
 * MIT Licensed
 */

'use strict'

/**
 * Module exports.
 * @public
 */

module.exports = onFinished
module.exports.isFinished = isFinished

/**
 * Module dependencies.
 * @private
 */

var asyncHooks = tryRequireAsyncHooks()
var first = require('ee-first')

/**
 * Variables.
 * @private
 */

/* istanbul ignore next */
var defer = typeof setImmediate === 'function'
  ? setImmediate
  : function (fn) ***REMOVED*** process.nextTick(fn.bind.apply(fn, arguments)) }

/**
 * Invoke callback when the response has finished, useful for
 * cleaning up resources afterwards.
 *
 * @param ***REMOVED***object} msg
 * @param ***REMOVED***function} listener
 * @return ***REMOVED***object}
 * @public
 */

function onFinished (msg, listener) ***REMOVED***
  if (isFinished(msg) !== false) ***REMOVED***
    defer(listener, null, msg)
    return msg
***REMOVED***

  // attach the listener to the message
  attachListener(msg, wrap(listener))

  return msg
}

/**
 * Determine if message is already finished.
 *
 * @param ***REMOVED***object} msg
 * @return ***REMOVED***boolean}
 * @public
 */

function isFinished (msg) ***REMOVED***
  var socket = msg.socket

  if (typeof msg.finished === 'boolean') ***REMOVED***
    // OutgoingMessage
    return Boolean(msg.finished || (socket && !socket.writable))
***REMOVED***

  if (typeof msg.complete === 'boolean') ***REMOVED***
    // IncomingMessage
    return Boolean(msg.upgrade || !socket || !socket.readable || (msg.complete && !msg.readable))
***REMOVED***

  // don't know
  return undefined
}

/**
 * Attach a finished listener to the message.
 *
 * @param ***REMOVED***object} msg
 * @param ***REMOVED***function} callback
 * @private
 */

function attachFinishedListener (msg, callback) ***REMOVED***
  var eeMsg
  var eeSocket
  var finished = false

  function onFinish (error) ***REMOVED***
    eeMsg.cancel()
    eeSocket.cancel()

    finished = true
    callback(error)
***REMOVED***

  // finished on first message event
  eeMsg = eeSocket = first([[msg, 'end', 'finish']], onFinish)

  function onSocket (socket) ***REMOVED***
    // remove listener
    msg.removeListener('socket', onSocket)

    if (finished) return
    if (eeMsg !== eeSocket) return

    // finished on first socket event
    eeSocket = first([[socket, 'error', 'close']], onFinish)
***REMOVED***

  if (msg.socket) ***REMOVED***
    // socket already assigned
    onSocket(msg.socket)
    return
***REMOVED***

  // wait for socket to be assigned
  msg.on('socket', onSocket)

  if (msg.socket === undefined) ***REMOVED***
    // istanbul ignore next: node.js 0.8 patch
    patchAssignSocket(msg, onSocket)
***REMOVED***
}

/**
 * Attach the listener to the message.
 *
 * @param ***REMOVED***object} msg
 * @return ***REMOVED***function}
 * @private
 */

function attachListener (msg, listener) ***REMOVED***
  var attached = msg.__onFinished

  // create a private single listener with queue
  if (!attached || !attached.queue) ***REMOVED***
    attached = msg.__onFinished = createListener(msg)
    attachFinishedListener(msg, attached)
***REMOVED***

  attached.queue.push(listener)
}

/**
 * Create listener on message.
 *
 * @param ***REMOVED***object} msg
 * @return ***REMOVED***function}
 * @private
 */

function createListener (msg) ***REMOVED***
  function listener (err) ***REMOVED***
    if (msg.__onFinished === listener) msg.__onFinished = null
    if (!listener.queue) return

    var queue = listener.queue
    listener.queue = null

    for (var i = 0; i < queue.length; i++) ***REMOVED***
      queue[i](err, msg)
  ***REMOVED***
***REMOVED***

  listener.queue = []

  return listener
}

/**
 * Patch ServerResponse.prototype.assignSocket for node.js 0.8.
 *
 * @param ***REMOVED***ServerResponse} res
 * @param ***REMOVED***function} callback
 * @private
 */

// istanbul ignore next: node.js 0.8 patch
function patchAssignSocket (res, callback) ***REMOVED***
  var assignSocket = res.assignSocket

  if (typeof assignSocket !== 'function') return

  // res.on('socket', callback) is broken in 0.8
  res.assignSocket = function _assignSocket (socket) ***REMOVED***
    assignSocket.call(this, socket)
    callback(socket)
***REMOVED***
}

/**
 * Try to require async_hooks
 * @private
 */

function tryRequireAsyncHooks () ***REMOVED***
  try ***REMOVED***
    return require('async_hooks')
***REMOVED*** catch (e) ***REMOVED***
    return ***REMOVED***}
***REMOVED***
}

/**
 * Wrap function with async resource, if possible.
 * AsyncResource.bind static method backported.
 * @private
 */

function wrap (fn) ***REMOVED***
  var res

  // create anonymous resource
  if (asyncHooks.AsyncResource) ***REMOVED***
    res = new asyncHooks.AsyncResource(fn.name || 'bound-anonymous-fn')
***REMOVED***

  // incompatible node.js
  if (!res || !res.runInAsyncScope) ***REMOVED***
    return fn
***REMOVED***

  // return bound function
  return res.runInAsyncScope.bind(res, fn, null)
}
