import Stylis, ***REMOVED***
  Context,
  PrefixContext,
  StylisUse,
  StylisSet
} from '@emotion/stylis'

new Stylis()
// $ExpectError
new Stylis(5)
// $ExpectError
new Stylis('abc')
// $ExpectError
new Stylis([])
new Stylis(***REMOVED***})
new Stylis(***REMOVED***
  // $ExpectError
  a: 5
})
new Stylis(***REMOVED***
  prefix: undefined
})
new Stylis(***REMOVED***
  prefix: true
})
new Stylis(***REMOVED***
  prefix: false
})
new Stylis(***REMOVED***
  prefix() ***REMOVED***
    return true
***REMOVED***
})
new Stylis(***REMOVED***
  prefix(key: string) ***REMOVED***
    return key === 'abc'
***REMOVED***
})
new Stylis(***REMOVED***
  prefix(key: string, value: string, context: PrefixContext) ***REMOVED***
    return value === 'world'
***REMOVED***
})
// $ExpectError
new Stylis(***REMOVED***
  prefix(key: string, value: string, context: PrefixContext) ***REMOVED***
    return 'hi'
***REMOVED***
})
// $ExpectError
new Stylis(***REMOVED***
  prefix(key: string, value: string, context: PrefixContext, a: any) ***REMOVED***
    return true
***REMOVED***
})

const stylis0 = new Stylis()
const stylis1: Stylis = stylis0
const stylis2: Stylis = new Stylis()
const stylis3 = new stylis2()
const stylis4: Stylis = stylis3

const stylis = new Stylis()

// $ExpectError
stylis.set()
stylis.set(***REMOVED***})
stylis.set(***REMOVED***
  prefix: true
})
stylis.set(***REMOVED***
  prefix(key: string) ***REMOVED***
    return false
***REMOVED***
})
stylis.set(***REMOVED***
  prefix(key: string, value: string, context: PrefixContext) ***REMOVED***
    return false
***REMOVED***
})
// $ExpectError
stylis.set(***REMOVED*** prefix: () => 'hi' })

stylis.use()
// $ExpectError
stylis.use(5)
// $ExpectError
stylis.use('ac')
stylis.use(true)
stylis.use(false)
stylis.use(null)
stylis.use(undefined)
// $ExpectError
stylis.use(***REMOVED***})
stylis.use([])
stylis.use(function() ***REMOVED***})
stylis.use(function(context) ***REMOVED***})
stylis.use(function(context: Context) ***REMOVED***
  return 'abc'
})
stylis.use(function(context: Context, content, selector, parent) ***REMOVED***
  const x: StylisUse = this.use
})
stylis.use(function(
  context: Context,
  content,
  selector,
  parent,
  line,
  column,
  length
) ***REMOVED***
  const x: StylisSet = this.set
})
stylis.use([
  function(
    context: Context,
    content,
    selector,
    parent,
    line,
    column,
    length,
    at,
    depth
  ) ***REMOVED***
    const x: StylisSet = this.set
***REMOVED***
])
// $ExpectError
stylis.use(function(
  context: Context,
  // $ExpectError
  content,
  // $ExpectError
  selector,
  // $ExpectError
  parent,
  // $ExpectError
  line,
  // $ExpectError
  column,
  // $ExpectError
  length,
  // $ExpectError
  at,
  // $ExpectError
  depth,
  // $ExpectError
  a
) ***REMOVED***
  return 'test'
})

// $ExpectError
stylis()
// $ExpectError
stylis('abc')
// $ExpectError
stylis('abc', 5)
// $ExpectError
stylis([], 'abc')
stylis('abc', 'cde')
