// gtoken/build/src/index.js
"use strict";
/**
 * Copyright 2018 Google LLC
 *
 * Distributed under MIT license.
 * See file LICENSE for detail or copy at https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.GoogleToken = void 0;
const fs = require("fs");
const gaxios_1 = require("gaxios");
const jws = require("jws");
const path = require("path");
const util_1 = require("util");
const readFile = fs.readFile
    ? (0, util_1.promisify)(fs.readFile)
    : async () => ***REMOVED***
        // if running in the web-browser, fs.readFile may not have been shimmed.
        throw new ErrorWithCode('use key rather than keyFile.', 'MISSING_CREDENTIALS');
  ***REMOVED***;
const GOOGLE_TOKEN_URL = 'https://www.googleapis.com/oauth2/v4/token';
const GOOGLE_REVOKE_TOKEN_URL = 'https://accounts.google.com/o/oauth2/revoke?token=';
class ErrorWithCode extends Error ***REMOVED***
    constructor(message, code) ***REMOVED***
        super(message);
        this.code = code;
  ***REMOVED***
}
let getPem;
class GoogleToken ***REMOVED***
    /**
     * Create a GoogleToken.
     *
     * @param options  Configuration object.
     */
    constructor(options) ***REMOVED***
        this.transporter = ***REMOVED***
            request: opts => (0, gaxios_1.request)(opts),
      ***REMOVED***;
        this.configure(options);
  ***REMOVED***
    get accessToken() ***REMOVED***
        return this.rawToken ? this.rawToken.access_token : undefined;
  ***REMOVED***
    get idToken() ***REMOVED***
        return this.rawToken ? this.rawToken.id_token : undefined;
  ***REMOVED***
    get tokenType() ***REMOVED***
        return this.rawToken ? this.rawToken.token_type : undefined;
  ***REMOVED***
    get refreshToken() ***REMOVED***
        return this.rawToken ? this.rawToken.refresh_token : undefined;
  ***REMOVED***
    /**
     * Returns whether the token has expired.
     *
     * @return true if the token has expired, false otherwise.
     */
    hasExpired() ***REMOVED***
        const now = new Date().getTime();
        if (this.rawToken && this.expiresAt) ***REMOVED***
            return now >= this.expiresAt;
      ***REMOVED***
        else ***REMOVED***
            return true;
      ***REMOVED***
  ***REMOVED***
    /**
     * Returns whether the token will expire within eagerRefreshThresholdMillis
     *
     * @return true if the token will be expired within eagerRefreshThresholdMillis, false otherwise.
     */
    isTokenExpiring() ***REMOVED***
        var _a;
        const now = new Date().getTime();
        const eagerRefreshThresholdMillis = (_a = this.eagerRefreshThresholdMillis) !== null && _a !== void 0 ? _a : 0;
        if (this.rawToken && this.expiresAt) ***REMOVED***
            return this.expiresAt <= now + eagerRefreshThresholdMillis;
      ***REMOVED***
        else ***REMOVED***
            return true;
      ***REMOVED***
  ***REMOVED***
    getToken(callback, opts = ***REMOVED***}) ***REMOVED***
        if (typeof callback === 'object') ***REMOVED***
            opts = callback;
            callback = undefined;
      ***REMOVED***
        opts = Object.assign(***REMOVED***
            forceRefresh: false,
      ***REMOVED***, opts);
        if (callback) ***REMOVED***
            const cb = callback;
            this.getTokenAsync(opts).then(t => cb(null, t), callback);
            return;
      ***REMOVED***
        return this.getTokenAsync(opts);
  ***REMOVED***
    /**
     * Given a keyFile, extract the key and client email if available
     * @param keyFile Path to a json, pem, or p12 file that contains the key.
     * @returns an object with privateKey and clientEmail properties
     */
    async getCredentials(keyFile) ***REMOVED***
        const ext = path.extname(keyFile);
        switch (ext) ***REMOVED***
            case '.json': ***REMOVED***
                const key = await readFile(keyFile, 'utf8');
                const body = JSON.parse(key);
                const privateKey = body.private_key;
                const clientEmail = body.client_email;
                if (!privateKey || !clientEmail) ***REMOVED***
                    throw new ErrorWithCode('private_key and client_email are required.', 'MISSING_CREDENTIALS');
              ***REMOVED***
                return ***REMOVED*** privateKey, clientEmail };
          ***REMOVED***
            case '.der':
            case '.crt':
            case '.pem': ***REMOVED***
                const privateKey = await readFile(keyFile, 'utf8');
                return ***REMOVED*** privateKey };
          ***REMOVED***
            case '.p12':
            case '.pfx': ***REMOVED***
                // NOTE:  The loading of `google-p12-pem` is deferred for performance
                // reasons.  The `node-forge` npm module in `google-p12-pem` adds a fair
                // bit time to overall module loading, and is likely not frequently
                // used.  In a future release, p12 support will be entirely removed.
                if (!getPem) ***REMOVED***
                    getPem = (await Promise.resolve().then(() => require('google-p12-pem'))).getPem;
              ***REMOVED***
                const privateKey = await getPem(keyFile);
                return ***REMOVED*** privateKey };
          ***REMOVED***
            default:
                throw new ErrorWithCode('Unknown certificate type. Type is determined based on file extension. ' +
                    'Current supported extensions are *.json, *.pem, and *.p12.', 'UNKNOWN_CERTIFICATE_TYPE');
      ***REMOVED***
  ***REMOVED***
    async getTokenAsync(opts) ***REMOVED***
        if (this.inFlightRequest && !opts.forceRefresh) ***REMOVED***
            return this.inFlightRequest;
      ***REMOVED***
        try ***REMOVED***
            return await (this.inFlightRequest = this.getTokenAsyncInner(opts));
      ***REMOVED***
        finally ***REMOVED***
            this.inFlightRequest = undefined;
      ***REMOVED***
  ***REMOVED***
    async getTokenAsyncInner(opts) ***REMOVED***
        if (this.isTokenExpiring() === false && opts.forceRefresh === false) ***REMOVED***
            return Promise.resolve(this.rawToken);
      ***REMOVED***
        if (!this.key && !this.keyFile) ***REMOVED***
            throw new Error('No key or keyFile set.');
      ***REMOVED***
        if (!this.key && this.keyFile) ***REMOVED***
            const creds = await this.getCredentials(this.keyFile);
            this.key = creds.privateKey;
            this.iss = creds.clientEmail || this.iss;
            if (!creds.clientEmail) ***REMOVED***
                this.ensureEmail();
          ***REMOVED***
      ***REMOVED***
        return this.requestToken();
  ***REMOVED***
    ensureEmail() ***REMOVED***
        if (!this.iss) ***REMOVED***
            throw new ErrorWithCode('email is required.', 'MISSING_CREDENTIALS');
      ***REMOVED***
  ***REMOVED***
    revokeToken(callback) ***REMOVED***
        if (callback) ***REMOVED***
            this.revokeTokenAsync().then(() => callback(), callback);
            return;
      ***REMOVED***
        return this.revokeTokenAsync();
  ***REMOVED***
    async revokeTokenAsync() ***REMOVED***
        if (!this.accessToken) ***REMOVED***
            throw new Error('No token to revoke.');
      ***REMOVED***
        const url = GOOGLE_REVOKE_TOKEN_URL + this.accessToken;
        await this.transporter.request(***REMOVED*** url });
        this.configure(***REMOVED***
            email: this.iss,
            sub: this.sub,
            key: this.key,
            keyFile: this.keyFile,
            scope: this.scope,
            additionalClaims: this.additionalClaims,
      ***REMOVED***);
  ***REMOVED***
    /**
     * Configure the GoogleToken for re-use.
     * @param  ***REMOVED***object} options Configuration object.
     */
    configure(options = ***REMOVED***}) ***REMOVED***
        this.keyFile = options.keyFile;
        this.key = options.key;
        this.rawToken = undefined;
        this.iss = options.email || options.iss;
        this.sub = options.sub;
        this.additionalClaims = options.additionalClaims;
        if (typeof options.scope === 'object') ***REMOVED***
            this.scope = options.scope.join(' ');
      ***REMOVED***
        else ***REMOVED***
            this.scope = options.scope;
      ***REMOVED***
        this.eagerRefreshThresholdMillis = options.eagerRefreshThresholdMillis;
        if (options.transporter) ***REMOVED***
            this.transporter = options.transporter;
      ***REMOVED***
  ***REMOVED***
    /**
     * Request the token from Google.
     */
    async requestToken() ***REMOVED***
        var _a, _b;
        const iat = Math.floor(new Date().getTime() / 1000);
        const additionalClaims = this.additionalClaims || ***REMOVED***};
        const payload = Object.assign(***REMOVED***
            iss: this.iss,
            scope: this.scope,
            aud: GOOGLE_TOKEN_URL,
            exp: iat + 3600,
            iat,
            sub: this.sub,
      ***REMOVED***, additionalClaims);
        const signedJWT = jws.sign(***REMOVED***
            header: ***REMOVED*** alg: 'RS256' },
            payload,
            secret: this.key,
      ***REMOVED***);
        try ***REMOVED***
            const r = await this.transporter.request(***REMOVED***
                method: 'POST',
                url: GOOGLE_TOKEN_URL,
                data: ***REMOVED***
                    grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
                    assertion: signedJWT,
              ***REMOVED***,
                headers: ***REMOVED*** 'Content-Type': 'application/x-www-form-urlencoded' },
                responseType: 'json',
          ***REMOVED***);
            this.rawToken = r.data;
            this.expiresAt =
                r.data.expires_in === null || r.data.expires_in === undefined
                    ? undefined
                    : (iat + r.data.expires_in) * 1000;
            return this.rawToken;
      ***REMOVED***
        catch (e) ***REMOVED***
            this.rawToken = undefined;
            this.tokenExpires = undefined;
            const body = e.response && ((_a = e.response) === null || _a === void 0 ? void 0 : _a.data)
                ? (_b = e.response) === null || _b === void 0 ? void 0 : _b.data
                : ***REMOVED***};
            if (body.error) ***REMOVED***
                const desc = body.error_description
                    ? `: $***REMOVED***body.error_description}`
                    : '';
                e.message = `$***REMOVED***body.error}$***REMOVED***desc}`;
          ***REMOVED***
            throw e;
      ***REMOVED***
  ***REMOVED***
}
exports.GoogleToken = GoogleToken;
