// websocket-driver/lib/websocket/driver.js
'use strict';

// Protocol references:
//
// * http://tools.ietf.org/html/draft-hixie-thewebsocketprotocol-75
// * http://tools.ietf.org/html/draft-hixie-thewebsocketprotocol-76
// * http://tools.ietf.org/html/draft-ietf-hybi-thewebsocketprotocol-17

var Base   = require('./driver/base'),
    Client = require('./driver/client'),
    Server = require('./driver/server');

var Driver = ***REMOVED***
  client: function(url, options) ***REMOVED***
    options = options || ***REMOVED***};
    if (options.masking === undefined) options.masking = true;
    return new Client(url, options);
***REMOVED***,

  server: function(options) ***REMOVED***
    options = options || ***REMOVED***};
    if (options.requireMasking === undefined) options.requireMasking = true;
    return new Server(options);
***REMOVED***,

  http: function() ***REMOVED***
    return Server.http.apply(Server, arguments);
***REMOVED***,

  isSecureRequest: function(request) ***REMOVED***
    return Server.isSecureRequest(request);
***REMOVED***,

  isWebSocket: function(request) ***REMOVED***
    return Base.isWebSocket(request);
***REMOVED***,

  validateOptions: function(options, validKeys) ***REMOVED***
    Base.validateOptions(options, validKeys);
***REMOVED***
};

module.exports = Driver;
