// firebase-admin/lib/firestore/firestore-internal.js
/*! firebase-admin v11.5.0 */
"use strict";
/*!
 * @license
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.getFirestoreOptions = exports.FirestoreService = void 0;
const error_1 = require("../utils/error");
const credential_internal_1 = require("../app/credential-internal");
const validator = require("../utils/validator");
const utils = require("../utils/index");
class FirestoreService ***REMOVED***
    constructor(app) ***REMOVED***
        this.databases = new Map();
        this.firestoreSettings = new Map();
        this.appInternal = app;
  ***REMOVED***
    initializeDatabase(databaseId, settings) ***REMOVED***
        const existingInstance = this.databases.get(databaseId);
        if (existingInstance) ***REMOVED***
            const initialSettings = this.firestoreSettings.get(databaseId) ?? ***REMOVED***};
            if (this.checkIfSameSettings(settings, initialSettings)) ***REMOVED***
                return existingInstance;
          ***REMOVED***
            throw new error_1.FirebaseFirestoreError(***REMOVED***
                code: 'failed-precondition',
                message: 'initializeFirestore() has already been called with ' +
                    'different options. To avoid this error, call initializeFirestore() with the ' +
                    'same options as when it was originally called, or call getFirestore() to return the' +
                    ' already initialized instance.'
          ***REMOVED***);
      ***REMOVED***
        const newInstance = initFirestore(this.app, databaseId, settings);
        this.databases.set(databaseId, newInstance);
        this.firestoreSettings.set(databaseId, settings);
        return newInstance;
  ***REMOVED***
    getDatabase(databaseId) ***REMOVED***
        let database = this.databases.get(databaseId);
        if (database === undefined) ***REMOVED***
            database = initFirestore(this.app, databaseId, ***REMOVED***});
            this.databases.set(databaseId, database);
            this.firestoreSettings.set(databaseId, ***REMOVED***});
      ***REMOVED***
        return database;
  ***REMOVED***
    checkIfSameSettings(settingsA, settingsB) ***REMOVED***
        const a = settingsA ?? ***REMOVED***};
        const b = settingsB ?? ***REMOVED***};
        // If we start passing more settings to Firestore constructor,
        // replace this with deep equality check.
        return (a.preferRest === b.preferRest);
  ***REMOVED***
    /**
     * Returns the app associated with this Storage instance.
     *
     * @returns The app associated with this Storage instance.
     */
    get app() ***REMOVED***
        return this.appInternal;
  ***REMOVED***
}
exports.FirestoreService = FirestoreService;
function getFirestoreOptions(app, firestoreSettings) ***REMOVED***
    if (!validator.isNonNullObject(app) || !('options' in app)) ***REMOVED***
        throw new error_1.FirebaseFirestoreError(***REMOVED***
            code: 'invalid-argument',
            message: 'First argument passed to admin.firestore() must be a valid Firebase app instance.',
      ***REMOVED***);
  ***REMOVED***
    const projectId = utils.getExplicitProjectId(app);
    const credential = app.options.credential;
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const ***REMOVED*** version: firebaseVersion } = require('../../package.json');
    const preferRest = firestoreSettings?.preferRest;
    if (credential instanceof credential_internal_1.ServiceAccountCredential) ***REMOVED***
        return ***REMOVED***
            credentials: ***REMOVED***
                private_key: credential.privateKey,
                client_email: credential.clientEmail,
          ***REMOVED***,
            // When the SDK is initialized with ServiceAccountCredentials an explicit projectId is
            // guaranteed to be available.
            projectId: projectId,
            firebaseVersion,
            preferRest,
      ***REMOVED***;
  ***REMOVED***
    else if ((0, credential_internal_1.isApplicationDefault)(app.options.credential)) ***REMOVED***
        // Try to use the Google application default credentials.
        // If an explicit project ID is not available, let Firestore client discover one from the
        // environment. This prevents the users from having to set GOOGLE_CLOUD_PROJECT in GCP runtimes.
        return validator.isNonEmptyString(projectId)
            ? ***REMOVED*** projectId, firebaseVersion, preferRest }
            : ***REMOVED*** firebaseVersion, preferRest };
  ***REMOVED***
    throw new error_1.FirebaseFirestoreError(***REMOVED***
        code: 'invalid-credential',
        message: 'Failed to initialize Google Cloud Firestore client with the available credentials. ' +
            'Must initialize the SDK with a certificate credential or application default credentials ' +
            'to use Cloud Firestore API.',
  ***REMOVED***);
}
exports.getFirestoreOptions = getFirestoreOptions;
function initFirestore(app, databaseId, firestoreSettings) ***REMOVED***
    const options = getFirestoreOptions(app, firestoreSettings);
    options.databaseId = databaseId;
    let firestoreDatabase;
    try ***REMOVED***
        // Lazy-load the Firestore implementation here, which in turns loads gRPC.
        firestoreDatabase = require('@google-cloud/firestore').Firestore;
  ***REMOVED***
    catch (err) ***REMOVED***
        throw new error_1.FirebaseFirestoreError(***REMOVED***
            code: 'missing-dependencies',
            message: 'Failed to import the Cloud Firestore client library for Node.js. '
                + 'Make sure to install the "@google-cloud/firestore" npm package. '
                + `Original error: $***REMOVED***err}`,
      ***REMOVED***);
  ***REMOVED***
    return new firestoreDatabase(options);
}
