// firebase-admin/lib/firestore/index.js
/*! firebase-admin v11.5.0 */
"use strict";
/*!
 * Copyright 2020 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.initializeFirestore = exports.getFirestore = exports.setLogFunction = exports.v1 = exports.WriteResult = exports.WriteBatch = exports.Transaction = exports.Timestamp = exports.QuerySnapshot = exports.QueryPartition = exports.QueryDocumentSnapshot = exports.Query = exports.GrpcStatus = exports.GeoPoint = exports.Firestore = exports.FieldValue = exports.FieldPath = exports.DocumentSnapshot = exports.DocumentReference = exports.CollectionReference = exports.CollectionGroup = exports.BundleBuilder = exports.BulkWriter = void 0;
const app_1 = require("../app");
const firestore_internal_1 = require("./firestore-internal");
const path_1 = require("@google-cloud/firestore/build/src/path");
var firestore_1 = require("@google-cloud/firestore");
Object.defineProperty(exports, "BulkWriter", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.BulkWriter; } });
Object.defineProperty(exports, "BundleBuilder", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.BundleBuilder; } });
Object.defineProperty(exports, "CollectionGroup", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.CollectionGroup; } });
Object.defineProperty(exports, "CollectionReference", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.CollectionReference; } });
Object.defineProperty(exports, "DocumentReference", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.DocumentReference; } });
Object.defineProperty(exports, "DocumentSnapshot", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.DocumentSnapshot; } });
Object.defineProperty(exports, "FieldPath", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.FieldPath; } });
Object.defineProperty(exports, "FieldValue", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.FieldValue; } });
Object.defineProperty(exports, "Firestore", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.Firestore; } });
Object.defineProperty(exports, "GeoPoint", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.GeoPoint; } });
Object.defineProperty(exports, "GrpcStatus", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.GrpcStatus; } });
Object.defineProperty(exports, "Query", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.Query; } });
Object.defineProperty(exports, "QueryDocumentSnapshot", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.QueryDocumentSnapshot; } });
Object.defineProperty(exports, "QueryPartition", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.QueryPartition; } });
Object.defineProperty(exports, "QuerySnapshot", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.QuerySnapshot; } });
Object.defineProperty(exports, "Timestamp", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.Timestamp; } });
Object.defineProperty(exports, "Transaction", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.Transaction; } });
Object.defineProperty(exports, "WriteBatch", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.WriteBatch; } });
Object.defineProperty(exports, "WriteResult", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.WriteResult; } });
Object.defineProperty(exports, "v1", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.v1; } });
Object.defineProperty(exports, "setLogFunction", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return firestore_1.setLogFunction; } });
function getFirestore(appOrDatabaseId, optionalDatabaseId) ***REMOVED***
    const app = typeof appOrDatabaseId === 'object' ? appOrDatabaseId : (0, app_1.getApp)();
    const databaseId = (typeof appOrDatabaseId === 'string' ? appOrDatabaseId : optionalDatabaseId) || path_1.DEFAULT_DATABASE_ID;
    const firebaseApp = app;
    const firestoreService = firebaseApp.getOrInitService('firestore', (app) => new firestore_internal_1.FirestoreService(app));
    return firestoreService.getDatabase(databaseId);
}
exports.getFirestore = getFirestore;
function initializeFirestore(app, settings, databaseId) ***REMOVED***
    settings ?? (settings = ***REMOVED***});
    databaseId ?? (databaseId = path_1.DEFAULT_DATABASE_ID);
    const firebaseApp = app;
    const firestoreService = firebaseApp.getOrInitService('firestore', (app) => new firestore_internal_1.FirestoreService(app));
    return firestoreService.initializeDatabase(databaseId, settings);
}
exports.initializeFirestore = initializeFirestore;
