// markdown-it/lib/rules_inline/strikethrough.js
// ~~strike through~~
//
'use strict';


// Insert each marker as a separate text token, and add it to delimiter list
//
module.exports.tokenize = function strikethrough(state, silent) ***REMOVED***
  var i, scanned, token, len, ch,
      start = state.pos,
      marker = state.src.charCodeAt(start);

  if (silent) ***REMOVED*** return false; }

  if (marker !== 0x7E/* ~ */) ***REMOVED*** return false; }

  scanned = state.scanDelims(state.pos, true);
  len = scanned.length;
  ch = String.fromCharCode(marker);

  if (len < 2) ***REMOVED*** return false; }

  if (len % 2) ***REMOVED***
    token         = state.push('text', '', 0);
    token.content = ch;
    len--;
***REMOVED***

  for (i = 0; i < len; i += 2) ***REMOVED***
    token         = state.push('text', '', 0);
    token.content = ch + ch;

    state.delimiters.push(***REMOVED***
      marker: marker,
      length: 0,     // disable "rule of 3" length checks meant for emphasis
      token:  state.tokens.length - 1,
      end:    -1,
      open:   scanned.can_open,
      close:  scanned.can_close
  ***REMOVED***);
***REMOVED***

  state.pos += scanned.length;

  return true;
};


function postProcess(state, delimiters) ***REMOVED***
  var i, j,
      startDelim,
      endDelim,
      token,
      loneMarkers = [],
      max = delimiters.length;

  for (i = 0; i < max; i++) ***REMOVED***
    startDelim = delimiters[i];

    if (startDelim.marker !== 0x7E/* ~ */) ***REMOVED***
      continue;
  ***REMOVED***

    if (startDelim.end === -1) ***REMOVED***
      continue;
  ***REMOVED***

    endDelim = delimiters[startDelim.end];

    token         = state.tokens[startDelim.token];
    token.type    = 's_open';
    token.tag     = 's';
    token.nesting = 1;
    token.markup  = '~~';
    token.content = '';

    token         = state.tokens[endDelim.token];
    token.type    = 's_close';
    token.tag     = 's';
    token.nesting = -1;
    token.markup  = '~~';
    token.content = '';

    if (state.tokens[endDelim.token - 1].type === 'text' &&
        state.tokens[endDelim.token - 1].content === '~') ***REMOVED***

      loneMarkers.push(endDelim.token - 1);
  ***REMOVED***
***REMOVED***

  // If a marker sequence has an odd number of characters, it's splitted
  // like this: `~~~~~` -> `~` + `~~` + `~~`, leaving one marker at the
  // start of the sequence.
  //
  // So, we have to move all those markers after subsequent s_close tags.
  //
  while (loneMarkers.length) ***REMOVED***
    i = loneMarkers.pop();
    j = i + 1;

    while (j < state.tokens.length && state.tokens[j].type === 's_close') ***REMOVED***
      j++;
  ***REMOVED***

    j--;

    if (i !== j) ***REMOVED***
      token = state.tokens[j];
      state.tokens[j] = state.tokens[i];
      state.tokens[i] = token;
  ***REMOVED***
***REMOVED***
}


// Walk through delimiter list and replace text tokens with tags
//
module.exports.postProcess = function strikethrough(state) ***REMOVED***
  var curr,
      tokens_meta = state.tokens_meta,
      max = state.tokens_meta.length;

  postProcess(state, state.delimiters);

  for (curr = 0; curr < max; curr++) ***REMOVED***
    if (tokens_meta[curr] && tokens_meta[curr].delimiters) ***REMOVED***
      postProcess(state, tokens_meta[curr].delimiters);
  ***REMOVED***
***REMOVED***
};
