// @fastify/busboy/lib/main.js
'use strict'

const WritableStream = require('stream').Writable
const ***REMOVED*** inherits } = require('util')
const Dicer = require('../deps/dicer/lib/Dicer')

const MultipartParser = require('./types/multipart')
const UrlencodedParser = require('./types/urlencoded')
const parseParams = require('./utils/parseParams')

function Busboy (opts) ***REMOVED***
  if (!(this instanceof Busboy)) ***REMOVED*** return new Busboy(opts) }

  if (typeof opts !== 'object') ***REMOVED***
    throw new TypeError('Busboy expected an options-Object.')
***REMOVED***
  if (typeof opts.headers !== 'object') ***REMOVED***
    throw new TypeError('Busboy expected an options-Object with headers-attribute.')
***REMOVED***
  if (typeof opts.headers['content-type'] !== 'string') ***REMOVED***
    throw new TypeError('Missing Content-Type-header.')
***REMOVED***

  const ***REMOVED***
    headers,
    ...streamOptions
***REMOVED*** = opts

  this.opts = ***REMOVED***
    autoDestroy: false,
    ...streamOptions
***REMOVED***
  WritableStream.call(this, this.opts)

  this._done = false
  this._parser = this.getParserByHeaders(headers)
  this._finished = false
}
inherits(Busboy, WritableStream)

Busboy.prototype.emit = function (ev) ***REMOVED***
  if (ev === 'finish') ***REMOVED***
    if (!this._done) ***REMOVED***
      this._parser && this._parser.end()
      return
  ***REMOVED*** else if (this._finished) ***REMOVED***
      return
  ***REMOVED***
    this._finished = true
***REMOVED***
  WritableStream.prototype.emit.apply(this, arguments)
}

Busboy.prototype.getParserByHeaders = function (headers) ***REMOVED***
  const parsed = parseParams(headers['content-type'])

  const cfg = ***REMOVED***
    defCharset: this.opts.defCharset,
    fileHwm: this.opts.fileHwm,
    headers,
    highWaterMark: this.opts.highWaterMark,
    isPartAFile: this.opts.isPartAFile,
    limits: this.opts.limits,
    parsedConType: parsed,
    preservePath: this.opts.preservePath
***REMOVED***

  if (MultipartParser.detect.test(parsed[0])) ***REMOVED***
    return new MultipartParser(this, cfg)
***REMOVED***
  if (UrlencodedParser.detect.test(parsed[0])) ***REMOVED***
    return new UrlencodedParser(this, cfg)
***REMOVED***
  throw new Error('Unsupported Content-Type.')
}

Busboy.prototype._write = function (chunk, encoding, cb) ***REMOVED***
  this._parser.write(chunk, cb)
}

module.exports = Busboy
module.exports.default = Busboy
module.exports.Busboy = Busboy

module.exports.Dicer = Dicer
