// @firebase/firestore-compat/dist/index.esm5.js
import firebase from '@firebase/app-compat';
import ***REMOVED*** __assign, __extends, __spreadArray } from 'tslib';
import ***REMOVED*** FirestoreError, Bytes, _isBase64Available, _logWarn, connectFirestoreEmulator, enableNetwork, disableNetwork, _validateIsNotUsedTogether, waitForPendingWrites, onSnapshotsInSync, collection, doc, collectionGroup, runTransaction, ensureFirestoreConfigured, WriteBatch as WriteBatch$1, executeWrite, loadBundle, namedQuery, AbstractUserDataWriter, DocumentReference as DocumentReference$1, _DocumentKey, refEqual, setDoc, updateDoc, deleteDoc, onSnapshot, getDocFromCache, getDocFromServer, getDoc, DocumentSnapshot as DocumentSnapshot$1, snapshotEqual, _debugAssert, addDoc, _DatabaseId, QueryDocumentSnapshot as QueryDocumentSnapshot$1, query, where, orderBy, limit, limitToLast, startAt, startAfter, endBefore, endAt, queryEqual, getDocsFromCache, getDocsFromServer, getDocs, QuerySnapshot as QuerySnapshot$1, _cast, enableIndexedDbPersistence, enableMultiTabIndexedDbPersistence, clearIndexedDbPersistence, setLogLevel as setLogLevel$1, _FieldPath, FieldPath as FieldPath$1, serverTimestamp, deleteField, arrayUnion, arrayRemove, increment, GeoPoint, Timestamp, CACHE_SIZE_UNLIMITED } from '@firebase/firestore';
import ***REMOVED*** getModularInstance } from '@firebase/util';
import ***REMOVED*** Component } from '@firebase/component';

const name = "@firebase/firestore-compat";
const version = "0.3.1";

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function validateSetOptions(methodName, options) ***REMOVED***
    if (options === undefined) ***REMOVED***
        return ***REMOVED***
            merge: false
      ***REMOVED***;
  ***REMOVED***
    if (options.mergeFields !== undefined && options.merge !== undefined) ***REMOVED***
        throw new FirestoreError('invalid-argument', "Invalid options passed to function ".concat(methodName, "(): You cannot ") +
            'specify both "merge" and "mergeFields".');
  ***REMOVED***
    return options;
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/** Helper function to assert Uint8Array is available at runtime. */
function assertUint8ArrayAvailable() ***REMOVED***
    if (typeof Uint8Array === 'undefined') ***REMOVED***
        throw new FirestoreError('unimplemented', 'Uint8Arrays are not available in this environment.');
  ***REMOVED***
}
/** Helper function to assert Base64 functions are available at runtime. */
function assertBase64Available() ***REMOVED***
    if (!_isBase64Available()) ***REMOVED***
        throw new FirestoreError('unimplemented', 'Blobs are unavailable in Firestore in this environment.');
  ***REMOVED***
}
/** Immutable class holding a blob (binary data) */
var Blob = /** @class */ (function () ***REMOVED***
    function Blob(_delegate) ***REMOVED***
        this._delegate = _delegate;
  ***REMOVED***
    Blob.fromBase64String = function (base64) ***REMOVED***
        assertBase64Available();
        return new Blob(Bytes.fromBase64String(base64));
  ***REMOVED***;
    Blob.fromUint8Array = function (array) ***REMOVED***
        assertUint8ArrayAvailable();
        return new Blob(Bytes.fromUint8Array(array));
  ***REMOVED***;
    Blob.prototype.toBase64 = function () ***REMOVED***
        assertBase64Available();
        return this._delegate.toBase64();
  ***REMOVED***;
    Blob.prototype.toUint8Array = function () ***REMOVED***
        assertUint8ArrayAvailable();
        return this._delegate.toUint8Array();
  ***REMOVED***;
    Blob.prototype.isEqual = function (other) ***REMOVED***
        return this._delegate.isEqual(other._delegate);
  ***REMOVED***;
    Blob.prototype.toString = function () ***REMOVED***
        return 'Blob(base64: ' + this.toBase64() + ')';
  ***REMOVED***;
    return Blob;
}());

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function isPartialObserver(obj) ***REMOVED***
    return implementsAnyMethods(obj, ['next', 'error', 'complete']);
}
/**
 * Returns true if obj is an object and contains at least one of the specified
 * methods.
 */
function implementsAnyMethods(obj, methods) ***REMOVED***
    if (typeof obj !== 'object' || obj === null) ***REMOVED***
        return false;
  ***REMOVED***
    var object = obj;
    for (var _i = 0, methods_1 = methods; _i < methods_1.length; _i++) ***REMOVED***
        var method = methods_1[_i];
        if (method in object && typeof object[method] === 'function') ***REMOVED***
            return true;
      ***REMOVED***
  ***REMOVED***
    return false;
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * The persistence provider included with the full Firestore SDK.
 */
var IndexedDbPersistenceProvider = /** @class */ (function () ***REMOVED***
    function IndexedDbPersistenceProvider() ***REMOVED***
  ***REMOVED***
    IndexedDbPersistenceProvider.prototype.enableIndexedDbPersistence = function (firestore, forceOwnership) ***REMOVED***
        return enableIndexedDbPersistence(firestore._delegate, ***REMOVED*** forceOwnership: forceOwnership });
  ***REMOVED***;
    IndexedDbPersistenceProvider.prototype.enableMultiTabIndexedDbPersistence = function (firestore) ***REMOVED***
        return enableMultiTabIndexedDbPersistence(firestore._delegate);
  ***REMOVED***;
    IndexedDbPersistenceProvider.prototype.clearIndexedDbPersistence = function (firestore) ***REMOVED***
        return clearIndexedDbPersistence(firestore._delegate);
  ***REMOVED***;
    return IndexedDbPersistenceProvider;
}());
/**
 * Compat class for Firestore. Exposes Firestore Legacy API, but delegates
 * to the functional API of firestore-exp.
 */
var Firestore = /** @class */ (function () ***REMOVED***
    function Firestore(databaseIdOrApp, _delegate, _persistenceProvider) ***REMOVED***
        var _this = this;
        this._delegate = _delegate;
        this._persistenceProvider = _persistenceProvider;
        this.INTERNAL = ***REMOVED***
            delete: function () ***REMOVED*** return _this.terminate(); }
      ***REMOVED***;
        if (!(databaseIdOrApp instanceof _DatabaseId)) ***REMOVED***
            this._appCompat = databaseIdOrApp;
      ***REMOVED***
  ***REMOVED***
    Object.defineProperty(Firestore.prototype, "_databaseId", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate._databaseId;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Firestore.prototype.settings = function (settingsLiteral) ***REMOVED***
        var currentSettings = this._delegate._getSettings();
        if (!settingsLiteral.merge &&
            currentSettings.host !== settingsLiteral.host) ***REMOVED***
            _logWarn('You are overriding the original host. If you did not intend ' +
                'to override your settings, use ***REMOVED***merge: true}.');
      ***REMOVED***
        if (settingsLiteral.merge) ***REMOVED***
            settingsLiteral = __assign(__assign(***REMOVED***}, currentSettings), settingsLiteral);
            // Remove the property from the settings once the merge is completed
            delete settingsLiteral.merge;
      ***REMOVED***
        this._delegate._setSettings(settingsLiteral);
  ***REMOVED***;
    Firestore.prototype.useEmulator = function (host, port, options) ***REMOVED***
        if (options === void 0) ***REMOVED*** options = ***REMOVED***}; }
        connectFirestoreEmulator(this._delegate, host, port, options);
  ***REMOVED***;
    Firestore.prototype.enableNetwork = function () ***REMOVED***
        return enableNetwork(this._delegate);
  ***REMOVED***;
    Firestore.prototype.disableNetwork = function () ***REMOVED***
        return disableNetwork(this._delegate);
  ***REMOVED***;
    Firestore.prototype.enablePersistence = function (settings) ***REMOVED***
        var synchronizeTabs = false;
        var experimentalForceOwningTab = false;
        if (settings) ***REMOVED***
            synchronizeTabs = !!settings.synchronizeTabs;
            experimentalForceOwningTab = !!settings.experimentalForceOwningTab;
            _validateIsNotUsedTogether('synchronizeTabs', synchronizeTabs, 'experimentalForceOwningTab', experimentalForceOwningTab);
      ***REMOVED***
        return synchronizeTabs
            ? this._persistenceProvider.enableMultiTabIndexedDbPersistence(this)
            : this._persistenceProvider.enableIndexedDbPersistence(this, experimentalForceOwningTab);
  ***REMOVED***;
    Firestore.prototype.clearPersistence = function () ***REMOVED***
        return this._persistenceProvider.clearIndexedDbPersistence(this);
  ***REMOVED***;
    Firestore.prototype.terminate = function () ***REMOVED***
        if (this._appCompat) ***REMOVED***
            this._appCompat._removeServiceInstance('firestore-compat');
            this._appCompat._removeServiceInstance('firestore');
      ***REMOVED***
        return this._delegate._delete();
  ***REMOVED***;
    Firestore.prototype.waitForPendingWrites = function () ***REMOVED***
        return waitForPendingWrites(this._delegate);
  ***REMOVED***;
    Firestore.prototype.onSnapshotsInSync = function (arg) ***REMOVED***
        return onSnapshotsInSync(this._delegate, arg);
  ***REMOVED***;
    Object.defineProperty(Firestore.prototype, "app", ***REMOVED***
        get: function () ***REMOVED***
            if (!this._appCompat) ***REMOVED***
                throw new FirestoreError('failed-precondition', "Firestore was not initialized using the Firebase SDK. 'app' is " +
                    'not available');
          ***REMOVED***
            return this._appCompat;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Firestore.prototype.collection = function (pathString) ***REMOVED***
        try ***REMOVED***
            return new CollectionReference(this, collection(this._delegate, pathString));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'collection()', 'Firestore.collection()');
      ***REMOVED***
  ***REMOVED***;
    Firestore.prototype.doc = function (pathString) ***REMOVED***
        try ***REMOVED***
            return new DocumentReference(this, doc(this._delegate, pathString));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'doc()', 'Firestore.doc()');
      ***REMOVED***
  ***REMOVED***;
    Firestore.prototype.collectionGroup = function (collectionId) ***REMOVED***
        try ***REMOVED***
            return new Query(this, collectionGroup(this._delegate, collectionId));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'collectionGroup()', 'Firestore.collectionGroup()');
      ***REMOVED***
  ***REMOVED***;
    Firestore.prototype.runTransaction = function (updateFunction) ***REMOVED***
        var _this = this;
        return runTransaction(this._delegate, function (transaction) ***REMOVED***
            return updateFunction(new Transaction(_this, transaction));
      ***REMOVED***);
  ***REMOVED***;
    Firestore.prototype.batch = function () ***REMOVED***
        var _this = this;
        ensureFirestoreConfigured(this._delegate);
        return new WriteBatch(new WriteBatch$1(this._delegate, function (mutations) ***REMOVED***
            return executeWrite(_this._delegate, mutations);
      ***REMOVED***));
  ***REMOVED***;
    Firestore.prototype.loadBundle = function (bundleData) ***REMOVED***
        return loadBundle(this._delegate, bundleData);
  ***REMOVED***;
    Firestore.prototype.namedQuery = function (name) ***REMOVED***
        var _this = this;
        return namedQuery(this._delegate, name).then(function (expQuery) ***REMOVED***
            if (!expQuery) ***REMOVED***
                return null;
          ***REMOVED***
            return new Query(_this, 
            // We can pass `expQuery` here directly since named queries don't have a UserDataConverter.
            // Otherwise, we would have to create a new ExpQuery and pass the old UserDataConverter.
            expQuery);
      ***REMOVED***);
  ***REMOVED***;
    return Firestore;
}());
var UserDataWriter = /** @class */ (function (_super) ***REMOVED***
    __extends(UserDataWriter, _super);
    function UserDataWriter(firestore) ***REMOVED***
        var _this = _super.call(this) || this;
        _this.firestore = firestore;
        return _this;
  ***REMOVED***
    UserDataWriter.prototype.convertBytes = function (bytes) ***REMOVED***
        return new Blob(new Bytes(bytes));
  ***REMOVED***;
    UserDataWriter.prototype.convertReference = function (name) ***REMOVED***
        var key = this.convertDocumentKey(name, this.firestore._databaseId);
        return DocumentReference.forKey(key, this.firestore, /* converter= */ null);
  ***REMOVED***;
    return UserDataWriter;
}(AbstractUserDataWriter));
function setLogLevel(level) ***REMOVED***
    setLogLevel$1(level);
}
/**
 * A reference to a transaction.
 */
var Transaction = /** @class */ (function () ***REMOVED***
    function Transaction(_firestore, _delegate) ***REMOVED***
        this._firestore = _firestore;
        this._delegate = _delegate;
        this._userDataWriter = new UserDataWriter(_firestore);
  ***REMOVED***
    Transaction.prototype.get = function (documentRef) ***REMOVED***
        var _this = this;
        var ref = castReference(documentRef);
        return this._delegate
            .get(ref)
            .then(function (result) ***REMOVED***
            return new DocumentSnapshot(_this._firestore, new DocumentSnapshot$1(_this._firestore._delegate, _this._userDataWriter, result._key, result._document, result.metadata, ref.converter));
      ***REMOVED***);
  ***REMOVED***;
    Transaction.prototype.set = function (documentRef, data, options) ***REMOVED***
        var ref = castReference(documentRef);
        if (options) ***REMOVED***
            validateSetOptions('Transaction.set', options);
            this._delegate.set(ref, data, options);
      ***REMOVED***
        else ***REMOVED***
            this._delegate.set(ref, data);
      ***REMOVED***
        return this;
  ***REMOVED***;
    Transaction.prototype.update = function (documentRef, dataOrField, value) ***REMOVED***
        var _a;
        var moreFieldsAndValues = [];
        for (var _i = 3; _i < arguments.length; _i++) ***REMOVED***
            moreFieldsAndValues[_i - 3] = arguments[_i];
      ***REMOVED***
        var ref = castReference(documentRef);
        if (arguments.length === 2) ***REMOVED***
            this._delegate.update(ref, dataOrField);
      ***REMOVED***
        else ***REMOVED***
            (_a = this._delegate).update.apply(_a, __spreadArray([ref,
                dataOrField,
                value], moreFieldsAndValues, false));
      ***REMOVED***
        return this;
  ***REMOVED***;
    Transaction.prototype.delete = function (documentRef) ***REMOVED***
        var ref = castReference(documentRef);
        this._delegate.delete(ref);
        return this;
  ***REMOVED***;
    return Transaction;
}());
var WriteBatch = /** @class */ (function () ***REMOVED***
    function WriteBatch(_delegate) ***REMOVED***
        this._delegate = _delegate;
  ***REMOVED***
    WriteBatch.prototype.set = function (documentRef, data, options) ***REMOVED***
        var ref = castReference(documentRef);
        if (options) ***REMOVED***
            validateSetOptions('WriteBatch.set', options);
            this._delegate.set(ref, data, options);
      ***REMOVED***
        else ***REMOVED***
            this._delegate.set(ref, data);
      ***REMOVED***
        return this;
  ***REMOVED***;
    WriteBatch.prototype.update = function (documentRef, dataOrField, value) ***REMOVED***
        var _a;
        var moreFieldsAndValues = [];
        for (var _i = 3; _i < arguments.length; _i++) ***REMOVED***
            moreFieldsAndValues[_i - 3] = arguments[_i];
      ***REMOVED***
        var ref = castReference(documentRef);
        if (arguments.length === 2) ***REMOVED***
            this._delegate.update(ref, dataOrField);
      ***REMOVED***
        else ***REMOVED***
            (_a = this._delegate).update.apply(_a, __spreadArray([ref,
                dataOrField,
                value], moreFieldsAndValues, false));
      ***REMOVED***
        return this;
  ***REMOVED***;
    WriteBatch.prototype.delete = function (documentRef) ***REMOVED***
        var ref = castReference(documentRef);
        this._delegate.delete(ref);
        return this;
  ***REMOVED***;
    WriteBatch.prototype.commit = function () ***REMOVED***
        return this._delegate.commit();
  ***REMOVED***;
    return WriteBatch;
}());
/**
 * Wraps a `PublicFirestoreDataConverter` translating the types from the
 * experimental SDK into corresponding types from the Classic SDK before passing
 * them to the wrapped converter.
 */
var FirestoreDataConverter = /** @class */ (function () ***REMOVED***
    function FirestoreDataConverter(_firestore, _userDataWriter, _delegate) ***REMOVED***
        this._firestore = _firestore;
        this._userDataWriter = _userDataWriter;
        this._delegate = _delegate;
  ***REMOVED***
    FirestoreDataConverter.prototype.fromFirestore = function (snapshot, options) ***REMOVED***
        var expSnapshot = new QueryDocumentSnapshot$1(this._firestore._delegate, this._userDataWriter, snapshot._key, snapshot._document, snapshot.metadata, 
        /* converter= */ null);
        return this._delegate.fromFirestore(new QueryDocumentSnapshot(this._firestore, expSnapshot), options !== null && options !== void 0 ? options : ***REMOVED***});
  ***REMOVED***;
    FirestoreDataConverter.prototype.toFirestore = function (modelObject, options) ***REMOVED***
        if (!options) ***REMOVED***
            return this._delegate.toFirestore(modelObject);
      ***REMOVED***
        else ***REMOVED***
            return this._delegate.toFirestore(modelObject, options);
      ***REMOVED***
  ***REMOVED***;
    // Use the same instance of `FirestoreDataConverter` for the given instances
    // of `Firestore` and `PublicFirestoreDataConverter` so that isEqual() will
    // compare equal for two objects created with the same converter instance.
    FirestoreDataConverter.getInstance = function (firestore, converter) ***REMOVED***
        var converterMapByFirestore = FirestoreDataConverter.INSTANCES;
        var untypedConverterByConverter = converterMapByFirestore.get(firestore);
        if (!untypedConverterByConverter) ***REMOVED***
            untypedConverterByConverter = new WeakMap();
            converterMapByFirestore.set(firestore, untypedConverterByConverter);
      ***REMOVED***
        var instance = untypedConverterByConverter.get(converter);
        if (!instance) ***REMOVED***
            instance = new FirestoreDataConverter(firestore, new UserDataWriter(firestore), converter);
            untypedConverterByConverter.set(converter, instance);
      ***REMOVED***
        return instance;
  ***REMOVED***;
    FirestoreDataConverter.INSTANCES = new WeakMap();
    return FirestoreDataConverter;
}());
/**
 * A reference to a particular document in a collection in the database.
 */
var DocumentReference = /** @class */ (function () ***REMOVED***
    function DocumentReference(firestore, _delegate) ***REMOVED***
        this.firestore = firestore;
        this._delegate = _delegate;
        this._userDataWriter = new UserDataWriter(firestore);
  ***REMOVED***
    DocumentReference.forPath = function (path, firestore, converter) ***REMOVED***
        if (path.length % 2 !== 0) ***REMOVED***
            throw new FirestoreError('invalid-argument', 'Invalid document reference. Document ' +
                'references must have an even number of segments, but ' +
                "".concat(path.canonicalString(), " has ").concat(path.length));
      ***REMOVED***
        return new DocumentReference(firestore, new DocumentReference$1(firestore._delegate, converter, new _DocumentKey(path)));
  ***REMOVED***;
    DocumentReference.forKey = function (key, firestore, converter) ***REMOVED***
        return new DocumentReference(firestore, new DocumentReference$1(firestore._delegate, converter, key));
  ***REMOVED***;
    Object.defineProperty(DocumentReference.prototype, "id", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.id;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(DocumentReference.prototype, "parent", ***REMOVED***
        get: function () ***REMOVED***
            return new CollectionReference(this.firestore, this._delegate.parent);
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(DocumentReference.prototype, "path", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.path;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    DocumentReference.prototype.collection = function (pathString) ***REMOVED***
        try ***REMOVED***
            return new CollectionReference(this.firestore, collection(this._delegate, pathString));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'collection()', 'DocumentReference.collection()');
      ***REMOVED***
  ***REMOVED***;
    DocumentReference.prototype.isEqual = function (other) ***REMOVED***
        other = getModularInstance(other);
        if (!(other instanceof DocumentReference$1)) ***REMOVED***
            return false;
      ***REMOVED***
        return refEqual(this._delegate, other);
  ***REMOVED***;
    DocumentReference.prototype.set = function (value, options) ***REMOVED***
        options = validateSetOptions('DocumentReference.set', options);
        try ***REMOVED***
            if (options) ***REMOVED***
                return setDoc(this._delegate, value, options);
          ***REMOVED***
            else ***REMOVED***
                return setDoc(this._delegate, value);
          ***REMOVED***
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'setDoc()', 'DocumentReference.set()');
      ***REMOVED***
  ***REMOVED***;
    DocumentReference.prototype.update = function (fieldOrUpdateData, value) ***REMOVED***
        var moreFieldsAndValues = [];
        for (var _i = 2; _i < arguments.length; _i++) ***REMOVED***
            moreFieldsAndValues[_i - 2] = arguments[_i];
      ***REMOVED***
        try ***REMOVED***
            if (arguments.length === 1) ***REMOVED***
                return updateDoc(this._delegate, fieldOrUpdateData);
          ***REMOVED***
            else ***REMOVED***
                return updateDoc.apply(void 0, __spreadArray([this._delegate,
                    fieldOrUpdateData,
                    value], moreFieldsAndValues, false));
          ***REMOVED***
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'updateDoc()', 'DocumentReference.update()');
      ***REMOVED***
  ***REMOVED***;
    DocumentReference.prototype.delete = function () ***REMOVED***
        return deleteDoc(this._delegate);
  ***REMOVED***;
    DocumentReference.prototype.onSnapshot = function () ***REMOVED***
        var _this = this;
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        var options = extractSnapshotOptions(args);
        var observer = wrapObserver(args, function (result) ***REMOVED***
            return new DocumentSnapshot(_this.firestore, new DocumentSnapshot$1(_this.firestore._delegate, _this._userDataWriter, result._key, result._document, result.metadata, _this._delegate.converter));
      ***REMOVED***);
        return onSnapshot(this._delegate, options, observer);
  ***REMOVED***;
    DocumentReference.prototype.get = function (options) ***REMOVED***
        var _this = this;
        var snap;
        if ((options === null || options === void 0 ? void 0 : options.source) === 'cache') ***REMOVED***
            snap = getDocFromCache(this._delegate);
      ***REMOVED***
        else if ((options === null || options === void 0 ? void 0 : options.source) === 'server') ***REMOVED***
            snap = getDocFromServer(this._delegate);
      ***REMOVED***
        else ***REMOVED***
            snap = getDoc(this._delegate);
      ***REMOVED***
        return snap.then(function (result) ***REMOVED***
            return new DocumentSnapshot(_this.firestore, new DocumentSnapshot$1(_this.firestore._delegate, _this._userDataWriter, result._key, result._document, result.metadata, _this._delegate.converter));
      ***REMOVED***);
  ***REMOVED***;
    DocumentReference.prototype.withConverter = function (converter) ***REMOVED***
        return new DocumentReference(this.firestore, converter
            ? this._delegate.withConverter(FirestoreDataConverter.getInstance(this.firestore, converter))
            : this._delegate.withConverter(null));
  ***REMOVED***;
    return DocumentReference;
}());
/**
 * Replaces the function name in an error thrown by the firestore-exp API
 * with the function names used in the classic API.
 */
function replaceFunctionName(e, original, updated) ***REMOVED***
    e.message = e.message.replace(original, updated);
    return e;
}
/**
 * Iterates the list of arguments from an `onSnapshot` call and returns the
 * first argument that may be an `SnapshotListenOptions` object. Returns an
 * empty object if none is found.
 */
function extractSnapshotOptions(args) ***REMOVED***
    for (var _i = 0, args_1 = args; _i < args_1.length; _i++) ***REMOVED***
        var arg = args_1[_i];
        if (typeof arg === 'object' && !isPartialObserver(arg)) ***REMOVED***
            return arg;
      ***REMOVED***
  ***REMOVED***
    return ***REMOVED***};
}
/**
 * Creates an observer that can be passed to the firestore-exp SDK. The
 * observer converts all observed values into the format expected by the classic
 * SDK.
 *
 * @param args - The list of arguments from an `onSnapshot` call.
 * @param wrapper - The function that converts the firestore-exp type into the
 * type used by this shim.
 */
function wrapObserver(args, wrapper) ***REMOVED***
    var _a, _b;
    var userObserver;
    if (isPartialObserver(args[0])) ***REMOVED***
        userObserver = args[0];
  ***REMOVED***
    else if (isPartialObserver(args[1])) ***REMOVED***
        userObserver = args[1];
  ***REMOVED***
    else if (typeof args[0] === 'function') ***REMOVED***
        userObserver = ***REMOVED***
            next: args[0],
            error: args[1],
            complete: args[2]
      ***REMOVED***;
  ***REMOVED***
    else ***REMOVED***
        userObserver = ***REMOVED***
            next: args[1],
            error: args[2],
            complete: args[3]
      ***REMOVED***;
  ***REMOVED***
    return ***REMOVED***
        next: function (val) ***REMOVED***
            if (userObserver.next) ***REMOVED***
                userObserver.next(wrapper(val));
          ***REMOVED***
      ***REMOVED***,
        error: (_a = userObserver.error) === null || _a === void 0 ? void 0 : _a.bind(userObserver),
        complete: (_b = userObserver.complete) === null || _b === void 0 ? void 0 : _b.bind(userObserver)
  ***REMOVED***;
}
var DocumentSnapshot = /** @class */ (function () ***REMOVED***
    function DocumentSnapshot(_firestore, _delegate) ***REMOVED***
        this._firestore = _firestore;
        this._delegate = _delegate;
  ***REMOVED***
    Object.defineProperty(DocumentSnapshot.prototype, "ref", ***REMOVED***
        get: function () ***REMOVED***
            return new DocumentReference(this._firestore, this._delegate.ref);
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(DocumentSnapshot.prototype, "id", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.id;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(DocumentSnapshot.prototype, "metadata", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.metadata;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(DocumentSnapshot.prototype, "exists", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.exists();
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    DocumentSnapshot.prototype.data = function (options) ***REMOVED***
        return this._delegate.data(options);
  ***REMOVED***;
    DocumentSnapshot.prototype.get = function (fieldPath, options
    // We are using `any` here to avoid an explicit cast by our users.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    ) ***REMOVED***
        return this._delegate.get(fieldPath, options);
  ***REMOVED***;
    DocumentSnapshot.prototype.isEqual = function (other) ***REMOVED***
        return snapshotEqual(this._delegate, other._delegate);
  ***REMOVED***;
    return DocumentSnapshot;
}());
var QueryDocumentSnapshot = /** @class */ (function (_super) ***REMOVED***
    __extends(QueryDocumentSnapshot, _super);
    function QueryDocumentSnapshot() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
  ***REMOVED***
    QueryDocumentSnapshot.prototype.data = function (options) ***REMOVED***
        var data = this._delegate.data(options);
        _debugAssert(data !== undefined, 'Document in a QueryDocumentSnapshot should exist');
        return data;
  ***REMOVED***;
    return QueryDocumentSnapshot;
}(DocumentSnapshot));
var Query = /** @class */ (function () ***REMOVED***
    function Query(firestore, _delegate) ***REMOVED***
        this.firestore = firestore;
        this._delegate = _delegate;
        this._userDataWriter = new UserDataWriter(firestore);
  ***REMOVED***
    Query.prototype.where = function (fieldPath, opStr, value) ***REMOVED***
        try ***REMOVED***
            // The "as string" cast is a little bit of a hack. `where` accepts the
            // FieldPath Compat type as input, but is not typed as such in order to
            // not expose this via our public typings file.
            return new Query(this.firestore, query(this._delegate, where(fieldPath, opStr, value)));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, /(orderBy|where)\(\)/, 'Query.$1()');
      ***REMOVED***
  ***REMOVED***;
    Query.prototype.orderBy = function (fieldPath, directionStr) ***REMOVED***
        try ***REMOVED***
            // The "as string" cast is a little bit of a hack. `orderBy` accepts the
            // FieldPath Compat type as input, but is not typed as such in order to
            // not expose this via our public typings file.
            return new Query(this.firestore, query(this._delegate, orderBy(fieldPath, directionStr)));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, /(orderBy|where)\(\)/, 'Query.$1()');
      ***REMOVED***
  ***REMOVED***;
    Query.prototype.limit = function (n) ***REMOVED***
        try ***REMOVED***
            return new Query(this.firestore, query(this._delegate, limit(n)));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'limit()', 'Query.limit()');
      ***REMOVED***
  ***REMOVED***;
    Query.prototype.limitToLast = function (n) ***REMOVED***
        try ***REMOVED***
            return new Query(this.firestore, query(this._delegate, limitToLast(n)));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'limitToLast()', 'Query.limitToLast()');
      ***REMOVED***
  ***REMOVED***;
    Query.prototype.startAt = function () ***REMOVED***
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        try ***REMOVED***
            return new Query(this.firestore, query(this._delegate, startAt.apply(void 0, args)));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'startAt()', 'Query.startAt()');
      ***REMOVED***
  ***REMOVED***;
    Query.prototype.startAfter = function () ***REMOVED***
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        try ***REMOVED***
            return new Query(this.firestore, query(this._delegate, startAfter.apply(void 0, args)));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'startAfter()', 'Query.startAfter()');
      ***REMOVED***
  ***REMOVED***;
    Query.prototype.endBefore = function () ***REMOVED***
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        try ***REMOVED***
            return new Query(this.firestore, query(this._delegate, endBefore.apply(void 0, args)));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'endBefore()', 'Query.endBefore()');
      ***REMOVED***
  ***REMOVED***;
    Query.prototype.endAt = function () ***REMOVED***
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        try ***REMOVED***
            return new Query(this.firestore, query(this._delegate, endAt.apply(void 0, args)));
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'endAt()', 'Query.endAt()');
      ***REMOVED***
  ***REMOVED***;
    Query.prototype.isEqual = function (other) ***REMOVED***
        return queryEqual(this._delegate, other._delegate);
  ***REMOVED***;
    Query.prototype.get = function (options) ***REMOVED***
        var _this = this;
        var query;
        if ((options === null || options === void 0 ? void 0 : options.source) === 'cache') ***REMOVED***
            query = getDocsFromCache(this._delegate);
      ***REMOVED***
        else if ((options === null || options === void 0 ? void 0 : options.source) === 'server') ***REMOVED***
            query = getDocsFromServer(this._delegate);
      ***REMOVED***
        else ***REMOVED***
            query = getDocs(this._delegate);
      ***REMOVED***
        return query.then(function (result) ***REMOVED***
            return new QuerySnapshot(_this.firestore, new QuerySnapshot$1(_this.firestore._delegate, _this._userDataWriter, _this._delegate, result._snapshot));
      ***REMOVED***);
  ***REMOVED***;
    Query.prototype.onSnapshot = function () ***REMOVED***
        var _this = this;
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        var options = extractSnapshotOptions(args);
        var observer = wrapObserver(args, function (snap) ***REMOVED***
            return new QuerySnapshot(_this.firestore, new QuerySnapshot$1(_this.firestore._delegate, _this._userDataWriter, _this._delegate, snap._snapshot));
      ***REMOVED***);
        return onSnapshot(this._delegate, options, observer);
  ***REMOVED***;
    Query.prototype.withConverter = function (converter) ***REMOVED***
        return new Query(this.firestore, converter
            ? this._delegate.withConverter(FirestoreDataConverter.getInstance(this.firestore, converter))
            : this._delegate.withConverter(null));
  ***REMOVED***;
    return Query;
}());
var DocumentChange = /** @class */ (function () ***REMOVED***
    function DocumentChange(_firestore, _delegate) ***REMOVED***
        this._firestore = _firestore;
        this._delegate = _delegate;
  ***REMOVED***
    Object.defineProperty(DocumentChange.prototype, "type", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.type;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(DocumentChange.prototype, "doc", ***REMOVED***
        get: function () ***REMOVED***
            return new QueryDocumentSnapshot(this._firestore, this._delegate.doc);
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(DocumentChange.prototype, "oldIndex", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.oldIndex;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(DocumentChange.prototype, "newIndex", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.newIndex;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    return DocumentChange;
}());
var QuerySnapshot = /** @class */ (function () ***REMOVED***
    function QuerySnapshot(_firestore, _delegate) ***REMOVED***
        this._firestore = _firestore;
        this._delegate = _delegate;
  ***REMOVED***
    Object.defineProperty(QuerySnapshot.prototype, "query", ***REMOVED***
        get: function () ***REMOVED***
            return new Query(this._firestore, this._delegate.query);
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(QuerySnapshot.prototype, "metadata", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.metadata;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(QuerySnapshot.prototype, "size", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.size;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(QuerySnapshot.prototype, "empty", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.empty;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(QuerySnapshot.prototype, "docs", ***REMOVED***
        get: function () ***REMOVED***
            var _this = this;
            return this._delegate.docs.map(function (doc) ***REMOVED*** return new QueryDocumentSnapshot(_this._firestore, doc); });
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    QuerySnapshot.prototype.docChanges = function (options) ***REMOVED***
        var _this = this;
        return this._delegate
            .docChanges(options)
            .map(function (docChange) ***REMOVED*** return new DocumentChange(_this._firestore, docChange); });
  ***REMOVED***;
    QuerySnapshot.prototype.forEach = function (callback, thisArg) ***REMOVED***
        var _this = this;
        this._delegate.forEach(function (snapshot) ***REMOVED***
            callback.call(thisArg, new QueryDocumentSnapshot(_this._firestore, snapshot));
      ***REMOVED***);
  ***REMOVED***;
    QuerySnapshot.prototype.isEqual = function (other) ***REMOVED***
        return snapshotEqual(this._delegate, other._delegate);
  ***REMOVED***;
    return QuerySnapshot;
}());
var CollectionReference = /** @class */ (function (_super) ***REMOVED***
    __extends(CollectionReference, _super);
    function CollectionReference(firestore, _delegate) ***REMOVED***
        var _this = _super.call(this, firestore, _delegate) || this;
        _this.firestore = firestore;
        _this._delegate = _delegate;
        return _this;
  ***REMOVED***
    Object.defineProperty(CollectionReference.prototype, "id", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.id;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(CollectionReference.prototype, "path", ***REMOVED***
        get: function () ***REMOVED***
            return this._delegate.path;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(CollectionReference.prototype, "parent", ***REMOVED***
        get: function () ***REMOVED***
            var docRef = this._delegate.parent;
            return docRef ? new DocumentReference(this.firestore, docRef) : null;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    CollectionReference.prototype.doc = function (documentPath) ***REMOVED***
        try ***REMOVED***
            if (documentPath === undefined) ***REMOVED***
                // Call `doc` without `documentPath` if `documentPath` is `undefined`
                // as `doc` validates the number of arguments to prevent users from
                // accidentally passing `undefined`.
                return new DocumentReference(this.firestore, doc(this._delegate));
          ***REMOVED***
            else ***REMOVED***
                return new DocumentReference(this.firestore, doc(this._delegate, documentPath));
          ***REMOVED***
      ***REMOVED***
        catch (e) ***REMOVED***
            throw replaceFunctionName(e, 'doc()', 'CollectionReference.doc()');
      ***REMOVED***
  ***REMOVED***;
    CollectionReference.prototype.add = function (data) ***REMOVED***
        var _this = this;
        return addDoc(this._delegate, data).then(function (docRef) ***REMOVED*** return new DocumentReference(_this.firestore, docRef); });
  ***REMOVED***;
    CollectionReference.prototype.isEqual = function (other) ***REMOVED***
        return refEqual(this._delegate, other._delegate);
  ***REMOVED***;
    CollectionReference.prototype.withConverter = function (converter) ***REMOVED***
        return new CollectionReference(this.firestore, converter
            ? this._delegate.withConverter(FirestoreDataConverter.getInstance(this.firestore, converter))
            : this._delegate.withConverter(null));
  ***REMOVED***;
    return CollectionReference;
}(Query));
function castReference(documentRef) ***REMOVED***
    return _cast(documentRef, DocumentReference$1);
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// The objects that are a part of this API are exposed to third-parties as
// compiled javascript so we want to flag our private members with a leading
// underscore to discourage their use.
/**
 * A `FieldPath` refers to a field in a document. The path may consist of a
 * single field name (referring to a top-level field in the document), or a list
 * of field names (referring to a nested field in the document).
 */
var FieldPath = /** @class */ (function () ***REMOVED***
    /**
     * Creates a FieldPath from the provided field names. If more than one field
     * name is provided, the path will point to a nested field in a document.
     *
     * @param fieldNames - A list of field names.
     */
    function FieldPath() ***REMOVED***
        var fieldNames = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            fieldNames[_i] = arguments[_i];
      ***REMOVED***
        this._delegate = new (FieldPath$1.bind.apply(FieldPath$1, __spreadArray([void 0], fieldNames, false)))();
  ***REMOVED***
    FieldPath.documentId = function () ***REMOVED***
        /**
         * Internal Note: The backend doesn't technically support querying by
         * document ID. Instead it queries by the entire document name (full path
         * included), but in the cases we currently support documentId(), the net
         * effect is the same.
         */
        return new FieldPath(_FieldPath.keyField().canonicalString());
  ***REMOVED***;
    FieldPath.prototype.isEqual = function (other) ***REMOVED***
        other = getModularInstance(other);
        if (!(other instanceof FieldPath$1)) ***REMOVED***
            return false;
      ***REMOVED***
        return this._delegate._internalPath.isEqual(other._internalPath);
  ***REMOVED***;
    return FieldPath;
}());

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var FieldValue = /** @class */ (function () ***REMOVED***
    function FieldValue(_delegate) ***REMOVED***
        this._delegate = _delegate;
  ***REMOVED***
    FieldValue.serverTimestamp = function () ***REMOVED***
        var delegate = serverTimestamp();
        delegate._methodName = 'FieldValue.serverTimestamp';
        return new FieldValue(delegate);
  ***REMOVED***;
    FieldValue.delete = function () ***REMOVED***
        var delegate = deleteField();
        delegate._methodName = 'FieldValue.delete';
        return new FieldValue(delegate);
  ***REMOVED***;
    FieldValue.arrayUnion = function () ***REMOVED***
        var elements = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            elements[_i] = arguments[_i];
      ***REMOVED***
        var delegate = arrayUnion.apply(void 0, elements);
        delegate._methodName = 'FieldValue.arrayUnion';
        return new FieldValue(delegate);
  ***REMOVED***;
    FieldValue.arrayRemove = function () ***REMOVED***
        var elements = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            elements[_i] = arguments[_i];
      ***REMOVED***
        var delegate = arrayRemove.apply(void 0, elements);
        delegate._methodName = 'FieldValue.arrayRemove';
        return new FieldValue(delegate);
  ***REMOVED***;
    FieldValue.increment = function (n) ***REMOVED***
        var delegate = increment(n);
        delegate._methodName = 'FieldValue.increment';
        return new FieldValue(delegate);
  ***REMOVED***;
    FieldValue.prototype.isEqual = function (other) ***REMOVED***
        return this._delegate.isEqual(other._delegate);
  ***REMOVED***;
    return FieldValue;
}());

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var firestoreNamespace = ***REMOVED***
    Firestore: Firestore,
    GeoPoint: GeoPoint,
    Timestamp: Timestamp,
    Blob: Blob,
    Transaction: Transaction,
    WriteBatch: WriteBatch,
    DocumentReference: DocumentReference,
    DocumentSnapshot: DocumentSnapshot,
    Query: Query,
    QueryDocumentSnapshot: QueryDocumentSnapshot,
    QuerySnapshot: QuerySnapshot,
    CollectionReference: CollectionReference,
    FieldPath: FieldPath,
    FieldValue: FieldValue,
    setLogLevel: setLogLevel,
    CACHE_SIZE_UNLIMITED: CACHE_SIZE_UNLIMITED
};
/**
 * Configures Firestore as part of the Firebase SDK by calling registerComponent.
 *
 * @param firebase - The FirebaseNamespace to register Firestore with
 * @param firestoreFactory - A factory function that returns a new Firestore
 *    instance.
 */
function configureForFirebase(firebase, firestoreFactory) ***REMOVED***
    firebase.INTERNAL.registerComponent(new Component('firestore-compat', function (container) ***REMOVED***
        var app = container.getProvider('app-compat').getImmediate();
        var firestoreExp = container.getProvider('firestore').getImmediate();
        return firestoreFactory(app, firestoreExp);
  ***REMOVED***, 'PUBLIC').setServiceProps(__assign(***REMOVED***}, firestoreNamespace)));
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Registers the main Firestore build with the components framework.
 * Persistence can be enabled via `firebase.firestore().enablePersistence()`.
 */
function registerFirestore(instance) ***REMOVED***
    configureForFirebase(instance, function (app, firestoreExp) ***REMOVED***
        return new Firestore(app, firestoreExp, new IndexedDbPersistenceProvider());
  ***REMOVED***);
    instance.registerVersion(name, version);
}
registerFirestore(firebase);

export ***REMOVED*** registerFirestore };
