// @babel/types/lib/builders/generated/uppercase.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
});
Object.defineProperty(exports, "AnyTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.anyTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "ArgumentPlaceholder", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.argumentPlaceholder;
***REMOVED***
});
Object.defineProperty(exports, "ArrayExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.arrayExpression;
***REMOVED***
});
Object.defineProperty(exports, "ArrayPattern", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.arrayPattern;
***REMOVED***
});
Object.defineProperty(exports, "ArrayTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.arrayTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "ArrowFunctionExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.arrowFunctionExpression;
***REMOVED***
});
Object.defineProperty(exports, "AssignmentExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.assignmentExpression;
***REMOVED***
});
Object.defineProperty(exports, "AssignmentPattern", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.assignmentPattern;
***REMOVED***
});
Object.defineProperty(exports, "AwaitExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.awaitExpression;
***REMOVED***
});
Object.defineProperty(exports, "BigIntLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.bigIntLiteral;
***REMOVED***
});
Object.defineProperty(exports, "BinaryExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.binaryExpression;
***REMOVED***
});
Object.defineProperty(exports, "BindExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.bindExpression;
***REMOVED***
});
Object.defineProperty(exports, "BlockStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.blockStatement;
***REMOVED***
});
Object.defineProperty(exports, "BooleanLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.booleanLiteral;
***REMOVED***
});
Object.defineProperty(exports, "BooleanLiteralTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.booleanLiteralTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "BooleanTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.booleanTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "BreakStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.breakStatement;
***REMOVED***
});
Object.defineProperty(exports, "CallExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.callExpression;
***REMOVED***
});
Object.defineProperty(exports, "CatchClause", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.catchClause;
***REMOVED***
});
Object.defineProperty(exports, "ClassAccessorProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.classAccessorProperty;
***REMOVED***
});
Object.defineProperty(exports, "ClassBody", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.classBody;
***REMOVED***
});
Object.defineProperty(exports, "ClassDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.classDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "ClassExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.classExpression;
***REMOVED***
});
Object.defineProperty(exports, "ClassImplements", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.classImplements;
***REMOVED***
});
Object.defineProperty(exports, "ClassMethod", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.classMethod;
***REMOVED***
});
Object.defineProperty(exports, "ClassPrivateMethod", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.classPrivateMethod;
***REMOVED***
});
Object.defineProperty(exports, "ClassPrivateProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.classPrivateProperty;
***REMOVED***
});
Object.defineProperty(exports, "ClassProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.classProperty;
***REMOVED***
});
Object.defineProperty(exports, "ConditionalExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.conditionalExpression;
***REMOVED***
});
Object.defineProperty(exports, "ContinueStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.continueStatement;
***REMOVED***
});
Object.defineProperty(exports, "DebuggerStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.debuggerStatement;
***REMOVED***
});
Object.defineProperty(exports, "DecimalLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.decimalLiteral;
***REMOVED***
});
Object.defineProperty(exports, "DeclareClass", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declareClass;
***REMOVED***
});
Object.defineProperty(exports, "DeclareExportAllDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declareExportAllDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "DeclareExportDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declareExportDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "DeclareFunction", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declareFunction;
***REMOVED***
});
Object.defineProperty(exports, "DeclareInterface", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declareInterface;
***REMOVED***
});
Object.defineProperty(exports, "DeclareModule", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declareModule;
***REMOVED***
});
Object.defineProperty(exports, "DeclareModuleExports", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declareModuleExports;
***REMOVED***
});
Object.defineProperty(exports, "DeclareOpaqueType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declareOpaqueType;
***REMOVED***
});
Object.defineProperty(exports, "DeclareTypeAlias", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declareTypeAlias;
***REMOVED***
});
Object.defineProperty(exports, "DeclareVariable", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declareVariable;
***REMOVED***
});
Object.defineProperty(exports, "DeclaredPredicate", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.declaredPredicate;
***REMOVED***
});
Object.defineProperty(exports, "Decorator", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.decorator;
***REMOVED***
});
Object.defineProperty(exports, "Directive", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.directive;
***REMOVED***
});
Object.defineProperty(exports, "DirectiveLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.directiveLiteral;
***REMOVED***
});
Object.defineProperty(exports, "DoExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.doExpression;
***REMOVED***
});
Object.defineProperty(exports, "DoWhileStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.doWhileStatement;
***REMOVED***
});
Object.defineProperty(exports, "EmptyStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.emptyStatement;
***REMOVED***
});
Object.defineProperty(exports, "EmptyTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.emptyTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "EnumBooleanBody", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.enumBooleanBody;
***REMOVED***
});
Object.defineProperty(exports, "EnumBooleanMember", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.enumBooleanMember;
***REMOVED***
});
Object.defineProperty(exports, "EnumDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.enumDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "EnumDefaultedMember", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.enumDefaultedMember;
***REMOVED***
});
Object.defineProperty(exports, "EnumNumberBody", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.enumNumberBody;
***REMOVED***
});
Object.defineProperty(exports, "EnumNumberMember", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.enumNumberMember;
***REMOVED***
});
Object.defineProperty(exports, "EnumStringBody", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.enumStringBody;
***REMOVED***
});
Object.defineProperty(exports, "EnumStringMember", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.enumStringMember;
***REMOVED***
});
Object.defineProperty(exports, "EnumSymbolBody", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.enumSymbolBody;
***REMOVED***
});
Object.defineProperty(exports, "ExistsTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.existsTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "ExportAllDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.exportAllDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "ExportDefaultDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.exportDefaultDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "ExportDefaultSpecifier", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.exportDefaultSpecifier;
***REMOVED***
});
Object.defineProperty(exports, "ExportNamedDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.exportNamedDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "ExportNamespaceSpecifier", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.exportNamespaceSpecifier;
***REMOVED***
});
Object.defineProperty(exports, "ExportSpecifier", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.exportSpecifier;
***REMOVED***
});
Object.defineProperty(exports, "ExpressionStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.expressionStatement;
***REMOVED***
});
Object.defineProperty(exports, "File", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.file;
***REMOVED***
});
Object.defineProperty(exports, "ForInStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.forInStatement;
***REMOVED***
});
Object.defineProperty(exports, "ForOfStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.forOfStatement;
***REMOVED***
});
Object.defineProperty(exports, "ForStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.forStatement;
***REMOVED***
});
Object.defineProperty(exports, "FunctionDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.functionDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "FunctionExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.functionExpression;
***REMOVED***
});
Object.defineProperty(exports, "FunctionTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.functionTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "FunctionTypeParam", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.functionTypeParam;
***REMOVED***
});
Object.defineProperty(exports, "GenericTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.genericTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "Identifier", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.identifier;
***REMOVED***
});
Object.defineProperty(exports, "IfStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.ifStatement;
***REMOVED***
});
Object.defineProperty(exports, "Import", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.import;
***REMOVED***
});
Object.defineProperty(exports, "ImportAttribute", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.importAttribute;
***REMOVED***
});
Object.defineProperty(exports, "ImportDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.importDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "ImportDefaultSpecifier", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.importDefaultSpecifier;
***REMOVED***
});
Object.defineProperty(exports, "ImportNamespaceSpecifier", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.importNamespaceSpecifier;
***REMOVED***
});
Object.defineProperty(exports, "ImportSpecifier", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.importSpecifier;
***REMOVED***
});
Object.defineProperty(exports, "IndexedAccessType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.indexedAccessType;
***REMOVED***
});
Object.defineProperty(exports, "InferredPredicate", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.inferredPredicate;
***REMOVED***
});
Object.defineProperty(exports, "InterfaceDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.interfaceDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "InterfaceExtends", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.interfaceExtends;
***REMOVED***
});
Object.defineProperty(exports, "InterfaceTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.interfaceTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "InterpreterDirective", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.interpreterDirective;
***REMOVED***
});
Object.defineProperty(exports, "IntersectionTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.intersectionTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "JSXAttribute", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxAttribute;
***REMOVED***
});
Object.defineProperty(exports, "JSXClosingElement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxClosingElement;
***REMOVED***
});
Object.defineProperty(exports, "JSXClosingFragment", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxClosingFragment;
***REMOVED***
});
Object.defineProperty(exports, "JSXElement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxElement;
***REMOVED***
});
Object.defineProperty(exports, "JSXEmptyExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxEmptyExpression;
***REMOVED***
});
Object.defineProperty(exports, "JSXExpressionContainer", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxExpressionContainer;
***REMOVED***
});
Object.defineProperty(exports, "JSXFragment", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxFragment;
***REMOVED***
});
Object.defineProperty(exports, "JSXIdentifier", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxIdentifier;
***REMOVED***
});
Object.defineProperty(exports, "JSXMemberExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxMemberExpression;
***REMOVED***
});
Object.defineProperty(exports, "JSXNamespacedName", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxNamespacedName;
***REMOVED***
});
Object.defineProperty(exports, "JSXOpeningElement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxOpeningElement;
***REMOVED***
});
Object.defineProperty(exports, "JSXOpeningFragment", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxOpeningFragment;
***REMOVED***
});
Object.defineProperty(exports, "JSXSpreadAttribute", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxSpreadAttribute;
***REMOVED***
});
Object.defineProperty(exports, "JSXSpreadChild", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxSpreadChild;
***REMOVED***
});
Object.defineProperty(exports, "JSXText", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.jsxText;
***REMOVED***
});
Object.defineProperty(exports, "LabeledStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.labeledStatement;
***REMOVED***
});
Object.defineProperty(exports, "LogicalExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.logicalExpression;
***REMOVED***
});
Object.defineProperty(exports, "MemberExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.memberExpression;
***REMOVED***
});
Object.defineProperty(exports, "MetaProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.metaProperty;
***REMOVED***
});
Object.defineProperty(exports, "MixedTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.mixedTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "ModuleExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.moduleExpression;
***REMOVED***
});
Object.defineProperty(exports, "NewExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.newExpression;
***REMOVED***
});
Object.defineProperty(exports, "Noop", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.noop;
***REMOVED***
});
Object.defineProperty(exports, "NullLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.nullLiteral;
***REMOVED***
});
Object.defineProperty(exports, "NullLiteralTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.nullLiteralTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "NullableTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.nullableTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "NumberLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.numberLiteral;
***REMOVED***
});
Object.defineProperty(exports, "NumberLiteralTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.numberLiteralTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "NumberTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.numberTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "NumericLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.numericLiteral;
***REMOVED***
});
Object.defineProperty(exports, "ObjectExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.objectExpression;
***REMOVED***
});
Object.defineProperty(exports, "ObjectMethod", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.objectMethod;
***REMOVED***
});
Object.defineProperty(exports, "ObjectPattern", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.objectPattern;
***REMOVED***
});
Object.defineProperty(exports, "ObjectProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.objectProperty;
***REMOVED***
});
Object.defineProperty(exports, "ObjectTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.objectTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "ObjectTypeCallProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.objectTypeCallProperty;
***REMOVED***
});
Object.defineProperty(exports, "ObjectTypeIndexer", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.objectTypeIndexer;
***REMOVED***
});
Object.defineProperty(exports, "ObjectTypeInternalSlot", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.objectTypeInternalSlot;
***REMOVED***
});
Object.defineProperty(exports, "ObjectTypeProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.objectTypeProperty;
***REMOVED***
});
Object.defineProperty(exports, "ObjectTypeSpreadProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.objectTypeSpreadProperty;
***REMOVED***
});
Object.defineProperty(exports, "OpaqueType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.opaqueType;
***REMOVED***
});
Object.defineProperty(exports, "OptionalCallExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.optionalCallExpression;
***REMOVED***
});
Object.defineProperty(exports, "OptionalIndexedAccessType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.optionalIndexedAccessType;
***REMOVED***
});
Object.defineProperty(exports, "OptionalMemberExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.optionalMemberExpression;
***REMOVED***
});
Object.defineProperty(exports, "ParenthesizedExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.parenthesizedExpression;
***REMOVED***
});
Object.defineProperty(exports, "PipelineBareFunction", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.pipelineBareFunction;
***REMOVED***
});
Object.defineProperty(exports, "PipelinePrimaryTopicReference", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.pipelinePrimaryTopicReference;
***REMOVED***
});
Object.defineProperty(exports, "PipelineTopicExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.pipelineTopicExpression;
***REMOVED***
});
Object.defineProperty(exports, "Placeholder", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.placeholder;
***REMOVED***
});
Object.defineProperty(exports, "PrivateName", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.privateName;
***REMOVED***
});
Object.defineProperty(exports, "Program", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.program;
***REMOVED***
});
Object.defineProperty(exports, "QualifiedTypeIdentifier", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.qualifiedTypeIdentifier;
***REMOVED***
});
Object.defineProperty(exports, "RecordExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.recordExpression;
***REMOVED***
});
Object.defineProperty(exports, "RegExpLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.regExpLiteral;
***REMOVED***
});
Object.defineProperty(exports, "RegexLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.regexLiteral;
***REMOVED***
});
Object.defineProperty(exports, "RestElement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.restElement;
***REMOVED***
});
Object.defineProperty(exports, "RestProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.restProperty;
***REMOVED***
});
Object.defineProperty(exports, "ReturnStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.returnStatement;
***REMOVED***
});
Object.defineProperty(exports, "SequenceExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.sequenceExpression;
***REMOVED***
});
Object.defineProperty(exports, "SpreadElement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.spreadElement;
***REMOVED***
});
Object.defineProperty(exports, "SpreadProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.spreadProperty;
***REMOVED***
});
Object.defineProperty(exports, "StaticBlock", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.staticBlock;
***REMOVED***
});
Object.defineProperty(exports, "StringLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.stringLiteral;
***REMOVED***
});
Object.defineProperty(exports, "StringLiteralTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.stringLiteralTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "StringTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.stringTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "Super", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.super;
***REMOVED***
});
Object.defineProperty(exports, "SwitchCase", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.switchCase;
***REMOVED***
});
Object.defineProperty(exports, "SwitchStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.switchStatement;
***REMOVED***
});
Object.defineProperty(exports, "SymbolTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.symbolTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "TSAnyKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsAnyKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSArrayType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsArrayType;
***REMOVED***
});
Object.defineProperty(exports, "TSAsExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsAsExpression;
***REMOVED***
});
Object.defineProperty(exports, "TSBigIntKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsBigIntKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSBooleanKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsBooleanKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSCallSignatureDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsCallSignatureDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "TSConditionalType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsConditionalType;
***REMOVED***
});
Object.defineProperty(exports, "TSConstructSignatureDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsConstructSignatureDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "TSConstructorType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsConstructorType;
***REMOVED***
});
Object.defineProperty(exports, "TSDeclareFunction", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsDeclareFunction;
***REMOVED***
});
Object.defineProperty(exports, "TSDeclareMethod", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsDeclareMethod;
***REMOVED***
});
Object.defineProperty(exports, "TSEnumDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsEnumDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "TSEnumMember", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsEnumMember;
***REMOVED***
});
Object.defineProperty(exports, "TSExportAssignment", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsExportAssignment;
***REMOVED***
});
Object.defineProperty(exports, "TSExpressionWithTypeArguments", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsExpressionWithTypeArguments;
***REMOVED***
});
Object.defineProperty(exports, "TSExternalModuleReference", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsExternalModuleReference;
***REMOVED***
});
Object.defineProperty(exports, "TSFunctionType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsFunctionType;
***REMOVED***
});
Object.defineProperty(exports, "TSImportEqualsDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsImportEqualsDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "TSImportType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsImportType;
***REMOVED***
});
Object.defineProperty(exports, "TSIndexSignature", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsIndexSignature;
***REMOVED***
});
Object.defineProperty(exports, "TSIndexedAccessType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsIndexedAccessType;
***REMOVED***
});
Object.defineProperty(exports, "TSInferType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsInferType;
***REMOVED***
});
Object.defineProperty(exports, "TSInstantiationExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsInstantiationExpression;
***REMOVED***
});
Object.defineProperty(exports, "TSInterfaceBody", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsInterfaceBody;
***REMOVED***
});
Object.defineProperty(exports, "TSInterfaceDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsInterfaceDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "TSIntersectionType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsIntersectionType;
***REMOVED***
});
Object.defineProperty(exports, "TSIntrinsicKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsIntrinsicKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSLiteralType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsLiteralType;
***REMOVED***
});
Object.defineProperty(exports, "TSMappedType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsMappedType;
***REMOVED***
});
Object.defineProperty(exports, "TSMethodSignature", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsMethodSignature;
***REMOVED***
});
Object.defineProperty(exports, "TSModuleBlock", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsModuleBlock;
***REMOVED***
});
Object.defineProperty(exports, "TSModuleDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsModuleDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "TSNamedTupleMember", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsNamedTupleMember;
***REMOVED***
});
Object.defineProperty(exports, "TSNamespaceExportDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsNamespaceExportDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "TSNeverKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsNeverKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSNonNullExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsNonNullExpression;
***REMOVED***
});
Object.defineProperty(exports, "TSNullKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsNullKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSNumberKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsNumberKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSObjectKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsObjectKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSOptionalType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsOptionalType;
***REMOVED***
});
Object.defineProperty(exports, "TSParameterProperty", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsParameterProperty;
***REMOVED***
});
Object.defineProperty(exports, "TSParenthesizedType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsParenthesizedType;
***REMOVED***
});
Object.defineProperty(exports, "TSPropertySignature", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsPropertySignature;
***REMOVED***
});
Object.defineProperty(exports, "TSQualifiedName", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsQualifiedName;
***REMOVED***
});
Object.defineProperty(exports, "TSRestType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsRestType;
***REMOVED***
});
Object.defineProperty(exports, "TSSatisfiesExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsSatisfiesExpression;
***REMOVED***
});
Object.defineProperty(exports, "TSStringKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsStringKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSSymbolKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsSymbolKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSThisType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsThisType;
***REMOVED***
});
Object.defineProperty(exports, "TSTupleType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTupleType;
***REMOVED***
});
Object.defineProperty(exports, "TSTypeAliasDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypeAliasDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "TSTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "TSTypeAssertion", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypeAssertion;
***REMOVED***
});
Object.defineProperty(exports, "TSTypeLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypeLiteral;
***REMOVED***
});
Object.defineProperty(exports, "TSTypeOperator", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypeOperator;
***REMOVED***
});
Object.defineProperty(exports, "TSTypeParameter", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypeParameter;
***REMOVED***
});
Object.defineProperty(exports, "TSTypeParameterDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypeParameterDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "TSTypeParameterInstantiation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypeParameterInstantiation;
***REMOVED***
});
Object.defineProperty(exports, "TSTypePredicate", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypePredicate;
***REMOVED***
});
Object.defineProperty(exports, "TSTypeQuery", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypeQuery;
***REMOVED***
});
Object.defineProperty(exports, "TSTypeReference", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsTypeReference;
***REMOVED***
});
Object.defineProperty(exports, "TSUndefinedKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsUndefinedKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSUnionType", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsUnionType;
***REMOVED***
});
Object.defineProperty(exports, "TSUnknownKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsUnknownKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TSVoidKeyword", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tsVoidKeyword;
***REMOVED***
});
Object.defineProperty(exports, "TaggedTemplateExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.taggedTemplateExpression;
***REMOVED***
});
Object.defineProperty(exports, "TemplateElement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.templateElement;
***REMOVED***
});
Object.defineProperty(exports, "TemplateLiteral", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.templateLiteral;
***REMOVED***
});
Object.defineProperty(exports, "ThisExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.thisExpression;
***REMOVED***
});
Object.defineProperty(exports, "ThisTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.thisTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "ThrowStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.throwStatement;
***REMOVED***
});
Object.defineProperty(exports, "TopicReference", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.topicReference;
***REMOVED***
});
Object.defineProperty(exports, "TryStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tryStatement;
***REMOVED***
});
Object.defineProperty(exports, "TupleExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tupleExpression;
***REMOVED***
});
Object.defineProperty(exports, "TupleTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.tupleTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "TypeAlias", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.typeAlias;
***REMOVED***
});
Object.defineProperty(exports, "TypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.typeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "TypeCastExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.typeCastExpression;
***REMOVED***
});
Object.defineProperty(exports, "TypeParameter", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.typeParameter;
***REMOVED***
});
Object.defineProperty(exports, "TypeParameterDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.typeParameterDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "TypeParameterInstantiation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.typeParameterInstantiation;
***REMOVED***
});
Object.defineProperty(exports, "TypeofTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.typeofTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "UnaryExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.unaryExpression;
***REMOVED***
});
Object.defineProperty(exports, "UnionTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.unionTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "UpdateExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.updateExpression;
***REMOVED***
});
Object.defineProperty(exports, "V8IntrinsicIdentifier", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.v8IntrinsicIdentifier;
***REMOVED***
});
Object.defineProperty(exports, "VariableDeclaration", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.variableDeclaration;
***REMOVED***
});
Object.defineProperty(exports, "VariableDeclarator", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.variableDeclarator;
***REMOVED***
});
Object.defineProperty(exports, "Variance", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.variance;
***REMOVED***
});
Object.defineProperty(exports, "VoidTypeAnnotation", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.voidTypeAnnotation;
***REMOVED***
});
Object.defineProperty(exports, "WhileStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.whileStatement;
***REMOVED***
});
Object.defineProperty(exports, "WithStatement", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.withStatement;
***REMOVED***
});
Object.defineProperty(exports, "YieldExpression", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _index.yieldExpression;
***REMOVED***
});
var _index = require("./index");
