// pgpass/lib/helper.js
'use strict';

var path = require('path')
  , Stream = require('stream').Stream
  , split = require('split2')
  , util = require('util')
  , defaultPort = 5432
  , isWin = (process.platform === 'win32')
  , warnStream = process.stderr
;


var S_IRWXG = 56     //    00070(8)
  , S_IRWXO = 7      //    00007(8)
  , S_IFMT  = 61440  // 00170000(8)
  , S_IFREG = 32768  //  0100000(8)
;
function isRegFile(mode) ***REMOVED***
    return ((mode & S_IFMT) == S_IFREG);
}

var fieldNames = [ 'host', 'port', 'database', 'user', 'password' ];
var nrOfFields = fieldNames.length;
var passKey = fieldNames[ nrOfFields -1 ];


function warn() ***REMOVED***
    var isWritable = (
        warnStream instanceof Stream &&
          true === warnStream.writable
    );

    if (isWritable) ***REMOVED***
        var args = Array.prototype.slice.call(arguments).concat("\n");
        warnStream.write( util.format.apply(util, args) );
  ***REMOVED***
}


Object.defineProperty(module.exports, 'isWin', ***REMOVED***
    get : function() ***REMOVED***
        return isWin;
  ***REMOVED*** ,
    set : function(val) ***REMOVED***
        isWin = val;
  ***REMOVED***
});


module.exports.warnTo = function(stream) ***REMOVED***
    var old = warnStream;
    warnStream = stream;
    return old;
};

module.exports.getFileName = function(rawEnv)***REMOVED***
    var env = rawEnv || process.env;
    var file = env.PGPASSFILE || (
        isWin ?
          path.join( env.APPDATA || './' , 'postgresql', 'pgpass.conf' ) :
          path.join( env.HOME || './', '.pgpass' )
    );
    return file;
};

module.exports.usePgPass = function(stats, fname) ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(process.env, 'PGPASSWORD')) ***REMOVED***
        return false;
  ***REMOVED***

    if (isWin) ***REMOVED***
        return true;
  ***REMOVED***

    fname = fname || '<unkn>';

    if (! isRegFile(stats.mode)) ***REMOVED***
        warn('WARNING: password file "%s" is not a plain file', fname);
        return false;
  ***REMOVED***

    if (stats.mode & (S_IRWXG | S_IRWXO)) ***REMOVED***
        /* If password file is insecure, alert the user and ignore it. */
        warn('WARNING: password file "%s" has group or world access; permissions should be u=rw (0600) or less', fname);
        return false;
  ***REMOVED***

    return true;
};


var matcher = module.exports.match = function(connInfo, entry) ***REMOVED***
    return fieldNames.slice(0, -1).reduce(function(prev, field, idx)***REMOVED***
        if (idx == 1) ***REMOVED***
            // the port
            if ( Number( connInfo[field] || defaultPort ) === Number( entry[field] ) ) ***REMOVED***
                return prev && true;
          ***REMOVED***
      ***REMOVED***
        return prev && (
            entry[field] === '*' ||
              entry[field] === connInfo[field]
        );
  ***REMOVED***, true);
};


module.exports.getPassword = function(connInfo, stream, cb) ***REMOVED***
    var pass;
    var lineStream = stream.pipe(split());

    function onLine(line) ***REMOVED***
        var entry = parseLine(line);
        if (entry && isValidEntry(entry) && matcher(connInfo, entry)) ***REMOVED***
            pass = entry[passKey];
            lineStream.end(); // -> calls onEnd(), but pass is set now
      ***REMOVED***
  ***REMOVED***

    var onEnd = function() ***REMOVED***
        stream.destroy();
        cb(pass);
  ***REMOVED***;

    var onErr = function(err) ***REMOVED***
        stream.destroy();
        warn('WARNING: error on reading file: %s', err);
        cb(undefined);
  ***REMOVED***;

    stream.on('error', onErr);
    lineStream
        .on('data', onLine)
        .on('end', onEnd)
        .on('error', onErr)
    ;

};


var parseLine = module.exports.parseLine = function(line) ***REMOVED***
    if (line.length < 11 || line.match(/^\s+#/)) ***REMOVED***
        return null;
  ***REMOVED***

    var curChar = '';
    var prevChar = '';
    var fieldIdx = 0;
    var startIdx = 0;
    var endIdx = 0;
    var obj = ***REMOVED***};
    var isLastField = false;
    var addToObj = function(idx, i0, i1) ***REMOVED***
        var field = line.substring(i0, i1);

        if (! Object.hasOwnProperty.call(process.env, 'PGPASS_NO_DEESCAPE')) ***REMOVED***
            field = field.replace(/\\([:\\])/g, '$1');
      ***REMOVED***

        obj[ fieldNames[idx] ] = field;
  ***REMOVED***;

    for (var i = 0 ; i < line.length-1 ; i += 1) ***REMOVED***
        curChar = line.charAt(i+1);
        prevChar = line.charAt(i);

        isLastField = (fieldIdx == nrOfFields-1);

        if (isLastField) ***REMOVED***
            addToObj(fieldIdx, startIdx);
            break;
      ***REMOVED***

        if (i >= 0 && curChar == ':' && prevChar !== '\\') ***REMOVED***
            addToObj(fieldIdx, startIdx, i+1);

            startIdx = i+2;
            fieldIdx += 1;
      ***REMOVED***
  ***REMOVED***

    obj = ( Object.keys(obj).length === nrOfFields ) ? obj : null;

    return obj;
};


var isValidEntry = module.exports.isValidEntry = function(entry)***REMOVED***
    var rules = ***REMOVED***
        // host
        0 : function(x)***REMOVED***
            return x.length > 0;
      ***REMOVED*** ,
        // port
        1 : function(x)***REMOVED***
            if (x === '*') ***REMOVED***
                return true;
          ***REMOVED***
            x = Number(x);
            return (
                isFinite(x) &&
                  x > 0 &&
                  x < 9007199254740992 &&
                  Math.floor(x) === x
            );
      ***REMOVED*** ,
        // database
        2 : function(x)***REMOVED***
            return x.length > 0;
      ***REMOVED*** ,
        // username
        3 : function(x)***REMOVED***
            return x.length > 0;
      ***REMOVED*** ,
        // password
        4 : function(x)***REMOVED***
            return x.length > 0;
      ***REMOVED***
  ***REMOVED***;

    for (var idx = 0 ; idx < fieldNames.length ; idx += 1) ***REMOVED***
        var rule = rules[idx];
        var value = entry[ fieldNames[idx] ] || '';

        var res = rule(value);
        if (!res) ***REMOVED***
            return false;
      ***REMOVED***
  ***REMOVED***

    return true;
};
