// @babel/generator/lib/node/parentheses.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
});
exports.ArrowFunctionExpression = ArrowFunctionExpression;
exports.AssignmentExpression = AssignmentExpression;
exports.Binary = Binary;
exports.BinaryExpression = BinaryExpression;
exports.ClassExpression = ClassExpression;
exports.ConditionalExpression = ConditionalExpression;
exports.DoExpression = DoExpression;
exports.FunctionExpression = FunctionExpression;
exports.FunctionTypeAnnotation = FunctionTypeAnnotation;
exports.Identifier = Identifier;
exports.LogicalExpression = LogicalExpression;
exports.NullableTypeAnnotation = NullableTypeAnnotation;
exports.ObjectExpression = ObjectExpression;
exports.OptionalIndexedAccessType = OptionalIndexedAccessType;
exports.OptionalCallExpression = exports.OptionalMemberExpression = OptionalMemberExpression;
exports.SequenceExpression = SequenceExpression;
exports.TSTypeAssertion = exports.TSSatisfiesExpression = exports.TSAsExpression = TSAsExpression;
exports.TSInferType = TSInferType;
exports.TSInstantiationExpression = TSInstantiationExpression;
exports.TSIntersectionType = exports.TSUnionType = TSUnionType;
exports.UnaryLike = UnaryLike;
exports.IntersectionTypeAnnotation = exports.UnionTypeAnnotation = UnionTypeAnnotation;
exports.UpdateExpression = UpdateExpression;
exports.AwaitExpression = exports.YieldExpression = YieldExpression;
var _t = require("@babel/types");
const ***REMOVED***
  isArrayTypeAnnotation,
  isArrowFunctionExpression,
  isAssignmentExpression,
  isAwaitExpression,
  isBinary,
  isBinaryExpression,
  isUpdateExpression,
  isCallExpression,
  isClass,
  isClassExpression,
  isConditional,
  isConditionalExpression,
  isExportDeclaration,
  isExportDefaultDeclaration,
  isExpressionStatement,
  isFor,
  isForInStatement,
  isForOfStatement,
  isForStatement,
  isFunctionExpression,
  isIfStatement,
  isIndexedAccessType,
  isIntersectionTypeAnnotation,
  isLogicalExpression,
  isMemberExpression,
  isNewExpression,
  isNullableTypeAnnotation,
  isObjectPattern,
  isOptionalCallExpression,
  isOptionalMemberExpression,
  isReturnStatement,
  isSequenceExpression,
  isSwitchStatement,
  isTSArrayType,
  isTSAsExpression,
  isTSInstantiationExpression,
  isTSIntersectionType,
  isTSNonNullExpression,
  isTSOptionalType,
  isTSRestType,
  isTSTypeAssertion,
  isTSUnionType,
  isTaggedTemplateExpression,
  isThrowStatement,
  isTypeAnnotation,
  isUnaryLike,
  isUnionTypeAnnotation,
  isVariableDeclarator,
  isWhileStatement,
  isYieldExpression,
  isTSSatisfiesExpression
} = _t;
const PRECEDENCE = ***REMOVED***
  "||": 0,
  "??": 0,
  "|>": 0,
  "&&": 1,
  "|": 2,
  "^": 3,
  "&": 4,
  "==": 5,
  "===": 5,
  "!=": 5,
  "!==": 5,
  "<": 6,
  ">": 6,
  "<=": 6,
  ">=": 6,
  in: 6,
  instanceof: 6,
  ">>": 7,
  "<<": 7,
  ">>>": 7,
  "+": 8,
  "-": 8,
  "*": 9,
  "/": 9,
  "%": 9,
  "**": 10
};
function isTSTypeExpression(node) ***REMOVED***
  return isTSAsExpression(node) || isTSSatisfiesExpression(node) || isTSTypeAssertion(node);
}
const isClassExtendsClause = (node, parent) => isClass(parent, ***REMOVED***
  superClass: node
});
const hasPostfixPart = (node, parent) => (isMemberExpression(parent) || isOptionalMemberExpression(parent)) && parent.object === node || (isCallExpression(parent) || isOptionalCallExpression(parent) || isNewExpression(parent)) && parent.callee === node || isTaggedTemplateExpression(parent) && parent.tag === node || isTSNonNullExpression(parent);
function NullableTypeAnnotation(node, parent) ***REMOVED***
  return isArrayTypeAnnotation(parent);
}
function FunctionTypeAnnotation(node, parent, printStack) ***REMOVED***
  if (printStack.length < 3) return;
  return (
    isUnionTypeAnnotation(parent) ||
    isIntersectionTypeAnnotation(parent) ||
    isArrayTypeAnnotation(parent) ||
    isTypeAnnotation(parent) &&
    isArrowFunctionExpression(printStack[printStack.length - 3])
  );
}
function UpdateExpression(node, parent) ***REMOVED***
  return hasPostfixPart(node, parent) || isClassExtendsClause(node, parent);
}
function ObjectExpression(node, parent, printStack) ***REMOVED***
  return isFirstInContext(printStack, 1 | 2);
}
function DoExpression(node, parent, printStack) ***REMOVED***
  return !node.async && isFirstInContext(printStack, 1);
}
function Binary(node, parent) ***REMOVED***
  if (node.operator === "**" && isBinaryExpression(parent, ***REMOVED***
    operator: "**"
***REMOVED***)) ***REMOVED***
    return parent.left === node;
***REMOVED***
  if (isClassExtendsClause(node, parent)) ***REMOVED***
    return true;
***REMOVED***
  if (hasPostfixPart(node, parent) || isUnaryLike(parent) || isAwaitExpression(parent)) ***REMOVED***
    return true;
***REMOVED***
  if (isBinary(parent)) ***REMOVED***
    const parentOp = parent.operator;
    const parentPos = PRECEDENCE[parentOp];
    const nodeOp = node.operator;
    const nodePos = PRECEDENCE[nodeOp];
    if (
    parentPos === nodePos && parent.right === node && !isLogicalExpression(parent) || parentPos > nodePos) ***REMOVED***
      return true;
  ***REMOVED***
***REMOVED***
}
function UnionTypeAnnotation(node, parent) ***REMOVED***
  return isArrayTypeAnnotation(parent) || isNullableTypeAnnotation(parent) || isIntersectionTypeAnnotation(parent) || isUnionTypeAnnotation(parent);
}
function OptionalIndexedAccessType(node, parent) ***REMOVED***
  return isIndexedAccessType(parent, ***REMOVED***
    objectType: node
***REMOVED***);
}
function TSAsExpression() ***REMOVED***
  return true;
}
function TSUnionType(node, parent) ***REMOVED***
  return isTSArrayType(parent) || isTSOptionalType(parent) || isTSIntersectionType(parent) || isTSUnionType(parent) || isTSRestType(parent);
}
function TSInferType(node, parent) ***REMOVED***
  return isTSArrayType(parent) || isTSOptionalType(parent);
}
function TSInstantiationExpression(node, parent) ***REMOVED***
  return (isCallExpression(parent) || isOptionalCallExpression(parent) || isNewExpression(parent) || isTSInstantiationExpression(parent)) && !!parent.typeParameters;
}
function BinaryExpression(node, parent) ***REMOVED***
  return node.operator === "in" && (isVariableDeclarator(parent) || isFor(parent));
}
function SequenceExpression(node, parent) ***REMOVED***
  if (
  isForStatement(parent) || isThrowStatement(parent) || isReturnStatement(parent) || isIfStatement(parent) && parent.test === node || isWhileStatement(parent) && parent.test === node || isForInStatement(parent) && parent.right === node || isSwitchStatement(parent) && parent.discriminant === node || isExpressionStatement(parent) && parent.expression === node) ***REMOVED***
    return false;
***REMOVED***

  return true;
}
function YieldExpression(node, parent) ***REMOVED***
  return isBinary(parent) || isUnaryLike(parent) || hasPostfixPart(node, parent) || isAwaitExpression(parent) && isYieldExpression(node) || isConditionalExpression(parent) && node === parent.test || isClassExtendsClause(node, parent);
}
function ClassExpression(node, parent, printStack) ***REMOVED***
  return isFirstInContext(printStack, 1 | 4);
}
function UnaryLike(node, parent) ***REMOVED***
  return hasPostfixPart(node, parent) || isBinaryExpression(parent, ***REMOVED***
    operator: "**",
    left: node
***REMOVED***) || isClassExtendsClause(node, parent);
}
function FunctionExpression(node, parent, printStack) ***REMOVED***
  return isFirstInContext(printStack, 1 | 4);
}
function ArrowFunctionExpression(node, parent) ***REMOVED***
  return isExportDeclaration(parent) || ConditionalExpression(node, parent);
}
function ConditionalExpression(node, parent) ***REMOVED***
  if (isUnaryLike(parent) || isBinary(parent) || isConditionalExpression(parent, ***REMOVED***
    test: node
***REMOVED***) || isAwaitExpression(parent) || isTSTypeExpression(parent)) ***REMOVED***
    return true;
***REMOVED***
  return UnaryLike(node, parent);
}
function OptionalMemberExpression(node, parent) ***REMOVED***
  return isCallExpression(parent, ***REMOVED***
    callee: node
***REMOVED***) || isMemberExpression(parent, ***REMOVED***
    object: node
***REMOVED***);
}
function AssignmentExpression(node, parent) ***REMOVED***
  if (isObjectPattern(node.left)) ***REMOVED***
    return true;
***REMOVED*** else ***REMOVED***
    return ConditionalExpression(node, parent);
***REMOVED***
}
function LogicalExpression(node, parent) ***REMOVED***
  if (isTSTypeExpression(parent)) return true;
  switch (node.operator) ***REMOVED***
    case "||":
      if (!isLogicalExpression(parent)) return false;
      return parent.operator === "??" || parent.operator === "&&";
    case "&&":
      return isLogicalExpression(parent, ***REMOVED***
        operator: "??"
    ***REMOVED***);
    case "??":
      return isLogicalExpression(parent) && parent.operator !== "??";
***REMOVED***
}
function Identifier(node, parent, printStack) ***REMOVED***
  var _node$extra;
  if ((_node$extra = node.extra) != null && _node$extra.parenthesized && isAssignmentExpression(parent, ***REMOVED***
    left: node
***REMOVED***) && (isFunctionExpression(parent.right) || isClassExpression(parent.right)) && parent.right.id == null) ***REMOVED***
    return true;
***REMOVED***
  if (node.name === "let") ***REMOVED***
    const isFollowedByBracket = isMemberExpression(parent, ***REMOVED***
      object: node,
      computed: true
  ***REMOVED***) || isOptionalMemberExpression(parent, ***REMOVED***
      object: node,
      computed: true,
      optional: false
  ***REMOVED***);
    return isFirstInContext(printStack, isFollowedByBracket ? 1 | 8 | 16 | 32 : 32);
***REMOVED***

  return node.name === "async" && isForOfStatement(parent) && node === parent.left;
}

function isFirstInContext(printStack, checkParam) ***REMOVED***
  const expressionStatement = checkParam & 1;
  const arrowBody = checkParam & 2;
  const exportDefault = checkParam & 4;
  const forHead = checkParam & 8;
  const forInHead = checkParam & 16;
  const forOfHead = checkParam & 32;
  let i = printStack.length - 1;
  if (i <= 0) return;
  let node = printStack[i];
  i--;
  let parent = printStack[i];
  while (i >= 0) ***REMOVED***
    if (expressionStatement && isExpressionStatement(parent, ***REMOVED***
      expression: node
  ***REMOVED***) || exportDefault && isExportDefaultDeclaration(parent, ***REMOVED***
      declaration: node
  ***REMOVED***) || arrowBody && isArrowFunctionExpression(parent, ***REMOVED***
      body: node
  ***REMOVED***) || forHead && isForStatement(parent, ***REMOVED***
      init: node
  ***REMOVED***) || forInHead && isForInStatement(parent, ***REMOVED***
      left: node
  ***REMOVED***) || forOfHead && isForOfStatement(parent, ***REMOVED***
      left: node
  ***REMOVED***)) ***REMOVED***
      return true;
  ***REMOVED***
    if (i > 0 && (hasPostfixPart(node, parent) && !isNewExpression(parent) || isSequenceExpression(parent) && parent.expressions[0] === node || isUpdateExpression(parent) && !parent.prefix || isConditional(parent, ***REMOVED***
      test: node
  ***REMOVED***) || isBinary(parent, ***REMOVED***
      left: node
  ***REMOVED***) || isAssignmentExpression(parent, ***REMOVED***
      left: node
  ***REMOVED***))) ***REMOVED***
      node = parent;
      i--;
      parent = printStack[i];
  ***REMOVED*** else ***REMOVED***
      return false;
  ***REMOVED***
***REMOVED***
  return false;
}
