// firebase-admin/lib/app-check/token-generator.js
/*! firebase-admin v11.5.0 */
"use strict";
/*!
 * @license
 * Copyright 2021 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.appCheckErrorFromCryptoSignerError = exports.AppCheckTokenGenerator = void 0;
const validator = require("../utils/validator");
const utils_1 = require("../utils");
const crypto_signer_1 = require("../utils/crypto-signer");
const app_check_api_client_internal_1 = require("./app-check-api-client-internal");
const ONE_MINUTE_IN_SECONDS = 60;
const ONE_MINUTE_IN_MILLIS = ONE_MINUTE_IN_SECONDS * 1000;
const ONE_DAY_IN_MILLIS = 24 * 60 * 60 * 1000;
// Audience to use for Firebase App Check Custom tokens
const FIREBASE_APP_CHECK_AUDIENCE = 'https://firebaseappcheck.googleapis.com/google.firebase.appcheck.v1.TokenExchangeService';
/**
 * Class for generating Firebase App Check tokens.
 *
 * @internal
 */
class AppCheckTokenGenerator ***REMOVED***
    /**
     * The AppCheckTokenGenerator class constructor.
     *
     * @param signer - The CryptoSigner instance for this token generator.
     * @constructor
     */
    constructor(signer) ***REMOVED***
        if (!validator.isNonNullObject(signer)) ***REMOVED***
            throw new app_check_api_client_internal_1.FirebaseAppCheckError('invalid-argument', 'INTERNAL ASSERT: Must provide a CryptoSigner to use AppCheckTokenGenerator.');
      ***REMOVED***
        this.signer = signer;
  ***REMOVED***
    /**
     * Creates a new custom token that can be exchanged to an App Check token.
     *
     * @param appId - The Application ID to use for the generated token.
     *
     * @returns A Promise fulfilled with a custom token signed with a service account key
     * that can be exchanged to an App Check token.
     */
    createCustomToken(appId, options) ***REMOVED***
        if (!validator.isNonEmptyString(appId)) ***REMOVED***
            throw new app_check_api_client_internal_1.FirebaseAppCheckError('invalid-argument', '`appId` must be a non-empty string.');
      ***REMOVED***
        let customOptions = ***REMOVED***};
        if (typeof options !== 'undefined') ***REMOVED***
            customOptions = this.validateTokenOptions(options);
      ***REMOVED***
        return this.signer.getAccountId().then((account) => ***REMOVED***
            const header = ***REMOVED***
                alg: this.signer.algorithm,
                typ: 'JWT',
          ***REMOVED***;
            const iat = Math.floor(Date.now() / 1000);
            const body = ***REMOVED***
                iss: account,
                sub: account,
                app_id: appId,
                aud: FIREBASE_APP_CHECK_AUDIENCE,
                exp: iat + (ONE_MINUTE_IN_SECONDS * 5),
                iat,
                ...customOptions,
          ***REMOVED***;
            const token = `$***REMOVED***this.encodeSegment(header)}.$***REMOVED***this.encodeSegment(body)}`;
            return this.signer.sign(Buffer.from(token))
                .then((signature) => ***REMOVED***
                return `$***REMOVED***token}.$***REMOVED***this.encodeSegment(signature)}`;
          ***REMOVED***);
      ***REMOVED***).catch((err) => ***REMOVED***
            throw appCheckErrorFromCryptoSignerError(err);
      ***REMOVED***);
  ***REMOVED***
    encodeSegment(segment) ***REMOVED***
        const buffer = (segment instanceof Buffer) ? segment : Buffer.from(JSON.stringify(segment));
        return (0, utils_1.toWebSafeBase64)(buffer).replace(/=+$/, '');
  ***REMOVED***
    /**
     * Checks if a given `AppCheckTokenOptions` object is valid. If successful, returns an object with
     * custom properties.
     *
     * @param options - An options object to be validated.
     * @returns A custom object with ttl converted to protobuf Duration string format.
     */
    validateTokenOptions(options) ***REMOVED***
        if (!validator.isNonNullObject(options)) ***REMOVED***
            throw new app_check_api_client_internal_1.FirebaseAppCheckError('invalid-argument', 'AppCheckTokenOptions must be a non-null object.');
      ***REMOVED***
        if (typeof options.ttlMillis !== 'undefined') ***REMOVED***
            if (!validator.isNumber(options.ttlMillis)) ***REMOVED***
                throw new app_check_api_client_internal_1.FirebaseAppCheckError('invalid-argument', 'ttlMillis must be a duration in milliseconds.');
          ***REMOVED***
            // ttlMillis must be between 30 minutes and 7 days (inclusive)
            if (options.ttlMillis < (ONE_MINUTE_IN_MILLIS * 30) || options.ttlMillis > (ONE_DAY_IN_MILLIS * 7)) ***REMOVED***
                throw new app_check_api_client_internal_1.FirebaseAppCheckError('invalid-argument', 'ttlMillis must be a duration in milliseconds between 30 minutes and 7 days (inclusive).');
          ***REMOVED***
            return ***REMOVED*** ttl: (0, utils_1.transformMillisecondsToSecondsString)(options.ttlMillis) };
      ***REMOVED***
        return ***REMOVED***};
  ***REMOVED***
}
exports.AppCheckTokenGenerator = AppCheckTokenGenerator;
/**
 * Creates a new `FirebaseAppCheckError` by extracting the error code, message and other relevant
 * details from a `CryptoSignerError`.
 *
 * @param err - The Error to convert into a `FirebaseAppCheckError` error
 * @returns A Firebase App Check error that can be returned to the user.
 */
function appCheckErrorFromCryptoSignerError(err) ***REMOVED***
    if (!(err instanceof crypto_signer_1.CryptoSignerError)) ***REMOVED***
        return err;
  ***REMOVED***
    if (err.code === crypto_signer_1.CryptoSignerErrorCode.SERVER_ERROR && validator.isNonNullObject(err.cause)) ***REMOVED***
        const httpError = err.cause;
        const errorResponse = httpError.response.data;
        if (errorResponse?.error) ***REMOVED***
            const status = errorResponse.error.status;
            const description = errorResponse.error.message || JSON.stringify(httpError.response);
            let code = 'unknown-error';
            if (status && status in app_check_api_client_internal_1.APP_CHECK_ERROR_CODE_MAPPING) ***REMOVED***
                code = app_check_api_client_internal_1.APP_CHECK_ERROR_CODE_MAPPING[status];
          ***REMOVED***
            return new app_check_api_client_internal_1.FirebaseAppCheckError(code, `Error returned from server while signing a custom token: $***REMOVED***description}`);
      ***REMOVED***
        return new app_check_api_client_internal_1.FirebaseAppCheckError('internal-error', 'Error returned from server: ' + JSON.stringify(errorResponse) + '.');
  ***REMOVED***
    return new app_check_api_client_internal_1.FirebaseAppCheckError(mapToAppCheckErrorCode(err.code), err.message);
}
exports.appCheckErrorFromCryptoSignerError = appCheckErrorFromCryptoSignerError;
function mapToAppCheckErrorCode(code) ***REMOVED***
    switch (code) ***REMOVED***
        case crypto_signer_1.CryptoSignerErrorCode.INVALID_CREDENTIAL:
            return 'invalid-credential';
        case crypto_signer_1.CryptoSignerErrorCode.INVALID_ARGUMENT:
            return 'invalid-argument';
        default:
            return 'internal-error';
  ***REMOVED***
}
