// packet-reader/index.js
var assert = require('assert')

var Reader = module.exports = function(options) ***REMOVED***
  //TODO - remove for version 1.0
  if(typeof options == 'number') ***REMOVED***
    options = ***REMOVED*** headerSize: options }
***REMOVED***
  options = options || ***REMOVED***}
  this.offset = 0
  this.lastChunk = false
  this.chunk = null
  this.chunkLength = 0
  this.headerSize = options.headerSize || 0
  this.lengthPadding = options.lengthPadding || 0
  this.header = null
  assert(this.headerSize < 2, 'pre-length header of more than 1 byte length not currently supported')
}

Reader.prototype.addChunk = function(chunk) ***REMOVED***
  if (!this.chunk || this.offset === this.chunkLength) ***REMOVED***
    this.chunk = chunk
    this.chunkLength = chunk.length
    this.offset = 0
    return
***REMOVED***

  var newChunkLength = chunk.length
  var newLength = this.chunkLength + newChunkLength

  if (newLength > this.chunk.length) ***REMOVED***
    var newBufferLength = this.chunk.length * 2
    while (newLength >= newBufferLength) ***REMOVED***
      newBufferLength *= 2
  ***REMOVED***
    var newBuffer = Buffer.alloc(newBufferLength)
    this.chunk.copy(newBuffer)
    this.chunk = newBuffer
***REMOVED***
  chunk.copy(this.chunk, this.chunkLength)
  this.chunkLength = newLength
}

Reader.prototype.read = function() ***REMOVED***
  if(this.chunkLength < (this.headerSize + 4 + this.offset)) ***REMOVED***
    return false
***REMOVED***

  if(this.headerSize) ***REMOVED***
    this.header = this.chunk[this.offset]
***REMOVED***

  //read length of next item
  var length = this.chunk.readUInt32BE(this.offset + this.headerSize) + this.lengthPadding

  //next item spans more chunks than we have
  var remaining = this.chunkLength - (this.offset + 4 + this.headerSize)
  if(length > remaining) ***REMOVED***
    return false
***REMOVED***

  this.offset += (this.headerSize + 4)
  var result = this.chunk.slice(this.offset, this.offset + length)
  this.offset += length
  return result
}
