// pg-protocol/src/messages.ts
export type Mode = 'text' | 'binary'

export type MessageName =
  | 'parseComplete'
  | 'bindComplete'
  | 'closeComplete'
  | 'noData'
  | 'portalSuspended'
  | 'replicationStart'
  | 'emptyQuery'
  | 'copyDone'
  | 'copyData'
  | 'rowDescription'
  | 'parameterDescription'
  | 'parameterStatus'
  | 'backendKeyData'
  | 'notification'
  | 'readyForQuery'
  | 'commandComplete'
  | 'dataRow'
  | 'copyInResponse'
  | 'copyOutResponse'
  | 'authenticationOk'
  | 'authenticationMD5Password'
  | 'authenticationCleartextPassword'
  | 'authenticationSASL'
  | 'authenticationSASLContinue'
  | 'authenticationSASLFinal'
  | 'error'
  | 'notice'

export interface BackendMessage ***REMOVED***
  name: MessageName
  length: number
}

export const parseComplete: BackendMessage = ***REMOVED***
  name: 'parseComplete',
  length: 5,
}

export const bindComplete: BackendMessage = ***REMOVED***
  name: 'bindComplete',
  length: 5,
}

export const closeComplete: BackendMessage = ***REMOVED***
  name: 'closeComplete',
  length: 5,
}

export const noData: BackendMessage = ***REMOVED***
  name: 'noData',
  length: 5,
}

export const portalSuspended: BackendMessage = ***REMOVED***
  name: 'portalSuspended',
  length: 5,
}

export const replicationStart: BackendMessage = ***REMOVED***
  name: 'replicationStart',
  length: 4,
}

export const emptyQuery: BackendMessage = ***REMOVED***
  name: 'emptyQuery',
  length: 4,
}

export const copyDone: BackendMessage = ***REMOVED***
  name: 'copyDone',
  length: 4,
}

interface NoticeOrError ***REMOVED***
  message: string | undefined
  severity: string | undefined
  code: string | undefined
  detail: string | undefined
  hint: string | undefined
  position: string | undefined
  internalPosition: string | undefined
  internalQuery: string | undefined
  where: string | undefined
  schema: string | undefined
  table: string | undefined
  column: string | undefined
  dataType: string | undefined
  constraint: string | undefined
  file: string | undefined
  line: string | undefined
  routine: string | undefined
}

export class DatabaseError extends Error implements NoticeOrError ***REMOVED***
  public severity: string | undefined
  public code: string | undefined
  public detail: string | undefined
  public hint: string | undefined
  public position: string | undefined
  public internalPosition: string | undefined
  public internalQuery: string | undefined
  public where: string | undefined
  public schema: string | undefined
  public table: string | undefined
  public column: string | undefined
  public dataType: string | undefined
  public constraint: string | undefined
  public file: string | undefined
  public line: string | undefined
  public routine: string | undefined
  constructor(message: string, public readonly length: number, public readonly name: MessageName) ***REMOVED***
    super(message)
***REMOVED***
}

export class CopyDataMessage ***REMOVED***
  public readonly name = 'copyData'
  constructor(public readonly length: number, public readonly chunk: Buffer) ***REMOVED***}
}

export class CopyResponse ***REMOVED***
  public readonly columnTypes: number[]
  constructor(
    public readonly length: number,
    public readonly name: MessageName,
    public readonly binary: boolean,
    columnCount: number
  ) ***REMOVED***
    this.columnTypes = new Array(columnCount)
***REMOVED***
}

export class Field ***REMOVED***
  constructor(
    public readonly name: string,
    public readonly tableID: number,
    public readonly columnID: number,
    public readonly dataTypeID: number,
    public readonly dataTypeSize: number,
    public readonly dataTypeModifier: number,
    public readonly format: Mode
  ) ***REMOVED***}
}

export class RowDescriptionMessage ***REMOVED***
  public readonly name: MessageName = 'rowDescription'
  public readonly fields: Field[]
  constructor(public readonly length: number, public readonly fieldCount: number) ***REMOVED***
    this.fields = new Array(this.fieldCount)
***REMOVED***
}

export class ParameterDescriptionMessage ***REMOVED***
  public readonly name: MessageName = 'parameterDescription'
  public readonly dataTypeIDs: number[]
  constructor(public readonly length: number, public readonly parameterCount: number) ***REMOVED***
    this.dataTypeIDs = new Array(this.parameterCount)
***REMOVED***
}

export class ParameterStatusMessage ***REMOVED***
  public readonly name: MessageName = 'parameterStatus'
  constructor(
    public readonly length: number,
    public readonly parameterName: string,
    public readonly parameterValue: string
  ) ***REMOVED***}
}

export class AuthenticationMD5Password implements BackendMessage ***REMOVED***
  public readonly name: MessageName = 'authenticationMD5Password'
  constructor(public readonly length: number, public readonly salt: Buffer) ***REMOVED***}
}

export class BackendKeyDataMessage ***REMOVED***
  public readonly name: MessageName = 'backendKeyData'
  constructor(public readonly length: number, public readonly processID: number, public readonly secretKey: number) ***REMOVED***}
}

export class NotificationResponseMessage ***REMOVED***
  public readonly name: MessageName = 'notification'
  constructor(
    public readonly length: number,
    public readonly processId: number,
    public readonly channel: string,
    public readonly payload: string
  ) ***REMOVED***}
}

export class ReadyForQueryMessage ***REMOVED***
  public readonly name: MessageName = 'readyForQuery'
  constructor(public readonly length: number, public readonly status: string) ***REMOVED***}
}

export class CommandCompleteMessage ***REMOVED***
  public readonly name: MessageName = 'commandComplete'
  constructor(public readonly length: number, public readonly text: string) ***REMOVED***}
}

export class DataRowMessage ***REMOVED***
  public readonly fieldCount: number
  public readonly name: MessageName = 'dataRow'
  constructor(public length: number, public fields: any[]) ***REMOVED***
    this.fieldCount = fields.length
***REMOVED***
}

export class NoticeMessage implements BackendMessage, NoticeOrError ***REMOVED***
  constructor(public readonly length: number, public readonly message: string | undefined) ***REMOVED***}
  public readonly name = 'notice'
  public severity: string | undefined
  public code: string | undefined
  public detail: string | undefined
  public hint: string | undefined
  public position: string | undefined
  public internalPosition: string | undefined
  public internalQuery: string | undefined
  public where: string | undefined
  public schema: string | undefined
  public table: string | undefined
  public column: string | undefined
  public dataType: string | undefined
  public constraint: string | undefined
  public file: string | undefined
  public line: string | undefined
  public routine: string | undefined
}
