// abort-controller/dist/abort-controller.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
'use strict';

Object.defineProperty(exports, '__esModule', ***REMOVED*** value: true });

var eventTargetShim = require('event-target-shim');

/**
 * The signal class.
 * @see https://dom.spec.whatwg.org/#abortsignal
 */
class AbortSignal extends eventTargetShim.EventTarget ***REMOVED***
    /**
     * AbortSignal cannot be constructed directly.
     */
    constructor() ***REMOVED***
        super();
        throw new TypeError("AbortSignal cannot be constructed directly");
  ***REMOVED***
    /**
     * Returns `true` if this `AbortSignal`'s `AbortController` has signaled to abort, and `false` otherwise.
     */
    get aborted() ***REMOVED***
        const aborted = abortedFlags.get(this);
        if (typeof aborted !== "boolean") ***REMOVED***
            throw new TypeError(`Expected 'this' to be an 'AbortSignal' object, but got $***REMOVED***this === null ? "null" : typeof this}`);
      ***REMOVED***
        return aborted;
  ***REMOVED***
}
eventTargetShim.defineEventAttribute(AbortSignal.prototype, "abort");
/**
 * Create an AbortSignal object.
 */
function createAbortSignal() ***REMOVED***
    const signal = Object.create(AbortSignal.prototype);
    eventTargetShim.EventTarget.call(signal);
    abortedFlags.set(signal, false);
    return signal;
}
/**
 * Abort a given signal.
 */
function abortSignal(signal) ***REMOVED***
    if (abortedFlags.get(signal) !== false) ***REMOVED***
        return;
  ***REMOVED***
    abortedFlags.set(signal, true);
    signal.dispatchEvent(***REMOVED*** type: "abort" });
}
/**
 * Aborted flag for each instances.
 */
const abortedFlags = new WeakMap();
// Properties should be enumerable.
Object.defineProperties(AbortSignal.prototype, ***REMOVED***
    aborted: ***REMOVED*** enumerable: true },
});
// `toString()` should return `"[object AbortSignal]"`
if (typeof Symbol === "function" && typeof Symbol.toStringTag === "symbol") ***REMOVED***
    Object.defineProperty(AbortSignal.prototype, Symbol.toStringTag, ***REMOVED***
        configurable: true,
        value: "AbortSignal",
  ***REMOVED***);
}

/**
 * The AbortController.
 * @see https://dom.spec.whatwg.org/#abortcontroller
 */
class AbortController ***REMOVED***
    /**
     * Initialize this controller.
     */
    constructor() ***REMOVED***
        signals.set(this, createAbortSignal());
  ***REMOVED***
    /**
     * Returns the `AbortSignal` object associated with this object.
     */
    get signal() ***REMOVED***
        return getSignal(this);
  ***REMOVED***
    /**
     * Abort and signal to any observers that the associated activity is to be aborted.
     */
    abort() ***REMOVED***
        abortSignal(getSignal(this));
  ***REMOVED***
}
/**
 * Associated signals.
 */
const signals = new WeakMap();
/**
 * Get the associated signal of a given controller.
 */
function getSignal(controller) ***REMOVED***
    const signal = signals.get(controller);
    if (signal == null) ***REMOVED***
        throw new TypeError(`Expected 'this' to be an 'AbortController' object, but got $***REMOVED***controller === null ? "null" : typeof controller}`);
  ***REMOVED***
    return signal;
}
// Properties should be enumerable.
Object.defineProperties(AbortController.prototype, ***REMOVED***
    signal: ***REMOVED*** enumerable: true },
    abort: ***REMOVED*** enumerable: true },
});
if (typeof Symbol === "function" && typeof Symbol.toStringTag === "symbol") ***REMOVED***
    Object.defineProperty(AbortController.prototype, Symbol.toStringTag, ***REMOVED***
        configurable: true,
        value: "AbortController",
  ***REMOVED***);
}

exports.AbortController = AbortController;
exports.AbortSignal = AbortSignal;
exports.default = AbortController;

module.exports = AbortController
module.exports.AbortController = module.exports["default"] = AbortController
module.exports.AbortSignal = AbortSignal
