// jose/dist/browser/runtime/pbes2kw.js
import random from './random.js';
import ***REMOVED*** p2s as concatSalt } from '../lib/buffer_utils.js';
import ***REMOVED*** encode as base64url } from './base64url.js';
import ***REMOVED*** wrap, unwrap } from './aeskw.js';
import checkP2s from '../lib/check_p2s.js';
import crypto, ***REMOVED*** isCryptoKey } from './webcrypto.js';
import ***REMOVED*** checkEncCryptoKey } from '../lib/crypto_key.js';
import invalidKeyInput from '../lib/invalid_key_input.js';
import ***REMOVED*** types } from './is_key_like.js';
function getCryptoKey(key, alg) ***REMOVED***
    if (key instanceof Uint8Array) ***REMOVED***
        return crypto.subtle.importKey('raw', key, 'PBKDF2', false, ['deriveBits']);
  ***REMOVED***
    if (isCryptoKey(key)) ***REMOVED***
        checkEncCryptoKey(key, alg, 'deriveBits', 'deriveKey');
        return key;
  ***REMOVED***
    throw new TypeError(invalidKeyInput(key, ...types, 'Uint8Array'));
}
async function deriveKey(p2s, alg, p2c, key) ***REMOVED***
    checkP2s(p2s);
    const salt = concatSalt(alg, p2s);
    const keylen = parseInt(alg.slice(13, 16), 10);
    const subtleAlg = ***REMOVED***
        hash: `SHA-$***REMOVED***alg.slice(8, 11)}`,
        iterations: p2c,
        name: 'PBKDF2',
        salt,
  ***REMOVED***;
    const wrapAlg = ***REMOVED***
        length: keylen,
        name: 'AES-KW',
  ***REMOVED***;
    const cryptoKey = await getCryptoKey(key, alg);
    if (cryptoKey.usages.includes('deriveBits')) ***REMOVED***
        return new Uint8Array(await crypto.subtle.deriveBits(subtleAlg, cryptoKey, keylen));
  ***REMOVED***
    if (cryptoKey.usages.includes('deriveKey')) ***REMOVED***
        return crypto.subtle.deriveKey(subtleAlg, cryptoKey, wrapAlg, false, ['wrapKey', 'unwrapKey']);
  ***REMOVED***
    throw new TypeError('PBKDF2 key "usages" must include "deriveBits" or "deriveKey"');
}
export const encrypt = async (alg, key, cek, p2c = 2048, p2s = random(new Uint8Array(16))) => ***REMOVED***
    const derived = await deriveKey(p2s, alg, p2c, key);
    const encryptedKey = await wrap(alg.slice(-6), derived, cek);
    return ***REMOVED*** encryptedKey, p2c, p2s: base64url(p2s) };
};
export const decrypt = async (alg, key, encryptedKey, p2c, p2s) => ***REMOVED***
    const derived = await deriveKey(p2s, alg, p2c, key);
    return unwrap(alg.slice(-6), derived, encryptedKey);
};
