// @babel/types/lib/definitions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
});
Object.defineProperty(exports, "ALIAS_KEYS", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _utils.ALIAS_KEYS;
***REMOVED***
});
Object.defineProperty(exports, "BUILDER_KEYS", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _utils.BUILDER_KEYS;
***REMOVED***
});
Object.defineProperty(exports, "DEPRECATED_KEYS", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _utils.DEPRECATED_KEYS;
***REMOVED***
});
Object.defineProperty(exports, "FLIPPED_ALIAS_KEYS", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _utils.FLIPPED_ALIAS_KEYS;
***REMOVED***
});
Object.defineProperty(exports, "NODE_FIELDS", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _utils.NODE_FIELDS;
***REMOVED***
});
Object.defineProperty(exports, "NODE_PARENT_VALIDATIONS", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _utils.NODE_PARENT_VALIDATIONS;
***REMOVED***
});
Object.defineProperty(exports, "PLACEHOLDERS", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _placeholders.PLACEHOLDERS;
***REMOVED***
});
Object.defineProperty(exports, "PLACEHOLDERS_ALIAS", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _placeholders.PLACEHOLDERS_ALIAS;
***REMOVED***
});
Object.defineProperty(exports, "PLACEHOLDERS_FLIPPED_ALIAS", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _placeholders.PLACEHOLDERS_FLIPPED_ALIAS;
***REMOVED***
});
exports.TYPES = void 0;
Object.defineProperty(exports, "VISITOR_KEYS", ***REMOVED***
  enumerable: true,
  get: function () ***REMOVED***
    return _utils.VISITOR_KEYS;
***REMOVED***
});
var _toFastProperties = require("to-fast-properties");
require("./core");
require("./flow");
require("./jsx");
require("./misc");
require("./experimental");
require("./typescript");
var _utils = require("./utils");
var _placeholders = require("./placeholders");
_toFastProperties(_utils.VISITOR_KEYS);
_toFastProperties(_utils.ALIAS_KEYS);
_toFastProperties(_utils.FLIPPED_ALIAS_KEYS);
_toFastProperties(_utils.NODE_FIELDS);
_toFastProperties(_utils.BUILDER_KEYS);
_toFastProperties(_utils.DEPRECATED_KEYS);
_toFastProperties(_placeholders.PLACEHOLDERS_ALIAS);
_toFastProperties(_placeholders.PLACEHOLDERS_FLIPPED_ALIAS);
const TYPES = [].concat(Object.keys(_utils.VISITOR_KEYS), Object.keys(_utils.FLIPPED_ALIAS_KEYS), Object.keys(_utils.DEPRECATED_KEYS));
exports.TYPES = TYPES;
