// jsdoc/lib/jsdoc/tag/inline.js
/**
 * @module jsdoc/tag/inline
 */
/**
 * Information about an inline tag that was found within a string.
 *
 * @typedef ***REMOVED***Object} InlineTagInfo
 * @memberof module:jsdoc/tag/inline
 * @property ***REMOVED***?string} completeTag - The entire inline tag, including its enclosing braces.
 * @property ***REMOVED***?string} tag - The tag whose text was found.
 * @property ***REMOVED***?string} text - The tag text that was found.
 */

/**
 * Information about the results of replacing inline tags within a string.
 *
 * @typedef ***REMOVED***Object} InlineTagResult
 * @memberof module:jsdoc/tag/inline
 * @property ***REMOVED***Array.<module:jsdoc/tag/inline.InlineTagInfo>} tags - The inline tags that were found.
 * @property ***REMOVED***string} newString - The updated text string after extracting or replacing the inline
 * tags.
 */

/**
 * Text-replacing function for strings that contain an inline tag.
 *
 * @callback InlineTagReplacer
 * @memberof module:jsdoc/tag/inline
 * @param ***REMOVED***string} string - The complete string containing the inline tag.
 * @param ***REMOVED***module:jsdoc/tag/inline.InlineTagInfo} tagInfo - Information about the inline tag.
 * @return ***REMOVED***string} An updated version of the complete string.
 */

/**
 * Create a regexp that matches a specific inline tag, or all inline tags.
 *
 * @private
 * @memberof module:jsdoc/tag/inline
 * @param ***REMOVED***?string} tagName - The inline tag that the regexp will match. May contain regexp
 * characters. If omitted, matches any string.
 * @param ***REMOVED***?string} prefix - A prefix for the regexp. Defaults to an empty string.
 * @param ***REMOVED***?string} suffix - A suffix for the regexp. Defaults to an empty string.
 * @returns ***REMOVED***RegExp} A regular expression that matches the requested inline tag.
 */
function regExpFactory(tagName = '\\S+', prefix = '', suffix = '') ***REMOVED***
    return new RegExp(`$***REMOVED***prefix}\\***REMOVED***@$***REMOVED***tagName}\\s+((?:.|\n)+?)\\}$***REMOVED***suffix}`, 'i');
}

/**
 * Check whether a string is an inline tag. You can check for a specific inline tag or for any valid
 * inline tag.
 *
 * @param ***REMOVED***string} string - The string to check.
 * @param ***REMOVED***?string} tagName - The inline tag to match. May contain regexp characters. If this
 * parameter is omitted, this method returns `true` for any valid inline tag.
 * @returns ***REMOVED***boolean} Set to `true` if the string is a valid inline tag or `false` in all other
 * cases.
 */
exports.isInlineTag = (string, tagName) => regExpFactory(tagName, '^', '$').test(string);

/**
 * Replace all instances of multiple inline tags with other text.
 *
 * @param ***REMOVED***string} string - The string in which to replace the inline tags.
 * @param ***REMOVED***Object} replacers - The functions that are used to replace text in the string. The keys
 * must contain tag names (for example, `link`), and the values must contain functions with the
 * type ***REMOVED***@link module:jsdoc/tag/inline.InlineTagReplacer}.
 * @return ***REMOVED***module:jsdoc/tag/inline.InlineTagResult} The updated string, as well as information
 * about the inline tags that were found.
 */
exports.replaceInlineTags = (string, replacers) => ***REMOVED***
    const tagInfo = [];

    function replaceMatch(replacer, tag, match, text) ***REMOVED***
        const matchedTag = ***REMOVED***
            completeTag: match,
            tag: tag,
            text: text
      ***REMOVED***;

        tagInfo.push(matchedTag);

        return replacer(string, matchedTag);
  ***REMOVED***

    string = string || '';

    Object.keys(replacers).forEach(replacer => ***REMOVED***
        const tagRegExp = regExpFactory(replacer);
        let matches;
        let previousString;

        // call the replacer once for each match
        do ***REMOVED***
            matches = tagRegExp.exec(string);
            if (matches) ***REMOVED***
                previousString = string;
                string = replaceMatch(replacers[replacer], replacer, matches[0], matches[1]);
          ***REMOVED***
      ***REMOVED*** while (matches && previousString !== string);
  ***REMOVED***);

    return ***REMOVED***
        tags: tagInfo,
        newString: string.trim()
  ***REMOVED***;
};

/**
 * Replace all instances of an inline tag with other text.
 *
 * @param ***REMOVED***string} string - The string in which to replace the inline tag.
 * @param ***REMOVED***string} tag - The name of the inline tag to replace.
 * @param ***REMOVED***module:jsdoc/tag/inline.InlineTagReplacer} replacer - The function that is used to
 * replace text in the string.
 * @return ***REMOVED***module:jsdoc/tag/inline.InlineTagResult} The updated string, as well as information
 * about the inline tags that were found.
 */
exports.replaceInlineTag = (string, tag, replacer) => ***REMOVED***
    const replacers = ***REMOVED***};

    replacers[tag] = replacer;

    return exports.replaceInlineTags(string, replacers);
};

/**
 * Extract inline tags from a string, replacing them with an empty string.
 *
 * @param ***REMOVED***string} string - The string from which to extract text.
 * @param ***REMOVED***?string} tag - The inline tag to extract.
 * @return ***REMOVED***module:jsdoc/tag/inline.InlineTagResult} The updated string, as well as information
 * about the inline tags that were found.
 */
exports.extractInlineTag = (string, tag) => exports.replaceInlineTag(string, tag, (str, ***REMOVED***completeTag}) => str.replace(completeTag, ''));
