// yargs/build/lib/command.js
import ***REMOVED*** assertNotStrictEqual, } from './typings/common-types.js';
import ***REMOVED*** isPromise } from './utils/is-promise.js';
import ***REMOVED*** applyMiddleware, commandMiddlewareFactory, } from './middleware.js';
import ***REMOVED*** parseCommand } from './parse-command.js';
import ***REMOVED*** isYargsInstance, } from './yargs-factory.js';
import whichModule from './utils/which-module.js';
const DEFAULT_MARKER = /(^\*)|(^\$0)/;
export function command(yargs, usage, validation, globalMiddleware = [], shim) ***REMOVED***
    const self = ***REMOVED***};
    let handlers = ***REMOVED***};
    let aliasMap = ***REMOVED***};
    let defaultCommand;
    self.addHandler = function addHandler(cmd, description, builder, handler, commandMiddleware, deprecated) ***REMOVED***
        let aliases = [];
        const middlewares = commandMiddlewareFactory(commandMiddleware);
        handler = handler || (() => ***REMOVED*** });
        if (Array.isArray(cmd)) ***REMOVED***
            if (isCommandAndAliases(cmd)) ***REMOVED***
                [cmd, ...aliases] = cmd;
          ***REMOVED***
            else ***REMOVED***
                for (const command of cmd) ***REMOVED***
                    self.addHandler(command);
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        else if (isCommandHandlerDefinition(cmd)) ***REMOVED***
            let command = Array.isArray(cmd.command) || typeof cmd.command === 'string'
                ? cmd.command
                : moduleName(cmd);
            if (cmd.aliases)
                command = [].concat(command).concat(cmd.aliases);
            self.addHandler(command, extractDesc(cmd), cmd.builder, cmd.handler, cmd.middlewares, cmd.deprecated);
            return;
      ***REMOVED***
        else if (isCommandBuilderDefinition(builder)) ***REMOVED***
            self.addHandler([cmd].concat(aliases), description, builder.builder, builder.handler, builder.middlewares, builder.deprecated);
            return;
      ***REMOVED***
        if (typeof cmd === 'string') ***REMOVED***
            const parsedCommand = parseCommand(cmd);
            aliases = aliases.map(alias => parseCommand(alias).cmd);
            let isDefault = false;
            const parsedAliases = [parsedCommand.cmd].concat(aliases).filter(c => ***REMOVED***
                if (DEFAULT_MARKER.test(c)) ***REMOVED***
                    isDefault = true;
                    return false;
              ***REMOVED***
                return true;
          ***REMOVED***);
            if (parsedAliases.length === 0 && isDefault)
                parsedAliases.push('$0');
            if (isDefault) ***REMOVED***
                parsedCommand.cmd = parsedAliases[0];
                aliases = parsedAliases.slice(1);
                cmd = cmd.replace(DEFAULT_MARKER, parsedCommand.cmd);
          ***REMOVED***
            aliases.forEach(alias => ***REMOVED***
                aliasMap[alias] = parsedCommand.cmd;
          ***REMOVED***);
            if (description !== false) ***REMOVED***
                usage.command(cmd, description, isDefault, aliases, deprecated);
          ***REMOVED***
            handlers[parsedCommand.cmd] = ***REMOVED***
                original: cmd,
                description,
                handler,
                builder: builder || ***REMOVED***},
                middlewares,
                deprecated,
                demanded: parsedCommand.demanded,
                optional: parsedCommand.optional,
          ***REMOVED***;
            if (isDefault)
                defaultCommand = handlers[parsedCommand.cmd];
      ***REMOVED***
  ***REMOVED***;
    self.addDirectory = function addDirectory(dir, context, req, callerFile, opts) ***REMOVED***
        opts = opts || ***REMOVED***};
        if (typeof opts.recurse !== 'boolean')
            opts.recurse = false;
        if (!Array.isArray(opts.extensions))
            opts.extensions = ['js'];
        const parentVisit = typeof opts.visit === 'function' ? opts.visit : (o) => o;
        opts.visit = function visit(obj, joined, filename) ***REMOVED***
            const visited = parentVisit(obj, joined, filename);
            if (visited) ***REMOVED***
                if (~context.files.indexOf(joined))
                    return visited;
                context.files.push(joined);
                self.addHandler(visited);
          ***REMOVED***
            return visited;
      ***REMOVED***;
        shim.requireDirectory(***REMOVED*** require: req, filename: callerFile }, dir, opts);
  ***REMOVED***;
    function moduleName(obj) ***REMOVED***
        const mod = whichModule(obj);
        if (!mod)
            throw new Error(`No command name given for module: $***REMOVED***shim.inspect(obj)}`);
        return commandFromFilename(mod.filename);
  ***REMOVED***
    function commandFromFilename(filename) ***REMOVED***
        return shim.path.basename(filename, shim.path.extname(filename));
  ***REMOVED***
    function extractDesc(***REMOVED*** describe, description, desc, }) ***REMOVED***
        for (const test of [describe, description, desc]) ***REMOVED***
            if (typeof test === 'string' || test === false)
                return test;
            assertNotStrictEqual(test, true, shim);
      ***REMOVED***
        return false;
  ***REMOVED***
    self.getCommands = () => Object.keys(handlers).concat(Object.keys(aliasMap));
    self.getCommandHandlers = () => handlers;
    self.hasDefaultCommand = () => !!defaultCommand;
    self.runCommand = function runCommand(command, yargs, parsed, commandIndex) ***REMOVED***
        let aliases = parsed.aliases;
        const commandHandler = handlers[command] || handlers[aliasMap[command]] || defaultCommand;
        const currentContext = yargs.getContext();
        let numFiles = currentContext.files.length;
        const parentCommands = currentContext.commands.slice();
        let innerArgv = parsed.argv;
        let positionalMap = ***REMOVED***};
        if (command) ***REMOVED***
            currentContext.commands.push(command);
            currentContext.fullCommands.push(commandHandler.original);
      ***REMOVED***
        const builder = commandHandler.builder;
        if (isCommandBuilderCallback(builder)) ***REMOVED***
            const builderOutput = builder(yargs.reset(parsed.aliases));
            const innerYargs = isYargsInstance(builderOutput) ? builderOutput : yargs;
            if (shouldUpdateUsage(innerYargs)) ***REMOVED***
                innerYargs
                    .getUsageInstance()
                    .usage(usageFromParentCommandsCommandHandler(parentCommands, commandHandler), commandHandler.description);
          ***REMOVED***
            innerArgv = innerYargs._parseArgs(null, null, true, commandIndex);
            aliases = innerYargs.parsed.aliases;
      ***REMOVED***
        else if (isCommandBuilderOptionDefinitions(builder)) ***REMOVED***
            const innerYargs = yargs.reset(parsed.aliases);
            if (shouldUpdateUsage(innerYargs)) ***REMOVED***
                innerYargs
                    .getUsageInstance()
                    .usage(usageFromParentCommandsCommandHandler(parentCommands, commandHandler), commandHandler.description);
          ***REMOVED***
            Object.keys(commandHandler.builder).forEach(key => ***REMOVED***
                innerYargs.option(key, builder[key]);
          ***REMOVED***);
            innerArgv = innerYargs._parseArgs(null, null, true, commandIndex);
            aliases = innerYargs.parsed.aliases;
      ***REMOVED***
        if (!yargs._hasOutput()) ***REMOVED***
            positionalMap = populatePositionals(commandHandler, innerArgv, currentContext);
      ***REMOVED***
        const middlewares = globalMiddleware
            .slice(0)
            .concat(commandHandler.middlewares);
        applyMiddleware(innerArgv, yargs, middlewares, true);
        if (!yargs._hasOutput()) ***REMOVED***
            yargs._runValidation(innerArgv, aliases, positionalMap, yargs.parsed.error, !command);
      ***REMOVED***
        if (commandHandler.handler && !yargs._hasOutput()) ***REMOVED***
            yargs._setHasOutput();
            const populateDoubleDash = !!yargs.getOptions().configuration['populate--'];
            yargs._postProcess(innerArgv, populateDoubleDash);
            innerArgv = applyMiddleware(innerArgv, yargs, middlewares, false);
            let handlerResult;
            if (isPromise(innerArgv)) ***REMOVED***
                handlerResult = innerArgv.then(argv => commandHandler.handler(argv));
          ***REMOVED***
            else ***REMOVED***
                handlerResult = commandHandler.handler(innerArgv);
          ***REMOVED***
            const handlerFinishCommand = yargs.getHandlerFinishCommand();
            if (isPromise(handlerResult)) ***REMOVED***
                yargs.getUsageInstance().cacheHelpMessage();
                handlerResult
                    .then(value => ***REMOVED***
                    if (handlerFinishCommand) ***REMOVED***
                        handlerFinishCommand(value);
                  ***REMOVED***
              ***REMOVED***)
                    .catch(error => ***REMOVED***
                    try ***REMOVED***
                        yargs.getUsageInstance().fail(null, error);
                  ***REMOVED***
                    catch (err) ***REMOVED***
                  ***REMOVED***
              ***REMOVED***)
                    .then(() => ***REMOVED***
                    yargs.getUsageInstance().clearCachedHelpMessage();
              ***REMOVED***);
          ***REMOVED***
            else ***REMOVED***
                if (handlerFinishCommand) ***REMOVED***
                    handlerFinishCommand(handlerResult);
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        if (command) ***REMOVED***
            currentContext.commands.pop();
            currentContext.fullCommands.pop();
      ***REMOVED***
        numFiles = currentContext.files.length - numFiles;
        if (numFiles > 0)
            currentContext.files.splice(numFiles * -1, numFiles);
        return innerArgv;
  ***REMOVED***;
    function shouldUpdateUsage(yargs) ***REMOVED***
        return (!yargs.getUsageInstance().getUsageDisabled() &&
            yargs.getUsageInstance().getUsage().length === 0);
  ***REMOVED***
    function usageFromParentCommandsCommandHandler(parentCommands, commandHandler) ***REMOVED***
        const c = DEFAULT_MARKER.test(commandHandler.original)
            ? commandHandler.original.replace(DEFAULT_MARKER, '').trim()
            : commandHandler.original;
        const pc = parentCommands.filter(c => ***REMOVED***
            return !DEFAULT_MARKER.test(c);
      ***REMOVED***);
        pc.push(c);
        return `$0 $***REMOVED***pc.join(' ')}`;
  ***REMOVED***
    self.runDefaultBuilderOn = function (yargs) ***REMOVED***
        assertNotStrictEqual(defaultCommand, undefined, shim);
        if (shouldUpdateUsage(yargs)) ***REMOVED***
            const commandString = DEFAULT_MARKER.test(defaultCommand.original)
                ? defaultCommand.original
                : defaultCommand.original.replace(/^[^[\]<>]*/, '$0 ');
            yargs.getUsageInstance().usage(commandString, defaultCommand.description);
      ***REMOVED***
        const builder = defaultCommand.builder;
        if (isCommandBuilderCallback(builder)) ***REMOVED***
            builder(yargs);
      ***REMOVED***
        else if (!isCommandBuilderDefinition(builder)) ***REMOVED***
            Object.keys(builder).forEach(key => ***REMOVED***
                yargs.option(key, builder[key]);
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***;
    function populatePositionals(commandHandler, argv, context) ***REMOVED***
        argv._ = argv._.slice(context.commands.length);
        const demanded = commandHandler.demanded.slice(0);
        const optional = commandHandler.optional.slice(0);
        const positionalMap = ***REMOVED***};
        validation.positionalCount(demanded.length, argv._.length);
        while (demanded.length) ***REMOVED***
            const demand = demanded.shift();
            populatePositional(demand, argv, positionalMap);
      ***REMOVED***
        while (optional.length) ***REMOVED***
            const maybe = optional.shift();
            populatePositional(maybe, argv, positionalMap);
      ***REMOVED***
        argv._ = context.commands.concat(argv._.map(a => '' + a));
        postProcessPositionals(argv, positionalMap, self.cmdToParseOptions(commandHandler.original));
        return positionalMap;
  ***REMOVED***
    function populatePositional(positional, argv, positionalMap) ***REMOVED***
        const cmd = positional.cmd[0];
        if (positional.variadic) ***REMOVED***
            positionalMap[cmd] = argv._.splice(0).map(String);
      ***REMOVED***
        else ***REMOVED***
            if (argv._.length)
                positionalMap[cmd] = [String(argv._.shift())];
      ***REMOVED***
  ***REMOVED***
    function postProcessPositionals(argv, positionalMap, parseOptions) ***REMOVED***
        const options = Object.assign(***REMOVED***}, yargs.getOptions());
        options.default = Object.assign(parseOptions.default, options.default);
        for (const key of Object.keys(parseOptions.alias)) ***REMOVED***
            options.alias[key] = (options.alias[key] || []).concat(parseOptions.alias[key]);
      ***REMOVED***
        options.array = options.array.concat(parseOptions.array);
        options.config = ***REMOVED***};
        const unparsed = [];
        Object.keys(positionalMap).forEach(key => ***REMOVED***
            positionalMap[key].map(value => ***REMOVED***
                if (options.configuration['unknown-options-as-args'])
                    options.key[key] = true;
                unparsed.push(`--$***REMOVED***key}`);
                unparsed.push(value);
          ***REMOVED***);
      ***REMOVED***);
        if (!unparsed.length)
            return;
        const config = Object.assign(***REMOVED***}, options.configuration, ***REMOVED***
            'populate--': true,
      ***REMOVED***);
        const parsed = shim.Parser.detailed(unparsed, Object.assign(***REMOVED***}, options, ***REMOVED***
            configuration: config,
      ***REMOVED***));
        if (parsed.error) ***REMOVED***
            yargs.getUsageInstance().fail(parsed.error.message, parsed.error);
      ***REMOVED***
        else ***REMOVED***
            const positionalKeys = Object.keys(positionalMap);
            Object.keys(positionalMap).forEach(key => ***REMOVED***
                positionalKeys.push(...parsed.aliases[key]);
          ***REMOVED***);
            Object.keys(parsed.argv).forEach(key => ***REMOVED***
                if (positionalKeys.indexOf(key) !== -1) ***REMOVED***
                    if (!positionalMap[key])
                        positionalMap[key] = parsed.argv[key];
                    argv[key] = parsed.argv[key];
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***
    self.cmdToParseOptions = function (cmdString) ***REMOVED***
        const parseOptions = ***REMOVED***
            array: [],
            default: ***REMOVED***},
            alias: ***REMOVED***},
            demand: ***REMOVED***},
      ***REMOVED***;
        const parsed = parseCommand(cmdString);
        parsed.demanded.forEach(d => ***REMOVED***
            const [cmd, ...aliases] = d.cmd;
            if (d.variadic) ***REMOVED***
                parseOptions.array.push(cmd);
                parseOptions.default[cmd] = [];
          ***REMOVED***
            parseOptions.alias[cmd] = aliases;
            parseOptions.demand[cmd] = true;
      ***REMOVED***);
        parsed.optional.forEach(o => ***REMOVED***
            const [cmd, ...aliases] = o.cmd;
            if (o.variadic) ***REMOVED***
                parseOptions.array.push(cmd);
                parseOptions.default[cmd] = [];
          ***REMOVED***
            parseOptions.alias[cmd] = aliases;
      ***REMOVED***);
        return parseOptions;
  ***REMOVED***;
    self.reset = () => ***REMOVED***
        handlers = ***REMOVED***};
        aliasMap = ***REMOVED***};
        defaultCommand = undefined;
        return self;
  ***REMOVED***;
    const frozens = [];
    self.freeze = () => ***REMOVED***
        frozens.push(***REMOVED***
            handlers,
            aliasMap,
            defaultCommand,
      ***REMOVED***);
  ***REMOVED***;
    self.unfreeze = () => ***REMOVED***
        const frozen = frozens.pop();
        assertNotStrictEqual(frozen, undefined, shim);
        (***REMOVED*** handlers, aliasMap, defaultCommand } = frozen);
  ***REMOVED***;
    return self;
}
export function isCommandBuilderDefinition(builder) ***REMOVED***
    return (typeof builder === 'object' &&
        !!builder.builder &&
        typeof builder.handler === 'function');
}
function isCommandAndAliases(cmd) ***REMOVED***
    if (cmd.every(c => typeof c === 'string')) ***REMOVED***
        return true;
  ***REMOVED***
    else ***REMOVED***
        return false;
  ***REMOVED***
}
export function isCommandBuilderCallback(builder) ***REMOVED***
    return typeof builder === 'function';
}
function isCommandBuilderOptionDefinitions(builder) ***REMOVED***
    return typeof builder === 'object';
}
export function isCommandHandlerDefinition(cmd) ***REMOVED***
    return typeof cmd === 'object' && !Array.isArray(cmd);
}
