// google-auth-library/build/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.GoogleAuth = exports.auth = exports.DefaultTransporter = exports.PluggableAuthClient = exports.DownscopedClient = exports.BaseExternalAccountClient = exports.ExternalAccountClient = exports.IdentityPoolClient = exports.AwsClient = exports.UserRefreshClient = exports.LoginTicket = exports.OAuth2Client = exports.CodeChallengeMethod = exports.Impersonated = exports.JWT = exports.JWTAccess = exports.IdTokenClient = exports.IAMAuth = exports.GCPEnv = exports.Compute = exports.AuthClient = void 0;
// Copyright 2017 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
const googleauth_1 = require("./auth/googleauth");
Object.defineProperty(exports, "GoogleAuth", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return googleauth_1.GoogleAuth; } });
var authclient_1 = require("./auth/authclient");
Object.defineProperty(exports, "AuthClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return authclient_1.AuthClient; } });
var computeclient_1 = require("./auth/computeclient");
Object.defineProperty(exports, "Compute", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return computeclient_1.Compute; } });
var envDetect_1 = require("./auth/envDetect");
Object.defineProperty(exports, "GCPEnv", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return envDetect_1.GCPEnv; } });
var iam_1 = require("./auth/iam");
Object.defineProperty(exports, "IAMAuth", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return iam_1.IAMAuth; } });
var idtokenclient_1 = require("./auth/idtokenclient");
Object.defineProperty(exports, "IdTokenClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return idtokenclient_1.IdTokenClient; } });
var jwtaccess_1 = require("./auth/jwtaccess");
Object.defineProperty(exports, "JWTAccess", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return jwtaccess_1.JWTAccess; } });
var jwtclient_1 = require("./auth/jwtclient");
Object.defineProperty(exports, "JWT", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return jwtclient_1.JWT; } });
var impersonated_1 = require("./auth/impersonated");
Object.defineProperty(exports, "Impersonated", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return impersonated_1.Impersonated; } });
var oauth2client_1 = require("./auth/oauth2client");
Object.defineProperty(exports, "CodeChallengeMethod", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return oauth2client_1.CodeChallengeMethod; } });
Object.defineProperty(exports, "OAuth2Client", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return oauth2client_1.OAuth2Client; } });
var loginticket_1 = require("./auth/loginticket");
Object.defineProperty(exports, "LoginTicket", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return loginticket_1.LoginTicket; } });
var refreshclient_1 = require("./auth/refreshclient");
Object.defineProperty(exports, "UserRefreshClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return refreshclient_1.UserRefreshClient; } });
var awsclient_1 = require("./auth/awsclient");
Object.defineProperty(exports, "AwsClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return awsclient_1.AwsClient; } });
var identitypoolclient_1 = require("./auth/identitypoolclient");
Object.defineProperty(exports, "IdentityPoolClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return identitypoolclient_1.IdentityPoolClient; } });
var externalclient_1 = require("./auth/externalclient");
Object.defineProperty(exports, "ExternalAccountClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return externalclient_1.ExternalAccountClient; } });
var baseexternalclient_1 = require("./auth/baseexternalclient");
Object.defineProperty(exports, "BaseExternalAccountClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return baseexternalclient_1.BaseExternalAccountClient; } });
var downscopedclient_1 = require("./auth/downscopedclient");
Object.defineProperty(exports, "DownscopedClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return downscopedclient_1.DownscopedClient; } });
var pluggable_auth_client_1 = require("./auth/pluggable-auth-client");
Object.defineProperty(exports, "PluggableAuthClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return pluggable_auth_client_1.PluggableAuthClient; } });
var transporters_1 = require("./transporters");
Object.defineProperty(exports, "DefaultTransporter", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return transporters_1.DefaultTransporter; } });
const auth = new googleauth_1.GoogleAuth();
exports.auth = auth;
