// @firebase/remote-config/dist/esm/index.esm.js
import ***REMOVED*** getApp, _getProvider, _registerComponent, registerVersion, SDK_VERSION } from '@firebase/app';
import ***REMOVED*** ErrorFactory, FirebaseError, getModularInstance, calculateBackoffMillis, isIndexedDBAvailable, validateIndexedDBOpenable } from '@firebase/util';
import ***REMOVED*** Component } from '@firebase/component';
import ***REMOVED*** LogLevel, Logger } from '@firebase/logger';
import ***REMOVED*** __awaiter, __generator, __assign } from 'tslib';
import '@firebase/installations';

var name = "@firebase/remote-config";
var version = "0.4.1";

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Shims a minimal AbortSignal.
 *
 * <p>AbortController's AbortSignal conveniently decouples fetch timeout logic from other aspects
 * of networking, such as retries. Firebase doesn't use AbortController enough to justify a
 * polyfill recommendation, like we do with the Fetch API, but this minimal shim can easily be
 * swapped out if/when we do.
 */
var RemoteConfigAbortSignal = /** @class */ (function () ***REMOVED***
    function RemoteConfigAbortSignal() ***REMOVED***
        this.listeners = [];
  ***REMOVED***
    RemoteConfigAbortSignal.prototype.addEventListener = function (listener) ***REMOVED***
        this.listeners.push(listener);
  ***REMOVED***;
    RemoteConfigAbortSignal.prototype.abort = function () ***REMOVED***
        this.listeners.forEach(function (listener) ***REMOVED*** return listener(); });
  ***REMOVED***;
    return RemoteConfigAbortSignal;
}());

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var RC_COMPONENT_NAME = 'remote-config';

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var _a;
var ERROR_DESCRIPTION_MAP = (_a = ***REMOVED***},
    _a["registration-window" /* ErrorCode.REGISTRATION_WINDOW */] = 'Undefined window object. This SDK only supports usage in a browser environment.',
    _a["registration-project-id" /* ErrorCode.REGISTRATION_PROJECT_ID */] = 'Undefined project identifier. Check Firebase app initialization.',
    _a["registration-api-key" /* ErrorCode.REGISTRATION_API_KEY */] = 'Undefined API key. Check Firebase app initialization.',
    _a["registration-app-id" /* ErrorCode.REGISTRATION_APP_ID */] = 'Undefined app identifier. Check Firebase app initialization.',
    _a["storage-open" /* ErrorCode.STORAGE_OPEN */] = 'Error thrown when opening storage. Original error: ***REMOVED***$originalErrorMessage}.',
    _a["storage-get" /* ErrorCode.STORAGE_GET */] = 'Error thrown when reading from storage. Original error: ***REMOVED***$originalErrorMessage}.',
    _a["storage-set" /* ErrorCode.STORAGE_SET */] = 'Error thrown when writing to storage. Original error: ***REMOVED***$originalErrorMessage}.',
    _a["storage-delete" /* ErrorCode.STORAGE_DELETE */] = 'Error thrown when deleting from storage. Original error: ***REMOVED***$originalErrorMessage}.',
    _a["fetch-client-network" /* ErrorCode.FETCH_NETWORK */] = 'Fetch client failed to connect to a network. Check Internet connection.' +
        ' Original error: ***REMOVED***$originalErrorMessage}.',
    _a["fetch-timeout" /* ErrorCode.FETCH_TIMEOUT */] = 'The config fetch request timed out. ' +
        ' Configure timeout using "fetchTimeoutMillis" SDK setting.',
    _a["fetch-throttle" /* ErrorCode.FETCH_THROTTLE */] = 'The config fetch request timed out while in an exponential backoff state.' +
        ' Configure timeout using "fetchTimeoutMillis" SDK setting.' +
        ' Unix timestamp in milliseconds when fetch request throttling ends: ***REMOVED***$throttleEndTimeMillis}.',
    _a["fetch-client-parse" /* ErrorCode.FETCH_PARSE */] = 'Fetch client could not parse response.' +
        ' Original error: ***REMOVED***$originalErrorMessage}.',
    _a["fetch-status" /* ErrorCode.FETCH_STATUS */] = 'Fetch server returned an HTTP error status. HTTP status: ***REMOVED***$httpStatus}.',
    _a["indexed-db-unavailable" /* ErrorCode.INDEXED_DB_UNAVAILABLE */] = 'Indexed DB is not supported by current browser',
    _a);
var ERROR_FACTORY = new ErrorFactory('remoteconfig' /* service */, 'Remote Config' /* service name */, ERROR_DESCRIPTION_MAP);
// Note how this is like typeof/instanceof, but for ErrorCode.
function hasErrorCode(e, errorCode) ***REMOVED***
    return e instanceof FirebaseError && e.code.indexOf(errorCode) !== -1;
}

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var DEFAULT_VALUE_FOR_BOOLEAN = false;
var DEFAULT_VALUE_FOR_STRING = '';
var DEFAULT_VALUE_FOR_NUMBER = 0;
var BOOLEAN_TRUTHY_VALUES = ['1', 'true', 't', 'yes', 'y', 'on'];
var Value = /** @class */ (function () ***REMOVED***
    function Value(_source, _value) ***REMOVED***
        if (_value === void 0) ***REMOVED*** _value = DEFAULT_VALUE_FOR_STRING; }
        this._source = _source;
        this._value = _value;
  ***REMOVED***
    Value.prototype.asString = function () ***REMOVED***
        return this._value;
  ***REMOVED***;
    Value.prototype.asBoolean = function () ***REMOVED***
        if (this._source === 'static') ***REMOVED***
            return DEFAULT_VALUE_FOR_BOOLEAN;
      ***REMOVED***
        return BOOLEAN_TRUTHY_VALUES.indexOf(this._value.toLowerCase()) >= 0;
  ***REMOVED***;
    Value.prototype.asNumber = function () ***REMOVED***
        if (this._source === 'static') ***REMOVED***
            return DEFAULT_VALUE_FOR_NUMBER;
      ***REMOVED***
        var num = Number(this._value);
        if (isNaN(num)) ***REMOVED***
            num = DEFAULT_VALUE_FOR_NUMBER;
      ***REMOVED***
        return num;
  ***REMOVED***;
    Value.prototype.getSource = function () ***REMOVED***
        return this._source;
  ***REMOVED***;
    return Value;
}());

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 *
 * @param app - The ***REMOVED***@link @firebase/app#FirebaseApp} instance.
 * @returns A ***REMOVED***@link RemoteConfig} instance.
 *
 * @public
 */
function getRemoteConfig(app) ***REMOVED***
    if (app === void 0) ***REMOVED*** app = getApp(); }
    app = getModularInstance(app);
    var rcProvider = _getProvider(app, RC_COMPONENT_NAME);
    return rcProvider.getImmediate();
}
/**
 * Makes the last fetched config available to the getters.
 * @param remoteConfig - The ***REMOVED***@link RemoteConfig} instance.
 * @returns A `Promise` which resolves to true if the current call activated the fetched configs.
 * If the fetched configs were already activated, the `Promise` will resolve to false.
 *
 * @public
 */
function activate(remoteConfig) ***REMOVED***
    return __awaiter(this, void 0, void 0, function () ***REMOVED***
        var rc, _a, lastSuccessfulFetchResponse, activeConfigEtag;
        return __generator(this, function (_b) ***REMOVED***
            switch (_b.label) ***REMOVED***
                case 0:
                    rc = getModularInstance(remoteConfig);
                    return [4 /*yield*/, Promise.all([
                            rc._storage.getLastSuccessfulFetchResponse(),
                            rc._storage.getActiveConfigEtag()
                        ])];
                case 1:
                    _a = _b.sent(), lastSuccessfulFetchResponse = _a[0], activeConfigEtag = _a[1];
                    if (!lastSuccessfulFetchResponse ||
                        !lastSuccessfulFetchResponse.config ||
                        !lastSuccessfulFetchResponse.eTag ||
                        lastSuccessfulFetchResponse.eTag === activeConfigEtag) ***REMOVED***
                        // Either there is no successful fetched config, or is the same as current active
                        // config.
                        return [2 /*return*/, false];
                  ***REMOVED***
                    return [4 /*yield*/, Promise.all([
                            rc._storageCache.setActiveConfig(lastSuccessfulFetchResponse.config),
                            rc._storage.setActiveConfigEtag(lastSuccessfulFetchResponse.eTag)
                        ])];
                case 2:
                    _b.sent();
                    return [2 /*return*/, true];
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***);
}
/**
 * Ensures the last activated config are available to the getters.
 * @param remoteConfig - The ***REMOVED***@link RemoteConfig} instance.
 *
 * @returns A `Promise` that resolves when the last activated config is available to the getters.
 * @public
 */
function ensureInitialized(remoteConfig) ***REMOVED***
    var rc = getModularInstance(remoteConfig);
    if (!rc._initializePromise) ***REMOVED***
        rc._initializePromise = rc._storageCache.loadFromStorage().then(function () ***REMOVED***
            rc._isInitializationComplete = true;
      ***REMOVED***);
  ***REMOVED***
    return rc._initializePromise;
}
/**
 * Fetches and caches configuration from the Remote Config service.
 * @param remoteConfig - The ***REMOVED***@link RemoteConfig} instance.
 * @public
 */
function fetchConfig(remoteConfig) ***REMOVED***
    return __awaiter(this, void 0, void 0, function () ***REMOVED***
        var rc, abortSignal, e_1, lastFetchStatus;
        var _this = this;
        return __generator(this, function (_a) ***REMOVED***
            switch (_a.label) ***REMOVED***
                case 0:
                    rc = getModularInstance(remoteConfig);
                    abortSignal = new RemoteConfigAbortSignal();
                    setTimeout(function () ***REMOVED*** return __awaiter(_this, void 0, void 0, function () ***REMOVED***
                        return __generator(this, function (_a) ***REMOVED***
                            // Note a very low delay, eg < 10ms, can elapse before listeners are initialized.
                            abortSignal.abort();
                            return [2 /*return*/];
                      ***REMOVED***);
                  ***REMOVED***); }, rc.settings.fetchTimeoutMillis);
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 4, , 6]);
                    return [4 /*yield*/, rc._client.fetch(***REMOVED***
                            cacheMaxAgeMillis: rc.settings.minimumFetchIntervalMillis,
                            signal: abortSignal
                      ***REMOVED***)];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, rc._storageCache.setLastFetchStatus('success')];
                case 3:
                    _a.sent();
                    return [3 /*break*/, 6];
                case 4:
                    e_1 = _a.sent();
                    lastFetchStatus = hasErrorCode(e_1, "fetch-throttle" /* ErrorCode.FETCH_THROTTLE */)
                        ? 'throttle'
                        : 'failure';
                    return [4 /*yield*/, rc._storageCache.setLastFetchStatus(lastFetchStatus)];
                case 5:
                    _a.sent();
                    throw e_1;
                case 6: return [2 /*return*/];
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***);
}
/**
 * Gets all config.
 *
 * @param remoteConfig - The ***REMOVED***@link RemoteConfig} instance.
 * @returns All config.
 *
 * @public
 */
function getAll(remoteConfig) ***REMOVED***
    var rc = getModularInstance(remoteConfig);
    return getAllKeys(rc._storageCache.getActiveConfig(), rc.defaultConfig).reduce(function (allConfigs, key) ***REMOVED***
        allConfigs[key] = getValue(remoteConfig, key);
        return allConfigs;
  ***REMOVED***, ***REMOVED***});
}
/**
 * Gets the value for the given key as a boolean.
 *
 * Convenience method for calling <code>remoteConfig.getValue(key).asBoolean()</code>.
 *
 * @param remoteConfig - The ***REMOVED***@link RemoteConfig} instance.
 * @param key - The name of the parameter.
 *
 * @returns The value for the given key as a boolean.
 * @public
 */
function getBoolean(remoteConfig, key) ***REMOVED***
    return getValue(getModularInstance(remoteConfig), key).asBoolean();
}
/**
 * Gets the value for the given key as a number.
 *
 * Convenience method for calling <code>remoteConfig.getValue(key).asNumber()</code>.
 *
 * @param remoteConfig - The ***REMOVED***@link RemoteConfig} instance.
 * @param key - The name of the parameter.
 *
 * @returns The value for the given key as a number.
 *
 * @public
 */
function getNumber(remoteConfig, key) ***REMOVED***
    return getValue(getModularInstance(remoteConfig), key).asNumber();
}
/**
 * Gets the value for the given key as a string.
 * Convenience method for calling <code>remoteConfig.getValue(key).asString()</code>.
 *
 * @param remoteConfig - The ***REMOVED***@link RemoteConfig} instance.
 * @param key - The name of the parameter.
 *
 * @returns The value for the given key as a string.
 *
 * @public
 */
function getString(remoteConfig, key) ***REMOVED***
    return getValue(getModularInstance(remoteConfig), key).asString();
}
/**
 * Gets the ***REMOVED***@link Value} for the given key.
 *
 * @param remoteConfig - The ***REMOVED***@link RemoteConfig} instance.
 * @param key - The name of the parameter.
 *
 * @returns The value for the given key.
 *
 * @public
 */
function getValue(remoteConfig, key) ***REMOVED***
    var rc = getModularInstance(remoteConfig);
    if (!rc._isInitializationComplete) ***REMOVED***
        rc._logger.debug("A value was requested for key \"".concat(key, "\" before SDK initialization completed.") +
            ' Await on ensureInitialized if the intent was to get a previously activated value.');
  ***REMOVED***
    var activeConfig = rc._storageCache.getActiveConfig();
    if (activeConfig && activeConfig[key] !== undefined) ***REMOVED***
        return new Value('remote', activeConfig[key]);
  ***REMOVED***
    else if (rc.defaultConfig && rc.defaultConfig[key] !== undefined) ***REMOVED***
        return new Value('default', String(rc.defaultConfig[key]));
  ***REMOVED***
    rc._logger.debug("Returning static value for key \"".concat(key, "\".") +
        ' Define a default or remote value if this is unintentional.');
    return new Value('static');
}
/**
 * Defines the log level to use.
 *
 * @param remoteConfig - The ***REMOVED***@link RemoteConfig} instance.
 * @param logLevel - The log level to set.
 *
 * @public
 */
function setLogLevel(remoteConfig, logLevel) ***REMOVED***
    var rc = getModularInstance(remoteConfig);
    switch (logLevel) ***REMOVED***
        case 'debug':
            rc._logger.logLevel = LogLevel.DEBUG;
            break;
        case 'silent':
            rc._logger.logLevel = LogLevel.SILENT;
            break;
        default:
            rc._logger.logLevel = LogLevel.ERROR;
  ***REMOVED***
}
/**
 * Dedupes and returns an array of all the keys of the received objects.
 */
function getAllKeys(obj1, obj2) ***REMOVED***
    if (obj1 === void 0) ***REMOVED*** obj1 = ***REMOVED***}; }
    if (obj2 === void 0) ***REMOVED*** obj2 = ***REMOVED***}; }
    return Object.keys(__assign(__assign(***REMOVED***}, obj1), obj2));
}

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Implements the ***REMOVED***@link RemoteConfigClient} abstraction with success response caching.
 *
 * <p>Comparable to the browser's Cache API for responses, but the Cache API requires a Service
 * Worker, which requires HTTPS, which would significantly complicate SDK installation. Also, the
 * Cache API doesn't support matching entries by time.
 */
var CachingClient = /** @class */ (function () ***REMOVED***
    function CachingClient(client, storage, storageCache, logger) ***REMOVED***
        this.client = client;
        this.storage = storage;
        this.storageCache = storageCache;
        this.logger = logger;
  ***REMOVED***
    /**
     * Returns true if the age of the cached fetched configs is less than or equal to
     * ***REMOVED***@link Settings#minimumFetchIntervalInSeconds}.
     *
     * <p>This is comparable to passing `headers = ***REMOVED*** 'Cache-Control': max-age <maxAge> }` to the
     * native Fetch API.
     *
     * <p>Visible for testing.
     */
    CachingClient.prototype.isCachedDataFresh = function (cacheMaxAgeMillis, lastSuccessfulFetchTimestampMillis) ***REMOVED***
        // Cache can only be fresh if it's populated.
        if (!lastSuccessfulFetchTimestampMillis) ***REMOVED***
            this.logger.debug('Config fetch cache check. Cache unpopulated.');
            return false;
      ***REMOVED***
        // Calculates age of cache entry.
        var cacheAgeMillis = Date.now() - lastSuccessfulFetchTimestampMillis;
        var isCachedDataFresh = cacheAgeMillis <= cacheMaxAgeMillis;
        this.logger.debug('Config fetch cache check.' +
            " Cache age millis: ".concat(cacheAgeMillis, ".") +
            " Cache max age millis (minimumFetchIntervalMillis setting): ".concat(cacheMaxAgeMillis, ".") +
            " Is cache hit: ".concat(isCachedDataFresh, "."));
        return isCachedDataFresh;
  ***REMOVED***;
    CachingClient.prototype.fetch = function (request) ***REMOVED***
        return __awaiter(this, void 0, void 0, function () ***REMOVED***
            var _a, lastSuccessfulFetchTimestampMillis, lastSuccessfulFetchResponse, response, storageOperations;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0: return [4 /*yield*/, Promise.all([
                            this.storage.getLastSuccessfulFetchTimestampMillis(),
                            this.storage.getLastSuccessfulFetchResponse()
                        ])];
                    case 1:
                        _a = _b.sent(), lastSuccessfulFetchTimestampMillis = _a[0], lastSuccessfulFetchResponse = _a[1];
                        // Exits early on cache hit.
                        if (lastSuccessfulFetchResponse &&
                            this.isCachedDataFresh(request.cacheMaxAgeMillis, lastSuccessfulFetchTimestampMillis)) ***REMOVED***
                            return [2 /*return*/, lastSuccessfulFetchResponse];
                      ***REMOVED***
                        // Deviates from pure decorator by not honoring a passed ETag since we don't have a public API
                        // that allows the caller to pass an ETag.
                        request.eTag =
                            lastSuccessfulFetchResponse && lastSuccessfulFetchResponse.eTag;
                        return [4 /*yield*/, this.client.fetch(request)];
                    case 2:
                        response = _b.sent();
                        storageOperations = [
                            // Uses write-through cache for consistency with synchronous public API.
                            this.storageCache.setLastSuccessfulFetchTimestampMillis(Date.now())
                        ];
                        if (response.status === 200) ***REMOVED***
                            // Caches response only if it has changed, ie non-304 responses.
                            storageOperations.push(this.storage.setLastSuccessfulFetchResponse(response));
                      ***REMOVED***
                        return [4 /*yield*/, Promise.all(storageOperations)];
                    case 3:
                        _b.sent();
                        return [2 /*return*/, response];
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***;
    return CachingClient;
}());

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Attempts to get the most accurate browser language setting.
 *
 * <p>Adapted from getUserLanguage in packages/auth/src/utils.js for TypeScript.
 *
 * <p>Defers default language specification to server logic for consistency.
 *
 * @param navigatorLanguage Enables tests to override read-only ***REMOVED***@link NavigatorLanguage}.
 */
function getUserLanguage(navigatorLanguage) ***REMOVED***
    if (navigatorLanguage === void 0) ***REMOVED*** navigatorLanguage = navigator; }
    return (
    // Most reliable, but only supported in Chrome/Firefox.
    (navigatorLanguage.languages && navigatorLanguage.languages[0]) ||
        // Supported in most browsers, but returns the language of the browser
        // UI, not the language set in browser settings.
        navigatorLanguage.language
    // Polyfill otherwise.
    );
}

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Implements the Client abstraction for the Remote Config REST API.
 */
var RestClient = /** @class */ (function () ***REMOVED***
    function RestClient(firebaseInstallations, sdkVersion, namespace, projectId, apiKey, appId) ***REMOVED***
        this.firebaseInstallations = firebaseInstallations;
        this.sdkVersion = sdkVersion;
        this.namespace = namespace;
        this.projectId = projectId;
        this.apiKey = apiKey;
        this.appId = appId;
  ***REMOVED***
    /**
     * Fetches from the Remote Config REST API.
     *
     * @throws a ***REMOVED***@link ErrorCode.FETCH_NETWORK} error if ***REMOVED***@link GlobalFetch#fetch} can't
     * connect to the network.
     * @throws a ***REMOVED***@link ErrorCode.FETCH_PARSE} error if ***REMOVED***@link Response#json} can't parse the
     * fetch response.
     * @throws a ***REMOVED***@link ErrorCode.FETCH_STATUS} error if the service returns an HTTP error status.
     */
    RestClient.prototype.fetch = function (request) ***REMOVED***
        return __awaiter(this, void 0, void 0, function () ***REMOVED***
            var _a, installationId, installationToken, urlBase, url, headers, requestBody, options, fetchPromise, timeoutPromise, response, originalError_1, errorCode, status, responseEtag, config, state, responseBody, originalError_2;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0: return [4 /*yield*/, Promise.all([
                            this.firebaseInstallations.getId(),
                            this.firebaseInstallations.getToken()
                        ])];
                    case 1:
                        _a = _b.sent(), installationId = _a[0], installationToken = _a[1];
                        urlBase = window.FIREBASE_REMOTE_CONFIG_URL_BASE ||
                            'https://firebaseremoteconfig.googleapis.com';
                        url = "".concat(urlBase, "/v1/projects/").concat(this.projectId, "/namespaces/").concat(this.namespace, ":fetch?key=").concat(this.apiKey);
                        headers = ***REMOVED***
                            'Content-Type': 'application/json',
                            'Content-Encoding': 'gzip',
                            // Deviates from pure decorator by not passing max-age header since we don't currently have
                            // service behavior using that header.
                            'If-None-Match': request.eTag || '*'
                      ***REMOVED***;
                        requestBody = ***REMOVED***
                            /* eslint-disable camelcase */
                            sdk_version: this.sdkVersion,
                            app_instance_id: installationId,
                            app_instance_id_token: installationToken,
                            app_id: this.appId,
                            language_code: getUserLanguage()
                            /* eslint-enable camelcase */
                      ***REMOVED***;
                        options = ***REMOVED***
                            method: 'POST',
                            headers: headers,
                            body: JSON.stringify(requestBody)
                      ***REMOVED***;
                        fetchPromise = fetch(url, options);
                        timeoutPromise = new Promise(function (_resolve, reject) ***REMOVED***
                            // Maps async event listener to Promise API.
                            request.signal.addEventListener(function () ***REMOVED***
                                // Emulates https://heycam.github.io/webidl/#aborterror
                                var error = new Error('The operation was aborted.');
                                error.name = 'AbortError';
                                reject(error);
                          ***REMOVED***);
                      ***REMOVED***);
                        _b.label = 2;
                    case 2:
                        _b.trys.push([2, 5, , 6]);
                        return [4 /*yield*/, Promise.race([fetchPromise, timeoutPromise])];
                    case 3:
                        _b.sent();
                        return [4 /*yield*/, fetchPromise];
                    case 4:
                        response = _b.sent();
                        return [3 /*break*/, 6];
                    case 5:
                        originalError_1 = _b.sent();
                        errorCode = "fetch-client-network" /* ErrorCode.FETCH_NETWORK */;
                        if ((originalError_1 === null || originalError_1 === void 0 ? void 0 : originalError_1.name) === 'AbortError') ***REMOVED***
                            errorCode = "fetch-timeout" /* ErrorCode.FETCH_TIMEOUT */;
                      ***REMOVED***
                        throw ERROR_FACTORY.create(errorCode, ***REMOVED***
                            originalErrorMessage: originalError_1 === null || originalError_1 === void 0 ? void 0 : originalError_1.message
                      ***REMOVED***);
                    case 6:
                        status = response.status;
                        responseEtag = response.headers.get('ETag') || undefined;
                        if (!(response.status === 200)) return [3 /*break*/, 11];
                        responseBody = void 0;
                        _b.label = 7;
                    case 7:
                        _b.trys.push([7, 9, , 10]);
                        return [4 /*yield*/, response.json()];
                    case 8:
                        responseBody = _b.sent();
                        return [3 /*break*/, 10];
                    case 9:
                        originalError_2 = _b.sent();
                        throw ERROR_FACTORY.create("fetch-client-parse" /* ErrorCode.FETCH_PARSE */, ***REMOVED***
                            originalErrorMessage: originalError_2 === null || originalError_2 === void 0 ? void 0 : originalError_2.message
                      ***REMOVED***);
                    case 10:
                        config = responseBody['entries'];
                        state = responseBody['state'];
                        _b.label = 11;
                    case 11:
                        // Normalizes based on legacy state.
                        if (state === 'INSTANCE_STATE_UNSPECIFIED') ***REMOVED***
                            status = 500;
                      ***REMOVED***
                        else if (state === 'NO_CHANGE') ***REMOVED***
                            status = 304;
                      ***REMOVED***
                        else if (state === 'NO_TEMPLATE' || state === 'EMPTY_CONFIG') ***REMOVED***
                            // These cases can be fixed remotely, so normalize to safe value.
                            config = ***REMOVED***};
                      ***REMOVED***
                        // Normalize to exception-based control flow for non-success cases.
                        // Encapsulates HTTP specifics in this class as much as possible. Status is still the best for
                        // differentiating success states (200 from 304; the state body param is undefined in a
                        // standard 304).
                        if (status !== 304 && status !== 200) ***REMOVED***
                            throw ERROR_FACTORY.create("fetch-status" /* ErrorCode.FETCH_STATUS */, ***REMOVED***
                                httpStatus: status
                          ***REMOVED***);
                      ***REMOVED***
                        return [2 /*return*/, ***REMOVED*** status: status, eTag: responseEtag, config: config }];
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***;
    return RestClient;
}());

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Supports waiting on a backoff by:
 *
 * <ul>
 *   <li>Promisifying setTimeout, so we can set a timeout in our Promise chain</li>
 *   <li>Listening on a signal bus for abort events, just like the Fetch API</li>
 *   <li>Failing in the same way the Fetch API fails, so timing out a live request and a throttled
 *       request appear the same.</li>
 * </ul>
 *
 * <p>Visible for testing.
 */
function setAbortableTimeout(signal, throttleEndTimeMillis) ***REMOVED***
    return new Promise(function (resolve, reject) ***REMOVED***
        // Derives backoff from given end time, normalizing negative numbers to zero.
        var backoffMillis = Math.max(throttleEndTimeMillis - Date.now(), 0);
        var timeout = setTimeout(resolve, backoffMillis);
        // Adds listener, rather than sets onabort, because signal is a shared object.
        signal.addEventListener(function () ***REMOVED***
            clearTimeout(timeout);
            // If the request completes before this timeout, the rejection has no effect.
            reject(ERROR_FACTORY.create("fetch-throttle" /* ErrorCode.FETCH_THROTTLE */, ***REMOVED***
                throttleEndTimeMillis: throttleEndTimeMillis
          ***REMOVED***));
      ***REMOVED***);
  ***REMOVED***);
}
/**
 * Returns true if the ***REMOVED***@link Error} indicates a fetch request may succeed later.
 */
function isRetriableError(e) ***REMOVED***
    if (!(e instanceof FirebaseError) || !e.customData) ***REMOVED***
        return false;
  ***REMOVED***
    // Uses string index defined by ErrorData, which FirebaseError implements.
    var httpStatus = Number(e.customData['httpStatus']);
    return (httpStatus === 429 ||
        httpStatus === 500 ||
        httpStatus === 503 ||
        httpStatus === 504);
}
/**
 * Decorates a Client with retry logic.
 *
 * <p>Comparable to CachingClient, but uses backoff logic instead of cache max age and doesn't cache
 * responses (because the SDK has no use for error responses).
 */
var RetryingClient = /** @class */ (function () ***REMOVED***
    function RetryingClient(client, storage) ***REMOVED***
        this.client = client;
        this.storage = storage;
  ***REMOVED***
    RetryingClient.prototype.fetch = function (request) ***REMOVED***
        return __awaiter(this, void 0, void 0, function () ***REMOVED***
            var throttleMetadata;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this.storage.getThrottleMetadata()];
                    case 1:
                        throttleMetadata = (_a.sent()) || ***REMOVED***
                            backoffCount: 0,
                            throttleEndTimeMillis: Date.now()
                      ***REMOVED***;
                        return [2 /*return*/, this.attemptFetch(request, throttleMetadata)];
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***;
    /**
     * A recursive helper for attempting a fetch request repeatedly.
     *
     * @throws any non-retriable errors.
     */
    RetryingClient.prototype.attemptFetch = function (request, _a) ***REMOVED***
        var throttleEndTimeMillis = _a.throttleEndTimeMillis, backoffCount = _a.backoffCount;
        return __awaiter(this, void 0, void 0, function () ***REMOVED***
            var response, e_1, throttleMetadata;
            return __generator(this, function (_b) ***REMOVED***
                switch (_b.label) ***REMOVED***
                    case 0: 
                    // Starts with a (potentially zero) timeout to support resumption from stored state.
                    // Ensures the throttle end time is honored if the last attempt timed out.
                    // Note the SDK will never make a request if the fetch timeout expires at this point.
                    return [4 /*yield*/, setAbortableTimeout(request.signal, throttleEndTimeMillis)];
                    case 1:
                        // Starts with a (potentially zero) timeout to support resumption from stored state.
                        // Ensures the throttle end time is honored if the last attempt timed out.
                        // Note the SDK will never make a request if the fetch timeout expires at this point.
                        _b.sent();
                        _b.label = 2;
                    case 2:
                        _b.trys.push([2, 5, , 7]);
                        return [4 /*yield*/, this.client.fetch(request)];
                    case 3:
                        response = _b.sent();
                        // Note the SDK only clears throttle state if response is success or non-retriable.
                        return [4 /*yield*/, this.storage.deleteThrottleMetadata()];
                    case 4:
                        // Note the SDK only clears throttle state if response is success or non-retriable.
                        _b.sent();
                        return [2 /*return*/, response];
                    case 5:
                        e_1 = _b.sent();
                        if (!isRetriableError(e_1)) ***REMOVED***
                            throw e_1;
                      ***REMOVED***
                        throttleMetadata = ***REMOVED***
                            throttleEndTimeMillis: Date.now() + calculateBackoffMillis(backoffCount),
                            backoffCount: backoffCount + 1
                      ***REMOVED***;
                        // Persists state.
                        return [4 /*yield*/, this.storage.setThrottleMetadata(throttleMetadata)];
                    case 6:
                        // Persists state.
                        _b.sent();
                        return [2 /*return*/, this.attemptFetch(request, throttleMetadata)];
                    case 7: return [2 /*return*/];
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***;
    return RetryingClient;
}());

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var DEFAULT_FETCH_TIMEOUT_MILLIS = 60 * 1000; // One minute
var DEFAULT_CACHE_MAX_AGE_MILLIS = 12 * 60 * 60 * 1000; // Twelve hours.
/**
 * Encapsulates business logic mapping network and storage dependencies to the public SDK API.
 *
 * See ***REMOVED***@link https://github.com/FirebasePrivate/firebase-js-sdk/blob/master/packages/firebase/index.d.ts|interface documentation} for method descriptions.
 */
var RemoteConfig = /** @class */ (function () ***REMOVED***
    function RemoteConfig(
    // Required by FirebaseServiceFactory interface.
    app, 
    // JS doesn't support private yet
    // (https://github.com/tc39/proposal-class-fields#private-fields), so we hint using an
    // underscore prefix.
    /**
     * @internal
     */
    _client, 
    /**
     * @internal
     */
    _storageCache, 
    /**
     * @internal
     */
    _storage, 
    /**
     * @internal
     */
    _logger) ***REMOVED***
        this.app = app;
        this._client = _client;
        this._storageCache = _storageCache;
        this._storage = _storage;
        this._logger = _logger;
        /**
         * Tracks completion of initialization promise.
         * @internal
         */
        this._isInitializationComplete = false;
        this.settings = ***REMOVED***
            fetchTimeoutMillis: DEFAULT_FETCH_TIMEOUT_MILLIS,
            minimumFetchIntervalMillis: DEFAULT_CACHE_MAX_AGE_MILLIS
      ***REMOVED***;
        this.defaultConfig = ***REMOVED***};
  ***REMOVED***
    Object.defineProperty(RemoteConfig.prototype, "fetchTimeMillis", ***REMOVED***
        get: function () ***REMOVED***
            return this._storageCache.getLastSuccessfulFetchTimestampMillis() || -1;
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    Object.defineProperty(RemoteConfig.prototype, "lastFetchStatus", ***REMOVED***
        get: function () ***REMOVED***
            return this._storageCache.getLastFetchStatus() || 'no-fetch-yet';
      ***REMOVED***,
        enumerable: false,
        configurable: true
  ***REMOVED***);
    return RemoteConfig;
}());

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Converts an error event associated with a ***REMOVED***@link IDBRequest} to a ***REMOVED***@link FirebaseError}.
 */
function toFirebaseError(event, errorCode) ***REMOVED***
    var originalError = event.target.error || undefined;
    return ERROR_FACTORY.create(errorCode, ***REMOVED***
        originalErrorMessage: originalError && (originalError === null || originalError === void 0 ? void 0 : originalError.message)
  ***REMOVED***);
}
/**
 * A general-purpose store keyed by app + namespace + ***REMOVED***@link
 * ProjectNamespaceKeyFieldValue}.
 *
 * <p>The Remote Config SDK can be used with multiple app installations, and each app can interact
 * with multiple namespaces, so this store uses app (ID + name) and namespace as common parent keys
 * for a set of key-value pairs. See ***REMOVED***@link Storage#createCompositeKey}.
 *
 * <p>Visible for testing.
 */
var APP_NAMESPACE_STORE = 'app_namespace_store';
var DB_NAME = 'firebase_remote_config';
var DB_VERSION = 1;
// Visible for testing.
function openDatabase() ***REMOVED***
    return new Promise(function (resolve, reject) ***REMOVED***
        try ***REMOVED***
            var request = indexedDB.open(DB_NAME, DB_VERSION);
            request.onerror = function (event) ***REMOVED***
                reject(toFirebaseError(event, "storage-open" /* ErrorCode.STORAGE_OPEN */));
          ***REMOVED***;
            request.onsuccess = function (event) ***REMOVED***
                resolve(event.target.result);
          ***REMOVED***;
            request.onupgradeneeded = function (event) ***REMOVED***
                var db = event.target.result;
                // We don't use 'break' in this switch statement, the fall-through
                // behavior is what we want, because if there are multiple versions between
                // the old version and the current version, we want ALL the migrations
                // that correspond to those versions to run, not only the last one.
                // eslint-disable-next-line default-case
                switch (event.oldVersion) ***REMOVED***
                    case 0:
                        db.createObjectStore(APP_NAMESPACE_STORE, ***REMOVED***
                            keyPath: 'compositeKey'
                      ***REMOVED***);
              ***REMOVED***
          ***REMOVED***;
      ***REMOVED***
        catch (error) ***REMOVED***
            reject(ERROR_FACTORY.create("storage-open" /* ErrorCode.STORAGE_OPEN */, ***REMOVED***
                originalErrorMessage: error === null || error === void 0 ? void 0 : error.message
          ***REMOVED***));
      ***REMOVED***
  ***REMOVED***);
}
/**
 * Abstracts data persistence.
 */
var Storage = /** @class */ (function () ***REMOVED***
    /**
     * @param appId enables storage segmentation by app (ID + name).
     * @param appName enables storage segmentation by app (ID + name).
     * @param namespace enables storage segmentation by namespace.
     */
    function Storage(appId, appName, namespace, openDbPromise) ***REMOVED***
        if (openDbPromise === void 0) ***REMOVED*** openDbPromise = openDatabase(); }
        this.appId = appId;
        this.appName = appName;
        this.namespace = namespace;
        this.openDbPromise = openDbPromise;
  ***REMOVED***
    Storage.prototype.getLastFetchStatus = function () ***REMOVED***
        return this.get('last_fetch_status');
  ***REMOVED***;
    Storage.prototype.setLastFetchStatus = function (status) ***REMOVED***
        return this.set('last_fetch_status', status);
  ***REMOVED***;
    // This is comparable to a cache entry timestamp. If we need to expire other data, we could
    // consider adding timestamp to all storage records and an optional max age arg to getters.
    Storage.prototype.getLastSuccessfulFetchTimestampMillis = function () ***REMOVED***
        return this.get('last_successful_fetch_timestamp_millis');
  ***REMOVED***;
    Storage.prototype.setLastSuccessfulFetchTimestampMillis = function (timestamp) ***REMOVED***
        return this.set('last_successful_fetch_timestamp_millis', timestamp);
  ***REMOVED***;
    Storage.prototype.getLastSuccessfulFetchResponse = function () ***REMOVED***
        return this.get('last_successful_fetch_response');
  ***REMOVED***;
    Storage.prototype.setLastSuccessfulFetchResponse = function (response) ***REMOVED***
        return this.set('last_successful_fetch_response', response);
  ***REMOVED***;
    Storage.prototype.getActiveConfig = function () ***REMOVED***
        return this.get('active_config');
  ***REMOVED***;
    Storage.prototype.setActiveConfig = function (config) ***REMOVED***
        return this.set('active_config', config);
  ***REMOVED***;
    Storage.prototype.getActiveConfigEtag = function () ***REMOVED***
        return this.get('active_config_etag');
  ***REMOVED***;
    Storage.prototype.setActiveConfigEtag = function (etag) ***REMOVED***
        return this.set('active_config_etag', etag);
  ***REMOVED***;
    Storage.prototype.getThrottleMetadata = function () ***REMOVED***
        return this.get('throttle_metadata');
  ***REMOVED***;
    Storage.prototype.setThrottleMetadata = function (metadata) ***REMOVED***
        return this.set('throttle_metadata', metadata);
  ***REMOVED***;
    Storage.prototype.deleteThrottleMetadata = function () ***REMOVED***
        return this.delete('throttle_metadata');
  ***REMOVED***;
    Storage.prototype.get = function (key) ***REMOVED***
        return __awaiter(this, void 0, void 0, function () ***REMOVED***
            var db;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this.openDbPromise];
                    case 1:
                        db = _a.sent();
                        return [2 /*return*/, new Promise(function (resolve, reject) ***REMOVED***
                                var transaction = db.transaction([APP_NAMESPACE_STORE], 'readonly');
                                var objectStore = transaction.objectStore(APP_NAMESPACE_STORE);
                                var compositeKey = _this.createCompositeKey(key);
                                try ***REMOVED***
                                    var request = objectStore.get(compositeKey);
                                    request.onerror = function (event) ***REMOVED***
                                        reject(toFirebaseError(event, "storage-get" /* ErrorCode.STORAGE_GET */));
                                  ***REMOVED***;
                                    request.onsuccess = function (event) ***REMOVED***
                                        var result = event.target.result;
                                        if (result) ***REMOVED***
                                            resolve(result.value);
                                      ***REMOVED***
                                        else ***REMOVED***
                                            resolve(undefined);
                                      ***REMOVED***
                                  ***REMOVED***;
                              ***REMOVED***
                                catch (e) ***REMOVED***
                                    reject(ERROR_FACTORY.create("storage-get" /* ErrorCode.STORAGE_GET */, ***REMOVED***
                                        originalErrorMessage: e === null || e === void 0 ? void 0 : e.message
                                  ***REMOVED***));
                              ***REMOVED***
                          ***REMOVED***)];
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***;
    Storage.prototype.set = function (key, value) ***REMOVED***
        return __awaiter(this, void 0, void 0, function () ***REMOVED***
            var db;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this.openDbPromise];
                    case 1:
                        db = _a.sent();
                        return [2 /*return*/, new Promise(function (resolve, reject) ***REMOVED***
                                var transaction = db.transaction([APP_NAMESPACE_STORE], 'readwrite');
                                var objectStore = transaction.objectStore(APP_NAMESPACE_STORE);
                                var compositeKey = _this.createCompositeKey(key);
                                try ***REMOVED***
                                    var request = objectStore.put(***REMOVED***
                                        compositeKey: compositeKey,
                                        value: value
                                  ***REMOVED***);
                                    request.onerror = function (event) ***REMOVED***
                                        reject(toFirebaseError(event, "storage-set" /* ErrorCode.STORAGE_SET */));
                                  ***REMOVED***;
                                    request.onsuccess = function () ***REMOVED***
                                        resolve();
                                  ***REMOVED***;
                              ***REMOVED***
                                catch (e) ***REMOVED***
                                    reject(ERROR_FACTORY.create("storage-set" /* ErrorCode.STORAGE_SET */, ***REMOVED***
                                        originalErrorMessage: e === null || e === void 0 ? void 0 : e.message
                                  ***REMOVED***));
                              ***REMOVED***
                          ***REMOVED***)];
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***;
    Storage.prototype.delete = function (key) ***REMOVED***
        return __awaiter(this, void 0, void 0, function () ***REMOVED***
            var db;
            var _this = this;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0: return [4 /*yield*/, this.openDbPromise];
                    case 1:
                        db = _a.sent();
                        return [2 /*return*/, new Promise(function (resolve, reject) ***REMOVED***
                                var transaction = db.transaction([APP_NAMESPACE_STORE], 'readwrite');
                                var objectStore = transaction.objectStore(APP_NAMESPACE_STORE);
                                var compositeKey = _this.createCompositeKey(key);
                                try ***REMOVED***
                                    var request = objectStore.delete(compositeKey);
                                    request.onerror = function (event) ***REMOVED***
                                        reject(toFirebaseError(event, "storage-delete" /* ErrorCode.STORAGE_DELETE */));
                                  ***REMOVED***;
                                    request.onsuccess = function () ***REMOVED***
                                        resolve();
                                  ***REMOVED***;
                              ***REMOVED***
                                catch (e) ***REMOVED***
                                    reject(ERROR_FACTORY.create("storage-delete" /* ErrorCode.STORAGE_DELETE */, ***REMOVED***
                                        originalErrorMessage: e === null || e === void 0 ? void 0 : e.message
                                  ***REMOVED***));
                              ***REMOVED***
                          ***REMOVED***)];
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***;
    // Facilitates composite key functionality (which is unsupported in IE).
    Storage.prototype.createCompositeKey = function (key) ***REMOVED***
        return [this.appId, this.appName, this.namespace, key].join();
  ***REMOVED***;
    return Storage;
}());

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * A memory cache layer over storage to support the SDK's synchronous read requirements.
 */
var StorageCache = /** @class */ (function () ***REMOVED***
    function StorageCache(storage) ***REMOVED***
        this.storage = storage;
  ***REMOVED***
    /**
     * Memory-only getters
     */
    StorageCache.prototype.getLastFetchStatus = function () ***REMOVED***
        return this.lastFetchStatus;
  ***REMOVED***;
    StorageCache.prototype.getLastSuccessfulFetchTimestampMillis = function () ***REMOVED***
        return this.lastSuccessfulFetchTimestampMillis;
  ***REMOVED***;
    StorageCache.prototype.getActiveConfig = function () ***REMOVED***
        return this.activeConfig;
  ***REMOVED***;
    /**
     * Read-ahead getter
     */
    StorageCache.prototype.loadFromStorage = function () ***REMOVED***
        return __awaiter(this, void 0, void 0, function () ***REMOVED***
            var lastFetchStatusPromise, lastSuccessfulFetchTimestampMillisPromise, activeConfigPromise, lastFetchStatus, lastSuccessfulFetchTimestampMillis, activeConfig;
            return __generator(this, function (_a) ***REMOVED***
                switch (_a.label) ***REMOVED***
                    case 0:
                        lastFetchStatusPromise = this.storage.getLastFetchStatus();
                        lastSuccessfulFetchTimestampMillisPromise = this.storage.getLastSuccessfulFetchTimestampMillis();
                        activeConfigPromise = this.storage.getActiveConfig();
                        return [4 /*yield*/, lastFetchStatusPromise];
                    case 1:
                        lastFetchStatus = _a.sent();
                        if (lastFetchStatus) ***REMOVED***
                            this.lastFetchStatus = lastFetchStatus;
                      ***REMOVED***
                        return [4 /*yield*/, lastSuccessfulFetchTimestampMillisPromise];
                    case 2:
                        lastSuccessfulFetchTimestampMillis = _a.sent();
                        if (lastSuccessfulFetchTimestampMillis) ***REMOVED***
                            this.lastSuccessfulFetchTimestampMillis =
                                lastSuccessfulFetchTimestampMillis;
                      ***REMOVED***
                        return [4 /*yield*/, activeConfigPromise];
                    case 3:
                        activeConfig = _a.sent();
                        if (activeConfig) ***REMOVED***
                            this.activeConfig = activeConfig;
                      ***REMOVED***
                        return [2 /*return*/];
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***;
    /**
     * Write-through setters
     */
    StorageCache.prototype.setLastFetchStatus = function (status) ***REMOVED***
        this.lastFetchStatus = status;
        return this.storage.setLastFetchStatus(status);
  ***REMOVED***;
    StorageCache.prototype.setLastSuccessfulFetchTimestampMillis = function (timestampMillis) ***REMOVED***
        this.lastSuccessfulFetchTimestampMillis = timestampMillis;
        return this.storage.setLastSuccessfulFetchTimestampMillis(timestampMillis);
  ***REMOVED***;
    StorageCache.prototype.setActiveConfig = function (activeConfig) ***REMOVED***
        this.activeConfig = activeConfig;
        return this.storage.setActiveConfig(activeConfig);
  ***REMOVED***;
    return StorageCache;
}());

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function registerRemoteConfig() ***REMOVED***
    _registerComponent(new Component(RC_COMPONENT_NAME, remoteConfigFactory, "PUBLIC" /* ComponentType.PUBLIC */).setMultipleInstances(true));
    registerVersion(name, version);
    // BUILD_TARGET will be replaced by values like esm5, esm2017, cjs5, etc during the compilation
    registerVersion(name, version, 'esm5');
    function remoteConfigFactory(container, _a) ***REMOVED***
        var namespace = _a.instanceIdentifier;
        /* Dependencies */
        // getImmediate for FirebaseApp will always succeed
        var app = container.getProvider('app').getImmediate();
        // The following call will always succeed because rc has `import '@firebase/installations'`
        var installations = container
            .getProvider('installations-internal')
            .getImmediate();
        // Guards against the SDK being used in non-browser environments.
        if (typeof window === 'undefined') ***REMOVED***
            throw ERROR_FACTORY.create("registration-window" /* ErrorCode.REGISTRATION_WINDOW */);
      ***REMOVED***
        // Guards against the SDK being used when indexedDB is not available.
        if (!isIndexedDBAvailable()) ***REMOVED***
            throw ERROR_FACTORY.create("indexed-db-unavailable" /* ErrorCode.INDEXED_DB_UNAVAILABLE */);
      ***REMOVED***
        // Normalizes optional inputs.
        var _b = app.options, projectId = _b.projectId, apiKey = _b.apiKey, appId = _b.appId;
        if (!projectId) ***REMOVED***
            throw ERROR_FACTORY.create("registration-project-id" /* ErrorCode.REGISTRATION_PROJECT_ID */);
      ***REMOVED***
        if (!apiKey) ***REMOVED***
            throw ERROR_FACTORY.create("registration-api-key" /* ErrorCode.REGISTRATION_API_KEY */);
      ***REMOVED***
        if (!appId) ***REMOVED***
            throw ERROR_FACTORY.create("registration-app-id" /* ErrorCode.REGISTRATION_APP_ID */);
      ***REMOVED***
        namespace = namespace || 'firebase';
        var storage = new Storage(appId, app.name, namespace);
        var storageCache = new StorageCache(storage);
        var logger = new Logger(name);
        // Sets ERROR as the default log level.
        // See RemoteConfig#setLogLevel for corresponding normalization to ERROR log level.
        logger.logLevel = LogLevel.ERROR;
        var restClient = new RestClient(installations, 
        // Uses the JS SDK version, by which the RC package version can be deduced, if necessary.
        SDK_VERSION, namespace, projectId, apiKey, appId);
        var retryingClient = new RetryingClient(restClient, storage);
        var cachingClient = new CachingClient(retryingClient, storage, storageCache, logger);
        var remoteConfigInstance = new RemoteConfig(app, cachingClient, storageCache, storage, logger);
        // Starts warming cache.
        // eslint-disable-next-line @typescript-eslint/no-floating-promises
        ensureInitialized(remoteConfigInstance);
        return remoteConfigInstance;
  ***REMOVED***
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// This API is put in a separate file, so we can stub fetchConfig and activate in tests.
// It's not possible to stub standalone functions from the same module.
/**
 *
 * Performs fetch and activate operations, as a convenience.
 *
 * @param remoteConfig - The ***REMOVED***@link RemoteConfig} instance.
 *
 * @returns A `Promise` which resolves to true if the current call activated the fetched configs.
 * If the fetched configs were already activated, the `Promise` will resolve to false.
 *
 * @public
 */
function fetchAndActivate(remoteConfig) ***REMOVED***
    return __awaiter(this, void 0, void 0, function () ***REMOVED***
        return __generator(this, function (_a) ***REMOVED***
            switch (_a.label) ***REMOVED***
                case 0:
                    remoteConfig = getModularInstance(remoteConfig);
                    return [4 /*yield*/, fetchConfig(remoteConfig)];
                case 1:
                    _a.sent();
                    return [2 /*return*/, activate(remoteConfig)];
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***);
}
/**
 * This method provides two different checks:
 *
 * 1. Check if IndexedDB exists in the browser environment.
 * 2. Check if the current browser context allows IndexedDB `open()` calls.
 *
 * @returns A `Promise` which resolves to true if a ***REMOVED***@link RemoteConfig} instance
 * can be initialized in this environment, or false if it cannot.
 * @public
 */
function isSupported() ***REMOVED***
    return __awaiter(this, void 0, void 0, function () ***REMOVED***
        var isDBOpenable;
        return __generator(this, function (_a) ***REMOVED***
            switch (_a.label) ***REMOVED***
                case 0:
                    if (!isIndexedDBAvailable()) ***REMOVED***
                        return [2 /*return*/, false];
                  ***REMOVED***
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 3, , 4]);
                    return [4 /*yield*/, validateIndexedDBOpenable()];
                case 2:
                    isDBOpenable = _a.sent();
                    return [2 /*return*/, isDBOpenable];
                case 3:
                    _a.sent();
                    return [2 /*return*/, false];
                case 4: return [2 /*return*/];
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***);
}

/**
 * Firebase Remote Config
 *
 * @packageDocumentation
 */
/** register component and version */
registerRemoteConfig();

export ***REMOVED*** activate, ensureInitialized, fetchAndActivate, fetchConfig, getAll, getBoolean, getNumber, getRemoteConfig, getString, getValue, isSupported, setLogLevel };
