// hoist-non-react-statics/dist/hoist-non-react-statics.js
(function (global, factory) ***REMOVED***
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global = global || self, global.hoistNonReactStatics = factory());
}(this, (function () ***REMOVED*** 'use strict';

	function unwrapExports (x) ***REMOVED***
		return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
	}

	function createCommonjsModule(fn, module) ***REMOVED***
		return module = ***REMOVED*** exports: ***REMOVED***} }, fn(module, module.exports), module.exports;
	}

	var reactIs_production_min = createCommonjsModule(function (module, exports) ***REMOVED***
	Object.defineProperty(exports,"__esModule",***REMOVED***value:!0});
	var b="function"===typeof Symbol&&Symbol.for,c=b?Symbol.for("react.element"):60103,d=b?Symbol.for("react.portal"):60106,e=b?Symbol.for("react.fragment"):60107,f=b?Symbol.for("react.strict_mode"):60108,g=b?Symbol.for("react.profiler"):60114,h=b?Symbol.for("react.provider"):60109,k=b?Symbol.for("react.context"):60110,l=b?Symbol.for("react.async_mode"):60111,m=b?Symbol.for("react.concurrent_mode"):60111,n=b?Symbol.for("react.forward_ref"):60112,p=b?Symbol.for("react.suspense"):60113,q=b?Symbol.for("react.suspense_list"):
	60120,r=b?Symbol.for("react.memo"):60115,t=b?Symbol.for("react.lazy"):60116,v=b?Symbol.for("react.fundamental"):60117,w=b?Symbol.for("react.responder"):60118,x=b?Symbol.for("react.scope"):60119;function y(a)***REMOVED***if("object"===typeof a&&null!==a)***REMOVED***var u=a.$$typeof;switch(u)***REMOVED***case c:switch(a=a.type,a)***REMOVED***case l:case m:case e:case g:case f:case p:return a;default:switch(a=a&&a.$$typeof,a)***REMOVED***case k:case n:case t:case r:case h:return a;default:return u}}case d:return u}}}function z(a)***REMOVED***return y(a)===m}
	exports.typeOf=y;exports.AsyncMode=l;exports.ConcurrentMode=m;exports.ContextConsumer=k;exports.ContextProvider=h;exports.Element=c;exports.ForwardRef=n;exports.Fragment=e;exports.Lazy=t;exports.Memo=r;exports.Portal=d;exports.Profiler=g;exports.StrictMode=f;exports.Suspense=p;
	exports.isValidElementType=function(a)***REMOVED***return "string"===typeof a||"function"===typeof a||a===e||a===m||a===g||a===f||a===p||a===q||"object"===typeof a&&null!==a&&(a.$$typeof===t||a.$$typeof===r||a.$$typeof===h||a.$$typeof===k||a.$$typeof===n||a.$$typeof===v||a.$$typeof===w||a.$$typeof===x)};exports.isAsyncMode=function(a)***REMOVED***return z(a)||y(a)===l};exports.isConcurrentMode=z;exports.isContextConsumer=function(a)***REMOVED***return y(a)===k};exports.isContextProvider=function(a)***REMOVED***return y(a)===h};
	exports.isElement=function(a)***REMOVED***return "object"===typeof a&&null!==a&&a.$$typeof===c};exports.isForwardRef=function(a)***REMOVED***return y(a)===n};exports.isFragment=function(a)***REMOVED***return y(a)===e};exports.isLazy=function(a)***REMOVED***return y(a)===t};exports.isMemo=function(a)***REMOVED***return y(a)===r};exports.isPortal=function(a)***REMOVED***return y(a)===d};exports.isProfiler=function(a)***REMOVED***return y(a)===g};exports.isStrictMode=function(a)***REMOVED***return y(a)===f};exports.isSuspense=function(a)***REMOVED***return y(a)===p};
	});

	unwrapExports(reactIs_production_min);
	var reactIs_production_min_1 = reactIs_production_min.typeOf;
	var reactIs_production_min_2 = reactIs_production_min.AsyncMode;
	var reactIs_production_min_3 = reactIs_production_min.ConcurrentMode;
	var reactIs_production_min_4 = reactIs_production_min.ContextConsumer;
	var reactIs_production_min_5 = reactIs_production_min.ContextProvider;
	var reactIs_production_min_6 = reactIs_production_min.Element;
	var reactIs_production_min_7 = reactIs_production_min.ForwardRef;
	var reactIs_production_min_8 = reactIs_production_min.Fragment;
	var reactIs_production_min_9 = reactIs_production_min.Lazy;
	var reactIs_production_min_10 = reactIs_production_min.Memo;
	var reactIs_production_min_11 = reactIs_production_min.Portal;
	var reactIs_production_min_12 = reactIs_production_min.Profiler;
	var reactIs_production_min_13 = reactIs_production_min.StrictMode;
	var reactIs_production_min_14 = reactIs_production_min.Suspense;
	var reactIs_production_min_15 = reactIs_production_min.isValidElementType;
	var reactIs_production_min_16 = reactIs_production_min.isAsyncMode;
	var reactIs_production_min_17 = reactIs_production_min.isConcurrentMode;
	var reactIs_production_min_18 = reactIs_production_min.isContextConsumer;
	var reactIs_production_min_19 = reactIs_production_min.isContextProvider;
	var reactIs_production_min_20 = reactIs_production_min.isElement;
	var reactIs_production_min_21 = reactIs_production_min.isForwardRef;
	var reactIs_production_min_22 = reactIs_production_min.isFragment;
	var reactIs_production_min_23 = reactIs_production_min.isLazy;
	var reactIs_production_min_24 = reactIs_production_min.isMemo;
	var reactIs_production_min_25 = reactIs_production_min.isPortal;
	var reactIs_production_min_26 = reactIs_production_min.isProfiler;
	var reactIs_production_min_27 = reactIs_production_min.isStrictMode;
	var reactIs_production_min_28 = reactIs_production_min.isSuspense;

	var reactIs_development = createCommonjsModule(function (module, exports) ***REMOVED***



	if (process.env.NODE_ENV !== "production") ***REMOVED***
	  (function() ***REMOVED***

	Object.defineProperty(exports, '__esModule', ***REMOVED*** value: true });

	// The Symbol used to tag the ReactElement-like types. If there is no native Symbol
	// nor polyfill, then a plain number is used for performance.
	var hasSymbol = typeof Symbol === 'function' && Symbol.for;
	var REACT_ELEMENT_TYPE = hasSymbol ? Symbol.for('react.element') : 0xeac7;
	var REACT_PORTAL_TYPE = hasSymbol ? Symbol.for('react.portal') : 0xeaca;
	var REACT_FRAGMENT_TYPE = hasSymbol ? Symbol.for('react.fragment') : 0xeacb;
	var REACT_STRICT_MODE_TYPE = hasSymbol ? Symbol.for('react.strict_mode') : 0xeacc;
	var REACT_PROFILER_TYPE = hasSymbol ? Symbol.for('react.profiler') : 0xead2;
	var REACT_PROVIDER_TYPE = hasSymbol ? Symbol.for('react.provider') : 0xeacd;
	var REACT_CONTEXT_TYPE = hasSymbol ? Symbol.for('react.context') : 0xeace; // TODO: We don't use AsyncMode or ConcurrentMode anymore. They were temporary
	// (unstable) APIs that have been removed. Can we remove the symbols?

	var REACT_ASYNC_MODE_TYPE = hasSymbol ? Symbol.for('react.async_mode') : 0xeacf;
	var REACT_CONCURRENT_MODE_TYPE = hasSymbol ? Symbol.for('react.concurrent_mode') : 0xeacf;
	var REACT_FORWARD_REF_TYPE = hasSymbol ? Symbol.for('react.forward_ref') : 0xead0;
	var REACT_SUSPENSE_TYPE = hasSymbol ? Symbol.for('react.suspense') : 0xead1;
	var REACT_SUSPENSE_LIST_TYPE = hasSymbol ? Symbol.for('react.suspense_list') : 0xead8;
	var REACT_MEMO_TYPE = hasSymbol ? Symbol.for('react.memo') : 0xead3;
	var REACT_LAZY_TYPE = hasSymbol ? Symbol.for('react.lazy') : 0xead4;
	var REACT_FUNDAMENTAL_TYPE = hasSymbol ? Symbol.for('react.fundamental') : 0xead5;
	var REACT_RESPONDER_TYPE = hasSymbol ? Symbol.for('react.responder') : 0xead6;
	var REACT_SCOPE_TYPE = hasSymbol ? Symbol.for('react.scope') : 0xead7;

	function isValidElementType(type) ***REMOVED***
	  return typeof type === 'string' || typeof type === 'function' || // Note: its typeof might be other than 'symbol' or 'number' if it's a polyfill.
	  type === REACT_FRAGMENT_TYPE || type === REACT_CONCURRENT_MODE_TYPE || type === REACT_PROFILER_TYPE || type === REACT_STRICT_MODE_TYPE || type === REACT_SUSPENSE_TYPE || type === REACT_SUSPENSE_LIST_TYPE || typeof type === 'object' && type !== null && (type.$$typeof === REACT_LAZY_TYPE || type.$$typeof === REACT_MEMO_TYPE || type.$$typeof === REACT_PROVIDER_TYPE || type.$$typeof === REACT_CONTEXT_TYPE || type.$$typeof === REACT_FORWARD_REF_TYPE || type.$$typeof === REACT_FUNDAMENTAL_TYPE || type.$$typeof === REACT_RESPONDER_TYPE || type.$$typeof === REACT_SCOPE_TYPE);
	}

	/**
	 * Forked from fbjs/warning:
	 * https://github.com/facebook/fbjs/blob/e66ba20ad5be433eb54423f2b097d829324d9de6/packages/fbjs/src/__forks__/warning.js
	 *
	 * Only change is we use console.warn instead of console.error,
	 * and do nothing when 'console' is not supported.
	 * This really simplifies the code.
	 * ---
	 * Similar to invariant but only logs a warning if the condition is not met.
	 * This can be used to log issues in development environments in critical
	 * paths. Removing the logging code for production environments will keep the
	 * same logic and follow the same code paths.
	 */
	var lowPriorityWarningWithoutStack = function () ***REMOVED***};

	***REMOVED***
	  var printWarning = function (format) ***REMOVED***
	    for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) ***REMOVED***
	      args[_key - 1] = arguments[_key];
	  ***REMOVED***

	    var argIndex = 0;
	    var message = 'Warning: ' + format.replace(/%s/g, function () ***REMOVED***
	      return args[argIndex++];
	  ***REMOVED***);

	    if (typeof console !== 'undefined') ***REMOVED***
	      console.warn(message);
	  ***REMOVED***

	    try ***REMOVED***
	      // --- Welcome to debugging React ---
	      // This error was thrown as a convenience so that you can use this stack
	      // to find the callsite that caused this warning to fire.
	      throw new Error(message);
	  ***REMOVED*** catch (x) ***REMOVED***}
	***REMOVED***;

	  lowPriorityWarningWithoutStack = function (condition, format) ***REMOVED***
	    if (format === undefined) ***REMOVED***
	      throw new Error('`lowPriorityWarningWithoutStack(condition, format, ...args)` requires a warning ' + 'message argument');
	  ***REMOVED***

	    if (!condition) ***REMOVED***
	      for (var _len2 = arguments.length, args = new Array(_len2 > 2 ? _len2 - 2 : 0), _key2 = 2; _key2 < _len2; _key2++) ***REMOVED***
	        args[_key2 - 2] = arguments[_key2];
	    ***REMOVED***

	      printWarning.apply(void 0, [format].concat(args));
	  ***REMOVED***
	***REMOVED***;
	}

	var lowPriorityWarningWithoutStack$1 = lowPriorityWarningWithoutStack;

	function typeOf(object) ***REMOVED***
	  if (typeof object === 'object' && object !== null) ***REMOVED***
	    var $$typeof = object.$$typeof;

	    switch ($$typeof) ***REMOVED***
	      case REACT_ELEMENT_TYPE:
	        var type = object.type;

	        switch (type) ***REMOVED***
	          case REACT_ASYNC_MODE_TYPE:
	          case REACT_CONCURRENT_MODE_TYPE:
	          case REACT_FRAGMENT_TYPE:
	          case REACT_PROFILER_TYPE:
	          case REACT_STRICT_MODE_TYPE:
	          case REACT_SUSPENSE_TYPE:
	            return type;

	          default:
	            var $$typeofType = type && type.$$typeof;

	            switch ($$typeofType) ***REMOVED***
	              case REACT_CONTEXT_TYPE:
	              case REACT_FORWARD_REF_TYPE:
	              case REACT_LAZY_TYPE:
	              case REACT_MEMO_TYPE:
	              case REACT_PROVIDER_TYPE:
	                return $$typeofType;

	              default:
	                return $$typeof;
	          ***REMOVED***

	      ***REMOVED***

	      case REACT_PORTAL_TYPE:
	        return $$typeof;
	  ***REMOVED***
	***REMOVED***

	  return undefined;
	} // AsyncMode is deprecated along with isAsyncMode

	var AsyncMode = REACT_ASYNC_MODE_TYPE;
	var ConcurrentMode = REACT_CONCURRENT_MODE_TYPE;
	var ContextConsumer = REACT_CONTEXT_TYPE;
	var ContextProvider = REACT_PROVIDER_TYPE;
	var Element = REACT_ELEMENT_TYPE;
	var ForwardRef = REACT_FORWARD_REF_TYPE;
	var Fragment = REACT_FRAGMENT_TYPE;
	var Lazy = REACT_LAZY_TYPE;
	var Memo = REACT_MEMO_TYPE;
	var Portal = REACT_PORTAL_TYPE;
	var Profiler = REACT_PROFILER_TYPE;
	var StrictMode = REACT_STRICT_MODE_TYPE;
	var Suspense = REACT_SUSPENSE_TYPE;
	var hasWarnedAboutDeprecatedIsAsyncMode = false; // AsyncMode should be deprecated

	function isAsyncMode(object) ***REMOVED***
	  ***REMOVED***
	    if (!hasWarnedAboutDeprecatedIsAsyncMode) ***REMOVED***
	      hasWarnedAboutDeprecatedIsAsyncMode = true;
	      lowPriorityWarningWithoutStack$1(false, 'The ReactIs.isAsyncMode() alias has been deprecated, ' + 'and will be removed in React 17+. Update your code to use ' + 'ReactIs.isConcurrentMode() instead. It has the exact same API.');
	  ***REMOVED***
	***REMOVED***

	  return isConcurrentMode(object) || typeOf(object) === REACT_ASYNC_MODE_TYPE;
	}
	function isConcurrentMode(object) ***REMOVED***
	  return typeOf(object) === REACT_CONCURRENT_MODE_TYPE;
	}
	function isContextConsumer(object) ***REMOVED***
	  return typeOf(object) === REACT_CONTEXT_TYPE;
	}
	function isContextProvider(object) ***REMOVED***
	  return typeOf(object) === REACT_PROVIDER_TYPE;
	}
	function isElement(object) ***REMOVED***
	  return typeof object === 'object' && object !== null && object.$$typeof === REACT_ELEMENT_TYPE;
	}
	function isForwardRef(object) ***REMOVED***
	  return typeOf(object) === REACT_FORWARD_REF_TYPE;
	}
	function isFragment(object) ***REMOVED***
	  return typeOf(object) === REACT_FRAGMENT_TYPE;
	}
	function isLazy(object) ***REMOVED***
	  return typeOf(object) === REACT_LAZY_TYPE;
	}
	function isMemo(object) ***REMOVED***
	  return typeOf(object) === REACT_MEMO_TYPE;
	}
	function isPortal(object) ***REMOVED***
	  return typeOf(object) === REACT_PORTAL_TYPE;
	}
	function isProfiler(object) ***REMOVED***
	  return typeOf(object) === REACT_PROFILER_TYPE;
	}
	function isStrictMode(object) ***REMOVED***
	  return typeOf(object) === REACT_STRICT_MODE_TYPE;
	}
	function isSuspense(object) ***REMOVED***
	  return typeOf(object) === REACT_SUSPENSE_TYPE;
	}

	exports.typeOf = typeOf;
	exports.AsyncMode = AsyncMode;
	exports.ConcurrentMode = ConcurrentMode;
	exports.ContextConsumer = ContextConsumer;
	exports.ContextProvider = ContextProvider;
	exports.Element = Element;
	exports.ForwardRef = ForwardRef;
	exports.Fragment = Fragment;
	exports.Lazy = Lazy;
	exports.Memo = Memo;
	exports.Portal = Portal;
	exports.Profiler = Profiler;
	exports.StrictMode = StrictMode;
	exports.Suspense = Suspense;
	exports.isValidElementType = isValidElementType;
	exports.isAsyncMode = isAsyncMode;
	exports.isConcurrentMode = isConcurrentMode;
	exports.isContextConsumer = isContextConsumer;
	exports.isContextProvider = isContextProvider;
	exports.isElement = isElement;
	exports.isForwardRef = isForwardRef;
	exports.isFragment = isFragment;
	exports.isLazy = isLazy;
	exports.isMemo = isMemo;
	exports.isPortal = isPortal;
	exports.isProfiler = isProfiler;
	exports.isStrictMode = isStrictMode;
	exports.isSuspense = isSuspense;
	***REMOVED***)();
	}
	});

	unwrapExports(reactIs_development);
	var reactIs_development_1 = reactIs_development.typeOf;
	var reactIs_development_2 = reactIs_development.AsyncMode;
	var reactIs_development_3 = reactIs_development.ConcurrentMode;
	var reactIs_development_4 = reactIs_development.ContextConsumer;
	var reactIs_development_5 = reactIs_development.ContextProvider;
	var reactIs_development_6 = reactIs_development.Element;
	var reactIs_development_7 = reactIs_development.ForwardRef;
	var reactIs_development_8 = reactIs_development.Fragment;
	var reactIs_development_9 = reactIs_development.Lazy;
	var reactIs_development_10 = reactIs_development.Memo;
	var reactIs_development_11 = reactIs_development.Portal;
	var reactIs_development_12 = reactIs_development.Profiler;
	var reactIs_development_13 = reactIs_development.StrictMode;
	var reactIs_development_14 = reactIs_development.Suspense;
	var reactIs_development_15 = reactIs_development.isValidElementType;
	var reactIs_development_16 = reactIs_development.isAsyncMode;
	var reactIs_development_17 = reactIs_development.isConcurrentMode;
	var reactIs_development_18 = reactIs_development.isContextConsumer;
	var reactIs_development_19 = reactIs_development.isContextProvider;
	var reactIs_development_20 = reactIs_development.isElement;
	var reactIs_development_21 = reactIs_development.isForwardRef;
	var reactIs_development_22 = reactIs_development.isFragment;
	var reactIs_development_23 = reactIs_development.isLazy;
	var reactIs_development_24 = reactIs_development.isMemo;
	var reactIs_development_25 = reactIs_development.isPortal;
	var reactIs_development_26 = reactIs_development.isProfiler;
	var reactIs_development_27 = reactIs_development.isStrictMode;
	var reactIs_development_28 = reactIs_development.isSuspense;

	var reactIs = createCommonjsModule(function (module) ***REMOVED***

	if (process.env.NODE_ENV === 'production') ***REMOVED***
	  module.exports = reactIs_production_min;
	} else ***REMOVED***
	  module.exports = reactIs_development;
	}
	});
	var reactIs_1 = reactIs.typeOf;
	var reactIs_2 = reactIs.AsyncMode;
	var reactIs_3 = reactIs.ConcurrentMode;
	var reactIs_4 = reactIs.ContextConsumer;
	var reactIs_5 = reactIs.ContextProvider;
	var reactIs_6 = reactIs.Element;
	var reactIs_7 = reactIs.ForwardRef;
	var reactIs_8 = reactIs.Fragment;
	var reactIs_9 = reactIs.Lazy;
	var reactIs_10 = reactIs.Memo;
	var reactIs_11 = reactIs.Portal;
	var reactIs_12 = reactIs.Profiler;
	var reactIs_13 = reactIs.StrictMode;
	var reactIs_14 = reactIs.Suspense;
	var reactIs_15 = reactIs.isValidElementType;
	var reactIs_16 = reactIs.isAsyncMode;
	var reactIs_17 = reactIs.isConcurrentMode;
	var reactIs_18 = reactIs.isContextConsumer;
	var reactIs_19 = reactIs.isContextProvider;
	var reactIs_20 = reactIs.isElement;
	var reactIs_21 = reactIs.isForwardRef;
	var reactIs_22 = reactIs.isFragment;
	var reactIs_23 = reactIs.isLazy;
	var reactIs_24 = reactIs.isMemo;
	var reactIs_25 = reactIs.isPortal;
	var reactIs_26 = reactIs.isProfiler;
	var reactIs_27 = reactIs.isStrictMode;
	var reactIs_28 = reactIs.isSuspense;

	/**
	 * Copyright 2015, Yahoo! Inc.
	 * Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
	 */
	var REACT_STATICS = ***REMOVED***
	  childContextTypes: true,
	  contextType: true,
	  contextTypes: true,
	  defaultProps: true,
	  displayName: true,
	  getDefaultProps: true,
	  getDerivedStateFromError: true,
	  getDerivedStateFromProps: true,
	  mixins: true,
	  propTypes: true,
	  type: true
	};
	var KNOWN_STATICS = ***REMOVED***
	  name: true,
	  length: true,
	  prototype: true,
	  caller: true,
	  callee: true,
	  arguments: true,
	  arity: true
	};
	var FORWARD_REF_STATICS = ***REMOVED***
	  '$$typeof': true,
	  render: true,
	  defaultProps: true,
	  displayName: true,
	  propTypes: true
	};
	var MEMO_STATICS = ***REMOVED***
	  '$$typeof': true,
	  compare: true,
	  defaultProps: true,
	  displayName: true,
	  propTypes: true,
	  type: true
	};
	var TYPE_STATICS = ***REMOVED***};
	TYPE_STATICS[reactIs_7] = FORWARD_REF_STATICS;
	TYPE_STATICS[reactIs_10] = MEMO_STATICS;

	function getStatics(component) ***REMOVED***
	  // React v16.11 and below
	  if (reactIs_24(component)) ***REMOVED***
	    return MEMO_STATICS;
	***REMOVED*** // React v16.12 and above


	  return TYPE_STATICS[component['$$typeof']] || REACT_STATICS;
	}

	var defineProperty = Object.defineProperty;
	var getOwnPropertyNames = Object.getOwnPropertyNames;
	var getOwnPropertySymbols = Object.getOwnPropertySymbols;
	var getOwnPropertyDescriptor = Object.getOwnPropertyDescriptor;
	var getPrototypeOf = Object.getPrototypeOf;
	var objectPrototype = Object.prototype;
	function hoistNonReactStatics(targetComponent, sourceComponent, blacklist) ***REMOVED***
	  if (typeof sourceComponent !== 'string') ***REMOVED***
	    // don't hoist over string (html) components
	    if (objectPrototype) ***REMOVED***
	      var inheritedComponent = getPrototypeOf(sourceComponent);

	      if (inheritedComponent && inheritedComponent !== objectPrototype) ***REMOVED***
	        hoistNonReactStatics(targetComponent, inheritedComponent, blacklist);
	    ***REMOVED***
	  ***REMOVED***

	    var keys = getOwnPropertyNames(sourceComponent);

	    if (getOwnPropertySymbols) ***REMOVED***
	      keys = keys.concat(getOwnPropertySymbols(sourceComponent));
	  ***REMOVED***

	    var targetStatics = getStatics(targetComponent);
	    var sourceStatics = getStatics(sourceComponent);

	    for (var i = 0; i < keys.length; ++i) ***REMOVED***
	      var key = keys[i];

	      if (!KNOWN_STATICS[key] && !(blacklist && blacklist[key]) && !(sourceStatics && sourceStatics[key]) && !(targetStatics && targetStatics[key])) ***REMOVED***
	        var descriptor = getOwnPropertyDescriptor(sourceComponent, key);

	        try ***REMOVED***
	          // Avoid failures from read-only properties
	          defineProperty(targetComponent, key, descriptor);
	      ***REMOVED*** catch (e) ***REMOVED***}
	    ***REMOVED***
	  ***REMOVED***
	***REMOVED***

	  return targetComponent;
	}

	return hoistNonReactStatics;

})));
