// hoist-non-react-statics/dist/hoist-non-react-statics.cjs.js
'use strict';

var reactIs = require('react-is');

/**
 * Copyright 2015, Yahoo! Inc.
 * Copyrights licensed under the New BSD License. See the accompanying LICENSE file for terms.
 */
var REACT_STATICS = ***REMOVED***
  childContextTypes: true,
  contextType: true,
  contextTypes: true,
  defaultProps: true,
  displayName: true,
  getDefaultProps: true,
  getDerivedStateFromError: true,
  getDerivedStateFromProps: true,
  mixins: true,
  propTypes: true,
  type: true
};
var KNOWN_STATICS = ***REMOVED***
  name: true,
  length: true,
  prototype: true,
  caller: true,
  callee: true,
  arguments: true,
  arity: true
};
var FORWARD_REF_STATICS = ***REMOVED***
  '$$typeof': true,
  render: true,
  defaultProps: true,
  displayName: true,
  propTypes: true
};
var MEMO_STATICS = ***REMOVED***
  '$$typeof': true,
  compare: true,
  defaultProps: true,
  displayName: true,
  propTypes: true,
  type: true
};
var TYPE_STATICS = ***REMOVED***};
TYPE_STATICS[reactIs.ForwardRef] = FORWARD_REF_STATICS;
TYPE_STATICS[reactIs.Memo] = MEMO_STATICS;

function getStatics(component) ***REMOVED***
  // React v16.11 and below
  if (reactIs.isMemo(component)) ***REMOVED***
    return MEMO_STATICS;
***REMOVED*** // React v16.12 and above


  return TYPE_STATICS[component['$$typeof']] || REACT_STATICS;
}

var defineProperty = Object.defineProperty;
var getOwnPropertyNames = Object.getOwnPropertyNames;
var getOwnPropertySymbols = Object.getOwnPropertySymbols;
var getOwnPropertyDescriptor = Object.getOwnPropertyDescriptor;
var getPrototypeOf = Object.getPrototypeOf;
var objectPrototype = Object.prototype;
function hoistNonReactStatics(targetComponent, sourceComponent, blacklist) ***REMOVED***
  if (typeof sourceComponent !== 'string') ***REMOVED***
    // don't hoist over string (html) components
    if (objectPrototype) ***REMOVED***
      var inheritedComponent = getPrototypeOf(sourceComponent);

      if (inheritedComponent && inheritedComponent !== objectPrototype) ***REMOVED***
        hoistNonReactStatics(targetComponent, inheritedComponent, blacklist);
    ***REMOVED***
  ***REMOVED***

    var keys = getOwnPropertyNames(sourceComponent);

    if (getOwnPropertySymbols) ***REMOVED***
      keys = keys.concat(getOwnPropertySymbols(sourceComponent));
  ***REMOVED***

    var targetStatics = getStatics(targetComponent);
    var sourceStatics = getStatics(sourceComponent);

    for (var i = 0; i < keys.length; ++i) ***REMOVED***
      var key = keys[i];

      if (!KNOWN_STATICS[key] && !(blacklist && blacklist[key]) && !(sourceStatics && sourceStatics[key]) && !(targetStatics && targetStatics[key])) ***REMOVED***
        var descriptor = getOwnPropertyDescriptor(sourceComponent, key);

        try ***REMOVED***
          // Avoid failures from read-only properties
          defineProperty(targetComponent, key, descriptor);
      ***REMOVED*** catch (e) ***REMOVED***}
    ***REMOVED***
  ***REMOVED***
***REMOVED***

  return targetComponent;
}

module.exports = hoistNonReactStatics;
