// @google-cloud/storage/build/src/file.js
"use strict";
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) ***REMOVED***
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _File_instances, _File_validateIntegrity;
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.File = exports.FileExceptionMessages = exports.STORAGE_POST_POLICY_BASE_URL = exports.ActionToHTTPMethod = void 0;
const nodejs_common_1 = require("./nodejs-common");
const promisify_1 = require("@google-cloud/promisify");
const compressible = require("compressible");
const crypto = require("crypto");
const extend = require("extend");
const fs = require("fs");
const mime = require("mime");
const resumableUpload = require("./resumable-upload");
const stream_1 = require("stream");
const zlib = require("zlib");
const storage_1 = require("./storage");
const bucket_1 = require("./bucket");
const acl_1 = require("./acl");
const signer_1 = require("./signer");
// eslint-disable-next-line @typescript-eslint/no-var-requires
const duplexify = require('duplexify');
const util_1 = require("./util");
const hash_stream_validator_1 = require("./hash-stream-validator");
const retry = require("async-retry");
var ActionToHTTPMethod;
(function (ActionToHTTPMethod) ***REMOVED***
    ActionToHTTPMethod["read"] = "GET";
    ActionToHTTPMethod["write"] = "PUT";
    ActionToHTTPMethod["delete"] = "DELETE";
    ActionToHTTPMethod["resumable"] = "POST";
})(ActionToHTTPMethod = exports.ActionToHTTPMethod || (exports.ActionToHTTPMethod = ***REMOVED***}));
/**
 * @private
 */
exports.STORAGE_POST_POLICY_BASE_URL = 'https://storage.googleapis.com';
/**
 * @private
 */
const GS_URL_REGEXP = /^gs:\/\/([a-z0-9_.-]+)\/(.+)$/;
class RequestError extends Error ***REMOVED***
}
const SEVEN_DAYS = 7 * 24 * 60 * 60;
var FileExceptionMessages;
(function (FileExceptionMessages) ***REMOVED***
    FileExceptionMessages["EXPIRATION_TIME_NA"] = "An expiration time is not available.";
    FileExceptionMessages["DESTINATION_NO_NAME"] = "Destination file should have a name.";
    FileExceptionMessages["INVALID_VALIDATION_FILE_RANGE"] = "Cannot use validation with file ranges (start/end).";
    FileExceptionMessages["MD5_NOT_AVAILABLE"] = "MD5 verification was specified, but is not available for the requested object. MD5 is not available for composite objects.";
    FileExceptionMessages["EQUALS_CONDITION_TWO_ELEMENTS"] = "Equals condition must be an array of 2 elements.";
    FileExceptionMessages["STARTS_WITH_TWO_ELEMENTS"] = "StartsWith condition must be an array of 2 elements.";
    FileExceptionMessages["CONTENT_LENGTH_RANGE_MIN_MAX"] = "ContentLengthRange must have numeric min & max fields.";
    FileExceptionMessages["DOWNLOAD_MISMATCH"] = "The downloaded data did not match the data from the server. To be sure the content is the same, you should download the file again.";
    FileExceptionMessages["UPLOAD_MISMATCH_DELETE_FAIL"] = "The uploaded data did not match the data from the server.\n    As a precaution, we attempted to delete the file, but it was not successful.\n    To be sure the content is the same, you should try removing the file manually,\n    then uploading the file again.\n    \n\nThe delete attempt failed with this message:\n\n  ";
    FileExceptionMessages["UPLOAD_MISMATCH"] = "The uploaded data did not match the data from the server.\n    As a precaution, the file has been deleted.\n    To be sure the content is the same, you should try uploading the file again.";
})(FileExceptionMessages = exports.FileExceptionMessages || (exports.FileExceptionMessages = ***REMOVED***}));
/**
 * A File object is created from your ***REMOVED***@link Bucket} object using
 * ***REMOVED***@link Bucket#file}.
 *
 * @class
 */
class File extends nodejs_common_1.ServiceObject ***REMOVED***
    /**
     * Cloud Storage uses access control lists (ACLs) to manage object and
     * bucket access. ACLs are the mechanism you use to share objects with other
     * users and allow other users to access your buckets and objects.
     *
     * An ACL consists of one or more entries, where each entry grants permissions
     * to an entity. Permissions define the actions that can be performed against
     * an object or bucket (for example, `READ` or `WRITE`); the entity defines
     * who the permission applies to (for example, a specific user or group of
     * users).
     *
     * The `acl` object on a File instance provides methods to get you a list of
     * the ACLs defined on your bucket, as well as set, update, and delete them.
     *
     * See ***REMOVED***@link http://goo.gl/6qBBPO| About Access Control lists}
     *
     * @name File#acl
     * @mixes Acl
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     * //-
     * // Make a file publicly readable.
     * //-
     * const options = ***REMOVED***
     *   entity: 'allUsers',
     *   role: storage.acl.READER_ROLE
     * };
     *
     * file.acl.add(options, function(err, aclObject) ***REMOVED***});
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.acl.add(options).then(function(data) ***REMOVED***
     *   const aclObject = data[0];
     *   const apiResponse = data[1];
     * });
     * ```
     */
    /**
     * The API-formatted resource description of the file.
     *
     * Note: This is not guaranteed to be up-to-date when accessed. To get the
     * latest record, call the `getMetadata()` method.
     *
     * @name File#metadata
     * @type ***REMOVED***object}
     */
    /**
     * The file's name.
     * @name File#name
     * @type ***REMOVED***string}
     */
    /**
     * @callback Crc32cGeneratorToStringCallback
     * A method returning the CRC32C as a base64-encoded string.
     *
     * @returns ***REMOVED***string}
     *
     * @example
     * Hashing the string 'data' should return 'rth90Q=='
     *
     * ```js
     * const buffer = Buffer.from('data');
     * crc32c.update(buffer);
     * crc32c.toString(); // 'rth90Q=='
     * ```
     **/
    /**
     * @callback Crc32cGeneratorValidateCallback
     * A method validating a base64-encoded CRC32C string.
     *
     * @param ***REMOVED***string} [value] base64-encoded CRC32C string to validate
     * @returns ***REMOVED***boolean}
     *
     * @example
     * Should return `true` if the value matches, `false` otherwise
     *
     * ```js
     * const buffer = Buffer.from('data');
     * crc32c.update(buffer);
     * crc32c.validate('DkjKuA=='); // false
     * crc32c.validate('rth90Q=='); // true
     * ```
     **/
    /**
     * @callback Crc32cGeneratorUpdateCallback
     * A method for passing `Buffer`s for CRC32C generation.
     *
     * @param ***REMOVED***Buffer} [data] data to update CRC32C value with
     * @returns ***REMOVED***undefined}
     *
     * @example
     * Hashing buffers from 'some ' and 'text\n'
     *
     * ```js
     * const buffer1 = Buffer.from('some ');
     * crc32c.update(buffer1);
     *
     * const buffer2 = Buffer.from('text\n');
     * crc32c.update(buffer2);
     *
     * crc32c.toString(); // 'DkjKuA=='
     * ```
     **/
    /**
     * @typedef ***REMOVED***object} CRC32CValidator
     * @property ***REMOVED***Crc32cGeneratorToStringCallback}
     * @property ***REMOVED***Crc32cGeneratorValidateCallback}
     * @property ***REMOVED***Crc32cGeneratorUpdateCallback}
     */
    /**
     * @callback Crc32cGeneratorCallback
     * @returns ***REMOVED***CRC32CValidator}
     */
    /**
     * @typedef ***REMOVED***object} FileOptions Options passed to the File constructor.
     * @property ***REMOVED***string} [encryptionKey] A custom encryption key.
     * @property ***REMOVED***number} [generation] Generation to scope the file to.
     * @property ***REMOVED***string} [kmsKeyName] Cloud KMS Key used to encrypt this
     *     object, if the object is encrypted by such a key. Limited availability;
     *     usable only by enabled projects.
     * @property ***REMOVED***string} [userProject] The ID of the project which will be
     *     billed for all requests made from File object.
     * @property ***REMOVED***Crc32cGeneratorCallback} [callback] A function that generates a CRC32C Validator. Defaults to ***REMOVED***@link CRC32C}
     */
    /**
     * Constructs a file object.
     *
     * @param ***REMOVED***Bucket} bucket The Bucket instance this file is
     *     attached to.
     * @param ***REMOVED***string} name The name of the remote file.
     * @param ***REMOVED***FileOptions} [options] Configuration options.
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     * ```
     */
    constructor(bucket, name, options = ***REMOVED***}) ***REMOVED***
        var _a, _b;
        const requestQueryObject = ***REMOVED***};
        let generation;
        if (options.generation !== null) ***REMOVED***
            if (typeof options.generation === 'string') ***REMOVED***
                generation = Number(options.generation);
          ***REMOVED***
            else ***REMOVED***
                generation = options.generation;
          ***REMOVED***
            if (!isNaN(generation)) ***REMOVED***
                requestQueryObject.generation = generation;
          ***REMOVED***
      ***REMOVED***
        Object.assign(requestQueryObject, options.preconditionOpts);
        const userProject = options.userProject || bucket.userProject;
        if (typeof userProject === 'string') ***REMOVED***
            requestQueryObject.userProject = userProject;
      ***REMOVED***
        const methods = ***REMOVED***
            /**
             * @typedef ***REMOVED***array} DeleteFileResponse
             * @property ***REMOVED***object} 0 The full API response.
             */
            /**
             * @callback DeleteFileCallback
             * @param ***REMOVED***?Error} err Request error, if any.
             * @param ***REMOVED***object} apiResponse The full API response.
             */
            /**
             * Delete the file.
             *
             * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/delete| Objects: delete API Documentation}
             *
             * @method File#delete
             * @param ***REMOVED***object} [options] Configuration options.
             * @param ***REMOVED***boolean} [options.ignoreNotFound = false] Ignore an error if
             *     the file does not exist.
             * @param ***REMOVED***string} [options.userProject] The ID of the project which will be
             *     billed for the request.
             * @param ***REMOVED***DeleteFileCallback} [callback] Callback function.
             * @returns ***REMOVED***Promise<DeleteFileResponse>}
             *
             * @example
             * ```
             * const ***REMOVED***Storage} = require('@google-cloud/storage');
             * const storage = new Storage();
             * const myBucket = storage.bucket('my-bucket');
             *
             * const file = myBucket.file('my-file');
             * file.delete(function(err, apiResponse) ***REMOVED***});
             *
             * //-
             * // If the callback is omitted, we'll return a Promise.
             * //-
             * file.delete().then(function(data) ***REMOVED***
             *   const apiResponse = data[0];
             * });
             *
             * ```
             * @example <caption>include:samples/files.js</caption>
             * region_tag:storage_delete_file
             * Another example:
             */
            delete: ***REMOVED***
                reqOpts: ***REMOVED***
                    qs: requestQueryObject,
              ***REMOVED***,
          ***REMOVED***,
            /**
             * @typedef ***REMOVED***array} FileExistsResponse
             * @property ***REMOVED***boolean} 0 Whether the ***REMOVED***@link File} exists.
             */
            /**
             * @callback FileExistsCallback
             * @param ***REMOVED***?Error} err Request error, if any.
             * @param ***REMOVED***boolean} exists Whether the ***REMOVED***@link File} exists.
             */
            /**
             * Check if the file exists.
             *
             * @method File#exists
             * @param ***REMOVED***options} [options] Configuration options.
             * @param ***REMOVED***string} [options.userProject] The ID of the project which will be
             *     billed for the request.
             * @param ***REMOVED***FileExistsCallback} [callback] Callback function.
             * @returns ***REMOVED***Promise<FileExistsResponse>}
             *
             * @example
             * ```
             * const ***REMOVED***Storage} = require('@google-cloud/storage');
             * const storage = new Storage();
             * const myBucket = storage.bucket('my-bucket');
             *
             * const file = myBucket.file('my-file');
             *
             * file.exists(function(err, exists) ***REMOVED***});
             *
             * //-
             * // If the callback is omitted, we'll return a Promise.
             * //-
             * file.exists().then(function(data) ***REMOVED***
             *   const exists = data[0];
             * });
             * ```
             */
            exists: ***REMOVED***
                reqOpts: ***REMOVED***
                    qs: requestQueryObject,
              ***REMOVED***,
          ***REMOVED***,
            /**
             * @typedef ***REMOVED***array} GetFileResponse
             * @property ***REMOVED***File} 0 The ***REMOVED***@link File}.
             * @property ***REMOVED***object} 1 The full API response.
             */
            /**
             * @callback GetFileCallback
             * @param ***REMOVED***?Error} err Request error, if any.
             * @param ***REMOVED***File} file The ***REMOVED***@link File}.
             * @param ***REMOVED***object} apiResponse The full API response.
             */
            /**
             * Get a file object and its metadata if it exists.
             *
             * @method File#get
             * @param ***REMOVED***options} [options] Configuration options.
             * @param ***REMOVED***string} [options.userProject] The ID of the project which will be
             *     billed for the request.
             * @param ***REMOVED***GetFileCallback} [callback] Callback function.
             * @returns ***REMOVED***Promise<GetFileResponse>}
             *
             * @example
             * ```
             * const ***REMOVED***Storage} = require('@google-cloud/storage');
             * const storage = new Storage();
             * const myBucket = storage.bucket('my-bucket');
             *
             * const file = myBucket.file('my-file');
             *
             * file.get(function(err, file, apiResponse) ***REMOVED***
             *   // file.metadata` has been populated.
             * });
             *
             * //-
             * // If the callback is omitted, we'll return a Promise.
             * //-
             * file.get().then(function(data) ***REMOVED***
             *   const file = data[0];
             *   const apiResponse = data[1];
             * });
             * ```
             */
            get: ***REMOVED***
                reqOpts: ***REMOVED***
                    qs: requestQueryObject,
              ***REMOVED***,
          ***REMOVED***,
            /**
             * @typedef ***REMOVED***array} GetFileMetadataResponse
             * @property ***REMOVED***object} 0 The ***REMOVED***@link File} metadata.
             * @property ***REMOVED***object} 1 The full API response.
             */
            /**
             * @callback GetFileMetadataCallback
             * @param ***REMOVED***?Error} err Request error, if any.
             * @param ***REMOVED***object} metadata The ***REMOVED***@link File} metadata.
             * @param ***REMOVED***object} apiResponse The full API response.
             */
            /**
             * Get the file's metadata.
             *
             * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/get| Objects: get API Documentation}
             *
             * @method File#getMetadata
             * @param ***REMOVED***object} [options] Configuration options.
             * @param ***REMOVED***string} [options.userProject] The ID of the project which will be
             *     billed for the request.
             * @param ***REMOVED***GetFileMetadataCallback} [callback] Callback function.
             * @returns ***REMOVED***Promise<GetFileMetadataResponse>}
             *
             * @example
             * ```
             * const ***REMOVED***Storage} = require('@google-cloud/storage');
             * const storage = new Storage();
             * const myBucket = storage.bucket('my-bucket');
             *
             * const file = myBucket.file('my-file');
             *
             * file.getMetadata(function(err, metadata, apiResponse) ***REMOVED***});
             *
             * //-
             * // If the callback is omitted, we'll return a Promise.
             * //-
             * file.getMetadata().then(function(data) ***REMOVED***
             *   const metadata = data[0];
             *   const apiResponse = data[1];
             * });
             *
             * ```
             * @example <caption>include:samples/files.js</caption>
             * region_tag:storage_get_metadata
             * Another example:
             */
            getMetadata: ***REMOVED***
                reqOpts: ***REMOVED***
                    qs: requestQueryObject,
              ***REMOVED***,
          ***REMOVED***,
            /**
             * @typedef ***REMOVED***object} SetFileMetadataOptions Configuration options for File#setMetadata().
             * @param ***REMOVED***string} [userProject] The ID of the project which will be billed for the request.
             */
            /**
             * @callback SetFileMetadataCallback
             * @param ***REMOVED***?Error} err Request error, if any.
             * @param ***REMOVED***object} apiResponse The full API response.
             */
            /**
             * @typedef ***REMOVED***array} SetFileMetadataResponse
             * @property ***REMOVED***object} 0 The full API response.
             */
            /**
             * Merge the given metadata with the current remote file's metadata. This
             * will set metadata if it was previously unset or update previously set
             * metadata. To unset previously set metadata, set its value to null.
             *
             * You can set custom key/value pairs in the metadata key of the given
             * object, however the other properties outside of this object must adhere
             * to the ***REMOVED***@link https://goo.gl/BOnnCK| official API documentation}.
             *
             *
             * See the examples below for more information.
             *
             * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/patch| Objects: patch API Documentation}
             *
             * @method File#setMetadata
             * @param ***REMOVED***object} [metadata] The metadata you wish to update.
             * @param ***REMOVED***SetFileMetadataOptions} [options] Configuration options.
             * @param ***REMOVED***SetFileMetadataCallback} [callback] Callback function.
             * @returns ***REMOVED***Promise<SetFileMetadataResponse>}
             *
             * @example
             * ```
             * const ***REMOVED***Storage} = require('@google-cloud/storage');
             * const storage = new Storage();
             * const myBucket = storage.bucket('my-bucket');
             *
             * const file = myBucket.file('my-file');
             *
             * const metadata = ***REMOVED***
             *   contentType: 'application/x-font-ttf',
             *   metadata: ***REMOVED***
             *     my: 'custom',
             *     properties: 'go here'
             * ***REMOVED***
             * };
             *
             * file.setMetadata(metadata, function(err, apiResponse) ***REMOVED***});
             *
             * // Assuming current metadata = ***REMOVED*** hello: 'world', unsetMe: 'will do' }
             * file.setMetadata(***REMOVED***
             *   metadata: ***REMOVED***
             *     abc: '123', // will be set.
             *     unsetMe: null, // will be unset (deleted).
             *     hello: 'goodbye' // will be updated from 'world' to 'goodbye'.
             * ***REMOVED***
             * }, function(err, apiResponse) ***REMOVED***
             *   // metadata should now be ***REMOVED*** abc: '123', hello: 'goodbye' }
             * });
             *
             * //-
             * // Set a temporary hold on this file from its bucket's retention period
             * // configuration.
             * //
             * file.setMetadata(***REMOVED***
             *   temporaryHold: true
             * }, function(err, apiResponse) ***REMOVED***});
             *
             * //-
             * // Alternatively, you may set a temporary hold. This will follow the
             * // same behavior as an event-based hold, with the exception that the
             * // bucket's retention policy will not renew for this file from the time
             * // the hold is released.
             * //-
             * file.setMetadata(***REMOVED***
             *   eventBasedHold: true
             * }, function(err, apiResponse) ***REMOVED***});
             *
             * //-
             * // If the callback is omitted, we'll return a Promise.
             * //-
             * file.setMetadata(metadata).then(function(data) ***REMOVED***
             *   const apiResponse = data[0];
             * });
             * ```
             */
            setMetadata: ***REMOVED***
                reqOpts: ***REMOVED***
                    qs: requestQueryObject,
              ***REMOVED***,
          ***REMOVED***,
      ***REMOVED***;
        super(***REMOVED***
            parent: bucket,
            baseUrl: '/o',
            id: encodeURIComponent(name),
            methods,
      ***REMOVED***);
        _File_instances.add(this);
        this.bucket = bucket;
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        this.storage = bucket.parent;
        // @TODO Can this duplicate code from above be avoided?
        if (options.generation !== null) ***REMOVED***
            let generation;
            if (typeof options.generation === 'string') ***REMOVED***
                generation = Number(options.generation);
          ***REMOVED***
            else ***REMOVED***
                generation = options.generation;
          ***REMOVED***
            if (!isNaN(generation)) ***REMOVED***
                this.generation = generation;
          ***REMOVED***
      ***REMOVED***
        this.kmsKeyName = options.kmsKeyName;
        this.userProject = userProject;
        this.name = name;
        if (options.encryptionKey) ***REMOVED***
            this.setEncryptionKey(options.encryptionKey);
      ***REMOVED***
        this.acl = new acl_1.Acl(***REMOVED***
            request: this.request.bind(this),
            pathPrefix: '/acl',
      ***REMOVED***);
        this.crc32cGenerator =
            options.crc32cGenerator || this.bucket.crc32cGenerator;
        this.instanceRetryValue = (_b = (_a = this.storage) === null || _a === void 0 ? void 0 : _a.retryOptions) === null || _b === void 0 ? void 0 : _b.autoRetry;
        this.instancePreconditionOpts = options === null || options === void 0 ? void 0 : options.preconditionOpts;
  ***REMOVED***
    /**
     * The object's Cloud Storage URI (`gs://`)
     *
     * @example
     * ```ts
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const bucket = storage.bucket('my-bucket');
     * const file = bucket.file('image.png');
     *
     * // `gs://my-bucket/image.png`
     * const href = file.cloudStorageURI.href;
     * ```
     */
    get cloudStorageURI() ***REMOVED***
        const uri = this.bucket.cloudStorageURI;
        uri.pathname = this.name;
        return uri;
  ***REMOVED***
    /**
     * A helper method for determining if a request should be retried based on preconditions.
     * This should only be used for methods where the idempotency is determined by
     * `ifGenerationMatch`
     * @private
     *
     * A request should not be retried under the following conditions:
     * - if precondition option `ifGenerationMatch` is not set OR
     * - if `idempotencyStrategy` is set to `RetryNever`
     */
    shouldRetryBasedOnPreconditionAndIdempotencyStrat(options) ***REMOVED***
        var _a;
        return !(((options === null || options === void 0 ? void 0 : options.ifGenerationMatch) === undefined &&
            ((_a = this.instancePreconditionOpts) === null || _a === void 0 ? void 0 : _a.ifGenerationMatch) === undefined &&
            this.storage.retryOptions.idempotencyStrategy ===
                storage_1.IdempotencyStrategy.RetryConditional) ||
            this.storage.retryOptions.idempotencyStrategy ===
                storage_1.IdempotencyStrategy.RetryNever);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} CopyResponse
     * @property ***REMOVED***File} 0 The copied ***REMOVED***@link File}.
     * @property ***REMOVED***object} 1 The full API response.
     */
    /**
     * @callback CopyCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***File} copiedFile The copied ***REMOVED***@link File}.
     * @param ***REMOVED***object} apiResponse The full API response.
     */
    /**
     * @typedef ***REMOVED***object} CopyOptions Configuration options for File#copy(). See an
     *     ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects#resource| Object resource}.
     * @property ***REMOVED***string} [cacheControl] The cacheControl setting for the new file.
     * @property ***REMOVED***string} [contentEncoding] The contentEncoding setting for the new file.
     * @property ***REMOVED***string} [contentType] The contentType setting for the new file.
     * @property ***REMOVED***string} [destinationKmsKeyName] Resource name of the Cloud
     *     KMS key, of the form
     *     `projects/my-project/locations/location/keyRings/my-kr/cryptoKeys/my-key`,
     *     that will be used to encrypt the object. Overwrites the object
     * metadata's `kms_key_name` value, if any.
     * @property ***REMOVED***Metadata} [metadata] Metadata to specify on the copied file.
     * @property ***REMOVED***string} [predefinedAcl] Set the ACL for the new file.
     * @property ***REMOVED***string} [token] A previously-returned `rewriteToken` from an
     *     unfinished rewrite request.
     * @property ***REMOVED***string} [userProject] The ID of the project which will be
     *     billed for the request.
     */
    /**
     * Copy this file to another file. By default, this will copy the file to the
     * same bucket, but you can choose to copy it to another Bucket by providing
     * a Bucket or File object or a URL starting with "gs://".
     * The generation of the file will not be preserved.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite| Objects: rewrite API Documentation}
     *
     * @throws ***REMOVED***Error} If the destination file is not provided.
     *
     * @param ***REMOVED***string|Bucket|File} destination Destination file.
     * @param ***REMOVED***CopyOptions} [options] Configuration options. See an
     * @param ***REMOVED***CopyCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<CopyResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     *
     * //-
     * // You can pass in a variety of types for the destination.
     * //
     * // For all of the below examples, assume we are working with the following
     * // Bucket and File objects.
     * //-
     * const bucket = storage.bucket('my-bucket');
     * const file = bucket.file('my-image.png');
     *
     * //-
     * // If you pass in a string for the destination, the file is copied to its
     * // current bucket, under the new name provided.
     * //-
     * file.copy('my-image-copy.png', function(err, copiedFile, apiResponse) ***REMOVED***
     *   // `my-bucket` now contains:
     *   // - "my-image.png"
     *   // - "my-image-copy.png"
     *
     *   // `copiedFile` is an instance of a File object that refers to your new
     *   // file.
     * });
     *
     * //-
     * // If you pass in a string starting with "gs://" for the destination, the
     * // file is copied to the other bucket and under the new name provided.
     * //-
     * const newLocation = 'gs://another-bucket/my-image-copy.png';
     * file.copy(newLocation, function(err, copiedFile, apiResponse) ***REMOVED***
     *   // `my-bucket` still contains:
     *   // - "my-image.png"
     *   //
     *   // `another-bucket` now contains:
     *   // - "my-image-copy.png"
     *
     *   // `copiedFile` is an instance of a File object that refers to your new
     *   // file.
     * });
     *
     * //-
     * // If you pass in a Bucket object, the file will be copied to that bucket
     * // using the same name.
     * //-
     * const anotherBucket = storage.bucket('another-bucket');
     * file.copy(anotherBucket, function(err, copiedFile, apiResponse) ***REMOVED***
     *   // `my-bucket` still contains:
     *   // - "my-image.png"
     *   //
     *   // `another-bucket` now contains:
     *   // - "my-image.png"
     *
     *   // `copiedFile` is an instance of a File object that refers to your new
     *   // file.
     * });
     *
     * //-
     * // If you pass in a File object, you have complete control over the new
     * // bucket and filename.
     * //-
     * const anotherFile = anotherBucket.file('my-awesome-image.png');
     * file.copy(anotherFile, function(err, copiedFile, apiResponse) ***REMOVED***
     *   // `my-bucket` still contains:
     *   // - "my-image.png"
     *   //
     *   // `another-bucket` now contains:
     *   // - "my-awesome-image.png"
     *
     *   // Note:
     *   // The `copiedFile` parameter is equal to `anotherFile`.
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.copy(newLocation).then(function(data) ***REMOVED***
     *   const newFile = data[0];
     *   const apiResponse = data[1];
     * });
     *
     * ```
     * @example <caption>include:samples/files.js</caption>
     * region_tag:storage_copy_file
     * Another example:
     */
    copy(destination, optionsOrCallback, callback) ***REMOVED***
        var _a, _b;
        const noDestinationError = new Error(FileExceptionMessages.DESTINATION_NO_NAME);
        if (!destination) ***REMOVED***
            throw noDestinationError;
      ***REMOVED***
        let options = ***REMOVED***};
        if (typeof optionsOrCallback === 'function') ***REMOVED***
            callback = optionsOrCallback;
      ***REMOVED***
        else if (optionsOrCallback) ***REMOVED***
            options = optionsOrCallback;
      ***REMOVED***
        options = extend(true, ***REMOVED***}, options);
        callback = callback || nodejs_common_1.util.noop;
        let destBucket;
        let destName;
        let newFile;
        if (typeof destination === 'string') ***REMOVED***
            const parsedDestination = GS_URL_REGEXP.exec(destination);
            if (parsedDestination !== null && parsedDestination.length === 3) ***REMOVED***
                destBucket = this.storage.bucket(parsedDestination[1]);
                destName = parsedDestination[2];
          ***REMOVED***
            else ***REMOVED***
                destBucket = this.bucket;
                destName = destination;
          ***REMOVED***
      ***REMOVED***
        else if (destination instanceof bucket_1.Bucket) ***REMOVED***
            destBucket = destination;
            destName = this.name;
      ***REMOVED***
        else if (destination instanceof File) ***REMOVED***
            destBucket = destination.bucket;
            destName = destination.name;
            newFile = destination;
      ***REMOVED***
        else ***REMOVED***
            throw noDestinationError;
      ***REMOVED***
        const query = ***REMOVED***};
        if (this.generation !== undefined) ***REMOVED***
            query.sourceGeneration = this.generation;
      ***REMOVED***
        if (options.token !== undefined) ***REMOVED***
            query.rewriteToken = options.token;
      ***REMOVED***
        if (options.userProject !== undefined) ***REMOVED***
            query.userProject = options.userProject;
            delete options.userProject;
      ***REMOVED***
        if (options.predefinedAcl !== undefined) ***REMOVED***
            query.destinationPredefinedAcl = options.predefinedAcl;
            delete options.predefinedAcl;
      ***REMOVED***
        newFile = newFile || destBucket.file(destName);
        const headers = ***REMOVED***};
        if (this.encryptionKey !== undefined) ***REMOVED***
            headers['x-goog-copy-source-encryption-algorithm'] = 'AES256';
            headers['x-goog-copy-source-encryption-key'] = this.encryptionKeyBase64;
            headers['x-goog-copy-source-encryption-key-sha256'] =
                this.encryptionKeyHash;
      ***REMOVED***
        if (newFile.encryptionKey !== undefined) ***REMOVED***
            this.setEncryptionKey(newFile.encryptionKey);
      ***REMOVED***
        else if (options.destinationKmsKeyName !== undefined) ***REMOVED***
            query.destinationKmsKeyName = options.destinationKmsKeyName;
            delete options.destinationKmsKeyName;
      ***REMOVED***
        else if (newFile.kmsKeyName !== undefined) ***REMOVED***
            query.destinationKmsKeyName = newFile.kmsKeyName;
      ***REMOVED***
        if (query.destinationKmsKeyName) ***REMOVED***
            this.kmsKeyName = query.destinationKmsKeyName;
            const keyIndex = this.interceptors.indexOf(this.encryptionKeyInterceptor);
            if (keyIndex > -1) ***REMOVED***
                this.interceptors.splice(keyIndex, 1);
          ***REMOVED***
      ***REMOVED***
        if (!this.shouldRetryBasedOnPreconditionAndIdempotencyStrat(options === null || options === void 0 ? void 0 : options.preconditionOpts)) ***REMOVED***
            this.storage.retryOptions.autoRetry = false;
      ***REMOVED***
        if (((_a = options.preconditionOpts) === null || _a === void 0 ? void 0 : _a.ifGenerationMatch) !== undefined) ***REMOVED***
            query.ifGenerationMatch = (_b = options.preconditionOpts) === null || _b === void 0 ? void 0 : _b.ifGenerationMatch;
            delete options.preconditionOpts;
      ***REMOVED***
        this.request(***REMOVED***
            method: 'POST',
            uri: `/rewriteTo/b/$***REMOVED***destBucket.name}/o/$***REMOVED***encodeURIComponent(newFile.name)}`,
            qs: query,
            json: options,
            headers,
      ***REMOVED***, (err, resp) => ***REMOVED***
            this.storage.retryOptions.autoRetry = this.instanceRetryValue;
            if (err) ***REMOVED***
                callback(err, null, resp);
                return;
          ***REMOVED***
            if (resp.rewriteToken) ***REMOVED***
                const options = ***REMOVED***
                    token: resp.rewriteToken,
              ***REMOVED***;
                if (query.userProject) ***REMOVED***
                    options.userProject = query.userProject;
              ***REMOVED***
                if (query.destinationKmsKeyName) ***REMOVED***
                    options.destinationKmsKeyName = query.destinationKmsKeyName;
              ***REMOVED***
                this.copy(newFile, options, callback);
                return;
          ***REMOVED***
            callback(null, newFile, resp);
      ***REMOVED***);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***object} CreateReadStreamOptions Configuration options for File#createReadStream.
     * @property ***REMOVED***string} [userProject] The ID of the project which will be
     *     billed for the request.
     * @property ***REMOVED***string|boolean} [validation] Possible values: `"md5"`,
     *     `"crc32c"`, or `false`. By default, data integrity is validated with a
     *     CRC32c checksum. You may use MD5 if preferred, but that hash is not
     *     supported for composite objects. An error will be raised if MD5 is
     *     specified but is not available. You may also choose to skip validation
     *     completely, however this is **not recommended**.
     * @property ***REMOVED***number} [start] A byte offset to begin the file's download
     *     from. Default is 0. NOTE: Byte ranges are inclusive; that is,
     *     `options.start = 0` and `options.end = 999` represent the first 1000
     *     bytes in a file or object. NOTE: when specifying a byte range, data
     *     integrity is not available.
     * @property ***REMOVED***number} [end] A byte offset to stop reading the file at.
     *     NOTE: Byte ranges are inclusive; that is, `options.start = 0` and
     *     `options.end = 999` represent the first 1000 bytes in a file or object.
     *     NOTE: when specifying a byte range, data integrity is not available.
     * @property ***REMOVED***boolean} [decompress=true] Disable auto decompression of the
     *     received data. By default this option is set to `true`.
     *     Applicable in cases where the data was uploaded with
     *     `gzip: true` option. See ***REMOVED***@link File#createWriteStream}.
     */
    /**
     * Create a readable stream to read the contents of the remote file. It can be
     * piped to a writable stream or listened to for 'data' events to read a
     * file's contents.
     *
     * In the unlikely event there is a mismatch between what you downloaded and
     * the version in your Bucket, your error handler will receive an error with
     * code "CONTENT_DOWNLOAD_MISMATCH". If you receive this error, the best
     * recourse is to try downloading the file again.
     *
     * NOTE: Readable streams will emit the `end` event when the file is fully
     * downloaded.
     *
     * @param ***REMOVED***CreateReadStreamOptions} [options] Configuration options.
     * @returns ***REMOVED***ReadableStream}
     *
     * @example
     * ```
     * //-
     * // <h4>Downloading a File</h4>
     * //
     * // The example below demonstrates how we can reference a remote file, then
     * // pipe its contents to a local file. This is effectively creating a local
     * // backup of your remote data.
     * //-
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const bucket = storage.bucket('my-bucket');
     *
     * const fs = require('fs');
     * const remoteFile = bucket.file('image.png');
     * const localFilename = '/Users/stephen/Photos/image.png';
     *
     * remoteFile.createReadStream()
     *   .on('error', function(err) ***REMOVED***})
     *   .on('response', function(response) ***REMOVED***
     *     // Server connected and responded with the specified status and headers.
     *  ***REMOVED***)
     *   .on('end', function() ***REMOVED***
     *     // The file is fully downloaded.
     * ***REMOVED***)
     *   .pipe(fs.createWriteStream(localFilename));
     *
     * //-
     * // To limit the downloaded data to only a byte range, pass an options
     * // object.
     * //-
     * const logFile = myBucket.file('access_log');
     * logFile.createReadStream(***REMOVED***
     *     start: 10000,
     *     end: 20000
     * ***REMOVED***)
     *   .on('error', function(err) ***REMOVED***})
     *   .pipe(fs.createWriteStream('/Users/stephen/logfile.txt'));
     *
     * //-
     * // To read a tail byte range, specify only `options.end` as a negative
     * // number.
     * //-
     * const logFile = myBucket.file('access_log');
     * logFile.createReadStream(***REMOVED***
     *     end: -100
     * ***REMOVED***)
     *   .on('error', function(err) ***REMOVED***})
     *   .pipe(fs.createWriteStream('/Users/stephen/logfile.txt'));
     * ```
     */
    createReadStream(options = ***REMOVED***}) ***REMOVED***
        options = Object.assign(***REMOVED*** decompress: true }, options);
        const rangeRequest = typeof options.start === 'number' || typeof options.end === 'number';
        const tailRequest = options.end < 0;
        let validateStream = undefined;
        const throughStream = new util_1.PassThroughShim();
        let isCompressed = true;
        let crc32c = true;
        let md5 = false;
        let safeToValidate = true;
        if (typeof options.validation === 'string') ***REMOVED***
            const value = options.validation.toLowerCase().trim();
            crc32c = value === 'crc32c';
            md5 = value === 'md5';
      ***REMOVED***
        else if (options.validation === false) ***REMOVED***
            crc32c = false;
      ***REMOVED***
        const shouldRunValidation = !rangeRequest && (crc32c || md5);
        if (rangeRequest) ***REMOVED***
            if (typeof options.validation === 'string' ||
                options.validation === true) ***REMOVED***
                throw new Error(FileExceptionMessages.INVALID_VALIDATION_FILE_RANGE);
          ***REMOVED***
            // Range requests can't receive data integrity checks.
            crc32c = false;
            md5 = false;
      ***REMOVED***
        // Authenticate the request, then pipe the remote API request to the stream
        // returned to the user.
        const makeRequest = () => ***REMOVED***
            const query = ***REMOVED*** alt: 'media' };
            if (this.generation) ***REMOVED***
                query.generation = this.generation;
          ***REMOVED***
            if (options.userProject) ***REMOVED***
                query.userProject = options.userProject;
          ***REMOVED***
            const headers = ***REMOVED***
                'Accept-Encoding': 'gzip',
                'Cache-Control': 'no-store',
          ***REMOVED***;
            if (rangeRequest) ***REMOVED***
                const start = typeof options.start === 'number' ? options.start : '0';
                const end = typeof options.end === 'number' ? options.end : '';
                headers.Range = `bytes=$***REMOVED***tailRequest ? end : `$***REMOVED***start}-$***REMOVED***end}`}`;
          ***REMOVED***
            const reqOpts = ***REMOVED***
                forever: false,
                uri: '',
                headers,
                qs: query,
          ***REMOVED***;
            const hashes = ***REMOVED***};
            this.requestStream(reqOpts)
                .on('error', err => ***REMOVED***
                throughStream.destroy(err);
          ***REMOVED***)
                .on('response', res => ***REMOVED***
                throughStream.emit('response', res);
                nodejs_common_1.util.handleResp(null, res, null, onResponse);
          ***REMOVED***)
                .resume();
            // We listen to the response event from the request stream so that we
            // can...
            //
            //   1) Intercept any data from going to the user if an error occurred.
            //   2) Calculate the hashes from the http.IncomingMessage response
            //   stream,
            //      which will return the bytes from the source without decompressing
            //      gzip'd content. We then send it through decompressed, if
            //      applicable, to the user.
            const onResponse = (err, _body, rawResponseStream) => ***REMOVED***
                if (err) ***REMOVED***
                    // Get error message from the body.
                    this.getBufferFromReadable(rawResponseStream).then(body => ***REMOVED***
                        err.message = body.toString('utf8');
                        throughStream.destroy(err);
                  ***REMOVED***);
                    return;
              ***REMOVED***
                rawResponseStream.on('error', onComplete);
                const headers = rawResponseStream.toJSON().headers;
                isCompressed = headers['content-encoding'] === 'gzip';
                // The object is safe to validate if:
                // 1. It was stored gzip and returned to us gzip OR
                // 2. It was never stored as gzip
                safeToValidate =
                    (headers['x-goog-stored-content-encoding'] === 'gzip' &&
                        isCompressed) ||
                        headers['x-goog-stored-content-encoding'] === 'identity';
                const transformStreams = [];
                if (shouldRunValidation) ***REMOVED***
                    // The x-goog-hash header should be set with a crc32c and md5 hash.
                    // ex: headers['x-goog-hash'] = 'crc32c=xxxx,md5=xxxx'
                    if (typeof headers['x-goog-hash'] === 'string') ***REMOVED***
                        headers['x-goog-hash']
                            .split(',')
                            .forEach((hashKeyValPair) => ***REMOVED***
                            const delimiterIndex = hashKeyValPair.indexOf('=');
                            const hashType = hashKeyValPair.substr(0, delimiterIndex);
                            const hashValue = hashKeyValPair.substr(delimiterIndex + 1);
                            hashes[hashType] = hashValue;
                      ***REMOVED***);
                  ***REMOVED***
                    validateStream = new hash_stream_validator_1.HashStreamValidator(***REMOVED***
                        crc32c,
                        md5,
                        crc32cGenerator: this.crc32cGenerator,
                  ***REMOVED***);
                    transformStreams.push(validateStream);
              ***REMOVED***
                if (isCompressed && options.decompress) ***REMOVED***
                    transformStreams.push(zlib.createGunzip());
              ***REMOVED***
                const handoffStream = new stream_1.PassThrough(***REMOVED***
                    final: cb => ***REMOVED***
                        // Preserving `onComplete`'s ability to
                        // destroy `throughStream` before pipeline
                        // attempts to.
                        onComplete(null)
                            .then(() => ***REMOVED***
                            cb();
                      ***REMOVED***)
                            .catch(cb);
                  ***REMOVED***,
              ***REMOVED***);
                (0, stream_1.pipeline)(rawResponseStream, ...transformStreams, handoffStream, throughStream, onComplete);
          ***REMOVED***;
            // This is hooked to the `complete` event from the request stream. This is
            // our chance to validate the data and let the user know if anything went
            // wrong.
            let onCompleteCalled = false;
            const onComplete = async (err) => ***REMOVED***
                if (onCompleteCalled) ***REMOVED***
                    return;
              ***REMOVED***
                onCompleteCalled = true;
                if (err) ***REMOVED***
                    throughStream.destroy(err);
                    return;
              ***REMOVED***
                if (rangeRequest || !shouldRunValidation) ***REMOVED***
                    return;
              ***REMOVED***
                // If we're doing validation, assume the worst-- a data integrity
                // mismatch. If not, these tests won't be performed, and we can assume
                // the best.
                // We must check if the server decompressed the data on serve because hash
                // validation is not possible in this case.
                let failed = (crc32c || md5) && safeToValidate;
                if (validateStream && safeToValidate) ***REMOVED***
                    if (crc32c && hashes.crc32c) ***REMOVED***
                        failed = !validateStream.test('crc32c', hashes.crc32c);
                  ***REMOVED***
                    if (md5 && hashes.md5) ***REMOVED***
                        failed = !validateStream.test('md5', hashes.md5);
                  ***REMOVED***
              ***REMOVED***
                if (md5 && !hashes.md5) ***REMOVED***
                    const hashError = new RequestError(FileExceptionMessages.MD5_NOT_AVAILABLE);
                    hashError.code = 'MD5_NOT_AVAILABLE';
                    throughStream.destroy(hashError);
              ***REMOVED***
                else if (failed) ***REMOVED***
                    const mismatchError = new RequestError(FileExceptionMessages.DOWNLOAD_MISMATCH);
                    mismatchError.code = 'CONTENT_DOWNLOAD_MISMATCH';
                    throughStream.destroy(mismatchError);
              ***REMOVED***
                else ***REMOVED***
                    return;
              ***REMOVED***
          ***REMOVED***;
      ***REMOVED***;
        throughStream.on('reading', makeRequest);
        return throughStream;
  ***REMOVED***
    /**
     * @callback CreateResumableUploadCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***string} uri The resumable upload's unique session URI.
     */
    /**
     * @typedef ***REMOVED***array} CreateResumableUploadResponse
     * @property ***REMOVED***string} 0 The resumable upload's unique session URI.
     */
    /**
     * @typedef ***REMOVED***object} CreateResumableUploadOptions
     * @property ***REMOVED***object} [metadata] Metadata to set on the file.
     * @property ***REMOVED***number} [offset] The starting byte of the upload stream for resuming an interrupted upload.
     * @property ***REMOVED***string} [origin] Origin header to set for the upload.
     * @property ***REMOVED***string} [predefinedAcl] Apply a predefined set of access
     * controls to this object.
     *
     * Acceptable values are:
     * - **`authenticatedRead`** - Object owner gets `OWNER` access, and
     *   `allAuthenticatedUsers` get `READER` access.
     *
     * - **`bucketOwnerFullControl`** - Object owner gets `OWNER` access, and
     *   project team owners get `OWNER` access.
     *
     * - **`bucketOwnerRead`** - Object owner gets `OWNER` access, and project
     *   team owners get `READER` access.
     *
     * - **`private`** - Object owner gets `OWNER` access.
     *
     * - **`projectPrivate`** - Object owner gets `OWNER` access, and project
     *   team members get access according to their roles.
     *
     * - **`publicRead`** - Object owner gets `OWNER` access, and `allUsers`
     *   get `READER` access.
     * @property ***REMOVED***boolean} [private] Make the uploaded file private. (Alias for
     *     `options.predefinedAcl = 'private'`)
     * @property ***REMOVED***boolean} [public] Make the uploaded file public. (Alias for
     *     `options.predefinedAcl = 'publicRead'`)
     * @property ***REMOVED***string} [userProject] The ID of the project which will be
     *     billed for the request.
     * @property ***REMOVED***string} [chunkSize] Create a separate request per chunk. This
     *     value is in bytes and should be a multiple of 256 KiB (2^18).
     *     ***REMOVED***@link https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload| We recommend using at least 8 MiB for the chunk size.}
     */
    /**
     * Create a unique resumable upload session URI. This is the first step when
     * performing a resumable upload.
     *
     * See the ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload| Resumable upload guide}
     * for more on how the entire process works.
     *
     * <h4>Note</h4>
     *
     * If you are just looking to perform a resumable upload without worrying
     * about any of the details, see ***REMOVED***@link File#createWriteStream}. Resumable
     * uploads are performed by default.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/how-tos/resumable-upload| Resumable upload guide}
     *
     * @param ***REMOVED***CreateResumableUploadOptions} [options] Configuration options.
     * @param ***REMOVED***CreateResumableUploadCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<CreateResumableUploadResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     * file.createResumableUpload(function(err, uri) ***REMOVED***
     *   if (!err) ***REMOVED***
     *     // `uri` can be used to PUT data to.
     * ***REMOVED***
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.createResumableUpload().then(function(data) ***REMOVED***
     *   const uri = data[0];
     * });
     * ```
     */
    createResumableUpload(optionsOrCallback, callback) ***REMOVED***
        var _a, _b;
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        callback =
            typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
        const retryOptions = this.storage.retryOptions;
        if ((((_a = options === null || options === void 0 ? void 0 : options.preconditionOpts) === null || _a === void 0 ? void 0 : _a.ifGenerationMatch) === undefined &&
            ((_b = this.instancePreconditionOpts) === null || _b === void 0 ? void 0 : _b.ifGenerationMatch) === undefined &&
            this.storage.retryOptions.idempotencyStrategy ===
                storage_1.IdempotencyStrategy.RetryConditional) ||
            this.storage.retryOptions.idempotencyStrategy ===
                storage_1.IdempotencyStrategy.RetryNever) ***REMOVED***
            retryOptions.autoRetry = false;
      ***REMOVED***
        resumableUpload.createURI(***REMOVED***
            authClient: this.storage.authClient,
            apiEndpoint: this.storage.apiEndpoint,
            bucket: this.bucket.name,
            customRequestOptions: this.getRequestInterceptors().reduce((reqOpts, interceptorFn) => interceptorFn(reqOpts), ***REMOVED***}),
            file: this.name,
            generation: this.generation,
            key: this.encryptionKey,
            kmsKeyName: this.kmsKeyName,
            metadata: options.metadata,
            offset: options.offset,
            origin: options.origin,
            predefinedAcl: options.predefinedAcl,
            private: options.private,
            public: options.public,
            userProject: options.userProject || this.userProject,
            retryOptions: retryOptions,
            params: (options === null || options === void 0 ? void 0 : options.preconditionOpts) || this.instancePreconditionOpts,
      ***REMOVED***, callback);
        this.storage.retryOptions.autoRetry = this.instanceRetryValue;
  ***REMOVED***
    /**
     * @typedef ***REMOVED***object} CreateWriteStreamOptions Configuration options for File#createWriteStream().
     * @property ***REMOVED***string} [contentType] Alias for
     *     `options.metadata.contentType`. If set to `auto`, the file name is used
     *     to determine the contentType.
     * @property ***REMOVED***string|boolean} [gzip] If true, automatically gzip the file.
     *     If set to `auto`, the contentType is used to determine if the file
     * should be gzipped. This will set `options.metadata.contentEncoding` to
     * `gzip` if necessary.
     * @property ***REMOVED***object} [metadata] See the examples below or
     *     ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/insert#request_properties_JSON| Objects: insert request body}
     *     for more details.
     * @property ***REMOVED***number} [offset] The starting byte of the upload stream, for
     *     resuming an interrupted upload. Defaults to 0.
     * @property ***REMOVED***string} [predefinedAcl] Apply a predefined set of access
     * controls to this object.
     *
     * Acceptable values are:
     * - **`authenticatedRead`** - Object owner gets `OWNER` access, and
     *   `allAuthenticatedUsers` get `READER` access.
     *
     * - **`bucketOwnerFullControl`** - Object owner gets `OWNER` access, and
     *   project team owners get `OWNER` access.
     *
     * - **`bucketOwnerRead`** - Object owner gets `OWNER` access, and project
     *   team owners get `READER` access.
     *
     * - **`private`** - Object owner gets `OWNER` access.
     *
     * - **`projectPrivate`** - Object owner gets `OWNER` access, and project
     *   team members get access according to their roles.
     *
     * - **`publicRead`** - Object owner gets `OWNER` access, and `allUsers`
     *   get `READER` access.
     * @property ***REMOVED***boolean} [private] Make the uploaded file private. (Alias for
     *     `options.predefinedAcl = 'private'`)
     * @property ***REMOVED***boolean} [public] Make the uploaded file public. (Alias for
     *     `options.predefinedAcl = 'publicRead'`)
     * @property ***REMOVED***boolean} [resumable] Force a resumable upload. NOTE: When
     *     working with streams, the file format and size is unknown until it's
     *     completely consumed. Because of this, it's best for you to be explicit
     *     for what makes sense given your input.
     * @property ***REMOVED***number} [timeout=60000] Set the HTTP request timeout in
     *     milliseconds. This option is not available for resumable uploads.
     *     Default: `60000`
     * @property ***REMOVED***string} [uri] The URI for an already-created resumable
     *     upload. See ***REMOVED***@link File#createResumableUpload}.
     * @property ***REMOVED***string} [userProject] The ID of the project which will be
     *     billed for the request.
     * @property ***REMOVED***string|boolean} [validation] Possible values: `"md5"`,
     *     `"crc32c"`, or `false`. By default, data integrity is validated with a
     *     CRC32c checksum. You may use MD5 if preferred, but that hash is not
     *     supported for composite objects. An error will be raised if MD5 is
     *     specified but is not available. You may also choose to skip validation
     *     completely, however this is **not recommended**. In addition to specifying
     *     validation type, providing `metadata.crc32c` or `metadata.md5Hash` will
     *     cause the server to perform validation in addition to client validation.
     *     NOTE: Validation is automatically skipped for objects that were
     *     uploaded using the `gzip` option and have already compressed content.
     */
    /**
     * Create a writable stream to overwrite the contents of the file in your
     * bucket.
     *
     * A File object can also be used to create files for the first time.
     *
     * Resumable uploads are automatically enabled and must be shut off explicitly
     * by setting `options.resumable` to `false`.
     *
     *
     * <p class="notice">
     *   There is some overhead when using a resumable upload that can cause
     *   noticeable performance degradation while uploading a series of small
     *   files. When uploading files less than 10MB, it is recommended that the
     *   resumable feature is disabled.
     * </p>
     *
     * NOTE: Writable streams will emit the `finish` event when the file is fully
     * uploaded.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/how-tos/upload| Upload Options (Simple or Resumable)}
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/insert| Objects: insert API Documentation}
     *
     * @param ***REMOVED***CreateWriteStreamOptions} [options] Configuration options.
     * @returns ***REMOVED***WritableStream}
     *
     * @example
     * ```
     * const fs = require('fs');
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     *
     * //-
     * // <h4>Uploading a File</h4>
     * //
     * // Now, consider a case where we want to upload a file to your bucket. You
     * // have the option of using ***REMOVED***@link Bucket#upload}, but that is just
     * // a convenience method which will do the following.
     * //-
     * fs.createReadStream('/Users/stephen/Photos/birthday-at-the-zoo/panda.jpg')
     *   .pipe(file.createWriteStream())
     *   .on('error', function(err) ***REMOVED***})
     *   .on('finish', function() ***REMOVED***
     *     // The file upload is complete.
     * ***REMOVED***);
     *
     * //-
     * // <h4>Uploading a File with gzip compression</h4>
     * //-
     * fs.createReadStream('/Users/stephen/site/index.html')
     *   .pipe(file.createWriteStream(***REMOVED*** gzip: true }))
     *   .on('error', function(err) ***REMOVED***})
     *   .on('finish', function() ***REMOVED***
     *     // The file upload is complete.
     * ***REMOVED***);
     *
     * //-
     * // Downloading the file with `createReadStream` will automatically decode
     * // the file.
     * //-
     *
     * //-
     * // <h4>Uploading a File with Metadata</h4>
     * //
     * // One last case you may run into is when you want to upload a file to your
     * // bucket and set its metadata at the same time. Like above, you can use
     * // ***REMOVED***@link Bucket#upload} to do this, which is just a wrapper around
     * // the following.
     * //-
     * fs.createReadStream('/Users/stephen/Photos/birthday-at-the-zoo/panda.jpg')
     *   .pipe(file.createWriteStream(***REMOVED***
     *     metadata: ***REMOVED***
     *       contentType: 'image/jpeg',
     *       metadata: ***REMOVED***
     *         custom: 'metadata'
     *     ***REMOVED***
     *   ***REMOVED***
     * ***REMOVED***))
     *   .on('error', function(err) ***REMOVED***})
     *   .on('finish', function() ***REMOVED***
     *     // The file upload is complete.
     * ***REMOVED***);
     * ```
     */
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    createWriteStream(options = ***REMOVED***}) ***REMOVED***
        options = extend(true, ***REMOVED*** metadata: ***REMOVED***} }, options);
        if (options.contentType) ***REMOVED***
            options.metadata.contentType = options.contentType;
      ***REMOVED***
        if (!options.metadata.contentType ||
            options.metadata.contentType === 'auto') ***REMOVED***
            const detectedContentType = mime.getType(this.name);
            if (detectedContentType) ***REMOVED***
                options.metadata.contentType = detectedContentType;
          ***REMOVED***
      ***REMOVED***
        let gzip = options.gzip;
        if (gzip === 'auto') ***REMOVED***
            gzip = compressible(options.metadata.contentType);
      ***REMOVED***
        if (gzip) ***REMOVED***
            options.metadata.contentEncoding = 'gzip';
      ***REMOVED***
        let crc32c = true;
        let md5 = false;
        if (typeof options.validation === 'string') ***REMOVED***
            options.validation = options.validation.toLowerCase();
            crc32c = options.validation === 'crc32c';
            md5 = options.validation === 'md5';
      ***REMOVED***
        else if (options.validation === false) ***REMOVED***
            crc32c = false;
      ***REMOVED***
        /**
         * A callback for determining when the underlying pipeline is complete.
         * It's possible the pipeline callback could error before the write stream
         * calls `final` so by default this will destroy the write stream unless the
         * write stream sets this callback via its `final` handler.
         * @param error An optional error
         */
        let pipelineCallback = error => ***REMOVED***
            writeStream.destroy(error || undefined);
      ***REMOVED***;
        // A stream for consumer to write to
        const writeStream = new stream_1.Writable(***REMOVED***
            final(cb) ***REMOVED***
                // Set the pipeline callback to this callback so the pipeline's results
                // can be populated to the consumer
                pipelineCallback = cb;
                emitStream.end();
          ***REMOVED***,
            write(chunk, encoding, cb) ***REMOVED***
                emitStream.write(chunk, encoding, cb);
          ***REMOVED***,
      ***REMOVED***);
        const emitStream = new util_1.PassThroughShim();
        const hashCalculatingStream = new hash_stream_validator_1.HashStreamValidator(***REMOVED***
            crc32c,
            md5,
            crc32cGenerator: this.crc32cGenerator,
      ***REMOVED***);
        const fileWriteStream = duplexify();
        let fileWriteStreamMetadataReceived = false;
        // Handing off emitted events to users
        emitStream.on('reading', () => writeStream.emit('reading'));
        emitStream.on('writing', () => writeStream.emit('writing'));
        fileWriteStream.on('progress', evt => writeStream.emit('progress', evt));
        fileWriteStream.on('response', resp => writeStream.emit('response', resp));
        fileWriteStream.once('metadata', () => ***REMOVED***
            fileWriteStreamMetadataReceived = true;
      ***REMOVED***);
        writeStream.on('writing', () => ***REMOVED***
            if (options.resumable === false) ***REMOVED***
                this.startSimpleUpload_(fileWriteStream, options);
          ***REMOVED***
            else ***REMOVED***
                this.startResumableUpload_(fileWriteStream, options);
          ***REMOVED***
            (0, stream_1.pipeline)(emitStream, gzip ? zlib.createGzip() : new stream_1.PassThrough(), hashCalculatingStream, fileWriteStream, async (e) => ***REMOVED***
                if (e) ***REMOVED***
                    return pipelineCallback(e);
              ***REMOVED***
                // We want to make sure we've received the metadata from the server in order
                // to properly validate the object's integrity. Depending on the type of upload,
                // the stream could close before the response is returned.
                if (!fileWriteStreamMetadataReceived) ***REMOVED***
                    try ***REMOVED***
                        await new Promise((resolve, reject) => ***REMOVED***
                            fileWriteStream.once('metadata', resolve);
                            fileWriteStream.once('error', reject);
                      ***REMOVED***);
                  ***REMOVED***
                    catch (e) ***REMOVED***
                        return pipelineCallback(e);
                  ***REMOVED***
              ***REMOVED***
                try ***REMOVED***
                    await __classPrivateFieldGet(this, _File_instances, "m", _File_validateIntegrity).call(this, hashCalculatingStream, ***REMOVED*** crc32c, md5 });
                    pipelineCallback();
              ***REMOVED***
                catch (e) ***REMOVED***
                    pipelineCallback(e);
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***);
        return writeStream;
  ***REMOVED***
    delete(optionsOrCallback, cb) ***REMOVED***
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        cb = typeof optionsOrCallback === 'function' ? optionsOrCallback : cb;
        this.disableAutoRetryConditionallyIdempotent_(this.methods.delete, bucket_1.AvailableServiceObjectMethods.delete, options);
        super
            .delete(options)
            .then(resp => cb(null, ...resp))
            .catch(cb)
            .finally(() => ***REMOVED***
            this.storage.retryOptions.autoRetry = this.instanceRetryValue;
      ***REMOVED***);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} DownloadResponse
     * @property [0] The contents of a File.
     */
    /**
     * @callback DownloadCallback
     * @param err Request error, if any.
     * @param contents The contents of a File.
     */
    /**
     * Convenience method to download a file into memory or to a local
     * destination.
     *
     * @param ***REMOVED***object} [options] Configuration options. The arguments match those
     *     passed to ***REMOVED***@link File#createReadStream}.
     * @param ***REMOVED***string} [options.destination] Local file path to write the file's
     *     contents to.
     * @param ***REMOVED***string} [options.userProject] The ID of the project which will be
     *     billed for the request.
     * @param ***REMOVED***DownloadCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<DownloadResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     *
     * //-
     * // Download a file into memory. The contents will be available as the
     * second
     * // argument in the demonstration below, `contents`.
     * //-
     * file.download(function(err, contents) ***REMOVED***});
     *
     * //-
     * // Download a file to a local destination.
     * //-
     * file.download(***REMOVED***
     *   destination: '/Users/me/Desktop/file-backup.txt'
     * }, function(err) ***REMOVED***});
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.download().then(function(data) ***REMOVED***
     *   const contents = data[0];
     * });
     *
     * ```
     * @example <caption>include:samples/files.js</caption>
     * region_tag:storage_download_file
     * Another example:
     *
     * @example <caption>include:samples/encryption.js</caption>
     * region_tag:storage_download_encrypted_file
     * Example of downloading an encrypted file:
     *
     * @example <caption>include:samples/requesterPays.js</caption>
     * region_tag:storage_download_file_requester_pays
     * Example of downloading a file where the requester pays:
     */
    download(optionsOrCallback, cb) ***REMOVED***
        let options;
        if (typeof optionsOrCallback === 'function') ***REMOVED***
            cb = optionsOrCallback;
            options = ***REMOVED***};
      ***REMOVED***
        else ***REMOVED***
            options = optionsOrCallback;
      ***REMOVED***
        let called = false;
        const callback = ((...args) => ***REMOVED***
            if (!called)
                cb(...args);
            called = true;
      ***REMOVED***);
        const destination = options.destination;
        delete options.destination;
        const fileStream = this.createReadStream(options);
        let receivedData = false;
        if (destination) ***REMOVED***
            fileStream
                .on('error', callback)
                .once('data', data => ***REMOVED***
                // We know that the file exists the server - now we can truncate/write to a file
                receivedData = true;
                const writable = fs.createWriteStream(destination);
                writable.write(data);
                fileStream
                    .pipe(writable)
                    .on('error', callback)
                    .on('finish', callback);
          ***REMOVED***)
                .on('end', () => ***REMOVED***
                // In the case of an empty file no data will be received before the end event fires
                if (!receivedData) ***REMOVED***
                    fs.openSync(destination, 'w');
                    callback(null, Buffer.alloc(0));
              ***REMOVED***
          ***REMOVED***);
      ***REMOVED***
        else ***REMOVED***
            this.getBufferFromReadable(fileStream)
                .then(contents => callback === null || callback === void 0 ? void 0 : callback(null, contents))
                .catch(callback);
      ***REMOVED***
  ***REMOVED***
    /**
     * The Storage API allows you to use a custom key for server-side encryption.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/encryption#customer-supplied| Customer-supplied Encryption Keys}
     *
     * @param ***REMOVED***string|buffer} encryptionKey An AES-256 encryption key.
     * @returns ***REMOVED***File}
     *
     * @example
     * ```
     * const crypto = require('crypto');
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const encryptionKey = crypto.randomBytes(32);
     *
     * const fileWithCustomEncryption = myBucket.file('my-file');
     * fileWithCustomEncryption.setEncryptionKey(encryptionKey);
     *
     * const fileWithoutCustomEncryption = myBucket.file('my-file');
     *
     * fileWithCustomEncryption.save('data', function(err) ***REMOVED***
     *   // Try to download with the File object that hasn't had
     *   // `setEncryptionKey()` called:
     *   fileWithoutCustomEncryption.download(function(err) ***REMOVED***
     *     // We will receive an error:
     *     //   err.message === 'Bad Request'
     *
     *     // Try again with the File object we called `setEncryptionKey()` on:
     *     fileWithCustomEncryption.download(function(err, contents) ***REMOVED***
     *       // contents.toString() === 'data'
     *   ***REMOVED***);
     * ***REMOVED***);
     * });
     *
     * ```
     * @example <caption>include:samples/encryption.js</caption>
     * region_tag:storage_upload_encrypted_file
     * Example of uploading an encrypted file:
     *
     * @example <caption>include:samples/encryption.js</caption>
     * region_tag:storage_download_encrypted_file
     * Example of downloading an encrypted file:
     */
    setEncryptionKey(encryptionKey) ***REMOVED***
        this.encryptionKey = encryptionKey;
        this.encryptionKeyBase64 = Buffer.from(encryptionKey).toString('base64');
        this.encryptionKeyHash = crypto
            .createHash('sha256')
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            .update(this.encryptionKeyBase64, 'base64')
            .digest('base64');
        this.encryptionKeyInterceptor = ***REMOVED***
            request: reqOpts => ***REMOVED***
                reqOpts.headers = reqOpts.headers || ***REMOVED***};
                reqOpts.headers['x-goog-encryption-algorithm'] = 'AES256';
                reqOpts.headers['x-goog-encryption-key'] = this.encryptionKeyBase64;
                reqOpts.headers['x-goog-encryption-key-sha256'] =
                    this.encryptionKeyHash;
                return reqOpts;
          ***REMOVED***,
      ***REMOVED***;
        this.interceptors.push(this.encryptionKeyInterceptor);
        return this;
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} GetExpirationDateResponse
     * @property ***REMOVED***date} 0 A Date object representing the earliest time this file's
     *     retention policy will expire.
     */
    /**
     * @callback GetExpirationDateCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***date} expirationDate A Date object representing the earliest time
     *     this file's retention policy will expire.
     */
    /**
     * If this bucket has a retention policy defined, use this method to get a
     * Date object representing the earliest time this file will expire.
     *
     * @param ***REMOVED***GetExpirationDateCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<GetExpirationDateResponse>}
     *
     * @example
     * ```
     * const storage = require('@google-cloud/storage')();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     *
     * file.getExpirationDate(function(err, expirationDate) ***REMOVED***
     *   // expirationDate is a Date object.
     * });
     * ```
     */
    getExpirationDate(callback) ***REMOVED***
        this.getMetadata((err, metadata, apiResponse) => ***REMOVED***
            if (err) ***REMOVED***
                callback(err, null, apiResponse);
                return;
          ***REMOVED***
            if (!metadata.retentionExpirationTime) ***REMOVED***
                const error = new Error(FileExceptionMessages.EXPIRATION_TIME_NA);
                callback(error, null, apiResponse);
                return;
          ***REMOVED***
            callback(null, new Date(metadata.retentionExpirationTime), apiResponse);
      ***REMOVED***);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} GenerateSignedPostPolicyV2Response
     * @property ***REMOVED***object} 0 The document policy.
     */
    /**
     * @callback GenerateSignedPostPolicyV2Callback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***object} policy The document policy.
     */
    /**
     * Get a signed policy document to allow a user to upload data with a POST
     * request.
     *
     * In Google Cloud Platform environments, such as Cloud Functions and App
     * Engine, you usually don't provide a `keyFilename` or `credentials` during
     * instantiation. In those environments, we call the
     * ***REMOVED***@link https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/signBlob| signBlob API}
     * to create a signed policy. That API requires either the
     * `https://www.googleapis.com/auth/iam` or
     * `https://www.googleapis.com/auth/cloud-platform` scope, so be sure they are
     * enabled.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/xml-api/post-object-v2| POST Object with the V2 signing process}
     *
     * @throws ***REMOVED***Error} If an expiration timestamp from the past is given.
     * @throws ***REMOVED***Error} If options.equals has an array with less or more than two
     *     members.
     * @throws ***REMOVED***Error} If options.startsWith has an array with less or more than two
     *     members.
     *
     * @param ***REMOVED***object} options Configuration options.
     * @param ***REMOVED***array|array[]} [options.equals] Array of request parameters and
     *     their expected value (e.g. [['$<field>', '<value>']]). Values are
     *     translated into equality constraints in the conditions field of the
     *     policy document (e.g. ['eq', '$<field>', '<value>']). If only one
     *     equality condition is to be specified, options.equals can be a one-
     *     dimensional array (e.g. ['$<field>', '<value>']).
     * @param ***REMOVED****} options.expires - A timestamp when this policy will expire. Any
     *     value given is passed to `new Date()`.
     * @param ***REMOVED***array|array[]} [options.startsWith] Array of request parameters and
     *     their expected prefixes (e.g. [['$<field>', '<value>']). Values are
     *     translated into starts-with constraints in the conditions field of the
     *     policy document (e.g. ['starts-with', '$<field>', '<value>']). If only
     *     one prefix condition is to be specified, options.startsWith can be a
     * one- dimensional array (e.g. ['$<field>', '<value>']).
     * @param ***REMOVED***string} [options.acl] ACL for the object from possibly predefined
     *     ACLs.
     * @param ***REMOVED***string} [options.successRedirect] The URL to which the user client
     *     is redirected if the upload is successful.
     * @param ***REMOVED***string} [options.successStatus] - The status of the Google Storage
     *     response if the upload is successful (must be string).
     * @param ***REMOVED***object} [options.contentLengthRange]
     * @param ***REMOVED***number} [options.contentLengthRange.min] Minimum value for the
     *     request's content length.
     * @param ***REMOVED***number} [options.contentLengthRange.max] Maximum value for the
     *     request's content length.
     * @param ***REMOVED***GenerateSignedPostPolicyV2Callback} [callback] Callback function.
     * @returns ***REMOVED***Promise<GenerateSignedPostPolicyV2Response>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     * const options = ***REMOVED***
     *   equals: ['$Content-Type', 'image/jpeg'],
     *   expires: '10-25-2022',
     *   contentLengthRange: ***REMOVED***
     *     min: 0,
     *     max: 1024
     * ***REMOVED***
     * };
     *
     * file.generateSignedPostPolicyV2(options, function(err, policy) ***REMOVED***
     *   // policy.string: the policy document in plain text.
     *   // policy.base64: the policy document in base64.
     *   // policy.signature: the policy signature in base64.
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.generateSignedPostPolicyV2(options).then(function(data) ***REMOVED***
     *   const policy = data[0];
     * });
     * ```
     */
    generateSignedPostPolicyV2(optionsOrCallback, cb) ***REMOVED***
        const args = (0, util_1.normalize)(optionsOrCallback, cb);
        let options = args.options;
        const callback = args.callback;
        const expires = new Date(options.expires);
        if (isNaN(expires.getTime())) ***REMOVED***
            throw new Error(storage_1.ExceptionMessages.EXPIRATION_DATE_INVALID);
      ***REMOVED***
        if (expires.valueOf() < Date.now()) ***REMOVED***
            throw new Error(storage_1.ExceptionMessages.EXPIRATION_DATE_PAST);
      ***REMOVED***
        options = Object.assign(***REMOVED***}, options);
        const conditions = [
            ['eq', '$key', this.name],
            ***REMOVED***
                bucket: this.bucket.name,
          ***REMOVED***,
        ];
        if (Array.isArray(options.equals)) ***REMOVED***
            if (!Array.isArray(options.equals[0])) ***REMOVED***
                options.equals = [options.equals];
          ***REMOVED***
            options.equals.forEach(condition => ***REMOVED***
                if (!Array.isArray(condition) || condition.length !== 2) ***REMOVED***
                    throw new Error(FileExceptionMessages.EQUALS_CONDITION_TWO_ELEMENTS);
              ***REMOVED***
                conditions.push(['eq', condition[0], condition[1]]);
          ***REMOVED***);
      ***REMOVED***
        if (Array.isArray(options.startsWith)) ***REMOVED***
            if (!Array.isArray(options.startsWith[0])) ***REMOVED***
                options.startsWith = [options.startsWith];
          ***REMOVED***
            options.startsWith.forEach(condition => ***REMOVED***
                if (!Array.isArray(condition) || condition.length !== 2) ***REMOVED***
                    throw new Error(FileExceptionMessages.STARTS_WITH_TWO_ELEMENTS);
              ***REMOVED***
                conditions.push(['starts-with', condition[0], condition[1]]);
          ***REMOVED***);
      ***REMOVED***
        if (options.acl) ***REMOVED***
            conditions.push(***REMOVED***
                acl: options.acl,
          ***REMOVED***);
      ***REMOVED***
        if (options.successRedirect) ***REMOVED***
            conditions.push(***REMOVED***
                success_action_redirect: options.successRedirect,
          ***REMOVED***);
      ***REMOVED***
        if (options.successStatus) ***REMOVED***
            conditions.push(***REMOVED***
                success_action_status: options.successStatus,
          ***REMOVED***);
      ***REMOVED***
        if (options.contentLengthRange) ***REMOVED***
            const min = options.contentLengthRange.min;
            const max = options.contentLengthRange.max;
            if (typeof min !== 'number' || typeof max !== 'number') ***REMOVED***
                throw new Error(FileExceptionMessages.CONTENT_LENGTH_RANGE_MIN_MAX);
          ***REMOVED***
            conditions.push(['content-length-range', min, max]);
      ***REMOVED***
        const policy = ***REMOVED***
            expiration: expires.toISOString(),
            conditions,
      ***REMOVED***;
        const policyString = JSON.stringify(policy);
        const policyBase64 = Buffer.from(policyString).toString('base64');
        this.storage.authClient.sign(policyBase64).then(signature => ***REMOVED***
            callback(null, ***REMOVED***
                string: policyString,
                base64: policyBase64,
                signature,
          ***REMOVED***);
      ***REMOVED***, err => ***REMOVED***
            callback(new signer_1.SigningError(err.message));
      ***REMOVED***);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***object} SignedPostPolicyV4Output
     * @property ***REMOVED***string} url The request URL.
     * @property ***REMOVED***object} fields The form fields to include in the POST request.
     */
    /**
     * @typedef ***REMOVED***array} GenerateSignedPostPolicyV4Response
     * @property ***REMOVED***SignedPostPolicyV4Output} 0 An object containing the request URL and form fields.
     */
    /**
     * @callback GenerateSignedPostPolicyV4Callback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***SignedPostPolicyV4Output} output An object containing the request URL and form fields.
     */
    /**
     * Get a v4 signed policy document to allow a user to upload data with a POST
     * request.
     *
     * In Google Cloud Platform environments, such as Cloud Functions and App
     * Engine, you usually don't provide a `keyFilename` or `credentials` during
     * instantiation. In those environments, we call the
     * ***REMOVED***@link https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/signBlob| signBlob API}
     * to create a signed policy. That API requires either the
     * `https://www.googleapis.com/auth/iam` or
     * `https://www.googleapis.com/auth/cloud-platform` scope, so be sure they are
     * enabled.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/xml-api/post-object#policydocument| Policy Document Reference}
     *
     * @param ***REMOVED***object} options Configuration options.
     * @param ***REMOVED***Date|number|string} options.expires - A timestamp when this policy will expire. Any
     *     value given is passed to `new Date()`.
     * @param ***REMOVED***boolean} [config.virtualHostedStyle=false] Use virtual hosted-style
     *     URLs ('https://mybucket.storage.googleapis.com/...') instead of path-style
     *     ('https://storage.googleapis.com/mybucket/...'). Virtual hosted-style URLs
     *     should generally be preferred instead of path-style URL.
     *     Currently defaults to `false` for path-style, although this may change in a
     *     future major-version release.
     * @param ***REMOVED***string} [config.bucketBoundHostname] The bucket-bound hostname to return in
     *     the result, e.g. "https://cdn.example.com".
     * @param ***REMOVED***object} [config.fields] [Form fields]***REMOVED***@link https://cloud.google.com/storage/docs/xml-api/post-object#policydocument}
     *     to include in the signed policy. Any fields with key beginning with 'x-ignore-'
     *     will not be included in the policy to be signed.
     * @param ***REMOVED***object[]} [config.conditions] [Conditions]***REMOVED***@link https://cloud.google.com/storage/docs/authentication/signatures#policy-document}
     *     to include in the signed policy. All fields given in `config.fields` are
     *     automatically included in the conditions array, adding the same entry
     *     in both `fields` and `conditions` will result in duplicate entries.
     *
     * @param ***REMOVED***GenerateSignedPostPolicyV4Callback} [callback] Callback function.
     * @returns ***REMOVED***Promise<GenerateSignedPostPolicyV4Response>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     * const options = ***REMOVED***
     *   expires: '10-25-2022',
     *   conditions: [
     *     ['eq', '$Content-Type', 'image/jpeg'],
     *     ['content-length-range', 0, 1024],
     *   ],
     *   fields: ***REMOVED***
     *     acl: 'public-read',
     *     'x-goog-meta-foo': 'bar',
     *     'x-ignore-mykey': 'data'
     * ***REMOVED***
     * };
     *
     * file.generateSignedPostPolicyV4(options, function(err, response) ***REMOVED***
     *   // response.url The request URL
     *   // response.fields The form fields (including the signature) to include
     *   //     to be used to upload objects by HTML forms.
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.generateSignedPostPolicyV4(options).then(function(data) ***REMOVED***
     *   const response = data[0];
     *   // response.url The request URL
     *   // response.fields The form fields (including the signature) to include
     *   //     to be used to upload objects by HTML forms.
     * });
     * ```
     */
    generateSignedPostPolicyV4(optionsOrCallback, cb) ***REMOVED***
        const args = (0, util_1.normalize)(optionsOrCallback, cb);
        let options = args.options;
        const callback = args.callback;
        const expires = new Date(options.expires);
        if (isNaN(expires.getTime())) ***REMOVED***
            throw new Error(storage_1.ExceptionMessages.EXPIRATION_DATE_INVALID);
      ***REMOVED***
        if (expires.valueOf() < Date.now()) ***REMOVED***
            throw new Error(storage_1.ExceptionMessages.EXPIRATION_DATE_PAST);
      ***REMOVED***
        if (expires.valueOf() - Date.now() > SEVEN_DAYS * 1000) ***REMOVED***
            throw new Error(`Max allowed expiration is seven days ($***REMOVED***SEVEN_DAYS} seconds).`);
      ***REMOVED***
        options = Object.assign(***REMOVED***}, options);
        let fields = Object.assign(***REMOVED***}, options.fields);
        const now = new Date();
        const nowISO = (0, util_1.formatAsUTCISO)(now, true);
        const todayISO = (0, util_1.formatAsUTCISO)(now);
        const sign = async () => ***REMOVED***
            const ***REMOVED*** client_email } = await this.storage.authClient.getCredentials();
            const credential = `$***REMOVED***client_email}/$***REMOVED***todayISO}/auto/storage/goog4_request`;
            fields = ***REMOVED***
                ...fields,
                bucket: this.bucket.name,
                key: this.name,
                'x-goog-date': nowISO,
                'x-goog-credential': credential,
                'x-goog-algorithm': 'GOOG4-RSA-SHA256',
          ***REMOVED***;
            const conditions = options.conditions || [];
            Object.entries(fields).forEach(([key, value]) => ***REMOVED***
                if (!key.startsWith('x-ignore-')) ***REMOVED***
                    conditions.push(***REMOVED*** [key]: value });
              ***REMOVED***
          ***REMOVED***);
            delete fields.bucket;
            const expiration = (0, util_1.formatAsUTCISO)(expires, true, '-', ':');
            const policy = ***REMOVED***
                conditions,
                expiration,
          ***REMOVED***;
            const policyString = (0, util_1.unicodeJSONStringify)(policy);
            const policyBase64 = Buffer.from(policyString).toString('base64');
            try ***REMOVED***
                const signature = await this.storage.authClient.sign(policyBase64);
                const signatureHex = Buffer.from(signature, 'base64').toString('hex');
                fields['policy'] = policyBase64;
                fields['x-goog-signature'] = signatureHex;
                let url;
                if (options.virtualHostedStyle) ***REMOVED***
                    url = `https://$***REMOVED***this.bucket.name}.storage.googleapis.com/`;
              ***REMOVED***
                else if (options.bucketBoundHostname) ***REMOVED***
                    url = `$***REMOVED***options.bucketBoundHostname}/`;
              ***REMOVED***
                else ***REMOVED***
                    url = `$***REMOVED***exports.STORAGE_POST_POLICY_BASE_URL}/$***REMOVED***this.bucket.name}/`;
              ***REMOVED***
                return ***REMOVED***
                    url,
                    fields,
              ***REMOVED***;
          ***REMOVED***
            catch (err) ***REMOVED***
                throw new signer_1.SigningError(err.message);
          ***REMOVED***
      ***REMOVED***;
        sign().then(res => callback(null, res), callback);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} GetSignedUrlResponse
     * @property ***REMOVED***object} 0 The signed URL.
     */
    /**
     * @callback GetSignedUrlCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***object} url The signed URL.
     */
    /**
     * Get a signed URL to allow limited time access to the file.
     *
     * In Google Cloud Platform environments, such as Cloud Functions and App
     * Engine, you usually don't provide a `keyFilename` or `credentials` during
     * instantiation. In those environments, we call the
     * ***REMOVED***@link https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/signBlob| signBlob API}
     * to create a signed URL. That API requires either the
     * `https://www.googleapis.com/auth/iam` or
     * `https://www.googleapis.com/auth/cloud-platform` scope, so be sure they are
     * enabled.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/access-control/signed-urls| Signed URLs Reference}
     *
     * @throws ***REMOVED***Error} if an expiration timestamp from the past is given.
     *
     * @param ***REMOVED***object} config Configuration object.
     * @param ***REMOVED***string} config.action "read" (HTTP: GET), "write" (HTTP: PUT), or
     *     "delete" (HTTP: DELETE), "resumable" (HTTP: POST).
     *     When using "resumable", the header `X-Goog-Resumable: start` has
     *     to be sent when making a request with the signed URL.
     * @param ***REMOVED****} config.expires A timestamp when this link will expire. Any value
     *     given is passed to `new Date()`.
     *     Note: 'v4' supports maximum duration of 7 days (604800 seconds) from now.
     *     See [reference]***REMOVED***@link https://cloud.google.com/storage/docs/access-control/signed-urls#example}
     * @param ***REMOVED***string} [config.version='v2'] The signing version to use, either
     *     'v2' or 'v4'.
     * @param ***REMOVED***boolean} [config.virtualHostedStyle=false] Use virtual hosted-style
     *     URLs ('https://mybucket.storage.googleapis.com/...') instead of path-style
     *     ('https://storage.googleapis.com/mybucket/...'). Virtual hosted-style URLs
     *     should generally be preferred instaed of path-style URL.
     *     Currently defaults to `false` for path-style, although this may change in a
     *     future major-version release.
     * @param ***REMOVED***string} [config.cname] The cname for this bucket, i.e.,
     *     "https://cdn.example.com".
     * @param ***REMOVED***string} [config.contentMd5] The MD5 digest value in base64. Just like
     *     if you provide this, the client must provide this HTTP header with this same
     *     value in its request, so to if this parameter is not provided here,
     *     the client must not provide any value for this HTTP header in its request.
     * @param ***REMOVED***string} [config.contentType] Just like if you provide this, the client
     *     must provide this HTTP header with this same value in its request, so to if
     *     this parameter is not provided here, the client must not provide any value
     *     for this HTTP header in its request.
     * @param ***REMOVED***object} [config.extensionHeaders] If these headers are used, the
     * server will check to make sure that the client provides matching
     * values. See ***REMOVED***@link https://cloud.google.com/storage/docs/access-control/signed-urls#about-canonical-extension-headers| Canonical extension headers}
     * for the requirements of this feature, most notably:
     * - The header name must be prefixed with `x-goog-`
     * - The header name must be all lowercase
     *
     * Note: Multi-valued header passed as an array in the extensionHeaders
     *       object is converted into a string, delimited by `,` with
     *       no space. Requests made using the signed URL will need to
     *       delimit multi-valued headers using a single `,` as well, or
     *       else the server will report a mismatched signature.
     * @param ***REMOVED***object} [config.queryParams] Additional query parameters to include
     *     in the signed URL.
     * @param ***REMOVED***string} [config.promptSaveAs] The filename to prompt the user to
     *     save the file as when the signed url is accessed. This is ignored if
     *     `config.responseDisposition` is set.
     * @param ***REMOVED***string} [config.responseDisposition] The
     *     ***REMOVED***@link http://goo.gl/yMWxQV| response-content-disposition parameter} of the
     *     signed url.
     * @param ***REMOVED****} [config.accessibleAt=Date.now()] A timestamp when this link became usable. Any value
     *     given is passed to `new Date()`.
     *     Note: Use for 'v4' only.
     * @param ***REMOVED***string} [config.responseType] The response-content-type parameter
     *     of the signed url.
     * @param ***REMOVED***GetSignedUrlCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<GetSignedUrlResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     *
     * //-
     * // Generate a URL that allows temporary access to download your file.
     * //-
     * const request = require('request');
     *
     * const config = ***REMOVED***
     *   action: 'read',
     *   expires: '03-17-2025',
     * };
     *
     * file.getSignedUrl(config, function(err, url) ***REMOVED***
     *   if (err) ***REMOVED***
     *     console.error(err);
     *     return;
     * ***REMOVED***
     *
     *   // The file is now available to read from this URL.
     *   request(url, function(err, resp) ***REMOVED***
     *     // resp.statusCode = 200
     * ***REMOVED***);
     * });
     *
     * //-
     * // Generate a URL that allows temporary access to download your file.
     * // Access will begin at accessibleAt and end at expires.
     * //-
     * const request = require('request');
     *
     * const config = ***REMOVED***
     *   action: 'read',
     *   expires: '03-17-2025',
     *   accessibleAt: '03-13-2025'
     * };
     *
     * file.getSignedUrl(config, function(err, url) ***REMOVED***
     *   if (err) ***REMOVED***
     *     console.error(err);
     *     return;
     * ***REMOVED***
     *
     *   // The file will be available to read from this URL from 03-13-2025 to 03-17-2025.
     *   request(url, function(err, resp) ***REMOVED***
     *     // resp.statusCode = 200
     * ***REMOVED***);
     * });
     *
     * //-
     * // Generate a URL to allow write permissions. This means anyone with this
     * URL
     * // can send a POST request with new data that will overwrite the file.
     * //-
     * file.getSignedUrl(***REMOVED***
     *   action: 'write',
     *   expires: '03-17-2025'
     * }, function(err, url) ***REMOVED***
     *   if (err) ***REMOVED***
     *     console.error(err);
     *     return;
     * ***REMOVED***
     *
     *   // The file is now available to be written to.
     *   const writeStream = request.put(url);
     *   writeStream.end('New data');
     *
     *   writeStream.on('complete', function(resp) ***REMOVED***
     *     // Confirm the new content was saved.
     *     file.download(function(err, fileContents) ***REMOVED***
     *       console.log('Contents:', fileContents.toString());
     *       // Contents: New data
     *   ***REMOVED***);
     * ***REMOVED***);
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.getSignedUrl(config).then(function(data) ***REMOVED***
     *   const url = data[0];
     * });
     *
     * ```
     * @example <caption>include:samples/files.js</caption>
     * region_tag:storage_generate_signed_url
     * Another example:
     */
    getSignedUrl(cfg, callback) ***REMOVED***
        const method = ActionToHTTPMethod[cfg.action];
        if (!method) ***REMOVED***
            throw new Error(storage_1.ExceptionMessages.INVALID_ACTION);
      ***REMOVED***
        const extensionHeaders = (0, util_1.objectKeyToLowercase)(cfg.extensionHeaders || ***REMOVED***});
        if (cfg.action === 'resumable') ***REMOVED***
            extensionHeaders['x-goog-resumable'] = 'start';
      ***REMOVED***
        const queryParams = Object.assign(***REMOVED***}, cfg.queryParams);
        if (typeof cfg.responseType === 'string') ***REMOVED***
            queryParams['response-content-type'] = cfg.responseType;
      ***REMOVED***
        if (typeof cfg.promptSaveAs === 'string') ***REMOVED***
            queryParams['response-content-disposition'] =
                'attachment; filename="' + cfg.promptSaveAs + '"';
      ***REMOVED***
        if (typeof cfg.responseDisposition === 'string') ***REMOVED***
            queryParams['response-content-disposition'] = cfg.responseDisposition;
      ***REMOVED***
        if (this.generation) ***REMOVED***
            queryParams['generation'] = this.generation.toString();
      ***REMOVED***
        const signConfig = ***REMOVED***
            method,
            expires: cfg.expires,
            accessibleAt: cfg.accessibleAt,
            extensionHeaders,
            queryParams,
            contentMd5: cfg.contentMd5,
            contentType: cfg.contentType,
      ***REMOVED***;
        if (cfg.cname) ***REMOVED***
            signConfig.cname = cfg.cname;
      ***REMOVED***
        if (cfg.version) ***REMOVED***
            signConfig.version = cfg.version;
      ***REMOVED***
        if (cfg.virtualHostedStyle) ***REMOVED***
            signConfig.virtualHostedStyle = cfg.virtualHostedStyle;
      ***REMOVED***
        if (!this.signer) ***REMOVED***
            this.signer = new signer_1.URLSigner(this.storage.authClient, this.bucket, this);
      ***REMOVED***
        this.signer
            .getSignedUrl(signConfig)
            .then(signedUrl => callback(null, signedUrl), callback);
  ***REMOVED***
    /**
     * @callback IsPublicCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***boolean} resp Whether file is public or not.
     */
    /**
     * @typedef ***REMOVED***array} IsPublicResponse
     * @property ***REMOVED***boolean} 0 Whether file is public or not.
     */
    /**
     * Check whether this file is public or not by sending
     * a HEAD request without credentials.
     * No errors from the server indicates that the current
     * file is public.
     * A 403-Forbidden error ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/status-codes#403_Forbidden}
     * indicates that file is private.
     * Any other non 403 error is propagated to user.
     *
     * @param ***REMOVED***IsPublicCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<IsPublicResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     *
     * //-
     * // Check whether the file is publicly accessible.
     * //-
     * file.isPublic(function(err, resp) ***REMOVED***
     *   if (err) ***REMOVED***
     *     console.error(err);
     *     return;
     * ***REMOVED***
     *   console.log(`the file $***REMOVED***file.id} is public: $***REMOVED***resp}`) ;
     * })
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.isPublic().then(function(data) ***REMOVED***
     *   const resp = data[0];
     * });
     * ```
     */
    isPublic(callback) ***REMOVED***
        var _a;
        // Build any custom headers based on the defined interceptors on the parent
        // storage object and this object
        const storageInterceptors = ((_a = this.storage) === null || _a === void 0 ? void 0 : _a.interceptors) || [];
        const fileInterceptors = this.interceptors || [];
        const allInterceptors = storageInterceptors.concat(fileInterceptors);
        const headers = allInterceptors.reduce((acc, curInterceptor) => ***REMOVED***
            const currentHeaders = curInterceptor.request(***REMOVED***
                uri: `$***REMOVED***this.storage.apiEndpoint}/$***REMOVED***this.bucket.name}/$***REMOVED***encodeURIComponent(this.name)}`,
          ***REMOVED***);
            Object.assign(acc, currentHeaders.headers);
            return acc;
      ***REMOVED***, ***REMOVED***});
        nodejs_common_1.util.makeRequest(***REMOVED***
            method: 'GET',
            uri: `$***REMOVED***this.storage.apiEndpoint}/$***REMOVED***this.bucket.name}/$***REMOVED***encodeURIComponent(this.name)}`,
            headers,
      ***REMOVED***, ***REMOVED***
            retryOptions: this.storage.retryOptions,
      ***REMOVED***, (err) => ***REMOVED***
            if (err) ***REMOVED***
                const apiError = err;
                if (apiError.code === 403) ***REMOVED***
                    callback(null, false);
              ***REMOVED***
                else ***REMOVED***
                    callback(err);
              ***REMOVED***
          ***REMOVED***
            else ***REMOVED***
                callback(null, true);
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***object} MakeFilePrivateOptions Configuration options for File#makePrivate().
     * @property ***REMOVED***Metadata} [metadata] Define custom metadata properties to define
     *     along with the operation.
     * @property ***REMOVED***boolean} [strict] If true, set the file to be private to
     *     only the owner user. Otherwise, it will be private to the project.
     * @property ***REMOVED***string} [userProject] The ID of the project which will be
     *     billed for the request.
     */
    /**
     * @callback MakeFilePrivateCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***object} apiResponse The full API response.
     */
    /**
     * @typedef ***REMOVED***array} MakeFilePrivateResponse
     * @property ***REMOVED***object} 0 The full API response.
     */
    /**
     * Make a file private to the project and remove all other permissions.
     * Set `options.strict` to true to make the file private to only the owner.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/patch| Objects: patch API Documentation}
     *
     * @param ***REMOVED***MakeFilePrivateOptions} [options] Configuration options.
     * @param ***REMOVED***MakeFilePrivateCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<MakeFilePrivateResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     *
     * //-
     * // Set the file private so only project maintainers can see and modify it.
     * //-
     * file.makePrivate(function(err) ***REMOVED***});
     *
     * //-
     * // Set the file private so only the owner can see and modify it.
     * //-
     * file.makePrivate(***REMOVED*** strict: true }, function(err) ***REMOVED***});
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.makePrivate().then(function(data) ***REMOVED***
     *   const apiResponse = data[0];
     * });
     * ```
     */
    makePrivate(optionsOrCallback, callback) ***REMOVED***
        var _a, _b;
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        callback =
            typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
        const query = ***REMOVED***
            predefinedAcl: options.strict ? 'private' : 'projectPrivate',
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
      ***REMOVED***;
        if (((_a = options.preconditionOpts) === null || _a === void 0 ? void 0 : _a.ifGenerationMatch) !== undefined) ***REMOVED***
            query.ifGenerationMatch = (_b = options.preconditionOpts) === null || _b === void 0 ? void 0 : _b.ifGenerationMatch;
            delete options.preconditionOpts;
      ***REMOVED***
        if (options.userProject) ***REMOVED***
            query.userProject = options.userProject;
      ***REMOVED***
        // You aren't allowed to set both predefinedAcl & acl properties on a file,
        // so acl must explicitly be nullified, destroying all previous acls on the
        // file.
        const metadata = extend(***REMOVED***}, options.metadata, ***REMOVED*** acl: null });
        this.setMetadata(metadata, query, callback);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} MakeFilePublicResponse
     * @property ***REMOVED***object} 0 The full API response.
     */
    /**
     * @callback MakeFilePublicCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***object} apiResponse The full API response.
     */
    /**
     * Set a file to be publicly readable and maintain all previous permissions.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objectAccessControls/insert| ObjectAccessControls: insert API Documentation}
     *
     * @param ***REMOVED***MakeFilePublicCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<MakeFilePublicResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     *
     * file.makePublic(function(err, apiResponse) ***REMOVED***});
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.makePublic().then(function(data) ***REMOVED***
     *   const apiResponse = data[0];
     * });
     *
     * ```
     * @example <caption>include:samples/files.js</caption>
     * region_tag:storage_make_public
     * Another example:
     */
    makePublic(callback) ***REMOVED***
        callback = callback || nodejs_common_1.util.noop;
        this.acl.add(***REMOVED***
            entity: 'allUsers',
            role: 'READER',
      ***REMOVED***, (err, acl, resp) => ***REMOVED***
            callback(err, resp);
      ***REMOVED***);
  ***REMOVED***
    /**
     * The public URL of this File
     * Use ***REMOVED***@link File#makePublic} to enable anonymous access via the returned URL.
     *
     * @returns ***REMOVED***string}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const bucket = storage.bucket('albums');
     * const file = bucket.file('my-file');
     *
     * // publicUrl will be "https://storage.googleapis.com/albums/my-file"
     * const publicUrl = file.publicUrl();
     * ```
     */
    publicUrl() ***REMOVED***
        return `$***REMOVED***this.storage.apiEndpoint}/$***REMOVED***this.bucket.name}/$***REMOVED***encodeURIComponent(this.name)}`;
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} MoveResponse
     * @property ***REMOVED***File} 0 The destination File.
     * @property ***REMOVED***object} 1 The full API response.
     */
    /**
     * @callback MoveCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***?File} destinationFile The destination File.
     * @param ***REMOVED***object} apiResponse The full API response.
     */
    /**
     * @typedef ***REMOVED***object} MoveOptions Configuration options for File#move(). See an
     *     ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects#resource| Object resource}.
     * @param ***REMOVED***string} [userProject] The ID of the project which will be
     *     billed for the request.
     */
    /**
     * Move this file to another location. By default, this will rename the file
     * and keep it in the same bucket, but you can choose to move it to another
     * Bucket by providing a Bucket or File object or a URL beginning with
     * "gs://".
     *
     * **Warning**:
     * There is currently no atomic `move` method in the Cloud Storage API,
     * so this method is a composition of ***REMOVED***@link File#copy} (to the new
     * location) and ***REMOVED***@link File#delete} (from the old location). While
     * unlikely, it is possible that an error returned to your callback could be
     * triggered from either one of these API calls failing, which could leave a
     * duplicate file lingering. The error message will indicate what operation
     * has failed.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects/copy| Objects: copy API Documentation}
     *
     * @throws ***REMOVED***Error} If the destination file is not provided.
     *
     * @param ***REMOVED***string|Bucket|File} destination Destination file.
     * @param ***REMOVED***MoveCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<MoveResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * //-
     * // You can pass in a variety of types for the destination.
     * //
     * // For all of the below examples, assume we are working with the following
     * // Bucket and File objects.
     * //-
     * const bucket = storage.bucket('my-bucket');
     * const file = bucket.file('my-image.png');
     *
     * //-
     * // If you pass in a string for the destination, the file is moved to its
     * // current bucket, under the new name provided.
     * //-
     * file.move('my-image-new.png', function(err, destinationFile, apiResponse) ***REMOVED***
     *   // `my-bucket` no longer contains:
     *   // - "my-image.png"
     *   // but contains instead:
     *   // - "my-image-new.png"
     *
     *   // `destinationFile` is an instance of a File object that refers to your
     *   // new file.
     * });
     *
     * //-
     * // If you pass in a string starting with "gs://" for the destination, the
     * // file is copied to the other bucket and under the new name provided.
     * //-
     * const newLocation = 'gs://another-bucket/my-image-new.png';
     * file.move(newLocation, function(err, destinationFile, apiResponse) ***REMOVED***
     *   // `my-bucket` no longer contains:
     *   // - "my-image.png"
     *   //
     *   // `another-bucket` now contains:
     *   // - "my-image-new.png"
     *
     *   // `destinationFile` is an instance of a File object that refers to your
     *   // new file.
     * });
     *
     * //-
     * // If you pass in a Bucket object, the file will be moved to that bucket
     * // using the same name.
     * //-
     * const anotherBucket = gcs.bucket('another-bucket');
     *
     * file.move(anotherBucket, function(err, destinationFile, apiResponse) ***REMOVED***
     *   // `my-bucket` no longer contains:
     *   // - "my-image.png"
     *   //
     *   // `another-bucket` now contains:
     *   // - "my-image.png"
     *
     *   // `destinationFile` is an instance of a File object that refers to your
     *   // new file.
     * });
     *
     * //-
     * // If you pass in a File object, you have complete control over the new
     * // bucket and filename.
     * //-
     * const anotherFile = anotherBucket.file('my-awesome-image.png');
     *
     * file.move(anotherFile, function(err, destinationFile, apiResponse) ***REMOVED***
     *   // `my-bucket` no longer contains:
     *   // - "my-image.png"
     *   //
     *   // `another-bucket` now contains:
     *   // - "my-awesome-image.png"
     *
     *   // Note:
     *   // The `destinationFile` parameter is equal to `anotherFile`.
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.move('my-image-new.png').then(function(data) ***REMOVED***
     *   const destinationFile = data[0];
     *   const apiResponse = data[1];
     * });
     *
     * ```
     * @example <caption>include:samples/files.js</caption>
     * region_tag:storage_move_file
     * Another example:
     */
    move(destination, optionsOrCallback, callback) ***REMOVED***
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        callback =
            typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
        callback = callback || nodejs_common_1.util.noop;
        this.copy(destination, options, (err, destinationFile, copyApiResponse) => ***REMOVED***
            if (err) ***REMOVED***
                err.message = 'file#copy failed with an error - ' + err.message;
                callback(err, null, copyApiResponse);
                return;
          ***REMOVED***
            if (this.name !== destinationFile.name ||
                this.bucket.name !== destinationFile.bucket.name) ***REMOVED***
                this.delete(options, (err, apiResponse) => ***REMOVED***
                    if (err) ***REMOVED***
                        err.message = 'file#delete failed with an error - ' + err.message;
                        callback(err, destinationFile, apiResponse);
                        return;
                  ***REMOVED***
                    callback(null, destinationFile, copyApiResponse);
              ***REMOVED***);
          ***REMOVED***
            else ***REMOVED***
                callback(null, destinationFile, copyApiResponse);
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} RenameResponse
     * @property ***REMOVED***File} 0 The destination File.
     * @property ***REMOVED***object} 1 The full API response.
     */
    /**
     * @callback RenameCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***?File} destinationFile The destination File.
     * @param ***REMOVED***object} apiResponse The full API response.
     */
    /**
     * @typedef ***REMOVED***object} RenameOptions Configuration options for File#move(). See an
     *     ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/objects#resource| Object resource}.
     * @param ***REMOVED***string} [userProject] The ID of the project which will be
     *     billed for the request.
     */
    /**
     * Rename this file.
     *
     * **Warning**:
     * There is currently no atomic `rename` method in the Cloud Storage API,
     * so this method is an alias of ***REMOVED***@link File#move}, which in turn is a
     * composition of ***REMOVED***@link File#copy} (to the new location) and
     * ***REMOVED***@link File#delete} (from the old location). While
     * unlikely, it is possible that an error returned to your callback could be
     * triggered from either one of these API calls failing, which could leave a
     * duplicate file lingering. The error message will indicate what operation
     * has failed.
     *
     * @param ***REMOVED***string|File} destinationFile Destination file.
     * @param ***REMOVED***RenameCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<RenameResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     *
     * //-
     * // You can pass in a string or a File object.
     * //
     * // For all of the below examples, assume we are working with the following
     * // Bucket and File objects.
     * //-
     *
     * const bucket = storage.bucket('my-bucket');
     * const file = bucket.file('my-image.png');
     *
     * //-
     * // You can pass in a string for the destinationFile.
     * //-
     * file.rename('renamed-image.png', function(err, renamedFile, apiResponse) ***REMOVED***
     *   // `my-bucket` no longer contains:
     *   // - "my-image.png"
     *   // but contains instead:
     *   // - "renamed-image.png"
     *
     *   // `renamedFile` is an instance of a File object that refers to your
     *   // renamed file.
     * });
     *
     * //-
     * // You can pass in a File object.
     * //-
     * const anotherFile = anotherBucket.file('my-awesome-image.png');
     *
     * file.rename(anotherFile, function(err, renamedFile, apiResponse) ***REMOVED***
     *   // `my-bucket` no longer contains:
     *   // - "my-image.png"
     *
     *   // Note:
     *   // The `renamedFile` parameter is equal to `anotherFile`.
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.rename('my-renamed-image.png').then(function(data) ***REMOVED***
     *   const renamedFile = data[0];
     *   const apiResponse = data[1];
     * });
     * ```
     */
    rename(destinationFile, optionsOrCallback, callback) ***REMOVED***
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        callback =
            typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
        callback = callback || nodejs_common_1.util.noop;
        this.move(destinationFile, options, callback);
  ***REMOVED***
    /**
     * Makes request and applies userProject query parameter if necessary.
     *
     * @private
     *
     * @param ***REMOVED***object} reqOpts - The request options.
     * @param ***REMOVED***function} callback - The callback function.
     */
    request(reqOpts, callback) ***REMOVED***
        return this.parent.request.call(this, reqOpts, callback);
  ***REMOVED***
    /**
     * @callback RotateEncryptionKeyCallback
     * @extends CopyCallback
     */
    /**
     * @typedef RotateEncryptionKeyResponse
     * @extends CopyResponse
     */
    /**
     * @param ***REMOVED***string|buffer|object} RotateEncryptionKeyOptions Configuration options
     *     for File#rotateEncryptionKey().
     * If a string or Buffer is provided, it is interpreted as an AES-256,
     * customer-supplied encryption key. If you'd like to use a Cloud KMS key
     * name, you must specify an options object with the property name:
     * `kmsKeyName`.
     * @param ***REMOVED***string|buffer} [options.encryptionKey] An AES-256 encryption key.
     * @param ***REMOVED***string} [options.kmsKeyName] A Cloud KMS key name.
     */
    /**
     * This method allows you to update the encryption key associated with this
     * file.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/encryption#customer-supplied| Customer-supplied Encryption Keys}
     *
     * @param ***REMOVED***RotateEncryptionKeyOptions} [options] - Configuration options.
     * @param ***REMOVED***RotateEncryptionKeyCallback} [callback]
     * @returns ***REMOVED***Promise<File>}
     *
     * @example <caption>include:samples/encryption.js</caption>
     * region_tag:storage_rotate_encryption_key
     * Example of rotating the encryption key for this file:
     */
    rotateEncryptionKey(optionsOrCallback, callback) ***REMOVED***
        var _a;
        callback =
            typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
        let options = ***REMOVED***};
        if (typeof optionsOrCallback === 'string' ||
            optionsOrCallback instanceof Buffer) ***REMOVED***
            options = ***REMOVED***
                encryptionKey: optionsOrCallback,
          ***REMOVED***;
      ***REMOVED***
        else if (typeof optionsOrCallback === 'object') ***REMOVED***
            options = optionsOrCallback;
      ***REMOVED***
        const newFile = this.bucket.file(this.id, options);
        const copyOptions = ((_a = options.preconditionOpts) === null || _a === void 0 ? void 0 : _a.ifGenerationMatch) !== undefined
            ? ***REMOVED*** preconditionOpts: options.preconditionOpts }
            : ***REMOVED***};
        this.copy(newFile, copyOptions, callback);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***object} SaveOptions
     * @extends CreateWriteStreamOptions
     */
    /**
     * @callback SaveCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     */
    /**
     * Write strings or buffers to a file.
     *
     * *This is a convenience method which wraps ***REMOVED***@link File#createWriteStream}.*
     * To upload arbitrary data to a file, please use ***REMOVED***@link File#createWriteStream} directly.
     *
     * Resumable uploads are automatically enabled and must be shut off explicitly
     * by setting `options.resumable` to `false`.
     *
     * Multipart uploads with retryable error codes will be retried 3 times with exponential backoff.
     *
     * <p class="notice">
     *   There is some overhead when using a resumable upload that can cause
     *   noticeable performance degradation while uploading a series of small
     * files. When uploading files less than 10MB, it is recommended that the
     * resumable feature is disabled.
     * </p>
     *
     * @param ***REMOVED***string | Buffer} data The data to write to a file.
     * @param ***REMOVED***SaveOptions} [options] See ***REMOVED***@link File#createWriteStream}'s `options`
     *     parameter.
     * @param ***REMOVED***SaveCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     *
     * const file = myBucket.file('my-file');
     * const contents = 'This is the contents of the file.';
     *
     * file.save(contents, function(err) ***REMOVED***
     *   if (!err) ***REMOVED***
     *     // File written successfully.
     * ***REMOVED***
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.save(contents).then(function() ***REMOVED***});
     * ```
     */
    save(data, optionsOrCallback, callback) ***REMOVED***
        // tslint:enable:no-any
        callback =
            typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        let maxRetries = this.storage.retryOptions.maxRetries;
        if (!this.shouldRetryBasedOnPreconditionAndIdempotencyStrat(options === null || options === void 0 ? void 0 : options.preconditionOpts)) ***REMOVED***
            maxRetries = 0;
      ***REMOVED***
        const returnValue = retry(async (bail) => ***REMOVED***
            await new Promise((resolve, reject) => ***REMOVED***
                if (maxRetries === 0) ***REMOVED***
                    this.storage.retryOptions.autoRetry = false;
              ***REMOVED***
                const writable = this.createWriteStream(options)
                    .on('error', err => ***REMOVED***
                    if (this.storage.retryOptions.autoRetry &&
                        this.storage.retryOptions.retryableErrorFn(err)) ***REMOVED***
                        return reject(err);
                  ***REMOVED***
                    else ***REMOVED***
                        return bail(err);
                  ***REMOVED***
              ***REMOVED***)
                    .on('finish', () => ***REMOVED***
                    return resolve();
              ***REMOVED***);
                if (options.onUploadProgress) ***REMOVED***
                    writable.on('progress', options.onUploadProgress);
              ***REMOVED***
                writable.end(data);
          ***REMOVED***);
      ***REMOVED***, ***REMOVED***
            retries: maxRetries,
            factor: this.storage.retryOptions.retryDelayMultiplier,
            maxTimeout: this.storage.retryOptions.maxRetryDelay * 1000,
            maxRetryTime: this.storage.retryOptions.totalTimeout * 1000, //convert to milliseconds
      ***REMOVED***);
        if (!callback) ***REMOVED***
            return returnValue;
      ***REMOVED***
        else ***REMOVED***
            return returnValue
                .then(() => ***REMOVED***
                if (callback) ***REMOVED***
                    return callback();
              ***REMOVED***
          ***REMOVED***)
                .catch(callback);
      ***REMOVED***
  ***REMOVED***
    setMetadata(metadata, optionsOrCallback, cb) ***REMOVED***
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        cb =
            typeof optionsOrCallback === 'function'
                ? optionsOrCallback
                : cb;
        this.disableAutoRetryConditionallyIdempotent_(this.methods.setMetadata, bucket_1.AvailableServiceObjectMethods.setMetadata, options);
        super
            .setMetadata(metadata, options)
            .then(resp => cb(null, ...resp))
            .catch(cb)
            .finally(() => ***REMOVED***
            this.storage.retryOptions.autoRetry = this.instanceRetryValue;
      ***REMOVED***);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} SetStorageClassResponse
     * @property ***REMOVED***object} 0 The full API response.
     */
    /**
     * @typedef ***REMOVED***object} SetStorageClassOptions Configuration options for File#setStorageClass().
     * @property ***REMOVED***string} [userProject] The ID of the project which will be
     *     billed for the request.
     */
    /**
     * @callback SetStorageClassCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***object} apiResponse The full API response.
     */
    /**
     * Set the storage class for this file.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/per-object-storage-class| Per-Object Storage Class}
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/storage-classes| Storage Classes}
     *
     * @param ***REMOVED***string} storageClass The new storage class. (`standard`,
     *     `nearline`, `coldline`, or `archive`)
     *     **Note:** The storage classes `multi_regional` and `regional`
     *     are now legacy and will be deprecated in the future.
     * @param ***REMOVED***SetStorageClassOptions} [options] Configuration options.
     * @param ***REMOVED***string} [options.userProject] The ID of the project which will be
     *     billed for the request.
     * @param ***REMOVED***SetStorageClassCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<SetStorageClassResponse>}
     *
     * @example
     * ```
     * file.setStorageClass('nearline', function(err, apiResponse) ***REMOVED***
     *   if (err) ***REMOVED***
     *     // Error handling omitted.
     * ***REMOVED***
     *
     *   // The storage class was updated successfully.
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * file.setStorageClass('nearline').then(function() ***REMOVED***});
     * ```
     */
    setStorageClass(storageClass, optionsOrCallback, callback) ***REMOVED***
        callback =
            typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        const req = extend(true, ***REMOVED***}, options);
        // In case we get input like `storageClass`, convert to `storage_class`.
        req.storageClass = storageClass
            .replace(/-/g, '_')
            .replace(/([a-z])([A-Z])/g, (_, low, up) => ***REMOVED***
            return low + '_' + up;
      ***REMOVED***)
            .toUpperCase();
        this.copy(this, req, (err, file, apiResponse) => ***REMOVED***
            if (err) ***REMOVED***
                callback(err, apiResponse);
                return;
          ***REMOVED***
            this.metadata = file.metadata;
            callback(null, apiResponse);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Set a user project to be billed for all requests made from this File
     * object.
     *
     * @param ***REMOVED***string} userProject The user project.
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const bucket = storage.bucket('albums');
     * const file = bucket.file('my-file');
     *
     * file.setUserProject('grape-spaceship-123');
     * ```
     */
    setUserProject(userProject) ***REMOVED***
        this.bucket.setUserProject.call(this, userProject);
  ***REMOVED***
    /**
     * This creates a resumable-upload upload stream.
     *
     * @param ***REMOVED***Duplexify} stream - Duplexify stream of data to pipe to the file.
     * @param ***REMOVED***object=} options - Configuration object.
     *
     * @private
     */
    startResumableUpload_(dup, options) ***REMOVED***
        options = extend(true, ***REMOVED***
            metadata: ***REMOVED***},
      ***REMOVED***, options);
        const retryOptions = this.storage.retryOptions;
        if (!this.shouldRetryBasedOnPreconditionAndIdempotencyStrat(options === null || options === void 0 ? void 0 : options.preconditionOpts)) ***REMOVED***
            retryOptions.autoRetry = false;
      ***REMOVED***
        const uploadStream = resumableUpload.upload(***REMOVED***
            authClient: this.storage.authClient,
            apiEndpoint: this.storage.apiEndpoint,
            bucket: this.bucket.name,
            customRequestOptions: this.getRequestInterceptors().reduce((reqOpts, interceptorFn) => interceptorFn(reqOpts), ***REMOVED***}),
            file: this.name,
            generation: this.generation,
            key: this.encryptionKey,
            kmsKeyName: this.kmsKeyName,
            metadata: options.metadata,
            offset: options.offset,
            predefinedAcl: options.predefinedAcl,
            private: options.private,
            public: options.public,
            uri: options.uri,
            userProject: options.userProject || this.userProject,
            retryOptions: ***REMOVED*** ...retryOptions },
            params: (options === null || options === void 0 ? void 0 : options.preconditionOpts) || this.instancePreconditionOpts,
            chunkSize: options === null || options === void 0 ? void 0 : options.chunkSize,
      ***REMOVED***);
        uploadStream
            .on('response', resp => ***REMOVED***
            dup.emit('response', resp);
      ***REMOVED***)
            .on('metadata', metadata => ***REMOVED***
            this.metadata = metadata;
            dup.emit('metadata');
      ***REMOVED***)
            .on('finish', () => ***REMOVED***
            dup.emit('complete');
      ***REMOVED***)
            .on('progress', evt => dup.emit('progress', evt));
        dup.setWritable(uploadStream);
        this.storage.retryOptions.autoRetry = this.instanceRetryValue;
  ***REMOVED***
    /**
     * Takes a readable stream and pipes it to a remote file. Unlike
     * `startResumableUpload_`, which uses the resumable upload technique, this
     * method uses a simple upload (all or nothing).
     *
     * @param ***REMOVED***Duplexify} dup - Duplexify stream of data to pipe to the file.
     * @param ***REMOVED***object=} options - Configuration object.
     *
     * @private
     */
    startSimpleUpload_(dup, options) ***REMOVED***
        options = extend(true, ***REMOVED***
            metadata: ***REMOVED***},
      ***REMOVED***, options);
        const apiEndpoint = this.storage.apiEndpoint;
        const bucketName = this.bucket.name;
        const uri = `$***REMOVED***apiEndpoint}/upload/storage/v1/b/$***REMOVED***bucketName}/o`;
        const reqOpts = ***REMOVED***
            qs: ***REMOVED***
                name: this.name,
          ***REMOVED***,
            uri: uri,
      ***REMOVED***;
        if (this.generation !== undefined) ***REMOVED***
            reqOpts.qs.ifGenerationMatch = this.generation;
      ***REMOVED***
        if (this.kmsKeyName !== undefined) ***REMOVED***
            reqOpts.qs.kmsKeyName = this.kmsKeyName;
      ***REMOVED***
        if (typeof options.timeout === 'number') ***REMOVED***
            reqOpts.timeout = options.timeout;
      ***REMOVED***
        if (options.userProject || this.userProject) ***REMOVED***
            reqOpts.qs.userProject = options.userProject || this.userProject;
      ***REMOVED***
        if (options.predefinedAcl) ***REMOVED***
            reqOpts.qs.predefinedAcl = options.predefinedAcl;
      ***REMOVED***
        else if (options.private) ***REMOVED***
            reqOpts.qs.predefinedAcl = 'private';
      ***REMOVED***
        else if (options.public) ***REMOVED***
            reqOpts.qs.predefinedAcl = 'publicRead';
      ***REMOVED***
        Object.assign(reqOpts.qs, this.instancePreconditionOpts, options.preconditionOpts);
        nodejs_common_1.util.makeWritableStream(dup, ***REMOVED***
            makeAuthenticatedRequest: (reqOpts) => ***REMOVED***
                this.request(reqOpts, (err, body, resp) => ***REMOVED***
                    if (err) ***REMOVED***
                        dup.destroy(err);
                        return;
                  ***REMOVED***
                    this.metadata = body;
                    dup.emit('metadata', body);
                    dup.emit('response', resp);
                    dup.emit('complete');
              ***REMOVED***);
          ***REMOVED***,
            metadata: options.metadata,
            request: reqOpts,
      ***REMOVED***);
  ***REMOVED***
    disableAutoRetryConditionallyIdempotent_(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    coreOpts, methodType, localPreconditionOptions) ***REMOVED***
        var _a, _b;
        if ((typeof coreOpts === 'object' &&
            ((_b = (_a = coreOpts === null || coreOpts === void 0 ? void 0 : coreOpts.reqOpts) === null || _a === void 0 ? void 0 : _a.qs) === null || _b === void 0 ? void 0 : _b.ifGenerationMatch) === undefined &&
            (localPreconditionOptions === null || localPreconditionOptions === void 0 ? void 0 : localPreconditionOptions.ifGenerationMatch) === undefined &&
            (methodType === bucket_1.AvailableServiceObjectMethods.setMetadata ||
                methodType === bucket_1.AvailableServiceObjectMethods.delete) &&
            this.storage.retryOptions.idempotencyStrategy ===
                storage_1.IdempotencyStrategy.RetryConditional) ||
            this.storage.retryOptions.idempotencyStrategy ===
                storage_1.IdempotencyStrategy.RetryNever) ***REMOVED***
            this.storage.retryOptions.autoRetry = false;
      ***REMOVED***
  ***REMOVED***
    async getBufferFromReadable(readable) ***REMOVED***
        const buf = [];
        for await (const chunk of readable) ***REMOVED***
            buf.push(chunk);
      ***REMOVED***
        return Buffer.concat(buf);
  ***REMOVED***
}
exports.File = File;
_File_instances = new WeakSet(), _File_validateIntegrity = 
/**
 *
 * @param hashCalculatingStream
 * @param verify
 * @returns ***REMOVED***boolean} Returns `true` if valid, throws with error otherwise
 */
async function _File_validateIntegrity(hashCalculatingStream, verify = ***REMOVED***}) ***REMOVED***
    const metadata = this.metadata;
    // If we're doing validation, assume the worst
    let dataMismatch = !!(verify.crc32c || verify.md5);
    if (verify.crc32c && metadata.crc32c) ***REMOVED***
        dataMismatch = !hashCalculatingStream.test('crc32c', metadata.crc32c);
  ***REMOVED***
    if (verify.md5 && metadata.md5Hash) ***REMOVED***
        dataMismatch = !hashCalculatingStream.test('md5', metadata.md5Hash);
  ***REMOVED***
    if (dataMismatch) ***REMOVED***
        const errors = [];
        let code = '';
        let message = '';
        try ***REMOVED***
            await this.delete();
            if (verify.md5 && !metadata.md5Hash) ***REMOVED***
                code = 'MD5_NOT_AVAILABLE';
                message = FileExceptionMessages.MD5_NOT_AVAILABLE;
          ***REMOVED***
            else ***REMOVED***
                code = 'FILE_NO_UPLOAD';
                message = FileExceptionMessages.UPLOAD_MISMATCH;
          ***REMOVED***
      ***REMOVED***
        catch (e) ***REMOVED***
            const error = e;
            code = 'FILE_NO_UPLOAD_DELETE';
            message = `$***REMOVED***FileExceptionMessages.UPLOAD_MISMATCH_DELETE_FAIL}$***REMOVED***error.message}`;
            errors.push(error);
      ***REMOVED***
        const error = new RequestError(message);
        error.code = code;
        error.errors = errors;
        throw error;
  ***REMOVED***
    return true;
};
/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
(0, promisify_1.promisifyAll)(File, ***REMOVED***
    exclude: [
        'cloudStorageURI',
        'publicUrl',
        'request',
        'save',
        'setEncryptionKey',
        'shouldRetryBasedOnPreconditionAndIdempotencyStrat',
        'getBufferFromReadable',
    ],
});
