// @google-cloud/storage/build/src/iam.js
"use strict";
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.Iam = exports.IAMExceptionMessages = void 0;
const promisify_1 = require("@google-cloud/promisify");
const util_1 = require("./util");
var IAMExceptionMessages;
(function (IAMExceptionMessages) ***REMOVED***
    IAMExceptionMessages["POLICY_OBJECT_REQUIRED"] = "A policy object is required.";
    IAMExceptionMessages["PERMISSIONS_REQUIRED"] = "Permissions are required.";
})(IAMExceptionMessages = exports.IAMExceptionMessages || (exports.IAMExceptionMessages = ***REMOVED***}));
/**
 * Get and set IAM policies for your Cloud Storage bucket.
 *
 * See ***REMOVED***@link https://cloud.google.com/storage/docs/access-control/iam#short_title_iam_management| Cloud Storage IAM Management}
 * See ***REMOVED***@link https://cloud.google.com/iam/docs/granting-changing-revoking-access| Granting, Changing, and Revoking Access}
 * See ***REMOVED***@link https://cloud.google.com/iam/docs/understanding-roles| IAM Roles}
 *
 * @constructor Iam
 *
 * @param ***REMOVED***Bucket} bucket The parent instance.
 * @example
 * ```
 * const ***REMOVED***Storage} = require('@google-cloud/storage');
 * const storage = new Storage();
 * const bucket = storage.bucket('my-bucket');
 * // bucket.iam
 * ```
 */
class Iam ***REMOVED***
    constructor(bucket) ***REMOVED***
        this.request_ = bucket.request.bind(bucket);
        this.resourceId_ = 'buckets/' + bucket.getId();
  ***REMOVED***
    /**
     * @typedef ***REMOVED***object} GetPolicyOptions Requested options for IAM#getPolicy().
     * @property ***REMOVED***number} [requestedPolicyVersion] The version of IAM policies to
     *     request. If a policy with a condition is requested without setting
     *     this, the server will return an error. This must be set to a value
     *     of 3 to retrieve IAM policies containing conditions. This is to
     *     prevent client code that isn't aware of IAM conditions from
     *     interpreting and modifying policies incorrectly. The service might
     *     return a policy with version lower than the one that was requested,
     *     based on the feature syntax in the policy fetched.
     *     See ***REMOVED***@link https://cloud.google.com/iam/docs/policies#versions| IAM Policy versions}
     * @property ***REMOVED***string} [userProject] The ID of the project which will be
     *     billed for the request.
     */
    /**
     * @typedef ***REMOVED***array} GetPolicyResponse
     * @property ***REMOVED***Policy} 0 The policy.
     * @property ***REMOVED***object} 1 The full API response.
     */
    /**
     * @typedef ***REMOVED***object} Policy
     * @property ***REMOVED***PolicyBinding[]} policy.bindings Bindings associate members with roles.
     * @property ***REMOVED***string} [policy.etag] Etags are used to perform a read-modify-write.
     * @property ***REMOVED***number} [policy.version] The syntax schema version of the Policy.
     *      To set an IAM policy with conditional binding, this field must be set to
     *      3 or greater.
     *     See ***REMOVED***@link https://cloud.google.com/iam/docs/policies#versions| IAM Policy versions}
     */
    /**
     * @typedef ***REMOVED***object} PolicyBinding
     * @property ***REMOVED***string} role Role that is assigned to members.
     * @property ***REMOVED***string[]} members Specifies the identities requesting access for the bucket.
     * @property ***REMOVED***Expr} [condition] The condition that is associated with this binding.
     */
    /**
     * @typedef ***REMOVED***object} Expr
     * @property ***REMOVED***string} [title] An optional title for the expression, i.e. a
     *     short string describing its purpose. This can be used e.g. in UIs
     *     which allow to enter the expression.
     * @property ***REMOVED***string} [description] An optional description of the
     *     expression. This is a longer text which describes the expression,
     *     e.g. when hovered over it in a UI.
     * @property ***REMOVED***string} expression Textual representation of an expression in
     *     Common Expression Language syntax. The application context of the
     *     containing message determines which well-known feature set of CEL
     *     is supported.The condition that is associated with this binding.
     *
     * @see [Condition] https://cloud.google.com/storage/docs/access-control/iam#conditions
     */
    /**
     * Get the IAM policy.
     *
     * @param ***REMOVED***GetPolicyOptions} [options] Request options.
     * @param ***REMOVED***GetPolicyCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<GetPolicyResponse>}
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/getIamPolicy| Buckets: setIamPolicy API Documentation}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const bucket = storage.bucket('my-bucket');
     *
     * bucket.iam.getPolicy(
     *     ***REMOVED***requestedPolicyVersion: 3},
     *     function(err, policy, apiResponse) ***REMOVED***
     *
     *   ***REMOVED***,
     * );
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * bucket.iam.getPolicy(***REMOVED***requestedPolicyVersion: 3})
     *   .then(function(data) ***REMOVED***
     *     const policy = data[0];
     *     const apiResponse = data[1];
     * ***REMOVED***);
     *
     * ```
     * @example <caption>include:samples/iam.js</caption>
     * region_tag:storage_view_bucket_iam_members
     * Example of retrieving a bucket's IAM policy:
     */
    getPolicy(optionsOrCallback, callback) ***REMOVED***
        const ***REMOVED*** options, callback: cb } = (0, util_1.normalize)(optionsOrCallback, callback);
        const qs = ***REMOVED***};
        if (options.userProject) ***REMOVED***
            qs.userProject = options.userProject;
      ***REMOVED***
        if (options.requestedPolicyVersion !== null &&
            options.requestedPolicyVersion !== undefined) ***REMOVED***
            qs.optionsRequestedPolicyVersion = options.requestedPolicyVersion;
      ***REMOVED***
        this.request_(***REMOVED***
            uri: '/iam',
            qs,
      ***REMOVED***, cb);
  ***REMOVED***
    /**
     * Set the IAM policy.
     *
     * @throws ***REMOVED***Error} If no policy is provided.
     *
     * @param ***REMOVED***Policy} policy The policy.
     * @param ***REMOVED***SetPolicyOptions} [options] Configuration options.
     * @param ***REMOVED***SetPolicyCallback} callback Callback function.
     * @returns ***REMOVED***Promise<SetPolicyResponse>}
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/setIamPolicy| Buckets: setIamPolicy API Documentation}
     * See ***REMOVED***@link https://cloud.google.com/iam/docs/understanding-roles| IAM Roles}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const bucket = storage.bucket('my-bucket');
     *
     * const myPolicy = ***REMOVED***
     *   bindings: [
     *     ***REMOVED***
     *       role: 'roles/storage.admin',
     *       members:
     * ['serviceAccount:myotherproject@appspot.gserviceaccount.com']
     *   ***REMOVED***
     *   ]
     * };
     *
     * bucket.iam.setPolicy(myPolicy, function(err, policy, apiResponse) ***REMOVED***});
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * bucket.iam.setPolicy(myPolicy).then(function(data) ***REMOVED***
     *   const policy = data[0];
     *   const apiResponse = data[1];
     * });
     *
     * ```
     * @example <caption>include:samples/iam.js</caption>
     * region_tag:storage_add_bucket_iam_member
     * Example of adding to a bucket's IAM policy:
     *
     * @example <caption>include:samples/iam.js</caption>
     * region_tag:storage_remove_bucket_iam_member
     * Example of removing from a bucket's IAM policy:
     */
    setPolicy(policy, optionsOrCallback, callback) ***REMOVED***
        if (policy === null || typeof policy !== 'object') ***REMOVED***
            throw new Error(IAMExceptionMessages.POLICY_OBJECT_REQUIRED);
      ***REMOVED***
        const ***REMOVED*** options, callback: cb } = (0, util_1.normalize)(optionsOrCallback, callback);
        let maxRetries;
        if (policy.etag === undefined) ***REMOVED***
            maxRetries = 0;
      ***REMOVED***
        this.request_(***REMOVED***
            method: 'PUT',
            uri: '/iam',
            maxRetries,
            json: Object.assign(***REMOVED***
                resourceId: this.resourceId_,
          ***REMOVED***, policy),
            qs: options,
      ***REMOVED***, cb);
  ***REMOVED***
    /**
     * Test a set of permissions for a resource.
     *
     * @throws ***REMOVED***Error} If permissions are not provided.
     *
     * @param ***REMOVED***string|string[]} permissions The permission(s) to test for.
     * @param ***REMOVED***TestIamPermissionsOptions} [options] Configuration object.
     * @param ***REMOVED***TestIamPermissionsCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<TestIamPermissionsResponse>}
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/testIamPermissions| Buckets: testIamPermissions API Documentation}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const bucket = storage.bucket('my-bucket');
     *
     * //-
     * // Test a single permission.
     * //-
     * const test = 'storage.buckets.delete';
     *
     * bucket.iam.testPermissions(test, function(err, permissions, apiResponse) ***REMOVED***
     *   console.log(permissions);
     *   // ***REMOVED***
     *   //   "storage.buckets.delete": true
     *   // }
     * });
     *
     * //-
     * // Test several permissions at once.
     * //-
     * const tests = [
     *   'storage.buckets.delete',
     *   'storage.buckets.get'
     * ];
     *
     * bucket.iam.testPermissions(tests, function(err, permissions) ***REMOVED***
     *   console.log(permissions);
     *   // ***REMOVED***
     *   //   "storage.buckets.delete": false,
     *   //   "storage.buckets.get": true
     *   // }
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * bucket.iam.testPermissions(test).then(function(data) ***REMOVED***
     *   const permissions = data[0];
     *   const apiResponse = data[1];
     * });
     * ```
     */
    testPermissions(permissions, optionsOrCallback, callback) ***REMOVED***
        if (!Array.isArray(permissions) && typeof permissions !== 'string') ***REMOVED***
            throw new Error(IAMExceptionMessages.PERMISSIONS_REQUIRED);
      ***REMOVED***
        const ***REMOVED*** options, callback: cb } = (0, util_1.normalize)(optionsOrCallback, callback);
        const permissionsArray = Array.isArray(permissions)
            ? permissions
            : [permissions];
        const req = Object.assign(***REMOVED***
            permissions: permissionsArray,
      ***REMOVED***, options);
        this.request_(***REMOVED***
            uri: '/iam/testPermissions',
            qs: req,
            useQuerystring: true,
      ***REMOVED***, (err, resp) => ***REMOVED***
            if (err) ***REMOVED***
                cb(err, null, resp);
                return;
          ***REMOVED***
            const availablePermissions = resp.permissions;
            const permissionsHash = permissionsArray.reduce((acc, permission) => ***REMOVED***
                acc[permission] = availablePermissions.indexOf(permission) > -1;
                return acc;
          ***REMOVED***, ***REMOVED***});
            cb(null, permissionsHash, resp);
      ***REMOVED***);
  ***REMOVED***
}
exports.Iam = Iam;
/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
(0, promisify_1.promisifyAll)(Iam);
