// @google-cloud/storage/build/src/notification.js
"use strict";
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.Notification = void 0;
const nodejs_common_1 = require("./nodejs-common");
const promisify_1 = require("@google-cloud/promisify");
/**
 * The API-formatted resource description of the notification.
 *
 * Note: This is not guaranteed to be up-to-date when accessed. To get the
 * latest record, call the `getMetadata()` method.
 *
 * @name Notification#metadata
 * @type ***REMOVED***object}
 */
/**
 * A Notification object is created from your ***REMOVED***@link Bucket} object using
 * ***REMOVED***@link Bucket#notification}. Use it to interact with Cloud Pub/Sub
 * notifications.
 *
 * See ***REMOVED***@link https://cloud.google.com/storage/docs/pubsub-notifications| Cloud Pub/Sub Notifications for Google Cloud Storage}
 *
 * @class
 * @hideconstructor
 *
 * @param ***REMOVED***Bucket} bucket The bucket instance this notification is attached to.
 * @param ***REMOVED***string} id The ID of the notification.
 *
 * @example
 * ```
 * const ***REMOVED***Storage} = require('@google-cloud/storage');
 * const storage = new Storage();
 * const myBucket = storage.bucket('my-bucket');
 *
 * const notification = myBucket.notification('1');
 * ```
 */
class Notification extends nodejs_common_1.ServiceObject ***REMOVED***
    constructor(bucket, id) ***REMOVED***
        const methods = ***REMOVED***
            /**
             * Creates a notification subscription for the bucket.
             *
             * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/notifications/insert| Notifications: insert}
             * @method Notification#create
             *
             * @param ***REMOVED***Topic|string} topic The Cloud PubSub topic to which this
             * subscription publishes. If the project ID is omitted, the current
             * project ID will be used.
             *
             * Acceptable formats are:
             * - `projects/grape-spaceship-123/topics/my-topic`
             *
             * - `my-topic`
             * @param ***REMOVED***CreateNotificationRequest} [options] Metadata to set for
             *     the notification.
             * @param ***REMOVED***CreateNotificationCallback} [callback] Callback function.
             * @returns ***REMOVED***Promise<CreateNotificationResponse>}
             * @throws ***REMOVED***Error} If a valid topic is not provided.
             *
             * @example
             * ```
             * const ***REMOVED***Storage} = require('@google-cloud/storage');
             * const storage = new Storage();
             * const myBucket = storage.bucket('my-bucket');
             * const notification = myBucket.notification('1');
             *
             * notification.create(function(err, notification, apiResponse) ***REMOVED***
             *   if (!err) ***REMOVED***
             *     // The notification was created successfully.
             * ***REMOVED***
             * });
             *
             * //-
             * // If the callback is omitted, we'll return a Promise.
             * //-
             * notification.create().then(function(data) ***REMOVED***
             *   const notification = data[0];
             *   const apiResponse = data[1];
             * });
             * ```
             */
            create: true,
            /**
             * @typedef ***REMOVED***array} NotificationExistsResponse
             * @property ***REMOVED***boolean} 0 Whether the notification exists or not.
             */
            /**
             * @callback NotificationExistsCallback
             * @param ***REMOVED***?Error} err Request error, if any.
             * @param ***REMOVED***boolean} exists Whether the notification exists or not.
             */
            /**
             * Check if the notification exists.
             *
             * @method Notification#exists
             * @param ***REMOVED***NotificationExistsCallback} [callback] Callback function.
             * @returns ***REMOVED***Promise<NotificationExistsResponse>}
             *
             * @example
             * ```
             * const ***REMOVED***Storage} = require('@google-cloud/storage');
             * const storage = new Storage();
             * const myBucket = storage.bucket('my-bucket');
             * const notification = myBucket.notification('1');
             *
             * notification.exists(function(err, exists) ***REMOVED***});
             *
             * //-
             * // If the callback is omitted, we'll return a Promise.
             * //-
             * notification.exists().then(function(data) ***REMOVED***
             *   const exists = data[0];
             * });
             * ```
             */
            exists: true,
      ***REMOVED***;
        super(***REMOVED***
            parent: bucket,
            baseUrl: '/notificationConfigs',
            id: id.toString(),
            createMethod: bucket.createNotification.bind(bucket),
            methods,
      ***REMOVED***);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} DeleteNotificationResponse
     * @property ***REMOVED***object} 0 The full API response.
     */
    /**
     * Permanently deletes a notification subscription.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/notifications/delete| Notifications: delete API Documentation}
     *
     * @param ***REMOVED***object} [options] Configuration options.
     * @param ***REMOVED***string} [options.userProject] The ID of the project which will be
     *     billed for the request.
     * @param ***REMOVED***DeleteNotificationCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<DeleteNotificationResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     * const notification = myBucket.notification('1');
     *
     * notification.delete(function(err, apiResponse) ***REMOVED***});
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * notification.delete().then(function(data) ***REMOVED***
     *   const apiResponse = data[0];
     * });
     *
     * ```
     * @example <caption>include:samples/deleteNotification.js</caption>
     * region_tag:storage_delete_bucket_notification
     * Another example:
     */
    delete(optionsOrCallback, callback) ***REMOVED***
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        callback =
            typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
        this.request(***REMOVED***
            method: 'DELETE',
            uri: '',
            qs: options,
      ***REMOVED***, callback || nodejs_common_1.util.noop);
  ***REMOVED***
    /**
     * Get a notification and its metadata if it exists.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/notifications/get| Notifications: get API Documentation}
     *
     * @param ***REMOVED***object} [options] Configuration options.
     *     See ***REMOVED***@link Bucket#createNotification} for create options.
     * @param ***REMOVED***boolean} [options.autoCreate] Automatically create the object if
     *     it does not exist. Default: `false`.
     * @param ***REMOVED***string} [options.userProject] The ID of the project which will be
     *     billed for the request.
     * @param ***REMOVED***GetNotificationCallback} [callback] Callback function.
     * @return ***REMOVED***Promise<GetNotificationCallback>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     * const notification = myBucket.notification('1');
     *
     * notification.get(function(err, notification, apiResponse) ***REMOVED***
     *   // `notification.metadata` has been populated.
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * notification.get().then(function(data) ***REMOVED***
     *   const notification = data[0];
     *   const apiResponse = data[1];
     * });
     * ```
     */
    get(optionsOrCallback, callback) ***REMOVED***
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        callback =
            typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
        const autoCreate = options.autoCreate;
        delete options.autoCreate;
        const onCreate = (err, notification, apiResponse) => ***REMOVED***
            if (err) ***REMOVED***
                if (err.code === 409) ***REMOVED***
                    this.get(options, callback);
                    return;
              ***REMOVED***
                callback(err, null, apiResponse);
                return;
          ***REMOVED***
            callback(null, notification, apiResponse);
      ***REMOVED***;
        this.getMetadata(options, (err, metadata) => ***REMOVED***
            if (err) ***REMOVED***
                if (err.code === 404 && autoCreate) ***REMOVED***
                    const args = [];
                    if (Object.keys(options).length > 0) ***REMOVED***
                        args.push(options);
                  ***REMOVED***
                    args.push(onCreate);
                    // eslint-disable-next-line
                    this.create.apply(this, args);
                    return;
              ***REMOVED***
                callback(err, null, metadata);
                return;
          ***REMOVED***
            callback(null, this, metadata);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Get the notification's metadata.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/notifications/get| Notifications: get API Documentation}
     *
     * @param ***REMOVED***object} [options] Configuration options.
     * @param ***REMOVED***string} [options.userProject] The ID of the project which will be
     *     billed for the request.
     * @param ***REMOVED***GetNotificationMetadataCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<GetNotificationMetadataResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const myBucket = storage.bucket('my-bucket');
     * const notification = myBucket.notification('1');
     *
     * notification.getMetadata(function(err, metadata, apiResponse) ***REMOVED***});
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * notification.getMetadata().then(function(data) ***REMOVED***
     *   const metadata = data[0];
     *   const apiResponse = data[1];
     * });
     *
     * ```
     * @example <caption>include:samples/getMetadataNotifications.js</caption>
     * region_tag:storage_print_pubsub_bucket_notification
     * Another example:
     */
    getMetadata(optionsOrCallback, callback) ***REMOVED***
        const options = typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***};
        callback =
            typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;
        this.request(***REMOVED***
            uri: '',
            qs: options,
      ***REMOVED***, (err, resp) => ***REMOVED***
            if (err) ***REMOVED***
                callback(err, null, resp);
                return;
          ***REMOVED***
            this.metadata = resp;
            callback(null, this.metadata, resp);
      ***REMOVED***);
  ***REMOVED***
}
exports.Notification = Notification;
/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
(0, promisify_1.promisifyAll)(Notification);
