// @google-cloud/storage/build/src/storage.js
"use strict";
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.Storage = exports.RETRYABLE_ERR_FN_DEFAULT = exports.MAX_RETRY_DELAY_DEFAULT = exports.TOTAL_TIMEOUT_DEFAULT = exports.RETRY_DELAY_MULTIPLIER_DEFAULT = exports.MAX_RETRY_DEFAULT = exports.AUTO_RETRY_DEFAULT = exports.PROTOCOL_REGEX = exports.StorageExceptionMessages = exports.ExceptionMessages = exports.IdempotencyStrategy = void 0;
const nodejs_common_1 = require("./nodejs-common");
const paginator_1 = require("@google-cloud/paginator");
const promisify_1 = require("@google-cloud/promisify");
const stream_1 = require("stream");
const bucket_1 = require("./bucket");
const channel_1 = require("./channel");
const file_1 = require("./file");
const util_1 = require("./util");
const hmacKey_1 = require("./hmacKey");
const crc32c_1 = require("./crc32c");
var IdempotencyStrategy;
(function (IdempotencyStrategy) ***REMOVED***
    IdempotencyStrategy[IdempotencyStrategy["RetryAlways"] = 0] = "RetryAlways";
    IdempotencyStrategy[IdempotencyStrategy["RetryConditional"] = 1] = "RetryConditional";
    IdempotencyStrategy[IdempotencyStrategy["RetryNever"] = 2] = "RetryNever";
})(IdempotencyStrategy = exports.IdempotencyStrategy || (exports.IdempotencyStrategy = ***REMOVED***}));
var ExceptionMessages;
(function (ExceptionMessages) ***REMOVED***
    ExceptionMessages["EXPIRATION_DATE_INVALID"] = "The expiration date provided was invalid.";
    ExceptionMessages["EXPIRATION_DATE_PAST"] = "An expiration date cannot be in the past.";
    ExceptionMessages["INVALID_ACTION"] = "The action is not provided or invalid.";
})(ExceptionMessages = exports.ExceptionMessages || (exports.ExceptionMessages = ***REMOVED***}));
var StorageExceptionMessages;
(function (StorageExceptionMessages) ***REMOVED***
    StorageExceptionMessages["BUCKET_NAME_REQUIRED"] = "A bucket name is needed to use Cloud Storage.";
    StorageExceptionMessages["BUCKET_NAME_REQUIRED_CREATE"] = "A name is required to create a bucket.";
    StorageExceptionMessages["HMAC_SERVICE_ACCOUNT"] = "The first argument must be a service account email to create an HMAC key.";
    StorageExceptionMessages["HMAC_ACCESS_ID"] = "An access ID is needed to create an HmacKey object.";
})(StorageExceptionMessages = exports.StorageExceptionMessages || (exports.StorageExceptionMessages = ***REMOVED***}));
exports.PROTOCOL_REGEX = /^(\w*):\/\//;
/**
 * Default behavior: Automatically retry retriable server errors.
 *
 * @const ***REMOVED***boolean}
 */
exports.AUTO_RETRY_DEFAULT = true;
/**
 * Default behavior: Only attempt to retry retriable errors 3 times.
 *
 * @const ***REMOVED***number}
 */
exports.MAX_RETRY_DEFAULT = 3;
/**
 * Default behavior: Wait twice as long as previous retry before retrying.
 *
 * @const ***REMOVED***number}
 */
exports.RETRY_DELAY_MULTIPLIER_DEFAULT = 2;
/**
 * Default behavior: If the operation doesn't succeed after 600 seconds,
 *  stop retrying.
 *
 * @const ***REMOVED***number}
 */
exports.TOTAL_TIMEOUT_DEFAULT = 600;
/**
 * Default behavior: Wait no more than 64 seconds between retries.
 *
 * @const ***REMOVED***number}
 */
exports.MAX_RETRY_DELAY_DEFAULT = 64;
/**
 * Default behavior: Retry conditionally idempotent operations if correct preconditions are set.
 *
 * @const ***REMOVED***enum}
 * @private
 */
const IDEMPOTENCY_STRATEGY_DEFAULT = IdempotencyStrategy.RetryConditional;
/**
 * Returns true if the API request should be retried, given the error that was
 * given the first time the request was attempted.
 * @const
 * @param ***REMOVED***error} err - The API error to check if it is appropriate to retry.
 * @return ***REMOVED***boolean} True if the API request should be retried, false otherwise.
 */
const RETRYABLE_ERR_FN_DEFAULT = function (err) ***REMOVED***
    var _a;
    const isConnectionProblem = (reason) => ***REMOVED***
        return (reason.includes('eai_again') || // DNS lookup error
            reason === 'econnreset' ||
            reason === 'unexpected connection closure' ||
            reason === 'epipe');
  ***REMOVED***;
    if (err) ***REMOVED***
        if ([408, 429, 500, 502, 503, 504].indexOf(err.code) !== -1) ***REMOVED***
            return true;
      ***REMOVED***
        if (typeof err.code === 'string') ***REMOVED***
            const reason = err.code.toLowerCase();
            if (isConnectionProblem(reason)) ***REMOVED***
                return true;
          ***REMOVED***
      ***REMOVED***
        if (err.errors) ***REMOVED***
            for (const e of err.errors) ***REMOVED***
                const reason = (_a = e === null || e === void 0 ? void 0 : e.reason) === null || _a === void 0 ? void 0 : _a.toString().toLowerCase();
                if (reason && isConnectionProblem(reason)) ***REMOVED***
                    return true;
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    return false;
};
exports.RETRYABLE_ERR_FN_DEFAULT = RETRYABLE_ERR_FN_DEFAULT;
/*! Developer Documentation
 *
 * Invoke this method to create a new Storage object bound with pre-determined
 * configuration options. For each object that can be created (e.g., a bucket),
 * there is an equivalent static and instance method. While they are classes,
 * they can be instantiated without use of the `new` keyword.
 */
/**
 * Cloud Storage uses access control lists (ACLs) to manage object and
 * bucket access. ACLs are the mechanism you use to share objects with other
 * users and allow other users to access your buckets and objects.
 *
 * This object provides constants to refer to the three permission levels that
 * can be granted to an entity:
 *
 *   - `gcs.acl.OWNER_ROLE` - ("OWNER")
 *   - `gcs.acl.READER_ROLE` - ("READER")
 *   - `gcs.acl.WRITER_ROLE` - ("WRITER")
 *
 * See ***REMOVED***@link https://cloud.google.com/storage/docs/access-control/lists| About Access Control Lists}
 *
 * @name Storage#acl
 * @type ***REMOVED***object}
 * @property ***REMOVED***string} OWNER_ROLE
 * @property ***REMOVED***string} READER_ROLE
 * @property ***REMOVED***string} WRITER_ROLE
 *
 * @example
 * ```
 * const ***REMOVED***Storage} = require('@google-cloud/storage');
 * const storage = new Storage();
 * const albums = storage.bucket('albums');
 *
 * //-
 * // Make all of the files currently in a bucket publicly readable.
 * //-
 * const options = ***REMOVED***
 *   entity: 'allUsers',
 *   role: storage.acl.READER_ROLE
 * };
 *
 * albums.acl.add(options, function(err, aclObject) ***REMOVED***});
 *
 * //-
 * // Make any new objects added to a bucket publicly readable.
 * //-
 * albums.acl.default.add(options, function(err, aclObject) ***REMOVED***});
 *
 * //-
 * // Grant a user ownership permissions to a bucket.
 * //-
 * albums.acl.add(***REMOVED***
 *   entity: 'user-useremail@example.com',
 *   role: storage.acl.OWNER_ROLE
 * }, function(err, aclObject) ***REMOVED***});
 *
 * //-
 * // If the callback is omitted, we'll return a Promise.
 * //-
 * albums.acl.add(options).then(function(data) ***REMOVED***
 *   const aclObject = data[0];
 *   const apiResponse = data[1];
 * });
 * ```
 */
/**
 * Get ***REMOVED***@link Bucket} objects for all of the buckets in your project as
 * a readable object stream.
 *
 * @method Storage#getBucketsStream
 * @param ***REMOVED***GetBucketsRequest} [query] Query object for listing buckets.
 * @returns ***REMOVED***ReadableStream} A readable stream that emits ***REMOVED***@link Bucket}
 *     instances.
 *
 * @example
 * ```
 * storage.getBucketsStream()
 *   .on('error', console.error)
 *   .on('data', function(bucket) ***REMOVED***
 *     // bucket is a Bucket object.
 * ***REMOVED***)
 *   .on('end', function() ***REMOVED***
 *     // All buckets retrieved.
 * ***REMOVED***);
 *
 * //-
 * // If you anticipate many results, you can end a stream early to prevent
 * // unnecessary processing and API requests.
 * //-
 * storage.getBucketsStream()
 *   .on('data', function(bucket) ***REMOVED***
 *     this.end();
 * ***REMOVED***);
 * ```
 */
/**
 * Get ***REMOVED***@link HmacKey} objects for all of the HMAC keys in the project in a
 * readable object stream.
 *
 * @method Storage#getHmacKeysStream
 * @param ***REMOVED***GetHmacKeysOptions} [options] Configuration options.
 * @returns ***REMOVED***ReadableStream} A readable stream that emits ***REMOVED***@link HmacKey}
 *     instances.
 *
 * @example
 * ```
 * storage.getHmacKeysStream()
 *   .on('error', console.error)
 *   .on('data', function(hmacKey) ***REMOVED***
 *     // hmacKey is an HmacKey object.
 * ***REMOVED***)
 *   .on('end', function() ***REMOVED***
 *     // All HmacKey retrieved.
 * ***REMOVED***);
 *
 * //-
 * // If you anticipate many results, you can end a stream early to prevent
 * // unnecessary processing and API requests.
 * //-
 * storage.getHmacKeysStream()
 *   .on('data', function(bucket) ***REMOVED***
 *     this.end();
 * ***REMOVED***);
 * ```
 */
/**
 * <h4>ACLs</h4>
 * Cloud Storage uses access control lists (ACLs) to manage object and
 * bucket access. ACLs are the mechanism you use to share files with other users
 * and allow other users to access your buckets and files.
 *
 * To learn more about ACLs, read this overview on
 * ***REMOVED***@link https://cloud.google.com/storage/docs/access-control| Access Control}.
 *
 * See ***REMOVED***@link https://cloud.google.com/storage/docs/overview| Cloud Storage overview}
 * See ***REMOVED***@link https://cloud.google.com/storage/docs/access-control| Access Control}
 *
 * @class
 */
class Storage extends nodejs_common_1.Service ***REMOVED***
    getBucketsStream() ***REMOVED***
        // placeholder body, overwritten in constructor
        return new stream_1.Readable();
  ***REMOVED***
    getHmacKeysStream() ***REMOVED***
        // placeholder body, overwritten in constructor
        return new stream_1.Readable();
  ***REMOVED***
    /**
     * @callback Crc32cGeneratorToStringCallback
     * A method returning the CRC32C as a base64-encoded string.
     *
     * @returns ***REMOVED***string}
     *
     * @example
     * Hashing the string 'data' should return 'rth90Q=='
     *
     * ```js
     * const buffer = Buffer.from('data');
     * crc32c.update(buffer);
     * crc32c.toString(); // 'rth90Q=='
     * ```
     **/
    /**
     * @callback Crc32cGeneratorValidateCallback
     * A method validating a base64-encoded CRC32C string.
     *
     * @param ***REMOVED***string} [value] base64-encoded CRC32C string to validate
     * @returns ***REMOVED***boolean}
     *
     * @example
     * Should return `true` if the value matches, `false` otherwise
     *
     * ```js
     * const buffer = Buffer.from('data');
     * crc32c.update(buffer);
     * crc32c.validate('DkjKuA=='); // false
     * crc32c.validate('rth90Q=='); // true
     * ```
     **/
    /**
     * @callback Crc32cGeneratorUpdateCallback
     * A method for passing `Buffer`s for CRC32C generation.
     *
     * @param ***REMOVED***Buffer} [data] data to update CRC32C value with
     * @returns ***REMOVED***undefined}
     *
     * @example
     * Hashing buffers from 'some ' and 'text\n'
     *
     * ```js
     * const buffer1 = Buffer.from('some ');
     * crc32c.update(buffer1);
     *
     * const buffer2 = Buffer.from('text\n');
     * crc32c.update(buffer2);
     *
     * crc32c.toString(); // 'DkjKuA=='
     * ```
     **/
    /**
     * @typedef ***REMOVED***object} CRC32CValidator
     * @property ***REMOVED***Crc32cGeneratorToStringCallback}
     * @property ***REMOVED***Crc32cGeneratorValidateCallback}
     * @property ***REMOVED***Crc32cGeneratorUpdateCallback}
     */
    /**
     * @callback Crc32cGeneratorCallback
     * @returns ***REMOVED***CRC32CValidator}
     */
    /**
     * @typedef ***REMOVED***object} StorageOptions
     * @property ***REMOVED***string} [projectId] The project ID from the Google Developer's
     *     Console, e.g. 'grape-spaceship-123'. We will also check the environment
     *     variable `GCLOUD_PROJECT` for your project ID. If your app is running
     * in an environment which supports ***REMOVED***@link
     * https://cloud.google.com/docs/authentication/production#providing_credentials_to_your_application
     * Application Default Credentials}, your project ID will be detected
     * automatically.
     * @property ***REMOVED***string} [keyFilename] Full path to the a .json, .pem, or .p12 key
     *     downloaded from the Google Developers Console. If you provide a path to
     * a JSON file, the `projectId` option above is not necessary. NOTE: .pem and
     *     .p12 require you to specify the `email` option as well.
     * @property ***REMOVED***string} [email] Account email address. Required when using a .pem
     *     or .p12 keyFilename.
     * @property ***REMOVED***object} [credentials] Credentials object.
     * @property ***REMOVED***string} [credentials.client_email]
     * @property ***REMOVED***string} [credentials.private_key]
     * @property ***REMOVED***object} [retryOptions] Options for customizing retries. Retriable server errors
     *     will be retried with exponential delay between them dictated by the formula
     *     max(maxRetryDelay, retryDelayMultiplier*retryNumber) until maxRetries or totalTimeout
     *     has been reached. Retries will only happen if autoRetry is set to true.
     * @property ***REMOVED***boolean} [retryOptions.autoRetry=true] Automatically retry requests if the
     *     response is related to rate limits or certain intermittent server
     * errors. We will exponentially backoff subsequent requests by default.
     * @property ***REMOVED***number} [retryOptions.retryDelayMultiplier = 2] the multiplier by which to
     *   increase the delay time between the completion of failed requests, and the
     *   initiation of the subsequent retrying request.
     * @property ***REMOVED***number} [retryOptions.totalTimeout = 600] The total time, starting from
     *  when the initial request is sent, after which an error will
     *   be returned, regardless of the retrying attempts made meanwhile.
     * @property ***REMOVED***number} [retryOptions.maxRetryDelay = 64] The maximum delay time between requests.
     *   When this value is reached, ``retryDelayMultiplier`` will no longer be used to
     *   increase delay time.
     * @property ***REMOVED***number} [retryOptions.maxRetries=3] Maximum number of automatic retries
     *     attempted before returning the error.
     * @property ***REMOVED***function} [retryOptions.retryableErrorFn] Function that returns true if a given
     *     error should be retried and false otherwise.
     * @property ***REMOVED***enum} [retryOptions.idempotencyStrategy=IdempotencyStrategy.RetryConditional] Enumeration
     *     controls how conditionally idempotent operations are retried. Possible values are: RetryAlways -
     *     will respect other retry settings and attempt to retry conditionally idempotent operations. RetryConditional -
     *     will retry conditionally idempotent operations if the correct preconditions are set. RetryNever - never
     *     retry a conditionally idempotent operation.
     * @property ***REMOVED***string} [userAgent] The value to be prepended to the User-Agent
     *     header in API requests.
     * @property ***REMOVED***object} [authClient] `AuthClient` or `GoogleAuth` client to reuse instead of creating a new one.
     * @property ***REMOVED***number} [timeout] The amount of time in milliseconds to wait per http request before timing out.
     * @property ***REMOVED***object[]} [interceptors_] Array of custom request interceptors to be returned in the order they were assigned.
     * @property ***REMOVED***string} [apiEndpoint = storage.google.com] The API endpoint of the service used to make requests.
     * @property ***REMOVED***boolean} [useAuthWithCustomEndpoint = false] Controls whether or not to use authentication when using a custom endpoint.
     * @property ***REMOVED***Crc32cGeneratorCallback} [callback] A function that generates a CRC32C Validator. Defaults to ***REMOVED***@link CRC32C}
     */
    /**
     * Constructs the Storage client.
     *
     * @example
     * Create a client that uses Application Default Credentials
     * (ADC)
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * ```
     *
     * @example
     * Create a client with explicit credentials
     * ```
     * const storage = new Storage(***REMOVED***
     *   projectId: 'your-project-id',
     *   keyFilename: '/path/to/keyfile.json'
     * });
     * ```
     *
     * @example
     * Create a client with credentials passed
     * by value as a JavaScript object
     * ```
     * const storage = new Storage(***REMOVED***
     *   projectId: 'your-project-id',
     *   credentials: ***REMOVED***
     *     type: 'service_account',
     *     project_id: 'xxxxxxx',
     *     private_key_id: 'xxxx',
     *     private_key:'-----BEGIN PRIVATE KEY-----xxxxxxx\n-----END PRIVATE KEY-----\n',
     *     client_email: 'xxxx',
     *     client_id: 'xxx',
     *     auth_uri: 'https://accounts.google.com/o/oauth2/auth',
     *     token_uri: 'https://oauth2.googleapis.com/token',
     *     auth_provider_x509_cert_url: 'https://www.googleapis.com/oauth2/v1/certs',
     *     client_x509_cert_url: 'xxx',
     *   ***REMOVED***
     * });
     * ```
     *
     * @example
     * Create a client with credentials passed
     * by loading a JSON file directly from disk
     * ```
     * const storage = new Storage(***REMOVED***
     *   projectId: 'your-project-id',
     *   credentials: require('/path/to-keyfile.json')
     * });
     * ```
     *
     * @example
     * Create a client with an `AuthClient` (e.g. `DownscopedClient`)
     * ```
     * const ***REMOVED***DownscopedClient} = require('google-auth-library');
     * const authClient = new DownscopedClient(***REMOVED***...});
     *
     * const storage = new Storage(***REMOVED***authClient});
     * ```
     *
     * Additional samples:
     * - https://github.com/googleapis/google-auth-library-nodejs#sample-usage-1
     * - https://github.com/googleapis/google-auth-library-nodejs/blob/main/samples/downscopedclient.js
     *
     * @param ***REMOVED***StorageOptions} [options] Configuration options.
     */
    constructor(options = ***REMOVED***}) ***REMOVED***
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p;
        let apiEndpoint = 'https://storage.googleapis.com';
        let customEndpoint = false;
        // Note: EMULATOR_HOST is an experimental configuration variable. Use apiEndpoint instead.
        const EMULATOR_HOST = process.env.STORAGE_EMULATOR_HOST;
        if (typeof EMULATOR_HOST === 'string') ***REMOVED***
            apiEndpoint = Storage.sanitizeEndpoint(EMULATOR_HOST);
            customEndpoint = true;
      ***REMOVED***
        if (options.apiEndpoint) ***REMOVED***
            apiEndpoint = Storage.sanitizeEndpoint(options.apiEndpoint);
            customEndpoint = true;
      ***REMOVED***
        options = Object.assign(***REMOVED***}, options, ***REMOVED*** apiEndpoint });
        // Note: EMULATOR_HOST is an experimental configuration variable. Use apiEndpoint instead.
        const baseUrl = EMULATOR_HOST || `$***REMOVED***options.apiEndpoint}/storage/v1`;
        const config = ***REMOVED***
            apiEndpoint: options.apiEndpoint,
            retryOptions: ***REMOVED***
                autoRetry: ((_a = options.retryOptions) === null || _a === void 0 ? void 0 : _a.autoRetry) !== undefined
                    ? (_b = options.retryOptions) === null || _b === void 0 ? void 0 : _b.autoRetry
                    : exports.AUTO_RETRY_DEFAULT,
                maxRetries: ((_c = options.retryOptions) === null || _c === void 0 ? void 0 : _c.maxRetries)
                    ? (_d = options.retryOptions) === null || _d === void 0 ? void 0 : _d.maxRetries
                    : exports.MAX_RETRY_DEFAULT,
                retryDelayMultiplier: ((_e = options.retryOptions) === null || _e === void 0 ? void 0 : _e.retryDelayMultiplier)
                    ? (_f = options.retryOptions) === null || _f === void 0 ? void 0 : _f.retryDelayMultiplier
                    : exports.RETRY_DELAY_MULTIPLIER_DEFAULT,
                totalTimeout: ((_g = options.retryOptions) === null || _g === void 0 ? void 0 : _g.totalTimeout)
                    ? (_h = options.retryOptions) === null || _h === void 0 ? void 0 : _h.totalTimeout
                    : exports.TOTAL_TIMEOUT_DEFAULT,
                maxRetryDelay: ((_j = options.retryOptions) === null || _j === void 0 ? void 0 : _j.maxRetryDelay)
                    ? (_k = options.retryOptions) === null || _k === void 0 ? void 0 : _k.maxRetryDelay
                    : exports.MAX_RETRY_DELAY_DEFAULT,
                retryableErrorFn: ((_l = options.retryOptions) === null || _l === void 0 ? void 0 : _l.retryableErrorFn)
                    ? (_m = options.retryOptions) === null || _m === void 0 ? void 0 : _m.retryableErrorFn
                    : exports.RETRYABLE_ERR_FN_DEFAULT,
                idempotencyStrategy: ((_o = options.retryOptions) === null || _o === void 0 ? void 0 : _o.idempotencyStrategy) !== undefined
                    ? (_p = options.retryOptions) === null || _p === void 0 ? void 0 : _p.idempotencyStrategy
                    : IDEMPOTENCY_STRATEGY_DEFAULT,
          ***REMOVED***,
            baseUrl,
            customEndpoint,
            useAuthWithCustomEndpoint: options === null || options === void 0 ? void 0 : options.useAuthWithCustomEndpoint,
            projectIdRequired: false,
            scopes: [
                'https://www.googleapis.com/auth/iam',
                'https://www.googleapis.com/auth/cloud-platform',
                'https://www.googleapis.com/auth/devstorage.full_control',
            ],
            packageJson: require('../../package.json'),
      ***REMOVED***;
        super(config, options);
        /**
         * Reference to ***REMOVED***@link Storage.acl}.
         *
         * @name Storage#acl
         * @see Storage.acl
         */
        this.acl = Storage.acl;
        this.crc32cGenerator =
            options.crc32cGenerator || crc32c_1.CRC32C_DEFAULT_VALIDATOR_GENERATOR;
        this.retryOptions = config.retryOptions;
        this.getBucketsStream = paginator_1.paginator.streamify('getBuckets');
        this.getHmacKeysStream = paginator_1.paginator.streamify('getHmacKeys');
  ***REMOVED***
    static sanitizeEndpoint(url) ***REMOVED***
        if (!exports.PROTOCOL_REGEX.test(url)) ***REMOVED***
            url = `https://$***REMOVED***url}`;
      ***REMOVED***
        return url.replace(/\/+$/, ''); // Remove trailing slashes
  ***REMOVED***
    /**
     * Get a reference to a Cloud Storage bucket.
     *
     * @param ***REMOVED***string} name Name of the bucket.
     * @param ***REMOVED***object} [options] Configuration object.
     * @param ***REMOVED***string} [options.kmsKeyName] A Cloud KMS key that will be used to
     *     encrypt objects inserted into this bucket, if no encryption method is
     *     specified.
     * @param ***REMOVED***string} [options.userProject] User project to be billed for all
     *     requests made from this Bucket object.
     * @returns ***REMOVED***Bucket}
     * @see Bucket
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const albums = storage.bucket('albums');
     * const photos = storage.bucket('photos');
     * ```
     */
    bucket(name, options) ***REMOVED***
        if (!name) ***REMOVED***
            throw new Error(StorageExceptionMessages.BUCKET_NAME_REQUIRED);
      ***REMOVED***
        return new bucket_1.Bucket(this, name, options);
  ***REMOVED***
    /**
     * Reference a channel to receive notifications about changes to your bucket.
     *
     * @param ***REMOVED***string} id The ID of the channel.
     * @param ***REMOVED***string} resourceId The resource ID of the channel.
     * @returns ***REMOVED***Channel}
     * @see Channel
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const channel = storage.channel('id', 'resource-id');
     * ```
     */
    channel(id, resourceId) ***REMOVED***
        return new channel_1.Channel(this, id, resourceId);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} CreateBucketResponse
     * @property ***REMOVED***Bucket} 0 The new ***REMOVED***@link Bucket}.
     * @property ***REMOVED***object} 1 The full API response.
     */
    /**
     * @callback CreateBucketCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***Bucket} bucket The new ***REMOVED***@link Bucket}.
     * @param ***REMOVED***object} apiResponse The full API response.
     */
    /**
     * Metadata to set for the bucket.
     *
     * @typedef ***REMOVED***object} CreateBucketRequest
     * @property ***REMOVED***boolean} [archive=false] Specify the storage class as Archive.
     * @property ***REMOVED***object} [autoclass.enabled=false] Specify whether Autoclass is
     *     enabled for the bucket.
     * @property ***REMOVED***boolean} [coldline=false] Specify the storage class as Coldline.
     * @property ***REMOVED***Cors[]} [cors=[]] Specify the CORS configuration to use.
     * @property ***REMOVED***CustomPlacementConfig} [customPlacementConfig=***REMOVED***}] Specify the bucket's regions for dual-region buckets.
     *     For more information, see ***REMOVED***@link https://cloud.google.com/storage/docs/locations| Bucket Locations}.
     * @property ***REMOVED***boolean} [dra=false] Specify the storage class as Durable Reduced
     *     Availability.
     * @property ***REMOVED***string} [location] Specify the bucket's location. If specifying
     *     a dual-region, the `customPlacementConfig` property should be set in conjunction.
     *     For more information, see ***REMOVED***@link https://cloud.google.com/storage/docs/locations| Bucket Locations}.
     * @property ***REMOVED***boolean} [multiRegional=false] Specify the storage class as
     *     Multi-Regional.
     * @property ***REMOVED***boolean} [nearline=false] Specify the storage class as Nearline.
     * @property ***REMOVED***boolean} [regional=false] Specify the storage class as Regional.
     * @property ***REMOVED***boolean} [requesterPays=false] **Early Access Testers Only**
     *     Force the use of the User Project metadata field to assign operational
     *     costs when an operation is made on a Bucket and its objects.
     * @property ***REMOVED***string} [rpo] For dual-region buckets, controls whether turbo
     *      replication is enabled (`ASYNC_TURBO`) or disabled (`DEFAULT`).
     * @property ***REMOVED***boolean} [standard=true] Specify the storage class as Standard.
     * @property ***REMOVED***string} [storageClass] The new storage class. (`standard`,
     *     `nearline`, `coldline`, or `archive`).
     *     **Note:** The storage classes `multi_regional`, `regional`, and
     *     `durable_reduced_availability` are now legacy and will be deprecated in
     *     the future.
     * @property ***REMOVED***Versioning} [versioning=undefined] Specify the versioning status.
     * @property ***REMOVED***string} [userProject] The ID of the project which will be billed
     *     for the request.
     */
    /**
     * Create a bucket.
     *
     * Cloud Storage uses a flat namespace, so you can't create a bucket with
     * a name that is already in use. For more information, see
     * ***REMOVED***@link https://cloud.google.com/storage/docs/bucketnaming.html#requirements| Bucket Naming Guidelines}.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/insert| Buckets: insert API Documentation}
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/storage-classes| Storage Classes}
     *
     * @param ***REMOVED***string} name Name of the bucket to create.
     * @param ***REMOVED***CreateBucketRequest} [metadata] Metadata to set for the bucket.
     * @param ***REMOVED***CreateBucketCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<CreateBucketResponse>}
     * @throws ***REMOVED***Error} If a name is not provided.
     * @see Bucket#create
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const callback = function(err, bucket, apiResponse) ***REMOVED***
     *   // `bucket` is a Bucket object.
     * };
     *
     * storage.createBucket('new-bucket', callback);
     *
     * //-
     * // Create a bucket in a specific location and region. <em>See the <a
     * // href="https://cloud.google.com/storage/docs/json_api/v1/buckets/insert">
     * // Official JSON API docs</a> for complete details on the `location`
     * option.
     * // </em>
     * //-
     * const metadata = ***REMOVED***
     *   location: 'US-CENTRAL1',
     *   regional: true
     * };
     *
     * storage.createBucket('new-bucket', metadata, callback);
     *
     * //-
     * // Create a bucket with a retention policy of 6 months.
     * //-
     * const metadata = ***REMOVED***
     *   retentionPolicy: ***REMOVED***
     *     retentionPeriod: 15780000 // 6 months in seconds.
     * ***REMOVED***
     * };
     *
     * storage.createBucket('new-bucket', metadata, callback);
     *
     * //-
     * // Enable versioning on a new bucket.
     * //-
     * const metadata = ***REMOVED***
     *   versioning: ***REMOVED***
     *     enabled: true
     * ***REMOVED***
     * };
     *
     * storage.createBucket('new-bucket', metadata, callback);
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * storage.createBucket('new-bucket').then(function(data) ***REMOVED***
     *   const bucket = data[0];
     *   const apiResponse = data[1];
     * });
     *
     * ```
     * @example <caption>include:samples/buckets.js</caption>
     * region_tag:storage_create_bucket
     * Another example:
     */
    createBucket(name, metadataOrCallback, callback) ***REMOVED***
        if (!name) ***REMOVED***
            throw new Error(StorageExceptionMessages.BUCKET_NAME_REQUIRED_CREATE);
      ***REMOVED***
        let metadata;
        if (!callback) ***REMOVED***
            callback = metadataOrCallback;
            metadata = ***REMOVED***};
      ***REMOVED***
        else ***REMOVED***
            metadata = metadataOrCallback;
      ***REMOVED***
        const body = ***REMOVED***
            ...metadata,
            name,
      ***REMOVED***;
        const storageClasses = ***REMOVED***
            archive: 'ARCHIVE',
            coldline: 'COLDLINE',
            dra: 'DURABLE_REDUCED_AVAILABILITY',
            multiRegional: 'MULTI_REGIONAL',
            nearline: 'NEARLINE',
            regional: 'REGIONAL',
            standard: 'STANDARD',
      ***REMOVED***;
        const storageClassKeys = Object.keys(storageClasses);
        for (const storageClass of storageClassKeys) ***REMOVED***
            if (body[storageClass]) ***REMOVED***
                if (metadata.storageClass && metadata.storageClass !== storageClass) ***REMOVED***
                    throw new Error(`Both \`$***REMOVED***storageClass}\` and \`storageClass\` were provided.`);
              ***REMOVED***
                body.storageClass = storageClasses[storageClass];
                delete body[storageClass];
          ***REMOVED***
      ***REMOVED***
        if (body.requesterPays) ***REMOVED***
            body.billing = ***REMOVED***
                requesterPays: body.requesterPays,
          ***REMOVED***;
            delete body.requesterPays;
      ***REMOVED***
        const query = ***REMOVED***
            project: this.projectId,
      ***REMOVED***;
        if (body.userProject) ***REMOVED***
            query.userProject = body.userProject;
            delete body.userProject;
      ***REMOVED***
        this.request(***REMOVED***
            method: 'POST',
            uri: '/b',
            qs: query,
            json: body,
      ***REMOVED***, (err, resp) => ***REMOVED***
            if (err) ***REMOVED***
                callback(err, null, resp);
                return;
          ***REMOVED***
            const bucket = this.bucket(name);
            bucket.metadata = resp;
            callback(null, bucket, resp);
      ***REMOVED***);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***object} CreateHmacKeyOptions
     * @property ***REMOVED***string} [projectId] The project ID of the project that owns
     *     the service account of the requested HMAC key. If not provided,
     *     the project ID used to instantiate the Storage client will be used.
     * @property ***REMOVED***string} [userProject] This parameter is currently ignored.
     */
    /**
     * @typedef ***REMOVED***object} HmacKeyMetadata
     * @property ***REMOVED***string} accessId The access id identifies which HMAC key was
     *     used to sign a request when authenticating with HMAC.
     * @property ***REMOVED***string} etag Used to perform a read-modify-write of the key.
     * @property ***REMOVED***string} id The resource name of the HMAC key.
     * @property ***REMOVED***string} projectId The project ID.
     * @property ***REMOVED***string} serviceAccountEmail The service account's email this
     *     HMAC key is created for.
     * @property ***REMOVED***string} state The state of this HMAC key. One of "ACTIVE",
     *     "INACTIVE" or "DELETED".
     * @property ***REMOVED***string} timeCreated The creation time of the HMAC key in
     *     RFC 3339 format.
     * @property ***REMOVED***string} [updated] The time this HMAC key was last updated in
     *     RFC 3339 format.
     */
    /**
     * @typedef ***REMOVED***array} CreateHmacKeyResponse
     * @property ***REMOVED***HmacKey} 0 The HmacKey instance created from API response.
     * @property ***REMOVED***string} 1 The HMAC key's secret used to access the XML API.
     * @property ***REMOVED***object} 3 The raw API response.
     */
    /**
     * @callback CreateHmacKeyCallback Callback function.
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***HmacKey} hmacKey The HmacKey instance created from API response.
     * @param ***REMOVED***string} secret The HMAC key's secret used to access the XML API.
     * @param ***REMOVED***object} apiResponse The raw API response.
     */
    /**
     * Create an HMAC key associated with an service account to authenticate
     * requests to the Cloud Storage XML API.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/authentication/hmackeys| HMAC keys documentation}
     *
     * @param ***REMOVED***string} serviceAccountEmail The service account's email address
     *     with which the HMAC key is created for.
     * @param ***REMOVED***CreateHmacKeyCallback} [callback] Callback function.
     * @return ***REMOVED***Promise<CreateHmacKeyResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('google-cloud/storage');
     * const storage = new Storage();
     *
     * // Replace with your service account's email address
     * const serviceAccountEmail =
     *   'my-service-account@appspot.gserviceaccount.com';
     *
     * storage.createHmacKey(serviceAccountEmail, function(err, hmacKey, secret) ***REMOVED***
     *   if (!err) ***REMOVED***
     *     // Securely store the secret for use with the XML API.
     * ***REMOVED***
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * storage.createHmacKey(serviceAccountEmail)
     *   .then((response) => ***REMOVED***
     *     const hmacKey = response[0];
     *     const secret = response[1];
     *     // Securely store the secret for use with the XML API.
     * ***REMOVED***);
     * ```
     */
    createHmacKey(serviceAccountEmail, optionsOrCb, cb) ***REMOVED***
        if (typeof serviceAccountEmail !== 'string') ***REMOVED***
            throw new Error(StorageExceptionMessages.HMAC_SERVICE_ACCOUNT);
      ***REMOVED***
        const ***REMOVED*** options, callback } = (0, util_1.normalize)(optionsOrCb, cb);
        const query = Object.assign(***REMOVED***}, options, ***REMOVED*** serviceAccountEmail });
        const projectId = query.projectId || this.projectId;
        delete query.projectId;
        this.request(***REMOVED***
            method: 'POST',
            uri: `/projects/$***REMOVED***projectId}/hmacKeys`,
            qs: query,
            maxRetries: 0, //explicitly set this value since this is a non-idempotent function
      ***REMOVED***, (err, resp) => ***REMOVED***
            if (err) ***REMOVED***
                callback(err, null, null, resp);
                return;
          ***REMOVED***
            const metadata = resp.metadata;
            const hmacKey = this.hmacKey(metadata.accessId, ***REMOVED***
                projectId: metadata.projectId,
          ***REMOVED***);
            hmacKey.metadata = resp.metadata;
            callback(null, hmacKey, resp.secret, resp);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Query object for listing buckets.
     *
     * @typedef ***REMOVED***object} GetBucketsRequest
     * @property ***REMOVED***boolean} [autoPaginate=true] Have pagination handled
     *     automatically.
     * @property ***REMOVED***number} [maxApiCalls] Maximum number of API calls to make.
     * @property ***REMOVED***number} [maxResults] Maximum number of items plus prefixes to
     *     return per call.
     *     Note: By default will handle pagination automatically
     *     if more than 1 page worth of results are requested per call.
     *     When `autoPaginate` is set to `false` the smaller of `maxResults`
     *     or 1 page of results will be returned per call.
     * @property ***REMOVED***string} [pageToken] A previously-returned page token
     *     representing part of the larger set of results to view.
     * @property ***REMOVED***string} [userProject] The ID of the project which will be billed
     *     for the request.
     */
    /**
     * @typedef ***REMOVED***array} GetBucketsResponse
     * @property ***REMOVED***Bucket[]} 0 Array of ***REMOVED***@link Bucket} instances.
     * @property ***REMOVED***object} 1 nextQuery A query object to receive more results.
     * @property ***REMOVED***object} 2 The full API response.
     */
    /**
     * @callback GetBucketsCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***Bucket[]} buckets Array of ***REMOVED***@link Bucket} instances.
     * @param ***REMOVED***object} nextQuery A query object to receive more results.
     * @param ***REMOVED***object} apiResponse The full API response.
     */
    /**
     * Get Bucket objects for all of the buckets in your project.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/buckets/list| Buckets: list API Documentation}
     *
     * @param ***REMOVED***GetBucketsRequest} [query] Query object for listing buckets.
     * @param ***REMOVED***GetBucketsCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<GetBucketsResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * storage.getBuckets(function(err, buckets) ***REMOVED***
     *   if (!err) ***REMOVED***
     *     // buckets is an array of Bucket objects.
     * ***REMOVED***
     * });
     *
     * //-
     * // To control how many API requests are made and page through the results
     * // manually, set `autoPaginate` to `false`.
     * //-
     * const callback = function(err, buckets, nextQuery, apiResponse) ***REMOVED***
     *   if (nextQuery) ***REMOVED***
     *     // More results exist.
     *     storage.getBuckets(nextQuery, callback);
     * ***REMOVED***
     *
     *   // The `metadata` property is populated for you with the metadata at the
     *   // time of fetching.
     *   buckets[0].metadata;
     *
     *   // However, in cases where you are concerned the metadata could have
     *   // changed, use the `getMetadata` method.
     *   buckets[0].getMetadata(function(err, metadata, apiResponse) ***REMOVED***});
     * };
     *
     * storage.getBuckets(***REMOVED***
     *   autoPaginate: false
     * }, callback);
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * storage.getBuckets().then(function(data) ***REMOVED***
     *   const buckets = data[0];
     * });
     *
     * ```
     * @example <caption>include:samples/buckets.js</caption>
     * region_tag:storage_list_buckets
     * Another example:
     */
    getBuckets(optionsOrCallback, cb) ***REMOVED***
        const ***REMOVED*** options, callback } = (0, util_1.normalize)(optionsOrCallback, cb);
        options.project = options.project || this.projectId;
        this.request(***REMOVED***
            uri: '/b',
            qs: options,
      ***REMOVED***, (err, resp) => ***REMOVED***
            if (err) ***REMOVED***
                callback(err, null, null, resp);
                return;
          ***REMOVED***
            const itemsArray = resp.items ? resp.items : [];
            const buckets = itemsArray.map((bucket) => ***REMOVED***
                const bucketInstance = this.bucket(bucket.id);
                bucketInstance.metadata = bucket;
                return bucketInstance;
          ***REMOVED***);
            const nextQuery = resp.nextPageToken
                ? Object.assign(***REMOVED***}, options, ***REMOVED*** pageToken: resp.nextPageToken })
                : null;
            callback(null, buckets, nextQuery, resp);
      ***REMOVED***);
  ***REMOVED***
    getHmacKeys(optionsOrCb, cb) ***REMOVED***
        const ***REMOVED*** options, callback } = (0, util_1.normalize)(optionsOrCb, cb);
        const query = Object.assign(***REMOVED***}, options);
        const projectId = query.projectId || this.projectId;
        delete query.projectId;
        this.request(***REMOVED***
            uri: `/projects/$***REMOVED***projectId}/hmacKeys`,
            qs: query,
      ***REMOVED***, (err, resp) => ***REMOVED***
            if (err) ***REMOVED***
                callback(err, null, null, resp);
                return;
          ***REMOVED***
            const itemsArray = resp.items ? resp.items : [];
            const hmacKeys = itemsArray.map((hmacKey) => ***REMOVED***
                const hmacKeyInstance = this.hmacKey(hmacKey.accessId, ***REMOVED***
                    projectId: hmacKey.projectId,
              ***REMOVED***);
                hmacKeyInstance.metadata = hmacKey;
                return hmacKeyInstance;
          ***REMOVED***);
            const nextQuery = resp.nextPageToken
                ? Object.assign(***REMOVED***}, options, ***REMOVED*** pageToken: resp.nextPageToken })
                : null;
            callback(null, hmacKeys, nextQuery, resp);
      ***REMOVED***);
  ***REMOVED***
    /**
     * @typedef ***REMOVED***array} GetServiceAccountResponse
     * @property ***REMOVED***object} 0 The service account resource.
     * @property ***REMOVED***object} 1 The full
     * ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/projects/serviceAccount#resource| API response}.
     */
    /**
     * @callback GetServiceAccountCallback
     * @param ***REMOVED***?Error} err Request error, if any.
     * @param ***REMOVED***object} serviceAccount The serviceAccount resource.
     * @param ***REMOVED***string} serviceAccount.emailAddress The service account email
     *     address.
     * @param ***REMOVED***object} apiResponse The full
     * ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/projects/serviceAccount#resource| API response}.
     */
    /**
     * Get the email address of this project's Google Cloud Storage service
     * account.
     *
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/projects/serviceAccount/get| Projects.serviceAccount: get API Documentation}
     * See ***REMOVED***@link https://cloud.google.com/storage/docs/json_api/v1/projects/serviceAccount#resource| Projects.serviceAccount Resource}
     *
     * @param ***REMOVED***object} [options] Configuration object.
     * @param ***REMOVED***string} [options.userProject] User project to be billed for this
     *     request.
     * @param ***REMOVED***GetServiceAccountCallback} [callback] Callback function.
     * @returns ***REMOVED***Promise<GetServiceAccountResponse>}
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     *
     * storage.getServiceAccount(function(err, serviceAccount, apiResponse) ***REMOVED***
     *   if (!err) ***REMOVED***
     *     const serviceAccountEmail = serviceAccount.emailAddress;
     * ***REMOVED***
     * });
     *
     * //-
     * // If the callback is omitted, we'll return a Promise.
     * //-
     * storage.getServiceAccount().then(function(data) ***REMOVED***
     *   const serviceAccountEmail = data[0].emailAddress;
     *   const apiResponse = data[1];
     * });
     * ```
     */
    getServiceAccount(optionsOrCallback, cb) ***REMOVED***
        const ***REMOVED*** options, callback } = (0, util_1.normalize)(optionsOrCallback, cb);
        this.request(***REMOVED***
            uri: `/projects/$***REMOVED***this.projectId}/serviceAccount`,
            qs: options,
      ***REMOVED***, (err, resp) => ***REMOVED***
            if (err) ***REMOVED***
                callback(err, null, resp);
                return;
          ***REMOVED***
            const camelCaseResponse = ***REMOVED***};
            for (const prop in resp) ***REMOVED***
                // eslint-disable-next-line no-prototype-builtins
                if (resp.hasOwnProperty(prop)) ***REMOVED***
                    const camelCaseProp = prop.replace(/_(\w)/g, (_, match) => match.toUpperCase());
                    camelCaseResponse[camelCaseProp] = resp[prop];
              ***REMOVED***
          ***REMOVED***
            callback(null, camelCaseResponse, resp);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Get a reference to an HmacKey object.
     * Note: this does not fetch the HMAC key's metadata. Use HmacKey#get() to
     * retrieve and populate the metadata.
     *
     * To get a reference to an HMAC key that's not created for a service
     * account in the same project used to instantiate the Storage client,
     * supply the project's ID as `projectId` in the `options` argument.
     *
     * @param ***REMOVED***string} accessId The HMAC key's access ID.
     * @param ***REMOVED***HmacKeyOptions} options HmacKey constructor options.
     * @returns ***REMOVED***HmacKey}
     * @see HmacKey
     *
     * @example
     * ```
     * const ***REMOVED***Storage} = require('@google-cloud/storage');
     * const storage = new Storage();
     * const hmacKey = storage.hmacKey('ACCESS_ID');
     * ```
     */
    hmacKey(accessId, options) ***REMOVED***
        if (!accessId) ***REMOVED***
            throw new Error(StorageExceptionMessages.HMAC_ACCESS_ID);
      ***REMOVED***
        return new hmacKey_1.HmacKey(this, accessId, options);
  ***REMOVED***
}
exports.Storage = Storage;
/**
 * ***REMOVED***@link Bucket} class.
 *
 * @name Storage.Bucket
 * @see Bucket
 * @type ***REMOVED***Constructor}
 */
Storage.Bucket = bucket_1.Bucket;
/**
 * ***REMOVED***@link Channel} class.
 *
 * @name Storage.Channel
 * @see Channel
 * @type ***REMOVED***Constructor}
 */
Storage.Channel = channel_1.Channel;
/**
 * ***REMOVED***@link File} class.
 *
 * @name Storage.File
 * @see File
 * @type ***REMOVED***Constructor}
 */
Storage.File = file_1.File;
/**
 * ***REMOVED***@link HmacKey} class.
 *
 * @name Storage.HmacKey
 * @see HmacKey
 * @type ***REMOVED***Constructor}
 */
Storage.HmacKey = hmacKey_1.HmacKey;
Storage.acl = ***REMOVED***
    OWNER_ROLE: 'OWNER',
    READER_ROLE: 'READER',
    WRITER_ROLE: 'WRITER',
};
/*! Developer Documentation
 *
 * These methods can be auto-paginated.
 */
paginator_1.paginator.extend(Storage, ['getBuckets', 'getHmacKeys']);
/*! Developer Documentation
 *
 * All async methods (except for streams) will return a Promise in the event
 * that a callback is omitted.
 */
(0, promisify_1.promisifyAll)(Storage, ***REMOVED***
    exclude: ['bucket', 'channel', 'hmacKey'],
});
