// @google-cloud/storage/build/src/util.js
"use strict";
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.PassThroughShim = exports.formatAsUTCISO = exports.convertObjKeysToSnakeCase = exports.unicodeJSONStringify = exports.objectKeyToLowercase = exports.qsStringify = exports.encodeURI = exports.fixedEncodeURIComponent = exports.objectEntries = exports.normalize = void 0;
const querystring = require("querystring");
const stream_1 = require("stream");
function normalize(optionsOrCallback, cb) ***REMOVED***
    const options = (typeof optionsOrCallback === 'object' ? optionsOrCallback : ***REMOVED***});
    const callback = (typeof optionsOrCallback === 'function' ? optionsOrCallback : cb);
    return ***REMOVED*** options, callback };
}
exports.normalize = normalize;
/**
 * Flatten an object into an Array of arrays, [[key, value], ..].
 * Implements Object.entries() for Node.js <8
 * @internal
 */
function objectEntries(obj) ***REMOVED***
    return Object.keys(obj).map(key => [key, obj[key]]);
}
exports.objectEntries = objectEntries;
/**
 * Encode `str` with encodeURIComponent, plus these
 * reserved characters: `! * ' ( )`.
 *
 * See ***REMOVED***@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent| MDN: fixedEncodeURIComponent}
 *
 * @param ***REMOVED***string} str The URI component to encode.
 * @return ***REMOVED***string} The encoded string.
 */
function fixedEncodeURIComponent(str) ***REMOVED***
    return encodeURIComponent(str).replace(/[!'()*]/g, c => '%' + c.charCodeAt(0).toString(16).toUpperCase());
}
exports.fixedEncodeURIComponent = fixedEncodeURIComponent;
/**
 * URI encode `uri` for generating signed URLs, using fixedEncodeURIComponent.
 *
 * Encode every byte except `A-Z a-Z 0-9 ~ - . _`.
 *
 * @param ***REMOVED***string} uri The URI to encode.
 * @param [boolean=false] encodeSlash If `true`, the "/" character is not encoded.
 * @return ***REMOVED***string} The encoded string.
 */
function encodeURI(uri, encodeSlash) ***REMOVED***
    // Split the string by `/`, and conditionally rejoin them with either
    // %2F if encodeSlash is `true`, or '/' if `false`.
    return uri
        .split('/')
        .map(fixedEncodeURIComponent)
        .join(encodeSlash ? '%2F' : '/');
}
exports.encodeURI = encodeURI;
/**
 * Serialize an object to a URL query string using util.encodeURI(uri, true).
 * @param ***REMOVED***string} url The object to serialize.
 * @return ***REMOVED***string} Serialized string.
 */
function qsStringify(qs) ***REMOVED***
    return querystring.stringify(qs, '&', '=', ***REMOVED***
        encodeURIComponent: (component) => encodeURI(component, true),
  ***REMOVED***);
}
exports.qsStringify = qsStringify;
function objectKeyToLowercase(object) ***REMOVED***
    const newObj = ***REMOVED***};
    for (let key of Object.keys(object)) ***REMOVED***
        const value = object[key];
        key = key.toLowerCase();
        newObj[key] = value;
  ***REMOVED***
    return newObj;
}
exports.objectKeyToLowercase = objectKeyToLowercase;
/**
 * JSON encode str, with unicode \u+ representation.
 * @param ***REMOVED***object} obj The object to encode.
 * @return ***REMOVED***string} Serialized string.
 */
function unicodeJSONStringify(obj) ***REMOVED***
    return JSON.stringify(obj).replace(/[\u0080-\uFFFF]/g, (char) => '\\u' + ('0000' + char.charCodeAt(0).toString(16)).slice(-4));
}
exports.unicodeJSONStringify = unicodeJSONStringify;
/**
 * Converts the given objects keys to snake_case
 * @param ***REMOVED***object} obj object to convert keys to snake case.
 * @returns ***REMOVED***object} object with keys converted to snake case.
 */
function convertObjKeysToSnakeCase(obj) ***REMOVED***
    if (obj instanceof Date || obj instanceof RegExp) ***REMOVED***
        return obj;
  ***REMOVED***
    if (Array.isArray(obj)) ***REMOVED***
        return obj.map(convertObjKeysToSnakeCase);
  ***REMOVED***
    if (obj instanceof Object) ***REMOVED***
        return Object.keys(obj).reduce((acc, cur) => ***REMOVED***
            const s = cur[0].toLocaleLowerCase() +
                cur.slice(1).replace(/([A-Z]+)/g, (match, p1) => ***REMOVED***
                    return `_$***REMOVED***p1.toLowerCase()}`;
              ***REMOVED***);
            acc[s] = convertObjKeysToSnakeCase(obj[cur]);
            return acc;
      ***REMOVED***, Object());
  ***REMOVED***
    return obj;
}
exports.convertObjKeysToSnakeCase = convertObjKeysToSnakeCase;
/**
 * Formats the provided date object as a UTC ISO string.
 * @param ***REMOVED***Date} dateTimeToFormat date object to be formatted.
 * @param ***REMOVED***boolean} includeTime flag to include hours, minutes, seconds in output.
 * @param ***REMOVED***string} dateDelimiter delimiter between date components.
 * @param ***REMOVED***string} timeDelimiter delimiter between time components.
 * @returns ***REMOVED***string} UTC ISO format of provided date obect.
 */
function formatAsUTCISO(dateTimeToFormat, includeTime = false, dateDelimiter = '', timeDelimiter = '') ***REMOVED***
    const year = dateTimeToFormat.getUTCFullYear();
    const month = dateTimeToFormat.getUTCMonth() + 1;
    const day = dateTimeToFormat.getUTCDate();
    const hour = dateTimeToFormat.getUTCHours();
    const minute = dateTimeToFormat.getUTCMinutes();
    const second = dateTimeToFormat.getUTCSeconds();
    let resultString = `$***REMOVED***year.toString().padStart(4, '0')}$***REMOVED***dateDelimiter}$***REMOVED***month
        .toString()
        .padStart(2, '0')}$***REMOVED***dateDelimiter}$***REMOVED***day.toString().padStart(2, '0')}`;
    if (includeTime) ***REMOVED***
        resultString = `$***REMOVED***resultString}T$***REMOVED***hour
            .toString()
            .padStart(2, '0')}$***REMOVED***timeDelimiter}$***REMOVED***minute
            .toString()
            .padStart(2, '0')}$***REMOVED***timeDelimiter}$***REMOVED***second.toString().padStart(2, '0')}Z`;
  ***REMOVED***
    return resultString;
}
exports.formatAsUTCISO = formatAsUTCISO;
class PassThroughShim extends stream_1.PassThrough ***REMOVED***
    constructor() ***REMOVED***
        super(...arguments);
        this.shouldEmitReading = true;
        this.shouldEmitWriting = true;
  ***REMOVED***
    _read(size) ***REMOVED***
        if (this.shouldEmitReading) ***REMOVED***
            this.emit('reading');
            this.shouldEmitReading = false;
      ***REMOVED***
        super._read(size);
  ***REMOVED***
    _write(chunk, encoding, callback) ***REMOVED***
        if (this.shouldEmitWriting) ***REMOVED***
            this.emit('writing');
            this.shouldEmitWriting = false;
      ***REMOVED***
        // Per the nodejs documention, callback must be invoked on the next tick
        process.nextTick(() => ***REMOVED***
            super._write(chunk, encoding, callback);
      ***REMOVED***);
  ***REMOVED***
    _final(callback) ***REMOVED***
        // If the stream is empty (i.e. empty file) final will be invoked before _read / _write
        // and we should still emit the proper events.
        if (this.shouldEmitReading) ***REMOVED***
            this.emit('reading');
            this.shouldEmitReading = false;
      ***REMOVED***
        if (this.shouldEmitWriting) ***REMOVED***
            this.emit('writing');
            this.shouldEmitWriting = false;
      ***REMOVED***
        callback(null);
  ***REMOVED***
}
exports.PassThroughShim = PassThroughShim;
