// entities/lib/decode.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) ***REMOVED***
    return (mod && mod.__esModule) ? mod : ***REMOVED*** "default": mod };
};
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.decodeHTML = exports.decodeHTMLStrict = exports.decodeXML = void 0;
var entities_json_1 = __importDefault(require("./maps/entities.json"));
var legacy_json_1 = __importDefault(require("./maps/legacy.json"));
var xml_json_1 = __importDefault(require("./maps/xml.json"));
var decode_codepoint_1 = __importDefault(require("./decode_codepoint"));
exports.decodeXML = getStrictDecoder(xml_json_1.default);
exports.decodeHTMLStrict = getStrictDecoder(entities_json_1.default);
function getStrictDecoder(map) ***REMOVED***
    var keys = Object.keys(map).join("|");
    var replace = getReplacer(map);
    keys += "|#[xX][\\da-fA-F]+|#\\d+";
    var re = new RegExp("&(?:" + keys + ");", "g");
    return function (str) ***REMOVED*** return String(str).replace(re, replace); };
}
var sorter = function (a, b) ***REMOVED*** return (a < b ? 1 : -1); };
exports.decodeHTML = (function () ***REMOVED***
    var legacy = Object.keys(legacy_json_1.default).sort(sorter);
    var keys = Object.keys(entities_json_1.default).sort(sorter);
    for (var i = 0, j = 0; i < keys.length; i++) ***REMOVED***
        if (legacy[j] === keys[i]) ***REMOVED***
            keys[i] += ";?";
            j++;
      ***REMOVED***
        else ***REMOVED***
            keys[i] += ";";
      ***REMOVED***
  ***REMOVED***
    var re = new RegExp("&(?:" + keys.join("|") + "|#[xX][\\da-fA-F]+;?|#\\d+;?)", "g");
    var replace = getReplacer(entities_json_1.default);
    function replacer(str) ***REMOVED***
        if (str.substr(-1) !== ";")
            str += ";";
        return replace(str);
  ***REMOVED***
    // TODO consider creating a merged map
    return function (str) ***REMOVED*** return String(str).replace(re, replacer); };
})();
function getReplacer(map) ***REMOVED***
    return function replace(str) ***REMOVED***
        if (str.charAt(1) === "#") ***REMOVED***
            var secondChar = str.charAt(2);
            if (secondChar === "X" || secondChar === "x") ***REMOVED***
                return decode_codepoint_1.default(parseInt(str.substr(3), 16));
          ***REMOVED***
            return decode_codepoint_1.default(parseInt(str.substr(2), 10));
      ***REMOVED***
        return map[str.slice(1, -1)];
  ***REMOVED***;
}
