// entities/lib/encode.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) ***REMOVED***
    return (mod && mod.__esModule) ? mod : ***REMOVED*** "default": mod };
};
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.escape = exports.encodeHTML = exports.encodeXML = void 0;
var xml_json_1 = __importDefault(require("./maps/xml.json"));
var inverseXML = getInverseObj(xml_json_1.default);
var xmlReplacer = getInverseReplacer(inverseXML);
exports.encodeXML = getInverse(inverseXML, xmlReplacer);
var entities_json_1 = __importDefault(require("./maps/entities.json"));
var inverseHTML = getInverseObj(entities_json_1.default);
var htmlReplacer = getInverseReplacer(inverseHTML);
exports.encodeHTML = getInverse(inverseHTML, htmlReplacer);
function getInverseObj(obj) ***REMOVED***
    return Object.keys(obj)
        .sort()
        .reduce(function (inverse, name) ***REMOVED***
        inverse[obj[name]] = "&" + name + ";";
        return inverse;
  ***REMOVED***, ***REMOVED***});
}
function getInverseReplacer(inverse) ***REMOVED***
    var single = [];
    var multiple = [];
    for (var _i = 0, _a = Object.keys(inverse); _i < _a.length; _i++) ***REMOVED***
        var k = _a[_i];
        if (k.length === 1) ***REMOVED***
            // Add value to single array
            single.push("\\" + k);
      ***REMOVED***
        else ***REMOVED***
            // Add value to multiple array
            multiple.push(k);
      ***REMOVED***
  ***REMOVED***
    // Add ranges to single characters.
    single.sort();
    for (var start = 0; start < single.length - 1; start++) ***REMOVED***
        // Find the end of a run of characters
        var end = start;
        while (end < single.length - 1 &&
            single[end].charCodeAt(1) + 1 === single[end + 1].charCodeAt(1)) ***REMOVED***
            end += 1;
      ***REMOVED***
        var count = 1 + end - start;
        // We want to replace at least three characters
        if (count < 3)
            continue;
        single.splice(start, count, single[start] + "-" + single[end]);
  ***REMOVED***
    multiple.unshift("[" + single.join("") + "]");
    return new RegExp(multiple.join("|"), "g");
}
var reNonASCII = /(?:[\x80-\uD7FF\uE000-\uFFFF]|[\uD800-\uDBFF][\uDC00-\uDFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF])/g;
function singleCharReplacer(c) ***REMOVED***
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return "&#x" + c.codePointAt(0).toString(16).toUpperCase() + ";";
}
function getInverse(inverse, re) ***REMOVED***
    return function (data) ***REMOVED***
        return data
            .replace(re, function (name) ***REMOVED*** return inverse[name]; })
            .replace(reNonASCII, singleCharReplacer);
  ***REMOVED***;
}
var reXmlChars = getInverseReplacer(inverseXML);
function escape(data) ***REMOVED***
    return data
        .replace(reXmlChars, singleCharReplacer)
        .replace(reNonASCII, singleCharReplacer);
}
exports.escape = escape;
