// protobufjs-cli/util.js
"use strict";
var fs       = require("fs"),
    path     = require("path"),
    protobuf = require("protobufjs");

function basenameCompare(a, b) ***REMOVED***
    var aa = String(a).replace(/\.\w+$/, "").split(/(-?\d*\.?\d+)/g),
        bb = String(b).replace(/\.\w+$/, "").split(/(-?\d*\.?\d+)/g);
    for (var i = 0, k = Math.min(aa.length, bb.length); i < k; i++) ***REMOVED***
        var x = parseFloat(aa[i]) || aa[i].toLowerCase(),
            y = parseFloat(bb[i]) || bb[i].toLowerCase();
        if (x < y)
            return -1;
        if (x > y)
            return 1;
  ***REMOVED***
    return a.length < b.length ? -1 : 0;
}

exports.requireAll = function requireAll(dirname) ***REMOVED***
    dirname   = path.join(__dirname, dirname);
    var files = fs.readdirSync(dirname).sort(basenameCompare),
        all = ***REMOVED***};
    files.forEach(function(file) ***REMOVED***
        var basename = path.basename(file, ".js"),
            extname  = path.extname(file);
        if (extname === ".js")
            all[basename] = require(path.join(dirname, file));
  ***REMOVED***);
    return all;
};

exports.traverse = function traverse(current, fn) ***REMOVED***
    fn(current);
    if (current.fieldsArray)
        current.fieldsArray.forEach(function(field) ***REMOVED***
            traverse(field, fn);
      ***REMOVED***);
    if (current.oneofsArray)
        current.oneofsArray.forEach(function(oneof) ***REMOVED***
            traverse(oneof, fn);
      ***REMOVED***);
    if (current.methodsArray)
        current.methodsArray.forEach(function(method) ***REMOVED***
            traverse(method, fn);
      ***REMOVED***);
    if (current.nestedArray)
        current.nestedArray.forEach(function(nested) ***REMOVED***
            traverse(nested, fn);
      ***REMOVED***);
};

exports.traverseResolved = function traverseResolved(current, fn) ***REMOVED***
    fn(current);
    if (current.resolvedType)
        traverseResolved(current.resolvedType, fn);
    if (current.resolvedKeyType)
        traverseResolved(current.resolvedKeyType, fn);
    if (current.resolvedRequestType)
        traverseResolved(current.resolvedRequestType, fn);
    if (current.resolvedResponseType)
        traverseResolved(current.resolvedResponseType, fn);
};

exports.inspect = function inspect(object, indent) ***REMOVED***
    if (!object)
        return "";
    var chalk = require("chalk");
    var sb = [];
    if (!indent)
        indent = "";
    var ind = indent ? indent.substring(0, indent.length - 2) + "└ " : "";
    sb.push(
        ind + chalk.bold(object.toString()) + (object.visible ? " (visible)" : ""),
        indent + chalk.gray("parent: ") + object.parent
    );
    if (object instanceof protobuf.Field) ***REMOVED***
        if (object.extend !== undefined)
            sb.push(indent + chalk.gray("extend: ") + object.extend);
        if (object.partOf)
            sb.push(indent + chalk.gray("oneof : ") + object.oneof);
  ***REMOVED***
    sb.push("");
    if (object.fieldsArray)
        object.fieldsArray.forEach(function(field) ***REMOVED***
            sb.push(inspect(field, indent + "  "));
      ***REMOVED***);
    if (object.oneofsArray)
        object.oneofsArray.forEach(function(oneof) ***REMOVED***
            sb.push(inspect(oneof, indent + "  "));
      ***REMOVED***);
    if (object.methodsArray)
        object.methodsArray.forEach(function(service) ***REMOVED***
            sb.push(inspect(service, indent + "  "));
      ***REMOVED***);
    if (object.nestedArray)
        object.nestedArray.forEach(function(nested) ***REMOVED***
            sb.push(inspect(nested, indent + "  "));
      ***REMOVED***);
    return sb.join("\n");
};

exports.wrap = function(OUTPUT, options) ***REMOVED***
    var name = options.wrap || "default";
    var wrap;
    try ***REMOVED***
        // try built-in wrappers first
        wrap = fs.readFileSync(path.join(__dirname, "wrappers", name + ".js")).toString("utf8");
  ***REMOVED*** catch (e) ***REMOVED***
        // otherwise fetch the custom one
        wrap = fs.readFileSync(path.resolve(process.cwd(), name)).toString("utf8");
  ***REMOVED***
    wrap = wrap.replace(/\$DEPENDENCY/g, JSON.stringify(options.dependency || "protobufjs"));
    wrap = wrap.replace(/( *)\$OUTPUT;/, function($0, $1) ***REMOVED***
        return $1.length ? OUTPUT.replace(/^/mg, $1) : OUTPUT;
  ***REMOVED***);
    if (options.lint !== "")
        wrap = "/*" + options.lint + "*/\n" + wrap;
    return wrap.replace(/\r?\n/g, "\n");
};

exports.pad = function(str, len, l) ***REMOVED***
    while (str.length < len)
        str = l ? str + " " : " " + str;
    return str;
};
