// lru-memoizer/lib/async.js
"use strict";
var __read = (this && this.__read) || function (o, n) ***REMOVED***
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try ***REMOVED***
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
  ***REMOVED***
    catch (error) ***REMOVED*** e = ***REMOVED*** error: error }; }
    finally ***REMOVED***
        try ***REMOVED***
            if (r && !r.done && (m = i["return"])) m.call(i);
      ***REMOVED***
        finally ***REMOVED*** if (e) throw e.error; }
  ***REMOVED***
    return ar;
};
var __spread = (this && this.__spread) || function () ***REMOVED***
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __values = (this && this.__values) || function (o) ***REMOVED***
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return ***REMOVED***
        next: function () ***REMOVED***
            if (o && i >= o.length) o = void 0;
            return ***REMOVED*** value: o && o[i++], done: !o };
      ***REMOVED***
  ***REMOVED***;
};
var __importDefault = (this && this.__importDefault) || function (mod) ***REMOVED***
    return (mod && mod.__esModule) ? mod : ***REMOVED*** "default": mod };
};
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
var lru_cache_1 = __importDefault(require("lru-cache"));
var events_1 = require("events");
var lodash_clonedeep_1 = __importDefault(require("lodash.clonedeep"));
var freeze_1 = require("./freeze");
var sync_1 = require("./sync");
function asyncMemoizer(options) ***REMOVED***
    var cache = new lru_cache_1.default(options);
    var load = options.load;
    var hash = options.hash;
    var bypass = options.bypass;
    var itemMaxAge = options.itemMaxAge;
    var freeze = options.freeze;
    var clone = options.clone;
    var queueMaxAge = options.queueMaxAge || 1000;
    var loading = new Map();
    var emitter = new events_1.EventEmitter();
    var memoizerMethods = Object.assign(***REMOVED***
        del: del,
        reset: function () ***REMOVED*** return cache.reset(); },
        keys: cache.keys.bind(cache),
        on: emitter.on.bind(emitter),
        once: emitter.once.bind(emitter)
  ***REMOVED***, options);
    if (options.disable) ***REMOVED***
        return Object.assign(load, memoizerMethods);
  ***REMOVED***
    function del() ***REMOVED***
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        var key = hash.apply(void 0, __spread(args));
        cache.del(key);
  ***REMOVED***
    function add(key, parameters, result) ***REMOVED***
        if (freeze) ***REMOVED***
            result.forEach(freeze_1.deepFreeze);
      ***REMOVED***
        if (itemMaxAge) ***REMOVED***
            cache.set(key, result, itemMaxAge.apply(void 0, __spread(parameters.concat(result))));
      ***REMOVED***
        else ***REMOVED***
            cache.set(key, result);
      ***REMOVED***
  ***REMOVED***
    function runCallbacks(callbacks, args) ***REMOVED***
        var e_1, _a;
        try ***REMOVED***
            for (var callbacks_1 = __values(callbacks), callbacks_1_1 = callbacks_1.next(); !callbacks_1_1.done; callbacks_1_1 = callbacks_1.next()) ***REMOVED***
                var callback = callbacks_1_1.value;
                // Simulate async call when returning from cache
                // and yield between callback resolution
                if (clone) ***REMOVED***
                    setImmediate.apply(void 0, __spread([callback], args.map(lodash_clonedeep_1.default)));
              ***REMOVED***
                else ***REMOVED***
                    setImmediate.apply(void 0, __spread([callback], args));
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        catch (e_1_1) ***REMOVED*** e_1 = ***REMOVED*** error: e_1_1 }; }
        finally ***REMOVED***
            try ***REMOVED***
                if (callbacks_1_1 && !callbacks_1_1.done && (_a = callbacks_1.return)) _a.call(callbacks_1);
          ***REMOVED***
            finally ***REMOVED*** if (e_1) throw e_1.error; }
      ***REMOVED***
  ***REMOVED***
    function emit(event) ***REMOVED***
        var parameters = [];
        for (var _i = 1; _i < arguments.length; _i++) ***REMOVED***
            parameters[_i - 1] = arguments[_i];
      ***REMOVED***
        emitter.emit.apply(emitter, __spread([event], parameters));
  ***REMOVED***
    function memoizedFunction() ***REMOVED***
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        var parameters = args.slice(0, -1);
        var callback = args.slice(-1).pop();
        var key;
        if (bypass && bypass.apply(void 0, __spread(parameters))) ***REMOVED***
            emit.apply(void 0, __spread(['miss'], parameters));
            return load.apply(void 0, __spread(args));
      ***REMOVED***
        if (parameters.length === 0 && !hash) ***REMOVED***
            //the load function only receives callback.
            key = '_';
      ***REMOVED***
        else ***REMOVED***
            key = hash.apply(void 0, __spread(parameters));
      ***REMOVED***
        var fromCache = cache.get(key);
        if (fromCache) ***REMOVED***
            emit.apply(void 0, __spread(['hit'], parameters));
            // found, invoke callback
            return runCallbacks([callback], [null].concat(fromCache));
      ***REMOVED***
        var pendingLoad = loading.get(key);
        if (pendingLoad && pendingLoad.expiresAt > Date.now()) ***REMOVED***
            // request already in progress, queue and return
            pendingLoad.queue.push(callback);
            emit.apply(void 0, __spread(['queue'], parameters));
            return;
      ***REMOVED***
        emit.apply(void 0, __spread(['miss'], parameters));
        var started = Date.now();
        // no pending request or not resolved before expiration
        // create a new queue and invoke load
        var queue = [callback];
        loading.set(key, ***REMOVED***
            queue: queue,
            expiresAt: started + queueMaxAge
      ***REMOVED***);
        var loadHandler = function () ***REMOVED***
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
                args[_i] = arguments[_i];
          ***REMOVED***
            var err = args[0];
            if (!err) ***REMOVED***
                add(key, parameters, args.slice(1));
          ***REMOVED***
            // this can potentially delete a different queue than `queue` if
            // this callback was called after expiration.
            // that will only cause a new call to be performed and a new queue to be
            // created
            loading.delete(key);
            emit.apply(void 0, __spread(['loaded', Date.now() - started], parameters));
            runCallbacks(queue, args);
      ***REMOVED***;
        load.apply(void 0, __spread(parameters, [loadHandler]));
  ***REMOVED***
    ;
    return Object.assign(memoizedFunction, memoizerMethods);
}
exports.asyncMemoizer = asyncMemoizer;
asyncMemoizer.sync = sync_1.syncMemoizer;
