// lru-memoizer/lib/sync.js
"use strict";
var __read = (this && this.__read) || function (o, n) ***REMOVED***
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try ***REMOVED***
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
  ***REMOVED***
    catch (error) ***REMOVED*** e = ***REMOVED*** error: error }; }
    finally ***REMOVED***
        try ***REMOVED***
            if (r && !r.done && (m = i["return"])) m.call(i);
      ***REMOVED***
        finally ***REMOVED*** if (e) throw e.error; }
  ***REMOVED***
    return ar;
};
var __spread = (this && this.__spread) || function () ***REMOVED***
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) ***REMOVED***
    return (mod && mod.__esModule) ? mod : ***REMOVED*** "default": mod };
};
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
var lru_cache_1 = __importDefault(require("lru-cache"));
var events_1 = require("events");
var lodash_clonedeep_1 = __importDefault(require("lodash.clonedeep"));
var freeze_1 = require("./freeze");
function syncMemoizer(options) ***REMOVED***
    var cache = new lru_cache_1.default(options);
    var load = options.load;
    var hash = options.hash;
    var bypass = options.bypass;
    var itemMaxAge = options.itemMaxAge;
    var freeze = options.freeze;
    var clone = options.clone;
    var emitter = new events_1.EventEmitter();
    var defaultResult = Object.assign(***REMOVED***
        del: del,
        reset: function () ***REMOVED*** return cache.reset(); },
        keys: cache.keys.bind(cache),
        on: emitter.on.bind(emitter),
        once: emitter.once.bind(emitter)
  ***REMOVED***, options);
    if (options.disable) ***REMOVED***
        return Object.assign(load, defaultResult);
  ***REMOVED***
    function del() ***REMOVED***
        var key = hash.apply(void 0, __spread(arguments));
        cache.del(key);
  ***REMOVED***
    function emit(event) ***REMOVED***
        var parameters = [];
        for (var _i = 1; _i < arguments.length; _i++) ***REMOVED***
            parameters[_i - 1] = arguments[_i];
      ***REMOVED***
        emitter.emit.apply(emitter, __spread([event], parameters));
  ***REMOVED***
    function isPromise(result) ***REMOVED***
        // detect native, bluebird, A+ promises
        return result && result.then && typeof result.then === 'function';
  ***REMOVED***
    function processResult(result) ***REMOVED***
        var res = result;
        if (clone) ***REMOVED***
            if (isPromise(res)) ***REMOVED***
                res = res.then(lodash_clonedeep_1.default);
          ***REMOVED***
            else ***REMOVED***
                res = lodash_clonedeep_1.default(res);
          ***REMOVED***
      ***REMOVED***
        if (freeze) ***REMOVED***
            if (isPromise(res)) ***REMOVED***
                res = res.then(freeze_1.deepFreeze);
          ***REMOVED***
            else ***REMOVED***
                freeze_1.deepFreeze(res);
          ***REMOVED***
      ***REMOVED***
        return res;
  ***REMOVED***
    var result = function () ***REMOVED***
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        if (bypass && bypass.apply(void 0, __spread(args))) ***REMOVED***
            emit.apply(void 0, __spread(['miss'], args));
            return load.apply(void 0, __spread(args));
      ***REMOVED***
        var key = hash.apply(void 0, __spread(args));
        var fromCache = cache.get(key);
        if (fromCache) ***REMOVED***
            emit.apply(void 0, __spread(['hit'], args));
            return processResult(fromCache);
      ***REMOVED***
        emit.apply(void 0, __spread(['miss'], args));
        var result = load.apply(void 0, __spread(args));
        if (itemMaxAge) ***REMOVED***
            // @ts-ignore
            cache.set(key, result, itemMaxAge.apply(void 0, __spread(args.concat([result]))));
      ***REMOVED***
        else ***REMOVED***
            cache.set(key, result);
      ***REMOVED***
        return processResult(result);
  ***REMOVED***;
    return Object.assign(result, defaultResult);
}
exports.syncMemoizer = syncMemoizer;
