// protobufjs/src/common.js
"use strict";
module.exports = common;

var commonRe = /\/|\./;

/**
 * Provides common type definitions.
 * Can also be used to provide additional google types or your own custom types.
 * @param ***REMOVED***string} name Short name as in `google/protobuf/[name].proto` or full file name
 * @param ***REMOVED***Object.<string,*>} json JSON definition within `google.protobuf` if a short name, otherwise the file's root definition
 * @returns ***REMOVED***undefined}
 * @property ***REMOVED***INamespace} google/protobuf/any.proto Any
 * @property ***REMOVED***INamespace} google/protobuf/duration.proto Duration
 * @property ***REMOVED***INamespace} google/protobuf/empty.proto Empty
 * @property ***REMOVED***INamespace} google/protobuf/field_mask.proto FieldMask
 * @property ***REMOVED***INamespace} google/protobuf/struct.proto Struct, Value, NullValue and ListValue
 * @property ***REMOVED***INamespace} google/protobuf/timestamp.proto Timestamp
 * @property ***REMOVED***INamespace} google/protobuf/wrappers.proto Wrappers
 * @example
 * // manually provides descriptor.proto (assumes google/protobuf/ namespace and .proto extension)
 * protobuf.common("descriptor", descriptorJson);
 *
 * // manually provides a custom definition (uses my.foo namespace)
 * protobuf.common("my/foo/bar.proto", myFooBarJson);
 */
function common(name, json) ***REMOVED***
    if (!commonRe.test(name)) ***REMOVED***
        name = "google/protobuf/" + name + ".proto";
        json = ***REMOVED*** nested: ***REMOVED*** google: ***REMOVED*** nested: ***REMOVED*** protobuf: ***REMOVED*** nested: json } } } } };
  ***REMOVED***
    common[name] = json;
}

// Not provided because of limited use (feel free to discuss or to provide yourself):
//
// google/protobuf/descriptor.proto
// google/protobuf/source_context.proto
// google/protobuf/type.proto
//
// Stripped and pre-parsed versions of these non-bundled files are instead available as part of
// the repository or package within the google/protobuf directory.

common("any", ***REMOVED***

    /**
     * Properties of a google.protobuf.Any message.
     * @interface IAny
     * @type ***REMOVED***Object}
     * @property ***REMOVED***string} [typeUrl]
     * @property ***REMOVED***Uint8Array} [bytes]
     * @memberof common
     */
    Any: ***REMOVED***
        fields: ***REMOVED***
            type_url: ***REMOVED***
                type: "string",
                id: 1
          ***REMOVED***,
            value: ***REMOVED***
                type: "bytes",
                id: 2
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
});

var timeType;

common("duration", ***REMOVED***

    /**
     * Properties of a google.protobuf.Duration message.
     * @interface IDuration
     * @type ***REMOVED***Object}
     * @property ***REMOVED***number|Long} [seconds]
     * @property ***REMOVED***number} [nanos]
     * @memberof common
     */
    Duration: timeType = ***REMOVED***
        fields: ***REMOVED***
            seconds: ***REMOVED***
                type: "int64",
                id: 1
          ***REMOVED***,
            nanos: ***REMOVED***
                type: "int32",
                id: 2
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
});

common("timestamp", ***REMOVED***

    /**
     * Properties of a google.protobuf.Timestamp message.
     * @interface ITimestamp
     * @type ***REMOVED***Object}
     * @property ***REMOVED***number|Long} [seconds]
     * @property ***REMOVED***number} [nanos]
     * @memberof common
     */
    Timestamp: timeType
});

common("empty", ***REMOVED***

    /**
     * Properties of a google.protobuf.Empty message.
     * @interface IEmpty
     * @memberof common
     */
    Empty: ***REMOVED***
        fields: ***REMOVED***}
  ***REMOVED***
});

common("struct", ***REMOVED***

    /**
     * Properties of a google.protobuf.Struct message.
     * @interface IStruct
     * @type ***REMOVED***Object}
     * @property ***REMOVED***Object.<string,IValue>} [fields]
     * @memberof common
     */
    Struct: ***REMOVED***
        fields: ***REMOVED***
            fields: ***REMOVED***
                keyType: "string",
                type: "Value",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.Value message.
     * @interface IValue
     * @type ***REMOVED***Object}
     * @property ***REMOVED***string} [kind]
     * @property ***REMOVED***0} [nullValue]
     * @property ***REMOVED***number} [numberValue]
     * @property ***REMOVED***string} [stringValue]
     * @property ***REMOVED***boolean} [boolValue]
     * @property ***REMOVED***IStruct} [structValue]
     * @property ***REMOVED***IListValue} [listValue]
     * @memberof common
     */
    Value: ***REMOVED***
        oneofs: ***REMOVED***
            kind: ***REMOVED***
                oneof: [
                    "nullValue",
                    "numberValue",
                    "stringValue",
                    "boolValue",
                    "structValue",
                    "listValue"
                ]
          ***REMOVED***
      ***REMOVED***,
        fields: ***REMOVED***
            nullValue: ***REMOVED***
                type: "NullValue",
                id: 1
          ***REMOVED***,
            numberValue: ***REMOVED***
                type: "double",
                id: 2
          ***REMOVED***,
            stringValue: ***REMOVED***
                type: "string",
                id: 3
          ***REMOVED***,
            boolValue: ***REMOVED***
                type: "bool",
                id: 4
          ***REMOVED***,
            structValue: ***REMOVED***
                type: "Struct",
                id: 5
          ***REMOVED***,
            listValue: ***REMOVED***
                type: "ListValue",
                id: 6
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    NullValue: ***REMOVED***
        values: ***REMOVED***
            NULL_VALUE: 0
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.ListValue message.
     * @interface IListValue
     * @type ***REMOVED***Object}
     * @property ***REMOVED***Array.<IValue>} [values]
     * @memberof common
     */
    ListValue: ***REMOVED***
        fields: ***REMOVED***
            values: ***REMOVED***
                rule: "repeated",
                type: "Value",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
});

common("wrappers", ***REMOVED***

    /**
     * Properties of a google.protobuf.DoubleValue message.
     * @interface IDoubleValue
     * @type ***REMOVED***Object}
     * @property ***REMOVED***number} [value]
     * @memberof common
     */
    DoubleValue: ***REMOVED***
        fields: ***REMOVED***
            value: ***REMOVED***
                type: "double",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.FloatValue message.
     * @interface IFloatValue
     * @type ***REMOVED***Object}
     * @property ***REMOVED***number} [value]
     * @memberof common
     */
    FloatValue: ***REMOVED***
        fields: ***REMOVED***
            value: ***REMOVED***
                type: "float",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.Int64Value message.
     * @interface IInt64Value
     * @type ***REMOVED***Object}
     * @property ***REMOVED***number|Long} [value]
     * @memberof common
     */
    Int64Value: ***REMOVED***
        fields: ***REMOVED***
            value: ***REMOVED***
                type: "int64",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.UInt64Value message.
     * @interface IUInt64Value
     * @type ***REMOVED***Object}
     * @property ***REMOVED***number|Long} [value]
     * @memberof common
     */
    UInt64Value: ***REMOVED***
        fields: ***REMOVED***
            value: ***REMOVED***
                type: "uint64",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.Int32Value message.
     * @interface IInt32Value
     * @type ***REMOVED***Object}
     * @property ***REMOVED***number} [value]
     * @memberof common
     */
    Int32Value: ***REMOVED***
        fields: ***REMOVED***
            value: ***REMOVED***
                type: "int32",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.UInt32Value message.
     * @interface IUInt32Value
     * @type ***REMOVED***Object}
     * @property ***REMOVED***number} [value]
     * @memberof common
     */
    UInt32Value: ***REMOVED***
        fields: ***REMOVED***
            value: ***REMOVED***
                type: "uint32",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.BoolValue message.
     * @interface IBoolValue
     * @type ***REMOVED***Object}
     * @property ***REMOVED***boolean} [value]
     * @memberof common
     */
    BoolValue: ***REMOVED***
        fields: ***REMOVED***
            value: ***REMOVED***
                type: "bool",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.StringValue message.
     * @interface IStringValue
     * @type ***REMOVED***Object}
     * @property ***REMOVED***string} [value]
     * @memberof common
     */
    StringValue: ***REMOVED***
        fields: ***REMOVED***
            value: ***REMOVED***
                type: "string",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***,

    /**
     * Properties of a google.protobuf.BytesValue message.
     * @interface IBytesValue
     * @type ***REMOVED***Object}
     * @property ***REMOVED***Uint8Array} [value]
     * @memberof common
     */
    BytesValue: ***REMOVED***
        fields: ***REMOVED***
            value: ***REMOVED***
                type: "bytes",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
});

common("field_mask", ***REMOVED***

    /**
     * Properties of a google.protobuf.FieldMask message.
     * @interface IDoubleValue
     * @type ***REMOVED***Object}
     * @property ***REMOVED***number} [value]
     * @memberof common
     */
    FieldMask: ***REMOVED***
        fields: ***REMOVED***
            paths: ***REMOVED***
                rule: "repeated",
                type: "string",
                id: 1
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
});

/**
 * Gets the root definition of the specified common proto file.
 *
 * Bundled definitions are:
 * - google/protobuf/any.proto
 * - google/protobuf/duration.proto
 * - google/protobuf/empty.proto
 * - google/protobuf/field_mask.proto
 * - google/protobuf/struct.proto
 * - google/protobuf/timestamp.proto
 * - google/protobuf/wrappers.proto
 *
 * @param ***REMOVED***string} file Proto file name
 * @returns ***REMOVED***INamespace|null} Root definition or `null` if not defined
 */
common.get = function get(file) ***REMOVED***
    return common[file] || null;
};
