// protobufjs/src/method.js
"use strict";
module.exports = Method;

// extends ReflectionObject
var ReflectionObject = require("./object");
((Method.prototype = Object.create(ReflectionObject.prototype)).constructor = Method).className = "Method";

var util = require("./util");

/**
 * Constructs a new service method instance.
 * @classdesc Reflected service method.
 * @extends ReflectionObject
 * @constructor
 * @param ***REMOVED***string} name Method name
 * @param ***REMOVED***string|undefined} type Method type, usually `"rpc"`
 * @param ***REMOVED***string} requestType Request message type
 * @param ***REMOVED***string} responseType Response message type
 * @param ***REMOVED***boolean|Object.<string,*>} [requestStream] Whether the request is streamed
 * @param ***REMOVED***boolean|Object.<string,*>} [responseStream] Whether the response is streamed
 * @param ***REMOVED***Object.<string,*>} [options] Declared options
 * @param ***REMOVED***string} [comment] The comment for this method
 * @param ***REMOVED***Object.<string,*>} [parsedOptions] Declared options, properly parsed into an object
 */
function Method(name, type, requestType, responseType, requestStream, responseStream, options, comment, parsedOptions) ***REMOVED***

    /* istanbul ignore next */
    if (util.isObject(requestStream)) ***REMOVED***
        options = requestStream;
        requestStream = responseStream = undefined;
  ***REMOVED*** else if (util.isObject(responseStream)) ***REMOVED***
        options = responseStream;
        responseStream = undefined;
  ***REMOVED***

    /* istanbul ignore if */
    if (!(type === undefined || util.isString(type)))
        throw TypeError("type must be a string");

    /* istanbul ignore if */
    if (!util.isString(requestType))
        throw TypeError("requestType must be a string");

    /* istanbul ignore if */
    if (!util.isString(responseType))
        throw TypeError("responseType must be a string");

    ReflectionObject.call(this, name, options);

    /**
     * Method type.
     * @type ***REMOVED***string}
     */
    this.type = type || "rpc"; // toJSON

    /**
     * Request type.
     * @type ***REMOVED***string}
     */
    this.requestType = requestType; // toJSON, marker

    /**
     * Whether requests are streamed or not.
     * @type ***REMOVED***boolean|undefined}
     */
    this.requestStream = requestStream ? true : undefined; // toJSON

    /**
     * Response type.
     * @type ***REMOVED***string}
     */
    this.responseType = responseType; // toJSON

    /**
     * Whether responses are streamed or not.
     * @type ***REMOVED***boolean|undefined}
     */
    this.responseStream = responseStream ? true : undefined; // toJSON

    /**
     * Resolved request type.
     * @type ***REMOVED***Type|null}
     */
    this.resolvedRequestType = null;

    /**
     * Resolved response type.
     * @type ***REMOVED***Type|null}
     */
    this.resolvedResponseType = null;

    /**
     * Comment for this method
     * @type ***REMOVED***string|null}
     */
    this.comment = comment;

    /**
     * Options properly parsed into an object
     */
    this.parsedOptions = parsedOptions;
}

/**
 * Method descriptor.
 * @interface IMethod
 * @property ***REMOVED***string} [type="rpc"] Method type
 * @property ***REMOVED***string} requestType Request type
 * @property ***REMOVED***string} responseType Response type
 * @property ***REMOVED***boolean} [requestStream=false] Whether requests are streamed
 * @property ***REMOVED***boolean} [responseStream=false] Whether responses are streamed
 * @property ***REMOVED***Object.<string,*>} [options] Method options
 * @property ***REMOVED***string} comment Method comments
 * @property ***REMOVED***Object.<string,*>} [parsedOptions] Method options properly parsed into an object
 */

/**
 * Constructs a method from a method descriptor.
 * @param ***REMOVED***string} name Method name
 * @param ***REMOVED***IMethod} json Method descriptor
 * @returns ***REMOVED***Method} Created method
 * @throws ***REMOVED***TypeError} If arguments are invalid
 */
Method.fromJSON = function fromJSON(name, json) ***REMOVED***
    return new Method(name, json.type, json.requestType, json.responseType, json.requestStream, json.responseStream, json.options, json.comment, json.parsedOptions);
};

/**
 * Converts this method to a method descriptor.
 * @param ***REMOVED***IToJSONOptions} [toJSONOptions] JSON conversion options
 * @returns ***REMOVED***IMethod} Method descriptor
 */
Method.prototype.toJSON = function toJSON(toJSONOptions) ***REMOVED***
    var keepComments = toJSONOptions ? Boolean(toJSONOptions.keepComments) : false;
    return util.toObject([
        "type"           , this.type !== "rpc" && /* istanbul ignore next */ this.type || undefined,
        "requestType"    , this.requestType,
        "requestStream"  , this.requestStream,
        "responseType"   , this.responseType,
        "responseStream" , this.responseStream,
        "options"        , this.options,
        "comment"        , keepComments ? this.comment : undefined,
        "parsedOptions"  , this.parsedOptions,
    ]);
};

/**
 * @override
 */
Method.prototype.resolve = function resolve() ***REMOVED***

    /* istanbul ignore if */
    if (this.resolved)
        return this;

    this.resolvedRequestType = this.parent.lookupType(this.requestType);
    this.resolvedResponseType = this.parent.lookupType(this.responseType);

    return ReflectionObject.prototype.resolve.call(this);
};
