// @grpc/grpc-js/build/src/backoff-timeout.js
"use strict";
/*
 * Copyright 2019 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.BackoffTimeout = void 0;
const INITIAL_BACKOFF_MS = 1000;
const BACKOFF_MULTIPLIER = 1.6;
const MAX_BACKOFF_MS = 120000;
const BACKOFF_JITTER = 0.2;
/**
 * Get a number uniformly at random in the range [min, max)
 * @param min
 * @param max
 */
function uniformRandom(min, max) ***REMOVED***
    return Math.random() * (max - min) + min;
}
class BackoffTimeout ***REMOVED***
    constructor(callback, options) ***REMOVED***
        this.callback = callback;
        /**
         * The delay time at the start, and after each reset.
         */
        this.initialDelay = INITIAL_BACKOFF_MS;
        /**
         * The exponential backoff multiplier.
         */
        this.multiplier = BACKOFF_MULTIPLIER;
        /**
         * The maximum delay time
         */
        this.maxDelay = MAX_BACKOFF_MS;
        /**
         * The maximum fraction by which the delay time can randomly vary after
         * applying the multiplier.
         */
        this.jitter = BACKOFF_JITTER;
        /**
         * Indicates whether the timer is currently running.
         */
        this.running = false;
        /**
         * Indicates whether the timer should keep the Node process running if no
         * other async operation is doing so.
         */
        this.hasRef = true;
        /**
         * The time that the currently running timer was started. Only valid if
         * running is true.
         */
        this.startTime = new Date();
        if (options) ***REMOVED***
            if (options.initialDelay) ***REMOVED***
                this.initialDelay = options.initialDelay;
          ***REMOVED***
            if (options.multiplier) ***REMOVED***
                this.multiplier = options.multiplier;
          ***REMOVED***
            if (options.jitter) ***REMOVED***
                this.jitter = options.jitter;
          ***REMOVED***
            if (options.maxDelay) ***REMOVED***
                this.maxDelay = options.maxDelay;
          ***REMOVED***
      ***REMOVED***
        this.nextDelay = this.initialDelay;
        this.timerId = setTimeout(() => ***REMOVED*** }, 0);
        clearTimeout(this.timerId);
  ***REMOVED***
    runTimer(delay) ***REMOVED***
        var _a, _b;
        clearTimeout(this.timerId);
        this.timerId = setTimeout(() => ***REMOVED***
            this.callback();
            this.running = false;
      ***REMOVED***, delay);
        if (!this.hasRef) ***REMOVED***
            (_b = (_a = this.timerId).unref) === null || _b === void 0 ? void 0 : _b.call(_a);
      ***REMOVED***
  ***REMOVED***
    /**
     * Call the callback after the current amount of delay time
     */
    runOnce() ***REMOVED***
        this.running = true;
        this.startTime = new Date();
        this.runTimer(this.nextDelay);
        const nextBackoff = Math.min(this.nextDelay * this.multiplier, this.maxDelay);
        const jitterMagnitude = nextBackoff * this.jitter;
        this.nextDelay =
            nextBackoff + uniformRandom(-jitterMagnitude, jitterMagnitude);
  ***REMOVED***
    /**
     * Stop the timer. The callback will not be called until `runOnce` is called
     * again.
     */
    stop() ***REMOVED***
        clearTimeout(this.timerId);
        this.running = false;
  ***REMOVED***
    /**
     * Reset the delay time to its initial value. If the timer is still running,
     * retroactively apply that reset to the current timer.
     */
    reset() ***REMOVED***
        this.nextDelay = this.initialDelay;
        if (this.running) ***REMOVED***
            const now = new Date();
            const newEndTime = this.startTime;
            newEndTime.setMilliseconds(newEndTime.getMilliseconds() + this.nextDelay);
            clearTimeout(this.timerId);
            if (now < newEndTime) ***REMOVED***
                this.runTimer(newEndTime.getTime() - now.getTime());
          ***REMOVED***
            else ***REMOVED***
                this.running = false;
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    /**
     * Check whether the timer is currently running.
     */
    isRunning() ***REMOVED***
        return this.running;
  ***REMOVED***
    /**
     * Set that while the timer is running, it should keep the Node process
     * running.
     */
    ref() ***REMOVED***
        var _a, _b;
        this.hasRef = true;
        (_b = (_a = this.timerId).ref) === null || _b === void 0 ? void 0 : _b.call(_a);
  ***REMOVED***
    /**
     * Set that while the timer is running, it should not keep the Node process
     * running.
     */
    unref() ***REMOVED***
        var _a, _b;
        this.hasRef = false;
        (_b = (_a = this.timerId).unref) === null || _b === void 0 ? void 0 : _b.call(_a);
  ***REMOVED***
}
exports.BackoffTimeout = BackoffTimeout;
