// @grpc/grpc-js/build/src/call-interface.js
"use strict";
/*
 * Copyright 2022 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.InterceptingListenerImpl = exports.isInterceptingListener = void 0;
function isInterceptingListener(listener) ***REMOVED***
    return (listener.onReceiveMetadata !== undefined &&
        listener.onReceiveMetadata.length === 1);
}
exports.isInterceptingListener = isInterceptingListener;
class InterceptingListenerImpl ***REMOVED***
    constructor(listener, nextListener) ***REMOVED***
        this.listener = listener;
        this.nextListener = nextListener;
        this.processingMetadata = false;
        this.hasPendingMessage = false;
        this.processingMessage = false;
        this.pendingStatus = null;
  ***REMOVED***
    processPendingMessage() ***REMOVED***
        if (this.hasPendingMessage) ***REMOVED***
            this.nextListener.onReceiveMessage(this.pendingMessage);
            this.pendingMessage = null;
            this.hasPendingMessage = false;
      ***REMOVED***
  ***REMOVED***
    processPendingStatus() ***REMOVED***
        if (this.pendingStatus) ***REMOVED***
            this.nextListener.onReceiveStatus(this.pendingStatus);
      ***REMOVED***
  ***REMOVED***
    onReceiveMetadata(metadata) ***REMOVED***
        this.processingMetadata = true;
        this.listener.onReceiveMetadata(metadata, (metadata) => ***REMOVED***
            this.processingMetadata = false;
            this.nextListener.onReceiveMetadata(metadata);
            this.processPendingMessage();
            this.processPendingStatus();
      ***REMOVED***);
  ***REMOVED***
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    onReceiveMessage(message) ***REMOVED***
        /* If this listener processes messages asynchronously, the last message may
          * be reordered with respect to the status */
        this.processingMessage = true;
        this.listener.onReceiveMessage(message, (msg) => ***REMOVED***
            this.processingMessage = false;
            if (this.processingMetadata) ***REMOVED***
                this.pendingMessage = msg;
                this.hasPendingMessage = true;
          ***REMOVED***
            else ***REMOVED***
                this.nextListener.onReceiveMessage(msg);
                this.processPendingStatus();
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***
    onReceiveStatus(status) ***REMOVED***
        this.listener.onReceiveStatus(status, (processedStatus) => ***REMOVED***
            if (this.processingMetadata || this.processingMessage) ***REMOVED***
                this.pendingStatus = processedStatus;
          ***REMOVED***
            else ***REMOVED***
                this.nextListener.onReceiveStatus(processedStatus);
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***
}
exports.InterceptingListenerImpl = InterceptingListenerImpl;
