// @grpc/grpc-js/build/src/channel-credentials.js
"use strict";
/*
 * Copyright 2019 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.ChannelCredentials = void 0;
const tls_1 = require("tls");
const call_credentials_1 = require("./call-credentials");
const tls_helpers_1 = require("./tls-helpers");
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function verifyIsBufferOrNull(obj, friendlyName) ***REMOVED***
    if (obj && !(obj instanceof Buffer)) ***REMOVED***
        throw new TypeError(`$***REMOVED***friendlyName}, if provided, must be a Buffer.`);
  ***REMOVED***
}
function bufferOrNullEqual(buf1, buf2) ***REMOVED***
    if (buf1 === null && buf2 === null) ***REMOVED***
        return true;
  ***REMOVED***
    else ***REMOVED***
        return buf1 !== null && buf2 !== null && buf1.equals(buf2);
  ***REMOVED***
}
/**
 * A class that contains credentials for communicating over a channel, as well
 * as a set of per-call credentials, which are applied to every method call made
 * over a channel initialized with an instance of this class.
 */
class ChannelCredentials ***REMOVED***
    constructor(callCredentials) ***REMOVED***
        this.callCredentials = callCredentials || call_credentials_1.CallCredentials.createEmpty();
  ***REMOVED***
    /**
     * Gets the set of per-call credentials associated with this instance.
     */
    _getCallCredentials() ***REMOVED***
        return this.callCredentials;
  ***REMOVED***
    /**
     * Return a new ChannelCredentials instance with a given set of credentials.
     * The resulting instance can be used to construct a Channel that communicates
     * over TLS.
     * @param rootCerts The root certificate data.
     * @param privateKey The client certificate private key, if available.
     * @param certChain The client certificate key chain, if available.
     * @param verifyOptions Additional options to modify certificate verification
     */
    static createSsl(rootCerts, privateKey, certChain, verifyOptions) ***REMOVED***
        var _a;
        verifyIsBufferOrNull(rootCerts, 'Root certificate');
        verifyIsBufferOrNull(privateKey, 'Private key');
        verifyIsBufferOrNull(certChain, 'Certificate chain');
        if (privateKey && !certChain) ***REMOVED***
            throw new Error('Private key must be given with accompanying certificate chain');
      ***REMOVED***
        if (!privateKey && certChain) ***REMOVED***
            throw new Error('Certificate chain must be given with accompanying private key');
      ***REMOVED***
        const secureContext = tls_1.createSecureContext(***REMOVED***
            ca: (_a = rootCerts !== null && rootCerts !== void 0 ? rootCerts : tls_helpers_1.getDefaultRootsData()) !== null && _a !== void 0 ? _a : undefined,
            key: privateKey !== null && privateKey !== void 0 ? privateKey : undefined,
            cert: certChain !== null && certChain !== void 0 ? certChain : undefined,
            ciphers: tls_helpers_1.CIPHER_SUITES,
      ***REMOVED***);
        return new SecureChannelCredentialsImpl(secureContext, verifyOptions !== null && verifyOptions !== void 0 ? verifyOptions : ***REMOVED***});
  ***REMOVED***
    /**
     * Return a new ChannelCredentials instance with credentials created using
     * the provided secureContext. The resulting instances can be used to
     * construct a Channel that communicates over TLS. gRPC will not override
     * anything in the provided secureContext, so the environment variables
     * GRPC_SSL_CIPHER_SUITES and GRPC_DEFAULT_SSL_ROOTS_FILE_PATH will
     * not be applied.
     * @param secureContext The return value of tls.createSecureContext()
     * @param verifyOptions Additional options to modify certificate verification
     */
    static createFromSecureContext(secureContext, verifyOptions) ***REMOVED***
        return new SecureChannelCredentialsImpl(secureContext, verifyOptions !== null && verifyOptions !== void 0 ? verifyOptions : ***REMOVED***});
  ***REMOVED***
    /**
     * Return a new ChannelCredentials instance with no credentials.
     */
    static createInsecure() ***REMOVED***
        return new InsecureChannelCredentialsImpl();
  ***REMOVED***
}
exports.ChannelCredentials = ChannelCredentials;
class InsecureChannelCredentialsImpl extends ChannelCredentials ***REMOVED***
    constructor(callCredentials) ***REMOVED***
        super(callCredentials);
  ***REMOVED***
    compose(callCredentials) ***REMOVED***
        throw new Error('Cannot compose insecure credentials');
  ***REMOVED***
    _getConnectionOptions() ***REMOVED***
        return null;
  ***REMOVED***
    _isSecure() ***REMOVED***
        return false;
  ***REMOVED***
    _equals(other) ***REMOVED***
        return other instanceof InsecureChannelCredentialsImpl;
  ***REMOVED***
}
class SecureChannelCredentialsImpl extends ChannelCredentials ***REMOVED***
    constructor(secureContext, verifyOptions) ***REMOVED***
        super();
        this.secureContext = secureContext;
        this.verifyOptions = verifyOptions;
        this.connectionOptions = ***REMOVED***
            secureContext
      ***REMOVED***;
        // Node asserts that this option is a function, so we cannot pass undefined
        if (verifyOptions === null || verifyOptions === void 0 ? void 0 : verifyOptions.checkServerIdentity) ***REMOVED***
            this.connectionOptions.checkServerIdentity = verifyOptions.checkServerIdentity;
      ***REMOVED***
  ***REMOVED***
    compose(callCredentials) ***REMOVED***
        const combinedCallCredentials = this.callCredentials.compose(callCredentials);
        return new ComposedChannelCredentialsImpl(this, combinedCallCredentials);
  ***REMOVED***
    _getConnectionOptions() ***REMOVED***
        // Copy to prevent callers from mutating this.connectionOptions
        return Object.assign(***REMOVED***}, this.connectionOptions);
  ***REMOVED***
    _isSecure() ***REMOVED***
        return true;
  ***REMOVED***
    _equals(other) ***REMOVED***
        if (this === other) ***REMOVED***
            return true;
      ***REMOVED***
        if (other instanceof SecureChannelCredentialsImpl) ***REMOVED***
            return (this.secureContext === other.secureContext &&
                this.verifyOptions.checkServerIdentity === other.verifyOptions.checkServerIdentity);
      ***REMOVED***
        else ***REMOVED***
            return false;
      ***REMOVED***
  ***REMOVED***
}
class ComposedChannelCredentialsImpl extends ChannelCredentials ***REMOVED***
    constructor(channelCredentials, callCreds) ***REMOVED***
        super(callCreds);
        this.channelCredentials = channelCredentials;
  ***REMOVED***
    compose(callCredentials) ***REMOVED***
        const combinedCallCredentials = this.callCredentials.compose(callCredentials);
        return new ComposedChannelCredentialsImpl(this.channelCredentials, combinedCallCredentials);
  ***REMOVED***
    _getConnectionOptions() ***REMOVED***
        return this.channelCredentials._getConnectionOptions();
  ***REMOVED***
    _isSecure() ***REMOVED***
        return true;
  ***REMOVED***
    _equals(other) ***REMOVED***
        if (this === other) ***REMOVED***
            return true;
      ***REMOVED***
        if (other instanceof ComposedChannelCredentialsImpl) ***REMOVED***
            return (this.channelCredentials._equals(other.channelCredentials) &&
                this.callCredentials._equals(other.callCredentials));
      ***REMOVED***
        else ***REMOVED***
            return false;
      ***REMOVED***
  ***REMOVED***
}
