// @grpc/grpc-js/build/src/index.js
"use strict";
/*
 * Copyright 2019 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.experimental = exports.StatusBuilder = exports.getClientChannel = exports.ServerCredentials = exports.Server = exports.setLogVerbosity = exports.setLogger = exports.load = exports.loadObject = exports.CallCredentials = exports.ChannelCredentials = exports.waitForClientReady = exports.closeClient = exports.Channel = exports.makeGenericClientConstructor = exports.makeClientConstructor = exports.loadPackageDefinition = exports.Client = exports.compressionAlgorithms = exports.propagate = exports.connectivityState = exports.status = exports.logVerbosity = exports.Metadata = exports.credentials = void 0;
const call_credentials_1 = require("./call-credentials");
Object.defineProperty(exports, "CallCredentials", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return call_credentials_1.CallCredentials; } });
const channel_1 = require("./channel");
Object.defineProperty(exports, "Channel", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return channel_1.ChannelImplementation; } });
const compression_algorithms_1 = require("./compression-algorithms");
Object.defineProperty(exports, "compressionAlgorithms", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return compression_algorithms_1.CompressionAlgorithms; } });
const connectivity_state_1 = require("./connectivity-state");
Object.defineProperty(exports, "connectivityState", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return connectivity_state_1.ConnectivityState; } });
const channel_credentials_1 = require("./channel-credentials");
Object.defineProperty(exports, "ChannelCredentials", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return channel_credentials_1.ChannelCredentials; } });
const client_1 = require("./client");
Object.defineProperty(exports, "Client", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return client_1.Client; } });
const constants_1 = require("./constants");
Object.defineProperty(exports, "logVerbosity", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return constants_1.LogVerbosity; } });
Object.defineProperty(exports, "status", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return constants_1.Status; } });
Object.defineProperty(exports, "propagate", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return constants_1.Propagate; } });
const logging = require("./logging");
const make_client_1 = require("./make-client");
Object.defineProperty(exports, "loadPackageDefinition", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return make_client_1.loadPackageDefinition; } });
Object.defineProperty(exports, "makeClientConstructor", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return make_client_1.makeClientConstructor; } });
Object.defineProperty(exports, "makeGenericClientConstructor", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return make_client_1.makeClientConstructor; } });
const metadata_1 = require("./metadata");
Object.defineProperty(exports, "Metadata", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return metadata_1.Metadata; } });
const server_1 = require("./server");
Object.defineProperty(exports, "Server", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return server_1.Server; } });
const server_credentials_1 = require("./server-credentials");
Object.defineProperty(exports, "ServerCredentials", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return server_credentials_1.ServerCredentials; } });
const status_builder_1 = require("./status-builder");
Object.defineProperty(exports, "StatusBuilder", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return status_builder_1.StatusBuilder; } });
/**** Client Credentials ****/
// Using assign only copies enumerable properties, which is what we want
exports.credentials = ***REMOVED***
    /**
     * Combine a ChannelCredentials with any number of CallCredentials into a
     * single ChannelCredentials object.
     * @param channelCredentials The ChannelCredentials object.
     * @param callCredentials Any number of CallCredentials objects.
     * @return The resulting ChannelCredentials object.
     */
    combineChannelCredentials: (channelCredentials, ...callCredentials) => ***REMOVED***
        return callCredentials.reduce((acc, other) => acc.compose(other), channelCredentials);
  ***REMOVED***,
    /**
     * Combine any number of CallCredentials into a single CallCredentials
     * object.
     * @param first The first CallCredentials object.
     * @param additional Any number of additional CallCredentials objects.
     * @return The resulting CallCredentials object.
     */
    combineCallCredentials: (first, ...additional) => ***REMOVED***
        return additional.reduce((acc, other) => acc.compose(other), first);
  ***REMOVED***,
    // from channel-credentials.ts
    createInsecure: channel_credentials_1.ChannelCredentials.createInsecure,
    createSsl: channel_credentials_1.ChannelCredentials.createSsl,
    createFromSecureContext: channel_credentials_1.ChannelCredentials.createFromSecureContext,
    // from call-credentials.ts
    createFromMetadataGenerator: call_credentials_1.CallCredentials.createFromMetadataGenerator,
    createFromGoogleCredential: call_credentials_1.CallCredentials.createFromGoogleCredential,
    createEmpty: call_credentials_1.CallCredentials.createEmpty,
};
/**
 * Close a Client object.
 * @param client The client to close.
 */
exports.closeClient = (client) => client.close();
exports.waitForClientReady = (client, deadline, callback) => client.waitForReady(deadline, callback);
/* eslint-enable @typescript-eslint/no-explicit-any */
/**** Unimplemented function stubs ****/
/* eslint-disable @typescript-eslint/no-explicit-any */
exports.loadObject = (value, options) => ***REMOVED***
    throw new Error('Not available in this library. Use @grpc/proto-loader and loadPackageDefinition instead');
};
exports.load = (filename, format, options) => ***REMOVED***
    throw new Error('Not available in this library. Use @grpc/proto-loader and loadPackageDefinition instead');
};
exports.setLogger = (logger) => ***REMOVED***
    logging.setLogger(logger);
};
exports.setLogVerbosity = (verbosity) => ***REMOVED***
    logging.setLoggerVerbosity(verbosity);
};
exports.getClientChannel = (client) => ***REMOVED***
    return client_1.Client.prototype.getChannel.call(client);
};
var client_interceptors_1 = require("./client-interceptors");
Object.defineProperty(exports, "ListenerBuilder", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return client_interceptors_1.ListenerBuilder; } });
Object.defineProperty(exports, "RequesterBuilder", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return client_interceptors_1.RequesterBuilder; } });
Object.defineProperty(exports, "InterceptingCall", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return client_interceptors_1.InterceptingCall; } });
Object.defineProperty(exports, "InterceptorConfigurationError", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return client_interceptors_1.InterceptorConfigurationError; } });
var channelz_1 = require("./channelz");
Object.defineProperty(exports, "getChannelzServiceDefinition", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return channelz_1.getChannelzServiceDefinition; } });
Object.defineProperty(exports, "getChannelzHandlers", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return channelz_1.getChannelzHandlers; } });
var admin_1 = require("./admin");
Object.defineProperty(exports, "addAdminServicesToServer", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return admin_1.addAdminServicesToServer; } });
const experimental = require("./experimental");
exports.experimental = experimental;
const resolver_dns = require("./resolver-dns");
const resolver_uds = require("./resolver-uds");
const resolver_ip = require("./resolver-ip");
const load_balancer_pick_first = require("./load-balancer-pick-first");
const load_balancer_round_robin = require("./load-balancer-round-robin");
const load_balancer_outlier_detection = require("./load-balancer-outlier-detection");
const channelz = require("./channelz");
const clientVersion = require('../../package.json').version;
(() => ***REMOVED***
    logging.trace(constants_1.LogVerbosity.DEBUG, 'index', 'Loading @grpc/grpc-js version ' + clientVersion);
    resolver_dns.setup();
    resolver_uds.setup();
    resolver_ip.setup();
    load_balancer_pick_first.setup();
    load_balancer_round_robin.setup();
    load_balancer_outlier_detection.setup();
    channelz.setup();
})();
