// @grpc/grpc-js/build/src/picker.js
"use strict";
/*
 * Copyright 2019 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.QueuePicker = exports.UnavailablePicker = exports.PickResultType = void 0;
const metadata_1 = require("./metadata");
const constants_1 = require("./constants");
var PickResultType;
(function (PickResultType) ***REMOVED***
    PickResultType[PickResultType["COMPLETE"] = 0] = "COMPLETE";
    PickResultType[PickResultType["QUEUE"] = 1] = "QUEUE";
    PickResultType[PickResultType["TRANSIENT_FAILURE"] = 2] = "TRANSIENT_FAILURE";
    PickResultType[PickResultType["DROP"] = 3] = "DROP";
})(PickResultType = exports.PickResultType || (exports.PickResultType = ***REMOVED***}));
/**
 * A standard picker representing a load balancer in the TRANSIENT_FAILURE
 * state. Always responds to every pick request with an UNAVAILABLE status.
 */
class UnavailablePicker ***REMOVED***
    constructor(status) ***REMOVED***
        if (status !== undefined) ***REMOVED***
            this.status = status;
      ***REMOVED***
        else ***REMOVED***
            this.status = ***REMOVED***
                code: constants_1.Status.UNAVAILABLE,
                details: 'No connection established',
                metadata: new metadata_1.Metadata(),
          ***REMOVED***;
      ***REMOVED***
  ***REMOVED***
    pick(pickArgs) ***REMOVED***
        return ***REMOVED***
            pickResultType: PickResultType.TRANSIENT_FAILURE,
            subchannel: null,
            status: this.status,
            extraFilterFactories: [],
            onCallStarted: null,
      ***REMOVED***;
  ***REMOVED***
}
exports.UnavailablePicker = UnavailablePicker;
/**
 * A standard picker representing a load balancer in the IDLE or CONNECTING
 * state. Always responds to every pick request with a QUEUE pick result
 * indicating that the pick should be tried again with the next `Picker`. Also
 * reports back to the load balancer that a connection should be established
 * once any pick is attempted.
 */
class QueuePicker ***REMOVED***
    // Constructed with a load balancer. Calls exitIdle on it the first time pick is called
    constructor(loadBalancer) ***REMOVED***
        this.loadBalancer = loadBalancer;
        this.calledExitIdle = false;
  ***REMOVED***
    pick(pickArgs) ***REMOVED***
        if (!this.calledExitIdle) ***REMOVED***
            process.nextTick(() => ***REMOVED***
                this.loadBalancer.exitIdle();
          ***REMOVED***);
            this.calledExitIdle = true;
      ***REMOVED***
        return ***REMOVED***
            pickResultType: PickResultType.QUEUE,
            subchannel: null,
            status: null,
            extraFilterFactories: [],
            onCallStarted: null,
      ***REMOVED***;
  ***REMOVED***
}
exports.QueuePicker = QueuePicker;
