// @grpc/grpc-js/build/src/resolver-dns.js
"use strict";
/*
 * Copyright 2019 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.setup = void 0;
const resolver_1 = require("./resolver");
const dns = require("dns");
const util = require("util");
const service_config_1 = require("./service-config");
const constants_1 = require("./constants");
const metadata_1 = require("./metadata");
const logging = require("./logging");
const constants_2 = require("./constants");
const uri_parser_1 = require("./uri-parser");
const net_1 = require("net");
const backoff_timeout_1 = require("./backoff-timeout");
const TRACER_NAME = 'dns_resolver';
function trace(text) ***REMOVED***
    logging.trace(constants_2.LogVerbosity.DEBUG, TRACER_NAME, text);
}
/**
 * The default TCP port to connect to if not explicitly specified in the target.
 */
const DEFAULT_PORT = 443;
const DEFAULT_MIN_TIME_BETWEEN_RESOLUTIONS_MS = 30000;
const resolveTxtPromise = util.promisify(dns.resolveTxt);
const dnsLookupPromise = util.promisify(dns.lookup);
/**
 * Merge any number of arrays into a single alternating array
 * @param arrays
 */
function mergeArrays(...arrays) ***REMOVED***
    const result = [];
    for (let i = 0; i <
        Math.max.apply(null, arrays.map((array) => array.length)); i++) ***REMOVED***
        for (const array of arrays) ***REMOVED***
            if (i < array.length) ***REMOVED***
                result.push(array[i]);
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    return result;
}
/**
 * Resolver implementation that handles DNS names and IP addresses.
 */
class DnsResolver ***REMOVED***
    constructor(target, listener, channelOptions) ***REMOVED***
        var _a, _b, _c;
        this.target = target;
        this.listener = listener;
        this.pendingLookupPromise = null;
        this.pendingTxtPromise = null;
        this.latestLookupResult = null;
        this.latestServiceConfig = null;
        this.latestServiceConfigError = null;
        this.continueResolving = false;
        this.isNextResolutionTimerRunning = false;
        trace('Resolver constructed for target ' + uri_parser_1.uriToString(target));
        const hostPort = uri_parser_1.splitHostPort(target.path);
        if (hostPort === null) ***REMOVED***
            this.ipResult = null;
            this.dnsHostname = null;
            this.port = null;
      ***REMOVED***
        else ***REMOVED***
            if (net_1.isIPv4(hostPort.host) || net_1.isIPv6(hostPort.host)) ***REMOVED***
                this.ipResult = [
                    ***REMOVED***
                        host: hostPort.host,
                        port: (_a = hostPort.port) !== null && _a !== void 0 ? _a : DEFAULT_PORT,
                  ***REMOVED***,
                ];
                this.dnsHostname = null;
                this.port = null;
          ***REMOVED***
            else ***REMOVED***
                this.ipResult = null;
                this.dnsHostname = hostPort.host;
                this.port = (_b = hostPort.port) !== null && _b !== void 0 ? _b : DEFAULT_PORT;
          ***REMOVED***
      ***REMOVED***
        this.percentage = Math.random() * 100;
        this.defaultResolutionError = ***REMOVED***
            code: constants_1.Status.UNAVAILABLE,
            details: `Name resolution failed for target $***REMOVED***uri_parser_1.uriToString(this.target)}`,
            metadata: new metadata_1.Metadata(),
      ***REMOVED***;
        const backoffOptions = ***REMOVED***
            initialDelay: channelOptions['grpc.initial_reconnect_backoff_ms'],
            maxDelay: channelOptions['grpc.max_reconnect_backoff_ms'],
      ***REMOVED***;
        this.backoff = new backoff_timeout_1.BackoffTimeout(() => ***REMOVED***
            if (this.continueResolving) ***REMOVED***
                this.startResolutionWithBackoff();
          ***REMOVED***
      ***REMOVED***, backoffOptions);
        this.backoff.unref();
        this.minTimeBetweenResolutionsMs = (_c = channelOptions['grpc.dns_min_time_between_resolutions_ms']) !== null && _c !== void 0 ? _c : DEFAULT_MIN_TIME_BETWEEN_RESOLUTIONS_MS;
        this.nextResolutionTimer = setTimeout(() => ***REMOVED*** }, 0);
        clearTimeout(this.nextResolutionTimer);
  ***REMOVED***
    /**
     * If the target is an IP address, just provide that address as a result.
     * Otherwise, initiate A, AAAA, and TXT lookups
     */
    startResolution() ***REMOVED***
        if (this.ipResult !== null) ***REMOVED***
            trace('Returning IP address for target ' + uri_parser_1.uriToString(this.target));
            setImmediate(() => ***REMOVED***
                this.listener.onSuccessfulResolution(this.ipResult, null, null, null, ***REMOVED***});
          ***REMOVED***);
            this.backoff.stop();
            this.backoff.reset();
            return;
      ***REMOVED***
        if (this.dnsHostname === null) ***REMOVED***
            trace('Failed to parse DNS address ' + uri_parser_1.uriToString(this.target));
            setImmediate(() => ***REMOVED***
                this.listener.onError(***REMOVED***
                    code: constants_1.Status.UNAVAILABLE,
                    details: `Failed to parse DNS address $***REMOVED***uri_parser_1.uriToString(this.target)}`,
                    metadata: new metadata_1.Metadata(),
              ***REMOVED***);
          ***REMOVED***);
            this.stopNextResolutionTimer();
      ***REMOVED***
        else ***REMOVED***
            if (this.pendingLookupPromise !== null) ***REMOVED***
                return;
          ***REMOVED***
            trace('Looking up DNS hostname ' + this.dnsHostname);
            /* We clear out latestLookupResult here to ensure that it contains the
             * latest result since the last time we started resolving. That way, the
             * TXT resolution handler can use it, but only if it finishes second. We
             * don't clear out any previous service config results because it's
             * better to use a service config that's slightly out of date than to
             * revert to an effectively blank one. */
            this.latestLookupResult = null;
            const hostname = this.dnsHostname;
            /* We lookup both address families here and then split them up later
             * because when looking up a single family, dns.lookup outputs an error
             * if the name exists but there are no records for that family, and that
             * error is indistinguishable from other kinds of errors */
            this.pendingLookupPromise = dnsLookupPromise(hostname, ***REMOVED*** all: true });
            this.pendingLookupPromise.then((addressList) => ***REMOVED***
                this.pendingLookupPromise = null;
                this.backoff.reset();
                this.backoff.stop();
                const ip4Addresses = addressList.filter((addr) => addr.family === 4);
                const ip6Addresses = addressList.filter((addr) => addr.family === 6);
                this.latestLookupResult = mergeArrays(ip6Addresses, ip4Addresses).map((addr) => (***REMOVED*** host: addr.address, port: +this.port }));
                const allAddressesString = '[' +
                    this.latestLookupResult
                        .map((addr) => addr.host + ':' + addr.port)
                        .join(',') +
                    ']';
                trace('Resolved addresses for target ' +
                    uri_parser_1.uriToString(this.target) +
                    ': ' +
                    allAddressesString);
                if (this.latestLookupResult.length === 0) ***REMOVED***
                    this.listener.onError(this.defaultResolutionError);
                    return;
              ***REMOVED***
                /* If the TXT lookup has not yet finished, both of the last two
                 * arguments will be null, which is the equivalent of getting an
                 * empty TXT response. When the TXT lookup does finish, its handler
                 * can update the service config by using the same address list */
                this.listener.onSuccessfulResolution(this.latestLookupResult, this.latestServiceConfig, this.latestServiceConfigError, null, ***REMOVED***});
          ***REMOVED***, (err) => ***REMOVED***
                trace('Resolution error for target ' +
                    uri_parser_1.uriToString(this.target) +
                    ': ' +
                    err.message);
                this.pendingLookupPromise = null;
                this.stopNextResolutionTimer();
                this.listener.onError(this.defaultResolutionError);
          ***REMOVED***);
            /* If there already is a still-pending TXT resolution, we can just use
             * that result when it comes in */
            if (this.pendingTxtPromise === null) ***REMOVED***
                /* We handle the TXT query promise differently than the others because
                 * the name resolution attempt as a whole is a success even if the TXT
                 * lookup fails */
                this.pendingTxtPromise = resolveTxtPromise(hostname);
                this.pendingTxtPromise.then((txtRecord) => ***REMOVED***
                    this.pendingTxtPromise = null;
                    try ***REMOVED***
                        this.latestServiceConfig = service_config_1.extractAndSelectServiceConfig(txtRecord, this.percentage);
                  ***REMOVED***
                    catch (err) ***REMOVED***
                        this.latestServiceConfigError = ***REMOVED***
                            code: constants_1.Status.UNAVAILABLE,
                            details: 'Parsing service config failed',
                            metadata: new metadata_1.Metadata(),
                      ***REMOVED***;
                  ***REMOVED***
                    if (this.latestLookupResult !== null) ***REMOVED***
                        /* We rely here on the assumption that calling this function with
                         * identical parameters will be essentialy idempotent, and calling
                         * it with the same address list and a different service config
                         * should result in a fast and seamless switchover. */
                        this.listener.onSuccessfulResolution(this.latestLookupResult, this.latestServiceConfig, this.latestServiceConfigError, null, ***REMOVED***});
                  ***REMOVED***
              ***REMOVED***, (err) => ***REMOVED***
                    /* If TXT lookup fails we should do nothing, which means that we
                     * continue to use the result of the most recent successful lookup,
                     * or the default null config object if there has never been a
                     * successful lookup. We do not set the latestServiceConfigError
                     * here because that is specifically used for response validation
                     * errors. We still need to handle this error so that it does not
                     * bubble up as an unhandled promise rejection. */
              ***REMOVED***);
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    startNextResolutionTimer() ***REMOVED***
        var _a, _b;
        clearTimeout(this.nextResolutionTimer);
        this.nextResolutionTimer = (_b = (_a = setTimeout(() => ***REMOVED***
            this.stopNextResolutionTimer();
            if (this.continueResolving) ***REMOVED***
                this.startResolutionWithBackoff();
          ***REMOVED***
      ***REMOVED***, this.minTimeBetweenResolutionsMs)).unref) === null || _b === void 0 ? void 0 : _b.call(_a);
        this.isNextResolutionTimerRunning = true;
  ***REMOVED***
    stopNextResolutionTimer() ***REMOVED***
        clearTimeout(this.nextResolutionTimer);
        this.isNextResolutionTimerRunning = false;
  ***REMOVED***
    startResolutionWithBackoff() ***REMOVED***
        if (this.pendingLookupPromise === null) ***REMOVED***
            this.continueResolving = false;
            this.startResolution();
            this.backoff.runOnce();
            this.startNextResolutionTimer();
      ***REMOVED***
  ***REMOVED***
    updateResolution() ***REMOVED***
        /* If there is a pending lookup, just let it finish. Otherwise, if the
         * nextResolutionTimer or backoff timer is running, set the
         * continueResolving flag to resolve when whichever of those timers
         * fires. Otherwise, start resolving immediately. */
        if (this.pendingLookupPromise === null) ***REMOVED***
            if (this.isNextResolutionTimerRunning || this.backoff.isRunning()) ***REMOVED***
                this.continueResolving = true;
          ***REMOVED***
            else ***REMOVED***
                this.startResolutionWithBackoff();
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
    destroy() ***REMOVED***
        this.continueResolving = false;
        this.backoff.stop();
        this.stopNextResolutionTimer();
  ***REMOVED***
    /**
     * Get the default authority for the given target. For IP targets, that is
     * the IP address. For DNS targets, it is the hostname.
     * @param target
     */
    static getDefaultAuthority(target) ***REMOVED***
        return target.path;
  ***REMOVED***
}
/**
 * Set up the DNS resolver class by registering it as the handler for the
 * "dns:" prefix and as the default resolver.
 */
function setup() ***REMOVED***
    resolver_1.registerResolver('dns', DnsResolver);
    resolver_1.registerDefaultScheme('dns');
}
exports.setup = setup;
