// @grpc/grpc-js/build/src/resolving-load-balancer.js
"use strict";
/*
 * Copyright 2019 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.ResolvingLoadBalancer = void 0;
const load_balancer_1 = require("./load-balancer");
const service_config_1 = require("./service-config");
const connectivity_state_1 = require("./connectivity-state");
const resolver_1 = require("./resolver");
const picker_1 = require("./picker");
const backoff_timeout_1 = require("./backoff-timeout");
const constants_1 = require("./constants");
const metadata_1 = require("./metadata");
const logging = require("./logging");
const constants_2 = require("./constants");
const uri_parser_1 = require("./uri-parser");
const load_balancer_child_handler_1 = require("./load-balancer-child-handler");
const TRACER_NAME = 'resolving_load_balancer';
function trace(text) ***REMOVED***
    logging.trace(constants_2.LogVerbosity.DEBUG, TRACER_NAME, text);
}
const DEFAULT_LOAD_BALANCER_NAME = 'pick_first';
function getDefaultConfigSelector(serviceConfig) ***REMOVED***
    return function defaultConfigSelector(methodName, metadata) ***REMOVED***
        var _a, _b;
        const splitName = methodName.split('/').filter((x) => x.length > 0);
        const service = (_a = splitName[0]) !== null && _a !== void 0 ? _a : '';
        const method = (_b = splitName[1]) !== null && _b !== void 0 ? _b : '';
        if (serviceConfig && serviceConfig.methodConfig) ***REMOVED***
            for (const methodConfig of serviceConfig.methodConfig) ***REMOVED***
                for (const name of methodConfig.name) ***REMOVED***
                    if (name.service === service &&
                        (name.method === undefined || name.method === method)) ***REMOVED***
                        return ***REMOVED***
                            methodConfig: methodConfig,
                            pickInformation: ***REMOVED***},
                            status: constants_1.Status.OK,
                            dynamicFilterFactories: []
                      ***REMOVED***;
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        return ***REMOVED***
            methodConfig: ***REMOVED*** name: [] },
            pickInformation: ***REMOVED***},
            status: constants_1.Status.OK,
            dynamicFilterFactories: []
      ***REMOVED***;
  ***REMOVED***;
}
class ResolvingLoadBalancer ***REMOVED***
    /**
     * Wrapper class that behaves like a `LoadBalancer` and also handles name
     * resolution internally.
     * @param target The address of the backend to connect to.
     * @param channelControlHelper `ChannelControlHelper` instance provided by
     *     this load balancer's owner.
     * @param defaultServiceConfig The default service configuration to be used
     *     if none is provided by the name resolver. A `null` value indicates
     *     that the default behavior should be the default unconfigured behavior.
     *     In practice, that means using the "pick first" load balancer
     *     implmentation
     */
    constructor(target, channelControlHelper, channelOptions, onSuccessfulResolution, onFailedResolution) ***REMOVED***
        this.target = target;
        this.channelControlHelper = channelControlHelper;
        this.channelOptions = channelOptions;
        this.onSuccessfulResolution = onSuccessfulResolution;
        this.onFailedResolution = onFailedResolution;
        this.latestChildState = connectivity_state_1.ConnectivityState.IDLE;
        this.latestChildPicker = new picker_1.QueuePicker(this);
        /**
         * This resolving load balancer's current connectivity state.
         */
        this.currentState = connectivity_state_1.ConnectivityState.IDLE;
        /**
         * The service config object from the last successful resolution, if
         * available. A value of null indicates that we have not yet received a valid
         * service config from the resolver.
         */
        this.previousServiceConfig = null;
        /**
         * Indicates whether we should attempt to resolve again after the backoff
         * timer runs out.
         */
        this.continueResolving = false;
        if (channelOptions['grpc.service_config']) ***REMOVED***
            this.defaultServiceConfig = service_config_1.validateServiceConfig(JSON.parse(channelOptions['grpc.service_config']));
      ***REMOVED***
        else ***REMOVED***
            this.defaultServiceConfig = ***REMOVED***
                loadBalancingConfig: [],
                methodConfig: [],
          ***REMOVED***;
      ***REMOVED***
        this.updateState(connectivity_state_1.ConnectivityState.IDLE, new picker_1.QueuePicker(this));
        this.childLoadBalancer = new load_balancer_child_handler_1.ChildLoadBalancerHandler(***REMOVED***
            createSubchannel: channelControlHelper.createSubchannel.bind(channelControlHelper),
            requestReresolution: () => ***REMOVED***
                /* If the backoffTimeout is running, we're still backing off from
                 * making resolve requests, so we shouldn't make another one here.
                 * In that case, the backoff timer callback will call
                 * updateResolution */
                if (this.backoffTimeout.isRunning()) ***REMOVED***
                    this.continueResolving = true;
              ***REMOVED***
                else ***REMOVED***
                    this.updateResolution();
              ***REMOVED***
          ***REMOVED***,
            updateState: (newState, picker) => ***REMOVED***
                this.latestChildState = newState;
                this.latestChildPicker = picker;
                this.updateState(newState, picker);
          ***REMOVED***,
            addChannelzChild: channelControlHelper.addChannelzChild.bind(channelControlHelper),
            removeChannelzChild: channelControlHelper.removeChannelzChild.bind(channelControlHelper)
      ***REMOVED***);
        this.innerResolver = resolver_1.createResolver(target, ***REMOVED***
            onSuccessfulResolution: (addressList, serviceConfig, serviceConfigError, configSelector, attributes) => ***REMOVED***
                var _a;
                let workingServiceConfig = null;
                /* This first group of conditionals implements the algorithm described
                 * in https://github.com/grpc/proposal/blob/master/A21-service-config-error-handling.md
                 * in the section called "Behavior on receiving a new gRPC Config".
                 */
                if (serviceConfig === null) ***REMOVED***
                    // Step 4 and 5
                    if (serviceConfigError === null) ***REMOVED***
                        // Step 5
                        this.previousServiceConfig = null;
                        workingServiceConfig = this.defaultServiceConfig;
                  ***REMOVED***
                    else ***REMOVED***
                        // Step 4
                        if (this.previousServiceConfig === null) ***REMOVED***
                            // Step 4.ii
                            this.handleResolutionFailure(serviceConfigError);
                      ***REMOVED***
                        else ***REMOVED***
                            // Step 4.i
                            workingServiceConfig = this.previousServiceConfig;
                      ***REMOVED***
                  ***REMOVED***
              ***REMOVED***
                else ***REMOVED***
                    // Step 3
                    workingServiceConfig = serviceConfig;
                    this.previousServiceConfig = serviceConfig;
              ***REMOVED***
                const workingConfigList = (_a = workingServiceConfig === null || workingServiceConfig === void 0 ? void 0 : workingServiceConfig.loadBalancingConfig) !== null && _a !== void 0 ? _a : [];
                const loadBalancingConfig = load_balancer_1.getFirstUsableConfig(workingConfigList, true);
                if (loadBalancingConfig === null) ***REMOVED***
                    // There were load balancing configs but none are supported. This counts as a resolution failure
                    this.handleResolutionFailure(***REMOVED***
                        code: constants_1.Status.UNAVAILABLE,
                        details: 'All load balancer options in service config are not compatible',
                        metadata: new metadata_1.Metadata(),
                  ***REMOVED***);
                    return;
              ***REMOVED***
                this.childLoadBalancer.updateAddressList(addressList, loadBalancingConfig, attributes);
                const finalServiceConfig = workingServiceConfig !== null && workingServiceConfig !== void 0 ? workingServiceConfig : this.defaultServiceConfig;
                this.onSuccessfulResolution(configSelector !== null && configSelector !== void 0 ? configSelector : getDefaultConfigSelector(finalServiceConfig));
          ***REMOVED***,
            onError: (error) => ***REMOVED***
                this.handleResolutionFailure(error);
          ***REMOVED***,
      ***REMOVED***, channelOptions);
        const backoffOptions = ***REMOVED***
            initialDelay: channelOptions['grpc.initial_reconnect_backoff_ms'],
            maxDelay: channelOptions['grpc.max_reconnect_backoff_ms'],
      ***REMOVED***;
        this.backoffTimeout = new backoff_timeout_1.BackoffTimeout(() => ***REMOVED***
            if (this.continueResolving) ***REMOVED***
                this.updateResolution();
                this.continueResolving = false;
          ***REMOVED***
            else ***REMOVED***
                this.updateState(this.latestChildState, this.latestChildPicker);
          ***REMOVED***
      ***REMOVED***, backoffOptions);
        this.backoffTimeout.unref();
  ***REMOVED***
    updateResolution() ***REMOVED***
        this.innerResolver.updateResolution();
        if (this.currentState === connectivity_state_1.ConnectivityState.IDLE) ***REMOVED***
            this.updateState(connectivity_state_1.ConnectivityState.CONNECTING, new picker_1.QueuePicker(this));
      ***REMOVED***
        this.backoffTimeout.runOnce();
  ***REMOVED***
    updateState(connectivityState, picker) ***REMOVED***
        trace(uri_parser_1.uriToString(this.target) +
            ' ' +
            connectivity_state_1.ConnectivityState[this.currentState] +
            ' -> ' +
            connectivity_state_1.ConnectivityState[connectivityState]);
        // Ensure that this.exitIdle() is called by the picker
        if (connectivityState === connectivity_state_1.ConnectivityState.IDLE) ***REMOVED***
            picker = new picker_1.QueuePicker(this);
      ***REMOVED***
        this.currentState = connectivityState;
        this.channelControlHelper.updateState(connectivityState, picker);
  ***REMOVED***
    handleResolutionFailure(error) ***REMOVED***
        if (this.latestChildState === connectivity_state_1.ConnectivityState.IDLE) ***REMOVED***
            this.updateState(connectivity_state_1.ConnectivityState.TRANSIENT_FAILURE, new picker_1.UnavailablePicker(error));
            this.onFailedResolution(error);
      ***REMOVED***
  ***REMOVED***
    exitIdle() ***REMOVED***
        if (this.currentState === connectivity_state_1.ConnectivityState.IDLE || this.currentState === connectivity_state_1.ConnectivityState.TRANSIENT_FAILURE) ***REMOVED***
            if (this.backoffTimeout.isRunning()) ***REMOVED***
                this.continueResolving = true;
          ***REMOVED***
            else ***REMOVED***
                this.updateResolution();
          ***REMOVED***
      ***REMOVED***
        this.childLoadBalancer.exitIdle();
  ***REMOVED***
    updateAddressList(addressList, lbConfig) ***REMOVED***
        throw new Error('updateAddressList not supported on ResolvingLoadBalancer');
  ***REMOVED***
    resetBackoff() ***REMOVED***
        this.backoffTimeout.reset();
        this.childLoadBalancer.resetBackoff();
  ***REMOVED***
    destroy() ***REMOVED***
        this.childLoadBalancer.destroy();
        this.innerResolver.destroy();
        this.updateState(connectivity_state_1.ConnectivityState.SHUTDOWN, new picker_1.UnavailablePicker());
  ***REMOVED***
    getTypeName() ***REMOVED***
        return 'resolving_load_balancer';
  ***REMOVED***
}
exports.ResolvingLoadBalancer = ResolvingLoadBalancer;
