// @grpc/grpc-js/build/src/subchannel-address.js
"use strict";
/*
 * Copyright 2021 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.stringToSubchannelAddress = exports.subchannelAddressToString = exports.subchannelAddressEqual = exports.isTcpSubchannelAddress = void 0;
const net_1 = require("net");
function isTcpSubchannelAddress(address) ***REMOVED***
    return 'port' in address;
}
exports.isTcpSubchannelAddress = isTcpSubchannelAddress;
function subchannelAddressEqual(address1, address2) ***REMOVED***
    if (isTcpSubchannelAddress(address1)) ***REMOVED***
        return (isTcpSubchannelAddress(address2) &&
            address1.host === address2.host &&
            address1.port === address2.port);
  ***REMOVED***
    else ***REMOVED***
        return !isTcpSubchannelAddress(address2) && address1.path === address2.path;
  ***REMOVED***
}
exports.subchannelAddressEqual = subchannelAddressEqual;
function subchannelAddressToString(address) ***REMOVED***
    if (isTcpSubchannelAddress(address)) ***REMOVED***
        return address.host + ':' + address.port;
  ***REMOVED***
    else ***REMOVED***
        return address.path;
  ***REMOVED***
}
exports.subchannelAddressToString = subchannelAddressToString;
const DEFAULT_PORT = 443;
function stringToSubchannelAddress(addressString, port) ***REMOVED***
    if (net_1.isIP(addressString)) ***REMOVED***
        return ***REMOVED***
            host: addressString,
            port: port !== null && port !== void 0 ? port : DEFAULT_PORT
      ***REMOVED***;
  ***REMOVED***
    else ***REMOVED***
        return ***REMOVED***
            path: addressString
      ***REMOVED***;
  ***REMOVED***
}
exports.stringToSubchannelAddress = stringToSubchannelAddress;
