// firebase-admin/lib/auth/tenant.js
/*! firebase-admin v11.5.0 */
"use strict";
/*!
 * Copyright 2019 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.Tenant = void 0;
const validator = require("../utils/validator");
const deep_copy_1 = require("../utils/deep-copy");
const error_1 = require("../utils/error");
const auth_config_1 = require("./auth-config");
/**
 * Represents a tenant configuration.
 *
 * Multi-tenancy support requires Google Cloud's Identity Platform
 * (GCIP). To learn more about GCIP, including pricing and features,
 * see the ***REMOVED***@link https://cloud.google.com/identity-platform | GCIP documentation}.
 *
 * Before multi-tenancy can be used on a Google Cloud Identity Platform project,
 * tenants must be allowed on that project via the Cloud Console UI.
 *
 * A tenant configuration provides information such as the display name, tenant
 * identifier and email authentication configuration.
 * For OIDC/SAML provider configuration management, `TenantAwareAuth` instances should
 * be used instead of a `Tenant` to retrieve the list of configured IdPs on a tenant.
 * When configuring these providers, note that tenants will inherit
 * whitelisted domains and authenticated redirect URIs of their parent project.
 *
 * All other settings of a tenant will also be inherited. These will need to be managed
 * from the Cloud Console UI.
 */
class Tenant ***REMOVED***
    /**
     * The Tenant object constructor.
     *
     * @param response - The server side response used to initialize the Tenant object.
     * @constructor
     * @internal
     */
    constructor(response) ***REMOVED***
        const tenantId = Tenant.getTenantIdFromResourceName(response.name);
        if (!tenantId) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INTERNAL_ERROR, 'INTERNAL ASSERT FAILED: Invalid tenant response');
      ***REMOVED***
        this.tenantId = tenantId;
        this.displayName = response.displayName;
        try ***REMOVED***
            this.emailSignInConfig_ = new auth_config_1.EmailSignInConfig(response);
      ***REMOVED***
        catch (e) ***REMOVED***
            // If allowPasswordSignup is undefined, it is disabled by default.
            this.emailSignInConfig_ = new auth_config_1.EmailSignInConfig(***REMOVED***
                allowPasswordSignup: false,
          ***REMOVED***);
      ***REMOVED***
        this.anonymousSignInEnabled = !!response.enableAnonymousUser;
        if (typeof response.mfaConfig !== 'undefined') ***REMOVED***
            this.multiFactorConfig_ = new auth_config_1.MultiFactorAuthConfig(response.mfaConfig);
      ***REMOVED***
        if (typeof response.testPhoneNumbers !== 'undefined') ***REMOVED***
            this.testPhoneNumbers = (0, deep_copy_1.deepCopy)(response.testPhoneNumbers || ***REMOVED***});
      ***REMOVED***
        if (typeof response.smsRegionConfig !== 'undefined') ***REMOVED***
            this.smsRegionConfig = (0, deep_copy_1.deepCopy)(response.smsRegionConfig);
      ***REMOVED***
  ***REMOVED***
    /**
     * Builds the corresponding server request for a TenantOptions object.
     *
     * @param tenantOptions - The properties to convert to a server request.
     * @param createRequest - Whether this is a create request.
     * @returns The equivalent server request.
     *
     * @internal
     */
    static buildServerRequest(tenantOptions, createRequest) ***REMOVED***
        Tenant.validate(tenantOptions, createRequest);
        let request = ***REMOVED***};
        if (typeof tenantOptions.emailSignInConfig !== 'undefined') ***REMOVED***
            request = auth_config_1.EmailSignInConfig.buildServerRequest(tenantOptions.emailSignInConfig);
      ***REMOVED***
        if (typeof tenantOptions.displayName !== 'undefined') ***REMOVED***
            request.displayName = tenantOptions.displayName;
      ***REMOVED***
        if (typeof tenantOptions.anonymousSignInEnabled !== 'undefined') ***REMOVED***
            request.enableAnonymousUser = tenantOptions.anonymousSignInEnabled;
      ***REMOVED***
        if (typeof tenantOptions.multiFactorConfig !== 'undefined') ***REMOVED***
            request.mfaConfig = auth_config_1.MultiFactorAuthConfig.buildServerRequest(tenantOptions.multiFactorConfig);
      ***REMOVED***
        if (typeof tenantOptions.testPhoneNumbers !== 'undefined') ***REMOVED***
            // null will clear existing test phone numbers. Translate to empty object.
            request.testPhoneNumbers = tenantOptions.testPhoneNumbers ?? ***REMOVED***};
      ***REMOVED***
        if (typeof tenantOptions.smsRegionConfig !== 'undefined') ***REMOVED***
            request.smsRegionConfig = tenantOptions.smsRegionConfig;
      ***REMOVED***
        return request;
  ***REMOVED***
    /**
     * Returns the tenant ID corresponding to the resource name if available.
     *
     * @param resourceName - The server side resource name
     * @returns The tenant ID corresponding to the resource, null otherwise.
     *
     * @internal
     */
    static getTenantIdFromResourceName(resourceName) ***REMOVED***
        // name is of form projects/project1/tenants/tenant1
        const matchTenantRes = resourceName.match(/\/tenants\/(.*)$/);
        if (!matchTenantRes || matchTenantRes.length < 2) ***REMOVED***
            return null;
      ***REMOVED***
        return matchTenantRes[1];
  ***REMOVED***
    /**
     * Validates a tenant options object. Throws an error on failure.
     *
     * @param request - The tenant options object to validate.
     * @param createRequest - Whether this is a create request.
     */
    static validate(request, createRequest) ***REMOVED***
        const validKeys = ***REMOVED***
            displayName: true,
            emailSignInConfig: true,
            anonymousSignInEnabled: true,
            multiFactorConfig: true,
            testPhoneNumbers: true,
            smsRegionConfig: true,
      ***REMOVED***;
        const label = createRequest ? 'CreateTenantRequest' : 'UpdateTenantRequest';
        if (!validator.isNonNullObject(request)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, `"$***REMOVED***label}" must be a valid non-null object.`);
      ***REMOVED***
        // Check for unsupported top level attributes.
        for (const key in request) ***REMOVED***
            if (!(key in validKeys)) ***REMOVED***
                throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, `"$***REMOVED***key}" is not a valid $***REMOVED***label} parameter.`);
          ***REMOVED***
      ***REMOVED***
        // Validate displayName type if provided.
        if (typeof request.displayName !== 'undefined' &&
            !validator.isNonEmptyString(request.displayName)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, `"$***REMOVED***label}.displayName" must be a valid non-empty string.`);
      ***REMOVED***
        // Validate emailSignInConfig type if provided.
        if (typeof request.emailSignInConfig !== 'undefined') ***REMOVED***
            // This will throw an error if invalid.
            auth_config_1.EmailSignInConfig.buildServerRequest(request.emailSignInConfig);
      ***REMOVED***
        // Validate test phone numbers if provided.
        if (typeof request.testPhoneNumbers !== 'undefined' &&
            request.testPhoneNumbers !== null) ***REMOVED***
            (0, auth_config_1.validateTestPhoneNumbers)(request.testPhoneNumbers);
      ***REMOVED***
        else if (request.testPhoneNumbers === null && createRequest) ***REMOVED***
            // null allowed only for update operations.
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, `"$***REMOVED***label}.testPhoneNumbers" must be a non-null object.`);
      ***REMOVED***
        // Validate multiFactorConfig type if provided.
        if (typeof request.multiFactorConfig !== 'undefined') ***REMOVED***
            // This will throw an error if invalid.
            auth_config_1.MultiFactorAuthConfig.buildServerRequest(request.multiFactorConfig);
      ***REMOVED***
        // Validate SMS Regions Config if provided.
        if (typeof request.smsRegionConfig != 'undefined') ***REMOVED***
            auth_config_1.SmsRegionsAuthConfig.validate(request.smsRegionConfig);
      ***REMOVED***
  ***REMOVED***
    /**
     * The email sign in provider configuration.
     */
    get emailSignInConfig() ***REMOVED***
        return this.emailSignInConfig_;
  ***REMOVED***
    /**
     * The multi-factor auth configuration on the current tenant.
     */
    get multiFactorConfig() ***REMOVED***
        return this.multiFactorConfig_;
  ***REMOVED***
    /**
     * Returns a JSON-serializable representation of this object.
     *
     * @returns A JSON-serializable representation of this object.
     */
    toJSON() ***REMOVED***
        const json = ***REMOVED***
            tenantId: this.tenantId,
            displayName: this.displayName,
            emailSignInConfig: this.emailSignInConfig_?.toJSON(),
            multiFactorConfig: this.multiFactorConfig_?.toJSON(),
            anonymousSignInEnabled: this.anonymousSignInEnabled,
            testPhoneNumbers: this.testPhoneNumbers,
            smsRegionConfig: (0, deep_copy_1.deepCopy)(this.smsRegionConfig),
      ***REMOVED***;
        if (typeof json.multiFactorConfig === 'undefined') ***REMOVED***
            delete json.multiFactorConfig;
      ***REMOVED***
        if (typeof json.testPhoneNumbers === 'undefined') ***REMOVED***
            delete json.testPhoneNumbers;
      ***REMOVED***
        if (typeof json.smsRegionConfig === 'undefined') ***REMOVED***
            delete json.smsRegionConfig;
      ***REMOVED***
        return json;
  ***REMOVED***
}
exports.Tenant = Tenant;
