// firebase-admin/lib/auth/token-verifier.js
/*! firebase-admin v11.5.0 */
"use strict";
/*!
 * Copyright 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.createSessionCookieVerifier = exports.createAuthBlockingTokenVerifier = exports.createIdTokenVerifier = exports.FirebaseTokenVerifier = exports.SESSION_COOKIE_INFO = exports.AUTH_BLOCKING_TOKEN_INFO = exports.ID_TOKEN_INFO = void 0;
const error_1 = require("../utils/error");
const util = require("../utils/index");
const validator = require("../utils/validator");
const jwt_1 = require("../utils/jwt");
// Audience to use for Firebase Auth Custom tokens
const FIREBASE_AUDIENCE = 'https://identitytoolkit.googleapis.com/google.identity.identitytoolkit.v1.IdentityToolkit';
// URL containing the public keys for the Google certs (whose private keys are used to sign Firebase
// Auth ID tokens)
const CLIENT_CERT_URL = 'https://www.googleapis.com/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com';
// URL containing the public keys for Firebase session cookies. This will be updated to a different URL soon.
const SESSION_COOKIE_CERT_URL = 'https://www.googleapis.com/identitytoolkit/v3/relyingparty/publicKeys';
const EMULATOR_VERIFIER = new jwt_1.EmulatorSignatureVerifier();
/**
 * User facing token information related to the Firebase ID token.
 *
 * @internal
 */
exports.ID_TOKEN_INFO = ***REMOVED***
    url: 'https://firebase.google.com/docs/auth/admin/verify-id-tokens',
    verifyApiName: 'verifyIdToken()',
    jwtName: 'Firebase ID token',
    shortName: 'ID token',
    expiredErrorCode: error_1.AuthClientErrorCode.ID_TOKEN_EXPIRED,
};
/**
 * User facing token information related to the Firebase Auth Blocking token.
 *
 * @internal
 */
exports.AUTH_BLOCKING_TOKEN_INFO = ***REMOVED***
    url: 'https://cloud.google.com/identity-platform/docs/blocking-functions',
    verifyApiName: '_verifyAuthBlockingToken()',
    jwtName: 'Firebase Auth Blocking token',
    shortName: 'Auth Blocking token',
    expiredErrorCode: error_1.AuthClientErrorCode.AUTH_BLOCKING_TOKEN_EXPIRED,
};
/**
 * User facing token information related to the Firebase session cookie.
 *
 * @internal
 */
exports.SESSION_COOKIE_INFO = ***REMOVED***
    url: 'https://firebase.google.com/docs/auth/admin/manage-cookies',
    verifyApiName: 'verifySessionCookie()',
    jwtName: 'Firebase session cookie',
    shortName: 'session cookie',
    expiredErrorCode: error_1.AuthClientErrorCode.SESSION_COOKIE_EXPIRED,
};
/**
 * Class for verifying general purpose Firebase JWTs. This verifies ID tokens and session cookies.
 *
 * @internal
 */
class FirebaseTokenVerifier ***REMOVED***
    constructor(clientCertUrl, issuer, tokenInfo, app) ***REMOVED***
        this.issuer = issuer;
        this.tokenInfo = tokenInfo;
        this.app = app;
        if (!validator.isURL(clientCertUrl)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, 'The provided public client certificate URL is an invalid URL.');
      ***REMOVED***
        else if (!validator.isURL(issuer)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, 'The provided JWT issuer is an invalid URL.');
      ***REMOVED***
        else if (!validator.isNonNullObject(tokenInfo)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, 'The provided JWT information is not an object or null.');
      ***REMOVED***
        else if (!validator.isURL(tokenInfo.url)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, 'The provided JWT verification documentation URL is invalid.');
      ***REMOVED***
        else if (!validator.isNonEmptyString(tokenInfo.verifyApiName)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, 'The JWT verify API name must be a non-empty string.');
      ***REMOVED***
        else if (!validator.isNonEmptyString(tokenInfo.jwtName)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, 'The JWT public full name must be a non-empty string.');
      ***REMOVED***
        else if (!validator.isNonEmptyString(tokenInfo.shortName)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, 'The JWT public short name must be a non-empty string.');
      ***REMOVED***
        else if (!validator.isNonNullObject(tokenInfo.expiredErrorCode) || !('code' in tokenInfo.expiredErrorCode)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, 'The JWT expiration error code must be a non-null ErrorInfo object.');
      ***REMOVED***
        this.shortNameArticle = tokenInfo.shortName.charAt(0).match(/[aeiou]/i) ? 'an' : 'a';
        this.signatureVerifier =
            jwt_1.PublicKeySignatureVerifier.withCertificateUrl(clientCertUrl, app.options.httpAgent);
        // For backward compatibility, the project ID is validated in the verification call.
  ***REMOVED***
    /**
     * Verifies the format and signature of a Firebase Auth JWT token.
     *
     * @param jwtToken - The Firebase Auth JWT token to verify.
     * @param isEmulator - Whether to accept Auth Emulator tokens.
     * @returns A promise fulfilled with the decoded claims of the Firebase Auth ID token.
     */
    verifyJWT(jwtToken, isEmulator = false) ***REMOVED***
        if (!validator.isString(jwtToken)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, `First argument to $***REMOVED***this.tokenInfo.verifyApiName} must be a $***REMOVED***this.tokenInfo.jwtName} string.`);
      ***REMOVED***
        return this.ensureProjectId()
            .then((projectId) => ***REMOVED***
            return this.decodeAndVerify(jwtToken, projectId, isEmulator);
      ***REMOVED***)
            .then((decoded) => ***REMOVED***
            const decodedIdToken = decoded.payload;
            decodedIdToken.uid = decodedIdToken.sub;
            return decodedIdToken;
      ***REMOVED***);
  ***REMOVED***
    /** @alpha */
    // eslint-disable-next-line @typescript-eslint/naming-convention
    _verifyAuthBlockingToken(jwtToken, isEmulator, audience) ***REMOVED***
        if (!validator.isString(jwtToken)) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, `First argument to $***REMOVED***this.tokenInfo.verifyApiName} must be a $***REMOVED***this.tokenInfo.jwtName} string.`);
      ***REMOVED***
        return this.ensureProjectId()
            .then((projectId) => ***REMOVED***
            if (typeof audience === 'undefined') ***REMOVED***
                audience = `$***REMOVED***projectId}.cloudfunctions.net/`;
          ***REMOVED***
            return this.decodeAndVerify(jwtToken, projectId, isEmulator, audience);
      ***REMOVED***)
            .then((decoded) => ***REMOVED***
            const decodedAuthBlockingToken = decoded.payload;
            decodedAuthBlockingToken.uid = decodedAuthBlockingToken.sub;
            return decodedAuthBlockingToken;
      ***REMOVED***);
  ***REMOVED***
    ensureProjectId() ***REMOVED***
        return util.findProjectId(this.app)
            .then((projectId) => ***REMOVED***
            if (!validator.isNonEmptyString(projectId)) ***REMOVED***
                throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_CREDENTIAL, 'Must initialize app with a cert credential or set your Firebase project ID as the ' +
                    `GOOGLE_CLOUD_PROJECT environment variable to call $***REMOVED***this.tokenInfo.verifyApiName}.`);
          ***REMOVED***
            return Promise.resolve(projectId);
      ***REMOVED***);
  ***REMOVED***
    decodeAndVerify(token, projectId, isEmulator, audience) ***REMOVED***
        return this.safeDecode(token)
            .then((decodedToken) => ***REMOVED***
            this.verifyContent(decodedToken, projectId, isEmulator, audience);
            return this.verifySignature(token, isEmulator)
                .then(() => decodedToken);
      ***REMOVED***);
  ***REMOVED***
    safeDecode(jwtToken) ***REMOVED***
        return (0, jwt_1.decodeJwt)(jwtToken)
            .catch((err) => ***REMOVED***
            if (err.code == jwt_1.JwtErrorCode.INVALID_ARGUMENT) ***REMOVED***
                const verifyJwtTokenDocsMessage = ` See $***REMOVED***this.tokenInfo.url} ` +
                    `for details on how to retrieve $***REMOVED***this.shortNameArticle} $***REMOVED***this.tokenInfo.shortName}.`;
                const errorMessage = `Decoding $***REMOVED***this.tokenInfo.jwtName} failed. Make sure you passed ` +
                    `the entire string JWT which represents $***REMOVED***this.shortNameArticle} ` +
                    `$***REMOVED***this.tokenInfo.shortName}.` + verifyJwtTokenDocsMessage;
                throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, errorMessage);
          ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INTERNAL_ERROR, err.message);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Verifies the content of a Firebase Auth JWT.
     *
     * @param fullDecodedToken - The decoded JWT.
     * @param projectId - The Firebase Project Id.
     * @param isEmulator - Whether the token is an Emulator token.
     */
    verifyContent(fullDecodedToken, projectId, isEmulator, audience) ***REMOVED***
        const header = fullDecodedToken && fullDecodedToken.header;
        const payload = fullDecodedToken && fullDecodedToken.payload;
        const projectIdMatchMessage = ` Make sure the $***REMOVED***this.tokenInfo.shortName} comes from the same ` +
            'Firebase project as the service account used to authenticate this SDK.';
        const verifyJwtTokenDocsMessage = ` See $***REMOVED***this.tokenInfo.url} ` +
            `for details on how to retrieve $***REMOVED***this.shortNameArticle} $***REMOVED***this.tokenInfo.shortName}.`;
        let errorMessage;
        if (!isEmulator && typeof header.kid === 'undefined') ***REMOVED***
            const isCustomToken = (payload.aud === FIREBASE_AUDIENCE);
            const isLegacyCustomToken = (header.alg === 'HS256' && payload.v === 0 && 'd' in payload && 'uid' in payload.d);
            if (isCustomToken) ***REMOVED***
                errorMessage = `$***REMOVED***this.tokenInfo.verifyApiName} expects $***REMOVED***this.shortNameArticle} ` +
                    `$***REMOVED***this.tokenInfo.shortName}, but was given a custom token.`;
          ***REMOVED***
            else if (isLegacyCustomToken) ***REMOVED***
                errorMessage = `$***REMOVED***this.tokenInfo.verifyApiName} expects $***REMOVED***this.shortNameArticle} ` +
                    `$***REMOVED***this.tokenInfo.shortName}, but was given a legacy custom token.`;
          ***REMOVED***
            else ***REMOVED***
                errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has no "kid" claim.`;
          ***REMOVED***
            errorMessage += verifyJwtTokenDocsMessage;
      ***REMOVED***
        else if (!isEmulator && header.alg !== jwt_1.ALGORITHM_RS256) ***REMOVED***
            errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has incorrect algorithm. Expected "` + jwt_1.ALGORITHM_RS256 + '" but got ' +
                '"' + header.alg + '".' + verifyJwtTokenDocsMessage;
      ***REMOVED***
        else if (typeof audience !== 'undefined' && !payload.aud.includes(audience)) ***REMOVED***
            errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has incorrect "aud" (audience) claim. Expected "` +
                audience + '" but got "' + payload.aud + '".' + verifyJwtTokenDocsMessage;
      ***REMOVED***
        else if (typeof audience === 'undefined' && payload.aud !== projectId) ***REMOVED***
            errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has incorrect "aud" (audience) claim. Expected "` +
                projectId + '" but got "' + payload.aud + '".' + projectIdMatchMessage +
                verifyJwtTokenDocsMessage;
      ***REMOVED***
        else if (payload.iss !== this.issuer + projectId) ***REMOVED***
            errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has incorrect "iss" (issuer) claim. Expected ` +
                `"$***REMOVED***this.issuer}` + projectId + '" but got "' +
                payload.iss + '".' + projectIdMatchMessage + verifyJwtTokenDocsMessage;
      ***REMOVED***
        else if (typeof payload.sub !== 'string') ***REMOVED***
            errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has no "sub" (subject) claim.` + verifyJwtTokenDocsMessage;
      ***REMOVED***
        else if (payload.sub === '') ***REMOVED***
            errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has an empty string "sub" (subject) claim.` + verifyJwtTokenDocsMessage;
      ***REMOVED***
        else if (payload.sub.length > 128) ***REMOVED***
            errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has "sub" (subject) claim longer than 128 characters.` +
                verifyJwtTokenDocsMessage;
      ***REMOVED***
        if (errorMessage) ***REMOVED***
            throw new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, errorMessage);
      ***REMOVED***
  ***REMOVED***
    verifySignature(jwtToken, isEmulator) ***REMOVED***
        const verifier = isEmulator ? EMULATOR_VERIFIER : this.signatureVerifier;
        return verifier.verify(jwtToken)
            .catch((error) => ***REMOVED***
            throw this.mapJwtErrorToAuthError(error);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Maps JwtError to FirebaseAuthError
     *
     * @param error - JwtError to be mapped.
     * @returns FirebaseAuthError or Error instance.
     */
    mapJwtErrorToAuthError(error) ***REMOVED***
        const verifyJwtTokenDocsMessage = ` See $***REMOVED***this.tokenInfo.url} ` +
            `for details on how to retrieve $***REMOVED***this.shortNameArticle} $***REMOVED***this.tokenInfo.shortName}.`;
        if (error.code === jwt_1.JwtErrorCode.TOKEN_EXPIRED) ***REMOVED***
            const errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has expired. Get a fresh $***REMOVED***this.tokenInfo.shortName}` +
                ` from your client app and try again (auth/$***REMOVED***this.tokenInfo.expiredErrorCode.code}).` +
                verifyJwtTokenDocsMessage;
            return new error_1.FirebaseAuthError(this.tokenInfo.expiredErrorCode, errorMessage);
      ***REMOVED***
        else if (error.code === jwt_1.JwtErrorCode.INVALID_SIGNATURE) ***REMOVED***
            const errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has invalid signature.` + verifyJwtTokenDocsMessage;
            return new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, errorMessage);
      ***REMOVED***
        else if (error.code === jwt_1.JwtErrorCode.NO_MATCHING_KID) ***REMOVED***
            const errorMessage = `$***REMOVED***this.tokenInfo.jwtName} has "kid" claim which does not ` +
                `correspond to a known public key. Most likely the $***REMOVED***this.tokenInfo.shortName} ` +
                'is expired, so get a fresh token from your client app and try again.';
            return new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, errorMessage);
      ***REMOVED***
        return new error_1.FirebaseAuthError(error_1.AuthClientErrorCode.INVALID_ARGUMENT, error.message);
  ***REMOVED***
}
exports.FirebaseTokenVerifier = FirebaseTokenVerifier;
/**
 * Creates a new FirebaseTokenVerifier to verify Firebase ID tokens.
 *
 * @internal
 * @param app - Firebase app instance.
 * @returns FirebaseTokenVerifier
 */
function createIdTokenVerifier(app) ***REMOVED***
    return new FirebaseTokenVerifier(CLIENT_CERT_URL, 'https://securetoken.google.com/', exports.ID_TOKEN_INFO, app);
}
exports.createIdTokenVerifier = createIdTokenVerifier;
/**
 * Creates a new FirebaseTokenVerifier to verify Firebase Auth Blocking tokens.
 *
 * @internal
 * @param app - Firebase app instance.
 * @returns FirebaseTokenVerifier
 */
function createAuthBlockingTokenVerifier(app) ***REMOVED***
    return new FirebaseTokenVerifier(CLIENT_CERT_URL, 'https://securetoken.google.com/', exports.AUTH_BLOCKING_TOKEN_INFO, app);
}
exports.createAuthBlockingTokenVerifier = createAuthBlockingTokenVerifier;
/**
 * Creates a new FirebaseTokenVerifier to verify Firebase session cookies.
 *
 * @internal
 * @param app - Firebase app instance.
 * @returns FirebaseTokenVerifier
 */
function createSessionCookieVerifier(app) ***REMOVED***
    return new FirebaseTokenVerifier(SESSION_COOKIE_CERT_URL, 'https://session.firebase.google.com/', exports.SESSION_COOKIE_INFO, app);
}
exports.createSessionCookieVerifier = createSessionCookieVerifier;
