// firebase-admin/lib/project-management/project-management-api-request-internal.js
/*! firebase-admin v11.5.0 */
"use strict";
/*!
 * Copyright 2018 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.ProjectManagementRequestHandler = exports.assertServerResponse = void 0;
const api_request_1 = require("../utils/api-request");
const error_1 = require("../utils/error");
const index_1 = require("../utils/index");
const validator = require("../utils/validator");
/** Project management backend host and port. */
const PROJECT_MANAGEMENT_HOST_AND_PORT = 'firebase.googleapis.com:443';
/** Project management backend path. */
const PROJECT_MANAGEMENT_PATH = '/v1/';
/** Project management beta backend path. */
const PROJECT_MANAGEMENT_BETA_PATH = '/v1beta1/';
/** Project management request header. */
const PROJECT_MANAGEMENT_HEADERS = ***REMOVED***
    'X-Client-Version': `Node/Admin/$***REMOVED***(0, index_1.getSdkVersion)()}`,
};
/** Project management request timeout duration in milliseconds. */
const PROJECT_MANAGEMENT_TIMEOUT_MILLIS = 10000;
const LIST_APPS_MAX_PAGE_SIZE = 100;
const CERT_TYPE_API_MAP = ***REMOVED***
    sha1: 'SHA_1',
    sha256: 'SHA_256',
};
function assertServerResponse(condition, responseData, message) ***REMOVED***
    if (!condition) ***REMOVED***
        throw new error_1.FirebaseProjectManagementError('invalid-server-response', `$***REMOVED***message} Response data: $***REMOVED***JSON.stringify(responseData, null, 2)}`);
  ***REMOVED***
}
exports.assertServerResponse = assertServerResponse;
/**
 * Class that provides mechanism to send requests to the Firebase project management backend
 * endpoints.
 *
 * @internal
 */
class ProjectManagementRequestHandler ***REMOVED***
    /**
     * @param app - The app used to fetch access tokens to sign API requests.
     * @constructor
     */
    constructor(app) ***REMOVED***
        this.baseUrl = `https://$***REMOVED***PROJECT_MANAGEMENT_HOST_AND_PORT}$***REMOVED***PROJECT_MANAGEMENT_PATH}`;
        this.baseBetaUrl = `https://$***REMOVED***PROJECT_MANAGEMENT_HOST_AND_PORT}$***REMOVED***PROJECT_MANAGEMENT_BETA_PATH}`;
        this.httpClient = new api_request_1.AuthorizedHttpClient(app);
  ***REMOVED***
    static wrapAndRethrowHttpError(errStatusCode, errText) ***REMOVED***
        let errorCode;
        let errorMessage;
        switch (errStatusCode) ***REMOVED***
            case 400:
                errorCode = 'invalid-argument';
                errorMessage = 'Invalid argument provided.';
                break;
            case 401:
            case 403:
                errorCode = 'authentication-error';
                errorMessage = 'An error occurred when trying to authenticate. Make sure the credential '
                    + 'used to authenticate this SDK has the proper permissions. See '
                    + 'https://firebase.google.com/docs/admin/setup for setup instructions.';
                break;
            case 404:
                errorCode = 'not-found';
                errorMessage = 'The specified entity could not be found.';
                break;
            case 409:
                errorCode = 'already-exists';
                errorMessage = 'The specified entity already exists.';
                break;
            case 500:
                errorCode = 'internal-error';
                errorMessage = 'An internal error has occurred. Please retry the request.';
                break;
            case 503:
                errorCode = 'service-unavailable';
                errorMessage = 'The server could not process the request in time. See the error '
                    + 'documentation for more details.';
                break;
            default:
                errorCode = 'unknown-error';
                errorMessage = 'An unknown server error was returned.';
      ***REMOVED***
        if (!errText) ***REMOVED***
            errText = '<missing>';
      ***REMOVED***
        throw new error_1.FirebaseProjectManagementError(errorCode, `$***REMOVED***errorMessage} Status code: $***REMOVED***errStatusCode}. Raw server response: "$***REMOVED***errText}".`);
  ***REMOVED***
    /**
     * @param parentResourceName - Fully-qualified resource name of the project whose Android
     *     apps you want to list.
     */
    listAndroidApps(parentResourceName) ***REMOVED***
        return this.invokeRequestHandler('GET', `$***REMOVED***parentResourceName}/androidApps?page_size=$***REMOVED***LIST_APPS_MAX_PAGE_SIZE}`, 
        /* requestData */ null, 'v1beta1');
  ***REMOVED***
    /**
     * @param parentResourceName - Fully-qualified resource name of the project whose iOS apps
     *     you want to list.
     */
    listIosApps(parentResourceName) ***REMOVED***
        return this.invokeRequestHandler('GET', `$***REMOVED***parentResourceName}/iosApps?page_size=$***REMOVED***LIST_APPS_MAX_PAGE_SIZE}`, 
        /* requestData */ null, 'v1beta1');
  ***REMOVED***
    /**
     * @param parentResourceName - Fully-qualified resource name of the project whose iOS apps
     *     you want to list.
     */
    listAppMetadata(parentResourceName) ***REMOVED***
        return this.invokeRequestHandler('GET', `$***REMOVED***parentResourceName}:searchApps?page_size=$***REMOVED***LIST_APPS_MAX_PAGE_SIZE}`, 
        /* requestData */ null, 'v1beta1');
  ***REMOVED***
    /**
     * @param parentResourceName - Fully-qualified resource name of the project that you want
     *     to create the Android app within.
     */
    createAndroidApp(parentResourceName, packageName, displayName) ***REMOVED***
        const requestData = ***REMOVED***
            packageName,
      ***REMOVED***;
        if (validator.isNonEmptyString(displayName)) ***REMOVED***
            requestData.displayName = displayName;
      ***REMOVED***
        return this
            .invokeRequestHandler('POST', `$***REMOVED***parentResourceName}/androidApps`, requestData, 'v1beta1')
            .then((responseData) => ***REMOVED***
            assertServerResponse(validator.isNonNullObject(responseData), responseData, 'createAndroidApp\'s responseData must be a non-null object.');
            assertServerResponse(validator.isNonEmptyString(responseData.name), responseData, 'createAndroidApp\'s responseData.name must be a non-empty string.');
            return this.pollRemoteOperationWithExponentialBackoff(responseData.name);
      ***REMOVED***);
  ***REMOVED***
    /**
     * @param parentResourceName - Fully-qualified resource name of the project that you want
     *     to create the iOS app within.
     */
    createIosApp(parentResourceName, bundleId, displayName) ***REMOVED***
        const requestData = ***REMOVED***
            bundleId,
      ***REMOVED***;
        if (validator.isNonEmptyString(displayName)) ***REMOVED***
            requestData.displayName = displayName;
      ***REMOVED***
        return this
            .invokeRequestHandler('POST', `$***REMOVED***parentResourceName}/iosApps`, requestData, 'v1beta1')
            .then((responseData) => ***REMOVED***
            assertServerResponse(validator.isNonNullObject(responseData), responseData, 'createIosApp\'s responseData must be a non-null object.');
            assertServerResponse(validator.isNonEmptyString(responseData.name), responseData, 'createIosApp\'s responseData.name must be a non-empty string.');
            return this.pollRemoteOperationWithExponentialBackoff(responseData.name);
      ***REMOVED***);
  ***REMOVED***
    /**
     * @param resourceName - Fully-qualified resource name of the entity whose display name you
     *     want to set.
     */
    setDisplayName(resourceName, newDisplayName) ***REMOVED***
        const requestData = ***REMOVED***
            displayName: newDisplayName,
      ***REMOVED***;
        return this
            .invokeRequestHandler('PATCH', `$***REMOVED***resourceName}?update_mask=display_name`, requestData, 'v1beta1')
            .then(() => undefined);
  ***REMOVED***
    /**
     * @param parentResourceName - Fully-qualified resource name of the Android app whose SHA
     *     certificates you want to get.
     */
    getAndroidShaCertificates(parentResourceName) ***REMOVED***
        return this.invokeRequestHandler('GET', `$***REMOVED***parentResourceName}/sha`, /* requestData */ null, 'v1beta1');
  ***REMOVED***
    /**
     * @param parentResourceName - Fully-qualified resource name of the Android app that you
     *     want to add the given SHA certificate to.
     */
    addAndroidShaCertificate(parentResourceName, certificate) ***REMOVED***
        const requestData = ***REMOVED***
            shaHash: certificate.shaHash,
            certType: CERT_TYPE_API_MAP[certificate.certType],
      ***REMOVED***;
        return this
            .invokeRequestHandler('POST', `$***REMOVED***parentResourceName}/sha`, requestData, 'v1beta1')
            .then(() => undefined);
  ***REMOVED***
    /**
     * @param parentResourceName - Fully-qualified resource name of the app whose config you
     *     want to get.
     */
    getConfig(parentResourceName) ***REMOVED***
        return this.invokeRequestHandler('GET', `$***REMOVED***parentResourceName}/config`, /* requestData */ null, 'v1beta1');
  ***REMOVED***
    /**
     * @param parentResourceName - Fully-qualified resource name of the entity that you want to
     *     get.
     */
    getResource(parentResourceName) ***REMOVED***
        return this.invokeRequestHandler('GET', parentResourceName, /* requestData */ null, 'v1beta1');
  ***REMOVED***
    /**
     * @param resourceName - Fully-qualified resource name of the entity that you want to
     *     delete.
     */
    deleteResource(resourceName) ***REMOVED***
        return this
            .invokeRequestHandler('DELETE', resourceName, /* requestData */ null, 'v1beta1')
            .then(() => undefined);
  ***REMOVED***
    pollRemoteOperationWithExponentialBackoff(operationResourceName) ***REMOVED***
        const poller = new api_request_1.ExponentialBackoffPoller();
        return poller.poll(() => ***REMOVED***
            return this.invokeRequestHandler('GET', operationResourceName, /* requestData */ null)
                .then((responseData) => ***REMOVED***
                if (responseData.error) ***REMOVED***
                    const errStatusCode = responseData.error.code || 500;
                    const errText = responseData.error.message || JSON.stringify(responseData.error);
                    ProjectManagementRequestHandler.wrapAndRethrowHttpError(errStatusCode, errText);
              ***REMOVED***
                if (!responseData.done) ***REMOVED***
                    // Continue polling.
                    return null;
              ***REMOVED***
                // Polling complete. Resolve with operation response JSON.
                return responseData.response;
          ***REMOVED***);
      ***REMOVED***);
  ***REMOVED***
    /**
     * Invokes the request handler with the provided request data.
     */
    invokeRequestHandler(method, path, requestData, apiVersion = 'v1') ***REMOVED***
        const baseUrlToUse = (apiVersion === 'v1') ? this.baseUrl : this.baseBetaUrl;
        const request = ***REMOVED***
            method,
            url: `$***REMOVED***baseUrlToUse}$***REMOVED***path}`,
            headers: PROJECT_MANAGEMENT_HEADERS,
            data: requestData,
            timeout: PROJECT_MANAGEMENT_TIMEOUT_MILLIS,
      ***REMOVED***;
        return this.httpClient.send(request)
            .then((response) => ***REMOVED***
            // Send non-JSON responses to the catch() below, where they will be treated as errors.
            if (!response.isJson()) ***REMOVED***
                throw new api_request_1.HttpError(response);
          ***REMOVED***
            return response.data;
      ***REMOVED***)
            .catch((err) => ***REMOVED***
            if (err instanceof api_request_1.HttpError) ***REMOVED***
                ProjectManagementRequestHandler.wrapAndRethrowHttpError(err.response.status, err.response.text);
          ***REMOVED***
            throw err;
      ***REMOVED***);
  ***REMOVED***
}
exports.ProjectManagementRequestHandler = ProjectManagementRequestHandler;
