// pg-types/index.js
var textParsers = require('./lib/textParsers');
var binaryParsers = require('./lib/binaryParsers');
var arrayParser = require('./lib/arrayParser');
var builtinTypes = require('./lib/builtins');

exports.getTypeParser = getTypeParser;
exports.setTypeParser = setTypeParser;
exports.arrayParser = arrayParser;
exports.builtins = builtinTypes;

var typeParsers = ***REMOVED***
  text: ***REMOVED***},
  binary: ***REMOVED***}
};

//the empty parse function
function noParse (val) ***REMOVED***
  return String(val);
};

//returns a function used to convert a specific type (specified by
//oid) into a result javascript type
//note: the oid can be obtained via the following sql query:
//SELECT oid FROM pg_type WHERE typname = 'TYPE_NAME_HERE';
function getTypeParser (oid, format) ***REMOVED***
  format = format || 'text';
  if (!typeParsers[format]) ***REMOVED***
    return noParse;
***REMOVED***
  return typeParsers[format][oid] || noParse;
};

function setTypeParser (oid, format, parseFn) ***REMOVED***
  if(typeof format == 'function') ***REMOVED***
    parseFn = format;
    format = 'text';
***REMOVED***
  typeParsers[format][oid] = parseFn;
};

textParsers.init(function(oid, converter) ***REMOVED***
  typeParsers.text[oid] = converter;
});

binaryParsers.init(function(oid, converter) ***REMOVED***
  typeParsers.binary[oid] = converter;
});
