// websocket-driver/lib/websocket/driver/base.js
'use strict';

var Buffer  = require('safe-buffer').Buffer,
    Emitter = require('events').EventEmitter,
    util    = require('util'),
    streams = require('../streams'),
    Headers = require('./headers'),
    Reader  = require('./stream_reader');

var Base = function(request, url, options) ***REMOVED***
  Emitter.call(this);
  Base.validateOptions(options || ***REMOVED***}, ['maxLength', 'masking', 'requireMasking', 'protocols']);

  this._request   = request;
  this._reader    = new Reader();
  this._options   = options || ***REMOVED***};
  this._maxLength = this._options.maxLength || this.MAX_LENGTH;
  this._headers   = new Headers();
  this.__queue    = [];
  this.readyState = 0;
  this.url        = url;

  this.io = new streams.IO(this);
  this.messages = new streams.Messages(this);
  this._bindEventListeners();
};
util.inherits(Base, Emitter);

Base.isWebSocket = function(request) ***REMOVED***
  var connection = request.headers.connection || '',
      upgrade    = request.headers.upgrade || '';

  return request.method === 'GET' &&
         connection.toLowerCase().split(/ *, */).indexOf('upgrade') >= 0 &&
         upgrade.toLowerCase() === 'websocket';
};

Base.validateOptions = function(options, validKeys) ***REMOVED***
  for (var key in options) ***REMOVED***
    if (validKeys.indexOf(key) < 0)
      throw new Error('Unrecognized option: ' + key);
***REMOVED***
};

var instance = ***REMOVED***
  // This is 64MB, small enough for an average VPS to handle without
  // crashing from process out of memory
  MAX_LENGTH: 0x3ffffff,

  STATES: ['connecting', 'open', 'closing', 'closed'],

  _bindEventListeners: function() ***REMOVED***
    var self = this;

    // Protocol errors are informational and do not have to be handled
    this.messages.on('error', function() ***REMOVED***});

    this.on('message', function(event) ***REMOVED***
      var messages = self.messages;
      if (messages.readable) messages.emit('data', event.data);
  ***REMOVED***);

    this.on('error', function(error) ***REMOVED***
      var messages = self.messages;
      if (messages.readable) messages.emit('error', error);
  ***REMOVED***);

    this.on('close', function() ***REMOVED***
      var messages = self.messages;
      if (!messages.readable) return;
      messages.readable = messages.writable = false;
      messages.emit('end');
  ***REMOVED***);
***REMOVED***,

  getState: function() ***REMOVED***
    return this.STATES[this.readyState] || null;
***REMOVED***,

  addExtension: function(extension) ***REMOVED***
    return false;
***REMOVED***,

  setHeader: function(name, value) ***REMOVED***
    if (this.readyState > 0) return false;
    this._headers.set(name, value);
    return true;
***REMOVED***,

  start: function() ***REMOVED***
    if (this.readyState !== 0) return false;

    if (!Base.isWebSocket(this._request))
      return this._failHandshake(new Error('Not a WebSocket request'));

    var response;

    try ***REMOVED***
      response = this._handshakeResponse();
  ***REMOVED*** catch (error) ***REMOVED***
      return this._failHandshake(error);
  ***REMOVED***

    this._write(response);
    if (this._stage !== -1) this._open();
    return true;
***REMOVED***,

  _failHandshake: function(error) ***REMOVED***
    var headers = new Headers();
    headers.set('Content-Type', 'text/plain');
    headers.set('Content-Length', Buffer.byteLength(error.message, 'utf8'));

    headers = ['HTTP/1.1 400 Bad Request', headers.toString(), error.message];
    this._write(Buffer.from(headers.join('\r\n'), 'utf8'));
    this._fail('protocol_error', error.message);

    return false;
***REMOVED***,

  text: function(message) ***REMOVED***
    return this.frame(message);
***REMOVED***,

  binary: function(message) ***REMOVED***
    return false;
***REMOVED***,

  ping: function() ***REMOVED***
    return false;
***REMOVED***,

  pong: function() ***REMOVED***
      return false;
***REMOVED***,

  close: function(reason, code) ***REMOVED***
    if (this.readyState !== 1) return false;
    this.readyState = 3;
    this.emit('close', new Base.CloseEvent(null, null));
    return true;
***REMOVED***,

  _open: function() ***REMOVED***
    this.readyState = 1;
    this.__queue.forEach(function(args) ***REMOVED*** this.frame.apply(this, args) }, this);
    this.__queue = [];
    this.emit('open', new Base.OpenEvent());
***REMOVED***,

  _queue: function(message) ***REMOVED***
    this.__queue.push(message);
    return true;
***REMOVED***,

  _write: function(chunk) ***REMOVED***
    var io = this.io;
    if (io.readable) io.emit('data', chunk);
***REMOVED***,

  _fail: function(type, message) ***REMOVED***
    this.readyState = 2;
    this.emit('error', new Error(message));
    this.close();
***REMOVED***
};

for (var key in instance)
  Base.prototype[key] = instance[key];


Base.ConnectEvent = function() ***REMOVED***};

Base.OpenEvent = function() ***REMOVED***};

Base.CloseEvent = function(code, reason) ***REMOVED***
  this.code   = code;
  this.reason = reason;
};

Base.MessageEvent = function(data) ***REMOVED***
  this.data = data;
};

Base.PingEvent = function(data) ***REMOVED***
  this.data = data;
};

Base.PongEvent = function(data) ***REMOVED***
  this.data = data;
};

module.exports = Base;
