// readable-stream/errors-browser.js
'use strict';

function _inheritsLoose(subClass, superClass) ***REMOVED*** subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

var codes = ***REMOVED***};

function createErrorType(code, message, Base) ***REMOVED***
  if (!Base) ***REMOVED***
    Base = Error;
***REMOVED***

  function getMessage(arg1, arg2, arg3) ***REMOVED***
    if (typeof message === 'string') ***REMOVED***
      return message;
  ***REMOVED*** else ***REMOVED***
      return message(arg1, arg2, arg3);
  ***REMOVED***
***REMOVED***

  var NodeError =
  /*#__PURE__*/
  function (_Base) ***REMOVED***
    _inheritsLoose(NodeError, _Base);

    function NodeError(arg1, arg2, arg3) ***REMOVED***
      return _Base.call(this, getMessage(arg1, arg2, arg3)) || this;
  ***REMOVED***

    return NodeError;
***REMOVED***(Base);

  NodeError.prototype.name = Base.name;
  NodeError.prototype.code = code;
  codes[code] = NodeError;
} // https://github.com/nodejs/node/blob/v10.8.0/lib/internal/errors.js


function oneOf(expected, thing) ***REMOVED***
  if (Array.isArray(expected)) ***REMOVED***
    var len = expected.length;
    expected = expected.map(function (i) ***REMOVED***
      return String(i);
  ***REMOVED***);

    if (len > 2) ***REMOVED***
      return "one of ".concat(thing, " ").concat(expected.slice(0, len - 1).join(', '), ", or ") + expected[len - 1];
  ***REMOVED*** else if (len === 2) ***REMOVED***
      return "one of ".concat(thing, " ").concat(expected[0], " or ").concat(expected[1]);
  ***REMOVED*** else ***REMOVED***
      return "of ".concat(thing, " ").concat(expected[0]);
  ***REMOVED***
***REMOVED*** else ***REMOVED***
    return "of ".concat(thing, " ").concat(String(expected));
***REMOVED***
} // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith


function startsWith(str, search, pos) ***REMOVED***
  return str.substr(!pos || pos < 0 ? 0 : +pos, search.length) === search;
} // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/endsWith


function endsWith(str, search, this_len) ***REMOVED***
  if (this_len === undefined || this_len > str.length) ***REMOVED***
    this_len = str.length;
***REMOVED***

  return str.substring(this_len - search.length, this_len) === search;
} // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes


function includes(str, search, start) ***REMOVED***
  if (typeof start !== 'number') ***REMOVED***
    start = 0;
***REMOVED***

  if (start + search.length > str.length) ***REMOVED***
    return false;
***REMOVED*** else ***REMOVED***
    return str.indexOf(search, start) !== -1;
***REMOVED***
}

createErrorType('ERR_INVALID_OPT_VALUE', function (name, value) ***REMOVED***
  return 'The value "' + value + '" is invalid for option "' + name + '"';
}, TypeError);
createErrorType('ERR_INVALID_ARG_TYPE', function (name, expected, actual) ***REMOVED***
  // determiner: 'must be' or 'must not be'
  var determiner;

  if (typeof expected === 'string' && startsWith(expected, 'not ')) ***REMOVED***
    determiner = 'must not be';
    expected = expected.replace(/^not /, '');
***REMOVED*** else ***REMOVED***
    determiner = 'must be';
***REMOVED***

  var msg;

  if (endsWith(name, ' argument')) ***REMOVED***
    // For cases like 'first argument'
    msg = "The ".concat(name, " ").concat(determiner, " ").concat(oneOf(expected, 'type'));
***REMOVED*** else ***REMOVED***
    var type = includes(name, '.') ? 'property' : 'argument';
    msg = "The \"".concat(name, "\" ").concat(type, " ").concat(determiner, " ").concat(oneOf(expected, 'type'));
***REMOVED***

  msg += ". Received type ".concat(typeof actual);
  return msg;
}, TypeError);
createErrorType('ERR_STREAM_PUSH_AFTER_EOF', 'stream.push() after EOF');
createErrorType('ERR_METHOD_NOT_IMPLEMENTED', function (name) ***REMOVED***
  return 'The ' + name + ' method is not implemented';
});
createErrorType('ERR_STREAM_PREMATURE_CLOSE', 'Premature close');
createErrorType('ERR_STREAM_DESTROYED', function (name) ***REMOVED***
  return 'Cannot call ' + name + ' after a stream was destroyed';
});
createErrorType('ERR_MULTIPLE_CALLBACK', 'Callback called multiple times');
createErrorType('ERR_STREAM_CANNOT_PIPE', 'Cannot pipe, not readable');
createErrorType('ERR_STREAM_WRITE_AFTER_END', 'write after end');
createErrorType('ERR_STREAM_NULL_VALUES', 'May not write null values to stream', TypeError);
createErrorType('ERR_UNKNOWN_ENCODING', function (arg) ***REMOVED***
  return 'Unknown encoding: ' + arg;
}, TypeError);
createErrorType('ERR_STREAM_UNSHIFT_AFTER_END_EVENT', 'stream.unshift() after end event');
module.exports.codes = codes;
