// google-gax/build/src/call.js
"use strict";
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.OngoingCallPromise = exports.OngoingCall = void 0;
const status_1 = require("./status");
const googleError_1 = require("./googleError");
class OngoingCall ***REMOVED***
    /**
     * OngoingCall manages callback, API calls, and cancellation
     * of the API calls.
     * @param ***REMOVED***APICallback=} callback
     *   The callback to be called asynchronously when the API call
     *   finishes.
     * @constructor
     * @property ***REMOVED***APICallback} callback
     *   The callback function to be called.
     * @private
     */
    constructor(callback) ***REMOVED***
        this.callback = callback;
        this.completed = false;
  ***REMOVED***
    /**
     * Cancels the ongoing promise.
     */
    cancel() ***REMOVED***
        if (this.completed) ***REMOVED***
            return;
      ***REMOVED***
        this.completed = true;
        if (this.cancelFunc) ***REMOVED***
            this.cancelFunc();
      ***REMOVED***
        else ***REMOVED***
            const error = new googleError_1.GoogleError('cancelled');
            error.code = status_1.Status.CANCELLED;
            this.callback(error);
      ***REMOVED***
  ***REMOVED***
    /**
     * Call calls the specified function. Result will be used to fulfill
     * the promise.
     *
     * @param ***REMOVED***SimpleCallbackFunction} func
     *   A function for an API call.
     * @param ***REMOVED***Object} argument
     *   A request object.
     */
    call(func, argument) ***REMOVED***
        if (this.completed) ***REMOVED***
            return;
      ***REMOVED***
        const canceller = func(argument, (err, response, next, rawResponse) => ***REMOVED***
            this.completed = true;
            setImmediate(this.callback, err, response, next, rawResponse);
      ***REMOVED***);
        if (canceller instanceof Promise) ***REMOVED***
            canceller.catch(err => ***REMOVED***
                setImmediate(this.callback, new googleError_1.GoogleError(err), null, null, null);
          ***REMOVED***);
      ***REMOVED***
        this.cancelFunc = () => canceller.cancel();
  ***REMOVED***
}
exports.OngoingCall = OngoingCall;
class OngoingCallPromise extends OngoingCall ***REMOVED***
    /**
     * GaxPromise is GRPCCallbackWrapper, but it holds a promise when
     * the API call finishes.
     * @constructor
     * @private
     */
    constructor() ***REMOVED***
        let resolveCallback;
        let rejectCallback;
        const callback = (err, response, next, rawResponse) => ***REMOVED***
            if (err) ***REMOVED***
                // If gRPC metadata exist, parsed google.rpc.status details.
                if (err.metadata) ***REMOVED***
                    rejectCallback(googleError_1.GoogleError.parseGRPCStatusDetails(err));
              ***REMOVED***
                else ***REMOVED***
                    rejectCallback(err);
              ***REMOVED***
          ***REMOVED***
            else if (response !== undefined) ***REMOVED***
                resolveCallback([response, next || null, rawResponse || null]);
          ***REMOVED***
            else ***REMOVED***
                throw new googleError_1.GoogleError('Neither error nor response are defined');
          ***REMOVED***
      ***REMOVED***;
        const promise = new Promise((resolve, reject) => ***REMOVED***
            resolveCallback = resolve;
            rejectCallback = reject;
      ***REMOVED***);
        super(callback);
        this.promise = promise;
        this.promise.cancel = () => ***REMOVED***
            this.cancel();
      ***REMOVED***;
  ***REMOVED***
}
exports.OngoingCallPromise = OngoingCallPromise;
