// google-gax/build/src/gax.js
"use strict";
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.createByteLengthFunction = exports.constructSettings = exports.createBundleOptions = exports.createMaxRetriesBackoffSettings = exports.createDefaultBackoffSettings = exports.createBackoffSettings = exports.createRetryOptions = exports.CallSettings = exports.RetryOptions = void 0;
const warnings_1 = require("./warnings");
const util_1 = require("./util");
/**
 * Encapsulates the overridable settings for a particular API call.
 *
 * ``CallOptions`` is an optional arg for all GAX API calls.  It is used to
 * configure the settings of a specific API call.
 *
 * When provided, its values override the GAX service defaults for that
 * particular call.
 *
 * Typically the API clients will accept this as the second to the last
 * argument. See the examples below.
 * @typedef ***REMOVED***Object} CallOptions
 * @property ***REMOVED***number=} timeout - The client-side timeout for API calls.
 * @property ***REMOVED***RetryOptions=} retry - determines whether and how to retry
 *   on transient errors. When set to null, the call will not retry.
 * @property ***REMOVED***boolean=} autoPaginate - If set to false and the call is
 *   configured for paged iteration, page unrolling is not performed, instead
 *   the callback will be called with the response object.
 * @property ***REMOVED***Object=} pageToken - If set and the call is configured for
 *   paged iteration, paged iteration is not performed and requested with this
 *   pageToken.
 * @property ***REMOVED***number} maxResults - If set and the call is configured for
 *   paged iteration, the call will stop when the number of response elements
 *   reaches to the specified size. By default, it will unroll the page to
 *   the end of the list.
 * @property ***REMOVED***boolean=} isBundling - If set to false and the call is configured
 *   for bundling, bundling is not performed.
 * @property ***REMOVED***BackoffSettings=} longrunning - BackoffSettings used for polling.
 * @example
 * // suppress bundling for bundled method.
 * api.bundlingMethod(
 *     param, ***REMOVED***optParam: aValue, isBundling: false}, function(err, response) ***REMOVED***
 *   // handle response.
 * });
 * @example
 * // suppress streaming for page-streaming method.
 * api.pageStreamingMethod(
 *     param, ***REMOVED***optParam: aValue, autoPaginate: false}, function(err, page) ***REMOVED***
 *   // not returning a stream, but callback is called with the paged response.
 * });
 */
/**
 * Per-call configurable settings for retrying upon transient failure.
 * @typedef ***REMOVED***Object} RetryOptions
 * @property ***REMOVED***String[]} retryCodes
 * @property ***REMOVED***BackoffSettings} backoffSettings
 */
class RetryOptions ***REMOVED***
    constructor(retryCodes, backoffSettings) ***REMOVED***
        this.retryCodes = retryCodes;
        this.backoffSettings = backoffSettings;
  ***REMOVED***
}
exports.RetryOptions = RetryOptions;
class CallSettings ***REMOVED***
    /**
     * @param ***REMOVED***Object} settings - An object containing parameters of this settings.
     * @param ***REMOVED***number} settings.timeout - The client-side timeout for API calls.
     *   This parameter is ignored for retrying calls.
     * @param ***REMOVED***RetryOptions} settings.retry - The configuration for retrying upon
     *   transient error. If set to null, this call will not retry.
     * @param ***REMOVED***boolean} settings.autoPaginate - If there is no `pageDescriptor`,
     *   this attrbute has no meaning. Otherwise, determines whether a page
     * streamed response should make the page structure transparent to the user by
     *   flattening the repeated field in the returned generator.
     * @param ***REMOVED***number} settings.pageToken - If there is no `pageDescriptor`,
     *   this attribute has no meaning. Otherwise, determines the page token used
     * in the page streaming request.
     * @param ***REMOVED***Object} settings.otherArgs - Additional arguments to be passed to
     *   the API calls.
     *
     * @constructor
     */
    constructor(settings) ***REMOVED***
        var _a;
        settings = settings || ***REMOVED***};
        this.timeout = settings.timeout || 30 * 1000;
        this.retry = settings.retry;
        this.autoPaginate =
            'autoPaginate' in settings ? settings.autoPaginate : true;
        this.maxResults = settings.maxResults;
        this.otherArgs = settings.otherArgs || ***REMOVED***};
        this.bundleOptions = settings.bundleOptions;
        this.isBundling = 'isBundling' in settings ? settings.isBundling : true;
        this.longrunning =
            'longrunning' in settings ? settings.longrunning : undefined;
        this.apiName = (_a = settings.apiName) !== null && _a !== void 0 ? _a : undefined;
        this.retryRequestOptions = settings.retryRequestOptions;
  ***REMOVED***
    /**
     * Returns a new CallSettings merged from this and a CallOptions object.
     *
     * @param ***REMOVED***CallOptions} options - an instance whose values override
     *   those in this object. If null, ``merge`` returns a copy of this
     *   object
     * @return ***REMOVED***CallSettings} The merged CallSettings instance.
     */
    merge(options) ***REMOVED***
        if (!options) ***REMOVED***
            return new CallSettings(this);
      ***REMOVED***
        let timeout = this.timeout;
        let retry = this.retry;
        let autoPaginate = this.autoPaginate;
        let maxResults = this.maxResults;
        let otherArgs = this.otherArgs;
        let isBundling = this.isBundling;
        let longrunning = this.longrunning;
        let apiName = this.apiName;
        let retryRequestOptions = this.retryRequestOptions;
        // If a method-specific timeout is set in the service config, and the retry codes for that
        // method are non-null, then that timeout value will be used to
        // override backoff settings.
        if (retry !== undefined &&
            retry !== null &&
            retry.retryCodes !== null &&
            retry.retryCodes.length > 0) ***REMOVED***
            retry.backoffSettings.initialRpcTimeoutMillis = timeout;
            retry.backoffSettings.maxRpcTimeoutMillis = timeout;
            retry.backoffSettings.totalTimeoutMillis = timeout;
      ***REMOVED***
        // If the user provides a timeout to the method, that timeout value will be used
        // to override the backoff settings.
        if ('timeout' in options) ***REMOVED***
            timeout = options.timeout;
            if (retry !== undefined &&
                retry !== null &&
                retry.retryCodes.length > 0) ***REMOVED***
                retry.backoffSettings.initialRpcTimeoutMillis = timeout;
                retry.backoffSettings.maxRpcTimeoutMillis = timeout;
                retry.backoffSettings.totalTimeoutMillis = timeout;
          ***REMOVED***
      ***REMOVED***
        if ('retry' in options) ***REMOVED***
            retry = mergeRetryOptions(retry || ***REMOVED***}, options.retry);
      ***REMOVED***
        if ('autoPaginate' in options && !options.autoPaginate) ***REMOVED***
            autoPaginate = false;
      ***REMOVED***
        if ('maxResults' in options) ***REMOVED***
            maxResults = options.maxResults;
      ***REMOVED***
        if ('otherArgs' in options) ***REMOVED***
            otherArgs = ***REMOVED***};
            for (const key in this.otherArgs) ***REMOVED***
                otherArgs[key] = this.otherArgs[key];
          ***REMOVED***
            for (const optionsKey in options.otherArgs) ***REMOVED***
                otherArgs[optionsKey] = options.otherArgs[optionsKey];
          ***REMOVED***
      ***REMOVED***
        if ('isBundling' in options) ***REMOVED***
            isBundling = options.isBundling;
      ***REMOVED***
        if ('maxRetries' in options) ***REMOVED***
            retry.backoffSettings.maxRetries = options.maxRetries;
            delete retry.backoffSettings.totalTimeoutMillis;
      ***REMOVED***
        if ('longrunning' in options) ***REMOVED***
            longrunning = options.longrunning;
      ***REMOVED***
        if ('apiName' in options) ***REMOVED***
            apiName = options.apiName;
      ***REMOVED***
        if ('retryRequestOptions' in options) ***REMOVED***
            retryRequestOptions = options.retryRequestOptions;
      ***REMOVED***
        return new CallSettings(***REMOVED***
            timeout,
            retry,
            bundleOptions: this.bundleOptions,
            longrunning,
            autoPaginate,
            maxResults,
            otherArgs,
            isBundling,
            apiName,
            retryRequestOptions,
      ***REMOVED***);
  ***REMOVED***
}
exports.CallSettings = CallSettings;
/**
 * Per-call configurable settings for retrying upon transient failure.
 *
 * @param ***REMOVED***number[]} retryCodes - a list of Google API canonical error codes
 *   upon which a retry should be attempted.
 * @param ***REMOVED***BackoffSettings} backoffSettings - configures the retry
 *   exponential backoff algorithm.
 * @return ***REMOVED***RetryOptions} A new RetryOptions object.
 *
 */
function createRetryOptions(retryCodes, backoffSettings) ***REMOVED***
    return ***REMOVED***
        retryCodes,
        backoffSettings,
  ***REMOVED***;
}
exports.createRetryOptions = createRetryOptions;
/**
 * Parameters to the exponential backoff algorithm for retrying.
 *
 * @param ***REMOVED***number} initialRetryDelayMillis - the initial delay time,
 *   in milliseconds, between the completion of the first failed request and the
 *   initiation of the first retrying request.
 * @param ***REMOVED***number} retryDelayMultiplier - the multiplier by which to
 *   increase the delay time between the completion of failed requests, and the
 *   initiation of the subsequent retrying request.
 * @param ***REMOVED***number} maxRetryDelayMillis - the maximum delay time, in
 *   milliseconds, between requests. When this value is reached,
 *   ``retryDelayMultiplier`` will no longer be used to increase delay time.
 * @param ***REMOVED***number} initialRpcTimeoutMillis - the initial timeout parameter
 *   to the request.
 * @param ***REMOVED***number} rpcTimeoutMultiplier - the multiplier by which to
 *   increase the timeout parameter between failed requests.
 * @param ***REMOVED***number} maxRpcTimeoutMillis - the maximum timeout parameter, in
 *   milliseconds, for a request. When this value is reached,
 *   ``rpcTimeoutMultiplier`` will no longer be used to increase the timeout.
 * @param ***REMOVED***number} totalTimeoutMillis - the total time, in milliseconds,
 *   starting from when the initial request is sent, after which an error will
 *   be returned, regardless of the retrying attempts made meanwhile.
 * @return ***REMOVED***BackoffSettings} a new settings.
 *
 */
function createBackoffSettings(initialRetryDelayMillis, retryDelayMultiplier, maxRetryDelayMillis, initialRpcTimeoutMillis, rpcTimeoutMultiplier, maxRpcTimeoutMillis, totalTimeoutMillis) ***REMOVED***
    return ***REMOVED***
        initialRetryDelayMillis,
        retryDelayMultiplier,
        maxRetryDelayMillis,
        initialRpcTimeoutMillis,
        rpcTimeoutMultiplier,
        maxRpcTimeoutMillis,
        totalTimeoutMillis,
  ***REMOVED***;
}
exports.createBackoffSettings = createBackoffSettings;
function createDefaultBackoffSettings() ***REMOVED***
    return createBackoffSettings(100, 1.3, 60000, null, null, null, null);
}
exports.createDefaultBackoffSettings = createDefaultBackoffSettings;
/**
 * Parameters to the exponential backoff algorithm for retrying.
 * This function is unsupported, and intended for internal use only.
 *
 * @param ***REMOVED***number} initialRetryDelayMillis - the initial delay time,
 *   in milliseconds, between the completion of the first failed request and the
 *   initiation of the first retrying request.
 * @param ***REMOVED***number} retryDelayMultiplier - the multiplier by which to
 *   increase the delay time between the completion of failed requests, and the
 *   initiation of the subsequent retrying request.
 * @param ***REMOVED***number} maxRetryDelayMillis - the maximum delay time, in
 *   milliseconds, between requests. When this value is reached,
 *   ``retryDelayMultiplier`` will no longer be used to increase delay time.
 * @param ***REMOVED***number} initialRpcTimeoutMillis - the initial timeout parameter
 *   to the request.
 * @param ***REMOVED***number} rpcTimeoutMultiplier - the multiplier by which to
 *   increase the timeout parameter between failed requests.
 * @param ***REMOVED***number} maxRpcTimeoutMillis - the maximum timeout parameter, in
 *   milliseconds, for a request. When this value is reached,
 *   ``rpcTimeoutMultiplier`` will no longer be used to increase the timeout.
 * @param ***REMOVED***number} maxRetries - the maximum number of retrying attempts that
 *   will be made. If reached, an error will be returned.
 * @return ***REMOVED***BackoffSettings} a new settings.
 *
 */
function createMaxRetriesBackoffSettings(initialRetryDelayMillis, retryDelayMultiplier, maxRetryDelayMillis, initialRpcTimeoutMillis, rpcTimeoutMultiplier, maxRpcTimeoutMillis, maxRetries) ***REMOVED***
    return ***REMOVED***
        initialRetryDelayMillis,
        retryDelayMultiplier,
        maxRetryDelayMillis,
        initialRpcTimeoutMillis,
        rpcTimeoutMultiplier,
        maxRpcTimeoutMillis,
        maxRetries,
  ***REMOVED***;
}
exports.createMaxRetriesBackoffSettings = createMaxRetriesBackoffSettings;
/**
 * Creates a new ***REMOVED***@link BundleOptions}.
 *
 * @private
 * @param ***REMOVED***Object} options - An object to hold optional parameters. See
 *   properties for the content of options.
 * @return ***REMOVED***BundleOptions} - A new options.
 */
function createBundleOptions(options) ***REMOVED***
    const params = [
        'element_count_threshold',
        'element_count_limit',
        'request_byte_threshold',
        'request_byte_limit',
        'delay_threshold_millis',
    ];
    params.forEach(param => ***REMOVED***
        if (param in options && typeof options[param] !== 'number') ***REMOVED***
            throw new Error(`$***REMOVED***param} should be a number`);
      ***REMOVED***
  ***REMOVED***);
    const elementCountThreshold = options.element_count_threshold || 0;
    const elementCountLimit = options.element_count_limit || 0;
    const requestByteThreshold = options.request_byte_threshold || 0;
    const requestByteLimit = options.request_byte_limit || 0;
    const delayThreshold = options.delay_threshold_millis || 0;
    if (elementCountThreshold === 0 &&
        requestByteThreshold === 0 &&
        delayThreshold === 0) ***REMOVED***
        throw new Error('one threshold should be > 0');
  ***REMOVED***
    return ***REMOVED***
        elementCountThreshold,
        elementCountLimit,
        requestByteThreshold,
        requestByteLimit,
        delayThreshold,
  ***REMOVED***;
}
exports.createBundleOptions = createBundleOptions;
/**
 * Helper for ***REMOVED***@link constructSettings}
 *
 * @private
 *
 * @param ***REMOVED***Object} methodConfig - A dictionary representing a single
 *   `methods` entry of the standard API client config file. (See
 *   ***REMOVED***@link constructSettings} for information on this yaml.)
 * @param ***REMOVED***?Object} retryCodes - A dictionary parsed from the
 *   `retry_codes_def` entry of the standard API client config
 *   file. (See ***REMOVED***@link constructSettings} for information on this yaml.)
 * @param ***REMOVED***Object} retryParams - A dictionary parsed from the
 *   `retry_params` entry of the standard API client config
 *   file. (See ***REMOVED***@link constructSettings} for information on this yaml.)
 * @param ***REMOVED***Object} retryNames - A dictionary mapping the string names
 *   used in the standard API client config file to API response
 *   status codes.
 * @return ***REMOVED***?RetryOptions} The new retry options.
 */
function constructRetry(methodConfig, retryCodes, retryParams, retryNames) ***REMOVED***
    if (!methodConfig) ***REMOVED***
        return null;
  ***REMOVED***
    let codes = null;
    if (retryCodes && 'retry_codes_name' in methodConfig) ***REMOVED***
        const retryCodesName = methodConfig['retry_codes_name'];
        codes = (retryCodes[retryCodesName] || []).map(name => ***REMOVED***
            return Number(retryNames[name]);
      ***REMOVED***);
  ***REMOVED***
    let backoffSettings = null;
    if (retryParams && 'retry_params_name' in methodConfig) ***REMOVED***
        const params = retryParams[methodConfig.retry_params_name];
        backoffSettings = createBackoffSettings(params.initial_retry_delay_millis, params.retry_delay_multiplier, params.max_retry_delay_millis, params.initial_rpc_timeout_millis, params.rpc_timeout_multiplier, params.max_rpc_timeout_millis, params.total_timeout_millis);
  ***REMOVED***
    return createRetryOptions(codes, backoffSettings);
}
/**
 * Helper for ***REMOVED***@link constructSettings}
 *
 * Takes two retry options, and merges them into a single RetryOption instance.
 *
 * @private
 *
 * @param ***REMOVED***RetryOptions} retry - The base RetryOptions.
 * @param ***REMOVED***RetryOptions} overrides - The RetryOptions used for overriding
 *   `retry`. Use the values if it is not null. If entire `overrides` is null,
 *   ignore the base retry and return null.
 * @return ***REMOVED***?RetryOptions} The merged RetryOptions.
 */
function mergeRetryOptions(retry, overrides) ***REMOVED***
    if (!overrides) ***REMOVED***
        return null;
  ***REMOVED***
    if (!overrides.retryCodes && !overrides.backoffSettings) ***REMOVED***
        return retry;
  ***REMOVED***
    const codes = overrides.retryCodes ? overrides.retryCodes : retry.retryCodes;
    const backoffSettings = overrides.backoffSettings
        ? overrides.backoffSettings
        : retry.backoffSettings;
    return createRetryOptions(codes, backoffSettings);
}
/**
 * Constructs a dictionary mapping method names to ***REMOVED***@link CallSettings}.
 *
 * The `clientConfig` parameter is parsed from a client configuration JSON
 * file of the form:
 *
 *     ***REMOVED***
 *       "interfaces": ***REMOVED***
 *         "google.fake.v1.ServiceName": ***REMOVED***
 *           "retry_codes": ***REMOVED***
 *             "idempotent": ["UNAVAILABLE", "DEADLINE_EXCEEDED"],
 *             "non_idempotent": []
 *         ***REMOVED***,
 *           "retry_params": ***REMOVED***
 *             "default": ***REMOVED***
 *               "initial_retry_delay_millis": 100,
 *               "retry_delay_multiplier": 1.2,
 *               "max_retry_delay_millis": 1000,
 *               "initial_rpc_timeout_millis": 2000,
 *               "rpc_timeout_multiplier": 1.5,
 *               "max_rpc_timeout_millis": 30000,
 *               "total_timeout_millis": 45000
 *           ***REMOVED***
 *         ***REMOVED***,
 *           "methods": ***REMOVED***
 *             "CreateFoo": ***REMOVED***
 *               "retry_codes_name": "idempotent",
 *               "retry_params_name": "default"
 *           ***REMOVED***,
 *             "Publish": ***REMOVED***
 *               "retry_codes_name": "non_idempotent",
 *               "retry_params_name": "default",
 *               "bundling": ***REMOVED***
 *                 "element_count_threshold": 40,
 *                 "element_count_limit": 200,
 *                 "request_byte_threshold": 90000,
 *                 "request_byte_limit": 100000,
 *                 "delay_threshold_millis": 100
 *             ***REMOVED***
 *           ***REMOVED***
 *         ***REMOVED***
 *       ***REMOVED***
 *     ***REMOVED***
 *   ***REMOVED***
 *
 * @param ***REMOVED***String} serviceName - The fully-qualified name of this
 *   service, used as a key into the client config file (in the
 *   example above, this value should be 'google.fake.v1.ServiceName').
 * @param ***REMOVED***Object} clientConfig - A dictionary parsed from the
 *   standard API client config file.
 * @param ***REMOVED***Object} configOverrides - A dictionary in the same structure of
 *   client_config to override the settings.
 * @param ***REMOVED***Object.<string, string[]>} retryNames - A dictionary mapping the strings
 *   referring to response status codes to objects representing
 *   those codes.
 * @param ***REMOVED***Object} otherArgs - the non-request arguments to be passed to the API
 *   calls.
 * @return ***REMOVED***Object} A mapping from method name to CallSettings, or null if the
 *   service is not found in the config.
 */
function constructSettings(serviceName, clientConfig, configOverrides, retryNames, otherArgs) ***REMOVED***
    otherArgs = otherArgs || ***REMOVED***};
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const defaults = ***REMOVED***};
    const serviceConfig = (clientConfig.interfaces || ***REMOVED***})[serviceName];
    if (!serviceConfig) ***REMOVED***
        return null;
  ***REMOVED***
    // users can override the config from client side, like bundling options.
    // The detailed structure of the clientConfig can be found here: https://github.com/googleapis/gax-nodejs/blob/main/src/gax.ts#L546
    // The way to override bundling options:
    //
    // const customConfig = ***REMOVED***"interfaces": ***REMOVED***"service": ***REMOVED***"methods": ***REMOVED***"methodName": ***REMOVED***"bundling": ***REMOVED***..}}}}}}
    // const client = new Client(***REMOVED*** projectId, customConfig });
    const overrides = (configOverrides.interfaces || ***REMOVED***})[serviceName] || ***REMOVED***};
    const methods = serviceConfig.methods;
    const overridingMethods = overrides.methods || ***REMOVED***};
    for (const methodName in methods) ***REMOVED***
        const methodConfig = methods[methodName];
        const jsName = (0, util_1.toLowerCamelCase)(methodName);
        let retry = constructRetry(methodConfig, serviceConfig.retry_codes, serviceConfig.retry_params, retryNames);
        let bundlingConfig = methodConfig.bundling;
        let timeout = methodConfig.timeout_millis;
        if (methodName in overridingMethods) ***REMOVED***
            const overridingMethod = overridingMethods[methodName];
            if (overridingMethod) ***REMOVED***
                if ('bundling' in overridingMethod) ***REMOVED***
                    bundlingConfig = overridingMethod.bundling;
              ***REMOVED***
                if ('timeout_millis' in overridingMethod) ***REMOVED***
                    timeout = overridingMethod.timeout_millis;
              ***REMOVED***
          ***REMOVED***
            retry = mergeRetryOptions(retry, constructRetry(overridingMethod, overrides.retry_codes, overrides.retry_params, retryNames));
      ***REMOVED***
        const apiName = serviceName;
        defaults[jsName] = new CallSettings(***REMOVED***
            timeout,
            retry,
            bundleOptions: bundlingConfig
                ? createBundleOptions(bundlingConfig)
                : null,
            otherArgs,
            apiName,
      ***REMOVED***);
  ***REMOVED***
    return defaults;
}
exports.constructSettings = constructSettings;
function createByteLengthFunction(message) ***REMOVED***
    return function getByteLength(obj) ***REMOVED***
        try ***REMOVED***
            return message.encode(obj).finish().length;
      ***REMOVED***
        catch (err) ***REMOVED***
            const stringified = JSON.stringify(obj);
            (0, warnings_1.warn)('error_encoding_protobufjs_object', `Cannot encode protobuf.js object: $***REMOVED***stringified}: $***REMOVED***err}`);
            // We failed to encode the object properly, let's just return an upper boundary of its length.
            // It's only needed for calculating the size of the batch, so it's safe if it's bigger than needed.
            return stringified.length;
      ***REMOVED***
  ***REMOVED***;
}
exports.createByteLengthFunction = createByteLengthFunction;
