// google-gax/build/src/googleError.js
"use strict";
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.GoogleErrorDecoder = exports.GoogleError = void 0;
const status_1 = require("./status");
const protobuf = require("protobufjs");
const serializer = require("proto3-json-serializer");
const fallback_1 = require("./fallback");
class GoogleError extends Error ***REMOVED***
    // Parse details field in google.rpc.status wire over gRPC medatadata.
    // Promote google.rpc.ErrorInfo if exist.
    static parseGRPCStatusDetails(err) ***REMOVED***
        const decoder = new GoogleErrorDecoder();
        try ***REMOVED***
            if (err.metadata && err.metadata.get('grpc-status-details-bin')) ***REMOVED***
                const statusDetailsObj = decoder.decodeGRPCStatusDetails(err.metadata.get('grpc-status-details-bin'));
                if (statusDetailsObj &&
                    statusDetailsObj.details &&
                    statusDetailsObj.details.length > 0) ***REMOVED***
                    err.statusDetails = statusDetailsObj.details;
              ***REMOVED***
                if (statusDetailsObj && statusDetailsObj.errorInfo) ***REMOVED***
                    err.reason = statusDetailsObj.errorInfo.reason;
                    err.domain = statusDetailsObj.errorInfo.domain;
                    err.errorInfoMetadata = statusDetailsObj.errorInfo.metadata;
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        catch (decodeErr) ***REMOVED***
            // ignoring the error
      ***REMOVED***
        return err;
  ***REMOVED***
    // Parse http JSON error and promote google.rpc.ErrorInfo if exist.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static parseHttpError(json) ***REMOVED***
        if (Array.isArray(json)) ***REMOVED***
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            json = json.find((obj) => ***REMOVED***
                return 'error' in obj;
          ***REMOVED***);
      ***REMOVED***
        // fallback logic.
        // related issue: https://github.com/googleapis/gax-nodejs/issues/1303
        // google error mapping: https://cloud.google.com/apis/design/errors
        // if input json doesn't have 'error' fields, wrap the whole object with 'error' field
        if (!json['error']) ***REMOVED***
            json['error'] = ***REMOVED***};
            Object.keys(json)
                .filter(key => key !== 'error')
                .forEach(key => ***REMOVED***
                json['error'][key] = json[key];
                delete json[key];
          ***REMOVED***);
      ***REMOVED***
        const decoder = new GoogleErrorDecoder();
        const proto3Error = decoder.decodeHTTPError(json['error']);
        const error = Object.assign(new GoogleError(json['error']['message']), proto3Error);
        // Map Http Status Code to gRPC Status Code
        if (json['error']['code']) ***REMOVED***
            error.code = (0, status_1.rpcCodeFromHttpStatusCode)(json['error']['code']);
      ***REMOVED***
        else ***REMOVED***
            // If error code is absent, proto3 message default value is 0. We should
            // keep error code as undefined.
            delete error.code;
      ***REMOVED***
        // Keep consistency with gRPC statusDetails fields. gRPC details has been occupied before.
        // Rename "details" to "statusDetails".
        if (error.details) ***REMOVED***
            try ***REMOVED***
                const statusDetailsObj = decoder.decodeHttpStatusDetails(error.details);
                if (statusDetailsObj &&
                    statusDetailsObj.details &&
                    statusDetailsObj.details.length > 0) ***REMOVED***
                    error.statusDetails = statusDetailsObj.details;
              ***REMOVED***
                if (statusDetailsObj && statusDetailsObj.errorInfo) ***REMOVED***
                    error.reason = statusDetailsObj.errorInfo.reason;
                    error.domain = statusDetailsObj.errorInfo.domain;
                    // error.metadata has been occupied for gRPC metadata, so we use
                    // errorInfoMetadata to represent ErrorInfo' metadata field. Keep
                    // consistency with gRPC ErrorInfo metadata field name.
                    error.errorInfoMetadata = statusDetailsObj.errorInfo.metadata;
              ***REMOVED***
          ***REMOVED***
            catch (decodeErr) ***REMOVED***
                // ignoring the error
          ***REMOVED***
      ***REMOVED***
        return error;
  ***REMOVED***
}
exports.GoogleError = GoogleError;
class GoogleErrorDecoder ***REMOVED***
    constructor() ***REMOVED***
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const errorProtoJson = require('../../build/protos/status.json');
        this.root = protobuf.Root.fromJSON(errorProtoJson);
        this.anyType = this.root.lookupType('google.protobuf.Any');
        this.statusType = this.root.lookupType('google.rpc.Status');
  ***REMOVED***
    decodeProtobufAny(anyValue) ***REMOVED***
        const match = anyValue.type_url.match(/^type.googleapis.com\/(.*)/);
        if (!match) ***REMOVED***
            throw new Error(`Unknown type encoded in google.protobuf.any: $***REMOVED***anyValue.type_url}`);
      ***REMOVED***
        const typeName = match[1];
        const type = this.root.lookupType(typeName);
        if (!type) ***REMOVED***
            throw new Error(`Cannot lookup type $***REMOVED***typeName}`);
      ***REMOVED***
        return type.decode(anyValue.value);
  ***REMOVED***
    // Decodes gRPC-fallback error which is an instance of google.rpc.Status.
    decodeRpcStatus(buffer) ***REMOVED***
        const uint8array = new Uint8Array(buffer);
        const status = this.statusType.decode(uint8array);
        // google.rpc.Status contains an array of google.protobuf.Any
        // which need a special treatment
        const details = [];
        let errorInfo;
        for (const detail of status.details) ***REMOVED***
            try ***REMOVED***
                const decodedDetail = this.decodeProtobufAny(detail);
                details.push(decodedDetail);
                if (detail.type_url === 'type.googleapis.com/google.rpc.ErrorInfo') ***REMOVED***
                    errorInfo = decodedDetail;
              ***REMOVED***
          ***REMOVED***
            catch (err) ***REMOVED***
                // cannot decode detail, likely because of the unknown type - just skip it
          ***REMOVED***
      ***REMOVED***
        const result = ***REMOVED***
            code: status.code,
            message: status.message,
            statusDetails: details,
            reason: errorInfo === null || errorInfo === void 0 ? void 0 : errorInfo.reason,
            domain: errorInfo === null || errorInfo === void 0 ? void 0 : errorInfo.domain,
            errorInfoMetadata: errorInfo === null || errorInfo === void 0 ? void 0 : errorInfo.metadata,
      ***REMOVED***;
        return result;
  ***REMOVED***
    // Construct an Error from a StatusObject.
    // Adapted from https://github.com/grpc/grpc-node/blob/main/packages/grpc-js/src/call.ts#L79
    callErrorFromStatus(status) ***REMOVED***
        status.message = `$***REMOVED***status.code} $***REMOVED***status_1.Status[status.code]}: $***REMOVED***status.message}`;
        return Object.assign(new GoogleError(status.message), status);
  ***REMOVED***
    // Decodes gRPC-fallback error which is an instance of google.rpc.Status,
    // and puts it into the object similar to gRPC ServiceError object.
    decodeErrorFromBuffer(buffer) ***REMOVED***
        return this.callErrorFromStatus(this.decodeRpcStatus(buffer));
  ***REMOVED***
    // Decodes gRPC metadata error details which is an instance of google.rpc.Status.
    decodeGRPCStatusDetails(bufferArr) ***REMOVED***
        const details = [];
        let errorInfo;
        bufferArr.forEach(buffer => ***REMOVED***
            const uint8array = new Uint8Array(buffer);
            const rpcStatus = this.statusType.decode(uint8array);
            for (const detail of rpcStatus.details) ***REMOVED***
                try ***REMOVED***
                    const decodedDetail = this.decodeProtobufAny(detail);
                    details.push(decodedDetail);
                    if (detail.type_url === 'type.googleapis.com/google.rpc.ErrorInfo') ***REMOVED***
                        errorInfo = decodedDetail;
                  ***REMOVED***
              ***REMOVED***
                catch (err) ***REMOVED***
                    // cannot decode detail, likely because of the unknown type - just skip it
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***);
        const result = ***REMOVED***
            details,
            errorInfo,
      ***REMOVED***;
        return result;
  ***REMOVED***
    // Decodes http error which is an instance of google.rpc.Status.
    decodeHTTPError(json) ***REMOVED***
        const errorMessage = serializer.fromProto3JSON(this.statusType, json);
        if (!errorMessage) ***REMOVED***
            throw new Error(`Received error message $***REMOVED***json}, but failed to serialize as proto3 message`);
      ***REMOVED***
        return this.statusType.toObject(errorMessage, fallback_1.defaultToObjectOptions);
  ***REMOVED***
    // Decodes http error details which is an instance of Array<google.protobuf.Any>.
    decodeHttpStatusDetails(rawDetails) ***REMOVED***
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const details = [];
        let errorInfo;
        for (const detail of rawDetails) ***REMOVED***
            try ***REMOVED***
                const decodedDetail = this.decodeProtobufAny(detail);
                details.push(decodedDetail);
                if (detail.type_url === 'type.googleapis.com/google.rpc.ErrorInfo') ***REMOVED***
                    errorInfo = decodedDetail;
              ***REMOVED***
          ***REMOVED***
            catch (err) ***REMOVED***
                // cannot decode detail, likely because of the unknown type - just skip it
          ***REMOVED***
      ***REMOVED***
        return ***REMOVED*** details, errorInfo };
  ***REMOVED***
}
exports.GoogleErrorDecoder = GoogleErrorDecoder;
