// google-gax/build/src/index.js
"use strict";
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.serializer = exports.warn = exports.ChannelCredentials = exports.fallback = exports.protobufMinimal = exports.protobuf = exports.version = exports.createByteLengthFunction = exports.LocationsClient = exports.IamClient = exports.OperationsClient = exports.LocationProtos = exports.IamProtos = exports.operationsProtos = exports.lro = exports.routingHeader = exports.StreamType = exports.Status = exports.PathTemplate = exports.operation = exports.Operation = exports.GrpcClient = exports.GoogleProtoFilesRoot = exports.ClientStub = exports.GoogleError = exports.createMaxRetriesBackoffSettings = exports.createDefaultBackoffSettings = exports.createBackoffSettings = exports.createBundleOptions = exports.createRetryOptions = exports.RetryOptions = exports.constructSettings = exports.CallSettings = exports.StreamDescriptor = exports.PageDescriptor = exports.LongrunningDescriptor = exports.BundleDescriptor = exports.createApiCall = exports.OngoingCall = exports.grpc = exports.GoogleAuth = void 0;
const grpc = require("@grpc/grpc-js");
exports.grpc = grpc;
const grpc_1 = require("./grpc");
const IamProtos = require("../protos/iam_service");
exports.IamProtos = IamProtos;
const LocationProtos = require("../protos/locations");
exports.LocationProtos = LocationProtos;
const operationsProtos = require("../protos/operations");
exports.operationsProtos = operationsProtos;
const operationsClient = require("./operationsClient");
const routingHeader = require("./routingHeader");
exports.routingHeader = routingHeader;
var google_auth_library_1 = require("google-auth-library");
Object.defineProperty(exports, "GoogleAuth", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return google_auth_library_1.GoogleAuth; } });
var call_1 = require("./call");
Object.defineProperty(exports, "OngoingCall", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return call_1.OngoingCall; } });
var createApiCall_1 = require("./createApiCall");
Object.defineProperty(exports, "createApiCall", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return createApiCall_1.createApiCall; } });
var descriptor_1 = require("./descriptor");
Object.defineProperty(exports, "BundleDescriptor", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return descriptor_1.BundleDescriptor; } });
Object.defineProperty(exports, "LongrunningDescriptor", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return descriptor_1.LongrunningDescriptor; } });
Object.defineProperty(exports, "PageDescriptor", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return descriptor_1.PageDescriptor; } });
Object.defineProperty(exports, "StreamDescriptor", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return descriptor_1.StreamDescriptor; } });
var gax_1 = require("./gax");
Object.defineProperty(exports, "CallSettings", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return gax_1.CallSettings; } });
Object.defineProperty(exports, "constructSettings", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return gax_1.constructSettings; } });
Object.defineProperty(exports, "RetryOptions", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return gax_1.RetryOptions; } });
Object.defineProperty(exports, "createRetryOptions", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return gax_1.createRetryOptions; } });
Object.defineProperty(exports, "createBundleOptions", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return gax_1.createBundleOptions; } });
Object.defineProperty(exports, "createBackoffSettings", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return gax_1.createBackoffSettings; } });
Object.defineProperty(exports, "createDefaultBackoffSettings", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return gax_1.createDefaultBackoffSettings; } });
Object.defineProperty(exports, "createMaxRetriesBackoffSettings", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return gax_1.createMaxRetriesBackoffSettings; } });
var googleError_1 = require("./googleError");
Object.defineProperty(exports, "GoogleError", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return googleError_1.GoogleError; } });
var grpc_2 = require("./grpc");
Object.defineProperty(exports, "ClientStub", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return grpc_2.ClientStub; } });
Object.defineProperty(exports, "GoogleProtoFilesRoot", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return grpc_2.GoogleProtoFilesRoot; } });
Object.defineProperty(exports, "GrpcClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return grpc_2.GrpcClient; } });
var longrunning_1 = require("./longRunningCalls/longrunning");
Object.defineProperty(exports, "Operation", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return longrunning_1.Operation; } });
Object.defineProperty(exports, "operation", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return longrunning_1.operation; } });
var pathTemplate_1 = require("./pathTemplate");
Object.defineProperty(exports, "PathTemplate", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return pathTemplate_1.PathTemplate; } });
var status_1 = require("./status");
Object.defineProperty(exports, "Status", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return status_1.Status; } });
var streaming_1 = require("./streamingCalls/streaming");
Object.defineProperty(exports, "StreamType", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return streaming_1.StreamType; } });
function lro(options) ***REMOVED***
    options = Object.assign(***REMOVED*** scopes: lro.ALL_SCOPES }, options);
    const gaxGrpc = new grpc_1.GrpcClient(options);
    return new operationsClient.OperationsClientBuilder(gaxGrpc);
}
exports.lro = lro;
lro.SERVICE_ADDRESS = operationsClient.SERVICE_ADDRESS;
lro.ALL_SCOPES = operationsClient.ALL_SCOPES;
var operationsClient_1 = require("./operationsClient");
Object.defineProperty(exports, "OperationsClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return operationsClient_1.OperationsClient; } });
var iamService_1 = require("./iamService");
Object.defineProperty(exports, "IamClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return iamService_1.IamClient; } });
var locationService_1 = require("./locationService");
Object.defineProperty(exports, "LocationsClient", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return locationService_1.LocationsClient; } });
exports.createByteLengthFunction = grpc_1.GrpcClient.createByteLengthFunction;
exports.version = require('../../package.json').version;
const protobuf = require("protobufjs");
exports.protobuf = protobuf;
exports.protobufMinimal = require("protobufjs/minimal");
const fallback = require("./fallback");
exports.fallback = fallback;
var grpc_js_1 = require("@grpc/grpc-js");
Object.defineProperty(exports, "ChannelCredentials", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return grpc_js_1.ChannelCredentials; } });
var warnings_1 = require("./warnings");
Object.defineProperty(exports, "warn", ***REMOVED*** enumerable: true, get: function () ***REMOVED*** return warnings_1.warn; } });
const serializer = require("proto3-json-serializer");
exports.serializer = serializer;
