// google-gax/build/src/locationService.js
"use strict";
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.LocationsClient = void 0;
/* global window */
const gax = require("./gax");
const warnings_1 = require("./warnings");
const createApiCall_1 = require("./createApiCall");
const routingHeader = require("./routingHeader");
const pageDescriptor_1 = require("./paginationCalls/pageDescriptor");
const jsonProtos = require("../protos/locations.json");
/**
 * This file defines retry strategy and timeouts for all API methods in this library.
 */
const gapicConfig = require("./locations_client_config.json");
const version = require('../../package.json').version;
/**
 *  Google Cloud Locations Client.
 *  This is manually written for providing methods [listLocations, getLocations] to the generated client.
 */
class LocationsClient ***REMOVED***
    /**
     * Construct an instance of LocationsClient.
     *
     * @param ***REMOVED***object} [options] - The configuration object.
     * The options accepted by the constructor are described in detail
     * in [this document](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#creating-the-client-instance).
     * The common options are:
     * @param ***REMOVED***object} [options.credentials] - Credentials object.
     * @param ***REMOVED***string} [options.credentials.client_email]
     * @param ***REMOVED***string} [options.credentials.private_key]
     * @param ***REMOVED***string} [options.email] - Account email address. Required when
     *     using a .pem or .p12 keyFilename.
     * @param ***REMOVED***string} [options.keyFilename] - Full path to the a .json, .pem, or
     *     .p12 key downloaded from the Google Developers Console. If you provide
     *     a path to a JSON file, the projectId option below is not necessary.
     *     NOTE: .pem and .p12 require you to specify options.email as well.
     * @param ***REMOVED***number} [options.port] - The port on which to connect to
     *     the remote host.
     * @param ***REMOVED***string} [options.projectId] - The project ID from the Google
     *     Developer's Console, e.g. 'grape-spaceship-123'. We will also check
     *     the environment variable GCLOUD_PROJECT for your project ID. If your
     *     app is running in an environment which supports
     *     ***REMOVED***@link https://developers.google.com/identity/protocols/application-default-credentials Application Default Credentials},
     *     your project ID will be detected automatically.
     * @param ***REMOVED***string} [options.apiEndpoint] - The domain name of the
     *     API remote host.
     * @param ***REMOVED***gax.ClientConfig} [options.clientConfig] - Client configuration override.
     *     Follows the structure of ***REMOVED***@link gapicConfig}.
     * @param ***REMOVED***boolean} [options.fallback] - Use HTTP fallback mode.
     *     In fallback mode, a special browser-compatible transport implementation is used
     *     instead of gRPC transport. In browser context (if the `window` object is defined)
     *     the fallback mode is enabled automatically; set `options.fallback` to `false`
     *     if you need to override this behavior.
     */
    constructor(gaxGrpc, 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    opts) ***REMOVED***
        var _a, _b;
        this._terminated = false;
        this.descriptors = ***REMOVED***
            page: ***REMOVED***},
            stream: ***REMOVED***},
            longrunning: ***REMOVED***},
            batching: ***REMOVED***},
      ***REMOVED***;
        // Ensure that options include all the required fields.
        this.gaxGrpc = gaxGrpc;
        const staticMembers = this.constructor;
        const servicePath = (opts === null || opts === void 0 ? void 0 : opts.servicePath) || (opts === null || opts === void 0 ? void 0 : opts.apiEndpoint) || staticMembers.servicePath;
        this._providedCustomServicePath = !!((opts === null || opts === void 0 ? void 0 : opts.servicePath) || (opts === null || opts === void 0 ? void 0 : opts.apiEndpoint));
        const port = (opts === null || opts === void 0 ? void 0 : opts.port) || staticMembers.port;
        const clientConfig = (_a = opts === null || opts === void 0 ? void 0 : opts.clientConfig) !== null && _a !== void 0 ? _a : ***REMOVED***};
        const fallback = (_b = opts === null || opts === void 0 ? void 0 : opts.fallback) !== null && _b !== void 0 ? _b : (typeof window !== 'undefined' && typeof (window === null || window === void 0 ? void 0 : window.fetch) === 'function');
        opts = Object.assign(***REMOVED*** servicePath, port, clientConfig, fallback }, opts);
        // If scopes are unset in options and we're connecting to a non-default endpoint, set scopes just in case.
        if (servicePath !== staticMembers.servicePath && !('scopes' in opts)) ***REMOVED***
            opts['scopes'] = staticMembers.scopes;
      ***REMOVED***
        // Save options to use in initialize() method.
        this._opts = opts;
        // Save the auth object to the client, for use by other methods.
        this.auth = gaxGrpc.auth;
        // Set the default scopes in auth client if needed.
        if (servicePath === staticMembers.servicePath) ***REMOVED***
            this.auth.defaultScopes = staticMembers.scopes;
      ***REMOVED***
        // Determine the client header string.
        const clientHeader = [`gax/$***REMOVED***version}`, `gapic/$***REMOVED***version}`];
        if (typeof process !== 'undefined' && 'versions' in process) ***REMOVED***
            clientHeader.push(`gl-node/$***REMOVED***process.versions.node}`);
      ***REMOVED***
        else ***REMOVED***
            clientHeader.push(`gl-web/$***REMOVED***version}`);
      ***REMOVED***
        if (!opts.fallback) ***REMOVED***
            clientHeader.push(`grpc/$***REMOVED***gaxGrpc.grpcVersion}`);
      ***REMOVED***
        else if (opts.fallback === 'rest') ***REMOVED***
            clientHeader.push(`rest/$***REMOVED***gaxGrpc.grpcVersion}`);
      ***REMOVED***
        if (opts.libName && opts.libVersion) ***REMOVED***
            clientHeader.push(`$***REMOVED***opts.libName}/$***REMOVED***opts.libVersion}`);
      ***REMOVED***
        // Load the applicable protos.
        this._protos = gaxGrpc.loadProtoJSON(jsonProtos);
        // Some of the methods on this service return "paged" results,
        // (e.g. 50 results at a time, with tokens to get subsequent
        // pages). Denote the keys used for pagination and results.
        this.descriptors.page = ***REMOVED***
            listLocations: new pageDescriptor_1.PageDescriptor('pageToken', 'nextPageToken', 'locations'),
      ***REMOVED***;
        // Put together the default options sent with requests.
        this._defaults = gaxGrpc.constructSettings('google.cloud.location.Locations', gapicConfig, opts.clientConfig || ***REMOVED***}, ***REMOVED*** 'x-goog-api-client': clientHeader.join(' ') });
        // Set up a dictionary of "inner API calls"; the core implementation
        // of calling the API is handled in `google-gax`, with this code
        // merely providing the destination and request information.
        this.innerApiCalls = ***REMOVED***};
        // Add a warn function to the client constructor so it can be easily tested.
        this.warn = warnings_1.warn;
  ***REMOVED***
    /**
     * Initialize the client.
     * Performs asynchronous operations (such as authentication) and prepares the client.
     * This function will be called automatically when any class method is called for the
     * first time, but if you need to initialize it before calling an actual method,
     * feel free to call initialize() directly.
     *
     * You can await on this method if you want to make sure the client is initialized.
     *
     * @returns ***REMOVED***Promise} A promise that resolves to an authenticated service stub.
     */
    initialize() ***REMOVED***
        // If the client stub promise is already initialized, return immediately.
        if (this.locationsStub) ***REMOVED***
            return this.locationsStub;
      ***REMOVED***
        // Put together the "service stub" for
        // google.cloud.location.Locations.
        this.locationsStub = this.gaxGrpc.createStub(this._opts.fallback
            ? this._protos.lookupService('google.cloud.location.Locations')
            : // eslint-disable-next-line @typescript-eslint/no-explicit-any
                this._protos.google.cloud.location.Locations, this._opts, this._providedCustomServicePath);
        // Iterate over each of the methods that the service provides
        // and create an API call method for each.
        const locationsStubMethods = ['listLocations', 'getLocation'];
        for (const methodName of locationsStubMethods) ***REMOVED***
            const callPromise = this.locationsStub.then(stub => (...args) => ***REMOVED***
                if (this._terminated) ***REMOVED***
                    return Promise.reject('The client has already been closed.');
              ***REMOVED***
                const func = stub[methodName];
                return func.apply(stub, args);
          ***REMOVED***, (err) => () => ***REMOVED***
                throw err;
          ***REMOVED***);
            const descriptor = this.descriptors.page[methodName] || undefined;
            const apiCall = (0, createApiCall_1.createApiCall)(callPromise, this._defaults[methodName], descriptor);
            this.innerApiCalls[methodName] = apiCall;
      ***REMOVED***
        return this.locationsStub;
  ***REMOVED***
    /**
     * The DNS address for this API service.
     * @returns ***REMOVED***string} The DNS address for this service.
     */
    static get servicePath() ***REMOVED***
        return 'cloud.googleapis.com';
  ***REMOVED***
    /**
     * The DNS address for this API service - same as servicePath(),
     * exists for compatibility reasons.
     * @returns ***REMOVED***string} The DNS address for this service.
     */
    static get apiEndpoint() ***REMOVED***
        return 'cloud.googleapis.com';
  ***REMOVED***
    /**
     * The port for this API service.
     * @returns ***REMOVED***number} The default port for this service.
     */
    static get port() ***REMOVED***
        return 443;
  ***REMOVED***
    /**
     * The scopes needed to make gRPC calls for every method defined
     * in this service.
     * @returns ***REMOVED***string[]} List of default scopes.
     */
    static get scopes() ***REMOVED***
        return ['https://www.googleapis.com/auth/cloud-platform'];
  ***REMOVED***
    getProjectId(callback) ***REMOVED***
        if (callback) ***REMOVED***
            this.auth.getProjectId(callback);
            return;
      ***REMOVED***
        return this.auth.getProjectId();
  ***REMOVED***
    /**
     * Gets information about a location.
     *
     * @param ***REMOVED***Object} request
     *   The request object that will be sent.
     * @param ***REMOVED***string} request.name
     *   Resource name for the location.
     * @param ***REMOVED***object} [options]
     *   Call options. See ***REMOVED***@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns ***REMOVED***Promise} - The promise which resolves to an array.
     *   The first element of the array is an object representing [Location]***REMOVED***@link google.cloud.location.Location}.
     *   Please see the
     *   [documentation](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#regular-methods)
     *   for more details and examples.
     * @example
     * const [response] = await client.getLocation(request);
     */
    getLocation(request, optionsOrCallback, callback) ***REMOVED***
        request = request || ***REMOVED***};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) ***REMOVED***
            callback = optionsOrCallback;
            options = ***REMOVED***};
      ***REMOVED***
        else ***REMOVED***
            options = optionsOrCallback;
      ***REMOVED***
        options = options || ***REMOVED***};
        options.otherArgs = options.otherArgs || ***REMOVED***};
        options.otherArgs.headers = options.otherArgs.headers || ***REMOVED***};
        options.otherArgs.headers['x-goog-request-params'] =
            routingHeader.fromParams(***REMOVED***
                name: request.name || '',
          ***REMOVED***);
        this.initialize();
        return this.innerApiCalls.getLocation(request, options, callback);
  ***REMOVED***
    /**
     * Lists information about the supported locations for this service.
     *
     * @param ***REMOVED***Object} request
     *   The request object that will be sent.
     * @param ***REMOVED***string} request.name
     *   The resource that owns the locations collection, if applicable.
     * @param ***REMOVED***string} request.filter
     *   The standard list filter.
     * @param ***REMOVED***number} request.pageSize
     *   The standard list page size.
     * @param ***REMOVED***string} request.pageToken
     *   The standard list page token.
     * @param ***REMOVED***object} [options]
     *   Call options. See ***REMOVED***@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns ***REMOVED***Promise} - The promise which resolves to an array.
     *   The first element of the array is Array of [Location]***REMOVED***@link google.cloud.location.Location}.
     *   The client library will perform auto-pagination by default: it will call the API as many
     *   times as needed and will merge results from all the pages into this array.
     *   Note that it can affect your quota.
     *   We recommend using `listLocationsAsync()`
     *   method described below for async iteration which you can stop as needed.
     *   Please see the
     *   [documentation](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#auto-pagination)
     *   for more details and examples.
     */
    listLocations(request, optionsOrCallback, callback) ***REMOVED***
        request = request || ***REMOVED***};
        let options;
        if (typeof optionsOrCallback === 'function' && callback === undefined) ***REMOVED***
            callback = optionsOrCallback;
            options = ***REMOVED***};
      ***REMOVED***
        else ***REMOVED***
            options = optionsOrCallback;
      ***REMOVED***
        options = options || ***REMOVED***};
        options.otherArgs = options.otherArgs || ***REMOVED***};
        options.otherArgs.headers = options.otherArgs.headers || ***REMOVED***};
        options.otherArgs.headers['x-goog-request-params'] =
            routingHeader.fromParams(***REMOVED***
                name: request.name || '',
          ***REMOVED***);
        this.initialize();
        return this.innerApiCalls.listLocations(request, options, callback);
  ***REMOVED***
    /**
     * Equivalent to `listLocations`, but returns an iterable object.
     *
     * `for`-`await`-`of` syntax is used with the iterable to get response elements on-demand.
     * @param ***REMOVED***Object} request
     *   The request object that will be sent.
     * @param ***REMOVED***string} request.name
     *   The resource that owns the locations collection, if applicable.
     * @param ***REMOVED***string} request.filter
     *   The standard list filter.
     * @param ***REMOVED***number} request.pageSize
     *   The standard list page size.
     * @param ***REMOVED***string} request.pageToken
     *   The standard list page token.
     * @param ***REMOVED***object} [options]
     *   Call options. See ***REMOVED***@link https://googleapis.dev/nodejs/google-gax/latest/interfaces/CallOptions.html|CallOptions} for more details.
     * @returns ***REMOVED***Object}
     *   An iterable Object that allows [async iteration](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols).
     *   When you iterate the returned iterable, each element will be an object representing
     *   [Location]***REMOVED***@link google.cloud.location.Location}. The API will be called under the hood as needed, once per the page,
     *   so you can stop the iteration when you don't need more results.
     *   Please see the
     *   [documentation](https://github.com/googleapis/gax-nodejs/blob/main/client-libraries.md#auto-pagination)
     *   for more details and examples.
     * @example
     * const iterable = client.listLocationsAsync(request);
     * for await (const response of iterable) ***REMOVED***
     *   // process response
     * }
     */
    listLocationsAsync(request, options) ***REMOVED***
        request = request || ***REMOVED***};
        options = options || ***REMOVED***};
        options.otherArgs = options.otherArgs || ***REMOVED***};
        options.otherArgs.headers = options.otherArgs.headers || ***REMOVED***};
        options.otherArgs.headers['x-goog-request-params'] =
            routingHeader.fromParams(***REMOVED***
                name: request.name || '',
          ***REMOVED***);
        options = options || ***REMOVED***};
        const callSettings = new gax.CallSettings(options);
        this.initialize();
        return this.descriptors.page.listLocations.asyncIterate(this.innerApiCalls['listLocations'], request, callSettings);
  ***REMOVED***
    /**
     * Terminate the gRPC channel and close the client.
     *
     * The client will no longer be usable and all future behavior is undefined.
     * @returns ***REMOVED***Promise} A promise that resolves when the client is closed.
     */
    close() ***REMOVED***
        this.initialize();
        if (!this._terminated) ***REMOVED***
            return this.locationsStub.then(stub => ***REMOVED***
                this._terminated = true;
                stub.close();
          ***REMOVED***);
      ***REMOVED***
        return Promise.resolve();
  ***REMOVED***
}
exports.LocationsClient = LocationsClient;
