// google-gax/build/src/operationsClient.js
"use strict";
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true });
exports.OperationsClientBuilder = exports.OperationsClient = exports.ALL_SCOPES = exports.SERVICE_ADDRESS = void 0;
const createApiCall_1 = require("./createApiCall");
const descriptor_1 = require("./descriptor");
const gax = require("./gax");
const configData = require("./operations_client_config.json");
const operationProtoJson = require("../protos/operations.json");
const transcoding_1 = require("./transcoding");
exports.SERVICE_ADDRESS = 'longrunning.googleapis.com';
const version = require('../../package.json').version;
const DEFAULT_SERVICE_PORT = 443;
const CODE_GEN_NAME_VERSION = 'gapic/0.7.1';
/**
 * The scopes needed to make gRPC calls to all of the methods defined in
 * this service.
 */
exports.ALL_SCOPES = [];
/**
 * Manages long-running operations with an API service.
 *
 * When an API method normally takes long time to complete, it can be designed
 * to return ***REMOVED***@link Operation} to the client, and the client can use this
 * interface to receive the real response asynchronously by polling the
 * operation resource, or pass the operation resource to another API (such as
 * Google Cloud Pub/Sub API) to receive the response.  Any API service that
 * returns long-running operations should implement the `Operations` interface
 * so developers can have a consistent client experience.
 *
 * This will be created through a builder function which can be obtained by the
 * module. See the following example of how to initialize the module and how to
 * access to the builder.
 * @see ***REMOVED***@link operationsClient}
 *
 * @class
 */
class OperationsClient ***REMOVED***
    constructor(gaxGrpc, 
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    operationsProtos, options) ***REMOVED***
        const opts = Object.assign(***REMOVED***
            servicePath: exports.SERVICE_ADDRESS,
            port: DEFAULT_SERVICE_PORT,
            clientConfig: ***REMOVED***},
      ***REMOVED***, options);
        const googleApiClient = ['gl-node/' + process.versions.node];
        if (opts.libName && opts.libVersion) ***REMOVED***
            googleApiClient.push(opts.libName + '/' + opts.libVersion);
      ***REMOVED***
        googleApiClient.push(CODE_GEN_NAME_VERSION, 'gax/' + version);
        if (opts.fallback) ***REMOVED***
            googleApiClient.push('gl-web/' + version);
      ***REMOVED***
        else ***REMOVED***
            googleApiClient.push('grpc/' + gaxGrpc.grpcVersion);
      ***REMOVED***
        const defaults = gaxGrpc.constructSettings('google.longrunning.Operations', configData, opts.clientConfig || ***REMOVED***}, ***REMOVED*** 'x-goog-api-client': googleApiClient.join(' ') });
        this.auth = gaxGrpc.auth;
        // Set up a dictionary of "inner API calls"; the core implementation
        // of calling the API is handled in `google-gax`, with this code
        // merely providing the destination and request information.
        this.innerApiCalls = ***REMOVED***};
        this.descriptor = ***REMOVED***
            listOperations: new descriptor_1.PageDescriptor('pageToken', 'nextPageToken', 'operations'),
      ***REMOVED***;
        // Put together the "service stub" for
        // google.longrunning.Operations.
        this.operationsStub = gaxGrpc.createStub(opts.fallback
            ? operationsProtos.lookupService('google.longrunning.Operations')
            : operationsProtos.google.longrunning.Operations, opts);
        const operationsStubMethods = [
            'getOperation',
            'listOperations',
            'cancelOperation',
            'deleteOperation',
        ];
        for (const methodName of operationsStubMethods) ***REMOVED***
            const innerCallPromise = this.operationsStub.then(stub => (...args) => ***REMOVED***
                const func = stub[methodName];
                return func.apply(stub, args);
          ***REMOVED***, err => () => ***REMOVED***
                throw err;
          ***REMOVED***);
            this.innerApiCalls[methodName] = (0, createApiCall_1.createApiCall)(innerCallPromise, defaults[methodName], this.descriptor[methodName]);
      ***REMOVED***
  ***REMOVED***
    /** Closes this operations client. */
    close() ***REMOVED***
        this.operationsStub.then(stub => stub.close());
  ***REMOVED***
    getProjectId(callback) ***REMOVED***
        if (this.auth && 'getProjectId' in this.auth) ***REMOVED***
            return this.auth.getProjectId(callback);
      ***REMOVED***
        if (callback) ***REMOVED***
            callback(new Error('Cannot determine project ID.'));
      ***REMOVED***
        else ***REMOVED***
            return Promise.reject('Cannot determine project ID.');
      ***REMOVED***
  ***REMOVED***
    // Service calls
    getOperationInternal(request, options, callback) ***REMOVED***
        request = request || ***REMOVED***};
        options = options || ***REMOVED***};
        return this.innerApiCalls.getOperation(request, options, callback);
  ***REMOVED***
    /**
     * Gets the latest state of a long-running operation.  Clients can use this
     * method to poll the operation result at intervals as recommended by the API
     * service.
     *
     * @param ***REMOVED***Object} request - The request object that will be sent.
     * @param ***REMOVED***string} request.name - The name of the operation resource.
     * @param ***REMOVED***Object=} options
     *   Optional parameters. You can override the default settings for this call,
     *   e.g, timeout, retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link
     *   https://googleapis.github.io/gax-nodejs/global.html#CallOptions} for the
     *   details.
     * @param ***REMOVED***function(?Error, ?Object)=} callback
     *   The function which will be called with the result of the API call.
     *
     *   The second parameter to the callback is an object representing
     * [google.longrunning.Operation]***REMOVED***@link
     * external:"google.longrunning.Operation"}.
     * @return ***REMOVED***Promise} - The promise which resolves to an array.
     *   The first element of the array is an object representing
     * [google.longrunning.Operation]***REMOVED***@link
     * external:"google.longrunning.Operation"}. The promise has a method named
     * "cancel" which cancels the ongoing API call.
     *
     * @example
     *
     * const client = longrunning.operationsClient();
     * const name = '';
     * const [response] = await client.getOperation(***REMOVED***name});
     * // doThingsWith(response)
     */
    getOperation(request, optionsOrCallback, callback) ***REMOVED***
        let options;
        if (optionsOrCallback instanceof Function && callback === undefined) ***REMOVED***
            callback = optionsOrCallback;
            options = ***REMOVED***};
      ***REMOVED***
        else ***REMOVED***
            options = optionsOrCallback;
      ***REMOVED***
        request = request || ***REMOVED***};
        options = options || ***REMOVED***};
        return this.innerApiCalls.getOperation(request, options, callback);
  ***REMOVED***
    /**
     * Lists operations that match the specified filter in the request. If the
     * server doesn't support this method, it returns `UNIMPLEMENTED`.
     *
     * NOTE: the `name` binding below allows API services to override the binding
     * to use different resource name schemes.
     *
     * @param ***REMOVED***Object} request - The request object that will be sent.
     * @param ***REMOVED***string} request.name - The name of the operation collection.
     * @param ***REMOVED***string} request.filter - The standard list filter.
     * @param ***REMOVED***number=} request.pageSize
     *   The maximum number of resources contained in the underlying API
     *   response. If page streaming is performed per-resource, this
     *   parameter does not affect the return value. If page streaming is
     *   performed per-page, this determines the maximum number of
     *   resources in a page.
     * @param ***REMOVED***Object=} options
     *   Optional parameters. You can override the default settings for this call,
     * e.g, timeout, retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link
     * https://googleapis.github.io/gax-nodejs/global.html#CallOptions} for the
     * details.
     * @param ***REMOVED***function(?Error, ?Array, ?Object, ?Object)=} callback
     *   The function which will be called with the result of the API call.
     *
     *   The second parameter to the callback is Array of
     * [google.longrunning.Operation]***REMOVED***@link
     * external:"google.longrunning.Operation"}.
     *
     *   When autoPaginate: false is specified through options, it contains the
     * result in a single response. If the response indicates the next page
     * exists, the third parameter is set to be used for the next request object.
     * The fourth parameter keeps the raw response object of an object
     * representing [google.longrunning.ListOperationsResponse]***REMOVED***@link
     * external:"google.longrunning.ListOperationsResponse"}.
     * @return ***REMOVED***Promise} - The promise which resolves to an array.
     *   The first element of the array is Array of
     * [google.longrunning.Operation]***REMOVED***@link
     * external:"google.longrunning.Operation"}.
     *
     *   When autoPaginate: false is specified through options, the array has
     * three elements. The first element is Array of
     * [google.longrunning.Operation]***REMOVED***@link
     * external:"google.longrunning.Operation"} in a single response. The second
     * element is the next request object if the response indicates the next page
     * exists, or null. The third element is an object representing
     * [google.longrunning.ListOperationsResponse]***REMOVED***@link
     * external:"google.longrunning.ListOperationsResponse"}.
     *
     *   The promise has a method named "cancel" which cancels the ongoing API
     * call.
     *
     * @example
     *
     * const client = longrunning.operationsClient();
     * const request = ***REMOVED***
     *     name: '',
     *     filter: ''
     * };
     * // Iterate over all elements.
     * const [resources] = await client.listOperations(request);
     * for (const resource of resources) ***REMOVED***
     *   console.log(resources);
     * }
     *
     * // Or obtain the paged response.
     * const options = ***REMOVED***autoPaginate: false};
     * let nextRequest = request;
     * while(nextRequest) ***REMOVED***
     *   const response = await client.listOperations(nextRequest, options);
     *   const resources = response[0];
     *   nextRequest = response[1];
     *   const rawResponse = response[2];
     *   for (const resource of resources) ***REMOVED***
     *     // doThingsWith(resource);
     * ***REMOVED***
     * };
     */
    listOperations(request, optionsOrCallback, callback) ***REMOVED***
        let options;
        if (optionsOrCallback instanceof Function && callback === undefined) ***REMOVED***
            callback = optionsOrCallback;
            options = ***REMOVED***};
      ***REMOVED***
        else ***REMOVED***
            options = optionsOrCallback;
      ***REMOVED***
        request = request || ***REMOVED***};
        options = options || ***REMOVED***};
        return this.innerApiCalls.listOperations(request, options, callback);
  ***REMOVED***
    /**
     * Equivalent to ***REMOVED***@link listOperations}, but returns a NodeJS Stream object.
     *
     * This fetches the paged responses for ***REMOVED***@link listOperations} continuously
     * and invokes the callback registered for 'data' event for each element in
     * the responses.
     *
     * The returned object has 'end' method when no more elements are required.
     *
     * autoPaginate option will be ignored.
     *
     * @see ***REMOVED***@link https://nodejs.org/api/stream.html}
     *
     * @param ***REMOVED***Object} request - The request object that will be sent.
     * @param ***REMOVED***string} request.name - The name of the operation collection.
     * @param ***REMOVED***string} request.filter - The standard list filter.
     * @param ***REMOVED***number=} request.pageSize -
     *   The maximum number of resources contained in the underlying API
     *   response. If page streaming is performed per-resource, this
     *   parameter does not affect the return value. If page streaming is
     *   performed per-page, this determines the maximum number of
     *   resources in a page.
     * @param ***REMOVED***Object=} options
     *   Optional parameters. You can override the default settings for this call,
     *   e.g, timeout, retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link
     *   https://googleapis.github.io/gax-nodejs/global.html#CallOptions} for the
     *   details.
     * @return ***REMOVED***Stream} - An object stream which emits an object representing [google.longrunning.Operation]***REMOVED***@link external:"google.longrunning.Operation"} on 'data' event.
     *
     * @example
     *
     * const client = longrunning.operationsClient();
     * const request = ***REMOVED***
     *   name: '',
     *   filter: ''
     * };
     * client.listOperationsStream(request)
     *   .on('data', element => ***REMOVED***
     *     // doThingsWith(element)
     * ***REMOVED***)
     *   .on('error', err => ***REMOVED***
     *     console.error(err);
     * ***REMOVED***);
     */
    listOperationsStream(request, options) ***REMOVED***
        const callSettings = new gax.CallSettings(options);
        return this.descriptor.listOperations.createStream(this.innerApiCalls.listOperations, request, callSettings);
  ***REMOVED***
    /**
     * Equivalent to ***REMOVED***@link listOperations}, but returns an iterable object.
     *
     * for-await-of syntax is used with the iterable to recursively get response element on-demand.
     *
     * @param ***REMOVED***Object} request - The request object that will be sent.
     * @param ***REMOVED***string} request.name - The name of the operation collection.
     * @param ***REMOVED***string} request.filter - The standard list filter.
     * @param ***REMOVED***number=} request.pageSize -
     *   The maximum number of resources contained in the underlying API
     *   response. If page streaming is performed per-resource, this
     *   parameter does not affect the return value. If page streaming is
     *   performed per-page, this determines the maximum number of
     *   resources in a page.
     * @param ***REMOVED***Object=} options
     *   Optional parameters. You can override the default settings for this call,
     *   e.g, timeout, retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link
     *   https://googleapis.github.io/gax-nodejs/global.html#CallOptions} for the
     *   details.
     * @returns ***REMOVED***Object}
     *   An iterable Object that conforms to @link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols.
     */
    listOperationsAsync(request, options) ***REMOVED***
        request = request || ***REMOVED***};
        options = options || ***REMOVED***};
        const callSettings = new gax.CallSettings(options);
        return this.descriptor.listOperations.asyncIterate(this.innerApiCalls.listOperations, request, callSettings);
  ***REMOVED***
    /**
     * Starts asynchronous cancellation on a long-running operation.  The server
     * makes a best effort to cancel the operation, but success is not
     * guaranteed.  If the server doesn't support this method, it returns
     * `google.rpc.Code.UNIMPLEMENTED`.  Clients can use
     * ***REMOVED***@link Operations.GetOperation} or
     * other methods to check whether the cancellation succeeded or whether the
     * operation completed despite cancellation. On successful cancellation,
     * the operation is not deleted; instead, it becomes an operation with
     * an ***REMOVED***@link Operation.error} value with a ***REMOVED***@link google.rpc.Status.code} of
     * 1, corresponding to `Code.CANCELLED`.
     *
     * @param ***REMOVED***Object} request - The request object that will be sent.
     * @param ***REMOVED***string} request.name - The name of the operation resource to be cancelled.
     * @param ***REMOVED***Object=} options
     *   Optional parameters. You can override the default settings for this call,
     * e.g, timeout, retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link
     * https://googleapis.github.io/gax-nodejs/global.html#CallOptions} for the
     * details.
     * @param ***REMOVED***function(?Error)=} callback
     *   The function which will be called with the result of the API call.
     * @return ***REMOVED***Promise} - The promise which resolves when API call finishes.
     *   The promise has a method named "cancel" which cancels the ongoing API
     * call.
     *
     * @example
     *
     * const client = longrunning.operationsClient();
     * await client.cancelOperation(***REMOVED***name: ''});
     */
    cancelOperation(request, optionsOrCallback, callback) ***REMOVED***
        let options;
        if (optionsOrCallback instanceof Function && callback === undefined) ***REMOVED***
            callback = optionsOrCallback;
            options = ***REMOVED***};
      ***REMOVED***
        else ***REMOVED***
            options = optionsOrCallback;
      ***REMOVED***
        request = request || ***REMOVED***};
        options = options || ***REMOVED***};
        return this.innerApiCalls.cancelOperation(request, options, callback);
  ***REMOVED***
    /**
     * Deletes a long-running operation. This method indicates that the client is
     * no longer interested in the operation result. It does not cancel the
     * operation. If the server doesn't support this method, it returns
     * `google.rpc.Code.UNIMPLEMENTED`.
     *
     * @param ***REMOVED***Object} request - The request object that will be sent.
     * @param ***REMOVED***string} request.name - The name of the operation resource to be deleted.
     * @param ***REMOVED***Object=} options
     *   Optional parameters. You can override the default settings for this call,
     * e.g, timeout, retries, paginations, etc. See [gax.CallOptions]***REMOVED***@link
     * https://googleapis.github.io/gax-nodejs/global.html#CallOptions} for the
     * details.
     * @param ***REMOVED***function(?Error)=} callback
     *   The function which will be called with the result of the API call.
     * @return ***REMOVED***Promise} - The promise which resolves when API call finishes.
     *   The promise has a method named "cancel" which cancels the ongoing API
     * call.
     *
     * @example
     *
     * const client = longrunning.operationsClient();
     * await client.deleteOperation(***REMOVED***name: ''});
     */
    deleteOperation(request, optionsOrCallback, callback) ***REMOVED***
        let options;
        if (optionsOrCallback instanceof Function && callback === undefined) ***REMOVED***
            callback = optionsOrCallback;
            options = ***REMOVED***};
      ***REMOVED***
        else ***REMOVED***
            options = optionsOrCallback;
      ***REMOVED***
        request = request || ***REMOVED***};
        options = options || ***REMOVED***};
        return this.innerApiCalls.deleteOperation(request, options, callback);
  ***REMOVED***
}
exports.OperationsClient = OperationsClient;
class OperationsClientBuilder ***REMOVED***
    /**
     * Builds a new Operations Client
     * @param gaxGrpc ***REMOVED***GrpcClient}
     */
    constructor(gaxGrpc, protoJson) ***REMOVED***
        if (protoJson && gaxGrpc.httpRules) ***REMOVED***
            // overwrite the http rules if provide in service yaml.
            (0, transcoding_1.overrideHttpRules)(gaxGrpc.httpRules, protoJson);
      ***REMOVED***
        const operationsProtos = protoJson !== null && protoJson !== void 0 ? protoJson : gaxGrpc.loadProtoJSON(operationProtoJson);
        /**
         * Build a new instance of ***REMOVED***@link OperationsClient}.
         *
         * @param ***REMOVED***Object=} opts - The optional parameters.
         * @param ***REMOVED***String=} opts.servicePath - Domain name of the API remote host.
         * @param ***REMOVED***number=} opts.port - The port on which to connect to the remote host.
         * @param ***REMOVED***grpc.ClientCredentials=} opts.sslCreds - A ClientCredentials for use with an SSL-enabled channel.
         * @param ***REMOVED***Object=} opts.clientConfig - The customized config to build the call settings. See ***REMOVED***@link gax.constructSettings} for the format.
         */
        this.operationsClient = opts => ***REMOVED***
            if (gaxGrpc.fallback) ***REMOVED***
                opts.fallback = gaxGrpc.fallback;
          ***REMOVED***
            return new OperationsClient(gaxGrpc, operationsProtos, opts);
      ***REMOVED***;
        Object.assign(this.operationsClient, OperationsClient);
  ***REMOVED***
}
exports.OperationsClientBuilder = OperationsClientBuilder;
