// @firebase/util/dist/index.node.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', ***REMOVED*** value: true });

var tslib = require('tslib');

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * @fileoverview Firebase constants.  Some of these (@defines) can be overridden at compile-time.
 */
var CONSTANTS = ***REMOVED***
    /**
     * @define ***REMOVED***boolean} Whether this is the client Node.js SDK.
     */
    NODE_CLIENT: false,
    /**
     * @define ***REMOVED***boolean} Whether this is the Admin Node.js SDK.
     */
    NODE_ADMIN: false,
    /**
     * Firebase SDK Version
     */
    SDK_VERSION: '$***REMOVED***JSCORE_VERSION}'
};

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Throws an error if the provided assertion is falsy
 */
var assert = function (assertion, message) ***REMOVED***
    if (!assertion) ***REMOVED***
        throw assertionError(message);
  ***REMOVED***
};
/**
 * Returns an Error object suitable for throwing.
 */
var assertionError = function (message) ***REMOVED***
    return new Error('Firebase Database (' +
        CONSTANTS.SDK_VERSION +
        ') INTERNAL ASSERT FAILED: ' +
        message);
};

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var stringToByteArray$1 = function (str) ***REMOVED***
    // TODO(user): Use native implementations if/when available
    var out = [];
    var p = 0;
    for (var i = 0; i < str.length; i++) ***REMOVED***
        var c = str.charCodeAt(i);
        if (c < 128) ***REMOVED***
            out[p++] = c;
      ***REMOVED***
        else if (c < 2048) ***REMOVED***
            out[p++] = (c >> 6) | 192;
            out[p++] = (c & 63) | 128;
      ***REMOVED***
        else if ((c & 0xfc00) === 0xd800 &&
            i + 1 < str.length &&
            (str.charCodeAt(i + 1) & 0xfc00) === 0xdc00) ***REMOVED***
            // Surrogate Pair
            c = 0x10000 + ((c & 0x03ff) << 10) + (str.charCodeAt(++i) & 0x03ff);
            out[p++] = (c >> 18) | 240;
            out[p++] = ((c >> 12) & 63) | 128;
            out[p++] = ((c >> 6) & 63) | 128;
            out[p++] = (c & 63) | 128;
      ***REMOVED***
        else ***REMOVED***
            out[p++] = (c >> 12) | 224;
            out[p++] = ((c >> 6) & 63) | 128;
            out[p++] = (c & 63) | 128;
      ***REMOVED***
  ***REMOVED***
    return out;
};
/**
 * Turns an array of numbers into the string given by the concatenation of the
 * characters to which the numbers correspond.
 * @param bytes Array of numbers representing characters.
 * @return Stringification of the array.
 */
var byteArrayToString = function (bytes) ***REMOVED***
    // TODO(user): Use native implementations if/when available
    var out = [];
    var pos = 0, c = 0;
    while (pos < bytes.length) ***REMOVED***
        var c1 = bytes[pos++];
        if (c1 < 128) ***REMOVED***
            out[c++] = String.fromCharCode(c1);
      ***REMOVED***
        else if (c1 > 191 && c1 < 224) ***REMOVED***
            var c2 = bytes[pos++];
            out[c++] = String.fromCharCode(((c1 & 31) << 6) | (c2 & 63));
      ***REMOVED***
        else if (c1 > 239 && c1 < 365) ***REMOVED***
            // Surrogate Pair
            var c2 = bytes[pos++];
            var c3 = bytes[pos++];
            var c4 = bytes[pos++];
            var u = (((c1 & 7) << 18) | ((c2 & 63) << 12) | ((c3 & 63) << 6) | (c4 & 63)) -
                0x10000;
            out[c++] = String.fromCharCode(0xd800 + (u >> 10));
            out[c++] = String.fromCharCode(0xdc00 + (u & 1023));
      ***REMOVED***
        else ***REMOVED***
            var c2 = bytes[pos++];
            var c3 = bytes[pos++];
            out[c++] = String.fromCharCode(((c1 & 15) << 12) | ((c2 & 63) << 6) | (c3 & 63));
      ***REMOVED***
  ***REMOVED***
    return out.join('');
};
// We define it as an object literal instead of a class because a class compiled down to es5 can't
// be treeshaked. https://github.com/rollup/rollup/issues/1691
// Static lookup maps, lazily populated by init_()
var base64 = ***REMOVED***
    /**
     * Maps bytes to characters.
     */
    byteToCharMap_: null,
    /**
     * Maps characters to bytes.
     */
    charToByteMap_: null,
    /**
     * Maps bytes to websafe characters.
     * @private
     */
    byteToCharMapWebSafe_: null,
    /**
     * Maps websafe characters to bytes.
     * @private
     */
    charToByteMapWebSafe_: null,
    /**
     * Our default alphabet, shared between
     * ENCODED_VALS and ENCODED_VALS_WEBSAFE
     */
    ENCODED_VALS_BASE: 'ABCDEFGHIJKLMNOPQRSTUVWXYZ' + 'abcdefghijklmnopqrstuvwxyz' + '0123456789',
    /**
     * Our default alphabet. Value 64 (=) is special; it means "nothing."
     */
    get ENCODED_VALS() ***REMOVED***
        return this.ENCODED_VALS_BASE + '+/=';
  ***REMOVED***,
    /**
     * Our websafe alphabet.
     */
    get ENCODED_VALS_WEBSAFE() ***REMOVED***
        return this.ENCODED_VALS_BASE + '-_.';
  ***REMOVED***,
    /**
     * Whether this browser supports the atob and btoa functions. This extension
     * started at Mozilla but is now implemented by many browsers. We use the
     * ASSUME_* variables to avoid pulling in the full useragent detection library
     * but still allowing the standard per-browser compilations.
     *
     */
    HAS_NATIVE_SUPPORT: typeof atob === 'function',
    /**
     * Base64-encode an array of bytes.
     *
     * @param input An array of bytes (numbers with
     *     value in [0, 255]) to encode.
     * @param webSafe Boolean indicating we should use the
     *     alternative alphabet.
     * @return The base64 encoded string.
     */
    encodeByteArray: function (input, webSafe) ***REMOVED***
        if (!Array.isArray(input)) ***REMOVED***
            throw Error('encodeByteArray takes an array as a parameter');
      ***REMOVED***
        this.init_();
        var byteToCharMap = webSafe
            ? this.byteToCharMapWebSafe_
            : this.byteToCharMap_;
        var output = [];
        for (var i = 0; i < input.length; i += 3) ***REMOVED***
            var byte1 = input[i];
            var haveByte2 = i + 1 < input.length;
            var byte2 = haveByte2 ? input[i + 1] : 0;
            var haveByte3 = i + 2 < input.length;
            var byte3 = haveByte3 ? input[i + 2] : 0;
            var outByte1 = byte1 >> 2;
            var outByte2 = ((byte1 & 0x03) << 4) | (byte2 >> 4);
            var outByte3 = ((byte2 & 0x0f) << 2) | (byte3 >> 6);
            var outByte4 = byte3 & 0x3f;
            if (!haveByte3) ***REMOVED***
                outByte4 = 64;
                if (!haveByte2) ***REMOVED***
                    outByte3 = 64;
              ***REMOVED***
          ***REMOVED***
            output.push(byteToCharMap[outByte1], byteToCharMap[outByte2], byteToCharMap[outByte3], byteToCharMap[outByte4]);
      ***REMOVED***
        return output.join('');
  ***REMOVED***,
    /**
     * Base64-encode a string.
     *
     * @param input A string to encode.
     * @param webSafe If true, we should use the
     *     alternative alphabet.
     * @return The base64 encoded string.
     */
    encodeString: function (input, webSafe) ***REMOVED***
        // Shortcut for Mozilla browsers that implement
        // a native base64 encoder in the form of "btoa/atob"
        if (this.HAS_NATIVE_SUPPORT && !webSafe) ***REMOVED***
            return btoa(input);
      ***REMOVED***
        return this.encodeByteArray(stringToByteArray$1(input), webSafe);
  ***REMOVED***,
    /**
     * Base64-decode a string.
     *
     * @param input to decode.
     * @param webSafe True if we should use the
     *     alternative alphabet.
     * @return string representing the decoded value.
     */
    decodeString: function (input, webSafe) ***REMOVED***
        // Shortcut for Mozilla browsers that implement
        // a native base64 encoder in the form of "btoa/atob"
        if (this.HAS_NATIVE_SUPPORT && !webSafe) ***REMOVED***
            return atob(input);
      ***REMOVED***
        return byteArrayToString(this.decodeStringToByteArray(input, webSafe));
  ***REMOVED***,
    /**
     * Base64-decode a string.
     *
     * In base-64 decoding, groups of four characters are converted into three
     * bytes.  If the encoder did not apply padding, the input length may not
     * be a multiple of 4.
     *
     * In this case, the last group will have fewer than 4 characters, and
     * padding will be inferred.  If the group has one or two characters, it decodes
     * to one byte.  If the group has three characters, it decodes to two bytes.
     *
     * @param input Input to decode.
     * @param webSafe True if we should use the web-safe alphabet.
     * @return bytes representing the decoded value.
     */
    decodeStringToByteArray: function (input, webSafe) ***REMOVED***
        this.init_();
        var charToByteMap = webSafe
            ? this.charToByteMapWebSafe_
            : this.charToByteMap_;
        var output = [];
        for (var i = 0; i < input.length;) ***REMOVED***
            var byte1 = charToByteMap[input.charAt(i++)];
            var haveByte2 = i < input.length;
            var byte2 = haveByte2 ? charToByteMap[input.charAt(i)] : 0;
            ++i;
            var haveByte3 = i < input.length;
            var byte3 = haveByte3 ? charToByteMap[input.charAt(i)] : 64;
            ++i;
            var haveByte4 = i < input.length;
            var byte4 = haveByte4 ? charToByteMap[input.charAt(i)] : 64;
            ++i;
            if (byte1 == null || byte2 == null || byte3 == null || byte4 == null) ***REMOVED***
                throw Error();
          ***REMOVED***
            var outByte1 = (byte1 << 2) | (byte2 >> 4);
            output.push(outByte1);
            if (byte3 !== 64) ***REMOVED***
                var outByte2 = ((byte2 << 4) & 0xf0) | (byte3 >> 2);
                output.push(outByte2);
                if (byte4 !== 64) ***REMOVED***
                    var outByte3 = ((byte3 << 6) & 0xc0) | byte4;
                    output.push(outByte3);
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        return output;
  ***REMOVED***,
    /**
     * Lazy static initialization function. Called before
     * accessing any of the static map variables.
     * @private
     */
    init_: function () ***REMOVED***
        if (!this.byteToCharMap_) ***REMOVED***
            this.byteToCharMap_ = ***REMOVED***};
            this.charToByteMap_ = ***REMOVED***};
            this.byteToCharMapWebSafe_ = ***REMOVED***};
            this.charToByteMapWebSafe_ = ***REMOVED***};
            // We want quick mappings back and forth, so we precompute two maps.
            for (var i = 0; i < this.ENCODED_VALS.length; i++) ***REMOVED***
                this.byteToCharMap_[i] = this.ENCODED_VALS.charAt(i);
                this.charToByteMap_[this.byteToCharMap_[i]] = i;
                this.byteToCharMapWebSafe_[i] = this.ENCODED_VALS_WEBSAFE.charAt(i);
                this.charToByteMapWebSafe_[this.byteToCharMapWebSafe_[i]] = i;
                // Be forgiving when decoding and correctly decode both encodings.
                if (i >= this.ENCODED_VALS_BASE.length) ***REMOVED***
                    this.charToByteMap_[this.ENCODED_VALS_WEBSAFE.charAt(i)] = i;
                    this.charToByteMapWebSafe_[this.ENCODED_VALS.charAt(i)] = i;
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
};
/**
 * URL-safe base64 encoding
 */
var base64Encode = function (str) ***REMOVED***
    var utf8Bytes = stringToByteArray$1(str);
    return base64.encodeByteArray(utf8Bytes, true);
};
/**
 * URL-safe base64 encoding (without "." padding in the end).
 * e.g. Used in JSON Web Token (JWT) parts.
 */
var base64urlEncodeWithoutPadding = function (str) ***REMOVED***
    // Use base64url encoding and remove padding in the end (dot characters).
    return base64Encode(str).replace(/\./g, '');
};
/**
 * URL-safe base64 decoding
 *
 * NOTE: DO NOT use the global atob() function - it does NOT support the
 * base64Url variant encoding.
 *
 * @param str To be decoded
 * @return Decoded result, if possible
 */
var base64Decode = function (str) ***REMOVED***
    try ***REMOVED***
        return base64.decodeString(str, true);
  ***REMOVED***
    catch (e) ***REMOVED***
        console.error('base64Decode failed: ', e);
  ***REMOVED***
    return null;
};

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Do a deep-copy of basic JavaScript Objects or Arrays.
 */
function deepCopy(value) ***REMOVED***
    return deepExtend(undefined, value);
}
/**
 * Copy properties from source to target (recursively allows extension
 * of Objects and Arrays).  Scalar values in the target are over-written.
 * If target is undefined, an object of the appropriate type will be created
 * (and returned).
 *
 * We recursively copy all child properties of plain Objects in the source- so
 * that namespace- like dictionaries are merged.
 *
 * Note that the target can be a function, in which case the properties in
 * the source Object are copied onto it as static properties of the Function.
 *
 * Note: we don't merge __proto__ to prevent prototype pollution
 */
function deepExtend(target, source) ***REMOVED***
    if (!(source instanceof Object)) ***REMOVED***
        return source;
  ***REMOVED***
    switch (source.constructor) ***REMOVED***
        case Date:
            // Treat Dates like scalars; if the target date object had any child
            // properties - they will be lost!
            var dateValue = source;
            return new Date(dateValue.getTime());
        case Object:
            if (target === undefined) ***REMOVED***
                target = ***REMOVED***};
          ***REMOVED***
            break;
        case Array:
            // Always copy the array source and overwrite the target.
            target = [];
            break;
        default:
            // Not a plain Object - treat it as a scalar.
            return source;
  ***REMOVED***
    for (var prop in source) ***REMOVED***
        // use isValidKey to guard against prototype pollution. See https://snyk.io/vuln/SNYK-JS-LODASH-450202
        if (!source.hasOwnProperty(prop) || !isValidKey(prop)) ***REMOVED***
            continue;
      ***REMOVED***
        target[prop] = deepExtend(target[prop], source[prop]);
  ***REMOVED***
    return target;
}
function isValidKey(key) ***REMOVED***
    return key !== '__proto__';
}

/**
 * @license
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Polyfill for `globalThis` object.
 * @returns the `globalThis` object for the given environment.
 * @public
 */
function getGlobal() ***REMOVED***
    if (typeof self !== 'undefined') ***REMOVED***
        return self;
  ***REMOVED***
    if (typeof window !== 'undefined') ***REMOVED***
        return window;
  ***REMOVED***
    if (typeof global !== 'undefined') ***REMOVED***
        return global;
  ***REMOVED***
    throw new Error('Unable to locate global object.');
}

/**
 * @license
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var getDefaultsFromGlobal = function () ***REMOVED***
    return getGlobal().__FIREBASE_DEFAULTS__;
};
/**
 * Attempt to read defaults from a JSON string provided to
 * process(.)env(.)__FIREBASE_DEFAULTS__ or a JSON file whose path is in
 * process(.)env(.)__FIREBASE_DEFAULTS_PATH__
 * The dots are in parens because certain compilers (Vite?) cannot
 * handle seeing that variable in comments.
 * See https://github.com/firebase/firebase-js-sdk/issues/6838
 */
var getDefaultsFromEnvVariable = function () ***REMOVED***
    if (typeof process === 'undefined' || typeof process.env === 'undefined') ***REMOVED***
        return;
  ***REMOVED***
    var defaultsJsonString = process.env.__FIREBASE_DEFAULTS__;
    if (defaultsJsonString) ***REMOVED***
        return JSON.parse(defaultsJsonString);
  ***REMOVED***
};
var getDefaultsFromCookie = function () ***REMOVED***
    if (typeof document === 'undefined') ***REMOVED***
        return;
  ***REMOVED***
    var match;
    try ***REMOVED***
        match = document.cookie.match(/__FIREBASE_DEFAULTS__=([^;]+)/);
  ***REMOVED***
    catch (e) ***REMOVED***
        // Some environments such as Angular Universal SSR have a
        // `document` object but error on accessing `document.cookie`.
        return;
  ***REMOVED***
    var decoded = match && base64Decode(match[1]);
    return decoded && JSON.parse(decoded);
};
/**
 * Get the __FIREBASE_DEFAULTS__ object. It checks in order:
 * (1) if such an object exists as a property of `globalThis`
 * (2) if such an object was provided on a shell environment variable
 * (3) if such an object exists in a cookie
 * @public
 */
var getDefaults = function () ***REMOVED***
    try ***REMOVED***
        return (getDefaultsFromGlobal() ||
            getDefaultsFromEnvVariable() ||
            getDefaultsFromCookie());
  ***REMOVED***
    catch (e) ***REMOVED***
        /**
         * Catch-all for being unable to get __FIREBASE_DEFAULTS__ due
         * to any environment case we have not accounted for. Log to
         * info instead of swallowing so we can find these unknown cases
         * and add paths for them if needed.
         */
        console.info("Unable to get __FIREBASE_DEFAULTS__ due to: ".concat(e));
        return;
  ***REMOVED***
};
/**
 * Returns emulator host stored in the __FIREBASE_DEFAULTS__ object
 * for the given product.
 * @returns a URL host formatted like `127.0.0.1:9999` or `[::1]:4000` if available
 * @public
 */
var getDefaultEmulatorHost = function (productName) ***REMOVED*** var _a, _b; return (_b = (_a = getDefaults()) === null || _a === void 0 ? void 0 : _a.emulatorHosts) === null || _b === void 0 ? void 0 : _b[productName]; };
/**
 * Returns emulator hostname and port stored in the __FIREBASE_DEFAULTS__ object
 * for the given product.
 * @returns a pair of hostname and port like `["::1", 4000]` if available
 * @public
 */
var getDefaultEmulatorHostnameAndPort = function (productName) ***REMOVED***
    var host = getDefaultEmulatorHost(productName);
    if (!host) ***REMOVED***
        return undefined;
  ***REMOVED***
    var separatorIndex = host.lastIndexOf(':'); // Finding the last since IPv6 addr also has colons.
    if (separatorIndex <= 0 || separatorIndex + 1 === host.length) ***REMOVED***
        throw new Error("Invalid host ".concat(host, " with no separate hostname and port!"));
  ***REMOVED***
    // eslint-disable-next-line no-restricted-globals
    var port = parseInt(host.substring(separatorIndex + 1), 10);
    if (host[0] === '[') ***REMOVED***
        // Bracket-quoted `[ipv6addr]:port` => return "ipv6addr" (without brackets).
        return [host.substring(1, separatorIndex - 1), port];
  ***REMOVED***
    else ***REMOVED***
        return [host.substring(0, separatorIndex), port];
  ***REMOVED***
};
/**
 * Returns Firebase app config stored in the __FIREBASE_DEFAULTS__ object.
 * @public
 */
var getDefaultAppConfig = function () ***REMOVED*** var _a; return (_a = getDefaults()) === null || _a === void 0 ? void 0 : _a.config; };
/**
 * Returns an experimental setting on the __FIREBASE_DEFAULTS__ object (properties
 * prefixed by "_")
 * @public
 */
var getExperimentalSetting = function (name) ***REMOVED*** var _a; return (_a = getDefaults()) === null || _a === void 0 ? void 0 : _a["_".concat(name)]; };

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var Deferred = /** @class */ (function () ***REMOVED***
    function Deferred() ***REMOVED***
        var _this = this;
        this.reject = function () ***REMOVED*** };
        this.resolve = function () ***REMOVED*** };
        this.promise = new Promise(function (resolve, reject) ***REMOVED***
            _this.resolve = resolve;
            _this.reject = reject;
      ***REMOVED***);
  ***REMOVED***
    /**
     * Our API internals are not promiseified and cannot because our callback APIs have subtle expectations around
     * invoking promises inline, which Promises are forbidden to do. This method accepts an optional node-style callback
     * and returns a node-style callback which will resolve or reject the Deferred's promise.
     */
    Deferred.prototype.wrapCallback = function (callback) ***REMOVED***
        var _this = this;
        return function (error, value) ***REMOVED***
            if (error) ***REMOVED***
                _this.reject(error);
          ***REMOVED***
            else ***REMOVED***
                _this.resolve(value);
          ***REMOVED***
            if (typeof callback === 'function') ***REMOVED***
                // Attaching noop handler just in case developer wasn't expecting
                // promises
                _this.promise.catch(function () ***REMOVED*** });
                // Some of our callbacks don't expect a value and our own tests
                // assert that the parameter length is 1
                if (callback.length === 1) ***REMOVED***
                    callback(error);
              ***REMOVED***
                else ***REMOVED***
                    callback(error, value);
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***;
  ***REMOVED***;
    return Deferred;
}());

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function createMockUserToken(token, projectId) ***REMOVED***
    if (token.uid) ***REMOVED***
        throw new Error('The "uid" field is no longer supported by mockUserToken. Please use "sub" instead for Firebase Auth User ID.');
  ***REMOVED***
    // Unsecured JWTs use "none" as the algorithm.
    var header = ***REMOVED***
        alg: 'none',
        type: 'JWT'
  ***REMOVED***;
    var project = projectId || 'demo-project';
    var iat = token.iat || 0;
    var sub = token.sub || token.user_id;
    if (!sub) ***REMOVED***
        throw new Error("mockUserToken must contain 'sub' or 'user_id' field!");
  ***REMOVED***
    var payload = tslib.__assign(***REMOVED*** 
        // Set all required fields to decent defaults
        iss: "https://securetoken.google.com/".concat(project), aud: project, iat: iat, exp: iat + 3600, auth_time: iat, sub: sub, user_id: sub, firebase: ***REMOVED***
            sign_in_provider: 'custom',
            identities: ***REMOVED***}
      ***REMOVED*** }, token);
    // Unsecured JWTs use the empty string as a signature.
    var signature = '';
    return [
        base64urlEncodeWithoutPadding(JSON.stringify(header)),
        base64urlEncodeWithoutPadding(JSON.stringify(payload)),
        signature
    ].join('.');
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Returns navigator.userAgent string or '' if it's not defined.
 * @return user agent string
 */
function getUA() ***REMOVED***
    if (typeof navigator !== 'undefined' &&
        typeof navigator['userAgent'] === 'string') ***REMOVED***
        return navigator['userAgent'];
  ***REMOVED***
    else ***REMOVED***
        return '';
  ***REMOVED***
}
/**
 * Detect Cordova / PhoneGap / Ionic frameworks on a mobile device.
 *
 * Deliberately does not rely on checking `file://` URLs (as this fails PhoneGap
 * in the Ripple emulator) nor Cordova `onDeviceReady`, which would normally
 * wait for a callback.
 */
function isMobileCordova() ***REMOVED***
    return (typeof window !== 'undefined' &&
        // @ts-ignore Setting up an broadly applicable index signature for Window
        // just to deal with this case would probably be a bad idea.
        !!(window['cordova'] || window['phonegap'] || window['PhoneGap']) &&
        /ios|iphone|ipod|ipad|android|blackberry|iemobile/i.test(getUA()));
}
/**
 * Detect Node.js.
 *
 * @return true if Node.js environment is detected or specified.
 */
// Node detection logic from: https://github.com/iliakan/detect-node/
function isNode() ***REMOVED***
    var _a;
    var forceEnvironment = (_a = getDefaults()) === null || _a === void 0 ? void 0 : _a.forceEnvironment;
    if (forceEnvironment === 'node') ***REMOVED***
        return true;
  ***REMOVED***
    else if (forceEnvironment === 'browser') ***REMOVED***
        return false;
  ***REMOVED***
    try ***REMOVED***
        return (Object.prototype.toString.call(global.process) === '[object process]');
  ***REMOVED***
    catch (e) ***REMOVED***
        return false;
  ***REMOVED***
}
/**
 * Detect Browser Environment
 */
function isBrowser() ***REMOVED***
    return typeof self === 'object' && self.self === self;
}
function isBrowserExtension() ***REMOVED***
    var runtime = typeof chrome === 'object'
        ? chrome.runtime
        : typeof browser === 'object'
            ? browser.runtime
            : undefined;
    return typeof runtime === 'object' && runtime.id !== undefined;
}
/**
 * Detect React Native.
 *
 * @return true if ReactNative environment is detected.
 */
function isReactNative() ***REMOVED***
    return (typeof navigator === 'object' && navigator['product'] === 'ReactNative');
}
/** Detects Electron apps. */
function isElectron() ***REMOVED***
    return getUA().indexOf('Electron/') >= 0;
}
/** Detects Internet Explorer. */
function isIE() ***REMOVED***
    var ua = getUA();
    return ua.indexOf('MSIE ') >= 0 || ua.indexOf('Trident/') >= 0;
}
/** Detects Universal Windows Platform apps. */
function isUWP() ***REMOVED***
    return getUA().indexOf('MSAppHost/') >= 0;
}
/**
 * Detect whether the current SDK build is the Node version.
 *
 * @return true if it's the Node SDK build.
 */
function isNodeSdk() ***REMOVED***
    return CONSTANTS.NODE_CLIENT === true || CONSTANTS.NODE_ADMIN === true;
}
/** Returns true if we are running in Safari. */
function isSafari() ***REMOVED***
    return (!isNode() &&
        navigator.userAgent.includes('Safari') &&
        !navigator.userAgent.includes('Chrome'));
}
/**
 * This method checks if indexedDB is supported by current browser/service worker context
 * @return true if indexedDB is supported by current browser/service worker context
 */
function isIndexedDBAvailable() ***REMOVED***
    try ***REMOVED***
        return typeof indexedDB === 'object';
  ***REMOVED***
    catch (e) ***REMOVED***
        return false;
  ***REMOVED***
}
/**
 * This method validates browser/sw context for indexedDB by opening a dummy indexedDB database and reject
 * if errors occur during the database open operation.
 *
 * @throws exception if current browser/sw context can't run idb.open (ex: Safari iframe, Firefox
 * private browsing)
 */
function validateIndexedDBOpenable() ***REMOVED***
    return new Promise(function (resolve, reject) ***REMOVED***
        try ***REMOVED***
            var preExist_1 = true;
            var DB_CHECK_NAME_1 = 'validate-browser-context-for-indexeddb-analytics-module';
            var request_1 = self.indexedDB.open(DB_CHECK_NAME_1);
            request_1.onsuccess = function () ***REMOVED***
                request_1.result.close();
                // delete database only when it doesn't pre-exist
                if (!preExist_1) ***REMOVED***
                    self.indexedDB.deleteDatabase(DB_CHECK_NAME_1);
              ***REMOVED***
                resolve(true);
          ***REMOVED***;
            request_1.onupgradeneeded = function () ***REMOVED***
                preExist_1 = false;
          ***REMOVED***;
            request_1.onerror = function () ***REMOVED***
                var _a;
                reject(((_a = request_1.error) === null || _a === void 0 ? void 0 : _a.message) || '');
          ***REMOVED***;
      ***REMOVED***
        catch (error) ***REMOVED***
            reject(error);
      ***REMOVED***
  ***REMOVED***);
}
/**
 *
 * This method checks whether cookie is enabled within current browser
 * @return true if cookie is enabled within current browser
 */
function areCookiesEnabled() ***REMOVED***
    if (typeof navigator === 'undefined' || !navigator.cookieEnabled) ***REMOVED***
        return false;
  ***REMOVED***
    return true;
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var ERROR_NAME = 'FirebaseError';
// Based on code from:
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#Custom_Error_Types
var FirebaseError = /** @class */ (function (_super) ***REMOVED***
    tslib.__extends(FirebaseError, _super);
    function FirebaseError(
    /** The error code for this error. */
    code, message, 
    /** Custom data for this error. */
    customData) ***REMOVED***
        var _this = _super.call(this, message) || this;
        _this.code = code;
        _this.customData = customData;
        /** The custom name for all FirebaseErrors. */
        _this.name = ERROR_NAME;
        // Fix For ES5
        // https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
        Object.setPrototypeOf(_this, FirebaseError.prototype);
        // Maintains proper stack trace for where our error was thrown.
        // Only available on V8.
        if (Error.captureStackTrace) ***REMOVED***
            Error.captureStackTrace(_this, ErrorFactory.prototype.create);
      ***REMOVED***
        return _this;
  ***REMOVED***
    return FirebaseError;
}(Error));
var ErrorFactory = /** @class */ (function () ***REMOVED***
    function ErrorFactory(service, serviceName, errors) ***REMOVED***
        this.service = service;
        this.serviceName = serviceName;
        this.errors = errors;
  ***REMOVED***
    ErrorFactory.prototype.create = function (code) ***REMOVED***
        var data = [];
        for (var _i = 1; _i < arguments.length; _i++) ***REMOVED***
            data[_i - 1] = arguments[_i];
      ***REMOVED***
        var customData = data[0] || ***REMOVED***};
        var fullCode = "".concat(this.service, "/").concat(code);
        var template = this.errors[code];
        var message = template ? replaceTemplate(template, customData) : 'Error';
        // Service Name: Error message (service/code).
        var fullMessage = "".concat(this.serviceName, ": ").concat(message, " (").concat(fullCode, ").");
        var error = new FirebaseError(fullCode, fullMessage, customData);
        return error;
  ***REMOVED***;
    return ErrorFactory;
}());
function replaceTemplate(template, data) ***REMOVED***
    return template.replace(PATTERN, function (_, key) ***REMOVED***
        var value = data[key];
        return value != null ? String(value) : "<".concat(key, "?>");
  ***REMOVED***);
}
var PATTERN = /\***REMOVED***\$([^}]+)}/g;

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Evaluates a JSON string into a javascript object.
 *
 * @param ***REMOVED***string} str A string containing JSON.
 * @return ***REMOVED****} The javascript object representing the specified JSON.
 */
function jsonEval(str) ***REMOVED***
    return JSON.parse(str);
}
/**
 * Returns JSON representing a javascript object.
 * @param ***REMOVED****} data Javascript object to be stringified.
 * @return ***REMOVED***string} The JSON contents of the object.
 */
function stringify(data) ***REMOVED***
    return JSON.stringify(data);
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Decodes a Firebase auth. token into constituent parts.
 *
 * Notes:
 * - May return with invalid / incomplete claims if there's no native base64 decoding support.
 * - Doesn't check if the token is actually valid.
 */
var decode = function (token) ***REMOVED***
    var header = ***REMOVED***}, claims = ***REMOVED***}, data = ***REMOVED***}, signature = '';
    try ***REMOVED***
        var parts = token.split('.');
        header = jsonEval(base64Decode(parts[0]) || '');
        claims = jsonEval(base64Decode(parts[1]) || '');
        signature = parts[2];
        data = claims['d'] || ***REMOVED***};
        delete claims['d'];
  ***REMOVED***
    catch (e) ***REMOVED*** }
    return ***REMOVED***
        header: header,
        claims: claims,
        data: data,
        signature: signature
  ***REMOVED***;
};
/**
 * Decodes a Firebase auth. token and checks the validity of its time-based claims. Will return true if the
 * token is within the time window authorized by the 'nbf' (not-before) and 'iat' (issued-at) claims.
 *
 * Notes:
 * - May return a false negative if there's no native base64 decoding support.
 * - Doesn't check if the token is actually valid.
 */
var isValidTimestamp = function (token) ***REMOVED***
    var claims = decode(token).claims;
    var now = Math.floor(new Date().getTime() / 1000);
    var validSince = 0, validUntil = 0;
    if (typeof claims === 'object') ***REMOVED***
        if (claims.hasOwnProperty('nbf')) ***REMOVED***
            validSince = claims['nbf'];
      ***REMOVED***
        else if (claims.hasOwnProperty('iat')) ***REMOVED***
            validSince = claims['iat'];
      ***REMOVED***
        if (claims.hasOwnProperty('exp')) ***REMOVED***
            validUntil = claims['exp'];
      ***REMOVED***
        else ***REMOVED***
            // token will expire after 24h by default
            validUntil = validSince + 86400;
      ***REMOVED***
  ***REMOVED***
    return (!!now &&
        !!validSince &&
        !!validUntil &&
        now >= validSince &&
        now <= validUntil);
};
/**
 * Decodes a Firebase auth. token and returns its issued at time if valid, null otherwise.
 *
 * Notes:
 * - May return null if there's no native base64 decoding support.
 * - Doesn't check if the token is actually valid.
 */
var issuedAtTime = function (token) ***REMOVED***
    var claims = decode(token).claims;
    if (typeof claims === 'object' && claims.hasOwnProperty('iat')) ***REMOVED***
        return claims['iat'];
  ***REMOVED***
    return null;
};
/**
 * Decodes a Firebase auth. token and checks the validity of its format. Expects a valid issued-at time.
 *
 * Notes:
 * - May return a false negative if there's no native base64 decoding support.
 * - Doesn't check if the token is actually valid.
 */
var isValidFormat = function (token) ***REMOVED***
    var decoded = decode(token), claims = decoded.claims;
    return !!claims && typeof claims === 'object' && claims.hasOwnProperty('iat');
};
/**
 * Attempts to peer into an auth token and determine if it's an admin auth token by looking at the claims portion.
 *
 * Notes:
 * - May return a false negative if there's no native base64 decoding support.
 * - Doesn't check if the token is actually valid.
 */
var isAdmin = function (token) ***REMOVED***
    var claims = decode(token).claims;
    return typeof claims === 'object' && claims['admin'] === true;
};

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function contains(obj, key) ***REMOVED***
    return Object.prototype.hasOwnProperty.call(obj, key);
}
function safeGet(obj, key) ***REMOVED***
    if (Object.prototype.hasOwnProperty.call(obj, key)) ***REMOVED***
        return obj[key];
  ***REMOVED***
    else ***REMOVED***
        return undefined;
  ***REMOVED***
}
function isEmpty(obj) ***REMOVED***
    for (var key in obj) ***REMOVED***
        if (Object.prototype.hasOwnProperty.call(obj, key)) ***REMOVED***
            return false;
      ***REMOVED***
  ***REMOVED***
    return true;
}
function map(obj, fn, contextObj) ***REMOVED***
    var res = ***REMOVED***};
    for (var key in obj) ***REMOVED***
        if (Object.prototype.hasOwnProperty.call(obj, key)) ***REMOVED***
            res[key] = fn.call(contextObj, obj[key], key, obj);
      ***REMOVED***
  ***REMOVED***
    return res;
}
/**
 * Deep equal two objects. Support Arrays and Objects.
 */
function deepEqual(a, b) ***REMOVED***
    if (a === b) ***REMOVED***
        return true;
  ***REMOVED***
    var aKeys = Object.keys(a);
    var bKeys = Object.keys(b);
    for (var _i = 0, aKeys_1 = aKeys; _i < aKeys_1.length; _i++) ***REMOVED***
        var k = aKeys_1[_i];
        if (!bKeys.includes(k)) ***REMOVED***
            return false;
      ***REMOVED***
        var aProp = a[k];
        var bProp = b[k];
        if (isObject(aProp) && isObject(bProp)) ***REMOVED***
            if (!deepEqual(aProp, bProp)) ***REMOVED***
                return false;
          ***REMOVED***
      ***REMOVED***
        else if (aProp !== bProp) ***REMOVED***
            return false;
      ***REMOVED***
  ***REMOVED***
    for (var _a = 0, bKeys_1 = bKeys; _a < bKeys_1.length; _a++) ***REMOVED***
        var k = bKeys_1[_a];
        if (!aKeys.includes(k)) ***REMOVED***
            return false;
      ***REMOVED***
  ***REMOVED***
    return true;
}
function isObject(thing) ***REMOVED***
    return thing !== null && typeof thing === 'object';
}

/**
 * @license
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Rejects if the given promise doesn't resolve in timeInMS milliseconds.
 * @internal
 */
function promiseWithTimeout(promise, timeInMS) ***REMOVED***
    if (timeInMS === void 0) ***REMOVED*** timeInMS = 2000; }
    var deferredPromise = new Deferred();
    setTimeout(function () ***REMOVED*** return deferredPromise.reject('timeout!'); }, timeInMS);
    promise.then(deferredPromise.resolve, deferredPromise.reject);
    return deferredPromise.promise;
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Returns a querystring-formatted string (e.g. &arg=val&arg2=val2) from a
 * params object (e.g. ***REMOVED***arg: 'val', arg2: 'val2'})
 * Note: You must prepend it with ? when adding it to a URL.
 */
function querystring(querystringParams) ***REMOVED***
    var params = [];
    var _loop_1 = function (key, value) ***REMOVED***
        if (Array.isArray(value)) ***REMOVED***
            value.forEach(function (arrayVal) ***REMOVED***
                params.push(encodeURIComponent(key) + '=' + encodeURIComponent(arrayVal));
          ***REMOVED***);
      ***REMOVED***
        else ***REMOVED***
            params.push(encodeURIComponent(key) + '=' + encodeURIComponent(value));
      ***REMOVED***
  ***REMOVED***;
    for (var _i = 0, _a = Object.entries(querystringParams); _i < _a.length; _i++) ***REMOVED***
        var _b = _a[_i], key = _b[0], value = _b[1];
        _loop_1(key, value);
  ***REMOVED***
    return params.length ? '&' + params.join('&') : '';
}
/**
 * Decodes a querystring (e.g. ?arg=val&arg2=val2) into a params object
 * (e.g. ***REMOVED***arg: 'val', arg2: 'val2'})
 */
function querystringDecode(querystring) ***REMOVED***
    var obj = ***REMOVED***};
    var tokens = querystring.replace(/^\?/, '').split('&');
    tokens.forEach(function (token) ***REMOVED***
        if (token) ***REMOVED***
            var _a = token.split('='), key = _a[0], value = _a[1];
            obj[decodeURIComponent(key)] = decodeURIComponent(value);
      ***REMOVED***
  ***REMOVED***);
    return obj;
}
/**
 * Extract the query string part of a URL, including the leading question mark (if present).
 */
function extractQuerystring(url) ***REMOVED***
    var queryStart = url.indexOf('?');
    if (!queryStart) ***REMOVED***
        return '';
  ***REMOVED***
    var fragmentStart = url.indexOf('#', queryStart);
    return url.substring(queryStart, fragmentStart > 0 ? fragmentStart : undefined);
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * @fileoverview SHA-1 cryptographic hash.
 * Variable names follow the notation in FIPS PUB 180-3:
 * http://csrc.nist.gov/publications/fips/fips180-3/fips180-3_final.pdf.
 *
 * Usage:
 *   var sha1 = new sha1();
 *   sha1.update(bytes);
 *   var hash = sha1.digest();
 *
 * Performance:
 *   Chrome 23:   ~400 Mbit/s
 *   Firefox 16:  ~250 Mbit/s
 *
 */
/**
 * SHA-1 cryptographic hash constructor.
 *
 * The properties declared here are discussed in the above algorithm document.
 * @constructor
 * @final
 * @struct
 */
var Sha1 = /** @class */ (function () ***REMOVED***
    function Sha1() ***REMOVED***
        /**
         * Holds the previous values of accumulated variables a-e in the compress_
         * function.
         * @private
         */
        this.chain_ = [];
        /**
         * A buffer holding the partially computed hash result.
         * @private
         */
        this.buf_ = [];
        /**
         * An array of 80 bytes, each a part of the message to be hashed.  Referred to
         * as the message schedule in the docs.
         * @private
         */
        this.W_ = [];
        /**
         * Contains data needed to pad messages less than 64 bytes.
         * @private
         */
        this.pad_ = [];
        /**
         * @private ***REMOVED***number}
         */
        this.inbuf_ = 0;
        /**
         * @private ***REMOVED***number}
         */
        this.total_ = 0;
        this.blockSize = 512 / 8;
        this.pad_[0] = 128;
        for (var i = 1; i < this.blockSize; ++i) ***REMOVED***
            this.pad_[i] = 0;
      ***REMOVED***
        this.reset();
  ***REMOVED***
    Sha1.prototype.reset = function () ***REMOVED***
        this.chain_[0] = 0x67452301;
        this.chain_[1] = 0xefcdab89;
        this.chain_[2] = 0x98badcfe;
        this.chain_[3] = 0x10325476;
        this.chain_[4] = 0xc3d2e1f0;
        this.inbuf_ = 0;
        this.total_ = 0;
  ***REMOVED***;
    /**
     * Internal compress helper function.
     * @param buf Block to compress.
     * @param offset Offset of the block in the buffer.
     * @private
     */
    Sha1.prototype.compress_ = function (buf, offset) ***REMOVED***
        if (!offset) ***REMOVED***
            offset = 0;
      ***REMOVED***
        var W = this.W_;
        // get 16 big endian words
        if (typeof buf === 'string') ***REMOVED***
            for (var i = 0; i < 16; i++) ***REMOVED***
                // TODO(user): [bug 8140122] Recent versions of Safari for Mac OS and iOS
                // have a bug that turns the post-increment ++ operator into pre-increment
                // during JIT compilation.  We have code that depends heavily on SHA-1 for
                // correctness and which is affected by this bug, so I've removed all uses
                // of post-increment ++ in which the result value is used.  We can revert
                // this change once the Safari bug
                // (https://bugs.webkit.org/show_bug.cgi?id=109036) has been fixed and
                // most clients have been updated.
                W[i] =
                    (buf.charCodeAt(offset) << 24) |
                        (buf.charCodeAt(offset + 1) << 16) |
                        (buf.charCodeAt(offset + 2) << 8) |
                        buf.charCodeAt(offset + 3);
                offset += 4;
          ***REMOVED***
      ***REMOVED***
        else ***REMOVED***
            for (var i = 0; i < 16; i++) ***REMOVED***
                W[i] =
                    (buf[offset] << 24) |
                        (buf[offset + 1] << 16) |
                        (buf[offset + 2] << 8) |
                        buf[offset + 3];
                offset += 4;
          ***REMOVED***
      ***REMOVED***
        // expand to 80 words
        for (var i = 16; i < 80; i++) ***REMOVED***
            var t = W[i - 3] ^ W[i - 8] ^ W[i - 14] ^ W[i - 16];
            W[i] = ((t << 1) | (t >>> 31)) & 0xffffffff;
      ***REMOVED***
        var a = this.chain_[0];
        var b = this.chain_[1];
        var c = this.chain_[2];
        var d = this.chain_[3];
        var e = this.chain_[4];
        var f, k;
        // TODO(user): Try to unroll this loop to speed up the computation.
        for (var i = 0; i < 80; i++) ***REMOVED***
            if (i < 40) ***REMOVED***
                if (i < 20) ***REMOVED***
                    f = d ^ (b & (c ^ d));
                    k = 0x5a827999;
              ***REMOVED***
                else ***REMOVED***
                    f = b ^ c ^ d;
                    k = 0x6ed9eba1;
              ***REMOVED***
          ***REMOVED***
            else ***REMOVED***
                if (i < 60) ***REMOVED***
                    f = (b & c) | (d & (b | c));
                    k = 0x8f1bbcdc;
              ***REMOVED***
                else ***REMOVED***
                    f = b ^ c ^ d;
                    k = 0xca62c1d6;
              ***REMOVED***
          ***REMOVED***
            var t = (((a << 5) | (a >>> 27)) + f + e + k + W[i]) & 0xffffffff;
            e = d;
            d = c;
            c = ((b << 30) | (b >>> 2)) & 0xffffffff;
            b = a;
            a = t;
      ***REMOVED***
        this.chain_[0] = (this.chain_[0] + a) & 0xffffffff;
        this.chain_[1] = (this.chain_[1] + b) & 0xffffffff;
        this.chain_[2] = (this.chain_[2] + c) & 0xffffffff;
        this.chain_[3] = (this.chain_[3] + d) & 0xffffffff;
        this.chain_[4] = (this.chain_[4] + e) & 0xffffffff;
  ***REMOVED***;
    Sha1.prototype.update = function (bytes, length) ***REMOVED***
        // TODO(johnlenz): tighten the function signature and remove this check
        if (bytes == null) ***REMOVED***
            return;
      ***REMOVED***
        if (length === undefined) ***REMOVED***
            length = bytes.length;
      ***REMOVED***
        var lengthMinusBlock = length - this.blockSize;
        var n = 0;
        // Using local instead of member variables gives ~5% speedup on Firefox 16.
        var buf = this.buf_;
        var inbuf = this.inbuf_;
        // The outer while loop should execute at most twice.
        while (n < length) ***REMOVED***
            // When we have no data in the block to top up, we can directly process the
            // input buffer (assuming it contains sufficient data). This gives ~25%
            // speedup on Chrome 23 and ~15% speedup on Firefox 16, but requires that
            // the data is provided in large chunks (or in multiples of 64 bytes).
            if (inbuf === 0) ***REMOVED***
                while (n <= lengthMinusBlock) ***REMOVED***
                    this.compress_(bytes, n);
                    n += this.blockSize;
              ***REMOVED***
          ***REMOVED***
            if (typeof bytes === 'string') ***REMOVED***
                while (n < length) ***REMOVED***
                    buf[inbuf] = bytes.charCodeAt(n);
                    ++inbuf;
                    ++n;
                    if (inbuf === this.blockSize) ***REMOVED***
                        this.compress_(buf);
                        inbuf = 0;
                        // Jump to the outer loop so we use the full-block optimization.
                        break;
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
            else ***REMOVED***
                while (n < length) ***REMOVED***
                    buf[inbuf] = bytes[n];
                    ++inbuf;
                    ++n;
                    if (inbuf === this.blockSize) ***REMOVED***
                        this.compress_(buf);
                        inbuf = 0;
                        // Jump to the outer loop so we use the full-block optimization.
                        break;
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***
        this.inbuf_ = inbuf;
        this.total_ += length;
  ***REMOVED***;
    /** @override */
    Sha1.prototype.digest = function () ***REMOVED***
        var digest = [];
        var totalBits = this.total_ * 8;
        // Add pad 0x80 0x00*.
        if (this.inbuf_ < 56) ***REMOVED***
            this.update(this.pad_, 56 - this.inbuf_);
      ***REMOVED***
        else ***REMOVED***
            this.update(this.pad_, this.blockSize - (this.inbuf_ - 56));
      ***REMOVED***
        // Add # bits.
        for (var i = this.blockSize - 1; i >= 56; i--) ***REMOVED***
            this.buf_[i] = totalBits & 255;
            totalBits /= 256; // Don't use bit-shifting here!
      ***REMOVED***
        this.compress_(this.buf_);
        var n = 0;
        for (var i = 0; i < 5; i++) ***REMOVED***
            for (var j = 24; j >= 0; j -= 8) ***REMOVED***
                digest[n] = (this.chain_[i] >> j) & 255;
                ++n;
          ***REMOVED***
      ***REMOVED***
        return digest;
  ***REMOVED***;
    return Sha1;
}());

/**
 * Helper to make a Subscribe function (just like Promise helps make a
 * Thenable).
 *
 * @param executor Function which can make calls to a single Observer
 *     as a proxy.
 * @param onNoObservers Callback when count of Observers goes to zero.
 */
function createSubscribe(executor, onNoObservers) ***REMOVED***
    var proxy = new ObserverProxy(executor, onNoObservers);
    return proxy.subscribe.bind(proxy);
}
/**
 * Implement fan-out for any number of Observers attached via a subscribe
 * function.
 */
var ObserverProxy = /** @class */ (function () ***REMOVED***
    /**
     * @param executor Function which can make calls to a single Observer
     *     as a proxy.
     * @param onNoObservers Callback when count of Observers goes to zero.
     */
    function ObserverProxy(executor, onNoObservers) ***REMOVED***
        var _this = this;
        this.observers = [];
        this.unsubscribes = [];
        this.observerCount = 0;
        // Micro-task scheduling by calling task.then().
        this.task = Promise.resolve();
        this.finalized = false;
        this.onNoObservers = onNoObservers;
        // Call the executor asynchronously so subscribers that are called
        // synchronously after the creation of the subscribe function
        // can still receive the very first value generated in the executor.
        this.task
            .then(function () ***REMOVED***
            executor(_this);
      ***REMOVED***)
            .catch(function (e) ***REMOVED***
            _this.error(e);
      ***REMOVED***);
  ***REMOVED***
    ObserverProxy.prototype.next = function (value) ***REMOVED***
        this.forEachObserver(function (observer) ***REMOVED***
            observer.next(value);
      ***REMOVED***);
  ***REMOVED***;
    ObserverProxy.prototype.error = function (error) ***REMOVED***
        this.forEachObserver(function (observer) ***REMOVED***
            observer.error(error);
      ***REMOVED***);
        this.close(error);
  ***REMOVED***;
    ObserverProxy.prototype.complete = function () ***REMOVED***
        this.forEachObserver(function (observer) ***REMOVED***
            observer.complete();
      ***REMOVED***);
        this.close();
  ***REMOVED***;
    /**
     * Subscribe function that can be used to add an Observer to the fan-out list.
     *
     * - We require that no event is sent to a subscriber sychronously to their
     *   call to subscribe().
     */
    ObserverProxy.prototype.subscribe = function (nextOrObserver, error, complete) ***REMOVED***
        var _this = this;
        var observer;
        if (nextOrObserver === undefined &&
            error === undefined &&
            complete === undefined) ***REMOVED***
            throw new Error('Missing Observer.');
      ***REMOVED***
        // Assemble an Observer object when passed as callback functions.
        if (implementsAnyMethods(nextOrObserver, [
            'next',
            'error',
            'complete'
        ])) ***REMOVED***
            observer = nextOrObserver;
      ***REMOVED***
        else ***REMOVED***
            observer = ***REMOVED***
                next: nextOrObserver,
                error: error,
                complete: complete
          ***REMOVED***;
      ***REMOVED***
        if (observer.next === undefined) ***REMOVED***
            observer.next = noop;
      ***REMOVED***
        if (observer.error === undefined) ***REMOVED***
            observer.error = noop;
      ***REMOVED***
        if (observer.complete === undefined) ***REMOVED***
            observer.complete = noop;
      ***REMOVED***
        var unsub = this.unsubscribeOne.bind(this, this.observers.length);
        // Attempt to subscribe to a terminated Observable - we
        // just respond to the Observer with the final error or complete
        // event.
        if (this.finalized) ***REMOVED***
            // eslint-disable-next-line @typescript-eslint/no-floating-promises
            this.task.then(function () ***REMOVED***
                try ***REMOVED***
                    if (_this.finalError) ***REMOVED***
                        observer.error(_this.finalError);
                  ***REMOVED***
                    else ***REMOVED***
                        observer.complete();
                  ***REMOVED***
              ***REMOVED***
                catch (e) ***REMOVED***
                    // nothing
              ***REMOVED***
                return;
          ***REMOVED***);
      ***REMOVED***
        this.observers.push(observer);
        return unsub;
  ***REMOVED***;
    // Unsubscribe is synchronous - we guarantee that no events are sent to
    // any unsubscribed Observer.
    ObserverProxy.prototype.unsubscribeOne = function (i) ***REMOVED***
        if (this.observers === undefined || this.observers[i] === undefined) ***REMOVED***
            return;
      ***REMOVED***
        delete this.observers[i];
        this.observerCount -= 1;
        if (this.observerCount === 0 && this.onNoObservers !== undefined) ***REMOVED***
            this.onNoObservers(this);
      ***REMOVED***
  ***REMOVED***;
    ObserverProxy.prototype.forEachObserver = function (fn) ***REMOVED***
        if (this.finalized) ***REMOVED***
            // Already closed by previous event....just eat the additional values.
            return;
      ***REMOVED***
        // Since sendOne calls asynchronously - there is no chance that
        // this.observers will become undefined.
        for (var i = 0; i < this.observers.length; i++) ***REMOVED***
            this.sendOne(i, fn);
      ***REMOVED***
  ***REMOVED***;
    // Call the Observer via one of it's callback function. We are careful to
    // confirm that the observe has not been unsubscribed since this asynchronous
    // function had been queued.
    ObserverProxy.prototype.sendOne = function (i, fn) ***REMOVED***
        var _this = this;
        // Execute the callback asynchronously
        // eslint-disable-next-line @typescript-eslint/no-floating-promises
        this.task.then(function () ***REMOVED***
            if (_this.observers !== undefined && _this.observers[i] !== undefined) ***REMOVED***
                try ***REMOVED***
                    fn(_this.observers[i]);
              ***REMOVED***
                catch (e) ***REMOVED***
                    // Ignore exceptions raised in Observers or missing methods of an
                    // Observer.
                    // Log error to console. b/31404806
                    if (typeof console !== 'undefined' && console.error) ***REMOVED***
                        console.error(e);
                  ***REMOVED***
              ***REMOVED***
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***;
    ObserverProxy.prototype.close = function (err) ***REMOVED***
        var _this = this;
        if (this.finalized) ***REMOVED***
            return;
      ***REMOVED***
        this.finalized = true;
        if (err !== undefined) ***REMOVED***
            this.finalError = err;
      ***REMOVED***
        // Proxy is no longer needed - garbage collect references
        // eslint-disable-next-line @typescript-eslint/no-floating-promises
        this.task.then(function () ***REMOVED***
            _this.observers = undefined;
            _this.onNoObservers = undefined;
      ***REMOVED***);
  ***REMOVED***;
    return ObserverProxy;
}());
/** Turn synchronous function into one called asynchronously. */
// eslint-disable-next-line @typescript-eslint/ban-types
function async(fn, onError) ***REMOVED***
    return function () ***REMOVED***
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
            args[_i] = arguments[_i];
      ***REMOVED***
        Promise.resolve(true)
            .then(function () ***REMOVED***
            fn.apply(void 0, args);
      ***REMOVED***)
            .catch(function (error) ***REMOVED***
            if (onError) ***REMOVED***
                onError(error);
          ***REMOVED***
      ***REMOVED***);
  ***REMOVED***;
}
/**
 * Return true if the object passed in implements any of the named methods.
 */
function implementsAnyMethods(obj, methods) ***REMOVED***
    if (typeof obj !== 'object' || obj === null) ***REMOVED***
        return false;
  ***REMOVED***
    for (var _i = 0, methods_1 = methods; _i < methods_1.length; _i++) ***REMOVED***
        var method = methods_1[_i];
        if (method in obj && typeof obj[method] === 'function') ***REMOVED***
            return true;
      ***REMOVED***
  ***REMOVED***
    return false;
}
function noop() ***REMOVED***
    // do nothing
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Check to make sure the appropriate number of arguments are provided for a public function.
 * Throws an error if it fails.
 *
 * @param fnName The function name
 * @param minCount The minimum number of arguments to allow for the function call
 * @param maxCount The maximum number of argument to allow for the function call
 * @param argCount The actual number of arguments provided.
 */
var validateArgCount = function (fnName, minCount, maxCount, argCount) ***REMOVED***
    var argError;
    if (argCount < minCount) ***REMOVED***
        argError = 'at least ' + minCount;
  ***REMOVED***
    else if (argCount > maxCount) ***REMOVED***
        argError = maxCount === 0 ? 'none' : 'no more than ' + maxCount;
  ***REMOVED***
    if (argError) ***REMOVED***
        var error = fnName +
            ' failed: Was called with ' +
            argCount +
            (argCount === 1 ? ' argument.' : ' arguments.') +
            ' Expects ' +
            argError +
            '.';
        throw new Error(error);
  ***REMOVED***
};
/**
 * Generates a string to prefix an error message about failed argument validation
 *
 * @param fnName The function name
 * @param argName The name of the argument
 * @return The prefix to add to the error thrown for validation.
 */
function errorPrefix(fnName, argName) ***REMOVED***
    return "".concat(fnName, " failed: ").concat(argName, " argument ");
}
/**
 * @param fnName
 * @param argumentNumber
 * @param namespace
 * @param optional
 */
function validateNamespace(fnName, namespace, optional) ***REMOVED***
    if (optional && !namespace) ***REMOVED***
        return;
  ***REMOVED***
    if (typeof namespace !== 'string') ***REMOVED***
        //TODO: I should do more validation here. We only allow certain chars in namespaces.
        throw new Error(errorPrefix(fnName, 'namespace') + 'must be a valid firebase namespace.');
  ***REMOVED***
}
function validateCallback(fnName, argumentName, 
// eslint-disable-next-line @typescript-eslint/ban-types
callback, optional) ***REMOVED***
    if (optional && !callback) ***REMOVED***
        return;
  ***REMOVED***
    if (typeof callback !== 'function') ***REMOVED***
        throw new Error(errorPrefix(fnName, argumentName) + 'must be a valid function.');
  ***REMOVED***
}
function validateContextObject(fnName, argumentName, context, optional) ***REMOVED***
    if (optional && !context) ***REMOVED***
        return;
  ***REMOVED***
    if (typeof context !== 'object' || context === null) ***REMOVED***
        throw new Error(errorPrefix(fnName, argumentName) + 'must be a valid context object.');
  ***REMOVED***
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// Code originally came from goog.crypt.stringToUtf8ByteArray, but for some reason they
// automatically replaced '\r\n' with '\n', and they didn't handle surrogate pairs,
// so it's been modified.
// Note that not all Unicode characters appear as single characters in JavaScript strings.
// fromCharCode returns the UTF-16 encoding of a character - so some Unicode characters
// use 2 characters in Javascript.  All 4-byte UTF-8 characters begin with a first
// character in the range 0xD800 - 0xDBFF (the first character of a so-called surrogate
// pair).
// See http://www.ecma-international.org/ecma-262/5.1/#sec-15.1.3
/**
 * @param ***REMOVED***string} str
 * @return ***REMOVED***Array}
 */
var stringToByteArray = function (str) ***REMOVED***
    var out = [];
    var p = 0;
    for (var i = 0; i < str.length; i++) ***REMOVED***
        var c = str.charCodeAt(i);
        // Is this the lead surrogate in a surrogate pair?
        if (c >= 0xd800 && c <= 0xdbff) ***REMOVED***
            var high = c - 0xd800; // the high 10 bits.
            i++;
            assert(i < str.length, 'Surrogate pair missing trail surrogate.');
            var low = str.charCodeAt(i) - 0xdc00; // the low 10 bits.
            c = 0x10000 + (high << 10) + low;
      ***REMOVED***
        if (c < 128) ***REMOVED***
            out[p++] = c;
      ***REMOVED***
        else if (c < 2048) ***REMOVED***
            out[p++] = (c >> 6) | 192;
            out[p++] = (c & 63) | 128;
      ***REMOVED***
        else if (c < 65536) ***REMOVED***
            out[p++] = (c >> 12) | 224;
            out[p++] = ((c >> 6) & 63) | 128;
            out[p++] = (c & 63) | 128;
      ***REMOVED***
        else ***REMOVED***
            out[p++] = (c >> 18) | 240;
            out[p++] = ((c >> 12) & 63) | 128;
            out[p++] = ((c >> 6) & 63) | 128;
            out[p++] = (c & 63) | 128;
      ***REMOVED***
  ***REMOVED***
    return out;
};
/**
 * Calculate length without actually converting; useful for doing cheaper validation.
 * @param ***REMOVED***string} str
 * @return ***REMOVED***number}
 */
var stringLength = function (str) ***REMOVED***
    var p = 0;
    for (var i = 0; i < str.length; i++) ***REMOVED***
        var c = str.charCodeAt(i);
        if (c < 128) ***REMOVED***
            p++;
      ***REMOVED***
        else if (c < 2048) ***REMOVED***
            p += 2;
      ***REMOVED***
        else if (c >= 0xd800 && c <= 0xdbff) ***REMOVED***
            // Lead surrogate of a surrogate pair.  The pair together will take 4 bytes to represent.
            p += 4;
            i++; // skip trail surrogate.
      ***REMOVED***
        else ***REMOVED***
            p += 3;
      ***REMOVED***
  ***REMOVED***
    return p;
};

/**
 * @license
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Copied from https://stackoverflow.com/a/2117523
 * Generates a new uuid.
 * @public
 */
var uuidv4 = function () ***REMOVED***
    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) ***REMOVED***
        var r = (Math.random() * 16) | 0, v = c === 'x' ? r : (r & 0x3) | 0x8;
        return v.toString(16);
  ***REMOVED***);
};

/**
 * @license
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * The amount of milliseconds to exponentially increase.
 */
var DEFAULT_INTERVAL_MILLIS = 1000;
/**
 * The factor to backoff by.
 * Should be a number greater than 1.
 */
var DEFAULT_BACKOFF_FACTOR = 2;
/**
 * The maximum milliseconds to increase to.
 *
 * <p>Visible for testing
 */
var MAX_VALUE_MILLIS = 4 * 60 * 60 * 1000; // Four hours, like iOS and Android.
/**
 * The percentage of backoff time to randomize by.
 * See
 * http://go/safe-client-behavior#step-1-determine-the-appropriate-retry-interval-to-handle-spike-traffic
 * for context.
 *
 * <p>Visible for testing
 */
var RANDOM_FACTOR = 0.5;
/**
 * Based on the backoff method from
 * https://github.com/google/closure-library/blob/master/closure/goog/math/exponentialbackoff.js.
 * Extracted here so we don't need to pass metadata and a stateful ExponentialBackoff object around.
 */
function calculateBackoffMillis(backoffCount, intervalMillis, backoffFactor) ***REMOVED***
    if (intervalMillis === void 0) ***REMOVED*** intervalMillis = DEFAULT_INTERVAL_MILLIS; }
    if (backoffFactor === void 0) ***REMOVED*** backoffFactor = DEFAULT_BACKOFF_FACTOR; }
    // Calculates an exponentially increasing value.
    // Deviation: calculates value from count and a constant interval, so we only need to save value
    // and count to restore state.
    var currBaseValue = intervalMillis * Math.pow(backoffFactor, backoffCount);
    // A random "fuzz" to avoid waves of retries.
    // Deviation: randomFactor is required.
    var randomWait = Math.round(
    // A fraction of the backoff value to add/subtract.
    // Deviation: changes multiplication order to improve readability.
    RANDOM_FACTOR *
        currBaseValue *
        // A random float (rounded to int by Math.round above) in the range [-1, 1]. Determines
        // if we add or subtract.
        (Math.random() - 0.5) *
        2);
    // Limits backoff to max to avoid effectively permanent backoff.
    return Math.min(MAX_VALUE_MILLIS, currBaseValue + randomWait);
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Provide English ordinal letters after a number
 */
function ordinal(i) ***REMOVED***
    if (!Number.isFinite(i)) ***REMOVED***
        return "".concat(i);
  ***REMOVED***
    return i + indicator(i);
}
function indicator(i) ***REMOVED***
    i = Math.abs(i);
    var cent = i % 100;
    if (cent >= 10 && cent <= 20) ***REMOVED***
        return 'th';
  ***REMOVED***
    var dec = i % 10;
    if (dec === 1) ***REMOVED***
        return 'st';
  ***REMOVED***
    if (dec === 2) ***REMOVED***
        return 'nd';
  ***REMOVED***
    if (dec === 3) ***REMOVED***
        return 'rd';
  ***REMOVED***
    return 'th';
}

/**
 * @license
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
function getModularInstance(service) ***REMOVED***
    if (service && service._delegate) ***REMOVED***
        return service._delegate;
  ***REMOVED***
    else ***REMOVED***
        return service;
  ***REMOVED***
}

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// Overriding the constant (we should be the only ones doing this)
CONSTANTS.NODE_CLIENT = true;

exports.CONSTANTS = CONSTANTS;
exports.Deferred = Deferred;
exports.ErrorFactory = ErrorFactory;
exports.FirebaseError = FirebaseError;
exports.MAX_VALUE_MILLIS = MAX_VALUE_MILLIS;
exports.RANDOM_FACTOR = RANDOM_FACTOR;
exports.Sha1 = Sha1;
exports.areCookiesEnabled = areCookiesEnabled;
exports.assert = assert;
exports.assertionError = assertionError;
exports.async = async;
exports.base64 = base64;
exports.base64Decode = base64Decode;
exports.base64Encode = base64Encode;
exports.base64urlEncodeWithoutPadding = base64urlEncodeWithoutPadding;
exports.calculateBackoffMillis = calculateBackoffMillis;
exports.contains = contains;
exports.createMockUserToken = createMockUserToken;
exports.createSubscribe = createSubscribe;
exports.decode = decode;
exports.deepCopy = deepCopy;
exports.deepEqual = deepEqual;
exports.deepExtend = deepExtend;
exports.errorPrefix = errorPrefix;
exports.extractQuerystring = extractQuerystring;
exports.getDefaultAppConfig = getDefaultAppConfig;
exports.getDefaultEmulatorHost = getDefaultEmulatorHost;
exports.getDefaultEmulatorHostnameAndPort = getDefaultEmulatorHostnameAndPort;
exports.getDefaults = getDefaults;
exports.getExperimentalSetting = getExperimentalSetting;
exports.getGlobal = getGlobal;
exports.getModularInstance = getModularInstance;
exports.getUA = getUA;
exports.isAdmin = isAdmin;
exports.isBrowser = isBrowser;
exports.isBrowserExtension = isBrowserExtension;
exports.isElectron = isElectron;
exports.isEmpty = isEmpty;
exports.isIE = isIE;
exports.isIndexedDBAvailable = isIndexedDBAvailable;
exports.isMobileCordova = isMobileCordova;
exports.isNode = isNode;
exports.isNodeSdk = isNodeSdk;
exports.isReactNative = isReactNative;
exports.isSafari = isSafari;
exports.isUWP = isUWP;
exports.isValidFormat = isValidFormat;
exports.isValidTimestamp = isValidTimestamp;
exports.issuedAtTime = issuedAtTime;
exports.jsonEval = jsonEval;
exports.map = map;
exports.ordinal = ordinal;
exports.promiseWithTimeout = promiseWithTimeout;
exports.querystring = querystring;
exports.querystringDecode = querystringDecode;
exports.safeGet = safeGet;
exports.stringLength = stringLength;
exports.stringToByteArray = stringToByteArray;
exports.stringify = stringify;
exports.uuidv4 = uuidv4;
exports.validateArgCount = validateArgCount;
exports.validateCallback = validateCallback;
exports.validateContextObject = validateContextObject;
exports.validateIndexedDBOpenable = validateIndexedDBOpenable;
exports.validateNamespace = validateNamespace;
