// uglify-js/tools/node.js
var fs = require("fs");

exports.FILES = [
    require.resolve("../lib/utils.js"),
    require.resolve("../lib/ast.js"),
    require.resolve("../lib/transform.js"),
    require.resolve("../lib/parse.js"),
    require.resolve("../lib/scope.js"),
    require.resolve("../lib/compress.js"),
    require.resolve("../lib/output.js"),
    require.resolve("../lib/sourcemap.js"),
    require.resolve("../lib/mozilla-ast.js"),
    require.resolve("../lib/propmangle.js"),
    require.resolve("../lib/minify.js"),
    require.resolve("./exports.js"),
];

new Function("domprops", "exports", function() ***REMOVED***
    var code = exports.FILES.map(function(file) ***REMOVED***
        return fs.readFileSync(file, "utf8");
  ***REMOVED***);
    code.push("exports.describe_ast = " + describe_ast.toString());
    return code.join("\n\n");
}())(require("./domprops.json"), exports);

function to_comment(value) ***REMOVED***
    if (typeof value != "string") value = JSON.stringify(value, function(key, value) ***REMOVED***
        return typeof value == "function" ? "<[ " + value + " ]>" : value;
  ***REMOVED***, 2);
    return "// " + value.replace(/\n/g, "\n// ");
}

if (+process.env["UGLIFY_BUG_REPORT"]) exports.minify = function(files, options) ***REMOVED***
    if (typeof options == "undefined") options = "<<undefined>>";
    var code = [
        "// UGLIFY_BUG_REPORT",
        to_comment(options),
    ];
    if (typeof files == "string") ***REMOVED***
        code.push("");
        code.push("//-------------------------------------------------------------")
        code.push("// INPUT CODE", files);
  ***REMOVED*** else for (var name in files) ***REMOVED***
        code.push("");
        code.push("//-------------------------------------------------------------")
        code.push(to_comment(name), files[name]);
  ***REMOVED***
    if (options.sourceMap && options.sourceMap.url) ***REMOVED***
        code.push("");
        code.push("//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IiJ9");
  ***REMOVED***
    var result = ***REMOVED*** code: code.join("\n") };
    if (options.sourceMap) result.map = '***REMOVED***"version":3,"sources":[],"names":[],"mappings":""}';
    return result;
};

function describe_ast() ***REMOVED***
    var out = OutputStream(***REMOVED*** beautify: true });
    doitem(AST_Node);
    return out.get() + "\n";

    function doitem(ctor) ***REMOVED***
        out.print("AST_" + ctor.TYPE);
        var props = ctor.SELF_PROPS.filter(function(prop) ***REMOVED***
            return !/^\$/.test(prop);
      ***REMOVED***);
        if (props.length > 0) ***REMOVED***
            out.space();
            out.with_parens(function() ***REMOVED***
                props.forEach(function(prop, i) ***REMOVED***
                    if (i) out.space();
                    out.print(prop);
              ***REMOVED***);
          ***REMOVED***);
      ***REMOVED***
        if (ctor.documentation) ***REMOVED***
            out.space();
            out.print_string(ctor.documentation);
      ***REMOVED***
        if (ctor.SUBCLASSES.length > 0) ***REMOVED***
            out.space();
            out.with_block(function() ***REMOVED***
                ctor.SUBCLASSES.sort(function(a, b) ***REMOVED***
                    return a.TYPE < b.TYPE ? -1 : 1;
              ***REMOVED***).forEach(function(ctor, i) ***REMOVED***
                    out.indent();
                    doitem(ctor);
                    out.newline();
              ***REMOVED***);
          ***REMOVED***);
      ***REMOVED***
  ***REMOVED***
}

function infer_options(options) ***REMOVED***
    var result = exports.minify("", options);
    return result.error && result.error.defs;
}

exports.default_options = function() ***REMOVED***
    var defs = infer_options(***REMOVED*** 0: 0 });
    Object.keys(defs).forEach(function(component) ***REMOVED***
        var options = ***REMOVED***};
        options[component] = ***REMOVED*** 0: 0 };
        if (options = infer_options(options)) ***REMOVED***
            defs[component] = options;
      ***REMOVED***
  ***REMOVED***);
    return defs;
};
